<!DOCTYPE html><html><head>
            <meta content="text/html;charset=utf-8" http-equiv="Content-Type">
            <meta content="utf-8" http-equiv="encoding">
          <link href="katex/katex.min.css" rel="stylesheet"><script src="anki_src/dist/anki.js"></script><title>offbynull Anki</title><meta name="viewport" content="width=device-width,initial-scale=1"><link href=".temp_githib_css/github-markdown.css" rel="stylesheet"><link href=".temp_highlightjs_css/default.css" rel="stylesheet"></head>
          <body class="markdown-body"><p> 
</p>
<p><span class="anki-infopanel"></span>
</p>
<!-- --><ul>
<li>
<p>Question 1 four and beep or boop. The word red should be hidden. <span class="anki-hidepattern" data-pattern="red"></span> <span class="anki-answerpattern" data-pattern="four"></span> <span class="anki-answerpattern" data-pattern="beep"></span> <span class="anki-answerpattern" data-pattern="boop"></span> <span class="anki-deadquestion"></span>
</p>
</li>
<li>
<p>Question 2 What are the colors? <span class="anki-ignorepattern" data-pattern="blueberries"></span> <span class="anki-answerpattern" data-pattern="green"></span> <span class="anki-answerpattern" data-pattern="yellow"></span> <span class="anki-answerpattern" data-pattern="(red)/i"></span> <span class="anki-answerpattern" data-pattern="blue"></span> <span class="anki-deadquestion"></span>
</p>
<ul>
<li>avocado = green <span class="anki-listrandomorder"></span>
</li>
<li>blueberries = blue</li>
<li>banana = yellow</li>
<li>strawberry = red</li>
</ul>
</li>
<li>
<p>Question 3 Which fruit has a red color? <span class="anki-ignorepattern" data-pattern="(true|false)_IGNORE/i"></span> <span class="anki-answerpattern" data-pattern="(false1?)/i"></span> <span class="anki-answerpattern" data-pattern="(false2?)/i"></span> <span class="anki-answerpattern" data-pattern="(false3?)/i"></span> <span class="anki-answerpattern" data-pattern="(true0?)/i"></span> <span class="anki-deadquestion"></span>
</p>
<p>Answer with true_IGNORE or false_IGNORE:</p>
<ul>
<li>avocado = false1 <span class="anki-listrandomorder"></span>
</li>
<li>blueberries = false2</li>
<li>banana = false3</li>
<li>strawberry = true0</li>
</ul>
</li>
<li>
<p>Question 4 What color is the item? 4 possibilities but only 1 will be shown. <span class="anki-answerpattern" data-pattern="red"></span> <span class="anki-deadquestion"></span>
</p>
<ul>
<li>cherry <span class="anki-listrandomselect"></span></li>
<li>strawberry</li>
<li>stop sign</li>
<li>lipstick</li>
</ul>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(k-mer|kmer)/i"></span> - A substring of length k within some larger biological sequence (e.g. DNA or amino acid chain). For example, in the DNA sequence GAAATC, the following k-mer's exist:</p>
<table>
<thead>
<tr>
<th>k</th>
<th>k-mers</th>
</tr>
</thead>
<tbody>
<tr>
<td>1</td>
<td>G A A A T C</td>
</tr>
<tr>
<td>2</td>
<td>GA AA AA AT TC</td>
</tr>
<tr>
<td>3</td>
<td>GAA AAA AAT ATC</td>
</tr>
<tr>
<td>4</td>
<td>GAAA AAAT AATC</td>
</tr>
<tr>
<td>5</td>
<td>GAAAT AAATC</td>
</tr>
<tr>
<td>6</td>
<td>GAAATC</td>
</tr>
</tbody>
</table>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(kd-mer|kdmer|\(k,\s*d\)-mer)/i"></span> <span class="anki-hidepattern" data-pattern="(\(\d+,\s*\d+\)-mer)/i"></span> - A substring of length 2k + d within some larger biological sequence (e.g. DNA or amino acid chain) where the first k elements and the last k elements are known but the d elements in between isn't known.</p>
<p>When identifying a kd-mer with a specific k and d, the proper syntax is (k, d)-mer. For example, (1, 2)-mer represents a kd-mer with k=1 and d=2. In the DNA sequence GAAATC, the following (1, 2)-mer's exist: <code>G--A</code>, <code>A--T</code>, <code>A--C</code>.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(reverse complement)/i"></span> - Given double-stranded DNA, each ...</p>
<ul>
<li>strand's direction is opposite of the other,</li>
<li>strand position has a nucleotide that complements the nucleotide at that same position on the other stand (A ⟷ T and C ⟷ G)</li>
</ul>
<p><img src="svgbob_b05df825cb791aaf67d96b4e629e42c02b75f509.svg" alt="Kroki diagram output"></p>
<p>The reverse complement means that a stretch of single-stranded DNA has its direction reversed (5' and 3' switched) and nucleotides complemented.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(5'|5 prime)/i"></span> <span class="anki-answerpattern" data-pattern="(3'|3 prime)/i"></span> - 5' (5 prime) and 3' (3 prime) describe the opposite ends of DNA. The chemical structure at each end is what defines if it's 5' or 3' -- each end is guaranteed to be different from the other. The forward direction on DNA is defined as 5' to 3', while the backwards direction is 3' to 5'.</p>
<p>Two complementing DNA strands will always be attached in opposite directions.</p>
<p><img src="svgbob_17a210d3487024b39998105bb2cf073cdd5d23a1.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="DNA polymerase"></span> - An enzyme that replicates a strand of DNA. That is, DNA polymerase walks over a single strand of DNA bases (not the strand of base pairs) and  generates a strand of complements. Before DNA polymerase can attach itself and start replicating DNA, it requires a primer.</p>
<p><img src="svgbob_4dc3db657aeeb2dd3bd7644d409f31aacad64536.svg" alt="Kroki diagram output"></p>
<p>DNA polymerase is unidirectional, meaning that it can only walk a DNA strand in one direction: reverse (3' to 5')</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="replication fork"></span> - The process of DNA replication requires that DNA's 2 complementing strands be unwound and split open. The area where the DNA starts to  split is called the replication fork. In bacteria, the replication fork starts at the replication origin and keeps expanding until it reaches the replication terminus.  Special enzymes called DNA polymerases walk over each unwound strand and create complementing strands.</p>
<p><img src="svgbob_fc7cf2c734fa0cab15b5746e25ad07e25f85ec64.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="replication origin"></span> <span class="anki-answerpattern" data-pattern="ori"></span> - The point in DNA at which replication starts. May be shortened to ori.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="replication terminus"></span> <span class="anki-answerpattern" data-pattern="ter"></span> - The point in DNA at which replication ends. May be shortened to ter.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="forward half-strand"></span> <span class="anki-answerpattern" data-pattern="(reverse half-strand|backward half-strand|backwards half-strand)/i"></span> <span class="anki-answerpattern" data-pattern="lagging half-strand"></span> <span class="anki-answerpattern" data-pattern="leading half-strand"></span> - Bacteria are known to have a single chromosome of circular / looping DNA. In this DNA, the replication origin (ori) is the region of DNA where replication starts, while the replication terminus (ter) is where replication ends.</p>
<p><img src="svgbob_fb9438945a109c884df7ccc79d1d8051fd358a87.svg" alt="Kroki diagram output"></p>
<p>If you split up the DNA based on ori and ter being cutting points, you end up with 4 distinct strands. Given that the direction of a strand is 5' to 3', if the direction of the strand starts at...</p>
<ul>
<li>
<p>ori and ends at ter, it's called the forward half-strand.</p>
<p><img src="svgbob_ab156edc66768e6c63e3a6c78c175004940c67ac.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>ter and ends at ori, it's called the reverse half-strand.</p>
<p><img src="svgbob_02b4b080274409daeaa7ecbe8dbc22df63279b65.svg" alt="Kroki diagram output"></p>
</li>
</ul>
<p>Given the 2 strands tha make up a DNA molecule, the strand that goes in the...</p>
<ul>
<li>forward half-strand (5' to 3') is called the lagging half-strand.</li>
<li>reverse half-strand (3' to 5') is called the leading half-strand.</li>
</ul>
<p>This nomenclature has to do with DNA polymerase. Since DNA polymerase can only walk in the reverse direction (3' to 5'), it synthesizes the leading half-strand in one shot. For the lagging half-strand (5' to 3'), multiple DNA polymerases have to used to synthesize DNA, each binding to the lagging strand and walking backwards a small amount to generate a small fragment of DNA (Okazaki fragment). the process is much slower for the lagging half-strand, that's why it's called lagging.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="Okazaki fragment"></span> - A small fragment of DNA generated by DNA polymerase for forward half-strands. DNA synthesis for the forward half-strands can only happen in small pieces. As the fork open ups every ~2000 nucleotides, DNA polymerase attaches to the end of the fork on the forward half-strand and walks in reverse to generate that small segment (DNA polymerase can only walk in the reverse direction).</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="DNA ligase"></span> - An enzyme that sews together short segments of DNA called Okazaki fragments by binding the phosphate group on the end of one strand with the deoxyribose group on the other strand.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="DnaA box"></span> - A sequence in the ori that the DnaA protein (responsible for DNA replication) binds to.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="reverse complement"></span> - Two strands of DNA bound together, where each strand is the reverse complement of the other.</p>
<p><img src="svgbob_b05df825cb791aaf67d96b4e629e42c02b75f509.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="gene"></span> - A segment of DNA that contains the instructions for either a protein or functional RNA.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="gene product"></span> - The final synthesized material resulting from the instructions that make up a gene. That synthesized material either being a protein or functional RNA.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(transcription|transcribed|transcribe)/i"></span> - The process of copying a gene to mRNA. Specifically, the enzyme RNA polymerase copies the segment of DNA that makes up that gene to a strand of RNA.</p>
<p><img src="svgbob_0fed5f4215fb0a5da3d93da589650b84df3f9c09.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(translation|translated|translate)/i"></span> - The process of turning mRNA to protein. Specifically, a ribosome takes in the mRNA generated by transcription and outputs the protein that it codes for.</p>
<p><img src="svgbob_aeaffd032a869fcce0f71f969f22f2191880a88c.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="gene expression"></span> - The process by which a gene is synthesized into a gene product. When the gene product is...</p>
<ul>
<li>a protein, the gene is transcribed to mRNA and translated to a protein.</li>
<li>functional RNA, the gene is transcribed to a type of RNA that isn't mRNA (only mRNA is translated to a protein).</li>
</ul>
<p><img src="svgbob_78b3833e37ddc45358bf3ebf79daba1a9e2604eb.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(regulatory gene|regulatory protein)/i"></span> - The proteins encoded by these genes effect gene expression for certain other genes. That is, a regulatory protein can cause certain other genes to be expressed more (promote gene expression) or less (repress gene expression).</p>
<p>Regulatory genes are often controlled by external factors (e.g. sunlight, nutrients, temperature, etc..)</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="negative feedback loop"></span> <span class="anki-answerpattern" data-pattern="positive feedback loop"></span> <span class="anki-answerpattern" data-pattern="feedback loop"></span> - A feedback loop is a system where the output (or some part of the output) is fed back into the system to either promote or repress further outputs.</p>
<p><img src="svgbob_3ba8fa4d05a4b9ef40e13dd6d9c7d60ecb9705ad.svg" alt="Kroki diagram output"></p>
<p>A positive feedback loop amplifies the output while a negative feedback loop regulates the output. Negative feedback loops in particular are important in biology because they allow organisms to maintain homeostasis / equilibrium (keep a consistent internal state). For example, the system that regulates core temperatures in a human is a negative feedback loop. If a human's core temperature gets too...</p>
<ul>
<li>low, they shiver to drive the temperature up.</li>
<li>high, they sweat to drive the temperature down.</li>
</ul>
<p>In the example above, the output is the core temperature. The body monitors its core temperature and employs mechanisms to bring it back to normal if it goes out of range (e.g. sweat, shiver). The outside temperature is influencing the body's core temperature as well as the internal shivering / sweating mechanisms the body employs.</p>
<p><img src="svgbob_894d108e9621a523aaed2c3a3d5d13d20871f1f3.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(circadian clock|circadian oscillator)/i"></span> - A biological clock that synchronizes roughly around the earth's day-night cycle. This internal clock helps many species regulate their physical and behavioural attributes. For example, hunt during the night vs sleep during the day (e.g. nocturnal owls).</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="upstream region"></span> - The area just before some interval of DNA. Since the direction of DNA is 5' to 3', this area is towards the 5' end (upper end).</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="downstream region"></span> - The area just after some interval of DNA. Since the direction of DNA is 5' to 3', this area is towards the 3' end (lower end).</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="transcription factor"></span> - A regulatory protein that controls the rate of transcription for some gene that it has influence over (the copying of DNA to mRNA). The protein binds to a specific sequence in the gene's upstream region.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="motif"></span> - A pattern that matches against many different k-mers, where those matched k-mers have some shared biological significance. The pattern matches a fixed k where each position may have alternate forms. The simplest way to think of a motif is a regex pattern without quantifiers. For example, the regex <code>[AT]TT[GC]C</code> may match to ATTGC, ATTCC, TTTGC, and TTTCC.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="motif member"></span> - A specific nucleotide sequence that matches a motif. For example, given a motif represented by the regex <code>[AT]TT[GC]C</code>, the sequences ATTGC, ATTCC, TTTGC, and TTTCC would be its motif members.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(motif matrix|motif matrices)/i"></span> - A set of k-mers stacked on top of each other in a matrix, where the k-mers are either...</p>
<ul>
<li>motif members of the same motif,</li>
<li>or suspected motif members of the same motif.</li>
</ul>
<p>For example, the motif <code>[AT]TT[GC]C</code> has the following matrix:</p>
<table>
<thead>
<tr>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
</tr>
</thead>
<tbody>
<tr>
<td>A</td>
<td>T</td>
<td>T</td>
<td>G</td>
<td>C</td>
</tr>
<tr>
<td>A</td>
<td>T</td>
<td>T</td>
<td>C</td>
<td>C</td>
</tr>
<tr>
<td>T</td>
<td>T</td>
<td>T</td>
<td>G</td>
<td>C</td>
</tr>
<tr>
<td>T</td>
<td>T</td>
<td>T</td>
<td>C</td>
<td>C</td>
</tr>
</tbody>
</table>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="regulatory motif"></span> - The motif of a transcription factor, typically 8 to 12 nucleotides in length.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="transcription factor binding site"></span> - The physical binding site for a transcription factor. A gene that's regulated by a transcription factor needs a sequence located in its upstream region that the transcription factor can bind to: a motif member of that transcription factor's regulatory motif.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>A gene's upstream region is the 600 to 1000 nucleotides preceding the start of the gene.</p>
</div>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(cDNA|complementary DNA)/i"></span> - A single strand of DNA generated from mRNA. The enzyme reverse transcriptase scans over the mRNA and creates the complementing single DNA strand.</p>
<p><img src="svgbob_a4a2743464f6da803a8d9d15a554dd4c6b62734d.svg" alt="Kroki diagram output"></p>
<p>The mRNA portion breaks off, leaving the single-stranded DNA.</p>
<p><img src="svgbob_e663fd58e09eb0df87cbfcdd76f15efdc0a142ec.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(DNA microarray|DNA array)/i"></span> - A device used to compare gene expression. This works by measuring 2 mRNA samples against each other: a control sample and an experimental sample. The samples could be from...</p>
<ul>
<li>the same organism but at different times.</li>
<li>diseased and healthy versions of the same organism.</li>
<li>etc..</li>
</ul>
<p>Both mRNA samples are converted to cDNA and are given fluorescent dyes. The control sample gets dyed green while the experimental sample gets dyed red.</p>
<p><img src="svgbob_6ce607591d81d0fd6c7507408e97fb8146e83baf.svg" alt="Kroki diagram output"></p>
<p>A sheet is broken up into multiple regions, where each region has the cDNA for one specific gene from the control sample printed.</p>
<p><img src="svgbob_421210658cf3b4d0040b4815e879948ebfd7ccc5.svg" alt="Kroki diagram output"></p>
<p>The idea is that once the experimental cDNA is introduced to that region, it should bind to the control cDNA that's been printed to form double-stranded DNA. The color emitted in a region should correspond to the amount of gene expression for the gene that region represents. For example, if a region on the sheet is fully yellow, it means that the gene expression for that gene is roughly equal (red mixed with green is yellow).</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="greedy algorithm"></span> - An algorithm that tries to speed things up by taking the locally optimal choice at each step. That is, the algorithm doesn't look more than 1 step ahead.</p>
<p>For example, imagine a chess playing AI that had a strategy of trying to eliminate the other player's most valuable piece at each turn. It would be considered greedy because it only looks 1 move ahead before taking action. Normal chess AIs / players look many moves ahead before taking action. As such, the greedy AI may be fast but it would very likely lose most matches.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(Cromwell'?s? rule)/i"></span> - When a probability is based off past events, 0.0 and 1.0 shouldn't be used. That is, if you've...</p>
<ul>
<li>never seen an even occur in the past, it doesn't mean that there's a 0.0 probability of it occurring next.</li>
<li>always seen an event occur in the past, it doesn't mean that there's a 1.0 probability of it occurring next.</li>
</ul>
<p>Unless you're dealing with hard logical statements where prior occurrences don't come in to play (e.g. 1+1=2), you should include a small chance that some extremely unlikely event may happen. The example tossed around is "the probability that the sun will not rise tomorrow." Prior recorded observations show that sun has always risen, but that doesn't mean that there's a 1.0 probability of the sun rising tomorrow (e.g. some extremely unlikely cataclysmic event may prevent the sun from rising).</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(Laplace'?s? rule of succession|Laplace'?s? rule)/i"></span> - If some independent true/false event occurs n times, and s of those n times were successes, it's natural for people to assume the probability of success is <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mfrac><mi>s</mi><mi>n</mi></mfrac></mrow><annotation encoding="application/x-tex">\frac{s}{n}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.040392em;vertical-align:-0.345em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.695392em;"><span style="top:-2.6550000000000002em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">n</span></span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.394em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">s</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.345em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span>. However, if the number of successes is 0, the probability would be 0.0. Cromwell's rule states that when a probability is based off past events, 0.0 and 1.0 shouldn't be used. As such, a more appropriate / meaningful measure of probability is <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mfrac><mrow><mi>s</mi><mo>+</mo><mn>1</mn></mrow><mrow><mi>n</mi><mo>+</mo><mn>2</mn></mrow></mfrac></mrow><annotation encoding="application/x-tex">\frac{s+1}{n+2}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.2484389999999999em;vertical-align:-0.403331em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.845108em;"><span style="top:-2.655em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">n</span><span class="mbin mtight">+</span><span class="mord mtight">2</span></span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.394em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">s</span><span class="mbin mtight">+</span><span class="mord mtight">1</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.403331em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span>.</p>
<p>For example, imagine you're sitting on a park bench having lunch. Of the 8 birds you've seen since starting your lunch, all have been pigeons. If you were to calculate the probability that the next bird you'll see a crow, <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mfrac><mn>0</mn><mn>8</mn></mfrac></mrow><annotation encoding="application/x-tex">\frac{0}{8}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.190108em;vertical-align:-0.345em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.845108em;"><span style="top:-2.6550000000000002em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">8</span></span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.394em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">0</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.345em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span> would be flawed because it states that there's no chance that the next bird will be a crow (there obviously is a chance, but it may be a small chance). Instead, applying Laplace's rule allows for the small probability that a crow may be seen next: <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mfrac><mrow><mn>0</mn><mo>+</mo><mn>1</mn></mrow><mrow><mn>8</mn><mo>+</mo><mn>2</mn></mrow></mfrac></mrow><annotation encoding="application/x-tex">\frac{0+1}{8+2}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.2484389999999999em;vertical-align:-0.403331em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.845108em;"><span style="top:-2.655em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">8</span><span class="mbin mtight">+</span><span class="mord mtight">2</span></span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.394em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">0</span><span class="mbin mtight">+</span><span class="mord mtight">1</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.403331em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span>.</p>
<p>Laplace's rule of succession is more meaningful when the number of trials (n) is small.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="pseudocount"></span> - When a zero is replaced with a small number to prevent unfair scoring. See Laplace's rule of succession.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="Las Vegas algorithm"></span> - A randomized algorithm that delivers a guaranteed exact solution. That is, even though the algorithm makes random decisions it is guaranteed to converge on the exact solution to the problem its trying to solve (not an approximate solution).</p>
<p>An example of a Las Vegas algorithm is randomized quicksort (randomness is applied when choosing the pivot).</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="Monte Carlo algorithm"></span> - A randomized algorithm that delivers an approximate solution. Because these algorithms are quick, they're typically run many times. The approximation considered the best out of all runs is the one that gets chosen as the solution.</p>
<p>An example of a Monte Carlo algorithm is a genetic algorithm to optimize the weights of a deep neural network. That is, a step of the optimization requires running n different neural networks to see which gives the best result, then replacing those n networks with n copies of the best performing network where each copy has randomly tweaked weights. At some point the algorithm will stop producing incrementally better results.</p>
<p>Perform the optimization (the entire thing, not just a single step) thousands of times and pick the best network.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(consensus string|consensus sequence|consensus)/i"></span> - The k-mer generated by selecting the most abundant column at each index of a motif matrix.</p>
<table>
<thead>
<tr>
<th></th>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
</tr>
</thead>
<tbody>
<tr>
<td>k-mer 1</td>
<td>A</td>
<td>T</td>
<td>T</td>
<td>G</td>
<td>C</td>
</tr>
<tr>
<td>k-mer 2</td>
<td>A</td>
<td>T</td>
<td>T</td>
<td>C</td>
<td>C</td>
</tr>
<tr>
<td>k-mer 3</td>
<td>T</td>
<td>T</td>
<td>T</td>
<td>G</td>
<td>C</td>
</tr>
<tr>
<td>k-mer 4</td>
<td>T</td>
<td>T</td>
<td>T</td>
<td>C</td>
<td>C</td>
</tr>
<tr>
<td>k-mer 5</td>
<td>A</td>
<td>T</td>
<td>T</td>
<td>C</td>
<td>G</td>
</tr>
<tr>
<td>consensus</td>
<td>A</td>
<td>T</td>
<td>T</td>
<td>C</td>
<td>C</td>
</tr>
</tbody>
</table>
<p>The generated k-mer may also use a hybrid alphabet. The consensus string for the same matrix above using IUPAC nucleotide codes: <code>WTTSS</code>.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="entropy"></span> - The uncertainty associated with a random variable. Given some set of outcomes for a variable, it's calculated as <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo>−</mo><msubsup><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></msubsup><mi>P</mi><mo stretchy="false">(</mo><msub><mi>x</mi><mi>i</mi></msub><mo stretchy="false">)</mo><mi>l</mi><mi>o</mi><mi>g</mi><mi>P</mi><mo stretchy="false">(</mo><msub><mi>x</mi><mi>i</mi></msub><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">-\sum_{i=1}^{n} P(x_i) log P(x_i)</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.104002em;vertical-align:-0.29971000000000003em;"></span><span class="mord">−</span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mop"><span class="mop op-symbol small-op" style="position:relative;top:-0.0000050000000000050004em;">∑</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.804292em;"><span style="top:-2.40029em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">i</span><span class="mrel mtight">=</span><span class="mord mtight">1</span></span></span></span><span style="top:-3.2029em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">n</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.29971000000000003em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord mathnormal" style="margin-right:0.13889em;">P</span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal">x</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mclose">)</span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord mathnormal">o</span><span class="mord mathnormal" style="margin-right:0.03588em;">g</span><span class="mord mathnormal" style="margin-right:0.13889em;">P</span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal">x</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mclose">)</span></span></span></span>.</p>
<p>This definition is for information theory. In other contexts (e.g. physics, economics), this term has a different meaning.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="genome"></span> - In the context of a ...</p>
<ul>
<li>specific organism, genome refers to all of the DNA for that organism (e.g. a specific E. coli cell).</li>
<li>specific species, genome refers to the idealized DNA for that species (e.g. all E. coli).</li>
</ul>
<p>DNA of individual cells mutate all the time. For example, even in multi-cell organism, two cells from the same mouse may not have the exact same DNA.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="sequence"></span> - The ordered elements that make up some biological entity. For example, a ...</p>
<ul>
<li>DNA sequence contains the set of nucleotides and their positions for that DNA strand.</li>
<li>peptide sequence contains the set of amino acids and their positions for that peptide.</li>
</ul>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(sequencing|sequenced)/i"></span> - The process of determining which nucleotides are assigned to which positions in a strand of DNA or RNA.</p>
<p>The machinery used for DNA sequencing is called a sequencer. A sequencer takes multiple copies of the same DNA, breaks that DNA up into smaller fragments, and scans in those fragments. Each fragment is typically the same length but has a unique starting offset. Because the starting offsets are all different, the original larger DNA sequence can be guessed at by finding fragment with overlapping regions and stitching them together.</p>
<table>
<thead>
<tr>
<th></th>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
<th>6</th>
<th>7</th>
<th>8</th>
<th>9</th>
</tr>
</thead>
<tbody>
<tr>
<td>read 1</td>
<td></td>
<td></td>
<td></td>
<td></td>
<td>C</td>
<td>T</td>
<td>T</td>
<td>C</td>
<td>T</td>
<td>T</td>
</tr>
<tr>
<td>read 2</td>
<td></td>
<td></td>
<td></td>
<td>G</td>
<td>C</td>
<td>T</td>
<td>T</td>
<td>C</td>
<td>T</td>
<td></td>
</tr>
<tr>
<td>read 3</td>
<td></td>
<td></td>
<td>T</td>
<td>G</td>
<td>C</td>
<td>T</td>
<td>T</td>
<td>C</td>
<td></td>
<td></td>
</tr>
<tr>
<td>read 4</td>
<td></td>
<td>T</td>
<td>T</td>
<td>G</td>
<td>C</td>
<td>T</td>
<td>T</td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td>read 5</td>
<td>A</td>
<td>T</td>
<td>T</td>
<td>G</td>
<td>C</td>
<td>T</td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td>reconstructed</td>
<td>A</td>
<td>T</td>
<td>T</td>
<td>G</td>
<td>C</td>
<td>T</td>
<td>T</td>
<td>C</td>
<td>T</td>
<td>T</td>
</tr>
</tbody>
</table>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="sequencer"></span> - A machine that performs DNA or RNA sequencing.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(sequencing error|sequencer error)/i"></span> - An error caused by a sequencer returning a fragment where a nucleotide was misinterpreted at one or more positions (e.g. offset 3 was actually a C but it got scanned in as a G).</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>This term may also be used in reference to homopolymer errors, known to happen with nanopore technology. From <a href="https://news.ycombinator.com/item?id=25459670">here</a>...</p>
<blockquote>
<p>A homopolymer is when you have stretches of the same nucleotide, and the error is miscounting the number of them. e.g: GAAAC could be called as "GAAC" or "GAAAAC" or even "GAAAAAAAC".</p>
</blockquote>
</div>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="read"></span> - A segment of genome scanned in during the process of sequencing.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(read-pair|read pair)/i"></span> - A segment of genome scanning in during the process of sequencing, where the middle of the segment is unknown. That is, the first k elements and the last k elements are known, but the d elements in between aren't known. The total size of the segment is 2k + d.</p>
<p>Sequencers provide read-pairs as an alternative to longer reads because the longer a read is the more errors it contains.</p>
<p>See kd-mer.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="fragment"></span> - A scanned sequence returned by a sequencer. Represented as either a read or a read-pair.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(assembly|assemble)/i"></span> - The process of stitching together overlapping fragments to guess the sequence of the original larger DNA sequence that those fragments came from.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(hybrid alphabet|alternate alphabet|alternative alphabet)/i"></span> - When representing a sequence that isn't fully conserved, it may be more appropriate to use an alphabet where each letter can represent more than 1 nucleotide. For example, the IUPAC nucleotide codes provides the following alphabet:</p>
<ul>
<li>A = A</li>
<li>C = C</li>
<li>T = T</li>
<li>G = G</li>
<li>W = A or T</li>
<li>S = G or C</li>
<li>K = G or T</li>
<li>Y = C or T</li>
<li>...</li>
</ul>
<p>If the sequence being represented can be either AAAC or AATT, it may be easier to represent a single string of AAWY.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(IUPAC nucleotide code|IUPAC)/i"></span> - A hybrid alphabet with the following mapping:</p>
<table>
<thead>
<tr>
<th>Letter</th>
<th>Base</th>
</tr>
</thead>
<tbody>
<tr>
<td>A</td>
<td>Adenine</td>
</tr>
<tr>
<td>C</td>
<td>Cytosine</td>
</tr>
<tr>
<td>G</td>
<td>Guanine</td>
</tr>
<tr>
<td>T (or U)</td>
<td>Thymine (or Uracil)</td>
</tr>
<tr>
<td>R</td>
<td>A or G</td>
</tr>
<tr>
<td>Y</td>
<td>C or T</td>
</tr>
<tr>
<td>S</td>
<td>G or C</td>
</tr>
<tr>
<td>W</td>
<td>A or T</td>
</tr>
<tr>
<td>K</td>
<td>G or T</td>
</tr>
<tr>
<td>M</td>
<td>A or C</td>
</tr>
<tr>
<td>B</td>
<td>C or G or T</td>
</tr>
<tr>
<td>D</td>
<td>A or G or T</td>
</tr>
<tr>
<td>H</td>
<td>A or C or T</td>
</tr>
<tr>
<td>V</td>
<td>A or C or G</td>
</tr>
<tr>
<td>N</td>
<td>any base</td>
</tr>
<tr>
<td>. or -</td>
<td>gap</td>
</tr>
</tbody>
</table>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(logo|sequence logo|motif logo)/i"></span> - A graphical representation of how conserved a sequence's positions are. Each position has its possible nucleotides stacked on top of each other, where the height of each nucleotide is based on how conserved it is. The more conserved a position is, the taller that column will be.</p>
<p>Typically applied to DNA or RNA, and May also be applied to other biological sequence types (e.g. amino acids).</p>
<p>The following is an example of a logo generated from a motif sequence:</p>
<p><img src="motif_logo.svg" alt="Example Image"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(transposon|transposable element|jumping gene)/i"></span> - A DNA sequence that can change its position within a genome, altering the genome size. They come in two flavours:</p>
<ul>
<li>Class I (retrotransposon) - Behaves similarly to copy-and-paste where the sequence is duplicated. DNA is transcribed to RNA, followed by that RNA being reverse transcribed back to DNA by an enzyme called reverse transcriptase.</li>
<li>Class II (DNA transposon) - Behaves similarly to cut-and-paste where the sequence is moved. DNA is physically cut out by an enzyme called transposases and placed back in at some other location.</li>
</ul>
<p>Often times, transposons cause disease. For example, ...</p>
<ul>
<li>insertion of a transposon into a gene will likely disable that gene.</li>
<li>after a transposon leaves a gene, the gap likely won't be repaired correctly.</li>
</ul>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="adjacency list"></span> - An internal representation of a graph where each node has a list of pointers to other nodes that it can forward to.</p>
<p><img src="svgbob_5b40f65c78645193c91742f02934854ae8846ede.svg" alt="Kroki diagram output"></p>
<p>The graph above represented as an adjacency list would be...</p>
<table>
<thead>
<tr>
<th>From</th>
<th>To</th>
</tr>
</thead>
<tbody>
<tr>
<td>A</td>
<td>B</td>
</tr>
<tr>
<td>B</td>
<td>C</td>
</tr>
<tr>
<td>C</td>
<td>D,E</td>
</tr>
<tr>
<td>D</td>
<td>F</td>
</tr>
<tr>
<td>E</td>
<td>D,F</td>
</tr>
<tr>
<td>F</td>
<td></td>
</tr>
</tbody>
</table>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="adjacency matrix"></span> - An internal representation of a graph where a matrix defines the number of times that each node forwards to every other node.</p>
<p><img src="svgbob_5b40f65c78645193c91742f02934854ae8846ede.svg" alt="Kroki diagram output"></p>
<p>The graph above represented as an adjacency matrix would be...</p>
<table>
<thead>
<tr>
<th></th>
<th>A</th>
<th>B</th>
<th>C</th>
<th>D</th>
<th>E</th>
<th>F</th>
</tr>
</thead>
<tbody>
<tr>
<td>A</td>
<td>0</td>
<td>1</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
<tr>
<td>B</td>
<td>0</td>
<td>0</td>
<td>1</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
<tr>
<td>C</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>1</td>
<td>1</td>
<td>0</td>
</tr>
<tr>
<td>D</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>1</td>
</tr>
<tr>
<td>E</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>1</td>
<td>0</td>
<td>1</td>
</tr>
<tr>
<td>F</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
</tbody>
</table>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(Hamiltonian path|Hamilton path)/i"></span> - A path in a graph that visits every node exactly once.</p>
<p>The graph below has the Hamiltonian path ABCEDF.</p>
<p><img src="svgbob_5b40f65c78645193c91742f02934854ae8846ede.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="Eulerian path"></span> - A path in a graph that visits every edge exactly once.</p>
<p>In the graph below, the Eulerian path is (A,B), (B,C), (C,D), (D,E), (E,C), (C,D), (D,F).</p>
<p><img src="svgbob_87660b191d813d39c396985391b70507c8c1236b.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="Eulerian cycle"></span> - An Eulerian path that forms a cycle. That is, a path in a graph that is a cycle and visits every edge exactly once.</p>
<p>The graph below has an Eulerian cycle of (A,B), (B,C) (C,D), (D,F), (F,C), (C,A).</p>
<p><img src="svgbob_244dd1930eb16ef58be2261f1ffb9277ad2d793a.svg" alt="Kroki diagram output"></p>
<p>If a graph contains an Eulerian cycle, it's said to be an Eulerian graph.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="Eulerian graph"></span> <span class="anki-answerpattern" data-pattern="Eulerian cycle"></span> - For a graph to be a Eulerian graph, it must have an Eulerian cycle: a path in a graph that is a cycle and visits every edge exactly once. For a graph to have an Eulerian cycle, it must be both balanced and strongly connected.</p>
<p><img src="svgbob_d0af75d07c2969afc5f4088d18d3b7a327f26ef2.svg" alt="Kroki diagram output"></p>
<p>Note how in the graph above, ...</p>
<ul>
<li>
<p>every node is reachable from every other node (strongly connected),</p>
</li>
<li>
<p>every node has an outdegree equal to its indegree (balanced).</p>
<table>
<thead>
<tr>
<th>Node</th>
<th>Indegree</th>
<th>Outdegree</th>
</tr>
</thead>
<tbody>
<tr>
<td>A</td>
<td>1</td>
<td>1</td>
</tr>
<tr>
<td>B</td>
<td>1</td>
<td>1</td>
</tr>
<tr>
<td>C</td>
<td>2</td>
<td>2</td>
</tr>
<tr>
<td>D</td>
<td>1</td>
<td>1</td>
</tr>
<tr>
<td>F</td>
<td>1</td>
<td>1</td>
</tr>
</tbody>
</table>
</li>
</ul>
<p>In contrast, the following graphs are not Eulerian graphs (no Eulerian cycles exist):</p>
<ul>
<li>
<p>Strongly connected but not balanced.</p>
<p><img src="svgbob_efdff1b3ec66000b9acf2ddc84c2d3684d318683.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>Balanced but not strongly connected.</p>
<p><img src="svgbob_48abea7e0377d91bfd845a85b2ee9615e9852bfd.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>Balanced but disconnected (not strongly connected).</p>
<p><img src="svgbob_1f6d7503fdb3f3654b3c64f041254352c8d9a43e.svg" alt="Kroki diagram output"></p>
</li>
</ul>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="disconnected"></span> <span class="anki-answerpattern" data-pattern="connected"></span> - A graph is disconnected if you can break it out into 2 or more distinct sub-graphs without breaking any paths. In other words, the graph contains at least two nodes which aren't contained in any path.</p>
<p>The graph below is disconnected because there is no path that contains E, F, G, or H and A, B, C, or D.</p>
<p><img src="svgbob_754a446a49b8209971f339d76f4fcc492acbb615.svg" alt="Kroki diagram output"></p>
<p>The graph below is connected.</p>
<p><img src="svgbob_028c9291a4412afebe0b605b464987e2c9bf75ef.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="strongly connected"></span> - A graph is strongly connected if every node is reachable from every other node.</p>
<p>The graph below is <strong>not</strong> strongly connected because neither A nor B is reachable by C, D, E, or F.</p>
<p><img src="svgbob_5b40f65c78645193c91742f02934854ae8846ede.svg" alt="Kroki diagram output"></p>
<p>The graph below is strongly connected because all nodes are reachable from all nodes.</p>
<p><img src="svgbob_244dd1930eb16ef58be2261f1ffb9277ad2d793a.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="indegree"></span> <span class="anki-answerpattern" data-pattern="outdegree"></span> - The number of edges leading into / out of a node of a directed graph.</p>
<p>The node below has an indegree of 3 and an outdegree of 1.</p>
<p><img src="svgbob_7159ebaf654fe57b46d2f14a29a73d50e280afe6.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="balanced node"></span> - A node of a directed graph that has an equal indegree and outdegree. That is, the number of edges coming in is equal to the number of edges going out.</p>
<p>The node below has an indegree and outdegree of 1. It is a balanced node.</p>
<p><img src="svgbob_7c49a4fbe8fc85173147545e7a459bca61c84f26.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="balanced graph"></span> - A directed graph where ever node is balanced.</p>
<p>The graph below is balanced graph because each node has an equal indegree and outdegree.</p>
<p><img src="svgbob_244dd1930eb16ef58be2261f1ffb9277ad2d793a.svg" alt="Kroki diagram output"></p>
<table>
<thead>
<tr>
<th>Node</th>
<th>Indegree</th>
<th>Outdegree</th>
</tr>
</thead>
<tbody>
<tr>
<td>A</td>
<td>1</td>
<td>1</td>
</tr>
<tr>
<td>B</td>
<td>1</td>
<td>1</td>
</tr>
<tr>
<td>C</td>
<td>2</td>
<td>2</td>
</tr>
<tr>
<td>D</td>
<td>1</td>
<td>1</td>
</tr>
<tr>
<td>F</td>
<td>1</td>
<td>1</td>
</tr>
</tbody>
</table>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="overlap graph"></span> - A graph representing the k-mers making up a string. Specifically, the graph is built in 2 steps:</p>
<ol>
<li>
<p>Each node is a fragment.</p>
<p><img src="svgbob_ba61508d2d4f3dfef58ce5783f8e22f49ddec891.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>Each edge is between overlapping fragments (nodes), where the ...</p>
<ul>
<li>source node has the overlap in its suffix .</li>
<li>destination node has the overlap in its prefix.</li>
</ul>
<p><img src="svgbob_679d24959920acb04c059d6e53dee545fff77300.svg" alt="Kroki diagram output"></p>
</li>
</ol>
<p>Overlap graphs used for genome assembly.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="de Bruijn graph"></span> - A special graph representing the k-mers making up a string. Specifically, the graph is built in 2 steps:</p>
<ol>
<li>
<p>Each k-mer is represented as an edge connecting 2 nodes. The ...</p>
<ul>
<li>source node represents the first 0 to n-1 elements of the k-mer,</li>
<li>destination node represents last 1 to n elements of the k-mer,</li>
<li>and edge represents the k-mer.</li>
</ul>
<p>For example, ...</p>
<p><img src="svgbob_8e6112b97be535a52f2be925bb5b56b546699aaa.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>Each node representing the same value is merged together to form the graph.</p>
<p>For example, ...</p>
<p><img src="svgbob_cfa519e3b0b53ea602d7a16720f54ea9c85f3f08.svg" alt="Kroki diagram output"></p>
</li>
</ol>
<p>De Bruijn graphs are used for genome assembly. It's much faster to assemble a genome from a de Bruijn graph than it is to from an overlap graphs.</p>
<p>De Bruijn graphs were originally invented to solve the k-universal string problem.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(k-universal string|k-universal)/i"></span> - For some alphabet and k, a string is considered k-universal if it contains every k-mer for that alphabet exactly once. For example, for an alphabet containing only 0 and 1 (binary) and k=3, a 3-universal string would be 0001110100 because it contains every 3-mer exactly once:</p>
<ul>
<li>000: <strong>000</strong>1110100</li>
<li>001: 0<strong>001</strong>110100</li>
<li>010: 000111<strong>010</strong>0</li>
<li>011: 00<strong>011</strong>10100</li>
<li>100: 0001110<strong>100</strong></li>
<li>101: 00011<strong>101</strong>00</li>
<li>110: 0001<strong>110</strong>100</li>
<li>111: 000<strong>111</strong>0100</li>
</ul>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>This is effectively assembly. There are a set of k-mers and they're being stitched together to form a larger string. The only difference is that the elements aren't nucleotides.</p>
</div>
<p>De Bruijn graphs were invented in an effort to construct k-universal strings for arbitrary values of k. For example, given the k-mers in the example above (000, 001, ...), a k-universal string can be found by constructing a de Bruijn graph from the k-mers and finding a Eulerian cycle in that graph.</p>
<p><img src="svgbob_e0f100e748698e3c9c3adfa857869f920a7fd19a.svg" alt="Kroki diagram output"></p>
<p>There are multiple Eulerian cycles in the graph, meaning that there are multiple 3-universal strings:</p>
<ul>
<li>0001110100</li>
<li>0011101000</li>
<li>1110001011</li>
<li>1100010111</li>
<li>...</li>
</ul>
<p>For larger values of k (e.g. 20), finding k-universal strings would be too computationally intensive without De Bruijn graphs and Eulerian cycles.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="coverage"></span> - Given a substring from some larger sequence that was reconstructed from a set of fragments, the coverage of that substring is the number of reads used to construct it. The substring length is typically 1: the coverage for each position of the sequence.</p>
<p><img src="svgbob_327ff8af96b87d3be0b99ef88935217e60fa80cf.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(read breaking|read-breaking|breaking reads)/i"></span> - The concept of taking multiple reads and breaking them up into smaller reads.</p>
<p><img src="svgbob_d9e103b16312cfe660d19df9305a9361ac1241ac.svg" alt="Kroki diagram output"></p>
<p>When read breaking, smaller k-mers result in better coverage but also make the de Bruijn graph more tangled. The more tangled the de Bruijn graph is, the harder it is to infer the full sequence.</p>
<p>In the example above, the average coverage...</p>
<ul>
<li>for the left-hand side (original) is 2.1.</li>
<li>for the right-hand side (broken) is 4.</li>
</ul>
<p>See also: read-pair breaking.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>What purpose does this actually serve? Mimicking 1 long read as n shorter reads isn't equivalent to actually having sequenced those n shorter reads. For example, what if the longer read being broken up has an error? That error replicates when breaking into n shorter reads, which gives a false sense of having good coverage and makes it seems as if it wasn't an error.</p>
</div>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(read-pair breaking|read pair breaking|breaking read-pairs|breaking read pairs)/i"></span> - The concept of taking multiple read-pairs and breaking them up into read-pairs with a smaller k.</p>
<p><img src="svgbob_eb8f1251b4e293a9662ade2c0a4452a7c7884f2f.svg" alt="Kroki diagram output"></p>
<p>When read-pair breaking, a smaller k results in better coverage but also make the de Bruijn graph more tangled. The more tangled the de Bruijn graph is, the harder it is to infer the full sequence.</p>
<p>In the example above, the average coverage...</p>
<ul>
<li>for the left-hand side (original) is 1.6.</li>
<li>for the right-hand side (broken) is 2.5.</li>
</ul>
<p>See also: read breaking.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>What purpose does this actually serve? Mimicking 1 long read-pair as n shorter read-pairs isn't equivalent to actually having sequenced those n shorter read-pairs. For example, what if the longer read-pair being broken up has an error? That error replicates when breaking into n shorter read-pairs, which gives a false sense of having good coverage and makes it seems as if it wasn't an error.</p>
</div>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="contig"></span> - An unambiguous stretch of DNA derived by searching an overlap graph / de Bruijn graph for paths that are the longest possible stretches of non-branching nodes (indegree and outdegree of 1). Each stretch will be a path that's either  ...</p>
<ul>
<li>
<p>a line: each node has an indegree and outdegree of 1.</p>
<p><img src="svgbob_8f50e668ad56e2eeff3c39e6aa89792f6f9444f9.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>a cycle: each node has an indegree and outdegree of 1 and it loops.</p>
<p><img src="svgbob_9a5c6dc10a8a016d6f43d4c9bdd6243d0ee5b332.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>a line sandwiched between branching nodes: nodes in between have an indegree and outdegree of 1 but either...</p>
<ul>
<li>starts at a node where indegree != 1 but outdegree == 1 (incoming branch),</li>
<li>or ends at a node where indegree == 1 but outdegree != 1 (outgoing branch),</li>
<li>or both.</li>
</ul>
<p><img src="svgbob_057bcd0e831c50ca2384b8da9ea9cd1f46186b4a.svg" alt="Kroki diagram output"></p>
</li>
</ul>
<p>Real-world complications with DNA sequencing make de Bruijn / overlap graphs too tangled to guess a full genome: both strands of double-stranded DNA are sequenced and mixed into the graph, sequencing errors make into the graph, repeats regions of the genome can't be reliably handled by the graph, poor coverage, etc.. As such, biologists / bioinformaticians have no choice but to settle on contigs.</p>
<p><img src="svgbob_db2c0ed38b3144b9db68bc5e39f4b97ce471f510.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="ribonucleotide"></span> - Elements that make up RNA, similar to how nucleotides are the elements that make up DNA.</p>
<ul>
<li>A = Adenine (same as nucleotide)</li>
<li>C = Cytosine (same as nucleotide)</li>
<li>G = Guanine (same as nucleotide)</li>
<li>U = Uracil (replace nucleotide Thymine)</li>
</ul>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="antibiotic"></span> - A substance (typically an enzyme) for killing, preventing, or inhibiting the grow of bacterial infections.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="amino acid"></span> - The building blocks of peptides / proteins, similar to how nucleotides are the building blocks of DNA.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>Answer shouldn't be plural.</p>
</div>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="proteinogenic amino acid"></span> - Amino acids that are used during translation. These are the 20 amino acids that the ribosome translates from codons. In contrast, there are many other non-proteinogenic amino acids that are used for non-ribosomal peptides.</p>
<p>The term "proteinogenic" means "protein creating".</p>
<table>
<thead>
<tr>
<th>1 Letter Code</th>
<th>3 Letter Code</th>
<th>Amino Acid</th>
<th>Mass (daltons)</th>
</tr>
</thead>
<tbody>
<tr>
<td>A</td>
<td>Ala</td>
<td>Alanine</td>
<td>71.04</td>
</tr>
<tr>
<td>C</td>
<td>Cys</td>
<td>Cysteine</td>
<td>103.01</td>
</tr>
<tr>
<td>D</td>
<td>Asp</td>
<td>Aspartic acid</td>
<td>115.03</td>
</tr>
<tr>
<td>E</td>
<td>Glu</td>
<td>Glutamic acid</td>
<td>129.04</td>
</tr>
<tr>
<td>F</td>
<td>Phe</td>
<td>Phenylalanine</td>
<td>147.07</td>
</tr>
<tr>
<td>G</td>
<td>Gly</td>
<td>Glycine</td>
<td>57.02</td>
</tr>
<tr>
<td>H</td>
<td>His</td>
<td>Histidine</td>
<td>137.06</td>
</tr>
<tr>
<td>I</td>
<td>Ile</td>
<td>Isoleucine</td>
<td>113.08</td>
</tr>
<tr>
<td>K</td>
<td>Lys</td>
<td>Lysine</td>
<td>128.09</td>
</tr>
<tr>
<td>L</td>
<td>Leu</td>
<td>Leucine</td>
<td>113.08</td>
</tr>
<tr>
<td>M</td>
<td>Met</td>
<td>Methionine</td>
<td>131.04</td>
</tr>
<tr>
<td>N</td>
<td>Asn</td>
<td>Asparagine</td>
<td>114.04</td>
</tr>
<tr>
<td>P</td>
<td>Pro</td>
<td>Proline</td>
<td>97.05</td>
</tr>
<tr>
<td>Q</td>
<td>Gln</td>
<td>Glutamine</td>
<td>128.06</td>
</tr>
<tr>
<td>R</td>
<td>Arg</td>
<td>Arginine</td>
<td>156.1</td>
</tr>
<tr>
<td>S</td>
<td>Ser</td>
<td>Serine</td>
<td>87.03</td>
</tr>
<tr>
<td>T</td>
<td>Thr</td>
<td>Threonine</td>
<td>101.05</td>
</tr>
<tr>
<td>V</td>
<td>Val</td>
<td>Valine</td>
<td>99.07</td>
</tr>
<tr>
<td>W</td>
<td>Trp</td>
<td>Tryptophan</td>
<td>186.08</td>
</tr>
<tr>
<td>Y</td>
<td>Tyr</td>
<td>Tyrosine</td>
<td>163.06</td>
</tr>
</tbody>
</table>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The masses are monoisotopic masses.</p>
</div>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="peptide"></span> - A short amino acid chain of at least size two. Peptides are considered miniature proteins, but when something should be called a peptide vs a protein is loosely defined: the cut-off is anywhere between 50 to 100 amino acids.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="polypeptide"></span> - A peptide of at least size 10.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(amino acid residue|residue)/i"></span> - The part of an amino acid that makes it unique from all others.</p>
<p>When two or more amino acids combine to make a peptide/protein, specific elements are removed from each amino acid. What remains of each amino acid is the amino acid residue.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(cyclopeptide|cyclic peptide)/i"></span> - A peptide that doesn't have a start / end. It loops.</p>
<p><img src="svgbob_0de8076d92d299650e3473fc869328891e1ef452.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="linear peptide"></span> - A peptide that has a start and an end. It doesn't loop.</p>
<p><img src="svgbob_1a5d3942fce4284a9ad77c8e4624b7b21573543d.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="subpeptide"></span> - A peptide derived taking some contiguous piece of a larger peptide. A subpeptide can have a length == 1 where a peptide must have a length &gt; 1.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="central dogma of molecular biology"></span> - The overall concept of transcription and translation: Instructions for making a protein are copied from DNA to RNA, then RNA feeds into the ribosome to make that protein (DNA → RNA → Protein).</p>
<p>Most, not all, peptides are synthesized as described above. Non-ribosomal peptides are synthesized outside of the transcription and translation.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="non-ribosomal peptide"></span> - A peptide that was synthesized by a protein called NRP synthetase rather than synthesized by a ribosome. NRP synthetase builds peptides one amino acid at a time without relying on transcription or translation.</p>
<p>Non-ribosomal peptides may be cyclic. Common use-cases for non-ribosomal peptides:</p>
<ul>
<li>antibiotics</li>
<li>anti-tumor agents</li>
<li>immunosuppressors</li>
<li>communication between bacteria (quorum sensing)</li>
</ul>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(non-ribosomal peptide synthetase|NRP synthetase)/i"></span> - A protein responsible for the production of a non-ribosomal peptide.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="adenylation domain"></span> <span class="anki-answerpattern" data-pattern="(A[\s-]domain)/i"></span> - A segment of an NRP synthetase protein responsible for the outputting a single amino acid. For example, the NRP synthetase responsible for producing Tyrocidine has 10 adenylation domains, each of which is responsible for outputting a single amino acid of Tyrocidine.</p>
<p>Adenylation domain may be shortened to A-domain.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(mass spectrometer|mass spectrometry)/i"></span> - A device that randomly shatters molecules into pieces and measures the mass-to-charge of those pieces. The output of the device is a plot called a spectrum.</p>
<p>Note that mass spectrometers have various real-world practical problems. Specifically, they ...</p>
<ul>
<li>may not capture all possible pieces from the intended molecules (missing mass-to-charge ratios).</li>
<li>may capture pieces from unintended molecules (faulty mass-to-charge ratios).</li>
<li>will likely introduce noise into the pieces they capture.</li>
</ul>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(spectrum)/i"></span> - The output of a mass spectrometer. The...</p>
<ul>
<li>x-axis is the mass-to-charge ratio.</li>
<li>y-axis is the intensity of that mass-to-charge ratio (how much more / less did that mass-to-charge appear compared to the others).</li>
</ul>
<p><img src="svgbob_2b5ef43ae7572c2c6bf63687c53860a71830d160.svg" alt="Kroki diagram output"></p>
<p>Note that mass spectrometers have various real-world practical problems. Specifically, they ...</p>
<ul>
<li>may not capture all possible pieces from the intended molecules (missing mass-to-charge ratios).</li>
<li>may capture pieces from unintended molecules (faulty mass-to-charge ratios).</li>
<li>will likely introduce noise into the pieces they capture.</li>
</ul>
<p>As such, these plots aren't exact.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(mass-?to-?charge ratio|mass-?to-?charge)/i"></span> <span class="anki-answerpattern" data-pattern="(intensity|relative abundance|abundance)/i"></span></p>
<p>In a mass spectrometer output, the...</p>
<ul>
<li>x-axis is the mass-to-charge ratio. <span class="anki-listrandomorder"></span>
</li>
<li>y-axis is the intensity.</li>
</ul>
<p><img src="svgbob_431bd1b9afda2c818011f2bb4b03a8d0c60b7fe5.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="experimental spectrum"></span> - List of potential fragment masses derived from a spectrum. That is, the molecules fed into the mass spectrometer were randomly fragmented and each fragment had its mass-to-charge ratio measured. From there, each mass-to-charge ratio was converted a set of potential masses.</p>
<p>The masses in an experimental spectrum ...</p>
<ul>
<li>may not capture all possible fragments for the intended molecule (missing masses).</li>
<li>may capture fragments from unintended molecules (faulty masses).</li>
<li>will likely contain noise.</li>
</ul>
<p>In the context of peptides, the mass spectrometer is expected to fragment based on the bonds holding the individual amino acids together. For example, given the linear peptide NQY, the experimental spectrum may include the masses for [N, Q, ?, ?, QY, ?, NQY] (? indicate faulty masses, Y and NQ missing).</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="theoretical spectrum"></span> - List of all of possible fragment masses for a molecule in addition to 0 and the mass of the entire molecule. This is what the experimental spectrum would be in a perfect world: no missing masses, no faulty masses, no noise, only a single possible mass for each mass-to-charge ratio.</p>
<p>In the context of peptides, the mass spectrometer is expected to fragment based on the bonds holding the individual amino acids together. For example, given the linear peptide NQY, the theoretical spectrum will include the masses for [0, N, Q, Y, NQ, QY, NQY]. It shouldn't include masses for partial amino acids. For example, it shouldn't include NQY breaking into 2 pieces by splitting Q, such that one half has N and part of Q, and the other has the remaining part of Q with Y.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="spectrum convolution"></span> - An operation used to derive amino acid masses that probably come from the peptide used to generate that experimental spectrum. That is, it generates a list of amino acid masses that could have been for the peptide that generated the experimental spectrum.</p>
<p>The operation derives amino acid masses by subtracting experimental spectrum masses from each other. For example, the following experimental spectrum is for the linear peptide NQY: [113.9, 115.1, 136.2, 162.9, 242.0, 311.1, 346.0, 405.2]. Performing 242.0 - 113.9 results in 128.1, which is very close to the mass for amino acid Y.</p>
<p>Note how the mass for Y was derived from the masses in experimental spectrum even though it's missing from the experimental spectrum itself:</p>
<ul>
<li>Mass of N is 114. 2 masses are close to 114 in the experimental spectrum: [113.9, 115.1].</li>
<li>Mass of Q is 163. 1 mass is close to 163 in the experimental spectrum: [162.9].</li>
<li>Mass of Y is 128. 0 masses are close to 128 in the experimental spectrum: [].</li>
</ul>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="dalton"></span> - A unit of measurement used in physics and chemistry. 1 Dalton is approximately the mass of a single proton / neutron, derived by taking the mass of a carbon-12 atom and dividing it by 12.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(codon)/i"></span> - A sequence of 3 ribonucleotides that maps to an amino acid or a stop marker. During translation, the ribosome translates the RNA to a protein 3 ribonucleotides at a time:</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The stop marker tells the ribosome to stop translating / the protein is complete.</p>
</div>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The codons are listed as ribonucleotides (RNA). For nucleotides (DNA), swap U with T.</p>
</div>
<table>
<thead>
<tr>
<th>1 Letter Code</th>
<th>3 Letter Code</th>
<th>Amino Acid</th>
<th>Codons</th>
</tr>
</thead>
<tbody>
<tr>
<td>A</td>
<td>Ala</td>
<td>Alanine</td>
<td>GCA, GCC, GCG, GCU</td>
</tr>
<tr>
<td>C</td>
<td>Cys</td>
<td>Cysteine</td>
<td>UGC, UGU</td>
</tr>
<tr>
<td>D</td>
<td>Asp</td>
<td>Aspartic acid</td>
<td>GAC, GAU</td>
</tr>
<tr>
<td>E</td>
<td>Glu</td>
<td>Glutamic acid</td>
<td>GAA, GAG</td>
</tr>
<tr>
<td>F</td>
<td>Phe</td>
<td>Phenylalanine</td>
<td>UUC, UUU</td>
</tr>
<tr>
<td>G</td>
<td>Gly</td>
<td>Glycine</td>
<td>GGA, GGC, GGG, GGU</td>
</tr>
<tr>
<td>H</td>
<td>His</td>
<td>Histidine</td>
<td>CAC, CAU</td>
</tr>
<tr>
<td>I</td>
<td>Ile</td>
<td>Isoleucine</td>
<td>AUA, AUC, AUU</td>
</tr>
<tr>
<td>K</td>
<td>Lys</td>
<td>Lysine</td>
<td>AAA, AAG</td>
</tr>
<tr>
<td>L</td>
<td>Leu</td>
<td>Leucine</td>
<td>CUA, CUC, CUG, CUU, UUA, UUG</td>
</tr>
<tr>
<td>M</td>
<td>Met</td>
<td>Methionine</td>
<td>AUG</td>
</tr>
<tr>
<td>N</td>
<td>Asn</td>
<td>Asparagine</td>
<td>AAC, AAU</td>
</tr>
<tr>
<td>P</td>
<td>Pro</td>
<td>Proline</td>
<td>CCA, CCC, CCG, CCU</td>
</tr>
<tr>
<td>Q</td>
<td>Gln</td>
<td>Glutamine</td>
<td>CAA, CAG</td>
</tr>
<tr>
<td>R</td>
<td>Arg</td>
<td>Arginine</td>
<td>AGA, AGG, CGA, CGC, CGG, CGU</td>
</tr>
<tr>
<td>S</td>
<td>Ser</td>
<td>Serine</td>
<td>AGC, AGU, UCA, UCC, UCG, UCU</td>
</tr>
<tr>
<td>T</td>
<td>Thr</td>
<td>Threonine</td>
<td>ACA, ACC, ACG, ACU</td>
</tr>
<tr>
<td>V</td>
<td>Val</td>
<td>Valine</td>
<td>GUA, GUC, GUG, GUU</td>
</tr>
<tr>
<td>W</td>
<td>Trp</td>
<td>Tryptophan</td>
<td>UGG</td>
</tr>
<tr>
<td>Y</td>
<td>Tyr</td>
<td>Tyrosine</td>
<td>UAC, UAU</td>
</tr>
<tr>
<td>*</td>
<td>*</td>
<td><strong>STOP</strong></td>
<td>UAA, UAG, UGA</td>
</tr>
</tbody>
</table>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="reading frame"></span> - The different ways of dividing a DNA string into codons. Specifically, there are 6 different ways that a DNA string can be divided into codons:</p>
<ul>
<li>You can start dividing at index 0, 1, or 2.</li>
<li>You can divide either the DNA string itself or the reverse complementing DNA string.</li>
</ul>
<p>For example, given the string ATGTTCCATTAA, the following codon division are possible:</p>
<table>
<thead>
<tr>
<th>DNA</th>
<th>Start Index</th>
<th>Discard Prefix</th>
<th>Codons</th>
<th>Discard Suffix</th>
</tr>
</thead>
<tbody>
<tr>
<td>ATGTTCCATTAA</td>
<td>0</td>
<td></td>
<td>ATG, TTC, CAT, TAA</td>
<td></td>
</tr>
<tr>
<td>ATGTTCCATTAA</td>
<td>1</td>
<td>A</td>
<td>TGT, TCC, ATT</td>
<td>AA</td>
</tr>
<tr>
<td>ATGTTCCATTAA</td>
<td>2</td>
<td>AT</td>
<td>GTT, CCA, TTA</td>
<td>A</td>
</tr>
<tr>
<td>TTAATGGAACAT</td>
<td>0</td>
<td></td>
<td>TTA, ATG, GAA, CAT</td>
<td></td>
</tr>
<tr>
<td>TTAATGGAACAT</td>
<td>1</td>
<td>T</td>
<td>TAA, TGG, AAC</td>
<td>AT</td>
</tr>
<tr>
<td>TTAATGGAACAT</td>
<td>2</td>
<td>TT</td>
<td>AAT, GGA, ACA</td>
<td>T</td>
</tr>
</tbody>
</table>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>TTAATGGAACAT is the reverse complement of ATGTTCCATTAA.</p>
</div>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(encode)/i"></span> - When a DNA string or its reverse complement is made up of the codons required for an amino acid sequence. For example, ACAGTA encodes for the amino acid sequence...</p>
<ul>
<li>Threonine-Valine</li>
<li>Tyrosine-Cysteine (derived from reverse complement)</li>
</ul>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(branch-and-bound algorithm|branch and bound algorithm|branch-and-bound|branch and bound|branch-and-bound)/i"></span> - A bruteforce algorithm that enumerates candidates to explore at each step but also discards untenable candidates using various checks. The enumeration of candidates is the branching step, while the culling of untenable candidates is the bounding step.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="subsequence"></span> - A sequence derived by traversing some other sequence in order and choosing which elements to keep vs delete. For example, can is a subsequence of cation.</p>
<p><img src="svgbob_c25309d3982681fa348879d4c3fe065899f82016.svg" alt="Kroki diagram output"></p>
<p>Not to be confused with substring. A substring may also be a subsequence, but a subsequence won't necessarily be a substring.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="substring"></span> - A sequence derived by taking a contiguous part of some other sequence (order of elements maintained). For example, cat is a substring of cation.</p>
<p><img src="svgbob_1c969351b840ffdbd3172ad117673e5506efbd17.svg" alt="Kroki diagram output"></p>
<p>Not to be confused with subsequence. A substring may also be a subsequence, but a subsequence won't necessarily be a substring.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="subsequence"></span> <span class="anki-answerpattern" data-pattern="substring"></span></p>
<ul>
<li>Subsequence: A sequence derived by traversing some other sequence in order and choosing which elements to keep vs delete. <span class="anki-listrandomorder"></span>
</li>
<li>Substring: A sequence derived by taking a contiguous part of some other sequence (order of elements maintained).</li>
</ul>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(topological ordering|topological order|topological)/i"></span> - A 1-dimensional ordering of nodes in a directed acyclic graph in which each node is ahead of all of its predecessors / parents. In other words, the node is ahead of all other nodes that connect to it.</p>
<p>For example, the graph ...</p>
<p><img src="svgbob_6a2f3c0925fa0e1b69f5e2c59e731dc23f8c54e5.svg" alt="Kroki diagram output"></p>
<p>... the topological order is either [A, B, C, D, E] or [A, B, C, E, D]. Both are correct.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="longest common subsequence"></span> - A common subsequence between a set of strings of which is the longest out of all possible common subsequences. There may be more than one per set.</p>
<p>For example, AACCTTGG and ACACTGTGA share a longest common subsequence of...</p>
<ul>
<li>
<p>ACCTGG...</p>
<p><img src="svgbob_f1eadf7cf7bbf51eeb3471a6a5610a06f4dc2d1b.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>AACTGG...</p>
<p><img src="svgbob_a9b0360f8eeb866179f567437d7cfc36c167251a.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>etc..</p>
</li>
</ul>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="sequence alignment"></span> <span class="anki-hidepattern" data-pattern="alignment"></span> <span class="anki-hidepattern" data-pattern="align"></span> - Given a set of sequences, a sequence alignment is a set of operations applied to each position in an effort to line up the sequences. These operations include:</p>
<ul>
<li>insert/delete (indel for short).</li>
<li>replace (also referred to as mismatch).</li>
<li>keep matching (also referred to as match).</li>
</ul>
<p>For example, the sequences MAPLE and TABLE may be aligned by performing...</p>
<table>
<thead>
<tr>
<th>String 1</th>
<th>String 2</th>
<th>Operation</th>
</tr>
</thead>
<tbody>
<tr>
<td>M</td>
<td></td>
<td>Insert/delete</td>
</tr>
<tr>
<td></td>
<td>T</td>
<td>Insert/delete</td>
</tr>
<tr>
<td>A</td>
<td>A</td>
<td>Keep matching</td>
</tr>
<tr>
<td>P</td>
<td>B</td>
<td>Replace</td>
</tr>
<tr>
<td>L</td>
<td>L</td>
<td>Keep matching</td>
</tr>
<tr>
<td>E</td>
<td>E</td>
<td>Keep matching</td>
</tr>
</tbody>
</table>
<p>Or, MAPLE and TABLE may be aligned by performing...</p>
<table>
<thead>
<tr>
<th>String 1</th>
<th>String 2</th>
<th>Operation</th>
</tr>
</thead>
<tbody>
<tr>
<td>M</td>
<td>T</td>
<td>Replace</td>
</tr>
<tr>
<td>A</td>
<td>A</td>
<td>Keep matching</td>
</tr>
<tr>
<td>P</td>
<td>B</td>
<td>Replace</td>
</tr>
<tr>
<td>L</td>
<td>L</td>
<td>Keep matching</td>
</tr>
<tr>
<td>E</td>
<td>E</td>
<td>Keep matching</td>
</tr>
</tbody>
</table>
<p>The names of these operations make more sense if you were to think of alignment instead as <strong>transformation</strong>. The first example above in the context of <strong>transforming</strong> MAPLE to TABLE may be thought of as:</p>
<table>
<thead>
<tr>
<th>From</th>
<th>To</th>
<th>Operation</th>
<th>Result</th>
</tr>
</thead>
<tbody>
<tr>
<td>M</td>
<td></td>
<td>Delete M</td>
<td></td>
</tr>
<tr>
<td></td>
<td>T</td>
<td>Insert T</td>
<td>T</td>
</tr>
<tr>
<td>A</td>
<td>A</td>
<td>Keep matching A</td>
<td>TA</td>
</tr>
<tr>
<td>P</td>
<td>B</td>
<td>Replace P to B</td>
<td>TAB</td>
</tr>
<tr>
<td>L</td>
<td>L</td>
<td>Keep matching L</td>
<td>TABL</td>
</tr>
<tr>
<td>E</td>
<td>E</td>
<td>Keep matching E</td>
<td>TABLE</td>
</tr>
</tbody>
</table>
<p>The shorthand form of representing sequence alignments is to stack each sequence. The example above may be written as...</p>
<table>
<thead>
<tr>
<th></th>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
</tr>
</thead>
<tbody>
<tr>
<td>String 1</td>
<td>M</td>
<td></td>
<td>A</td>
<td>P</td>
<td>L</td>
<td>E</td>
</tr>
<tr>
<td>String 2</td>
<td></td>
<td>T</td>
<td>A</td>
<td>B</td>
<td>L</td>
<td>E</td>
</tr>
</tbody>
</table>
<p>All possible sequence alignments are represented using an alignment graph. A path through the alignment graph (called alignment path) represents one possible way to align the set of sequences.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(alignment graph|sequence alignment graph)/i"></span> - A directed graph representing all possible sequence alignments for some set of sequences. For example, the graph showing all the different ways that MAPLE and TABLE may be aligned ...</p>
<p><img src="svgbob_ce1ac3439fefb865a1d70920bc1a29d0ee7c2308.svg" alt="Kroki diagram output"></p>
<p>A path in this graph from source (top-left) to sink (bottom-right) represents an alignment.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(sequence alignment graph path|sequence alignment path|alignment graph path|alignment path)/i"></span> - A path in an alignment graph that represents one possible sequence alignment. For example, given the following alignment path ...</p>
<p><img src="svgbob_6133c15854b867ec6d44f9a55c046bd9f5afd9d8.svg" alt="Kroki diagram output"></p>
<p>is represent as the alignment...</p>
<table>
<thead>
<tr>
<th></th>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
<th>6</th>
<th>7</th>
</tr>
</thead>
<tbody>
<tr>
<td>String 1</td>
<td>-</td>
<td>-</td>
<td>M</td>
<td>A</td>
<td>P</td>
<td>-</td>
<td>L</td>
<td>E</td>
</tr>
<tr>
<td>String 2</td>
<td>T</td>
<td>A</td>
<td>-</td>
<td>B</td>
<td>L</td>
<td>E</td>
<td>-</td>
<td>-</td>
</tr>
</tbody>
</table>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="indel"></span> - In the context of sequence alignment, indel is short-hand for insert/delete. For example, the following sequence alignment has 2 indels in the very beginning...</p>
<table>
<thead>
<tr>
<th>Left</th>
<th>Right</th>
<th>Operation</th>
</tr>
</thead>
<tbody>
<tr>
<td>M</td>
<td></td>
<td>Indel</td>
</tr>
<tr>
<td></td>
<td>T</td>
<td>Indel</td>
</tr>
<tr>
<td>A</td>
<td>A</td>
<td>Keep matching</td>
</tr>
<tr>
<td>P</td>
<td>B</td>
<td>Replace</td>
</tr>
<tr>
<td>L</td>
<td>L</td>
<td>Keep matching</td>
</tr>
<tr>
<td>E</td>
<td>E</td>
<td>Keep matching</td>
</tr>
</tbody>
</table>
<p>The term insert/delete makes sense if you were to think of the set of operations as a <strong>transformation</strong> rather than an alignment. For example, the example above in the context of <strong>transforming</strong> MAPLE to TABLE:</p>
<table>
<thead>
<tr>
<th>From</th>
<th>To</th>
<th>Operation</th>
<th>Result</th>
</tr>
</thead>
<tbody>
<tr>
<td>M</td>
<td></td>
<td>Delete M</td>
<td></td>
</tr>
<tr>
<td></td>
<td>T</td>
<td>Insert T</td>
<td>T</td>
</tr>
<tr>
<td>A</td>
<td>A</td>
<td>Keep matching A</td>
<td>TA</td>
</tr>
<tr>
<td>P</td>
<td>B</td>
<td>Replace P to B</td>
<td>TAB</td>
</tr>
<tr>
<td>L</td>
<td>L</td>
<td>Keep matching L</td>
<td>TABL</td>
</tr>
<tr>
<td>E</td>
<td>E</td>
<td>Keep matching E</td>
<td>TABLE</td>
</tr>
</tbody>
</table>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="oncogene"></span> - A gene that has the potential to cause cancer. In tumor cells, these genes are often mutated or expressed at higher levels.</p>
<p>Most normal cells will undergo apoptosis when critical functions are altered and malfunctioning. Activated oncogenes may cause those cells to survive and proliferate instead.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="hamming distance"></span> - Given two strings, the hamming distance is the number of positional mismatches between them. For example, the hamming distance between TABLE and MAPLE is 2.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(dynamic programming algorithm|dynamic programming)/i"></span> - An algorithm that solves a problem by recursively breaking it down into smaller sub-problems, where the result of each recurrence computation is stored in some lookup table such that it can be re-used if it were ever encountered again (essentially trading space for speed). The lookup table may be created before hand or as a cache that gets filled as the algorithm runs.</p>
<p>For example, imagine a money system where coins are represented in 1, 12, and 13 cent denominations. You can use recursion to find the minimum number of coins to represent some monetary value such as $0.17:</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">min_coins</span>(<span class="hljs-params">value</span>):
  <span class="hljs-keyword">if</span> value == <span class="hljs-number">0.01</span> <span class="hljs-keyword">or</span> value == <span class="hljs-number">0.12</span> <span class="hljs-keyword">or</span> value == <span class="hljs-number">0.13</span>:
    <span class="hljs-keyword">return</span> <span class="hljs-number">1</span>
  <span class="hljs-keyword">else</span>:
    <span class="hljs-keyword">return</span> <span class="hljs-built_in">min</span>([
      min_coins(value - <span class="hljs-number">0.01</span>) + <span class="hljs-number">1</span>,
      min_coins(value - <span class="hljs-number">0.12</span>) + <span class="hljs-number">1</span>,
      min_coins(value - <span class="hljs-number">0.13</span>) + <span class="hljs-number">1</span>
    ])
</code></pre>
<p><img src="svgbob_011d5f745d519fd5c1a5354faa9961f538f726fc.svg" alt="Kroki diagram output"></p>
<p>The recursive graph above shows how $0.17 can be produced from a minimum of 5 coins: 1 x 13 cent denomination and 4 x 1 cent denomination. However, it recomputes identical parts of the graph multiple times. For example, <code>min_coins(3)</code> is independently computed 5 times. With dynamic programming, it would only be computed once and the result would be re-used each subsequent time <code>min_coins(3)</code> gets encountered.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="manhattan tourist problem"></span> <span class="anki-answerpattern" data-pattern="(sequence alignment graph|sequence alignment)/i"></span> <span class="anki-hidepattern" data-pattern="global alignment"></span> - The Manhattan tourist problem is an allegory to help explain sequence alignment graphs. Where as in sequence alignments you're finding a path through the graph from source to sink that has the maximum weight, in the Manhattan tourist problem you're finding a path from 59th St and 8th Ave to 42nd St and 3rd Ave with the most tourist sights to see. It's essentially almost the same problem as global alignment (diagonal edges are missing):</p>
<ul>
<li>The graph is the street layout of Manhattan.</li>
<li>The only options at each intersection are to move right or down.</li>
<li>The source node is the intersection of 59th St and 8th Ave.</li>
<li>The sink node is the intersection of 42nd St and 3rd Ave.</li>
<li>The number of tourist sights per street is the weight of an edge.</li>
</ul>
<p><img src="svgbob_d97ecce5abf34fc49b4c08c082e77841e4fd50b3.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(point accepted mutation|percent accepted mutation)/i"></span> <span class="anki-answerpattern" data-pattern="\b(PAM)\d*\b//false/true"></span> <span class="anki-hidepattern" data-pattern="BLOSUM"></span> - A scoring matrix used for sequence alignments of proteins. The scoring matrix is calculated by inspecting / extrapolating mutations as homologous proteins evolve. Specifically, mutations in the DNA sequence that encode some protein may change the resulting amino acid sequence for that protein. Those mutations that...</p>
<ul>
<li>impair the ability of the protein to function aren't likely to survive, and as such are given a low score.</li>
<li>keep the protein functional are likely to survive, and as such are given a normal or high score.</li>
</ul>
<p>Point accepted mutation, abbreviated to PAM.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(blocks substitution matrix|blocks substitution matrices|blocks amino acid substitution matrices|blocks amino acid substitution matrix)/i"></span> <span class="anki-answerpattern" data-pattern="\b(BLOSUM)\d*\b//false/true"></span> - A scoring matrix used for sequence alignments of proteins. The scoring matrix is calculated by scanning a protein database for highly conserved regions between similar proteins, where the mutations between those highly conserved regions define the scores. Specifically, those highly conserved regions are identified based on local alignments without support for indels (gaps not allowed). Non-matching positions in that alignment define potentially acceptable mutations.</p>
<p>Blocks substitution matrix, abbreviated to BLOSUM.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="point mutation"></span> - A mutation in DNA (or RNA) where a single nucleotide base is either changed, inserted, or deleted.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="directed acyclic graph"></span> <span class="anki-answerpattern" data-pattern="\b(DAG)\b/"></span> - A graph where the edges are directed (have a direction) and no cycles exist in the graph.</p>
<p>For example, the following is a directed acyclic graph...</p>
<p><img src="svgbob_d6ba29b704329aea40b096d1beeeec7da7d062f5.svg" alt="Kroki diagram output"></p>
<p>The following graph isn't a directed acyclic graph because the edges don't have a direction (no arrowhead means you can travel in either direction)...</p>
<p><img src="svgbob_64d614efe5a1433237c817bfba588e85082acf69.svg" alt="Kroki diagram output"></p>
<p>The following graph isn't a directed acyclic graph because it contains a cycle between D and B...</p>
<p><img src="svgbob_36208e0955503cf8020cbe27be7ca286ebcf55e2.svg" alt="Kroki diagram output"></p>
<p>Directed acyclic graph, abbreviated to DAG.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(divide-and-conquer algorithm|divide and conquer algorithm)/i"></span> <span class="anki-hidepattern" data-pattern="dynamic programming"></span> - An algorithm that solves a problem by recursively breaking it down into two or more smaller sub-problems, up until the point where each sub-problem is small enough / simple enough to solve. Examples include quicksort and merge sort.</p>
<p>See dynamic programming.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(global alignment|global sequence alignment)/i"></span> - A form of sequence alignment that finds the highest scoring alignment between a set of sequences. The sequences are aligned in their entirety. For example, the sequences TRELLO and MELLOW have the following global alignment...</p>
<table>
<thead>
<tr>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
<th>6</th>
</tr>
</thead>
<tbody>
<tr>
<td>T</td>
<td>R</td>
<td>E</td>
<td>L</td>
<td>L</td>
<td>O</td>
<td>-</td>
</tr>
<tr>
<td>-</td>
<td>M</td>
<td>E</td>
<td>L</td>
<td>L</td>
<td>O</td>
<td>W</td>
</tr>
</tbody>
</table>
<p>This is the form of sequence alignment that most people think about when they hear "sequence alignment."</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(local alignment|local sequence alignment)/i"></span> - A form of sequence alignment that isolates the alignment to a substring of each sequence. The substrings that score the highest are the ones selected. For example, the sequences TRELLO and MELLOW have the following local alignment...</p>
<table>
<thead>
<tr>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
</tr>
</thead>
<tbody>
<tr>
<td>E</td>
<td>L</td>
<td>L</td>
<td>O</td>
</tr>
<tr>
<td>E</td>
<td>L</td>
<td>L</td>
<td>O</td>
</tr>
</tbody>
</table>
<p>... because out of all substrings in TRELLO and all substrings in MELLOW, ELLO (from TRELLO) scores the highest against ELLO (from MELLOW).</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(fitting alignment|fitting sequence alignment)/i"></span> - A form of 2-way sequence alignment that isolates the alignment such that the entirety of one sequence is aligned against a substring of the other sequence. The substring producing the highest score is the one that's selected. For example, the sequences ELO and MELLOW have the following fitting alignment...</p>
<table>
<thead>
<tr>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
</tr>
</thead>
<tbody>
<tr>
<td>E</td>
<td>L</td>
<td>-</td>
<td>O</td>
</tr>
<tr>
<td>E</td>
<td>L</td>
<td>L</td>
<td>O</td>
</tr>
</tbody>
</table>
<p>... because out of all the substrings in MELLOW, the substring ELLO scores the highest against ELO.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(overlap alignment|overlap sequence alignment)/i"></span> - A form of 2-way sequence alignment that isolates the alignment to a suffix of the first sequences and a prefix of the second sequence. The prefix and suffix producing the highest score are the ones selected. For example, the sequences BURRITO and RICOTTA have the following overlap alignment...</p>
<table>
<thead>
<tr>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
</tr>
</thead>
<tbody>
<tr>
<td>R</td>
<td>I</td>
<td>T</td>
<td>-</td>
<td>O</td>
</tr>
<tr>
<td>R</td>
<td>I</td>
<td>-</td>
<td>C</td>
<td>O</td>
</tr>
</tbody>
</table>
<p>... because out of all the suffixes in BURRITO and the prefixes in RICOTTA, RITO and RICO score the highest.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(global alignment|global sequence alignment)/i"></span> <span class="anki-answerpattern" data-pattern="(local alignment|local sequence alignment)/i"></span> <span class="anki-answerpattern" data-pattern="(fitting alignment|fitting sequence alignment)/i"></span> <span class="anki-answerpattern" data-pattern="(overlap alignment|overlap sequence alignment)/i"></span></p>
<p>Name the following forms of sequence alignment:</p>
<ul>
<li>
<p>Global alignment: Finds the highest scoring alignment between a set of sequences. The sequences are aligned in their entirety. For example, the sequences TRELLO and MELLOW have the following global alignment... <span class="anki-listrandomorder"></span>
</p>
<table>
<thead>
<tr>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
<th>6</th>
</tr>
</thead>
<tbody>
<tr>
<td>T</td>
<td>R</td>
<td>E</td>
<td>L</td>
<td>L</td>
<td>O</td>
<td>-</td>
</tr>
<tr>
<td>-</td>
<td>M</td>
<td>E</td>
<td>L</td>
<td>L</td>
<td>O</td>
<td>W</td>
</tr>
</tbody>
</table>
</li>
<li>
<p>Local alignment: Isolates the alignment to a substring of each sequence. The substrings that score the highest are the ones selected. For example, the sequences TRELLO and MELLOW have the following local alignment...</p>
<table>
<thead>
<tr>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
</tr>
</thead>
<tbody>
<tr>
<td>E</td>
<td>L</td>
<td>L</td>
<td>O</td>
</tr>
<tr>
<td>E</td>
<td>L</td>
<td>L</td>
<td>O</td>
</tr>
</tbody>
</table>
</li>
<li>
<p>Fitting alignment: Isolates the alignment such that the entirety of one sequence is aligned against a substring of the other sequence. The substring producing the highest score is the one that's selected. For example, the sequences ELO and MELLOW have the following fitting alignment...</p>
<table>
<thead>
<tr>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
</tr>
</thead>
<tbody>
<tr>
<td>E</td>
<td>L</td>
<td>-</td>
<td>O</td>
</tr>
<tr>
<td>E</td>
<td>L</td>
<td>L</td>
<td>O</td>
</tr>
</tbody>
</table>
</li>
<li>
<p>Overlap alignment: Isolates the alignment to a suffix of the first sequences and a prefix of the second sequence. The prefix and suffix producing the highest score are the ones selected. For example, the sequences BURRITO and RICOTTA have the following overlap alignment...</p>
<table>
<thead>
<tr>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
</tr>
</thead>
<tbody>
<tr>
<td>R</td>
<td>I</td>
<td>T</td>
<td>-</td>
<td>O</td>
</tr>
<tr>
<td>R</td>
<td>I</td>
<td>-</td>
<td>C</td>
<td>O</td>
</tr>
</tbody>
</table>
</li>
</ul>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(levenshtein distance|string distance)/i"></span> <span class="anki-answerpattern" data-pattern="0"></span> <span class="anki-answerpattern" data-pattern="-1"></span> <span class="anki-ignorepattern" data-pattern="(0)_IGNORE/i"></span> - An application of global alignment where the final weight represents the minimum number of operations required to transform one sequence to another (via swaps, insertions, and deletions). Matches are scored as 0, while mismatches and indels are scored at -1. For example, TRELLO and MELLOW have the Levenshtein distance of 3...</p>
<table>
<thead>
<tr>
<th></th>
<th>0_IGNORE</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
<th>6</th>
<th></th>
</tr>
</thead>
<tbody>
<tr>
<td></td>
<td>T</td>
<td>R</td>
<td>E</td>
<td>L</td>
<td>L</td>
<td>O</td>
<td>-</td>
<td></td>
</tr>
<tr>
<td></td>
<td>-</td>
<td>M</td>
<td>E</td>
<td>L</td>
<td>L</td>
<td>O</td>
<td>W</td>
<td></td>
</tr>
<tr>
<td>Score</td>
<td>-1</td>
<td>-1</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>-1</td>
<td>Total: -3</td>
</tr>
</tbody>
</table>
<p>Negate the total score to get the minimum number of operations. In the example above, the final score of -3 maps to a minimum of 3 operations.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(genome rearrangement|chromosomal rearrangement|chromosome rearrangement)/i"></span> - A type of mutation where chromosomes go through structural changes, typically caused by either</p>
<ul>
<li>breakages in the chromosome, where the broken ends possibly rejoin but in a different order / direction.</li>
<li>DNA replication error.</li>
<li>DNA repair error.</li>
</ul>
<p>The different classes of rearrangement include...</p>
<ul>
<li>
<p>reversal / inversion: A break at two different locations followed by rejoining of the broken ends in different order. and rejoin in different order <span class="anki-listrandomorder"></span>
</p>
<p><img src="svgbob_c74a74fa3c923d49b533b6a31889fba8c857805a.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>translocation:</p>
<p><img src="svgbob_1a60b8a84c9cf4c8e262eb77c4db7fd6cdcb613c.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>deletion:</p>
<p><img src="svgbob_99d9b70d106d2088d8c2b416ae27d9528b9bd80f.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>duplication:</p>
<p><img src="svgbob_89a0a11786d7852db04220f4567d8243d5c22975.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>chromosome fusion:</p>
<p><img src="svgbob_5dda88b1068989e48adff91a638bd6e8122a5269.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>chromosome fission:</p>
<p><img src="svgbob_2bf9b82f60b2bbaf95ba7e19855dd43c0b57008f.svg" alt="Kroki diagram output"></p>
</li>
</ul>
<p>The segments of the genome that were moved around are referred to as synteny blocks.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="chimeric gene"></span> - A gene born from two separate genes getting fused together. A chimeric gene may have been created via genome rearrangement translocations.</p>
<p>An example of a chimeric gene is the gene coding for ABL-BCR fusion protein: A fusion of two smaller genes (coding for ABL and BCR individually) caused by chromosomes 9 and 22 breaking and re-attaching in a different order. The ABL-BCR fusion protein has been implicated in the development of a cancer known as chronic myeloid leukemia.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="reversal distance"></span> - The <em>minimum</em> number of genome rearrangement reversals required to transform genome P to genome Q. The minimum is chosen because of parsimony.</p>
<p>The short-hand for this is <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>d</mi><mrow><mi>r</mi><mi>e</mi><mi>v</mi></mrow></msub><mo stretchy="false">(</mo><mi>P</mi><mo separator="true">,</mo><mi>Q</mi><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">d_{rev}(P, Q)</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord"><span class="mord mathnormal">d</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.151392em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.02778em;">r</span><span class="mord mathnormal mtight">e</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.13889em;">P</span><span class="mpunct">,</span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord mathnormal">Q</span><span class="mclose">)</span></span></span></span>.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(sex-chromosome dosage compensation|sex chromosome dosage compensation|dosage compensation)/i"></span> - The mechanism by which sex chromosome gene expression is equalized between different sexes of the same species.</p>
<p>For example, mammals have two sexes...</p>
<ul>
<li>males, identified by one X chromosome and one Y chromosome. <span class="anki-listrandomorder"></span>
</li>
<li>females, identified by two X chromosomes.</li>
</ul>
<p>Since females have two X chromosomes, it would make sense for females to have double the gene expression for X chromosome genes. However, many X chromosome genes have nothing to do with sex and if their expression were doubled it would lead to disease. As such, female mammals randomly shut down one of two X chromosomes so as to keep X chromosome gene expression levels roughly equivalent to that of males.</p>
<p>For mammals, this mechanism means that X chromosomes are mostly conserved because an X chromosome that's gone through genome rearrangement likely won't survive: If a gene jumps off an X chromosome its gene expression may double, leading to problems.</p>
<p>Different species have different mechanisms for equalization. For example, some species will double the gene expression on the male's single X chromosome rather than deactivating one of the female's two X chromosomes. Other hermaphroditic species may scale down X chromosome gene expression when multiple X chromosomes are present.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(synteny block|synteny region|syntenic region|synteny|syntenic|syntenies)/i"></span> - Intervals within two sets of chromosomes that have similar genes which are either in ...</p>
<ul>
<li>
<p>the same order.</p>
<p><img src="svgbob_d3e4eb729d41d644a981794c1b1d200e2ac32fac.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>reverse order, where each gene's sequence is also reversed.</p>
<p><img src="svgbob_feb089f8283ebeb0a53ce28c3d58c61f4291646f.svg" alt="Kroki diagram output"></p>
</li>
</ul>
<p>The idea is that as evolution branches out a single ancestor species to different sub-species, genome rearrangements (reversals, translocations, etc..) are responsible for some of those mutations. As chromosomes break and rejoin back together in different order, the stretches between breakage points remain largely the same. For example, it's assumed that mice and humans have the same ancestor species because of the high number of synteny blocks between their genomes (most human genes have a mouse counterpart).</p>
<p><img src="svgbob_f4076143294d6480fb4b883c6cf2d02cdbf9117b.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(parsimony|parsimonious)/i"></span> - The scientific concept of choosing the fewest number of steps / shortest path / simplest scenario / simplest explanation that fits the evidence available.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(genomic dot-plot|genomic dot plot)/i"></span> - Given two genomes, create a 2D plot where each axis is assigned to one of the genomes and a dot is placed at each coordinate containing a match, where a match is either a shared k-mer or a k-mer and its reverse complement. Matches may also be fuzzily found (e.g. within some hamming distance rather).</p>
<p>For example, ...</p>
<p><img src="svgbob_3e7a2fb0e16cdd00ff9eaefd665998612d1db15c.svg" alt="Kroki diagram output"></p>
<p>Genomic dot-plots are typically used in building synteny graphs: Graphs that reveal shared synteny blocks (shared stretches of DNA). Synteny blocks exist because genome rearrangements account for a large percentage of mutations between two species that branched off from the same parent (given that they aren't too far removed -- e.g. mouse vs human).</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(synteny graph)/i"></span> - Given the genomic dot-plot for two genomes, cluster together points so as to reveal synteny blocks. For example, ...</p>
<p><img src="svgbob_eb50ab9882ae13623c74006244614b059e22fa47.svg" alt="Kroki diagram output"></p>
<p>... reveals that 4 synteny blocks are shared between the genomes. One of the synteny blocks is a normal match (C) while three are matching against their reverse complements (A, B, and D)...</p>
<p><img src="svgbob_725cfe74948aa476c335f99e888801f679a0059a.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(adjacent|adjacency|adjacencies|breakpoint)/i"></span> - Given two genomes that share synteny blocks, where one genome has the synteny blocks in desired order and direction while the other does not, an ...</p>
<ul>
<li>
<p>adjacency is when two neighbouring synteny blocks in the undesired genome are following each other just as they do in the desired genome. <span class="anki-listrandomorder"></span>
</p>
<p><img src="svgbob_ef5578eeb87986c967edea2a617d034be77e5477.svg" alt="Kroki diagram output"></p>
<p><img src="svgbob_076d065748dd4e638132b3f29fbb34d75bb184f3.svg" alt="Kroki diagram output"></p>
<p><img src="svgbob_d0bde3ada227e34f8571467009b7ef0bcaa51c57.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>breakpoint is when two neighbouring synteny blocks in the undesired genome don't fit the definition of an adjacency: They aren't following each other just as they do in the desired genome.</p>
<p><img src="svgbob_ab356811559d698349cda64e9f7a2358677b0ead.svg" alt="Kroki diagram output"></p>
<p><img src="svgbob_50bbe22a5291932cc5b9550a809e89a71b893090.svg" alt="Kroki diagram output"></p>
</li>
</ul>
<p>Breakpoints and adjacencies are useful because they identify desirable points for reversals (genome rearrangement), giving way to algorithms that find / estimate the reversal distance. For example, a contiguous train of adjacencies in an undesired genome may identify the boundaries for a single reversal that gets the undesired genome closer to the desired genome.</p>
<p><img src="svgbob_f265e51b0204855ebea5dad68bc33bf63c310885.svg" alt="Kroki diagram output"></p>
<p>The number of breakpoints and adjacencies always equals one less than the number of synteny blocks.</p>
<p><img src="svgbob_c6541b471afebaadca2c80a09a04248761d667e2.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="breakpoint graph"></span> - An undirected graph representing the order and orientation of synteny blocks shared between two genomes.</p>
<p>For example, the following two genomes share the synteny blocks A, B, C, and D...</p>
<p><img src="svgbob_7e9c33210faa97dd82efb19ecbdd7b9300236fba.svg" alt="Kroki diagram output"></p>
<p>The breakpoint graph for the above two genomes is basically just a merge of the above diagrams. The set of synteny blocks shared between both genomes (A, B, C, and D) become dashed edges where each edge's...</p>
<ul>
<li>arrow end is a tail node (t suffix).</li>
<li>non-arrow end is a head node (h suffix).</li>
</ul>
<p>Gap regions between synteny blocks are represented by solid colored edges, either red or blue depending on which genome it is.</p>
<p>If the genomes are linear, gap region edges are created between the nodes and the edges and a special termination node.</p>
<p><img src="dot_68956ed2f489ac4cb837efb6bd1d2a65.svg" alt="Dot diagram"></p>
<p>In the above breakpoint graph, the blue edges represent genome 2's gap regions while the red edges represent genome 1's gap regions. The set of edges representing synteny blocks is shared between them.</p>
<p>Breakpoint graphs build on the concept of breakpoints to compute a parsimonious path of fusion, fission, and reversal mutations (genome rearrangements) that transforms one genome into the other (see 2-break). Conventionally, blue edges represent the final desired path while red edges represent the path being transformed. As such, breakpoint graphs typically order synteny blocks so that blue edges are uniformly sandwiched between synteny blocks / red edges get chaotically scattered around.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="\b(2-break|2 break|two break|two-break)s?\b/i"></span> - Given a breakpoint graph, a 2-break operation breaks the two red edges at a synteny block boundary and re-wires them such that one of the red edges matches the blue edge at that boundary.</p>
<p>For example, the two red edges highlighted below share the same synteny block boundary and can be re-wired such that one of the edges matches the blue edge at that synteny boundary ...</p>
<p><img src="dot_2ca5243fc24500a47a2859ee21a4a61a.svg" alt="Dot diagram"></p>
<p>Each 2-break operation on a breakpoint graph represents a fusion, fission, or reversal mutation (genome rearrangement). Continually applying 2-breaks until all red edges match blue edges reveals a parsimonious path of such mutations that transforms the red genome to the blue genome.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(permutation)/i"></span> - A list representing a single chromosome in one of the two genomes that make up a breakpoint graph. The entire breakpoint graph is representable as two sets of permutations, where each genome in the breakpoint graph is a set.</p>
<p>Permutation sets are commonly used for tersely representing breakpoint graphs as text. For example, given the following breakpoint graph ...</p>
<p><img src="dot_bed1d61715ed515600f51c8a4b395968.svg" alt="Dot diagram"></p>
<p>... , the permutation set representing the red genome may be any of the following ...</p>
<ul>
<li><code>{[-D, -B, +C, -A]}</code></li>
<li><code>{[+A, -C, +B, +D]}</code></li>
<li><code>{[-B, +C, -A, -D]}</code></li>
<li><code>{[-C, +B, +D, +A]}</code></li>
<li><code>{[+C, -A, -D, -B]}</code></li>
<li>...</li>
</ul>
<p>All representations above are equivalent.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>See Algorithms/Synteny/Reversal Path/Breakpoint List Algorithm for a full explanation of how to read_NORM permutations / how to convert from and to breakpoint graphs.</p>
</div>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(\bfusion|\bfuse)/i"></span> - Joining two or more things together to form a single entity. For example, two chromosomes may join together to form a single chromosome (genome rearrangement).</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="fission"></span> - Splitting a single entity into two or more parts. For example, a single chromosome may break into multiple pieces where each piece becomes its own chromosome (genome rearrangement).</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="translocation"></span> - Changing location. For example, part of a chromosome may transfer to another chromosome (genome rearrangement).</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="severe acute respiratory syndrome"></span> - A deadly coronavirus that emerged from China around early 2003. The virus transmits itself through droplets that enter the air when someone with the disease coughs.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="coronavirus"></span> - A family of viruses that attack the respiratory tracts of mammals and birds. The name comes from the fact that the outer spikes of the virus resemble the corona of the sun (crown of the sun / outermost part of the sun's atmosphere protruding out).</p>
<p>The common cold, SARS, and COVID-19 are examples of coronaviruses.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="human immunodeficiency virus"></span> - A virus that over time causes acquired immunodeficiency syndrome (AIDS).</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="acquired immunodeficiency syndrome"></span> - The disease caused by the human immunodeficiency virus (HIV).</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(immunodeficiency|immunocompromisation|immunocompromised)/i"></span> - A state in which the immune system's ability to fight infectious disease and cancer is compromised or entirely absent.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="DNA virus"></span> - A virus with a DNA genome. Depending on the type of virus, the genome may be single-stranded DNA or double-stranded DNA.</p>
<p>Herpes, chickenpox, and smallpox are examples of DNA viruses.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="RNA virus"></span> - A virus with a RNA genome. RNA replication has a higher rate than DNA replication, meaning that RNA viruses mutate faster than DNA viruses.</p>
<p>Coronaviruses, HIV, and influenza are examples of RNA viruses.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(phylogeny|phylogenetic)/i"></span> - The concept of inferring the evolutionary history among some set of species (shared ancestry) by inspecting properties of those species (e.g. relatedness of phenotypic or genotypic characteristics).</p>
<p><img src="svgbob_cabf6297c9c4eca18762400d14f3c05340a62b31.svg" alt="Kroki diagram output"></p>
<p>In the example above, cat and lion are descendants of some shared ancestor species. Likewise, that ancestor and bears are likely descendants from some other higher up species.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(phylogenetic tree|phylogeny tree|phylogenies|evolutionary tree)/i"></span> - A tree showing the degree in which biological species or entities (e.g. viruses) are related. Such trees help infer relationships such as common ancestry or which animal a virus jumped to humans from (e.g. virus A and B are related but A is only present in bats while B just showed up in humans).</p>
<p><img src="svgbob_f7b1705852841b00a5ecb758d6e39af634d6898e.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="distance metric"></span> - A metric used to measure how <em>different</em> a pair of entities are to each other. Examples include...</p>
<ul>
<li>hamming distance between DNA / protein sequences.</li>
<li>global alignment score between DNA / protein sequences.</li>
<li>two-break count (reversal distance).</li>
<li>number of similar physical or behavioural attributes.</li>
<li>euclidean distance between two vectors.</li>
<li>pearson distance between two vectors.</li>
<li>etc..</li>
</ul>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(distance matrix|distance matrices)/i"></span> - Given a set of n different entities, a distance matrix is an n-by-n matrix where each element contains the distance between the entities for that cell. For example, for the species snake, lizard, bird, and crocodile ...</p>
<table>
<thead>
<tr>
<th></th>
<th>Snake</th>
<th>Lizard</th>
<th>Bird</th>
<th>Crocodile</th>
</tr>
</thead>
<tbody>
<tr>
<td>Snake</td>
<td>0</td>
<td>2</td>
<td>6</td>
<td>4</td>
</tr>
<tr>
<td>Lizard</td>
<td>2</td>
<td>0</td>
<td>6</td>
<td>4</td>
</tr>
<tr>
<td>Bird</td>
<td>6</td>
<td>6</td>
<td>0</td>
<td>5</td>
</tr>
<tr>
<td>Crocodile</td>
<td>4</td>
<td>4</td>
<td>5</td>
<td>0</td>
</tr>
</tbody>
</table>
<p>The distance metric can be anything so long as it meets the following properties:</p>
<ul>
<li>Must produce a non-negative distance -- dist(A,B) &gt;= 0</li>
<li>Must produce the same distance regardless of species order -- dist(A,B) == dist(B,A)</li>
<li>Must produce a distance that satisfies the triangle inequality property -- if dist(B,C) = x, then dist(A,B) + dist(A,C) &gt;= x</li>
</ul>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>I think what the last bullet point means is that the distance will be &gt;= if you travel to it indirectly (hop over to it instead of taking a straight path). For example, if dist(B,C) = 5, then dist(A,B) + dist(A,C) must be &gt;= 5.</p>
<p><img src="svgbob_3f0c5795b674de5612a12db473d621ed2ed53e92.svg" alt="Kroki diagram output"></p>
<p>A, B, and C are species.</p>
</div>
<p>Common distance metrics include...</p>
<ul>
<li>hamming distance between the DNA sequences.</li>
<li>levenshtein distance between the DNA sequences.</li>
<li>two-break count (reversal distance).</li>
</ul>
<p>Distance matrices are used to generate phylogenetic trees. A single distance matrix may fit many different trees or it's possible that it fits no tree at all. For example, the distance matrix above fits the tree...</p>
<p><img src="svgbob_5a2ddd5247c86bd4e9d05f814cb689158dca59b0.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="tree"></span> - In graph theory, a tree is an acyclic undirected graph in which any two nodes are connected by exactly one path (nodes branch outward / never converge).</p>
<p>Trees come in two forms: rooted trees and unrooted trees. In graph theory, a tree typically refers to an unrooted tree.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>This is different from the computer science definition of tree, which is an abstract data type representing a hierarchy (always a single root that flows downwards), typically generalized as a <em>directed</em> acyclic graph as opposed to an undirected acyclic graph.</p>
</div>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(un-rooted tree|unrooted tree)/i"></span> - A tree without a root node...</p>
<p><img src="svgbob_3ae717329571359a328c5697332e94ce675003fb.svg" alt="Kroki diagram output"></p>
<p>An unrooted tree may be turned into a rooted tree by choosing any non-leaf node (internal node) to be the root node</p>
<pre><code class="language-{note{">If injecting a node is a possibility, you can also convert an unrooted tree to a rooted tree by injecting a root node along one of its edges.
</code></pre>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="rooted tree"></span> - A tree with a root node...</p>
<p><img src="svgbob_4a12a5e83db150bd3a7d62bc553c9ece9b5d54b2.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(subtree|sub-tree)/i"></span> - Given a node in a tree, that node and all of its descendants comprise a subtree. For example, the following tree has the subtree ...</p>
<p><img src="svgbob_f55a16a060f12e0666677a9e75ab0f385bbd7793.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="degree"></span> - The number of edges leading into / out of a node of an undirected graph.</p>
<p>The node below has a degree of 3.</p>
<p><img src="svgbob_60f0bcaccb25a424cce90b06db0fd197ae921dda.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="simple tree"></span> - An unrooted tree where ...</p>
<ul>
<li>every internal node has a degree &gt; 2.</li>
<li>every edge has a weight of &gt; 0.</li>
</ul>
<p><img src="svgbob_80dd4249a34f25e70e5294c8e3a572adbe096671.svg" alt="Kroki diagram output"></p>
<p>In the context of phylogeny, a simple tree's ...</p>
<ul>
<li>leaf nodes represent known entities.</li>
<li>internal nodes represent inferred ancestor entities.</li>
<li>edge weights represent distances between entities.</li>
</ul>
<p>The restrictions placed on simple trees simplify the process of working backwards from a distance matrix to a phylogenetic tree.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(additive matrix|additive matrices|additive distance matrix|additive distance matrices)/i"></span> <span class="anki-hidepattern" data-pattern="additive"></span> - Given a distance matrix, if there exists a tree with edge weights that satisfy that distance matrix (referred to as fit), that distance matrix is said to be an additive distance matrix.</p>
<p>For example, the following tree fits the following distance matrix ...</p>
<table>
<thead>
<tr>
<th></th>
<th>Cat</th>
<th>Lion</th>
<th>Bear</th>
</tr>
</thead>
<tbody>
<tr>
<td>Cat</td>
<td>0</td>
<td>2</td>
<td>3</td>
</tr>
<tr>
<td>Lion</td>
<td>2</td>
<td>0</td>
<td>3</td>
</tr>
<tr>
<td>Bear</td>
<td>3</td>
<td>3</td>
<td>0</td>
</tr>
</tbody>
</table>
<p><img src="svgbob_e5ec5970ccdbca871edbc896779d5d1131ed9055.svg" alt="Kroki diagram output"></p>
<p>The term additive is used because the weights of all edges along the path between leaves (i, j) add to <code>dist(i, j)</code> in the distance matrix. Not all distance matrices are additive. For example, no simple tree exists that satisfies the following distance matrix...</p>
<table>
<thead>
<tr>
<th></th>
<th>S1</th>
<th>S2</th>
<th>S3</th>
<th>S4</th>
</tr>
</thead>
<tbody>
<tr>
<td>S1</td>
<td>0</td>
<td>3</td>
<td>4</td>
<td>3</td>
</tr>
<tr>
<td>S2</td>
<td>3</td>
<td>0</td>
<td>4</td>
<td>5</td>
</tr>
<tr>
<td>S3</td>
<td>4</td>
<td>4</td>
<td>0</td>
<td>2</td>
</tr>
<tr>
<td>S4</td>
<td>3</td>
<td>5</td>
<td>2</td>
<td>0</td>
</tr>
</tbody>
</table>
<ul>
<li>
<p>Test simple tree 1:</p>
<p><img src="svgbob_74cdc79c5f9b82937ba473b3e3e632aab5cb6b4e.svg" alt="Kroki diagram output"></p>
<pre><code>dist(S1, S2) is 3 = w + x
dist(S1, S3) is 4 = w + y
dist(S1, S4) is 3 = w + z
dist(S2, S3) is 4 = x + y
dist(S2, S4) is 5 = x + z
dist(S3, S4) is 2 = y + z
</code></pre>
<p>Attempting to solve this produces inconsistent results. Solved values for each variable don't work across all equations present.</p>
</li>
<li>
<p>Test simple tree 2:</p>
<p><img src="svgbob_310a023ae046f0c94fc708941d7e494797cf787b.svg" alt="Kroki diagram output"></p>
<pre><code>dist(S1, S2) is 3 = w + x
dist(S1, S3) is 4 = w + u + y
dist(S1, S4) is 3 = w + u + z
dist(S2, S3) is 4 = x + u + y
dist(S2, S4) is 5 = x + u + z
dist(S3, S4) is 2 = y + z
</code></pre>
<p>Attempting to solve this produces inconsistent results. Solved values for each variable don't work across all equations present.</p>
</li>
<li>
<p>Test simple tree 3:</p>
<p><img src="svgbob_27b6be0c6cb6193b60f5164ee5e83ce0f0389d38.svg" alt="Kroki diagram output"></p>
<pre><code>dist(S1, S2) is 4 = w + u + y
dist(S1, S3) is 3 = w + x
dist(S1, S4) is 3 = w + u + z
dist(S2, S3) is 4 = x + u + y
dist(S2, S4) is 2 = y + z
dist(S3, S4) is 5 = x + u + z
</code></pre>
<p>Attempting to solve this produces inconsistent results. Solved values for each variable don't work across all equations present.</p>
</li>
<li>
<p>etc..</p>
</li>
</ul>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="neighbour"></span> - Given two leaf nodes in a tree, those leaf nodes are said to be neighbours if they share they connect to the same internal node. For example, leaf nodes A and B are neighbours in the following tree because they both connect to internal node D ...</p>
<p><img src="svgbob_80372a14f2a248e2ffdfdbfef223b5b15c39f7b1.svg" alt="Kroki diagram output"></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>A leaf node will only ever have 1 parent, by definition of a tree.</p>
</div>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="limb"></span> - Given a leaf node in a tree, that leaf node's limb is the edge between it and its parent (node it's connected to). For example, the following tree has the following limbs ...</p>
<p><img src="svgbob_80372a14f2a248e2ffdfdbfef223b5b15c39f7b1.svg" alt="Kroki diagram output"></p>
<ul>
<li>(A, D)</li>
<li>(B, D)</li>
<li>(G, I)</li>
</ul>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>A leaf node will only ever have 1 parent, by definition of a tree.</p>
</div>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(limb length|limb weight)/i"></span> - Given a leaf node in a tree, the leaf node's limb length is the weight assigned to its limb. For example, node A has a limb length of 2 in the following tree...</p>
<p><img src="svgbob_54ab99476bd3bb4646b1899b68122441df1dc495.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(four points? condition|four point theorem)/i"></span> - An algorithm for determining if a distance matrix is an additive distance matrix. Given four leaf nodes, the algorithm checks different permutations of those leaf nodes to see if any pass a test, where that test builds node pairings from the quartet and checks their distances to see if they meet a specific condition...</p>
<pre class="hljs"><code><span class="hljs-keyword">for</span> a, b, c, d <span class="hljs-keyword">in</span> permutations(quartet, r=<span class="hljs-number">4</span>):  <span class="hljs-comment"># find one perm that passes the following test</span>
    s1 = dist_mat[a][b] + dist_mat[c][d]  <span class="hljs-comment"># sum of dists for (a,b) and (c,d)</span>
    s2 = dist_mat[a][c] + dist_mat[b][d]  <span class="hljs-comment"># sum of dists for (a,c) and (b,d)</span>
    s3 = dist_mat[a][d] + dist_mat[b][c]  <span class="hljs-comment"># sum of dists for (a,d) and (b,c)</span>
    <span class="hljs-keyword">if</span> s1 &lt;= s2 == s3:
        <span class="hljs-keyword">return</span> <span class="hljs-literal">True</span>
<span class="hljs-keyword">return</span> <span class="hljs-literal">False</span>
</code></pre>
<p>If all possible leaf node quartets pass the above test, the distance matrix is an additive distance matrix (was derived from a tree / fits a tree).</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>See Algorithms/Phylogeny/Test Additive Distance Matrix for a full explanation of how this algorithm works.</p>
</div>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(trim distance matrix|trimmed distance matrix|trim distance matrices|trimmed distance matrices)/i"></span> - A distance matrix where a leaf node's row and column have been removed. This is equivalent to removing the leaf node's limb in the corresponding simple tree and merging together any edges connected by nodes of degree 2.</p>
<p>For example, removing v2 from the following distance matrix...</p>
<table>
<thead>
<tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th><span style="background-color: red; color: white; padding: 3px">v2</span></th>
<th>v3</th>
</tr>
</thead>
<tbody>
<tr>
<td>v0</td>
<td>0</td>
<td>13</td>
<td>21</td>
<td>22</td>
</tr>
<tr>
<td>v1</td>
<td>13</td>
<td>0</td>
<td>12</td>
<td>13</td>
</tr>
<tr>
<td><span style="background-color: red; color: white; padding: 3px">v2</span></td>
<td>21</td>
<td>12</td>
<td>0</td>
<td>13</td>
</tr>
<tr>
<td>v3</td>
<td>22</td>
<td>13</td>
<td>13</td>
<td>0</td>
</tr>
</tbody>
</table>
<p><img src="dot_4e5daffbacf71191e85a5881e8367790.svg" alt="Dot diagram"></p>
<p>... results in v2's row and column being removed ...</p>
<table>
<thead>
<tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v3</th>
</tr>
</thead>
<tbody>
<tr>
<td>v0</td>
<td>0</td>
<td>13</td>
<td>22</td>
</tr>
<tr>
<td>v1</td>
<td>13</td>
<td>0</td>
<td>13</td>
</tr>
<tr>
<td>v3</td>
<td>22</td>
<td>13</td>
<td>0</td>
</tr>
</tbody>
</table>
<p><img src="dot_8970363f8d351a28a22328fbfd029f13.svg" alt="Dot diagram"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(bald distance matrix|balded distance matrix|bald distance matrices|balded distance matrices)/i"></span> - An additive distance matrix where the distances in a leaf node's row and column have been subtracted by that leaf node's limb length. This is equivalent to setting the leaf node's limb length to 0 in the corresponding simple tree.</p>
<p>For example, balding v5's limb length in the following distance matrix ...</p>
<table>
<thead>
<tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr>
</thead>
<tbody>
<tr>
<td>v0</td>
<td>0</td>
<td>13</td>
<td>21</td>
<td>21</td>
<td>22</td>
<td>22</td>
</tr>
<tr>
<td>v1</td>
<td>13</td>
<td>0</td>
<td>12</td>
<td>12</td>
<td>13</td>
<td>13</td>
</tr>
<tr>
<td>v2</td>
<td>21</td>
<td>12</td>
<td>0</td>
<td>20</td>
<td>21</td>
<td>21</td>
</tr>
<tr>
<td>v3</td>
<td>21</td>
<td>12</td>
<td>20</td>
<td>0</td>
<td>7</td>
<td>13</td>
</tr>
<tr>
<td>v4</td>
<td>22</td>
<td>13</td>
<td>21</td>
<td>7</td>
<td>0</td>
<td>14</td>
</tr>
<tr>
<td>v5</td>
<td>22</td>
<td>13</td>
<td>21</td>
<td>13</td>
<td>14</td>
<td>0</td>
</tr>
</tbody>
</table>
<p><img src="dot_e304cd9917c7e5b9d72c4e876e260cba.svg" alt="Dot diagram"></p>
<p>... results in ...</p>
<table>
<thead>
<tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr>
</thead>
<tbody>
<tr>
<td>v0</td>
<td>0</td>
<td>13</td>
<td>21</td>
<td>21</td>
<td>22</td>
<td>15</td>
</tr>
<tr>
<td>v1</td>
<td>13</td>
<td>0</td>
<td>12</td>
<td>12</td>
<td>13</td>
<td>6</td>
</tr>
<tr>
<td>v2</td>
<td>21</td>
<td>12</td>
<td>0</td>
<td>20</td>
<td>21</td>
<td>14</td>
</tr>
<tr>
<td>v3</td>
<td>21</td>
<td>12</td>
<td>20</td>
<td>0</td>
<td>7</td>
<td>6</td>
</tr>
<tr>
<td>v4</td>
<td>22</td>
<td>13</td>
<td>21</td>
<td>7</td>
<td>0</td>
<td>7</td>
</tr>
<tr>
<td>v5</td>
<td>15</td>
<td>6</td>
<td>14</td>
<td>6</td>
<td>7</td>
<td>0</td>
</tr>
</tbody>
</table>
<p><img src="dot_504f11240750f6a2add064fe593ca205.svg" alt="Dot diagram"></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>Technically, an edge weight of 0 is a violation of the simple tree requirement of having edge weights &gt; 0. This is a special case.</p>
</div>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>How do you know the limb length from just the distance matrix? See the algorithm to determine limb length for any leaf from just the distance matrix.</p>
</div>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="additive phylogeny"></span> - A recursive algorithm that finds the unique simple tree for some additive distance matrix. The algorithm trims a single leaf node at each recursive step until the distance matrix has a size of two. The simple tree for any two leaf nodes is those two nodes connected by a single edge. Using that tree as its base, the algorithm recurses out of each step by finding where that step's trimmed node exists on the tree and attaching it on.</p>
<p>At the end, the algorithm will have constructed the entire simple tree for the additive distance matrix. For example, ...</p>
<ul>
<li>
<p>Initial distance matrix ...</p>
<table>
<thead>
<tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
</tr>
</thead>
<tbody>
<tr>
<td>v0</td>
<td>0</td>
<td>13</td>
<td>21</td>
<td>22</td>
</tr>
<tr>
<td>v1</td>
<td>13</td>
<td>0</td>
<td>12</td>
<td>13</td>
</tr>
<tr>
<td>v2</td>
<td>21</td>
<td>12</td>
<td>0</td>
<td>13</td>
</tr>
<tr>
<td>v3</td>
<td>22</td>
<td>13</td>
<td>13</td>
<td>0</td>
</tr>
</tbody>
</table>
</li>
<li>
<p>Trim v1 to produce distance matrix ...</p>
<table>
<thead>
<tr>
<th></th>
<th>v0</th>
<th>v2</th>
<th>v3</th>
</tr>
</thead>
<tbody>
<tr>
<td>v0</td>
<td>0</td>
<td>21</td>
<td>22</td>
</tr>
<tr>
<td>v2</td>
<td>21</td>
<td>0</td>
<td>13</td>
</tr>
<tr>
<td>v3</td>
<td>22</td>
<td>13</td>
<td>0</td>
</tr>
</tbody>
</table>
</li>
<li>
<p>Trim v0 to produce distance matrix ...</p>
<table>
<thead>
<tr>
<th></th>
<th>v2</th>
<th>v3</th>
</tr>
</thead>
<tbody>
<tr>
<td>v2</td>
<td>0</td>
<td>13</td>
</tr>
<tr>
<td>v3</td>
<td>13</td>
<td>0</td>
</tr>
</tbody>
</table>
</li>
<li>
<p>Distance matrix maps to the obvious simple tree...</p>
<p><img src="dot_17295d277521c7d7445601414800311c.svg" alt="Dot diagram"></p>
</li>
<li>
<p>Attach v0 to produce tree...</p>
<p><img src="dot_ee7b26791f6d3ee197982f570c83e294.svg" alt="Dot diagram"></p>
</li>
<li>
<p>Attach v1 to produce tree...</p>
<p><img src="dot_a9a8b9059c8bf67a1d36b651e907a38a.svg" alt="Dot diagram"></p>
</li>
</ul>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>See Algorithms/Phylogeny/Distance Matrix to Tree/Additive Phylogeny Algorithm for a full explanation of how this algorithm works.</p>
</div>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(sum of squared errors|sum of square error|sum of errors squared)/i"></span> - Sum of errors squared is an algorithm used to quantify how far off some estimation / prediction is.</p>
<p>Given a set of real values and a set of predicted values, the error is the difference between the real and predicted values at each data point. For example...</p>
<table>
<thead>
<tr>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
</tr>
</thead>
<tbody>
<tr>
<td>Real</td>
<td>5</td>
<td>4</td>
<td>7</td>
<td>8</td>
<td>5</td>
<td>4</td>
</tr>
<tr>
<td>Predicted</td>
<td>4</td>
<td>5</td>
<td>7</td>
<td>6</td>
<td>4</td>
<td>4</td>
</tr>
<tr>
<td>Error</td>
<td>1</td>
<td>-1</td>
<td>0</td>
<td>2</td>
<td>1</td>
<td>0</td>
</tr>
</tbody>
</table>
<p>The algorithm squares each error and sums them together:</p>
<pre class="hljs"><code>res = <span class="hljs-number">0</span>
<span class="hljs-keyword">for</span> r_val, p_val <span class="hljs-keyword">in</span> <span class="hljs-built_in">zip</span>(real, predicted):
  err = r_val - p_val
  res += err ** <span class="hljs-number">2</span>
</code></pre>
<p>The algorithm as a formula: <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msubsup><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>0</mn></mrow><mi>n</mi></msubsup><mrow><mo stretchy="false">(</mo><msub><mi>r</mi><mi>i</mi></msub><mo>−</mo><msub><mi>p</mi><mi>i</mi></msub><msup><mo stretchy="false">)</mo><mn>2</mn></msup></mrow></mrow><annotation encoding="application/x-tex">\sum_{i=0}^{n}{(r_i - p_i)^2}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.113818em;vertical-align:-0.29971000000000003em;"></span><span class="mop"><span class="mop op-symbol small-op" style="position:relative;top:-0.0000050000000000050004em;">∑</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.804292em;"><span style="top:-2.40029em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">i</span><span class="mrel mtight">=</span><span class="mord mtight">0</span></span></span></span><span style="top:-3.2029em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">n</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.29971000000000003em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord"><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">r</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal">p</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mclose"><span class="mclose">)</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8141079999999999em;"><span style="top:-3.063em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span></span></span></span></span></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="speciation"></span> - The evolutionary process by which a species splits into distinct child species.</p>
<p>In phylogenetic trees, internal nodes branching are assumed to be speciation events. That is, an event where the ancestral species represented by that node splits into distinct child species.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="unrooted binary tree"></span> - In the context of phylogeny, an unrooted binary tree is a simple tree where internal nodes must have a degree of 3...</p>
<p><img src="svgbob_b29e6fe881c0b485a2b8b1eac26f15987cb45a9a.svg" alt="Kroki diagram output"></p>
<p>In other words, an edge leading to an internal node is guaranteed to branch <em>exactly</em> twice.</p>
<p><img src="svgbob_43a3ec6298299c551b148002f27fa4910a77a171.svg" alt="Kroki diagram output"></p>
<p>Contrast that to normal simple trees where internal nodes can have any degree greater than 2...</p>
<p><img src="svgbob_17ac9bddd64ed68c7e7e56c709f3a2967622a61a.svg" alt="Kroki diagram output"></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>Recall that simple trees are unrooted to begin with and can't have nodes with degree 2 (train of non-branching edges not allowed).</p>
</div>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="rooted binary tree"></span> - In the context of phylogeny, a rooted binary tree is an unrooted binary tree with a root node injected in between one of its edges. The injected root node will always end up as an internal node of degree of 2, breaking the constraint of ...</p>
<ul>
<li>unrooted binary trees that states internal nodes must have degree of exactly 3.</li>
<li>simple trees that states internal nodes must have degree of greater than 2.</li>
</ul>
<p><img src="svgbob_9d8f5b3c109dc96946bf1d24e21ac3304a29070c.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(ultrametric tree|ultrametric)/i"></span> - A rooted tree where all leaf nodes are equidistant from the root.</p>
<p><img src="svgbob_10fa2bf98dc55b5346b2ebb7f45c2e56a81730fe.svg" alt="Kroki diagram output"></p>
<p>In the example above, all leaf nodes are a distance of 4 from the root.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>Does an ultrametric tree have to be a rooted binary tree? I think the answer is no: UPGMA generated rooted binary trees, but ultrametric trees in general just have to be rooted trees / they don't have to be binary.</p>
</div>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="molecular clock"></span> - The assumption that the rate of mutation is more-or-less consistent. For example, ...</p>
<ul>
<li>every minute, around n of m nucleotides get mutated.</li>
<li>every hour, around n genome rearrangement reversals occur per chromosome segment of size m.</li>
<li>etc..</li>
</ul>
<p>This assumption is used for some phylogeny algorithms (e.g. UPGMA).</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(unweighted pair group method with arithmetic mean)/i"></span> <span class="anki-hidepattern" data-pattern="UPGMA"></span> - A heuristic algorithm used to estimate a binary ultrametric tree for some distance matrix.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>A binary ultrametric tree is an ultrametric tree where each internal node only branches to two children. In other words, a binary ultrametric tree is a rooted binary tree where all leaf nodes are equidistant from the root.</p>
</div>
<p>The algorithm assumes that the rate of mutation is consistent (molecular clock). This assumption is what makes the tree ultrametric. A set of present day species (leaf nodes) are assumed to all have the same amount of mutation (distance) from their shared ancestor (shared internal node).</p>
<p><img src="svgbob_9c45e3a8fedf42d570c570b7c417173e7ed57970.svg" alt="Kroki diagram output"></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>See Algorithms/Phylogeny/Distance Matrix to Tree/UPGMA Algorithm for a full explanation of how this algorithm works.</p>
</div>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="neighbour joining matrix"></span> - A matrix produced by transforming a distance matrix such that each element is calculated as total_dist(a) + total_dist(b) - (n - 2) * dist(a, b), where...</p>
<ul>
<li>n is the number of leaf nodes in the distance matrix.</li>
<li>a and b are the leaf nodes being referenced in the distance matrix.</li>
<li>dist(a, b) returns the distance between leaf nodes a and b in the distance matrix.</li>
<li>total dist(a) returns the sum of all distances to / from leaf node a.</li>
<li>total dist(b) returns the sum of all distances to / from leaf node b.</li>
</ul>
<p>The maximum element in the neighbour joining matrix is guaranteed to be for two neighbouring leaf nodes. For example, the following distance matrix produces the following neighbour joining matrix...</p>
<table>
<thead>
<tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr>
</thead>
<tbody>
<tr>
<td>v0</td>
<td>0</td>
<td>13</td>
<td>21</td>
<td>21</td>
<td>22</td>
<td>22</td>
</tr>
<tr>
<td>v1</td>
<td>13</td>
<td>0</td>
<td>12</td>
<td>12</td>
<td>13</td>
<td>13</td>
</tr>
<tr>
<td>v2</td>
<td>21</td>
<td>12</td>
<td>0</td>
<td>20</td>
<td>21</td>
<td>21</td>
</tr>
<tr>
<td>v3</td>
<td>21</td>
<td>12</td>
<td>20</td>
<td>0</td>
<td>7</td>
<td>13</td>
</tr>
<tr>
<td>v4</td>
<td>22</td>
<td>13</td>
<td>21</td>
<td>7</td>
<td>0</td>
<td>14</td>
</tr>
<tr>
<td>v5</td>
<td>22</td>
<td>13</td>
<td>21</td>
<td>13</td>
<td>14</td>
<td>0</td>
</tr>
</tbody>
</table>
<table>
<thead>
<tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr>
</thead>
<tbody>
<tr>
<td>v0</td>
<td>0</td>
<td>110</td>
<td>110</td>
<td>88</td>
<td>88</td>
<td>94</td>
</tr>
<tr>
<td>v1</td>
<td>110</td>
<td>0</td>
<td>110</td>
<td>88</td>
<td>88</td>
<td>94</td>
</tr>
<tr>
<td>v2</td>
<td>110</td>
<td>110</td>
<td>0</td>
<td>88</td>
<td>88</td>
<td>94</td>
</tr>
<tr>
<td>v3</td>
<td>88</td>
<td>88</td>
<td>88</td>
<td>0</td>
<td>122</td>
<td>104</td>
</tr>
<tr>
<td>v4</td>
<td>88</td>
<td>88</td>
<td>88</td>
<td>122</td>
<td>0</td>
<td>104</td>
</tr>
<tr>
<td>v5</td>
<td>94</td>
<td>94</td>
<td>94</td>
<td>104</td>
<td>104</td>
<td>0</td>
</tr>
</tbody>
</table>
<p>The maximum element is for (v3, v4), meaning that v3 and v4 are neighbouring leaf nodes.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>See Algorithms/Phylogeny/Find Neighbours for a full explanation of how this algorithm works.</p>
</div>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="neighbour joining phylogeny"></span> - A recursive algorithm that can either...</p>
<ul>
<li>find the unique simple tree for an additive distance matrix.</li>
<li>approximate a simple tree for a non-additive distance matrix.</li>
</ul>
<p>The algorithm finds and replaces a pair of neighbouring leaf nodes in the distance matrix with their shared parent at each recursive step (parent is promoted to a leaf node) until the distance matrix has a size of two. The simple tree for any two leaf nodes is those two nodes connected by a single edge. Using that tree as its base, the algorithm recurses out of each step by attaching the neighbours removed from the distance at that step to their parent in the tree.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The term neighbouring means having a shared parent in the tree, not next to each other in the distance matrix.</p>
</div>
<p>At the end, the algorithm will have constructed the entire simple tree for the additive distance matrix. For example, ...</p>
<ul>
<li>
<p>Initial non-additive distance matrix ...</p>
<table>
<thead>
<tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
</tr>
</thead>
<tbody>
<tr>
<td>v0</td>
<td>0</td>
<td>16</td>
<td>22</td>
<td>22</td>
</tr>
<tr>
<td>v1</td>
<td>16</td>
<td>0</td>
<td>13</td>
<td>12</td>
</tr>
<tr>
<td>v2</td>
<td>22</td>
<td>13</td>
<td>0</td>
<td>11</td>
</tr>
<tr>
<td>v3</td>
<td>22</td>
<td>12</td>
<td>11</td>
<td>0</td>
</tr>
</tbody>
</table>
</li>
<li>
<p>Replace neighbours (v1, v0) with their parent N1 to produce distance matrix ...</p>
<table>
<thead>
<tr>
<th></th>
<th>N1</th>
<th>v2</th>
<th>v3</th>
</tr>
</thead>
<tbody>
<tr>
<td>N1</td>
<td>0</td>
<td>22</td>
<td>22</td>
</tr>
<tr>
<td>v2</td>
<td>9.5</td>
<td>0</td>
<td>11</td>
</tr>
<tr>
<td>v3</td>
<td>9</td>
<td>11</td>
<td>0</td>
</tr>
</tbody>
</table>
</li>
<li>
<p>Replace neighbours (v2, v3) with their parent N2 to produce distance matrix ...</p>
<table>
<thead>
<tr>
<th></th>
<th>N1</th>
<th>N2</th>
</tr>
</thead>
<tbody>
<tr>
<td>N1</td>
<td>0</td>
<td>3.75</td>
</tr>
<tr>
<td>N2</td>
<td>3.75</td>
<td>0</td>
</tr>
</tbody>
</table>
</li>
<li>
<p>Distance matrix maps to the obvious simple tree...</p>
<p><img src="dot_9f4177ab91402e113243c9985f953563.svg" alt="Dot diagram"></p>
</li>
<li>
<p>Attach (v2, v3) to N2 to produce tree...</p>
<p><img src="dot_3bcebffa84a180db2bd3ef759fd52701.svg" alt="Dot diagram"></p>
</li>
<li>
<p>Attach (v1, v0) to N1 to produce tree...</p>
<p><img src="dot_9a3e57b6b197a70fe4edf94c50972151.svg" alt="Dot diagram"></p>
</li>
</ul>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>See Algorithms/Phylogeny/Distance Matrix to Tree/Neighbour Joining Phylogeny Algorithm for a full explanation of how this algorithm works.</p>
</div>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(paleontology|palaeontology|palæontology)/i"></span> - The scientific study of ancient organisms: dinosaurs, prehistoric plants, prehistoric insects, prehistoric fungi, etc...</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(anatomy|anatomical)/i"></span> - The study of the identification and description of structures in organisms.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(physiology|physiological)/i"></span> - The study of organism function.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="character table"></span> <span class="anki-answerpattern" data-pattern="character vector"></span> - A matrix where the columns represent biological entities and the rows represent characteristics of those entities, where those characteristics are typically anatomically or physiologically.</p>
<table>
<thead>
<tr>
<th></th>
<th>wings</th>
<th>sucks blood</th>
<th>number of legs</th>
</tr>
</thead>
<tbody>
<tr>
<td>house fly</td>
<td>2</td>
<td>no</td>
<td>6</td>
</tr>
<tr>
<td>mosquito</td>
<td>2</td>
<td>yes</td>
<td>6</td>
</tr>
<tr>
<td>snail</td>
<td>0</td>
<td>no</td>
<td>0</td>
</tr>
</tbody>
</table>
<p>Character tables were commonly used for phylogeny before discovering that DNA can be used to compare the relatedness of organisms.</p>
<p>A row in a character table is referred to as a character vector. Prior to the advent of sequencers, scientists would treat character vectors as sequences for generating phylogenetic trees or doing comparisons between organisms.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(mitochondrial DNA|mtDNA)/i"></span> - DNA unique to the mitochondria. This DNA is unique to the mitochondria, different from the DNA of the cell that the mitochondria lives in. The mitochondria is suspected of being bacteria that made it into the cell and survived, forming a symbiotic relationship.</p>
<p>Mitochondrial DNA is inherited fully from the mother. It isn't a mix of parental DNA as the cell DNA is.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(small parsimony|large parsimony)/i"></span> - In the context of phylogenetic trees, ...</p>
<ul>
<li>small parsimony: When a tree structure and its leaf node sequences are given, derive the internal node sequences with the lowest possible distance (most parsimonious).</li>
<li>large parsimony: When only the leaf node sequences are given, derive the combination of tree structure and internal node sequences with the lowest possible distance (most parsimonious).</li>
</ul>
<p><img src="svgbob_4869b4c66c4290aef524561ee17190d9cff0640c.svg" alt="Kroki diagram output"></p>
<p>Large parsimony isn't a process that's normally done because the search space explodes in size (e.g. NP-complete). Instead, small parsimony is used on a tree generated using an algorithm like UPGMA or neighbour joining phylogeny.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The parsimony score algorithm is what's typically used to evaluate how well a combination of tree structure + ancestral sequences do.</p>
</div>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="parsimony score"></span> - Given a phylogenetic tree with sequences for both leaf nodes (known entities) and internal nodes (inferred ancestral entities), the parsimony score is a measure of how far off a parent's sequence is from its children (and vice versa). The idea is that the most parsimonious evolutionary path is the one that's most likely to have occurred. As such, the less far off sequences are, the more likely it is that the actual ancestral lineage and ancestral sequences match what's depicted in the tree.</p>
<p><img src="svgbob_68590e7d4bd005a4f80364ba3967a7124b6b6527.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="Saccharomyces"></span> - A genus of yeast used for producing alcohol and bread.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="diauxic shift"></span> - A change in the metabolism of Saccharomyces cerevisiae. When glucose is present, Saccharomyces cerevisiae consumes that glucose for energy and produces ethanol as a byproduct. Once all available glucose in the environment has been depleted, it inverts its metabolism to instead consume the ethanol byproduct it produced earlier.</p>
<p>The consumption of ethanol only happens in the presence of oxygen. Without oxygen, Saccharomyces cerevisiae enters hibernation until either glucose or oxygen become available. This property is what allows for the making of wine: Wine production typically involves sealed barrels that prevent oxygen from entering.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="whole genome duplication"></span> - A rare evolutionary event in which the entire genome of an organism is duplicated.</p>
<p>After duplication, much of the functionality provided by the genome becomes redundant. The organism evolves much more rapidly because a mutation in one copy of a gene won't necessarily make the organism less fit (duplicate copy still exists). It's typical for a whole genome duplication to be quickly followed up by massive amounts of gene mutations, gene loss, and genome rearrangements.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(gene expression matrix|expression matrix|gene expression matrices|expression matrices|gene expression data|expression data|gene expression vector|expression vector)/i"></span> - A matrix where each column represents a point in time, each row represents a gene, and each cell is a number representing the amount of gene expression taking place for that gene (row) at that time (column).</p>
<table>
<thead>
<tr>
<th></th>
<th>5 AM</th>
<th>6 AM</th>
<th>7 AM</th>
</tr>
</thead>
<tbody>
<tr>
<td>Gene 1</td>
<td>1.0</td>
<td>1.0</td>
<td>1.0</td>
</tr>
<tr>
<td>Gene 2</td>
<td>1.0</td>
<td>0.7</td>
<td>0.5</td>
</tr>
<tr>
<td>Gene 3</td>
<td>1.0</td>
<td>1.1</td>
<td>1.4</td>
</tr>
</tbody>
</table>
<p>Each row in a gene expression matrix is called a gene expression vector.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="co-regulate"></span> - A set of genes whose gene expression is regulated by the same transcription factor.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(RNA transcript|transcript)/i"></span> - RNA output of transcription.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="transcriptome"></span> - All RNA transcripts within a cell at a specific point in time.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="good clustering principle"></span> - The idea that items within the same cluster should be more similar to each other than items in other clusters.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>This was originally introduced as "every pair of points within the same cluster should be closer to each other than any points each from different clusters", where "closer" was implying euclidean distance. I think the idea was intended to be abstracted out to the definition above since the items may not be thought of as "points" but as vectors or sequences + you can choose a similarity metric other than euclidean distance.</p>
</div>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="euclidean distance"></span> - The distance between two points if traveling directly from one to the other in a straight line.</p>
<p><img src="svgbob_812b661882246e5c5673044d47b2a743122f1f75.svg" alt="Kroki diagram output"></p>
<p>In 2 dimensional space, this is calculated as <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msqrt><mrow><mo stretchy="false">(</mo><msub><mi>x</mi><mn>2</mn></msub><mo>−</mo><msub><mi>x</mi><mn>1</mn></msub><msup><mo stretchy="false">)</mo><mn>2</mn></msup><mo>+</mo><mo stretchy="false">(</mo><msub><mi>y</mi><mn>2</mn></msub><mo>−</mo><msub><mi>y</mi><mn>1</mn></msub><msup><mo stretchy="false">)</mo><mn>2</mn></msup></mrow></msqrt></mrow><annotation encoding="application/x-tex">\sqrt{(x_2 - x_1)^2 + (y_2 - y_1)^2}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.24em;vertical-align:-0.30499999999999994em;"></span><span class="mord sqrt"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.935em;"><span class="svg-align" style="top:-3.2em;"><span class="pstrut" style="height:3.2em;"></span><span class="mord" style="padding-left:1em;"><span class="mopen">(</span><span class="mord"><span class="mord mathnormal">x</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal">x</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mclose"><span class="mclose">)</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.740108em;"><span style="top:-2.9890000000000003em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.03588em;">y</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.03588em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.03588em;">y</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.03588em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mclose"><span class="mclose">)</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.740108em;"><span style="top:-2.9890000000000003em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span></span></span><span style="top:-2.8950000000000005em;"><span class="pstrut" style="height:3.2em;"></span><span class="hide-tail" style="min-width:1.02em;height:1.28em;"><svg width="400em" height="1.28em" viewBox="0 0 400000 1296" preserveAspectRatio="xMinYMin slice"><path d="M263,681c0.7,0,18,39.7,52,119
c34,79.3,68.167,158.7,102.5,238c34.3,79.3,51.8,119.3,52.5,120
c340,-704.7,510.7,-1060.3,512,-1067
l0 -0
c4.7,-7.3,11,-11,19,-11
H40000v40H1012.3
s-271.3,567,-271.3,567c-38.7,80.7,-84,175,-136,283c-52,108,-89.167,185.3,-111.5,232
c-22.3,46.7,-33.8,70.3,-34.5,71c-4.7,4.7,-12.3,7,-23,7s-12,-1,-12,-1
s-109,-253,-109,-253c-72.7,-168,-109.3,-252,-110,-252c-10.7,8,-22,16.7,-34,26
c-22,17.3,-33.3,26,-34,26s-26,-26,-26,-26s76,-59,76,-59s76,-60,76,-60z
M1001 80h400000v40h-400000z"></path></svg></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.30499999999999994em;"><span></span></span></span></span></span></span></span></span>.</p>
<p>In 3 dimensional space, this is calculated as <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msqrt><mrow><mo stretchy="false">(</mo><msub><mi>x</mi><mn>2</mn></msub><mo>−</mo><msub><mi>x</mi><mn>1</mn></msub><msup><mo stretchy="false">)</mo><mn>2</mn></msup><mo>+</mo><mo stretchy="false">(</mo><msub><mi>y</mi><mn>2</mn></msub><mo>−</mo><msub><mi>y</mi><mn>1</mn></msub><msup><mo stretchy="false">)</mo><mn>2</mn></msup><mo>+</mo><mo stretchy="false">(</mo><msub><mi>z</mi><mn>2</mn></msub><mo>−</mo><msub><mi>z</mi><mn>1</mn></msub><msup><mo stretchy="false">)</mo><mn>2</mn></msup></mrow></msqrt></mrow><annotation encoding="application/x-tex">\sqrt{(x_2 - x_1)^2 + (y_2 - y_1)^2 + (z_2 - z_1)^2}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.24em;vertical-align:-0.30499999999999994em;"></span><span class="mord sqrt"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.935em;"><span class="svg-align" style="top:-3.2em;"><span class="pstrut" style="height:3.2em;"></span><span class="mord" style="padding-left:1em;"><span class="mopen">(</span><span class="mord"><span class="mord mathnormal">x</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal">x</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mclose"><span class="mclose">)</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.740108em;"><span style="top:-2.9890000000000003em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.03588em;">y</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.03588em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.03588em;">y</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.03588em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mclose"><span class="mclose">)</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.740108em;"><span style="top:-2.9890000000000003em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.04398em;">z</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.04398em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.04398em;">z</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.04398em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mclose"><span class="mclose">)</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.740108em;"><span style="top:-2.9890000000000003em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span></span></span><span style="top:-2.8950000000000005em;"><span class="pstrut" style="height:3.2em;"></span><span class="hide-tail" style="min-width:1.02em;height:1.28em;"><svg width="400em" height="1.28em" viewBox="0 0 400000 1296" preserveAspectRatio="xMinYMin slice"><path d="M263,681c0.7,0,18,39.7,52,119
c34,79.3,68.167,158.7,102.5,238c34.3,79.3,51.8,119.3,52.5,120
c340,-704.7,510.7,-1060.3,512,-1067
l0 -0
c4.7,-7.3,11,-11,19,-11
H40000v40H1012.3
s-271.3,567,-271.3,567c-38.7,80.7,-84,175,-136,283c-52,108,-89.167,185.3,-111.5,232
c-22.3,46.7,-33.8,70.3,-34.5,71c-4.7,4.7,-12.3,7,-23,7s-12,-1,-12,-1
s-109,-253,-109,-253c-72.7,-168,-109.3,-252,-110,-252c-10.7,8,-22,16.7,-34,26
c-22,17.3,-33.3,26,-34,26s-26,-26,-26,-26s76,-59,76,-59s76,-60,76,-60z
M1001 80h400000v40h-400000z"></path></svg></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.30499999999999994em;"><span></span></span></span></span></span></span></span></span>.</p>
<p>In n dimensional space, this is calculated as <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msqrt><mrow><msubsup><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></msubsup><mrow><mo stretchy="false">(</mo><msub><mi>v</mi><mi>i</mi></msub><mo>−</mo><msub><mi>w</mi><mi>i</mi></msub><msup><mo stretchy="false">)</mo><mn>2</mn></msup></mrow></mrow></msqrt></mrow><annotation encoding="application/x-tex">\sqrt{\sum_{i=1}^n{(v_i - w_i)^2}}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.24em;vertical-align:-0.302709em;"></span><span class="mord sqrt"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.937291em;"><span class="svg-align" style="top:-3.2em;"><span class="pstrut" style="height:3.2em;"></span><span class="mord" style="padding-left:1em;"><span class="mop"><span class="mop op-symbol small-op" style="position:relative;top:-0.0000050000000000050004em;">∑</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.804292em;"><span style="top:-2.40029em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">i</span><span class="mrel mtight">=</span><span class="mord mtight">1</span></span></span></span><span style="top:-3.2029em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">n</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.29971000000000003em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord"><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.03588em;">v</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:-0.03588em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02691em;">w</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:-0.02691em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mclose"><span class="mclose">)</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.740108em;"><span style="top:-2.9890000000000003em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span></span></span></span><span style="top:-2.897291em;"><span class="pstrut" style="height:3.2em;"></span><span class="hide-tail" style="min-width:1.02em;height:1.28em;"><svg width="400em" height="1.28em" viewBox="0 0 400000 1296" preserveAspectRatio="xMinYMin slice"><path d="M263,681c0.7,0,18,39.7,52,119
c34,79.3,68.167,158.7,102.5,238c34.3,79.3,51.8,119.3,52.5,120
c340,-704.7,510.7,-1060.3,512,-1067
l0 -0
c4.7,-7.3,11,-11,19,-11
H40000v40H1012.3
s-271.3,567,-271.3,567c-38.7,80.7,-84,175,-136,283c-52,108,-89.167,185.3,-111.5,232
c-22.3,46.7,-33.8,70.3,-34.5,71c-4.7,4.7,-12.3,7,-23,7s-12,-1,-12,-1
s-109,-253,-109,-253c-72.7,-168,-109.3,-252,-110,-252c-10.7,8,-22,16.7,-34,26
c-22,17.3,-33.3,26,-34,26s-26,-26,-26,-26s76,-59,76,-59s76,-60,76,-60z
M1001 80h400000v40h-400000z"></path></svg></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.302709em;"><span></span></span></span></span></span></span></span></span>, where v and w are two n dimensional points.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(manhattan distance|taxi[\s\-]?cab distance|city[\s\-]block distance)/i"></span> - The distance between two points if traveling only via the axis of the coordinate system.</p>
<p><img src="svgbob_19c8269cf6dc61112d64c53c1045bfe1e1f518e3.svg" alt="Kroki diagram output"></p>
<p>In n dimensional space, this is calculated as <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msubsup><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></msubsup><mrow><mi mathvariant="normal">∣</mi><msub><mi>v</mi><mi>i</mi></msub><mo>−</mo><msub><mi>w</mi><mi>i</mi></msub><mi mathvariant="normal">∣</mi></mrow></mrow><annotation encoding="application/x-tex">\sum_{i=1}^n{|v_i - w_i|}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.104002em;vertical-align:-0.29971000000000003em;"></span><span class="mop"><span class="mop op-symbol small-op" style="position:relative;top:-0.0000050000000000050004em;">∑</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.804292em;"><span style="top:-2.40029em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">i</span><span class="mrel mtight">=</span><span class="mord mtight">1</span></span></span></span><span style="top:-3.2029em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">n</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.29971000000000003em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord"><span class="mord">∣</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.03588em;">v</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:-0.03588em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02691em;">w</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:-0.02691em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mord">∣</span></span></span></span></span>, where v and w are two n dimensional points.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The name manhattan distance is an allegory to the city blocks of manhattan, where your options (most of the time) are to move either left/right or up/down. Other names for this same metric are taxi-cab distance and city block distance.</p>
</div>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(cluster)/i"></span> - Grouping together a set of objects such that objects within the same group are more similar to each other than objects in other groups. Each group is referred to as a cluster.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(membership|member)/i"></span> - An object assigned to a cluster is said to be a member of that cluster.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(k-centers? clustering|\d+-centers? clustering|k-centers?|\d+-centers?)/i"></span> - A form of clustering where a point, called a center, defines cluster membership_NORM. k different centers are chosen (one for each cluster), and the points closest to each center (euclidean distance) make up members of that cluster. The goal is to choose centers such that, out of all possible cluster center to member distances, the farthest distance is the minimum it could possibly be out of all possible choices for centers.</p>
<p>In terms of a scoring function, the score being minimized is ...</p>
<p><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><mi>s</mi><mi>c</mi><mi>o</mi><mi>r</mi><mi>e</mi><mo>=</mo><mi>m</mi><mi>a</mi><mi>x</mi><mo stretchy="false">(</mo><mi>d</mi><mo stretchy="false">(</mo><msub><mi>P</mi><mn>1</mn></msub><mo separator="true">,</mo><mi>C</mi><mo stretchy="false">)</mo><mo separator="true">,</mo><mi>d</mi><mo stretchy="false">(</mo><msub><mi>P</mi><mn>2</mn></msub><mo separator="true">,</mo><mi>C</mi><mo stretchy="false">)</mo><mo separator="true">,</mo><mi mathvariant="normal">.</mi><mi mathvariant="normal">.</mi><mi mathvariant="normal">.</mi><mo separator="true">,</mo><mi>d</mi><mo stretchy="false">(</mo><msub><mi>P</mi><mi>n</mi></msub><mo separator="true">,</mo><mi>C</mi><mo stretchy="false">)</mo><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">score = max(d(P_1, C), d(P_2, C), ..., d(P_n, C))
</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.43056em;vertical-align:0em;"></span><span class="mord mathnormal">s</span><span class="mord mathnormal">c</span><span class="mord mathnormal">o</span><span class="mord mathnormal" style="margin-right:0.02778em;">r</span><span class="mord mathnormal">e</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal">m</span><span class="mord mathnormal">a</span><span class="mord mathnormal">x</span><span class="mopen">(</span><span class="mord mathnormal">d</span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.13889em;">P</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.13889em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mpunct">,</span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord mathnormal" style="margin-right:0.07153em;">C</span><span class="mclose">)</span><span class="mpunct">,</span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord mathnormal">d</span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.13889em;">P</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.13889em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mpunct">,</span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord mathnormal" style="margin-right:0.07153em;">C</span><span class="mclose">)</span><span class="mpunct">,</span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord">.</span><span class="mord">.</span><span class="mord">.</span><span class="mpunct">,</span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord mathnormal">d</span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.13889em;">P</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.151392em;"><span style="top:-2.5500000000000003em;margin-left:-0.13889em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">n</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mpunct">,</span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord mathnormal" style="margin-right:0.07153em;">C</span><span class="mclose">)</span><span class="mclose">)</span></span></span></span></span></p>
<ul>
<li>P is the set of points.</li>
<li>C is the set of centers.</li>
<li>n is the number of points in P</li>
<li>d returns the euclidean distance between a point and its closest center.</li>
<li>max is the maximum function.</li>
</ul>
<pre class="hljs"><code><span class="hljs-comment"># d() function from the formula</span>
<span class="hljs-keyword">def</span> <span class="hljs-title function_">dist_to_closest_center</span>(<span class="hljs-params">data_pt, center_pts</span>):
  center_pt = <span class="hljs-built_in">min</span>(
    center_pts,
    key=<span class="hljs-keyword">lambda</span> cp: dist(data_pt, cp)
  )
  <span class="hljs-keyword">return</span> dist(center_pt, data_pt)

<span class="hljs-comment"># scoring function (what's trying to be minimized)</span>
<span class="hljs-keyword">def</span> <span class="hljs-title function_">k_centers_score</span>(<span class="hljs-params">data_pts, center_pts</span>):
  <span class="hljs-keyword">return</span> <span class="hljs-built_in">max</span>(dist_to_closest_center(p, center_pts) <span class="hljs-keyword">for</span> p <span class="hljs-keyword">in</span> data_pts)
</code></pre>
<p><img src="svgbob_5fa140a687d17ef8490f92bae2a74d597d2d56f9.svg" alt="Kroki diagram output"></p>
<p>For a non-trivial input, the search space is too massive for a straight-forward algorithm to work. As such, heuristics are commonly used instead.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>See farthest first traversal heuristic.</p>
</div>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="farthest first traversal"></span> - A heuristic commonly used for k-centers clustering. The algorithm iteratively builds out more centers by inspecting the euclidean distances from points to existing centers. At each step, the algorithm ...</p>
<ol>
<li>gets the closest center for each point,</li>
<li>picks the point with the farthest euclidean distance and sets it as the new center.</li>
</ol>
<p>The algorithm initially primes the list of centers with a randomly chosen point and stops executing once it has k points.</p>
<p><img src="svgbob_52c67bd6a373b1ba175fc4205574ee9cfc99b6fb.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(k-means?|\d+-means?)/i"></span> <span class="anki-answerpattern" data-pattern="(squared error distortion)/i"></span> - A form of clustering where a point, called a center, defines cluster membership_NORM. k different centers are chosen (one for each cluster), and the points closest to each center (euclidean distance) make up members of that cluster. The goal is to choose centers such that, out of all possible cluster center to member distances, the formula below is the minimum it could possibly be out of all possible choices for centers.</p>
<p>The formula below is referred to as squared error distortion.</p>
<p><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><mi>s</mi><mi>c</mi><mi>o</mi><mi>r</mi><mi>e</mi><mo>=</mo><mfrac><mrow><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><mrow><mi>d</mi><mo stretchy="false">(</mo><msub><mi>P</mi><mi>i</mi></msub><mo separator="true">,</mo><mi>C</mi><msup><mo stretchy="false">)</mo><mn>2</mn></msup></mrow></mrow><mi>n</mi></mfrac></mrow><annotation encoding="application/x-tex">score = \frac{\sum_{i=1}^{n} {d(P_i, C)^2}}{n}
</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.43056em;vertical-align:0em;"></span><span class="mord mathnormal">s</span><span class="mord mathnormal">c</span><span class="mord mathnormal">o</span><span class="mord mathnormal" style="margin-right:0.02778em;">r</span><span class="mord mathnormal">e</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:2.189818em;vertical-align:-0.686em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.5038179999999999em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord mathnormal">n</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.6897100000000003em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mop"><span class="mop op-symbol small-op" style="position:relative;top:-0.0000050000000000050004em;">∑</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.804292em;"><span style="top:-2.40029em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">i</span><span class="mrel mtight">=</span><span class="mord mtight">1</span></span></span></span><span style="top:-3.2029em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">n</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.29971000000000003em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord"><span class="mord mathnormal">d</span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.13889em;">P</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:-0.13889em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mpunct">,</span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord mathnormal" style="margin-right:0.07153em;">C</span><span class="mclose"><span class="mclose">)</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8141079999999999em;"><span style="top:-3.063em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span></span></p>
<ul>
<li>P is the set of points.</li>
<li>C is the set of centers.</li>
<li>n is the number of points in P</li>
<li>d returns the euclidean distance between a point and its closest center.</li>
<li>max is the maximum function.</li>
</ul>
<pre class="hljs"><code><span class="hljs-comment"># d() function from the formula</span>
<span class="hljs-keyword">def</span> <span class="hljs-title function_">find_closest_center</span>(<span class="hljs-params">data_pt, center_pts</span>):
  center_pt = <span class="hljs-built_in">min</span>(
    center_pts,
    key=<span class="hljs-keyword">lambda</span> cp: dist(data_pt, cp)
  )
  <span class="hljs-keyword">return</span> center_pt, dist(center_pt, data_pt)

<span class="hljs-comment"># scoring function (what's trying to be minimized) -- taking squares of d() and averaging</span>
<span class="hljs-keyword">def</span> <span class="hljs-title function_">squared_error_distortion</span>(<span class="hljs-params">data_pts, center_pts</span>):
  res = []
  <span class="hljs-keyword">for</span> data_pt <span class="hljs-keyword">in</span> data_pts:
    closest_center_pt, dist_to = find_closest_center(data_pt, center_pts)
    res.append(dist_to ** <span class="hljs-number">2</span>)
  <span class="hljs-keyword">return</span> <span class="hljs-built_in">sum</span>(res) / <span class="hljs-built_in">len</span>(res)
</code></pre>
<p>Compared to k-centers, k-means more appropriately handles outliers. If outliers are present, k-centers's metric causes the cluster center to be wildly offset while k-mean's metric will only be mildly offset. In the example below, the best scoring cluster center for k-centers is wildly offset by outlier Z while k-means isn't offset as much.</p>
<p><img src="svgbob_50f9ef0bee677ed328acdaec5260e8530cf4b1be.svg" alt="Kroki diagram output"></p>
<p>For a non-trivial input, the search space is too massive for a straight-forward algorithm to work. As such, heuristics are commonly used instead.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>See Lloyd's algorithm.</p>
</div>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The Pevzner book states that k-means clustering often doesn't produce good results if the clumps have a non-globular shape (e.g. elongated) or have widely different densities. This probably applies to k-centers as well</p>
</div>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(lloyd algorithm|lloyd's algorithm)/i"></span> <span class="anki-answerpattern" data-pattern="(k-means?\+\+ initializer|k-means?\+\+)/i"></span> - A heuristic for determining centers in k-means clustering. The algorithm begins by choosing k arbitrary points from the points being clustered as the initial centers, then ...</p>
<ol>
<li>
<p>derives clusters from centers: Each point is assigned to its nearest center (ties broken arbitrarily).</p>
</li>
<li>
<p>derives centers from clusters: Each center is updated to be the "center of gravity" of its points.</p>
<p>The center of gravity for a set of points is the average of each individual coordinate. For example, for the 2D points (0,5), (1,3), and (2,2), the center of gravity is (1, 3.3333). The ...</p>
<ul>
<li>first coordinate is calculated as <code>0+1+2/3 = 1</code>.</li>
<li>second coordinate is calculated as <code>5+3+2/3 = 3.3333</code>.</li>
</ul>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">center_of_gravity</span>(<span class="hljs-params">data_pts, dim</span>):
   center = []
   <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(dim):
     val = mean(pt[i] <span class="hljs-keyword">for</span> pt <span class="hljs-keyword">in</span> data_pts)
     center.append(val)
   <span class="hljs-keyword">return</span> center
</code></pre>
</li>
</ol>
<p>The above two steps loop repeatedly until the new centers are the same as the centers from the previous iteration (stops converging).</p>
<p>Since this algorithm is a heuristic, it doesn't always converge to a good solution. The algorithm typically runs multiple times, where the run producing centers with the lowest squared error distortion is accepted. An enhancement to the algorithm, called k-means++ initializer, increases the chances of converging to a good solution by <em>probabilistically</em> selecting initial centers that are far from each other:</p>
<ol>
<li>The 1st center is chosen from the list of points at random.</li>
<li>The 2nd center is chosen by selecting a point that's likely to be much farther away from center 1 than most other points.</li>
<li>The 3rd center is chosen by selecting a point that's likely to be much farther away from center 1 and 2 than most other points.</li>
<li>...</li>
</ol>
<p>The probability of selecting a point as the next center is proportional to its squared distance to the existing centers.</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">select_initial_centers</span>(<span class="hljs-params">data_pts, k</span>):
  centers = [random.choice(data_pts)]
  <span class="hljs-keyword">while</span> <span class="hljs-built_in">len</span>(centers) &lt; k:
    choice_points = []
    choice_weights = []
    <span class="hljs-keyword">for</span> pt <span class="hljs-keyword">in</span> data_pts:
      <span class="hljs-keyword">if</span> pt <span class="hljs-keyword">in</span> centers:
        <span class="hljs-keyword">continue</span>
      _, d = find_closest_center(pt, centers)
      choice_weights.append(d)
      choice_points.append(pt)
    total = <span class="hljs-built_in">sum</span>(choice_weights)
    choice_weights = [w / total <span class="hljs-keyword">for</span> w <span class="hljs-keyword">in</span> choice_weights]
    c_pt = random.choices(choice_points, weights=choice_weights, k=<span class="hljs-number">1</span>).pop(<span class="hljs-number">0</span>)
    centers.append(c_pt)
  <span class="hljs-keyword">return</span> centers
</code></pre>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="hard clustering"></span> <span class="anki-answerpattern" data-pattern="soft clustering"></span> - In the context of clustering, ...</p>
<ul>
<li>
<p>soft clustering algorithms assign each object to a set of probabilities, where each probability is how likely it is for that object to be assigned to a cluster. <span class="anki-listrandomorder"></span>
</p>
<p><img src="svgbob_a9ecf872b60c1d4716510c88c16191c37ea93595.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>hard clustering algorithms assign each object to exactly one cluster.</p>
<p><img src="svgbob_59de88c1a6509780a865b5dee966022210b55bd7.svg" alt="Kroki diagram output"></p>
</li>
</ul>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="dot product"></span> - Given two equal sized vectors, the dot product of those vectors is calculated by first multiplying the pair at each index, then summing the result of those multiplications together. For example, the dot product (1, 2, 3) and (4, 5, 6) is 1*4+2*5+3*6.</p>
<p>The notation for dot product is a central dot in between the two vector: <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>a</mi><mo>⋅</mo><mi>b</mi></mrow><annotation encoding="application/x-tex">a \cdot b</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.44445em;vertical-align:0em;"></span><span class="mord mathnormal">a</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.69444em;vertical-align:0em;"></span><span class="mord mathnormal">b</span></span></span></span>.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>Central dot is also commonplace for standard multiplication.</p>
<p>In geometry, the dot product of two vectors is used to get the angle between those vectors.</p>
</div>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="conditional probability"></span> - The probability of an event occurring given that another event has already occurred.</p>
<p>The notation for conditional probability is Pr(B|A), where B is the event occurring and A is the event that's already occurred.</p>
<p>If B and A are...</p>
<ul>
<li>independent events, Pr(B|A) is simply Pr(B).</li>
<li>dependent events, Pr(B|A) is calculated as the probability that both B and A happen divided by the probability that just A happens: Pr(A∩B) / Pr(A).</li>
</ul>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="similarity metric"></span> - A metric used to measure how similar a pair of entities are to each other. Whereas a distance metric must start at 0 for total similarity and grows based on how different the entities are, a similarity metric has no requirements for bounds on similarity or dissimilarity. Examples of similarity metrics include ...</p>
<ul>
<li>pearson similarity for gene expression vectors.</li>
<li>cosine similarity for gene expression vectors.</li>
<li>BLOSUM / PAM matrices for protein sequence alignments.</li>
<li>etc..</li>
</ul>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>This topic was only briefly discussed, so I don't know for sure what the properties/requirements are for a similarity metric other than higher = more similar. Contrast this to distance metrics, where it explicitly mentions the requirements that need to be followed (e.g. triangle inequality property). For similarity metrics, it didn't say if there's some upper-bound to similarity or if totally similar entities have to score the same. For example, does <code>similarity(snake,snake) == similarity(bird,bird)</code> have to be true or can it be that <code>similarity(snake,snake) &gt; similarity(bird,bird)</code>?</p>
<p>I saw on Wikipedia that sequence alignment scoring matrices like PAM and BLOSUM are similarity matrices, so that implies that totally similar entities don't have to be the same score. For example, in BLOSUM62 <code>similarity(A,A) = 4</code> but <code>similarity(R,R) = 5</code>.</p>
<p>Also, does a similarity metric have to be symmetric? For example, <code>similarity(snake,bird) == similarity(bird,snake)</code>. I think it does have to be symmetric.</p>
</div>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(similarity matrix|similarity matrices)/i"></span> - Given a set of n different entities, a similarity matrix is an n-by-n matrix where each element contains the similarity measure between the entities for that cell. For example, for the species snake, lizard, bird, and crocodile ...</p>
<table>
<thead>
<tr>
<th></th>
<th>Snake</th>
<th>Lizard</th>
<th>Bird</th>
<th>Crocodile</th>
</tr>
</thead>
<tbody>
<tr>
<td>Snake</td>
<td>1.0</td>
<td>0.8</td>
<td>0.4</td>
<td>0.6</td>
</tr>
<tr>
<td>Lizard</td>
<td>0.8</td>
<td>1.0</td>
<td>0.4</td>
<td>0.6</td>
</tr>
<tr>
<td>Bird</td>
<td>0.4</td>
<td>0.4</td>
<td>1.0</td>
<td>0.5</td>
</tr>
<tr>
<td>Crocodile</td>
<td>0.6</td>
<td>0.6</td>
<td>0.5</td>
<td>1.0</td>
</tr>
</tbody>
</table>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>This topic was only briefly discussed, so I have no idea what properties are required other than: 0 = completely dissimilar / orthogonal and anything higher than that is more similar. It didn't say if there's some upper-bound to similarity or if totally similar entities have to score the same. For example, does <code>similarity(snake,snake) == similarity(bird,bird)</code> have to be true or can it be that <code>similarity(snake,snake) &gt; similarity(bird,bird)</code>? I saw on Wikipedia that sequence alignment scoring matrices like PAM and BLOSUM are similarity matrices, so that implies that totally similar entities don't have to be the same score. For example, in BLOSUM62 <code>similarity(A,A) = 4</code> but <code>similarity(R,R) =5</code>.</p>
<p>There may be other properties involved, such as how the triangle inequality property is a thing for distance matrices / distance metrics.</p>
</div>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(pearson correlation coefficient|pearson correlation)/i"></span> <span class="anki-answerpattern" data-pattern="pearson distance"></span> <span class="anki-answerpattern" data-pattern="pearson similarity"></span> - A metric used to quantify how correlated two vectors are.</p>
<p><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><mfrac><mrow><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>m</mi></munderover><mo stretchy="false">(</mo><msub><mi>x</mi><mi>i</mi></msub><mo>−</mo><mi>a</mi><mi>v</mi><mi>g</mi><mo stretchy="false">(</mo><mi>x</mi><mo stretchy="false">)</mo><mo stretchy="false">)</mo><mo>⋅</mo><mo stretchy="false">(</mo><msub><mi>y</mi><mi>i</mi></msub><mo>−</mo><mi>a</mi><mi>v</mi><mi>g</mi><mo stretchy="false">(</mo><mi>y</mi><mo stretchy="false">)</mo><mo stretchy="false">)</mo></mrow><msqrt><mrow><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>m</mi></munderover><mo stretchy="false">(</mo><msub><mi>x</mi><mi>i</mi></msub><mo>−</mo><mi>a</mi><mi>v</mi><mi>g</mi><mo stretchy="false">(</mo><mi>x</mi><mo stretchy="false">)</mo><msup><mo stretchy="false">)</mo><mn>2</mn></msup><mo>⋅</mo><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>m</mi></munderover><mo stretchy="false">(</mo><msub><mi>y</mi><mi>i</mi></msub><mo>−</mo><mi>a</mi><mi>v</mi><mi>g</mi><mo stretchy="false">(</mo><mi>y</mi><mo stretchy="false">)</mo><msup><mo stretchy="false">)</mo><mn>2</mn></msup></mrow></msqrt></mfrac></mrow><annotation encoding="application/x-tex">\frac{
\sum_{i=1}^{m}(x_i-avg(x))\cdot(y_i-avg(y))
}{
\sqrt{
\sum_{i=1}^{m}(x_i-avg(x))^2\cdot\sum_{i=1}^{m}(y_i-avg(y))^2
}
}
</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:2.624002em;vertical-align:-1.13em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.494002em;"><span style="top:-2.1727090000000002em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord sqrt"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.937291em;"><span class="svg-align" style="top:-3.2em;"><span class="pstrut" style="height:3.2em;"></span><span class="mord" style="padding-left:1em;"><span class="mop"><span class="mop op-symbol small-op" style="position:relative;top:-0.0000050000000000050004em;">∑</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.804292em;"><span style="top:-2.40029em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">i</span><span class="mrel mtight">=</span><span class="mord mtight">1</span></span></span></span><span style="top:-3.2029em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">m</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.29971000000000003em;"><span></span></span></span></span></span></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal">x</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord mathnormal">a</span><span class="mord mathnormal" style="margin-right:0.03588em;">v</span><span class="mord mathnormal" style="margin-right:0.03588em;">g</span><span class="mopen">(</span><span class="mord mathnormal">x</span><span class="mclose">)</span><span class="mclose"><span class="mclose">)</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.740108em;"><span style="top:-2.9890000000000003em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mop"><span class="mop op-symbol small-op" style="position:relative;top:-0.0000050000000000050004em;">∑</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.804292em;"><span style="top:-2.40029em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">i</span><span class="mrel mtight">=</span><span class="mord mtight">1</span></span></span></span><span style="top:-3.2029em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">m</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.29971000000000003em;"><span></span></span></span></span></span></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.03588em;">y</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:-0.03588em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord mathnormal">a</span><span class="mord mathnormal" style="margin-right:0.03588em;">v</span><span class="mord mathnormal" style="margin-right:0.03588em;">g</span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.03588em;">y</span><span class="mclose">)</span><span class="mclose"><span class="mclose">)</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.740108em;"><span style="top:-2.9890000000000003em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span></span></span><span style="top:-2.897291em;"><span class="pstrut" style="height:3.2em;"></span><span class="hide-tail" style="min-width:1.02em;height:1.28em;"><svg width="400em" height="1.28em" viewBox="0 0 400000 1296" preserveAspectRatio="xMinYMin slice"><path d="M263,681c0.7,0,18,39.7,52,119
c34,79.3,68.167,158.7,102.5,238c34.3,79.3,51.8,119.3,52.5,120
c340,-704.7,510.7,-1060.3,512,-1067
l0 -0
c4.7,-7.3,11,-11,19,-11
H40000v40H1012.3
s-271.3,567,-271.3,567c-38.7,80.7,-84,175,-136,283c-52,108,-89.167,185.3,-111.5,232
c-22.3,46.7,-33.8,70.3,-34.5,71c-4.7,4.7,-12.3,7,-23,7s-12,-1,-12,-1
s-109,-253,-109,-253c-72.7,-168,-109.3,-252,-110,-252c-10.7,8,-22,16.7,-34,26
c-22,17.3,-33.3,26,-34,26s-26,-26,-26,-26s76,-59,76,-59s76,-60,76,-60z
M1001 80h400000v40h-400000z"></path></svg></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.302709em;"><span></span></span></span></span></span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.6897100000000003em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mop"><span class="mop op-symbol small-op" style="position:relative;top:-0.0000050000000000050004em;">∑</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.804292em;"><span style="top:-2.40029em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">i</span><span class="mrel mtight">=</span><span class="mord mtight">1</span></span></span></span><span style="top:-3.2029em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">m</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.29971000000000003em;"><span></span></span></span></span></span></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal">x</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord mathnormal">a</span><span class="mord mathnormal" style="margin-right:0.03588em;">v</span><span class="mord mathnormal" style="margin-right:0.03588em;">g</span><span class="mopen">(</span><span class="mord mathnormal">x</span><span class="mclose">)</span><span class="mclose">)</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.03588em;">y</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:-0.03588em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord mathnormal">a</span><span class="mord mathnormal" style="margin-right:0.03588em;">v</span><span class="mord mathnormal" style="margin-right:0.03588em;">g</span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.03588em;">y</span><span class="mclose">)</span><span class="mclose">)</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.13em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span></span></p>
<p>In the above formula, x and y are the two input vectors and avg is the average function. The result of the formula is a number between -1 and 1, where ...</p>
<ul>
<li>-1 represents a total negative correlation.</li>
<li>0 represents no correlation.</li>
<li>1 represents a total positive correlation.</li>
</ul>
<p>The formula may be modified to become a distance metric as follows: <code>1 - metric(x, y)</code>. Whereas the pearson correlation coefficient varies between -1 and 1, the pearson distance varies between 0 (totally similar) and 2 (totally dissimilar).</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="similarity graph"></span> - A transformation of a similarity matrix into a graph, where the entities that make up the similarity matrix are represented as nodes and edges between nodes are only made if the similarity exceeds a certain threshold.</p>
<p>The similarity graph below was generated using the accompanying similarity matrix and threshold of 7.</p>
<table>
<thead>
<tr>
<th></th>
<th>a</th>
<th>b</th>
<th>c</th>
<th>d</th>
<th>e</th>
<th>f</th>
<th>g</th>
</tr>
</thead>
<tbody>
<tr>
<td>a</td>
<td>9</td>
<td>8</td>
<td>9</td>
<td>1</td>
<td>0</td>
<td>1</td>
<td>1</td>
</tr>
<tr>
<td>b</td>
<td>8</td>
<td>9</td>
<td>9</td>
<td>1</td>
<td>1</td>
<td>0</td>
<td>2</td>
</tr>
<tr>
<td>c</td>
<td>9</td>
<td>9</td>
<td>8</td>
<td>2</td>
<td>1</td>
<td>1</td>
<td>1</td>
</tr>
<tr>
<td>d</td>
<td>1</td>
<td>1</td>
<td>2</td>
<td>9</td>
<td>8</td>
<td>9</td>
<td>9</td>
</tr>
<tr>
<td>e</td>
<td>0</td>
<td>1</td>
<td>1</td>
<td>8</td>
<td>8</td>
<td>8</td>
<td>9</td>
</tr>
<tr>
<td>f</td>
<td>1</td>
<td>0</td>
<td>1</td>
<td>9</td>
<td>8</td>
<td>9</td>
<td>9</td>
</tr>
<tr>
<td>g</td>
<td>1</td>
<td>2</td>
<td>1</td>
<td>9</td>
<td>9</td>
<td>9</td>
<td>8</td>
</tr>
</tbody>
</table>
<p><img src="svgbob_735293a6de7bf4348ff0eebf1cb2effc798a8442.svg" alt="Kroki diagram output"></p>
<p>Similarity graphs are used for clustering (e.g. gene expression vectors). Assuming clusters exist and the similarity metric used captures them, there should be some threshold where the edges produced in the similarity graph form cliques as in the example above.</p>
<p>Since real-world data often has complications (e.g. noisy) / the similarity metric used may have complications, it could be that corrupted cliques are formed instead. Heuristic algorithms are often used to correct corrupted cliques.</p>
<p><img src="svgbob_728047af231fb5d360931860850f13d1a4207105.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="clique"></span> - A set of nodes in a graph where every possible node pairing has an edge.</p>
<p><img src="svgbob_331f1a046e4a700b7f330a8c4875cfa3415603f4.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="corrupted clique"></span> - A set of nodes and edges in a graph that almost form a clique. Some edges may be missing or extraneous.</p>
<p><img src="svgbob_fdc15f9a1702553abc7978949e98181d60744807.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="clique graph"></span> - A graph consisting only of cliques.</p>
<p><img src="svgbob_146037c85efe776db75ed0bb94cf76542654c3df.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="cluster affinity search technique"></span> - A heuristic algorithm that corrects the corrupted cliques in a similarity graph.</p>
<p><img src="svgbob_5519c35ef5dcc12c07a7fd7ea6dadc034d343056.svg" alt="Kroki diagram output"></p>
<p>The algorithm attempts to re-create each corrupted clique in its corrected form by iteratively finding the ...</p>
<ol>
<li>closest node not in the clique/cluster and including it <em>if it exceeds the similarity graph threshold</em>.</li>
<li>farthest node within the clique/cluster and removing it <em>if it doesn't exceed the similarity graph threshold</em>.</li>
</ol>
<p>How close or far a gene is from the clique/cluster is defined as the average similarity between that node and all nodes in the clique/cluster.</p>
<p>While the similarity graph has nodes, the algorithm picks the node with the highest degree from the similarity graph to prime a clique/cluster. It then loops the add and remove process described above until there's an iteration where nothing changes. At that point, that cluster/clique is said to be consistent and its nodes are removed from the original similarity graph.</p>
<p><img src="svgbob_851034e77512ba13719b9e5bf9b1f6235e4ef63b.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="RNA sequencing"></span> - A technique which uses next-generation sequencing to reveal the presence and quantity of RNA in a biological sample at some given moment.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="hierarchical cluster"></span> - A form of tiered clustering where clusters are represented as a tree. Each node represents a cluster (leaf nodes being a cluster of size 1), where the clusters represented by a parent node is the combination of the clusters represented by its children.</p>
<p><img src="svgbob_238058b9ed61fe867a16e2ee4c8854111e20582a.svg" alt="Kroki diagram output"></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>Hierarchical clustering has its roots in phylogeny. The similarity metric to build clusters is massaged into a distance metric, which is then used to form a tree that represents the clusters.</p>
</div>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="cosine similarity"></span> <span class="anki-answerpattern" data-pattern="cosine distance"></span> - A similarity metric that measures if two vectors grew/shrunk in a similar trajectories (similar angles).</p>
<p><img src="svgbob_b7fe91c303f6d60163c6d99da6f196ea8e1f0d4d.svg" alt="Kroki diagram output"></p>
<p>The metric computes the trajectory as the cosine of an angle between vectors. In the example above, T and U have different magnitudes than A and B, but the angle between T and U is exactly the same as the angle between A and B: 20deg. The cosine similarity of both pairs is cos(20deg) = 0.939. Had the angle been ...</p>
<ul>
<li>smaller (more similar trajectory), the cosine would get closer to 1.</li>
<li>larger (less similar trajectory), the cosine would get closer to -1.</li>
</ul>
<p>The formula may be modified to become a distance metric as follows: <code>1 - cosine_similarity(x, y)</code>. Whereas the cosine similarity varies between -1 and 1, the cosine distance varies between 0 (totally similar) and 2 (totally dissimilar).</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="dendrogram"></span> - A diagram of a tree. The term is most often used in the context of hierarchical clustering, where the tree that makes up the hierarchy of clusters is referred to as a dendrogram.</p>
<p><img src="svgbob_a475538fb2c51584ff27ec3a0225b8017824a0b9.svg" alt="Kroki diagram output"></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>It's tough to get a handle on what the requirements are, if any, to call a tree a dendrogram: Is it restricted to 2 children per internal node (or can there be more)? Do the edges extending from an internal node have to be of equal weight (e.g. equidistant)? Does the tree have to be ultrametric? Does it have to be a rooted tree (or can it be an unrooted tree)?</p>
<p>It seems like you can call any tree, even unrooted trees, a dendrogram. This seems like a gate keeping term. "Draw the tree that makes up the hierarchical cluster" vs "Draw the dendrogram that makes up the hierarchical cluster".</p>
</div>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(differential expression|differential gene expression)/i"></span> - Given a set of transcriptome snapshots, where each snapshot is for the same species but in a different state, differential gene expression analyzes transcript abundances across the transcriptomes to see ...</p>
<ol>
<li>which genes are responsible for / influenced by the state,</li>
<li>what the pattern of gene expression change is in those genes (transcript abundances).</li>
</ol>
<p>For example, differential expression analysis may be used to compare cancerous and non-cancerous blood cells to identify which genes are responsible for the cancer and their gene expression levels.</p>
<table>
<thead>
<tr>
<th></th>
<th>patient1 (cancer)</th>
<th>patient2 (cancer)</th>
<th>patient3 (non-cancer)</th>
<th>...</th>
</tr>
</thead>
<tbody>
<tr>
<td>Gene A</td>
<td>100</td>
<td>100</td>
<td>100</td>
<td>...</td>
</tr>
<tr>
<td>Gene B</td>
<td>100</td>
<td>110</td>
<td>50</td>
<td>...</td>
</tr>
<tr>
<td>Gene C</td>
<td>100</td>
<td>110</td>
<td>140</td>
<td>...</td>
</tr>
<tr>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
</tbody>
</table>
<p>In the example above, gene B has roughly double the expression when cancerous.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>Recall that genes are transcribed from DNA to mRNA, then translated to a protein. A transcript in a transcriptome is essentially a gene currently undergoing the process of gene expression.</p>
</div>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>I suspect the term transcript abundance is used instead of transcript count because oftentimes the counts are processed / normalized into some other form in an effort to denoise / de-bias (RNA sequencing is a noisy process).</p>
</div>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="Ohdo syndrome"></span> - A rare disease causing learning disabilities and distinct facial features. The disease is caused by a single nucleotide polymorphism resulting in a truncated protein (see codons).</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="single nucleotide polymorphism"></span> <span class="anki-hidepattern" data-pattern="(SNP)/i"></span> - A nucleotide variation at a specific location in a DNA sequence (e.g. position 15 has a SNP where it's A vs a SNP where it's T). While a single nucleotide polymorphism technically qualifies as a change in DNA, it occurs frequently enough that it's considered a variation rather than a mutation. Specifically, across the entire population, if the frequency of the change occurring is ...</p>
<ul>
<li>less than 1%, it's considered a point mutation.</li>
<li>at least 1%, it's considered a single nucleotide polymorphism.</li>
</ul>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(read mapping|read mapped|read map)/i"></span> - The alignment of DNA sequences (e.g. read_SEQs, contigs, etc..) to some larger DNA sequence (e.g. reference genome).</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(reference genome|reference human genome)/i"></span> - A genome assembled from multiple organisms of the same species, represented as the idealized genome for that species. Sequenced DNA fragment_SEQs / contigs of an organism are often read mapped against the reference genome for that organism's species, such that ...</p>
<ul>
<li>a clearer picture of the organism's genome is produced.</li>
<li>single nucleotide polymorphisms can be found.</li>
</ul>
<p>Reference genomes don't capture genomic nuances such as genome rearrangement, areas of high mutation, or single nucleotide polymorphisms. For example, roughly 0.1% of an individual human's genome can't be read mapped to the human reference genome (e.g. major histocompatibility complex).</p>
<p>A new type of reference genome, called a pan-genome, attempts to capture such nuances.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="pan-genome|pangenome)/i"></span> - A graph representing the relationships between a set of genomes. Pan-genomes are intended to be a new form of reference genome where nuances like genome rearrangements are retained.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="major histocompatibility complex"></span> - A region of DNA containing genes linked to the immune system. The genes in this region are highly diverse, to the point that it's unlikely for two individuals to have the genes in the exact same form.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(trie)/i"></span> - A rooted tree that holds a set of sequences. Shared prefixes between those sequences are collapsed into a single path while the non-shared remainders split out as deviating paths.</p>
<p><img src="svgbob_a84f0cb9a8ddb31c2df6974e512858573d8cde82.svg" alt="Kroki diagram output"></p>
<p>To disambiguate scenarios where one sequence is a prefix of the other, a trie typically either includes a ...</p>
<ul>
<li>a flag on each node to indicate if its the end of a sequence.</li>
<li>a special "end-of-sequence" at the end of each sequence.</li>
</ul>
<p><img src="svgbob_5a01644a6141c1fd7745e9eeb6f5cbb7102bee06.svg" alt="Kroki diagram output"></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>End of sequence marker is the preferred mechanism.</p>
</div>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(Aho[\s-]?Corasick|Aho[\s-]?Corasick trie)/i"></span> - A trie with special hop edges that eliminates redundant scanning during searches.</p>
<p>Given a trie containing sequence prefixes P1 and P2, a special hop edge (P1, P2) is added if P2 is equal to P1 but with its first element chopped off (<code>P2 = P1[1:]</code>). In the example below, a special hop edge connects "arat" to "rat".</p>
<p><img src="svgbob_d6e202f055e470240a3f70bee8cb98c573f8141b.svg" alt="Kroki diagram output"></p>
<p>If a scan walks the trie to "arat", the next scan <em>must</em> contain "rat". As such, a special edge connects "arat" to "rat" such that the next scan can start directly past "rat".</p>
<p><img src="svgbob_83da5124b5f21260416ff061e8154726d6f8d6e7.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="suffix trie"></span> - A trie of all suffixes within a sequence.</p>
<p><img src="svgbob_325556b783b02fe0931376c9510a2074bf716cc1.svg" alt="Kroki diagram output"></p>
<p>Suffix tries are used to efficiently determine if a string contains a substring. The string is converted to a suffix trie, then the trie is searched from the root node to see if a specific substring exists.</p>
<p><img src="svgbob_e98da221de14299799efa96f49772a00dff3993f.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="suffix tree"></span> - A suffix trie where paths of nodes with indegree and outdegree of 1 are combined into a single edge. The elements at the edges being combined are concatenated together.</p>
<p><img src="svgbob_4eea217255c04c65a6a80ce653c8cb82e2ec2aab.svg" alt="Kroki diagram output"></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>Implementations typically represent edge strings as pointers / string views back to the original string.</p>
</div>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="suffix array"></span> - A memory-efficient representation of a suffix tree as an array of pointers.</p>
<p>The suffixes of a sequence are sorted lexicographically, where each suffix includes the same end marker that's included in the suffix tree. The end marker comes first in the lexicographical sort order. The example below is the suffix array for the word banana.</p>
<table>
<thead>
<tr>
<th>Index</th>
<th>Pointer</th>
<th>Suffix</th>
</tr>
</thead>
<tbody>
<tr>
<td>0</td>
<td>6</td>
<td>¶</td>
</tr>
<tr>
<td>1</td>
<td>5</td>
<td>a¶</td>
</tr>
<tr>
<td>2</td>
<td>3</td>
<td>ana¶</td>
</tr>
<tr>
<td>3</td>
<td>1</td>
<td>anana¶</td>
</tr>
<tr>
<td>4</td>
<td>0</td>
<td>banana¶</td>
</tr>
<tr>
<td>5</td>
<td>4</td>
<td>na¶</td>
</tr>
<tr>
<td>6</td>
<td>2</td>
<td>nana¶</td>
</tr>
</tbody>
</table>
<p>The common prefix between two neighbouring suffixes represents a shared branch point in the suffix tree.</p>
<p><img src="svgbob_f4ea07d5e51e3b023280a458096fd65017e35d3b.svg" alt="Kroki diagram output"></p>
<p>Sliding a window of size two down the suffix array, the changes in common prefix from one pair of suffixes to the next defines the suffix tree structure. If a pair's common prefix ...</p>
<ul>
<li>has the same length vs the previous pair, it means the branch point is the same as the previous pair's branch point.</li>
<li>increases in length vs the previous pair, it means the branch point extends from the previous pair's branch point.</li>
<li>decreases in length vs the previous pair, it means the branch point reverts to that of the last pair with that length.</li>
</ul>
<p>In the example above, the common prefix length between index ...</p>
<ul>
<li>(0, 1) is 0, meaning it branches from the root node.</li>
<li>(1, 2) is 1 ("a"), meaning the branch point extends from the last branch point (root node) with an edge representing the "a".</li>
<li>(2, 3) is 3 ("ana"), meaning the branch point extends from the last branch point ("a") with an edge representing "na".</li>
<li>(3, 4) is 0, meaning it branches from the root node.</li>
<li>(4, 5) is 0, meaning it branches from the root node.</li>
<li>(5, 6) is 2 ("na"), meaning the branch point extends from the last branch point (root node) with an edge representing "na".</li>
</ul>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The entire point of the suffix array is that it's just an array of pointers to the suffix in the source sequence. Since the pointers are sorted (sorted by the suffixes they point to), you can quickly find if a substring exists just by doing a binary search on the suffix array (if a substring exists, it must be a prefix of one of the suffixes).</p>
</div>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(burrows[-\s]wheeler transform)/i"></span> <code>{ankiHied} (BWT)/</code> - A matrix formed by combining all cyclic rotations of a sequence and sorting lexicographically. The sequence must have an end marker, where the end marker comes first in the lexicographical sort order (similar to suffix arrays).</p>
<p>The example below is the burrows-wheeler transform of "banana¶", where <em>¶</em> is the end marker.</p>
<ol>
<li>
<p>Cyclic rotations.</p>
<table>
<thead>
<tr>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
</tr>
</thead>
<tbody>
<tr>
<td>b</td>
<td>a</td>
<td>n</td>
<td>a</td>
<td>n</td>
<td>a</td>
<td>¶</td>
</tr>
<tr>
<td>¶</td>
<td>b</td>
<td>a</td>
<td>n</td>
<td>a</td>
<td>n</td>
<td>a</td>
</tr>
<tr>
<td>a</td>
<td>¶</td>
<td>b</td>
<td>a</td>
<td>n</td>
<td>a</td>
<td>n</td>
</tr>
<tr>
<td>n</td>
<td>a</td>
<td>¶</td>
<td>b</td>
<td>a</td>
<td>n</td>
<td>a</td>
</tr>
<tr>
<td>a</td>
<td>n</td>
<td>a</td>
<td>¶</td>
<td>b</td>
<td>a</td>
<td>n</td>
</tr>
<tr>
<td>n</td>
<td>a</td>
<td>n</td>
<td>a</td>
<td>¶</td>
<td>b</td>
<td>a</td>
</tr>
<tr>
<td>a</td>
<td>n</td>
<td>a</td>
<td>n</td>
<td>a</td>
<td>¶</td>
<td>b</td>
</tr>
</tbody>
</table>
</li>
<li>
<p>Lexicographically sort the cyclic rotations.</p>
<table>
<thead>
<tr>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
</tr>
</thead>
<tbody>
<tr>
<td>¶</td>
<td>b</td>
<td>a</td>
<td>n</td>
<td>a</td>
<td>n</td>
<td>a</td>
</tr>
<tr>
<td>a</td>
<td>¶</td>
<td>b</td>
<td>a</td>
<td>n</td>
<td>a</td>
<td>n</td>
</tr>
<tr>
<td>a</td>
<td>n</td>
<td>a</td>
<td>¶</td>
<td>b</td>
<td>a</td>
<td>n</td>
</tr>
<tr>
<td>a</td>
<td>n</td>
<td>a</td>
<td>n</td>
<td>a</td>
<td>¶</td>
<td>b</td>
</tr>
<tr>
<td>b</td>
<td>a</td>
<td>n</td>
<td>a</td>
<td>n</td>
<td>a</td>
<td>¶</td>
</tr>
<tr>
<td>n</td>
<td>a</td>
<td>¶</td>
<td>b</td>
<td>a</td>
<td>n</td>
<td>a</td>
</tr>
<tr>
<td>n</td>
<td>a</td>
<td>n</td>
<td>a</td>
<td>¶</td>
<td>b</td>
<td>a</td>
</tr>
</tbody>
</table>
</li>
</ol>
<p>BWT matrices have a special property called the first-last property which makes them suitable for quickly determining if and how many times a substring exists in the original sequence.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(first[-\s]last property)/i"></span> - The property of BWT matrices that guarantees consistent ordering of a symbol's instances between the first and last columns of a BWT matrix.</p>
<p>Consider the sequence "banana¶": The symbols in "banana¶" are {¶, a, b, n}. At index ...</p>
<ol start="0">
<li>the first <em>b</em> occurs: b<sub>1</sub></li>
<li>the first <em>a</em> occurs: a<sub>1</sub></li>
<li>the first <em>n</em> occurs: n<sub>1</sub></li>
<li>the second <em>a</em> occurs: a<sub>2</sub></li>
<li>the second <em>n</em> occurs: n<sub>2</sub></li>
<li>the third <em>a</em> occurs: a<sub>3</sub></li>
<li>the first <em>¶</em> occurs: ¶<sub>1</sub></li>
</ol>
<p>With these occurrence counts, the sequence becomes b<sub>1</sub>a<sub>1</sub>n<sub>1</sub>a<sub>2</sub>n<sub>2</sub>a<sub>3</sub>¶<sub>1</sub>. In the BWT matrix, for each symbol, even though the position of symbol instances are different between the first and last columns, the order in which those instances appear in are the same. For example, ...</p>
<ul>
<li>symbol <em>a</em> instances are ordered as [<span style="background-color: #f00; color: white; padding: 3px">a<sub>3</sub></span>, <span style="background-color: #b00; color: white; padding: 3px">a<sub>2</sub></span>, <span style="background-color: #800; color: white; padding: 3px">a<sub>1</sub></span>] in both the first and last column.</li>
<li>symbol <em>n</em> instances are ordered as [<span style="background-color: #00f; color: white; padding: 3px">n<sub>2</sub></span>, <span style="background-color: #00b; color: white; padding: 3px">n<sub>1</sub></span>] in both the first and last column.</li>
</ul>
<table>
<thead>
<tr>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
</tr>
</thead>
<tbody>
<tr>
<td>¶<sub>1</sub></td>
<td>b<sub>1</sub></td>
<td>a<sub>1</sub></td>
<td>n<sub>1</sub></td>
<td>a<sub>2</sub></td>
<td>n<sub>2</sub></td>
<td><span style="background-color: #f00; color: white; padding: 3px">a<sub>3</sub></span></td>
</tr>
<tr>
<td><span style="background-color: #f00; color: white; padding: 3px">a<sub>3</sub></span></td>
<td>¶<sub>1</sub></td>
<td>b<sub>1</sub></td>
<td>a<sub>1</sub></td>
<td>n<sub>1</sub></td>
<td>a<sub>2</sub></td>
<td><span style="background-color: #00f; color: white; padding: 3px">n<sub>2</sub></span></td>
</tr>
<tr>
<td><span style="background-color: #b00; color: white; padding: 3px">a<sub>2</sub></span></td>
<td>n<sub>2</sub></td>
<td>a<sub>3</sub></td>
<td>¶<sub>1</sub></td>
<td>b<sub>1</sub></td>
<td>a<sub>1</sub></td>
<td><span style="background-color: #00b; color: white; padding: 3px">n<sub>1</sub></span></td>
</tr>
<tr>
<td><span style="background-color: #800; color: white; padding: 3px">a<sub>1</sub></span></td>
<td>n<sub>1</sub></td>
<td>a<sub>2</sub></td>
<td>n<sub>2</sub></td>
<td>a<sub>3</sub></td>
<td>¶<sub>1</sub></td>
<td>b<sub>1</sub></td>
</tr>
<tr>
<td>b<sub>1</sub></td>
<td>a<sub>1</sub></td>
<td>n<sub>1</sub></td>
<td>a<sub>2</sub></td>
<td>n<sub>2</sub></td>
<td>a<sub>3</sub></td>
<td>¶<sub>1</sub></td>
</tr>
<tr>
<td><span style="background-color: #00f; color: white; padding: 3px">n<sub>2</sub></span></td>
<td>a<sub>3</sub></td>
<td>¶<sub>1</sub></td>
<td>b<sub>1</sub></td>
<td>a<sub>1</sub></td>
<td>n<sub>1</sub></td>
<td><span style="background-color: #b00; color: white; padding: 3px">a<sub>2</sub></span></td>
</tr>
<tr>
<td><span style="background-color: #00b; color: white; padding: 3px">n<sub>1</sub></span></td>
<td>a<sub>2</sub></td>
<td>n<sub>2</sub></td>
<td>a<sub>3</sub></td>
<td>¶<sub>1</sub></td>
<td>b<sub>1</sub></td>
<td><span style="background-color: #800; color: white; padding: 3px">a<sub>1</sub></span></td>
</tr>
</tbody>
</table>
<p>The first-last property comes from lexicographic sorting. In the example matrix above, isolating the matrix to those rows starting with <em>a</em> shows that, the second column is also lexicographically sorted in the isolated matrix.</p>
<table>
<thead>
<tr>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
</tr>
</thead>
<tbody>
<tr>
<td><span style="background-color: #f00; color: white; padding: 3px">a<sub>3</sub></span></td>
<td>¶<sub>1</sub></td>
<td>b<sub>1</sub></td>
<td>a<sub>1</sub></td>
<td>n<sub>1</sub></td>
<td>a<sub>2</sub></td>
<td>n<sub>2</sub></td>
</tr>
<tr>
<td><span style="background-color: #b00; color: white; padding: 3px">a<sub>2</sub></span></td>
<td>n<sub>2</sub></td>
<td>a<sub>3</sub></td>
<td>¶<sub>1</sub></td>
<td>b<sub>1</sub></td>
<td>a<sub>1</sub></td>
<td>n<sub>1</sub></td>
</tr>
<tr>
<td><span style="background-color: #800; color: white; padding: 3px">a<sub>1</sub></span></td>
<td>n<sub>1</sub></td>
<td>a<sub>2</sub></td>
<td>n<sub>2</sub></td>
<td>a<sub>3</sub></td>
<td>¶<sub>1</sub></td>
<td>b<sub>1</sub></td>
</tr>
</tbody>
</table>
<p>In other words, cyclically rotating each row right by 1 moves each corresponding <em>a</em> to the end but doesn't change the lexicographic ordering of the rows.</p>
<table>
<thead>
<tr>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
</tr>
</thead>
<tbody>
<tr>
<td>¶<sub>1</sub></td>
<td>b<sub>1</sub></td>
<td>a<sub>1</sub></td>
<td>n<sub>1</sub></td>
<td>a<sub>2</sub></td>
<td>n<sub>2</sub></td>
<td><span style="background-color: #f00; color: white; padding: 3px">a<sub>3</sub></span></td>
</tr>
<tr>
<td>n<sub>2</sub></td>
<td>a<sub>3</sub></td>
<td>¶<sub>1</sub></td>
<td>b<sub>1</sub></td>
<td>a<sub>1</sub></td>
<td>n<sub>1</sub></td>
<td><span style="background-color: #b00; color: white; padding: 3px">a<sub>2</sub></span></td>
</tr>
<tr>
<td>n<sub>1</sub></td>
<td>a<sub>2</sub></td>
<td>n<sub>2</sub></td>
<td>a<sub>3</sub></td>
<td>¶<sub>1</sub></td>
<td>b<sub>1</sub></td>
<td><span style="background-color: #800; color: white; padding: 3px">a<sub>1</sub></span></td>
</tr>
</tbody>
</table>
<p>Once rotated, the rows in the isolated matrix become other rows from the original matrix. Since the rows in the isolated matrix are still lexicographically sorted, they're ordered as they appear in that original matrix.</p>
<table>
<thead>
<tr>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
</tr>
</thead>
<tbody>
<tr>
<td>¶<sub>1</sub></td>
<td>b<sub>1</sub></td>
<td>a<sub>1</sub></td>
<td>n<sub>1</sub></td>
<td>a<sub>2</sub></td>
<td>n<sub>2</sub></td>
<td><span style="background-color: #f00; color: white; padding: 3px">a<sub>3</sub></span></td>
</tr>
<tr>
<td>a<sub>3</sub></td>
<td>¶<sub>1</sub></td>
<td>b<sub>1</sub></td>
<td>a<sub>1</sub></td>
<td>n<sub>1</sub></td>
<td>a<sub>2</sub></td>
<td>n<sub>2</sub></td>
</tr>
<tr>
<td>a<sub>2</sub></td>
<td>n<sub>2</sub></td>
<td>a<sub>3</sub></td>
<td>¶<sub>1</sub></td>
<td>b<sub>1</sub></td>
<td>a<sub>1</sub></td>
<td>n<sub>1</sub></td>
</tr>
<tr>
<td>a<sub>1</sub></td>
<td>n<sub>1</sub></td>
<td>a<sub>2</sub></td>
<td>n<sub>2</sub></td>
<td>a<sub>3</sub></td>
<td>¶<sub>1</sub></td>
<td>b<sub>1</sub></td>
</tr>
<tr>
<td>b<sub>1</sub></td>
<td>a<sub>1</sub></td>
<td>n<sub>1</sub></td>
<td>a<sub>2</sub></td>
<td>n<sub>2</sub></td>
<td>a<sub>3</sub></td>
<td>¶<sub>1</sub></td>
</tr>
<tr>
<td>n<sub>2</sub></td>
<td>a<sub>3</sub></td>
<td>¶<sub>1</sub></td>
<td>b<sub>1</sub></td>
<td>a<sub>1</sub></td>
<td>n<sub>1</sub></td>
<td><span style="background-color: #b00; color: white; padding: 3px">a<sub>2</sub></span></td>
</tr>
<tr>
<td>n<sub>1</sub></td>
<td>a<sub>2</sub></td>
<td>n<sub>2</sub></td>
<td>a<sub>3</sub></td>
<td>¶<sub>1</sub></td>
<td>b<sub>1</sub></td>
<td><span style="background-color: #800; color: white; padding: 3px">a<sub>1</sub></span></td>
</tr>
</tbody>
</table>
<p>Given just the first and last column of a BWT matrix, the original sequence can be pulled out by walking between those columns from last-to-first. Since it's known that ...</p>
<ul>
<li>the row containing the end marker (<em>¶</em>) within the first column has the sequence's last element within the last column,</li>
<li>that end marker (<em>¶</em>) only appears once in the sequence and is always lexicographically sorted to the top of the first column,</li>
</ul>
<p>... the walk always starts from the top row.</p>
<p><img src="svgbob_57e0f1bae5a018cfdeaca8f0bd21352d8d510b91.svg" alt="Kroki diagram output"></p>
<p>Likewise, given just the first and last column of a BWT matrix, it's possible to quickly identify if and how many instances of some substring exists in the original sequence.</p>
<p><img src="svgbob_6daab13e146d8f4524d3b6d51eeebfa62689062a.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(pre[-\s]?order traversal)/i"></span> <span class="anki-answerpattern" data-pattern="(post[-\s]?order traversal)/i"></span> <span class="anki-answerpattern" data-pattern="(in[-\s]?order traversal)/i"></span></p>
<ul>
<li>
<p>pre-order traversal - A form of depth-first traversal for binary trees where, starting from the root node, ...  <span class="anki-listrandomorder"></span>
</p>
<ol>
<li>visit current node.</li>
<li>recursively visit current node's left subtree.</li>
<li>recursively visit current node's right subtree.</li>
</ol>
<p><img src="svgbob_d8441d729e89e94e305e729ce7089b194c8e2e1a.svg" alt="Kroki diagram output"></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>Pre-order traversal is sometimes referred to as NLR (node-left-right).</p>
<p>For reverse pre-order traversal, swap steps 2 and 3: NRL (node-right-left).</p>
</div>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>This is a form of topological order traversal because the parent node is traversed before its children.</p>
</div>
<p>The term pre-order traversal also applies to non-binary trees (variable number of children per node): If the children have a specific order, pre-order traversal recursively visits each from first (left-most) to last (right-most).</p>
</li>
<li>
<p>post-order traversal  - A form of depth-first traversal for binary trees where, starting from the root node, ...</p>
<ol>
<li>recursively visit current node's left subtree.</li>
<li>recursively visit current node's right subtree.</li>
<li>visit current node.</li>
</ol>
<p><img src="svgbob_6d34eb672d5564c2b4ce0a9bb152ff226d2be918.svg" alt="Kroki diagram output"></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>Post-order traversal is sometimes referred to as LRN (left-right-node).</p>
<p>For reverse post-order traversal, swap steps 1 and 2: RLN (right-left-node).</p>
</div>
<p>The term post-order traversal also applies to non-binary trees (variable number of children per node): If the children have a specific order, post-order traversal recursively visits each from last (right-most) to first (left-most).</p>
</li>
<li>
<p>in-order traversal - A form of depth-first traversal for binary trees where, starting from the root node, ...</p>
<ol>
<li>recursively visit current node's left subtree.</li>
<li>visit current node.</li>
<li>recursively visit current node's right subtree.</li>
</ol>
<p><img src="svgbob_7914992283ad27b17cf8643939fe25a5fb79f78a.svg" alt="Kroki diagram output"></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>In-order is sometimes referred to as LNR (left-node-right).</p>
<p>For reverse in-order traversal, swap steps 1 and 3: RNL (right-node-left).</p>
</div>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>It's unclear if there's an analog for this for non-binary trees (variable number of children per node). Maybe if the children have a specific order, it recursively visits the first half (left children), then visits the parent node, then recursively visits the last half (right children). But, how would this work if there were an odd number of children? The middle child wouldn't be in the left-half or right-half.</p>
</div>
</li>
</ul>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="Basic Local Alignment Search Tool"></span> <span class="anki-answerpattern" data-pattern="(high[-\s]scoring segment pair)/i"></span> <span class="anki-hidepattern" data-pattern="(BLAST)/"></span> <span class="anki-hidepattern" data-pattern="(HSP)/"></span> - A heuristic algorithm that quickly finds shared regions between a query sequence and a database of sequences, where those shared regions are called high-scoring segment pairs. High-scoring segment pairs may be identified even in the presence of mutations, potentially even if mutated to the point where all elements are different in the shared region (e.g. BLOSUM scoring may deem two peptides to be highly related but they may not actually share any amino acids between them).</p>
<p><img src="svgbob_bb1cb211935586eee5a511739d14c699a85f2179.svg" alt="Kroki diagram output"></p>
<p>BLAST works by preprocessing the database of sequences into a hash table of k-mers, where other k-mers similar to those k-mers are included in the hash table as well. Similarity is determined by performing sequence alignments (the higher the score, the more similar the k-mer is).</p>
<p><img src="svgbob_0a9e1f0fc7f14b37640e40326681e0e3da12652b.svg" alt="Kroki diagram output"></p>
<p>K-mers from a query sequence are then looked up one-by-one in the hash table. Found matches are extended left and right until some criteria is met (e.g. the score drops below some threshold). The final product of the extension is called a high-scoring segment pair.</p>
<p><img src="svgbob_d1064c02dd24172be71e0d3855b32de7deef66b7.svg" alt="Kroki diagram output"></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The Pevzner book and documentation online refers to k-mers from the query sequence as seeds and the extension left-and-right as seed extension.</p>
</div>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="seed extension"></span> <span class="anki-answerpattern" data-pattern="seed"></span> - A substring of a string which is specifically used for mismatch tolerant searches.</p>
<p>The example below searches for GCCGTTTT with a mismatch tolerance of 1 by first breaking GCCGTTTT into two non-overlapping seeds (GCCG and TTTT), then searching for each seed independently. Since GCCGTTTT can only contain a single mismatch, that mismatch has to be either in the 1st seed (GCCG) or the 2nd seed (TTTT), not both.</p>
<p><img src="svgbob_c62030de2ac380d4b7ac8eaa697a6fb1805ca807.svg" alt="Kroki diagram output"></p>
<p>Each found seed is then extended to cover the entirety of GCCGTTT and tested in full, called seed extension. If the hamming distance of the extended seed is within the mismatch tolerance of 1, it's considered a match.</p>
<p><img src="svgbob_9f10c1db9ba47e4e26e56bfedd44f424bc226362.svg" alt="Kroki diagram output"></p>
<p>It's impossible for d mismatches to exist across d + 1 seeds. There are more seeds than there are mismatches — at least one of the seeds must match exactly.</p>
<p><img src="svgbob_27be0ab62d6a6964135cfc8727e10d7ede8f2e5f.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="retrovirus"></span> - A virus that inserts a DNA copy of its RNA genome into the DNA of the host cell that it invades.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(anti[\-]?viral|anti[\-]?retroviral)/i"></span> - A class of medication used for treating viral infections.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The term antiretroviral therapy is commonly used to refer to HIV treatments, although retroviruses other than HIV exist (e.g. human T-lymphotropic virus)</p>
</div>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="surface protein"></span> - Protein embedded into a cell surface or viral envelope. See glycoprotein / glycan.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(viral envelope|viral envelope protein|envelope protein)/i"></span> - One of the proteins making up the outermost layer of a virus, called the viral envelope.</p>
<p>A viral envelope often has one or more spikes which facilitate the entry of the virus's genetic material into the host cell.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="glycoprotein"></span> - A protein containing glycans.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="glycan"></span> - A carbohydrate portion of some glycoconjugate (e.g. glycoprotein or glycolipid). Cells have a dense coating of glycans on their surface, which are used for modulating interactions with other cells and biological entities (e.g. communication between the cells of a human, interactions between bacterial cells and human cells, interactions between a human cells and viruses, etc..).</p>
<p>Glycans may also coat viral envelope proteins, which can make those viruses invisible to the human immune system (e.g. HIV).</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="glycosylation"></span> - A modification to a protein, applied after it's already been translated out of the ribosome, that turns it into a glycoprotein.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(Red Queen effect|Red Queen hypothesis)/i"></span> - The hypothesis that organisms must constantly evolve in order to survive due to predator-prey dynamics within an environment. The name comes from Lewis Carroll's novel Through the Look-Glass, where the Red Queen tells Alice "Now, here, you see, it takes all the running you can do, to keep in the same place."</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(syncytium|symplasm)/i"></span> - A cytoplasmic mass containing several nuclei formed by the fusion of multiple cells. Certain HIV phenotypes embed their viral envelope proteins into the host cell's surface upon infection, which ends up causing neighbouring cells to fuse into a non-functional syncytium.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(odds[\-\s]ratio)/i"></span> - A measure of the chance of success, defined as the probability of some event occurring divided by the probability that event doesn't occur. For example, given a based coin, a fair coin, and a sequence of flips, the probability the sequence of flips was generated by the ...</p>
<ul>
<li>the fair coin is stated as <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>P</mi><msub><mi>r</mi><mi>F</mi></msub><mo stretchy="false">(</mo><mi>f</mi><mi>l</mi><mi>i</mi><mi>p</mi><mi>s</mi><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">Pr_F(flips)</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal" style="margin-right:0.13889em;">P</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">r</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.32833099999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight" style="margin-right:0.13889em;">F</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.10764em;">f</span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord mathnormal">i</span><span class="mord mathnormal">p</span><span class="mord mathnormal">s</span><span class="mclose">)</span></span></span></span>.</li>
<li>the biased coin is stated as <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>P</mi><msub><mi>r</mi><mi>B</mi></msub><mo stretchy="false">(</mo><mi>f</mi><mi>l</mi><mi>i</mi><mi>p</mi><mi>s</mi><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">Pr_B(flips)</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal" style="margin-right:0.13889em;">P</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">r</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.32833099999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight" style="margin-right:0.05017em;">B</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.10764em;">f</span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord mathnormal">i</span><span class="mord mathnormal">p</span><span class="mord mathnormal">s</span><span class="mclose">)</span></span></span></span>.</li>
</ul>
<p>The odds ratio that the flips were generated by the fair coin is <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mfrac><mrow><mi>P</mi><msub><mi>r</mi><mi>F</mi></msub><mo stretchy="false">(</mo><mi>f</mi><mi>l</mi><mi>i</mi><mi>p</mi><mi>s</mi><mo stretchy="false">)</mo></mrow><mrow><mi>P</mi><msub><mi>r</mi><mi>B</mi></msub><mo stretchy="false">(</mo><mi>f</mi><mi>l</mi><mi>i</mi><mi>p</mi><mi>s</mi><mo stretchy="false">)</mo></mrow></mfrac></mrow><annotation encoding="application/x-tex">\frac{Pr_F(flips)}{Pr_B(flips)}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.53em;vertical-align:-0.52em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.01em;"><span style="top:-2.655em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.13889em;">P</span><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.02778em;">r</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3448em;"><span style="top:-2.3567071428571427em;margin-left:-0.02778em;margin-right:0.07142857142857144em;"><span class="pstrut" style="height:2.5em;"></span><span class="sizing reset-size3 size1 mtight"><span class="mord mathnormal mtight" style="margin-right:0.05017em;">B</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.14329285714285717em;"><span></span></span></span></span></span></span><span class="mopen mtight">(</span><span class="mord mathnormal mtight" style="margin-right:0.10764em;">f</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mathnormal mtight">i</span><span class="mord mathnormal mtight">p</span><span class="mord mathnormal mtight">s</span><span class="mclose mtight">)</span></span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.485em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.13889em;">P</span><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.02778em;">r</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3448em;"><span style="top:-2.3567071428571427em;margin-left:-0.02778em;margin-right:0.07142857142857144em;"><span class="pstrut" style="height:2.5em;"></span><span class="sizing reset-size3 size1 mtight"><span class="mord mathnormal mtight" style="margin-right:0.13889em;">F</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.14329285714285717em;"><span></span></span></span></span></span></span><span class="mopen mtight">(</span><span class="mord mathnormal mtight" style="margin-right:0.10764em;">f</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mathnormal mtight">i</span><span class="mord mathnormal mtight">p</span><span class="mord mathnormal mtight">s</span><span class="mclose mtight">)</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.52em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span>. Likewise, the odds ratio that the flips were generated by the biased coin is <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mfrac><mrow><mi>P</mi><msub><mi>r</mi><mi>B</mi></msub><mo stretchy="false">(</mo><mi>f</mi><mi>l</mi><mi>i</mi><mi>p</mi><mi>s</mi><mo stretchy="false">)</mo></mrow><mrow><mi>P</mi><msub><mi>r</mi><mi>F</mi></msub><mo stretchy="false">(</mo><mi>f</mi><mi>l</mi><mi>i</mi><mi>p</mi><mi>s</mi><mo stretchy="false">)</mo></mrow></mfrac></mrow><annotation encoding="application/x-tex">\frac{Pr_B(flips)}{Pr_F(flips)}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.53em;vertical-align:-0.52em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.01em;"><span style="top:-2.655em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.13889em;">P</span><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.02778em;">r</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3448em;"><span style="top:-2.3567071428571427em;margin-left:-0.02778em;margin-right:0.07142857142857144em;"><span class="pstrut" style="height:2.5em;"></span><span class="sizing reset-size3 size1 mtight"><span class="mord mathnormal mtight" style="margin-right:0.13889em;">F</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.14329285714285717em;"><span></span></span></span></span></span></span><span class="mopen mtight">(</span><span class="mord mathnormal mtight" style="margin-right:0.10764em;">f</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mathnormal mtight">i</span><span class="mord mathnormal mtight">p</span><span class="mord mathnormal mtight">s</span><span class="mclose mtight">)</span></span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.485em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.13889em;">P</span><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.02778em;">r</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3448em;"><span style="top:-2.3567071428571427em;margin-left:-0.02778em;margin-right:0.07142857142857144em;"><span class="pstrut" style="height:2.5em;"></span><span class="sizing reset-size3 size1 mtight"><span class="mord mathnormal mtight" style="margin-right:0.05017em;">B</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.14329285714285717em;"><span></span></span></span></span></span></span><span class="mopen mtight">(</span><span class="mord mathnormal mtight" style="margin-right:0.10764em;">f</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mathnormal mtight">i</span><span class="mord mathnormal mtight">p</span><span class="mord mathnormal mtight">s</span><span class="mclose mtight">)</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.52em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span>.</p>
<p>The result of an odds ratio is how much more likely the top of the fraction is vs the bottom. For example, if the odds ratio for <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mfrac><mrow><mi>P</mi><msub><mi>r</mi><mi>F</mi></msub><mo stretchy="false">(</mo><mi>f</mi><mi>l</mi><mi>i</mi><mi>p</mi><mi>s</mi><mo stretchy="false">)</mo></mrow><mrow><mi>P</mi><msub><mi>r</mi><mi>B</mi></msub><mo stretchy="false">(</mo><mi>f</mi><mi>l</mi><mi>i</mi><mi>p</mi><mi>s</mi><mo stretchy="false">)</mo></mrow></mfrac></mrow><annotation encoding="application/x-tex">\frac{Pr_F(flips)}{Pr_B(flips)}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.53em;vertical-align:-0.52em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.01em;"><span style="top:-2.655em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.13889em;">P</span><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.02778em;">r</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3448em;"><span style="top:-2.3567071428571427em;margin-left:-0.02778em;margin-right:0.07142857142857144em;"><span class="pstrut" style="height:2.5em;"></span><span class="sizing reset-size3 size1 mtight"><span class="mord mathnormal mtight" style="margin-right:0.05017em;">B</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.14329285714285717em;"><span></span></span></span></span></span></span><span class="mopen mtight">(</span><span class="mord mathnormal mtight" style="margin-right:0.10764em;">f</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mathnormal mtight">i</span><span class="mord mathnormal mtight">p</span><span class="mord mathnormal mtight">s</span><span class="mclose mtight">)</span></span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.485em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.13889em;">P</span><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.02778em;">r</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3448em;"><span style="top:-2.3567071428571427em;margin-left:-0.02778em;margin-right:0.07142857142857144em;"><span class="pstrut" style="height:2.5em;"></span><span class="sizing reset-size3 size1 mtight"><span class="mord mathnormal mtight" style="margin-right:0.13889em;">F</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.14329285714285717em;"><span></span></span></span></span></span></span><span class="mopen mtight">(</span><span class="mord mathnormal mtight" style="margin-right:0.10764em;">f</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mathnormal mtight">i</span><span class="mord mathnormal mtight">p</span><span class="mord mathnormal mtight">s</span><span class="mclose mtight">)</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.52em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span> results in 2, it means that it's two times more likely for the fair coin to have been used vs the biased coin.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(log[-\s]?odds[-\s]?ratio)/i"></span> - The logarithm of the odds ratio: <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>l</mi><mi>o</mi><msub><mi>g</mi><mn>2</mn></msub><mo stretchy="false">(</mo><mfrac><mrow><mi>P</mi><msub><mi>r</mi><mi>B</mi></msub><mo stretchy="false">(</mo><mi>f</mi><mi>l</mi><mi>i</mi><mi>p</mi><mi>s</mi><mo stretchy="false">)</mo></mrow><mrow><mi>P</mi><msub><mi>r</mi><mi>F</mi></msub><mo stretchy="false">(</mo><mi>f</mi><mi>l</mi><mi>i</mi><mi>p</mi><mi>s</mi><mo stretchy="false">)</mo></mrow></mfrac><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">log_2(\frac{Pr_B(flips)}{Pr_F(flips)})</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.53em;vertical-align:-0.52em;"></span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord mathnormal">o</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.03588em;">g</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.03588em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mopen">(</span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.01em;"><span style="top:-2.655em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.13889em;">P</span><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.02778em;">r</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3448em;"><span style="top:-2.3567071428571427em;margin-left:-0.02778em;margin-right:0.07142857142857144em;"><span class="pstrut" style="height:2.5em;"></span><span class="sizing reset-size3 size1 mtight"><span class="mord mathnormal mtight" style="margin-right:0.13889em;">F</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.14329285714285717em;"><span></span></span></span></span></span></span><span class="mopen mtight">(</span><span class="mord mathnormal mtight" style="margin-right:0.10764em;">f</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mathnormal mtight">i</span><span class="mord mathnormal mtight">p</span><span class="mord mathnormal mtight">s</span><span class="mclose mtight">)</span></span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.485em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.13889em;">P</span><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.02778em;">r</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3448em;"><span style="top:-2.3567071428571427em;margin-left:-0.02778em;margin-right:0.07142857142857144em;"><span class="pstrut" style="height:2.5em;"></span><span class="sizing reset-size3 size1 mtight"><span class="mord mathnormal mtight" style="margin-right:0.05017em;">B</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.14329285714285717em;"><span></span></span></span></span></span></span><span class="mopen mtight">(</span><span class="mord mathnormal mtight" style="margin-right:0.10764em;">f</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mathnormal mtight">i</span><span class="mord mathnormal mtight">p</span><span class="mord mathnormal mtight">s</span><span class="mclose mtight">)</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.52em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mclose">)</span></span></span></span>. Log-odds ratio is just another representation of odds ratio, typically used in cases when odds ratio generates a very small / large result.</p>
<table>
<thead>
<tr>
<th>odds ratio</th>
<th>log-odds ratio</th>
</tr>
</thead>
<tbody>
<tr>
<td>0.015625 (1/64)</td>
<td>-6</td>
</tr>
<tr>
<td>0.03125  (1/32)</td>
<td>-5</td>
</tr>
<tr>
<td>0.0625   (1/16)</td>
<td>-4</td>
</tr>
<tr>
<td>0.125    (1/8)</td>
<td>-3</td>
</tr>
<tr>
<td>0.25     (1/4)</td>
<td>-2</td>
</tr>
<tr>
<td>0.5      (1/2)</td>
<td>-1</td>
</tr>
<tr>
<td>1</td>
<td>0</td>
</tr>
<tr>
<td>2</td>
<td>1</td>
</tr>
<tr>
<td>4</td>
<td>2</td>
</tr>
<tr>
<td>8</td>
<td>3</td>
</tr>
<tr>
<td>16</td>
<td>4</td>
</tr>
<tr>
<td>32</td>
<td>5</td>
</tr>
<tr>
<td>64</td>
<td>6</td>
</tr>
</tbody>
</table>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="methylation"></span> - The addition of a methyl group (CH3) to a cytosine or guanine nucleotide.</p>
<p>DNA methylation is an important part of cell development. Specifically, when a stem cell converts into a specialized cell (cell differentiation), DNA methylation is an important factor in the change:</p>
<ul>
<li>DNA methylation often alters the expression of nearby genes.</li>
<li>When a gene's upstream regions become highly methylated, its expression is suppressed.</li>
</ul>
<p>DNA methylation is typically permanent (specialized cell cannot convert back to stem cell) and inherited during cell division, except in the case of zygote formation. Also, various cancers have been linked to both DNA hypermethylation and DNA hypomethylation.</p>
<p>When cytosine goes through DNA methylation, it has a tendency to deaminate to thymine. However, DNA methylation is often suppressed in areas of DNA dubbed CG-islands, where CG appears more frequently than the rest of the genome.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(C[p\-]?G island|C[p\-]?G site)/i"></span> - Regions of DNA with a high frequency of cytosine followed by guanine. The reverse complementing strand will have equal regions with equally high frequencies of guanine followed by cytosine.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="Hidden Markov Model"></span> - A model of a machine that outputs a sequence.</p>
<p><img src="svgbob_1b85dba6b2e5054c30423fb3fdfe17c1abf1ea80.svg" alt="Kroki diagram output"></p>
<p>The machine being modeled can be in one of many hidden states (called hidden because that state is unobservable). For example, the machine above could be in one of two hidden states: Gene or Non-gene. If in the ...</p>
<ul>
<li>gene hidden state, it's outputting DNA for a region of DNA that's a gene.</li>
<li>non-gene hidden state, it's outputting DNA for a region of DNA that isn't a gene (e.g. telomeres).</li>
</ul>
<p>At each step, the machine transitions from its existing hidden state to another hidden state and emits a symbol (transitions to the same hidden state are possible). For the example machine above, the emitted symbols are nucleotides (A, C, T, and G).</p>
<p>An HMM models such a machine by using four pieces of information:</p>
<ol>
<li>
<p>Set of hidden states the machine can be in: <code>{gene, non-gene}</code></p>
</li>
<li>
<p>Set of symbols that the machine can emit: <code>{A, C, T, G}</code></p>
</li>
<li>
<p>Set of hidden state-to-hidden state transition probabilities:</p>
<pre class="hljs"><code>{
  [<span class="hljs-string">gene</span>, <span class="hljs-string">gene</span>]<span class="hljs-string">:</span> <span class="hljs-number">0.9</span>,
  [<span class="hljs-string">gene</span>, <span class="hljs-string">non-gene</span>]<span class="hljs-string">:</span> <span class="hljs-number">0.1</span>,
  [<span class="hljs-string">non-gene</span>, <span class="hljs-string">gene</span>]<span class="hljs-string">:</span> <span class="hljs-number">0.1</span>,
  [<span class="hljs-string">non-gene</span>, <span class="hljs-string">non-gene</span>]<span class="hljs-string">:</span> <span class="hljs-number">0.9</span>
}
</code></pre>
</li>
<li>
<p>set of hidden state-to-symbol emission probabilities.</p>
<pre class="hljs"><code>{
  <span class="hljs-attr">gene:</span> {<span class="hljs-attr">A:</span> <span class="hljs-number">0.2</span>, <span class="hljs-attr">B:</span> <span class="hljs-number">0.3</span>, <span class="hljs-attr">C:</span> <span class="hljs-number">0.1</span>, <span class="hljs-attr">D:</span> <span class="hljs-number">0.4</span>},
  <span class="hljs-attr">non-gene:</span> {<span class="hljs-attr">A:</span> <span class="hljs-number">0.3</span>, <span class="hljs-attr">B:</span> <span class="hljs-number">0.2</span>, <span class="hljs-attr">C:</span> <span class="hljs-number">0.2</span>, <span class="hljs-attr">D:</span> <span class="hljs-number">0.3</span>}
}
</code></pre>
</li>
</ol>
<p>HMMs are often represented using HMM diagrams.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The probabilities above are totally made up. The example machine above is a bad example to model as an HMM. Only 2 hidden states and emitting a single nucleotide will result in a useless HMM model. The machine should be modeled as emitting 5-mers or something else and would likely need more than 2 hidden states?</p>
</div>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="Hidden Markov Model diagram"></span> - A visualization of an HMM as a directed graph.</p>
<ul>
<li>Solid nodes represent hidden states.</li>
<li>Solid edges represent hidden state transitions.</li>
<li>Dashed nodes represent symbols.</li>
<li>Dashed edges represent symbol emissions.</li>
</ul>
<p>Edges are labeled with the probability of the hidden state transition / symbol emission occurring.</p>
<p><img src="svgbob_ed178e9723e37b38c4fa700ec23c44ae8fd2e594.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="hidden state"></span> - A state within an HMM. At any given time, a HMM will be in one of n different hidden states. Unless a hidden state is a non-emitting hidden state, ...</p>
<ul>
<li>a HMM transitioning to that hidden state will result in a symbol emission (e.g. [foul, miss, hit] in the HMM diagram below).</li>
<li>the probability of symbol emission depends on the hidden state (e.g. emitting foul from hitter bat has 0.15 probability vs fouler bat has 0.4 probability, in the HMM diagram below).</li>
</ul>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>I think the word "hidden" is used because the machine that an HMM models typically has unobservable state (as in you can't observe its state, hence the word hidden).</p>
</div>
<p>In the HMM diagram below, the hidden states are [SOURCE, hitter bat, quitter bat].</p>
<p><img src="svgbob_ed178e9723e37b38c4fa700ec23c44ae8fd2e594.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="emitting hidden state"></span> - A hidden state that emits a symbol. An HMM typically emits a symbol after transitioning between hidden states. However, if the hidden state being transitioned to is a non-emitting hidden state, it doesn't emit a symbol.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(non-emitting hidden state|silent hidden state)/i"></span> - A hidden state that doesn't emit symbols. An HMM typically emits a symbol after transitioning between hidden states. However, if the hidden state being transitioned to is a non-emitting hidden state, it doesn't emit a symbol.</p>
<p>An HMM ...</p>
<ul>
<li>typically has a non-emitting hidden state that represents the start state of the machine.</li>
<li>may have a non-emitting hidden state that represents the termination state of the machine.</li>
<li>may have several other non-emitting hidden states (not related to machine start or termination).</li>
</ul>
<p>The HMM diagram below has the non-emitting hidden state SOURCE, which represents the machine's start state.</p>
<p><img src="svgbob_ed178e9723e37b38c4fa700ec23c44ae8fd2e594.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="hidden path"></span> - A sequence of hidden state transitions that a HMM passes through. For example, in the HMM diagram below, one possible hidden path could be as follows:</p>
<ol>
<li>SOURCE → fouler bat</li>
<li>fouler bat → hitter bat</li>
<li>hitter bat → hitter bat</li>
<li>hitter bat → hitter bat</li>
<li>hitter bat → hitter bat</li>
<li>hitter bat → hitter bat</li>
</ol>
<p><img src="svgbob_ed178e9723e37b38c4fa700ec23c44ae8fd2e594.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(symbol emission|emitted symbol)/i"></span> - A symbol emitted after a hidden state transition. The HMM diagram below can emit the symbols [hit, miss, foul].</p>
<p><img src="svgbob_ed178e9723e37b38c4fa700ec23c44ae8fd2e594.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(symbol emission sequence|emission sequence|emitted sequence)/i"></span> - A sequence of symbol emissions, where those symbols are emitted from an HMM. The HMM diagram below can produce the emitted sequence ...</p>
<ul>
<li>[hit, hit hit]</li>
<li>[hit, hit hit, hit]</li>
<li>[hit, hit hit, hit, hit]</li>
<li>[hit, hit hit, hit, foul]</li>
<li>[hit, hit hit, hit, miss]</li>
<li>[hit, hit hit, foul, hit]</li>
<li>...</li>
</ul>
<p><img src="svgbob_ed178e9723e37b38c4fa700ec23c44ae8fd2e594.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="(Viterbi algorithm|Viterbi graph|Viterbi)/i"></span> - An algorithm that determines the most probable hidden path in an HMM for some emitted sequence.</p>
<p>The algorithm begins by transforming an HMM to an exploded HMM.</p>
<p><img src="svgbob_33b91cac3246b7b9a9f43d0dd77f926b78067699.svg" alt="Kroki diagram output"></p>
<p>Each edge in the exploded HMM represents a hidden state transition (e.g. fouler bat → hitter bat) followed by a symbol emission (e.g. hit emitted after reaching hitter bat). The algorithm sets each exploded HMM edge's weight to that probability of that edge's transition-emission occurring: Pr(symbol|transition) = Pr(transition) * Pr(symbol). For example, Pr(fouler bat → hitter bat) is 0.1 in the HMM diagram above, and Pr(hit) once entered into the hitter bat hidden state is 0.75, so Pr(hit|fouler bat → hitter bat) = 0.1 * 0.75 = 0.075.</p>
<table>
<thead>
<tr>
<th></th>
<th>Pr(hit)</th>
<th>Pr(miss)</th>
<th>Pr(foul)</th>
<th>NON-EMITTABLE</th>
</tr>
</thead>
<tbody>
<tr>
<td>Pr(hitter bat → hitter bat)</td>
<td>0.9 * 0.75 = 0.675</td>
<td>0.9 * 0.1 = 0.09</td>
<td>0.9 * 0.15 = 0.135</td>
<td></td>
</tr>
<tr>
<td>Pr(hitter bat → fouler bat)</td>
<td>0.1 * 0.5  = 0.05</td>
<td>0.1 * 0.1 = 0.01</td>
<td>0.1 * 0.4  = 0.04</td>
<td></td>
</tr>
<tr>
<td>Pr(fouler bat → hitter bat)</td>
<td>0.1 * 0.75 = 0.075</td>
<td>0.1 * 0.1 = 0.01</td>
<td>0.1 * 0.15 = 0.015</td>
<td></td>
</tr>
<tr>
<td>Pr(fouler bat → fouler bat)</td>
<td>0.9 * 0.5  = 0.45</td>
<td>0.9 * 0.1 = 0.09</td>
<td>0.9 * 0.4  = 0.36</td>
<td></td>
</tr>
<tr>
<td>Pr(SOURCE     → hitter bat)</td>
<td>0.5 * 0.75 = 0.375</td>
<td>0.5 * 0.1 = 0.05</td>
<td>0.5 * 0.15 = 0.075</td>
<td></td>
</tr>
<tr>
<td>Pr(SOURCE     → fouler bat)</td>
<td>0.5 * 0.5  = 0.25</td>
<td>0.5 * 0.1 = 0.05</td>
<td>0.5 * 0.4  = 0.2</td>
<td></td>
</tr>
<tr>
<td>Pr(hitter bat → SINK)</td>
<td></td>
<td></td>
<td></td>
<td>1.0</td>
</tr>
<tr>
<td>Pr(fouler bat → SINK)</td>
<td></td>
<td></td>
<td></td>
<td>1.0</td>
</tr>
</tbody>
</table>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The transitions to the SINK node are set to 1.0 because, once the emitted sequence ends, there's a 100% chance of going to the SINK node (no other options are available).</p>
</div>
<p><img src="svgbob_1bc36df6664449b1ec1fd21f5b522cbc73bb636e.svg" alt="Kroki diagram output"></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The example above doesn't cover non-emitting hidden states. A non-emitting hidden state's probability will be Pr(transition) because transitioning to it won't result in a symbol emission.</p>
<p>See Algorithms/Discriminator Hidden Markov Models/Most Probable Hidden Path/Viterbi Non-emitting Hidden States Algorithm_TOPIC</p>
</div>
<p>The directed graph above is called a Viterbi graph. The goal with a Viterbi graph is to determine the most likely set of hidden state transitions that resulted in the emitted symbols, which is the path from source node to sink node with the highest product (multiplication) of edge weights. In the example above, that path is ...</p>
<ol>
<li>Pr(foul|SOURCE → fouler bat): 0.02</li>
<li>Pr(hit|fouler bat → fouler bat): 0.45</li>
<li>Pr(miss|fouler bat → fouler bat): 0.09</li>
<li>Pr(hit|fouler bat → fouler bat): 0.45</li>
<li>Pr(hit|fouler bat → fouler bat): 0.45</li>
<li>Pr(fouler bat → SINK): 1.0</li>
</ol>
<p>0.2 * 0.45 * 0.09 * 0.45 * 0.45 = 0.0016.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="exploded HMM"></span> - A transformation of an HMM such that hidden state transitions are enumerated based on an emitted sequence (HMM cycles removed).</p>
<p><img src="svgbob_8ee5698233ab79365c89d5469fb70e9a6ae7c0b0.svg" alt="Kroki diagram output"></p>
<p>The HMM above is transformed to an exploded HMM based on the emitted sequence [foul, hit, miss, hit]. Each column in the exploded HMM represents an index within the emitted sequence, where that column replicates all possible hidden state transitions that lead to that index (both nodes and edges).</p>
<ul>
<li>SOURCE: Possible transitions are to [hitter bat, fouler bat], both of which lead to index 0 being emitted: [SOURCE → hitter bat0, SOURCE → fouler bat0].</li>
<li>hitter bat0: Possible transitions are to [hitter bat, fouler bat], both of which lead to index 1 being emitted: [hitter bat0 → hitter bat1, hitter bat0 → fouler bat1].</li>
<li>fouler bat0: Possible transitions are to [hitter bat, fouler bat], both of which lead to index 1 being emitted: [fouler bat0 → hitter bat1, fouler bat0 → fouler bat1].</li>
<li>hitter bat1: Possible transitions are to [hitter bat, fouler bat], both of which lead to index 2 being emitted: [hitter bat1 → hitter bat2, hitter bat1 → fouler bat2].</li>
<li>fouler bat1: Possible transitions are to [hitter bat, fouler bat], both of which lead to index 2 being emitted: [fouler bat1 → hitter bat2, fouler bat1 → fouler bat2].</li>
<li>...</li>
<li>hitter bat3: Emissions have ended, leading to an artificially placed sink node: [hitter bat3 → SINK, hitter bat3 → SINK].</li>
<li>fouler bat3: Emissions have ended, leading to an artificially placed sink node: [fouler bat3 → SINK, fouler bat3 → SINK].</li>
</ul>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>In certain algorithms, the sink node may exist in the HMM (meaning it isn't artificial).</p>
</div>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The example above doesn't include non-emitting hidden states. A non-emitting hidden state means that a transition to that hidden state doesn't result in a symbol emission. In other words, the emission index wouldn't increment, meaning the exploded HMM node would end up under the same column as the node that's pointing to it.</p>
<p>Consider if fouler bat in the example above had a transition to a non-emitting hidden state called bingo. At ...</p>
<ul>
<li>fouler bat0, the transition to bingo would be fouler bat0 → bingo0.</li>
<li>fouler bat1, the transition to bingo would be fouler bat1 → bingo1.</li>
<li>etc...</li>
</ul>
<p>Exploded bingo nodes maintain the same index as their exploded fouler bat predecessor because bingo is a non-emitting hidden state (nothing gets emitted when you transition to it, meaning you stay at the same index).</p>
</div>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="Viterbi learning"></span> - A Monte Carlo algorithm that uses the Viterbi algorithm to derive an HMM's probabilities from an emitted sequence.</p>
</li>
<li>
<p><span class="anki-answerpattern" data-pattern="Baum-Welch learning"></span> - A Monte Carlo algorithm that uses confidence measurements to derive an HMM's probabilities from an emitted sequence.</p>
</li>
</ul>

        </body></html>