<!DOCTYPE html><html><head>
            <meta content="text/html;charset=utf-8" http-equiv="Content-Type">
            <meta content="utf-8" http-equiv="encoding">
          <link href="katex/katex.min.css" rel="stylesheet"><link href="katex/CUSTOMIZED_NO_CENTER.css" rel="stylesheet"><title>Bioinformatics Algorithms
</title><meta name="viewport" content="width=device-width,initial-scale=1"><link href=".temp_githib_css/github-markdown.css" rel="stylesheet"><link href=".temp_highlightjs_css/default.css" rel="stylesheet"></head>
          <body class="markdown-body"><div class="toc">
<ul>
<li><a href="#H_Introduction">Introduction</a></li>
<li><a href="#H_Algorithms">Algorithms</a></li>
<ul>
<li><a href="#H_K-mer">K-mer</a></li>
<ul>
<li><a href="#H_Reverse%20Complement">Reverse Complement</a></li>
<li><a href="#H_Hamming%20Distance">Hamming Distance</a></li>
<li><a href="#H_Hamming%20Distance%20Neighbourhood">Hamming Distance Neighbourhood</a></li>
<li><a href="#H_Find%20Locations">Find Locations</a></li>
<li><a href="#H_Find%20Clumps">Find Clumps</a></li>
<li><a href="#H_Find%20Repeating">Find Repeating</a></li>
<li><a href="#H_Find%20Repeating%20in%20Window">Find Repeating in Window</a></li>
<li><a href="#H_Probability%20of%20Appearance">Probability of Appearance</a></li>
<ul>
<li><a href="#H_Bruteforce%20Algorithm">Bruteforce Algorithm</a></li>
<li><a href="#H_Selection%20Estimate%20Algorithm">Selection Estimate Algorithm</a></li>
</ul>
</ul>
<li><a href="#H_GC%20Skew">GC Skew</a></li>
<li><a href="#H_Motif">Motif</a></li>
<ul>
<li><a href="#H_Consensus%20String">Consensus String</a></li>
<li><a href="#H_Motif%20Matrix%20Count">Motif Matrix Count</a></li>
<li><a href="#H_Motif%20Matrix%20Profile">Motif Matrix Profile</a></li>
<li><a href="#H_Motif%20Matrix%20Score">Motif Matrix Score</a></li>
<ul>
<li><a href="#H_Popularity%20Algorithm">Popularity Algorithm</a></li>
<li><a href="#H_Entropy%20Algorithm">Entropy Algorithm</a></li>
<li><a href="#H_Relative%20Entropy%20Algorithm">Relative Entropy Algorithm</a></li>
</ul>
<li><a href="#H_Motif%20Logo">Motif Logo</a></li>
<li><a href="#H_K-mer%20Match%20Probability">K-mer Match Probability</a></li>
<li><a href="#H_Find%20Motif%20Matrix">Find Motif Matrix</a></li>
<ul>
<li><a href="#H1_Bruteforce%20Algorithm">Bruteforce Algorithm</a></li>
<li><a href="#H_Median%20String%20Algorithm">Median String Algorithm</a></li>
<li><a href="#H_Greedy%20Algorithm">Greedy Algorithm</a></li>
<li><a href="#H_Randomized%20Algorithm">Randomized Algorithm</a></li>
<li><a href="#H_Gibbs%20Sampling%20Algorithm">Gibbs Sampling Algorithm</a></li>
</ul>
<li><a href="#H_Motif%20Matrix%20Hybrid%20Alphabet">Motif Matrix Hybrid Alphabet</a></li>
</ul>
<li><a href="#H_DNA%20Assembly">DNA Assembly</a></li>
<ul>
<li><a href="#H_Stitch%20Reads">Stitch Reads</a></li>
<li><a href="#H_Stitch%20Read-Pairs">Stitch Read-Pairs</a></li>
<li><a href="#H_Break%20Reads">Break Reads</a></li>
<li><a href="#H_Break%20Read-Pairs">Break Read-Pairs</a></li>
<li><a href="#H_Probability%20of%20Fragment%20Occurrence">Probability of Fragment Occurrence</a></li>
<li><a href="#H_Overlap%20Graph">Overlap Graph</a></li>
<li><a href="#H_De%20Bruijn%20Graph">De Bruijn Graph</a></li>
<li><a href="#H_Find%20Bubbles">Find Bubbles</a></li>
<li><a href="#H_Find%20Contigs">Find Contigs</a></li>
</ul>
<li><a href="#H_Peptide%20Sequence">Peptide Sequence</a></li>
<ul>
<li><a href="#H_Codon%20Encode">Codon Encode</a></li>
<li><a href="#H_Codon%20Decode">Codon Decode</a></li>
<li><a href="#H_Experimental%20Spectrum">Experimental Spectrum</a></li>
<li><a href="#H_Theoretical%20Spectrum">Theoretical Spectrum</a></li>
<ul>
<li><a href="#H2_Bruteforce%20Algorithm">Bruteforce Algorithm</a></li>
<li><a href="#H_Prefix%20Sum%20Algorithm">Prefix Sum Algorithm</a></li>
</ul>
<li><a href="#H_Spectrum%20Convolution">Spectrum Convolution</a></li>
<li><a href="#H_Spectrum%20Score">Spectrum Score</a></li>
<li><a href="#H_Spectrum%20Sequence">Spectrum Sequence</a></li>
<ul>
<li><a href="#H3_Bruteforce%20Algorithm">Bruteforce Algorithm</a></li>
<li><a href="#H_Branch-and-bound%20Algorithm">Branch-and-bound Algorithm</a></li>
<li><a href="#H_Leaderboard%20Algorithm">Leaderboard Algorithm</a></li>
</ul>
</ul>
<li><a href="#H_Sequence%20Alignment">Sequence Alignment</a></li>
<ul>
<li><a href="#H_Find%20Maximum%20Path">Find Maximum Path</a></li>
<ul>
<li><a href="#H4_Bruteforce%20Algorithm">Bruteforce Algorithm</a></li>
<li><a href="#H_Cache%20Algorithm">Cache Algorithm</a></li>
<li><a href="#H_Backtrack%20Algorithm">Backtrack Algorithm</a></li>
</ul>
<li><a href="#H_Global%20Alignment">Global Alignment</a></li>
<ul>
<li><a href="#H_Graph%20Algorithm">Graph Algorithm</a></li>
<li><a href="#H_Matrix%20Algorithm">Matrix Algorithm</a></li>
<li><a href="#H_Divide-and-Conquer%20Algorithm">Divide-and-Conquer Algorithm</a></li>
</ul>
<li><a href="#H_Fitting%20Alignment">Fitting Alignment</a></li>
<ul>
<li><a href="#H1_Graph%20Algorithm">Graph Algorithm</a></li>
<li><a href="#H1_Matrix%20Algorithm">Matrix Algorithm</a></li>
</ul>
<li><a href="#H_Overlap%20Alignment">Overlap Alignment</a></li>
<ul>
<li><a href="#H2_Graph%20Algorithm">Graph Algorithm</a></li>
<li><a href="#H2_Matrix%20Algorithm">Matrix Algorithm</a></li>
</ul>
<li><a href="#H_Local%20Alignment">Local Alignment</a></li>
<ul>
<li><a href="#H3_Graph%20Algorithm">Graph Algorithm</a></li>
<li><a href="#H3_Matrix%20Algorithm">Matrix Algorithm</a></li>
</ul>
<li><a href="#H_Protein%20Scoring">Protein Scoring</a></li>
<ul>
<li><a href="#H_PAM%20Scoring%20Matrix">PAM Scoring Matrix</a></li>
<li><a href="#H_BLOSUM%20Scoring%20Matrix">BLOSUM Scoring Matrix</a></li>
</ul>
<li><a href="#H_Extended%20Gap%20Scoring">Extended Gap Scoring</a></li>
<ul>
<li><a href="#H_Naive%20Algorithm">Naive Algorithm</a></li>
<li><a href="#H_Layer%20Algorithm">Layer Algorithm</a></li>
</ul>
<li><a href="#H_Multiple%20Alignment">Multiple Alignment</a></li>
<ul>
<li><a href="#H4_Graph%20Algorithm">Graph Algorithm</a></li>
<li><a href="#H4_Matrix%20Algorithm">Matrix Algorithm</a></li>
<li><a href="#H1_Greedy%20Algorithm">Greedy Algorithm</a></li>
</ul>
<li><a href="#H_Sum-of-Pairs%20Scoring">Sum-of-Pairs Scoring</a></li>
<li><a href="#H_Entropy%20Scoring">Entropy Scoring</a></li>
</ul>
<li><a href="#H_Synteny">Synteny</a></li>
<ul>
<li><a href="#H_Genomic%20Dot%20Plot">Genomic Dot Plot</a></li>
<li><a href="#H_Synteny%20Graph">Synteny Graph</a></li>
<li><a href="#H_Reversal%20Path">Reversal Path</a></li>
<ul>
<li><a href="#H_Breakpoint%20List%20Algorithm">Breakpoint List Algorithm</a></li>
<li><a href="#H_Breakpoint%20Graph%20Algorithm">Breakpoint Graph Algorithm</a></li>
</ul>
</ul>
<li><a href="#H_Phylogeny">Phylogeny</a></li>
<ul>
<li><a href="#H_Tree%20to%20Additive%20Distance%20Matrix">Tree to Additive Distance Matrix</a></li>
<li><a href="#H_Tree%20to%20Simple%20Tree">Tree to Simple Tree</a></li>
<li><a href="#H_Additive%20Distance%20Matrix%20Cardinality">Additive Distance Matrix Cardinality</a></li>
<li><a href="#H_Test%20Additive%20Distance%20Matrix">Test Additive Distance Matrix</a></li>
<li><a href="#H_Find%20Limb%20Length">Find Limb Length</a></li>
<li><a href="#H_Test%20Same%20Subtree">Test Same Subtree</a></li>
<li><a href="#H_Trim">Trim</a></li>
<li><a href="#H_Bald">Bald</a></li>
<li><a href="#H_Un-trim%20Tree">Un-trim Tree</a></li>
<li><a href="#H_Find%20Neighbours">Find Neighbours</a></li>
<li><a href="#H_Find%20Neighbour%20Limb%20Lengths">Find Neighbour Limb Lengths</a></li>
<ul>
<li><a href="#H_Average%20Algorithm">Average Algorithm</a></li>
<li><a href="#H_Optimized%20Average%20Algorithm">Optimized Average Algorithm</a></li>
</ul>
<li><a href="#H_Expose%20Neighbour%20Parent">Expose Neighbour Parent</a></li>
<ul>
<li><a href="#H1_Average%20Algorithm">Average Algorithm</a></li>
<li><a href="#H_Inverse%20Algorithm">Inverse Algorithm</a></li>
</ul>
<li><a href="#H_Distance%20Matrix%20to%20Tree">Distance Matrix to Tree</a></li>
<ul>
<li><a href="#H_UPGMA%20Algorithm">UPGMA Algorithm</a></li>
<li><a href="#H_Additive%20Phylogeny%20Algorithm">Additive Phylogeny Algorithm</a></li>
<li><a href="#H_Neighbour%20Joining%20Phylogeny%20Algorithm">Neighbour Joining Phylogeny Algorithm</a></li>
<li><a href="#H_Evolutionary%20Algorithm">Evolutionary Algorithm</a></li>
</ul>
<li><a href="#H_Sequence%20Inference">Sequence Inference</a></li>
<ul>
<li><a href="#H_Small%20Parsimony%20Algorithm">Small Parsimony Algorithm</a></li>
<li><a href="#H_Nearest%20Neighbour%20Interchange%20Algorithm">Nearest Neighbour Interchange Algorithm</a></li>
</ul>
</ul>
<li><a href="#H_Gene%20Expression">Gene Expression</a></li>
<ul>
<li><a href="#H_Euclidean%20Distance%20Metric">Euclidean Distance Metric</a></li>
<li><a href="#H_Manhattan%20Distance%20Metric">Manhattan Distance Metric</a></li>
<li><a href="#H_Cosine%20Similarity%20Metric">Cosine Similarity Metric</a></li>
<li><a href="#H_Pearson%20Similarity%20Metric">Pearson Similarity Metric</a></li>
<li><a href="#H_K-Centers%20Clustering">K-Centers Clustering</a></li>
<li><a href="#H_K-Means%20Clustering">K-Means Clustering</a></li>
<li><a href="#H_Soft%20K-Means%20Clustering">Soft K-Means Clustering</a></li>
<li><a href="#H_Hierarchial%20Clustering">Hierarchial Clustering</a></li>
<li><a href="#H_Soft%20Hierarchial%20Clustering">Soft Hierarchial Clustering</a></li>
<li><a href="#H_CAST%20Clustering">CAST Clustering</a></li>
</ul>
</ul>
<li><a href="#H_Stories">Stories</a></li>
<ul>
<li><a href="#H_Bacterial%20Genome%20Replication">Bacterial Genome Replication</a></li>
<ul>
<li><a href="#H_Find%20Ori%20and%20Ter">Find Ori and Ter</a></li>
<li><a href="#H_Find%20the%20DnaA%20Box">Find the DnaA Box</a></li>
</ul>
<li><a href="#H_Transcription%20Factors">Transcription Factors</a></li>
<ul>
<li><a href="#H_Find%20Regulatory%20Motif">Find Regulatory Motif</a></li>
</ul>
<li><a href="#H_Non-ribosomal%20Peptides">Non-ribosomal Peptides</a></li>
<ul>
<li><a href="#H_Find%20Sequence">Find Sequence</a></li>
</ul>
<li><a href="#H_Genome%20Rearrangement">Genome Rearrangement</a></li>
<ul>
<li><a href="#H_Find%20Synteny%20Blocks">Find Synteny Blocks</a></li>
<li><a href="#H_Find%20Reversal%20Path">Find Reversal Path</a></li>
</ul>
<li><a href="#H_Evolutionary%20History">Evolutionary History</a></li>
<ul>
<li><a href="#H_Find%20Evolutionary%20Tree">Find Evolutionary Tree</a></li>
<li><a href="#H_Find%20Ancestral%20Features">Find Ancestral Features</a></li>
</ul>
</ul>
<li><a href="#H_Ideas">Ideas</a></li>
<li><a href="#H_Terminology">Terminology</a></li>
</ul>
</div>
<a name="H_Introduction"></a><h1>Introduction</h1>
<p>Bioinformatics is the science of transforming and processing biological data to gain new insights, particularly omics data: genomics, proteomics, metabolomics, etc.. Bioinformatics is mostly a mix of biology, computer science, and statistics / data science.</p>
<a name="H_Algorithms"></a><h1>Algorithms</h1>
<a name="H_K-mer"></a><h2>K-mer</h2>
<p><a name="BM_(Algorithms%5C%2FK-mer)_TOPIC%2F"></a><strong></strong></p>
<p>A <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> is a <a href="#BM_(substring)%2Fi">substring</a> of length k within some larger biological <a href="#BM_(sequence)%2Fi">sequence</a> (e.g. DNA or <a href="#BM_(amino%20acid)%2Fi">amino acid</a> chain). For example, in the DNA <a href="#BM_(sequence)%2Fi">sequence</a> GAAATC, the following <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>'s exist:</p>
<table>
<thead>
<tr>
<th>k</th>
<th><a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s</th>
</tr>
</thead>
<tbody>
<tr>
<td>1</td>
<td>G A A A T C</td>
</tr>
<tr>
<td>2</td>
<td>GA AA AA AT TC</td>
</tr>
<tr>
<td>3</td>
<td>GAA AAA AAT ATC</td>
</tr>
<tr>
<td>4</td>
<td>GAAA AAAT AATC</td>
</tr>
<tr>
<td>5</td>
<td>GAAAT AAATC</td>
</tr>
<tr>
<td>6</td>
<td>GAAATC</td>
</tr>
</tbody>
</table>
<p>Common scenarios involving <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s:</p>
<ul>
<li>Search for an exact <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>.</li>
<li>Search for an approximate <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> (fuzzy search).</li>
<li>Find <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s of interest in a <a href="#BM_(sequence)%2Fi">sequence</a> (e.g. repeating <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s).</li>
</ul>
<a name="H_Reverse%20Complement"></a><h3>Reverse Complement</h3>
<p><a name="BM_(Algorithms%5C%2FK-mer%2FReverse%20Complement)_TOPIC%2F"></a><strong></strong></p>
<p><strong>WHAT</strong>: Given a DNA <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>, calculate its <a href="#BM_(reverse%20complement)%2Fi">reverse complement</a>.</p>
<p><strong>WHY</strong>: Depending on the type of biological <a href="#BM_(sequence)%2Fi">sequence</a>, a <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> may have one or more alternatives. For DNA <a href="#BM_(sequence)%2Fi">sequence</a>s specifically, a <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> of interest may have an alternate form. Since the DNA molecule comes as 2 strands, where ...</p>
<ul>
<li>each strand's direction is opposite of the other,</li>
<li>each strand position has a nucleotide that complements the nucleotide at that same position on the other stand:
<ul>
<li>A ⟷ T</li>
<li>C ⟷ G</li>
</ul>
</li>
</ul>
<p><img src="svgbob_c44b72fd8e03cfe92b63a0a3b83bfe37c5551666.svg" alt="Kroki diagram output"></p>
<p>, ... the <a href="#BM_(reverse%20complement)%2Fi">reverse complement</a> of that <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> may be just as valid as the original <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>. For example, if an enzyme is known to bind to a specific DNA <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>, it's possible that it might also bind to the <a href="#BM_(reverse%20complement)%2Fi">reverse complement</a> of that <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>.</p>
<p><strong>ALGORITHM</strong>:</p>
<p><a href="ch1_code/src/ReverseComplementADnaKmer.py">ch1_code/src/ReverseComplementADnaKmer.py</a> (lines 5 to 22):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">reverse_complement</span>(<span class="hljs-params">strand: <span class="hljs-built_in">str</span></span>):
    ret = <span class="hljs-string">''</span>
    <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(<span class="hljs-number">0</span>, <span class="hljs-built_in">len</span>(strand)):
        base = strand[i]
        <span class="hljs-keyword">if</span> base == <span class="hljs-string">'A'</span> <span class="hljs-keyword">or</span> base == <span class="hljs-string">'a'</span>:
            base = <span class="hljs-string">'T'</span>
        <span class="hljs-keyword">elif</span> base == <span class="hljs-string">'T'</span> <span class="hljs-keyword">or</span> base == <span class="hljs-string">'t'</span>:
            base = <span class="hljs-string">'A'</span>
        <span class="hljs-keyword">elif</span> base == <span class="hljs-string">'C'</span> <span class="hljs-keyword">or</span> base == <span class="hljs-string">'c'</span>:
            base = <span class="hljs-string">'G'</span>
        <span class="hljs-keyword">elif</span> base == <span class="hljs-string">'G'</span> <span class="hljs-keyword">or</span> base == <span class="hljs-string">'g'</span>:
            base = <span class="hljs-string">'C'</span>
        <span class="hljs-keyword">else</span>:
            <span class="hljs-keyword">raise</span> Exception(<span class="hljs-string">'Unexpected base: '</span> + base)

        ret += base
    <span class="hljs-keyword">return</span> ret[::-<span class="hljs-number">1</span>]
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Original: TAATCCG</p>
<p>Reverse Complement: CGGATTA</p>
</div>
<p></p>
<a name="H_Hamming%20Distance"></a><h3>Hamming Distance</h3>
<p><a name="BM_(Algorithms%5C%2FK-mer%2FHamming%20Distance)_TOPIC%2F"></a><strong></strong></p>
<p><strong>WHAT</strong>: Given 2 <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s, the <a href="#BM_(hamming%20distance)%2Fi">hamming distance</a> is the number of positional mismatches between them.</p>
<p><strong>WHY</strong>: Imagine an enzyme that looks for a specific DNA <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> pattern to bind to. Since DNA is known to mutate, it may be that enzyme can also bind to other <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> patterns that are slight variations of the original. For example, that enzyme may be able to bind to both AAACTG and AAAGTG.</p>
<p><strong>ALGORITHM</strong>:</p>
<p><a href="ch1_code/src/HammingDistanceBetweenKmers.py">ch1_code/src/HammingDistanceBetweenKmers.py</a> (lines 5 to 13):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">hamming_distance</span>(<span class="hljs-params">kmer1: <span class="hljs-built_in">str</span>, kmer2: <span class="hljs-built_in">str</span></span>) -&gt; <span class="hljs-built_in">int</span>:
    mismatch = <span class="hljs-number">0</span>

    <span class="hljs-keyword">for</span> ch1, ch2 <span class="hljs-keyword">in</span> <span class="hljs-built_in">zip</span>(kmer1, kmer2):
        <span class="hljs-keyword">if</span> ch1 != ch2:
            mismatch += <span class="hljs-number">1</span>

    <span class="hljs-keyword">return</span> mismatch
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Kmer1: ACTTTGTT</p>
<p>Kmer2: AGTTTCTT</p>
<p>Hamming Distance: 2</p>
</div>
<p></p>
<a name="H_Hamming%20Distance%20Neighbourhood"></a><h3>Hamming Distance Neighbourhood</h3>
<p><a name="BM_(Algorithms%5C%2FK-mer%2FHamming%20Distance%20Neighbourhood)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FK-mer%2FHamming%20Distance)_TOPIC%2F">Algorithms/K-mer/Hamming Distance</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: Given a source <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> and a minimum <a href="#BM_(hamming%20distance)%2Fi">hamming distance</a>, find all <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s such within the <a href="#BM_(hamming%20distance)%2Fi">hamming distance</a> of the source <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>. In other words, find all <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s such that <code>hamming_distance(source_kmer, kmer) &lt;= min_distance</code>.</p>
<p><strong>WHY</strong>: Imagine an enzyme that looks for a specific DNA <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> pattern to bind to. Since DNA is known to mutate, it may be that enzyme can also bind to other <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> patterns that are slight variations of the original. This algorithm finds all such variations.</p>
<p><strong>ALGORITHM</strong>:</p>
<p><a href="ch1_code/src/FindAllDnaKmersWithinHammingDistance.py">ch1_code/src/FindAllDnaKmersWithinHammingDistance.py</a> (lines 5 to 20):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">find_all_dna_kmers_within_hamming_distance</span>(<span class="hljs-params">kmer: <span class="hljs-built_in">str</span>, hamming_dist: <span class="hljs-built_in">int</span></span>) -&gt; <span class="hljs-built_in">set</span>[<span class="hljs-built_in">str</span>]:
    <span class="hljs-keyword">def</span> <span class="hljs-title function_">recurse</span>(<span class="hljs-params">kmer: <span class="hljs-built_in">str</span>, hamming_dist: <span class="hljs-built_in">int</span>, output: <span class="hljs-built_in">set</span>[<span class="hljs-built_in">str</span>]</span>) -&gt; <span class="hljs-literal">None</span>:
        <span class="hljs-keyword">if</span> hamming_dist == <span class="hljs-number">0</span>:
            output.add(kmer)
            <span class="hljs-keyword">return</span>

        <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(<span class="hljs-number">0</span>, <span class="hljs-built_in">len</span>(kmer)):
            <span class="hljs-keyword">for</span> ch <span class="hljs-keyword">in</span> <span class="hljs-string">'ACTG'</span>:
                neighbouring_kmer = kmer[:i] + ch + kmer[i + <span class="hljs-number">1</span>:]
                recurse(neighbouring_kmer, hamming_dist - <span class="hljs-number">1</span>, output)

    output = <span class="hljs-built_in">set</span>()
    recurse(kmer, hamming_dist, output)

    <span class="hljs-keyword">return</span> output
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Kmers within hamming distance 1 of AAAA: {'ATAA', 'AACA', 'AAAC', 'GAAA', 'ACAA', 'AAAT', 'CAAA', 'AAAG', 'AGAA', 'AAGA', 'AATA', 'TAAA', 'AAAA'}</p>
</div>
<p></p>
<a name="H_Find%20Locations"></a><h3>Find Locations</h3>
<p><a name="BM_(Algorithms%5C%2FK-mer%2FFind%20Locations)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FK-mer%2FHamming%20Distance)_TOPIC%2F">Algorithms/K-mer/Hamming Distance</a></li>
<li><a href="#BM_(Algorithms%5C%2FK-mer%2FReverse%20Complement)_TOPIC%2F">Algorithms/K-mer/Reverse Complement</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: Given a <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>, find where that <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> occurs in some larger <a href="#BM_(sequence)%2Fi">sequence</a>. The search may potentially include the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>'s variants (e.g. <a href="#BM_(reverse%20complement)%2Fi">reverse complement</a>).</p>
<p><strong>WHY</strong>: Imagine that you know of a specific <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> pattern that serves some function in an organism. If you see that same <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> pattern appearing in some other related organism, it could be a sign that <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> pattern serves a similar function. For example, the same <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> pattern could be used by 2 related types of bacteria as a <a href="#BM_(DnaA%20box)%2Fi">DnaA box</a>.</p>
<p>The enzyme that operates on that <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> may also operate on its <a href="#BM_(reverse%20complement)%2Fi">reverse complement</a> as well as slight variations on that <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>. For example, if an enzyme binds to AAAAAAAAA, it may also bind to its...</p>
<ul>
<li><a href="#BM_(reverse%20complement)%2Fi">reverse complement</a>: TTTTTTTTT</li>
<li>approximate variants: AAAAAAAAA, AAATAAAAA, AAAAAGAAA, ...</li>
<li>approximate variants of its <a href="#BM_(reverse%20complement)%2Fi">reverse complement</a>s: TTTTTTTTT, TTTTTTATT, TTCTTTTTT, ...</li>
</ul>
<p><strong>ALGORITHM</strong>:</p>
<p><a href="ch1_code/src/FindLocations.py">ch1_code/src/FindLocations.py</a> (lines 11 to 32):</p>
<pre class="hljs"><code><span class="hljs-keyword">class</span> <span class="hljs-title class_">Options</span>(<span class="hljs-title class_ inherited__">NamedTuple</span>):
    hamming_distance: <span class="hljs-built_in">int</span> = <span class="hljs-number">0</span>
    reverse_complement: <span class="hljs-built_in">bool</span> = <span class="hljs-literal">False</span>


<span class="hljs-keyword">def</span> <span class="hljs-title function_">find_kmer_locations</span>(<span class="hljs-params">sequence: <span class="hljs-built_in">str</span>, kmer: <span class="hljs-built_in">str</span>, options: Options = Options(<span class="hljs-params"></span>)</span>) -&gt; <span class="hljs-type">List</span>[<span class="hljs-built_in">int</span>]:
    <span class="hljs-comment"># Construct test kmers</span>
    test_kmers = <span class="hljs-built_in">set</span>()
    test_kmers.add(kmer)
    [test_kmers.add(alt_kmer) <span class="hljs-keyword">for</span> alt_kmer <span class="hljs-keyword">in</span> find_all_dna_kmers_within_hamming_distance(kmer, options.hamming_distance)]
    <span class="hljs-keyword">if</span> options.reverse_complement:
        rc_kmer = reverse_complement(kmer)
        [test_kmers.add(alt_rc_kmer) <span class="hljs-keyword">for</span> alt_rc_kmer <span class="hljs-keyword">in</span> find_all_dna_kmers_within_hamming_distance(rc_kmer, options.hamming_distance)]

    <span class="hljs-comment"># Slide over the sequence's kmers and check for matches against test kmers</span>
    k = <span class="hljs-built_in">len</span>(kmer)
    idxes = []
    <span class="hljs-keyword">for</span> seq_kmer, i <span class="hljs-keyword">in</span> slide_window(sequence, k):
        <span class="hljs-keyword">if</span> seq_kmer <span class="hljs-keyword">in</span> test_kmers:
            idxes.append(i)
    <span class="hljs-keyword">return</span> idxes
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Found AAAA in AAAAGAACCTAATCTTAAAGGAGATGATGATTCTAA at index [0, 1, 2, 3, 12, 15, 16, 30]</p>
</div>
<p></p>
<a name="H_Find%20Clumps"></a><h3>Find Clumps</h3>
<p><a name="BM_(Algorithms%5C%2FK-mer%2FFind%20Clumps)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FK-mer%2FFind%20Locations)_TOPIC%2F">Algorithms/K-mer/Find Locations</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: Given a <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>, find where that <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> <a href="#BM_(cluster)%2Fi">cluster</a>s in some larger <a href="#BM_(sequence)%2Fi">sequence</a>. The search may potentially include the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>'s variants (e.g. <a href="#BM_(reverse%20complement)%2Fi">reverse complement</a>).</p>
<p><strong>WHY</strong>: An enzyme may need to bind to a specific region of DNA to begin doing its job. That is, it looks for a specific <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> pattern to bind to, where that <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> represents the beginning of some larger DNA region that it operates on. Since DNA is known to mutate, often times you'll find multiple copies of the same <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> pattern <a href="#BM_(cluster)%2Fi">cluster</a>ed together -- if one copy mutated to become unusable, the other copies are still around.</p>
<p>For example, the <a href="#BM_(DnaA%20box)%2Fi">DnaA box</a> is a special <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> pattern used by enzymes during DNA replication. Since DNA is known to mutate, the <a href="#BM_(DnaA%20box)%2Fi">DnaA box</a> can be found repeating multiple times in the region of DNA known as the <a href="#BM_(replication%20origin)%2Fi">replication origin</a>. Finding the <a href="#BM_(DnaA%20box)%2Fi">DnaA box</a> <a href="#BM_(cluster)%2Fi">cluster</a>ed in a small region is a good indicator that you've found the <a href="#BM_(replication%20origin)%2Fi">replication origin</a>.</p>
<p><strong>ALGORITHM</strong>:</p>
<p><a href="ch1_code/src/FindClumps.py">ch1_code/src/FindClumps.py</a> (lines 10 to 26):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">find_kmer_clusters</span>(<span class="hljs-params">sequence: <span class="hljs-built_in">str</span>, kmer: <span class="hljs-built_in">str</span>, min_occurrence_in_cluster: <span class="hljs-built_in">int</span>, cluster_window_size: <span class="hljs-built_in">int</span>, options: Options = Options(<span class="hljs-params"></span>)</span>) -&gt; <span class="hljs-type">List</span>[<span class="hljs-built_in">int</span>]:
    cluster_locs = []

    locs = find_kmer_locations(sequence, kmer, options)
    start_i = <span class="hljs-number">0</span>
    occurrence_count = <span class="hljs-number">1</span>
    <span class="hljs-keyword">for</span> end_i <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(<span class="hljs-number">1</span>, <span class="hljs-built_in">len</span>(locs)):
        <span class="hljs-keyword">if</span> locs[end_i] - locs[start_i] &lt; cluster_window_size:  <span class="hljs-comment"># within a cluster window?</span>
            occurrence_count += <span class="hljs-number">1</span>
        <span class="hljs-keyword">else</span>:
            <span class="hljs-keyword">if</span> occurrence_count &gt;= min_occurrence_in_cluster:  <span class="hljs-comment"># did the last cluster meet the min ocurr requirement?</span>
                cluster_locs.append(locs[start_i])
            start_i = end_i
            occurrence_count = <span class="hljs-number">1</span>

    <span class="hljs-keyword">return</span> cluster_locs
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Found clusters of GGG (at least 3 occurrences in window of 13) in GGGACTGAACAAACAAATTTGGGAGGGCACGGGTTAAAGGAGATGATGATTCAAAGGGT at index [19, 37]</p>
</div>
<p></p>
<a name="H_Find%20Repeating"></a><h3>Find Repeating</h3>
<p><a name="BM_(Algorithms%5C%2FK-mer%2FFind%20Repeating)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FK-mer%2FReverse%20Complement)_TOPIC%2F">Algorithms/K-mer/Reverse Complement</a></li>
<li><a href="#BM_(Algorithms%5C%2FK-mer%2FHamming%20Distance%20Neighbourhood)_TOPIC%2F">Algorithms/K-mer/Hamming Distance Neighbourhood</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: Given a <a href="#BM_(sequence)%2Fi">sequence</a>, find <a href="#BM_(cluster)%2Fi">cluster</a>s of unique <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s within that <a href="#BM_(sequence)%2Fi">sequence</a>. In other words, for each unique <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> that exists in the <a href="#BM_(sequence)%2Fi">sequence</a>, see if it <a href="#BM_(cluster)%2Fi">cluster</a>s in the <a href="#BM_(sequence)%2Fi">sequence</a>. The search may potentially include variants of <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> variants (e.g. <a href="#BM_(reverse%20complement)%2Fi">reverse complement</a>s of the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s).</p>
<p><strong>WHY</strong>: An enzyme may need to bind to a specific region of DNA to begin doing its job. That is, it looks for a specific <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> pattern to bind to, where that <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> represents the beginning of some larger DNA region that it operates on. Since DNA is known to mutate, often times you'll find multiple copies of the same <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> pattern <a href="#BM_(cluster)%2Fi">cluster</a>ed together -- if one copy mutated to become unusable, the other copies are still around.</p>
<p>For example, the <a href="#BM_(DnaA%20box)%2Fi">DnaA box</a> is a special <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> pattern used by enzymes during DNA replication. Since DNA is known to mutate, the <a href="#BM_(DnaA%20box)%2Fi">DnaA box</a> can be found repeating multiple times in the region of DNA known as the <a href="#BM_(replication%20origin)%2Fi">replication origin</a>. Given that you don't know the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> pattern for the <a href="#BM_(DnaA%20box)%2Fi">DnaA box</a> but you do know the <a href="#BM_(replication%20origin)%2Fi">replication origin</a>, you can scan through the <a href="#BM_(replication%20origin)%2Fi">replication origin</a> for repeating <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> patterns. If a pattern is found to heavily repeat, it's a good candidate that it's the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> pattern for the <a href="#BM_(DnaA%20box)%2Fi">DnaA box</a>.</p>
<p><strong>ALGORITHM</strong>:</p>
<p><a href="ch1_code/src/FindRepeating.py">ch1_code/src/FindRepeating.py</a> (lines 12 to 41):</p>
<pre class="hljs"><code><span class="hljs-keyword">from</span> Utils <span class="hljs-keyword">import</span> slide_window


<span class="hljs-keyword">def</span> <span class="hljs-title function_">count_kmers</span>(<span class="hljs-params">data: <span class="hljs-built_in">str</span>, k: <span class="hljs-built_in">int</span>, options: Options = Options(<span class="hljs-params"></span>)</span>) -&gt; Counter[<span class="hljs-built_in">str</span>]:
    counter = Counter()
    <span class="hljs-keyword">for</span> kmer, i <span class="hljs-keyword">in</span> slide_window(data, k):
        neighbourhood = find_all_dna_kmers_within_hamming_distance(kmer, options.hamming_distance)
        <span class="hljs-keyword">for</span> neighbouring_kmer <span class="hljs-keyword">in</span> neighbourhood:
            counter[neighbouring_kmer] += <span class="hljs-number">1</span>

        <span class="hljs-keyword">if</span> options.reverse_complement:
            kmer_rc = reverse_complement(kmer)
            neighbourhood = find_all_dna_kmers_within_hamming_distance(kmer_rc, options.hamming_distance)
            <span class="hljs-keyword">for</span> neighbouring_kmer <span class="hljs-keyword">in</span> neighbourhood:
                counter[neighbouring_kmer] += <span class="hljs-number">1</span>

    <span class="hljs-keyword">return</span> counter


<span class="hljs-keyword">def</span> <span class="hljs-title function_">top_repeating_kmers</span>(<span class="hljs-params">data: <span class="hljs-built_in">str</span>, k: <span class="hljs-built_in">int</span>, options: Options = Options(<span class="hljs-params"></span>)</span>) -&gt; <span class="hljs-type">Set</span>[<span class="hljs-built_in">str</span>]:
    counts = count_kmers(data, k, options)

    _, top_count = counts.most_common(<span class="hljs-number">1</span>)[<span class="hljs-number">0</span>]

    top_kmers = <span class="hljs-built_in">set</span>()
    <span class="hljs-keyword">for</span> kmer, count <span class="hljs-keyword">in</span> counts.items():
        <span class="hljs-keyword">if</span> count == top_count:
            top_kmers.add((kmer, count))
    <span class="hljs-keyword">return</span> top_kmers
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Top 5-mer frequencies for GGGACTGAACAAACAAATTTGGGAGGGCACGGGTTAAAGGAGATGATGATTCAAAGGGT:</p>
<ul>
<li>TGATC = 6 occurrences</li>
<li>TTTAA = 6 occurrences</li>
<li>GATCA = 6 occurrences</li>
<li>CCCTT = 6 occurrences</li>
<li>AAGGG = 6 occurrences</li>
<li>TTAAA = 6 occurrences</li>
</ul>
</div>
<p></p>
<a name="H_Find%20Repeating%20in%20Window"></a><h3>Find Repeating in Window</h3>
<p><a name="BM_(Algorithms%5C%2FK-mer%2FFind%20Repeating%20in%20Window)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FK-mer%2FFind%20Repeating)_TOPIC%2F">Algorithms/K-mer/Find Repeating</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: Given a <a href="#BM_(sequence)%2Fi">sequence</a>, find regions within that <a href="#BM_(sequence)%2Fi">sequence</a> that contain <a href="#BM_(cluster)%2Fi">cluster</a>s of unique <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s. In other words, ...</p>
<ul>
<li>slide a window over the <a href="#BM_(cluster)%2Fi">cluster</a>.</li>
<li>for each unique <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> that exists in the window, see if it <a href="#BM_(cluster)%2Fi">cluster</a>s in the <a href="#BM_(sequence)%2Fi">sequence</a>.</li>
</ul>
<p>The search may potentially include variants of <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> variants (e.g. <a href="#BM_(reverse%20complement)%2Fi">reverse complement</a>s of the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s).</p>
<p><strong>WHY</strong>: An enzyme may need to bind to a specific region of DNA to begin doing its job. That is, it looks for a specific <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> pattern to bind to, where that <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> represents the beginning of some larger DNA region that it operates on. Since DNA is known to mutate, often times you'll find multiple copies of the same <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> pattern <a href="#BM_(cluster)%2Fi">cluster</a>ed together -- if one copy mutated to become unusable, the other copies are still around.</p>
<p>For example, the <a href="#BM_(DnaA%20box)%2Fi">DnaA box</a> is a special <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> pattern used by enzymes during DNA replication. Since DNA is known to mutate, the <a href="#BM_(DnaA%20box)%2Fi">DnaA box</a> can be found repeating multiple times in the region of DNA known as the <a href="#BM_(replication%20origin)%2Fi">replication origin</a>. Given that you don't know the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> pattern for the <a href="#BM_(DnaA%20box)%2Fi">DnaA box</a> but you do know the <a href="#BM_(replication%20origin)%2Fi">replication origin</a>, you can scan through the <a href="#BM_(replication%20origin)%2Fi">replication origin</a> for repeating <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> patterns. If a pattern is found to heavily repeat, it's a good candidate that it's the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> pattern for the <a href="#BM_(DnaA%20box)%2Fi">DnaA box</a>.</p>
<p><strong>ALGORITHM</strong>:</p>
<p><a href="ch1_code/src/FindRepeatingInWindow.py">ch1_code/src/FindRepeatingInWindow.py</a> (lines 20 to 67):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">scan_for_repeating_kmers_in_clusters</span>(<span class="hljs-params">sequence: <span class="hljs-built_in">str</span>, k: <span class="hljs-built_in">int</span>, min_occurrence_in_cluster: <span class="hljs-built_in">int</span>, cluster_window_size: <span class="hljs-built_in">int</span>, options: Options = Options(<span class="hljs-params"></span>)</span>) -&gt; <span class="hljs-type">Set</span>[KmerCluster]:
    <span class="hljs-keyword">def</span> <span class="hljs-title function_">neighborhood</span>(<span class="hljs-params">kmer: <span class="hljs-built_in">str</span></span>) -&gt; <span class="hljs-type">Set</span>[<span class="hljs-built_in">str</span>]:
        neighbourhood = find_all_dna_kmers_within_hamming_distance(kmer, options.hamming_distance)
        <span class="hljs-keyword">if</span> options.reverse_complement:
            kmer_rc = reverse_complement(kmer)
            neighbourhood = find_all_dna_kmers_within_hamming_distance(kmer_rc, options.hamming_distance)
        <span class="hljs-keyword">return</span> neighbourhood

    kmer_counter = {}

    <span class="hljs-keyword">def</span> <span class="hljs-title function_">add_kmer</span>(<span class="hljs-params">kmer: <span class="hljs-built_in">str</span>, loc: <span class="hljs-built_in">int</span></span>) -&gt; <span class="hljs-literal">None</span>:
        <span class="hljs-keyword">if</span> kmer <span class="hljs-keyword">not</span> <span class="hljs-keyword">in</span> kmer_counter:
            kmer_counter[kmer] = <span class="hljs-built_in">set</span>()
        kmer_counter[kmer].add(window_idx + kmer_idx)

    <span class="hljs-keyword">def</span> <span class="hljs-title function_">remove_kmer</span>(<span class="hljs-params">kmer: <span class="hljs-built_in">str</span>, loc: <span class="hljs-built_in">int</span></span>) -&gt; <span class="hljs-literal">None</span>:
        kmer_counter[kmer].remove(window_idx - <span class="hljs-number">1</span>)
        <span class="hljs-keyword">if</span> <span class="hljs-built_in">len</span>(kmer_counter[kmer]) == <span class="hljs-number">0</span>:
            <span class="hljs-keyword">del</span> kmer_counter[kmer]

    clustered_kmers = <span class="hljs-built_in">set</span>()

    old_first_kmer = <span class="hljs-literal">None</span>
    <span class="hljs-keyword">for</span> window, window_idx <span class="hljs-keyword">in</span> slide_window(sequence, cluster_window_size):
        first_kmer = window[<span class="hljs-number">0</span>:k]
        last_kmer = window[-k:]

        <span class="hljs-comment"># If first iteration, add all kmers</span>
        <span class="hljs-keyword">if</span> window_idx == <span class="hljs-number">0</span>:
            <span class="hljs-keyword">for</span> kmer, kmer_idx <span class="hljs-keyword">in</span> slide_window(window, k):
                <span class="hljs-keyword">for</span> alt_kmer <span class="hljs-keyword">in</span> neighborhood(kmer):
                    add_kmer(alt_kmer, window_idx + kmer_idx)
        <span class="hljs-keyword">else</span>:
            <span class="hljs-comment"># Add kmer that was walked in to</span>
            <span class="hljs-keyword">for</span> new_last_kmer <span class="hljs-keyword">in</span> neighborhood(last_kmer):
                add_kmer(new_last_kmer, window_idx + cluster_window_size - k)
            <span class="hljs-comment"># Remove kmer that was walked out of</span>
            <span class="hljs-keyword">if</span> old_first_kmer <span class="hljs-keyword">is</span> <span class="hljs-keyword">not</span> <span class="hljs-literal">None</span>:
                <span class="hljs-keyword">for</span> alt_kmer <span class="hljs-keyword">in</span> neighborhood(old_first_kmer):
                    remove_kmer(alt_kmer, window_idx - <span class="hljs-number">1</span>)

        old_first_kmer = first_kmer

        <span class="hljs-comment"># Find clusters within window -- tuple is k-mer, start_idx, occurrence_count</span>
        [clustered_kmers.add(KmerCluster(k, <span class="hljs-built_in">min</span>(v), <span class="hljs-built_in">len</span>(v))) <span class="hljs-keyword">for</span> k, v <span class="hljs-keyword">in</span> kmer_counter.items() <span class="hljs-keyword">if</span> <span class="hljs-built_in">len</span>(v) &gt;= min_occurrence_in_cluster]

    <span class="hljs-keyword">return</span> clustered_kmers
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Found clusters of k=9 (at least 6 occurrences in window of 20) in TTTTTTTTTTTTTCCCTTTTTTTTTCCCTTTTTTTTTTTTT at...</p>
<ul>
<li>KmerCluster(kmer='CAAAAAAAA', start_index=0, occurrence_count=6)</li>
<li>KmerCluster(kmer='AGAAAAAAA', start_index=0, occurrence_count=6)</li>
<li>KmerCluster(kmer='AAAAAAAAA', start_index=0, occurrence_count=6)</li>
<li>KmerCluster(kmer='TAAAAAAAA', start_index=0, occurrence_count=6)</li>
<li>KmerCluster(kmer='GAAAAAAAA', start_index=0, occurrence_count=7)</li>
<li>KmerCluster(kmer='GAAAAAAAA', start_index=1, occurrence_count=6)</li>
<li>KmerCluster(kmer='AAAAAAAAG', start_index=26, occurrence_count=6)</li>
<li>KmerCluster(kmer='AAAAAAAGA', start_index=26, occurrence_count=6)</li>
<li>KmerCluster(kmer='AAAAAAAAG', start_index=26, occurrence_count=7)</li>
<li>KmerCluster(kmer='AAAAAAAAA', start_index=27, occurrence_count=6)</li>
<li>KmerCluster(kmer='AAAAAAAAC', start_index=27, occurrence_count=6)</li>
<li>KmerCluster(kmer='AAAAAAAAT', start_index=27, occurrence_count=6)</li>
</ul>
</div>
<p></p>
<a name="H_Probability%20of%20Appearance"></a><h3>Probability of Appearance</h3>
<p><a name="BM_(Algorithms%5C%2FK-mer%2FProbability%20of%20Appearance)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FK-mer%2FFind%20Locations)_TOPIC%2F">Algorithms/K-mer/Find Locations</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: Given ...</p>
<ul>
<li>the length of a <a href="#BM_(sequence)%2Fi">sequence</a> (n)</li>
<li>a <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a></li>
<li>a count (c)</li>
</ul>
<p>... find the probability of that <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> appearing at least c times within an arbitrary <a href="#BM_(sequence)%2Fi">sequence</a> of length n. For example, the probability that the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">2-mer</a> AA appears at least 2 times in a <a href="#BM_(sequence)%2Fi">sequence</a> of length 4:</p>
<ul>
<li>AAAA - yes</li>
<li>AAAT - yes</li>
<li>AAAC - yes</li>
<li>AAAG - yes</li>
<li>AATA - no</li>
<li>AATT - no</li>
<li>AATC - no</li>
<li>AATG - no</li>
<li>...</li>
<li>TAAA - yes</li>
<li>...</li>
<li>CAAA - yes</li>
<li>...</li>
<li>GAAA - yes</li>
<li>...</li>
<li>GGGA - no</li>
<li>GGGT - no</li>
<li>GGGC - no</li>
<li>GGGG - no</li>
</ul>
<p>The probability is 7/256.</p>
<p>This isn't trivial to accurately compute because the occurrences of a <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> within a <a href="#BM_(sequence)%2Fi">sequence</a> may overlap. For example, the number of times AA appears in AAAA is 3 while in CAAA it's 2.</p>
<p><strong>WHY</strong>: When a <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> is found within a <a href="#BM_(sequence)%2Fi">sequence</a>, knowing the probability of that <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> being found within an arbitrary <a href="#BM_(sequence)%2Fi">sequence</a> of the same length hints at the significance of the find. For example, if some <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">10-mer</a> has a 0.2 chance of appearing in an arbitrary <a href="#BM_(sequence)%2Fi">sequence</a> of length 50, that's too high of a chance to consider it a significant find -- 0.2 means 1 in 5 chance that the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">10-mer</a> just randomly happens to appear.</p>
<a name="H_Bruteforce%20Algorithm"></a><h4>Bruteforce Algorithm</h4>
<p><strong>ALGORITHM</strong>:</p>
<p>This algorithm tries every possible combination of <a href="#BM_(sequence)%2Fi">sequence</a> to find the probability. It falls over once the length of the <a href="#BM_(sequence)%2Fi">sequence</a> extends into the double digits. It's intended to help conceptualize what's going on.</p>
<p><a href="ch1_code/src/BruteforceProbabilityOfKmerInArbitrarySequence.py">ch1_code/src/BruteforceProbabilityOfKmerInArbitrarySequence.py</a> (lines 9 to 39):</p>
<pre class="hljs"><code><span class="hljs-comment"># Of the X sequence combinations tried, Y had the k-mer. The probability is Y/X.</span>
<span class="hljs-keyword">def</span> <span class="hljs-title function_">bruteforce_probability</span>(<span class="hljs-params">searchspace_len: <span class="hljs-built_in">int</span>, searchspace_symbol_count: <span class="hljs-built_in">int</span>, search_for: <span class="hljs-type">List</span>[<span class="hljs-built_in">int</span>], min_occurrence: <span class="hljs-built_in">int</span></span>) -&gt; (<span class="hljs-built_in">int</span>, <span class="hljs-built_in">int</span>):
    found = <span class="hljs-number">0</span>
    found_max = searchspace_symbol_count ** searchspace_len

    str_to_search = [<span class="hljs-number">0</span>] * searchspace_len

    <span class="hljs-keyword">def</span> <span class="hljs-title function_">count_instances</span>():
        ret = <span class="hljs-number">0</span>
        <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(<span class="hljs-number">0</span>, searchspace_len - <span class="hljs-built_in">len</span>(search_for) + <span class="hljs-number">1</span>):
            <span class="hljs-keyword">if</span> str_to_search[i:i + <span class="hljs-built_in">len</span>(search_for)] == search_for:
                ret += <span class="hljs-number">1</span>
        <span class="hljs-keyword">return</span> ret

    <span class="hljs-keyword">def</span> <span class="hljs-title function_">walk</span>(<span class="hljs-params">idx: <span class="hljs-built_in">int</span></span>):
        <span class="hljs-keyword">nonlocal</span> found

        <span class="hljs-keyword">if</span> idx == searchspace_len:
            count = count_instances()
            <span class="hljs-keyword">if</span> count &gt;= min_occurrence:
                found += <span class="hljs-number">1</span>
        <span class="hljs-keyword">else</span>:
            <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(<span class="hljs-number">0</span>, searchspace_symbol_count):
                walk(idx + <span class="hljs-number">1</span>)
                str_to_search[idx] += <span class="hljs-number">1</span>
            str_to_search[idx] = <span class="hljs-number">0</span>

    walk(<span class="hljs-number">0</span>)

    <span class="hljs-keyword">return</span> found, found_max
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Brute-forcing probability of ACTG in arbitrary sequence of length 8</p>
<p>Probability: 0.0195159912109375 (1279/65536)</p>
</div>
<p></p>
<a name="H_Selection%20Estimate%20Algorithm"></a><h4>Selection Estimate Algorithm</h4>
<p><strong>ALGORITHM</strong>:</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The explanation in the comments below are a bastardization of "1.13 Detour: Probabilities of Patterns in a String" in the Pevzner book...</p>
</div>
<p>This algorithm tries estimating the probability by ignoring the fact that the occurrences of a <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> in a <a href="#BM_(sequence)%2Fi">sequence</a> may overlap. For example, searching for the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">2-mer</a> AA in the <a href="#BM_(sequence)%2Fi">sequence</a> AAAT yields 2 instances of AA:</p>
<ul>
<li>[AA]AT</li>
<li>A[AA]T</li>
</ul>
<p>If you go ahead and ignore overlaps, you can think of the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s occurring in a string as insertions. For example, imagine a <a href="#BM_(sequence)%2Fi">sequence</a> of length 7 and the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">2-mer</a> AA. If you were to inject 2 instances of AA into the <a href="#BM_(sequence)%2Fi">sequence</a> to get it to reach length 7, how would that look?</p>
<p>2 instances of a <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">2-mer</a> is 4 characters has a length of 5. To get the <a href="#BM_(sequence)%2Fi">sequence</a> to end up with a length of 7 after the insertions, the <a href="#BM_(sequence)%2Fi">sequence</a> needs to start with a length of 3:</p>
<pre><code>SSS
</code></pre>
<p>Given that you're changing reality to say that the instances WON'T overlap in the <a href="#BM_(sequence)%2Fi">sequence</a>, you can treat each instance of the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">2-mer</a> AA as a single entity being inserted. The number of ways that these 2 instances can be inserted into the <a href="#BM_(sequence)%2Fi">sequence</a> is 10:</p>
<pre><code>I = insertion of AA, S = arbitrary sequence character

IISSS  ISISS  ISSIS  ISSSI
SIISS  SISIS  SISSI
SSIIS  SSISI
SSSII
</code></pre>
<p>Another way to think of the above insertions is that they aren't insertions. Rather, you have 5 items in total and you're selecting 2 of them. How many ways can you select 2 of those 5 items? 10.</p>
<p>The number of ways to insert can be counted via the "binomial coefficient": <code>bc(m, k) = m!/(k!(m-k)!)</code>, where m is the total number of items (5 in the example above) and k is the number of selections (2 in the example above). For the example above:</p>
<pre><code>bc(5, 2) = 5!/(2!(5-2)!) = 10
</code></pre>
<p>Since the SSS can be any arbitrary nucleotide <a href="#BM_(sequence)%2Fi">sequence</a> of 3, count the number of different representations that are possible for SSS: <code>4^3 = 4*4*4 = 64</code> (4^3, 4 because a nucleotide can be one of ACTG, 3 because the length is 3). In each of these representations, the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">2-mer</a> AA can be inserted in 10 different ways:</p>
<pre><code>64*10 = 640
</code></pre>
<p>Since the total length of the <a href="#BM_(sequence)%2Fi">sequence</a> is 7, count the number of different representations that are possible:</p>
<pre><code>4^7 = 4*4*4*4*4*4*4 = 16384
</code></pre>
<p>The estimated probability is 640/16384. For...</p>
<ul>
<li>non-overlapping <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s, the estimation will actually be relatively accurate.</li>
<li>overlapping <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s, the estimation won't be as accurate.</li>
</ul>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>Maybe try training a deep learning model to see if it can provide better estimates?</p>
</div>
<p><a href="ch1_code/src/EstimateProbabilityOfKmerInArbitrarySequence.py">ch1_code/src/EstimateProbabilityOfKmerInArbitrarySequence.py</a> (lines 57 to 70):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">estimate_probability</span>(<span class="hljs-params">searchspace_len: <span class="hljs-built_in">int</span>, searchspace_symbol_count: <span class="hljs-built_in">int</span>, search_for: <span class="hljs-type">List</span>[<span class="hljs-built_in">int</span>], min_occurrence: <span class="hljs-built_in">int</span></span>) -&gt; <span class="hljs-built_in">float</span>:
    <span class="hljs-keyword">def</span> <span class="hljs-title function_">factorial</span>(<span class="hljs-params">num</span>):
        <span class="hljs-keyword">if</span> num == <span class="hljs-number">1</span>:
            <span class="hljs-keyword">return</span> num
        <span class="hljs-keyword">else</span>:
            <span class="hljs-keyword">return</span> num * factorial(num - <span class="hljs-number">1</span>)

    <span class="hljs-keyword">def</span> <span class="hljs-title function_">bc</span>(<span class="hljs-params">m, k</span>):
        <span class="hljs-keyword">return</span> factorial(m) / (factorial(k) * factorial(m - k))

    k = <span class="hljs-built_in">len</span>(search_for)
    n = (searchspace_len - min_occurrence * k)
    <span class="hljs-keyword">return</span> bc(n + min_occurrence, min_occurrence) * (searchspace_symbol_count ** n) / searchspace_symbol_count ** searchspace_len
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Estimating probability of ACTG in arbitrary sequence of length 8</p>
<p>Probability: 0.01953125</p>
</div>
<p></p>
<a name="H_GC%20Skew"></a><h2>GC Skew</h2>
<p><a name="BM_(Algorithms%5C%2FGC%20Skew)_TOPIC%2F"></a><strong></strong></p>
<p><strong>WHAT</strong>: Given a <a href="#BM_(sequence)%2Fi">sequence</a>, create a counter and walk over the <a href="#BM_(sequence)%2Fi">sequence</a>. Whenever a ...</p>
<ul>
<li>G is encountered, increment the counter.</li>
<li>C is encountered, decrement the counter.</li>
</ul>
<p><strong>WHY</strong>: Given the DNA <a href="#BM_(sequence)%2Fi">sequence</a> of an organism, some segments may have lower count of Gs vs Cs.</p>
<p>During replication, some segments of DNA stay single-stranded for a much longer time than other segments. <a href="#BM_(single%20stranded%20DNA%7Csingle-stranded%20DNA)%2Fi">Single-stranded DNA</a> is 100 times more susceptible to mutations than <a href="#BM_(double%20stranded%20DNA%7Cdouble-stranded%20DNA)%2Fi">double-stranded DNA</a>. Specifically, in <a href="#BM_(single%20stranded%20DNA%7Csingle-stranded%20DNA)%2Fi">single-stranded DNA</a>, C has a greater tendency to mutate to T. When that <a href="#BM_(single%20stranded%20DNA%7Csingle-stranded%20DNA)%2Fi">single-stranded DNA</a> re-binds to a <a href="#BM_(neighbour)%2Fi">neighbour</a>ing strand, the positions of any nucleotides that mutated from C to T will change on the <a href="#BM_(neighbour)%2Fi">neighbour</a>ing strand from G to A.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>Recall that the <a href="#BM_(reverse%20complement)%2Fi">reverse complement</a>s of ...</p>
<ul>
<li>C is G</li>
<li>A is T</li>
</ul>
<p>It mutated from C to T. Since its now T, its complement is A.</p>
</div>
<p>Plotting the skew shows roughly which segments of DNA stayed single-stranded for a longer period of time. That information hints at special / useful locations in the organism's DNA <a href="#BM_(sequence)%2Fi">sequence</a> (<a href="#BM_(replication%20origin)%2Fi">replication origin</a> / <a href="#BM_(replication%20terminus)%2Fi">replication terminus</a>).</p>
<p><strong>ALGORITHM</strong>:</p>
<p><a href="ch1_code/src/GCSkew.py">ch1_code/src/GCSkew.py</a> (lines 8 to 21):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">gc_skew</span>(<span class="hljs-params">seq: <span class="hljs-built_in">str</span></span>):
    counter = <span class="hljs-number">0</span>
    skew = [counter]
    <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(<span class="hljs-built_in">len</span>(seq)):
        <span class="hljs-keyword">if</span> seq[i] == <span class="hljs-string">'G'</span>:
            counter += <span class="hljs-number">1</span>
            skew.append(counter)
        <span class="hljs-keyword">elif</span> seq[i] == <span class="hljs-string">'C'</span>:
            counter -= <span class="hljs-number">1</span>
            skew.append(counter)
        <span class="hljs-keyword">else</span>:
            skew.append(counter)
    <span class="hljs-keyword">return</span> skew
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Calculating skew for: ...</p>
<p>Result: [0, -1, -1,...</p>
<p><img src="skew_babf857bad6dd935f0d1ddbbb8b5005f.png" alt="GC Skew Plot"></p>
</div>
<p></p>
<a name="H_Motif"></a><h2>Motif</h2>
<p><a name="BM_(Algorithms%5C%2FMotif)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FK-mer)_TOPIC%2F">Algorithms/K-mer</a></li>
</ul>
</div>
<p>A <a href="#BM_(motif)%2Fi">motif</a> is a pattern that matches many different <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s, where those matched <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s have some shared biological significance. The pattern matches a fixed k where each position may have alternate forms. The simplest way to think of a <a href="#BM_(motif)%2Fi">motif</a> is a regex pattern without quantifiers. For example, the regex <code>[AT]TT[GC]C</code> may match to ATTGC, ATTCC, TTTGC, and TTTCC.</p>
<p>A common scenario involving <a href="#BM_(motif)%2Fi">motif</a>s is to search through a set of DNA <a href="#BM_(sequence)%2Fi">sequence</a>s for an unknown <a href="#BM_(motif)%2Fi">motif</a>: Given a set of <a href="#BM_(sequence)%2Fi">sequence</a>s, it's suspected that each <a href="#BM_(sequence)%2Fi">sequence</a> contains a <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> that matches some <a href="#BM_(motif)%2Fi">motif</a>. But, that <a href="#BM_(motif)%2Fi">motif</a> isn't known beforehand. Both the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s and the <a href="#BM_(motif)%2Fi">motif</a> they match need to be found.</p>
<p>For example, each of the following <a href="#BM_(sequence)%2Fi">sequence</a>s contains a <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> that matches some <a href="#BM_(motif)%2Fi">motif</a>:</p>
<table>
<thead>
<tr>
<th><a href="#BM_(sequence)%2Fi">Sequence</a>s</th>
</tr>
</thead>
<tbody>
<tr>
<td>ATTGTTACCATAACCTTATTGCTAG</td>
</tr>
<tr>
<td>ATTCCTTTAGGACCACCCCAAACCC</td>
</tr>
<tr>
<td>CCCCAGGAGGGAACCTTTGCACACA</td>
</tr>
<tr>
<td>TATATATTTCCCACCCCAAGGGGGG</td>
</tr>
</tbody>
</table>
<p>That <a href="#BM_(motif)%2Fi">motif</a> is the one described above (<code>[AT]TT[GC]C</code>):</p>
<table>
<thead>
<tr>
<th><a href="#BM_(sequence)%2Fi">Sequence</a>s</th>
</tr>
</thead>
<tbody>
<tr>
<td>ATTGTTACCATAACCTT<strong>ATTGC</strong>TAG</td>
</tr>
<tr>
<td><strong>ATTCC</strong>TTTAGGACCACCCCAAACCC</td>
</tr>
<tr>
<td>CCCCAGGAGGGAACC<strong>TTTGC</strong>ACACA</td>
</tr>
<tr>
<td>TATATA<strong>TTTCC</strong>CACCCCAAGGGGGG</td>
</tr>
</tbody>
</table>
<p>A <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> is a matrix of <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s where each <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> matches a <a href="#BM_(motif)%2Fi">motif</a>. In the example <a href="#BM_(sequence)%2Fi">sequence</a>s above, the <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> would be:</p>
<table>
<thead>
<tr>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
</tr>
</thead>
<tbody>
<tr>
<td>A</td>
<td>T</td>
<td>T</td>
<td>G</td>
<td>C</td>
</tr>
<tr>
<td>A</td>
<td>T</td>
<td>T</td>
<td>C</td>
<td>C</td>
</tr>
<tr>
<td>T</td>
<td>T</td>
<td>T</td>
<td>G</td>
<td>C</td>
</tr>
<tr>
<td>T</td>
<td>T</td>
<td>T</td>
<td>C</td>
<td>C</td>
</tr>
</tbody>
</table>
<p>A <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> that matches a <a href="#BM_(motif)%2Fi">motif</a> may be referred to as a <a href="#BM_(motif%20member)%2Fi">motif member</a>.</p>
<a name="H_Consensus%20String"></a><h3>Consensus String</h3>
<p><a name="BM_(Algorithms%5C%2FMotif%2FConsensus%20String)_TOPIC%2F"></a><strong></strong></p>
<p><strong>WHAT</strong>: Given a <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>, generate a <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> where each position is the nucleotide most abundant at that column of the matrix.</p>
<p><strong>WHY</strong>: Given a set of <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s that are suspected to be part of a <a href="#BM_(motif)%2Fi">motif</a> (<a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>), the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> generated by selecting the most abundant column at each index is the "ideal" <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> for the <a href="#BM_(motif)%2Fi">motif</a>. It's a concise way of describing the <a href="#BM_(motif)%2Fi">motif</a>, especially if the columns in the <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> are highly conserved.</p>
<p><strong>ALGORITHM</strong>:</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>It may be more appropriate to use a <a href="#BM_(hybrid%20alphabet%7Calternate%20alphabet%7Calternative%20alphabet)%2Fi">hybrid alphabet</a> when representing <a href="#BM_(consensus%20string%7Cconsensus%20sequence)%2Fi">consensus string</a> because alternate nucleotides could be represented as a single letter. The Pevzner book doesn't mention this specifically but multiple online sources discuss it.</p>
</div>
<p><a href="ch2_code/src/ConsensusString.py">ch2_code/src/ConsensusString.py</a> (lines 5 to 15):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">get_consensus_string</span>(<span class="hljs-params">kmers: <span class="hljs-type">List</span>[<span class="hljs-built_in">str</span>]</span>) -&gt; <span class="hljs-built_in">str</span>:
    count = <span class="hljs-built_in">len</span>(kmers[<span class="hljs-number">0</span>]);
    out = <span class="hljs-string">''</span>
    <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(<span class="hljs-number">0</span>, count):
        c = Counter()
        <span class="hljs-keyword">for</span> kmer <span class="hljs-keyword">in</span> kmers:
            c[kmer[i]] += <span class="hljs-number">1</span>
        ch = c.most_common(<span class="hljs-number">1</span>)
        out += ch[<span class="hljs-number">0</span>][<span class="hljs-number">0</span>]
    <span class="hljs-keyword">return</span> out
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Consensus is TTTCC in <br><br> ATTGC<br>ATTCC<br>TTTGC<br>TTTCC<br>TTTCA</p>
</div>
<p></p>
<a name="H_Motif%20Matrix%20Count"></a><h3>Motif Matrix Count</h3>
<p><a name="BM_(Algorithms%5C%2FMotif%2FMotif%20Matrix%20Count)_TOPIC%2F"></a><strong></strong></p>
<p><strong>WHAT</strong>: Given a <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>, count how many of each nucleotide are in each column.</p>
<p><strong>WHY</strong>: Having a count of the number of nucleotides in each column is a basic statistic that gets used further down the line for tasks such as scoring a <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>.</p>
<p><strong>ALGORITHM</strong>:</p>
<p><a href="ch2_code/src/MotifMatrixCount.py">ch2_code/src/MotifMatrixCount.py</a> (lines 7 to 21):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">motif_matrix_count</span>(<span class="hljs-params">motif_matrix: <span class="hljs-type">List</span>[<span class="hljs-built_in">str</span>], elements=<span class="hljs-string">'ACGT'</span></span>) -&gt; <span class="hljs-type">Dict</span>[<span class="hljs-built_in">str</span>, <span class="hljs-type">List</span>[<span class="hljs-built_in">int</span>]]:
    rows = <span class="hljs-built_in">len</span>(motif_matrix)
    cols = <span class="hljs-built_in">len</span>(motif_matrix[<span class="hljs-number">0</span>])

    ret = {}
    <span class="hljs-keyword">for</span> ch <span class="hljs-keyword">in</span> elements:
        ret[ch] = [<span class="hljs-number">0</span>] * cols
    
    <span class="hljs-keyword">for</span> c <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(<span class="hljs-number">0</span>, cols):
        <span class="hljs-keyword">for</span> r <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(<span class="hljs-number">0</span>, rows):
            item = motif_matrix[r][c]
            ret[item][c] += <span class="hljs-number">1</span>
            
    <span class="hljs-keyword">return</span> ret
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Counting nucleotides at each column of the motif matrix...</p>
<p>ATTGC<br>TTTGC<br>TTTGG<br>ATTGC</p>
<p>Result...</p>
<p>('A', [2, 0, 0, 0, 0])<br>('C', [0, 0, 0, 0, 3])<br>('G', [0, 0, 0, 4, 1])<br>('T', [2, 4, 4, 0, 0])</p>
</div>
<p></p>
<a name="H_Motif%20Matrix%20Profile"></a><h3>Motif Matrix Profile</h3>
<p><a name="BM_(Algorithms%5C%2FMotif%2FMotif%20Matrix%20Profile)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FMotif%2FMotif%20Matrix%20Count)_TOPIC%2F">Algorithms/Motif/Motif Matrix Count</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: Given a <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>, for each column calculate how often A, C, G, and T occur as percentages.</p>
<p><strong>WHY</strong>: The percentages for each column represent a probability distribution for that column. For example, in column 1 of...</p>
<table>
<thead>
<tr>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
</tr>
</thead>
<tbody>
<tr>
<td>A</td>
<td>T</td>
<td>T</td>
<td>C</td>
<td>G</td>
</tr>
<tr>
<td>C</td>
<td>T</td>
<td>T</td>
<td>C</td>
<td>G</td>
</tr>
<tr>
<td>T</td>
<td>T</td>
<td>T</td>
<td>C</td>
<td>G</td>
</tr>
<tr>
<td>T</td>
<td>T</td>
<td>T</td>
<td>T</td>
<td>G</td>
</tr>
</tbody>
</table>
<ul>
<li>A appears 25% of the time.</li>
<li>C appears 25% of the time.</li>
<li>T appears 50% of the time.</li>
<li>G appears 0% of the time.</li>
</ul>
<p>These probability distributions can be used further down the line for tasks such as determining the probability that some arbitrary <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> conforms to the same <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>.</p>
<p><strong>ALGORITHM</strong>:</p>
<p><a href="ch2_code/src/MotifMatrixProfile.py">ch2_code/src/MotifMatrixProfile.py</a> (lines 8 to 22):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">motif_matrix_profile</span>(<span class="hljs-params">motif_matrix_counts: <span class="hljs-type">Dict</span>[<span class="hljs-built_in">str</span>, <span class="hljs-type">List</span>[<span class="hljs-built_in">int</span>]]</span>) -&gt; <span class="hljs-type">Dict</span>[<span class="hljs-built_in">str</span>, <span class="hljs-type">List</span>[<span class="hljs-built_in">float</span>]]:
    ret = {}
    <span class="hljs-keyword">for</span> elem, counts <span class="hljs-keyword">in</span> motif_matrix_counts.items():
        ret[elem] = [<span class="hljs-number">0.0</span>] * <span class="hljs-built_in">len</span>(counts)

    cols = <span class="hljs-built_in">len</span>(counts)  <span class="hljs-comment"># all elems should have the same len, so just grab the last one that was walked over</span>
    <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(cols):
        total = <span class="hljs-number">0</span>
        <span class="hljs-keyword">for</span> elem <span class="hljs-keyword">in</span> motif_matrix_counts.keys():
            total += motif_matrix_counts[elem][i]
        <span class="hljs-keyword">for</span> elem <span class="hljs-keyword">in</span> motif_matrix_counts.keys():
            ret[elem][i] = motif_matrix_counts[elem][i] / total

    <span class="hljs-keyword">return</span> ret
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Profiling nucleotides at each column of the motif matrix...</p>
<p>ATTCG<br>CTTCG<br>TTTCG<br>TTTTG</p>
<p>Result...</p>
<p>('A', [0.25, 0.0, 0.0, 0.0, 0.0])<br>('C', [0.25, 0.0, 0.0, 0.75, 0.0])<br>('G', [0.0, 0.0, 0.0, 0.0, 1.0])<br>('T', [0.5, 1.0, 1.0, 0.25, 0.0])</p>
</div>
<p></p>
<a name="H_Motif%20Matrix%20Score"></a><h3>Motif Matrix Score</h3>
<p><a name="BM_(Algorithms%5C%2FMotif%2FMotif%20Matrix%20Score)_TOPIC%2F"></a><strong></strong></p>
<p><strong>WHAT</strong>: Given a <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>, assign it a score based on how similar the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s that make up the matrix are to each other. Specifically, how conserved the nucleotides at each column are.</p>
<p><strong>WHY</strong>: Given a set of <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s that are suspected to be part of a <a href="#BM_(motif)%2Fi">motif</a> (<a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>), the more similar those <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s are to each other the more likely it is that those <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s are <a href="#BM_%5Cb(membership%7Cmember)_MOTIF%2Fi">member</a>s of the same <a href="#BM_(motif)%2Fi">motif</a>. This seems to be the case for many enzymes that bind to DNA based on a <a href="#BM_(motif)%2Fi">motif</a> (e.g. <a href="#BM_(transcription%20factor)%2Fi">transcription factor</a>s).</p>
<a name="H_Popularity%20Algorithm"></a><h4>Popularity Algorithm</h4>
<p><strong>ALGORITHM</strong>:</p>
<p>This algorithm scores a <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> by summing up the number of unpopular items in a column. For example, imagine a column has 7 Ts, 2 Cs, and 1A. The Ts are the most popular (7 items), meaning that the 3 items (2 Cs and 1 A) are unpopular -- the score for the column is 3.</p>
<p>Sum up each of the column scores to the get the final score for the <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>. A lower score is better.</p>
<p><a href="ch2_code/src/ScoreMotif.py">ch2_code/src/ScoreMotif.py</a> (lines 17 to 39):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">score_motif</span>(<span class="hljs-params">motif_matrix: <span class="hljs-type">List</span>[<span class="hljs-built_in">str</span>]</span>) -&gt; <span class="hljs-built_in">int</span>:
    rows = <span class="hljs-built_in">len</span>(motif_matrix)
    cols = <span class="hljs-built_in">len</span>(motif_matrix[<span class="hljs-number">0</span>])

    <span class="hljs-comment"># count up each column</span>
    counter_per_col = []
    <span class="hljs-keyword">for</span> c <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(<span class="hljs-number">0</span>, cols):
        counter = Counter()
        <span class="hljs-keyword">for</span> r <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(<span class="hljs-number">0</span>, rows):
            counter[motif_matrix[r][c]] += <span class="hljs-number">1</span>
        counter_per_col.append(counter)

    <span class="hljs-comment"># sum counts for each column AFTER removing the top-most count -- that is, consider the top-most count as the</span>
    <span class="hljs-comment"># most popular char, so you're summing the counts of all the UNPOPULAR chars</span>
    unpopular_sums = []
    <span class="hljs-keyword">for</span> counter <span class="hljs-keyword">in</span> counter_per_col:
        most_popular_item = counter.most_common(<span class="hljs-number">1</span>)[<span class="hljs-number">0</span>][<span class="hljs-number">0</span>]
        <span class="hljs-keyword">del</span> counter[most_popular_item]
        unpopular_sum = <span class="hljs-built_in">sum</span>(counter.values())
        unpopular_sums.append(unpopular_sum)

    <span class="hljs-keyword">return</span> <span class="hljs-built_in">sum</span>(unpopular_sums)
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Scoring...</p>
<p>ATTGC<br>TTTGC<br>TTTGG<br>ATTGC</p>
<p>3</p>
</div>
<p></p>
<a name="H_Entropy%20Algorithm"></a><h4>Entropy Algorithm</h4>
<p><a name="BM_(Algorithms%5C%2FMotif%2FMotif%20Matrix%20Score%2FEntropy%20Algorithm)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FMotif%2FMotif%20Matrix%20Profile)_TOPIC%2F">Algorithms/Motif/Motif Matrix Profile</a></li>
</ul>
</div>
<p><strong>ALGORITHM</strong>:</p>
<p>This algorithm scores a <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> by calculating the <a href="#BM_(entropy)%2Fi">entropy</a> of each column in the <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>. <a href="#BM_(entropy)%2Fi">Entropy</a> is defined as the level of uncertainty for some variable. The more uncertain the nucleotides are in the column of a <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>, the higher (worse) the score. For example, given a <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> with 10 rows, a column with ...</p>
<ul>
<li>10 A nucleotides has low <a href="#BM_(entropy)%2Fi">entropy</a> because it's highly conserved,</li>
<li>6 A and 4 T nucleotides has a higher <a href="#BM_(entropy)%2Fi">entropy</a> because it's less highly conserved.</li>
</ul>
<p>Sum the output for each column to get the final score for the <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>. A lower score is better.</p>
<p><a href="ch2_code/src/ScoreMotifUsingEntropy.py">ch2_code/src/ScoreMotifUsingEntropy.py</a> (lines 10 to 38):</p>
<pre class="hljs"><code><span class="hljs-comment"># According to the book, method of scoring a motif matrix as defined in ScoreMotif.py isn't the method used in the</span>
<span class="hljs-comment"># real-world. The method used in the real-world is this method, where...</span>
<span class="hljs-comment"># 1. each column has its probability distribution calculated (prob of A vs prob C vs prob of T vs prob of G)</span>
<span class="hljs-comment"># 2. the entropy of each of those prob dist are calculated</span>
<span class="hljs-comment"># 3. those entropies are summed up to get the ENTROPY OF THE MOTIF MATRIX</span>
<span class="hljs-keyword">def</span> <span class="hljs-title function_">calculate_entropy</span>(<span class="hljs-params">values: <span class="hljs-type">List</span>[<span class="hljs-built_in">float</span>]</span>) -&gt; <span class="hljs-built_in">float</span>:
    ret = <span class="hljs-number">0.0</span>
    <span class="hljs-keyword">for</span> value <span class="hljs-keyword">in</span> values:
        ret += value * (log(value, <span class="hljs-number">2.0</span>) <span class="hljs-keyword">if</span> value &gt; <span class="hljs-number">0.0</span> <span class="hljs-keyword">else</span> <span class="hljs-number">0.0</span>)
    ret = -ret
    <span class="hljs-keyword">return</span> ret

<span class="hljs-keyword">def</span> <span class="hljs-title function_">score_motify_entropy</span>(<span class="hljs-params">motif_matrix: <span class="hljs-type">List</span>[<span class="hljs-built_in">str</span>]</span>) -&gt; <span class="hljs-built_in">float</span>:
    rows = <span class="hljs-built_in">len</span>(motif_matrix)
    cols = <span class="hljs-built_in">len</span>(motif_matrix[<span class="hljs-number">0</span>])

    <span class="hljs-comment"># count up each column</span>
    counts = motif_matrix_count(motif_matrix)
    profile = motif_matrix_profile(counts)

    <span class="hljs-comment"># prob dist to entropy</span>
    entropy_per_col = []
    <span class="hljs-keyword">for</span> c <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(cols):
        entropy = calculate_entropy([profile[<span class="hljs-string">'A'</span>][c], profile[<span class="hljs-string">'C'</span>][c], profile[<span class="hljs-string">'G'</span>][c], profile[<span class="hljs-string">'T'</span>][c]])
        entropy_per_col.append(entropy)

    <span class="hljs-comment"># sum up entropies to get entropy of motif</span>
    <span class="hljs-keyword">return</span> <span class="hljs-built_in">sum</span>(entropy_per_col)
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Scoring...</p>
<p>ATTGC<br>TTTGC<br>TTTGG<br>ATTGC</p>
<p>1.811278124459133</p>
</div>
<p></p>
<a name="H_Relative%20Entropy%20Algorithm"></a><h4>Relative Entropy Algorithm</h4>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FMotif%2FMotif%20Matrix%20Score%2FEntropy%20Algorithm)_TOPIC%2F">Algorithms/Motif/Motif Matrix Score/Entropy Algorithm</a></li>
</ul>
</div>
<p><strong>ALGORITHM</strong>:</p>
<p>This algorithm scores a <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> by calculating the <a href="#BM_(entropy)%2Fi">entropy</a> of each column relative to the overall nucleotide distribution of the <a href="#BM_(sequence)%2Fi">sequence</a>s from which each <a href="#BM_(motif%20member)%2Fi">motif member</a> came from. This is important when finding <a href="#BM_(motif%20member)%2Fi">motif member</a>s across a set of <a href="#BM_(sequence)%2Fi">sequence</a>s. For example, the following <a href="#BM_(sequence)%2Fi">sequence</a>s have a nucleotide distribution highly skewed towards C...</p>
<table>
<thead>
<tr>
<th><a href="#BM_(sequence)%2Fi">Sequence</a>s</th>
</tr>
</thead>
<tbody>
<tr>
<td>CCCCCCCCCCCCCCCCCATTGCCCC</td>
</tr>
<tr>
<td>ATTCCCCCCCCCCCCCCCCCCCCCC</td>
</tr>
<tr>
<td>CCCCCCCCCCCCCCCTTTGCCCCCC</td>
</tr>
<tr>
<td>CCCCCCTTTCTCCCCCCCCCCCCCC</td>
</tr>
</tbody>
</table>
<p>Given the <a href="#BM_(sequence)%2Fi">sequence</a>s in the example above, of all <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrices</a> possible for k=5, basic <a href="#BM_(entropy)%2Fi">entropy</a> scoring will always lead to a matrix filled with Cs:</p>
<table>
<thead>
<tr>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
</tr>
</thead>
<tbody>
<tr>
<td>C</td>
<td>C</td>
<td>C</td>
<td>C</td>
<td>C</td>
</tr>
<tr>
<td>C</td>
<td>C</td>
<td>C</td>
<td>C</td>
<td>C</td>
</tr>
<tr>
<td>C</td>
<td>C</td>
<td>C</td>
<td>C</td>
<td>C</td>
</tr>
<tr>
<td>C</td>
<td>C</td>
<td>C</td>
<td>C</td>
<td>C</td>
</tr>
</tbody>
</table>
<p>Even though the above <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> scores perfect, it's likely junk. <a href="#BM_%5Cb(membership%7Cmember)_MOTIF%2Fi">Member</a>s containing all Cs score better because the <a href="#BM_(sequence)%2Fi">sequence</a>s they come from are biased (saturated with Cs), not because they share some higher biological significance.</p>
<p>To reduce bias, the nucleotide distributions from which the <a href="#BM_%5Cb(membership%7Cmember)_MOTIF%2Fi">member</a>s came from need to be factored in to the <a href="#BM_(entropy)%2Fi">entropy</a> calculation: relative <a href="#BM_(entropy)%2Fi">entropy</a>.</p>
<p><a href="ch2_code/src/ScoreMotifUsingRelativeEntropy.py">ch2_code/src/ScoreMotifUsingRelativeEntropy.py</a> (lines 10 to 84):</p>
<pre class="hljs"><code><span class="hljs-comment"># <span class="hljs-doctag">NOTE:</span> This is different from the traditional version of entropy -- it doesn't negate the sum before returning it.</span>
<span class="hljs-keyword">def</span> <span class="hljs-title function_">calculate_entropy</span>(<span class="hljs-params">probabilities_for_nuc: <span class="hljs-type">List</span>[<span class="hljs-built_in">float</span>]</span>) -&gt; <span class="hljs-built_in">float</span>:
    ret = <span class="hljs-number">0.0</span>
    <span class="hljs-keyword">for</span> value <span class="hljs-keyword">in</span> probabilities_for_nuc:
        ret += value * (log(value, <span class="hljs-number">2.0</span>) <span class="hljs-keyword">if</span> value &gt; <span class="hljs-number">0.0</span> <span class="hljs-keyword">else</span> <span class="hljs-number">0.0</span>)
    <span class="hljs-keyword">return</span> ret


<span class="hljs-keyword">def</span> <span class="hljs-title function_">calculate_cross_entropy</span>(<span class="hljs-params">probabilities_for_nuc: <span class="hljs-type">List</span>[<span class="hljs-built_in">float</span>], total_frequencies_for_nucs: <span class="hljs-type">List</span>[<span class="hljs-built_in">float</span>]</span>) -&gt; <span class="hljs-built_in">float</span>:
    ret = <span class="hljs-number">0.0</span>
    <span class="hljs-keyword">for</span> prob, total_freq <span class="hljs-keyword">in</span> <span class="hljs-built_in">zip</span>(probabilities_for_nuc, total_frequencies_for_nucs):
        ret += prob * (log(total_freq, <span class="hljs-number">2.0</span>) <span class="hljs-keyword">if</span> total_freq &gt; <span class="hljs-number">0.0</span> <span class="hljs-keyword">else</span> <span class="hljs-number">0.0</span>)
    <span class="hljs-keyword">return</span> ret


<span class="hljs-keyword">def</span> <span class="hljs-title function_">score_motif_relative_entropy</span>(<span class="hljs-params">motif_matrix: <span class="hljs-type">List</span>[<span class="hljs-built_in">str</span>], source_strs: <span class="hljs-type">List</span>[<span class="hljs-built_in">str</span>]</span>) -&gt; <span class="hljs-built_in">float</span>:
    <span class="hljs-comment"># calculate frequency of nucleotide across all source strings</span>
    nuc_counter = Counter()
    nuc_total = <span class="hljs-number">0</span>
    <span class="hljs-keyword">for</span> source_str <span class="hljs-keyword">in</span> source_strs:
        <span class="hljs-keyword">for</span> nuc <span class="hljs-keyword">in</span> source_str:
            nuc_counter[nuc] += <span class="hljs-number">1</span>
        nuc_total += <span class="hljs-built_in">len</span>(source_str)
    nuc_freqs = <span class="hljs-built_in">dict</span>([(k, v / nuc_total) <span class="hljs-keyword">for</span> k, v <span class="hljs-keyword">in</span> nuc_counter.items()])

    rows = <span class="hljs-built_in">len</span>(motif_matrix)
    cols = <span class="hljs-built_in">len</span>(motif_matrix[<span class="hljs-number">0</span>])

    <span class="hljs-comment"># count up each column</span>
    counts = motif_matrix_count(motif_matrix)
    profile = motif_matrix_profile(counts)
    relative_entropy_per_col = []
    <span class="hljs-keyword">for</span> c <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(cols):
        <span class="hljs-comment"># get entropy of column in motif</span>
        entropy = calculate_entropy(
            [
                profile[<span class="hljs-string">'A'</span>][c],
                profile[<span class="hljs-string">'C'</span>][c],
                profile[<span class="hljs-string">'G'</span>][c],
                profile[<span class="hljs-string">'T'</span>][c]
            ]
        )
        <span class="hljs-comment"># get cross entropy of column in motif (mixes in global nucleotide frequencies)</span>
        cross_entropy = calculate_cross_entropy(
            [
                profile[<span class="hljs-string">'A'</span>][c],
                profile[<span class="hljs-string">'C'</span>][c],
                profile[<span class="hljs-string">'G'</span>][c],
                profile[<span class="hljs-string">'T'</span>][c]
            ],
            [
                nuc_freqs[<span class="hljs-string">'A'</span>],
                nuc_freqs[<span class="hljs-string">'C'</span>],
                nuc_freqs[<span class="hljs-string">'G'</span>],
                nuc_freqs[<span class="hljs-string">'T'</span>]
            ]
        )
        relative_entropy = entropy - cross_entropy
        <span class="hljs-comment"># Right now relative_entropy is calculated by subtracting cross_entropy from (a negated) entropy. But, according</span>
        <span class="hljs-comment"># to the Pevzner book, the calculation of relative_entropy can be simplified to just...</span>
        <span class="hljs-comment"># def calculate_relative_entropy(probabilities_for_nuc: List[float], total_frequencies_for_nucs: List[float]) -&gt; float:</span>
        <span class="hljs-comment">#     ret = 0.0</span>
        <span class="hljs-comment">#     for prob, total_freq in zip(probabilities_for_nuc, total_frequency_for_nucs):</span>
        <span class="hljs-comment">#         ret += value * (log(value / total_freq, 2.0) if value &gt; 0.0 else 0.0)</span>
        <span class="hljs-comment">#     return ret</span>
        relative_entropy_per_col.append(relative_entropy)

    <span class="hljs-comment"># sum up entropies to get entropy of motif</span>
    ret = <span class="hljs-built_in">sum</span>(relative_entropy_per_col)

    <span class="hljs-comment"># All of the other score_motif algorithms try to MINIMIZE score. In the case of relative entropy (this algorithm),</span>
    <span class="hljs-comment"># the greater the score is the better of a match it is. As such, negate this score so the existing algorithms can</span>
    <span class="hljs-comment"># still try to minimize.</span>
    <span class="hljs-keyword">return</span> -ret
</code></pre>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>In the outputs below, the score in the second output should be less than (better) the score in the first output.</p>
</div>
<div style="border:1px solid black;">
<p></p>
<p>Scoring...</p>
<p>CCCCC<br>CCCCC<br>CCCCC<br>CCCCC</p>
<p>... which was pulled from ...</p>
<p>CCCCCCCCCCCCCCCCCATTGCCCC<br>ATTCCCCCCCCCCCCCCCCCCCCCC<br>CCCCCCCCCCCCCCCTTTGCCCCCC<br>CCCCCCTTTCTCCCCCCCCCCCCCC</p>
<p>-1.172326268185115</p>
</div>
<p></p>
<div style="border:1px solid black;">
<p></p>
<p>Scoring...</p>
<p>ATTGC<br>ATTCC<br>CTTTG<br>TTTCT</p>
<p>... which was pulled from ...</p>
<p>CCCCCCCCCCCCCCCCCATTGCCCC<br>ATTCCCCCCCCCCCCCCCCCCCCCC<br>CCCCCCCCCCCCCCCTTTGCCCCCC<br>CCCCCCTTTCTCCCCCCCCCCCCCC</p>
<p>-10.194105327448927</p>
</div>
<p></p>
<a name="H_Motif%20Logo"></a><h3>Motif Logo</h3>
<p><a name="BM_(Algorithms%5C%2FMotif%2FMotif%20Logo)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FMotif%2FMotif%20Matrix%20Score%2FEntropy%20Algorithm)_TOPIC%2F">Algorithms/Motif/Motif Matrix Score/Entropy Algorithm</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: Given a <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>, generate a graphical representation showing how conserved the <a href="#BM_(motif)%2Fi">motif</a> is. Each position has its possible nucleotides stacked on top of each other, where the height of each nucleotide is based on how conserved it is. The more conserved a position is, the taller that column will be. This type of graphical representation is called a <a href="#BM_(%5Cblogo%7Csequence%20logo)%2Fi">sequence logo</a>.</p>
<p><strong>WHY</strong>: A <a href="#BM_(%5Cblogo%7Csequence%20logo)%2Fi">sequence logo</a> helps more quickly convey the characteristics of the <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> it's for.</p>
<p><strong>ALGORITHM</strong>:</p>
<p>For this particular <a href="#BM_(%5Cblogo%7Csequence%20logo)%2Fi">logo</a> implementation, a lower <a href="#BM_(entropy)%2Fi">entropy</a> results in a taller overall column.</p>
<p><a href="ch2_code/src/MotifLogo.py">ch2_code/src/MotifLogo.py</a> (lines 15 to 39):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">calculate_entropy</span>(<span class="hljs-params">values: <span class="hljs-type">List</span>[<span class="hljs-built_in">float</span>]</span>) -&gt; <span class="hljs-built_in">float</span>:
    ret = <span class="hljs-number">0.0</span>
    <span class="hljs-keyword">for</span> value <span class="hljs-keyword">in</span> values:
        ret += value * (log(value, <span class="hljs-number">2.0</span>) <span class="hljs-keyword">if</span> value &gt; <span class="hljs-number">0.0</span> <span class="hljs-keyword">else</span> <span class="hljs-number">0.0</span>)
    ret = -ret
    <span class="hljs-keyword">return</span> ret

<span class="hljs-keyword">def</span> <span class="hljs-title function_">create_logo</span>(<span class="hljs-params">motif_matrix_profile: <span class="hljs-type">Dict</span>[<span class="hljs-built_in">str</span>, <span class="hljs-type">List</span>[<span class="hljs-built_in">float</span>]]</span>) -&gt; Logo:
    columns = <span class="hljs-built_in">list</span>(motif_matrix_profile.keys())
    data = [motif_matrix_profile[k] <span class="hljs-keyword">for</span> k <span class="hljs-keyword">in</span> motif_matrix_profile.keys()]
    data = <span class="hljs-built_in">list</span>(<span class="hljs-built_in">zip</span>(*data))  <span class="hljs-comment"># trick to transpose data</span>

    entropies = <span class="hljs-built_in">list</span>(<span class="hljs-built_in">map</span>(<span class="hljs-keyword">lambda</span> x: <span class="hljs-number">2</span> - calculate_entropy(x), data))

    data_scaledby_entropies = [[p * e <span class="hljs-keyword">for</span> p <span class="hljs-keyword">in</span> d] <span class="hljs-keyword">for</span> d, e <span class="hljs-keyword">in</span> <span class="hljs-built_in">zip</span>(data, entropies)]

    df = pd.DataFrame(
        columns=columns,
        data=data_scaledby_entropies
    )
    logo = lm.Logo(df)
    logo.ax.set_ylabel(<span class="hljs-string">'information (bits)'</span>)
    logo.ax.set_xlim([-<span class="hljs-number">1</span>, <span class="hljs-built_in">len</span>(df)])
    <span class="hljs-keyword">return</span> logo
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Generating logo for the following motif matrix...</p>
<p>TCGGGGGTTTTT<br>CCGGTGACTTAC<br>ACGGGGATTTTC<br>TTGGGGACTTTT<br>AAGGGGACTTCC<br>TTGGGGACTTCC<br>TCGGGGATTCAT<br>TCGGGGATTCCT<br>TAGGGGAACTAC<br>TCGGGTATAACC</p>
<p>Result...</p>
<p><img src="motif_logo_01b7e8763a84973751af52163d52f5bd.svg" alt="Motif Logo"></p>
</div>
<p></p>
<a name="H_K-mer%20Match%20Probability"></a><h3>K-mer Match Probability</h3>
<p><a name="BM_(Algorithms%5C%2FMotif%2FK-mer%20Match%20Probability)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FMotif%2FMotif%20Matrix%20Count)_TOPIC%2F">Algorithms/Motif/Motif Matrix Count</a></li>
<li><a href="#BM_(Algorithms%5C%2FMotif%2FMotif%20Matrix%20Profile)_TOPIC%2F">Algorithms/Motif/Motif Matrix Profile</a></li>
<li><a href="#BM_(Algorithms%5C%2FK-mer)_TOPIC%2F">Algorithms/K-mer</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: Given a <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> and a <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>, calculate the probability of that <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> being a <a href="#BM_%5Cb(membership%7Cmember)_MOTIF%2Fi">member</a> of that <a href="#BM_(motif)%2Fi">motif</a>.</p>
<p><strong>WHY</strong>: Being able to determine if a <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> is potentially a <a href="#BM_%5Cb(membership%7Cmember)_MOTIF%2Fi">member</a> of a <a href="#BM_(motif)%2Fi">motif</a> can help speed up experiments. For example, imagine that you suspect 21 different <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">genes</a> of being regulated by the same <a href="#BM_(transcription%20factor)%2Fi">transcription factor</a>. You isolate the <a href="#BM_(transcription%20factor%20binding%20site)%2Fi">transcription factor binding site</a> for 6 of those <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">genes</a> and use their <a href="#BM_(sequence)%2Fi">sequence</a>s as the underlying <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s for a <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>. That <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> doesn't represent the <a href="#BM_(transcription%20factor)%2Fi">transcription factor</a>'s <a href="#BM_(motif)%2Fi">motif</a> exactly, but it's close enough that you can use it to scan through the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s in the remaining 15 <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">genes</a> and calculate the probability of them being <a href="#BM_%5Cb(membership%7Cmember)_MOTIF%2Fi">member</a>s of the same <a href="#BM_(motif)%2Fi">motif</a>.</p>
<p>If a <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> exists such that it conforms to the <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> with a high probability, it likely is a <a href="#BM_%5Cb(membership%7Cmember)_MOTIF%2Fi">member</a> of the <a href="#BM_(motif)%2Fi">motif</a>.</p>
<p><strong>ALGORITHM</strong>:</p>
<p>Imagine the following <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>:</p>
<table>
<thead>
<tr>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
</tr>
</thead>
<tbody>
<tr>
<td>A</td>
<td>T</td>
<td>G</td>
<td>C</td>
<td>A</td>
<td>C</td>
</tr>
<tr>
<td>A</td>
<td>T</td>
<td>G</td>
<td>C</td>
<td>A</td>
<td>C</td>
</tr>
<tr>
<td>A</td>
<td>T</td>
<td>C</td>
<td>C</td>
<td>A</td>
<td>C</td>
</tr>
<tr>
<td>A</td>
<td>T</td>
<td>C</td>
<td>C</td>
<td>A</td>
<td>C</td>
</tr>
</tbody>
</table>
<p>Calculating the counts for that <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> results in:</p>
<table>
<thead>
<tr>
<th></th>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
</tr>
</thead>
<tbody>
<tr>
<td>A</td>
<td>4</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>4</td>
<td>0</td>
</tr>
<tr>
<td>C</td>
<td>0</td>
<td>0</td>
<td>2</td>
<td>4</td>
<td>0</td>
<td>4</td>
</tr>
<tr>
<td>T</td>
<td>0</td>
<td>4</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
<tr>
<td>G</td>
<td>0</td>
<td>0</td>
<td>2</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
</tbody>
</table>
<p>Calculating the profile from those counts results in:</p>
<table>
<thead>
<tr>
<th></th>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
</tr>
</thead>
<tbody>
<tr>
<td>A</td>
<td>1</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>1</td>
<td>0</td>
</tr>
<tr>
<td>C</td>
<td>0</td>
<td>0</td>
<td>0.5</td>
<td>1</td>
<td>0</td>
<td>1</td>
</tr>
<tr>
<td>T</td>
<td>0</td>
<td>1</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
<tr>
<td>G</td>
<td>0</td>
<td>0</td>
<td>0.5</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
</tbody>
</table>
<p>Using this profile, the probability that a <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> conforms to the <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> is calculated by mapping the nucleotide at each position of the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> to the corresponding nucleotide in the corresponding position of the profile and multiplying them together. For example, the probability that the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>...</p>
<ul>
<li>ATGCAC conforms to the example profile above is calculated as 1*1*0.5*1*1*1 = 0.5</li>
<li>TTGCAC conforms to the example profile above is calculated as 0*1*0.5*1*1*1 = 0</li>
</ul>
<p>Of these two <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s, ...</p>
<ul>
<li>all positions in the first (ATGCAC) have been seen before in the <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>.</li>
<li>all but one position in the second (TTGCAC) have been seen before in the <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> (index 0).</li>
</ul>
<p>Both of these <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s should have a reasonable probability of being <a href="#BM_%5Cb(membership%7Cmember)_MOTIF%2Fi">member</a>s of the <a href="#BM_(motif)%2Fi">motif</a>. However, notice how the second <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> ends up with a 0 probability. The reason has to do with the underlying concept behind <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrices</a>: the entire point of a <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> is to use the known <a href="#BM_%5Cb(membership%7Cmember)_MOTIF%2Fi">member</a>s of a <a href="#BM_(motif)%2Fi">motif</a> to find other potential <a href="#BM_%5Cb(membership%7Cmember)_MOTIF%2Fi">member</a>s of that same <a href="#BM_(motif)%2Fi">motif</a>. The second <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> contains a T at index 0, but none of the known <a href="#BM_%5Cb(membership%7Cmember)_MOTIF%2Fi">member</a>s of the <a href="#BM_(motif)%2Fi">motif</a> have a T at that index. As such, its probability gets reduced to 0 even though the rest of the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> conforms.</p>
<p><a href="#BM_(Cromwell's%20rule)%2Fi">Cromwell's rule</a> says that when a probability is based off past events, a hard 0 or 1 values shouldn't be used. As such, a quick workaround to the 0% probability problem described above is to artificially inflate the counts that lead to the profile such that no count is 0 (<a href="#BM_(pseudocount)%2Fi">pseudocount</a>s). For example, for the same <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>, incrementing the counts by 1 results in:</p>
<table>
<thead>
<tr>
<th></th>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
</tr>
</thead>
<tbody>
<tr>
<td>A</td>
<td>5</td>
<td>1</td>
<td>1</td>
<td>1</td>
<td>5</td>
<td>1</td>
</tr>
<tr>
<td>C</td>
<td>1</td>
<td>1</td>
<td>3</td>
<td>5</td>
<td>1</td>
<td>5</td>
</tr>
<tr>
<td>T</td>
<td>1</td>
<td>5</td>
<td>1</td>
<td>1</td>
<td>1</td>
<td>1</td>
</tr>
<tr>
<td>G</td>
<td>1</td>
<td>1</td>
<td>3</td>
<td>1</td>
<td>1</td>
<td>1</td>
</tr>
</tbody>
</table>
<p>Calculating the profile from those inflated counts results in:</p>
<table>
<thead>
<tr>
<th></th>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
</tr>
</thead>
<tbody>
<tr>
<td>A</td>
<td>0.625</td>
<td>0.125</td>
<td>0.125</td>
<td>0.125</td>
<td>0.625</td>
<td>0.125</td>
</tr>
<tr>
<td>C</td>
<td>0.125</td>
<td>0.125</td>
<td>0.375</td>
<td>0.625</td>
<td>0.125</td>
<td>0.625</td>
</tr>
<tr>
<td>T</td>
<td>0.125</td>
<td>0.625</td>
<td>0.125</td>
<td>0.125</td>
<td>0.125</td>
<td>0.125</td>
</tr>
<tr>
<td>G</td>
<td>0.125</td>
<td>0.125</td>
<td>0.375</td>
<td>0.125</td>
<td>0.125</td>
<td>0.125</td>
</tr>
</tbody>
</table>
<p>Using this new profile, the probability that the previous <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s conform are:</p>
<ul>
<li>ATGCAC is calculated as 0.625*0.625*0.325*0.625*0.625*0.625 = 0.031</li>
<li>TTGCAC is calculated as 0.125*0.625*0.325*0.625*0.625*0.625 = 0.0062</li>
</ul>
<p>Although the probabilities seem low, it's all relative. The probability calculated for the first <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> (ATGCAC) is the highest probability possible -- each position in the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> maps to the highest probability nucleotide of the corresponding position of the profile.</p>
<p><a href="ch2_code/src/FindMostProbableKmerUsingProfileMatrix.py">ch2_code/src/FindMostProbableKmerUsingProfileMatrix.py</a> (lines 9 to 46):</p>
<pre class="hljs"><code><span class="hljs-comment"># Run this on the counts before generating the profile to avoid the 0 probability problem.</span>
<span class="hljs-keyword">def</span> <span class="hljs-title function_">apply_psuedocounts_to_count_matrix</span>(<span class="hljs-params">counts: <span class="hljs-type">Dict</span>[<span class="hljs-built_in">str</span>, <span class="hljs-type">List</span>[<span class="hljs-built_in">int</span>]], extra_count: <span class="hljs-built_in">int</span> = <span class="hljs-number">1</span></span>):
    <span class="hljs-keyword">for</span> elem, elem_counts <span class="hljs-keyword">in</span> counts.items():
        <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(<span class="hljs-built_in">len</span>(elem_counts)):
            elem_counts[i] += extra_count


<span class="hljs-comment"># Recall that a profile matrix is a matrix of probabilities. Each row represents a single element (e.g. nucleotide) and</span>
<span class="hljs-comment"># each column represents the probability distribution for that position.</span>
<span class="hljs-comment">#</span>
<span class="hljs-comment"># So for example, imagine the following probability distribution...</span>
<span class="hljs-comment">#</span>
<span class="hljs-comment">#     1   2   3   4</span>
<span class="hljs-comment"># A: 0.2 0.2 0.0 0.0</span>
<span class="hljs-comment"># C: 0.1 0.6 0.0 0.0</span>
<span class="hljs-comment"># G: 0.1 0.0 1.0 1.0</span>
<span class="hljs-comment"># T: 0.7 0.2 0.0 0.0</span>
<span class="hljs-comment">#</span>
<span class="hljs-comment"># At position 2, the probability that the element will be C is 0.6 while the probability that it'll be T is 0.2. Note</span>
<span class="hljs-comment"># how each column sums to 1.</span>
<span class="hljs-keyword">def</span> <span class="hljs-title function_">determine_probability_of_match_using_profile_matrix</span>(<span class="hljs-params">profile: <span class="hljs-type">Dict</span>[<span class="hljs-built_in">str</span>, <span class="hljs-type">List</span>[<span class="hljs-built_in">float</span>]], kmer: <span class="hljs-built_in">str</span></span>):
    prob = <span class="hljs-number">1.0</span>
    <span class="hljs-keyword">for</span> idx, elem <span class="hljs-keyword">in</span> <span class="hljs-built_in">enumerate</span>(kmer):
        prob = prob * profile[elem][idx]
    <span class="hljs-keyword">return</span> prob


<span class="hljs-keyword">def</span> <span class="hljs-title function_">find_most_probable_kmer_using_profile_matrix</span>(<span class="hljs-params">profile: <span class="hljs-type">Dict</span>[<span class="hljs-built_in">str</span>, <span class="hljs-type">List</span>[<span class="hljs-built_in">float</span>]], dna: <span class="hljs-built_in">str</span></span>):
    k = <span class="hljs-built_in">len</span>(<span class="hljs-built_in">list</span>(profile.values())[<span class="hljs-number">0</span>])

    most_probable: <span class="hljs-type">Tuple</span>[<span class="hljs-built_in">str</span>, <span class="hljs-built_in">float</span>] = <span class="hljs-literal">None</span>  <span class="hljs-comment"># [kmer, probability]</span>
    <span class="hljs-keyword">for</span> kmer, _ <span class="hljs-keyword">in</span> slide_window(dna, k):
        prob = determine_probability_of_match_using_profile_matrix(profile, kmer)
        <span class="hljs-keyword">if</span> most_probable <span class="hljs-keyword">is</span> <span class="hljs-literal">None</span> <span class="hljs-keyword">or</span> prob &gt; most_probable[<span class="hljs-number">1</span>]:
            most_probable = (kmer, prob)

    <span class="hljs-keyword">return</span> most_probable
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Motif matrix...</p>
<p>ATGCAC<br>ATGCAC<br>ATCCAC</p>
<p>Probability that TTGCAC matches the motif 0.0...</p>
</div>
<p></p>
<a name="H_Find%20Motif%20Matrix"></a><h3>Find Motif Matrix</h3>
<p><a name="BM_(Algorithms%5C%2FMotif%2FFind%20Motif%20Matrix)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FMotif%2FK-mer%20Match%20Probability)_TOPIC%2F">Algorithms/Motif/K-mer Match Probability</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: Given a set of <a href="#BM_(sequence)%2Fi">sequence</a>s, find <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s in those <a href="#BM_(sequence)%2Fi">sequence</a>s that may be <a href="#BM_%5Cb(membership%7Cmember)_MOTIF%2Fi">member</a>s of the same <a href="#BM_(motif)%2Fi">motif</a>.</p>
<p><strong>WHY</strong>: A <a href="#BM_(transcription%20factor)%2Fi">transcription factor</a> is an enzyme that either increases or decreases a <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a>'s <a href="#BM_(transcription%7Ctranscribed%7Ctranscribe)%2Fi">transcription</a> rate. It does so by binding to a specific part of the <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a>'s <a href="#BM_(upstream%20region)%2Fi">upstream region</a> called the <a href="#BM_(transcription%20factor%20binding%20site)%2Fi">transcription factor binding site</a>. That <a href="#BM_(transcription%20factor%20binding%20site)%2Fi">transcription factor binding site</a> consists of a <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> that matches the <a href="#BM_(motif)%2Fi">motif</a> expected by that <a href="#BM_(transcription%20factor)%2Fi">transcription factor</a>, called a <a href="#BM_(regulatory%20motif)%2Fi">regulatory motif</a>.</p>
<p>A single <a href="#BM_(transcription%20factor)%2Fi">transcription factor</a> may operate on many different <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">genes</a>. Often times a scientist will identify a set of <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">genes</a> that are suspected to be regulated by a single <a href="#BM_(transcription%20factor)%2Fi">transcription factor</a>, but that scientist won't know ...</p>
<ul>
<li>what the <a href="#BM_(regulatory%20motif)%2Fi">regulatory motif</a> is (the pattern expected by the enzyme).</li>
<li>where the <a href="#BM_(transcription%20factor%20binding%20site)%2Fi">transcription factor binding site</a>s are (which <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s the enzyme is targeting).</li>
<li>how long the <a href="#BM_(transcription%20factor%20binding%20site)%2Fi">transcription factor binding site</a>s are (which k the enzyme is targeting).</li>
</ul>
<p>The <a href="#BM_(regulatory%20motif)%2Fi">regulatory motif</a> expected by a <a href="#BM_(transcription%20factor)%2Fi">transcription factor</a> typically expects <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s that have the same length and are similar to each other (short <a href="#BM_(hamming%20distance)%2Fi">hamming distance</a>). As such, potential <a href="#BM_(motif)%2Fi">motif</a> candidates can be derived by finding <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s across the set of <a href="#BM_(sequence)%2Fi">sequence</a>s that are similar to each other.</p>
<a name="H1_Bruteforce%20Algorithm"></a><h4>Bruteforce Algorithm</h4>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FK-mer%2FHamming%20Distance%20Neighbourhood)_TOPIC%2F">Algorithms/K-mer/Hamming Distance Neighbourhood</a></li>
<li><a href="#BM_(Algorithms%5C%2FMotif%2FMotif%20Matrix%20Score)_TOPIC%2F">Algorithms/Motif/Motif Matrix Score</a></li>
</ul>
</div>
<p><strong>ALGORITHM</strong>:</p>
<p>This algorithm scans over all <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s in a set of DNA <a href="#BM_(sequence)%2Fi">sequence</a>s, enumerates the <a href="#BM_(hamming%20distance)%2Fi">hamming distance</a> <a href="#BM_(neighbour)%2Fi">neighbour</a>hood of each <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>, and uses the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s from the <a href="#BM_(hamming%20distance)%2Fi">hamming distance</a> <a href="#BM_(neighbour)%2Fi">neighbour</a>hood to build out possible <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrices</a>. Of all the <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrices</a> built, it selects the one with the lowest score.</p>
<p>Neither k nor the mismatches allowed by the <a href="#BM_(motif)%2Fi">motif</a> is known. As such, the algorithm may need to be repeated multiple times with different value combinations.</p>
<p>Even for trivial inputs, this algorithm falls over very quickly. It's intended to help conceptualize the problem of <a href="#BM_(motif)%2Fi">motif</a> finding.</p>
<p><a href="ch2_code/src/ExhaustiveMotifMatrixSearch.py">ch2_code/src/ExhaustiveMotifMatrixSearch.py</a> (lines 9 to 41):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">enumerate_hamming_distance_neighbourhood_for_all_kmer</span>(<span class="hljs-params">
        dna: <span class="hljs-built_in">str</span>,             <span class="hljs-comment"># dna strings to search in for motif</span>
        k: <span class="hljs-built_in">int</span>,               <span class="hljs-comment"># k-mer length</span>
        max_mismatches: <span class="hljs-built_in">int</span>   <span class="hljs-comment"># max num of mismatches for motif (hamming dist)</span>
</span>) -&gt; <span class="hljs-type">Set</span>[<span class="hljs-built_in">str</span>]:
    kmers_to_check = <span class="hljs-built_in">set</span>()
    <span class="hljs-keyword">for</span> kmer, _ <span class="hljs-keyword">in</span> slide_window(dna, k):
        neighbouring_kmers = find_all_dna_kmers_within_hamming_distance(kmer, max_mismatches)
        kmers_to_check |= neighbouring_kmers

    <span class="hljs-keyword">return</span> kmers_to_check


<span class="hljs-keyword">def</span> <span class="hljs-title function_">exhaustive_motif_search</span>(<span class="hljs-params">dnas: <span class="hljs-type">List</span>[<span class="hljs-built_in">str</span>], k: <span class="hljs-built_in">int</span>, max_mismatches: <span class="hljs-built_in">int</span></span>):
    kmers_for_dnas = [enumerate_hamming_distance_neighbourhood_for_all_kmer(dna, k, max_mismatches) <span class="hljs-keyword">for</span> dna <span class="hljs-keyword">in</span> dnas]

    <span class="hljs-keyword">def</span> <span class="hljs-title function_">build_next_matrix</span>(<span class="hljs-params">out_matrix: <span class="hljs-type">List</span>[<span class="hljs-built_in">str</span>]</span>):
        idx = <span class="hljs-built_in">len</span>(out_matrix)
        <span class="hljs-keyword">if</span> <span class="hljs-built_in">len</span>(kmers_for_dnas) == idx:
            <span class="hljs-keyword">yield</span> out_matrix[:]
        <span class="hljs-keyword">else</span>:
            <span class="hljs-keyword">for</span> kmer <span class="hljs-keyword">in</span> kmers_for_dnas[idx]:
                out_matrix.append(kmer)
                <span class="hljs-keyword">yield</span> <span class="hljs-keyword">from</span> build_next_matrix(out_matrix)
                out_matrix.pop()

    best_motif_matrix = <span class="hljs-literal">None</span>
    <span class="hljs-keyword">for</span> next_motif_matrix <span class="hljs-keyword">in</span> build_next_matrix([]):
        <span class="hljs-keyword">if</span> best_motif_matrix <span class="hljs-keyword">is</span> <span class="hljs-literal">None</span> <span class="hljs-keyword">or</span> score_motif(next_motif_matrix) &lt; score_motif(best_motif_matrix):
            best_motif_matrix = next_motif_matrix

    <span class="hljs-keyword">return</span> best_motif_matrix
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Searching for motif of k=5 and a max of 1 mismatches in the following...</p>
<p>ATAAAGGGATA<br>ACAGAAATGAT<br>TGAAATAACCT</p>
<p>Found the motif matrix...</p>
<p>ATAAT<br>ATAAT<br>ATAAT</p>
</div>
<p></p>
<a name="H_Median%20String%20Algorithm"></a><h4>Median String Algorithm</h4>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FMotif%2FConsensus%20String)_TOPIC%2F">Algorithms/Motif/Consensus String</a></li>
<li><a href="#BM_(Algorithms%5C%2FMotif%2FMotif%20Matrix%20Score)_TOPIC%2F">Algorithms/Motif/Motif Matrix Score</a></li>
<li><a href="#BM_(Algorithms%5C%2FK-mer%2FHamming%20Distance)_TOPIC%2F">Algorithms/K-mer/Hamming Distance</a></li>
</ul>
</div>
<p><strong>ALGORITHM</strong>:</p>
<p>This algorithm takes advantage of the fact that the same score can be derived by scoring a <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> either row-by-row or column-by-column. For example, the score for the following <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> is 3...</p>
<table>
<thead>
<tr>
<th></th>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
<th></th>
</tr>
</thead>
<tbody>
<tr>
<td></td>
<td>A</td>
<td>T</td>
<td><strong>G</strong></td>
<td>C</td>
<td>A</td>
<td>C</td>
<td></td>
</tr>
<tr>
<td></td>
<td>A</td>
<td>T</td>
<td><strong>G</strong></td>
<td>C</td>
<td>A</td>
<td>C</td>
<td></td>
</tr>
<tr>
<td></td>
<td>A</td>
<td>T</td>
<td>C</td>
<td>C</td>
<td><strong>T</strong></td>
<td>C</td>
<td></td>
</tr>
<tr>
<td></td>
<td>A</td>
<td>T</td>
<td>C</td>
<td>C</td>
<td>A</td>
<td>C</td>
<td></td>
</tr>
<tr>
<td>Score</td>
<td>0</td>
<td>0</td>
<td>2</td>
<td>0</td>
<td>1</td>
<td>0</td>
<td>3</td>
</tr>
</tbody>
</table>
<p>For each column, the number of unpopular nucleotides is counted. Then, those counts are summed to get the score: 0 + 0 + 2 + 0 + 1 + 0 = 3.</p>
<p>That exact same score scan be calculated by working through the <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> row-by-row...</p>
<table>
<thead>
<tr>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
<th>Score</th>
</tr>
</thead>
<tbody>
<tr>
<td>A</td>
<td>T</td>
<td><strong>G</strong></td>
<td>C</td>
<td>A</td>
<td>C</td>
<td>1</td>
</tr>
<tr>
<td>A</td>
<td>T</td>
<td><strong>G</strong></td>
<td>C</td>
<td>A</td>
<td>C</td>
<td>1</td>
</tr>
<tr>
<td>A</td>
<td>T</td>
<td>C</td>
<td>C</td>
<td><strong>T</strong></td>
<td>C</td>
<td>1</td>
</tr>
<tr>
<td>A</td>
<td>T</td>
<td>C</td>
<td>C</td>
<td>A</td>
<td>C</td>
<td>0</td>
</tr>
<tr>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td>3</td>
</tr>
</tbody>
</table>
<p>For each row, the number of unpopular nucleotides is counted. Then, those counts are summed to get the score: 1 + 1 + 1 + 0 = 3.</p>
<table>
<thead>
<tr>
<th></th>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
<th>Score</th>
</tr>
</thead>
<tbody>
<tr>
<td></td>
<td>A</td>
<td>T</td>
<td><strong>G</strong></td>
<td>C</td>
<td>A</td>
<td>C</td>
<td>1</td>
</tr>
<tr>
<td></td>
<td>A</td>
<td>T</td>
<td><strong>G</strong></td>
<td>C</td>
<td>A</td>
<td>C</td>
<td>1</td>
</tr>
<tr>
<td></td>
<td>A</td>
<td>T</td>
<td>C</td>
<td>C</td>
<td><strong>T</strong></td>
<td>C</td>
<td>1</td>
</tr>
<tr>
<td></td>
<td>A</td>
<td>T</td>
<td>C</td>
<td>C</td>
<td>A</td>
<td>C</td>
<td>0</td>
</tr>
<tr>
<td>Score</td>
<td>0</td>
<td>0</td>
<td>2</td>
<td>0</td>
<td>1</td>
<td>0</td>
<td>3</td>
</tr>
</tbody>
</table>
<p>Notice how each row's score is equivalent to the <a href="#BM_(hamming%20distance)%2Fi">hamming distance</a> between the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> at that row and the <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>'s <a href="#BM_(consensus%20string%7Cconsensus%20sequence)%2Fi">consensus string</a>. Specifically, the <a href="#BM_(consensus%20string%7Cconsensus%20sequence)%2Fi">consensus string</a> for the <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> is ATCCAC. For each row, ...</p>
<ul>
<li>hamming_distance(ATGCAC, ATCCAC) = 1</li>
<li>hamming_distance(ATGCAC, ATCCAC) = 1</li>
<li>hamming_distance(ATCCTC, ATCCAC) = 1</li>
<li>hamming_distance(ATCCAC, ATCCAC) = 0</li>
</ul>
<p>Given these facts, this algorithm constructs a set of <a href="#BM_(consensus%20string%7Cconsensus%20sequence)%2Fi">consensus string</a>s by enumerating through all possible <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s for some k. Then, for each <a href="#BM_(consensus%20string%7Cconsensus%20sequence)%2Fi">consensus string</a>, it scans over each <a href="#BM_(sequence)%2Fi">sequence</a> to find the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> that minimizes the <a href="#BM_(hamming%20distance)%2Fi">hamming distance</a> for that <a href="#BM_(consensus%20string%7Cconsensus%20sequence)%2Fi">consensus string</a>. These <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s are used as the <a href="#BM_%5Cb(membership%7Cmember)_MOTIF%2Fi">member</a>s of a <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>.</p>
<p>Of all the <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrices</a> built, the one with the lowest score is selected.</p>
<p>Since the k for the <a href="#BM_(motif)%2Fi">motif</a> is unknown, this algorithm may need to be repeated multiple times with different k values. This algorithm also doesn't scale very well. For k=10, 1048576 different <a href="#BM_(consensus%20string%7Cconsensus%20sequence)%2Fi">consensus string</a>s are possible.</p>
<p><a href="ch2_code/src/MedianStringSearch.py">ch2_code/src/MedianStringSearch.py</a> (lines 8 to 33):</p>
<pre class="hljs"><code><span class="hljs-comment"># The name is slightly confusing. What this actually does...</span>
<span class="hljs-comment">#   For each dna string:</span>
<span class="hljs-comment">#     Find the k-mer with the min hamming distance between the k-mers that make up the DNA string and pattern</span>
<span class="hljs-comment">#   Sum up the min hamming distances of the found k-mers (equivalent to the motif matrix score)</span>
<span class="hljs-keyword">def</span> <span class="hljs-title function_">distance_between_pattern_and_strings</span>(<span class="hljs-params">pattern: <span class="hljs-built_in">str</span>, dnas: <span class="hljs-type">List</span>[<span class="hljs-built_in">str</span>]</span>) -&gt; <span class="hljs-built_in">int</span>:
    min_hds = []

    k = <span class="hljs-built_in">len</span>(pattern)
    <span class="hljs-keyword">for</span> dna <span class="hljs-keyword">in</span> dnas:
        min_hd = <span class="hljs-literal">None</span>
        <span class="hljs-keyword">for</span> dna_kmer, _ <span class="hljs-keyword">in</span> slide_window(dna, k):
            hd = hamming_distance(pattern, dna_kmer)
            <span class="hljs-keyword">if</span> min_hd <span class="hljs-keyword">is</span> <span class="hljs-literal">None</span> <span class="hljs-keyword">or</span> hd &lt; min_hd:
                min_hd = hd
        min_hds.append(min_hd)
    <span class="hljs-keyword">return</span> <span class="hljs-built_in">sum</span>(min_hds)


<span class="hljs-keyword">def</span> <span class="hljs-title function_">median_string</span>(<span class="hljs-params">k: <span class="hljs-built_in">int</span>, dnas: <span class="hljs-type">List</span>[<span class="hljs-built_in">str</span>]</span>):
    last_best: <span class="hljs-type">Tuple</span>[<span class="hljs-built_in">str</span>, <span class="hljs-built_in">int</span>] = <span class="hljs-literal">None</span>  <span class="hljs-comment"># last found consensus string and its score</span>
    <span class="hljs-keyword">for</span> kmer <span class="hljs-keyword">in</span> enumerate_patterns(k):
        score = distance_between_pattern_and_strings(kmer, dnas)  <span class="hljs-comment"># find score of best motif matrix where consensus str is kmer</span>
        <span class="hljs-keyword">if</span> last_best <span class="hljs-keyword">is</span> <span class="hljs-literal">None</span> <span class="hljs-keyword">or</span> score &lt; last_best[<span class="hljs-number">1</span>]:
            last_best = kmer, score
    <span class="hljs-keyword">return</span> last_best
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Searching for motif of k=3 in the following...</p>
<p>AAATTGACGCAT<br>GACGACCACGTT<br>CGTCAGCGCCTG<br>GCTGAGCACCGG<br>AGTTCGGGACAG</p>
<p>Found the consensus string GAC with a score of 2</p>
</div>
<p></p>
<a name="H_Greedy%20Algorithm"></a><h4>Greedy Algorithm</h4>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FMotif%2FMotif%20Matrix%20Score)_TOPIC%2F">Algorithms/Motif/Motif Matrix Score</a></li>
<li><a href="#BM_(Algorithms%5C%2FMotif%2FK-mer%20Match%20Probability)_TOPIC%2F">Algorithms/Motif/K-mer Match Probability</a></li>
</ul>
</div>
<p><strong>ALGORITHM</strong>:</p>
<p>This algorithm begins by constructing a <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> where the only <a href="#BM_%5Cb(membership%7Cmember)_MOTIF%2Fi">member</a> is a <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> picked from the first <a href="#BM_(sequence)%2Fi">sequence</a>. From there, it goes through the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s in the ...</p>
<ol>
<li>second <a href="#BM_(sequence)%2Fi">sequence</a> to find the one that has the highest match probability to the <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> and adds it as a <a href="#BM_%5Cb(membership%7Cmember)_MOTIF%2Fi">member</a> to the <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>.</li>
<li>third <a href="#BM_(sequence)%2Fi">sequence</a> to find the one that has the highest match probability to the <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> and adds it as a <a href="#BM_%5Cb(membership%7Cmember)_MOTIF%2Fi">member</a> to the <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>.</li>
<li>fourth <a href="#BM_(sequence)%2Fi">sequence</a> to find the one that has the highest match probability to the <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> and adds it as a <a href="#BM_%5Cb(membership%7Cmember)_MOTIF%2Fi">member</a> to the <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>.</li>
<li>...</li>
</ol>
<p>This process repeats once for every <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> in the first <a href="#BM_(sequence)%2Fi">sequence</a>. Each repetition produces a <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>. Of all the <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrices</a> built, the one with the lowest score is selected.</p>
<p>This is a <a href="#BM_(greedy%20algorithm)%2Fi">greedy algorithm</a>. It builds out potential <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrices</a> by selecting the locally optimal <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> from each <a href="#BM_(sequence)%2Fi">sequence</a>. While this may not lead to the globally optimal <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>, it's fast and has a higher than normal likelihood of picking out the correct <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>.</p>
<p><a href="ch2_code/src/GreedyMotifMatrixSearchWithPsuedocounts.py">ch2_code/src/GreedyMotifMatrixSearchWithPsuedocounts.py</a> (lines 12 to 33):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">greedy_motif_search_with_psuedocounts</span>(<span class="hljs-params">k: <span class="hljs-built_in">int</span>, dnas: <span class="hljs-type">List</span>[<span class="hljs-built_in">str</span>]</span>):
    best_motif_matrix = [dna[<span class="hljs-number">0</span>:k] <span class="hljs-keyword">for</span> dna <span class="hljs-keyword">in</span> dnas]

    <span class="hljs-keyword">for</span> motif, _ <span class="hljs-keyword">in</span> slide_window(dnas[<span class="hljs-number">0</span>], k):
        motif_matrix = [motif]
        counts = motif_matrix_count(motif_matrix)
        apply_psuedocounts_to_count_matrix(counts)
        profile = motif_matrix_profile(counts)

        <span class="hljs-keyword">for</span> dna <span class="hljs-keyword">in</span> dnas[<span class="hljs-number">1</span>:]:
            next_motif, _ = find_most_probable_kmer_using_profile_matrix(profile, dna)
            <span class="hljs-comment"># push in closest kmer as a motif member and recompute profile for the next iteration</span>
            motif_matrix.append(next_motif)
            counts = motif_matrix_count(motif_matrix)
            apply_psuedocounts_to_count_matrix(counts)
            profile = motif_matrix_profile(counts)

        <span class="hljs-keyword">if</span> score_motif(motif_matrix) &lt; score_motif(best_motif_matrix):
            best_motif_matrix = motif_matrix

    <span class="hljs-keyword">return</span> best_motif_matrix
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Searching for motif of k=3 in the following...</p>
<p>AAATTGACGCAT<br>GACGACCACGTT<br>CGTCAGCGCCTG<br>GCTGAGCACCGG<br>AGTTCGGGACAG</p>
<p>Found the motif matrix...</p>
<p>GAC<br>GAC<br>GTC<br>GAG<br>GAC</p>
</div>
<p></p>
<a name="H_Randomized%20Algorithm"></a><h4>Randomized Algorithm</h4>
<p><a name="BM_(Algorithms%5C%2FMotif%2FFind%20Motif%20Matrix%2FRandomized%20Algorithm)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FMotif%2FMotif%20Matrix%20Score)_TOPIC%2F">Algorithms/Motif/Motif Matrix Score</a></li>
<li><a href="#BM_(Algorithms%5C%2FMotif%2FMotif%20Matrix%20Profile)_TOPIC%2F">Algorithms/Motif/Motif Matrix Profile</a></li>
<li><a href="#BM_(Algorithms%5C%2FMotif%2FK-mer%20Match%20Probability)_TOPIC%2F">Algorithms/Motif/K-mer Match Probability</a></li>
</ul>
</div>
<p><strong>ALGORITHM</strong>:</p>
<p>This algorithm selects a random <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> from each <a href="#BM_(sequence)%2Fi">sequence</a> to form an initial <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>. Then, for each <a href="#BM_(sequence)%2Fi">sequence</a>, it finds the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> that has the highest probability of matching that <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>. Those <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s form the <a href="#BM_%5Cb(membership%7Cmember)_MOTIF%2Fi">member</a>s of a new <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>. If the new <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> scores better than the existing <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>, the existing <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> gets replaced with the new <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> and the process repeats. Otherwise, the existing <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> is selected.</p>
<p>In theory, this algorithm works because all <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s in a <a href="#BM_(sequence)%2Fi">sequence</a> other than the <a href="#BM_(motif%20member)%2Fi">motif member</a> are considered to be random noise. As such, if no <a href="#BM_(motif%20member)%2Fi">motif member</a>s were selected when creating the initial <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>, the profile of that initial <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> would be more or less uniform:</p>
<table>
<thead>
<tr>
<th></th>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
</tr>
</thead>
<tbody>
<tr>
<td>A</td>
<td>0.25</td>
<td>0.25</td>
<td>0.25</td>
<td>0.25</td>
<td>0.25</td>
<td>0.25</td>
</tr>
<tr>
<td>C</td>
<td>0.25</td>
<td>0.25</td>
<td>0.25</td>
<td>0.25</td>
<td>0.25</td>
<td>0.25</td>
</tr>
<tr>
<td>T</td>
<td>0.25</td>
<td>0.25</td>
<td>0.25</td>
<td>0.25</td>
<td>0.25</td>
<td>0.25</td>
</tr>
<tr>
<td>G</td>
<td>0.25</td>
<td>0.25</td>
<td>0.25</td>
<td>0.25</td>
<td>0.25</td>
<td>0.25</td>
</tr>
</tbody>
</table>
<p>Such a profile wouldn't allow for converging to a vastly better scoring <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>.</p>
<p>However, if at least one <a href="#BM_(motif%20member)%2Fi">motif member</a> were selected when creating the initial <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>, the profile of that initial <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> would skew towards the <a href="#BM_(motif)%2Fi">motif</a>:</p>
<table>
<thead>
<tr>
<th></th>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
</tr>
</thead>
<tbody>
<tr>
<td>A</td>
<td><strong>0.333</strong></td>
<td>0.233</td>
<td>0.233</td>
<td>0.233</td>
<td><strong>0.333</strong></td>
<td>0.233</td>
</tr>
<tr>
<td>C</td>
<td>0.233</td>
<td>0.233</td>
<td><strong>0.333</strong></td>
<td><strong>0.333</strong></td>
<td>0.233</td>
<td><strong>0.333</strong></td>
</tr>
<tr>
<td>T</td>
<td>0.233</td>
<td><strong>0.333</strong></td>
<td>0.233</td>
<td>0.233</td>
<td>0.233</td>
<td>0.233</td>
</tr>
<tr>
<td>G</td>
<td>0.233</td>
<td>0.233</td>
<td>0.233</td>
<td>0.233</td>
<td>0.233</td>
<td>0.233</td>
</tr>
</tbody>
</table>
<p>Such a profile would lead to a better scoring <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> where that better scoring <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> contains the other <a href="#BM_%5Cb(membership%7Cmember)_MOTIF%2Fi">member</a>s of the <a href="#BM_(motif)%2Fi">motif</a>.</p>
<p>In practice, this algorithm may trip up on real-world data. Real-world <a href="#BM_(sequence)%2Fi">sequence</a>s don't actually contain random noise. The hope is that the only <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s that are highly similar to each other in the <a href="#BM_(sequence)%2Fi">sequence</a>s are <a href="#BM_%5Cb(membership%7Cmember)_MOTIF%2Fi">member</a>s of the <a href="#BM_(motif)%2Fi">motif</a>. It's possible that the <a href="#BM_(sequence)%2Fi">sequence</a>s contain other sets of <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s that are similar to each other but vastly different from the <a href="#BM_(motif%20member)%2Fi">motif member</a>s. In such cases, even if a <a href="#BM_(motif%20member)%2Fi">motif member</a> were to be selected when creating the initial <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>, the algorithm may converge to a <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> that isn't for the <a href="#BM_(motif)%2Fi">motif</a>.</p>
<p>This is a <a href="#BM_(Monte%20Carlo%20algorithm)%2Fi">monte carlo algorithm</a>. It uses randomness to deliver an approximate solution. While this may not lead to the globally optimal <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>, it's fast and as such can be run multiple times. The run with the best <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> will likely be a good enough solution (it captures most of the <a href="#BM_(motif%20member)%2Fi">motif member</a>s, or parts of the <a href="#BM_(motif%20member)%2Fi">motif member</a>s if k was too small, or etc..).</p>
<p><a href="ch2_code/src/RandomizedMotifMatrixSearchWithPsuedocounts.py">ch2_code/src/RandomizedMotifMatrixSearchWithPsuedocounts.py</a> (lines 13 to 32):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">randomized_motif_search_with_psuedocounts</span>(<span class="hljs-params">k: <span class="hljs-built_in">int</span>, dnas: <span class="hljs-type">List</span>[<span class="hljs-built_in">str</span>]</span>) -&gt; <span class="hljs-type">List</span>[<span class="hljs-built_in">str</span>]:
        motif_matrix = []
        <span class="hljs-keyword">for</span> dna <span class="hljs-keyword">in</span> dnas:
            start = randrange(<span class="hljs-built_in">len</span>(dna) - k + <span class="hljs-number">1</span>)
            kmer = dna[start:start + k]
            motif_matrix.append(kmer)

        best_motif_matrix = motif_matrix

        <span class="hljs-keyword">while</span> <span class="hljs-literal">True</span>:
            counts = motif_matrix_count(motif_matrix)
            apply_psuedocounts_to_count_matrix(counts)
            profile = motif_matrix_profile(counts)

            motif_matrix = [find_most_probable_kmer_using_profile_matrix(profile, dna)[<span class="hljs-number">0</span>] <span class="hljs-keyword">for</span> dna <span class="hljs-keyword">in</span> dnas]
            <span class="hljs-keyword">if</span> score_motif(motif_matrix) &lt; score_motif(best_motif_matrix):
                best_motif_matrix = motif_matrix
            <span class="hljs-keyword">else</span>:
                <span class="hljs-keyword">return</span> best_motif_matrix
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Searching for motif of k=3 in the following...</p>
<p>AAATTGACGCAT<br>GACGACCACGTT<br>CGTCAGCGCCTG<br>GCTGAGCACCGG<br>AGTTCGGGACAG</p>
<p>Running 1000 iterations...</p>
<p>Best found the motif matrix...</p>
<p>GAC<br>GAC<br>GCC<br>GAG<br>GAC</p>
</div>
<p></p>
<a name="H_Gibbs%20Sampling%20Algorithm"></a><h4>Gibbs Sampling Algorithm</h4>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FMotif%2FMotif%20Matrix%20Score)_TOPIC%2F">Algorithms/Motif/Motif Matrix Score</a></li>
<li><a href="#BM_(Algorithms%5C%2FMotif%2FK-mer%20Match%20Probability)_TOPIC%2F">Algorithms/Motif/K-mer Match Probability</a></li>
<li><a href="#BM_(Algorithms%5C%2FMotif%2FFind%20Motif%20Matrix%2FRandomized%20Algorithm)_TOPIC%2F">Algorithms/Motif/Find Motif Matrix/Randomized Algorithm</a></li>
</ul>
</div>
<p><strong>ALGORITHM</strong>:</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The Pevzner book mentions there's more to Gibbs Sampling than what it discussed. I looked up the topic but couldn't make much sense of it.</p>
</div>
<p>This algorithm selects a random <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> from each <a href="#BM_(sequence)%2Fi">sequence</a> to form an initial <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>. Then, one of the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s from the <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> is randomly chosen and replaced with another <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> from the same <a href="#BM_(sequence)%2Fi">sequence</a> that the removed <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> came from. The replacement is selected by using a weighted random number algorithm, where how likely a <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> is to be chosen as a replacement has to do with how probable of a match it is to the <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>.</p>
<p>This process of replacement is repeated for some user-defined number of cycles, at which point the algorithm has hopefully homed in on the desired <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>.</p>
<p>This is a <a href="#BM_(Monte%20Carlo%20algorithm)%2Fi">monte carlo algorithm</a>. It uses randomness to deliver an approximate solution. While this may not lead to the globally optimal <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>, it's fast and as such can be run multiple times. The run with the best <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> will likely be a good enough solution (it captures most of the <a href="#BM_(motif%20member)%2Fi">motif member</a>s, or parts of the <a href="#BM_(motif%20member)%2Fi">motif member</a>s if k was too small, or etc..).</p>
<p>The idea behind this algorithm is similar to the idea behind the <a href="#BM_(randomized%20algorithm)%2Fi">randomized algorithm</a> for <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> finding, except that this algorithm is more conservative in how it converges on a <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> and the weighted random selection allows it to potentially break out if stuck in a local optima.</p>
<p><a href="ch2_code/src/GibbsSamplerMotifMatrixSearchWithPsuedocounts.py">ch2_code/src/GibbsSamplerMotifMatrixSearchWithPsuedocounts.py</a> (lines 14 to 59):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">gibbs_rand</span>(<span class="hljs-params">prob_dist: <span class="hljs-type">List</span>[<span class="hljs-built_in">float</span>]</span>) -&gt; <span class="hljs-built_in">int</span>:
    <span class="hljs-comment"># normalize prob_dist -- just incase sum(prob_dist) != 1.0</span>
    prob_dist_sum = <span class="hljs-built_in">sum</span>(prob_dist)
    prob_dist = [p / prob_dist_sum <span class="hljs-keyword">for</span> p <span class="hljs-keyword">in</span> prob_dist]

    <span class="hljs-keyword">while</span> <span class="hljs-literal">True</span>:
        selection = randrange(<span class="hljs-number">0</span>, <span class="hljs-built_in">len</span>(prob_dist))
        <span class="hljs-keyword">if</span> random() &lt; prob_dist[selection]:
            <span class="hljs-keyword">return</span> selection


<span class="hljs-keyword">def</span> <span class="hljs-title function_">determine_probabilities_of_all_kmers_in_dna</span>(<span class="hljs-params">profile_matrix: <span class="hljs-type">Dict</span>[<span class="hljs-built_in">str</span>, <span class="hljs-type">List</span>[<span class="hljs-built_in">float</span>]], dna: <span class="hljs-built_in">str</span>, k: <span class="hljs-built_in">int</span></span>) -&gt; <span class="hljs-type">List</span>[<span class="hljs-built_in">int</span>]:
    ret = []
    <span class="hljs-keyword">for</span> kmer, _ <span class="hljs-keyword">in</span> slide_window(dna, k):
        prob = determine_probability_of_match_using_profile_matrix(profile_matrix, kmer)
        ret.append(prob)
    <span class="hljs-keyword">return</span> ret


<span class="hljs-keyword">def</span> <span class="hljs-title function_">gibbs_sampler_motif_search_with_psuedocounts</span>(<span class="hljs-params">k: <span class="hljs-built_in">int</span>, dnas: <span class="hljs-type">List</span>[<span class="hljs-built_in">str</span>], cycles: <span class="hljs-built_in">int</span></span>) -&gt; <span class="hljs-type">List</span>[<span class="hljs-built_in">str</span>]:
    motif_matrix = []
    <span class="hljs-keyword">for</span> dna <span class="hljs-keyword">in</span> dnas:
        start = randrange(<span class="hljs-built_in">len</span>(dna) - k + <span class="hljs-number">1</span>)
        kmer = dna[start:start + k]
        motif_matrix.append(kmer)

    best_motif_matrix = motif_matrix[:]  <span class="hljs-comment"># create a copy, otherwise you'll be modifying both motif and best_motif</span>

    <span class="hljs-keyword">for</span> j <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(<span class="hljs-number">0</span>, cycles):
        i = randrange(<span class="hljs-built_in">len</span>(dnas))  <span class="hljs-comment"># pick a dna</span>
        <span class="hljs-keyword">del</span> motif_matrix[i]  <span class="hljs-comment"># remove the kmer for that dna from the motif str</span>

        counts = motif_matrix_count(motif_matrix)
        apply_psuedocounts_to_count_matrix(counts)
        profile = motif_matrix_profile(counts)

        new_motif_kmer_probs = determine_probabilities_of_all_kmers_in_dna(profile, dnas[i], k)
        new_motif_kmer_idx = gibbs_rand(new_motif_kmer_probs)
        new_motif_kmer = dnas[i][new_motif_kmer_idx:new_motif_kmer_idx+k]
        motif_matrix.insert(i, new_motif_kmer)

        <span class="hljs-keyword">if</span> score_motif(motif_matrix) &lt; score_motif(best_motif_matrix):
            best_motif_matrix = motif_matrix[:]  <span class="hljs-comment"># create a copy, otherwise you'll be modifying both motif and best_motif</span>

    <span class="hljs-keyword">return</span> best_motif_matrix
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Searching for motif of k=3 in the following...</p>
<p>AAATTGACGCAT<br>GACGACCACGTT<br>CGTCAGCGCCTG<br>GCTGAGCACCGG<br>AGTTCGGGACAG</p>
<p>Running 1000 iterations...</p>
<p>Best found the motif matrix...</p>
<p>GAC<br>GAC<br>GCC<br>GAG<br>GAC</p>
</div>
<p></p>
<a name="H_Motif%20Matrix%20Hybrid%20Alphabet"></a><h3>Motif Matrix Hybrid Alphabet</h3>
<p><a name="BM_(Algorithms%5C%2FMotif%2FMotif%20Matrix%20Hybrid%20Alphabet)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FMotif%2FConsensus%20String)_TOPIC%2F">Algorithms/Motif/Consensus String</a></li>
<li><a href="#BM_(Algorithms%5C%2FMotif%2FMotif%20Matrix%20Score)_TOPIC%2F">Algorithms/Motif/Motif Matrix Score</a></li>
<li><a href="#BM_(Algorithms%5C%2FMotif%2FFind%20Motif%20Matrix)_TOPIC%2F">Algorithms/Motif/Find Motif Matrix</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: When creating finding a <a href="#BM_(motif)%2Fi">motif</a>, it may be beneficial to use a <a href="#BM_(hybrid%20alphabet%7Calternate%20alphabet%7Calternative%20alphabet)%2Fi">hybrid alphabet</a> rather than the standard nucleotides (A, C, T, and G). For example, the following <a href="#BM_(hybrid%20alphabet%7Calternate%20alphabet%7Calternative%20alphabet)%2Fi">hybrid alphabet</a> marks certain combinations of nucleotides as a single letter:</p>
<ul>
<li>A = A</li>
<li>C = C</li>
<li>T = T</li>
<li>G = G</li>
<li>W = A or T</li>
<li>S = G or C</li>
<li>K = G or T</li>
<li>Y = C or T</li>
</ul>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The alphabet above was pulled from the Pevzner book section 2.16: Complications in <a href="#BM_(motif)%2Fi">Motif</a> Finding. It's a subset of the <a href="#BM_(IUPAC%20nucleotide%20code)%2Fi">IUPAC nucleotide code</a>s alphabet. The author didn't mention if the alphabet was explicitly chosen for <a href="#BM_(regulatory%20motif)%2Fi">regulatory motif</a> finding. If it was, it may have been derived from running probabilities over already discovered <a href="#BM_(regulatory%20motif)%2Fi">regulatory motif</a>s: e.g. for the <a href="#BM_(motif)%2Fi">motif</a>s already discovered, if a position has 2 possible nucleotides then G/C (S), G/T (K), C/T (Y), and A/T (W) are likely but other combinations aren't.</p>
</div>
<p><strong>WHY</strong>: <a href="#BM_(hybrid%20alphabet%7Calternate%20alphabet%7Calternative%20alphabet)%2Fi">Hybrid alphabet</a>s may make it easier for <a href="#BM_(motif)%2Fi">motif</a> finding algorithms to converge on a <a href="#BM_(motif)%2Fi">motif</a>. For example, when scoring a <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>, treat the position as a single letter if the distinct nucleotides at that position map to one of the combinations in the <a href="#BM_(hybrid%20alphabet%7Calternate%20alphabet%7Calternative%20alphabet)%2Fi">hybrid alphabet</a>.</p>
<p><a href="#BM_(hybrid%20alphabet%7Calternate%20alphabet%7Calternative%20alphabet)%2Fi">Hybrid alphabet</a>s may make more sense for representing a <a href="#BM_(consensus%20string%7Cconsensus%20sequence)%2Fi">consensus string</a>. Rather than picking out the most popular nucleotide, the <a href="#BM_(hybrid%20alphabet%7Calternate%20alphabet%7Calternative%20alphabet)%2Fi">hybrid alphabet</a> can be used to describe alternating nucleotides at each position.</p>
<p><strong>ALGORITHM</strong>:</p>
<p><a href="ch2_code/src/HybridAlphabetMatrix.py">ch2_code/src/HybridAlphabetMatrix.py</a> (lines 5 to 26):</p>
<pre class="hljs"><code>PEVZNER_2_16_ALPHABET = <span class="hljs-built_in">dict</span>()
PEVZNER_2_16_ALPHABET[<span class="hljs-built_in">frozenset</span>({<span class="hljs-string">'A'</span>, <span class="hljs-string">'T'</span>})] = <span class="hljs-string">'W'</span>
PEVZNER_2_16_ALPHABET[<span class="hljs-built_in">frozenset</span>({<span class="hljs-string">'G'</span>, <span class="hljs-string">'C'</span>})] = <span class="hljs-string">'S'</span>
PEVZNER_2_16_ALPHABET[<span class="hljs-built_in">frozenset</span>({<span class="hljs-string">'G'</span>, <span class="hljs-string">'T'</span>})] = <span class="hljs-string">'K'</span>
PEVZNER_2_16_ALPHABET[<span class="hljs-built_in">frozenset</span>({<span class="hljs-string">'C'</span>, <span class="hljs-string">'T'</span>})] = <span class="hljs-string">'Y'</span>


<span class="hljs-keyword">def</span> <span class="hljs-title function_">to_hybrid_alphabet_motif_matrix</span>(<span class="hljs-params">motif_matrix: <span class="hljs-type">List</span>[<span class="hljs-built_in">str</span>], hybrid_alphabet: <span class="hljs-type">Dict</span>[FrozenSet[<span class="hljs-built_in">str</span>], <span class="hljs-built_in">str</span>]</span>) -&gt; <span class="hljs-type">List</span>[<span class="hljs-built_in">str</span>]:
    rows = <span class="hljs-built_in">len</span>(motif_matrix)
    cols = <span class="hljs-built_in">len</span>(motif_matrix[<span class="hljs-number">0</span>])

    motif_matrix = motif_matrix[:]  <span class="hljs-comment"># make a copy</span>
    <span class="hljs-keyword">for</span> c <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(cols):
        distinct_nucs_at_c = <span class="hljs-built_in">frozenset</span>([motif_matrix[r][c] <span class="hljs-keyword">for</span> r <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(rows)])
        <span class="hljs-keyword">if</span> distinct_nucs_at_c <span class="hljs-keyword">in</span> hybrid_alphabet:
            <span class="hljs-keyword">for</span> r <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(rows):
                motif_member = motif_matrix[r]
                motif_member = motif_member[:c] + hybrid_alphabet[distinct_nucs_at_c] + motif_member[c+<span class="hljs-number">1</span>:]
                motif_matrix[r] = motif_member

    <span class="hljs-keyword">return</span> motif_matrix
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Converted...</p>
<p>CATCCG<br>CTTCCT<br>CATCTT</p>
<p>to...</p>
<p>CWTCYK<br>CWTCYK<br>CWTCYK</p>
<p>using...</p>
<p>{frozenset({'A', 'T'}): 'W', frozenset({'G', 'C'}): 'S', frozenset({'G', 'T'}): 'K', frozenset({'C', 'T'}): 'Y'}</p>
</div>
<p></p>
<a name="H_DNA%20Assembly"></a><h2>DNA Assembly</h2>
<p><a name="BM_(Algorithms%5C%2FDNA%20Assembly)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FK-mer)_TOPIC%2F">Algorithms/K-mer</a></li>
</ul>
</div>
<p>DNA <a href="#BM_(sequencer)%2Fi">sequencer</a>s work by taking many copies of an organism's <a href="#BM_(genome)%2Fi">genome</a>, breaking up those copies into fragments, then scanning in those fragments. <a href="#BM_(sequencer)%2Fi">Sequencer</a>s typically scan fragments in 1 of 2 ways:</p>
<ul>
<li>
<p><a href="#BM_%5Cb(read)_SEQ%2Fi">read</a>s - small DNA fragments of equal size (represented as <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s).</p>
<p><img src="svgbob_0466ab6a6dc49c17c2bac89fed628823f1b8d580.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a href="#BM_(read-pair%7Cread%20pair)%2Fi">read-pair</a>s - small DNA fragments of equal size where the bases in the middle part of the fragment aren't known (represented as <a href="#BM_(%5C(%5Cd%2B%2C%5Cs*%5Cd%2B%5C)-mer%7Ckd-mer%7Ckdmer%7C%5C(k%2C%5Cs*d%5C)-mer)%2Fi">kd-mer</a>s).</p>
<p><img src="svgbob_246ba3d13372ae267d9a596e030d11d7550b036b.svg" alt="Kroki diagram output"></p>
</li>
</ul>
<p><a href="#BM_(assembly%7Cassemble)%2Fi">Assembly</a> is the process of reconstructing an organism's <a href="#BM_(genome)%2Fi">genome</a> from the <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s returned by a <a href="#BM_(sequencer)%2Fi">sequencer</a>. Since the <a href="#BM_(sequencer)%2Fi">sequencer</a> breaks up many copies of the same <a href="#BM_(genome)%2Fi">genome</a> and each <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>'s start position is random, the original <a href="#BM_(genome)%2Fi">genome</a> can be reconstructed by finding overlaps between <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s and stitching them back together.</p>
<p><img src="svgbob_c14e6c546e8e4369db69bd39a7c6e934e7ddefe7.svg" alt="Kroki diagram output"></p>
<p>A typical problem with <a href="#BM_(sequencing%7Csequenced)%2Fi">sequencing</a> is that the number of errors in a <a href="#BM_(fragment)_SEQ%2Fi">fragment</a> increase as the number of scanned bases increases. As such, <a href="#BM_(read-pair%7Cread%20pair)%2Fi">read-pair</a>s are preferred over <a href="#BM_%5Cb(read)_SEQ%2Fi">read</a>s: by only scanning in the head and tail of a long <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>, the scan won't contain as many errors as a <a href="#BM_%5Cb(read)_SEQ%2Fi">read</a> of the same length but will still contain extra information which helps with <a href="#BM_(assembly%7Cassemble)%2Fi">assembly</a> (length of unknown nucleotides in between the prefix and suffix).</p>
<p><a href="#BM_(assembly%7Cassemble)%2Fi">Assembly</a> has many practical complications that prevent full <a href="#BM_(genome)%2Fi">genome</a> reconstruction from <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s:</p>
<ul>
<li>
<p>Which strand of <a href="#BM_(double%20stranded%20DNA%7Cdouble-stranded%20DNA)%2Fi">double stranded DNA</a> that a <a href="#BM_%5Cb(read)_SEQ%2Fi">read</a> / <a href="#BM_(read-pair%7Cread%20pair)%2Fi">read-pair</a> comes from isn't known, which means the overlaps you find may not be accurate.</p>
<p><img src="svgbob_9a1e8981d7c6da721c87d87057f266ca1aeee3d2.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>The <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s may not cover the entire <a href="#BM_(genome)%2Fi">genome</a>, which prevents full reconstruction.</p>
<p><img src="svgbob_b6294ec3ffaf661c338522ed9985ae78672bd1a0.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>The <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s may have errors (e.g. wrong nucleotides scanned in), which may prevent finding overlaps.</p>
<p><img src="svgbob_51205be80cb357bdfa476acb058e6136e81a0051.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>The <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s for repetitive parts of the <a href="#BM_(genome)%2Fi">genome</a> (e.g. <a href="#BM_(transposon%7Ctransposable%20element%7Cjumping%20gene)%2Fi">transposon</a>s) likely can't be accurately <a href="#BM_(assembly%7Cassemble)%2Fi">assemble</a>d.</p>
<p><img src="svgbob_b6caa3257c0c8b7d997763ca20474e872861ae48.svg" alt="Kroki diagram output"></p>
</li>
</ul>
<a name="H_Stitch%20Reads"></a><h3>Stitch Reads</h3>
<p><a name="BM_(Algorithms%5C%2FDNA%20Assembly%2FStitch%20Reads)_TOPIC%2F"></a><strong></strong></p>
<p><strong>WHAT</strong>: Given a list of overlapping <a href="#BM_%5Cb(read)_SEQ%2Fi">read</a>s where ...</p>
<ul>
<li>all <a href="#BM_%5Cb(read)_SEQ%2Fi">read</a>s are of the same k,</li>
<li>all overlap regions are of the same length,</li>
<li>and each <a href="#BM_%5Cb(read)_SEQ%2Fi">read</a> in the list overlaps with the next <a href="#BM_%5Cb(read)_SEQ%2Fi">read</a> in the list</li>
</ul>
<p>... , stitch them together. For example, in the <a href="#BM_%5Cb(read)_SEQ%2Fi">read</a> list [GAAA, AAAT, AATC] each <a href="#BM_%5Cb(read)_SEQ%2Fi">read</a> overlaps the subsequent <a href="#BM_%5Cb(read)_SEQ%2Fi">read</a> by an offset of 1: GAAATC.</p>
<table>
<thead>
<tr>
<th></th>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
</tr>
</thead>
<tbody>
<tr>
<td>R1</td>
<td>G</td>
<td>A</td>
<td>A</td>
<td>A</td>
<td></td>
<td></td>
</tr>
<tr>
<td>R2</td>
<td></td>
<td>A</td>
<td>A</td>
<td>A</td>
<td>T</td>
<td></td>
</tr>
<tr>
<td>R3</td>
<td></td>
<td></td>
<td>A</td>
<td>A</td>
<td>T</td>
<td>C</td>
</tr>
<tr>
<td>Stitched</td>
<td>G</td>
<td>A</td>
<td>A</td>
<td>A</td>
<td>T</td>
<td>C</td>
</tr>
</tbody>
</table>
<p><strong>WHY</strong>: Since the <a href="#BM_(sequencer)%2Fi">sequencer</a> breaks up many copies of the same DNA and each <a href="#BM_%5Cb(read)_SEQ%2Fi">read</a>'s start position is random, larger parts of the original DNA can be reconstructed by finding overlaps between <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s and stitching them back together.</p>
<p><strong>ALGORITHM</strong>:</p>
<p><a href="ch3_code/src/Read.py">ch3_code/src/Read.py</a> (lines 55 to 76):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">append_overlap</span>(<span class="hljs-params">self: Read, other: Read, skip: <span class="hljs-built_in">int</span> = <span class="hljs-number">1</span></span>) -&gt; Read:
    offset = <span class="hljs-built_in">len</span>(self.data) - <span class="hljs-built_in">len</span>(other.data)
    data_head = self.data[:offset]
    data = self.data[offset:]

    prefix = data[:skip]
    overlap1 = data[skip:]
    overlap2 = other.data[:-skip]
    suffix = other.data[-skip:]
    ret = data_head + prefix
    <span class="hljs-keyword">for</span> ch1, ch2 <span class="hljs-keyword">in</span> <span class="hljs-built_in">zip</span>(overlap1, overlap2):
        ret += ch1 <span class="hljs-keyword">if</span> ch1 == ch2 <span class="hljs-keyword">else</span> <span class="hljs-string">'?'</span>  <span class="hljs-comment"># for failure, use IUPAC nucleotide codes instead of question mark?</span>
    ret += suffix
    <span class="hljs-keyword">return</span> Read(ret, source=(<span class="hljs-string">'overlap'</span>, [self, other]))

<span class="hljs-meta">@staticmethod</span>
<span class="hljs-keyword">def</span> <span class="hljs-title function_">stitch</span>(<span class="hljs-params">items: <span class="hljs-type">List</span>[Read], skip: <span class="hljs-built_in">int</span> = <span class="hljs-number">1</span></span>) -&gt; <span class="hljs-built_in">str</span>:
    <span class="hljs-keyword">assert</span> <span class="hljs-built_in">len</span>(items) &gt; <span class="hljs-number">0</span>
    ret = items[<span class="hljs-number">0</span>]
    <span class="hljs-keyword">for</span> other <span class="hljs-keyword">in</span> items[<span class="hljs-number">1</span>:]:
        ret = ret.append_overlap(other, skip)
    <span class="hljs-keyword">return</span> ret.data
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Stitched [GAAA, AAAT, AATC] to GAAATC</p>
</div>
<p></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>See also: <a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FOverlap%20Alignment)_TOPIC%2F">Algorithms/Sequence Alignment/Overlap Alignment</a></p>
</div>
<a name="H_Stitch%20Read-Pairs"></a><h3>Stitch Read-Pairs</h3>
<p><a name="BM_(Algorithms%5C%2FDNA%20Assembly%2FStitch%20Read-Pairs)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FDNA%20Assembly%2FStitch%20Reads)_TOPIC%2F">Algorithms/DNA Assembly/Stitch Reads</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: Given a list of overlapping <a href="#BM_(read-pair%7Cread%20pair)%2Fi">read-pair</a>s where ...</p>
<ul>
<li>all <a href="#BM_(read-pair%7Cread%20pair)%2Fi">read-pair</a>s are of the same k and d,</li>
<li>all overlap regions are of the same length,</li>
<li>and each <a href="#BM_(read-pair%7Cread%20pair)%2Fi">read-pair</a> in the list overlaps with the next <a href="#BM_(read-pair%7Cread%20pair)%2Fi">read-pair</a> in the list</li>
</ul>
<p>... , stitch them together. For example, in the <a href="#BM_(read-pair%7Cread%20pair)%2Fi">read-pair</a> list [ATG---CCG, TGT---CGT, GTT---GTT, TTA---TTC] each <a href="#BM_(read-pair%7Cread%20pair)%2Fi">read-pair</a> overlaps the subsequent <a href="#BM_(read-pair%7Cread%20pair)%2Fi">read-pair</a> by an offset of 1: ATGTTACCGTTC.</p>
<table>
<thead>
<tr>
<th></th>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
<th>6</th>
<th>7</th>
<th>8</th>
<th>9</th>
<th>10</th>
<th>11</th>
</tr>
</thead>
<tbody>
<tr>
<td>R1</td>
<td>A</td>
<td>T</td>
<td>G</td>
<td>-</td>
<td>-</td>
<td>-</td>
<td>C</td>
<td>C</td>
<td>G</td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td>R2</td>
<td></td>
<td>T</td>
<td>G</td>
<td>T</td>
<td>-</td>
<td>-</td>
<td>-</td>
<td>C</td>
<td>G</td>
<td>T</td>
<td></td>
<td></td>
</tr>
<tr>
<td>R3</td>
<td></td>
<td></td>
<td>G</td>
<td>T</td>
<td>T</td>
<td>-</td>
<td>-</td>
<td>-</td>
<td>G</td>
<td>T</td>
<td>T</td>
<td></td>
</tr>
<tr>
<td>R4</td>
<td></td>
<td></td>
<td></td>
<td>T</td>
<td>T</td>
<td>A</td>
<td>-</td>
<td>-</td>
<td>-</td>
<td>T</td>
<td>T</td>
<td>C</td>
</tr>
<tr>
<td>Stitched</td>
<td>A</td>
<td>T</td>
<td>G</td>
<td>T</td>
<td>T</td>
<td>A</td>
<td>C</td>
<td>C</td>
<td>G</td>
<td>T</td>
<td>T</td>
<td>C</td>
</tr>
</tbody>
</table>
<p><strong>WHY</strong>: Since the <a href="#BM_(sequencer)%2Fi">sequencer</a> breaks up many copies of the same DNA and each <a href="#BM_%5Cb(read)_SEQ%2Fi">read</a>'s start position is random, larger parts of the original DNA can be reconstructed by finding overlaps between <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s and stitching them back together.</p>
<p><strong>ALGORITHM</strong>:</p>
<p>Overlapping <a href="#BM_(read-pair%7Cread%20pair)%2Fi">read-pair</a>s are stitched by taking the first <a href="#BM_(read-pair%7Cread%20pair)%2Fi">read-pair</a> and iterating through the remaining <a href="#BM_(read-pair%7Cread%20pair)%2Fi">read-pair</a>s where ...</p>
<ul>
<li>the suffix from each remaining <a href="#BM_(read-pair%7Cread%20pair)%2Fi">read-pair</a>'s head k is appended to the first <a href="#BM_(read-pair%7Cread%20pair)%2Fi">read-pair</a>'s head k.</li>
<li>the suffix from each remaining <a href="#BM_(read-pair%7Cread%20pair)%2Fi">read-pair</a>'s tail k is appended to the first <a href="#BM_(read-pair%7Cread%20pair)%2Fi">read-pair</a>'s tail k.</li>
</ul>
<p>For example, to stitch [ATG---CCG, TGT---CGT], ...</p>
<ol>
<li>stitch the heads as if they were <a href="#BM_%5Cb(read)_SEQ%2Fi">read</a>s: [ATG, TGT] results in ATGT,</li>
<li>stitch the tails as if they were <a href="#BM_%5Cb(read)_SEQ%2Fi">read</a>s: [CCG, CGT] results in CCGT.</li>
</ol>
<table>
<thead>
<tr>
<th></th>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
<th>6</th>
<th>7</th>
<th>8</th>
<th>9</th>
</tr>
</thead>
<tbody>
<tr>
<td>R1</td>
<td>A</td>
<td>T</td>
<td>G</td>
<td>-</td>
<td>-</td>
<td>-</td>
<td>C</td>
<td>C</td>
<td>G</td>
<td></td>
</tr>
<tr>
<td>R2</td>
<td></td>
<td>T</td>
<td>G</td>
<td>T</td>
<td>-</td>
<td>-</td>
<td>-</td>
<td>C</td>
<td>G</td>
<td>T</td>
</tr>
<tr>
<td>Stitched</td>
<td>A</td>
<td>T</td>
<td>G</td>
<td>T</td>
<td>-</td>
<td>-</td>
<td>C</td>
<td>C</td>
<td>G</td>
<td>T</td>
</tr>
</tbody>
</table>
<p><a href="ch3_code/src/ReadPair.py">ch3_code/src/ReadPair.py</a> (lines 82 to 110):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">append_overlap</span>(<span class="hljs-params">self: ReadPair, other: ReadPair, skip: <span class="hljs-built_in">int</span> = <span class="hljs-number">1</span></span>) -&gt; ReadPair:
    self_head = Read(self.data.head)
    other_head = Read(other.data.head)
    new_head = self_head.append_overlap(other_head)
    new_head = new_head.data

    self_tail = Read(self.data.tail)
    other_tail = Read(other.data.tail)
    new_tail = self_tail.append_overlap(other_tail)
    new_tail = new_tail.data

    <span class="hljs-comment"># WARNING: new_d may go negative -- In the event of a negative d, it means that rather than there being a gap</span>
    <span class="hljs-comment"># in between the head and tail, there's an OVERLAP in between the head and tail. To get rid of the overlap, you</span>
    <span class="hljs-comment"># need to remove either the last d chars from head or first d chars from tail.</span>
    new_d = self.d - skip
    kdmer = Kdmer(new_head, new_tail, new_d)

    <span class="hljs-keyword">return</span> ReadPair(kdmer, source=(<span class="hljs-string">'overlap'</span>, [self, other]))

<span class="hljs-meta">@staticmethod</span>
<span class="hljs-keyword">def</span> <span class="hljs-title function_">stitch</span>(<span class="hljs-params">items: <span class="hljs-type">List</span>[ReadPair], skip: <span class="hljs-built_in">int</span> = <span class="hljs-number">1</span></span>) -&gt; <span class="hljs-built_in">str</span>:
    <span class="hljs-keyword">assert</span> <span class="hljs-built_in">len</span>(items) &gt; <span class="hljs-number">0</span>
    ret = items[<span class="hljs-number">0</span>]
    <span class="hljs-keyword">for</span> other <span class="hljs-keyword">in</span> items[<span class="hljs-number">1</span>:]:
        ret = ret.append_overlap(other, skip)
    <span class="hljs-keyword">assert</span> ret.d &lt;= <span class="hljs-number">0</span>, <span class="hljs-string">"Gap still exists -- not enough to stitch"</span>
    overlap_count = -ret.d
    <span class="hljs-keyword">return</span> ret.data.head + ret.data.tail[overlap_count:]
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Stitched [ATG---CCG, TGT---CGT, GTT---GTT, TTA---TTC] to ATGTTACCGTTC</p>
</div>
<p></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>See also: <a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FOverlap%20Alignment)_TOPIC%2F">Algorithms/Sequence Alignment/Overlap Alignment</a></p>
</div>
<a name="H_Break%20Reads"></a><h3>Break Reads</h3>
<p><a name="BM_(Algorithms%5C%2FDNA%20Assembly%2FBreak%20Reads)_TOPIC%2F"></a><strong></strong></p>
<p><strong>WHAT</strong>: Given a set of <a href="#BM_%5Cb(read)_SEQ%2Fi">read</a>s that arbitrarily overlap, each <a href="#BM_%5Cb(read)_SEQ%2Fi">read</a> can be broken into many smaller <a href="#BM_%5Cb(read)_SEQ%2Fi">read</a>s that overlap better. For example, given 4 <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">10-mer</a>s that arbitrarily overlap, you can break them into better overlapping <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">5-mer</a>s...</p>
<p><img src="svgbob_d9e103b16312cfe660d19df9305a9361ac1241ac.svg" alt="Kroki diagram output"></p>
<p><strong>WHY</strong>: <a href="#BM_(read%20breaking%7Cread-breaking%7Cbreaking%20reads)%2Fi">Breaking reads</a> may cause more ambiguity in overlaps. At the same time, <a href="#BM_(read%20breaking%7Cread-breaking%7Cbreaking%20reads)%2Fi">read breaking</a> makes it easier to find overlaps by bringing the overlaps closer together and provides (artificially) increased <a href="#BM_(coverage)_SEQ%2Fi">coverage</a>.</p>
<p><strong>ALGORITHM</strong>:</p>
<p><a href="ch3_code/src/Read.py">ch3_code/src/Read.py</a> (lines 80 to 87):</p>
<pre class="hljs"><code><span class="hljs-comment"># This is read breaking -- why not just call it break? because break is a reserved keyword.</span>
<span class="hljs-keyword">def</span> <span class="hljs-title function_">shatter</span>(<span class="hljs-params">self: Read, k: <span class="hljs-built_in">int</span></span>) -&gt; <span class="hljs-type">List</span>[Read]:
    ret = []
    <span class="hljs-keyword">for</span> kmer, _ <span class="hljs-keyword">in</span> slide_window(self.data, k):
        r = Read(kmer, source=(<span class="hljs-string">'shatter'</span>, [self]))
        ret.append(r)
    <span class="hljs-keyword">return</span> ret
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Broke ACTAAGAACC to [ACTAA, CTAAG, TAAGA, AAGAA, AGAAC, GAACC]</p>
</div>
<p></p>
<a name="H_Break%20Read-Pairs"></a><h3>Break Read-Pairs</h3>
<p><a name="BM_(Algorithms%5C%2FDNA%20Assembly%2FBreak%20Read-Pairs)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FDNA%20Assembly%2FBreak%20Reads)_TOPIC%2F">Algorithms/DNA Assembly/Break Reads</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: Given a set of <a href="#BM_(read-pair%7Cread%20pair)%2Fi">read-pair</a>s that arbitrarily overlap, each <a href="#BM_(read-pair%7Cread%20pair)%2Fi">read-pair</a> can be broken into many <a href="#BM_(read-pair%7Cread%20pair)%2Fi">read-pair</a>s with a smaller k that overlap better. For example, given 4 <a href="#BM_(%5C(%5Cd%2B%2C%5Cs*%5Cd%2B%5C)-mer%7Ckd-mer%7Ckdmer%7C%5C(k%2C%5Cs*d%5C)-mer)%2Fi">(4,2)-mer</a>s that arbitrarily overlap, you can break them into better overlapping <a href="#BM_(%5C(%5Cd%2B%2C%5Cs*%5Cd%2B%5C)-mer%7Ckd-mer%7Ckdmer%7C%5C(k%2C%5Cs*d%5C)-mer)%2Fi">(2,4)-mer</a>s...</p>
<p><img src="svgbob_eb8f1251b4e293a9662ade2c0a4452a7c7884f2f.svg" alt="Kroki diagram output"></p>
<p><strong>WHY</strong>: <a href="#BM_(read-pair%20breaking%7Cread%20pair%20breaking%7Cbreaking%20read-pairs%7Cbreaking%20read%20pairs)%2Fi">Breaking read-pairs</a> may cause more ambiguity in overlaps. At the same time, <a href="#BM_(read-pair%20breaking%7Cread%20pair%20breaking%7Cbreaking%20read-pairs%7Cbreaking%20read%20pairs)%2Fi">read-pair breaking</a> makes it easier to find overlaps by bringing the overlaps closer together and provides (artificially) increased <a href="#BM_(coverage)_SEQ%2Fi">coverage</a>.</p>
<p><strong>ALGORITHM</strong>:</p>
<p><a href="ch3_code/src/ReadPair.py">ch3_code/src/ReadPair.py</a> (lines 113 to 124):</p>
<pre class="hljs"><code><span class="hljs-comment"># This is read breaking -- why not just call it break? because break is a reserved keyword.</span>
<span class="hljs-keyword">def</span> <span class="hljs-title function_">shatter</span>(<span class="hljs-params">self: ReadPair, k: <span class="hljs-built_in">int</span></span>) -&gt; <span class="hljs-type">List</span>[ReadPair]:
    ret = []
    d = (self.k - k) + self.d
    <span class="hljs-keyword">for</span> window_head, window_tail <span class="hljs-keyword">in</span> <span class="hljs-built_in">zip</span>(slide_window(self.data.head, k), slide_window(self.data.tail, k)):
        kmer_head, _ = window_head
        kmer_tail, _ = window_tail
        kdmer = Kdmer(kmer_head, kmer_tail, d)
        rp = ReadPair(kdmer, source=(<span class="hljs-string">'shatter'</span>, [self]))
        ret.append(rp)
    <span class="hljs-keyword">return</span> ret
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Broke ACTA--AACC to [AC----AA, CT----AC, TA----CC]</p>
</div>
<p></p>
<a name="H_Probability%20of%20Fragment%20Occurrence"></a><h3>Probability of Fragment Occurrence</h3>
<p><a name="BM_(Algorithms%5C%2FDNA%20Assembly%2FProbability%20of%20Fragment%20Occurrence)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FDNA%20Assembly%2FStitch%20Reads)_TOPIC%2F">Algorithms/DNA Assembly/Stitch Reads</a></li>
<li><a href="#BM_(Algorithms%5C%2FDNA%20Assembly%2FStitch%20Read-Pairs)_TOPIC%2F">Algorithms/DNA Assembly/Stitch Read-Pairs</a></li>
<li><a href="#BM_(Algorithms%5C%2FDNA%20Assembly%2FBreak%20Reads)_TOPIC%2F">Algorithms/DNA Assembly/Break Reads</a></li>
<li><a href="#BM_(Algorithms%5C%2FDNA%20Assembly%2FBreak%20Read-Pairs)_TOPIC%2F">Algorithms/DNA Assembly/Break Read-Pairs</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: <a href="#BM_(sequencer)%2Fi">Sequencer</a>s work by taking many copies of an organism's <a href="#BM_(genome)%2Fi">genome</a>, randomly breaking up those <a href="#BM_(genome)%2Fi">genome</a>s into smaller pieces, and randomly scanning in those pieces (<a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s). As such, it isn't immediately obvious how many times each <a href="#BM_(fragment)_SEQ%2Fi">fragment</a> actually appears in the <a href="#BM_(genome)%2Fi">genome</a>.</p>
<p>Imagine that you're <a href="#BM_(sequencing%7Csequenced)%2Fi">sequencing</a> an organism's <a href="#BM_(genome)%2Fi">genome</a>. Given that ...</p>
<ul>
<li>there's good <a href="#BM_(coverage)_SEQ%2Fi">coverage</a> of the <a href="#BM_(genome)%2Fi">genome</a> (e.g. ~30x as many <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s as the length of the <a href="#BM_(genome)%2Fi">genome</a>),</li>
<li>the <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s scanned in are chosen at random (unbiased),</li>
<li>the <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s scanned in start at random offsets in the <a href="#BM_(genome)%2Fi">genome</a> (unbiased),</li>
<li>and the majority of <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s are for non-repeating parts of the <a href="#BM_(genome)%2Fi">genome</a>.</li>
</ul>
<p>... you can use probabilities to hint at how many times a <a href="#BM_(fragment)_SEQ%2Fi">fragment</a> appears in the <a href="#BM_(genome)%2Fi">genome</a>.</p>
<p><strong>WHY</strong>:</p>
<p>Determining how many times a <a href="#BM_(fragment)_SEQ%2Fi">fragment</a> appears in a <a href="#BM_(genome)%2Fi">genome</a> helps with <a href="#BM_(assembly%7Cassemble)%2Fi">assembly</a>. Specifically, ...</p>
<ul>
<li><a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s for repeat regions of the <a href="#BM_(genome)%2Fi">genome</a> can be accounted for during <a href="#BM_(assembly%7Cassemble)%2Fi">assembly</a>.</li>
<li><a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s containing <a href="#BM_(sequencing%20error)%2Fi">sequencing error</a>s may be detectable and filtered out prior to <a href="#BM_(assembly%7Cassemble)%2Fi">assembly</a>.</li>
</ul>
<p><strong>ALGORITHM</strong>:</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>For simplicity's sake, the <a href="#BM_(genome)%2Fi">genome</a> is single-stranded (not <a href="#BM_(double%20stranded%20DNA%7Cdouble-stranded%20DNA)%2Fi">double-stranded DNA</a> / no <a href="#BM_(reverse%20complement)%2Fi">reverse complement</a>ing stand).</p>
</div>
<p>Imagine a <a href="#BM_(genome)%2Fi">genome</a> of ATGGATGC. A <a href="#BM_(sequencer)%2Fi">sequencer</a> runs over that single strand and generates <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">3-mer</a> <a href="#BM_%5Cb(read)_SEQ%2Fi">read</a>s with roughly 30x <a href="#BM_(coverage)_SEQ%2Fi">coverage</a>. The resulting <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s are ...</p>
<table>
<thead>
<tr>
<th><a href="#BM_%5Cb(read)_SEQ%2Fi">Read</a></th>
<th># of Copies</th>
</tr>
</thead>
<tbody>
<tr>
<td>ATG</td>
<td>61</td>
</tr>
<tr>
<td>TGG</td>
<td>30</td>
</tr>
<tr>
<td>GAT</td>
<td>31</td>
</tr>
<tr>
<td>TGC</td>
<td>29</td>
</tr>
<tr>
<td>TGT</td>
<td>1</td>
</tr>
</tbody>
</table>
<p>Since the <a href="#BM_(genome)%2Fi">genome</a> is known to have less than 50% repeats, the dominate number of copies likely maps to 1 instance of that <a href="#BM_%5Cb(read)_SEQ%2Fi">read</a> appearing in the <a href="#BM_(genome)%2Fi">genome</a>. Since the dominate number is ~30, divide the number of copies for each <a href="#BM_%5Cb(read)_SEQ%2Fi">read</a> by ~30 to find out roughly how many times each <a href="#BM_%5Cb(read)_SEQ%2Fi">read</a> appears in the <a href="#BM_(genome)%2Fi">genome</a> ...</p>
<table>
<thead>
<tr>
<th><a href="#BM_%5Cb(read)_SEQ%2Fi">Read</a></th>
<th># of Copies</th>
<th># of Appearances in <a href="#BM_(genome)%2Fi">Genome</a></th>
</tr>
</thead>
<tbody>
<tr>
<td>ATG</td>
<td>61</td>
<td>2</td>
</tr>
<tr>
<td>TGG</td>
<td>30</td>
<td>1</td>
</tr>
<tr>
<td>GAT</td>
<td>31</td>
<td>1</td>
</tr>
<tr>
<td>TGC</td>
<td>29</td>
<td>1</td>
</tr>
<tr>
<td>TGT</td>
<td>1</td>
<td>0.03</td>
</tr>
</tbody>
</table>
<p>Note the last <a href="#BM_%5Cb(read)_SEQ%2Fi">read</a> (TGT) has 0.03 appearances, meaning it's a <a href="#BM_%5Cb(read)_SEQ%2Fi">read</a> that it either</p>
<ul>
<li>contains a <a href="#BM_(sequencing%20error)%2Fi">sequencing error</a>,</li>
<li>or it has poor <a href="#BM_(coverage)_SEQ%2Fi">coverage</a> (likely because it's at the head / tail of the <a href="#BM_(genome)%2Fi">genome</a> so it got scanned in less than other <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s).</li>
</ul>
<p>In this case, it's an error because it doesn't appear in the original <a href="#BM_(genome)%2Fi">genome</a>: TGT is not in ATGGATGC.</p>
<p><a href="ch3_code/src/FragmentOccurrenceProbabilityCalculator.py">ch3_code/src/FragmentOccurrenceProbabilityCalculator.py</a> (lines 15 to 29):</p>
<pre class="hljs"><code><span class="hljs-comment"># If less than 50% of the reads are from repeats, this attempts to count and normalize such that it can hint at which</span>
<span class="hljs-comment"># reads may contain errors (= ~0) and which reads are for repeat regions (&gt; 1.0).</span>
<span class="hljs-keyword">def</span> <span class="hljs-title function_">calculate_fragment_occurrence_probabilities</span>(<span class="hljs-params">fragments: <span class="hljs-type">List</span>[T]</span>) -&gt; <span class="hljs-type">Dict</span>[T, <span class="hljs-built_in">float</span>]:
    counter = Counter(fragments)
    max_digit_count = <span class="hljs-built_in">max</span>([<span class="hljs-built_in">len</span>(<span class="hljs-built_in">str</span>(count)) <span class="hljs-keyword">for</span> count <span class="hljs-keyword">in</span> counter.values()])
    <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(max_digit_count):
        rounded_counter = Counter(<span class="hljs-built_in">dict</span>([(k, <span class="hljs-built_in">round</span>(count, -i)) <span class="hljs-keyword">for</span> k, count <span class="hljs-keyword">in</span> counter.items()]))
        <span class="hljs-keyword">for</span> k, orig_count <span class="hljs-keyword">in</span> counter.items():
            <span class="hljs-keyword">if</span> rounded_counter[k] == <span class="hljs-number">0</span>:
                rounded_counter[k] = orig_count
        most_occurring_count, times_counted = Counter(rounded_counter.values()).most_common(<span class="hljs-number">1</span>)[<span class="hljs-number">0</span>]
        <span class="hljs-keyword">if</span> times_counted &gt;= <span class="hljs-built_in">len</span>(rounded_counter) * <span class="hljs-number">0.5</span>:
            <span class="hljs-keyword">return</span> <span class="hljs-built_in">dict</span>([(key, value / most_occurring_count) <span class="hljs-keyword">for</span> key, value <span class="hljs-keyword">in</span> rounded_counter.items()])
    <span class="hljs-keyword">raise</span> ValueError(<span class="hljs-string">f'Failed to find a common count: <span class="hljs-subst">{counter}</span>'</span>)
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Sequenced fragments:</p>
<ul>
<li>ATG was scanned in 61 times.</li>
<li>TGG was scanned in 30 times.</li>
<li>GAT was scanned in 31 times.</li>
<li>TGC was scanned in 29 times.</li>
<li>TGT was scanned in 1 times.</li>
</ul>
<p>Probability of occurrence in genome:</p>
<ul>
<li>ATG probably has 2.0 appearances in the genome.</li>
<li>TGG probably has 1.0 appearances in the genome.</li>
<li>GAT probably has 1.0 appearances in the genome.</li>
<li>TGC probably has 1.0 appearances in the genome.</li>
<li>TGT probably has 0.03333333333333333 appearances in the genome.</li>
</ul>
</div>
<p></p>
<a name="H_Overlap%20Graph"></a><h3>Overlap Graph</h3>
<p><a name="BM_(Algorithms%5C%2FDNA%20Assembly%2FOverlap%20Graph)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FDNA%20Assembly%2FStitch%20Reads)_TOPIC%2F">Algorithms/DNA Assembly/Stitch Reads</a></li>
<li><a href="#BM_(Algorithms%5C%2FDNA%20Assembly%2FStitch%20Read-Pairs)_TOPIC%2F">Algorithms/DNA Assembly/Stitch Read-Pairs</a></li>
<li><a href="#BM_(Algorithms%5C%2FDNA%20Assembly%2FBreak%20Reads)_TOPIC%2F">Algorithms/DNA Assembly/Break Reads</a></li>
<li><a href="#BM_(Algorithms%5C%2FDNA%20Assembly%2FBreak%20Read-Pairs)_TOPIC%2F">Algorithms/DNA Assembly/Break Read-Pairs</a></li>
<li><a href="#BM_(Algorithms%5C%2FDNA%20Assembly%2FProbability%20of%20Fragment%20Occurrence)_TOPIC%2F">Algorithms/DNA Assembly/Probability of Fragment Occurrence</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: Given the <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s for a single strand of DNA, create a directed graph where ...</p>
<ol>
<li>
<p>each node is a <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>.</p>
<p><img src="svgbob_ba61508d2d4f3dfef58ce5783f8e22f49ddec891.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>each edge is between overlapping <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s (nodes), where the ...</p>
<ul>
<li>source node has the overlap in its suffix .</li>
<li>destination node has the overlap in its prefix.</li>
</ul>
<p><img src="svgbob_62843d070e55883049a5a0f4881f1ec17928f780.svg" alt="Kroki diagram output"></p>
</li>
</ol>
<p>This is called an <a href="#BM_(overlap%20graph)%2Fi">overlap graph</a>.</p>
<p><strong>WHY</strong>: An <a href="#BM_(overlap%20graph)%2Fi">overlap graph</a> shows the different ways that <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s can be stitched together. A path in an <a href="#BM_(overlap%20graph)%2Fi">overlap graph</a> that touches each node exactly once is one possibility for the original <a href="#BM_(single%20stranded%20DNA%7Csingle-stranded%20DNA)%2Fi">single stranded DNA</a> that the <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s came from. For example...</p>
<ul>
<li>[TTA, TAG, AGT, GTT, TTA, TAC, ACT, CTT] ⟶ TTAGTTACTT</li>
<li>[TTA, TAC, ACT, CTT, TTA, TAG, AGT, GTT] ⟶ TTACTTAGTT</li>
<li>[ACT, CTT, TTA, TAG, AGT, GTT, TTA, TAC] ⟶ ACTTAGTTAC</li>
<li>[CTT, TTA, TAG, AGT, GTT, TTA, TAC, ACT] ⟶ CTTAGTTACT</li>
<li>...</li>
</ul>
<p>These paths are referred to as <a href="#BM_(Hamiltonian%20path%7CHamilton%20path)%2Fi">Hamiltonian path</a>s.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>Notice that the example graph is circular. If the organism <a href="#BM_(genome)%2Fi">genome</a> itself were also circular (e.g. bacterial <a href="#BM_(genome)%2Fi">genome</a>), the <a href="#BM_(genome)%2Fi">genome</a> guesses above are all actually the same because circular <a href="#BM_(genome)%2Fi">genome</a>s don't have a beginning / end.</p>
</div>
<p><strong>ALGORITHM</strong>:</p>
<p><a href="#BM_(sequencer)%2Fi">Sequencer</a>s produce <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s, but <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s by themselves typically aren't enough for most experiments / algorithms. In theory, stitching overlapping <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s for a single-strand of DNA should reveal that single-strand of DNA. In practice, real-world complications make revealing that single-strand of DNA nearly impossible:</p>
<ul>
<li><a href="#BM_(fragment)_SEQ%2Fi">Fragment</a>s are for both strands (strand of <a href="#BM_(double%20stranded%20DNA%7Cdouble-stranded%20DNA)%2Fi">double-stranded DNA</a> a <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>'s from isn't known).</li>
<li><a href="#BM_(fragment)_SEQ%2Fi">Fragment</a>s may be missing (<a href="#BM_(sequencer)%2Fi">sequencer</a> didn't capture it).</li>
<li><a href="#BM_(fragment)_SEQ%2Fi">Fragment</a>s may have inconsistent <a href="#BM_(coverage)_SEQ%2Fi">coverage</a> (<a href="#BM_(sequencer)%2Fi">sequencer</a> captured it too many/few times).</li>
<li><a href="#BM_(fragment)_SEQ%2Fi">Fragment</a>s may be repeated (regions of the <a href="#BM_(genome)%2Fi">genome</a> may repeat).</li>
<li><a href="#BM_(fragment)_SEQ%2Fi">Fragment</a>s may have errors (<a href="#BM_(sequencer)%2Fi">sequencer</a> produced <a href="#BM_(sequencing%20error)%2Fi">sequencing error</a>s).</li>
<li><a href="#BM_(fragment)_SEQ%2Fi">Fragment</a>s may be stitch-able in more than one way (multiple <a href="#BM_(genome)%2Fi">genome</a> reconstruction guesses).</li>
<li><a href="#BM_(fragment)_SEQ%2Fi">Fragment</a>s may take a long time to stitch (computationally intensive).</li>
</ul>
<p>Nevertheless, in an ideal world where most of these problems don't exist, an <a href="#BM_(overlap%20graph)%2Fi">overlap graph</a> is a good way to guess the single-strand of DNA that a set of <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s came from. An <a href="#BM_(overlap%20graph)%2Fi">overlap graph</a> assumes that the <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s it's operating on ...</p>
<ul>
<li>are from a single-strand of DNA,</li>
<li>have correct occurrence counts (no missing or extra),</li>
<li>and contain no errors.</li>
</ul>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>Although the complications discussed above make it impossible to get the original <a href="#BM_(genome)%2Fi">genome</a> in its entirety, it's still possible to pull out large parts of the original <a href="#BM_(genome)%2Fi">genome</a>. This is discussed in <a href="#BM_(Algorithms%5C%2FDNA%20Assembly%2FFind%20Contigs)_TOPIC%2F">Algorithms/DNA Assembly/Find Contigs</a>.</p>
</div>
<p>To construct an <a href="#BM_(overlap%20graph)%2Fi">overlap graph</a>, create an edge between <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s that have an overlap.</p>
<p>For each <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>, add that <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>'s ...</p>
<ul>
<li>prefix to a hash table.</li>
<li>suffix to a hash table.</li>
</ul>
<p>Then, join the hash tables together to find overlapping <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s.</p>
<p><a href="ch3_code/src/ToOverlapGraphHash.py">ch3_code/src/ToOverlapGraphHash.py</a> (lines 13 to 36):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">to_overlap_graph</span>(<span class="hljs-params">items: <span class="hljs-type">List</span>[T], skip: <span class="hljs-built_in">int</span> = <span class="hljs-number">1</span></span>) -&gt; Graph[T]:
    ret = Graph()

    prefixes = <span class="hljs-built_in">dict</span>()
    suffixes = <span class="hljs-built_in">dict</span>()
    <span class="hljs-keyword">for</span> i, item <span class="hljs-keyword">in</span> <span class="hljs-built_in">enumerate</span>(items):
        prefix = item.prefix(skip)
        prefixes.setdefault(prefix, <span class="hljs-built_in">set</span>()).add(i)
        suffix = item.suffix(skip)
        suffixes.setdefault(suffix, <span class="hljs-built_in">set</span>()).add(i)

    <span class="hljs-keyword">for</span> key, indexes <span class="hljs-keyword">in</span> suffixes.items():
        other_indexes = prefixes.get(key)
        <span class="hljs-keyword">if</span> other_indexes <span class="hljs-keyword">is</span> <span class="hljs-literal">None</span>:
            <span class="hljs-keyword">continue</span>
        <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> indexes:
            item = items[i]
            <span class="hljs-keyword">for</span> j <span class="hljs-keyword">in</span> other_indexes:
                <span class="hljs-keyword">if</span> i == j:
                    <span class="hljs-keyword">continue</span>
                other_item = items[j]
                ret.insert_edge(item, other_item)
    <span class="hljs-keyword">return</span> ret
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the fragments ['TTA', 'TTA', 'TAG', 'AGT', 'GTT', 'TAC', 'ACT', 'CTT'], the overlap graph is...</p>
<p><img src="dot_8ca9fb7886e86704dc01b1bffd1d6dd1.svg" alt="Dot diagram"></p>
</div>
<p></p>
<p>A path that touches each node of an graph exactly once is a <a href="#BM_(Hamiltonian%20path%7CHamilton%20path)%2Fi">Hamiltonian path</a>. Each  The <a href="#BM_(Hamiltonian%20path%7CHamilton%20path)%2Fi">Hamiltonian path</a> in an <a href="#BM_(overlap%20graph)%2Fi">overlap graph</a> is a guess as to the original single strand of DNA that the <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s for the graph came from.</p>
<p>The code shown below recursively walks all paths. Of all the paths it walks over, the ones that walk every node of the graph exactly once are selected.</p>
<p>This algorithm will likely fall over on non-trivial <a href="#BM_(overlap%20graph)%2Fi">overlap graph</a>s. Even finding one <a href="#BM_(Hamiltonian%20path%7CHamilton%20path)%2Fi">Hamiltonian path</a> is computationally intensive.</p>
<p><a href="ch3_code/src/WalkAllHamiltonianPaths.py">ch3_code/src/WalkAllHamiltonianPaths.py</a> (lines 15 to 38):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">exhaustively_walk_until_all_nodes_touched_exactly_one</span>(<span class="hljs-params">
        graph: Graph[T],
        from_node: T,
        current_path: <span class="hljs-type">List</span>[T]
</span>) -&gt; <span class="hljs-type">List</span>[<span class="hljs-type">List</span>[T]]:
    current_path.append(from_node)

    <span class="hljs-keyword">if</span> <span class="hljs-built_in">len</span>(current_path) == <span class="hljs-built_in">len</span>(graph):
        found_paths = [current_path.copy()]
    <span class="hljs-keyword">else</span>:
        found_paths = []
        <span class="hljs-keyword">for</span> to_node <span class="hljs-keyword">in</span> graph.get_outputs(from_node):
            <span class="hljs-keyword">if</span> to_node <span class="hljs-keyword">in</span> <span class="hljs-built_in">set</span>(current_path):
                <span class="hljs-keyword">continue</span>
            found_paths += exhaustively_walk_until_all_nodes_touched_exactly_one(graph, to_node, current_path)

    current_path.pop()
    <span class="hljs-keyword">return</span> found_paths


<span class="hljs-comment"># walk each node exactly once</span>
<span class="hljs-keyword">def</span> <span class="hljs-title function_">walk_hamiltonian_paths</span>(<span class="hljs-params">graph: Graph[T], from_node: T</span>) -&gt; <span class="hljs-type">List</span>[<span class="hljs-type">List</span>[T]]:
    <span class="hljs-keyword">return</span> exhaustively_walk_until_all_nodes_touched_exactly_one(graph, from_node, [])
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the fragments ['TTA', 'TTA', 'TAG', 'AGT', 'GTT', 'TAC', 'ACT', 'CTT'], the overlap graph is...</p>
<p><img src="dot_8ca9fb7886e86704dc01b1bffd1d6dd1.svg" alt="Dot diagram"></p>
<p>... and the Hamiltonian paths are ...</p>
<ul>
<li>CTT -&gt; TTA -&gt; TAG -&gt; AGT -&gt; GTT -&gt; TTA -&gt; TAC -&gt; ACT</li>
<li>TAG -&gt; AGT -&gt; GTT -&gt; TTA -&gt; TAC -&gt; ACT -&gt; CTT -&gt; TTA</li>
<li>TTA -&gt; TAC -&gt; ACT -&gt; CTT -&gt; TTA -&gt; TAG -&gt; AGT -&gt; GTT</li>
<li>TAC -&gt; ACT -&gt; CTT -&gt; TTA -&gt; TAG -&gt; AGT -&gt; GTT -&gt; TTA</li>
<li>TTA -&gt; TAG -&gt; AGT -&gt; GTT -&gt; TTA -&gt; TAC -&gt; ACT -&gt; CTT</li>
<li>TAG -&gt; AGT -&gt; GTT -&gt; TTA -&gt; TAC -&gt; ACT -&gt; CTT -&gt; TTA</li>
<li>AGT -&gt; GTT -&gt; TTA -&gt; TAC -&gt; ACT -&gt; CTT -&gt; TTA -&gt; TAG</li>
<li>GTT -&gt; TTA -&gt; TAC -&gt; ACT -&gt; CTT -&gt; TTA -&gt; TAG -&gt; AGT</li>
<li>CTT -&gt; TTA -&gt; TAG -&gt; AGT -&gt; GTT -&gt; TTA -&gt; TAC -&gt; ACT</li>
<li>ACT -&gt; CTT -&gt; TTA -&gt; TAG -&gt; AGT -&gt; GTT -&gt; TTA -&gt; TAC</li>
<li>ACT -&gt; CTT -&gt; TTA -&gt; TAG -&gt; AGT -&gt; GTT -&gt; TTA -&gt; TAC</li>
<li>GTT -&gt; TTA -&gt; TAC -&gt; ACT -&gt; CTT -&gt; TTA -&gt; TAG -&gt; AGT</li>
<li>TAC -&gt; ACT -&gt; CTT -&gt; TTA -&gt; TAG -&gt; AGT -&gt; GTT -&gt; TTA</li>
<li>AGT -&gt; GTT -&gt; TTA -&gt; TAC -&gt; ACT -&gt; CTT -&gt; TTA -&gt; TAG</li>
<li>TTA -&gt; TAG -&gt; AGT -&gt; GTT -&gt; TTA -&gt; TAC -&gt; ACT -&gt; CTT</li>
<li>TTA -&gt; TAC -&gt; ACT -&gt; CTT -&gt; TTA -&gt; TAG -&gt; AGT -&gt; GTT</li>
</ul>
</div>
<p></p>
<a name="H_De%20Bruijn%20Graph"></a><h3>De Bruijn Graph</h3>
<p><a name="BM_(Algorithms%5C%2FDNA%20Assembly%2FDe%20Bruijn%20Graph)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FDNA%20Assembly%2FStitch%20Reads)_TOPIC%2F">Algorithms/DNA Assembly/Stitch Reads</a></li>
<li><a href="#BM_(Algorithms%5C%2FDNA%20Assembly%2FStitch%20Read-Pairs)_TOPIC%2F">Algorithms/DNA Assembly/Stitch Read-Pairs</a></li>
<li><a href="#BM_(Algorithms%5C%2FDNA%20Assembly%2FBreak%20Reads)_TOPIC%2F">Algorithms/DNA Assembly/Break Reads</a></li>
<li><a href="#BM_(Algorithms%5C%2FDNA%20Assembly%2FBreak%20Read-Pairs)_TOPIC%2F">Algorithms/DNA Assembly/Break Read-Pairs</a></li>
<li><a href="#BM_(Algorithms%5C%2FDNA%20Assembly%2FProbability%20of%20Fragment%20Occurrence)_TOPIC%2F">Algorithms/DNA Assembly/Probability of Fragment Occurrence</a></li>
<li><a href="#BM_(Algorithms%5C%2FDNA%20Assembly%2FOverlap%20Graph)_TOPIC%2F">Algorithms/DNA Assembly/Overlap Graph</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: Given the <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s for a single strand of DNA, create a directed graph where ...</p>
<ol>
<li>
<p>each <a href="#BM_(fragment)_SEQ%2Fi">fragment</a> is represented as an edge connecting 2 nodes, where the ...</p>
<ul>
<li>source node is the prefix of the <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>.</li>
<li>destination node is the suffix of the <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>.</li>
</ul>
<p><img src="svgbob_843b8aaedb289d3d3ad5aef2fcd9bf5a504ad5af.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>duplicate nodes are merged into a single node.</p>
<p><img src="svgbob_7dea4cc54e9eac3d3a657922085849157a9cf869.svg" alt="Kroki diagram output"></p>
</li>
</ol>
<p>This graph is called a <a href="#BM_(de%20Bruijn%20graph)%2Fi">de Bruijn graph</a>: a <a href="#BM_(balanced)_GRAPH%2Fi">balanced</a> and <a href="#BM_(strongly%20connected)%2Fi">strongly connected</a> graph where the <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s are represented as edges.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The example graph above is <a href="#BM_(balanced)_GRAPH%2Fi">balanced</a>. But, depending on the <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s used, the graph may not be totally <a href="#BM_(balanced)_GRAPH%2Fi">balanced</a>. A technique for dealing with this is detailed below. For now, just assume that the graph will be <a href="#BM_(balanced)_GRAPH%2Fi">balanced</a>.</p>
</div>
<p><strong>WHY</strong>:  Similar to an <a href="#BM_(overlap%20graph)%2Fi">overlap graph</a>, a <a href="#BM_(de%20Bruijn%20graph)%2Fi">de Bruijn graph</a> shows the different ways that <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s can be stitched together. However, unlike an <a href="#BM_(overlap%20graph)%2Fi">overlap graph</a>, the <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s are represented as edges rather than nodes. Where in an <a href="#BM_(overlap%20graph)%2Fi">overlap graph</a> you need to find paths that touch every node exactly once (<a href="#BM_(Hamiltonian%20path%7CHamilton%20path)%2Fi">Hamiltonian path</a>), in a <a href="#BM_(de%20Bruijn%20graph)%2Fi">de Bruijn graph</a> you need to find paths that walk over every edge exactly once (<a href="#BM_(Eulerian%20cycle)%2Fi">Eulerian cycle</a>).</p>
<p>A path in a <a href="#BM_(de%20Bruijn%20graph)%2Fi">de Bruijn graph</a> that walks over each edge exactly once is one possibility for the original <a href="#BM_(single%20stranded%20DNA%7Csingle-stranded%20DNA)%2Fi">single stranded DNA</a> that the <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s came from: it starts and ends at the same node (a cycle), and walks over every edge in the graph.</p>
<p>In contrast to finding a <a href="#BM_(Hamiltonian%20path%7CHamilton%20path)%2Fi">Hamiltonian path</a> in an <a href="#BM_(overlap%20graph)%2Fi">overlap graph</a>, it's much faster to find an <a href="#BM_(Eulerian%20cycle)%2Fi">Eulerian cycle</a> in a <a href="#BM_(de%20Bruijn%20graph)%2Fi">de Bruijn graph</a>.</p>
<p><a href="#BM_(de%20Bruijn%20graph)%2Fi">De Bruijn graph</a>s were originally invented to solve the <a href="#BM_(k-universal%7C%5Cd%2B-universal)%2Fi">k-universal</a> string problem, which is effectively the same concept as <a href="#BM_(assembly%7Cassemble)%2Fi">assembly</a>.</p>
<p><strong>ALGORITHM</strong>:</p>
<p><a href="#BM_(sequencer)%2Fi">Sequencer</a>s produce <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s, but <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s by themselves typically aren't enough for most experiments / algorithms. In theory, stitching overlapping <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s for a single-strand of DNA should reveal that single-strand of DNA. In practice, real-world complications make revealing that single-strand of DNA nearly impossible:</p>
<ul>
<li><a href="#BM_(fragment)_SEQ%2Fi">Fragment</a>s are for both strands (strand of <a href="#BM_(double%20stranded%20DNA%7Cdouble-stranded%20DNA)%2Fi">double-stranded DNA</a> a <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>'s from isn't known).</li>
<li><a href="#BM_(fragment)_SEQ%2Fi">Fragment</a>s may be missing (<a href="#BM_(sequencer)%2Fi">sequencer</a> didn't capture it).</li>
<li><a href="#BM_(fragment)_SEQ%2Fi">Fragment</a>s may have inconsistent <a href="#BM_(coverage)_SEQ%2Fi">coverage</a> (<a href="#BM_(sequencer)%2Fi">sequencer</a> captured it too many/few times).</li>
<li><a href="#BM_(fragment)_SEQ%2Fi">Fragment</a>s may be repeated (regions of the <a href="#BM_(genome)%2Fi">genome</a> may repeat).</li>
<li><a href="#BM_(fragment)_SEQ%2Fi">Fragment</a>s may have errors (<a href="#BM_(sequencer)%2Fi">sequencer</a> produced <a href="#BM_(sequencing%20error)%2Fi">sequencing error</a>s).</li>
<li><a href="#BM_(fragment)_SEQ%2Fi">Fragment</a>s may be stitch-able in more than one way (multiple <a href="#BM_(genome)%2Fi">genome</a> reconstruction guesses).</li>
<li><a href="#BM_(fragment)_SEQ%2Fi">Fragment</a>s may take a long time to stitch (computationally intensive).</li>
</ul>
<p>Nevertheless, in an ideal world where most of these problems don't exist, a <a href="#BM_(de%20Bruijn%20graph)%2Fi">de Bruijn graph</a> is a good way to guess the single-strand of DNA that a set of <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s came from. A <a href="#BM_(de%20Bruijn%20graph)%2Fi">de Bruijn graph</a> assumes that the <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s it's operating on ...</p>
<ul>
<li>are from a single-strand of DNA,</li>
<li>have correct occurrence counts (no missing or extra),</li>
<li>and contain no errors.</li>
</ul>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>Although the complications discussed above make it impossible to get the original <a href="#BM_(genome)%2Fi">genome</a> in its entirety, it's still possible to pull out large parts of the original <a href="#BM_(genome)%2Fi">genome</a>. This is discussed in <a href="#BM_(Algorithms%5C%2FDNA%20Assembly%2FFind%20Contigs)_TOPIC%2F">Algorithms/DNA Assembly/Find Contigs</a>.</p>
</div>
<p>To construct a <a href="#BM_(de%20Bruijn%20graph)%2Fi">de Bruijn graph</a>, add an edge for each <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>, creating missing nodes as required.</p>
<p><a href="ch3_code/src/ToDeBruijnGraph.py">ch3_code/src/ToDeBruijnGraph.py</a> (lines 13 to 20):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">to_debruijn_graph</span>(<span class="hljs-params">reads: <span class="hljs-type">List</span>[T], skip: <span class="hljs-built_in">int</span> = <span class="hljs-number">1</span></span>) -&gt; Graph[T]:
    graph = Graph()
    <span class="hljs-keyword">for</span> read <span class="hljs-keyword">in</span> reads:
        from_node = read.prefix(skip)
        to_node = read.suffix(skip)
        graph.insert_edge(from_node, to_node)
    <span class="hljs-keyword">return</span> graph
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the fragments ['TTAG', 'TAGT', 'AGTT', 'GTTA', 'TTAC', 'TACT', 'ACTT', 'CTTA'], the de Bruijn graph is...</p>
<p><img src="dot_fc6c4f4bf004838f81f7a6343b7d2e19.svg" alt="Dot diagram"></p>
</div>
<p></p>
<p>Note how the graph above is both <a href="#BM_(balanced)_GRAPH%2Fi">balanced</a> and <a href="#BM_(strongly%20connected)%2Fi">strongly connected</a>. In most cases, non-circular <a href="#BM_(genome)%2Fi">genome</a>s won't generate a <a href="#BM_(balanced%20graph)%2Fi">balanced graph</a> like the one above. Instead, a non-circular <a href="#BM_(genome)%2Fi">genome</a> will very likely generate a graph that's nearly <a href="#BM_(balanced)_GRAPH%2Fi">balanced</a>: Nearly <a href="#BM_(balanced%20graph)%2Fi">balanced graph</a>s are graphs that would be <a href="#BM_(balanced)_GRAPH%2Fi">balanced</a> if not for a few un<a href="#BM_(balanced%20node)%2Fi">balanced node</a>s (usually root and tail nodes). They can artificially be made to become <a href="#BM_(balanced)_GRAPH%2Fi">balanced</a> by finding im<a href="#BM_(balanced%20node)%2Fi">balanced node</a>s and creating artificial edges between them until they become <a href="#BM_(balanced%20node)%2Fi">balanced node</a>s.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>Circular <a href="#BM_(genome)%2Fi">genome</a>s are <a href="#BM_(genome)%2Fi">genome</a>s that wrap around (e.g. bacterial <a href="#BM_(genome)%2Fi">genome</a>s). They don't have a beginning / end.</p>
</div>
<p><a href="ch3_code/src/BalanceNearlyBalancedGraph.py">ch3_code/src/BalanceNearlyBalancedGraph.py</a> (lines 15 to 44):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">find_unbalanced_nodes</span>(<span class="hljs-params">graph: Graph[T]</span>) -&gt; <span class="hljs-type">List</span>[<span class="hljs-type">Tuple</span>[T, <span class="hljs-built_in">int</span>, <span class="hljs-built_in">int</span>]]:
    unbalanced_nodes = []
    <span class="hljs-keyword">for</span> node <span class="hljs-keyword">in</span> graph.get_nodes():
        in_degree = graph.get_in_degree(node)
        out_degree = graph.get_out_degree(node)
        <span class="hljs-keyword">if</span> in_degree != out_degree:
            unbalanced_nodes.append((node, in_degree, out_degree))
    <span class="hljs-keyword">return</span> unbalanced_nodes


<span class="hljs-comment"># creates a balanced graph from a nearly balanced graph -- nearly balanced means the graph has an equal number of</span>
<span class="hljs-comment"># missing outputs and missing inputs.</span>
<span class="hljs-keyword">def</span> <span class="hljs-title function_">balance_graph</span>(<span class="hljs-params">graph: Graph[T]</span>) -&gt; <span class="hljs-type">Tuple</span>[Graph[T], <span class="hljs-type">Set</span>[T], <span class="hljs-type">Set</span>[T]]:
    unbalanced_nodes = find_unbalanced_nodes(graph)
    nodes_with_missing_ins = <span class="hljs-built_in">filter</span>(<span class="hljs-keyword">lambda</span> x: x[<span class="hljs-number">1</span>] &lt; x[<span class="hljs-number">2</span>], unbalanced_nodes)
    nodes_with_missing_outs = <span class="hljs-built_in">filter</span>(<span class="hljs-keyword">lambda</span> x: x[<span class="hljs-number">1</span>] &gt; x[<span class="hljs-number">2</span>], unbalanced_nodes)

    graph = graph.copy()

    <span class="hljs-comment"># create 1 copy per missing input / per missing output</span>
    n_per_need_in = [_n <span class="hljs-keyword">for</span> n, in_degree, out_degree <span class="hljs-keyword">in</span> nodes_with_missing_ins <span class="hljs-keyword">for</span> _n <span class="hljs-keyword">in</span> [n] * (out_degree - in_degree)]
    n_per_need_out = [_n <span class="hljs-keyword">for</span> n, in_degree, out_degree <span class="hljs-keyword">in</span> nodes_with_missing_outs <span class="hljs-keyword">for</span> _n <span class="hljs-keyword">in</span> [n] * (in_degree - out_degree)]
    <span class="hljs-keyword">assert</span> <span class="hljs-built_in">len</span>(n_per_need_in) == <span class="hljs-built_in">len</span>(n_per_need_out)  <span class="hljs-comment"># need an equal count of missing ins and missing outs to balance</span>

    <span class="hljs-comment"># balance</span>
    <span class="hljs-keyword">for</span> n_need_in, n_need_out <span class="hljs-keyword">in</span> <span class="hljs-built_in">zip</span>(n_per_need_in, n_per_need_out):
        graph.insert_edge(n_need_out, n_need_in)

    <span class="hljs-keyword">return</span> graph, <span class="hljs-built_in">set</span>(n_per_need_in), <span class="hljs-built_in">set</span>(n_per_need_out)  <span class="hljs-comment"># return graph with cycle, orig root nodes, orig tail nodes</span>
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the fragments ['TTAC', 'TACC', 'ACCC', 'CCCT'], the artificially balanced de Bruijn graph is...</p>
<p><img src="dot_060b6d958ad63bcbb55a116f699b26f5.svg" alt="Dot diagram"></p>
<p>... with original head nodes at {TTA} and tail nodes at {CCT}.</p>
</div>
<p></p>
<p>Given a <a href="#BM_(de%20Bruijn%20graph)%2Fi">de Bruijn graph</a> (<a href="#BM_(strongly%20connected)%2Fi">strongly connected</a> and <a href="#BM_(balanced)_GRAPH%2Fi">balanced</a>), you can find a <a href="#BM_(Eulerian%20cycle)%2Fi">Eulerian cycle</a> by randomly walking unexplored edges in the graph. Pick a starting node and randomly walk edges until you end up back at that same node, ignoring all edges that were previously walked over. Of the nodes that were walked over, pick one that still has unexplored edges and repeat the process: Walk edges from that node until you end up back at that same node, ignoring edges all edges that were previously walked over (including those in the past iteration). Continue this until you run out of unexplored edges.</p>
<p><a href="ch3_code/src/WalkRandomEulerianCycle.py">ch3_code/src/WalkRandomEulerianCycle.py</a> (lines 14 to 64):</p>
<pre class="hljs"><code><span class="hljs-comment"># (6, 8), (8, 7), (7, 9), (9, 6)  ----&gt;  68796</span>
<span class="hljs-keyword">def</span> <span class="hljs-title function_">edge_list_to_node_list</span>(<span class="hljs-params">edges: <span class="hljs-type">List</span>[<span class="hljs-type">Tuple</span>[T, T]]</span>) -&gt; <span class="hljs-type">List</span>[T]:
    ret = [edges[<span class="hljs-number">0</span>][<span class="hljs-number">0</span>]]
    <span class="hljs-keyword">for</span> e <span class="hljs-keyword">in</span> edges:
        ret.append(e[<span class="hljs-number">1</span>])
    <span class="hljs-keyword">return</span> ret


<span class="hljs-keyword">def</span> <span class="hljs-title function_">randomly_walk_and_remove_edges_until_cycle</span>(<span class="hljs-params">graph: Graph[T], node: T</span>) -&gt; <span class="hljs-type">List</span>[T]:
    end_node = node
    edge_list = []
    from_node = node
    <span class="hljs-keyword">while</span> <span class="hljs-built_in">len</span>(graph) &gt; <span class="hljs-number">0</span>:
        to_nodes = graph.get_outputs(from_node)
        to_node = <span class="hljs-built_in">next</span>(to_nodes, <span class="hljs-literal">None</span>)
        <span class="hljs-keyword">assert</span> to_node <span class="hljs-keyword">is</span> <span class="hljs-keyword">not</span> <span class="hljs-literal">None</span>  <span class="hljs-comment"># eularian graphs are strongly connected, meaning we should never hit dead-end nodes</span>

        graph.delete_edge(from_node, to_node, <span class="hljs-literal">True</span>, <span class="hljs-literal">True</span>)

        edge = (from_node, to_node)
        edge_list.append(edge)
        from_node = to_node
        <span class="hljs-keyword">if</span> from_node == end_node:
            <span class="hljs-keyword">return</span> edge_list_to_node_list(edge_list)

    <span class="hljs-keyword">assert</span> <span class="hljs-literal">False</span>  <span class="hljs-comment"># eularian graphs are strongly connected and balanced, meaning we should never run out of nodes</span>


<span class="hljs-comment"># graph must be strongly connected</span>
<span class="hljs-comment"># graph must be balanced</span>
<span class="hljs-comment"># if the 2 conditions above are met, the graph will be eularian (a eulerian cycle exists)</span>
<span class="hljs-keyword">def</span> <span class="hljs-title function_">walk_eulerian_cycle</span>(<span class="hljs-params">graph: Graph[T], start_node: T</span>) -&gt; <span class="hljs-type">List</span>[T]:
    graph = graph.copy()

    node_cycle = randomly_walk_and_remove_edges_until_cycle(graph, start_node)
    node_cycle_ptr = <span class="hljs-number">0</span>
    <span class="hljs-keyword">while</span> <span class="hljs-built_in">len</span>(graph) &gt; <span class="hljs-number">0</span>:
        new_node_cycle = <span class="hljs-literal">None</span>
        <span class="hljs-keyword">for</span> local_ptr, node <span class="hljs-keyword">in</span> <span class="hljs-built_in">enumerate</span>(node_cycle[node_cycle_ptr:]):
            <span class="hljs-keyword">if</span> node <span class="hljs-keyword">not</span> <span class="hljs-keyword">in</span> graph:
                <span class="hljs-keyword">continue</span>
            node_cycle_ptr += local_ptr
            inject_node_cycle = randomly_walk_and_remove_edges_until_cycle(graph, node)
            new_node_cycle = node_cycle[:]
            new_node_cycle[node_cycle_ptr:node_cycle_ptr+<span class="hljs-number">1</span>] = inject_node_cycle
            <span class="hljs-keyword">break</span>
        <span class="hljs-keyword">assert</span> new_node_cycle <span class="hljs-keyword">is</span> <span class="hljs-keyword">not</span> <span class="hljs-literal">None</span>
        node_cycle = new_node_cycle

    <span class="hljs-keyword">return</span> node_cycle
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the fragments ['TTA', 'TAT', 'ATT', 'TTC', 'TCT', 'CTT'], the de Bruijn graph is...</p>
<p><img src="dot_3a93e1cbe061465a28efa39eadb1f973.svg" alt="Dot diagram"></p>
<p>... and a Eulerian cycle is ...</p>
<p>TT -&gt; TC -&gt; CT -&gt; TT -&gt; TA -&gt; AT -&gt; TT</p>
</div>
<p></p>
<p>Note that the graph above is naturally <a href="#BM_(balanced)_GRAPH%2Fi">balanced</a> (no artificial edges have been added in to make it <a href="#BM_(balanced)_GRAPH%2Fi">balanced</a>). If the graph you're finding a <a href="#BM_(Eulerian%20cycle)%2Fi">Eulerian cycle</a> on has been artificially <a href="#BM_(balanced)_GRAPH%2Fi">balanced</a>, simply start the search for a <a href="#BM_(Eulerian%20cycle)%2Fi">Eulerian cycle</a> from one of the original head node. The artificial edge will show up at the end of the <a href="#BM_(Eulerian%20cycle)%2Fi">Eulerian cycle</a>, and as such can be dropped from the path.</p>
<p><img src="svgbob_cdf5dbb7db5be8e11adb7461507ac8e7ae647e3a.svg" alt="Kroki diagram output"></p>
<p>This algorithm picks one <a href="#BM_(Eulerian%20cycle)%2Fi">Eulerian cycle</a> in a graph. Most graph have multiple <a href="#BM_(Eulerian%20cycle)%2Fi">Eulerian cycle</a>s, likely too many to enumerate all of them.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>See the section on <a href="#BM_(k-universal%7C%5Cd%2B-universal)%2Fi">k-universal</a> strings to see a real-world application of <a href="#BM_(Eulerian%20graph)%2Fi">Eulerian graph</a>s. For something like k=20, good luck trying to enumerate all <a href="#BM_(Eulerian%20cycle)%2Fi">Eulerian cycle</a>s.</p>
</div>
<a name="H_Find%20Bubbles"></a><h3>Find Bubbles</h3>
<p><a name="BM_(Algorithms%5C%2FDNA%20Assembly%2FFind%20Bubbles)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FDNA%20Assembly%2FOverlap%20Graph)_TOPIC%2F">Algorithms/DNA Assembly/Overlap Graph</a></li>
<li><a href="#BM_(Algorithms%5C%2FDNA%20Assembly%2FDe%20Bruijn%20Graph)_TOPIC%2F">Algorithms/DNA Assembly/De Bruijn Graph</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: Given a set of a <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s that have been broken to k (<a href="#BM_(read%20breaking%7Cread-breaking%7Cbreaking%20reads)%2Fi">read breaking</a> / <a href="#BM_(read-pair%20breaking%7Cread%20pair%20breaking%7Cbreaking%20read-pairs%7Cbreaking%20read%20pairs)%2Fi">read-pair breaking</a>), any ...</p>
<ul>
<li>forked prefixes,</li>
<li>forked suffixes,</li>
<li>or bubbles</li>
</ul>
<p>... of length ...</p>
<ul>
<li>k in the <a href="#BM_(overlap%20graph)%2Fi">overlap graph</a>,</li>
<li>or k-1 in the <a href="#BM_(de%20Bruijn%20graph)%2Fi">de Bruijn graph</a></li>
</ul>
<p>... may have been from a <a href="#BM_(sequencing%20error)%2Fi">sequencing error</a>.</p>
<p><img src="svgbob_986faa465630c014228747bd8c1b6750b9c7d5a7.svg" alt="Kroki diagram output"></p>
<p><strong>WHY</strong>: When <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s returned by a <a href="#BM_(sequencer)%2Fi">sequencer</a> get broken (<a href="#BM_(read%20breaking%7Cread-breaking%7Cbreaking%20reads)%2Fi">read breaking</a> / <a href="#BM_(read-pair%20breaking%7Cread%20pair%20breaking%7Cbreaking%20read-pairs%7Cbreaking%20read%20pairs)%2Fi">read-pair breaking</a>), any <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s containing <a href="#BM_(sequencing%20error)%2Fi">sequencing error</a>s may show up in the graph as one of 3 structures: forked prefix, forked suffix, or bubble. As such, it may be possible to detect these structures and flatten them (by removing bad branches) to get a cleaner graph.</p>
<p>For example, imagine the <a href="#BM_%5Cb(read)_SEQ%2Fi">read</a> ATTGG. <a href="#BM_(read%20breaking%7Cread-breaking%7Cbreaking%20reads)%2Fi">Read breaking</a> it into <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">2-mer</a> <a href="#BM_%5Cb(read)_SEQ%2Fi">read</a>s results in: [AT, TT, TG, GG].</p>
<p><img src="svgbob_14d21a2810cefc8cfab333aadbe576260b565d48.svg" alt="Kroki diagram output"></p>
<p>Now, imagine that the <a href="#BM_(sequencer)%2Fi">sequencer</a> captures that same part of the <a href="#BM_(genome)%2Fi">genome</a> again, but this time the <a href="#BM_%5Cb(read)_SEQ%2Fi">read</a> contains a <a href="#BM_(sequencing%20error)%2Fi">sequencing error</a>. Depending on where the incorrect nucleotide is, one of the 3 structures will get introduced into the graph:</p>
<ul>
<li>
<p>ATTGG vs A<strong>C</strong>TGG (within first 2 elements)</p>
<p><img src="svgbob_0c0a09e604164ee4ffcdc4f3fcfa420c2035cdd5.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>ATTGG vs ATT<strong>C</strong>G (within last 2 elements)</p>
<p><img src="svgbob_8749bd635048fbfa4bd483e9c5425e73ff41a8d1.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>ATTGG vs AT<strong>C</strong>GG (sandwiched after first 2 elements and before last 2 elements)</p>
<p><img src="svgbob_3cc83e04b59a481951a59a2edd92cab8b200b193.svg" alt="Kroki diagram output"></p>
</li>
</ul>
<p>Note that just because these structures exist doesn't mean that the <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s they represent definitively have <a href="#BM_(sequencing%20error)%2Fi">sequencing error</a>s. These structures could have been caused by other problems / may not be problems at all:</p>
<ul>
<li>Bubbles may be caused by repetitive regions of DNA: <a href="#BM_(fragment)_SEQ%2Fi">Fragment</a>s from different parts of the <a href="#BM_(genome)%2Fi">genome</a> that are the same except for a few positions will show up as bubbles.</li>
<li>Bubbles / forks may be caused when <a href="#BM_(sequencing%7Csequenced)%2Fi">sequencing</a> <a href="#BM_(double%20stranded%20DNA%7Cdouble-stranded%20DNA)%2Fi">double-stranded DNA</a>: When both strands of DNA get tangled into the same graph, it's possible that <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s from different strands form bubbles or forks.</li>
</ul>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The Pevzner book says that bubble removal is a common feature in modern <a href="#BM_(assembly%7Cassemble)%2Fi">assemble</a>rs. My assumption is that, before pulling out <a href="#BM_(contig)s%3F%5Cb%2Fi">contig</a>s (described later on), basic probabilities are used to try and suss out if a branch in a bubble / prefix fork / suffix fork is bad and remove it if it is. This (hopefully) results in longer <a href="#BM_(contig)s%3F%5Cb%2Fi">contig</a>s.</p>
</div>
<p><strong>ALGORITHM</strong>:</p>
<p><a href="ch3_code/src/FindGraphAnomalies.py">ch3_code/src/FindGraphAnomalies.py</a> (lines 53 to 105):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">find_head_convergences</span>(<span class="hljs-params">graph: Graph[T], branch_len: <span class="hljs-built_in">int</span></span>) -&gt; <span class="hljs-type">List</span>[<span class="hljs-type">Tuple</span>[<span class="hljs-type">Optional</span>[T], <span class="hljs-type">List</span>[T], <span class="hljs-type">Optional</span>[T]]]:
    root_nodes = <span class="hljs-built_in">filter</span>(<span class="hljs-keyword">lambda</span> n: graph.get_in_degree(n) == <span class="hljs-number">0</span>, graph.get_nodes())

    ret = []
    <span class="hljs-keyword">for</span> n <span class="hljs-keyword">in</span> root_nodes:
        <span class="hljs-keyword">for</span> child <span class="hljs-keyword">in</span> graph.get_outputs(n):
            path_from_child = walk_outs_until_converge(graph, child)
            <span class="hljs-keyword">if</span> path_from_child <span class="hljs-keyword">is</span> <span class="hljs-literal">None</span>:
                <span class="hljs-keyword">continue</span>
            diverging_node = <span class="hljs-literal">None</span>
            branch_path = [n] + path_from_child[:-<span class="hljs-number">1</span>]
            converging_node = path_from_child[-<span class="hljs-number">1</span>]
            path = (diverging_node, branch_path, converging_node)
            <span class="hljs-keyword">if</span> <span class="hljs-built_in">len</span>(branch_path) &lt;= branch_len:
                ret.append(path)
    <span class="hljs-keyword">return</span> ret


<span class="hljs-keyword">def</span> <span class="hljs-title function_">find_tail_divergences</span>(<span class="hljs-params">graph: Graph[T], branch_len: <span class="hljs-built_in">int</span></span>) -&gt; <span class="hljs-type">List</span>[<span class="hljs-type">Tuple</span>[<span class="hljs-type">Optional</span>[T], <span class="hljs-type">List</span>[T], <span class="hljs-type">Optional</span>[T]]]:
    tail_nodes = <span class="hljs-built_in">filter</span>(<span class="hljs-keyword">lambda</span> n: graph.get_out_degree(n) == <span class="hljs-number">0</span>, graph.get_nodes())

    ret = []
    <span class="hljs-keyword">for</span> n <span class="hljs-keyword">in</span> tail_nodes:
        <span class="hljs-keyword">for</span> child <span class="hljs-keyword">in</span> graph.get_inputs(n):
            path_from_child = walk_ins_until_diverge(graph, child)
            <span class="hljs-keyword">if</span> path_from_child <span class="hljs-keyword">is</span> <span class="hljs-literal">None</span>:
                <span class="hljs-keyword">continue</span>
            diverging_node = path_from_child[<span class="hljs-number">0</span>]
            branch_path = path_from_child[<span class="hljs-number">1</span>:] + [n]
            converging_node = <span class="hljs-literal">None</span>
            path = (diverging_node, branch_path, converging_node)
            <span class="hljs-keyword">if</span> <span class="hljs-built_in">len</span>(branch_path) &lt;= branch_len:
                ret.append(path)
    <span class="hljs-keyword">return</span> ret


<span class="hljs-keyword">def</span> <span class="hljs-title function_">find_bubbles</span>(<span class="hljs-params">graph: Graph[T], branch_len: <span class="hljs-built_in">int</span></span>) -&gt; <span class="hljs-type">List</span>[<span class="hljs-type">Tuple</span>[<span class="hljs-type">Optional</span>[T], <span class="hljs-type">List</span>[T], <span class="hljs-type">Optional</span>[T]]]:
    branching_nodes = <span class="hljs-built_in">filter</span>(<span class="hljs-keyword">lambda</span> n: graph.get_out_degree(n) &gt; <span class="hljs-number">1</span>, graph.get_nodes())

    ret = []
    <span class="hljs-keyword">for</span> n <span class="hljs-keyword">in</span> branching_nodes:
        <span class="hljs-keyword">for</span> child <span class="hljs-keyword">in</span> graph.get_outputs(n):
            path_from_child = walk_outs_until_converge(graph, child)
            <span class="hljs-keyword">if</span> path_from_child <span class="hljs-keyword">is</span> <span class="hljs-literal">None</span>:
                <span class="hljs-keyword">continue</span>
            diverging_node = n
            branch_path = path_from_child[:-<span class="hljs-number">1</span>]
            converging_node = path_from_child[-<span class="hljs-number">1</span>]
            path = (diverging_node, branch_path, converging_node)
            <span class="hljs-keyword">if</span> <span class="hljs-built_in">len</span>(branch_path) &lt;= branch_len:
                ret.append(path)
    <span class="hljs-keyword">return</span> ret
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Fragments from sequencer:</p>
<ul>
<li>ATAGGAC scanned in 1.</li>
<li>ATTGGAC scanned in 55.</li>
<li>TTGGACA scanned in 30.</li>
<li>TGGACAA scanned in 30.</li>
<li>GGACAAT scanned in 30.</li>
<li>GACAATC scanned in 30.</li>
<li>ACAATCT scanned in 30.</li>
<li>ACAGTCT scanned in 1.</li>
<li>CAATCTC scanned in 30.</li>
<li>AATCTCG scanned in 30.</li>
<li>ATCTCGG scanned in 30.</li>
<li>TCTCGGG scanned in 30.</li>
<li>CTCGGGC scanned in 55.</li>
<li>CTCGTGC scanned in 1.</li>
</ul>
<p>Fragments after being broken to k=4:</p>
<ul>
<li>ATAG broken out 1 times, so it probably appears in the genome 0.01 times.</li>
<li>TAGG broken out 1 times, so it probably appears in the genome 0.01 times.</li>
<li>AGGA broken out 1 times, so it probably appears in the genome 0.01 times.</li>
<li>GGAC broken out 146 times, so it probably appears in the genome 1.0 times.</li>
<li>ATTG broken out 55 times, so it probably appears in the genome 1.0 times.</li>
<li>TTGG broken out 85 times, so it probably appears in the genome 1.0 times.</li>
<li>TGGA broken out 115 times, so it probably appears in the genome 1.0 times.</li>
<li>GACA broken out 120 times, so it probably appears in the genome 1.0 times.</li>
<li>ACAA broken out 120 times, so it probably appears in the genome 1.0 times.</li>
<li>CAAT broken out 120 times, so it probably appears in the genome 1.0 times.</li>
<li>AATC broken out 120 times, so it probably appears in the genome 1.0 times.</li>
<li>ATCT broken out 120 times, so it probably appears in the genome 1.0 times.</li>
<li>ACAG broken out 1 times, so it probably appears in the genome 0.01 times.</li>
<li>CAGT broken out 1 times, so it probably appears in the genome 0.01 times.</li>
<li>AGTC broken out 1 times, so it probably appears in the genome 0.01 times.</li>
<li>GTCT broken out 1 times, so it probably appears in the genome 0.01 times.</li>
<li>TCTC broken out 120 times, so it probably appears in the genome 1.0 times.</li>
<li>CTCG broken out 146 times, so it probably appears in the genome 1.0 times.</li>
<li>TCGG broken out 115 times, so it probably appears in the genome 1.0 times.</li>
<li>CGGG broken out 85 times, so it probably appears in the genome 1.0 times.</li>
<li>GGGC broken out 55 times, so it probably appears in the genome 1.0 times.</li>
<li>TCGT broken out 1 times, so it probably appears in the genome 0.01 times.</li>
<li>CGTG broken out 1 times, so it probably appears in the genome 0.01 times.</li>
<li>GTGC broken out 1 times, so it probably appears in the genome 0.01 times.</li>
</ul>
<p>De Bruijn graph:</p>
<p><img src="dot_58d24202dc65f3967dedac820f12af62.svg" alt="Dot diagram"></p>
<p>Problem paths:</p>
<ul>
<li>Src: ACA, Dst: TCT, Branch: CAA-&gt;AAT-&gt;ATC</li>
<li>Src: ACA, Dst: TCT, Branch: CAG-&gt;AGT-&gt;GTC</li>
<li>Src: None, Dst: GGA, Branch: ATA-&gt;TAG-&gt;AGG</li>
<li>Src: None, Dst: GGA, Branch: ATT-&gt;TTG-&gt;TGG</li>
<li>Src: TCG, Dst: None, Branch: CGG-&gt;GGG-&gt;GGC</li>
<li>Src: TCG, Dst: None, Branch: CGT-&gt;GTG-&gt;TGC</li>
</ul>
</div>
<p></p>
<a name="H_Find%20Contigs"></a><h3>Find Contigs</h3>
<p><a name="BM_(Algorithms%5C%2FDNA%20Assembly%2FFind%20Contigs)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FDNA%20Assembly%2FOverlap%20Graph)_TOPIC%2F">Algorithms/DNA Assembly/Overlap Graph</a></li>
<li><a href="#BM_(Algorithms%5C%2FDNA%20Assembly%2FDe%20Bruijn%20Graph)_TOPIC%2F">Algorithms/DNA Assembly/De Bruijn Graph</a></li>
<li><a href="#BM_(Algorithms%5C%2FDNA%20Assembly%2FFind%20Bubbles)_TOPIC%2F">Algorithms/DNA Assembly/Find Bubbles</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: Given an <a href="#BM_(overlap%20graph)%2Fi">overlap graph</a> or <a href="#BM_(de%20Bruijn%20graph)%2Fi">de Bruijn graph</a>, find the longest possible stretches of non-branching nodes. Each stretch will be a path that's either  ...</p>
<ul>
<li>
<p>a line: each node has an <a href="#BM_(indegree)%2Fi">indegree</a> and <a href="#BM_(outdegree)%2Fi">outdegree</a> of 1.</p>
<p><img src="svgbob_8f50e668ad56e2eeff3c39e6aa89792f6f9444f9.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>a cycle: each node has an <a href="#BM_(indegree)%2Fi">indegree</a> and <a href="#BM_(outdegree)%2Fi">outdegree</a> of 1 and it loops.</p>
<p><img src="svgbob_9a5c6dc10a8a016d6f43d4c9bdd6243d0ee5b332.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>a line sandwiched between branching nodes: nodes in between have an <a href="#BM_(indegree)%2Fi">indegree</a> and <a href="#BM_(outdegree)%2Fi">outdegree</a> of 1 but either...</p>
<ul>
<li>starts at a node where <a href="#BM_(indegree)%2Fi">indegree</a> != 1 but <a href="#BM_(outdegree)%2Fi">outdegree</a> == 1 (incoming branch),</li>
<li>or ends at a node where <a href="#BM_(indegree)%2Fi">indegree</a> == 1 but <a href="#BM_(outdegree)%2Fi">outdegree</a> != 1 (outgoing branch),</li>
<li>or both.</li>
</ul>
<p><img src="svgbob_057bcd0e831c50ca2384b8da9ea9cd1f46186b4a.svg" alt="Kroki diagram output"></p>
</li>
</ul>
<p>Each found path is called a <a href="#BM_(contig)s%3F%5Cb%2Fi">contig</a>: a contiguous piece of the graph. For example, ...</p>
<p><img src="svgbob_f186de7e9649b7a39a09b48a6c55ea97a221dce1.svg" alt="Kroki diagram output"></p>
<p><strong>WHY</strong>: An <a href="#BM_(overlap%20graph)%2Fi">overlap graph</a> / <a href="#BM_(de%20Bruijn%20graph)%2Fi">de Bruijn graph</a> represents all the possible ways a set of <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s may be stitched together to infer the full <a href="#BM_(genome)%2Fi">genome</a>. However, real-world complications make it impractical to guess the full <a href="#BM_(genome)%2Fi">genome</a>:</p>
<ul>
<li><a href="#BM_(fragment)_SEQ%2Fi">Fragment</a>s are for both strands (strand of <a href="#BM_(double%20stranded%20DNA%7Cdouble-stranded%20DNA)%2Fi">double-stranded DNA</a> a <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>'s from isn't known).</li>
<li><a href="#BM_(fragment)_SEQ%2Fi">Fragment</a>s may be missing (<a href="#BM_(sequencer)%2Fi">sequencer</a> didn't capture it).</li>
<li><a href="#BM_(fragment)_SEQ%2Fi">Fragment</a>s may have inconsistent <a href="#BM_(coverage)_SEQ%2Fi">coverage</a> (<a href="#BM_(sequencer)%2Fi">sequencer</a> captured it too many/few times).</li>
<li><a href="#BM_(fragment)_SEQ%2Fi">Fragment</a>s may be repeated (regions of the <a href="#BM_(genome)%2Fi">genome</a> may repeat).</li>
<li><a href="#BM_(fragment)_SEQ%2Fi">Fragment</a>s may have errors (<a href="#BM_(sequencer)%2Fi">sequencer</a> produced <a href="#BM_(sequencing%20error)%2Fi">sequencing error</a>s).</li>
<li><a href="#BM_(fragment)_SEQ%2Fi">Fragment</a>s may be stitch-able in more than one way (multiple <a href="#BM_(genome)%2Fi">genome</a> reconstruction guesses).</li>
<li><a href="#BM_(fragment)_SEQ%2Fi">Fragment</a>s may take a long time to stitch (computationally intensive).</li>
</ul>
<p>These complications result in graphs that are too tangled, <a href="#BM_(disconnected)%2Fi">disconnected</a>, etc... As such, the best someone can do is to pull out the <a href="#BM_(contig)s%3F%5Cb%2Fi">contig</a>s in the graph: unambiguous stretches of DNA.</p>
<p><strong>ALGORITHM</strong>:</p>
<p><a href="ch3_code/src/FindContigs.py">ch3_code/src/FindContigs.py</a> (lines 14 to 82):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">walk_until_non_1_to_1</span>(<span class="hljs-params">graph: Graph[T], node: T</span>) -&gt; <span class="hljs-type">Optional</span>[<span class="hljs-type">List</span>[T]]:
    ret = [node]
    ret_quick_lookup = {node}
    <span class="hljs-keyword">while</span> <span class="hljs-literal">True</span>:
        out_degree = graph.get_out_degree(node)
        in_degree = graph.get_in_degree(node)
        <span class="hljs-keyword">if</span> <span class="hljs-keyword">not</span>(in_degree == <span class="hljs-number">1</span> <span class="hljs-keyword">and</span> out_degree == <span class="hljs-number">1</span>):
            <span class="hljs-keyword">return</span> ret

        children = graph.get_outputs(node)
        child = <span class="hljs-built_in">next</span>(children)
        <span class="hljs-keyword">if</span> child <span class="hljs-keyword">in</span> ret_quick_lookup:
            <span class="hljs-keyword">return</span> ret

        node = child
        ret.append(node)
        ret_quick_lookup.add(node)


<span class="hljs-keyword">def</span> <span class="hljs-title function_">walk_until_loop</span>(<span class="hljs-params">graph: Graph[T], node: T</span>) -&gt; <span class="hljs-type">Optional</span>[<span class="hljs-type">List</span>[T]]:
    ret = [node]
    ret_quick_lookup = {node}
    <span class="hljs-keyword">while</span> <span class="hljs-literal">True</span>:
        out_degree = graph.get_out_degree(node)
        <span class="hljs-keyword">if</span> out_degree &gt; <span class="hljs-number">1</span> <span class="hljs-keyword">or</span> out_degree == <span class="hljs-number">0</span>:
            <span class="hljs-keyword">return</span> <span class="hljs-literal">None</span>

        children = graph.get_outputs(node)
        child = <span class="hljs-built_in">next</span>(children)
        <span class="hljs-keyword">if</span> child <span class="hljs-keyword">in</span> ret_quick_lookup:
            <span class="hljs-keyword">return</span> ret

        node = child
        ret.append(node)
        ret_quick_lookup.add(node)


<span class="hljs-keyword">def</span> <span class="hljs-title function_">find_maximal_non_branching_paths</span>(<span class="hljs-params">graph: Graph[T]</span>) -&gt; <span class="hljs-type">List</span>[<span class="hljs-type">List</span>[T]]:
    paths = []

    <span class="hljs-keyword">for</span> node <span class="hljs-keyword">in</span> graph.get_nodes():
        out_degree = graph.get_out_degree(node)
        in_degree = graph.get_in_degree(node)
        <span class="hljs-keyword">if</span> (in_degree == <span class="hljs-number">1</span> <span class="hljs-keyword">and</span> out_degree == <span class="hljs-number">1</span>) <span class="hljs-keyword">or</span> out_degree == <span class="hljs-number">0</span>:
            <span class="hljs-keyword">continue</span>
        <span class="hljs-keyword">for</span> child <span class="hljs-keyword">in</span> graph.get_outputs(node):
            path_from_child = walk_until_non_1_to_1(graph, child)
            <span class="hljs-keyword">if</span> path_from_child <span class="hljs-keyword">is</span> <span class="hljs-literal">None</span>:
                <span class="hljs-keyword">continue</span>
            path = [node] + path_from_child
            paths.append(path)

    skip_nodes = <span class="hljs-built_in">set</span>()
    <span class="hljs-keyword">for</span> node <span class="hljs-keyword">in</span> graph.get_nodes():
        <span class="hljs-keyword">if</span> node <span class="hljs-keyword">in</span> skip_nodes:
            <span class="hljs-keyword">continue</span>
        out_degree = graph.get_out_degree(node)
        in_degree = graph.get_in_degree(node)
        <span class="hljs-keyword">if</span> <span class="hljs-keyword">not</span> (in_degree == <span class="hljs-number">1</span> <span class="hljs-keyword">and</span> out_degree == <span class="hljs-number">1</span>) <span class="hljs-keyword">or</span> out_degree == <span class="hljs-number">0</span>:
            <span class="hljs-keyword">continue</span>
        path = walk_until_loop(graph, node)
        <span class="hljs-keyword">if</span> path <span class="hljs-keyword">is</span> <span class="hljs-literal">None</span>:
            <span class="hljs-keyword">continue</span>
        path = path + [node]
        paths.append(path)
        skip_nodes |= <span class="hljs-built_in">set</span>(path)

    <span class="hljs-keyword">return</span> paths
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the fragments ['TGG', 'GGT', 'GGT', 'GTG', 'CAC', 'ACC', 'CCA'], the de Bruijn graph is...</p>
<p><img src="dot_c3d570f99942f291a0b7bb8aa20edc6d.svg" alt="Dot diagram"></p>
<p>The following contigs were found...</p>
<p>GG-&gt;GT</p>
<p>GG-&gt;GT</p>
<p>GT-&gt;TG-&gt;GG</p>
<p>CA-&gt;AC-&gt;CC-&gt;CA</p>
</div>
<p></p>
<a name="H_Peptide%20Sequence"></a><h2>Peptide Sequence</h2>
<p><a name="BM_(Algorithms%5C%2FPeptide%20Sequence)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FK-mer)_TOPIC%2F">Algorithms/K-mer</a></li>
</ul>
</div>
<p>A <a href="#BM_(peptide)%2Fi">peptide</a> is a miniature protein consisting of a chain of <a href="#BM_(amino%20acid)%2Fi">amino acid</a>s anywhere between 2 to 100 <a href="#BM_(amino%20acid)%2Fi">amino acid</a>s in length. <a href="#BM_(peptide)%2Fi">Peptide</a>s are created through two mechanisms:</p>
<ol>
<li>
<p>ribosomal <a href="#BM_(peptide)%2Fi">peptide</a>s: DNA gets <a href="#BM_(transcription%7Ctranscribed%7Ctranscribe)%2Fi">transcribed</a> to mRNA (<a href="#BM_(transcription%7Ctranscribed%7Ctranscribe)%2Fi">transcription</a>), which in turn gets <a href="#BM_(translation%7Ctranslated%7Ctranslate)%2Fi">translated</a> by the ribosome into a <a href="#BM_(peptide)%2Fi">peptide</a> (<a href="#BM_(translation%7Ctranslated%7Ctranslate)%2Fi">translation</a>).</p>
<p><img src="svgbob_3b69a590a67c8d8db6ca97d50bdeec9cb4c1ffee.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a href="#BM_(non-ribosomal%20peptide)%2Fi">non-ribosomal peptide</a>s: proteins called <a href="#BM_(non-ribosomal%20peptide%20synthetase%7CNRP%20synthetase)%2Fi">NRP synthetase</a> construct <a href="#BM_(peptide)%2Fi">peptide</a>s one <a href="#BM_(amino%20acid)%2Fi">amino acid</a> at a time.</p>
<p><img src="svgbob_beb44b4a49d8f6713eefec469487c6b717b811f4.svg" alt="Kroki diagram output"></p>
</li>
</ol>
<p>For ribosomal <a href="#BM_(peptide)%2Fi">peptide</a>s, each <a href="#BM_(amino%20acid)%2Fi">amino acid</a> is <a href="#BM_(encode%7Cencoding%7Cdecode%7Cdecoding)%2Fi">encode</a>d as a DNA <a href="#BM_(sequence)%2Fi">sequence</a> of length 3. This 3 length DNA <a href="#BM_(sequence)%2Fi">sequence</a> is referred to as a <a href="#BM_(codon%7Cgenetic%20code)%2Fi">codon</a>. By knowing which <a href="#BM_(codon%7Cgenetic%20code)%2Fi">codon</a>s map to which <a href="#BM_(amino%20acid)%2Fi">amino acid</a>s, the ...</p>
<ul>
<li><a href="#BM_(peptide)%2Fi">peptide</a> <a href="#BM_(sequence)%2Fi">sequence</a> can be determined by mapping from DNA to <a href="#BM_(codon%7Cgenetic%20code)%2Fi">codon</a>s (you know the <a href="#BM_(peptide)%2Fi">peptide</a> just by looking at the DNA).</li>
<li><a href="#BM_(peptide)%2Fi">peptide</a> <a href="#BM_(sequence)%2Fi">sequence</a> can be searched for in DNA by finding <a href="#BM_(codon%7Cgenetic%20code)%2Fi">codon</a>s (you can see if the <a href="#BM_(peptide)%2Fi">peptide</a> is <a href="#BM_(encode%7Cencoding%7Cdecode%7Cdecoding)%2Fi">encode</a>d in a <a href="#BM_(genome)%2Fi">genome</a>).</li>
</ul>
<p>For <a href="#BM_(non-ribosomal%20peptide)%2Fi">non-ribosomal peptide</a>s, a sample of the <a href="#BM_(peptide)%2Fi">peptide</a> needs to be isolated and passed through a <a href="#BM_(mass%20spectrometer%7Cmass%20spectrometry)%2Fi">mass spectrometer</a>. A <a href="#BM_(mass%20spectrometer%7Cmass%20spectrometry)%2Fi">mass spectrometer</a> is a device that shatters and bins molecules by their mass-to-charge ratio: Given a sample of molecules, the device randomly shatters each molecule in the sample (forming ions), then bins each ion by its mass-to-charge ratio (<span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mfrac><mi>m</mi><mi>z</mi></mfrac></mrow><annotation encoding="application/x-tex">\frac{m}{z}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.040392em;vertical-align:-0.345em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.695392em;"><span style="top:-2.6550000000000002em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.04398em;">z</span></span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.394em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">m</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.345em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span>).</p>
<p>The output of a <a href="#BM_(mass%20spectrometer%7Cmass%20spectrometry)%2Fi">mass spectrometer</a> is a plot called a <a href="#BM_(spectrum)_MS%2Fi">spectrum</a>. The plot's ...</p>
<ul>
<li>x-axis is the mass-to-charge ratio.</li>
<li>y-axis is the intensity of that mass-to-charge ratio (how much more / less did that mass-to-charge appear compared to the others).</li>
</ul>
<p><img src="svgbob_2b5ef43ae7572c2c6bf63687c53860a71830d160.svg" alt="Kroki diagram output"></p>
<p>For example, given a sample containing multiple instances of the <a href="#BM_(linear%20peptide)%2Fi">linear peptide</a> NQY, the <a href="#BM_(mass%20spectrometer%7Cmass%20spectrometry)%2Fi">mass spectrometer</a> will take each instance of NQY and randomly break the bonds between its <a href="#BM_(amino%20acid)%2Fi">amino acid</a>s:</p>
<p><img src="svgbob_33c109cffe376e25dfa6ba85b256a1bb307dcd76.svg" alt="Kroki diagram output"></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>How does it know to break the bonds holding <a href="#BM_(amino%20acid)%2Fi">amino acid</a>s together and not bonds within the <a href="#BM_(amino%20acid)%2Fi">amino acid</a>s themselves? My guess is that the bonds coupling one <a href="#BM_(amino%20acid)%2Fi">amino acid</a> to another are much weaker than the bonds holding an individual <a href="#BM_(amino%20acid)%2Fi">amino acid</a> together -- it's more likely that the weaker bonds will be broken.</p>
</div>
<p>Each <a href="#BM_(subpeptide)%2Fi">subpeptide</a> then will have its mass-to-charge ratio measured, which in turn gets converted to a set of potential masses by performing basic math. With these potential masses, it's possible to infer the <a href="#BM_(sequence)%2Fi">sequence</a> of the <a href="#BM_(peptide)%2Fi">peptide</a>.</p>
<p>Special consideration needs to be given to the real-world practical problems with <a href="#BM_(mass%20spectrometer%7Cmass%20spectrometry)%2Fi">mass spectrometry</a>. Specifically, the <a href="#BM_(spectrum)_MS%2Fi">spectrum</a> given back by a <a href="#BM_(mass%20spectrometer%7Cmass%20spectrometry)%2Fi">mass spectrometer</a> will very likely ...</p>
<ul>
<li>miss mass-to-charge ratios for some fragments of the intended molecule (missing entries).</li>
<li>include mass-to-charge ratios for fragments of unintended molecules (faulty entries).</li>
<li>have noisy mass-to-charge ratios.</li>
</ul>
<p>The following table contains a list of <a href="#BM_(proteinogenic%20amino%20acid)%2Fi">proteinogenic amino acid</a>s with their masses and <a href="#BM_(codon%7Cgenetic%20code)%2Fi">codon</a> mappings:</p>
<table>
<thead>
<tr>
<th>1 Letter Code</th>
<th>3 Letter Code</th>
<th><a href="#BM_(amino%20acid)%2Fi">Amino Acid</a></th>
<th><a href="#BM_(codon%7Cgenetic%20code)%2Fi">Codon</a>s</th>
<th>Monoisotopic Mass (<a href="#BM_(dalton)%2Fi">dalton</a>s)</th>
</tr>
</thead>
<tbody>
<tr>
<td>A</td>
<td>Ala</td>
<td>Alanine</td>
<td>GCA, GCC, GCG, GCU</td>
<td>71.04</td>
</tr>
<tr>
<td>C</td>
<td>Cys</td>
<td>Cysteine</td>
<td>UGC, UGU</td>
<td>103.01</td>
</tr>
<tr>
<td>D</td>
<td>Asp</td>
<td>Aspartic acid</td>
<td>GAC, GAU</td>
<td>115.03</td>
</tr>
<tr>
<td>E</td>
<td>Glu</td>
<td>Glutamic acid</td>
<td>GAA, GAG</td>
<td>129.04</td>
</tr>
<tr>
<td>F</td>
<td>Phe</td>
<td>Phenylalanine</td>
<td>UUC, UUU</td>
<td>147.07</td>
</tr>
<tr>
<td>G</td>
<td>Gly</td>
<td>Glycine</td>
<td>GGA, GGC, GGG, GGU</td>
<td>57.02</td>
</tr>
<tr>
<td>H</td>
<td>His</td>
<td>Histidine</td>
<td>CAC, CAU</td>
<td>137.06</td>
</tr>
<tr>
<td>I</td>
<td>Ile</td>
<td>Isoleucine</td>
<td>AUA, AUC, AUU</td>
<td>113.08</td>
</tr>
<tr>
<td>K</td>
<td>Lys</td>
<td>Lysine</td>
<td>AAA, AAG</td>
<td>128.09</td>
</tr>
<tr>
<td>L</td>
<td>Leu</td>
<td>Leucine</td>
<td>CUA, CUC, CUG, CUU, UUA, UUG</td>
<td>113.08</td>
</tr>
<tr>
<td>M</td>
<td>Met</td>
<td>Methionine</td>
<td>AUG</td>
<td>131.04</td>
</tr>
<tr>
<td>N</td>
<td>Asn</td>
<td>Asparagine</td>
<td>AAC, AAU</td>
<td>114.04</td>
</tr>
<tr>
<td>P</td>
<td>Pro</td>
<td>Proline</td>
<td>CCA, CCC, CCG, CCU</td>
<td>97.05</td>
</tr>
<tr>
<td>Q</td>
<td>Gln</td>
<td>Glutamine</td>
<td>CAA, CAG</td>
<td>128.06</td>
</tr>
<tr>
<td>R</td>
<td>Arg</td>
<td>Arginine</td>
<td>AGA, AGG, CGA, CGC, CGG, CGU</td>
<td>156.1</td>
</tr>
<tr>
<td>S</td>
<td>Ser</td>
<td>Serine</td>
<td>AGC, AGU, UCA, UCC, UCG, UCU</td>
<td>87.03</td>
</tr>
<tr>
<td>T</td>
<td>Thr</td>
<td>Threonine</td>
<td>ACA, ACC, ACG, ACU</td>
<td>101.05</td>
</tr>
<tr>
<td>V</td>
<td>Val</td>
<td>Valine</td>
<td>GUA, GUC, GUG, GUU</td>
<td>99.07</td>
</tr>
<tr>
<td>W</td>
<td>Trp</td>
<td>Tryptophan</td>
<td>UGG</td>
<td>186.08</td>
</tr>
<tr>
<td>Y</td>
<td>Tyr</td>
<td>Tyrosine</td>
<td>UAC, UAU</td>
<td>163.06</td>
</tr>
<tr>
<td>*</td>
<td>*</td>
<td><strong>STOP</strong></td>
<td>UAA, UAG, UGA</td>
<td></td>
</tr>
</tbody>
</table>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The stop marker tells the ribosome to stop translating / the protein is complete. The <a href="#BM_(codon%7Cgenetic%20code)%2Fi">codon</a>s are listed as <a href="#BM_(ribonucleotide)%2Fi">ribonucleotide</a>s (RNA). For nucleotides (DNA), swap U with T.</p>
</div>
<a name="H_Codon%20Encode"></a><h3>Codon Encode</h3>
<p><strong>WHAT</strong>: Given a DNA <a href="#BM_(sequence)%2Fi">sequence</a>, map each <a href="#BM_(codon%7Cgenetic%20code)%2Fi">codon</a> to the <a href="#BM_(amino%20acid)%2Fi">amino acid</a> it represents. In total, there are 6 different ways that a DNA <a href="#BM_(sequence)%2Fi">sequence</a> could be <a href="#BM_(translation%7Ctranslated%7Ctranslate)%2Fi">translated</a>:</p>
<ol>
<li>Since the length of a <a href="#BM_(codon%7Cgenetic%20code)%2Fi">codon</a> is 3, the <a href="#BM_(encode%7Cencoding%7Cdecode%7Cdecoding)%2Fi">encoding</a> of the <a href="#BM_(peptide)%2Fi">peptide</a> could start from offset 0, 1, or 2 (referred to as <a href="#BM_(reading%20frame)%2Fi">reading frame</a>s).</li>
<li>Since DNA is double stranded, either the DNA <a href="#BM_(sequence)%2Fi">sequence</a> or its <a href="#BM_(reverse%20complement)%2Fi">reverse complement</a> could represent the <a href="#BM_(peptide)%2Fi">peptide</a>.</li>
</ol>
<p><strong>WHY</strong>: The composition of a <a href="#BM_(peptide)%2Fi">peptide</a> can be determined from the DNA <a href="#BM_(sequence)%2Fi">sequence</a> that <a href="#BM_(encode%7Cencoding%7Cdecode%7Cdecoding)%2Fi">encode</a>s it.</p>
<p><strong>ALGORITHM</strong>:</p>
<p><a href="ch4_code/src/helpers/AminoAcidUtils.py">ch4_code/src/helpers/AminoAcidUtils.py</a> (lines 4 to 24):</p>
<pre class="hljs"><code>_codon_to_amino_acid = {<span class="hljs-string">'AAA'</span>: <span class="hljs-string">'K'</span>, <span class="hljs-string">'AAC'</span>: <span class="hljs-string">'N'</span>, <span class="hljs-string">'AAG'</span>: <span class="hljs-string">'K'</span>, <span class="hljs-string">'AAU'</span>: <span class="hljs-string">'N'</span>, <span class="hljs-string">'ACA'</span>: <span class="hljs-string">'T'</span>, <span class="hljs-string">'ACC'</span>: <span class="hljs-string">'T'</span>, <span class="hljs-string">'ACG'</span>: <span class="hljs-string">'T'</span>, <span class="hljs-string">'ACU'</span>: <span class="hljs-string">'T'</span>,
                        <span class="hljs-string">'AGA'</span>: <span class="hljs-string">'R'</span>, <span class="hljs-string">'AGC'</span>: <span class="hljs-string">'S'</span>, <span class="hljs-string">'AGG'</span>: <span class="hljs-string">'R'</span>, <span class="hljs-string">'AGU'</span>: <span class="hljs-string">'S'</span>, <span class="hljs-string">'AUA'</span>: <span class="hljs-string">'I'</span>, <span class="hljs-string">'AUC'</span>: <span class="hljs-string">'I'</span>, <span class="hljs-string">'AUG'</span>: <span class="hljs-string">'M'</span>, <span class="hljs-string">'AUU'</span>: <span class="hljs-string">'I'</span>,
                        <span class="hljs-string">'CAA'</span>: <span class="hljs-string">'Q'</span>, <span class="hljs-string">'CAC'</span>: <span class="hljs-string">'H'</span>, <span class="hljs-string">'CAG'</span>: <span class="hljs-string">'Q'</span>, <span class="hljs-string">'CAU'</span>: <span class="hljs-string">'H'</span>, <span class="hljs-string">'CCA'</span>: <span class="hljs-string">'P'</span>, <span class="hljs-string">'CCC'</span>: <span class="hljs-string">'P'</span>, <span class="hljs-string">'CCG'</span>: <span class="hljs-string">'P'</span>, <span class="hljs-string">'CCU'</span>: <span class="hljs-string">'P'</span>,
                        <span class="hljs-string">'CGA'</span>: <span class="hljs-string">'R'</span>, <span class="hljs-string">'CGC'</span>: <span class="hljs-string">'R'</span>, <span class="hljs-string">'CGG'</span>: <span class="hljs-string">'R'</span>, <span class="hljs-string">'CGU'</span>: <span class="hljs-string">'R'</span>, <span class="hljs-string">'CUA'</span>: <span class="hljs-string">'L'</span>, <span class="hljs-string">'CUC'</span>: <span class="hljs-string">'L'</span>, <span class="hljs-string">'CUG'</span>: <span class="hljs-string">'L'</span>, <span class="hljs-string">'CUU'</span>: <span class="hljs-string">'L'</span>,
                        <span class="hljs-string">'GAA'</span>: <span class="hljs-string">'E'</span>, <span class="hljs-string">'GAC'</span>: <span class="hljs-string">'D'</span>, <span class="hljs-string">'GAG'</span>: <span class="hljs-string">'E'</span>, <span class="hljs-string">'GAU'</span>: <span class="hljs-string">'D'</span>, <span class="hljs-string">'GCA'</span>: <span class="hljs-string">'A'</span>, <span class="hljs-string">'GCC'</span>: <span class="hljs-string">'A'</span>, <span class="hljs-string">'GCG'</span>: <span class="hljs-string">'A'</span>, <span class="hljs-string">'GCU'</span>: <span class="hljs-string">'A'</span>,
                        <span class="hljs-string">'GGA'</span>: <span class="hljs-string">'G'</span>, <span class="hljs-string">'GGC'</span>: <span class="hljs-string">'G'</span>, <span class="hljs-string">'GGG'</span>: <span class="hljs-string">'G'</span>, <span class="hljs-string">'GGU'</span>: <span class="hljs-string">'G'</span>, <span class="hljs-string">'GUA'</span>: <span class="hljs-string">'V'</span>, <span class="hljs-string">'GUC'</span>: <span class="hljs-string">'V'</span>, <span class="hljs-string">'GUG'</span>: <span class="hljs-string">'V'</span>, <span class="hljs-string">'GUU'</span>: <span class="hljs-string">'V'</span>,
                        <span class="hljs-string">'UAA'</span>: <span class="hljs-string">'*'</span>, <span class="hljs-string">'UAC'</span>: <span class="hljs-string">'Y'</span>, <span class="hljs-string">'UAG'</span>: <span class="hljs-string">'*'</span>, <span class="hljs-string">'UAU'</span>: <span class="hljs-string">'Y'</span>, <span class="hljs-string">'UCA'</span>: <span class="hljs-string">'S'</span>, <span class="hljs-string">'UCC'</span>: <span class="hljs-string">'S'</span>, <span class="hljs-string">'UCG'</span>: <span class="hljs-string">'S'</span>, <span class="hljs-string">'UCU'</span>: <span class="hljs-string">'S'</span>,
                        <span class="hljs-string">'UGA'</span>: <span class="hljs-string">'*'</span>, <span class="hljs-string">'UGC'</span>: <span class="hljs-string">'C'</span>, <span class="hljs-string">'UGG'</span>: <span class="hljs-string">'W'</span>, <span class="hljs-string">'UGU'</span>: <span class="hljs-string">'C'</span>, <span class="hljs-string">'UUA'</span>: <span class="hljs-string">'L'</span>, <span class="hljs-string">'UUC'</span>: <span class="hljs-string">'F'</span>, <span class="hljs-string">'UUG'</span>: <span class="hljs-string">'L'</span>, <span class="hljs-string">'UUU'</span>: <span class="hljs-string">'F'</span>}

_amino_acid_to_codons = <span class="hljs-built_in">dict</span>()
<span class="hljs-keyword">for</span> k, v <span class="hljs-keyword">in</span> _codon_to_amino_acid.items():
    _amino_acid_to_codons.setdefault(v, []).append(k)


<span class="hljs-keyword">def</span> <span class="hljs-title function_">codon_to_amino_acid</span>(<span class="hljs-params">rna: <span class="hljs-built_in">str</span></span>) -&gt; <span class="hljs-type">Optional</span>[<span class="hljs-built_in">str</span>]:
    <span class="hljs-keyword">return</span> _codon_to_amino_acid.get(rna)


<span class="hljs-keyword">def</span> <span class="hljs-title function_">amino_acid_to_codons</span>(<span class="hljs-params">codon: <span class="hljs-built_in">str</span></span>) -&gt; <span class="hljs-type">Optional</span>[<span class="hljs-type">List</span>[<span class="hljs-built_in">str</span>]]:
    <span class="hljs-keyword">return</span> _amino_acid_to_codons.get(codon)
</code></pre>
<p><a href="ch4_code/src/EncodePeptide.py">ch4_code/src/EncodePeptide.py</a> (lines 9 to 26):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">encode_peptide</span>(<span class="hljs-params">dna: <span class="hljs-built_in">str</span></span>) -&gt; <span class="hljs-built_in">str</span>:
    rna = dna_to_rna(dna)
    protein_seq = <span class="hljs-string">''</span>
    <span class="hljs-keyword">for</span> codon <span class="hljs-keyword">in</span> split_to_size(rna, <span class="hljs-number">3</span>):
        codon_str = <span class="hljs-string">''</span>.join(codon)
        protein_seq += codon_to_amino_acid(codon_str)
    <span class="hljs-keyword">return</span> protein_seq


<span class="hljs-keyword">def</span> <span class="hljs-title function_">encode_peptides_all_readingframes</span>(<span class="hljs-params">dna: <span class="hljs-built_in">str</span></span>) -&gt; <span class="hljs-type">List</span>[<span class="hljs-built_in">str</span>]:
    ret = []
    <span class="hljs-keyword">for</span> dna_ <span class="hljs-keyword">in</span> (dna, dna_reverse_complement(dna)):
        <span class="hljs-keyword">for</span> rf_start <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(<span class="hljs-number">3</span>):
            rf_end = <span class="hljs-built_in">len</span>(dna_) - ((<span class="hljs-built_in">len</span>(dna_) - rf_start) % <span class="hljs-number">3</span>)
            peptide = encode_peptide(dna_[rf_start:rf_end])
            ret.append(peptide)
    <span class="hljs-keyword">return</span> ret
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given AAAAGAACCTAATCTTAAAGGAGATGATGATTCTAA, the possible peptide encodings are...</p>
<ul>
<li>KRT*S*RR**F*</li>
<li>KEPNLKGDDDS</li>
<li>KNLILKEMMIL</li>
<li>LESSSPLRLGSF</li>
<li>*NHHLL*D*VL</li>
<li>RIIISFKIRFF</li>
</ul>
</div>
<p></p>
<a name="H_Codon%20Decode"></a><h3>Codon Decode</h3>
<p><strong>WHAT</strong>: Given a <a href="#BM_(peptide)%2Fi">peptide</a>, map each <a href="#BM_(amino%20acid)%2Fi">amino acid</a> to the DNA <a href="#BM_(sequence)%2Fi">sequence</a>s it represents. Since each <a href="#BM_(amino%20acid)%2Fi">amino acid</a> can map to multiple <a href="#BM_(codon%7Cgenetic%20code)%2Fi">codon</a>s, there may be multiple DNA <a href="#BM_(sequence)%2Fi">sequence</a>s for a single <a href="#BM_(peptide)%2Fi">peptide</a>.</p>
<p><strong>WHY</strong>: The DNA <a href="#BM_(sequence)%2Fi">sequence</a>s that <a href="#BM_(encode%7Cencoding%7Cdecode%7Cdecoding)%2Fi">encode</a> a <a href="#BM_(peptide)%2Fi">peptide</a> can be determined from the <a href="#BM_(peptide)%2Fi">peptide</a> itself.</p>
<p><strong>ALGORITHM</strong>:</p>
<p><a href="ch4_code/src/helpers/AminoAcidUtils.py">ch4_code/src/helpers/AminoAcidUtils.py</a> (lines 4 to 24):</p>
<pre class="hljs"><code>_codon_to_amino_acid = {<span class="hljs-string">'AAA'</span>: <span class="hljs-string">'K'</span>, <span class="hljs-string">'AAC'</span>: <span class="hljs-string">'N'</span>, <span class="hljs-string">'AAG'</span>: <span class="hljs-string">'K'</span>, <span class="hljs-string">'AAU'</span>: <span class="hljs-string">'N'</span>, <span class="hljs-string">'ACA'</span>: <span class="hljs-string">'T'</span>, <span class="hljs-string">'ACC'</span>: <span class="hljs-string">'T'</span>, <span class="hljs-string">'ACG'</span>: <span class="hljs-string">'T'</span>, <span class="hljs-string">'ACU'</span>: <span class="hljs-string">'T'</span>,
                        <span class="hljs-string">'AGA'</span>: <span class="hljs-string">'R'</span>, <span class="hljs-string">'AGC'</span>: <span class="hljs-string">'S'</span>, <span class="hljs-string">'AGG'</span>: <span class="hljs-string">'R'</span>, <span class="hljs-string">'AGU'</span>: <span class="hljs-string">'S'</span>, <span class="hljs-string">'AUA'</span>: <span class="hljs-string">'I'</span>, <span class="hljs-string">'AUC'</span>: <span class="hljs-string">'I'</span>, <span class="hljs-string">'AUG'</span>: <span class="hljs-string">'M'</span>, <span class="hljs-string">'AUU'</span>: <span class="hljs-string">'I'</span>,
                        <span class="hljs-string">'CAA'</span>: <span class="hljs-string">'Q'</span>, <span class="hljs-string">'CAC'</span>: <span class="hljs-string">'H'</span>, <span class="hljs-string">'CAG'</span>: <span class="hljs-string">'Q'</span>, <span class="hljs-string">'CAU'</span>: <span class="hljs-string">'H'</span>, <span class="hljs-string">'CCA'</span>: <span class="hljs-string">'P'</span>, <span class="hljs-string">'CCC'</span>: <span class="hljs-string">'P'</span>, <span class="hljs-string">'CCG'</span>: <span class="hljs-string">'P'</span>, <span class="hljs-string">'CCU'</span>: <span class="hljs-string">'P'</span>,
                        <span class="hljs-string">'CGA'</span>: <span class="hljs-string">'R'</span>, <span class="hljs-string">'CGC'</span>: <span class="hljs-string">'R'</span>, <span class="hljs-string">'CGG'</span>: <span class="hljs-string">'R'</span>, <span class="hljs-string">'CGU'</span>: <span class="hljs-string">'R'</span>, <span class="hljs-string">'CUA'</span>: <span class="hljs-string">'L'</span>, <span class="hljs-string">'CUC'</span>: <span class="hljs-string">'L'</span>, <span class="hljs-string">'CUG'</span>: <span class="hljs-string">'L'</span>, <span class="hljs-string">'CUU'</span>: <span class="hljs-string">'L'</span>,
                        <span class="hljs-string">'GAA'</span>: <span class="hljs-string">'E'</span>, <span class="hljs-string">'GAC'</span>: <span class="hljs-string">'D'</span>, <span class="hljs-string">'GAG'</span>: <span class="hljs-string">'E'</span>, <span class="hljs-string">'GAU'</span>: <span class="hljs-string">'D'</span>, <span class="hljs-string">'GCA'</span>: <span class="hljs-string">'A'</span>, <span class="hljs-string">'GCC'</span>: <span class="hljs-string">'A'</span>, <span class="hljs-string">'GCG'</span>: <span class="hljs-string">'A'</span>, <span class="hljs-string">'GCU'</span>: <span class="hljs-string">'A'</span>,
                        <span class="hljs-string">'GGA'</span>: <span class="hljs-string">'G'</span>, <span class="hljs-string">'GGC'</span>: <span class="hljs-string">'G'</span>, <span class="hljs-string">'GGG'</span>: <span class="hljs-string">'G'</span>, <span class="hljs-string">'GGU'</span>: <span class="hljs-string">'G'</span>, <span class="hljs-string">'GUA'</span>: <span class="hljs-string">'V'</span>, <span class="hljs-string">'GUC'</span>: <span class="hljs-string">'V'</span>, <span class="hljs-string">'GUG'</span>: <span class="hljs-string">'V'</span>, <span class="hljs-string">'GUU'</span>: <span class="hljs-string">'V'</span>,
                        <span class="hljs-string">'UAA'</span>: <span class="hljs-string">'*'</span>, <span class="hljs-string">'UAC'</span>: <span class="hljs-string">'Y'</span>, <span class="hljs-string">'UAG'</span>: <span class="hljs-string">'*'</span>, <span class="hljs-string">'UAU'</span>: <span class="hljs-string">'Y'</span>, <span class="hljs-string">'UCA'</span>: <span class="hljs-string">'S'</span>, <span class="hljs-string">'UCC'</span>: <span class="hljs-string">'S'</span>, <span class="hljs-string">'UCG'</span>: <span class="hljs-string">'S'</span>, <span class="hljs-string">'UCU'</span>: <span class="hljs-string">'S'</span>,
                        <span class="hljs-string">'UGA'</span>: <span class="hljs-string">'*'</span>, <span class="hljs-string">'UGC'</span>: <span class="hljs-string">'C'</span>, <span class="hljs-string">'UGG'</span>: <span class="hljs-string">'W'</span>, <span class="hljs-string">'UGU'</span>: <span class="hljs-string">'C'</span>, <span class="hljs-string">'UUA'</span>: <span class="hljs-string">'L'</span>, <span class="hljs-string">'UUC'</span>: <span class="hljs-string">'F'</span>, <span class="hljs-string">'UUG'</span>: <span class="hljs-string">'L'</span>, <span class="hljs-string">'UUU'</span>: <span class="hljs-string">'F'</span>}

_amino_acid_to_codons = <span class="hljs-built_in">dict</span>()
<span class="hljs-keyword">for</span> k, v <span class="hljs-keyword">in</span> _codon_to_amino_acid.items():
    _amino_acid_to_codons.setdefault(v, []).append(k)


<span class="hljs-keyword">def</span> <span class="hljs-title function_">codon_to_amino_acid</span>(<span class="hljs-params">rna: <span class="hljs-built_in">str</span></span>) -&gt; <span class="hljs-type">Optional</span>[<span class="hljs-built_in">str</span>]:
    <span class="hljs-keyword">return</span> _codon_to_amino_acid.get(rna)


<span class="hljs-keyword">def</span> <span class="hljs-title function_">amino_acid_to_codons</span>(<span class="hljs-params">codon: <span class="hljs-built_in">str</span></span>) -&gt; <span class="hljs-type">Optional</span>[<span class="hljs-type">List</span>[<span class="hljs-built_in">str</span>]]:
    <span class="hljs-keyword">return</span> _amino_acid_to_codons.get(codon)
</code></pre>
<p><a href="ch4_code/src/DecodePeptide.py">ch4_code/src/DecodePeptide.py</a> (lines 8 to 27):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">decode_peptide</span>(<span class="hljs-params">peptide: <span class="hljs-built_in">str</span></span>) -&gt; <span class="hljs-type">List</span>[<span class="hljs-built_in">str</span>]:
    <span class="hljs-keyword">def</span> <span class="hljs-title function_">dfs</span>(<span class="hljs-params">subpeptide: <span class="hljs-built_in">str</span>, dna: <span class="hljs-built_in">str</span>, ret: <span class="hljs-type">List</span>[<span class="hljs-built_in">str</span>]</span>) -&gt; <span class="hljs-literal">None</span>:
        <span class="hljs-keyword">if</span> <span class="hljs-built_in">len</span>(subpeptide) == <span class="hljs-number">0</span>:
            ret.append(dna)
            <span class="hljs-keyword">return</span>
        aa = subpeptide[<span class="hljs-number">0</span>]
        <span class="hljs-keyword">for</span> codon <span class="hljs-keyword">in</span> amino_acid_to_codons(aa):
            dfs(subpeptide[<span class="hljs-number">1</span>:], dna + rna_to_dna(codon), ret)
    dnas = []
    dfs(peptide, <span class="hljs-string">''</span>, dnas)
    <span class="hljs-keyword">return</span> dnas


<span class="hljs-keyword">def</span> <span class="hljs-title function_">decode_peptide_count</span>(<span class="hljs-params">peptide: <span class="hljs-built_in">str</span></span>) -&gt; <span class="hljs-built_in">int</span>:
    count = <span class="hljs-number">1</span>
    <span class="hljs-keyword">for</span> ch <span class="hljs-keyword">in</span> peptide:
        vals = amino_acid_to_codons(ch)
        count *= <span class="hljs-built_in">len</span>(vals)
    <span class="hljs-keyword">return</span> count
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given NQY, the possible DNA encodings are...</p>
<ul>
<li>AACCAATAC</li>
<li>AACCAATAT</li>
<li>AACCAGTAC</li>
<li>AACCAGTAT</li>
<li>AATCAATAC</li>
<li>AATCAATAT</li>
<li>AATCAGTAC</li>
<li>AATCAGTAT</li>
</ul>
</div>
<p></p>
<a name="H_Experimental%20Spectrum"></a><h3>Experimental Spectrum</h3>
<p><a name="BM_(Algorithms%5C%2FPeptide%20Sequence%2FExperimental%20Spectrum)_TOPIC%2F"></a><strong></strong></p>
<p><strong>WHAT</strong>: Given a <a href="#BM_(spectrum)_MS%2Fi">spectrum</a> for a <a href="#BM_(peptide)%2Fi">peptide</a>, derive a set of potential masses from the mass-to-charge ratios. These potential masses are referred to as an <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>.</p>
<p><strong>WHY</strong>: A <a href="#BM_(peptide)%2Fi">peptide</a>'s <a href="#BM_(sequence)%2Fi">sequence</a> can be inferred from a list of its potential <a href="#BM_(subpeptide)%2Fi">subpeptide</a> masses.</p>
<p><strong>ALGORITHM</strong>:</p>
<p>Prior to deriving masses from a <a href="#BM_(spectrum)_MS%2Fi">spectrum</a>, filter out low intensity mass-to-charge ratios. The remaining mass-to-charge ratios are converted to potential masses using <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mfrac><mi>m</mi><mi>z</mi></mfrac><mo>⋅</mo><mi>z</mi><mo>=</mo><mi>m</mi></mrow><annotation encoding="application/x-tex">\frac{m}{z} \cdot z = m</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.040392em;vertical-align:-0.345em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.695392em;"><span style="top:-2.6550000000000002em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.04398em;">z</span></span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.394em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">m</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.345em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.43056em;vertical-align:0em;"></span><span class="mord mathnormal" style="margin-right:0.04398em;">z</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:0.43056em;vertical-align:0em;"></span><span class="mord mathnormal">m</span></span></span></span>.</p>
<p>For example, consider a <a href="#BM_(mass%20spectrometer%7Cmass%20spectrometry)%2Fi">mass spectrometer</a> that has a tendency to produce +1 and +2 ions. This <a href="#BM_(mass%20spectrometer%7Cmass%20spectrometry)%2Fi">mass spectrometer</a> produces the following mass-to-charge ratios: [100, 150, 250]. Each mass-to-charge ratio from this <a href="#BM_(mass%20spectrometer%7Cmass%20spectrometry)%2Fi">mass spectrometer</a> will be converted to two possible masses:</p>
<ul>
<li>100 ⟶ [100<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 200<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>]</li>
<li>150 ⟶ [150<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 300<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>]</li>
<li>250 ⟶ [250<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 500<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>]</li>
</ul>
<p>It's impossible to know which mass is correct, so all masses are included in the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>:</p>
<p>[100<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 150<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 200<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 250<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 300<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 500<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>].</p>
<p><a href="ch4_code/src/ExperimentalSpectrum.py">ch4_code/src/ExperimentalSpectrum.py</a> (lines 6 to 14):</p>
<pre class="hljs"><code><span class="hljs-comment"># Its expected that low intensity mass_charge_ratios have already been filtered out prior to invoking this func.</span>
<span class="hljs-keyword">def</span> <span class="hljs-title function_">experimental_spectrum</span>(<span class="hljs-params">mass_charge_ratios: <span class="hljs-type">List</span>[<span class="hljs-built_in">float</span>], charge_tendencies: <span class="hljs-type">Set</span>[<span class="hljs-built_in">float</span>]</span>) -&gt; <span class="hljs-type">List</span>[<span class="hljs-built_in">float</span>]:
    ret = [<span class="hljs-number">0.0</span>]  <span class="hljs-comment"># implied -- subpeptide of length 0</span>
    <span class="hljs-keyword">for</span> mcr <span class="hljs-keyword">in</span> mass_charge_ratios:
        <span class="hljs-keyword">for</span> charge <span class="hljs-keyword">in</span> charge_tendencies:
            ret.append(mcr * charge)
    ret.sort()
    <span class="hljs-keyword">return</span> ret
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>The experimental spectrum for the mass-to-charge ratios...</p>
<p>[100.0, 150.0, 250.0]</p>
<p>... and charge tendencies...</p>
<p>{1.0, 2.0}</p>
<p>... is...</p>
<p>[0.0, 100.0, 150.0, 200.0, 250.0, 300.0, 500.0]</p>
</div>
<p></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The following section isn't from the Pevzner book or any online resources. I came up with it in an effort to solve the final assignment for Chapter 4 (the chapter on <a href="#BM_(non-ribosomal%20peptide)%2Fi">non-ribosomal peptide</a> <a href="#BM_(sequencing%7Csequenced)%2Fi">sequencing</a>). As such, it might not be entirely correct / there may be better ways to do this.</p>
</div>
<p>Just as a <a href="#BM_(spectrum)_MS%2Fi">spectrum</a> is noisy, the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> derived from a <a href="#BM_(spectrum)_MS%2Fi">spectrum</a> is also noisy. For example, consider a <a href="#BM_(mass%20spectrometer%7Cmass%20spectrometry)%2Fi">mass spectrometer</a> that produces up to ±0.5 noise per mass-to-charge ratio and has a tendency to produce +1 and +2 charges. A real mass of 100<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> measured by this <a href="#BM_(mass%20spectrometer%7Cmass%20spectrometry)%2Fi">mass spectrometer</a> will end up in the <a href="#BM_(spectrum)_MS%2Fi">spectrum</a> as a mass-to-charge ratio of either...</p>
<ul>
<li>for +1 charge, anywhere between 99.5 to 100.5 (calculated as <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mfrac><mn>100</mn><mn>1</mn></mfrac><mo>−</mo><mn>0.5</mn></mrow><annotation encoding="application/x-tex">\frac{100}{1} - 0.5</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.190108em;vertical-align:-0.345em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.845108em;"><span style="top:-2.6550000000000002em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">1</span></span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.394em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">1</span><span class="mord mtight">0</span><span class="mord mtight">0</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.345em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.64444em;vertical-align:0em;"></span><span class="mord">0</span><span class="mord">.</span><span class="mord">5</span></span></span></span> to <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mfrac><mn>100</mn><mn>1</mn></mfrac><mo>+</mo><mn>0.5</mn></mrow><annotation encoding="application/x-tex">\frac{100}{1} + 0.5</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.190108em;vertical-align:-0.345em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.845108em;"><span style="top:-2.6550000000000002em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">1</span></span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.394em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">1</span><span class="mord mtight">0</span><span class="mord mtight">0</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.345em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.64444em;vertical-align:0em;"></span><span class="mord">0</span><span class="mord">.</span><span class="mord">5</span></span></span></span>).</li>
<li>for +2 charge, anywhere between 49.5 to 50.5 (calculated as <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mfrac><mn>100</mn><mn>2</mn></mfrac><mo>−</mo><mn>0.5</mn></mrow><annotation encoding="application/x-tex">\frac{100}{2} - 0.5</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.190108em;vertical-align:-0.345em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.845108em;"><span style="top:-2.6550000000000002em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">2</span></span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.394em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">1</span><span class="mord mtight">0</span><span class="mord mtight">0</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.345em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.64444em;vertical-align:0em;"></span><span class="mord">0</span><span class="mord">.</span><span class="mord">5</span></span></span></span> to <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mfrac><mn>100</mn><mn>2</mn></mfrac><mo>+</mo><mn>0.5</mn></mrow><annotation encoding="application/x-tex">\frac{100}{2} + 0.5</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.190108em;vertical-align:-0.345em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.845108em;"><span style="top:-2.6550000000000002em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">2</span></span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.394em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">1</span><span class="mord mtight">0</span><span class="mord mtight">0</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.345em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.64444em;vertical-align:0em;"></span><span class="mord">0</span><span class="mord">.</span><span class="mord">5</span></span></span></span>).</li>
</ul>
<p>Converting these mass-to-charge ratio ranges to mass ranges...</p>
<ul>
<li>for +1 charge, anywhere between 99.5<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> to 100.5<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> (calculated as <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>99.5</mn><mo>⋅</mo><mn>1</mn></mrow><annotation encoding="application/x-tex">99.5 \cdot 1</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.64444em;vertical-align:0em;"></span><span class="mord">9</span><span class="mord">9</span><span class="mord">.</span><span class="mord">5</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.64444em;vertical-align:0em;"></span><span class="mord">1</span></span></span></span> to <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>100.5</mn><mo>⋅</mo><mn>1</mn></mrow><annotation encoding="application/x-tex">100.5 \cdot 1</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.64444em;vertical-align:0em;"></span><span class="mord">1</span><span class="mord">0</span><span class="mord">0</span><span class="mord">.</span><span class="mord">5</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.64444em;vertical-align:0em;"></span><span class="mord">1</span></span></span></span>).</li>
<li>for +2 charge, anywhere between 99<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> to 101<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> (calculated as <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>49.5</mn><mo>⋅</mo><mn>2</mn></mrow><annotation encoding="application/x-tex">49.5 \cdot 2</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.64444em;vertical-align:0em;"></span><span class="mord">4</span><span class="mord">9</span><span class="mord">.</span><span class="mord">5</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.64444em;vertical-align:0em;"></span><span class="mord">2</span></span></span></span> to <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>50.5</mn><mo>⋅</mo><mn>2</mn></mrow><annotation encoding="application/x-tex">50.5 \cdot 2</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.64444em;vertical-align:0em;"></span><span class="mord">5</span><span class="mord">0</span><span class="mord">.</span><span class="mord">5</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.64444em;vertical-align:0em;"></span><span class="mord">2</span></span></span></span>).</li>
</ul>
<p>Note how the +2 charge conversion produces the widest range: 100<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> ± 1<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>. Any real mass measured by this <a href="#BM_(mass%20spectrometer%7Cmass%20spectrometry)%2Fi">mass spectrometer</a> will end up in the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> with up to ±1<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> noise. For example, a real mass of ...</p>
<ul>
<li>99<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> will show up in the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> anywhere between 98<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> and 100<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>.</li>
<li>100<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> will show up in the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> anywhere between 99<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> to 101<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>.</li>
<li>101<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> will show up in the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> anywhere between 100<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> to 102<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>.</li>
</ul>
<p><img src="svgbob_d6d8a41fe65a2e957bcaeb54e9dc6794ee0de080.svg" alt="Kroki diagram output"></p>
<p>Similarly, any mass in the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> could have come from a real mass within ±1<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> of it. For example, an <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> mass of 100<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> could have come from a real mass of anywhere between 99<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> to 101<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>: At a real mass of ...</p>
<ul>
<li>99<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, the corresponding <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> mass range's maximum is 100<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> (98<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> to 100<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>).</li>
<li>100<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, the corresponding <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> mass range's middle is 100<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> (99<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> to 101<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>).</li>
<li>101<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, the corresponding <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> mass range's minimum is 100<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>: (100<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> to 102<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>).</li>
</ul>
<p><img src="svgbob_3b5e914f5add7debee6f9ce052bf324c6d6dcf4f.svg" alt="Kroki diagram output"></p>
<p>As such, the maximum amount of noise for a real mass that made its way into the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> is the same as the tolerance required for mapping an <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> mass back to the real mass it came from. This tolerance can also be considered noise: the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> mass is offset from the real mass that it came from.</p>
<p><a href="ch4_code/src/ExperimentalSpectrumNoise.py">ch4_code/src/ExperimentalSpectrumNoise.py</a> (lines 6 to 8):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">experimental_spectrum_noise</span>(<span class="hljs-params">max_mass_charge_ratio_noise: <span class="hljs-built_in">float</span>, charge_tendencies: <span class="hljs-type">Set</span>[<span class="hljs-built_in">float</span>]</span>) -&gt; <span class="hljs-built_in">float</span>:
    <span class="hljs-keyword">return</span> max_mass_charge_ratio_noise * <span class="hljs-built_in">abs</span>(<span class="hljs-built_in">max</span>(charge_tendencies))
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given a max mass-to-charge ratio noise of ±0.5 and charge tendencies {1.0, 2.0}, the maximum noise per experimental spectrum mass is ±1.0</p>
</div>
<p></p>
<a name="H_Theoretical%20Spectrum"></a><h3>Theoretical Spectrum</h3>
<p><a name="BM_(Algorithms%5C%2FPeptide%20Sequence%2FTheoretical%20Spectrum)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FPeptide%20Sequence%2FExperimental%20Spectrum)_TOPIC%2F">Algorithms/Peptide Sequence/Experimental Spectrum</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: A <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a> is an algorithmically generated list of all <a href="#BM_(subpeptide)%2Fi">subpeptide</a> masses for a known <a href="#BM_(peptide)%2Fi">peptide</a> <a href="#BM_(sequence)%2Fi">sequence</a> (including 0 and the full <a href="#BM_(peptide)%2Fi">peptide</a>'s mass).</p>
<p>For example, <a href="#BM_(linear%20peptide)%2Fi">linear peptide</a> NQY has the <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a>...</p>
<pre class="hljs"><code>theo_spec = [
  <span class="hljs-number">0</span>,    <span class="hljs-comment"># &lt;empty&gt;</span>
  <span class="hljs-number">114</span>,  <span class="hljs-comment"># N</span>
  <span class="hljs-number">128</span>,  <span class="hljs-comment"># Q</span>
  <span class="hljs-number">163</span>,  <span class="hljs-comment"># Y</span>
  <span class="hljs-number">242</span>,  <span class="hljs-comment"># NQ</span>
  <span class="hljs-number">291</span>,  <span class="hljs-comment"># QY</span>
  <span class="hljs-number">405</span>   <span class="hljs-comment"># NQY</span>
]
</code></pre>
<p>... while <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> produced by feeding NQY to a <a href="#BM_(mass%20spectrometer%7Cmass%20spectrometry)%2Fi">mass spectrometer</a> may look something like...</p>
<pre class="hljs"><code>exp_spec = [
  <span class="hljs-number">0.0</span>,    <span class="hljs-comment"># &lt;empty&gt; (implied)</span>
  <span class="hljs-number">113.9</span>,  <span class="hljs-comment"># N</span>
  <span class="hljs-number">115.1</span>,  <span class="hljs-comment"># N</span>
          <span class="hljs-comment"># Q missing</span>
  <span class="hljs-number">136.2</span>,  <span class="hljs-comment"># faulty</span>
  <span class="hljs-number">162.9</span>,  <span class="hljs-comment"># Y</span>
  <span class="hljs-number">242.0</span>,  <span class="hljs-comment"># NQ</span>
          <span class="hljs-comment"># QY missing</span>
  <span class="hljs-number">311.1</span>,  <span class="hljs-comment"># faulty</span>
  <span class="hljs-number">346.0</span>,  <span class="hljs-comment"># faulty</span>
  <span class="hljs-number">405.2</span>   <span class="hljs-comment"># NQY</span>
]
</code></pre>
<p>The <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a> is what the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> would be in a perfect world...</p>
<ul>
<li>only a single possible mass for each mass-to-charge ratio.</li>
<li>no missing masses.</li>
<li>no faulty masses.</li>
<li>no noise.</li>
</ul>
<p><strong>WHY</strong>: The closer a <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a> is to an <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>, the more likely it is that the <a href="#BM_(peptide)%2Fi">peptide</a> <a href="#BM_(sequence)%2Fi">sequence</a> used to generate that <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a> is related to the <a href="#BM_(peptide)%2Fi">peptide</a> <a href="#BM_(sequence)%2Fi">sequence</a> that produced that <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>. This is the basis for how <a href="#BM_(non-ribosomal%20peptide)%2Fi">non-ribosomal peptide</a>s are <a href="#BM_(sequencing%7Csequenced)%2Fi">sequenced</a>: an <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> is produced by a <a href="#BM_(mass%20spectrometer%7Cmass%20spectrometry)%2Fi">mass spectrometer</a>, then that <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> is compared against a set of <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a>s.</p>
<a name="H2_Bruteforce%20Algorithm"></a><h4>Bruteforce Algorithm</h4>
<p><a name="BM_(Algorithms%5C%2FPeptide%20Sequence%2FTheoretical%20Spectrum%2FBruteforce%20Algorithm)_TOPIC%2F"></a><strong></strong></p>
<p><strong>ALGORITHM</strong>:</p>
<p>The following algorithm generates a <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a> in the most obvious way: iterate over each <a href="#BM_(subpeptide)%2Fi">subpeptide</a> and calculate its mass.</p>
<p><a href="ch4_code/src/TheoreticalSpectrum_Bruteforce.py">ch4_code/src/TheoreticalSpectrum_Bruteforce.py</a> (lines 10 to 26):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">theoretical_spectrum</span>(<span class="hljs-params">
        peptide: <span class="hljs-type">List</span>[AA],
        peptide_type: PeptideType,
        mass_table: <span class="hljs-type">Dict</span>[AA, <span class="hljs-built_in">float</span>]
</span>) -&gt; <span class="hljs-type">List</span>[<span class="hljs-built_in">int</span>]:
    <span class="hljs-comment"># add subpeptide of length 0's mass</span>
    ret = [<span class="hljs-number">0.0</span>]
    <span class="hljs-comment"># add subpeptide of length 1 to k-1's mass</span>
    <span class="hljs-keyword">for</span> k <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(<span class="hljs-number">1</span>, <span class="hljs-built_in">len</span>(peptide)):
        <span class="hljs-keyword">for</span> subpeptide, _ <span class="hljs-keyword">in</span> slide_window(peptide, k, cyclic=peptide_type == PeptideType.CYCLIC):
            ret.append(<span class="hljs-built_in">sum</span>([mass_table[ch] <span class="hljs-keyword">for</span> ch <span class="hljs-keyword">in</span> subpeptide]))
    <span class="hljs-comment"># add subpeptide of length k's mass</span>
    ret.append(<span class="hljs-built_in">sum</span>([mass_table[aa] <span class="hljs-keyword">for</span> aa <span class="hljs-keyword">in</span> peptide]))
    <span class="hljs-comment"># sort and return</span>
    ret.sort()
    <span class="hljs-keyword">return</span> ret
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>The theoretical spectrum for the linear peptide NQY is [0.0, 114.0, 128.0, 163.0, 242.0, 291.0, 405.0]</p>
</div>
<p></p>
<a name="H_Prefix%20Sum%20Algorithm"></a><h4>Prefix Sum Algorithm</h4>
<p><a name="BM_(Algorithms%5C%2FPeptide%20Sequence%2FTheoretical%20Spectrum%2FPrefix%20Sum%20Algorithm)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FPeptide%20Sequence%2FTheoretical%20Spectrum%2FBruteforce%20Algorithm)_TOPIC%2F">Algorithms/Peptide Sequence/Theoretical Spectrum/Bruteforce Algorithm</a></li>
</ul>
</div>
<p><strong>ALGORITHM</strong>:</p>
<p>The algorithm starts by calculating the prefix sum of the mass at each position of the <a href="#BM_(peptide)%2Fi">peptide</a>. The prefix sum is calculated by summing all <a href="#BM_(amino%20acid)%2Fi">amino acid</a> masses up until that position. For example, the <a href="#BM_(peptide)%2Fi">peptide</a> GASP has the following masses at the following positions...</p>
<table>
<thead>
<tr>
<th>G</th>
<th>A</th>
<th>S</th>
<th>P</th>
</tr>
</thead>
<tbody>
<tr>
<td>57</td>
<td>71</td>
<td>87</td>
<td>97</td>
</tr>
</tbody>
</table>
<p>As such, the prefix sum at each position is...</p>
<table>
<thead>
<tr>
<th></th>
<th>G</th>
<th>A</th>
<th>S</th>
<th>P</th>
</tr>
</thead>
<tbody>
<tr>
<td>Mass</td>
<td>57</td>
<td>71</td>
<td>87</td>
<td>97</td>
</tr>
<tr>
<td>Prefix sum of mass</td>
<td>57=57</td>
<td>57+71=128</td>
<td>57+71+87=215</td>
<td>57+71+87+97=312</td>
</tr>
</tbody>
</table>
<pre class="hljs"><code>prefixsum_masses[<span class="hljs-number">0</span>] = mass[<span class="hljs-string">''</span>]     = <span class="hljs-number">0</span>             = <span class="hljs-number">0</span>   <span class="hljs-comment"># Artificially added</span>
prefixsum_masses[<span class="hljs-number">1</span>] = mass[<span class="hljs-string">'G'</span>]    = <span class="hljs-number">0</span>+<span class="hljs-number">57</span>          = <span class="hljs-number">57</span>
prefixsum_masses[<span class="hljs-number">2</span>] = mass[<span class="hljs-string">'GA'</span>]   = <span class="hljs-number">0</span>+<span class="hljs-number">57</span>+<span class="hljs-number">71</span>       = <span class="hljs-number">128</span>
prefixsum_masses[<span class="hljs-number">3</span>] = mass[<span class="hljs-string">'GAS'</span>]  = <span class="hljs-number">0</span>+<span class="hljs-number">57</span>+<span class="hljs-number">71</span>+<span class="hljs-number">87</span>    = <span class="hljs-number">215</span>
prefixsum_masses[<span class="hljs-number">4</span>] = mass[<span class="hljs-string">'GASP'</span>] = <span class="hljs-number">0</span>+<span class="hljs-number">57</span>+<span class="hljs-number">71</span>+<span class="hljs-number">87</span>+<span class="hljs-number">97</span> = <span class="hljs-number">312</span>
</code></pre>
<p>The mass for each <a href="#BM_(subpeptide)%2Fi">subpeptide</a> can be derived from just these prefix sums. For example, ...</p>
<pre class="hljs"><code>mass[<span class="hljs-string">'GASP'</span>] = mass[<span class="hljs-string">'GASP'</span>] - mass[<span class="hljs-string">''</span>]    = prefixsum_masses[<span class="hljs-number">4</span>] - prefixsum_masses[<span class="hljs-number">0</span>]
mass[<span class="hljs-string">'ASP'</span>]  = mass[<span class="hljs-string">'GASP'</span>] - mass[<span class="hljs-string">'G'</span>]   = prefixsum_masses[<span class="hljs-number">4</span>] - prefixsum_masses[<span class="hljs-number">1</span>]
mass[<span class="hljs-string">'AS'</span>]   = mass[<span class="hljs-string">'GAS'</span>]  - mass[<span class="hljs-string">'G'</span>]   = prefixsum_masses[<span class="hljs-number">3</span>] - prefixsum_masses[<span class="hljs-number">1</span>]
mass[<span class="hljs-string">'A'</span>]    = mass[<span class="hljs-string">'GA'</span>]   - mass[<span class="hljs-string">'G'</span>]   = prefixsum_masses[<span class="hljs-number">2</span>] - prefixsum_masses[<span class="hljs-number">1</span>]
mass[<span class="hljs-string">'S'</span>]    = mass[<span class="hljs-string">'GAS'</span>]  - mass[<span class="hljs-string">'GA'</span>]  = prefixsum_masses[<span class="hljs-number">3</span>] - prefixsum_masses[<span class="hljs-number">2</span>]
mass[<span class="hljs-string">'P'</span>]    = mass[<span class="hljs-string">'GASP'</span>] - mass[<span class="hljs-string">'GAS'</span>] = prefixsum_masses[<span class="hljs-number">4</span>] - prefixsum_masses[<span class="hljs-number">3</span>]
<span class="hljs-comment"># etc...</span>
</code></pre>
<p>If the <a href="#BM_(peptide)%2Fi">peptide</a> is a <a href="#BM_(cyclopeptide%7Ccyclic%20peptide)%2Fi">cyclic peptide</a>, some <a href="#BM_(subpeptide)%2Fi">subpeptide</a>s will wrap around. For example, PG is a valid <a href="#BM_(subpeptide)%2Fi">subpeptide</a> if GASP is a <a href="#BM_(cyclopeptide%7Ccyclic%20peptide)%2Fi">cyclic peptide</a>:</p>
<p><img src="svgbob_d514f964b7cded0cda5a87cba46c1b551682073d.svg" alt="Kroki diagram output"></p>
<p>The prefix sum can be used to calculate these wrapping <a href="#BM_(subpeptide)%2Fi">subpeptide</a>s as well. For example...</p>
<pre class="hljs"><code>mass[<span class="hljs-string">'PG'</span>] = mass[<span class="hljs-string">'GASP'</span>] - mass[<span class="hljs-string">'AS'</span>]
           = mass[<span class="hljs-string">'GASP'</span>] - (mass[<span class="hljs-string">'GAS'</span>] - mass[<span class="hljs-string">'G'</span>])    <span class="hljs-comment"># SUBSTITUTE IN mass['AS'] CALC FROM ABOVE</span>
           = prefixsum_masses[<span class="hljs-number">4</span>] - (prefixsum_masses[<span class="hljs-number">3</span>] - prefixsum_masses[<span class="hljs-number">1</span>])
</code></pre>
<p>This algorithm is faster than the bruteforce algorithm, but most use-cases won't notice a performance improvement unless either the...</p>
<ul>
<li><a href="#BM_(peptide)%2Fi">peptide</a> is very long (likely won't happen since <a href="#BM_(peptide)%2Fi">peptide</a>s by definition aren't larger than 50 to 100 <a href="#BM_(amino%20acid)%2Fi">amino acid</a>s)</li>
<li>algorithm runs often.</li>
</ul>
<p><a href="ch4_code/src/TheoreticalSpectrum_PrefixSum.py">ch4_code/src/TheoreticalSpectrum_PrefixSum.py</a> (lines 37 to 53):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">theoretical_spectrum</span>(<span class="hljs-params">
        peptide: <span class="hljs-type">List</span>[AA],
        peptide_type: PeptideType,
        mass_table: <span class="hljs-type">Dict</span>[AA, <span class="hljs-built_in">float</span>]
</span>) -&gt; <span class="hljs-type">List</span>[<span class="hljs-built_in">float</span>]:
    prefixsum_masses = <span class="hljs-built_in">list</span>(accumulate([mass_table[aa] <span class="hljs-keyword">for</span> aa <span class="hljs-keyword">in</span> peptide], initial=<span class="hljs-number">0.0</span>))
    ret = [<span class="hljs-number">0.0</span>]
    <span class="hljs-keyword">for</span> end_idx <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(<span class="hljs-number">0</span>, <span class="hljs-built_in">len</span>(prefixsum_masses)):
        <span class="hljs-keyword">for</span> start_idx <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(<span class="hljs-number">0</span>, end_idx):
            min_mass = prefixsum_masses[start_idx]
            max_mass = prefixsum_masses[end_idx]
            ret.append(max_mass - min_mass)
            <span class="hljs-keyword">if</span> peptide_type == PeptideType.CYCLIC <span class="hljs-keyword">and</span> start_idx &gt; <span class="hljs-number">0</span> <span class="hljs-keyword">and</span> end_idx &lt; <span class="hljs-built_in">len</span>(peptide):
                ret.append(prefixsum_masses[-<span class="hljs-number">1</span>] - (prefixsum_masses[end_idx] - prefixsum_masses[start_idx]))
    ret.sort()
    <span class="hljs-keyword">return</span> ret
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>The theoretical spectrum for the linear peptide NQY is [0.0, 114.0, 128.0, 163.0, 242.0, 291.0, 405.0]</p>
</div>
<p></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The algorithm above is serial, but it can be made parallel to get even more speed:</p>
<ol>
<li>Parallelized prefix sum (e.g. Hillis-Steele / Blelloch).</li>
<li>Parallelized iteration instead of nested for-loops.</li>
<li>Parallelized sorting (e.g. Parallel merge sort / Parallel brick sort / Bitonic sort).</li>
</ol>
</div>
<a name="H_Spectrum%20Convolution"></a><h3>Spectrum Convolution</h3>
<p><a name="BM_(Algorithms%5C%2FPeptide%20Sequence%2FSpectrum%20Convolution)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FPeptide%20Sequence%2FExperimental%20Spectrum)_TOPIC%2F">Algorithms/Peptide Sequence/Experimental Spectrum</a></li>
<li><a href="#BM_(Algorithms%5C%2FPeptide%20Sequence%2FTheoretical%20Spectrum)_TOPIC%2F">Algorithms/Peptide Sequence/Theoretical Spectrum</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: Given an <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>, subtract its masses from each other. The differences are a set of potential <a href="#BM_(amino%20acid)%2Fi">amino acid</a> masses for the <a href="#BM_(peptide)%2Fi">peptide</a> that generated that <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>.</p>
<p>For example, the following <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> is for the <a href="#BM_(linear%20peptide)%2Fi">linear peptide</a> NQY:</p>
<p>[0.0<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 113.9<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 115.1<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 136.2<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 162.9<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 242.0<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 311.1<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 346.0<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 405.2<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>]</p>
<p>Performing 242.0 - 113.9 results in 128.1, which is very close to the mass for <a href="#BM_(amino%20acid)%2Fi">amino acid</a> Q. The mass for Q was derived even though no <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> masses are near Q's mass:</p>
<ul>
<li>Mass of N is 114<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 2 <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> masses are near: [113.9, 115.1]</li>
<li>Mass of Q is 128<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 0 <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> masses are near: []</li>
<li>Mass of Y is 163<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 1 <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> mass is near: [162.9]</li>
</ul>
<p><strong>WHY</strong>: The closer a <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a> is to an <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>, the more likely it is that the <a href="#BM_(peptide)%2Fi">peptide</a> <a href="#BM_(sequence)%2Fi">sequence</a> used to generate that <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a> is related to the <a href="#BM_(peptide)%2Fi">peptide</a> <a href="#BM_(sequence)%2Fi">sequence</a> that produced that <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>. However, before being able to build a <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a>, a list of potential <a href="#BM_(amino%20acid)%2Fi">amino acid</a>s need to be inferred from the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>. In addition to the 20 <a href="#BM_(proteinogenic%20amino%20acid)%2Fi">proteinogenic amino acid</a>s, there are many other non-<a href="#BM_(proteinogenic%20amino%20acid)%2Fi">proteinogenic amino acid</a>s that may be part of the <a href="#BM_(peptide)%2Fi">peptide</a>.</p>
<p>This operation infers a list of potential <a href="#BM_(amino%20acid)%2Fi">amino acid</a> masses, which can be mapped back to <a href="#BM_(amino%20acid)%2Fi">amino acid</a>s themselves.</p>
<p><strong>ALGORITHM</strong>:</p>
<p>Consider an <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> with masses that don't contain any noise. That is, the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> may have faulty masses and may be missing masses, but any correct masses it does have are exact / noise-free. To derive a list of potential <a href="#BM_(amino%20acid)%2Fi">amino acid</a> masses for this <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>:</p>
<ol>
<li>Subtract <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> masses from each other (each mass gets subtracted from every mass).</li>
<li>Filter differences to those between 57<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> and 200<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> (generally accepted range for the mass of an <a href="#BM_(amino%20acid)%2Fi">amino acid</a>).</li>
<li>Filter differences to that don't occur at least n times (n is user-defined).</li>
</ol>
<p>The result is a list of potential <a href="#BM_(amino%20acid)%2Fi">amino acid</a> masses for the <a href="#BM_(peptide)%2Fi">peptide</a> that produced that <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>. For example, consider the following <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> for the <a href="#BM_(linear%20peptide)%2Fi">linear peptide</a> NQY:</p>
<p>[0<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 114<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 136<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 163<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 242<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 311<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 346<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 405<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>]</p>
<p>The <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> masses...</p>
<ul>
<li>[163<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 291<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>] are missing.</li>
<li>[136<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 311<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 346<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>] are faulty.</li>
<li>[114<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 163<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 242<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 405<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>] are correct and free of noise.</li>
</ul>
<p>Subtract the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> masses:</p>
<table>
<thead>
<tr>
<th></th>
<th>0</th>
<th>114</th>
<th>136</th>
<th>163</th>
<th>242</th>
<th>311</th>
<th>346</th>
<th>405</th>
</tr>
</thead>
<tbody>
<tr>
<td>0</td>
<td>0</td>
<td>-114</td>
<td>-136</td>
<td>-163</td>
<td>-242</td>
<td>-311</td>
<td>-346</td>
<td>-405</td>
</tr>
<tr>
<td>114</td>
<td>114</td>
<td>0</td>
<td>-22</td>
<td>-49</td>
<td>-128</td>
<td>-197</td>
<td>-231</td>
<td>-291</td>
</tr>
<tr>
<td>136</td>
<td>136</td>
<td>22</td>
<td>0</td>
<td>-27</td>
<td>-106</td>
<td>-175</td>
<td>-210</td>
<td>-269</td>
</tr>
<tr>
<td>163</td>
<td>163</td>
<td>49</td>
<td>27</td>
<td>0</td>
<td>-79</td>
<td>-148</td>
<td>-183</td>
<td>-242</td>
</tr>
<tr>
<td>242</td>
<td>242</td>
<td>128</td>
<td>106</td>
<td>79</td>
<td>0</td>
<td>-69</td>
<td>-104</td>
<td>-163</td>
</tr>
<tr>
<td>311</td>
<td>311</td>
<td>197</td>
<td>175</td>
<td>148</td>
<td>69</td>
<td>0</td>
<td>-35</td>
<td>-94</td>
</tr>
<tr>
<td>346</td>
<td>346</td>
<td>232</td>
<td>210</td>
<td>183</td>
<td>104</td>
<td>35</td>
<td>0</td>
<td>-59</td>
</tr>
<tr>
<td>405</td>
<td>405</td>
<td>291</td>
<td>269</td>
<td>242</td>
<td>163</td>
<td>94</td>
<td>59</td>
<td>0</td>
</tr>
</tbody>
</table>
<p>Then, remove differences that aren't between 57<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> and 200<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>:</p>
<table>
<thead>
<tr>
<th></th>
<th>0</th>
<th>114</th>
<th>136</th>
<th>163</th>
<th>242</th>
<th>311</th>
<th>346</th>
<th>405</th>
</tr>
</thead>
<tbody>
<tr>
<td>0</td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td>114</td>
<td>114</td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td>136</td>
<td>136</td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td>163</td>
<td>163</td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td>242</td>
<td></td>
<td>128</td>
<td>106</td>
<td>79</td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td>311</td>
<td></td>
<td>197</td>
<td>175</td>
<td>148</td>
<td>69</td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td>346</td>
<td></td>
<td></td>
<td></td>
<td>183</td>
<td>104</td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td>405</td>
<td></td>
<td></td>
<td></td>
<td></td>
<td>163</td>
<td>94</td>
<td>59</td>
<td></td>
</tr>
</tbody>
</table>
<p>Then, filter out any differences occurring less than than n times. In this case, it makes sense to set n to 1 because almost all of the differences occur only once.</p>
<p>The final result is a list of potential <a href="#BM_(amino%20acid)%2Fi">amino acid</a> masses for the <a href="#BM_(peptide)%2Fi">peptide</a> that produced the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>:</p>
<p>[59<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 69<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 79<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 94<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 104<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 106<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 114<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 128<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 136<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 148<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 163<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 175<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 183<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 197<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>]</p>
<p>Note that the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> is for the <a href="#BM_(linear%20peptide)%2Fi">linear peptide</a> NQY. The <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> contained the masses for N (114<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>) and Y (163<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>), but not Q (128<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>). This operation was able to pull out the mass for Q: 128<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> is in the final list of differences.</p>
<p><a href="ch4_code/src/SpectrumConvolution_NoNoise.py">ch4_code/src/SpectrumConvolution_NoNoise.py</a> (lines 6 to 16):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">spectrum_convolution</span>(<span class="hljs-params">experimental_spectrum: <span class="hljs-type">List</span>[<span class="hljs-built_in">float</span>], min_mass=<span class="hljs-number">57.0</span>, max_mass=<span class="hljs-number">200.0</span></span>) -&gt; <span class="hljs-type">List</span>[<span class="hljs-built_in">float</span>]:
    <span class="hljs-comment"># it's expected that experimental_spectrum is sorted smallest to largest</span>
    diffs = []
    <span class="hljs-keyword">for</span> row_idx, row_mass <span class="hljs-keyword">in</span> <span class="hljs-built_in">enumerate</span>(experimental_spectrum):
        <span class="hljs-keyword">for</span> col_idx, col_mass <span class="hljs-keyword">in</span> <span class="hljs-built_in">enumerate</span>(experimental_spectrum):
            mass_diff = row_mass - col_mass
            <span class="hljs-keyword">if</span> min_mass &lt;= mass_diff &lt;= max_mass:
                diffs.append(mass_diff)
    diffs.sort()
    <span class="hljs-keyword">return</span> diffs
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>The spectrum convolution for [0.0, 114.0, 136.0, 163.0, 242.0, 311.0, 346.0, 405.0] is ...</p>
<ul>
<li>2x163.0</li>
<li>1x59.0</li>
<li>1x69.0</li>
<li>1x79.0</li>
<li>1x94.0</li>
<li>1x104.0</li>
<li>1x106.0</li>
<li>1x114.0</li>
<li>1x128.0</li>
<li>1x136.0</li>
<li>1x148.0</li>
<li>1x175.0</li>
<li>1x183.0</li>
<li>1x197.0</li>
</ul>
</div>
<p></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The following section isn't from the Pevzner book or any online resources. I came up with it in an effort to solve the final assignment for Chapter 4 (the chapter on <a href="#BM_(non-ribosomal%20peptide)%2Fi">non-ribosomal peptide</a> <a href="#BM_(sequencing%7Csequenced)%2Fi">sequencing</a>). As such, it might not be entirely correct / there may be better ways to do this.</p>
</div>
<p>The algorithm described above is for <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>s that have exact masses (no noise). However, real <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>s will have noisy masses. Since a real <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> has noisy masses, the <a href="#BM_(amino%20acid)%2Fi">amino acid</a> masses derived from it will also be noisy. For example, consider an <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> that has ±1<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> noise per mass. A real mass of...</p>
<ul>
<li>242<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> will show up in the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> anywhere between 241<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> to 243<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>.</li>
<li>114<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> will show up in the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> anywhere between 113<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> to 115<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>.</li>
</ul>
<p>Subtract the opposite extremes from these two ranges: 243<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> - 113<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> = 130<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>. That's 2<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> away from the real mass difference: 128<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>. As such, the maximum noise per <a href="#BM_(amino%20acid)%2Fi">amino acid</a> mass is 2 times the maximum noise for the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> that it was derived from: ±2<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> for this example.</p>
<p><a href="ch4_code/src/SpectrumConvolutionNoise.py">ch4_code/src/SpectrumConvolutionNoise.py</a> (lines 7 to 9):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">spectrum_convolution_noise</span>(<span class="hljs-params">exp_spec_mass_noise: <span class="hljs-built_in">float</span></span>) -&gt; <span class="hljs-built_in">float</span>:
    <span class="hljs-keyword">return</span> <span class="hljs-number">2.0</span> * exp_spec_mass_noise
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given a max experimental spectrum mass noise of ±1.0, the maximum noise per amino acid derived from an experimental spectrum is ±2.0</p>
</div>
<p></p>
<p>Extending the algorithm to handle noisy <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> masses requires one extra step: group together differences that are within some tolerance of each other, where this tolerance is the maximum <a href="#BM_(amino%20acid)%2Fi">amino acid</a> mass noise calculation described above. For example, consider the following <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> for <a href="#BM_(linear%20peptide)%2Fi">linear peptide</a> NQY that has up to ±1<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> noise per mass:</p>
<p>[0.0<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 113.9<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 115.1<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 136.2<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 162.9<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 242.0<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 311.1<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 346.0<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 405.2<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>]</p>
<p>Just as before, subtract the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> masses and differences that aren't between 57<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> and 200<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>:</p>
<table>
<thead>
<tr>
<th></th>
<th>0.0</th>
<th>113.9</th>
<th>115.1</th>
<th>136.2</th>
<th>162.9</th>
<th>242.0</th>
<th>311.1</th>
<th>346.0</th>
<th>405.2</th>
</tr>
</thead>
<tbody>
<tr>
<td>0.0</td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td>113.9</td>
<td>113.9</td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td>115.1</td>
<td>115.1</td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td>136.2</td>
<td>136.2</td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td>162.9</td>
<td>162.9</td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td>242.0</td>
<td></td>
<td>128.1</td>
<td>126.9</td>
<td>105.8</td>
<td>79.1</td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td>311.1</td>
<td></td>
<td>197.2</td>
<td>196.0</td>
<td>174.9</td>
<td>142.9</td>
<td>69.1</td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td>346.0</td>
<td></td>
<td></td>
<td></td>
<td></td>
<td>183.1</td>
<td>104.0</td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td>405.2</td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td>163.0</td>
<td>94.1</td>
<td>59.2</td>
<td></td>
</tr>
</tbody>
</table>
<p>Then, group differences that are within ±2<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> of each other (2 times the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>'s maximum mass noise):</p>
<ul>
<li>[104.0, 105.8]</li>
<li>[113.9, 115.1]</li>
<li>[128.1, 126.9]</li>
<li>[162.9, 163.0]</li>
<li>[196.0, 197.2]</li>
<li>[59.2]</li>
<li>[69.1]</li>
<li>[79.1]</li>
<li>[94.1]</li>
<li>[136.2]</li>
<li>[142.9]</li>
<li>[174.9]</li>
</ul>
<p>Then, filter out any groups that have less than n occurrences. In this case, filtering to n=2 occurrences reveals that all <a href="#BM_(amino%20acid)%2Fi">amino acid</a> masses are captured for NQY:</p>
<ul>
<li>[104.0, 105.8] (junk)</li>
<li>[113.9, 115.1] (mass of N is 114)</li>
<li>[128.1, 126.9] (mass of Q is 128)</li>
<li>[162.9, 163.0] (mass of Y is 163)</li>
<li>[196.0, 197.2] (junk)</li>
</ul>
<p>Note that the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> is for the <a href="#BM_(linear%20peptide)%2Fi">linear peptide</a> NQY. The <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> contained the masses near N (113.Da and 115.1<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>) and Y (162.9<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>), but not Q. This operation was able to pull out masses near Q: [128.1, 126.9] is in the final list of differences.</p>
<p><a href="ch4_code/src/SpectrumConvolution.py">ch4_code/src/SpectrumConvolution.py</a> (lines 7 to 58):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">group_masses_by_tolerance</span>(<span class="hljs-params">masses: <span class="hljs-type">List</span>[<span class="hljs-built_in">float</span>], tolerance: <span class="hljs-built_in">float</span></span>) -&gt; typing.Counter[<span class="hljs-built_in">float</span>]:
    masses = <span class="hljs-built_in">sorted</span>(masses)
    length = <span class="hljs-built_in">len</span>(masses)
    ret = Counter()
    <span class="hljs-keyword">for</span> i, m1 <span class="hljs-keyword">in</span> <span class="hljs-built_in">enumerate</span>(masses):
        <span class="hljs-keyword">if</span> m1 <span class="hljs-keyword">in</span> ret:
            <span class="hljs-keyword">continue</span>
        <span class="hljs-comment"># search backwards</span>
        left_limit = <span class="hljs-number">0</span>
        <span class="hljs-keyword">for</span> j <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(i, -<span class="hljs-number">1</span>, -<span class="hljs-number">1</span>):
            m2 = masses[j]
            <span class="hljs-keyword">if</span> <span class="hljs-built_in">abs</span>(m2 - m1) &gt; tolerance:
                <span class="hljs-keyword">break</span>
            left_limit = j
        <span class="hljs-comment"># search forwards</span>
        right_limit = length - <span class="hljs-number">1</span>
        <span class="hljs-keyword">for</span> j <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(i, length):
            m2 = masses[j]
            <span class="hljs-keyword">if</span> <span class="hljs-built_in">abs</span>(m2 - m1) &gt; tolerance:
                <span class="hljs-keyword">break</span>
            right_limit = j
        count = right_limit - left_limit + <span class="hljs-number">1</span>
        ret[m1] = count
    <span class="hljs-keyword">return</span> ret


<span class="hljs-keyword">def</span> <span class="hljs-title function_">spectrum_convolution</span>(<span class="hljs-params">
        exp_spec: <span class="hljs-type">List</span>[<span class="hljs-built_in">float</span>],  <span class="hljs-comment"># must be sorted smallest to largest</span>
        tolerance: <span class="hljs-built_in">float</span>,
        min_mass: <span class="hljs-built_in">float</span> = <span class="hljs-number">57.0</span>,
        max_mass: <span class="hljs-built_in">float</span> = <span class="hljs-number">200.0</span>,
        round_digits: <span class="hljs-built_in">int</span> = -<span class="hljs-number">1</span>,  <span class="hljs-comment"># if set, rounds to this many digits past decimal point</span>
        implied_zero: <span class="hljs-built_in">bool</span> = <span class="hljs-literal">False</span>  <span class="hljs-comment"># if set, run as if 0.0 were added to exp_spec</span>
</span>) -&gt; typing.Counter[<span class="hljs-built_in">float</span>]:
    min_mass -= tolerance
    max_mass += tolerance
    diffs = []
    <span class="hljs-keyword">for</span> row_idx, row_mass <span class="hljs-keyword">in</span> <span class="hljs-built_in">enumerate</span>(exp_spec):
        <span class="hljs-keyword">for</span> col_idx, col_mass <span class="hljs-keyword">in</span> <span class="hljs-built_in">enumerate</span>(exp_spec):
            mass_diff = row_mass - col_mass
            <span class="hljs-keyword">if</span> round_digits != -<span class="hljs-number">1</span>:
                mass_diff = <span class="hljs-built_in">round</span>(mass_diff, round_digits)
            <span class="hljs-keyword">if</span> min_mass &lt;= mass_diff &lt;= max_mass:
                diffs.append(mass_diff)
    <span class="hljs-keyword">if</span> implied_zero:
        <span class="hljs-keyword">for</span> mass <span class="hljs-keyword">in</span> exp_spec:
            <span class="hljs-keyword">if</span> min_mass &lt;= mass &lt;= max_mass:
                diffs.append(mass)
            <span class="hljs-keyword">if</span> mass &gt; max_mass:
                <span class="hljs-keyword">break</span>
    <span class="hljs-keyword">return</span> group_masses_by_tolerance(diffs, tolerance)
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>The spectrum convolution for [113.9, 115.1, 136.2, 162.9, 242.0, 311.1, 346.0, 405.2] is ...</p>
<ul>
<li>2x104.0</li>
<li>2x105.8</li>
<li>2x113.9</li>
<li>2x115.1</li>
<li>2x126.9</li>
<li>2x128.1</li>
<li>2x162.9</li>
<li>2x163.2</li>
<li>2x196.0</li>
<li>2x197.2</li>
<li>1x59.2</li>
<li>1x69.1</li>
<li>1x79.1</li>
<li>1x94.1</li>
<li>1x136.2</li>
<li>1x148.2</li>
<li>1x174.9</li>
<li>1x183.1</li>
</ul>
</div>
<p></p>
<a name="H_Spectrum%20Score"></a><h3>Spectrum Score</h3>
<p><a name="BM_(Algorithms%5C%2FPeptide%20Sequence%2FSpectrum%20Score)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FPeptide%20Sequence%2FExperimental%20Spectrum)_TOPIC%2F">Algorithms/Peptide Sequence/Experimental Spectrum</a></li>
<li><a href="#BM_(Algorithms%5C%2FPeptide%20Sequence%2FTheoretical%20Spectrum)_TOPIC%2F">Algorithms/Peptide Sequence/Theoretical Spectrum</a></li>
<li><a href="#BM_(Algorithms%5C%2FPeptide%20Sequence%2FSpectrum%20Convolution)_TOPIC%2F">Algorithms/Peptide Sequence/Spectrum Convolution</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: Given an <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> and a <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a>, score them against each other by counting how many masses match between them.</p>
<p><strong>WHY</strong>: The more matching masses between a <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a> and an <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>, the more likely it is that the <a href="#BM_(peptide)%2Fi">peptide</a> <a href="#BM_(sequence)%2Fi">sequence</a> used to generate that <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a> is related to the <a href="#BM_(peptide)%2Fi">peptide</a> <a href="#BM_(sequence)%2Fi">sequence</a> that produced that <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>. This is the basis for how <a href="#BM_(non-ribosomal%20peptide)%2Fi">non-ribosomal peptide</a>s are <a href="#BM_(sequencing%7Csequenced)%2Fi">sequenced</a>: an <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> is produced by a <a href="#BM_(mass%20spectrometer%7Cmass%20spectrometry)%2Fi">mass spectrometer</a>, then that <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> is compared against a set of <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a>s.</p>
<p><strong>ALGORITHM</strong>:</p>
<p>Consider an <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> with masses that don't contain any noise. That is, the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> may have faulty masses and may be missing masses, but any correct masses it does have are exact / noise-free. Scoring this <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> against a <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a> is simple: count the number of matching masses.</p>
<p><a href="ch4_code/src/SpectrumScore_NoNoise.py">ch4_code/src/SpectrumScore_NoNoise.py</a> (lines 9 to 28):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">score_spectrums</span>(<span class="hljs-params">
        s1: <span class="hljs-type">List</span>[<span class="hljs-built_in">float</span>],  <span class="hljs-comment"># must be sorted ascending</span>
        s2: <span class="hljs-type">List</span>[<span class="hljs-built_in">float</span>]   <span class="hljs-comment"># must be sorted ascending</span>
</span>) -&gt; <span class="hljs-built_in">int</span>:
    idx_s1 = <span class="hljs-number">0</span>
    idx_s2 = <span class="hljs-number">0</span>
    score = <span class="hljs-number">0</span>
    <span class="hljs-keyword">while</span> idx_s1 &lt; <span class="hljs-built_in">len</span>(s1) <span class="hljs-keyword">and</span> idx_s2 &lt; <span class="hljs-built_in">len</span>(s2):
        s1_mass = s1[idx_s1]
        s2_mass = s2[idx_s2]
        <span class="hljs-keyword">if</span> s1_mass &lt; s2_mass:
            idx_s1 += <span class="hljs-number">1</span>
        <span class="hljs-keyword">elif</span> s1_mass &gt; s2_mass:
            idx_s2 += <span class="hljs-number">1</span>
        <span class="hljs-keyword">else</span>:
            idx_s1 += <span class="hljs-number">1</span>
            idx_s2 += <span class="hljs-number">1</span>
            score += <span class="hljs-number">1</span>
    <span class="hljs-keyword">return</span> score
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>The spectrum score for...</p>
<p>[0.0, 57.0, 71.0, 128.0, 199.0, 256.0]</p>
<p>... vs ...</p>
<p>[0.0, 57.0, 71.0, 128.0, 128.0, 199.0, 256.0]</p>
<p>... is 6</p>
</div>
<p></p>
<p>Note that a <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a> may have multiple masses with the same value but an <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> won't. For example, the <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a> for GAK is ...</p>
<table>
<thead>
<tr>
<th></th>
<th></th>
<th>G</th>
<th>A</th>
<th>K</th>
<th>GA</th>
<th>AK</th>
<th>GAK</th>
</tr>
</thead>
<tbody>
<tr>
<td>Mass</td>
<td>0<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a></td>
<td>57D a</td>
<td>71<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a></td>
<td>128<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a></td>
<td>128<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a></td>
<td>199<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a></td>
<td>256<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a></td>
</tr>
</tbody>
</table>
<p>K and GA both have a mass of 128<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>. Since <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>s don't distinguish between where masses come from, an <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> for this <a href="#BM_(linear%20peptide)%2Fi">linear peptide</a> will only have 1 entry for 128<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The following section isn't from the Pevzner book or any online resources. I came up with it in an effort to solve the final assignment for Chapter 4 (the chapter on <a href="#BM_(non-ribosomal%20peptide)%2Fi">non-ribosomal peptide</a> <a href="#BM_(sequencing%7Csequenced)%2Fi">sequencing</a>). As such, it might not be entirely correct / there may be better ways to do this.</p>
</div>
<p>The algorithm described above is for <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>s that have exact masses (no noise). However, real <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>s have noisy masses. That noise needs to be accounted for when identifying matches.</p>
<p>Recall that each <a href="#BM_(amino%20acid)%2Fi">amino acid</a> mass captured by a <a href="#BM_(spectrum%20convolution)%2Fi">spectrum convolution</a> has up to some amount of noise. This is what defines the tolerance for a matching mass between the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> and the <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a>. Specifically, the maximum amount of noise for a captured <a href="#BM_(amino%20acid)%2Fi">amino acid</a> mass is multiplied by the <a href="#BM_(amino%20acid)%2Fi">amino acid</a> count of the <a href="#BM_(subpeptide)%2Fi">subpeptide</a> to determine the tolerance.</p>
<p>For example, imagine a case where it's determined that the noise tolerance for each captured <a href="#BM_(amino%20acid)%2Fi">amino acid</a> mass is ±2<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>. Given the <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a> for <a href="#BM_(linear%20peptide)%2Fi">linear peptide</a> NQY, the tolerances would be as follows:</p>
<table>
<thead>
<tr>
<th></th>
<th></th>
<th>N</th>
<th>Q</th>
<th>Y</th>
<th>NQ</th>
<th>QY</th>
<th>NQY</th>
</tr>
</thead>
<tbody>
<tr>
<td>Mass</td>
<td>0<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a></td>
<td>114<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a></td>
<td>128<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a></td>
<td>163<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a></td>
<td>242<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a></td>
<td>291<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a></td>
<td>405<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a></td>
</tr>
<tr>
<td>Tolerance</td>
<td>0<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a></td>
<td>±2<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a></td>
<td>±2<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a></td>
<td>±2<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a></td>
<td>±4<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a></td>
<td>±4<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a></td>
<td>±6<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a></td>
</tr>
</tbody>
</table>
<p><a href="ch4_code/src/TheoreticalSpectrumTolerances.py">ch4_code/src/TheoreticalSpectrumTolerances.py</a> (lines 7 to 26):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">theoretical_spectrum_tolerances</span>(<span class="hljs-params">
        peptide_len: <span class="hljs-built_in">int</span>,
        peptide_type: PeptideType,
        amino_acid_mass_tolerance: <span class="hljs-built_in">float</span>
</span>) -&gt; <span class="hljs-type">List</span>[<span class="hljs-built_in">float</span>]:
    ret = [<span class="hljs-number">0.0</span>]
    <span class="hljs-keyword">if</span> peptide_type == PeptideType.LINEAR:
        <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(peptide_len):
            tolerance = (i + <span class="hljs-number">1</span>) * amino_acid_mass_tolerance
            ret += [tolerance] * (peptide_len - i)
    <span class="hljs-keyword">elif</span> peptide_type == PeptideType.CYCLIC:
        <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(peptide_len - <span class="hljs-number">1</span>):
            tolerance = (i + <span class="hljs-number">1</span>) * amino_acid_mass_tolerance
            ret += [tolerance] * peptide_len
        <span class="hljs-keyword">if</span> peptide_len != <span class="hljs-number">0</span>:
            ret.append(peptide_len * amino_acid_mass_tolerance)
    <span class="hljs-keyword">else</span>:
        <span class="hljs-keyword">raise</span> ValueError()
    <span class="hljs-keyword">return</span> ret
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>The theoretical spectrum for linear peptide NQY with amino acid mass tolerance of 2.0...</p>
<p>[0.0, 2.0, 2.0, 2.0, 4.0, 4.0, 6.0]</p>
</div>
<p></p>
<p>Given a <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a> with tolerances, each <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> mass is checked to see if it fits within a <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a> mass tolerance. If it fits, it's considered a match. The score includes both the number of matches and how closely each match was to the ideal <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a> mass.</p>
<p><a href="ch4_code/src/SpectrumScore.py">ch4_code/src/SpectrumScore.py</a> (lines 10 to 129):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">scan_left</span>(<span class="hljs-params">
        exp_spec: <span class="hljs-type">List</span>[<span class="hljs-built_in">float</span>],
        exp_spec_lo_idx: <span class="hljs-built_in">int</span>,
        exp_spec_start_idx: <span class="hljs-built_in">int</span>,
        theo_mid_mass: <span class="hljs-built_in">float</span>,
        theo_min_mass: <span class="hljs-built_in">float</span>
</span>) -&gt; <span class="hljs-type">Optional</span>[<span class="hljs-built_in">int</span>]:
    found_dist = <span class="hljs-literal">None</span>
    found_idx = <span class="hljs-literal">None</span>
    <span class="hljs-keyword">for</span> idx <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(exp_spec_start_idx, exp_spec_lo_idx - <span class="hljs-number">1</span>, -<span class="hljs-number">1</span>):
        exp_mass = exp_spec[idx]
        <span class="hljs-keyword">if</span> exp_mass &lt; theo_min_mass:
            <span class="hljs-keyword">break</span>
        dist_to_theo_mid_mass = <span class="hljs-built_in">abs</span>(exp_mass - theo_mid_mass)
        <span class="hljs-keyword">if</span> found_dist <span class="hljs-keyword">is</span> <span class="hljs-literal">None</span> <span class="hljs-keyword">or</span> dist_to_theo_mid_mass &lt; found_dist:
            found_idx = idx
            found_dist = dist_to_theo_mid_mass
    <span class="hljs-keyword">return</span> found_idx


<span class="hljs-keyword">def</span> <span class="hljs-title function_">scan_right</span>(<span class="hljs-params">
        exp_spec: <span class="hljs-type">List</span>[<span class="hljs-built_in">float</span>],
        exp_spec_hi_idx: <span class="hljs-built_in">int</span>,
        exp_spec_start_idx: <span class="hljs-built_in">int</span>,
        theo_mid_mass: <span class="hljs-built_in">float</span>,
        theo_max_mass: <span class="hljs-built_in">float</span>
</span>) -&gt; <span class="hljs-type">Optional</span>[<span class="hljs-built_in">int</span>]:
    found_dist = <span class="hljs-literal">None</span>
    found_idx = <span class="hljs-literal">None</span>
    <span class="hljs-keyword">for</span> idx <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(exp_spec_start_idx, exp_spec_hi_idx):
        exp_mass = exp_spec[idx]
        <span class="hljs-keyword">if</span> exp_mass &gt; theo_max_mass:
            <span class="hljs-keyword">break</span>
        dist_to_theo_mid_mass = <span class="hljs-built_in">abs</span>(exp_mass - theo_mid_mass)
        <span class="hljs-keyword">if</span> found_dist <span class="hljs-keyword">is</span> <span class="hljs-literal">None</span> <span class="hljs-keyword">or</span> dist_to_theo_mid_mass &lt; found_dist:
            found_idx = idx
            found_dist = dist_to_theo_mid_mass
    <span class="hljs-keyword">return</span> found_idx


<span class="hljs-keyword">def</span> <span class="hljs-title function_">find_closest_within_tolerance</span>(<span class="hljs-params">
        exp_spec: <span class="hljs-type">List</span>[<span class="hljs-built_in">float</span>],
        exp_spec_lo_idx: <span class="hljs-built_in">int</span>,
        exp_spec_hi_idx: <span class="hljs-built_in">int</span>,
        theo_exact_mass: <span class="hljs-built_in">float</span>,
        theo_min_mass: <span class="hljs-built_in">float</span>,
        theo_max_mass: <span class="hljs-built_in">float</span>
</span>) -&gt; <span class="hljs-type">Optional</span>[<span class="hljs-built_in">int</span>]:
    <span class="hljs-comment"># Binary search exp_spec for the where theo_mid_mass would be inserted (left-most index chosen if already there).</span>
    start_idx = bisect_left(exp_spec, theo_exact_mass, lo=exp_spec_lo_idx, hi=exp_spec_hi_idx)
    <span class="hljs-keyword">if</span> start_idx == exp_spec_hi_idx:
        start_idx -= <span class="hljs-number">1</span>
    <span class="hljs-comment"># From start_idx - 1, walk left to find the closest possible value to theo_mid_mass</span>
    left_idx = scan_left(exp_spec, exp_spec_lo_idx, start_idx - <span class="hljs-number">1</span>, theo_exact_mass, theo_min_mass)
    <span class="hljs-comment"># From start_idx, walk right to find the closest possible value to theo_mid_mass</span>
    right_idx = scan_right(exp_spec, exp_spec_hi_idx, start_idx, theo_exact_mass, theo_max_mass)
    <span class="hljs-keyword">if</span> left_idx <span class="hljs-keyword">is</span> <span class="hljs-literal">None</span> <span class="hljs-keyword">and</span> right_idx <span class="hljs-keyword">is</span> <span class="hljs-literal">None</span>:  <span class="hljs-comment"># If nothing found, return None</span>
        <span class="hljs-keyword">return</span> <span class="hljs-literal">None</span>
    <span class="hljs-keyword">if</span> left_idx <span class="hljs-keyword">is</span> <span class="hljs-literal">None</span>:  <span class="hljs-comment"># If found something while walking left but not while walking right, return left</span>
        <span class="hljs-keyword">return</span> right_idx
    <span class="hljs-keyword">if</span> right_idx <span class="hljs-keyword">is</span> <span class="hljs-literal">None</span>:  <span class="hljs-comment"># If found something while walking right but not while walking left, return right</span>
        <span class="hljs-keyword">return</span> left_idx
    <span class="hljs-comment"># Otherwise, compare left and right to see which is close to theo_mid_mass and return that</span>
    left_exp_mass = exp_spec[left_idx]
    left_dist_to_theo_mid_mass = <span class="hljs-built_in">abs</span>(left_exp_mass - theo_exact_mass)
    right_exp_mass = exp_spec[left_idx]
    right_dist_to_theo_mid_mass = <span class="hljs-built_in">abs</span>(right_exp_mass - theo_exact_mass)
    <span class="hljs-keyword">if</span> left_dist_to_theo_mid_mass &lt; right_dist_to_theo_mid_mass:
        <span class="hljs-keyword">return</span> left_idx
    <span class="hljs-keyword">else</span>:
        <span class="hljs-keyword">return</span> right_idx


<span class="hljs-keyword">def</span> <span class="hljs-title function_">score_spectrums</span>(<span class="hljs-params">
        exp_spec: <span class="hljs-type">List</span>[<span class="hljs-built_in">float</span>],  <span class="hljs-comment"># must be sorted asc</span>
        theo_spec_with_tolerances: <span class="hljs-type">List</span>[<span class="hljs-type">Tuple</span>[<span class="hljs-built_in">float</span>, <span class="hljs-built_in">float</span>, <span class="hljs-built_in">float</span>]]  <span class="hljs-comment"># must be sorted asc, items are (expected,min,max)</span>
</span>) -&gt; <span class="hljs-type">Tuple</span>[<span class="hljs-built_in">int</span>, <span class="hljs-built_in">float</span>, <span class="hljs-built_in">float</span>]:
    dist_score = <span class="hljs-number">0.0</span>
    within_score = <span class="hljs-number">0</span>
    exp_spec_lo_idx = <span class="hljs-number">0</span>
    exp_spec_hi_idx = <span class="hljs-built_in">len</span>(exp_spec)
    <span class="hljs-keyword">for</span> theo_mass <span class="hljs-keyword">in</span> theo_spec_with_tolerances:
        <span class="hljs-comment"># Find closest exp_spec mass for theo_mass</span>
        theo_exact_mass, theo_min_mass, theo_max_mass = theo_mass
        exp_idx = find_closest_within_tolerance(
            exp_spec,
            exp_spec_lo_idx,
            exp_spec_hi_idx,
            theo_exact_mass,
            theo_min_mass,
            theo_max_mass
        )
        <span class="hljs-keyword">if</span> exp_idx <span class="hljs-keyword">is</span> <span class="hljs-literal">None</span>:
            <span class="hljs-keyword">continue</span>
        <span class="hljs-comment"># Calculate how far the found mass is from the ideal mass (theo_exact_mass) -- a perfect match will add 1.0 to</span>
        <span class="hljs-comment"># score, the farther out it is away the less gets added to score (min added will be 0.5).</span>
        exp_mass = exp_spec[exp_idx]
        dist = <span class="hljs-built_in">abs</span>(exp_mass - theo_exact_mass)
        max_dist = theo_max_mass - theo_min_mass
        <span class="hljs-keyword">if</span> max_dist &gt; <span class="hljs-number">0.0</span>:
            closeness = <span class="hljs-number">1.0</span> - (dist / max_dist)
        <span class="hljs-keyword">else</span>:
            closeness = <span class="hljs-number">1.0</span>
        dist_score += closeness
        <span class="hljs-comment"># Increment within_score for each match. The above block increases dist_score as the found mass gets closer to</span>
        <span class="hljs-comment"># theo_exact_mass. There may be a case where a peptide with 6 of 10 AAs matches exactly (6 * 1.0) while another</span>
        <span class="hljs-comment"># peptide with 10 of 10 AAs matching very loosely (10 * 0.5) -- the first peptide will incorrectly win out if</span>
        <span class="hljs-comment"># only dist_score were used.</span>
        within_score += <span class="hljs-number">1</span>
        <span class="hljs-comment"># Move up the lower bound for what to consider in exp_spec such that it it's after the exp_spec mass found</span>
        <span class="hljs-comment"># in this cycle. That is, the next cycle won't consider anything lower than the mass that was found here. This</span>
        <span class="hljs-comment"># is done because theo_spec may contain multiple copies of the same mass, but a real experimental spectrum won't</span>
        <span class="hljs-comment"># do that (e.g. a peptide containing 57 twice will have two entries for 57 in its theoretical spectrum, but a</span>
        <span class="hljs-comment"># real experimental spectrum for that same peptide will only contain 57 -- anything with mass of 57 will be</span>
        <span class="hljs-comment"># collected into the same bin).</span>
        exp_spec_lo_idx = exp_idx + <span class="hljs-number">1</span>
        <span class="hljs-keyword">if</span> exp_spec_lo_idx == exp_spec_hi_idx:
            <span class="hljs-keyword">break</span>
    <span class="hljs-keyword">return</span> within_score, dist_score, <span class="hljs-number">0.0</span> <span class="hljs-keyword">if</span> within_score == <span class="hljs-number">0</span> <span class="hljs-keyword">else</span> dist_score / within_score
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>The spectrum score for...</p>
<p>[0.0, 56.1, 71.9, 126.8, 200.6, 250.9]</p>
<p>... vs ...</p>
<p>[0.0, 57.0, 71.0, 128.0, 128.0, 199.0, 256.0]</p>
<p>... with 2.0 amino acid tolerance is...</p>
<p>(6, 4.624999999999999, 0.7708333333333331)</p>
</div>
<p></p>
<a name="H_Spectrum%20Sequence"></a><h3>Spectrum Sequence</h3>
<p><a name="BM_(Algorithms%5C%2FPeptide%20Sequence%2FSpectrum%20Sequence)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FPeptide%20Sequence%2FExperimental%20Spectrum)_TOPIC%2F">Algorithms/Peptide Sequence/Experimental Spectrum</a></li>
<li><a href="#BM_(Algorithms%5C%2FPeptide%20Sequence%2FTheoretical%20Spectrum)_TOPIC%2F">Algorithms/Peptide Sequence/Theoretical Spectrum</a></li>
<li><a href="#BM_(Algorithms%5C%2FPeptide%20Sequence%2FSpectrum%20Convolution)_TOPIC%2F">Algorithms/Peptide Sequence/Spectrum Convolution</a></li>
<li><a href="#BM_(Algorithms%5C%2FPeptide%20Sequence%2FSpectrum%20Score)_TOPIC%2F">Algorithms/Peptide Sequence/Spectrum Score</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: Given an <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> and a set of <a href="#BM_(amino%20acid)%2Fi">amino acid</a> masses, generate <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a>s and score them against the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> in an effort to infer the <a href="#BM_(peptide)%2Fi">peptide</a> <a href="#BM_(sequence)%2Fi">sequence</a> of the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>.</p>
<p><strong>WHY</strong>: The more matching masses between a <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a> and an <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>, the more likely it is that the <a href="#BM_(peptide)%2Fi">peptide</a> <a href="#BM_(sequence)%2Fi">sequence</a> used to generate that <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a> is related to the <a href="#BM_(peptide)%2Fi">peptide</a> <a href="#BM_(sequence)%2Fi">sequence</a> that produced that <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>.</p>
<a name="H3_Bruteforce%20Algorithm"></a><h4>Bruteforce Algorithm</h4>
<p><a name="BM_(Algorithms%5C%2FPeptide%20Sequence%2FSpectrum%20Sequence%2FBruteforce%20Algorithm)_TOPIC%2F"></a><strong></strong></p>
<p><strong>ALGORITHM</strong>:</p>
<p>Imagine if <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>s were perfect just like <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a>s: no missing masses, no faulty masses, no noise, and preserved repeat masses. To bruteforce the <a href="#BM_(peptide)%2Fi">peptide</a> that produced such an <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>, generate candidate <a href="#BM_(peptide)%2Fi">peptide</a>s by branching out <a href="#BM_(amino%20acid)%2Fi">amino acid</a>s at each position and compare each candidate <a href="#BM_(peptide)%2Fi">peptide</a>'s <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a> to the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>. If the <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a> matches the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>, it's reasonable to assume that <a href="#BM_(peptide)%2Fi">peptide</a> is the same as the <a href="#BM_(peptide)%2Fi">peptide</a> that generated the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>.</p>
<p>The algorithm stops branching out once the mass of the candidate <a href="#BM_(peptide)%2Fi">peptide</a> exceeds the final mass in the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>. For a perfect <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>, the final mass is always the mass of the <a href="#BM_(peptide)%2Fi">peptide</a> that produced it. For example, for the <a href="#BM_(linear%20peptide)%2Fi">linear peptide</a> GAK ...</p>
<table>
<thead>
<tr>
<th></th>
<th></th>
<th>G</th>
<th>A</th>
<th>K</th>
<th>GA</th>
<th>AK</th>
<th>GAK</th>
</tr>
</thead>
<tbody>
<tr>
<td>Mass</td>
<td>0<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a></td>
<td>57<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a></td>
<td>71<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a></td>
<td>128<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a></td>
<td>128<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a></td>
<td>199<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a></td>
<td>256<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a></td>
</tr>
</tbody>
</table>
<p><a href="ch4_code/src/SequencePeptide_Naive_Bruteforce.py">ch4_code/src/SequencePeptide_Naive_Bruteforce.py</a> (lines 10 to 30):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">sequence_peptide</span>(<span class="hljs-params">
        exp_spec: <span class="hljs-type">List</span>[<span class="hljs-built_in">float</span>],  <span class="hljs-comment"># must be sorted asc</span>
        peptide_type: PeptideType,
        aa_mass_table: <span class="hljs-type">Dict</span>[AA, <span class="hljs-built_in">float</span>]
</span>) -&gt; <span class="hljs-type">List</span>[<span class="hljs-type">List</span>[AA]]:
    peptide_mass = exp_spec[-<span class="hljs-number">1</span>]
    candidate_peptides = [[]]
    final_peptides = []
    <span class="hljs-keyword">while</span> <span class="hljs-built_in">len</span>(candidate_peptides) &gt; <span class="hljs-number">0</span>:
        new_candidate_peptides = []
        <span class="hljs-keyword">for</span> p <span class="hljs-keyword">in</span> candidate_peptides:
            <span class="hljs-keyword">for</span> m <span class="hljs-keyword">in</span> aa_mass_table.keys():
                new_p = p[:] + [m]
                new_p_mass = <span class="hljs-built_in">sum</span>([aa_mass_table[aa] <span class="hljs-keyword">for</span> aa <span class="hljs-keyword">in</span> new_p])
                <span class="hljs-keyword">if</span> new_p_mass == peptide_mass <span class="hljs-keyword">and</span> theoretical_spectrum(new_p, peptide_type, aa_mass_table) == exp_spec:
                    final_peptides.append(new_p)
                <span class="hljs-keyword">elif</span> new_p_mass &lt; peptide_mass:
                    new_candidate_peptides.append(new_p)
        candidate_peptides = new_candidate_peptides
    <span class="hljs-keyword">return</span> final_peptides
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>The linear peptides matching the experimental spectrum [0.0, 57.0, 71.0, 128.0, 128.0, 199.0, 256.0] are...</p>
<ul>
<li>GAK</li>
<li>GAQ</li>
<li>KAG</li>
<li>QAG</li>
</ul>
</div>
<p></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The following section isn't from the Pevzner book or any online resources. I came up with it in an effort to solve the final assignment for Chapter 4 (the chapter on <a href="#BM_(non-ribosomal%20peptide)%2Fi">non-ribosomal peptide</a> <a href="#BM_(sequencing%7Csequenced)%2Fi">sequencing</a>). As such, it might not be entirely correct / there may be better ways to do this.</p>
</div>
<p>Even though real <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>s aren't perfect, the high-level algorithm remains the same: Create candidate <a href="#BM_(peptide)%2Fi">peptide</a>s by branching out <a href="#BM_(amino%20acid)%2Fi">amino acid</a>s and capture the best scoring ones until the mass goes too high. However, various low-level aspects of the algorithm need to be modified to handle the problems with real <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>s.</p>
<p>For starters, since there are no preset <a href="#BM_(amino%20acid)%2Fi">amino acid</a>s to build candidate <a href="#BM_(peptide)%2Fi">peptide</a>s with, <a href="#BM_(amino%20acid)%2Fi">amino acid</a> masses are captured using <a href="#BM_(spectrum%20convolution)%2Fi">spectrum convolution</a> and used directly. For example, instead of representing a <a href="#BM_(peptide)%2Fi">peptide</a> as GAK, it's represented as 57-71-128.</p>
<table>
<thead>
<tr>
<th>G</th>
<th>A</th>
<th>K</th>
</tr>
</thead>
<tbody>
<tr>
<td>57<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a></td>
<td>71<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a></td>
<td>128<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a></td>
</tr>
</tbody>
</table>
<p>Next, the last mass in a real <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> isn't guaranteed to be the mass of the <a href="#BM_(peptide)%2Fi">peptide</a> that produced it. Since real <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>s have faulty masses and may be missing masses, it's possible that either the <a href="#BM_(peptide)%2Fi">peptide</a>'s mass wasn't captured at all or was captured but at an index that isn't the last element.</p>
<p>If the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>'s <a href="#BM_(peptide)%2Fi">peptide</a> mass was captured and found, it'll have noise. For example, imagine an <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> for the <a href="#BM_(peptide)%2Fi">peptide</a> 57-57 with ±1<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> noise. The exact mass of the <a href="#BM_(peptide)%2Fi">peptide</a> 57-57 is 114<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, but if that mass gets placed into the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> it will show up as anywhere between 113<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> to 115<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>.</p>
<p>Given that same <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>, running a <a href="#BM_(spectrum%20convolution)%2Fi">spectrum convolution</a> to derive the <a href="#BM_(amino%20acid)%2Fi">amino acid</a> masses ends up giving back <a href="#BM_(amino%20acid)%2Fi">amino acid</a> masses with ±2<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> noise. For example, the mass 57<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> may be derived as anywhere between 55<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> to 59<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>. Assuming that you're building the <a href="#BM_(peptide)%2Fi">peptide</a> 57-57 with the low end of that range (55<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>), its mass will be 55<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> + 55<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> = 110<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>. Compared against the high end of the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>'s <a href="#BM_(peptide)%2Fi">peptide</a> mass (115<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>), it's 5<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> away.</p>
<p><a href="ch4_code/src/ExperimentalSpectrumPeptideMassNoise.py">ch4_code/src/ExperimentalSpectrumPeptideMassNoise.py</a> (lines 18 to 21):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">experimental_spectrum_peptide_mass_noise</span>(<span class="hljs-params">exp_spec_mass_noise: <span class="hljs-built_in">float</span>, peptide_len: <span class="hljs-built_in">int</span></span>) -&gt; <span class="hljs-built_in">float</span>:
    aa_mass_noise = spectrum_convolution_noise(exp_spec_mass_noise)
    <span class="hljs-keyword">return</span> aa_mass_noise * peptide_len + exp_spec_mass_noise
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given an experimental spectrum mass noise of ±1.0 and expected peptide length of 2, the maximum noise for an experimental spectrum's peptide mass is ±5.0</p>
</div>
<p></p>
<p>Finally, given that real <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>s contain faulty masses and may be missing masses, more often than not the <a href="#BM_(peptide)%2Fi">peptide</a>s that score the best aren't the best candidates. <a href="#BM_(theoretical%20spectrum)%2Fi">Theoretical spectrum</a> masses that are ...</p>
<ul>
<li>incorrect but match faulty <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> masses</li>
<li>correct but are missing in the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a></li>
</ul>
<p>... may push poor <a href="#BM_(peptide)%2Fi">peptide</a> candidates forward. As such, it makes sense to keep a backlog of the last m scoring <a href="#BM_(peptide)%2Fi">peptide</a>s. Any of these backlog <a href="#BM_(peptide)%2Fi">peptide</a>s may be the correct <a href="#BM_(peptide)%2Fi">peptide</a> for the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>.</p>
<p><a href="ch4_code/src/SequenceTester.py">ch4_code/src/SequenceTester.py</a> (lines 21 to 86):</p>
<pre class="hljs"><code><span class="hljs-keyword">class</span> <span class="hljs-title class_">SequenceTester</span>:
    <span class="hljs-keyword">def</span> <span class="hljs-title function_">__init__</span>(<span class="hljs-params">
            self,
            exp_spec: <span class="hljs-type">List</span>[<span class="hljs-built_in">float</span>],           <span class="hljs-comment"># must be sorted asc</span>
            aa_mass_table: <span class="hljs-type">Dict</span>[AA, <span class="hljs-built_in">float</span>],  <span class="hljs-comment"># amino acid mass table</span>
            aa_mass_tolerance: <span class="hljs-built_in">float</span>,        <span class="hljs-comment"># amino acid mass tolerance</span>
            peptide_min_mass: <span class="hljs-built_in">float</span>,         <span class="hljs-comment"># min mass that the peptide could be</span>
            peptide_max_mass: <span class="hljs-built_in">float</span>,         <span class="hljs-comment"># max mass that the peptide could be</span>
            peptide_type: PeptideType,       <span class="hljs-comment"># linear or cyclic</span>
            score_backlog: <span class="hljs-built_in">int</span> = <span class="hljs-number">0</span>           <span class="hljs-comment"># keep this many previous scores</span>
    </span>):
        self.exp_spec = exp_spec
        self.aa_mass_table = aa_mass_table
        self.aa_mass_tolerance = aa_mass_tolerance
        self.peptide_min_mass = peptide_min_mass
        self.peptide_max_mass = peptide_max_mass
        self.peptide_type = peptide_type
        self.score_backlog = score_backlog
        self.leader_peptides_top_score = <span class="hljs-number">0</span>
        self.leader_peptides = {<span class="hljs-number">0</span>: []}

<span class="hljs-meta">    @staticmethod</span>
    <span class="hljs-keyword">def</span> <span class="hljs-title function_">generate_theroetical_spectrum_with_tolerances</span>(<span class="hljs-params">
            peptide: <span class="hljs-type">List</span>[AA],
            peptide_type: PeptideType,
            aa_mass_table: <span class="hljs-type">Dict</span>[AA, <span class="hljs-built_in">float</span>],
            aa_mass_tolerance: <span class="hljs-built_in">float</span>
    </span>) -&gt; <span class="hljs-type">List</span>[<span class="hljs-type">Tuple</span>[<span class="hljs-built_in">float</span>, <span class="hljs-built_in">float</span>, <span class="hljs-built_in">float</span>]]:
        theo_spec_raw = theoretical_spectrum(peptide, peptide_type, aa_mass_table)
        theo_spec_tols = theoretical_spectrum_tolerances(<span class="hljs-built_in">len</span>(peptide), peptide_type, aa_mass_tolerance)
        theo_spec = [(m, m - t, m + t) <span class="hljs-keyword">for</span> m, t <span class="hljs-keyword">in</span> <span class="hljs-built_in">zip</span>(theo_spec_raw, theo_spec_tols)]
        <span class="hljs-keyword">return</span> theo_spec

    <span class="hljs-keyword">def</span> <span class="hljs-title function_">test</span>(<span class="hljs-params">
            self,
            peptide: <span class="hljs-type">List</span>[AA],
            theo_spec: <span class="hljs-type">Optional</span>[<span class="hljs-type">List</span>[<span class="hljs-type">Tuple</span>[<span class="hljs-built_in">float</span>, <span class="hljs-built_in">float</span>, <span class="hljs-built_in">float</span>]]] = <span class="hljs-literal">None</span>
    </span>) -&gt; TestResult:
        <span class="hljs-keyword">if</span> theo_spec <span class="hljs-keyword">is</span> <span class="hljs-literal">None</span>:
            theo_spec = SequenceTester.generate_theroetical_spectrum_with_tolerances(
                peptide,
                self.peptide_type,
                self.aa_mass_table,
                self.aa_mass_tolerance
            )
        <span class="hljs-comment"># Don't add if mass out of range</span>
        _, tp_min_mass, tp_max_mass = theo_spec[-<span class="hljs-number">1</span>]  <span class="hljs-comment"># last element of theo spec is the mass of the theo spec peptide</span>
        <span class="hljs-keyword">if</span> tp_min_mass &lt; self.peptide_min_mass:
            <span class="hljs-keyword">return</span> TestResult.MASS_TOO_SMALL
        <span class="hljs-keyword">elif</span> tp_max_mass &gt; self.peptide_max_mass:
            <span class="hljs-keyword">return</span> TestResult.MASS_TOO_LARGE
        <span class="hljs-comment"># Don't add if the score is lower than the previous n best scores</span>
        peptide_score = score_spectrums(self.exp_spec, theo_spec)[<span class="hljs-number">0</span>]
        min_acceptable_score = self.leader_peptides_top_score - self.score_backlog
        <span class="hljs-keyword">if</span> peptide_score &lt; min_acceptable_score:
            <span class="hljs-keyword">return</span> TestResult.SCORE_TOO_LOW
        <span class="hljs-comment"># Add, but also remove any previous test peptides that are no longer within the acceptable score threshold</span>
        leaders = self.leader_peptides.setdefault(peptide_score, [])
        leaders.append(peptide)
        <span class="hljs-keyword">if</span> peptide_score &gt; self.leader_peptides_top_score:
            self.leader_peptides_top_score = peptide_score
            <span class="hljs-keyword">if</span> <span class="hljs-built_in">len</span>(self.leader_peptides) &gt;= self.score_backlog:
                smallest_leader_score = <span class="hljs-built_in">min</span>(self.leader_peptides.keys())
                self.leader_peptides.pop(smallest_leader_score)
        <span class="hljs-keyword">return</span> TestResult.ADDED
</code></pre>
<p><a href="ch4_code/src/SequencePeptide_Bruteforce.py">ch4_code/src/SequencePeptide_Bruteforce.py</a> (lines 13 to 41):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">sequence_peptide</span>(<span class="hljs-params">
        exp_spec: <span class="hljs-type">List</span>[<span class="hljs-built_in">float</span>],                               <span class="hljs-comment"># must be sorted asc</span>
        aa_mass_table: <span class="hljs-type">Dict</span>[AA, <span class="hljs-built_in">float</span>],                      <span class="hljs-comment"># amino acid mass table</span>
        aa_mass_tolerance: <span class="hljs-built_in">float</span>,                            <span class="hljs-comment"># amino acid mass tolerance</span>
        peptide_mass_candidates: <span class="hljs-type">List</span>[<span class="hljs-type">Tuple</span>[<span class="hljs-built_in">float</span>, <span class="hljs-built_in">float</span>]],  <span class="hljs-comment"># mass range candidates for mass of peptide</span>
        peptide_type: PeptideType,                           <span class="hljs-comment"># linear or cyclic</span>
        score_backlog: <span class="hljs-built_in">int</span>                                   <span class="hljs-comment"># backlog of top scores</span>
</span>) -&gt; SequenceTesterSet:
    tester_set = SequenceTesterSet(
        exp_spec,
        aa_mass_table,
        aa_mass_tolerance,
        peptide_mass_candidates,
        peptide_type,
        score_backlog
    )
    candidates = [[]]
    <span class="hljs-keyword">while</span> <span class="hljs-built_in">len</span>(candidates) &gt; <span class="hljs-number">0</span>:
        new_candidate_peptides = []
        <span class="hljs-keyword">for</span> p <span class="hljs-keyword">in</span> candidates:
            <span class="hljs-keyword">for</span> m <span class="hljs-keyword">in</span> aa_mass_table.keys():
                new_p = p[:]
                new_p.append(m)
                res = <span class="hljs-built_in">set</span>(tester_set.test(new_p))
                <span class="hljs-keyword">if</span> res != {TestResult.MASS_TOO_LARGE}:
                    new_candidate_peptides.append(new_p)
        candidates = new_candidate_peptides
    <span class="hljs-keyword">return</span> tester_set
</code></pre>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> in the example below is for the <a href="#BM_(peptide)%2Fi">peptide</a> 114-128-163, which has the <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a> [0, 114, 128, 163, 242, 291, 405].</p>
</div>
<div style="border:1px solid black;">
<p></p>
<p>Given the ...</p>
<ul>
<li>experimental spectrum: [0.0, 112.5, 127.1, 242.9, 290.0, 404.0]</li>
<li>experimental spectrum mass noise: ±1.0</li>
<li>assumed peptide type: linear</li>
<li>assumed peptide length: 3</li>
<li>assumed peptide mass: any of the last 1 experimental spectrum masses</li>
<li>score backlog: 0</li>
</ul>
<p>Top 10 captured mino acid masses (rounded to 1): [114.0, 112.5, 115.8, 161.1, 162.9, 127.1, 130.4, 177.5]</p>
<p>For peptides between 397.0 and 411.0...</p>
<ul>
<li>Score 6: 114.0-127.1-162.9</li>
<li>Score 6: 162.9-127.1-114.0</li>
</ul>
</div>
<p></p>
<a name="H_Branch-and-bound%20Algorithm"></a><h4>Branch-and-bound Algorithm</h4>
<p><a name="BM_(Algorithms%5C%2FPeptide%20Sequence%2FSpectrum%20Sequence%2FBranch-and-bound%20Algorithm)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FPeptide%20Sequence%2FSpectrum%20Sequence%2FBruteforce%20Algorithm)_TOPIC%2F">Algorithms/Peptide Sequence/Spectrum Sequence/Bruteforce Algorithm</a></li>
</ul>
</div>
<p><strong>ALGORITHM</strong>:</p>
<p>This algorithm extends the bruteforce algorithm into a more efficient <a href="#BM_(branch-and-bound%20algorithm%7Cbranch%20and%20bound%20algorithm)%2Fi">branch-and-bound algorithm</a> by adding one extra step: After each branch, any candidate <a href="#BM_(peptide)%2Fi">peptide</a>s deemed to be untenable are discarded. In this case, untenable means that there's no chance / little chance of the <a href="#BM_(peptide)%2Fi">peptide</a> branching out to a correct solution.</p>
<p>Imagine if <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>s were perfect just like <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a>s: no missing masses, no faulty masses, no noise, and preserved repeat masses. For such an <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>, an untenable candidate <a href="#BM_(peptide)%2Fi">peptide</a> has a <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a> with at least one mass that don't exist in the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>. For example, the <a href="#BM_(peptide)%2Fi">peptide</a> 57-71-128 has the <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a> [0<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 57<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 71<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 128<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 128<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 199<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>, 256<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a>]. If 71<a href="#BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi">Da</a> were missing from the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>, that <a href="#BM_(peptide)%2Fi">peptide</a> would be untenable (won't move forward).</p>
<p>When testing if a candidate <a href="#BM_(peptide)%2Fi">peptide</a> should move forward, the candidate <a href="#BM_(peptide)%2Fi">peptide</a> be treated as a <a href="#BM_(linear%20peptide)%2Fi">linear peptide</a> even if the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> is for a <a href="#BM_(cyclopeptide%7Ccyclic%20peptide)%2Fi">cyclic peptide</a>. For example, testing the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> for <a href="#BM_(cyclopeptide%7Ccyclic%20peptide)%2Fi">cyclic peptide</a> NQYQ against the <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a> for candidate <a href="#BM_(cyclopeptide%7Ccyclic%20peptide)%2Fi">cyclic peptide</a> NQY...</p>
<table>
<thead>
<tr>
<th><a href="#BM_(peptide)%2Fi">Peptide</a></th>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
<th>6</th>
<th>7</th>
<th>8</th>
<th>9</th>
<th>10</th>
<th>11</th>
<th>12</th>
<th>13</th>
<th>14</th>
</tr>
</thead>
<tbody>
<tr>
<td>NQYQ</td>
<td>0</td>
<td>114</td>
<td>128</td>
<td>128</td>
<td>163</td>
<td>242</td>
<td>242</td>
<td></td>
<td>291</td>
<td>291</td>
<td>370</td>
<td>405</td>
<td>405</td>
<td>419</td>
<td>533</td>
</tr>
<tr>
<td>NQY</td>
<td>0</td>
<td>114</td>
<td>128</td>
<td></td>
<td>163</td>
<td>242</td>
<td></td>
<td>277</td>
<td>291</td>
<td></td>
<td></td>
<td>405</td>
<td></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<p>The <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a> contains 277, but the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> doesn't. That means NQY won't branch out any further even though it should. As such, even if the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> is for a <a href="#BM_(cyclopeptide%7Ccyclic%20peptide)%2Fi">cyclic peptide</a>, treat candidate <a href="#BM_(peptide)%2Fi">peptide</a>s as if they're linear segments of a <a href="#BM_(cyclopeptide%7Ccyclic%20peptide)%2Fi">cyclic peptide</a> (essentially the same as <a href="#BM_(linear%20peptide)%2Fi">linear peptide</a>s). If the <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a> for candidate <a href="#BM_(linear%20peptide)%2Fi">linear peptide</a> NQY were used...</p>
<table>
<thead>
<tr>
<th><a href="#BM_(peptide)%2Fi">Peptide</a></th>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
<th>6</th>
<th>7</th>
<th>8</th>
<th>9</th>
<th>10</th>
<th>11</th>
<th>12</th>
<th>13</th>
</tr>
</thead>
<tbody>
<tr>
<td>NQYQ</td>
<td>0</td>
<td>114</td>
<td>128</td>
<td>128</td>
<td>163</td>
<td>242</td>
<td>242</td>
<td>291</td>
<td>291</td>
<td>370</td>
<td>405</td>
<td>405</td>
<td>419</td>
<td>533</td>
</tr>
<tr>
<td>NQY</td>
<td>0</td>
<td>114</td>
<td>128</td>
<td></td>
<td>163</td>
<td>242</td>
<td></td>
<td>291</td>
<td></td>
<td></td>
<td>405</td>
<td></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<p>All <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a> masses are in the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>. As such, the candidate NQY would move forward.</p>
<p><a href="ch4_code/src/SequencePeptide_Naive_BranchAndBound.py">ch4_code/src/SequencePeptide_Naive_BranchAndBound.py</a> (lines 11 to 61):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">sequence_peptide</span>(<span class="hljs-params">
        exp_spec: <span class="hljs-type">List</span>[<span class="hljs-built_in">float</span>],  <span class="hljs-comment"># must be sorted asc</span>
        peptide_type: PeptideType,
        aa_mass_table: <span class="hljs-type">Dict</span>[AA, <span class="hljs-built_in">float</span>]
</span>) -&gt; <span class="hljs-type">List</span>[<span class="hljs-type">List</span>[AA]]:
    peptide_mass = exp_spec[-<span class="hljs-number">1</span>]
    candidate_peptides = [[]]
    final_peptides = []
    <span class="hljs-keyword">while</span> <span class="hljs-built_in">len</span>(candidate_peptides) &gt; <span class="hljs-number">0</span>:
        <span class="hljs-comment"># Branch candidates</span>
        new_candidate_peptides = []
        <span class="hljs-keyword">for</span> p <span class="hljs-keyword">in</span> candidate_peptides:
            <span class="hljs-keyword">for</span> m <span class="hljs-keyword">in</span> aa_mass_table:
                new_p = p[:] + [m]
                new_candidate_peptides.append(new_p)
        candidate_peptides = new_candidate_peptides
        <span class="hljs-comment"># Test candidates to see if they match exp_spec or if they should keep being branched</span>
        removal_idxes = <span class="hljs-built_in">set</span>()
        <span class="hljs-keyword">for</span> i, p <span class="hljs-keyword">in</span> <span class="hljs-built_in">enumerate</span>(candidate_peptides):
            p_mass = <span class="hljs-built_in">sum</span>([aa_mass_table[aa] <span class="hljs-keyword">for</span> aa <span class="hljs-keyword">in</span> p])
            <span class="hljs-keyword">if</span> p_mass == peptide_mass:
                theo_spec = theoretical_spectrum(p, peptide_type, aa_mass_table)
                <span class="hljs-keyword">if</span> theo_spec == exp_spec:
                    final_peptides.append(p)
                removal_idxes.add(i)
            <span class="hljs-keyword">else</span>:
                <span class="hljs-comment"># Why get the theo spec of the linear version even if the peptide is cyclic? Think about what's</span>
                <span class="hljs-comment"># happening here. If the exp spec is for cyclic peptide NQYQ, and you're checking to see if the</span>
                <span class="hljs-comment"># candidate NQY should continue to be branched out...</span>
                <span class="hljs-comment">#</span>
                <span class="hljs-comment"># Exp spec  cyclic NQYQ: [0, 114, 128, 128, 163, 242, 242,      291, 291, 370, 405, 405, 419, 533]</span>
                <span class="hljs-comment"># Theo spec cyclic NQY:  [0, 114, 128,      163, 242,      277, 291,           405]</span>
                <span class="hljs-comment">#                                                           ^</span>
                <span class="hljs-comment">#                                                           |</span>
                <span class="hljs-comment">#                                                        mass(YN)</span>
                <span class="hljs-comment">#</span>
                <span class="hljs-comment"># Since NQY is being treated as a cyclic peptide, it has the subpeptide YN (mass of 277). However, the</span>
                <span class="hljs-comment"># cyclic peptide NQYQ doesn't have the subpeptide YN. That means NQY won't be branched out any further</span>
                <span class="hljs-comment"># even though it should. As such, even if the exp spec is for a cyclic peptide, treat the candidates as</span>
                <span class="hljs-comment"># linear segments of that cyclic peptide (essentially linear peptides).</span>
                <span class="hljs-comment">#</span>
                <span class="hljs-comment"># Exp spec  cyclic NQYQ: [0, 114, 128, 128, 163, 242, 242, 291, 291, 370, 405, 405, 419, 533]</span>
                <span class="hljs-comment"># Theo spec linear NQY:  [0, 114, 128,      163, 242,      291,           405]</span>
                <span class="hljs-comment">#</span>
                <span class="hljs-comment"># Given the specs above, the exp spec contains all masses in the theo spec.</span>
                theo_spec = theoretical_spectrum(p, PeptideType.LINEAR, aa_mass_table)
                <span class="hljs-keyword">if</span> <span class="hljs-keyword">not</span> contains_all_sorted(theo_spec, exp_spec):
                    removal_idxes.add(i)
        candidate_peptides = [p <span class="hljs-keyword">for</span> i, p <span class="hljs-keyword">in</span> <span class="hljs-built_in">enumerate</span>(candidate_peptides) <span class="hljs-keyword">if</span> i <span class="hljs-keyword">not</span> <span class="hljs-keyword">in</span> removal_idxes]
    <span class="hljs-keyword">return</span> final_peptides
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>The cyclic peptides matching the experimental spectrum [0.0, 114.0, 128.0, 128.0, 163.0, 242.0, 242.0, 291.0, 291.0, 370.0, 405.0, 405.0, 419.0, 533.0] are...</p>
<ul>
<li>NKYK</li>
<li>NKYQ</li>
<li>NQYK</li>
<li>NQYQ</li>
<li>KNKY</li>
<li>KNQY</li>
<li>KYKN</li>
<li>KYQN</li>
<li>QNKY</li>
<li>QNQY</li>
<li>QYKN</li>
<li>QYQN</li>
<li>YKNK</li>
<li>YKNQ</li>
<li>YQNK</li>
<li>YQNQ</li>
</ul>
</div>
<p></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The following section isn't from the Pevzner book or any online resources. I came up with it in an effort to solve the final assignment for Chapter 4 (the chapter on <a href="#BM_(non-ribosomal%20peptide)%2Fi">non-ribosomal peptide</a> <a href="#BM_(sequencing%7Csequenced)%2Fi">sequencing</a>). As such, it might not be entirely correct / there may be better ways to do this.</p>
</div>
<p>The bounding step described above won't work for real <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>s. For example, a real <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> may ...</p>
<ul>
<li>have a faulty mass that allows candidate <a href="#BM_(peptide)%2Fi">peptide</a>s that should be untenable.</li>
<li>be missing a mass that drops candidate <a href="#BM_(peptide)%2Fi">peptide</a>s that should be good.</li>
<li>have noise that causes good candidate <a href="#BM_(peptide)%2Fi">peptide</a> to be dropped / untenable candidate <a href="#BM_(peptide)%2Fi">peptide</a> through.</li>
</ul>
<p>A possible bounding step for real <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>s is to mark a candidate <a href="#BM_(peptide)%2Fi">peptide</a> as untenable if it has a certain number or percentage of mismatches. This is a heuristic, meaning that it won't always lead to the correct <a href="#BM_(peptide)%2Fi">peptide</a>. In contrast, the algorithm described above for perfect <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>s always leads to the correct <a href="#BM_(peptide)%2Fi">peptide</a>.</p>
<p><a href="ch4_code/src/SequencePeptide_BranchAndBound.py">ch4_code/src/SequencePeptide_BranchAndBound.py</a> (lines 14 to 78):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">sequence_peptide</span>(<span class="hljs-params">
        exp_spec: <span class="hljs-type">List</span>[<span class="hljs-built_in">float</span>],                               <span class="hljs-comment"># must be sorted asc</span>
        aa_mass_table: <span class="hljs-type">Dict</span>[AA, <span class="hljs-built_in">float</span>],                      <span class="hljs-comment"># amino acid mass table</span>
        aa_mass_tolerance: <span class="hljs-built_in">float</span>,                            <span class="hljs-comment"># amino acid mass tolerance</span>
        peptide_mass_candidates: <span class="hljs-type">List</span>[<span class="hljs-type">Tuple</span>[<span class="hljs-built_in">float</span>, <span class="hljs-built_in">float</span>]],  <span class="hljs-comment"># mass range candidates for mass of peptide</span>
        peptide_type: PeptideType,                           <span class="hljs-comment"># linear or cyclic</span>
        score_backlog: <span class="hljs-built_in">int</span>,                                  <span class="hljs-comment"># backlog of top scores</span>
        candidate_threshold: <span class="hljs-built_in">float</span>                           <span class="hljs-comment"># if &lt; 1 then min % match, else min count match</span>
</span>) -&gt; SequenceTesterSet:
    tester_set = SequenceTesterSet(
        exp_spec,
        aa_mass_table,
        aa_mass_tolerance,
        peptide_mass_candidates,
        peptide_type,
        score_backlog
    )
    candidate_peptides = [[]]
    <span class="hljs-keyword">while</span> <span class="hljs-built_in">len</span>(candidate_peptides) &gt; <span class="hljs-number">0</span>:
        <span class="hljs-comment"># Branch candidates</span>
        new_candidate_peptides = []
        <span class="hljs-keyword">for</span> p <span class="hljs-keyword">in</span> candidate_peptides:
            <span class="hljs-keyword">for</span> m <span class="hljs-keyword">in</span> aa_mass_table:
                new_p = p[:] + [m]
                new_candidate_peptides.append(new_p)
        candidate_peptides = new_candidate_peptides
        <span class="hljs-comment"># Test candidates to see if they match exp_spec or if they should keep being branched</span>
        removal_idxes = <span class="hljs-built_in">set</span>()
        <span class="hljs-keyword">for</span> i, p <span class="hljs-keyword">in</span> <span class="hljs-built_in">enumerate</span>(candidate_peptides):
            res = <span class="hljs-built_in">set</span>(tester_set.test(p))
            <span class="hljs-keyword">if</span> {TestResult.MASS_TOO_LARGE} == res:
                removal_idxes.add(i)
            <span class="hljs-keyword">else</span>:
                <span class="hljs-comment"># Why get the theo spec of the linear version even if the peptide is cyclic? Think about what's</span>
                <span class="hljs-comment"># happening here. If the exp spec is for cyclic peptide NQYQ, and you're checking to see if the</span>
                <span class="hljs-comment"># candidate NQY should continue to be branched out...</span>
                <span class="hljs-comment">#</span>
                <span class="hljs-comment"># Exp spec  cyclic NQYQ: [0, 114, 128, 128, 163, 242, 242,      291, 291, 370, 405, 405, 419, 533]</span>
                <span class="hljs-comment"># Theo spec cyclic NQY:  [0, 114, 128,      163, 242,      277, 291,           405]</span>
                <span class="hljs-comment">#                                                           ^</span>
                <span class="hljs-comment">#                                                           |</span>
                <span class="hljs-comment">#                                                        mass(YN)</span>
                <span class="hljs-comment">#</span>
                <span class="hljs-comment"># Since NQY is being treated as a cyclic peptide, it has the subpeptide YN (mass of 277). However, the</span>
                <span class="hljs-comment"># cyclic peptide NQYQ doesn't have the subpeptide YN. That means NQY won't be branched out any further</span>
                <span class="hljs-comment"># even though it should. As such, even if the exp spec is for a cyclic peptide, treat the candidates as</span>
                <span class="hljs-comment"># linear segments of that cyclic peptide (essentially linear peptides).</span>
                <span class="hljs-comment">#</span>
                <span class="hljs-comment"># Exp spec  cyclic NQYQ: [0, 114, 128, 128, 163, 242, 242, 291, 291, 370, 405, 405, 419, 533]</span>
                <span class="hljs-comment"># Theo spec linear NQY:  [0, 114, 128,      163, 242,      291,           405]</span>
                <span class="hljs-comment">#</span>
                <span class="hljs-comment"># Given the specs above, the exp spec contains all masses in the theo spec.</span>
                theo_spec = SequenceTester.generate_theroetical_spectrum_with_tolerances(
                    p,
                    PeptideType.LINEAR,
                    aa_mass_table,
                    aa_mass_tolerance
                )
                score = score_spectrums(exp_spec, theo_spec)
                <span class="hljs-keyword">if</span> (candidate_threshold &lt; <span class="hljs-number">1.0</span> <span class="hljs-keyword">and</span> score[<span class="hljs-number">0</span>] / <span class="hljs-built_in">len</span>(theo_spec) &lt; candidate_threshold)\
                        <span class="hljs-keyword">or</span> score[<span class="hljs-number">0</span>] &lt; candidate_threshold:
                    removal_idxes.add(i)
        candidate_peptides = [p <span class="hljs-keyword">for</span> i, p <span class="hljs-keyword">in</span> <span class="hljs-built_in">enumerate</span>(candidate_peptides) <span class="hljs-keyword">if</span> i <span class="hljs-keyword">not</span> <span class="hljs-keyword">in</span> removal_idxes]
    <span class="hljs-keyword">return</span> tester_set
</code></pre>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> in the example below is for the <a href="#BM_(peptide)%2Fi">peptide</a> 114-128-163, which has the <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a> [0, 114, 128, 163, 242, 291, 405].</p>
</div>
<div style="border:1px solid black;">
<p></p>
<p>Given the ...</p>
<ul>
<li>experimental spectrum: [0.0, 112.5, 127.1, 242.9, 290.0, 404.0]</li>
<li>experimental spectrum mass noise: ±1.0</li>
<li>assumed peptide type: linear</li>
<li>assumed peptide length: 3</li>
<li>assumed peptide mass: any of the last 1 experimental spectrum masses</li>
<li>score backlog: 0</li>
<li>candidate threshold: 75.0% mass matches per iteration</li>
</ul>
<p>Top 10 captured mino acid masses (rounded to 1): [114.0, 112.5, 115.8, 161.1, 162.9, 127.1, 130.4, 177.5]</p>
<p>For peptides between 397.0 and 411.0...</p>
<ul>
<li>Score 6: 114.0-127.1-162.9</li>
</ul>
</div>
<p></p>
<a name="H_Leaderboard%20Algorithm"></a><h4>Leaderboard Algorithm</h4>
<p><strong>ALGORITHM</strong>:</p>
<p><a name="BM_(Algorithms%5C%2FPeptide%20Sequence%2FSpectrum%20Sequence%2FLeaderboard%20Algorithm)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FPeptide%20Sequence%2FSpectrum%20Sequence%2FBruteforce%20Algorithm)_TOPIC%2F">Algorithms/Peptide Sequence/Spectrum Sequence/Bruteforce Algorithm</a></li>
<li><a href="#BM_(Algorithms%5C%2FPeptide%20Sequence%2FSpectrum%20Sequence%2FBranch-and-bound%20Algorithm)_TOPIC%2F">Algorithms/Peptide Sequence/Spectrum Sequence/Branch-and-bound Algorithm</a></li>
</ul>
</div>
<p>This algorithm is similar to the <a href="#BM_(branch-and-bound%20algorithm%7Cbranch%20and%20bound%20algorithm)%2Fi">branch-and-bound algorithm</a>, but the bounding step is slightly different: At each branch, rather than removing untenable candidate <a href="#BM_(peptide)%2Fi">peptide</a>s, it only moves forward the best n scoring candidate <a href="#BM_(peptide)%2Fi">peptide</a>s. These best scoring <a href="#BM_(peptide)%2Fi">peptide</a>s are referred to as the leaderboard.</p>
<p>For a perfect <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> (no missing masses, no faulty masses, no noise, and preserved repeat masses), this algorithm isn't much different than the <a href="#BM_(branch-and-bound%20algorithm%7Cbranch%20and%20bound%20algorithm)%2Fi">branch-and-bound algorithm</a>. However, imagine if the perfect <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> wasn't exactly perfect in that it could have faulty masses and could be missing masses. In such a case, the <a href="#BM_(branch-and-bound%20algorithm%7Cbranch%20and%20bound%20algorithm)%2Fi">branch-and-bound algorithm</a> would always fail while this algorithm could still converge to the correct <a href="#BM_(peptide)%2Fi">peptide</a> -- it's a heuristic, meaning that it isn't guaranteed to lead to the correct <a href="#BM_(peptide)%2Fi">peptide</a>.</p>
<p><a href="ch4_code/src/SequencePeptide_Naive_Leaderboard.py">ch4_code/src/SequencePeptide_Naive_Leaderboard.py</a> (lines 11 to 95):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">sequence_peptide</span>(<span class="hljs-params">
        exp_spec: <span class="hljs-type">List</span>[<span class="hljs-built_in">float</span>],  <span class="hljs-comment"># must be sorted</span>
        peptide_type: PeptideType,
        peptide_mass: <span class="hljs-type">Optional</span>[<span class="hljs-built_in">float</span>],
        aa_mass_table: <span class="hljs-type">Dict</span>[AA, <span class="hljs-built_in">float</span>],
        leaderboard_size: <span class="hljs-built_in">int</span>
</span>) -&gt; <span class="hljs-type">List</span>[<span class="hljs-type">List</span>[AA]]:
    <span class="hljs-comment"># Exp_spec could be missing masses / have faulty masses, but even so assume the last mass in exp_spec is the peptide</span>
    <span class="hljs-comment"># mass if the user didn't supply one. This may not be correct -- it's a best guess.</span>
    <span class="hljs-keyword">if</span> peptide_mass <span class="hljs-keyword">is</span> <span class="hljs-literal">None</span>:
        peptide_mass = exp_spec[-<span class="hljs-number">1</span>]
    leaderboard = [[]]
    final_peptides = [<span class="hljs-built_in">next</span>(<span class="hljs-built_in">iter</span>(leaderboard))]
    final_score = score_spectrums(
        theoretical_spectrum(final_peptides[<span class="hljs-number">0</span>], peptide_type, aa_mass_table),
        exp_spec
    )
    <span class="hljs-keyword">while</span> <span class="hljs-built_in">len</span>(leaderboard) &gt; <span class="hljs-number">0</span>:
        <span class="hljs-comment"># Branch leaderboard</span>
        expanded_leaderboard = []
        <span class="hljs-keyword">for</span> p <span class="hljs-keyword">in</span> leaderboard:
            <span class="hljs-keyword">for</span> m <span class="hljs-keyword">in</span> aa_mass_table:
                new_p = p[:] + [m]
                expanded_leaderboard.append(new_p)
        <span class="hljs-comment"># Pull out any expanded_leaderboard peptides with mass &gt;= peptide_mass</span>
        removal_idxes = <span class="hljs-built_in">set</span>()
        <span class="hljs-keyword">for</span> i, p <span class="hljs-keyword">in</span> <span class="hljs-built_in">enumerate</span>(expanded_leaderboard):
            p_mass = <span class="hljs-built_in">sum</span>([aa_mass_table[aa] <span class="hljs-keyword">for</span> aa <span class="hljs-keyword">in</span> p])
            <span class="hljs-keyword">if</span> p_mass == peptide_mass:
                <span class="hljs-comment"># The peptide's mass is equal to the expected mass. Check if the score against the current top score. If</span>
                <span class="hljs-comment"># it's ...</span>
                <span class="hljs-comment">#  * a higher score, reset the final peptides to it.</span>
                <span class="hljs-comment">#  * the same score, add it to the final peptides.</span>
                theo_spec = theoretical_spectrum(p, peptide_type, aa_mass_table)
                score = score_spectrums(theo_spec, exp_spec)
                <span class="hljs-keyword">if</span> score &gt; final_score:
                    final_peptides = [p]
                    final_score = score_spectrums(
                        theoretical_spectrum(final_peptides[<span class="hljs-number">0</span>], peptide_type, aa_mass_table),
                        exp_spec
                    )
                <span class="hljs-keyword">elif</span> score == final_score:
                    final_peptides.append(p)
                <span class="hljs-comment"># p should be removed at this point (the line below should be uncommented). Not removing it means that</span>
                <span class="hljs-comment"># it may end up in the leaderboard for the next cycle. If that happens, it'll get branched out into new</span>
                <span class="hljs-comment"># candidate peptides where each has an amino acids appended.</span>
                <span class="hljs-comment">#</span>
                <span class="hljs-comment"># The problem with branching p out further is that p's mass already matches the expected peptide mass.</span>
                <span class="hljs-comment"># Once p gets branched out, those branched out candidate peptides will have masses that EXCEED the</span>
                <span class="hljs-comment"># expected peptide mass, meaning they'll all get removed anyway. This would be fine, except that by</span>
                <span class="hljs-comment"># moving p into the leaderboard for the next cycle you're potentially preventing other viable</span>
                <span class="hljs-comment"># candidate peptides from making it in.</span>
                <span class="hljs-comment">#</span>
                <span class="hljs-comment"># So why isn't p being removed here (why was the line below commented out)? The questions on Stepik</span>
                <span class="hljs-comment"># expect no removal at this point. Uncommenting it will cause more peptides than are expected to show up</span>
                <span class="hljs-comment"># for some questions, meaning the answer will be rejected by Stepik.</span>
                <span class="hljs-comment">#</span>
                <span class="hljs-comment"># removal_idxes.add(i)</span>
            <span class="hljs-keyword">elif</span> p_mass &gt; peptide_mass:
                <span class="hljs-comment"># The peptide's mass exceeds the expected mass, meaning that there's no chance that this peptide can be</span>
                <span class="hljs-comment"># a match for exp_spec. Discard it.</span>
                removal_idxes.add(i)
        expanded_leaderboard = [p <span class="hljs-keyword">for</span> i, p <span class="hljs-keyword">in</span> <span class="hljs-built_in">enumerate</span>(expanded_leaderboard) <span class="hljs-keyword">if</span> i <span class="hljs-keyword">not</span> <span class="hljs-keyword">in</span> removal_idxes]
        <span class="hljs-comment"># Set leaderboard to the top n scoring peptides from expanded_leaderboard, but include peptides past n as long</span>
        <span class="hljs-comment"># as those peptides have a score equal to the nth peptide. The reason for this is that because they score the</span>
        <span class="hljs-comment"># same, there's just as much of a chance that they'll end up as a winner as it is that the nth peptide will.</span>
            <span class="hljs-comment"># <span class="hljs-doctag">NOTE:</span> Why get the theo spec of the linear version even if the peptide is cyclic? For similar reasons as to</span>
            <span class="hljs-comment"># why it's done in the branch-and-bound variant: If we treat candidate peptides as cyclic, their theo spec</span>
            <span class="hljs-comment"># will include masses for wrapping subpeptides of the candidate peptide. These wrapping subpeptide masses</span>
            <span class="hljs-comment"># may end up inadvertently matching masses in the experimental spectrum, meaning that the candidate may get</span>
            <span class="hljs-comment"># a better score than it should, potentially pushing it forward over other candidates that would ultimately</span>
            <span class="hljs-comment"># branch out  to a more optimal final solution. As such, even if the exp  spec is  for a cyclic peptide,</span>
            <span class="hljs-comment"># treat the candidates as linear segments of that cyclic peptide (essentially linear  peptides). If you're</span>
            <span class="hljs-comment"># confused go see the comment in the branch-and-bound variant.</span>
        theo_specs = [theoretical_spectrum(p, PeptideType.LINEAR, aa_mass_table) <span class="hljs-keyword">for</span> p <span class="hljs-keyword">in</span> expanded_leaderboard]
        scores = [score_spectrums(theo_spec, exp_spec) <span class="hljs-keyword">for</span> theo_spec <span class="hljs-keyword">in</span> theo_specs]
        scores_paired = <span class="hljs-built_in">sorted</span>(<span class="hljs-built_in">zip</span>(expanded_leaderboard, scores), key=<span class="hljs-keyword">lambda</span> x: x[<span class="hljs-number">1</span>], reverse=<span class="hljs-literal">True</span>)
        leaderboard_trim_to_size = <span class="hljs-built_in">len</span>(expanded_leaderboard)
        <span class="hljs-keyword">for</span> j <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(leaderboard_size + <span class="hljs-number">1</span>, <span class="hljs-built_in">len</span>(scores_paired)):
            <span class="hljs-keyword">if</span> scores_paired[leaderboard_size][<span class="hljs-number">1</span>] &gt; scores_paired[j][<span class="hljs-number">1</span>]:
                leaderboard_trim_to_size = j - <span class="hljs-number">1</span>
                <span class="hljs-keyword">break</span>
        leaderboard = [p <span class="hljs-keyword">for</span> p, _ <span class="hljs-keyword">in</span> scores_paired[:leaderboard_trim_to_size]]
    <span class="hljs-keyword">return</span> final_peptides
</code></pre>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> in the example below is for the <a href="#BM_(peptide)%2Fi">peptide</a> NQYQ, which has the <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a> [0, 114, 128, 128, 163, 242, 242, 291, 291, 370, 405, 405, 419, 533].</p>
</div>
<div style="border:1px solid black;">
<p></p>
<p>The cyclic peptides matching the experimental spectrum [0.0, 114.0, 163.0, 242.0, 291.0, 370.0, 405.0, 419.0, 480.0, 533.0] are with leaderboard size of 10...</p>
<ul>
<li>NKYK</li>
<li>NKYQ</li>
<li>NQYK</li>
<li>NQYQ</li>
<li>YKNK</li>
<li>YKNQ</li>
<li>YQNK</li>
<li>YQNQ</li>
<li>NYKK</li>
<li>NYKQ</li>
<li>NYQK</li>
<li>NYQQ</li>
<li>YNKK</li>
<li>YNKQ</li>
<li>YNQK</li>
<li>YNQQ</li>
<li>KNYK</li>
<li>KNYQ</li>
<li>KYNK</li>
<li>KYNQ</li>
<li>QNYK</li>
<li>QNYQ</li>
</ul>
</div>
<p></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The following section isn't from the Pevzner book or any online resources. I came up with it in an effort to solve the final assignment for Chapter 4 (the chapter on <a href="#BM_(non-ribosomal%20peptide)%2Fi">non-ribosomal peptide</a> <a href="#BM_(sequencing%7Csequenced)%2Fi">sequencing</a>). As such, it might not be entirely correct / there may be better ways to do this.</p>
</div>
<p>For real <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>s, the algorithm is very similar to the real <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> version of the <a href="#BM_(branch-and-bound%20algorithm%7Cbranch%20and%20bound%20algorithm)%2Fi">branch-and-bound algorithm</a>. The only difference is the bounding heuristic: At each branch, rather than moving forward candidate <a href="#BM_(peptide)%2Fi">peptide</a>s that meet a certain score threshold, move forward the best n scoring candidate <a href="#BM_(peptide)%2Fi">peptide</a>s. These best scoring <a href="#BM_(peptide)%2Fi">peptide</a>s are referred to as the leaderboard.</p>
<p><a href="ch4_code/src/SequencePeptide_Leaderboard.py">ch4_code/src/SequencePeptide_Leaderboard.py</a> (lines 14 to 79):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">sequence_peptide</span>(<span class="hljs-params">
        exp_spec: <span class="hljs-type">List</span>[<span class="hljs-built_in">float</span>],                               <span class="hljs-comment"># must be sorted asc</span>
        aa_mass_table: <span class="hljs-type">Dict</span>[AA, <span class="hljs-built_in">float</span>],                      <span class="hljs-comment"># amino acid mass table</span>
        aa_mass_tolerance: <span class="hljs-built_in">float</span>,                            <span class="hljs-comment"># amino acid mass tolerance</span>
        peptide_mass_candidates: <span class="hljs-type">List</span>[<span class="hljs-type">Tuple</span>[<span class="hljs-built_in">float</span>, <span class="hljs-built_in">float</span>]],  <span class="hljs-comment"># mass range candidates for mass of peptide</span>
        peptide_type: PeptideType,                           <span class="hljs-comment"># linear or cyclic</span>
        score_backlog: <span class="hljs-built_in">int</span>,                                  <span class="hljs-comment"># backlog of top scores</span>
        leaderboard_size: <span class="hljs-built_in">int</span>,
        leaderboard_initial: <span class="hljs-type">List</span>[<span class="hljs-type">List</span>[AA]] = <span class="hljs-literal">None</span>           <span class="hljs-comment"># bootstrap candidate peptides for leaderboard</span>
</span>) -&gt; SequenceTesterSet:
    tester_set = SequenceTesterSet(
        exp_spec,
        aa_mass_table,
        aa_mass_tolerance,
        peptide_mass_candidates,
        peptide_type,
        score_backlog
    )
    <span class="hljs-keyword">if</span> leaderboard_initial <span class="hljs-keyword">is</span> <span class="hljs-literal">None</span>:
        leaderboard = [[]]
    <span class="hljs-keyword">else</span>:
        leaderboard = leaderboard_initial[:]
    <span class="hljs-keyword">while</span> <span class="hljs-built_in">len</span>(leaderboard) &gt; <span class="hljs-number">0</span>:
        <span class="hljs-comment"># Branch candidates</span>
        expanded_leaderboard = []
        <span class="hljs-keyword">for</span> p <span class="hljs-keyword">in</span> leaderboard:
            <span class="hljs-keyword">for</span> m <span class="hljs-keyword">in</span> aa_mass_table:
                new_p = p[:] + [m]
                expanded_leaderboard.append(new_p)
        <span class="hljs-comment"># Test candidates to see if they match exp_spec or if they should keep being branched</span>
        removal_idxes = <span class="hljs-built_in">set</span>()
        <span class="hljs-keyword">for</span> i, p <span class="hljs-keyword">in</span> <span class="hljs-built_in">enumerate</span>(expanded_leaderboard):
            res = <span class="hljs-built_in">set</span>(tester_set.test(p))
            <span class="hljs-keyword">if</span> {TestResult.MASS_TOO_LARGE} == res:
                removal_idxes.add(i)
        expanded_leaderboard = [p <span class="hljs-keyword">for</span> i, p <span class="hljs-keyword">in</span> <span class="hljs-built_in">enumerate</span>(expanded_leaderboard) <span class="hljs-keyword">if</span> i <span class="hljs-keyword">not</span> <span class="hljs-keyword">in</span> removal_idxes]
        <span class="hljs-comment"># Set leaderboard to the top n scoring peptides from expanded_leaderboard, but include peptides past n as long</span>
        <span class="hljs-comment"># as those peptides have a score equal to the nth peptide. The reason for this is that because they score the</span>
        <span class="hljs-comment"># same, there's just as much of a chance that they'll end up as the winner as it is that the nth peptide will.</span>
            <span class="hljs-comment"># <span class="hljs-doctag">NOTE:</span> Why get the theo spec of the linear version even if the peptide is cyclic? For similar reasons as to</span>
            <span class="hljs-comment"># why it's done in the branch-and-bound variant: If we treat candidate peptides as cyclic, their theo spec</span>
            <span class="hljs-comment"># will include masses for wrapping subpeptides of the candidate peptide. These wrapping subpeptide masses</span>
            <span class="hljs-comment"># may end up inadvertently matching masses in the experimental spectrum, meaning that the candidate may get</span>
            <span class="hljs-comment"># a better score than it should, potentially pushing it forward over other candidates that would ultimately</span>
            <span class="hljs-comment"># branch out  to a more optimal final solution. As such, even if the exp  spec is  for a cyclic peptide,</span>
            <span class="hljs-comment"># treat the candidates as linear segments of that cyclic peptide (essentially linear  peptides).</span>
        theo_specs = [
            SequenceTester.generate_theroetical_spectrum_with_tolerances(
                p,
                peptide_type,
                aa_mass_table,
                aa_mass_tolerance
            )
            <span class="hljs-keyword">for</span> p <span class="hljs-keyword">in</span> expanded_leaderboard
        ]
        scores = [score_spectrums(exp_spec, theo_spec) <span class="hljs-keyword">for</span> theo_spec <span class="hljs-keyword">in</span> theo_specs]
        scores_paired = <span class="hljs-built_in">sorted</span>(<span class="hljs-built_in">zip</span>(expanded_leaderboard, scores), key=<span class="hljs-keyword">lambda</span> x: x[<span class="hljs-number">1</span>], reverse=<span class="hljs-literal">True</span>)
        leaderboard_tail_idx = <span class="hljs-built_in">min</span>(leaderboard_size, <span class="hljs-built_in">len</span>(scores_paired)) - <span class="hljs-number">1</span>
        leaderboard_tail_score = <span class="hljs-number">0</span> <span class="hljs-keyword">if</span> leaderboard_tail_idx == -<span class="hljs-number">1</span> <span class="hljs-keyword">else</span> scores_paired[leaderboard_tail_idx][<span class="hljs-number">1</span>]
        <span class="hljs-keyword">for</span> j <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(leaderboard_tail_idx + <span class="hljs-number">1</span>, <span class="hljs-built_in">len</span>(scores_paired)):
            <span class="hljs-keyword">if</span> scores_paired[j][<span class="hljs-number">1</span>] &lt; leaderboard_tail_score:
                leaderboard_tail_idx = j - <span class="hljs-number">1</span>
                <span class="hljs-keyword">break</span>
        leaderboard = [p <span class="hljs-keyword">for</span> p, _ <span class="hljs-keyword">in</span> scores_paired[:leaderboard_tail_idx + <span class="hljs-number">1</span>]]
    <span class="hljs-keyword">return</span> tester_set
</code></pre>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> in the example below is for the <a href="#BM_(peptide)%2Fi">peptide</a> 114-128-163, which has the <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a> [0, 114, 128, 163, 242, 291, 405].</p>
</div>
<div style="border:1px solid black;">
<p></p>
<p>Given the ...</p>
<ul>
<li>experimental spectrum: [0.0, 112.5, 127.1, 242.9, 290.0, 404.0]</li>
<li>experimental spectrum mass noise: ±1.0</li>
<li>assumed peptide type: linear</li>
<li>assumed peptide length: 3</li>
<li>assumed peptide mass: any of the last 1 experimental spectrum masses</li>
<li>score backlog: 0</li>
<li>leaderboard size: 100</li>
</ul>
<p>Top 10 captured mino acid masses (rounded to 1): [114.0, 112.5, 115.8, 161.1, 162.9, 127.1, 130.4, 177.5]</p>
<p>For peptides between 397.0 and 411.0...</p>
<ul>
<li>Score 6: 114.0-127.1-162.9</li>
<li>Score 6: 162.9-127.1-114.0</li>
</ul>
</div>
<p></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>This was the version of the algorithm used to solve chapter 4's final assignment (<a href="#BM_(sequence)%2Fi">sequence</a> a real <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> for some unknown variant of Tyrocidine). Note how the parameters into <a href="#BM_(sequence)%2Fi">sequence</a>_<a href="#BM_(peptide)%2Fi">peptide</a> take an initial leaderboard. This initial leaderboard was primed with <a href="#BM_(subpeptide)%2Fi">subpeptide</a> <a href="#BM_(sequence)%2Fi">sequence</a>s from other Tyrocidine variants discusses in chapter 4. The problem wasn't solvable without these <a href="#BM_(subpeptide)%2Fi">subpeptide</a> <a href="#BM_(sequence)%2Fi">sequence</a>s. More information on this can be found in the Python file for the final assignment.</p>
<p>Before coming up with the above solution, I came up with another heuristic that I tried: Use basic genetic algorithms / evolutionary algorithms as the heuristic to move forward <a href="#BM_(peptide)%2Fi">peptide</a>s. This performed even worse than leaderboard: If the mutation rate is too low, the candidates converge to a local optima and can't break out. If the mutation rate is too high, the candidates never converge to a solution. As such, it was removed from the code.</p>
</div>
<a name="H_Sequence%20Alignment"></a><h2>Sequence Alignment</h2>
<p><a name="BM_(Algorithms%5C%2FSequence%20Alignment)_TOPIC%2F"></a><strong></strong></p>
<p>Many core biology constructs are represented as <a href="#BM_(sequence)%2Fi">sequence</a>s. For example, ...</p>
<ul>
<li>DNA strands are represented as a <a href="#BM_(sequence)%2Fi">sequence</a> (chained nucleotides),</li>
<li>proteins are represented as a <a href="#BM_(sequence)%2Fi">sequence</a> (chained <a href="#BM_(amino%20acid)%2Fi">amino acid</a>s),</li>
<li>etc..</li>
</ul>
<p>Performing a <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a> on a set of <a href="#BM_(sequence)%2Fi">sequence</a>s means to match up the elements of those <a href="#BM_(sequence)%2Fi">sequence</a>s against each other using a set of basic operations:</p>
<ul>
<li>insert/delete (also referred to as <a href="#BM_(indel)%2Fi">indel</a>).</li>
<li>replace (also referred to as mismatch).</li>
<li>keep matching (also referred to as match).</li>
</ul>
<p>There are many ways that a set of <a href="#BM_(sequence)%2Fi">sequence</a>s can be aligned. For example, the <a href="#BM_(sequence)%2Fi">sequence</a>s MAPLE and TABLE may be aligned by performing...</p>
<table>
<thead>
<tr>
<th>String 1</th>
<th>String 2</th>
<th>Operation</th>
</tr>
</thead>
<tbody>
<tr>
<td>M</td>
<td></td>
<td>Insert/delete</td>
</tr>
<tr>
<td></td>
<td>T</td>
<td>Insert/delete</td>
</tr>
<tr>
<td>A</td>
<td>A</td>
<td>Keep matching</td>
</tr>
<tr>
<td>P</td>
<td>B</td>
<td>Replace</td>
</tr>
<tr>
<td>L</td>
<td>L</td>
<td>Keep matching</td>
</tr>
<tr>
<td>E</td>
<td>E</td>
<td>Keep matching</td>
</tr>
</tbody>
</table>
<p>Or, MAPLE and TABLE may be aligned by performing...</p>
<table>
<thead>
<tr>
<th>String 1</th>
<th>String 2</th>
<th>Operation</th>
</tr>
</thead>
<tbody>
<tr>
<td>M</td>
<td>T</td>
<td>Replace</td>
</tr>
<tr>
<td>A</td>
<td>A</td>
<td>Keep matching</td>
</tr>
<tr>
<td>P</td>
<td>B</td>
<td>Replace</td>
</tr>
<tr>
<td>L</td>
<td>L</td>
<td>Keep matching</td>
</tr>
<tr>
<td>E</td>
<td>E</td>
<td>Keep matching</td>
</tr>
</tbody>
</table>
<p>Typically the highest scoring <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a> is the one that's chosen, where the score is some custom function that best represents the type of <a href="#BM_(sequence)%2Fi">sequence</a> being worked with (e.g. proteins are scored differently than DNA). In the example above, if replacements are scored better than <a href="#BM_(indel)%2Fi">indel</a>s, the latter alignment would be the highest scoring. <a href="#BM_(sequence)%2Fi">Sequence</a>s that strongly align are thought of as being related / similar (e.g. proteins that came from the same parent but diverged to 2 separate evolutionary paths).</p>
<p>The names of these operations make more sense if you were to think of alignment instead as <strong>transformation</strong>. The example above's first alignment in the context of <strong>transforming</strong> MAPLE to TABLE may be thought of as:</p>
<table>
<thead>
<tr>
<th>From</th>
<th>To</th>
<th>Operation</th>
<th>Result</th>
</tr>
</thead>
<tbody>
<tr>
<td>M</td>
<td></td>
<td>Delete M</td>
<td></td>
</tr>
<tr>
<td></td>
<td>T</td>
<td>Insert T</td>
<td>T</td>
</tr>
<tr>
<td>A</td>
<td>A</td>
<td>Keep matching A</td>
<td>TA</td>
</tr>
<tr>
<td>P</td>
<td>B</td>
<td>Replace P to B</td>
<td>TAB</td>
</tr>
<tr>
<td>L</td>
<td>L</td>
<td>Keep matching L</td>
<td>TABL</td>
</tr>
<tr>
<td>E</td>
<td>E</td>
<td>Keep matching E</td>
<td>TABLE</td>
</tr>
</tbody>
</table>
<p>The shorthand form of representing <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a>s is to stack each <a href="#BM_(sequence)%2Fi">sequence</a>. The example above may be written as...</p>
<table>
<thead>
<tr>
<th></th>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
</tr>
</thead>
<tbody>
<tr>
<td>String 1</td>
<td>M</td>
<td></td>
<td>A</td>
<td>P</td>
<td>L</td>
<td>E</td>
</tr>
<tr>
<td>String 2</td>
<td></td>
<td>T</td>
<td>A</td>
<td>B</td>
<td>L</td>
<td>E</td>
</tr>
</tbody>
</table>
<p>Typically, all possible <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a>s are represented using an <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a>: a graph that represents all possible alignments for a set of <a href="#BM_(sequence)%2Fi">sequence</a>s. A path through an <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a> from source node to sink node is called an <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a>: a path that represents one specific way the set of <a href="#BM_(sequence)%2Fi">sequence</a>s may be aligned. For example, the <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a> and <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a>s for the alignments above (MAPLE vs TABLE) ...</p>
<p><img src="svgbob_2a74202f0a773a8996e2e61a1f7290c867847b9e.svg" alt="Kroki diagram output"></p>
<p>The example above is just one of many <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a> types. There are different types of <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a>s, applications of <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a>s, and different scoring models used in bioinformatics.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The Pevzner book mentions a non-biology related problem to help illustrate <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a>s: the <a href="#BM_(manhattan%20tourist%20problem)%2Fi">Manhattan Tourist problem</a>. Look it up if you're confused.</p>
</div>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The Pevzner book, in a later chapter (ch7 -- <a href="#BM_(phylogeny%7Cphylogenetic)%2Fi">phylogeny</a>), spends an entire section talking about <a href="#BM_(character%20table%7Ccharacter%20vector)%2Fi">character table</a>s and how they can be thought of as <a href="#BM_(sequence)%2Fi">sequence</a>s (<a href="#BM_(character%20table%7Ccharacter%20vector)%2Fi">character vector</a>s). There's no good place to put this information. It seems non-critical so the only place it exists is in the terminology section.</p>
</div>
<a name="H_Find%20Maximum%20Path"></a><h3>Find Maximum Path</h3>
<p><a name="BM_(Algorithms%5C%2FSequence%20Alignment%2FFind%20Maximum%20Path)_TOPIC%2F"></a><strong></strong></p>
<p><strong>WHAT</strong>: Given an arbitrary <a href="#BM_(directed%20acyclic%20graph)%2Fi">directed acyclic graph</a> where each edge has a weight, find the path with the maximum weight between two nodes.</p>
<p><strong>WHY</strong>: Finding a maximum path between nodes is fundamental to <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a>s. That is, regardless of what type of <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a> is being performed, at its core it boils down to finding the maximum weight between two nodes in an <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a>.</p>
<a name="H4_Bruteforce%20Algorithm"></a><h4>Bruteforce Algorithm</h4>
<p><a name="BM_(Algorithms%5C%2FSequence%20Alignment%2FFind%20Maximum%20Path%2FBruteforce%20Algorithm)_TOPIC%2F"></a><strong></strong></p>
<p><strong>ALGORITHM</strong>:</p>
<p>This algorithm finds a maximum path using recursion. To calculate the maximum path between two nodes, iterate over each of the source node's children and calculate <code>edge_weight + max_path(child, destination).weight</code>. The iteration with the highest value is the one with the maximum path to the destination node.</p>
<p>This is too slow to be used on anything but small DAGs.</p>
<p><a href="ch5_code/src/find_max_path/FindMaxPath_Bruteforce.py">ch5_code/src/find_max_path/FindMaxPath_Bruteforce.py</a> (lines 21 to 50):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">find_max_path</span>(<span class="hljs-params">
        graph: Graph[N, ND, E, ED],
        current_node: N,
        end_node: N,
        get_edge_weight_func: GET_EDGE_WEIGHT_FUNC_TYPE
</span>) -&gt; <span class="hljs-type">Optional</span>[<span class="hljs-type">Tuple</span>[<span class="hljs-type">List</span>[E], <span class="hljs-built_in">float</span>]]:
    <span class="hljs-keyword">if</span> current_node == end_node:
        <span class="hljs-keyword">return</span> [], <span class="hljs-number">0.0</span>
    alternatives = []
    <span class="hljs-keyword">for</span> edge_id <span class="hljs-keyword">in</span> graph.get_outputs(current_node):
        edge_weight = get_edge_weight_func(edge_id)
        child_n = graph.get_edge_to(edge_id)
        res = find_max_path(
            graph,
            child_n,
            end_node,
            get_edge_weight_func
        )
        <span class="hljs-keyword">if</span> res <span class="hljs-keyword">is</span> <span class="hljs-literal">None</span>:
            <span class="hljs-keyword">continue</span>
        path, weight = res
        path = [edge_id] + path
        weight = edge_weight + weight
        res = path, weight
        alternatives.append(res)
    <span class="hljs-keyword">if</span> <span class="hljs-built_in">len</span>(alternatives) == <span class="hljs-number">0</span>:
        <span class="hljs-keyword">return</span> <span class="hljs-literal">None</span>  <span class="hljs-comment"># no path to end, so return None</span>
    <span class="hljs-keyword">else</span>:
        <span class="hljs-keyword">return</span> <span class="hljs-built_in">max</span>(alternatives, key=<span class="hljs-keyword">lambda</span> x: x[<span class="hljs-number">1</span>])  <span class="hljs-comment"># choose path to end with max weight</span>
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the following graph...</p>
<p><img src="dot_000bacb11a1651241e16e505830ab31e.svg" alt="Dot diagram"></p>
<p>... the path with the max weight between A and E ...</p>
<ul>
<li>Maximum path = A -&gt; B -&gt; C -&gt; E</li>
<li>Maximum weight = 3.0</li>
</ul>
</div>
<p></p>
<a name="H_Cache%20Algorithm"></a><h4>Cache Algorithm</h4>
<p><a name="BM_(Algorithms%5C%2FSequence%20Alignment%2FFind%20Maximum%20Path%2FCache%20Algorithm)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FFind%20Maximum%20Path%2FBruteforce%20Algorithm)_TOPIC%2F">Algorithms/Sequence Alignment/Find Maximum Path/Bruteforce Algorithm</a></li>
</ul>
</div>
<p><strong>ALGORITHM</strong>:</p>
<p>This algorithm extends the bruteforce algorithm using <a href="#BM_(dynamic%20programming%20algorithm%7Cdynamic%20programming)%2Fi">dynamic programming</a>: A technique that breaks down a problem into recursive sub-problems, where the result of each sub-problem is stored in some lookup table (cache) such that it can be re-used if that sub-problem were ever encountered again. The bruteforce algorithm already breaks down into recursive sub-problems. As such, the only change here is that the result of each sub-problem computation is cached such that it can be re-used if it were ever encountered again.</p>
<p><a href="ch5_code/src/find_max_path/FindMaxPath_DPCache.py">ch5_code/src/find_max_path/FindMaxPath_DPCache.py</a> (lines 21 to 56):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">find_max_path</span>(<span class="hljs-params">
        graph: Graph[N, ND, E, ED],
        current_node: N,
        end_node: N,
        cache: <span class="hljs-type">Dict</span>[N, <span class="hljs-type">Optional</span>[<span class="hljs-type">Tuple</span>[<span class="hljs-type">List</span>[E], <span class="hljs-built_in">float</span>]]],
        get_edge_weight_func: GET_EDGE_WEIGHT_FUNC_TYPE
</span>) -&gt; <span class="hljs-type">Optional</span>[<span class="hljs-type">Tuple</span>[<span class="hljs-type">List</span>[E], <span class="hljs-built_in">float</span>]]:
    <span class="hljs-keyword">if</span> current_node == end_node:
        <span class="hljs-keyword">return</span> [], <span class="hljs-number">0.0</span>
    alternatives = []
    <span class="hljs-keyword">for</span> edge_id <span class="hljs-keyword">in</span> graph.get_outputs(current_node):
        edge_weight = get_edge_weight_func(edge_id)
        child_n = graph.get_edge_to(edge_id)
        <span class="hljs-keyword">if</span> child_n <span class="hljs-keyword">in</span> cache:
            res = cache[child_n]
        <span class="hljs-keyword">else</span>:
            res = find_max_path(
                graph,
                child_n,
                end_node,
                cache,
                get_edge_weight_func
            )
            cache[child_n] = res
        <span class="hljs-keyword">if</span> res <span class="hljs-keyword">is</span> <span class="hljs-literal">None</span>:
            <span class="hljs-keyword">continue</span>
        path, weight = res
        path = [edge_id] + path
        weight = edge_weight + weight
        res = path, weight
        alternatives.append(res)
    <span class="hljs-keyword">if</span> <span class="hljs-built_in">len</span>(alternatives) == <span class="hljs-number">0</span>:
        <span class="hljs-keyword">return</span> <span class="hljs-literal">None</span>  <span class="hljs-comment"># no path to end, so return None</span>
    <span class="hljs-keyword">else</span>:
        <span class="hljs-keyword">return</span> <span class="hljs-built_in">max</span>(alternatives, key=<span class="hljs-keyword">lambda</span> x: x[<span class="hljs-number">1</span>])  <span class="hljs-comment"># choose path to end with max weight</span>
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the following graph...</p>
<p><img src="dot_315426b869e6c4fb63ef68e48f851ab2.svg" alt="Dot diagram"></p>
<p>... the path with the max weight between A and E ...</p>
<ul>
<li>Maximum path = A -&gt; B -&gt; C -&gt; E</li>
<li>Maximum weight = 3.0</li>
</ul>
</div>
<p></p>
<a name="H_Backtrack%20Algorithm"></a><h4>Backtrack Algorithm</h4>
<p><a name="BM_(Algorithms%5C%2FSequence%20Alignment%2FFind%20Maximum%20Path%2FBacktrack%20Algorithm)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FFind%20Maximum%20Path%2FCache%20Algorithm)_TOPIC%2F">Algorithms/Sequence Alignment/Find Maximum Path/Cache Algorithm</a></li>
</ul>
</div>
<p><strong>ALGORITHM</strong>:</p>
<p>This algorithm is a better but less obvious <a href="#BM_(dynamic%20programming%20algorithm%7Cdynamic%20programming)%2Fi">dynamic programming</a> approach. The previous <a href="#BM_(dynamic%20programming%20algorithm%7Cdynamic%20programming)%2Fi">dynamic programming algorithm</a> builds a cache containing the maximum path from each node encountered to the destination node. This <a href="#BM_(dynamic%20programming%20algorithm%7Cdynamic%20programming)%2Fi">dynamic programming algorithm</a> instead builds out a smaller cache from the source node fanning out one step at a time.</p>
<p>In this less obvious algorithm, there are edge weights just as before but each node also has a weight and a selected incoming edge. The <a href="#BM_%5Cb(DAG)%5Cb%2F">DAG</a> starts off with all node weights and incoming edge selections being unset. The source node has its weight set to 0. Then, for any node where all of its parents have a weight set, select the incoming edge where <code>parent_weight + edge_weight</code> is the highest. That highest <code>parent_weight + edge_weight</code> becomes the weight of that node and the edge responsible for it becomes the selected incoming edge (backtracking edge).</p>
<p>Repeat until all nodes have a weight and backtracking edge set.</p>
<p>For example, imagine the following <a href="#BM_%5Cb(DAG)%5Cb%2F">DAG</a>...</p>
<p><img src="svgbob_17a86f29e93ea0ac10d87eff5b54f515afa543aa.svg" alt="Kroki diagram output"></p>
<p>Set source nodes to have a weight of 0...</p>
<p><img src="svgbob_b46f186f6a89af1bdf29d6d58eb4eb86228b5a41.svg" alt="Kroki diagram output"></p>
<p>Then, iteratively set the weights and backtracking edges...</p>
<p><img src="svgbob_b50016ce027bdf6cec626ccea86f9351f4d043d4.svg" alt="Kroki diagram output"></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>This process is walking the <a href="#BM_%5Cb(DAG)%5Cb%2F">DAG</a> in <a href="#BM_(topological%20order)%2Fi">topological order</a>.</p>
</div>
<p>To find the path with the maximum weight, simply walk backward using the backtracking edges from the destination node to the source node. For example, in the <a href="#BM_%5Cb(DAG)%5Cb%2F">DAG</a> above the maximum path that ends at E can be determined by following the backtracking edges from E until A is reached...</p>
<ul>
<li>E came from C</li>
<li>C came from A</li>
</ul>
<p>The maximum path from A to E is A -&gt; C -&gt; E and the weight of that path is 4 (the weight of E).</p>
<p>This variant of the <a href="#BM_(dynamic%20programming%20algorithm%7Cdynamic%20programming)%2Fi">dynamic programming algorithm</a> uses less memory than the first. For each node encountered, ...</p>
<ul>
<li>the first variant caches a path to the destination.</li>
<li>this variant only caches a weight and backtracking edge.</li>
</ul>
<p><a href="ch5_code/src/find_max_path/FindMaxPath_DPBacktrack.py">ch5_code/src/find_max_path/FindMaxPath_DPBacktrack.py</a> (lines 41 to 143):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">populate_weights_and_backtrack_pointers</span>(<span class="hljs-params">
        g: Graph[N, ND, E, ED],
        from_node: N,
        set_node_data_func: SET_NODE_DATA_FUNC_TYPE,
        get_node_data_func: GET_NODE_DATA_FUNC_TYPE,
        get_edge_weight_func: GET_EDGE_WEIGHT_FUNC_TYPE
</span>):
    processed_nodes = <span class="hljs-built_in">set</span>()          <span class="hljs-comment"># nodes where all parents have been processed AND it has been processed</span>
    waiting_nodes = <span class="hljs-built_in">set</span>()            <span class="hljs-comment"># nodes where all parents have been processed BUT it has yet to be processed</span>
    unprocessable_nodes = Counter()  <span class="hljs-comment"># nodes that have some parents remaining to be processed (value=# of parents left)</span>
    <span class="hljs-comment"># For all root nodes, add to processed_nodes and set None weight and None backtracking edge.</span>
    <span class="hljs-keyword">for</span> node <span class="hljs-keyword">in</span> g.get_nodes():
        <span class="hljs-keyword">if</span> g.get_in_degree(node) == <span class="hljs-number">0</span>:
            set_node_data_func(node, <span class="hljs-literal">None</span>, <span class="hljs-literal">None</span>)
            processed_nodes |= {node}
    <span class="hljs-comment"># For all root nodes, add any children where its the only parent to waiting_nodes.</span>
    <span class="hljs-keyword">for</span> node <span class="hljs-keyword">in</span> processed_nodes:
        <span class="hljs-keyword">for</span> e <span class="hljs-keyword">in</span> g.get_outputs(node):
            dst_node = g.get_edge_to(e)
            <span class="hljs-keyword">if</span> {g.get_edge_from(e) <span class="hljs-keyword">for</span> e <span class="hljs-keyword">in</span> g.get_inputs(dst_node)}.issubset(processed_nodes):
                waiting_nodes |= {dst_node}
    <span class="hljs-comment"># Make sure from_node is a root and set its weight to 0.</span>
    <span class="hljs-keyword">assert</span> from_node <span class="hljs-keyword">in</span> processed_nodes
    set_node_data_func(from_node, <span class="hljs-number">0.0</span>, <span class="hljs-literal">None</span>)
    <span class="hljs-comment"># Track how many remaining parents each node in the graph has. Note that the graph's root nodes were already marked</span>
    <span class="hljs-comment"># as processed above.</span>
    <span class="hljs-keyword">for</span> node <span class="hljs-keyword">in</span> g.get_nodes():
        incoming_nodes = {g.get_edge_from(e) <span class="hljs-keyword">for</span> e <span class="hljs-keyword">in</span> g.get_inputs(node)}
        incoming_nodes -= processed_nodes
        unprocessable_nodes[node] = <span class="hljs-built_in">len</span>(incoming_nodes)
    <span class="hljs-comment"># Any nodes in waiting_nodes have had all their parents already processed (in processed_nodes). As such, they can</span>
    <span class="hljs-comment"># have their weights and backtracking pointers calculated. They can then be placed into processed_nodes themselves.</span>
    <span class="hljs-keyword">while</span> <span class="hljs-built_in">len</span>(waiting_nodes) &gt; <span class="hljs-number">0</span>:
        node = <span class="hljs-built_in">next</span>(<span class="hljs-built_in">iter</span>(waiting_nodes))
        incoming_nodes = {g.get_edge_from(e) <span class="hljs-keyword">for</span> e <span class="hljs-keyword">in</span> g.get_inputs(node)}
        <span class="hljs-keyword">if</span> <span class="hljs-keyword">not</span> incoming_nodes.issubset(processed_nodes):
            <span class="hljs-keyword">continue</span>
        incoming_accum_weights = {}
        <span class="hljs-keyword">for</span> edge <span class="hljs-keyword">in</span> g.get_inputs(node):
            src_node = g.get_edge_from(edge)
            src_node_weight, _ = get_node_data_func(src_node)
            edge_weight = get_edge_weight_func(edge)
            <span class="hljs-comment"># Roots that aren't from_node were initialized to a weight of None -- if you see them, skip them.</span>
            <span class="hljs-keyword">if</span> src_node_weight <span class="hljs-keyword">is</span> <span class="hljs-keyword">not</span> <span class="hljs-literal">None</span>:
                incoming_accum_weights[edge] = src_node_weight + edge_weight
        <span class="hljs-keyword">if</span> <span class="hljs-built_in">len</span>(incoming_accum_weights) == <span class="hljs-number">0</span>:
            max_edge = <span class="hljs-literal">None</span>
            max_weight = <span class="hljs-literal">None</span>
        <span class="hljs-keyword">else</span>:
            max_edge = <span class="hljs-built_in">max</span>(incoming_accum_weights, key=<span class="hljs-keyword">lambda</span> e: incoming_accum_weights[e])
            max_weight = incoming_accum_weights[max_edge]
        set_node_data_func(node, max_weight, max_edge)
        <span class="hljs-comment"># This node has been processed, move it over to processed_nodes.</span>
        waiting_nodes.remove(node)
        processed_nodes.add(node)
        <span class="hljs-comment"># For outgoing nodes this node points to, if that outgoing node has all of its dependencies in processed_nodes,</span>
        <span class="hljs-comment"># then add it to waiting_nodes (so it can be processed).</span>
        outgoing_nodes = {g.get_edge_to(e) <span class="hljs-keyword">for</span> e <span class="hljs-keyword">in</span> g.get_outputs(node)}
        <span class="hljs-keyword">for</span> output_node <span class="hljs-keyword">in</span> outgoing_nodes:
            unprocessable_nodes[output_node] -= <span class="hljs-number">1</span>
            <span class="hljs-keyword">if</span> unprocessable_nodes[output_node] == <span class="hljs-number">0</span>:
                waiting_nodes.add(output_node)


<span class="hljs-keyword">def</span> <span class="hljs-title function_">backtrack</span>(<span class="hljs-params">
        g: Graph[N, ND, E, ED],
        end_node: N,
        get_node_data_func: GET_NODE_DATA_FUNC_TYPE
</span>) -&gt; <span class="hljs-type">List</span>[E]:
    next_node = end_node
    reverse_path = []
    <span class="hljs-keyword">while</span> <span class="hljs-literal">True</span>:
        node = next_node
        weight, backtracking_edge = get_node_data_func(node)
        <span class="hljs-keyword">if</span> backtracking_edge <span class="hljs-keyword">is</span> <span class="hljs-literal">None</span>:
            <span class="hljs-keyword">break</span>
        <span class="hljs-keyword">else</span>:
            reverse_path.append(backtracking_edge)
        next_node = g.get_edge_from(backtracking_edge)
    <span class="hljs-keyword">return</span> reverse_path[::-<span class="hljs-number">1</span>]  <span class="hljs-comment"># this is the path in reverse -- reverse it to get it in the correct order</span>


<span class="hljs-keyword">def</span> <span class="hljs-title function_">find_max_path</span>(<span class="hljs-params">
        graph: Graph[N, ND, E, ED],
        start_node: N,
        end_node: N,
        set_node_data_func: SET_NODE_DATA_FUNC_TYPE,
        get_node_data_func: GET_NODE_DATA_FUNC_TYPE,
        get_edge_weight_func: GET_EDGE_WEIGHT_FUNC_TYPE
</span>) -&gt; <span class="hljs-type">Optional</span>[<span class="hljs-type">Tuple</span>[<span class="hljs-type">List</span>[E], <span class="hljs-built_in">float</span>]]:
    populate_weights_and_backtrack_pointers(
        graph,
        start_node,
        set_node_data_func,
        get_node_data_func,
        get_edge_weight_func
    )
    path = backtrack(graph, end_node, get_node_data_func)
    <span class="hljs-keyword">if</span> <span class="hljs-keyword">not</span> path:
        <span class="hljs-keyword">return</span> <span class="hljs-literal">None</span>
    weight, _ = get_node_data_func(end_node)
    <span class="hljs-keyword">return</span> path, weight
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the following graph...</p>
<p><img src="dot_178cd453d55fd535b5488d28b529a0c2.svg" alt="Dot diagram"></p>
<p>... the path with the max weight between A and E ...</p>
<ul>
<li>Maximum path = A -&gt; C -&gt; E</li>
<li>Maximum weight = 4.0</li>
</ul>
<p><img src="dot_8f5145677b3b43173a1c8ef136163426.svg" alt="Dot diagram"></p>
<p>The edges in blue signify the incoming edge that was selected for that node.</p>
</div>
<p></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>Note how ...</p>
<ul>
<li>the first variant's cache is built out such that it's possible to find the maximum path between any walked node and the destination node.</li>
<li>this variant's cache is built out such that it's possible to find the maximum path between the source node and any walked node.</li>
</ul>
<p>It's easy to flip this around by reversing the direction the algorithm walks.</p>
</div>
<a name="H_Global%20Alignment"></a><h3>Global Alignment</h3>
<p><a name="BM_(Algorithms%5C%2FSequence%20Alignment%2FGlobal%20Alignment)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FFind%20Maximum%20Path)_TOPIC%2F">Algorithms/Sequence Alignment/Find Maximum Path</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: Given two <a href="#BM_(sequence)%2Fi">sequence</a>s, perform <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a> and pull out the highest scoring alignment.</p>
<p><strong>WHY</strong>: A strong <a href="#BM_(global%20alignment%7Cglobal%20sequence%20alignment)%2Fi">global alignment</a> indicates that the <a href="#BM_(sequence)%2Fi">sequence</a>s are likely homologous / related.</p>
<a name="H_Graph%20Algorithm"></a><h4>Graph Algorithm</h4>
<p><a name="BM_(Algorithms%5C%2FSequence%20Alignment%2FGlobal%20Alignment%2FGraph%20Algorithm)_TOPIC%2F"></a><strong></strong></p>
<p><strong>ALGORITHM</strong>:</p>
<p>Determining the best scoring pairwise alignment can be done by generating a <a href="#BM_%5Cb(DAG)%5Cb%2F">DAG</a> of all possible operations at all possible positions in each <a href="#BM_(sequence)%2Fi">sequence</a>. Specifically, each operation (<a href="#BM_(indel)%2Fi">indel</a>, match, mismatch) is represented as an edge in the graph, where that edge has a weight. Operations with higher weights are more desirable operations compared to operations with lower weights (e.g. a match is typically more favourable than an <a href="#BM_(indel)%2Fi">indel</a>).</p>
<p>For example, consider a <a href="#BM_%5Cb(DAG)%5Cb%2F">DAG</a> that pits FOUR against CHOIR...</p>
<p><img src="svgbob_7e9d5fecdf93cf9ff7f6e6f7938c78ce6bc03d89.svg" alt="Kroki diagram output"></p>
<p>Given this graph, each ...</p>
<ul>
<li>diagonal edge is a replacement / keep matching.</li>
<li>horizontal edge is an <a href="#BM_(indel)%2Fi">indel</a> where the top is kept.</li>
<li>vertical edge is an <a href="#BM_(indel)%2Fi">indel</a> where the left is kept.</li>
</ul>
<div style="border:1px solid black;">
<p></p>
<p><img src="latex_ff32c897627430505ba188703be1d05e.svg" alt="Latex diagram"></p>
<p>NOTE: Each edge is labeled with the elements selected from the 1st sequence, 2nd sequence, and edge weight.</p>
</div>
<p></p>
<p>This graph is called an <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a>. A path through the <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a> from source (top-left) to sink (bottom-right) represents a single alignment, referred to as an <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a>. For example the <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a> representing...</p>
<pre><code>CH-OIR
--FOUR
</code></pre>
<p>... is as follows...</p>
<div style="border:1px solid black;">
<p></p>
<p><img src="latex_b15f446204c8f90af3fa526133e05a11.svg" alt="Latex diagram"></p>
<p>NOTE: Each edge is labeled with the elements selected from the 1st sequence, 2nd sequence, and edge weight.</p>
</div>
<p></p>
<p>The weight of an <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a> is the sum of its operation weights. Since operations with higher weights are more desirable than those with lower weights, <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a>s with higher weights are more desirable than those with lower weights. As such, out of all the <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a>s possible, the one with the highest weight is the one with the most desirable set of operations.</p>
<p>The highlighted path in the example path above has a weight of -1: -1 + -1 + -1 + 1 + 0 + 1.</p>
<p><a href="ch5_code/src/global_alignment/GlobalAlignment_Graph.py">ch5_code/src/global_alignment/GlobalAlignment_Graph.py</a> (lines 37 to 78):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">create_global_alignment_graph</span>(<span class="hljs-params">
        v: <span class="hljs-type">List</span>[ELEM],
        w: <span class="hljs-type">List</span>[ELEM],
        weight_lookup: WeightLookup
</span>) -&gt; Graph[<span class="hljs-type">Tuple</span>[<span class="hljs-built_in">int</span>, ...], NodeData, <span class="hljs-built_in">str</span>, EdgeData]:
    graph = create_grid_graph(
        [v, w],
        <span class="hljs-keyword">lambda</span> n_id: NodeData(),
        <span class="hljs-keyword">lambda</span> src_n_id, dst_n_id, offset, elems: EdgeData(elems[<span class="hljs-number">0</span>], elems[<span class="hljs-number">1</span>], weight_lookup.lookup(*elems))
    )
    <span class="hljs-keyword">return</span> graph


<span class="hljs-keyword">def</span> <span class="hljs-title function_">global_alignment</span>(<span class="hljs-params">
        v: <span class="hljs-type">List</span>[ELEM],
        w: <span class="hljs-type">List</span>[ELEM],
        weight_lookup: WeightLookup
</span>) -&gt; <span class="hljs-type">Tuple</span>[<span class="hljs-built_in">float</span>, <span class="hljs-type">List</span>[<span class="hljs-built_in">str</span>], <span class="hljs-type">List</span>[<span class="hljs-type">Tuple</span>[ELEM, ELEM]]]:
    v_node_count = <span class="hljs-built_in">len</span>(v) + <span class="hljs-number">1</span>
    w_node_count = <span class="hljs-built_in">len</span>(w) + <span class="hljs-number">1</span>
    graph = create_global_alignment_graph(v, w, weight_lookup)
    from_node = (<span class="hljs-number">0</span>, <span class="hljs-number">0</span>)
    to_node = (v_node_count - <span class="hljs-number">1</span>, w_node_count - <span class="hljs-number">1</span>)
    populate_weights_and_backtrack_pointers(
        graph,
        from_node,
        <span class="hljs-keyword">lambda</span> n_id, weight, e_id: graph.get_node_data(n_id).set_weight_and_backtracking_edge(weight, e_id),
        <span class="hljs-keyword">lambda</span> n_id: graph.get_node_data(n_id).get_weight_and_backtracking_edge(),
        <span class="hljs-keyword">lambda</span> e_id: graph.get_edge_data(e_id).weight
    )
    final_weight = graph.get_node_data(to_node).weight
    edges = backtrack(
        graph,
        to_node,
        <span class="hljs-keyword">lambda</span> n_id: graph.get_node_data(n_id).get_weight_and_backtracking_edge()
    )
    alignment = []
    <span class="hljs-keyword">for</span> e <span class="hljs-keyword">in</span> edges:
        ed = graph.get_edge_data(e)
        alignment.append((ed.v_elem, ed.w_elem))
    <span class="hljs-keyword">return</span> final_weight, edges, alignment
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the sequences TAAT and GAT and the score matrix...</p>
<pre><code>INDEL=-1.0
   A  C  T  G
A  1  0  0  0
C  0  1  0  0
T  0  0  1  0
G  0  0  0  1

</code></pre>
<p>... the global alignment is...</p>
<p><img src="latex_148a493a710b097feca4ff9287b8a20d.svg" alt="Latex diagram"></p>
<pre><code>TAAT
GA-T
</code></pre>
<p>Weight: 1.0</p>
</div>
<p></p>
<a name="H_Matrix%20Algorithm"></a><h4>Matrix Algorithm</h4>
<p><a name="BM_(Algorithms%5C%2FSequence%20Alignment%2FGlobal%20Alignment%2FMatrix%20Algorithm)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FGlobal%20Alignment%2FGraph%20Algorithm)_TOPIC%2F">Algorithms/Sequence Alignment/Global Alignment/Graph Algorithm</a></li>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FFind%20Maximum%20Path%2FBacktrack%20Algorithm)_TOPIC%2F">Algorithms/Sequence Alignment/Find Maximum Path/Backtrack Algorithm</a></li>
</ul>
</div>
<p><strong>ALGORITHM</strong>:</p>
<p>The following algorithm is essentially the same as the graph algorithm, except that the implementation is much more sympathetic to modern hardware. The <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a> is represented as a 2D matrix where each element in the matrix represents a node in the <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a>. The elements are then populated in a predefined <a href="#BM_(topological%20order)%2Fi">topological order</a>, where each element gets populated with the node weight, the chosen backtracking edge, and the elements from that backtracking edge.</p>
<p>Since the <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a> is a grid, the node weights may be populated either...</p>
<ul>
<li>row-by-row, starting from the left column and walking to the right column.</li>
<li>column-by-column, starting from the top row and walking to the bottom row.</li>
</ul>
<p>In either case, the nodes being walked are guaranteed to have their parent node weights already set.</p>
<p><img src="svgbob_a0c3678c1817dbc609f6127b83938d7cb4bcd9d1.svg" alt="Kroki diagram output"></p>
<p><a href="ch5_code/src/global_alignment/GlobalAlignment_Matrix.py">ch5_code/src/global_alignment/GlobalAlignment_Matrix.py</a> (lines 10 to 73):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">backtrack</span>(<span class="hljs-params">
        node_matrix: <span class="hljs-type">List</span>[<span class="hljs-type">List</span>[<span class="hljs-type">Any</span>]]
</span>) -&gt; <span class="hljs-type">Tuple</span>[<span class="hljs-built_in">float</span>, <span class="hljs-type">List</span>[<span class="hljs-type">Tuple</span>[ELEM, ELEM]]]:
    v_node_idx = <span class="hljs-built_in">len</span>(node_matrix) - <span class="hljs-number">1</span>
    w_node_idx = <span class="hljs-built_in">len</span>(node_matrix[<span class="hljs-number">0</span>]) - <span class="hljs-number">1</span>
    final_weight = node_matrix[v_node_idx][w_node_idx][<span class="hljs-number">0</span>]
    alignment = []
    <span class="hljs-keyword">while</span> v_node_idx != <span class="hljs-number">0</span> <span class="hljs-keyword">or</span> w_node_idx != <span class="hljs-number">0</span>:
        _, elems, backtrack_ptr = node_matrix[v_node_idx][w_node_idx]
        <span class="hljs-keyword">if</span> backtrack_ptr == <span class="hljs-string">'↓'</span>:
            v_node_idx -= <span class="hljs-number">1</span>
        <span class="hljs-keyword">elif</span> backtrack_ptr == <span class="hljs-string">'→'</span>:
            w_node_idx -= <span class="hljs-number">1</span>
        <span class="hljs-keyword">elif</span> backtrack_ptr == <span class="hljs-string">'↘'</span>:
            v_node_idx -= <span class="hljs-number">1</span>
            w_node_idx -= <span class="hljs-number">1</span>
        alignment.append(elems)
    <span class="hljs-keyword">return</span> final_weight, alignment[::-<span class="hljs-number">1</span>]


<span class="hljs-keyword">def</span> <span class="hljs-title function_">global_alignment</span>(<span class="hljs-params">
        v: <span class="hljs-type">List</span>[ELEM],
        w: <span class="hljs-type">List</span>[ELEM],
        weight_lookup: WeightLookup
</span>) -&gt; <span class="hljs-type">Tuple</span>[<span class="hljs-built_in">float</span>, <span class="hljs-type">List</span>[<span class="hljs-type">Tuple</span>[ELEM, ELEM]]]:
    v_node_count = <span class="hljs-built_in">len</span>(v) + <span class="hljs-number">1</span>
    w_node_count = <span class="hljs-built_in">len</span>(w) + <span class="hljs-number">1</span>
    node_matrix = []
    <span class="hljs-keyword">for</span> v_node_idx <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(v_node_count):
        row = []
        <span class="hljs-keyword">for</span> w_node_idx <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(w_node_count):
            row.append([-<span class="hljs-number">1.0</span>, (<span class="hljs-literal">None</span>, <span class="hljs-literal">None</span>), <span class="hljs-string">'?'</span>])
        node_matrix.append(row)
    node_matrix[<span class="hljs-number">0</span>][<span class="hljs-number">0</span>][<span class="hljs-number">0</span>] = <span class="hljs-number">0.0</span>           <span class="hljs-comment"># source node weight</span>
    node_matrix[<span class="hljs-number">0</span>][<span class="hljs-number">0</span>][<span class="hljs-number">1</span>] = (<span class="hljs-literal">None</span>, <span class="hljs-literal">None</span>)  <span class="hljs-comment"># source node elements (elements don't matter for source node)</span>
    node_matrix[<span class="hljs-number">0</span>][<span class="hljs-number">0</span>][<span class="hljs-number">2</span>] = <span class="hljs-string">'↘'</span>           <span class="hljs-comment"># source node backtracking edge (direction doesn't matter for source node)</span>
    <span class="hljs-keyword">for</span> v_node_idx, w_node_idx <span class="hljs-keyword">in</span> product(<span class="hljs-built_in">range</span>(v_node_count), <span class="hljs-built_in">range</span>(w_node_count)):
        parents = []
        <span class="hljs-keyword">if</span> v_node_idx &gt; <span class="hljs-number">0</span> <span class="hljs-keyword">and</span> w_node_idx &gt; <span class="hljs-number">0</span>:
            v_elem = v[v_node_idx - <span class="hljs-number">1</span>]
            w_elem = w[w_node_idx - <span class="hljs-number">1</span>]
            parents.append([
                node_matrix[v_node_idx - <span class="hljs-number">1</span>][w_node_idx - <span class="hljs-number">1</span>][<span class="hljs-number">0</span>] + weight_lookup.lookup(v_elem, w_elem),
                (v_elem, w_elem),
                <span class="hljs-string">'↘'</span>
            ])
        <span class="hljs-keyword">if</span> v_node_idx &gt; <span class="hljs-number">0</span>:
            v_elem = v[v_node_idx - <span class="hljs-number">1</span>]
            parents.append([
                node_matrix[v_node_idx - <span class="hljs-number">1</span>][w_node_idx][<span class="hljs-number">0</span>] + weight_lookup.lookup(v_elem, <span class="hljs-literal">None</span>),
                (v_elem, <span class="hljs-literal">None</span>),
                <span class="hljs-string">'↓'</span>
            ])
        <span class="hljs-keyword">if</span> w_node_idx &gt; <span class="hljs-number">0</span>:
            w_elem = w[w_node_idx - <span class="hljs-number">1</span>]
            parents.append([
                node_matrix[v_node_idx][w_node_idx - <span class="hljs-number">1</span>][<span class="hljs-number">0</span>] + weight_lookup.lookup(<span class="hljs-literal">None</span>, w_elem),
                (<span class="hljs-literal">None</span>, w_elem),
                <span class="hljs-string">'→'</span>
            ])
        <span class="hljs-keyword">if</span> parents:  <span class="hljs-comment"># parents wil be empty if v_node_idx and w_node_idx were both 0</span>
            node_matrix[v_node_idx][w_node_idx] = <span class="hljs-built_in">max</span>(parents, key=<span class="hljs-keyword">lambda</span> x: x[<span class="hljs-number">0</span>])
    <span class="hljs-keyword">return</span> backtrack(node_matrix)
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the sequences TATTATTAT and AAA and the score matrix...</p>
<pre><code>INDEL=-1.0
   A  C  T  G
A  1  0  0  0
C  0  1  0  0
T  0  0  1  0
G  0  0  0  1

</code></pre>
<p>... the global alignment is...</p>
<pre><code>TATTATTAT
-A--A--A-
</code></pre>
<p>Weight: -3.0</p>
</div>
<p></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The standard <a href="#BM_(levenshtein%20distance%7Cstring%20distance)%2Fi">Levenshtein distance</a> algorithm using a 2D array you remember from over a decade ago is this algorithm: Matrix-based <a href="#BM_(global%20alignment%7Cglobal%20sequence%20alignment)%2Fi">global alignment</a> where matches score 0 but mismatches and <a href="#BM_(indel)%2Fi">indel</a>s score -1. The final weight of the alignment is the minimum number of operations required to convert one <a href="#BM_(sequence)%2Fi">sequence</a> to another (e.g. swap, insert, delete) -- it'll be negative, ignore the sign.</p>
</div>
<a name="H_Divide-and-Conquer%20Algorithm"></a><h4>Divide-and-Conquer Algorithm</h4>
<p><a name="BM_(Algorithms%5C%2FSequence%20Alignment%2FGlobal%20Alignment%2FDivide-and-Conquer%20Algorithm)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FGlobal%20Alignment%2FMatrix%20Algorithm)_TOPIC%2F">Algorithms/Sequence Alignment/Global Alignment/Matrix Algorithm</a></li>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FFind%20Maximum%20Path%2FBacktrack%20Algorithm)_TOPIC%2F">Algorithms/Sequence Alignment/Find Maximum Path/Backtrack Algorithm</a></li>
</ul>
</div>
<p><strong>ALGORITHM</strong>:</p>
<p>The following algorithm extends the matrix algorithm such that it can process much larger graphs at the expense of duplicating some computation work (trading time for space). It relies on two ideas.</p>
<p>Recall that in the matrix implementation of <a href="#BM_(global%20alignment%7Cglobal%20sequence%20alignment)%2Fi">global alignment</a>, node weights are populated in a pre-defined <a href="#BM_(topological%20order)%2Fi">topological order</a> (either row-by-row or column-by-column). Imagine that you've chosen to populate the matrix column-by-column.</p>
<p>The first idea is that, if all you care about is the final weight of the sink node, the matrix implementation technically only needs to keep 2 columns in memory: the column having its node weights populated as well as the previous column.</p>
<p>In other words, the only data needed to calculate the weights of the next column is the weights in the previous column.</p>
<p><img src="svgbob_a778e17c20ba10aa76e0c0375a70cc19614cd870.svg" alt="Kroki diagram output"></p>
<p><a href="ch5_code/src/global_alignment/Global_ForwardSweeper.py">ch5_code/src/global_alignment/Global_ForwardSweeper.py</a> (lines 9 to 51):</p>
<pre class="hljs"><code><span class="hljs-keyword">class</span> <span class="hljs-title class_">ForwardSweeper</span>:
    <span class="hljs-keyword">def</span> <span class="hljs-title function_">__init__</span>(<span class="hljs-params">self, v: <span class="hljs-type">List</span>[ELEM], w: <span class="hljs-type">List</span>[ELEM], weight_lookup: WeightLookup, col_backtrack: <span class="hljs-built_in">int</span> = <span class="hljs-number">2</span></span>):
        self.v = v
        self.v_node_count = <span class="hljs-built_in">len</span>(v) + <span class="hljs-number">1</span>
        self.w = w
        self.w_node_count = <span class="hljs-built_in">len</span>(w) + <span class="hljs-number">1</span>
        self.weight_lookup = weight_lookup
        self.col_backtrack = col_backtrack
        self.matrix_v_start_idx = <span class="hljs-number">0</span>  <span class="hljs-comment"># col</span>
        self.matrix = []
        self._reset()

    <span class="hljs-keyword">def</span> <span class="hljs-title function_">_reset</span>(<span class="hljs-params">self</span>):
        self.matrix_v_start_idx = <span class="hljs-number">0</span>  <span class="hljs-comment"># col</span>
        col = [-<span class="hljs-number">1.0</span>] * self.w_node_count
        col[<span class="hljs-number">0</span>] = <span class="hljs-number">0.0</span>  <span class="hljs-comment"># source node weight is 0</span>
        <span class="hljs-keyword">for</span> w_idx <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(<span class="hljs-number">1</span>, self.w_node_count):
            col[w_idx] = col[w_idx - <span class="hljs-number">1</span>] + self.weight_lookup.lookup(<span class="hljs-literal">None</span>, self.w[w_idx - <span class="hljs-number">1</span>])
        self.matrix = [col]

    <span class="hljs-keyword">def</span> <span class="hljs-title function_">_step</span>(<span class="hljs-params">self</span>):
        next_col = [-<span class="hljs-number">1.0</span>] * self.w_node_count
        next_v_idx = self.matrix_v_start_idx + <span class="hljs-built_in">len</span>(self.matrix)
        <span class="hljs-keyword">if</span> <span class="hljs-built_in">len</span>(self.matrix) == self.col_backtrack:
            self.matrix.pop(<span class="hljs-number">0</span>)
            self.matrix_v_start_idx += <span class="hljs-number">1</span>
        self.matrix += [next_col]
        self.matrix[-<span class="hljs-number">1</span>][<span class="hljs-number">0</span>] = self.matrix[-<span class="hljs-number">2</span>][<span class="hljs-number">0</span>] + self.weight_lookup.lookup(self.v[next_v_idx - <span class="hljs-number">1</span>], <span class="hljs-literal">None</span>)  <span class="hljs-comment"># right penalty for first row of new col</span>
        <span class="hljs-keyword">for</span> w_idx <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(<span class="hljs-number">1</span>, <span class="hljs-built_in">len</span>(self.w) + <span class="hljs-number">1</span>):
            self.matrix[-<span class="hljs-number">1</span>][w_idx] = <span class="hljs-built_in">max</span>(
                self.matrix[-<span class="hljs-number">2</span>][w_idx] + self.weight_lookup.lookup(<span class="hljs-literal">None</span>, self.w[w_idx - <span class="hljs-number">1</span>]),                     <span class="hljs-comment"># right score</span>
                self.matrix[-<span class="hljs-number">1</span>][w_idx-<span class="hljs-number">1</span>] + self.weight_lookup.lookup(self.v[next_v_idx - <span class="hljs-number">1</span>], <span class="hljs-literal">None</span>),              <span class="hljs-comment"># down score</span>
                self.matrix[-<span class="hljs-number">2</span>][w_idx-<span class="hljs-number">1</span>] + self.weight_lookup.lookup(self.v[next_v_idx - <span class="hljs-number">1</span>], self.w[w_idx - <span class="hljs-number">1</span>])  <span class="hljs-comment"># diag score</span>
            )

    <span class="hljs-keyword">def</span> <span class="hljs-title function_">get_col</span>(<span class="hljs-params">self, idx: <span class="hljs-built_in">int</span></span>):
        <span class="hljs-keyword">if</span> idx &lt; self.matrix_v_start_idx:
            self._reset()
        furthest_stored_idx = self.matrix_v_start_idx + <span class="hljs-built_in">len</span>(self.matrix) - <span class="hljs-number">1</span>
        <span class="hljs-keyword">for</span> _ <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(furthest_stored_idx, idx):
            self._step()
        <span class="hljs-keyword">return</span> <span class="hljs-built_in">list</span>(self.matrix[idx - self.matrix_v_start_idx])
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the sequences TACT and GACGT and the score matrix...</p>
<pre><code>INDEL=-1.0
   A  C  T  G
A  1  0  0  0
C  0  1  0  0
T  0  0  1  0
G  0  0  0  1

</code></pre>
<p>... the node weights are ...</p>
<pre><code>   0.0  -1.0  -2.0  -3.0  -4.0
  -1.0   0.0  -1.0  -2.0  -3.0
  -2.0  -1.0   1.0   0.0  -1.0
  -3.0  -2.0   0.0   2.0   1.0
  -4.0  -3.0  -1.0   1.0   2.0
  -5.0  -3.0  -2.0   0.0   2.0
</code></pre>
<p>The sink node weight (maximum alignment path weight) is 2.0</p>
</div>
<p></p>
<p>The second idea is that, for a column, it's possible to find out which node in that column a maximum <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a> travels through without knowing that path beforehand.</p>
<p><img src="svgbob_a4f324dc13bf528a19be5bedcd757c10e373b382.svg" alt="Kroki diagram output"></p>
<p>Knowing this, a <a href="#BM_(divide-and-conquer%20algorithm%7Cdivide%20and%20conquer%20algorithm)%2Fi">divide-and-conquer algorithm</a> may be used to find that maximum <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a>. Any <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a> must travel from the source node (top-left) to the sink node (bottom-right). If you're able to find a node between the source node and sink node that a maximum <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a> travels through, you can sub-divide the <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a> into 2.</p>
<p>That is, if you know that a maximum <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a> travels through some node, it's guaranteed that...</p>
<ul>
<li>prior parts of the path travel through the region that's to the top-left of that node.</li>
<li>subsequent parts of that path travel through the region that's to the bottom-right of that node.</li>
</ul>
<p><img src="svgbob_a1e221ef894490d6a8b4252ca52d056518189ffd.svg" alt="Kroki diagram output"></p>
<p>By recursively performing this operation, you can pull out all nodes that make up a maximum <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a>:</p>
<ul>
<li>Pick a column, find a node, and sub-divide.</li>
<li>For each sub-division from last step: Pick a column, find a node, and sub-divide.</li>
<li>For each sub-division from last step: Pick a column, find a node, and sub-divide.</li>
<li>etc..</li>
</ul>
<p>Finding the edges between these nodes yields the maximum <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a>. To find the edges between the node found at column n and the node found at column n + 1, isolate the <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a> between those nodes and perform the standard matrix variant of <a href="#BM_(global%20alignment%7Cglobal%20sequence%20alignment)%2Fi">global alignment</a>. The graph will likely be very small, so the computation and space requirements will likely be very low.</p>
<p><a href="ch5_code/src/global_alignment/GlobalAlignment_DivideAndConquer_NodeVariant.py">ch5_code/src/global_alignment/GlobalAlignment_DivideAndConquer_NodeVariant.py</a> (lines 11 to 40):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">find_max_alignment_path_nodes</span>(<span class="hljs-params">
        v: <span class="hljs-type">List</span>[ELEM],
        w: <span class="hljs-type">List</span>[ELEM],
        weight_lookup: WeightLookup,
        buffer: <span class="hljs-type">List</span>[<span class="hljs-type">Tuple</span>[<span class="hljs-built_in">int</span>, <span class="hljs-built_in">int</span>]],
        v_offset: <span class="hljs-built_in">int</span> = <span class="hljs-number">0</span>,
        w_offset: <span class="hljs-built_in">int</span> = <span class="hljs-number">0</span></span>) -&gt; <span class="hljs-literal">None</span>:
    <span class="hljs-keyword">if</span> <span class="hljs-built_in">len</span>(v) == <span class="hljs-number">0</span> <span class="hljs-keyword">or</span> <span class="hljs-built_in">len</span>(w) == <span class="hljs-number">0</span>:
        <span class="hljs-keyword">return</span>
    c, r = find_node_that_max_alignment_path_travels_through_at_middle_col(v, w, weight_lookup)
    find_max_alignment_path_nodes(v[:c-<span class="hljs-number">1</span>], w[:r-<span class="hljs-number">1</span>], weight_lookup, buffer, v_offset=<span class="hljs-number">0</span>, w_offset=<span class="hljs-number">0</span>)
    buffer.append((v_offset + c, w_offset + r))
    find_max_alignment_path_nodes(v[c:], w[r:], weight_lookup, buffer, v_offset=v_offset+c, w_offset=v_offset+r)


<span class="hljs-keyword">def</span> <span class="hljs-title function_">global_alignment</span>(<span class="hljs-params">
        v: <span class="hljs-type">List</span>[ELEM],
        w: <span class="hljs-type">List</span>[ELEM],
        weight_lookup: WeightLookup
</span>) -&gt; <span class="hljs-type">Tuple</span>[<span class="hljs-built_in">float</span>, <span class="hljs-type">List</span>[<span class="hljs-type">Tuple</span>[ELEM, ELEM]]]:
    nodes = [(<span class="hljs-number">0</span>, <span class="hljs-number">0</span>)]
    find_max_alignment_path_nodes(v, w, weight_lookup, nodes)
    weight = <span class="hljs-number">0.0</span>
    alignment = []
    <span class="hljs-keyword">for</span> (v_idx1, w_idx1), (v_idx2, w_idx2) <span class="hljs-keyword">in</span> <span class="hljs-built_in">zip</span>(nodes, nodes[<span class="hljs-number">1</span>:]):
        sub_weight, sub_alignment = GlobalAlignment_Matrix.global_alignment(v[v_idx1:v_idx2], w[w_idx1:w_idx2], weight_lookup)
        weight += sub_weight
        alignment += sub_alignment
    <span class="hljs-keyword">return</span> weight, alignment
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the sequences TACT and GACGT and the score matrix...</p>
<pre><code>INDEL=-1.0
   A  C  T  G
A  1  0  0  0
C  0  1  0  0
T  0  0  1  0
G  0  0  0  1

</code></pre>
<p>... the global alignment is...</p>
<pre><code>TAC-T
GACGT
</code></pre>
<p>Weight: 2.0</p>
</div>
<p></p>
<p>To understand how to find which node in a column a maximum <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a> travels through, consider what happens when edge directions are reversed in an <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a>. When edge directions are reversed, the <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a> essentially becomes the <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a> for the reversed <a href="#BM_(sequence)%2Fi">sequence</a>s. For example, reversing the edges for the <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a> of SNACK and AJAX is essentially the same as the <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a> for KCANS (reverse of SNACK) and XAJA (reverse of AJAX)...</p>
<p><img src="svgbob_da8db26052b32c600c5e011bfc6c5cac06910718.svg" alt="Kroki diagram output"></p>
<p>Between an <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a> and its reversed edge variant, a maximum <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a> should travel through the same set of nodes. Notice how in the following example, ...</p>
<ol>
<li>
<p>the maximum <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a> in both <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a>s have the same edges.</p>
</li>
<li>
<p>the sink node weight in both <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a>s are the same.</p>
</li>
<li>
<p>for any node that the maximum <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a> travels through, taking the weight of that node from both <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a>s and adding them together results in the sink node weight.</p>
</li>
<li>
<p>for any node that the maximum <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a> <em>DOES NOT</em> travel through, taking the weight of that node from both <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a>s and adding them together results in <em>LESS THAN</em> the sink node weight.</p>
</li>
</ol>
<div style="border:1px solid black;">
<p></p>
<p><img src="latex_539aa4c9412be2f54aea834e08f31d2f.svg" alt="Latex diagram"></p>
</div>
<p></p>
<p>Insights #3 and #4 in the list above are the key for this algorithm. Consider an <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a> getting split down a column into two. The first half has edges traveling in the normal direction but the second half has its edges reversed...</p>
<p><img src="svgbob_072c713f2836b876275a7dbf010669c022d1b6a1.svg" alt="Kroki diagram output"></p>
<p>Populate node weights for both halves. Then, pair up half 1's last column with half 2's first column. For each row in the pair, add together the node weights in that row. The row with the maximum sum is for a node that a maximum <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a> travels through (insight #4 above). That maximum sum will always end up being the weight of the sink node in the original non-split <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a> (insight #3 above).</p>
<div style="border:1px solid black;">
<p></p>
<p><img src="latex_7306d547cf3d4eb48cb21caab56b4a60.svg" alt="Latex diagram"></p>
</div>
<p></p>
<p>One way to think about what's happening above is that the algorithm is converging on to the same answer but at a different spot in the <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a> (the same edge weights are being added). Normally the algorithm converges on to the bottom-right node of the <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a>. If it were to instead converge on the column just before, the answer would be the same, but the node's position in that column may be different -- it may be any node that ultimately drives to the bottom-right node.</p>
<p>Given that there may be multiple maximum <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a>s for an <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a>, there may be multiple nodes found per column. Each found node may be for a different maximum <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a> or the same maximum <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a>.</p>
<p>Ultimately, this entire process may be combined with the first idea (only need the previous column in memory to calculate the next column) such that the algorithm requires much lower memory requirements. That is, to find the nodes in a column which maximum <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a>s travel through, the...</p>
<ul>
<li>forward sweep only requires holding on to the weights from column n-1.</li>
<li>reverse sweep only requires holding on to the weights from column n+1.</li>
</ul>
<p><a href="ch5_code/src/global_alignment/Global_SweepCombiner.py">ch5_code/src/global_alignment/Global_SweepCombiner.py</a> (lines 10 to 19):</p>
<pre class="hljs"><code><span class="hljs-keyword">class</span> <span class="hljs-title class_">SweepCombiner</span>:
    <span class="hljs-keyword">def</span> <span class="hljs-title function_">__init__</span>(<span class="hljs-params">self, v: <span class="hljs-type">List</span>[ELEM], w: <span class="hljs-type">List</span>[ELEM], weight_lookup: WeightLookup</span>):
        self.forward_sweeper = ForwardSweeper(v, w, weight_lookup)
        self.reverse_sweeper = ReverseSweeper(v, w, weight_lookup)

    <span class="hljs-keyword">def</span> <span class="hljs-title function_">get_col</span>(<span class="hljs-params">self, idx: <span class="hljs-built_in">int</span></span>):
        fcol = self.forward_sweeper.get_col(idx)
        rcol = self.reverse_sweeper.get_col(idx)
        <span class="hljs-keyword">return</span> [a + b <span class="hljs-keyword">for</span> a, b <span class="hljs-keyword">in</span> <span class="hljs-built_in">zip</span>(fcol, rcol)]
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the sequences TACT and GACGT and the score matrix...</p>
<pre><code>INDEL=-1.0
   A  C  T  G
A  1  0  0  0
C  0  1  0  0
T  0  0  1  0
G  0  0  0  1

</code></pre>
<p>... the combined node weights at column 3 are ...</p>
<pre><code>  -6.0
  -4.0
  -1.0
   2.0
   2.0
  -1.0
</code></pre>
</div>
<p></p>
<p>To recap, the full <a href="#BM_(divide-and-conquer%20algorithm%7Cdivide%20and%20conquer%20algorithm)%2Fi">divide-and-conquer algorithm</a> is as follows: For the middle column in an <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a>, find a node that a maximum <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a> travels through. Then, sub-divide the <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a> based on that node. Recursively repeat this process on each sub-division until you have a node from each column -- these are the nodes in a maximum <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a>. The edges between these found nodes can be determined by finding a maximum <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a> between each found node and its <a href="#BM_(neighbour)%2Fi">neighbour</a>ing found node. Concatenate these edges to construct the path.</p>
<p><a href="ch5_code/src/global_alignment/Global_FindNodeThatMaxAlignmentPathTravelsThroughAtColumn.py">ch5_code/src/global_alignment/Global_FindNodeThatMaxAlignmentPathTravelsThroughAtColumn.py</a> (lines 10 to 29):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">find_node_that_max_alignment_path_travels_through_at_col</span>(<span class="hljs-params">
        v: <span class="hljs-type">List</span>[ELEM],
        w: <span class="hljs-type">List</span>[ELEM],
        weight_lookup: WeightLookup,
        col: <span class="hljs-built_in">int</span>
</span>) -&gt; <span class="hljs-type">Tuple</span>[<span class="hljs-built_in">int</span>, <span class="hljs-built_in">int</span>]:
    col_vals = SweepCombiner(v, w, weight_lookup).get_col(col)
    row, _ = <span class="hljs-built_in">max</span>(<span class="hljs-built_in">enumerate</span>(col_vals), key=<span class="hljs-keyword">lambda</span> x: x[<span class="hljs-number">1</span>])
    <span class="hljs-keyword">return</span> col, row


<span class="hljs-keyword">def</span> <span class="hljs-title function_">find_node_that_max_alignment_path_travels_through_at_middle_col</span>(<span class="hljs-params">
        v: <span class="hljs-type">List</span>[ELEM],
        w: <span class="hljs-type">List</span>[ELEM],
        weight_lookup: WeightLookup
</span>) -&gt; <span class="hljs-type">Tuple</span>[<span class="hljs-built_in">int</span>, <span class="hljs-built_in">int</span>]:
    v_node_count = <span class="hljs-built_in">len</span>(v) + <span class="hljs-number">1</span>
    middle_col_idx = v_node_count // <span class="hljs-number">2</span>
    <span class="hljs-keyword">return</span> find_node_that_max_alignment_path_travels_through_at_col(v, w, weight_lookup, middle_col_idx)
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the sequences TACT and GACGT and the score matrix...</p>
<pre><code>INDEL=-1.0
   A  C  T  G
A  1  0  0  0
C  0  1  0  0
T  0  0  1  0
G  0  0  0  1

</code></pre>
<p>... a maximum alignment path is guaranteed to travel through (3, 3).</p>
</div>
<p></p>
<p><a href="ch5_code/src/global_alignment/GlobalAlignment_DivideAndConquer_NodeVariant.py">ch5_code/src/global_alignment/GlobalAlignment_DivideAndConquer_NodeVariant.py</a> (lines 11 to 40):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">find_max_alignment_path_nodes</span>(<span class="hljs-params">
        v: <span class="hljs-type">List</span>[ELEM],
        w: <span class="hljs-type">List</span>[ELEM],
        weight_lookup: WeightLookup,
        buffer: <span class="hljs-type">List</span>[<span class="hljs-type">Tuple</span>[<span class="hljs-built_in">int</span>, <span class="hljs-built_in">int</span>]],
        v_offset: <span class="hljs-built_in">int</span> = <span class="hljs-number">0</span>,
        w_offset: <span class="hljs-built_in">int</span> = <span class="hljs-number">0</span></span>) -&gt; <span class="hljs-literal">None</span>:
    <span class="hljs-keyword">if</span> <span class="hljs-built_in">len</span>(v) == <span class="hljs-number">0</span> <span class="hljs-keyword">or</span> <span class="hljs-built_in">len</span>(w) == <span class="hljs-number">0</span>:
        <span class="hljs-keyword">return</span>
    c, r = find_node_that_max_alignment_path_travels_through_at_middle_col(v, w, weight_lookup)
    find_max_alignment_path_nodes(v[:c-<span class="hljs-number">1</span>], w[:r-<span class="hljs-number">1</span>], weight_lookup, buffer, v_offset=<span class="hljs-number">0</span>, w_offset=<span class="hljs-number">0</span>)
    buffer.append((v_offset + c, w_offset + r))
    find_max_alignment_path_nodes(v[c:], w[r:], weight_lookup, buffer, v_offset=v_offset+c, w_offset=v_offset+r)


<span class="hljs-keyword">def</span> <span class="hljs-title function_">global_alignment</span>(<span class="hljs-params">
        v: <span class="hljs-type">List</span>[ELEM],
        w: <span class="hljs-type">List</span>[ELEM],
        weight_lookup: WeightLookup
</span>) -&gt; <span class="hljs-type">Tuple</span>[<span class="hljs-built_in">float</span>, <span class="hljs-type">List</span>[<span class="hljs-type">Tuple</span>[ELEM, ELEM]]]:
    nodes = [(<span class="hljs-number">0</span>, <span class="hljs-number">0</span>)]
    find_max_alignment_path_nodes(v, w, weight_lookup, nodes)
    weight = <span class="hljs-number">0.0</span>
    alignment = []
    <span class="hljs-keyword">for</span> (v_idx1, w_idx1), (v_idx2, w_idx2) <span class="hljs-keyword">in</span> <span class="hljs-built_in">zip</span>(nodes, nodes[<span class="hljs-number">1</span>:]):
        sub_weight, sub_alignment = GlobalAlignment_Matrix.global_alignment(v[v_idx1:v_idx2], w[w_idx1:w_idx2], weight_lookup)
        weight += sub_weight
        alignment += sub_alignment
    <span class="hljs-keyword">return</span> weight, alignment
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the sequences TACT and GACGT and the score matrix...</p>
<pre><code>INDEL=-1.0
   A  C  T  G
A  1  0  0  0
C  0  1  0  0
T  0  0  1  0
G  0  0  0  1

</code></pre>
<p>... the global alignment is...</p>
<pre><code>TAC-T
GACGT
</code></pre>
<p>Weight: 2.0</p>
</div>
<p></p>
<p>A slightly more complicated but also more elegant / efficient solution is to extend the algorithm to find the edges for the nodes that it finds. In other words, rather than finding just nodes that maximum <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a>s travel through, find the edges where those nodes are the edge source (node that the edge starts from).</p>
<p>The algorithm finds all nodes that a maximum <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a> travels through at both column n and column n + 1. For a found node in column n, it's guaranteed that at least one of its immediate <a href="#BM_(neighbour)%2Fi">neighbour</a>s is also a found node. It may be that the node immediately to the ...</p>
<ul>
<li>right of it (same row but column n + 1) is also a found node.</li>
<li>bottom of it (1 row down and column n) is also a found node.</li>
<li>bottom-right of it (1 row down and column n + 1) is also a found node.</li>
</ul>
<p>Of the immediate <a href="#BM_(neighbour)%2Fi">neighbour</a>s that are also found nodes, the one forming the edge with the highest weight is the edge that a maximum <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a> travels through.</p>
<p><a href="ch5_code/src/global_alignment/Global_FindEdgeThatMaxAlignmentPathTravelsThroughAtColumn.py">ch5_code/src/global_alignment/Global_FindEdgeThatMaxAlignmentPathTravelsThroughAtColumn.py</a> (lines 10 to 65):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">find_edge_that_max_alignment_path_travels_through_at_col</span>(<span class="hljs-params">
        v: <span class="hljs-type">List</span>[ELEM],
        w: <span class="hljs-type">List</span>[ELEM],
        weight_lookup: WeightLookup,
        col: <span class="hljs-built_in">int</span>
</span>) -&gt; <span class="hljs-type">Tuple</span>[<span class="hljs-type">Tuple</span>[<span class="hljs-built_in">int</span>, <span class="hljs-built_in">int</span>], <span class="hljs-type">Tuple</span>[<span class="hljs-built_in">int</span>, <span class="hljs-built_in">int</span>]]:
    v_node_count = <span class="hljs-built_in">len</span>(v) + <span class="hljs-number">1</span>
    w_node_count = <span class="hljs-built_in">len</span>(w) + <span class="hljs-number">1</span>
    sc = SweepCombiner(v, w, weight_lookup)
    <span class="hljs-comment"># Get max node in column -- max alignment path guaranteed to go through here.</span>
    col_vals = sc.get_col(col)
    row, _ = <span class="hljs-built_in">max</span>(<span class="hljs-built_in">enumerate</span>(col_vals), key=<span class="hljs-keyword">lambda</span> x: x[<span class="hljs-number">1</span>])
    <span class="hljs-comment"># Check node immediately to the right, down, right-down (diag) -- the ones with the max value MAY form the edge that</span>
    <span class="hljs-comment"># the max alignment path goes through. Recall that the max value will be the same max value as the one from col_vals</span>
    <span class="hljs-comment"># (weight of the final alignment path / sink node in the full alignment graph).</span>
    <span class="hljs-comment">#</span>
    <span class="hljs-comment"># Of the ones WITH the max value, check the weights formed by each edge. The one with the highest edge weight is the</span>
    <span class="hljs-comment"># edge that the max alignment path goes through (if there's more than 1, it means there's more than 1 max alignment</span>
    <span class="hljs-comment"># path -- one is picked at random).</span>
    neighbours = []
    next_col_vals = sc.get_col(col + <span class="hljs-number">1</span>) <span class="hljs-keyword">if</span> col + <span class="hljs-number">1</span> &lt; v_node_count <span class="hljs-keyword">else</span> <span class="hljs-literal">None</span>  <span class="hljs-comment"># very quick due to prev call to get_col()</span>
    <span class="hljs-keyword">if</span> col + <span class="hljs-number">1</span> &lt; v_node_count:
        right_weight = next_col_vals[row]
        right_node = (col + <span class="hljs-number">1</span>, row)
        v_elem = v[col - <span class="hljs-number">1</span>]
        w_elem = <span class="hljs-literal">None</span>
        edge_weight = weight_lookup.lookup(v_elem, w_elem)
        neighbours += [(right_weight, edge_weight, right_node)]
    <span class="hljs-keyword">if</span> row + <span class="hljs-number">1</span> &lt; w_node_count:
        down_weight = col_vals[row + <span class="hljs-number">1</span>]
        down_node = (col, row + <span class="hljs-number">1</span>)
        v_elem = <span class="hljs-literal">None</span>
        w_elem = w[row - <span class="hljs-number">1</span>]
        edge_weight = weight_lookup.lookup(v_elem, w_elem)
        neighbours += [(down_weight, edge_weight, down_node)]
    <span class="hljs-keyword">if</span> col + <span class="hljs-number">1</span> &lt; v_node_count <span class="hljs-keyword">and</span> row + <span class="hljs-number">1</span> &lt; w_node_count:
        downright_weight = next_col_vals[row + <span class="hljs-number">1</span>]
        downright_node = (col + <span class="hljs-number">1</span>, row + <span class="hljs-number">1</span>)
        v_elem = v[col - <span class="hljs-number">1</span>]
        w_elem = w[row - <span class="hljs-number">1</span>]
        edge_weight = weight_lookup.lookup(v_elem, w_elem)
        neighbours += [(downright_weight, edge_weight, downright_node)]
    neighbours.sort(key=<span class="hljs-keyword">lambda</span> x: x[:<span class="hljs-number">2</span>])  <span class="hljs-comment"># sort by weight, then edge weight</span>
    _, _, (col2, row2) = neighbours[-<span class="hljs-number">1</span>]
    <span class="hljs-keyword">return</span> (col, row), (col2, row2)


<span class="hljs-keyword">def</span> <span class="hljs-title function_">find_edge_that_max_alignment_path_travels_through_at_middle_col</span>(<span class="hljs-params">
        v: <span class="hljs-type">List</span>[ELEM],
        w: <span class="hljs-type">List</span>[ELEM],
        weight_lookup: WeightLookup
</span>) -&gt; <span class="hljs-type">Tuple</span>[<span class="hljs-type">Tuple</span>[<span class="hljs-built_in">int</span>, <span class="hljs-built_in">int</span>], <span class="hljs-type">Tuple</span>[<span class="hljs-built_in">int</span>, <span class="hljs-built_in">int</span>]]:
    v_node_count = <span class="hljs-built_in">len</span>(v) + <span class="hljs-number">1</span>
    middle_col_idx = (v_node_count - <span class="hljs-number">1</span>) // <span class="hljs-number">2</span>
    <span class="hljs-keyword">return</span> find_edge_that_max_alignment_path_travels_through_at_col(v, w, weight_lookup, middle_col_idx)
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the sequences TACT and GACGT and the score matrix...</p>
<pre><code>INDEL=-1.0
   A  C  T  G
A  1  0  0  0
C  0  1  0  0
T  0  0  1  0
G  0  0  0  1

</code></pre>
<p>... a maximum alignment path is guaranteed to travel through the edge (3, 3), (3, 4).</p>
</div>
<p></p>
<p>The recursive sub-division process happens just as before, but this time with edges. Finding the maximum <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a> from edges provides two distinct advantages over the previous method of finding the maximum <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a> from nodes:</p>
<ul>
<li>
<p>Each sub-division results in one of the sub-graphs being smaller.</p>
<p><img src="svgbob_b2c41c83c7024803003127e32de4f185dec7bdbe.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>Since edges are being pulled out, the step that path finds between two <a href="#BM_(neighbour)%2Fi">neighbour</a>ing found nodes is no longer required. This is because sub-division of the <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a> happens on edges rather than nodes -- eventually all edges in the path will be walked as part of the recursive subdivision.</p>
</li>
</ul>
<p><a href="ch5_code/src/global_alignment/GlobalAlignment_DivideAndConquer_EdgeVariant.py">ch5_code/src/global_alignment/GlobalAlignment_DivideAndConquer_EdgeVariant.py</a> (lines 10 to 80):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">find_max_alignment_path_edges</span>(<span class="hljs-params">
        v: <span class="hljs-type">List</span>[ELEM],
        w: <span class="hljs-type">List</span>[ELEM],
        weight_lookup: WeightLookup,
        top: <span class="hljs-built_in">int</span>,
        bottom: <span class="hljs-built_in">int</span>,
        left: <span class="hljs-built_in">int</span>,
        right: <span class="hljs-built_in">int</span>,
        output: <span class="hljs-type">List</span>[<span class="hljs-built_in">str</span>]</span>):
    <span class="hljs-keyword">if</span> left == right:
        <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(top, bottom):
            output += [<span class="hljs-string">'↓'</span>]
        <span class="hljs-keyword">return</span>
    <span class="hljs-keyword">if</span> top == bottom:
        <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(left, right):
            output += [<span class="hljs-string">'→'</span>]
        <span class="hljs-keyword">return</span>

    (col1, row1), (col2, row2) = find_edge_that_max_alignment_path_travels_through_at_middle_col(v[left:right], w[top:bottom], weight_lookup)
    middle_col = left + col1
    middle_row = top + row1
    find_max_alignment_path_edges(v, w, weight_lookup, top, middle_row, left, middle_col, output)
    <span class="hljs-keyword">if</span> row1 + <span class="hljs-number">1</span> == row2 <span class="hljs-keyword">and</span> col1 + <span class="hljs-number">1</span> == col2:
        edge_dir = <span class="hljs-string">'↘'</span>
    <span class="hljs-keyword">elif</span> row1 == row2 <span class="hljs-keyword">and</span> col1 + <span class="hljs-number">1</span> == col2:
        edge_dir = <span class="hljs-string">'→'</span>
    <span class="hljs-keyword">elif</span> row1 + <span class="hljs-number">1</span> == row2 <span class="hljs-keyword">and</span> col1 == col2:
        edge_dir = <span class="hljs-string">'↓'</span>
    <span class="hljs-keyword">else</span>:
        <span class="hljs-keyword">raise</span> ValueError()
    <span class="hljs-keyword">if</span> edge_dir == <span class="hljs-string">'→'</span> <span class="hljs-keyword">or</span> edge_dir == <span class="hljs-string">'↘'</span>:
        middle_col += <span class="hljs-number">1</span>
    <span class="hljs-keyword">if</span> edge_dir == <span class="hljs-string">'↓'</span> <span class="hljs-keyword">or</span> edge_dir == <span class="hljs-string">'↘'</span>:
        middle_row += <span class="hljs-number">1</span>
    output += [edge_dir]
    find_max_alignment_path_edges(v, w, weight_lookup, middle_row, bottom, middle_col, right, output)


<span class="hljs-keyword">def</span> <span class="hljs-title function_">global_alignment</span>(<span class="hljs-params">
        v: <span class="hljs-type">List</span>[ELEM],
        w: <span class="hljs-type">List</span>[ELEM],
        weight_lookup: WeightLookup
</span>) -&gt; <span class="hljs-type">Tuple</span>[<span class="hljs-built_in">float</span>, <span class="hljs-type">List</span>[<span class="hljs-type">Tuple</span>[ELEM, ELEM]]]:
    edges = []
    find_max_alignment_path_edges(v, w, weight_lookup, <span class="hljs-number">0</span>, <span class="hljs-built_in">len</span>(w), <span class="hljs-number">0</span>, <span class="hljs-built_in">len</span>(v), edges)
    weight = <span class="hljs-number">0.0</span>
    alignment = []
    v_idx = <span class="hljs-number">0</span>
    w_idx = <span class="hljs-number">0</span>
    <span class="hljs-keyword">for</span> edge <span class="hljs-keyword">in</span> edges:
        <span class="hljs-keyword">if</span> edge == <span class="hljs-string">'→'</span>:
            v_elem = v[v_idx]
            w_elem = <span class="hljs-literal">None</span>
            alignment.append((v_elem, w_elem))
            weight += weight_lookup.lookup(v_elem, w_elem)
            v_idx += <span class="hljs-number">1</span>
        <span class="hljs-keyword">elif</span> edge == <span class="hljs-string">'↓'</span>:
            v_elem = <span class="hljs-literal">None</span>
            w_elem = w[w_idx]
            alignment.append((v_elem, w_elem))
            weight += weight_lookup.lookup(v_elem, w_elem)
            w_idx += <span class="hljs-number">1</span>
        <span class="hljs-keyword">elif</span> edge == <span class="hljs-string">'↘'</span>:
            v_elem = v[v_idx]
            w_elem = w[w_idx]
            alignment.append((v_elem, w_elem))
            weight += weight_lookup.lookup(v_elem, w_elem)
            v_idx += <span class="hljs-number">1</span>
            w_idx += <span class="hljs-number">1</span>
    <span class="hljs-keyword">return</span> weight, alignment
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the sequences TACT and GACGT and the score matrix...</p>
<pre><code>INDEL=-1.0
   A  C  T  G
A  1  0  0  0
C  0  1  0  0
T  0  0  1  0
G  0  0  0  1

</code></pre>
<p>... the global alignment is...</p>
<pre><code>TAC-T
GACGT
</code></pre>
<p>Weight: 2.0</p>
</div>
<p></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The other types of <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a> detailed in the sibling sections below don't implement a version of this algorithm. It's fairly straight forward to adapt this algorithm to support those <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a> types, but I didn't have the time to do it -- I almost completed a <a href="#BM_(local%20alignment%7Clocal%20sequence%20alignment)%2Fi">local alignment</a> version but backed out. The same high-level logic applies to those other alignment types: Converge on positions to find nodes/edges in the maximal <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a> and sub-divide on those positions.</p>
</div>
<a name="H_Fitting%20Alignment"></a><h3>Fitting Alignment</h3>
<p><a name="BM_(Algorithms%5C%2FSequence%20Alignment%2FFitting%20Alignment)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FFind%20Maximum%20Path%2FBacktrack%20Algorithm)_TOPIC%2F">Algorithms/Sequence Alignment/Find Maximum Path/Backtrack Algorithm</a></li>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FGlobal%20Alignment)_TOPIC%2F">Algorithms/Sequence Alignment/Global Alignment</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: Given two <a href="#BM_(sequence)%2Fi">sequence</a>s, for all possible <a href="#BM_(substring)%2Fi">substring</a>s of the first <a href="#BM_(sequence)%2Fi">sequence</a>, pull out the highest scoring alignment between that <a href="#BM_(substring)%2Fi">substring</a> that the second <a href="#BM_(sequence)%2Fi">sequence</a>.</p>
<p>In other words, find the <a href="#BM_(substring)%2Fi">substring</a> within the second <a href="#BM_(sequence)%2Fi">sequence</a> that produces the highest scoring alignment with the first <a href="#BM_(sequence)%2Fi">sequence</a>. For example, given the <a href="#BM_(sequence)%2Fi">sequence</a>s GGTTTTTAA and TTCTT, it may be that TTCTT (second <a href="#BM_(sequence)%2Fi">sequence</a>) has the highest scoring alignment with TTTTT (<a href="#BM_(substring)%2Fi">substring</a> of the first <a href="#BM_(sequence)%2Fi">sequence</a>)...</p>
<pre><code>TTC-TT
TT-TTT
</code></pre>
<p><strong>WHY</strong>: Searching for a <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a>'s <a href="#BM_(sequence)%2Fi">sequence</a> in some larger <a href="#BM_(genome)%2Fi">genome</a> may be problematic because of mutation. The <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a> <a href="#BM_(sequence)%2Fi">sequence</a> being searched for may be slightly off from the <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a> <a href="#BM_(sequence)%2Fi">sequence</a> in the <a href="#BM_(genome)%2Fi">genome</a>.</p>
<p>In the presence of minor mutations, a standard search will fail where a <a href="#BM_(fitting%20alignment%7Cfitting%20sequence%20alignment)%2Fi">fitting alignment</a> will still be able to find that <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a>.</p>
<a name="H1_Graph%20Algorithm"></a><h4>Graph Algorithm</h4>
<p><a name="BM_(Algorithms%5C%2FSequence%20Alignment%2FFitting%20Alignment%2FGraph%20Algorithm)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FFind%20Maximum%20Path)_TOPIC%2F">Algorithms/Sequence Alignment/Find Maximum Path</a></li>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FGlobal%20Alignment%2FGraph%20Algorithm)_TOPIC%2F">Algorithms/Sequence Alignment/Global Alignment/Graph Algorithm</a></li>
</ul>
</div>
<p>The graph algorithm for <a href="#BM_(fitting%20alignment%7Cfitting%20sequence%20alignment)%2Fi">fitting alignment</a> is an extension of the graph algorithm for <a href="#BM_(global%20alignment%7Cglobal%20sequence%20alignment)%2Fi">global alignment</a>. Construct the <a href="#BM_%5Cb(DAG)%5Cb%2F">DAG</a> as you would for <a href="#BM_(global%20alignment%7Cglobal%20sequence%20alignment)%2Fi">global alignment</a>, but for each node...</p>
<ul>
<li>in the first column that isn't the source node, construct a 0 weight edge from the source node to that node.</li>
<li>in the last column that isn't the sink node, construct a 0 weight edge from that node to the sink node.</li>
</ul>
<div style="border:1px solid black;">
<p></p>
<p><img src="latex_3df63dfc0d5f411eb21b54ea671cbe69.svg" alt="Latex diagram"></p>
<p>NOTE: Orange edges are "free rides" from source / Purple edges are "free rides" to sink.</p>
</div>
<p></p>
<p>These newly added edges represent hops in the graph -- 0 weight "free rides" to other nodes. The nodes at the destination of each one of these edges will never go below 0: When selecting a backtracking edge, the "free ride" edge will always be chosen over other edges that make the node weight negative.</p>
<p>When finding a maximum <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a>, these "free rides" make it so that the path ...</p>
<ul>
<li>starts from the most appropriate part of the second <a href="#BM_(sequence)%2Fi">sequence</a></li>
<li>stops at the most appropriate part of the second <a href="#BM_(sequence)%2Fi">sequence</a></li>
</ul>
<p>such that if the first <a href="#BM_(sequence)%2Fi">sequence</a> is wedged somewhere within the second <a href="#BM_(sequence)%2Fi">sequence</a>, that maximum <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a> will be targeted in such a way that it homes in on it.</p>
<p><a href="ch5_code/src/fitting_alignment/FittingAlignment_Graph.py">ch5_code/src/fitting_alignment/FittingAlignment_Graph.py</a> (lines 37 to 95):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">create_fitting_alignment_graph</span>(<span class="hljs-params">
        v: <span class="hljs-type">List</span>[ELEM],
        w: <span class="hljs-type">List</span>[ELEM],
        weight_lookup: WeightLookup
</span>) -&gt; Graph[<span class="hljs-type">Tuple</span>[<span class="hljs-built_in">int</span>, ...], NodeData, <span class="hljs-built_in">str</span>, EdgeData]:
    graph = create_grid_graph(
        [v, w],
        <span class="hljs-keyword">lambda</span> n_id: NodeData(),
        <span class="hljs-keyword">lambda</span> src_n_id, dst_n_id, offset, elems: EdgeData(elems[<span class="hljs-number">0</span>], elems[<span class="hljs-number">1</span>], weight_lookup.lookup(*elems))
    )
    v_node_count = <span class="hljs-built_in">len</span>(v) + <span class="hljs-number">1</span>
    w_node_count = <span class="hljs-built_in">len</span>(w) + <span class="hljs-number">1</span>
    source_node = <span class="hljs-number">0</span>, <span class="hljs-number">0</span>
    source_create_free_ride_edge_id_func = unique_id_generator(<span class="hljs-string">'FREE_RIDE_SOURCE'</span>)
    <span class="hljs-keyword">for</span> node <span class="hljs-keyword">in</span> product([<span class="hljs-number">0</span>], <span class="hljs-built_in">range</span>(w_node_count)):
        <span class="hljs-keyword">if</span> node == source_node:
            <span class="hljs-keyword">continue</span>
        e = source_create_free_ride_edge_id_func()
        graph.insert_edge(e, source_node, node, EdgeData(<span class="hljs-literal">None</span>, <span class="hljs-literal">None</span>, <span class="hljs-number">0.0</span>))
    sink_node = v_node_count - <span class="hljs-number">1</span>, w_node_count - <span class="hljs-number">1</span>
    sink_create_free_ride_edge_id_func = unique_id_generator(<span class="hljs-string">'FREE_RIDE_SINK'</span>)
    <span class="hljs-keyword">for</span> node <span class="hljs-keyword">in</span> product([v_node_count - <span class="hljs-number">1</span>], <span class="hljs-built_in">range</span>(w_node_count)):
        <span class="hljs-keyword">if</span> node == sink_node:
            <span class="hljs-keyword">continue</span>
        e = sink_create_free_ride_edge_id_func()
        graph.insert_edge(e, node, sink_node, EdgeData(<span class="hljs-literal">None</span>, <span class="hljs-literal">None</span>, <span class="hljs-number">0.0</span>))
    <span class="hljs-keyword">return</span> graph


<span class="hljs-keyword">def</span> <span class="hljs-title function_">fitting_alignment</span>(<span class="hljs-params">
        v: <span class="hljs-type">List</span>[ELEM],
        w: <span class="hljs-type">List</span>[ELEM],
        weight_lookup: WeightLookup
</span>) -&gt; <span class="hljs-type">Tuple</span>[<span class="hljs-built_in">float</span>, <span class="hljs-type">List</span>[<span class="hljs-built_in">str</span>], <span class="hljs-type">List</span>[<span class="hljs-type">Tuple</span>[ELEM, ELEM]]]:
    v_node_count = <span class="hljs-built_in">len</span>(v) + <span class="hljs-number">1</span>
    w_node_count = <span class="hljs-built_in">len</span>(w) + <span class="hljs-number">1</span>
    graph = create_fitting_alignment_graph(v, w, weight_lookup)
    from_node = (<span class="hljs-number">0</span>, <span class="hljs-number">0</span>)
    to_node = (v_node_count - <span class="hljs-number">1</span>, w_node_count - <span class="hljs-number">1</span>)
    populate_weights_and_backtrack_pointers(
        graph,
        from_node,
        <span class="hljs-keyword">lambda</span> n_id, weight, e_id: graph.get_node_data(n_id).set_weight_and_backtracking_edge(weight, e_id),
        <span class="hljs-keyword">lambda</span> n_id: graph.get_node_data(n_id).get_weight_and_backtracking_edge(),
        <span class="hljs-keyword">lambda</span> e_id: graph.get_edge_data(e_id).weight
    )
    final_weight = graph.get_node_data(to_node).weight
    edges = backtrack(
        graph,
        to_node,
        <span class="hljs-keyword">lambda</span> n_id: graph.get_node_data(n_id).get_weight_and_backtracking_edge()
    )
    alignment_edges = <span class="hljs-built_in">list</span>(<span class="hljs-built_in">filter</span>(<span class="hljs-keyword">lambda</span> e: <span class="hljs-keyword">not</span> e.startswith(<span class="hljs-string">'FREE_RIDE'</span>), edges))  <span class="hljs-comment"># remove free rides from list</span>
    alignment = []
    <span class="hljs-keyword">for</span> e <span class="hljs-keyword">in</span> alignment_edges:
        ed = graph.get_edge_data(e)
        alignment.append((ed.v_elem, ed.w_elem))
    <span class="hljs-keyword">return</span> final_weight, edges, alignment
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the sequences AGAC and TAAGAACT and the score matrix...</p>
<pre><code>INDEL=-1.0
    A   C   T   G
A   1  -1  -1  -1
C  -1   1  -1  -1
T  -1  -1   1  -1
G  -1  -1  -1   1

</code></pre>
<p>... the global alignment is...</p>
<p><img src="latex_abd17c5b8aa2a4703a255c00e6c578ac.svg" alt="Latex diagram"></p>
<pre><code>AG-AC
AGAAC
</code></pre>
<p>Weight: 3.0</p>
</div>
<p></p>
<a name="H1_Matrix%20Algorithm"></a><h4>Matrix Algorithm</h4>
<p><a name="BM_(Algorithms%5C%2FSequence%20Alignment%2FFitting%20Alignment%2FMatrix%20Algorithm)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FFitting%20Alignment%2FGraph%20Algorithm)_TOPIC%2F">Algorithms/Sequence Alignment/Fitting Alignment/Graph Algorithm</a></li>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FGlobal%20Alignment%2FMatrix%20Algorithm)_TOPIC%2F">Algorithms/Sequence Alignment/Global Alignment/Matrix Algorithm</a></li>
</ul>
</div>
<p><strong>ALGORITHM</strong>:</p>
<p>The following algorithm is an extension to <a href="#BM_(global%20alignment%7Cglobal%20sequence%20alignment)%2Fi">global alignment</a>'s matrix algorithm to properly account for the "free ride" edges required by <a href="#BM_(fitting%20alignment%7Cfitting%20sequence%20alignment)%2Fi">fitting alignment</a>. It's essentially the same as the graph algorithm, except that the implementation is much more sympathetic to modern hardware.</p>
<p><a href="ch5_code/src/fitting_alignment/FittingAlignment_Matrix.py">ch5_code/src/fitting_alignment/FittingAlignment_Matrix.py</a> (lines 10 to 93):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">backtrack</span>(<span class="hljs-params">
        node_matrix: <span class="hljs-type">List</span>[<span class="hljs-type">List</span>[<span class="hljs-type">Any</span>]]
</span>) -&gt; <span class="hljs-type">Tuple</span>[<span class="hljs-built_in">float</span>, <span class="hljs-type">List</span>[<span class="hljs-type">Tuple</span>[ELEM, ELEM]]]:
    v_node_idx = <span class="hljs-built_in">len</span>(node_matrix) - <span class="hljs-number">1</span>
    w_node_idx = <span class="hljs-built_in">len</span>(node_matrix[<span class="hljs-number">0</span>]) - <span class="hljs-number">1</span>
    final_weight = node_matrix[v_node_idx][w_node_idx][<span class="hljs-number">0</span>]
    alignment = []
    <span class="hljs-keyword">while</span> v_node_idx != <span class="hljs-number">0</span> <span class="hljs-keyword">or</span> w_node_idx != <span class="hljs-number">0</span>:
        _, elems, backtrack_ptr = node_matrix[v_node_idx][w_node_idx]
        <span class="hljs-keyword">if</span> backtrack_ptr == <span class="hljs-string">'↓'</span>:
            v_node_idx -= <span class="hljs-number">1</span>
            alignment.append(elems)
        <span class="hljs-keyword">elif</span> backtrack_ptr == <span class="hljs-string">'→'</span>:
            w_node_idx -= <span class="hljs-number">1</span>
            alignment.append(elems)
        <span class="hljs-keyword">elif</span> backtrack_ptr == <span class="hljs-string">'↘'</span>:
            v_node_idx -= <span class="hljs-number">1</span>
            w_node_idx -= <span class="hljs-number">1</span>
            alignment.append(elems)
        <span class="hljs-keyword">elif</span> <span class="hljs-built_in">isinstance</span>(backtrack_ptr, <span class="hljs-built_in">tuple</span>):
            v_node_idx = backtrack_ptr[<span class="hljs-number">0</span>]
            w_node_idx = backtrack_ptr[<span class="hljs-number">1</span>]
    <span class="hljs-keyword">return</span> final_weight, alignment[::-<span class="hljs-number">1</span>]


<span class="hljs-keyword">def</span> <span class="hljs-title function_">fitting_alignment</span>(<span class="hljs-params">
        v: <span class="hljs-type">List</span>[ELEM],
        w: <span class="hljs-type">List</span>[ELEM],
        weight_lookup: WeightLookup
</span>) -&gt; <span class="hljs-type">Tuple</span>[<span class="hljs-built_in">float</span>, <span class="hljs-type">List</span>[<span class="hljs-type">Tuple</span>[ELEM, ELEM]]]:
    v_node_count = <span class="hljs-built_in">len</span>(v) + <span class="hljs-number">1</span>
    w_node_count = <span class="hljs-built_in">len</span>(w) + <span class="hljs-number">1</span>
    node_matrix = []
    <span class="hljs-keyword">for</span> v_node_idx <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(v_node_count):
        row = []
        <span class="hljs-keyword">for</span> w_node_idx <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(w_node_count):
            row.append([-<span class="hljs-number">1.0</span>, (<span class="hljs-literal">None</span>, <span class="hljs-literal">None</span>), <span class="hljs-string">'?'</span>])
        node_matrix.append(row)
    node_matrix[<span class="hljs-number">0</span>][<span class="hljs-number">0</span>][<span class="hljs-number">0</span>] = <span class="hljs-number">0.0</span>           <span class="hljs-comment"># source node weight</span>
    node_matrix[<span class="hljs-number">0</span>][<span class="hljs-number">0</span>][<span class="hljs-number">1</span>] = (<span class="hljs-literal">None</span>, <span class="hljs-literal">None</span>)  <span class="hljs-comment"># source node elements (elements don't matter for source node)</span>
    node_matrix[<span class="hljs-number">0</span>][<span class="hljs-number">0</span>][<span class="hljs-number">2</span>] = <span class="hljs-string">'↘'</span>           <span class="hljs-comment"># source node backtracking edge (direction doesn't matter for source node)</span>
    <span class="hljs-keyword">for</span> v_node_idx, w_node_idx <span class="hljs-keyword">in</span> product(<span class="hljs-built_in">range</span>(v_node_count), <span class="hljs-built_in">range</span>(w_node_count)):
        parents = []
        <span class="hljs-keyword">if</span> v_node_idx &gt; <span class="hljs-number">0</span> <span class="hljs-keyword">and</span> w_node_idx &gt; <span class="hljs-number">0</span>:
            v_elem = v[v_node_idx - <span class="hljs-number">1</span>]
            w_elem = w[w_node_idx - <span class="hljs-number">1</span>]
            parents.append([
                node_matrix[v_node_idx - <span class="hljs-number">1</span>][w_node_idx - <span class="hljs-number">1</span>][<span class="hljs-number">0</span>] + weight_lookup.lookup(v_elem, w_elem),
                (v_elem, w_elem),
                <span class="hljs-string">'↘'</span>
            ])
        <span class="hljs-keyword">if</span> v_node_idx &gt; <span class="hljs-number">0</span>:
            v_elem = v[v_node_idx - <span class="hljs-number">1</span>]
            parents.append([
                node_matrix[v_node_idx - <span class="hljs-number">1</span>][w_node_idx][<span class="hljs-number">0</span>] + weight_lookup.lookup(v_elem, <span class="hljs-literal">None</span>),
                (v_elem, <span class="hljs-literal">None</span>),
                <span class="hljs-string">'↓'</span>
            ])
        <span class="hljs-keyword">if</span> w_node_idx &gt; <span class="hljs-number">0</span>:
            w_elem = w[w_node_idx - <span class="hljs-number">1</span>]
            parents.append([
                node_matrix[v_node_idx][w_node_idx - <span class="hljs-number">1</span>][<span class="hljs-number">0</span>] + weight_lookup.lookup(<span class="hljs-literal">None</span>, w_elem),
                (<span class="hljs-literal">None</span>, w_elem),
                <span class="hljs-string">'→'</span>
            ])
        <span class="hljs-comment"># If first column but not source node, consider free-ride from source node</span>
        <span class="hljs-keyword">if</span> v_node_idx == <span class="hljs-number">0</span> <span class="hljs-keyword">and</span> w_node_idx != <span class="hljs-number">0</span>:
            parents.append([
                <span class="hljs-number">0.0</span>,
                (<span class="hljs-literal">None</span>, <span class="hljs-literal">None</span>),
                (<span class="hljs-number">0</span>, <span class="hljs-number">0</span>)  <span class="hljs-comment"># jump to source</span>
            ])
        <span class="hljs-comment"># If sink node, consider free-rides coming from every node in last column that isn't sink node</span>
        <span class="hljs-keyword">if</span> v_node_idx == v_node_count - <span class="hljs-number">1</span> <span class="hljs-keyword">and</span> w_node_idx == w_node_count - <span class="hljs-number">1</span>:
            <span class="hljs-keyword">for</span> w_node_idx_from <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(w_node_count - <span class="hljs-number">1</span>):
                parents.append([
                    node_matrix[v_node_idx][w_node_idx_from][<span class="hljs-number">0</span>] + <span class="hljs-number">0.0</span>,
                    (<span class="hljs-literal">None</span>, <span class="hljs-literal">None</span>),
                    (v_node_idx, w_node_idx_from)  <span class="hljs-comment"># jump to this position</span>
                ])
        <span class="hljs-keyword">if</span> parents:  <span class="hljs-comment"># parents will be empty if v_node_idx and w_node_idx were both 0</span>
            node_matrix[v_node_idx][w_node_idx] = <span class="hljs-built_in">max</span>(parents, key=<span class="hljs-keyword">lambda</span> x: x[<span class="hljs-number">0</span>])
    <span class="hljs-keyword">return</span> backtrack(node_matrix)
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the sequences AGAC and TAAGAACT and the score matrix...</p>
<pre><code>INDEL=-1.0
    A   C   T   G
A   1  -1  -1  -1
C  -1   1  -1  -1
T  -1  -1   1  -1
G  -1  -1  -1   1

</code></pre>
<p>... the fitting alignment is...</p>
<pre><code>AG-AC
AGAAC
</code></pre>
<p>Weight: 3.0</p>
</div>
<p></p>
<a name="H_Overlap%20Alignment"></a><h3>Overlap Alignment</h3>
<p><a name="BM_(Algorithms%5C%2FSequence%20Alignment%2FOverlap%20Alignment)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FFind%20Maximum%20Path%2FBacktrack%20Algorithm)_TOPIC%2F">Algorithms/Sequence Alignment/Find Maximum Path/Backtrack Algorithm</a></li>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FGlobal%20Alignment)_TOPIC%2F">Algorithms/Sequence Alignment/Global Alignment</a></li>
<li><a href="#BM_(Algorithms%5C%2FDNA%20Assembly)_TOPIC%2F">Algorithms/DNA Assembly</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: Given two <a href="#BM_(sequence)%2Fi">sequence</a>s, for all possible <a href="#BM_(substring)%2Fi">substring</a>s that ...</p>
<ul>
<li>end at the first <a href="#BM_(sequence)%2Fi">sequence</a> (tail)</li>
<li>start at the second <a href="#BM_(sequence)%2Fi">sequence</a> (head)</li>
</ul>
<p>... , pull out the highest scoring alignment.</p>
<p>In other words, find the overlap between the two <a href="#BM_(sequence)%2Fi">sequence</a>s that produces the highest scoring alignment. For example, given the <a href="#BM_(sequence)%2Fi">sequence</a>s CCAAGGCT and GGTTTTTAA, it may be that the <a href="#BM_(substring)%2Fi">substring</a>s with the highest scoring alignment are GGCT (tail of the first <a href="#BM_(sequence)%2Fi">sequence</a>) and GGT (head of the second <a href="#BM_(sequence)%2Fi">sequence</a>)...</p>
<pre><code>GGCT
GG-T
</code></pre>
<p><strong>WHY</strong>: DNA <a href="#BM_(sequencer)%2Fi">sequencer</a>s frequently produce <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s with <a href="#BM_(sequencing%20error)%2Fi">sequencing error</a>s. <a href="#BM_(overlap%20alignment%7Coverlap%20sequence%20alignment)%2Fi">Overlap alignment</a>s may be used to detect if those <a href="#BM_(fragment)_SEQ%2Fi">fragment</a> overlap even in the presence of <a href="#BM_(sequencing%20error)%2Fi">sequencing error</a>s and minor mutations, making <a href="#BM_(assembly%7Cassemble)%2Fi">assembly</a> less tedious (<a href="#BM_(overlap%20graph)%2Fi">overlap graph</a>s / <a href="#BM_(de%20Bruijn%20graph)%2Fi">de Bruijn graph</a>s may turn out less tangled).</p>
<a name="H2_Graph%20Algorithm"></a><h4>Graph Algorithm</h4>
<p><a name="BM_(Algorithms%5C%2FSequence%20Alignment%2FOverlap%20Alignment%2FGraph%20Algorithm)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FFind%20Maximum%20Path)_TOPIC%2F">Algorithms/Sequence Alignment/Find Maximum Path</a></li>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FGlobal%20Alignment%2FGraph%20Algorithm)_TOPIC%2F">Algorithms/Sequence Alignment/Global Alignment/Graph Algorithm</a></li>
</ul>
</div>
<p>The graph algorithm for <a href="#BM_(overlap%20alignment%7Coverlap%20sequence%20alignment)%2Fi">overlap alignment</a> is an extension of the graph algorithm for <a href="#BM_(global%20alignment%7Cglobal%20sequence%20alignment)%2Fi">global alignment</a>. Construct the <a href="#BM_%5Cb(DAG)%5Cb%2F">DAG</a> as you would for <a href="#BM_(global%20alignment%7Cglobal%20sequence%20alignment)%2Fi">global alignment</a>, but for each node...</p>
<ul>
<li>in the first column that isn't the source node, construct a 0 weight edge from the source node to that node.</li>
<li>in the last row that isn't the sink node, construct a 0 weight edge from that node to the sink node.</li>
</ul>
<div style="border:1px solid black;">
<p></p>
<p><img src="latex_f7218e70493ad6c1fdceb71dda5eb519.svg" alt="Latex diagram"></p>
<p>NOTE: Orange edges are "free rides" from source / Purple edges are "free rides" to sink.</p>
</div>
<p></p>
<p>These newly added edges represent hops in the graph -- 0 weight "free rides" to other nodes. The nodes at the destination of each one of these edges will never go below 0: When selecting a backtracking edge, the "free ride" edge will always be chosen over other edges that make the node weight negative.</p>
<p>When finding a maximum <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a>, these "free rides" make it so that the path ...</p>
<ul>
<li>starts from the most appropriate part of the second <a href="#BM_(sequence)%2Fi">sequence</a></li>
<li>stops at the most appropriate part of the first <a href="#BM_(sequence)%2Fi">sequence</a></li>
</ul>
<p>such that if there is a matching overlap between the <a href="#BM_(sequence)%2Fi">sequence</a>s, that maximum <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a> will be targeted in such a way that maximizes that overlap.</p>
<p><a href="ch5_code/src/overlap_alignment/OverlapAlignment_Graph.py">ch5_code/src/overlap_alignment/OverlapAlignment_Graph.py</a> (lines 37 to 95):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">create_overlap_alignment_graph</span>(<span class="hljs-params">
        v: <span class="hljs-type">List</span>[ELEM],
        w: <span class="hljs-type">List</span>[ELEM],
        weight_lookup: WeightLookup
</span>) -&gt; Graph[<span class="hljs-type">Tuple</span>[<span class="hljs-built_in">int</span>, ...], NodeData, <span class="hljs-built_in">str</span>, EdgeData]:
    graph = create_grid_graph(
        [v, w],
        <span class="hljs-keyword">lambda</span> n_id: NodeData(),
        <span class="hljs-keyword">lambda</span> src_n_id, dst_n_id, offset, elems: EdgeData(elems[<span class="hljs-number">0</span>], elems[<span class="hljs-number">1</span>], weight_lookup.lookup(*elems))
    )
    v_node_count = <span class="hljs-built_in">len</span>(v) + <span class="hljs-number">1</span>
    w_node_count = <span class="hljs-built_in">len</span>(w) + <span class="hljs-number">1</span>
    source_node = <span class="hljs-number">0</span>, <span class="hljs-number">0</span>
    source_create_free_ride_edge_id_func = unique_id_generator(<span class="hljs-string">'FREE_RIDE_SOURCE'</span>)
    <span class="hljs-keyword">for</span> node <span class="hljs-keyword">in</span> product([<span class="hljs-number">0</span>], <span class="hljs-built_in">range</span>(w_node_count)):
        <span class="hljs-keyword">if</span> node == source_node:
            <span class="hljs-keyword">continue</span>
        e = source_create_free_ride_edge_id_func()
        graph.insert_edge(e, source_node, node, EdgeData(<span class="hljs-literal">None</span>, <span class="hljs-literal">None</span>, <span class="hljs-number">0.0</span>))
    sink_node = v_node_count - <span class="hljs-number">1</span>, w_node_count - <span class="hljs-number">1</span>
    sink_create_free_ride_edge_id_func = unique_id_generator(<span class="hljs-string">'FREE_RIDE_SINK'</span>)
    <span class="hljs-keyword">for</span> node <span class="hljs-keyword">in</span> product(<span class="hljs-built_in">range</span>(v_node_count), [w_node_count - <span class="hljs-number">1</span>]):
        <span class="hljs-keyword">if</span> node == sink_node:
            <span class="hljs-keyword">continue</span>
        e = sink_create_free_ride_edge_id_func()
        graph.insert_edge(e, node, sink_node, EdgeData(<span class="hljs-literal">None</span>, <span class="hljs-literal">None</span>, <span class="hljs-number">0.0</span>))
    <span class="hljs-keyword">return</span> graph


<span class="hljs-keyword">def</span> <span class="hljs-title function_">overlap_alignment</span>(<span class="hljs-params">
        v: <span class="hljs-type">List</span>[ELEM],
        w: <span class="hljs-type">List</span>[ELEM],
        weight_lookup: WeightLookup
</span>) -&gt; <span class="hljs-type">Tuple</span>[<span class="hljs-built_in">float</span>, <span class="hljs-type">List</span>[<span class="hljs-built_in">str</span>], <span class="hljs-type">List</span>[<span class="hljs-type">Tuple</span>[ELEM, ELEM]]]:
    v_node_count = <span class="hljs-built_in">len</span>(v) + <span class="hljs-number">1</span>
    w_node_count = <span class="hljs-built_in">len</span>(w) + <span class="hljs-number">1</span>
    graph = create_overlap_alignment_graph(v, w, weight_lookup)
    from_node = (<span class="hljs-number">0</span>, <span class="hljs-number">0</span>)
    to_node = (v_node_count - <span class="hljs-number">1</span>, w_node_count - <span class="hljs-number">1</span>)
    populate_weights_and_backtrack_pointers(
        graph,
        from_node,
        <span class="hljs-keyword">lambda</span> n_id, weight, e_id: graph.get_node_data(n_id).set_weight_and_backtracking_edge(weight, e_id),
        <span class="hljs-keyword">lambda</span> n_id: graph.get_node_data(n_id).get_weight_and_backtracking_edge(),
        <span class="hljs-keyword">lambda</span> e_id: graph.get_edge_data(e_id).weight
    )
    final_weight = graph.get_node_data(to_node).weight
    edges = backtrack(
        graph,
        to_node,
        <span class="hljs-keyword">lambda</span> n_id: graph.get_node_data(n_id).get_weight_and_backtracking_edge()
    )
    alignment_edges = <span class="hljs-built_in">list</span>(<span class="hljs-built_in">filter</span>(<span class="hljs-keyword">lambda</span> e: <span class="hljs-keyword">not</span> e.startswith(<span class="hljs-string">'FREE_RIDE'</span>), edges))  <span class="hljs-comment"># remove free rides from list</span>
    alignment = []
    <span class="hljs-keyword">for</span> e <span class="hljs-keyword">in</span> alignment_edges:
        ed = graph.get_edge_data(e)
        alignment.append((ed.v_elem, ed.w_elem))
    <span class="hljs-keyword">return</span> final_weight, edges, alignment
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the sequences AGACAAAT and GGGGAAAC and the score matrix...</p>
<pre><code>INDEL=-1.0
    A   C   T   G
A   1  -1  -1  -1
C  -1   1  -1  -1
T  -1  -1   1  -1
G  -1  -1  -1   1

</code></pre>
<p>... the global alignment is...</p>
<p><img src="latex_734ca090b4f68670a15f7449773b9dc5.svg" alt="Latex diagram"></p>
<pre><code>AGAC
A-AC
</code></pre>
<p>Weight: 2.0</p>
</div>
<p></p>
<a name="H2_Matrix%20Algorithm"></a><h4>Matrix Algorithm</h4>
<p><a name="BM_(Algorithms%5C%2FSequence%20Alignment%2FOverlap%20Alignment%2FMatrix%20Algorithm)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FOverlap%20Alignment%2FGraph%20Algorithm)_TOPIC%2F">Algorithms/Sequence Alignment/Overlap Alignment/Graph Algorithm</a></li>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FGlobal%20Alignment%2FMatrix%20Algorithm)_TOPIC%2F">Algorithms/Sequence Alignment/Global Alignment/Matrix Algorithm</a></li>
</ul>
</div>
<p><strong>ALGORITHM</strong>:</p>
<p>The following algorithm is an extension to <a href="#BM_(global%20alignment%7Cglobal%20sequence%20alignment)%2Fi">global alignment</a>'s matrix algorithm to properly account for the "free ride" edges required by <a href="#BM_(overlap%20alignment%7Coverlap%20sequence%20alignment)%2Fi">overlap alignment</a>. It's essentially the same as the graph algorithm, except that the implementation is much more sympathetic to modern hardware.</p>
<p><a href="ch5_code/src/overlap_alignment/OverlapAlignment_Matrix.py">ch5_code/src/overlap_alignment/OverlapAlignment_Matrix.py</a> (lines 10 to 93):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">backtrack</span>(<span class="hljs-params">
        node_matrix: <span class="hljs-type">List</span>[<span class="hljs-type">List</span>[<span class="hljs-type">Any</span>]]
</span>) -&gt; <span class="hljs-type">Tuple</span>[<span class="hljs-built_in">float</span>, <span class="hljs-type">List</span>[<span class="hljs-type">Tuple</span>[ELEM, ELEM]]]:
    v_node_idx = <span class="hljs-built_in">len</span>(node_matrix) - <span class="hljs-number">1</span>
    w_node_idx = <span class="hljs-built_in">len</span>(node_matrix[<span class="hljs-number">0</span>]) - <span class="hljs-number">1</span>
    final_weight = node_matrix[v_node_idx][w_node_idx][<span class="hljs-number">0</span>]
    alignment = []
    <span class="hljs-keyword">while</span> v_node_idx != <span class="hljs-number">0</span> <span class="hljs-keyword">or</span> w_node_idx != <span class="hljs-number">0</span>:
        _, elems, backtrack_ptr = node_matrix[v_node_idx][w_node_idx]
        <span class="hljs-keyword">if</span> backtrack_ptr == <span class="hljs-string">'↓'</span>:
            v_node_idx -= <span class="hljs-number">1</span>
            alignment.append(elems)
        <span class="hljs-keyword">elif</span> backtrack_ptr == <span class="hljs-string">'→'</span>:
            w_node_idx -= <span class="hljs-number">1</span>
            alignment.append(elems)
        <span class="hljs-keyword">elif</span> backtrack_ptr == <span class="hljs-string">'↘'</span>:
            v_node_idx -= <span class="hljs-number">1</span>
            w_node_idx -= <span class="hljs-number">1</span>
            alignment.append(elems)
        <span class="hljs-keyword">elif</span> <span class="hljs-built_in">isinstance</span>(backtrack_ptr, <span class="hljs-built_in">tuple</span>):
            v_node_idx = backtrack_ptr[<span class="hljs-number">0</span>]
            w_node_idx = backtrack_ptr[<span class="hljs-number">1</span>]
    <span class="hljs-keyword">return</span> final_weight, alignment[::-<span class="hljs-number">1</span>]


<span class="hljs-keyword">def</span> <span class="hljs-title function_">overlap_alignment</span>(<span class="hljs-params">
        v: <span class="hljs-type">List</span>[ELEM],
        w: <span class="hljs-type">List</span>[ELEM],
        weight_lookup: WeightLookup
</span>) -&gt; <span class="hljs-type">Tuple</span>[<span class="hljs-built_in">float</span>, <span class="hljs-type">List</span>[<span class="hljs-type">Tuple</span>[ELEM, ELEM]]]:
    v_node_count = <span class="hljs-built_in">len</span>(v) + <span class="hljs-number">1</span>
    w_node_count = <span class="hljs-built_in">len</span>(w) + <span class="hljs-number">1</span>
    node_matrix = []
    <span class="hljs-keyword">for</span> v_node_idx <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(v_node_count):
        row = []
        <span class="hljs-keyword">for</span> w_node_idx <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(w_node_count):
            row.append([-<span class="hljs-number">1.0</span>, (<span class="hljs-literal">None</span>, <span class="hljs-literal">None</span>), <span class="hljs-string">'?'</span>])
        node_matrix.append(row)
    node_matrix[<span class="hljs-number">0</span>][<span class="hljs-number">0</span>][<span class="hljs-number">0</span>] = <span class="hljs-number">0.0</span>           <span class="hljs-comment"># source node weight</span>
    node_matrix[<span class="hljs-number">0</span>][<span class="hljs-number">0</span>][<span class="hljs-number">1</span>] = (<span class="hljs-literal">None</span>, <span class="hljs-literal">None</span>)  <span class="hljs-comment"># source node elements (elements don't matter for source node)</span>
    node_matrix[<span class="hljs-number">0</span>][<span class="hljs-number">0</span>][<span class="hljs-number">2</span>] = <span class="hljs-string">'↘'</span>           <span class="hljs-comment"># source node backtracking edge (direction doesn't matter for source node)</span>
    <span class="hljs-keyword">for</span> v_node_idx, w_node_idx <span class="hljs-keyword">in</span> product(<span class="hljs-built_in">range</span>(v_node_count), <span class="hljs-built_in">range</span>(w_node_count)):
        parents = []
        <span class="hljs-keyword">if</span> v_node_idx &gt; <span class="hljs-number">0</span> <span class="hljs-keyword">and</span> w_node_idx &gt; <span class="hljs-number">0</span>:
            v_elem = v[v_node_idx - <span class="hljs-number">1</span>]
            w_elem = w[w_node_idx - <span class="hljs-number">1</span>]
            parents.append([
                node_matrix[v_node_idx - <span class="hljs-number">1</span>][w_node_idx - <span class="hljs-number">1</span>][<span class="hljs-number">0</span>] + weight_lookup.lookup(v_elem, w_elem),
                (v_elem, w_elem),
                <span class="hljs-string">'↘'</span>
            ])
        <span class="hljs-keyword">if</span> v_node_idx &gt; <span class="hljs-number">0</span>:
            v_elem = v[v_node_idx - <span class="hljs-number">1</span>]
            parents.append([
                node_matrix[v_node_idx - <span class="hljs-number">1</span>][w_node_idx][<span class="hljs-number">0</span>] + weight_lookup.lookup(v_elem, <span class="hljs-literal">None</span>),
                (v_elem, <span class="hljs-literal">None</span>),
                <span class="hljs-string">'↓'</span>
            ])
        <span class="hljs-keyword">if</span> w_node_idx &gt; <span class="hljs-number">0</span>:
            w_elem = w[w_node_idx - <span class="hljs-number">1</span>]
            parents.append([
                node_matrix[v_node_idx][w_node_idx - <span class="hljs-number">1</span>][<span class="hljs-number">0</span>] + weight_lookup.lookup(<span class="hljs-literal">None</span>, w_elem),
                (<span class="hljs-literal">None</span>, w_elem),
                <span class="hljs-string">'→'</span>
            ])
        <span class="hljs-comment"># If first column but not source node, consider free-ride from source node</span>
        <span class="hljs-keyword">if</span> v_node_idx == <span class="hljs-number">0</span> <span class="hljs-keyword">and</span> w_node_idx != <span class="hljs-number">0</span>:
            parents.append([
                <span class="hljs-number">0.0</span>,
                (<span class="hljs-literal">None</span>, <span class="hljs-literal">None</span>),
                (<span class="hljs-number">0</span>, <span class="hljs-number">0</span>)  <span class="hljs-comment"># jump to source</span>
            ])
        <span class="hljs-comment"># If sink node, consider free-rides coming from every node in last row that isn't sink node</span>
        <span class="hljs-keyword">if</span> v_node_idx == v_node_count - <span class="hljs-number">1</span> <span class="hljs-keyword">and</span> w_node_idx == w_node_count - <span class="hljs-number">1</span>:
            <span class="hljs-keyword">for</span> v_node_idx_from <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(v_node_count - <span class="hljs-number">1</span>):
                parents.append([
                    node_matrix[v_node_idx_from][w_node_idx][<span class="hljs-number">0</span>] + <span class="hljs-number">0.0</span>,
                    (<span class="hljs-literal">None</span>, <span class="hljs-literal">None</span>),
                    (v_node_idx_from, w_node_idx)  <span class="hljs-comment"># jump to this position</span>
                ])
        <span class="hljs-keyword">if</span> parents:  <span class="hljs-comment"># parents will be empty if v_node_idx and w_node_idx were both 0</span>
            node_matrix[v_node_idx][w_node_idx] = <span class="hljs-built_in">max</span>(parents, key=<span class="hljs-keyword">lambda</span> x: x[<span class="hljs-number">0</span>])
    <span class="hljs-keyword">return</span> backtrack(node_matrix)
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the sequences AGACAAAT and GGGGAAAC and the score matrix...</p>
<pre><code>INDEL=-1.0
    A   C   T   G
A   1  -1  -1  -1
C  -1   1  -1  -1
T  -1  -1   1  -1
G  -1  -1  -1   1

</code></pre>
<p>... the overlap alignment is...</p>
<pre><code>AGAC
AAAC
</code></pre>
<p>Weight: 2.0</p>
</div>
<p></p>
<a name="H_Local%20Alignment"></a><h3>Local Alignment</h3>
<p><a name="BM_(Algorithms%5C%2FSequence%20Alignment%2FLocal%20Alignment)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FFind%20Maximum%20Path%2FBacktrack%20Algorithm)_TOPIC%2F">Algorithms/Sequence Alignment/Find Maximum Path/Backtrack Algorithm</a></li>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FGlobal%20Alignment)_TOPIC%2F">Algorithms/Sequence Alignment/Global Alignment</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: Given two <a href="#BM_(sequence)%2Fi">sequence</a>s, for all possible <a href="#BM_(substring)%2Fi">substring</a>s of those <a href="#BM_(sequence)%2Fi">sequence</a>s, pull out the highest scoring alignment. For example, given the <a href="#BM_(sequence)%2Fi">sequence</a>s GGTTTTTAA and CCTTCTTAA, it may be that the <a href="#BM_(substring)%2Fi">substring</a>s with the highest scoring alignment are TTTTT (<a href="#BM_(substring)%2Fi">substring</a> of first <a href="#BM_(sequence)%2Fi">sequence</a>) and TTCTT (<a href="#BM_(substring)%2Fi">substring</a> of second <a href="#BM_(sequence)%2Fi">sequence</a>) ...</p>
<pre><code>TTC-TT
TT-TTT
</code></pre>
<p><strong>WHY</strong>: Two biological <a href="#BM_(sequence)%2Fi">sequence</a>s may have strongly related parts rather than being strongly related in their entirety. For example, a class of proteins called <a href="#BM_(non-ribosomal%20peptide%20synthetase%7CNRP%20synthetase)%2Fi">NRP synthetase</a> creates <a href="#BM_(peptide)%2Fi">peptide</a>s without going through a ribosome (<a href="#BM_(non-ribosomal%20peptide)%2Fi">non-ribosomal peptide</a>s). Each <a href="#BM_(non-ribosomal%20peptide%20synthetase%7CNRP%20synthetase)%2Fi">NRP synthetase</a> outputs a specific <a href="#BM_(peptide)%2Fi">peptide</a>, where each <a href="#BM_(amino%20acid)%2Fi">amino acid</a> in that <a href="#BM_(peptide)%2Fi">peptide</a> is pumped out by the unique part of the <a href="#BM_(non-ribosomal%20peptide%20synthetase%7CNRP%20synthetase)%2Fi">NRP synthetase</a> responsible for it.</p>
<p>These unique parts are referred to <a href="#BM_(adenylation%20domain%7CA%20domain%7CA-domain)%2Fi">adenylation domain</a>s (multiple <a href="#BM_(adenylation%20domain%7CA%20domain%7CA-domain)%2Fi">adenylation domain</a>s, 1 per <a href="#BM_(amino%20acid)%2Fi">amino acid</a> in created <a href="#BM_(peptide)%2Fi">peptide</a>). While the overall <a href="#BM_(sequence)%2Fi">sequence</a> between two types of <a href="#BM_(non-ribosomal%20peptide%20synthetase%7CNRP%20synthetase)%2Fi">NRP synthetase</a> differ greatly, the <a href="#BM_(sequence)%2Fi">sequence</a>s between their <a href="#BM_(adenylation%20domain%7CA%20domain%7CA-domain)%2Fi">adenylation domain</a>s are similar -- only a handful of positions in an <a href="#BM_(adenylation%20domain%7CA%20domain%7CA-domain)%2Fi">adenylation domain</a> <a href="#BM_(sequence)%2Fi">sequence</a> define the type of <a href="#BM_(amino%20acid)%2Fi">amino acid</a> it pumps out. As such, <a href="#BM_(local%20alignment%7Clocal%20sequence%20alignment)%2Fi">local alignment</a> may be used to identify these <a href="#BM_(adenylation%20domain%7CA%20domain%7CA-domain)%2Fi">adenylation domain</a>s across different types of <a href="#BM_(non-ribosomal%20peptide%20synthetase%7CNRP%20synthetase)%2Fi">NRP synthetase</a>.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The WHY section above is giving a high-level use-case for <a href="#BM_(local%20alignment%7Clocal%20sequence%20alignment)%2Fi">local alignment</a>. If you actually want to perform that use-case you need to get familiar with the protein scoring section: <a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FProtein%20Scoring)_TOPIC%2F">Algorithms/Sequence Alignment/Protein Scoring</a>.</p>
</div>
<a name="H3_Graph%20Algorithm"></a><h4>Graph Algorithm</h4>
<p><a name="BM_(Algorithms%5C%2FSequence%20Alignment%2FLocal%20Alignment%2FGraph%20Algorithm)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FFind%20Maximum%20Path)_TOPIC%2F">Algorithms/Sequence Alignment/Find Maximum Path</a></li>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FGlobal%20Alignment%2FGraph%20Algorithm)_TOPIC%2F">Algorithms/Sequence Alignment/Global Alignment/Graph Algorithm</a></li>
</ul>
</div>
<p><strong>ALGORITHM</strong>:</p>
<p>The graph algorithm for <a href="#BM_(local%20alignment%7Clocal%20sequence%20alignment)%2Fi">local alignment</a> is an extension of the graph algorithm for <a href="#BM_(global%20alignment%7Cglobal%20sequence%20alignment)%2Fi">global alignment</a>. Construct the <a href="#BM_%5Cb(DAG)%5Cb%2F">DAG</a> as you would for <a href="#BM_(global%20alignment%7Cglobal%20sequence%20alignment)%2Fi">global alignment</a>, but for each node...</p>
<ul>
<li>that isn't the source node, construct a 0 weight edge from the source node to that node.</li>
<li>that isn't the sink node, construct a 0 weight edge from that node to the sink node.</li>
</ul>
<div style="border:1px solid black;">
<p></p>
<p><img src="latex_aeacfe8c6a66ecab6242374110af4cc7.svg" alt="Latex diagram"></p>
<p>NOTE: Orange edges are "free rides" from source / Purple edges are "free rides" to sink.</p>
</div>
<p></p>
<p>These newly added edges represent hops in the graph -- 0 weight "free rides" to other nodes. The nodes at the destination of each one of these edges will never go below 0: When selecting a backtracking edge, the "free ride" edge will always be chosen over other edges that make the node weight negative.</p>
<p>When finding a maximum <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a>, these "free rides" make it so that if either the...</p>
<ul>
<li>prefix of the path sinks the path's weight below zero, there's a "free ride" from the source node that'll supersede it and as such that prefix will get skipped.</li>
<li>suffix of the path sinks the path's weight below zero, there's a "free ride" to the sink node that'll supersede it and as such that suffix will get skipped.</li>
</ul>
<p>The maximum <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a> will be targeted in such a way that it homes on the <a href="#BM_(substring)%2Fi">substring</a> within each <a href="#BM_(sequence)%2Fi">sequence</a> that produces the highest scoring alignment.</p>
<p><a href="ch5_code/src/local_alignment/LocalAlignment_Graph.py">ch5_code/src/local_alignment/LocalAlignment_Graph.py</a> (lines 38 to 96):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">create_local_alignment_graph</span>(<span class="hljs-params">
        v: <span class="hljs-type">List</span>[ELEM],
        w: <span class="hljs-type">List</span>[ELEM],
        weight_lookup: WeightLookup
</span>) -&gt; Graph[<span class="hljs-type">Tuple</span>[<span class="hljs-built_in">int</span>, ...], NodeData, <span class="hljs-built_in">str</span>, EdgeData]:
    graph = create_grid_graph(
        [v, w],
        <span class="hljs-keyword">lambda</span> n_id: NodeData(),
        <span class="hljs-keyword">lambda</span> src_n_id, dst_n_id, offset, elems: EdgeData(elems[<span class="hljs-number">0</span>], elems[<span class="hljs-number">1</span>], weight_lookup.lookup(*elems))
    )
    v_node_count = <span class="hljs-built_in">len</span>(v) + <span class="hljs-number">1</span>
    w_node_count = <span class="hljs-built_in">len</span>(w) + <span class="hljs-number">1</span>
    source_node = <span class="hljs-number">0</span>, <span class="hljs-number">0</span>
    source_create_free_ride_edge_id_func = unique_id_generator(<span class="hljs-string">'FREE_RIDE_SOURCE'</span>)
    <span class="hljs-keyword">for</span> node <span class="hljs-keyword">in</span> product(<span class="hljs-built_in">range</span>(v_node_count), <span class="hljs-built_in">range</span>(w_node_count)):
        <span class="hljs-keyword">if</span> node == source_node:
            <span class="hljs-keyword">continue</span>
        e = source_create_free_ride_edge_id_func()
        graph.insert_edge(e, source_node, node, EdgeData(<span class="hljs-literal">None</span>, <span class="hljs-literal">None</span>, <span class="hljs-number">0.0</span>))
    sink_node = v_node_count - <span class="hljs-number">1</span>, w_node_count - <span class="hljs-number">1</span>
    sink_create_free_ride_edge_id_func = unique_id_generator(<span class="hljs-string">'FREE_RIDE_SINK'</span>)
    <span class="hljs-keyword">for</span> node <span class="hljs-keyword">in</span> product(<span class="hljs-built_in">range</span>(v_node_count), <span class="hljs-built_in">range</span>(w_node_count)):
        <span class="hljs-keyword">if</span> node == sink_node:
            <span class="hljs-keyword">continue</span>
        e = sink_create_free_ride_edge_id_func()
        graph.insert_edge(e, node, sink_node, EdgeData(<span class="hljs-literal">None</span>, <span class="hljs-literal">None</span>, <span class="hljs-number">0.0</span>))
    <span class="hljs-keyword">return</span> graph


<span class="hljs-keyword">def</span> <span class="hljs-title function_">local_alignment</span>(<span class="hljs-params">
        v: <span class="hljs-type">List</span>[ELEM],
        w: <span class="hljs-type">List</span>[ELEM],
        weight_lookup: WeightLookup
</span>) -&gt; <span class="hljs-type">Tuple</span>[<span class="hljs-built_in">float</span>, <span class="hljs-type">List</span>[<span class="hljs-built_in">str</span>], <span class="hljs-type">List</span>[<span class="hljs-type">Tuple</span>[ELEM, ELEM]]]:
    v_node_count = <span class="hljs-built_in">len</span>(v) + <span class="hljs-number">1</span>
    w_node_count = <span class="hljs-built_in">len</span>(w) + <span class="hljs-number">1</span>
    graph = create_local_alignment_graph(v, w, weight_lookup)
    from_node = (<span class="hljs-number">0</span>, <span class="hljs-number">0</span>)
    to_node = (v_node_count - <span class="hljs-number">1</span>, w_node_count - <span class="hljs-number">1</span>)
    populate_weights_and_backtrack_pointers(
        graph,
        from_node,
        <span class="hljs-keyword">lambda</span> n_id, weight, e_id: graph.get_node_data(n_id).set_weight_and_backtracking_edge(weight, e_id),
        <span class="hljs-keyword">lambda</span> n_id: graph.get_node_data(n_id).get_weight_and_backtracking_edge(),
        <span class="hljs-keyword">lambda</span> e_id: graph.get_edge_data(e_id).weight
    )
    final_weight = graph.get_node_data(to_node).weight
    edges = backtrack(
        graph,
        to_node,
        <span class="hljs-keyword">lambda</span> n_id: graph.get_node_data(n_id).get_weight_and_backtracking_edge()
    )
    alignment_edges = <span class="hljs-built_in">list</span>(<span class="hljs-built_in">filter</span>(<span class="hljs-keyword">lambda</span> e: <span class="hljs-keyword">not</span> e.startswith(<span class="hljs-string">'FREE_RIDE'</span>), edges))  <span class="hljs-comment"># remove free rides from list</span>
    alignment = []
    <span class="hljs-keyword">for</span> e <span class="hljs-keyword">in</span> alignment_edges:
        ed = graph.get_edge_data(e)
        alignment.append((ed.v_elem, ed.w_elem))
    <span class="hljs-keyword">return</span> final_weight, edges, alignment
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the sequences TAGAACT and CGAAG and the score matrix...</p>
<pre><code>INDEL=-1.0
    A   C   T   G
A   1  -1  -1  -1
C  -1   1  -1  -1
T  -1  -1   1  -1
G  -1  -1  -1   1

</code></pre>
<p>... the local alignment is...</p>
<p><img src="latex_b6d25589def0992460b5c0d95c9c1f17.svg" alt="Latex diagram"></p>
<pre><code>GAA
GAA
</code></pre>
<p>Weight: 3.0</p>
</div>
<p></p>
<a name="H3_Matrix%20Algorithm"></a><h4>Matrix Algorithm</h4>
<p><a name="BM_(Algorithms%5C%2FSequence%20Alignment%2FLocal%20Alignment%2FMatrix%20Algorithm)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FLocal%20Alignment%2FGraph%20Algorithm)_TOPIC%2F">Algorithms/Sequence Alignment/Local Alignment/Graph Algorithm</a></li>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FGlobal%20Alignment%2FMatrix%20Algorithm)_TOPIC%2F">Algorithms/Sequence Alignment/Global Alignment/Matrix Algorithm</a></li>
</ul>
</div>
<p><strong>ALGORITHM</strong>:</p>
<p>The following algorithm is an extension to <a href="#BM_(global%20alignment%7Cglobal%20sequence%20alignment)%2Fi">global alignment</a>'s matrix algorithm to properly account for the "free ride" edges required by <a href="#BM_(local%20alignment%7Clocal%20sequence%20alignment)%2Fi">local alignment</a>. It's essentially the same as the graph algorithm, except that the implementation is much more sympathetic to modern hardware.</p>
<p><a href="ch5_code/src/local_alignment/LocalAlignment_Matrix.py">ch5_code/src/local_alignment/LocalAlignment_Matrix.py</a> (lines 10 to 95):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">backtrack</span>(<span class="hljs-params">
        node_matrix: <span class="hljs-type">List</span>[<span class="hljs-type">List</span>[<span class="hljs-type">Any</span>]]
</span>) -&gt; <span class="hljs-type">Tuple</span>[<span class="hljs-built_in">float</span>, <span class="hljs-type">List</span>[<span class="hljs-type">Tuple</span>[ELEM, ELEM]]]:
    v_node_idx = <span class="hljs-built_in">len</span>(node_matrix) - <span class="hljs-number">1</span>
    w_node_idx = <span class="hljs-built_in">len</span>(node_matrix[<span class="hljs-number">0</span>]) - <span class="hljs-number">1</span>
    final_weight = node_matrix[v_node_idx][w_node_idx][<span class="hljs-number">0</span>]
    alignment = []
    <span class="hljs-keyword">while</span> v_node_idx != <span class="hljs-number">0</span> <span class="hljs-keyword">or</span> w_node_idx != <span class="hljs-number">0</span>:
        _, elems, backtrack_ptr = node_matrix[v_node_idx][w_node_idx]
        <span class="hljs-keyword">if</span> backtrack_ptr == <span class="hljs-string">'↓'</span>:
            v_node_idx -= <span class="hljs-number">1</span>
            alignment.append(elems)
        <span class="hljs-keyword">elif</span> backtrack_ptr == <span class="hljs-string">'→'</span>:
            w_node_idx -= <span class="hljs-number">1</span>
            alignment.append(elems)
        <span class="hljs-keyword">elif</span> backtrack_ptr == <span class="hljs-string">'↘'</span>:
            v_node_idx -= <span class="hljs-number">1</span>
            w_node_idx -= <span class="hljs-number">1</span>
            alignment.append(elems)
        <span class="hljs-keyword">elif</span> <span class="hljs-built_in">isinstance</span>(backtrack_ptr, <span class="hljs-built_in">tuple</span>):
            v_node_idx = backtrack_ptr[<span class="hljs-number">0</span>]
            w_node_idx = backtrack_ptr[<span class="hljs-number">1</span>]
    <span class="hljs-keyword">return</span> final_weight, alignment[::-<span class="hljs-number">1</span>]


<span class="hljs-keyword">def</span> <span class="hljs-title function_">local_alignment</span>(<span class="hljs-params">
        v: <span class="hljs-type">List</span>[ELEM],
        w: <span class="hljs-type">List</span>[ELEM],
        weight_lookup: WeightLookup
</span>) -&gt; <span class="hljs-type">Tuple</span>[<span class="hljs-built_in">float</span>, <span class="hljs-type">List</span>[<span class="hljs-type">Tuple</span>[ELEM, ELEM]]]:
    v_node_count = <span class="hljs-built_in">len</span>(v) + <span class="hljs-number">1</span>
    w_node_count = <span class="hljs-built_in">len</span>(w) + <span class="hljs-number">1</span>
    node_matrix = []
    <span class="hljs-keyword">for</span> v_node_idx <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(v_node_count):
        row = []
        <span class="hljs-keyword">for</span> w_node_idx <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(w_node_count):
            row.append([-<span class="hljs-number">1.0</span>, (<span class="hljs-literal">None</span>, <span class="hljs-literal">None</span>), <span class="hljs-string">'?'</span>])
        node_matrix.append(row)
    node_matrix[<span class="hljs-number">0</span>][<span class="hljs-number">0</span>][<span class="hljs-number">0</span>] = <span class="hljs-number">0.0</span>           <span class="hljs-comment"># source node weight</span>
    node_matrix[<span class="hljs-number">0</span>][<span class="hljs-number">0</span>][<span class="hljs-number">1</span>] = (<span class="hljs-literal">None</span>, <span class="hljs-literal">None</span>)  <span class="hljs-comment"># source node elements (elements don't matter for source node)</span>
    node_matrix[<span class="hljs-number">0</span>][<span class="hljs-number">0</span>][<span class="hljs-number">2</span>] = <span class="hljs-string">'↘'</span>           <span class="hljs-comment"># source node backtracking edge (direction doesn't matter for source node)</span>
    <span class="hljs-keyword">for</span> v_node_idx, w_node_idx <span class="hljs-keyword">in</span> product(<span class="hljs-built_in">range</span>(v_node_count), <span class="hljs-built_in">range</span>(w_node_count)):
        parents = []
        <span class="hljs-keyword">if</span> v_node_idx &gt; <span class="hljs-number">0</span> <span class="hljs-keyword">and</span> w_node_idx &gt; <span class="hljs-number">0</span>:
            v_elem = v[v_node_idx - <span class="hljs-number">1</span>]
            w_elem = w[w_node_idx - <span class="hljs-number">1</span>]
            parents.append([
                node_matrix[v_node_idx - <span class="hljs-number">1</span>][w_node_idx - <span class="hljs-number">1</span>][<span class="hljs-number">0</span>] + weight_lookup.lookup(v_elem, w_elem),
                (v_elem, w_elem),
                <span class="hljs-string">'↘'</span>
            ])
        <span class="hljs-keyword">if</span> v_node_idx &gt; <span class="hljs-number">0</span>:
            v_elem = v[v_node_idx - <span class="hljs-number">1</span>]
            parents.append([
                node_matrix[v_node_idx - <span class="hljs-number">1</span>][w_node_idx][<span class="hljs-number">0</span>] + weight_lookup.lookup(v_elem, <span class="hljs-literal">None</span>),
                (v_elem, <span class="hljs-literal">None</span>),
                <span class="hljs-string">'↓'</span>
            ])
        <span class="hljs-keyword">if</span> w_node_idx &gt; <span class="hljs-number">0</span>:
            w_elem = w[w_node_idx - <span class="hljs-number">1</span>]
            parents.append([
                node_matrix[v_node_idx][w_node_idx - <span class="hljs-number">1</span>][<span class="hljs-number">0</span>] + weight_lookup.lookup(<span class="hljs-literal">None</span>, w_elem),
                (<span class="hljs-literal">None</span>, w_elem),
                <span class="hljs-string">'→'</span>
            ])
        <span class="hljs-comment"># If not source node, consider free-ride from source node</span>
        <span class="hljs-keyword">if</span> v_node_idx != <span class="hljs-number">0</span> <span class="hljs-keyword">and</span> w_node_idx != <span class="hljs-number">0</span>:
            parents.append([
                <span class="hljs-number">0.0</span>,
                (<span class="hljs-literal">None</span>, <span class="hljs-literal">None</span>),
                (<span class="hljs-number">0</span>, <span class="hljs-number">0</span>)  <span class="hljs-comment"># jump to source</span>
            ])
        <span class="hljs-comment"># If sink node, consider free-rides coming from every node that isn't sink node</span>
        <span class="hljs-keyword">if</span> v_node_idx == v_node_count - <span class="hljs-number">1</span> <span class="hljs-keyword">and</span> w_node_idx == w_node_count - <span class="hljs-number">1</span>:
            <span class="hljs-keyword">for</span> v_node_idx_from, w_node_idx_from <span class="hljs-keyword">in</span> product(<span class="hljs-built_in">range</span>(v_node_count), <span class="hljs-built_in">range</span>(w_node_count)):
                <span class="hljs-keyword">if</span> v_node_idx_from == v_node_count - <span class="hljs-number">1</span> <span class="hljs-keyword">and</span> w_node_idx_from == w_node_count - <span class="hljs-number">1</span>:
                    <span class="hljs-keyword">continue</span>
                parents.append([
                    node_matrix[v_node_idx_from][w_node_idx_from][<span class="hljs-number">0</span>] + <span class="hljs-number">0.0</span>,
                    (<span class="hljs-literal">None</span>, <span class="hljs-literal">None</span>),
                    (v_node_idx_from, w_node_idx_from)  <span class="hljs-comment"># jump to this position</span>
                ])
        <span class="hljs-keyword">if</span> parents:  <span class="hljs-comment"># parents will be empty if v_node_idx and w_node_idx were both 0</span>
            node_matrix[v_node_idx][w_node_idx] = <span class="hljs-built_in">max</span>(parents, key=<span class="hljs-keyword">lambda</span> x: x[<span class="hljs-number">0</span>])
    <span class="hljs-keyword">return</span> backtrack(node_matrix)
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the sequences TAGAACT and CGAAG and the score matrix...</p>
<pre><code>INDEL=-1.0
    A   C   T   G
A   1  -1  -1  -1
C  -1   1  -1  -1
T  -1  -1   1  -1
G  -1  -1  -1   1

</code></pre>
<p>... the local alignment is...</p>
<pre><code>GAA
GAA
</code></pre>
<p>Weight: 3.0</p>
</div>
<p></p>
<a name="H_Protein%20Scoring"></a><h3>Protein Scoring</h3>
<p><a name="BM_(Algorithms%5C%2FSequence%20Alignment%2FProtein%20Scoring)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FGlobal%20Alignment)_TOPIC%2F">Algorithms/Sequence Alignment/Global Alignment</a></li>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FLocal%20Alignment)_TOPIC%2F">Algorithms/Sequence Alignment/Local Alignment</a></li>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FFitting%20Alignment)_TOPIC%2F">Algorithms/Sequence Alignment/Fitting Alignment</a></li>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FOverlap%20Alignment)_TOPIC%2F">Algorithms/Sequence Alignment/Overlap Alignment</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: Given a pair of protein <a href="#BM_(sequence)%2Fi">sequence</a>s, score those <a href="#BM_(sequence)%2Fi">sequence</a>s against each other based on the similarity of the <a href="#BM_(amino%20acid)%2Fi">amino acid</a>s. In this case, similarity is defined as how probable it is that one <a href="#BM_(amino%20acid)%2Fi">amino acid</a> mutates to the other while still having the protein remain functional.</p>
<p><strong>WHY</strong>: Before performing a pair-wise <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a>, there needs to be some baseline for how elements within those <a href="#BM_(sequence)%2Fi">sequence</a>s measure up against each other. In the simplest case, elements are compared using equality: matching elements score 1, while mismatches or <a href="#BM_(indel)%2Fi">indel</a>s score 0. However, there are many other cases where element equality isn't a good measure.</p>
<p>Protein <a href="#BM_(sequence)%2Fi">sequence</a>s are one such case. Biological <a href="#BM_(sequence)%2Fi">sequence</a>s such as proteins and DNA undergo mutation. Two proteins may be very closely related (e.g. evolved from same parent protein, perform the same function, etc..) but their <a href="#BM_(sequence)%2Fi">sequence</a>s may have mutated to a point where they appear as being wildly different. To appropriately align protein <a href="#BM_(sequence)%2Fi">sequence</a>s, <a href="#BM_(amino%20acid)%2Fi">amino acid</a> mutation probabilities need to be derived and factored into scoring. For example, there may be good odds that some random protein would still continue to function as-is if some of its Y <a href="#BM_(amino%20acid)%2Fi">amino acid</a>s were swapped with F.</p>
<a name="H_PAM%20Scoring%20Matrix"></a><h4>PAM Scoring Matrix</h4>
<p><a name="BM_(Algorithms%5C%2FSequence%20Alignment%2FProtein%20Scoring%2FPAM%20Scoring%20Matrix)_TOPIC%2F"></a><strong></strong></p>
<p><a href="#BM_(point%20accepted%20mutation%7Cpercent%20accepted%20mutation)%2Fi">Point accepted mutation</a> (<a href="#BM_%5Cb(PAM)%5Cd*%5Cb%2F">PAM</a>) is a scoring matrix used for <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a>s of proteins. The scoring matrix is calculated by inspecting / extrapolating mutations as homologous proteins evolve. Specifically, mutations in the DNA <a href="#BM_(sequence)%2Fi">sequence</a> that <a href="#BM_(encode%7Cencoding%7Cdecode%7Cdecoding)%2Fi">encode</a> some protein may change the resulting <a href="#BM_(amino%20acid)%2Fi">amino acid</a> <a href="#BM_(sequence)%2Fi">sequence</a> for that protein. Those mutations that...</p>
<ul>
<li>impair the ability of the protein to function aren't likely to survive, and as such are given a low score.</li>
<li>keep the protein functional are likely to survive, and as such are given a normal or high score.</li>
</ul>
<p><a href="#BM_%5Cb(PAM)%5Cd*%5Cb%2F">PAM</a> matrices are developed iteratively. An initial <a href="#BM_%5Cb(PAM)%5Cd*%5Cb%2F">PAM</a> matrix is calculated by aligning extremely similar protein <a href="#BM_(sequence)%2Fi">sequence</a>s using a simple scoring model (1 for match, 0 for mismatch / <a href="#BM_(indel)%2Fi">indel</a>). That <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a> then provides the scoring model for the next iteration. For example, the alignment for the initial iteration may have determined that D may be a suitable substitution for W. As such, the <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a> for the next iteration will score more than 0 (e.g. 1) if it encounters D being compared to W.</p>
<p>Other factors are also brought into the mix when developing scores for <a href="#BM_%5Cb(PAM)%5Cd*%5Cb%2F">PAM</a> matrices. For example, the ...</p>
<ul>
<li>likelihood of <a href="#BM_(amino%20acid)%2Fi">amino acid</a> mutations (e.g. Cys and Trp are the least mutable <a href="#BM_(amino%20acid)%2Fi">amino acid</a>s).</li>
<li>speed of evolution (e.g. some mutations were more probably in species 100 million years ago vs 1 million years ago).</li>
</ul>
<p>It's said that <a href="#BM_%5Cb(PAM)%5Cd*%5Cb%2F">PAM</a> is focused on tracking the evolutionary origins of proteins. <a href="#BM_(sequence)%2Fi">Sequence</a>s that are 99% similar are said to be 1 <a href="#BM_%5Cb(PAM)%5Cd*%5Cb%2F">PAM</a> unit diverged, where a <a href="#BM_%5Cb(PAM)%5Cd*%5Cb%2F">PAM</a> unit is the amount of time it takes an "average" protein to mutate 1% of its <a href="#BM_(amino%20acid)%2Fi">amino acid</a>s. <a href="#BM_%5Cb(PAM)%5Cd*%5Cb%2F">PAM</a>1 (the initial scoring matrix) was defined by performing many <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a>s between proteins that are 99% similar (1 <a href="#BM_%5Cb(PAM)%5Cd*%5Cb%2F">PAM</a> unit diverged).</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p><a href="http://www.compbio.dundee.ac.uk/papers/rev93_1/subsection3_3_5.html">Here</a> and <a href="https://en.wikipedia.org/w/index.php?title=Point_accepted_mutation&amp;oldid=1002281881#Comparing_PAM_and_BLOSUM">here</a> both seem to say that <a href="#BM_%5Cb(BLOSUM)%5Cd*%5Cb%2F">BLOSUM</a> supersedes <a href="#BM_%5Cb(PAM)%5Cd*%5Cb%2F">PAM</a> as a scoring matrix for protein <a href="#BM_(sequence)%2Fi">sequence</a>s.</p>
<blockquote>
<p>Although both matrices produce similar scoring outcomes they were generated using differing methodologies. The <a href="#BM_%5Cb(BLOSUM)%5Cd*%5Cb%2F">BLOSUM</a> matrices were generated directly from the <a href="#BM_(amino%20acid)%2Fi">amino acid</a> differences in aligned blocks that have diverged to varying degrees the <a href="#BM_%5Cb(PAM)%5Cd*%5Cb%2F">PAM</a> matrices reflect the extrapolation of evolutionary information based on closely related <a href="#BM_(sequence)%2Fi">sequence</a>s to longer timescales</p>
</blockquote>
<blockquote>
<p>Henikoff and Henikoff [16] have compared the <a href="#BM_%5Cb(BLOSUM)%5Cd*%5Cb%2F">BLOSUM</a> matrices to <a href="#BM_%5Cb(PAM)%5Cd*%5Cb%2F">PAM</a>, PET, Overington, Gonnet [17] and multiple <a href="#BM_%5Cb(PAM)%5Cd*%5Cb%2F">PAM</a> matrices by evaluating how effectively the matrices can detect known members of a protein family from a database when searching with the ungapped <a href="#BM_(local%20alignment%7Clocal%20sequence%20alignment)%2Fi">local alignment</a> program BLAST [18]. They conclude that overall the <a href="#BM_%5Cb(BLOSUM)%5Cd*%5Cb%2F">BLOSUM</a> 62 matrix is the most effective.</p>
</blockquote>
</div>
<p><a href="#BM_%5Cb(PAM)%5Cd*%5Cb%2F">PAM</a>250 is the most commonly used variant:</p>
<p><a href="ch5_code/src/PAM250.txt">ch5_code/src/PAM250.txt</a> (lines 2 to 22):</p>
<pre><code class="language-([\s\S]+)">   A  C  D  E  F  G  H  I  K  L  M  N  P  Q  R  S  T  V  W  Y
A  2 -2  0  0 -3  1 -1 -1 -1 -2 -1  0  1  0 -2  1  1  0 -6 -3
C -2 12 -5 -5 -4 -3 -3 -2 -5 -6 -5 -4 -3 -5 -4  0 -2 -2 -8  0
D  0 -5  4  3 -6  1  1 -2  0 -4 -3  2 -1  2 -1  0  0 -2 -7 -4
E  0 -5  3  4 -5  0  1 -2  0 -3 -2  1 -1  2 -1  0  0 -2 -7 -4
F -3 -4 -6 -5  9 -5 -2  1 -5  2  0 -3 -5 -5 -4 -3 -3 -1  0  7
G  1 -3  1  0 -5  5 -2 -3 -2 -4 -3  0  0 -1 -3  1  0 -1 -7 -5
H -1 -3  1  1 -2 -2  6 -2  0 -2 -2  2  0  3  2 -1 -1 -2 -3  0
I -1 -2 -2 -2  1 -3 -2  5 -2  2  2 -2 -2 -2 -2 -1  0  4 -5 -1
K -1 -5  0  0 -5 -2  0 -2  5 -3  0  1 -1  1  3  0  0 -2 -3 -4
L -2 -6 -4 -3  2 -4 -2  2 -3  6  4 -3 -3 -2 -3 -3 -2  2 -2 -1
M -1 -5 -3 -2  0 -3 -2  2  0  4  6 -2 -2 -1  0 -2 -1  2 -4 -2
N  0 -4  2  1 -3  0  2 -2  1 -3 -2  2  0  1  0  1  0 -2 -4 -2
P  1 -3 -1 -1 -5  0  0 -2 -1 -3 -2  0  6  0  0  1  0 -1 -6 -5
Q  0 -5  2  2 -5 -1  3 -2  1 -2 -1  1  0  4  1 -1 -1 -2 -5 -4
R -2 -4 -1 -1 -4 -3  2 -2  3 -3  0  0  0  1  6  0 -1 -2  2 -4
S  1  0  0  0 -3  1 -1 -1  0 -3 -2  1  1 -1  0  2  1 -1 -2 -3
T  1 -2  0  0 -3  0 -1  0  0 -2 -1  0  0 -1 -1  1  3  0 -5 -3
V  0 -2 -2 -2 -1 -1 -2  4 -2  2  2 -2 -1 -2 -2 -1  0  4 -6 -2
W -6 -8 -7 -7  0 -7 -3 -5 -3 -2 -4 -4 -6 -5  2 -2 -5 -6 17  0
Y -3  0 -4 -4  7 -5  0 -1 -4 -1 -2 -2 -5 -4 -4 -3 -3 -2  0 10
</code></pre>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The above matrix was supplied by the Pevzner book. You can find it online <a href="https://swift.cmbi.umcn.nl/teach/aainfo/pam250.shtml">here</a>, but the <a href="#BM_(indel)%2Fi">indel</a> scores on that website are set to -8 Whereas in the Pevzner book I've also seen them set to -5. I don't know which is correct. I don't know if <a href="#BM_%5Cb(PAM)%5Cd*%5Cb%2F">PAM</a>250 defines a constant for <a href="#BM_(indel)%2Fi">indel</a>s.</p>
</div>
<a name="H_BLOSUM%20Scoring%20Matrix"></a><h4>BLOSUM Scoring Matrix</h4>
<p><a name="BM_(Algorithms%5C%2FSequence%20Alignment%2FProtein%20Scoring%2FBLOSUM%20Scoring%20Matrix)_TOPIC%2F"></a><strong></strong></p>
<p><a href="#BM_(blocks%20substitution%20matrix%7Cblocks%20substitution%20matrices%7Cblocks%20amino%20acid%20substitution%20matrices%7Cblocks%20amino%20acid%20substitution%20matrix)%2Fi">Blocks amino acid substitution matrix</a> (<a href="#BM_%5Cb(BLOSUM)%5Cd*%5Cb%2F">BLOSUM</a>) is a scoring matrix used for <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a>s of proteins. The scoring matrix is calculated by scanning a protein database for highly conserved regions between similar proteins, where the mutations between those highly conserved regions define the scores. Specifically, those highly conserved regions are identified based on <a href="#BM_(local%20alignment%7Clocal%20sequence%20alignment)%2Fi">local alignment</a>s without support for <a href="#BM_(indel)%2Fi">indel</a>s (gaps not allowed). Non-matching positions in that alignment define potentially acceptable mutations.</p>
<p>Several sets of <a href="#BM_%5Cb(BLOSUM)%5Cd*%5Cb%2F">BLOSUM</a> matrices exist, each identified by a different number. This number defines the similarity of the <a href="#BM_(sequence)%2Fi">sequence</a>s used to create the matrix: The protein database <a href="#BM_(sequence)%2Fi">sequence</a>s used to derive the matrix are filtered such that only those with &gt;= n% similarity are used, where n is the number. For example, ...</p>
<ul>
<li><a href="#BM_%5Cb(BLOSUM)%5Cd*%5Cb%2F">BLOSUM</a>80 is created from <a href="#BM_(sequence)%2Fi">sequence</a>s that are &gt;= 80% similar.</li>
<li><a href="#BM_%5Cb(BLOSUM)%5Cd*%5Cb%2F">BLOSUM</a>45 is created from <a href="#BM_(sequence)%2Fi">sequence</a>s that are &gt;= 45% similar.</li>
</ul>
<p>As such, <a href="#BM_%5Cb(BLOSUM)%5Cd*%5Cb%2F">BLOSUM</a> matrices with higher numbers are designed to compare more closely related <a href="#BM_(sequence)%2Fi">sequence</a>s while those with lower numbers are designed to score more distant related <a href="#BM_(sequence)%2Fi">sequence</a>s.</p>
<p><a href="#BM_%5Cb(BLOSUM)%5Cd*%5Cb%2F">BLOSUM</a>62 is the most commonly used variant since "experimentation has shown that it's among the best for detecting weak similarities":</p>
<p><a href="ch5_code/src/BLOSUM62.txt">ch5_code/src/BLOSUM62.txt</a> (lines 2 to 22):</p>
<pre><code class="language-([\s\S]+)">   A  C  D  E  F  G  H  I  K  L  M  N  P  Q  R  S  T  V  W  Y
A  4  0 -2 -1 -2  0 -2 -1 -1 -1 -1 -2 -1 -1 -1  1  0  0 -3 -2
C  0  9 -3 -4 -2 -3 -3 -1 -3 -1 -1 -3 -3 -3 -3 -1 -1 -1 -2 -2
D -2 -3  6  2 -3 -1 -1 -3 -1 -4 -3  1 -1  0 -2  0 -1 -3 -4 -3
E -1 -4  2  5 -3 -2  0 -3  1 -3 -2  0 -1  2  0  0 -1 -2 -3 -2
F -2 -2 -3 -3  6 -3 -1  0 -3  0  0 -3 -4 -3 -3 -2 -2 -1  1  3
G  0 -3 -1 -2 -3  6 -2 -4 -2 -4 -3  0 -2 -2 -2  0 -2 -3 -2 -3
H -2 -3 -1  0 -1 -2  8 -3 -1 -3 -2  1 -2  0  0 -1 -2 -3 -2  2
I -1 -1 -3 -3  0 -4 -3  4 -3  2  1 -3 -3 -3 -3 -2 -1  3 -3 -1
K -1 -3 -1  1 -3 -2 -1 -3  5 -2 -1  0 -1  1  2  0 -1 -2 -3 -2
L -1 -1 -4 -3  0 -4 -3  2 -2  4  2 -3 -3 -2 -2 -2 -1  1 -2 -1
M -1 -1 -3 -2  0 -3 -2  1 -1  2  5 -2 -2  0 -1 -1 -1  1 -1 -1
N -2 -3  1  0 -3  0  1 -3  0 -3 -2  6 -2  0  0  1  0 -3 -4 -2
P -1 -3 -1 -1 -4 -2 -2 -3 -1 -3 -2 -2  7 -1 -2 -1 -1 -2 -4 -3
Q -1 -3  0  2 -3 -2  0 -3  1 -2  0  0 -1  5  1  0 -1 -2 -2 -1
R -1 -3 -2  0 -3 -2  0 -3  2 -2 -1  0 -2  1  5 -1 -1 -3 -3 -2
S  1 -1  0  0 -2  0 -1 -2  0 -2 -1  1 -1  0 -1  4  1 -2 -3 -2
T  0 -1 -1 -1 -2 -2 -2 -1 -1 -1 -1  0 -1 -1 -1  1  5  0 -2 -2
V  0 -1 -3 -2 -1 -3 -3  3 -2  1  1 -3 -2 -2 -3 -2  0  4 -3 -1
W -3 -2 -4 -3  1 -2 -2 -3 -3 -2 -1 -4 -4 -2 -3 -3 -2 -3 11  2
Y -2 -2 -3 -2  3 -3  2 -1 -2 -1 -1 -2 -3 -1 -2 -2 -2 -1  2  7
</code></pre>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The above matrix was supplied by the Pevzner book. You can find it online <a href="https://www.ncbi.nlm.nih.gov/Class/FieldGuide/BLOSUM62.txt">here</a>, but the <a href="#BM_(indel)%2Fi">indel</a> scores on that website are set to -4 whereas in the Pevzner book I've seen them set to -5. I don't know which is correct. I don't know if <a href="#BM_%5Cb(BLOSUM)%5Cd*%5Cb%2F">BLOSUM</a>62 defines a constant for <a href="#BM_(indel)%2Fi">indel</a>s.</p>
</div>
<a name="H_Extended%20Gap%20Scoring"></a><h3>Extended Gap Scoring</h3>
<p><a name="BM_(Algorithms%5C%2FSequence%20Alignment%2FExtended%20Gap%20Scoring)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FGlobal%20Alignment)_TOPIC%2F">Algorithms/Sequence Alignment/Global Alignment</a></li>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FLocal%20Alignment)_TOPIC%2F">Algorithms/Sequence Alignment/Local Alignment</a></li>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FFitting%20Alignment)_TOPIC%2F">Algorithms/Sequence Alignment/Fitting Alignment</a></li>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FOverlap%20Alignment)_TOPIC%2F">Algorithms/Sequence Alignment/Overlap Alignment</a></li>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FProtein%20Scoring)_TOPIC%2F">Algorithms/Sequence Alignment/Protein Scoring</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: When performing <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a>, prefer contiguous <a href="#BM_(indel)%2Fi">indel</a>s in a <a href="#BM_(sequence)%2Fi">sequence</a> vs individual <a href="#BM_(indel)%2Fi">indel</a>s. This is done by scoring contiguous <a href="#BM_(indel)%2Fi">indel</a>s differently than individual <a href="#BM_(indel)%2Fi">indel</a>s:</p>
<ul>
<li>Individual <a href="#BM_(indel)%2Fi">indel</a>s are penalized by choosing the normal <a href="#BM_(indel)%2Fi">indel</a> score (e.g. score of -5).</li>
<li>Contiguous <a href="#BM_(indel)%2Fi">indel</a>s are penalized by choosing the normal <a href="#BM_(indel)%2Fi">indel</a> score for the first <a href="#BM_(indel)%2Fi">indel</a> in the list (e.g. score of -5), then all other <a href="#BM_(indel)%2Fi">indel</a>s are scored using a better <em>extended</em> <a href="#BM_(indel)%2Fi">indel</a> score (e.g. score of -0.1).</li>
</ul>
<p>For example, given an alignment region where one of the <a href="#BM_(sequence)%2Fi">sequence</a>s has 3 contiguous <a href="#BM_(indel)%2Fi">indel</a>s, the traditional method would assign a score of -15 (-5 for each <a href="#BM_(indel)%2Fi">indel</a>) while this method would assign a score of -5.2 (-5 for starting <a href="#BM_(indel)%2Fi">indel</a>, -0.1 for each subsequent <a href="#BM_(indel)%2Fi">indel</a>)...</p>
<pre><code>AAATTTAATA
AAA---AA-A

Score from indels using traditional scoring:   -5   + -5   + -5   + -5   = -20
Score from indels using extended gap scoring:  -5   + -0.1 + -0.1 + -5   = -10.2
</code></pre>
<p><strong>WHY</strong>: DNA mutations are more likely to happen in chunks rather than <a href="#BM_(point%20mutation)%2Fi">point mutation</a>s (e.g. <a href="#BM_(transposon%7Ctransposable%20element%7Cjumping%20gene)%2Fi">transposon</a>s). Extended gap scoring helps account for that fact. Since DNA <a href="#BM_(encode%7Cencoding%7Cdecode%7Cdecoding)%2Fi">encode</a> proteins (<a href="#BM_(codon%7Cgenetic%20code)%2Fi">codon</a>s), this effects proteins as well.</p>
<a name="H_Naive%20Algorithm"></a><h4>Naive Algorithm</h4>
<p><a name="BM_(Algorithms%5C%2FSequence%20Alignment%2FExtended%20Gap%20Scoring%2FNaive%20Algorithm)_TOPIC%2F"></a><strong></strong></p>
<p><strong>ALGORITHM</strong>:</p>
<p>The naive way to perform extended gap scoring is to introduce a new edge for each contiguous <a href="#BM_(indel)%2Fi">indel</a>. For example, given the <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a>...</p>
<p><img src="svgbob_a78c6b3a04fd87c8fe19f92513598ddfb1734507.svg" alt="Kroki diagram output"></p>
<ul>
<li>
<p>each row would have an edge added to represent a contiguous <a href="#BM_(indel)%2Fi">indel</a>s.</p>
<p><img src="svgbob_d61e1d9fea9db6aca581f7969df6c3ff3bb50887.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>each column would have an edge added to represent a contiguous <a href="#BM_(indel)%2Fi">indel</a>s.</p>
<p><img src="svgbob_39b1828786374ffdfc0a696685b7bbb7c95fc275.svg" alt="Kroki diagram output"></p>
</li>
</ul>
<p>Each added edge represents a contiguous set of <a href="#BM_(indel)%2Fi">indel</a>s. Contiguous <a href="#BM_(indel)%2Fi">indel</a>s are penalized by choosing the normal <a href="#BM_(indel)%2Fi">indel</a> score for the first <a href="#BM_(indel)%2Fi">indel</a> in the list (e.g. score of -5), then all other <a href="#BM_(indel)%2Fi">indel</a>s are scored using a better <em>extended</em> <a href="#BM_(indel)%2Fi">indel</a> score (e.g. score of -0.1). As such, the maximum <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a> will choose one of these contiguous <a href="#BM_(indel)%2Fi">indel</a> edges over individual <a href="#BM_(indel)%2Fi">indel</a> edges or poor substitution choices such as those in <a href="#BM_%5Cb(PAM)%5Cd*%5Cb%2F">PAM</a> / <a href="#BM_%5Cb(BLOSUM)%5Cd*%5Cb%2F">BLOSUM</a> scoring matrices.</p>
<div style="border:1px solid black;">
<p></p>
<p><img src="latex_730222afedc77c01943fc315e42992aa.svg" alt="Latex diagram"></p>
<p>NOTE: Purple and red edges are extended indel edges.</p>
</div>
<p></p>
<p>The problem with this algorithm is that as the <a href="#BM_(sequence)%2Fi">sequence</a> lengths grow, the number of added edges explodes. It isn't practical for anything other than short <a href="#BM_(sequence)%2Fi">sequence</a>s.</p>
<p><a href="ch5_code/src/affine_gap_alignment/AffineGapAlignment_Basic_Graph.py">ch5_code/src/affine_gap_alignment/AffineGapAlignment_Basic_Graph.py</a> (lines 37 to 104):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">create_affine_gap_alignment_graph</span>(<span class="hljs-params">
        v: <span class="hljs-type">List</span>[ELEM],
        w: <span class="hljs-type">List</span>[ELEM],
        weight_lookup: WeightLookup,
        extended_gap_weight: <span class="hljs-built_in">float</span>
</span>) -&gt; Graph[<span class="hljs-type">Tuple</span>[<span class="hljs-built_in">int</span>, ...], NodeData, <span class="hljs-built_in">str</span>, EdgeData]:
    graph = create_grid_graph(
        [v, w],
        <span class="hljs-keyword">lambda</span> n_id: NodeData(),
        <span class="hljs-keyword">lambda</span> src_n_id, dst_n_id, offset, elems: EdgeData(elems[<span class="hljs-number">0</span>], elems[<span class="hljs-number">1</span>], weight_lookup.lookup(*elems))
    )
    v_node_count = <span class="hljs-built_in">len</span>(v) + <span class="hljs-number">1</span>
    w_node_count = <span class="hljs-built_in">len</span>(w) + <span class="hljs-number">1</span>
    horizontal_indel_hop_edge_id_func = unique_id_generator(<span class="hljs-string">'HORIZONTAL_INDEL_HOP'</span>)
    <span class="hljs-keyword">for</span> from_c, r <span class="hljs-keyword">in</span> product(<span class="hljs-built_in">range</span>(v_node_count), <span class="hljs-built_in">range</span>(w_node_count)):
        from_node_id = from_c, r
        <span class="hljs-keyword">for</span> to_c <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(from_c + <span class="hljs-number">2</span>, v_node_count):
            to_node_id = to_c, r
            edge_id = horizontal_indel_hop_edge_id_func()
            v_elems = v[from_c:to_c]
            w_elems = [<span class="hljs-literal">None</span>] * <span class="hljs-built_in">len</span>(v_elems)
            hop_count = to_c - from_c
            weight = weight_lookup.lookup(v_elems[<span class="hljs-number">0</span>], w_elems[<span class="hljs-number">0</span>]) + (hop_count - <span class="hljs-number">1</span>) * extended_gap_weight
            graph.insert_edge(edge_id, from_node_id, to_node_id, EdgeData(v_elems, w_elems, weight))
    vertical_indel_hop_edge_id_func = unique_id_generator(<span class="hljs-string">'VERTICAL_INDEL_HOP'</span>)
    <span class="hljs-keyword">for</span> c, from_r <span class="hljs-keyword">in</span> product(<span class="hljs-built_in">range</span>(v_node_count), <span class="hljs-built_in">range</span>(w_node_count)):
        from_node_id = c, from_r
        <span class="hljs-keyword">for</span> to_r <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(from_r + <span class="hljs-number">2</span>, w_node_count):
            to_node_id = c, to_r
            edge_id = vertical_indel_hop_edge_id_func()
            w_elems = w[from_r:to_r]
            v_elems = [<span class="hljs-literal">None</span>] * <span class="hljs-built_in">len</span>(w_elems)
            hop_count = to_r - from_r
            weight = weight_lookup.lookup(v_elems[<span class="hljs-number">0</span>], w_elems[<span class="hljs-number">0</span>]) + (hop_count - <span class="hljs-number">1</span>) * extended_gap_weight
            graph.insert_edge(edge_id, from_node_id, to_node_id, EdgeData(v_elems, w_elems, weight))
    <span class="hljs-keyword">return</span> graph


<span class="hljs-keyword">def</span> <span class="hljs-title function_">affine_gap_alignment</span>(<span class="hljs-params">
        v: <span class="hljs-type">List</span>[ELEM],
        w: <span class="hljs-type">List</span>[ELEM],
        weight_lookup: WeightLookup,
        extended_gap_weight: <span class="hljs-built_in">float</span>
</span>) -&gt; <span class="hljs-type">Tuple</span>[<span class="hljs-built_in">float</span>, <span class="hljs-type">List</span>[<span class="hljs-built_in">str</span>], <span class="hljs-type">List</span>[<span class="hljs-type">Tuple</span>[ELEM, ELEM]]]:
    v_node_count = <span class="hljs-built_in">len</span>(v) + <span class="hljs-number">1</span>
    w_node_count = <span class="hljs-built_in">len</span>(w) + <span class="hljs-number">1</span>
    graph = create_affine_gap_alignment_graph(v, w, weight_lookup, extended_gap_weight)
    from_node = (<span class="hljs-number">0</span>, <span class="hljs-number">0</span>)
    to_node = (v_node_count - <span class="hljs-number">1</span>, w_node_count - <span class="hljs-number">1</span>)
    populate_weights_and_backtrack_pointers(
        graph,
        from_node,
        <span class="hljs-keyword">lambda</span> n_id, weight, e_id: graph.get_node_data(n_id).set_weight_and_backtracking_edge(weight, e_id),
        <span class="hljs-keyword">lambda</span> n_id: graph.get_node_data(n_id).get_weight_and_backtracking_edge(),
        <span class="hljs-keyword">lambda</span> e_id: graph.get_edge_data(e_id).weight
    )
    final_weight = graph.get_node_data(to_node).weight
    edges = backtrack(
        graph,
        to_node,
        <span class="hljs-keyword">lambda</span> n_id: graph.get_node_data(n_id).get_weight_and_backtracking_edge()
    )
    alignment = []
    <span class="hljs-keyword">for</span> e <span class="hljs-keyword">in</span> edges:
        ed = graph.get_edge_data(e)
        alignment.append((ed.v_elem, ed.w_elem))
    <span class="hljs-keyword">return</span> final_weight, edges, alignment
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the sequences TAGGCGGAT and TACCCCCAT and the score matrix...</p>
<pre><code>INDEL=-1.0
    A   C   T   G
A   1  -1  -1  -1
C  -1   1  -1  -1
T  -1  -1   1  -1
G  -1  -1  -1   1

</code></pre>
<p>... the global alignment is...</p>
<p><img src="latex_fae3b6e18027cce31f274cf8759d3d2a.svg" alt="Latex diagram"></p>
<pre><code>TA----GGCGGAT
TACCCC--C--AT
</code></pre>
<p>Weight: 1.4999999999999998</p>
</div>
<p></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The algorithm above was applied on <a href="#BM_(global%20alignment%7Cglobal%20sequence%20alignment)%2Fi">global alignment</a>, but it should be obvious how to apply it to the other alignment types discussed.</p>
</div>
<a name="H_Layer%20Algorithm"></a><h4>Layer Algorithm</h4>
<p><a name="BM_(Algorithms%5C%2FSequence%20Alignment%2FExtended%20Gap%20Scoring%2FLayer%20Algorithm)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FExtended%20Gap%20Scoring%2FNaive%20Algorithm)_TOPIC%2F">Algorithms/Sequence Alignment/Extended Gap Scoring/Naive Algorithm</a></li>
</ul>
</div>
<p><strong>ALGORITHM</strong>:</p>
<p>Recall that the problem with the naive algorithm algorithm is that as the <a href="#BM_(sequence)%2Fi">sequence</a> lengths grow, the number of added edges explodes. It isn't practical for anything other than short <a href="#BM_(sequence)%2Fi">sequence</a>s. A better algorithm that achieves the exact same result is the layer algorithm. The layer algorithm breaks an <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a> into 3 distinct layers:</p>
<ol>
<li>horizontal edges go into their own layer.</li>
<li>diagonal edges go into their own layer.</li>
<li>vertical edges go into their own layer.</li>
</ol>
<p>The edge weights in the horizontal and diagonal layers are updated such that they use the <em>extended</em> <a href="#BM_(indel)%2Fi">indel</a> score (e.g. -0.1). Then, for each node (x, y) in the diagonal layer, ...</p>
<ul>
<li>an edge is added to node (x+1, y) in the horizontal layer with a normal <a href="#BM_(indel)%2Fi">indel</a> score (e.g. -5).</li>
<li>an edge is added to node (x, y+1) in the vertical layer with a normal <a href="#BM_(indel)%2Fi">indel</a> score (e.g. -5).</li>
</ul>
<p>Similarly, for each node (x, y) in both the horizontal and vertical layers that an edge from the diagonal layer points to, create a 0 weight "free ride" edge back to node (x, y) in the diagonal layer. These "free ride" edges are the same as the "free ride" edges in <a href="#BM_(local%20alignment%7Clocal%20sequence%20alignment)%2Fi">local alignment</a> / <a href="#BM_(fitting%20alignment%7Cfitting%20sequence%20alignment)%2Fi">fitting alignment</a> / <a href="#BM_(overlap%20alignment%7Coverlap%20sequence%20alignment)%2Fi">overlap alignment</a> -- they hop across the <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a> without adding anything to the <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a>.</p>
<p>The source node and sink node are at the top-left node and bottom-right node (respectively) of the diagonal layer.</p>
<div style="border:1px solid black;">
<p></p>
<p><img src="latex_ee884a61370d303627cb798cd15bc48b.svg" alt="Latex diagram"></p>
<p>NOTE: Orange edges are "free rides" from source / Purple edges are "free rides" to sink.</p>
</div>
<p></p>
<p>The way to think about this layered structure <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a> is that the hop from a node in the diagonal layer to a node in the horizontal/vertical layer will always have a normal <a href="#BM_(indel)%2Fi">indel</a> score (e.g. -5). From there it either has the option to hop back to the diagonal layer (via the "free ride" edge) or continue pushing through <a href="#BM_(indel)%2Fi">indel</a>s using the less penalizing <em>extended</em> <a href="#BM_(indel)%2Fi">indel</a> score (e.g. -0.1).</p>
<p>This layered structure produces 3 times the number of nodes, but for longer <a href="#BM_(sequence)%2Fi">sequence</a>s it has far less edges than the naive method.</p>
<p><a href="ch5_code/src/affine_gap_alignment/AffineGapAlignment_Layer_Graph.py">ch5_code/src/affine_gap_alignment/AffineGapAlignment_Layer_Graph.py</a> (lines 37 to 135):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">create_affine_gap_alignment_graph</span>(<span class="hljs-params">
        v: <span class="hljs-type">List</span>[ELEM],
        w: <span class="hljs-type">List</span>[ELEM],
        weight_lookup: WeightLookup,
        extended_gap_weight: <span class="hljs-built_in">float</span>
</span>) -&gt; Graph[<span class="hljs-type">Tuple</span>[<span class="hljs-built_in">int</span>, ...], NodeData, <span class="hljs-built_in">str</span>, EdgeData]:
    graph_low = create_grid_graph(
        [v, w],
        <span class="hljs-keyword">lambda</span> n_id: NodeData(),
        <span class="hljs-keyword">lambda</span> src_n_id, dst_n_id, offset, elems: EdgeData(elems[<span class="hljs-number">0</span>], elems[<span class="hljs-number">1</span>], extended_gap_weight) <span class="hljs-keyword">if</span> offset == (<span class="hljs-number">1</span>, <span class="hljs-number">0</span>) <span class="hljs-keyword">else</span> <span class="hljs-literal">None</span>
    )
    graph_mid = create_grid_graph(
        [v, w],
        <span class="hljs-keyword">lambda</span> n_id: NodeData(),
        <span class="hljs-keyword">lambda</span> src_n_id, dst_n_id, offset, elems: EdgeData(elems[<span class="hljs-number">0</span>], elems[<span class="hljs-number">1</span>], weight_lookup.lookup(*elems)) <span class="hljs-keyword">if</span> offset == (<span class="hljs-number">1</span>, <span class="hljs-number">1</span>) <span class="hljs-keyword">else</span> <span class="hljs-literal">None</span>
    )
    graph_high = create_grid_graph(
        [v, w],
        <span class="hljs-keyword">lambda</span> n_id: NodeData(),
        <span class="hljs-keyword">lambda</span> src_n_id, dst_n_id, offset, elems: EdgeData(elems[<span class="hljs-number">0</span>], elems[<span class="hljs-number">1</span>], extended_gap_weight) <span class="hljs-keyword">if</span> offset == (<span class="hljs-number">0</span>, <span class="hljs-number">1</span>) <span class="hljs-keyword">else</span> <span class="hljs-literal">None</span>
    )

    graph_merged = Graph()
    create_edge_id_func = unique_id_generator(<span class="hljs-string">'E'</span>)

    <span class="hljs-keyword">def</span> <span class="hljs-title function_">merge</span>(<span class="hljs-params">from_graph, n_prefix</span>):
        <span class="hljs-keyword">for</span> n_id <span class="hljs-keyword">in</span> from_graph.get_nodes():
            n_data = from_graph.get_node_data(n_id)
            graph_merged.insert_node(n_prefix + n_id, n_data)
        <span class="hljs-keyword">for</span> e_id <span class="hljs-keyword">in</span> from_graph.get_edges():
            from_n_id, to_n_id, e_data = from_graph.get_edge(e_id)
            graph_merged.insert_edge(create_edge_id_func(), n_prefix + from_n_id, n_prefix + to_n_id, e_data)

    merge(graph_low, (<span class="hljs-string">'high'</span>, ))
    merge(graph_mid, (<span class="hljs-string">'mid'</span>,))
    merge(graph_high, (<span class="hljs-string">'low'</span>,))

    v_node_count = <span class="hljs-built_in">len</span>(v) + <span class="hljs-number">1</span>
    w_node_count = <span class="hljs-built_in">len</span>(w) + <span class="hljs-number">1</span>
    mid_to_low_edge_id_func = unique_id_generator(<span class="hljs-string">'MID_TO_LOW'</span>)
    <span class="hljs-keyword">for</span> r, c <span class="hljs-keyword">in</span> product(<span class="hljs-built_in">range</span>(v_node_count - <span class="hljs-number">1</span>), <span class="hljs-built_in">range</span>(w_node_count)):
        from_n_id = <span class="hljs-string">'mid'</span>, r, c
        to_n_id = <span class="hljs-string">'high'</span>, r + <span class="hljs-number">1</span>, c
        e = mid_to_low_edge_id_func()
        graph_merged.insert_edge(e, from_n_id, to_n_id, EdgeData(v[r], <span class="hljs-literal">None</span>, weight_lookup.lookup(v[r], <span class="hljs-literal">None</span>)))
    low_to_mid_edge_id_func = unique_id_generator(<span class="hljs-string">'HIGH_TO_MID'</span>)
    <span class="hljs-keyword">for</span> r, c <span class="hljs-keyword">in</span> product(<span class="hljs-built_in">range</span>(<span class="hljs-number">1</span>, v_node_count), <span class="hljs-built_in">range</span>(w_node_count)):
        from_n_id = <span class="hljs-string">'high'</span>, r, c
        to_n_id = <span class="hljs-string">'mid'</span>, r, c
        e = low_to_mid_edge_id_func()
        graph_merged.insert_edge(e, from_n_id, to_n_id, EdgeData(<span class="hljs-literal">None</span>, <span class="hljs-literal">None</span>, <span class="hljs-number">0.0</span>))
    mid_to_high_edge_id_func = unique_id_generator(<span class="hljs-string">'MID_TO_HIGH'</span>)
    <span class="hljs-keyword">for</span> r, c <span class="hljs-keyword">in</span> product(<span class="hljs-built_in">range</span>(v_node_count), <span class="hljs-built_in">range</span>(w_node_count - <span class="hljs-number">1</span>)):
        from_n_id = <span class="hljs-string">'mid'</span>, r, c
        to_n_id = <span class="hljs-string">'low'</span>, r, c + <span class="hljs-number">1</span>
        e = mid_to_high_edge_id_func()
        graph_merged.insert_edge(e, from_n_id, to_n_id, EdgeData(<span class="hljs-literal">None</span>, w[c], weight_lookup.lookup(<span class="hljs-literal">None</span>, w[c])))
    high_to_mid_edge_id_func = unique_id_generator(<span class="hljs-string">'LOW_TO_MID'</span>)
    <span class="hljs-keyword">for</span> r, c <span class="hljs-keyword">in</span> product(<span class="hljs-built_in">range</span>(v_node_count), <span class="hljs-built_in">range</span>(<span class="hljs-number">1</span>, w_node_count)):
        from_n_id = <span class="hljs-string">'low'</span>, r, c
        to_n_id = <span class="hljs-string">'mid'</span>, r, c
        e = high_to_mid_edge_id_func()
        graph_merged.insert_edge(e, from_n_id, to_n_id, EdgeData(<span class="hljs-literal">None</span>, <span class="hljs-literal">None</span>, <span class="hljs-number">0.0</span>))

    <span class="hljs-keyword">return</span> graph_merged


<span class="hljs-keyword">def</span> <span class="hljs-title function_">affine_gap_alignment</span>(<span class="hljs-params">
        v: <span class="hljs-type">List</span>[ELEM],
        w: <span class="hljs-type">List</span>[ELEM],
        weight_lookup: WeightLookup,
        extended_gap_weight: <span class="hljs-built_in">float</span>
</span>) -&gt; <span class="hljs-type">Tuple</span>[<span class="hljs-built_in">float</span>, <span class="hljs-type">List</span>[<span class="hljs-built_in">str</span>], <span class="hljs-type">List</span>[<span class="hljs-type">Tuple</span>[ELEM, ELEM]]]:
    v_node_count = <span class="hljs-built_in">len</span>(v) + <span class="hljs-number">1</span>
    w_node_count = <span class="hljs-built_in">len</span>(w) + <span class="hljs-number">1</span>
    graph = create_affine_gap_alignment_graph(v, w, weight_lookup, extended_gap_weight)
    from_node = (<span class="hljs-string">'mid'</span>, <span class="hljs-number">0</span>, <span class="hljs-number">0</span>)
    to_node = (<span class="hljs-string">'mid'</span>, v_node_count - <span class="hljs-number">1</span>, w_node_count - <span class="hljs-number">1</span>)
    populate_weights_and_backtrack_pointers(
        graph,
        from_node,
        <span class="hljs-keyword">lambda</span> n_id, weight, e_id: graph.get_node_data(n_id).set_weight_and_backtracking_edge(weight, e_id),
        <span class="hljs-keyword">lambda</span> n_id: graph.get_node_data(n_id).get_weight_and_backtracking_edge(),
        <span class="hljs-keyword">lambda</span> e_id: graph.get_edge_data(e_id).weight
    )
    final_weight = graph.get_node_data(to_node).weight
    edges = backtrack(
        graph,
        to_node,
        <span class="hljs-keyword">lambda</span> n_id: graph.get_node_data(n_id).get_weight_and_backtracking_edge()
    )
    edges = <span class="hljs-built_in">list</span>(<span class="hljs-built_in">filter</span>(<span class="hljs-keyword">lambda</span> e: <span class="hljs-keyword">not</span> e.startswith(<span class="hljs-string">'LOW_TO_MID'</span>), edges))  <span class="hljs-comment"># remove free rides from list</span>
    edges = <span class="hljs-built_in">list</span>(<span class="hljs-built_in">filter</span>(<span class="hljs-keyword">lambda</span> e: <span class="hljs-keyword">not</span> e.startswith(<span class="hljs-string">'HIGH_TO_MID'</span>), edges))  <span class="hljs-comment"># remove free rides from list</span>
    alignment = []
    <span class="hljs-keyword">for</span> e <span class="hljs-keyword">in</span> edges:
        ed = graph.get_edge_data(e)
        alignment.append((ed.v_elem, ed.w_elem))
    <span class="hljs-keyword">return</span> final_weight, edges, alignment
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the sequences TGGCGG and TCCCCC and the score matrix...</p>
<pre><code>INDEL=-1.0
    A   C   T   G
A   1  -1  -1  -1
C  -1   1  -1  -1
T  -1  -1   1  -1
G  -1  -1  -1   1

</code></pre>
<p>... the global alignment is...</p>
<p><img src="latex_e3d930cac12ca600babd456891386935.svg" alt="Latex diagram"></p>
<pre><code>TGGC----GG
T--CCCCC--
</code></pre>
<p>Weight: -1.5</p>
</div>
<p></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The algorithm above was applied on <a href="#BM_(global%20alignment%7Cglobal%20sequence%20alignment)%2Fi">global alignment</a>, but it should be obvious how to apply it to the other alignment types discussed.</p>
</div>
<a name="H_Multiple%20Alignment"></a><h3>Multiple Alignment</h3>
<p><a name="BM_(Algorithms%5C%2FSequence%20Alignment%2FMultiple%20Alignment)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FGlobal%20Alignment)_TOPIC%2F">Algorithms/Sequence Alignment/Global Alignment</a></li>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FLocal%20Alignment)_TOPIC%2F">Algorithms/Sequence Alignment/Local Alignment</a></li>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FFitting%20Alignment)_TOPIC%2F">Algorithms/Sequence Alignment/Fitting Alignment</a></li>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FOverlap%20Alignment)_TOPIC%2F">Algorithms/Sequence Alignment/Overlap Alignment</a></li>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FProtein%20Scoring)_TOPIC%2F">Algorithms/Sequence Alignment/Protein Scoring</a></li>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FExtended%20Gap%20Scoring)_TOPIC%2F">Algorithms/Sequence Alignment/Extended Gap Scoring</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: Given <em>more than</em> two <a href="#BM_(sequence)%2Fi">sequence</a>s, perform <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a> and pull out the highest scoring alignment.</p>
<p><strong>WHY</strong>: Proteins that perform the same function but are distantly related are likely to have similar regions. The problem is that a 2-way <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a> may have a hard time identifying those similar regions, where as an n-way <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a> (n &gt; 2) will likely reveal much more / more accurate regions.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>Quote from Pevzner book: "Bioinformaticians sometimes say that pairwise alignment whispers and multiple alignment shouts."</p>
</div>
<a name="H4_Graph%20Algorithm"></a><h4>Graph Algorithm</h4>
<p><a name="BM_(Algorithms%5C%2FSequence%20Alignment%2FMultiple%20Alignment%2FGraph%20Algorithm)_TOPIC%2F"></a><strong></strong></p>
<p><strong>ALGORITHM</strong>:</p>
<p>Thinking about <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a> geometrically, adding another <a href="#BM_(sequence)%2Fi">sequence</a> to a <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">sequence alignment graph</a> is akin to adding a new dimension. For example, a <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">sequence alignment graph</a> with...</p>
<ul>
<li>2 <a href="#BM_(sequence)%2Fi">sequence</a>s is represented as a 2D square.</li>
<li>3 <a href="#BM_(sequence)%2Fi">sequence</a>s is represented as a 3D cube.</li>
<li>4 <a href="#BM_(sequence)%2Fi">sequence</a>s is represented as a 4D hypercube.</li>
<li>etc..</li>
</ul>
<p><img src="svgbob_31f114194be61c9e0ce6265f3e9c720b06ef504a.svg" alt="Kroki diagram output"></p>
<p>The alignment possibilities at each step of a <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a> may be thought of as a vertex shooting out edges to all other vertices in the geometry. For example, in a <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a> with 2 <a href="#BM_(sequence)%2Fi">sequence</a>s, the vertex (0, 0) shoots out an edge to vertices ...</p>
<ul>
<li>(1, 0) - represents keep first but skip second.</li>
<li>(0, 1) - represents skip first but keep second.</li>
<li>(1, 1) - represents keep both.</li>
</ul>
<p>The vertex coordinates may be thought of as analogs of whether to keep or skip an element. Each coordinate position corresponds to a <a href="#BM_(sequence)%2Fi">sequence</a> element (first coordinate = first <a href="#BM_(sequence)%2Fi">sequence</a>'s element, second coordinate = second <a href="#BM_(sequence)%2Fi">sequence</a>'s element). If a coordinate is set to ...</p>
<ul>
<li>1, the element is kept.</li>
<li>0, the element is skipped.</li>
</ul>
<p><img src="latex_5f4a81bfd6561ec2c18f08969db5daf0.svg" alt="Latex diagram"></p>
<p>This same logic extends to <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a> with 3 or more <a href="#BM_(sequence)%2Fi">sequence</a>s. For example, in a <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a> with 3 <a href="#BM_(sequence)%2Fi">sequence</a>s, the vertex (0, 0, 0) shoots out an edge to all other vertices in the cube. The vertex coordinates define which <a href="#BM_(sequence)%2Fi">sequence</a> elements should be kept or skipped based on the same rules described above.</p>
<p><img src="latex_fb6e6fee4fbb05905be8f4a36924e862.svg" alt="Latex diagram"></p>
<p><a href="ch5_code/src/graph/GraphGridCreate.py">ch5_code/src/graph/GraphGridCreate.py</a> (lines 31 to 58):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">create_grid_graph</span>(<span class="hljs-params">
        sequences: <span class="hljs-type">List</span>[<span class="hljs-type">List</span>[ELEM]],
        on_new_node: ON_NEW_NODE_FUNC_TYPE,
        on_new_edge: ON_NEW_EDGE_FUNC_TYPE
</span>) -&gt; Graph[<span class="hljs-type">Tuple</span>[<span class="hljs-built_in">int</span>, ...], ND, <span class="hljs-built_in">str</span>, ED]:
    create_edge_id_func = unique_id_generator(<span class="hljs-string">'E'</span>)
    graph = Graph()
    axes = [[<span class="hljs-literal">None</span>] + av <span class="hljs-keyword">for</span> av <span class="hljs-keyword">in</span> sequences]
    axes_len = [<span class="hljs-built_in">range</span>(<span class="hljs-built_in">len</span>(axis)) <span class="hljs-keyword">for</span> axis <span class="hljs-keyword">in</span> axes]
    <span class="hljs-keyword">for</span> grid_coord <span class="hljs-keyword">in</span> product(*axes_len):
        node_data = on_new_node(grid_coord)
        <span class="hljs-keyword">if</span> node_data <span class="hljs-keyword">is</span> <span class="hljs-keyword">not</span> <span class="hljs-literal">None</span>:
            graph.insert_node(grid_coord, node_data)
    <span class="hljs-keyword">for</span> src_grid_coord <span class="hljs-keyword">in</span> graph.get_nodes():
        <span class="hljs-keyword">for</span> grid_coord_offsets <span class="hljs-keyword">in</span> product([<span class="hljs-number">0</span>, <span class="hljs-number">1</span>], repeat=<span class="hljs-built_in">len</span>(sequences)):
            dst_grid_coord = <span class="hljs-built_in">tuple</span>(axis + offset <span class="hljs-keyword">for</span> axis, offset <span class="hljs-keyword">in</span> <span class="hljs-built_in">zip</span>(src_grid_coord, grid_coord_offsets))
            <span class="hljs-keyword">if</span> src_grid_coord == dst_grid_coord:  <span class="hljs-comment"># skip if making a connection to self</span>
                <span class="hljs-keyword">continue</span>
            <span class="hljs-keyword">if</span> <span class="hljs-keyword">not</span> graph.has_node(dst_grid_coord):  <span class="hljs-comment"># skip if neighbouring node doesn't exist</span>
                <span class="hljs-keyword">continue</span>
            elements = <span class="hljs-built_in">tuple</span>(<span class="hljs-literal">None</span> <span class="hljs-keyword">if</span> src_idx == dst_idx <span class="hljs-keyword">else</span> axes[axis_idx][dst_idx]
                             <span class="hljs-keyword">for</span> axis_idx, (src_idx, dst_idx) <span class="hljs-keyword">in</span> <span class="hljs-built_in">enumerate</span>(<span class="hljs-built_in">zip</span>(src_grid_coord, dst_grid_coord)))
            edge_data = on_new_edge(src_grid_coord, dst_grid_coord, grid_coord_offsets, elements)
            <span class="hljs-keyword">if</span> edge_data <span class="hljs-keyword">is</span> <span class="hljs-keyword">not</span> <span class="hljs-literal">None</span>:
                edge_id = create_edge_id_func()
                graph.insert_edge(edge_id, src_grid_coord, dst_grid_coord, edge_data)
    <span class="hljs-keyword">return</span> graph
</code></pre>
<p><a href="ch5_code/src/global_alignment/GlobalMultipleAlignment_Graph.py">ch5_code/src/global_alignment/GlobalMultipleAlignment_Graph.py</a> (lines 33 to 71):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">create_global_alignment_graph</span>(<span class="hljs-params">
        seqs: <span class="hljs-type">List</span>[<span class="hljs-type">List</span>[ELEM]],
        weight_lookup: WeightLookup
</span>) -&gt; Graph[<span class="hljs-type">Tuple</span>[<span class="hljs-built_in">int</span>, ...], NodeData, <span class="hljs-built_in">str</span>, EdgeData]:
    graph = create_grid_graph(
        seqs,
        <span class="hljs-keyword">lambda</span> n_id: NodeData(),
        <span class="hljs-keyword">lambda</span> src_n_id, dst_n_id, offset, elems: EdgeData(elems, weight_lookup.lookup(*elems))
    )
    <span class="hljs-keyword">return</span> graph


<span class="hljs-keyword">def</span> <span class="hljs-title function_">global_alignment</span>(<span class="hljs-params">
        seqs: <span class="hljs-type">List</span>[<span class="hljs-type">List</span>[ELEM]],
        weight_lookup: WeightLookup
</span>) -&gt; <span class="hljs-type">Tuple</span>[<span class="hljs-built_in">float</span>, <span class="hljs-type">List</span>[<span class="hljs-built_in">str</span>], <span class="hljs-type">List</span>[<span class="hljs-type">Tuple</span>[ELEM, ...]]]:
    seq_node_counts = [<span class="hljs-built_in">len</span>(s) <span class="hljs-keyword">for</span> s <span class="hljs-keyword">in</span> seqs]
    graph = create_global_alignment_graph(seqs, weight_lookup)
    from_node = <span class="hljs-built_in">tuple</span>([<span class="hljs-number">0</span>] * <span class="hljs-built_in">len</span>(seqs))
    to_node = <span class="hljs-built_in">tuple</span>(seq_node_counts)
    populate_weights_and_backtrack_pointers(
        graph,
        from_node,
        <span class="hljs-keyword">lambda</span> n_id, weight, e_id: graph.get_node_data(n_id).set_weight_and_backtracking_edge(weight, e_id),
        <span class="hljs-keyword">lambda</span> n_id: graph.get_node_data(n_id).get_weight_and_backtracking_edge(),
        <span class="hljs-keyword">lambda</span> e_id: graph.get_edge_data(e_id).weight
    )
    final_weight = graph.get_node_data(to_node).weight
    edges = backtrack(
        graph,
        to_node,
        <span class="hljs-keyword">lambda</span> n_id: graph.get_node_data(n_id).get_weight_and_backtracking_edge()
    )
    alignment = []
    <span class="hljs-keyword">for</span> e <span class="hljs-keyword">in</span> edges:
        ed = graph.get_edge_data(e)
        alignment.append(ed.elems)
    <span class="hljs-keyword">return</span> final_weight, edges, alignment
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the sequences ['TATTATTAT', 'GATTATGATTAT', 'TACCATTACAT'] and the score matrix...</p>
<pre><code>INDEL=-1.0
A A A 1
A A C -1
A A T -1
A A G -1
A C A -1
A C C -1
A C T -1
A C G -1
A T A -1
A T C -1
A T T -1
A T G -1
A G A -1
A G C -1
A G T -1
A G G -1
C A A -1
C A C -1
C A T -1
C A G -1
C C A -1
C C C 1
C C T -1
C C G -1
C T A -1
C T C -1
C T T -1
C T G -1
C G A -1
C G C -1
C G T -1
C G G -1
T A A -1
T A C -1
T A T -1
T A G -1
T C A -1
T C C -1
T C T -1
T C G -1
T T A -1
T T C -1
T T T 1
T T G -1
T G A -1
T G C -1
T G T -1
T G G -1
G A A -1
G A C -1
G A T -1
G A G -1
G C A -1
G C C -1
G C T -1
G C G -1
G T A -1
G T C -1
G T T -1
G T G -1
G G A -1
G G C -1
G G T -1
G G G 1
</code></pre>
<p>... the global alignment is...</p>
<pre><code>--T-ATTATTA--T
GATTATGATTA--T
--T-ACCATTACAT
</code></pre>
<p>Weight: 0.0</p>
</div>
<p></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The multiple alignment algorithm displayed above was specifically for on <a href="#BM_(global%20alignment%7Cglobal%20sequence%20alignment)%2Fi">global alignment</a> on a graph implementation, but it should be obvious how to apply it to most of the other alignment types (e.g. <a href="#BM_(local%20alignment%7Clocal%20sequence%20alignment)%2Fi">local alignment</a>).</p>
</div>
<a name="H4_Matrix%20Algorithm"></a><h4>Matrix Algorithm</h4>
<p><a name="BM_(Algorithms%5C%2FSequence%20Alignment%2FMultiple%20Alignment%2FMatrix%20Algorithm)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FMultiple%20Alignment%2FGraph%20Algorithm)_TOPIC%2F">Algorithms/Sequence Alignment/Multiple Alignment/Graph Algorithm</a></li>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FGlobal%20Alignment%2FMatrix%20Algorithm)_TOPIC%2F">Algorithms/Sequence Alignment/Global Alignment/Matrix Algorithm</a></li>
</ul>
</div>
<p>The following algorithm is essentially the same as the graph algorithm, except that the implementation is much more sympathetic to modern hardware. The <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a> is represented as an N-dimensional matrix where each element in the matrix represents a node in the <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a>. This is similar to the 2D matrix used for <a href="#BM_(global%20alignment%7Cglobal%20sequence%20alignment)%2Fi">global alignment</a>'s matrix implementation.</p>
<p><a href="ch5_code/src/global_alignment/GlobalMultipleAlignment_Matrix.py">ch5_code/src/global_alignment/GlobalMultipleAlignment_Matrix.py</a> (lines 12 to 79):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">generate_matrix</span>(<span class="hljs-params">seq_node_counts: <span class="hljs-type">List</span>[<span class="hljs-built_in">int</span>]</span>) -&gt; <span class="hljs-type">List</span>[<span class="hljs-type">Any</span>]:
    last_buffer = [[-<span class="hljs-number">1.0</span>, (<span class="hljs-literal">None</span>, <span class="hljs-literal">None</span>), <span class="hljs-string">'?'</span>] <span class="hljs-keyword">for</span> _ <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(seq_node_counts[-<span class="hljs-number">1</span>])]  <span class="hljs-comment"># row</span>
    <span class="hljs-keyword">for</span> dim <span class="hljs-keyword">in</span> <span class="hljs-built_in">reversed</span>(seq_node_counts[:-<span class="hljs-number">1</span>]):
        <span class="hljs-comment"># DON'T USE DEEPCOPY -- VERY SLOW: https://stackoverflow.com/a/29385667</span>
        <span class="hljs-comment"># last_buffer = [deepcopy(last_buffer) for _ in range(dim)]</span>
        last_buffer = [pickle.loads(pickle.dumps(last_buffer, -<span class="hljs-number">1</span>)) <span class="hljs-keyword">for</span> _ <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(dim)]
    <span class="hljs-keyword">return</span> last_buffer


<span class="hljs-keyword">def</span> <span class="hljs-title function_">get_cell</span>(<span class="hljs-params">matrix: <span class="hljs-type">List</span>[<span class="hljs-type">Any</span>], idxes: Iterable[<span class="hljs-built_in">int</span>]</span>):
    buffer = matrix
    <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> idxes:
        buffer = buffer[i]
    <span class="hljs-keyword">return</span> buffer


<span class="hljs-keyword">def</span> <span class="hljs-title function_">set_cell</span>(<span class="hljs-params">matrix: <span class="hljs-type">List</span>[<span class="hljs-type">Any</span>], idxes: Iterable[<span class="hljs-built_in">int</span>], value: <span class="hljs-type">Any</span></span>):
    buffer = matrix
    <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> idxes[:-<span class="hljs-number">1</span>]:
        buffer = buffer[i]
    buffer[idxes[-<span class="hljs-number">1</span>]] = value


<span class="hljs-keyword">def</span> <span class="hljs-title function_">backtrack</span>(<span class="hljs-params">
        node_matrix: <span class="hljs-type">List</span>[<span class="hljs-type">List</span>[<span class="hljs-type">Any</span>]],
        dimensions: <span class="hljs-type">List</span>[<span class="hljs-built_in">int</span>]
</span>) -&gt; <span class="hljs-type">Tuple</span>[<span class="hljs-built_in">float</span>, <span class="hljs-type">List</span>[<span class="hljs-type">Tuple</span>[ELEM, ELEM]]]:
    node_idxes = [d - <span class="hljs-number">1</span> <span class="hljs-keyword">for</span> d <span class="hljs-keyword">in</span> dimensions]
    final_weight = get_cell(node_matrix, node_idxes)[<span class="hljs-number">0</span>]
    alignment = []
    <span class="hljs-keyword">while</span> <span class="hljs-built_in">set</span>(node_idxes) != {<span class="hljs-number">0</span>}:
        _, elems, backtrack_ptr = get_cell(node_matrix, node_idxes)
        node_idxes = backtrack_ptr
        alignment.append(elems)
    <span class="hljs-keyword">return</span> final_weight, alignment[::-<span class="hljs-number">1</span>]


<span class="hljs-keyword">def</span> <span class="hljs-title function_">global_alignment</span>(<span class="hljs-params">
        seqs: <span class="hljs-type">List</span>[<span class="hljs-type">List</span>[ELEM]],
        weight_lookup: WeightLookup
</span>) -&gt; <span class="hljs-type">Tuple</span>[<span class="hljs-built_in">float</span>, <span class="hljs-type">List</span>[<span class="hljs-type">Tuple</span>[ELEM, ...]]]:
    seq_node_counts = [<span class="hljs-built_in">len</span>(s) + <span class="hljs-number">1</span> <span class="hljs-keyword">for</span> s <span class="hljs-keyword">in</span> seqs]
    node_matrix = generate_matrix(seq_node_counts)
    src_node = get_cell(node_matrix, [<span class="hljs-number">0</span>] * <span class="hljs-built_in">len</span>(seqs))
    src_node[<span class="hljs-number">0</span>] = <span class="hljs-number">0.0</span>                   <span class="hljs-comment"># source node weight</span>
    src_node[<span class="hljs-number">1</span>] = (<span class="hljs-literal">None</span>, ) * <span class="hljs-built_in">len</span>(seqs)  <span class="hljs-comment"># source node elements (elements don't matter for source node)</span>
    src_node[<span class="hljs-number">2</span>] = (<span class="hljs-literal">None</span>, ) * <span class="hljs-built_in">len</span>(seqs)  <span class="hljs-comment"># source node parent (direction doesn't matter for source node)</span>
    <span class="hljs-keyword">for</span> to_node <span class="hljs-keyword">in</span> product(*(<span class="hljs-built_in">range</span>(sc) <span class="hljs-keyword">for</span> sc <span class="hljs-keyword">in</span> seq_node_counts)):
        parents = []
        parent_idx_ranges = []
        <span class="hljs-keyword">for</span> idx <span class="hljs-keyword">in</span> to_node:
            vals = [idx]
            <span class="hljs-keyword">if</span> idx &gt; <span class="hljs-number">0</span>:
                vals += [idx-<span class="hljs-number">1</span>]
            parent_idx_ranges.append(vals)
        <span class="hljs-keyword">for</span> from_node <span class="hljs-keyword">in</span> product(*parent_idx_ranges):
            <span class="hljs-keyword">if</span> from_node == to_node:  <span class="hljs-comment"># we want indexes of parent nodes, not self</span>
                <span class="hljs-keyword">continue</span>
            edge_elems = <span class="hljs-built_in">tuple</span>(<span class="hljs-literal">None</span> <span class="hljs-keyword">if</span> f == t <span class="hljs-keyword">else</span> s[t-<span class="hljs-number">1</span>] <span class="hljs-keyword">for</span> s, f, t <span class="hljs-keyword">in</span> <span class="hljs-built_in">zip</span>(seqs, from_node, to_node))
            parents.append([
                get_cell(node_matrix, from_node)[<span class="hljs-number">0</span>] + weight_lookup.lookup(*edge_elems),
                edge_elems,
                from_node
            ])
        <span class="hljs-keyword">if</span> parents:  <span class="hljs-comment"># parents will be empty if source node</span>
            set_cell(node_matrix, to_node, <span class="hljs-built_in">max</span>(parents, key=<span class="hljs-keyword">lambda</span> x: x[<span class="hljs-number">0</span>]))
    <span class="hljs-keyword">return</span> backtrack(node_matrix, seq_node_counts)
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the sequences ['TATTATTAT', 'GATTATGATTAT', 'TACCATTACAT'] and the score matrix...</p>
<pre><code>INDEL=-1.0
A A A 1
A A C -1
A A T -1
A A G -1
A C A -1
A C C -1
A C T -1
A C G -1
A T A -1
A T C -1
A T T -1
A T G -1
A G A -1
A G C -1
A G T -1
A G G -1
C A A -1
C A C -1
C A T -1
C A G -1
C C A -1
C C C 1
C C T -1
C C G -1
C T A -1
C T C -1
C T T -1
C T G -1
C G A -1
C G C -1
C G T -1
C G G -1
T A A -1
T A C -1
T A T -1
T A G -1
T C A -1
T C C -1
T C T -1
T C G -1
T T A -1
T T C -1
T T T 1
T T G -1
T G A -1
T G C -1
T G T -1
T G G -1
G A A -1
G A C -1
G A T -1
G A G -1
G C A -1
G C C -1
G C T -1
G C G -1
G T A -1
G T C -1
G T T -1
G T G -1
G G A -1
G G C -1
G G T -1
G G G 1
</code></pre>
<p>... the global alignment is...</p>
<pre><code>--T-ATTATTA--T
GATTATGATTA--T
--T-ACCATTACAT
</code></pre>
<p>Weight: 0.0</p>
</div>
<p></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The multiple alignment algorithm displayed above was specifically for on <a href="#BM_(global%20alignment%7Cglobal%20sequence%20alignment)%2Fi">global alignment</a> on a graph implementation, but it should be obvious how to apply it to most of the other alignment types (e.g. <a href="#BM_(local%20alignment%7Clocal%20sequence%20alignment)%2Fi">local alignment</a>). With a little bit of effort it can also be converted to using the <a href="#BM_(divide-and-conquer%20algorithm%7Cdivide%20and%20conquer%20algorithm)%2Fi">divide-and-conquer algorithm</a> discussed earlier (there aren't that many leaps in logic).</p>
</div>
<a name="H1_Greedy%20Algorithm"></a><h4>Greedy Algorithm</h4>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FMultiple%20Alignment%2FGraph%20Algorithm)_TOPIC%2F">Algorithms/Sequence Alignment/Multiple Alignment/Graph Algorithm</a></li>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FMultiple%20Alignment%2FMatrix%20Algorithm)_TOPIC%2F">Algorithms/Sequence Alignment/Multiple Alignment/Matrix Algorithm</a></li>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FSum-of-Pairs%20Scoring)_TOPIC%2F">Algorithms/Sequence Alignment/Sum-of-Pairs Scoring</a></li>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FGlobal%20Alignment%2FDivide-and-Conquer%20Algorithm)_TOPIC%2F">Algorithms/Sequence Alignment/Global Alignment/Divide-and-Conquer Algorithm</a></li>
<li><a href="#BM_(Algorithms%5C%2FMotif%2FMotif%20Matrix%20Profile)_TOPIC%2F">Algorithms/Motif/Motif Matrix Profile</a></li>
</ul>
</div>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The Pevzner book challenged you to come up with a <a href="#BM_(greedy%20algorithm)%2Fi">greedy algorithm</a> for multiple alignment using profile matrices. This is what I was able to come up with. I have no idea if my logic is correct / optimal, but with toy <a href="#BM_(sequence)%2Fi">sequence</a>s that are highly related it seems to perform well.</p>
<p>UPDATE: This algorithm seems to work well for the final assignment. ~380 <a href="#BM_(adenylation%20domain%7CA%20domain%7CA-domain)%2Fi">a-domain</a> <a href="#BM_(sequence)%2Fi">sequence</a>s were aligned in about 2 days and it produced an okay/good looking alignment. Aligning those <a href="#BM_(sequence)%2Fi">sequence</a>s using normal multiple alignment would be impossible -- nowhere near enough memory or speed available.</p>
</div>
<p>For an n-way <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a>, the <a href="#BM_(greedy%20algorithm)%2Fi">greedy algorithm</a> starts by finding the 2 <a href="#BM_(sequence)%2Fi">sequence</a>s that produce the highest scoring 2-way <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a>. That alignment is then used to build a profile matrix. For example, the alignment of TRELLO and MELLOW results in the following alignment:</p>
<table>
<thead>
<tr>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
<th>6</th>
</tr>
</thead>
<tbody>
<tr>
<td>T</td>
<td>R</td>
<td>E</td>
<td>L</td>
<td>L</td>
<td>O</td>
<td>-</td>
</tr>
<tr>
<td>-</td>
<td>M</td>
<td>E</td>
<td>L</td>
<td>L</td>
<td>O</td>
<td>W</td>
</tr>
</tbody>
</table>
<p>That alignment then turns into the following profile matrix:</p>
<table>
<thead>
<tr>
<th></th>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
<th>6</th>
</tr>
</thead>
<tbody>
<tr>
<td>Probability of T</td>
<td>0.5</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr>
<td>Probability of R</td>
<td>0.0</td>
<td>0.5</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr>
<td>Probability of M</td>
<td>0.0</td>
<td>0.5</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr>
<td>Probability of E</td>
<td>0.0</td>
<td>0.0</td>
<td>1.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr>
<td>Probability of L</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>1.0</td>
<td>1.0</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr>
<td>Probability of O</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>1.0</td>
<td>0.0</td>
</tr>
<tr>
<td>Probability of W</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.5</td>
</tr>
</tbody>
</table>
<p>Then, 2-way <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a>s are performed between the <em>profile matrix</em> and the remaining <a href="#BM_(sequence)%2Fi">sequence</a>s. For example, if the letter V is scored against column 1 of the profile matrix, the algorithm would score W against each letter stored in the profile matrix using the same scoring matrix as the initial 2-way <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a>. Each score would then get weighted by the corresponding probability in column 2 and the highest one would be chosen as the final score.</p>
<pre class="hljs"><code><span class="hljs-built_in">max</span>(
    score(<span class="hljs-string">'W'</span>, <span class="hljs-string">'T'</span>) * profile_mat[<span class="hljs-number">1</span>][<span class="hljs-string">'T'</span>],
    score(<span class="hljs-string">'W'</span>, <span class="hljs-string">'R'</span>) * profile_mat[<span class="hljs-number">1</span>][<span class="hljs-string">'R'</span>],
    score(<span class="hljs-string">'W'</span>, <span class="hljs-string">'M'</span>) * profile_mat[<span class="hljs-number">1</span>][<span class="hljs-string">'M'</span>],
    score(<span class="hljs-string">'W'</span>, <span class="hljs-string">'E'</span>) * profile_mat[<span class="hljs-number">1</span>][<span class="hljs-string">'E'</span>],
    score(<span class="hljs-string">'W'</span>, <span class="hljs-string">'L'</span>) * profile_mat[<span class="hljs-number">1</span>][<span class="hljs-string">'L'</span>],
    score(<span class="hljs-string">'W'</span>, <span class="hljs-string">'O'</span>) * profile_mat[<span class="hljs-number">1</span>][<span class="hljs-string">'O'</span>],
    score(<span class="hljs-string">'W'</span>, <span class="hljs-string">'W'</span>) * profile_mat[<span class="hljs-number">1</span>][<span class="hljs-string">'W'</span>]
)
</code></pre>
<p>Of all the remaining <a href="#BM_(sequence)%2Fi">sequence</a>s, the one with the highest scoring alignment is removed and its alignment is added to the profile matrix. The process repeats until no more <a href="#BM_(sequence)%2Fi">sequence</a>s are left.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The logic above is what was used to solve the final assignment. But, after thinking about it some more it probably isn't entirely correct. Elements that haven't been encountered yet should be left unset in the profile matrix. If this change were applied, the example above would end up looking more like this...</p>
<table>
<thead>
<tr>
<th></th>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
<th>6</th>
</tr>
</thead>
<tbody>
<tr>
<td>Probability of T</td>
<td>0.5</td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td>Probability of R</td>
<td></td>
<td>0.5</td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td>Probability of M</td>
<td></td>
<td>0.5</td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td>Probability of E</td>
<td></td>
<td></td>
<td>1.0</td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td>Probability of L</td>
<td></td>
<td></td>
<td></td>
<td>1.0</td>
<td>1.0</td>
<td></td>
<td></td>
</tr>
<tr>
<td>Probability of O</td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td>1.0</td>
<td></td>
</tr>
<tr>
<td>Probability of W</td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td>0.5</td>
</tr>
</tbody>
</table>
<p>Then, when scoring an element against a column in the profile matrix, ignore the unset elements in the column. The score calculation in the example above would end up being...</p>
<pre class="hljs"><code><span class="hljs-built_in">max</span>(
    score(<span class="hljs-string">'W'</span>, <span class="hljs-string">'R'</span>) * profile_mat[<span class="hljs-number">1</span>][<span class="hljs-string">'R'</span>],
    score(<span class="hljs-string">'W'</span>, <span class="hljs-string">'M'</span>) * profile_mat[<span class="hljs-number">1</span>][<span class="hljs-string">'M'</span>]
)
</code></pre>
</div>
<p>For n-way <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a>s where n is large (e.g. n=300) and the <a href="#BM_(sequence)%2Fi">sequence</a>s are highly related, the <a href="#BM_(greedy%20algorithm)%2Fi">greedy algorithm</a> performs well but it may produce sub-optimal results. In contrast, the amount of memory and computation required for an n-way <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a> using the standard graph algorithm goes up exponentially as n grows linearly. For realistic biological <a href="#BM_(sequence)%2Fi">sequence</a>s, the normal algorithm will likely fail for any n past 3 or 4. Adapting the <a href="#BM_(divide-and-conquer%20algorithm%7Cdivide%20and%20conquer%20algorithm)%2Fi">divide-and-conquer algorithm</a> for n-way <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a> will help, but even that only allows for targeting a slightly larger n (e.g. n=6).</p>
<p><a href="ch5_code/src/global_alignment/GlobalMultipleAlignment_Greedy.py">ch5_code/src/global_alignment/GlobalMultipleAlignment_Greedy.py</a> (lines 17 to 84):</p>
<pre class="hljs"><code><span class="hljs-keyword">class</span> <span class="hljs-title class_">ProfileWeightLookup</span>(<span class="hljs-title class_ inherited__">WeightLookup</span>):
    <span class="hljs-keyword">def</span> <span class="hljs-title function_">__init__</span>(<span class="hljs-params">self, total_seqs: <span class="hljs-built_in">int</span>, backing_2d_lookup: WeightLookup</span>):
        self.total_seqs = total_seqs
        self.backing_wl = backing_2d_lookup

    <span class="hljs-keyword">def</span> <span class="hljs-title function_">lookup</span>(<span class="hljs-params">self, *elements: <span class="hljs-type">Tuple</span>[ELEM_OR_COLUMN, ...]</span>):
        col: <span class="hljs-type">Tuple</span>[ELEM, ...] = elements[<span class="hljs-number">0</span>]
        elem: ELEM = elements[<span class="hljs-number">1</span>]

        <span class="hljs-keyword">if</span> col <span class="hljs-keyword">is</span> <span class="hljs-literal">None</span>:
            <span class="hljs-keyword">return</span> self.backing_wl.lookup(elem, <span class="hljs-literal">None</span>)  <span class="hljs-comment"># should map to indel score</span>
        <span class="hljs-keyword">elif</span> elem <span class="hljs-keyword">is</span> <span class="hljs-literal">None</span>:
            <span class="hljs-keyword">return</span> self.backing_wl.lookup(<span class="hljs-literal">None</span>, col[<span class="hljs-number">0</span>])  <span class="hljs-comment"># should map to indel score</span>
        <span class="hljs-keyword">else</span>:
            probs = {elem: count / self.total_seqs <span class="hljs-keyword">for</span> elem, count <span class="hljs-keyword">in</span> Counter(e <span class="hljs-keyword">for</span> e <span class="hljs-keyword">in</span> col <span class="hljs-keyword">if</span> e <span class="hljs-keyword">is</span> <span class="hljs-keyword">not</span> <span class="hljs-literal">None</span>).items()}
            ret = <span class="hljs-number">0.0</span>
            <span class="hljs-keyword">for</span> p_elem, prob <span class="hljs-keyword">in</span> probs.items():
                val = self.backing_wl.lookup(elem, p_elem) * prob
                ret = <span class="hljs-built_in">max</span>(val, ret)
            <span class="hljs-keyword">return</span> ret


<span class="hljs-keyword">def</span> <span class="hljs-title function_">global_alignment</span>(<span class="hljs-params">
        seqs: <span class="hljs-type">List</span>[<span class="hljs-type">List</span>[ELEM]],
        weight_lookup_2way: WeightLookup,
        weight_lookup_multi: WeightLookup
</span>) -&gt; <span class="hljs-type">Tuple</span>[<span class="hljs-built_in">float</span>, <span class="hljs-type">List</span>[<span class="hljs-type">Tuple</span>[ELEM, ...]]]:
    seqs = seqs[:]  <span class="hljs-comment"># copy</span>
    <span class="hljs-comment"># Get initial best 2-way alignment</span>
    highest_res = <span class="hljs-literal">None</span>
    highest_seqs = <span class="hljs-literal">None</span>
    <span class="hljs-keyword">for</span> s1, s2 <span class="hljs-keyword">in</span> combinations(seqs, r=<span class="hljs-number">2</span>):
        <span class="hljs-keyword">if</span> s1 <span class="hljs-keyword">is</span> s2:
            <span class="hljs-keyword">continue</span>
        res = GlobalAlignment_Matrix.global_alignment(s1, s2, weight_lookup_2way)
        <span class="hljs-keyword">if</span> highest_res <span class="hljs-keyword">is</span> <span class="hljs-literal">None</span> <span class="hljs-keyword">or</span> res[<span class="hljs-number">0</span>] &gt; highest_res[<span class="hljs-number">0</span>]:
            highest_res = res
            highest_seqs = s1, s2
    seqs.remove(highest_seqs[<span class="hljs-number">0</span>])
    seqs.remove(highest_seqs[<span class="hljs-number">1</span>])
    total_seqs = <span class="hljs-number">2</span>
    final_alignment = highest_res[<span class="hljs-number">1</span>]
    <span class="hljs-comment"># Build out profile matrix from alignment and continually add to it using 2-way alignment</span>
    <span class="hljs-keyword">if</span> seqs:
        s1 = highest_res[<span class="hljs-number">1</span>]
        <span class="hljs-keyword">while</span> seqs:
            profile_weight_lookup = ProfileWeightLookup(total_seqs, weight_lookup_2way)
            _, alignment = <span class="hljs-built_in">max</span>(
                [GlobalAlignment_Matrix.global_alignment(s1, s2, profile_weight_lookup) <span class="hljs-keyword">for</span> s2 <span class="hljs-keyword">in</span> seqs],
                key=<span class="hljs-keyword">lambda</span> x: x[<span class="hljs-number">0</span>]
            )
            <span class="hljs-comment"># pull out s1 from alignment and flatten for next cycle</span>
            s1 = []
            <span class="hljs-keyword">for</span> e <span class="hljs-keyword">in</span> alignment:
                <span class="hljs-keyword">if</span> e[<span class="hljs-number">0</span>] <span class="hljs-keyword">is</span> <span class="hljs-literal">None</span>:
                    s1 += [((<span class="hljs-literal">None</span>, ) * total_seqs) + (e[<span class="hljs-number">1</span>], )]
                <span class="hljs-keyword">else</span>:
                    s1 += [(*e[<span class="hljs-number">0</span>], e[<span class="hljs-number">1</span>])]
            <span class="hljs-comment"># pull out s2 from alignment and remove from seqs</span>
            s2 = [e <span class="hljs-keyword">for</span> _, e <span class="hljs-keyword">in</span> alignment <span class="hljs-keyword">if</span> e <span class="hljs-keyword">is</span> <span class="hljs-keyword">not</span> <span class="hljs-literal">None</span>]
            seqs.remove(s2)
            <span class="hljs-comment"># increase seq count</span>
            total_seqs += <span class="hljs-number">1</span>
        final_alignment = s1
    <span class="hljs-comment"># Recalculate score based on multi weight lookup</span>
    final_weight = <span class="hljs-built_in">sum</span>(weight_lookup_multi.lookup(*elems) <span class="hljs-keyword">for</span> elems <span class="hljs-keyword">in</span> final_alignment)
    <span class="hljs-keyword">return</span> final_weight, final_alignment
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the sequences ['TATTATTAT', 'GATTATGATTAT', 'TACCATTACAT', 'CTATTAGGAT'] and the score matrix...</p>
<pre><code>INDEL=-1.0
    A   C   T   G
A   1  -1  -1  -1
C  -1   1  -1  -1
T  -1  -1   1  -1
G  -1  -1  -1   1

</code></pre>
<p>... the global alignment is...</p>
<pre><code>---TATTATTAT
GATTATGATTAT
TACCATTA-CAT
--CTATTAGGAT
</code></pre>
<p>Weight: 8.0</p>
</div>
<p></p>
<a name="H_Sum-of-Pairs%20Scoring"></a><h3>Sum-of-Pairs Scoring</h3>
<p><a name="BM_(Algorithms%5C%2FSequence%20Alignment%2FSum-of-Pairs%20Scoring)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FProtein%20Scoring)_TOPIC%2F">Algorithms/Sequence Alignment/Protein Scoring</a></li>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FMultiple%20Alignment)_TOPIC%2F">Algorithms/Sequence Alignment/Multiple Alignment</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: If a scoring model already exists for 2-way <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a>s, that scoring model can be used as the basis for n-way <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a>s (where n &gt; 2). For a possible alignment position, generate all possible pairs between the elements at that position and score them. Then, sum those scores to get the final score for that alignment position.</p>
<p><strong>WHY</strong>: Traditionally, scoring an n-way alignment requires an n-dimensional scoring matrix. For example, protein <a href="#BM_(sequence)%2Fi">sequence</a>s have 20 possible element types (1 for each <a href="#BM_(proteinogenic%20amino%20acid)%2Fi">proteinogenic amino acid</a>). That means a...</p>
<ul>
<li>2-way alignment with protein <a href="#BM_(sequence)%2Fi">sequence</a>s requires <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>2</mn><msup><mn>0</mn><mn>2</mn></msup></mrow><annotation encoding="application/x-tex">20^2</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.8141079999999999em;vertical-align:0em;"></span><span class="mord">2</span><span class="mord"><span class="mord">0</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8141079999999999em;"><span style="top:-3.063em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span></span></span></span> scores.</li>
<li>3-way alignment with protein <a href="#BM_(sequence)%2Fi">sequence</a>s requires <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>2</mn><msup><mn>0</mn><mn>3</mn></msup></mrow><annotation encoding="application/x-tex">20^3</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.8141079999999999em;vertical-align:0em;"></span><span class="mord">2</span><span class="mord"><span class="mord">0</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8141079999999999em;"><span style="top:-3.063em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">3</span></span></span></span></span></span></span></span></span></span></span> scores.</li>
<li>4-way alignment with protein <a href="#BM_(sequence)%2Fi">sequence</a>s requires <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>2</mn><msup><mn>0</mn><mn>4</mn></msup></mrow><annotation encoding="application/x-tex">20^4</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.8141079999999999em;vertical-align:0em;"></span><span class="mord">2</span><span class="mord"><span class="mord">0</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8141079999999999em;"><span style="top:-3.063em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">4</span></span></span></span></span></span></span></span></span></span></span> scores.</li>
<li>5-way alignment with protein <a href="#BM_(sequence)%2Fi">sequence</a>s requires <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>2</mn><msup><mn>0</mn><mn>5</mn></msup></mrow><annotation encoding="application/x-tex">20^5</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.8141079999999999em;vertical-align:0em;"></span><span class="mord">2</span><span class="mord"><span class="mord">0</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8141079999999999em;"><span style="top:-3.063em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">5</span></span></span></span></span></span></span></span></span></span></span> scores.</li>
<li>...</li>
</ul>
<p>Creating probabilistic scoring models such a <a href="#BM_%5Cb(BLOSUM)%5Cd*%5Cb%2F">BLOSUM</a> and <a href="#BM_%5Cb(PAM)%5Cd*%5Cb%2F">PAM</a> for n-way alignments where n &gt; 2 is impractical. Sum-of-pairs scoring is a viable alternative.</p>
<p><strong>ALGORITHM</strong>:</p>
<p><a href="ch5_code/src/scoring/SumOfPairsWeightLookup.py">ch5_code/src/scoring/SumOfPairsWeightLookup.py</a> (lines 8 to 14):</p>
<pre class="hljs"><code><span class="hljs-keyword">class</span> <span class="hljs-title class_">SumOfPairsWeightLookup</span>(<span class="hljs-title class_ inherited__">WeightLookup</span>):
    <span class="hljs-keyword">def</span> <span class="hljs-title function_">__init__</span>(<span class="hljs-params">self, backing_2d_lookup: WeightLookup</span>):
        self.backing_wl = backing_2d_lookup

    <span class="hljs-keyword">def</span> <span class="hljs-title function_">lookup</span>(<span class="hljs-params">self, *elements: <span class="hljs-type">Tuple</span>[<span class="hljs-type">Optional</span>[ELEM], ...]</span>):
        <span class="hljs-keyword">return</span> <span class="hljs-built_in">sum</span>(self.backing_wl.lookup(a, b) <span class="hljs-keyword">for</span> a, b <span class="hljs-keyword">in</span> combinations(elements, r=<span class="hljs-number">2</span>))
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the elements ['M', 'E', 'A', None, 'L', 'Y'] and the backing score matrix...</p>
<pre><code>INDEL=-1.0
   A  C  D  E  F  G  H  I  K  L  M  N  P  Q  R  S  T  V  W  Y
A  4  0 -2 -1 -2  0 -2 -1 -1 -1 -1 -2 -1 -1 -1  1  0  0 -3 -2
C  0  9 -3 -4 -2 -3 -3 -1 -3 -1 -1 -3 -3 -3 -3 -1 -1 -1 -2 -2
D -2 -3  6  2 -3 -1 -1 -3 -1 -4 -3  1 -1  0 -2  0 -1 -3 -4 -3
E -1 -4  2  5 -3 -2  0 -3  1 -3 -2  0 -1  2  0  0 -1 -2 -3 -2
F -2 -2 -3 -3  6 -3 -1  0 -3  0  0 -3 -4 -3 -3 -2 -2 -1  1  3
G  0 -3 -1 -2 -3  6 -2 -4 -2 -4 -3  0 -2 -2 -2  0 -2 -3 -2 -3
H -2 -3 -1  0 -1 -2  8 -3 -1 -3 -2  1 -2  0  0 -1 -2 -3 -2  2
I -1 -1 -3 -3  0 -4 -3  4 -3  2  1 -3 -3 -3 -3 -2 -1  3 -3 -1
K -1 -3 -1  1 -3 -2 -1 -3  5 -2 -1  0 -1  1  2  0 -1 -2 -3 -2
L -1 -1 -4 -3  0 -4 -3  2 -2  4  2 -3 -3 -2 -2 -2 -1  1 -2 -1
M -1 -1 -3 -2  0 -3 -2  1 -1  2  5 -2 -2  0 -1 -1 -1  1 -1 -1
N -2 -3  1  0 -3  0  1 -3  0 -3 -2  6 -2  0  0  1  0 -3 -4 -2
P -1 -3 -1 -1 -4 -2 -2 -3 -1 -3 -2 -2  7 -1 -2 -1 -1 -2 -4 -3
Q -1 -3  0  2 -3 -2  0 -3  1 -2  0  0 -1  5  1  0 -1 -2 -2 -1
R -1 -3 -2  0 -3 -2  0 -3  2 -2 -1  0 -2  1  5 -1 -1 -3 -3 -2
S  1 -1  0  0 -2  0 -1 -2  0 -2 -1  1 -1  0 -1  4  1 -2 -3 -2
T  0 -1 -1 -1 -2 -2 -2 -1 -1 -1 -1  0 -1 -1 -1  1  5  0 -2 -2
V  0 -1 -3 -2 -1 -3 -3  3 -2  1  1 -3 -2 -2 -3 -2  0  4 -3 -1
W -3 -2 -4 -3  1 -2 -2 -3 -3 -2 -1 -4 -4 -2 -3 -3 -2 -3 11  2
Y -2 -2 -3 -2  3 -3  2 -1 -2 -1 -1 -2 -3 -1 -2 -2 -2 -1  2  7
</code></pre>
<p>... the sum-of-pairs score for these elements is -17.0.</p>
</div>
<p></p>
<a name="H_Entropy%20Scoring"></a><h3>Entropy Scoring</h3>
<p><a name="BM_(Algorithms%5C%2FSequence%20Alignment%2FEntropy%20Scoring)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FMotif%2FMotif%20Matrix%20Score%2FEntropy%20Algorithm)_TOPIC%2F">Algorithms/Motif/Motif Matrix Score/Entropy Algorithm</a></li>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FMultiple%20Alignment)_TOPIC%2F">Algorithms/Sequence Alignment/Multiple Alignment</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: When performing an n-way <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a>, score each possible alignment position based on <a href="#BM_(entropy)%2Fi">entropy</a>.</p>
<p><strong>WHY</strong>: <a href="#BM_(entropy)%2Fi">Entropy</a> is a measure of uncertainty. The idea is that the more "certain" an alignment position is, the more likely it is to be correct.</p>
<p><strong>ALGORITHM</strong>:</p>
<p><a href="ch5_code/src/scoring/EntropyWeightLookup.py">ch5_code/src/scoring/EntropyWeightLookup.py</a> (lines 9 to 31):</p>
<pre class="hljs"><code><span class="hljs-keyword">class</span> <span class="hljs-title class_">EntropyWeightLookup</span>(<span class="hljs-title class_ inherited__">WeightLookup</span>):
    <span class="hljs-keyword">def</span> <span class="hljs-title function_">__init__</span>(<span class="hljs-params">self, indel_weight: <span class="hljs-built_in">float</span></span>):
        self.indel_weight = indel_weight

<span class="hljs-meta">    @staticmethod</span>
    <span class="hljs-keyword">def</span> <span class="hljs-title function_">_calculate_entropy</span>(<span class="hljs-params">values: <span class="hljs-type">Tuple</span>[<span class="hljs-built_in">float</span>, ...]</span>) -&gt; <span class="hljs-built_in">float</span>:
        ret = <span class="hljs-number">0.0</span>
        <span class="hljs-keyword">for</span> value <span class="hljs-keyword">in</span> values:
            ret += value * (log(value, <span class="hljs-number">2.0</span>) <span class="hljs-keyword">if</span> value &gt; <span class="hljs-number">0.0</span> <span class="hljs-keyword">else</span> <span class="hljs-number">0.0</span>)
        ret = -ret
        <span class="hljs-keyword">return</span> ret

    <span class="hljs-keyword">def</span> <span class="hljs-title function_">lookup</span>(<span class="hljs-params">self, *elements: <span class="hljs-type">Tuple</span>[<span class="hljs-type">Optional</span>[ELEM], ...]</span>):
        <span class="hljs-keyword">if</span> <span class="hljs-literal">None</span> <span class="hljs-keyword">in</span> elements:
            <span class="hljs-keyword">return</span> self.indel_weight

        counts = Counter(elements)
        total = <span class="hljs-built_in">len</span>(elements)
        probs = <span class="hljs-built_in">tuple</span>(v / total <span class="hljs-keyword">for</span> k, v <span class="hljs-keyword">in</span> counts.most_common())
        entropy = EntropyWeightLookup._calculate_entropy(probs)

        <span class="hljs-keyword">return</span> -entropy
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the elements ['A', 'A', 'A', 'A', 'C'], the entropy score for these elements is -0.7219280948873623 (INDEL=-2.0).</p>
</div>
<p></p>
<a name="H_Synteny"></a><h2>Synteny</h2>
<p><a name="BM_(Algorithms%5C%2FSynteny)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FK-mer)_TOPIC%2F">Algorithms/K-mer</a></li>
</ul>
</div>
<p>A form of DNA mutation, called <a href="#BM_(genome%20rearrangement%7Cchromosomal%20rearrangement%7Cchromosome%20rearrangement)%2Fi">genome rearrangement</a>, is when chromosomes go through structural changes such as ...</p>
<ul>
<li>shuffling segments (<a href="#BM_(translocation)%2Fi">translocation</a>, <a href="#BM_(fission)%2Fi">fission</a>, <a href="#BM_(%5Cbfusion%7C%5Cbfuse)%2Fi">fusion</a>)</li>
<li>flipping direction (reversal).</li>
<li>duplicating segments.</li>
<li>deleting segments.</li>
</ul>
<p>When a new species branches off from an existing one, <a href="#BM_(genome%20rearrangement%7Cchromosomal%20rearrangement%7Cchromosome%20rearrangement)%2Fi">genome rearrangement</a>s are responsible for at least some of the divergence. That is, the two related <a href="#BM_(genome)%2Fi">genome</a>s will share long stretches of similar <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">genes</a>, but these long stretches will appear as if they had been randomly cut-and-paste and / or randomly reversed when compared to the other.</p>
<p><img src="svgbob_d5f827e20974efd41c9ac222d75bd5443f3d190d.svg" alt="Kroki diagram output"></p>
<p>These long stretches of similar <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">genes</a> are called <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s. The example above has 4 <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s:</p>
<ul>
<li>[G1, G2]</li>
<li>[G3]</li>
<li>[G4, G5, G6], although they're reversed</li>
<li>[G7, G8, G9]</li>
</ul>
<p><img src="svgbob_f80dc1439c34746ba44f61b057365ebb68442846.svg" alt="Kroki diagram output"></p>
<p>Real-life examples of species that share <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s include ...</p>
<ul>
<li>mouse and human, which share ~280 <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s across their chromosomes.</li>
<li>Escherichia coli and Salmonella enterica, which share ~5 <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s.</li>
</ul>
<a name="H_Genomic%20Dot%20Plot"></a><h3>Genomic Dot Plot</h3>
<p><a name="BM_(Algorithms%5C%2FSynteny%2FGenomic%20Dot%20Plot)_TOPIC%2F"></a><strong></strong></p>
<p><strong>WHAT</strong>: Given two <a href="#BM_(genome)%2Fi">genome</a>s, create a 2D plot where each axis is assigned to one of the <a href="#BM_(genome)%2Fi">genome</a>s and a dot is placed at each coordinate containing a match, where a match is either a shared <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> or a <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> and its <a href="#BM_(reverse%20complement)%2Fi">reverse complement</a>. These plots are called <a href="#BM_(genomic%20dot-plot%7Cgenomic%20dot%20plot)%2Fi">genomic dot plot</a>s.</p>
<p><img src="svgbob_3e7a2fb0e16cdd00ff9eaefd665998612d1db15c.svg" alt="Kroki diagram output"></p>
<p><strong>WHY</strong>: <a href="#BM_(genomic%20dot-plot%7Cgenomic%20dot%20plot)%2Fi">Genomic dot plot</a>s are used for identifying <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s between two <a href="#BM_(genome)%2Fi">genome</a>s.</p>
<p><strong>ALGORITHM</strong>:</p>
<p>The following algorithm finds direct matches. However, a better solution may be to consider anything with some <a href="#BM_(hamming%20distance)%2Fi">hamming distance</a> as a match. Doing so would require non-trivial changes to the algorithm (e.g. modifying the lookup to use bloom filters).</p>
<p><a href="ch6_code/src/synteny_graph/Match.py">ch6_code/src/synteny_graph/Match.py</a> (lines 176 to 232):</p>
<pre class="hljs"><code><span class="hljs-meta">@staticmethod</span>
<span class="hljs-keyword">def</span> <span class="hljs-title function_">create_from_genomes</span>(<span class="hljs-params">
        k: <span class="hljs-built_in">int</span>,
        cyclic: <span class="hljs-built_in">bool</span>,  <span class="hljs-comment"># True if chromosomes are cyclic</span>
        genome1: <span class="hljs-type">Dict</span>[<span class="hljs-built_in">str</span>, <span class="hljs-built_in">str</span>],  <span class="hljs-comment"># chromosome id -&gt; dna string</span>
        genome2: <span class="hljs-type">Dict</span>[<span class="hljs-built_in">str</span>, <span class="hljs-built_in">str</span>]   <span class="hljs-comment"># chromosome id -&gt; dna string</span>
</span>) -&gt; <span class="hljs-type">List</span>[Match]:
    <span class="hljs-comment"># lookup tables for data1</span>
    fwd_kmers1 = defaultdict(<span class="hljs-built_in">list</span>)
    rev_kmers1 = defaultdict(<span class="hljs-built_in">list</span>)
    <span class="hljs-keyword">for</span> chr_name, chr_data <span class="hljs-keyword">in</span> genome1.items():
        <span class="hljs-keyword">for</span> kmer, idx <span class="hljs-keyword">in</span> slide_window(chr_data, k, cyclic):
            fwd_kmers1[kmer].append((chr_name, idx))
            rev_kmers1[dna_reverse_complement(kmer)].append((chr_name, idx))
    <span class="hljs-comment"># lookup tables for data2</span>
    fwd_kmers2 = defaultdict(<span class="hljs-built_in">list</span>)
    rev_kmers2 = defaultdict(<span class="hljs-built_in">list</span>)
    <span class="hljs-keyword">for</span> chr_name, chr_data <span class="hljs-keyword">in</span> genome2.items():
        <span class="hljs-keyword">for</span> kmer, idx <span class="hljs-keyword">in</span> slide_window(chr_data, k, cyclic):
            fwd_kmers2[kmer].append((chr_name, idx))
            rev_kmers2[dna_reverse_complement(kmer)].append((chr_name, idx))
    <span class="hljs-comment"># match</span>
    matches = []
    fwd_key_matches = <span class="hljs-built_in">set</span>(fwd_kmers1.keys())
    fwd_key_matches.intersection_update(fwd_kmers2.keys())
    <span class="hljs-keyword">for</span> kmer <span class="hljs-keyword">in</span> fwd_key_matches:
        idxes1 = fwd_kmers1.get(kmer, [])
        idxes2 = fwd_kmers2.get(kmer, [])
        <span class="hljs-keyword">for</span> (chr_name1, idx1), (chr_name2, idx2) <span class="hljs-keyword">in</span> product(idxes1, idxes2):
            m = Match(
                y_axis_chromosome=chr_name1,
                y_axis_chromosome_min_idx=idx1,
                y_axis_chromosome_max_idx=idx1 + k - <span class="hljs-number">1</span>,
                x_axis_chromosome=chr_name2,
                x_axis_chromosome_min_idx=idx2,
                x_axis_chromosome_max_idx=idx2 + k - <span class="hljs-number">1</span>,
                <span class="hljs-built_in">type</span>=MatchType.NORMAL
            )
            matches.append(m)
    rev_key_matches = <span class="hljs-built_in">set</span>(fwd_kmers1.keys())
    rev_key_matches.intersection_update(rev_kmers2.keys())
    <span class="hljs-keyword">for</span> kmer <span class="hljs-keyword">in</span> rev_key_matches:
        idxes1 = fwd_kmers1.get(kmer, [])
        idxes2 = rev_kmers2.get(kmer, [])
        <span class="hljs-keyword">for</span> (chr_name1, idx1), (chr_name2, idx2) <span class="hljs-keyword">in</span> product(idxes1, idxes2):
            m = Match(
                y_axis_chromosome=chr_name1,
                y_axis_chromosome_min_idx=idx1,
                y_axis_chromosome_max_idx=idx1 + k - <span class="hljs-number">1</span>,
                x_axis_chromosome=chr_name2,
                x_axis_chromosome_min_idx=idx2,
                x_axis_chromosome_max_idx=idx2 + k - <span class="hljs-number">1</span>,
                <span class="hljs-built_in">type</span>=MatchType.REVERSE_COMPLEMENT
            )
            matches.append(m)
    <span class="hljs-keyword">return</span> matches
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Generating genomic dot plot for...</p>
<ul>
<li>k=5</li>
<li>cyclic=False</li>
<li>genome1={'A0': 'GGGGGAAACCCCATGCACTGG'}</li>
<li>genome2={'B0': 'GGTTTAGGTGACTTACTGGAACATGCTTGGGGG'}</li>
</ul>
<p>Result...</p>
<p><img src="genomicdotplot_c6498a2362a7e631585651f772932d3a.svg" alt="Genomic Dot Plot"></p>
<ul>
<li>{'y': ('A0', 0, 4), 'x': ('B0', 28, 32), 'type': 'NORMAL'}</li>
<li>{'y': ('A0', 11, 15), 'x': ('B0', 21, 25), 'type': 'NORMAL'}</li>
<li>{'y': ('A0', 16, 20), 'x': ('B0', 14, 18), 'type': 'NORMAL'}</li>
<li>{'y': ('A0', 8, 12), 'x': ('B0', 27, 31), 'type': 'REVERSE_COMPLEMENT'}</li>
<li>{'y': ('A0', 5, 9), 'x': ('B0', 0, 4), 'type': 'REVERSE_COMPLEMENT'}</li>
</ul>
</div>
<p></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>Rather than just showing dots at matches, the plot below draws a line over the entire match.</p>
</div>
<a name="H_Synteny%20Graph"></a><h3>Synteny Graph</h3>
<p><a name="BM_(Algorithms%5C%2FSynteny%2FSynteny%20Graph)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FSynteny%2FGenomic%20Dot%20Plot)_TOPIC%2F">Algorithms/Synteny/Genomic Dot Plot</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: Given the <a href="#BM_(genomic%20dot-plot%7Cgenomic%20dot%20plot)%2Fi">genomic dot-plot</a> for two <a href="#BM_(genome)%2Fi">genome</a>s, connect dots that are close together and going in the same direction. This process is commonly referred to as <a href="#BM_(cluster)%2Fi">cluster</a>ing. A <a href="#BM_(cluster)%2Fi">cluster</a>ed <a href="#BM_(genomic%20dot-plot%7Cgenomic%20dot%20plot)%2Fi">genomic dot plot</a> is called a <a href="#BM_(synteny%20graph)%2Fi">synteny graph</a>.</p>
<p><strong>WHY</strong>:  <a href="#BM_(cluster)%2Fi">Cluster</a>ing together matches reveals <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s.</p>
<p><img src="svgbob_1727f3faed87310ccd83cb2a23c798dcd921a972.svg" alt="Kroki diagram output"></p>
<p><img src="svgbob_4670331fbded82ef34c83903754da40e233aa22d.svg" alt="Kroki diagram output"></p>
<p><strong>ALGORITHM</strong>:</p>
<p>The following <a href="#BM_(synteny%20graph)%2Fi">synteny graph</a> algorithm relies on three non-trivial components:</p>
<ol>
<li>A spatial indexing algorithm bins points that are close together, such that it's fast to look up the set of dots that are within the proximity of some other dot. The spatial indexing algorithm used by this implementation is called a quad <a href="#BM_(tree)%2Fi">tree</a>.</li>
<li>A <a href="#BM_(cluster)%2Fi">cluster</a>ing algorithm connects dots going in the same direction to reveal <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s. The <a href="#BM_(cluster)%2Fi">cluster</a>ing algorithm used by this implementation is iterative, doing multiple rounds of connecting within a set of constraints (e.g. the <a href="#BM_(neighbour)%2Fi">neighbour</a>ing dot has to be within some limit / in some angle for it to connect).</li>
<li>A filtering algorithm that trims/merges overlapping <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s as well as removes superfluous <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s returned by the <a href="#BM_(cluster)%2Fi">cluster</a>ing algorithm. The filtering algorithm used by this implementation is simple off-the-cuff heuristics.</li>
</ol>
<p>These components are complicated and not specific to bioinformatics. As such, this section doesn't discuss them in detail but the source code is available (entrypoint is displayed below)).</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>This is code I came up with to solve the ch 6 final assignment in the Pevzner book. I came up with / fleshed out the ideas myself -- the book only hinted at specific bits. I believe the fundamentals are correct but the implementation is finicky and requires a lot of knob twisting to get decent results.</p>
</div>
<p><a href="ch6_code/src/synteny_graph/MatchMerger.py">ch6_code/src/synteny_graph/MatchMerger.py</a> (lines 18 to 65):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">distance_merge</span>(<span class="hljs-params">matches: Iterable[Match], radius: <span class="hljs-built_in">int</span>, angle_half_maw: <span class="hljs-built_in">int</span> = <span class="hljs-number">45</span></span>) -&gt; <span class="hljs-type">List</span>[Match]:
    min_x = <span class="hljs-built_in">min</span>(m.x_axis_chromosome_min_idx <span class="hljs-keyword">for</span> m <span class="hljs-keyword">in</span> matches)
    max_x = <span class="hljs-built_in">max</span>(m.x_axis_chromosome_max_idx <span class="hljs-keyword">for</span> m <span class="hljs-keyword">in</span> matches)
    min_y = <span class="hljs-built_in">min</span>(m.y_axis_chromosome_min_idx <span class="hljs-keyword">for</span> m <span class="hljs-keyword">in</span> matches)
    max_y = <span class="hljs-built_in">max</span>(m.y_axis_chromosome_max_idx <span class="hljs-keyword">for</span> m <span class="hljs-keyword">in</span> matches)
    indexer = MatchSpatialIndexer(min_x, max_x, min_y, max_y)
    <span class="hljs-keyword">for</span> m <span class="hljs-keyword">in</span> matches:
        indexer.index(m)
    ret = []
    remaining = <span class="hljs-built_in">set</span>(matches)
    <span class="hljs-keyword">while</span> remaining:
        m = <span class="hljs-built_in">next</span>(<span class="hljs-built_in">iter</span>(remaining))
        found = indexer.scan(m, radius, angle_half_maw)
        merged = Match.merge(found)
        <span class="hljs-keyword">for</span> _m <span class="hljs-keyword">in</span> found:
            indexer.unindex(_m)
            remaining.remove(_m)
        ret.append(merged)
    <span class="hljs-keyword">return</span> ret


<span class="hljs-keyword">def</span> <span class="hljs-title function_">overlap_filter</span>(<span class="hljs-params">
        matches: Iterable[Match],
        max_filter_length: <span class="hljs-built_in">float</span>,
        max_merge_distance: <span class="hljs-built_in">float</span>
</span>) -&gt; <span class="hljs-type">List</span>[Match]:
    clipper = MatchOverlapClipper(max_filter_length, max_merge_distance)
    <span class="hljs-keyword">for</span> m <span class="hljs-keyword">in</span> matches:
        <span class="hljs-keyword">while</span> <span class="hljs-literal">True</span>:
            <span class="hljs-comment"># When you attempt to add a match to the clipper, the clipper may instead ask you to make a set of changes</span>
            <span class="hljs-comment"># before it'll accept it. Specifically, the clipper may ask you to replace a bunch of existing matches that</span>
            <span class="hljs-comment"># it's already indexed and then give you a MODIFIED version of m that it'll accept once you've applied</span>
            <span class="hljs-comment"># those replacements</span>
            changes_requested = clipper.index(m)
            <span class="hljs-keyword">if</span> <span class="hljs-keyword">not</span> changes_requested:
                <span class="hljs-keyword">break</span>
            <span class="hljs-comment"># replace existing entries in clipper</span>
            <span class="hljs-keyword">for</span> from_m, to_m <span class="hljs-keyword">in</span> changes_requested.existing_matches_to_replace.items():
                clipper.unindex(from_m)
                <span class="hljs-keyword">if</span> to_m:
                    res = clipper.index(to_m)
                    <span class="hljs-keyword">assert</span> res <span class="hljs-keyword">is</span> <span class="hljs-literal">None</span>
            <span class="hljs-comment"># replace m with a revised version -- if None it means m isn't needed (its been filtered out)</span>
            m = changes_requested.revised_match
            <span class="hljs-keyword">if</span> <span class="hljs-keyword">not</span> m:
                <span class="hljs-keyword">break</span>
    <span class="hljs-keyword">return</span> <span class="hljs-built_in">list</span>(clipper.get())
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Generating synteny graph for...</p>
<ul>
<li>k=7</li>
<li>cyclic=False</li>
<li>genome1={'A0': 'GGATGGTGTCCTCATCTAATGATGTCGGTAAAGAGTCTACCCCGAATGATTATCTGAGTCTCCCATGAACCAAGTCCGTGGTATAGTCCATACTCTGAACCAAAACAGATAAACCAGCAAGATACATTGCAGAAGCTTGCCACCTTAGCAGGTTGTCAGATATCCGTTTCTGGAACTCCCGGGAGGACGATCGGAAGTTGAGCACAGGTACAAACACTTCAGGAATGATCTACTAAACTTTAGGGTCCGTACCTTTTATAATCCTTGCTAGCATCATGTTGAAGGTTAGAGGATTCCGAAACCAGAAGTGGCGATCTCGCTAAAGCAGGTCACCACGGTCAGCGGGTGGCCATTTACTCGTGAAAACCATAGTCCGTGAAAGCTGGGCAACTTTAGTTGGGACCCTTAAGGCGACTGAGGGAAGCAACTATCGGAAGTATCGTACAGGTCGTAAAGTACCAGTACGGAAGAAGCAGGGAGTTATAATATTCACTACCACAATTACCCGAGTTCACTTGTTTCAATCGCCCTCCCTTGACAGAACGTGCGTTACGTAGGAGTGCTTGACATACGGCGGCCGTCTGAGCTAGGACTATCGGAGCGTAATAATGGGATTTCAAATTTACCAGTTCCAGGTTGTCCAAGGGCTTGGCGGTGAGTCGACATGGAAAGATAAATTCCTCAGGTGCTGGCGCTCCCGTGGGGCCGCAGACACTACCTATTGGAGGGTGCTTAAACTATACAGCGCGCTAATTGTTAACTACTCCTTTGTGTCATAAGGGAGGGGAAACACGCGAGGACCGCCTTTGATCTGGTTCAAACGCCTAGAAGTATCTCCATTCTGTCCATTACGCCACCGCCCCGTCGAATGGTACCGGTATCGCTTGACATCTGCTTCTATACTAGAACAACTAATGCCGGCTTCTGGAGTGAAGGCACCATCCCACCAGAGCATTGAAGATTCGCTCGTTGGATTGATAGGAGTGAATATTCTGTCATCTCCTAACTTTTTGGGCACAGCTAG'}</li>
<li>genome2={'B0': 'CGGCATGGTGTCCTTCATGTGACCTGATGTCCGATAAGGGGGTTCTACGAAGGGCCCTCCACAGGTCCTTTGCCTAAGGATTGTTGGGTCGCATTCAACTGTTACGGAGACGTTACTAGGACGACCTAATAGAACACAACCAAGTTACGTACGCTATATCCTGTCCTGACCCAGTACCCTCTGGGTCTATATAAGTAAGCGGGTACGATTCGAGAGGGGAGCAACCAGTTACAAACACTTCAGGAATCGATCTTACTTAAACTTTTGGGTCCGATACTTTATAATCCTTGCTAGCCTACGATGTTGAGTTGAGGATTCGCGAACCAGAACAATTGCCGATACTCGCTATTAGAGGTCTCCAACGGTCACCGGGTGGGCCATTGACTCGTGAAACCAATAACCGGTGCCATTCGGACAGGGTGCTGTGGCTAGTGAAGTGAATGGCAGATTACGTCTACTGCGTTTGCAACCCAGATCCAAAGGCGTGGCTTCTACGCGTGTTTCCCATCCCTTATCACACAAGAGGGAGTAGTTAACAATTAGCGTGCTGAAGTAGTAAGCCACCCCAATGTTTAGTTCTGCGTGCCCACGGGAGCCAAGCCATCCTGAGGATTTTATGTGTCCATGTCGACAACTACACGGCAAGCACTTAGACAAGCCTGGCAACTGCGTAGAATTATGAAAGCCCACTTATTGCTCCGTATGGTCCGAGCTCAGACGGCCGCCGCTATGTCAAGCACCCCTACGGTTAACGCACGTGTTGTCCAAGGGAAGAGCGGATTGGAGACAAACGTGATCTGCGGGTTAATTGTGGTTAGTGACTATTATTAACCTCCCGTCTTCCTTCCGTCACGGGTAACCTTTACGACCTCATTACGATACTCTCCGATAGGTTGCCTTACCTTCAGTCGCCTAACGGGTCCCAATAAAGTTGCCCGCGTTCTAGGCGAATCATCGCTTGACATCTTGCTTCTTATATACAACCAACCAAATCCCGGCTTCTGGAGCTGAGGCACGCATCCCACCCAGAGGCATTGAAGATTACGCCTCGTTCGGATTGATAGTAGTGCGATATTCTGTATCTCCCTAACTATTTTCGGGCACACTACG'}</li>
</ul>
<p>Original genomic dot plot...</p>
<p><img src="syntenygraphpre_cf5ef177eb7cd4bb5a185d50761cf7e9.svg" alt="Genomic Dot Plot"></p>
<p>Merging radius=10 angle_half_maw=45...</p>
<p>Merging radius=15 angle_half_maw=45...</p>
<p>Merging radius=25 angle_half_maw=45...</p>
<p>Merging radius=35 angle_half_maw=45...</p>
<p>Filtering max_filter_length=35.0 max_merge_distance=35.0...</p>
<p>Merging radius=100 angle_half_maw=45...</p>
<p>Filtering max_filter_length=65.0 max_merge_distance=65.0...</p>
<p>Culling below length=15.0...</p>
<p>Final synteny graph...</p>
<p><img src="syntenygraphpost_cf5ef177eb7cd4bb5a185d50761cf7e9.svg" alt="Synteny Graph"></p>
<ul>
<li>{'y': ('A0', 212, 364), 'x': ('B0', 233, 392), 'type': 'NORMAL'}</li>
<li>{'y': ('A0', 879, 1018), 'x': ('B0', 953, 1104), 'type': 'NORMAL'}</li>
<li>{'y': ('A0', 384, 587), 'x': ('B0', 710, 936), 'type': 'REVERSE_COMPLEMENT'}</li>
<li>{'y': ('A0', 661, 823), 'x': ('B0', 459, 628), 'type': 'REVERSE_COMPLEMENT'}</li>
</ul>
</div>
<p></p>
<a name="H_Reversal%20Path"></a><h3>Reversal Path</h3>
<p><a name="BM_(Algorithms%5C%2FSynteny%2FReversal%20Path)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FSynteny%2FSynteny%20Graph)_TOPIC%2F">Algorithms/Synteny/Synteny Graph</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: Given two <a href="#BM_(genome)%2Fi">genome</a>s that share <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s, where one <a href="#BM_(genome)%2Fi">genome</a> has the <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s in desired form while the other does not, determine the minimum number of <a href="#BM_(genome%20rearrangement%7Cchromosomal%20rearrangement%7Cchromosome%20rearrangement)%2Fi">genome rearrangement</a> reversals (<a href="#BM_(reversal%20distance)%2Fi">reversal distance</a>) required to get the undesired <a href="#BM_(genome)%2Fi">genome</a>'s <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s to match those in the desired <a href="#BM_(genome)%2Fi">genome</a>.</p>
<p><img src="svgbob_0abfe253869a0df3178afa7a4929438ea7bbdafc.svg" alt="Kroki diagram output"></p>
<p><strong>WHY</strong>: The theory is that the <a href="#BM_(genome%20rearrangement%7Cchromosomal%20rearrangement%7Cchromosome%20rearrangement)%2Fi">genome rearrangement</a>s between two species take the <a href="#BM_(parsimony%7Cparsimonious)%2Fi">parsimonious</a> path (or close to it). Since <a href="#BM_(genome)%2Fi">genome</a> reversals are the most common form of <a href="#BM_(genome%20rearrangement%7Cchromosomal%20rearrangement%7Cchromosome%20rearrangement)%2Fi">genome rearrangement</a> mutation, by calculating a <a href="#BM_(parsimony%7Cparsimonious)%2Fi">parsimonious</a> reversal path (smallest set of reversals) it's possible to get an idea of how the two species branched off. In the example above, it may be that one of the <a href="#BM_(genome)%2Fi">genome</a>s in the reversal path is the parent that both <a href="#BM_(genome)%2Fi">genome</a>s are based off of.</p>
<p><img src="svgbob_f5e080c97a55e0d2d02be92cc43ad7aab9c45612.svg" alt="Kroki diagram output"></p>
<a name="H_Breakpoint%20List%20Algorithm"></a><h4>Breakpoint List Algorithm</h4>
<p><a name="BM_(Algorithms%5C%2FSynteny%2FReversal%20Path%2FBreakpoint%20List%20Algorithm)_TOPIC%2F"></a><strong></strong></p>
<p><strong>ALGORITHM</strong>:</p>
<p>This algorithm is a simple best effort heuristic to estimate the <a href="#BM_(parsimony%7Cparsimonious)%2Fi">parsimonious</a> reversal path. It isn't guaranteed to generate a reversal path in every case: The point of this algorithm isn't so much to be a robust solution as much as it is to be a foundation / provide intuition for better algorithms that determine reversal paths.</p>
<p>The algorithm relies on the concept of <a href="#BM_(adjacent%7Cadjacency%7Cadjacencies%7Cbreakpoint)_GR%2Fi">breakpoint</a>s and <a href="#BM_(adjacent%7Cadjacency%7Cadjacencies%7Cbreakpoint)_GR%2Fi">adjacencies</a>...</p>
<ul>
<li>
<p><a href="#BM_(adjacent%7Cadjacency%7Cadjacencies%7Cbreakpoint)_GR%2Fi">Adjacency</a>: Two <a href="#BM_(neighbour)%2Fi">neighbour</a>ing <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s in the undesired <a href="#BM_(genome)%2Fi">genome</a> that follow each other just as they do in the desired <a href="#BM_(genome)%2Fi">genome</a>. For example, ...</p>
<ul>
<li>
<p>this undesired <a href="#BM_(genome)%2Fi">genome</a> has B and C next to each other and the tail of B is followed by the head of C, just as in the desired <a href="#BM_(genome)%2Fi">genome</a>.</p>
<p><img src="svgbob_1c5addbf6716295da12222513f57fc2387b3f511.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>this undesired <a href="#BM_(genome)%2Fi">genome</a> has B and C next to each other and the tail of B is followed by the tail of C, just as in the desired <a href="#BM_(genome)%2Fi">genome</a>.</p>
<p><img src="svgbob_43f23cce2e445cfbf859bcbea9285a0f8aa498b0.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>this undesired <a href="#BM_(genome)%2Fi">genome</a> has B and C next to each other and the tail of B is followed by the head of C, just as in the desired <a href="#BM_(genome)%2Fi">genome</a>. Note that their placement has been swapped when compared to the desired <a href="#BM_(genome)%2Fi">genome</a>. As long as they follow each other as they do in the desired <a href="#BM_(genome)%2Fi">genome</a>, it's considered an <a href="#BM_(adjacent%7Cadjacency%7Cadjacencies%7Cbreakpoint)_GR%2Fi">adjacency</a>.</p>
<p><img src="svgbob_bc4f827dc42ff57e2e09e35b1d3509e9cdf87a01.svg" alt="Kroki diagram output"></p>
</li>
</ul>
</li>
<li>
<p><a href="#BM_(adjacent%7Cadjacency%7Cadjacencies%7Cbreakpoint)_GR%2Fi">Breakpoint</a>: Two <a href="#BM_(neighbour)%2Fi">neighbour</a>ing <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s in the undesired <a href="#BM_(genome)%2Fi">genome</a> don't fit the definition of an <a href="#BM_(adjacent%7Cadjacency%7Cadjacencies%7Cbreakpoint)_GR%2Fi">adjacency</a>. For example, ...</p>
<ul>
<li>
<p>this undesired <a href="#BM_(genome)%2Fi">genome</a> has B and C next to each other but the tail of B is NOT followed by the head of C, as it is in the desired <a href="#BM_(genome)%2Fi">genome</a>.</p>
<p><img src="svgbob_2a99d6be551b3ac1d08f8b2cdcf366e44cc7d668.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>this undesired <a href="#BM_(genome)%2Fi">genome</a> does NOT have B and C next to each other.</p>
<p><img src="svgbob_d95247e1c15a5d187a7fb0f5dde5bc0d132effec.svg" alt="Kroki diagram output"></p>
</li>
</ul>
</li>
</ul>
<p><a href="#BM_(adjacent%7Cadjacency%7Cadjacencies%7Cbreakpoint)_GR%2Fi">Breakpoint</a>s and <a href="#BM_(adjacent%7Cadjacency%7Cadjacencies%7Cbreakpoint)_GR%2Fi">adjacencies</a> are useful because they identify desirable points for reversals. This algorithm takes advantage of that fact to estimate the <a href="#BM_(reversal%20distance)%2Fi">reversal distance</a>. For example, a contiguous train of <a href="#BM_(adjacent%7Cadjacency%7Cadjacencies%7Cbreakpoint)_GR%2Fi">adjacencies</a> in an undesired <a href="#BM_(genome)%2Fi">genome</a> may identify the boundaries for a single reversal that gets the undesired <a href="#BM_(genome)%2Fi">genome</a> closer to the desired <a href="#BM_(genome)%2Fi">genome</a>.</p>
<p><img src="svgbob_2d57bfb808013a8f02bbf4b63633b470bd14ef9d.svg" alt="Kroki diagram output"></p>
<p>The algorithm starts by assigning integers to <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s. The <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s in the...</p>
<ul>
<li>desired <a href="#BM_(genome)%2Fi">genome</a> are represented as +1 to +n.</li>
<li>undesired <a href="#BM_(genome)%2Fi">genome</a> are represented by the integers of the corresponding <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a> in the desired <a href="#BM_(genome)%2Fi">genome</a>, where the integer is negated if the <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a> is reversed.</li>
</ul>
<p>For example, ...</p>
<p><img src="svgbob_b2af46c8723d0c24b7cc07e0a98008be700c7b60.svg" alt="Kroki diagram output"></p>
<p>The <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s in each <a href="#BM_(genome)%2Fi">genome</a>s of the above example may be represented as lists...</p>
<ul>
<li><code>[+1, +2, +3, +4, +5]</code> (DESIRED)</li>
<li><code>[+1, -4, -3, -2, -5]</code> (UNDESIRED)</li>
</ul>
<p>Artificially add a 0 prefix and a length + 1 suffix to both lists. In the above example, the length is 5, so each list gets a prefix of 0 and a suffix of 6...</p>
<ul>
<li><code>[0, +1, +2, +3, +4, +5, +6]</code> (DESIRED)</li>
<li><code>[0, +1, -4, -3, -2, -5, +6]</code> (UNDESIRED)</li>
</ul>
<p>In this modified list, consecutive elements <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo stretchy="false">(</mo><msub><mi>p</mi><mi>i</mi></msub><mo separator="true">,</mo><msub><mi>p</mi><mrow><mi>i</mi><mo>+</mo><mn>1</mn></mrow></msub><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">(p_i, p_{i+1})</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal">p</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mpunct">,</span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord"><span class="mord mathnormal">p</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.311664em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">i</span><span class="mbin mtight">+</span><span class="mord mtight">1</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.208331em;"><span></span></span></span></span></span></span><span class="mclose">)</span></span></span></span> are considered a...</p>
<ul>
<li><a href="#BM_(adjacent%7Cadjacency%7Cadjacencies%7Cbreakpoint)_GR%2Fi">adjacency</a> if <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>p</mi><mi>i</mi></msub><mo>+</mo><mn>1</mn><mo>=</mo><msub><mi>p</mi><mrow><mi>i</mi><mo>+</mo><mn>1</mn></mrow></msub></mrow><annotation encoding="application/x-tex">p_i + 1 = p_{i+1}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.7777700000000001em;vertical-align:-0.19444em;"></span><span class="mord"><span class="mord mathnormal">p</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.64444em;vertical-align:0em;"></span><span class="mord">1</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:0.638891em;vertical-align:-0.208331em;"></span><span class="mord"><span class="mord mathnormal">p</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.311664em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">i</span><span class="mbin mtight">+</span><span class="mord mtight">1</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.208331em;"><span></span></span></span></span></span></span></span></span></span>.</li>
<li><a href="#BM_(adjacent%7Cadjacency%7Cadjacencies%7Cbreakpoint)_GR%2Fi">breakpoint</a> if <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>p</mi><mi>i</mi></msub><mo>+</mo><mn>1</mn><mo mathvariant="normal">≠</mo><msub><mi>p</mi><mrow><mi>i</mi><mo>+</mo><mn>1</mn></mrow></msub></mrow><annotation encoding="application/x-tex">p_i + 1 \neq p_{i+1}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.7777700000000001em;vertical-align:-0.19444em;"></span><span class="mord"><span class="mord mathnormal">p</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.8888799999999999em;vertical-align:-0.19444em;"></span><span class="mord">1</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel"><span class="mrel"><span class="mord vbox"><span class="thinbox"><span class="rlap"><span class="strut" style="height:0.8888799999999999em;vertical-align:-0.19444em;"></span><span class="inner"><span class="mrel"></span></span><span class="fix"></span></span></span></span></span><span class="mrel">=</span></span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:0.638891em;vertical-align:-0.208331em;"></span><span class="mord"><span class="mord mathnormal">p</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.311664em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">i</span><span class="mbin mtight">+</span><span class="mord mtight">1</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.208331em;"><span></span></span></span></span></span></span></span></span></span>.</li>
</ul>
<p>In the undesired version of the example above, the <a href="#BM_(adjacent%7Cadjacency%7Cadjacencies%7Cbreakpoint)_GR%2Fi">breakpoint</a>s and <a href="#BM_(adjacent%7Cadjacency%7Cadjacencies%7Cbreakpoint)_GR%2Fi">adjacencies</a> are...</p>
<p><img src="svgbob_f6cc3ba996f4c83a71898d444054303e8a0891a6.svg" alt="Kroki diagram output"></p>
<p>This algorithm continually applies <a href="#BM_(genome%20rearrangement%7Cchromosomal%20rearrangement%7Cchromosome%20rearrangement)%2Fi">genome rearrangement</a> reversal operations on portions of the list in the hopes of reducing the number of <a href="#BM_(adjacent%7Cadjacency%7Cadjacencies%7Cbreakpoint)_GR%2Fi">breakpoint</a>s at each reversal, ultimately hoping to get it to the desired list. It targets portions of contiguous <a href="#BM_(adjacent%7Cadjacency%7Cadjacencies%7Cbreakpoint)_GR%2Fi">adjacencies</a> sandwiched between <a href="#BM_(adjacent%7Cadjacency%7Cadjacencies%7Cbreakpoint)_GR%2Fi">breakpoint</a>s. In the example above, the reversal of <code>[-4, -3, -2]</code> reduces the number of <a href="#BM_(adjacent%7Cadjacency%7Cadjacencies%7Cbreakpoint)_GR%2Fi">breakpoint</a>s by 1...</p>
<p><img src="svgbob_4dd4daeb521f7dba32a19b4112269bcab50e6385.svg" alt="Kroki diagram output"></p>
<p>Following that up with a reversal of <code>[-5]</code> reduces the number of <a href="#BM_(adjacent%7Cadjacency%7Cadjacencies%7Cbreakpoint)_GR%2Fi">breakpoint</a>s by 2...</p>
<p><img src="svgbob_76a2c560a0ba7f50f490ce3dc41d554cdbde0e3c.svg" alt="Kroki diagram output"></p>
<p>Leaving the undesired list in the same state as the desired list. As such, the <a href="#BM_(reversal%20distance)%2Fi">reversal distance</a> for this example is 2 reversals.</p>
<p>In the best case, a single reversal will remove 2 <a href="#BM_(adjacent%7Cadjacency%7Cadjacencies%7Cbreakpoint)_GR%2Fi">breakpoint</a>s (one on each side of the reversal). In the worst case, there is no single reversal that drives down the number of <a href="#BM_(adjacent%7Cadjacency%7Cadjacencies%7Cbreakpoint)_GR%2Fi">breakpoint</a>s. For example, there is no single reversal for the list <code>[+2, +1]</code> that reduces the number of <a href="#BM_(adjacent%7Cadjacency%7Cadjacencies%7Cbreakpoint)_GR%2Fi">breakpoint</a>s...</p>
<p><img src="svgbob_deb655097c2a34c25e662fcb9b31bbf5ae45f169.svg" alt="Kroki diagram output"></p>
<p>In such worst case scenarios, the algorithm fails. However, the point of this algorithm isn't so much to be a robust solution as much as it is to be a foundation for better algorithms that determine reversal paths.</p>
<p><a href="ch6_code/src/breakpoint_list/BreakpointList.py">ch6_code/src/breakpoint_list/BreakpointList.py</a> (lines 7 to 26):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">find_adjacencies_sandwiched_between_breakpoints</span>(<span class="hljs-params">augmented_blocks: <span class="hljs-type">List</span>[<span class="hljs-built_in">int</span>]</span>) -&gt; <span class="hljs-type">List</span>[<span class="hljs-built_in">int</span>]:
    <span class="hljs-keyword">assert</span> augmented_blocks[<span class="hljs-number">0</span>] == <span class="hljs-number">0</span>
    <span class="hljs-keyword">assert</span> augmented_blocks[-<span class="hljs-number">1</span>] == <span class="hljs-built_in">len</span>(augmented_blocks) - <span class="hljs-number">1</span>
    ret = []
    <span class="hljs-keyword">for</span> (x1, x2), idx <span class="hljs-keyword">in</span> slide_window(augmented_blocks, <span class="hljs-number">2</span>):
        <span class="hljs-keyword">if</span> x1 + <span class="hljs-number">1</span> != x2:
            ret.append(idx)
    <span class="hljs-keyword">return</span> ret


<span class="hljs-keyword">def</span> <span class="hljs-title function_">find_and_reverse_section</span>(<span class="hljs-params">augmented_blocks: <span class="hljs-type">List</span>[<span class="hljs-built_in">int</span>]</span>) -&gt; <span class="hljs-type">Optional</span>[<span class="hljs-type">List</span>[<span class="hljs-built_in">int</span>]]:
    bp_idxes = find_adjacencies_sandwiched_between_breakpoints(augmented_blocks)
    <span class="hljs-keyword">for</span> (bp_i1, bp_i2), _ <span class="hljs-keyword">in</span> slide_window(bp_idxes, <span class="hljs-number">2</span>):
        <span class="hljs-keyword">if</span> augmented_blocks[bp_i1] + <span class="hljs-number">1</span> == -augmented_blocks[bp_i2] <span class="hljs-keyword">or</span>\
                augmented_blocks[bp_i2 + <span class="hljs-number">1</span>] == -augmented_blocks[bp_i1 + <span class="hljs-number">1</span>] + <span class="hljs-number">1</span>:
            <span class="hljs-keyword">return</span> augmented_blocks[:bp_i1 + <span class="hljs-number">1</span>]\
                   + [-x <span class="hljs-keyword">for</span> x <span class="hljs-keyword">in</span> <span class="hljs-built_in">reversed</span>(augmented_blocks[bp_i1 + <span class="hljs-number">1</span>:bp_i2 + <span class="hljs-number">1</span>])]\
                   + augmented_blocks[bp_i2 + <span class="hljs-number">1</span>:]
    <span class="hljs-keyword">return</span> <span class="hljs-literal">None</span>
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Reversing on breakpoint boundaries...</p>
<ul>
<li><code>[0, +1, -4, -3, -2, -5, +6]</code></li>
<li><code>[0, +1, +2, +3, +4, -5, +6]</code></li>
<li><code>[0, +1, +2, +3, +4, +5, +6]</code></li>
</ul>
<p>No more reversals possible.</p>
</div>
<p></p>
<p>Since each reversal can at most reduce the number of <a href="#BM_(adjacent%7Cadjacency%7Cadjacencies%7Cbreakpoint)_GR%2Fi">breakpoint</a>s by 2, the <a href="#BM_(reversal%20distance)%2Fi">reversal distance</a> must be at least half the number of <a href="#BM_(adjacent%7Cadjacency%7Cadjacencies%7Cbreakpoint)_GR%2Fi">breakpoint</a>s (lower bound): <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>d</mi><mrow><mi>r</mi><mi>e</mi><mi>v</mi></mrow></msub><mo stretchy="false">(</mo><mi>p</mi><mo stretchy="false">)</mo><mo>&gt;</mo><mo>=</mo><mfrac><mrow><mi>b</mi><mi>p</mi><mo stretchy="false">(</mo><mi>p</mi><mo stretchy="false">)</mo></mrow><mn>2</mn></mfrac></mrow><annotation encoding="application/x-tex">d_{rev}(p) &gt;= \frac{bp(p)}{2}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord"><span class="mord mathnormal">d</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.151392em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.02778em;">r</span><span class="mord mathnormal mtight">e</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mopen">(</span><span class="mord mathnormal">p</span><span class="mclose">)</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">&gt;</span></span><span class="base"><span class="strut" style="height:0.36687em;vertical-align:0em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:1.355em;vertical-align:-0.345em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.01em;"><span style="top:-2.6550000000000002em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">2</span></span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.485em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">b</span><span class="mord mathnormal mtight">p</span><span class="mopen mtight">(</span><span class="mord mathnormal mtight">p</span><span class="mclose mtight">)</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.345em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span>. In other words, the minimum number of reversals to transform a <a href="#BM_(permutation)_GR%2Fi">permutation</a>s to an identity <a href="#BM_(permutation)_GR%2Fi">permutation</a> will never be less than <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mfrac><mrow><mi>b</mi><mi>p</mi><mo stretchy="false">(</mo><mi>p</mi><mo stretchy="false">)</mo></mrow><mn>2</mn></mfrac></mrow><annotation encoding="application/x-tex">\frac{bp(p)}{2}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.355em;vertical-align:-0.345em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.01em;"><span style="top:-2.6550000000000002em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">2</span></span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.485em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">b</span><span class="mord mathnormal mtight">p</span><span class="mopen mtight">(</span><span class="mord mathnormal mtight">p</span><span class="mclose mtight">)</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.345em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span>.</p>
<a name="H_Breakpoint%20Graph%20Algorithm"></a><h4>Breakpoint Graph Algorithm</h4>
<p><a name="BM_(Algorithms%5C%2FSynteny%2FReversal%20Path%2FBreakpoint%20Graph%20Algorithm)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FSynteny%2FReversal%20Path%2FBreakpoint%20List%20Algorithm)_TOPIC%2F">Algorithms/Synteny/Reversal Path/Breakpoint List Algorithm</a></li>
</ul>
</div>
<p><strong>ALGORITHM</strong>:</p>
<p>This algorithm calculates a <a href="#BM_(parsimony%7Cparsimonious)%2Fi">parsimonious</a> reversal path by constructing an undirected graph representing the <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s between <a href="#BM_(genome)%2Fi">genome</a>s. Unlike the <a href="#BM_(adjacent%7Cadjacency%7Cadjacencies%7Cbreakpoint)_GR%2Fi">breakpoint</a> list algorithm, this algorithm...</p>
<ul>
<li>determines a full <a href="#BM_(parsimony%7Cparsimonious)%2Fi">parsimonious</a> reversal path every time (there may be multiple such paths).</li>
<li>handles multiple chromosomes (<a href="#BM_(genome%20rearrangement%7Cchromosomal%20rearrangement%7Cchromosome%20rearrangement)%2Fi">genome rearrangement</a> chromosomal <a href="#BM_(%5Cbfusion%7C%5Cbfuse)%2Fi">fusion</a>s and <a href="#BM_(fission)%2Fi">fission</a>s).</li>
<li>handles both linear and circular chromosomes.</li>
</ul>
<p>This algorithm begins by constructing an undirected graphs containing both the desired and undesired <a href="#BM_(genome)%2Fi">genome</a>s, referred to as a <a href="#BM_(breakpoint%20graph)_GR%2Fi">breakpoint graph</a>. It then performs a set of re-wiring operations on the <a href="#BM_(breakpoint%20graph)_GR%2Fi">breakpoint graph</a> to determine a <a href="#BM_(parsimony%7Cparsimonious)%2Fi">parsimonious</a> reversal path (including <a href="#BM_(%5Cbfusion%7C%5Cbfuse)%2Fi">fusion</a> and <a href="#BM_(fission)%2Fi">fission</a>), where each re-wiring operation is referred to as a <a href="#BM_%5Cb(2-break%7C2%20break%7Ctwo%20break%7Ctwo-break)s%3F%5Cb%2Fi">two-break</a>.</p>
<p><strong><a href="#BM_(breakpoint%20graph)_GR%2Fi">BREAKPOINT GRAPH</a> REPRESENTATION</strong></p>
<p>Construction of a <a href="#BM_(breakpoint%20graph)_GR%2Fi">breakpoint graph</a> is as follows:</p>
<ol>
<li>
<p>Set the ends of <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s as nodes. The arrow end should have a <em>t</em> suffix (for tail) while the non-arrow end should have a <em>h</em> suffix (for head)...</p>
<p><img src="dot_0a2b976df1fe819b59828770f2a8a50d.svg" alt="Dot diagram"></p>
<p>If the <a href="#BM_(genome)%2Fi">genome</a> has linear chromosomes, add a termination node as well to represent chromosome ends. Only one termination node is needed -- all chromosome ends are represented by the same termination node.</p>
<p><img src="dot_9e751bbe7943da6f671b2b2f75c6d011.svg" alt="Dot diagram"></p>
</li>
<li>
<p>Set the <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s themselves as <em>undirected</em> edges, represented by dashed edges.</p>
<p><img src="dot_b5f4f9926e729d81240bfd9413d3bcb0.svg" alt="Dot diagram"></p>
<p>Note that the arrow heads on these dashed edges represent the direction of the <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny</a> match (e.g. head-to-tail for a normal match vs tail-to-head for a <a href="#BM_(reverse%20complement)%2Fi">reverse complement</a> match), not edge directions in the graph (graph is undirected). Since the <em>h</em> and <em>t</em> suffixes on nodes already convey the match direction information, the arrows may be omitted to reduce confusion.</p>
<p><img src="dot_19da0e0853121f133813afea783f11f6.svg" alt="Dot diagram"></p>
</li>
<li>
<p>Set the regions between <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s as <em>undirected</em> edges, represented by colored edges. Regions of ...</p>
<ul>
<li>desired <a href="#BM_(genome)%2Fi">genome</a> that border a pair of <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s are represented as blue edges.</li>
<li>undesired <a href="#BM_(genome)%2Fi">genome</a> that border a pair of <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s are represented as red edges.</li>
</ul>
<p><img src="dot_4740a5d5e6a50b683018382598a77f39.svg" alt="Dot diagram"></p>
<p>For linear chromosomes, the region between a chromosome end and the <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny</a> node just before it is also represented by the appropriate colored edge.</p>
<p><img src="dot_59403de9d680cd87815da4435b70bb70.svg" alt="Dot diagram"></p>
</li>
</ol>
<p>For example, the following two <a href="#BM_(genome)%2Fi">genome</a>s share the <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s A, B, C, and D between them ...</p>
<p><img src="svgbob_7d1d037f9aff25774105ac2ef33cc4cea5f9af58.svg" alt="Kroki diagram output"></p>
<p>Converting the above <a href="#BM_(genome)%2Fi">genome</a>s to both a circular and linear <a href="#BM_(breakpoint%20graph)_GR%2Fi">breakpoint graph</a> is as follows...</p>
<p><img src="dot_265a2cff663a78c6cb113f54e2ded77e.svg" alt="Dot diagram"></p>
<p>As shown in the example above, the convention for drawing a <a href="#BM_(breakpoint%20graph)_GR%2Fi">breakpoint graph</a> is to position nodes and edges as they appear in the desired <a href="#BM_(genome)%2Fi">genome</a> (<a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny</a> edges should be neatly sandwiched between blue edges). Note how both <a href="#BM_(breakpoint%20graph)_GR%2Fi">breakpoint graph</a>s in the example above are just another representation of their linear diagram counterparts. The ...</p>
<ul>
<li>dashed edges represent the <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s (suffix of <em>t</em> denotes arrow end / <em>h</em> denotes opposite end).</li>
<li>blue edges represent the gaps between <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a> and chromosome end nodes in the desired <a href="#BM_(genome)%2Fi">genome</a>.</li>
<li>red edges represent the gaps between <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a> and chromosome end nodes in the undesired <a href="#BM_(genome)%2Fi">genome</a>.</li>
</ul>
<p>The reason for this convention is that it helps conceptualize the algorithms that operate on <a href="#BM_(breakpoint%20graph)_GR%2Fi">breakpoint graph</a>s (described further down). Ultimately, a <a href="#BM_(breakpoint%20graph)_GR%2Fi">breakpoint graph</a> is simply a merged version of the linear diagrams for both the desired and undesired <a href="#BM_(genome)%2Fi">genome</a>s.</p>
<p>For example, if the circular <a href="#BM_(genome)%2Fi">genome</a> version of the <a href="#BM_(breakpoint%20graph)_GR%2Fi">breakpoint graph</a> example above were flattened based on the blue edges (desired <a href="#BM_(genome)%2Fi">genome</a>), the <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s would be ordered as they are in the linear diagram for the desired <a href="#BM_(genome)%2Fi">genome</a>...</p>
<p><img src="svgbob_2db636d6a9dca8ec58a87760d2bc130a878996e9.svg" alt="Kroki diagram output"></p>
<p><img src="dot_f8f42a4724463ac1f8ff20a8ab0dc7c9.svg" alt="Dot diagram"></p>
<p>Likewise, if the circular <a href="#BM_(genome)%2Fi">genome</a> version of the <a href="#BM_(breakpoint%20graph)_GR%2Fi">breakpoint graph</a> example above were flattened based on red edges (undesired <a href="#BM_(genome)%2Fi">genome</a>), the <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s would be ordered as they are in the linear diagram for the undesired <a href="#BM_(genome)%2Fi">genome</a>...</p>
<p><img src="svgbob_dcf88bd11a7c7ce3d07c62be91cbadd6b1b47aad.svg" alt="Kroki diagram output"></p>
<p><img src="dot_8e8ffca4fb4c3dabe9ce45d4a7bcada6.svg" alt="Dot diagram"></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>If you're confused at this point, don't continue. Go back and make sure you understand because in the next section builds on the above content.</p>
</div>
<p><strong>DATA STRUCTURE REPRESENTATION</strong></p>
<p>The data structure used to represent a <a href="#BM_(breakpoint%20graph)_GR%2Fi">breakpoint graph</a> can simply be two <a href="#BM_(adjacency%20list)%2Fi">adjacency list</a>s: one for the red edges and one for the blue edges.</p>
<p><a href="ch6_code/src/breakpoint_graph/ColoredEdgeSet.py">ch6_code/src/breakpoint_graph/ColoredEdgeSet.py</a> (lines 16 to 35):</p>
<pre class="hljs"><code><span class="hljs-comment"># Represents a single genome in a breakpoint graph</span>
<span class="hljs-keyword">class</span> <span class="hljs-title class_">ColoredEdgeSet</span>:
    <span class="hljs-keyword">def</span> <span class="hljs-title function_">__init__</span>(<span class="hljs-params">self</span>):
        self.by_node: <span class="hljs-type">Dict</span>[SyntenyNode, ColoredEdge] = {}

<span class="hljs-meta">    @staticmethod</span>
    <span class="hljs-keyword">def</span> <span class="hljs-title function_">create</span>(<span class="hljs-params">ce_list: Iterable[ColoredEdge]</span>) -&gt; ColoredEdgeSet:
        ret = ColoredEdgeSet()
        <span class="hljs-keyword">for</span> ce <span class="hljs-keyword">in</span> ce_list:
            ret.insert(ce)
        <span class="hljs-keyword">return</span> ret

    <span class="hljs-keyword">def</span> <span class="hljs-title function_">insert</span>(<span class="hljs-params">self, e: ColoredEdge</span>):
        <span class="hljs-keyword">if</span> e.n1 <span class="hljs-keyword">in</span> self.by_node <span class="hljs-keyword">or</span> e.n2 <span class="hljs-keyword">in</span> self.by_node:
            <span class="hljs-keyword">raise</span> ValueError(<span class="hljs-string">f'Node already occupied: <span class="hljs-subst">{e}</span>'</span>)
        <span class="hljs-keyword">if</span> <span class="hljs-keyword">not</span> <span class="hljs-built_in">isinstance</span>(e.n1, TerminalNode):
            self.by_node[e.n1] = e
        <span class="hljs-keyword">if</span> <span class="hljs-keyword">not</span> <span class="hljs-built_in">isinstance</span>(e.n2, TerminalNode):
            self.by_node[e.n2] = e
</code></pre>
<p>The edges representing <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s technically don't need to be tracked because they're easily derived from either set of colored edges (red or blue). For example, given the following circular <a href="#BM_(breakpoint%20graph)_GR%2Fi">breakpoint graph</a> ...</p>
<p><img src="dot_bed1d61715ed515600f51c8a4b395968.svg" alt="Dot diagram"></p>
<p>..., walk the blue edges starting from the node B_t. The opposite end of the blue edge at B_t is C_h. The next edge to walk must be a <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny</a> edge, but <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny</a> edges aren't tracked in this data structure. However, since it's known that the nodes of a <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny</a> edge...</p>
<ul>
<li>must either end in <em>t</em> or <em>h</em></li>
<li>share the same name</li>
</ul>
<p>, ... it's easy to derive that the opposite end of the <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny</a> edge at node C_h is node C_t. As such, get the blue edge for C_t and repeat. Keep repeating until a cycle is detected.</p>
<p>For linear <a href="#BM_(breakpoint%20graph)_GR%2Fi">breakpoint graph</a>s, the process must start and end at the termination node (no cycle).</p>
<p><a href="ch6_code/src/breakpoint_graph/ColoredEdgeSet.py">ch6_code/src/breakpoint_graph/ColoredEdgeSet.py</a> (lines 80 to 126):</p>
<pre class="hljs"><code><span class="hljs-comment"># Walks the colored edges, spliced with synteny edges.</span>
<span class="hljs-keyword">def</span> <span class="hljs-title function_">walk</span>(<span class="hljs-params">self</span>) -&gt; <span class="hljs-type">List</span>[<span class="hljs-type">List</span>[<span class="hljs-type">Union</span>[ColoredEdge, SyntenyEdge]]]:
    ret = []
    all_edges = self.edges()
    term_edges = <span class="hljs-built_in">set</span>()
    <span class="hljs-keyword">for</span> ce <span class="hljs-keyword">in</span> all_edges:
        <span class="hljs-keyword">if</span> ce.has_term():
            term_edges.add(ce)
    <span class="hljs-comment"># handle linear chromosomes</span>
    <span class="hljs-keyword">while</span> term_edges:
        ce = term_edges.pop()
        n = ce.non_term()
        all_edges.remove(ce)
        edges = []
        <span class="hljs-keyword">while</span> <span class="hljs-literal">True</span>:
            se_n1 = n
            se_n2 = se_n1.swap_end()
            se = SyntenyEdge(se_n1, se_n2)
            edges += [ce, se]
            ce = self.by_node[se_n2]
            <span class="hljs-keyword">if</span> ce.has_term():
                edges += [ce]
                term_edges.remove(ce)
                all_edges.remove(ce)
                <span class="hljs-keyword">break</span>
            n = ce.other_end(se_n2)
            all_edges.remove(ce)
        ret.append(edges)
    <span class="hljs-comment"># handle cyclic chromosomes</span>
    <span class="hljs-keyword">while</span> all_edges:
        start_ce = all_edges.pop()
        ce = start_ce
        n = ce.n1
        edges = []
        <span class="hljs-keyword">while</span> <span class="hljs-literal">True</span>:
            se_n1 = n
            se_n2 = se_n1.swap_end()
            se = SyntenyEdge(se_n1, se_n2)
            edges += [ce, se]
            ce = self.by_node[se_n2]
            <span class="hljs-keyword">if</span> ce == start_ce:
                <span class="hljs-keyword">break</span>
            n = ce.other_end(se_n2)
            all_edges.remove(ce)
        ret.append(edges)
    <span class="hljs-keyword">return</span> ret
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the colored edges...</p>
<ul>
<li>CE(C_h, D_t)</li>
<li>CE(B_h, C_t)</li>
<li>CE(A_h, B_t)</li>
<li>CE(A_t, D_h)</li>
<li>CE(X_h, Z_t)</li>
<li>CE(X_t, Z_h)</li>
</ul>
<p>Synteny edges spliced in...</p>
<ul>
<li>START
<ul>
<li>CE(X_h, Z_t)</li>
<li>SE(X_h, X_t)</li>
<li>CE(X_t, Z_h)</li>
<li>SE(Z_h, Z_t)</li>
</ul>
</li>
<li>START
<ul>
<li>CE(C_h, D_t)</li>
<li>SE(C_h, C_t)</li>
<li>CE(B_h, C_t)</li>
<li>SE(B_h, B_t)</li>
<li>CE(A_h, B_t)</li>
<li>SE(A_h, A_t)</li>
<li>CE(A_t, D_h)</li>
<li>SE(D_h, D_t)</li>
</ul>
</li>
</ul>
<p>CE means colored edge / SE means synteny edge.</p>
</div>
<p></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>If you're confused at this point, don't continue. Go back and make sure you understand because in the next section builds on the above content.</p>
</div>
<p><strong><a href="#BM_(permutation)_GR%2Fi">PERMUTATION</a> REPRESENTATION</strong></p>
<p>A common textual representation of a <a href="#BM_(breakpoint%20graph)_GR%2Fi">breakpoint graph</a> is writing out each of the two <a href="#BM_(genome)%2Fi">genome</a>s as a set of lists. Each list, referred to as a <a href="#BM_(permutation)_GR%2Fi">permutation</a>, describes one of the chromosomes in a <a href="#BM_(genome)%2Fi">genome</a>.</p>
<p>To convert a chromosome within a <a href="#BM_(breakpoint%20graph)_GR%2Fi">breakpoint graph</a> to a <a href="#BM_(permutation)_GR%2Fi">permutation</a>, simply walk the edges for that chromosome...</p>
<ul>
<li>desired <a href="#BM_(genome)%2Fi">genome</a>: blue edges and <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny</a> edges.</li>
<li>undesired <a href="#BM_(genome)%2Fi">genome</a>: red edges and <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny</a> edges.</li>
</ul>
<p>Each <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny</a> edge walked is appended to the list with a prefix of ...</p>
<ul>
<li><em>+</em> if it's walked from head to tail.</li>
<li><em>-</em> if it's walked from tail to head.</li>
</ul>
<p>For example, given the following <a href="#BM_(breakpoint%20graph)_GR%2Fi">breakpoint graph</a> ...</p>
<p><img src="dot_30145ecbe97993fdba2f7bb06a36749d.svg" alt="Dot diagram"></p>
<p>, ... walking the edges for the undesired <a href="#BM_(genome)%2Fi">genome</a> (red) from node D_t in the ...</p>
<ul>
<li>clockwise direction results in the <a href="#BM_(permutation)_GR%2Fi">permutation</a> <code>[-D, -C]</code>.</li>
<li>counter-clockwise direction results in the <a href="#BM_(permutation)_GR%2Fi">permutation</a> <code>[+C, +D]</code>.</li>
</ul>
<p>For circular chromosomes, the walk direction is irrelevant, meaning that both example <a href="#BM_(permutation)_GR%2Fi">permutation</a>s above represent the same chromosome. Likewise, the starting node is also irrelevant, meaning that the following <a href="#BM_(permutation)_GR%2Fi">permutation</a>s are all equivalent to the ones in the above example: <code>[+C, +D]</code>, and <code>[+D, +C]</code>.</p>
<p>For linear chromosomes, the walk direction is irrelevant but the walk must start from and end at the termination node (representing the ends of the chromosome). The termination nodes aren't included in the <a href="#BM_(permutation)_GR%2Fi">permutation</a>.</p>
<p>In the example <a href="#BM_(breakpoint%20graph)_GR%2Fi">breakpoint graph</a> above, the <a href="#BM_(permutation)_GR%2Fi">permutation</a> set representing the undesired <a href="#BM_(genome)%2Fi">genome</a> (red) may be written as either...</p>
<ul>
<li><code>{[+C, +D], [+A, +B]}</code></li>
<li><code>{[+A, +B], [-C, _D]}</code></li>
<li><code>{[-A, -B], [-C, -D]}</code></li>
<li>...</li>
</ul>
<p>Likewise, the <a href="#BM_(permutation)_GR%2Fi">permutation</a> set representing the desired <a href="#BM_(genome)%2Fi">genome</a> (blue) in the example above may be written as either...</p>
<ul>
<li><code>{[+A, +B, +C, +D]}</code></li>
<li><code>{[-D, -C, -B, -A]}</code></li>
<li><code>{[+B, +C, +D, +A]}</code></li>
<li>...</li>
</ul>
<p><a href="ch6_code/src/breakpoint_graph/Permutation.py">ch6_code/src/breakpoint_graph/Permutation.py</a> (lines 158 to 196):</p>
<pre class="hljs"><code><span class="hljs-meta">@staticmethod</span>
<span class="hljs-keyword">def</span> <span class="hljs-title function_">from_colored_edges</span>(<span class="hljs-params">
        colored_edges: ColoredEdgeSet,
        start_n: SyntenyNode,
        cyclic: <span class="hljs-built_in">bool</span>
</span>) -&gt; <span class="hljs-type">Tuple</span>[Permutation, <span class="hljs-type">Set</span>[ColoredEdge]]:
    <span class="hljs-comment"># if not cyclic, it's expected that start_n is either from or to a term node</span>
    <span class="hljs-keyword">if</span> <span class="hljs-keyword">not</span> cyclic:
        ce = colored_edges.find(start_n)
        <span class="hljs-keyword">assert</span> ce.has_term(), <span class="hljs-string">"Start node must be for a terminal colored edge"</span>
    <span class="hljs-comment"># if cyclic stop once you detect a loop, otherwise  stop once you encounter a term node</span>
    <span class="hljs-keyword">if</span> cyclic:
        walked = <span class="hljs-built_in">set</span>()
        <span class="hljs-keyword">def</span> <span class="hljs-title function_">stop_test</span>(<span class="hljs-params">x</span>):
            ret = x <span class="hljs-keyword">in</span> walked
            walked.add(next_n)
            <span class="hljs-keyword">return</span> ret
    <span class="hljs-keyword">else</span>:
        <span class="hljs-keyword">def</span> <span class="hljs-title function_">stop_test</span>(<span class="hljs-params">x</span>):
            <span class="hljs-keyword">return</span> x == TerminalNode.INST
    <span class="hljs-comment"># begin loop</span>
    blocks = []
    start_ce = colored_edges.find(start_n)
    walked_ce_set = {start_ce}
    next_n = start_n
    <span class="hljs-keyword">while</span> <span class="hljs-keyword">not</span> stop_test(next_n):
        <span class="hljs-keyword">if</span> next_n.end == SyntenyEnd.HEAD:
            b = Block(Direction.FORWARD, next_n.<span class="hljs-built_in">id</span>)
        <span class="hljs-keyword">elif</span> next_n.end == SyntenyEnd.TAIL:
            b = Block(Direction.BACKWARD, next_n.<span class="hljs-built_in">id</span>)
        <span class="hljs-keyword">else</span>:
            <span class="hljs-keyword">raise</span> ValueError(<span class="hljs-string">'???'</span>)
        blocks.append(b)
        swapped_n = next_n.swap_end()
        next_ce = colored_edges.find(swapped_n)
        next_n = next_ce.other_end(swapped_n)
        walked_ce_set.add(next_ce)
    <span class="hljs-keyword">return</span> Permutation(blocks, cyclic), walked_ce_set
</code></pre>
<p>Converting from a <a href="#BM_(permutation)_GR%2Fi">permutation</a> set back to a <a href="#BM_(breakpoint%20graph)_GR%2Fi">breakpoint graph</a> is basically just reversing the above process. For each <a href="#BM_(permutation)_GR%2Fi">permutation</a>, slide a window of size two to determine the colored edges that <a href="#BM_(permutation)_GR%2Fi">permutation</a> is for. The node chosen for the window element at index ...</p>
<ol start="0">
<li>should be tail if sign is <em>-</em> or head if sign is <em>+</em>.</li>
<li>should be head if sign is <em>-</em> or tail if sign is <em>+</em>.</li>
</ol>
<p>For circular chromosomes, the sliding window is cyclic. For example, sliding the window over <a href="#BM_(permutation)_GR%2Fi">permutation</a> <code>[+A, +C, -B, +D]</code> results in ...</p>
<ul>
<li><code>[+A, +C]</code> which produces the colored edge <code>(A_h, C_t)</code>.</li>
<li><code>[+C, -B]</code> which produces the colored edge <code>(C_h, B_h)</code>.</li>
<li><code>[-B, +D]</code> which produces the colored edge <code>(B_t, D_t)</code>.</li>
<li><code>[+D, +A]</code> which produces the colored edge <code>(D_h, A_t)</code>.</li>
</ul>
<p>For linear chromosomes, the sliding window is not cyclic and the chromosomes always start and end at the termination node. For example, the <a href="#BM_(permutation)_GR%2Fi">permutation</a> <code>[+A, +C, -B, +D]</code> would actually be treated as <code>[TERM, +A, +C, -B, +D, TERM]</code>, resulting in ...</p>
<ul>
<li><code>[TERM, +A]</code> which produces the colored edge <code>(TERM, A_h)</code>.</li>
<li><code>[+A, +C]</code> which produces the colored edge <code>(A_h, C_t)</code>.</li>
<li><code>[+C, -B]</code> which produces the colored edge <code>(C_h, B_h)</code>.</li>
<li><code>[-B, +D]</code> which produces the colored edge <code>(B_t, D_t)</code>.</li>
<li><code>[+D, TERM]</code> which produces the colored edge <code>(D_h, TERM)</code>.</li>
</ul>
<p><a href="ch6_code/src/breakpoint_graph/Permutation.py">ch6_code/src/breakpoint_graph/Permutation.py</a> (lines 111 to 146):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">to_colored_edges</span>(<span class="hljs-params">self</span>) -&gt; <span class="hljs-type">List</span>[ColoredEdge]:
    ret = []
    <span class="hljs-comment"># add link to dummy head if linear</span>
    <span class="hljs-keyword">if</span> <span class="hljs-keyword">not</span> self.cyclic:
        b = self.blocks[<span class="hljs-number">0</span>]
        ret.append(
            ColoredEdge(TerminalNode.INST, b.to_synteny_edge().n1)
        )
    <span class="hljs-comment"># add normal edges</span>
    <span class="hljs-keyword">for</span> (b1, b2), idx <span class="hljs-keyword">in</span> slide_window(self.blocks, <span class="hljs-number">2</span>, cyclic=self.cyclic):
        <span class="hljs-keyword">if</span> b1.<span class="hljs-built_in">dir</span> == Direction.BACKWARD <span class="hljs-keyword">and</span> b2.<span class="hljs-built_in">dir</span> == Direction.FORWARD:
            n1 = SyntenyNode(b1.<span class="hljs-built_in">id</span>, SyntenyEnd.HEAD)
            n2 = SyntenyNode(b2.<span class="hljs-built_in">id</span>, SyntenyEnd.HEAD)
        <span class="hljs-keyword">elif</span> b1.<span class="hljs-built_in">dir</span> == Direction.FORWARD <span class="hljs-keyword">and</span> b2.<span class="hljs-built_in">dir</span> == Direction.BACKWARD:
            n1 = SyntenyNode(b1.<span class="hljs-built_in">id</span>, SyntenyEnd.TAIL)
            n2 = SyntenyNode(b2.<span class="hljs-built_in">id</span>, SyntenyEnd.TAIL)
        <span class="hljs-keyword">elif</span> b1.<span class="hljs-built_in">dir</span> == Direction.FORWARD <span class="hljs-keyword">and</span> b2.<span class="hljs-built_in">dir</span> == Direction.FORWARD:
            n1 = SyntenyNode(b1.<span class="hljs-built_in">id</span>, SyntenyEnd.TAIL)
            n2 = SyntenyNode(b2.<span class="hljs-built_in">id</span>, SyntenyEnd.HEAD)
        <span class="hljs-keyword">elif</span> b1.<span class="hljs-built_in">dir</span> == Direction.BACKWARD <span class="hljs-keyword">and</span> b2.<span class="hljs-built_in">dir</span> == Direction.BACKWARD:
            n1 = SyntenyNode(b1.<span class="hljs-built_in">id</span>, SyntenyEnd.HEAD)
            n2 = SyntenyNode(b2.<span class="hljs-built_in">id</span>, SyntenyEnd.TAIL)
        <span class="hljs-keyword">else</span>:
            <span class="hljs-keyword">raise</span> ValueError(<span class="hljs-string">'???'</span>)
        ret.append(
            ColoredEdge(n1, n2)
        )
    <span class="hljs-comment"># add link to dummy tail if linear</span>
    <span class="hljs-keyword">if</span> <span class="hljs-keyword">not</span> self.cyclic:
        b = self.blocks[-<span class="hljs-number">1</span>]
        ret.append(
            ColoredEdge(b.to_synteny_edge().n2, TerminalNode.INST)
        )
    <span class="hljs-comment"># return</span>
    <span class="hljs-keyword">return</span> ret
</code></pre>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>If you're confused at this point, don't continue. Go back and make sure you understand because in the next section builds on the above content.</p>
</div>
<p><strong><a href="#BM_%5Cb(2-break%7C2%20break%7Ctwo%20break%7Ctwo-break)s%3F%5Cb%2Fi">TWO-BREAK</a> ALGORITHM</strong></p>
<p>Now that <a href="#BM_(breakpoint%20graph)_GR%2Fi">breakpoint graph</a>s have been adequately described, the goal of this algorithm is to iteratively re-wire the red edges of a <a href="#BM_(breakpoint%20graph)_GR%2Fi">breakpoint graph</a> such that they match its blue edges. At each step, the algorithm finds a pair of red edges that share nodes with a blue edge and re-wires those red edges such that one of them matches the blue edge.</p>
<p>For example, the two red edges highlighted below share the same nodes as a blue edge (D_h and C_t). These two red edges can be broken and re-wired such that one of them matches the blue edge...</p>
<p><img src="dot_2ca5243fc24500a47a2859ee21a4a61a.svg" alt="Dot diagram"></p>
<p>Each re-wiring operation is called a <a href="#BM_%5Cb(2-break%7C2%20break%7Ctwo%20break%7Ctwo-break)s%3F%5Cb%2Fi">2-break</a> and represents either a chromosome <a href="#BM_(%5Cbfusion%7C%5Cbfuse)%2Fi">fusion</a>, chromosome <a href="#BM_(fission)%2Fi">fission</a>, or reversal mutation (<a href="#BM_(genome%20rearrangement%7Cchromosomal%20rearrangement%7Cchromosome%20rearrangement)%2Fi">genome rearrangement</a>). For example, ...</p>
<ul>
<li>
<p><a href="#BM_(%5Cbfusion%7C%5Cbfuse)%2Fi">fusion</a>:</p>
<p><img src="svgbob_100dedbf586bc55bfbd2b51362b42edf00acaa9f.svg" alt="Kroki diagram output"></p>
<p><img src="dot_36054e551339873280bfdb261b65fed3.svg" alt="Dot diagram"></p>
</li>
<li>
<p><a href="#BM_(fission)%2Fi">fission</a>:</p>
<p><img src="svgbob_3c65208a8a1bb5ee1eaf4e56e21e05932b8900e5.svg" alt="Kroki diagram output"></p>
<p><img src="dot_da1cc3fbe7558a51c166e6f3562d4272.svg" alt="Dot diagram"></p>
</li>
<li>
<p>reversal:</p>
<p><img src="svgbob_9e15af9240b70ac8fce6c21446c8b507ecb50037.svg" alt="Kroki diagram output"></p>
<p><img src="dot_ad962f4590c01b25914ecf117d94ef2d.svg" alt="Dot diagram"></p>
</li>
</ul>
<p><a href="#BM_(genome%20rearrangement%7Cchromosomal%20rearrangement%7Cchromosome%20rearrangement)%2Fi">Genome rearrangement</a> duplications and deletions aren't representable as <a href="#BM_%5Cb(2-break%7C2%20break%7Ctwo%20break%7Ctwo-break)s%3F%5Cb%2Fi">2-break</a>s. <a href="#BM_(genome%20rearrangement%7Cchromosomal%20rearrangement%7Cchromosome%20rearrangement)%2Fi">Genome rearrangement</a> <a href="#BM_(translocation)%2Fi">translocation</a>s can't be reliably represented as a single <a href="#BM_%5Cb(2-break%7C2%20break%7Ctwo%20break%7Ctwo-break)s%3F%5Cb%2Fi">2-break</a> either. For example, the following <a href="#BM_(translocation)%2Fi">translocation</a> gets modeled as two <a href="#BM_%5Cb(2-break%7C2%20break%7Ctwo%20break%7Ctwo-break)s%3F%5Cb%2Fi">2-break</a>s, one that breaks the undesired chromosome (<a href="#BM_(fission)%2Fi">fission</a>) and another that glues it back together (<a href="#BM_(%5Cbfusion%7C%5Cbfuse)%2Fi">fusion</a>)...</p>
<p><img src="svgbob_5029f3007e1cfcf23146d31495be8429cfee8bc9.svg" alt="Kroki diagram output"></p>
<p><img src="dot_3768cf3fdb4dc533017bae0f868d5b87.svg" alt="Dot diagram"></p>
<p><img src="dot_63a4de777771e77adb6bc5786313e468.svg" alt="Dot diagram"></p>
<p><a href="ch6_code/src/breakpoint_graph/ColoredEdge.py">ch6_code/src/breakpoint_graph/ColoredEdge.py</a> (lines 46 to 86):</p>
<pre class="hljs"><code><span class="hljs-comment"># Takes e1 and e2 and swaps the ends, such that one of the swapped edges becomes desired_e. That is, e1 should have</span>
<span class="hljs-comment"># an end matching one of desired_e's ends while e2 should have an end matching desired_e's other end.</span>
<span class="hljs-comment">#</span>
<span class="hljs-comment"># This is basically a 2-break.</span>
<span class="hljs-meta">@staticmethod</span>
<span class="hljs-keyword">def</span> <span class="hljs-title function_">swap_ends</span>(<span class="hljs-params">
        e1: <span class="hljs-type">Optional</span>[ColoredEdge],
        e2: <span class="hljs-type">Optional</span>[ColoredEdge],
        desired_e: ColoredEdge
</span>) -&gt; <span class="hljs-type">Optional</span>[ColoredEdge]:
    <span class="hljs-keyword">if</span> e1 <span class="hljs-keyword">is</span> <span class="hljs-literal">None</span> <span class="hljs-keyword">and</span> e2 <span class="hljs-keyword">is</span> <span class="hljs-literal">None</span>:
        <span class="hljs-keyword">raise</span> ValueError(<span class="hljs-string">'Both edges can\'t be None'</span>)
    <span class="hljs-keyword">if</span> TerminalNode.INST <span class="hljs-keyword">in</span> desired_e:
        <span class="hljs-comment"># In this case, one of desired_e's ends is TERM (they can't both be TERM). That means either e1 or e2 will</span>
        <span class="hljs-comment"># be None because there's only one valid end (non-TERM end) to swap with.</span>
        _e = <span class="hljs-built_in">next</span>(<span class="hljs-built_in">filter</span>(<span class="hljs-keyword">lambda</span> x: x <span class="hljs-keyword">is</span> <span class="hljs-keyword">not</span> <span class="hljs-literal">None</span>, [e1, e2]), <span class="hljs-literal">None</span>)
        <span class="hljs-keyword">if</span> _e <span class="hljs-keyword">is</span> <span class="hljs-literal">None</span>:
            <span class="hljs-keyword">raise</span> ValueError(<span class="hljs-string">'If the desired edge has a terminal node, one of the edges must be None'</span>)
        <span class="hljs-keyword">if</span> desired_e.non_term() <span class="hljs-keyword">not</span> <span class="hljs-keyword">in</span> {_e.n1, _e.n2}:
            <span class="hljs-keyword">raise</span> ValueError(<span class="hljs-string">'Unexpected edge node(s) encountered'</span>)
        <span class="hljs-keyword">if</span> desired_e == _e:
            <span class="hljs-keyword">raise</span> ValueError(<span class="hljs-string">'Edge is already desired edge'</span>)
        other_n1 = _e.other_end(desired_e.non_term())
        other_n2 = TerminalNode.INST
        <span class="hljs-keyword">return</span> ColoredEdge(other_n1, other_n2)
    <span class="hljs-keyword">else</span>:
        <span class="hljs-comment"># In this case, neither of desired_e's ends is TERM. That means both e1 and e2 will be NOT None.</span>
        <span class="hljs-keyword">if</span> desired_e <span class="hljs-keyword">in</span> {e1, e2}:
            <span class="hljs-keyword">raise</span> ValueError(<span class="hljs-string">'Edge is already desired edge'</span>)
        <span class="hljs-keyword">if</span> desired_e.n1 <span class="hljs-keyword">in</span> e1 <span class="hljs-keyword">and</span> desired_e.n2 <span class="hljs-keyword">in</span> e2:
            other_n1 = e1.other_end(desired_e.n1)
            other_n2 = e2.other_end(desired_e.n2)
        <span class="hljs-keyword">elif</span> desired_e.n1 <span class="hljs-keyword">in</span> e2 <span class="hljs-keyword">and</span> desired_e.n2 <span class="hljs-keyword">in</span> e1:
            other_n1 = e2.other_end(desired_e.n1)
            other_n2 = e1.other_end(desired_e.n2)
        <span class="hljs-keyword">else</span>:
            <span class="hljs-keyword">raise</span> ValueError(<span class="hljs-string">'Unexpected edge node(s) encountered'</span>)
        <span class="hljs-keyword">if</span> {other_n1, other_n2} == {TerminalNode.INST}:  <span class="hljs-comment"># if both term edges, there is no other edge</span>
            <span class="hljs-keyword">return</span> <span class="hljs-literal">None</span>
        <span class="hljs-keyword">return</span> ColoredEdge(other_n1, other_n2)
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Applying 2-breaks on circular genome until red_p_list=[['+A', '-B', '-C', '+D'], ['+E']] matches blue_p_list=[['+A', '+B', '-D'], ['-C', '-E']] (show_graph=True)...</p>
<ul>
<li>
<p>Initial red_p_list=[['+A', '-B', '-C', '+D'], ['+E']]</p>
<p><img src="dot_897226d2234aef0c3564387d3f0438de.svg" alt="Dot diagram"></p>
</li>
<li>
<p>red_p_list=[['+A', '+B', '-C', '+D'], ['+E']]</p>
<p><img src="dot_99fb7bbbeb92ee0d4cd070244cf398c4.svg" alt="Dot diagram"></p>
</li>
<li>
<p>red_p_list=[['+A', '+B', '-D', '+C'], ['+E']]</p>
<p><img src="dot_d861cf60d293adaac49b9768348dee68.svg" alt="Dot diagram"></p>
</li>
<li>
<p>red_p_list=[['+A', '+B', '-D', '+C', '+E']]</p>
<p><img src="dot_71045e440d2232f9af8d0f048c44b1ed.svg" alt="Dot diagram"></p>
</li>
<li>
<p>red_p_list=[['+C', '+E'], ['+A', '+B', '-D']]</p>
<p><img src="dot_78d9644adba20b05ba3fb2d127adc01e.svg" alt="Dot diagram"></p>
</li>
</ul>
<p>Recall that the breakpoint graph is undirected. A permutation may have been walked in either direction (clockwise vs counter-clockwise) and there are multiple nodes to start walking from. If the output looks like it's going backwards, that's just as correct as if it looked like it's going forward.</p>
<p>Also, recall that a genome is represented as a set of permutations -- sets are not ordered.</p>
</div>
<p></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>It isn't discussed here, but the Pevzner book put an emphasis on calculating the <a href="#BM_(parsimony%7Cparsimonious)%2Fi">parsimonious</a> number of reversals (<a href="#BM_(reversal%20distance)%2Fi">reversal distance</a>) without having to go through and apply <a href="#BM_%5Cb(2-break%7C2%20break%7Ctwo%20break%7Ctwo-break)s%3F%5Cb%2Fi">two-break</a>s in the <a href="#BM_(breakpoint%20graph)_GR%2Fi">breakpoint graph</a>. The basic idea is to count the number of red-blue cycles in the graph.</p>
<p>For a cyclic <a href="#BM_(breakpoint%20graph)_GR%2Fi">breakpoint graph</a>s, a single red-blue cycle is when you pick a node, follow the blue edge, then follow the red edge, then follow the blue edge, then follow the red edge, ..., until you arrive back at the same node. If the blue and red <a href="#BM_(genome)%2Fi">genome</a>s match perfectly, the number of red-blue cycles should equal the number of <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s. Otherwise, you can calculate the number of reversals needed to get them to equal by subtracting the number of red-blue cycles by the number of <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s.</p>
<p>For a linear <a href="#BM_(breakpoint%20graph)_GR%2Fi">breakpoint graph</a>s, a single red-blue cycle isn't actually a cycle because: Pick the termination node, follow a blue edge, then follow the red edge, then follow the blue edge, then follow the red edge, ... until you arrive back at the termination node (what if there are actual cyclic red-blue loops as well like in cyclic <a href="#BM_(breakpoint%20graph)_GR%2Fi">breakpoint graph</a>s?). If the blue and red <a href="#BM_(genome)%2Fi">genome</a>s match perfectly, the number of red-blue cycles should equal the number of <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s + 1. Otherwise, you can <strong>ESTIMATE</strong> the number of reversals needed to get them to equal by subtracting the number of red-blue cycles by the number of <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s + 1.</p>
<p>To calculate the real number of reversals need for linear <a href="#BM_(breakpoint%20graph)_GR%2Fi">breakpoint graph</a>s (not estimate), there's a <a href="https://dl.acm.org/doi/10.1145/300515.300516">paper on ACM DL</a> that goes over the algorithm. I glanced through it but I don't have the time / wherewithal to go through it. Maybe do it in the future.</p>
<p>UPDATE: Calculating the number of reversals quickly is important because the number of reversals can be used as a <a href="#BM_(distance%20metric)%2Fi">distance metric</a> when computing a <a href="#BM_(phylogenetic%20tree%7Cphylogeny%20tree%7Cphylogenies%7Cevolutionary%20tree)%2Fi">phylogenetic tree</a> across a set of species (a <a href="#BM_(tree)%2Fi">tree</a> that shows how closely a set of species are related / how they branched out). See <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> definition.</p>
</div>
<a name="H_Phylogeny"></a><h2>Phylogeny</h2>
<p><a name="BM_(Algorithms%5C%2FPhylogeny)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FK-mer)_TOPIC%2F">Algorithms/K-mer</a></li>
</ul>
</div>
<p><a href="#BM_(phylogeny%7Cphylogenetic)%2Fi">Phylogeny</a> is the concept of inferring the evolutionary history of a set of biological entities (e.g. animal species, viruses, etc..) by inspecting properties of those entities for relatedness (e.g. phenotypic, genotypic, etc..).</p>
<p><img src="svgbob_9ad92bcc26081dc1ee059aa771d26eb82d6e4ad4.svg" alt="Kroki diagram output"></p>
<p>Evolutionary history is often displayed as a <a href="#BM_(tree)%2Fi">tree</a> called a <a href="#BM_(phylogenetic%20tree%7Cphylogeny%20tree%7Cphylogenies%7Cevolutionary%20tree)%2Fi">phylogenetic tree</a>, where leaf nodes represent known entities and internal nodes represent inferred ancestor entities. The example above shows a <a href="#BM_(phylogenetic%20tree%7Cphylogeny%20tree%7Cphylogenies%7Cevolutionary%20tree)%2Fi">phylogenetic tree</a> for the species cat, lion, and bear based on phenotypic inspection. Cats and lions are inferred as descending from the same ancestor because both have deeply shared physical and behavioural characteristics (felines). Similarly, that feline ancestor and bears are inferred as descending from the same ancestor because all descendants walk on 4 legs.</p>
<p>The typical process for <a href="#BM_(phylogeny%7Cphylogenetic)%2Fi">phylogeny</a> is to first measure how related a set of entities are to each other, where each measure is referred to as a distance (e.g. <code>dist(cat, lion) = 2</code>), then work backwards to find a <a href="#BM_(phylogenetic%20tree%7Cphylogeny%20tree%7Cphylogenies%7Cevolutionary%20tree)%2Fi">phylogenetic tree</a> that fits / maps to those distances. The distance may be any metric so long as ...</p>
<ul>
<li>it computes the distance to itself as 0 (e.g. <code>dist(cat, cat) = 0</code>)</li>
<li>it computes the distance to any other entity as &gt; 0 (e.g. <code>dist(cat, lion) = 2</code>)</li>
<li>it computes the same distance for the same pair regardless of order (e.g. <code>dist(cat, lion) = dist(lion, cat)</code>)</li>
<li>computed distances don't leapfrog each other (e.g. <code>dist(cat, lion) + dist(lion, dog) &gt;= dist(cat, dog)</code>)</li>
</ul>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The leapfrogging point may be confusing. All it's saying is that taking an indirect path between two species should produce a distance that's &gt;= the direct path. For example, the direct path between cat and dog is 6: <code>dist(cat, dog) = 6</code>. If you were to instead jump from cat to lion <code>dist(cat, lion) = 2</code>, then from lion to dog <code>dist(lion, dog) = 5</code>, that combined distance should be &gt;= to 6...</p>
<pre><code>dist(cat, dog)  = 6
dist(cat, lion) = 2
dist(lion, dog) = 5

dist(cat, lion) + dist(lion, dog) &gt;= dist(cat, dog)
        2       +        5        &gt;=       6
                7                 &gt;=       6
</code></pre>
<p>The Pevzner book refers to this as the triangle inequality.</p>
<p><img src="svgbob_a535ea37908c1e912c089d9ea621240bae63f541.svg" alt="Kroki diagram output"></p>
<p>Later on non-conforming <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrices</a> are discussed called non-<a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrices</a>. I don't know if non-<a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrices</a> are required to have this specific property, but they should have all others.</p>
</div>
<p>Examples of metrics that may be used as distance, referred to as <a href="#BM_(distance%20metric)%2Fi">distance metric</a>s, include...</p>
<ul>
<li><a href="#BM_(hamming%20distance)%2Fi">hamming distance</a> between DNA / protein <a href="#BM_(sequence)%2Fi">sequence</a>s.</li>
<li><a href="#BM_(global%20alignment%7Cglobal%20sequence%20alignment)%2Fi">global alignment</a> score between DNA / protein <a href="#BM_(sequence)%2Fi">sequence</a>s.</li>
<li><a href="#BM_%5Cb(2-break%7C2%20break%7Ctwo%20break%7Ctwo-break)s%3F%5Cb%2Fi">two-break</a> count (<a href="#BM_(reversal%20distance)%2Fi">reversal distance</a>).</li>
<li>number of similar physical or behavioural attributes.</li>
<li>etc..</li>
</ul>
<p>Distances for a set of entities are typically represented as a 2D matrix that contains all possible pairings, called a <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a>. The <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> for the example Cat/Lion/Bear <a href="#BM_(phylogenetic%20tree%7Cphylogeny%20tree%7Cphylogenies%7Cevolutionary%20tree)%2Fi">phylogenetic tree</a> is ...</p>
<table>
<thead>
<tr>
<th></th>
<th>Cat</th>
<th>Lion</th>
<th>Bear</th>
</tr>
</thead>
<tbody>
<tr>
<td>Cat</td>
<td>0</td>
<td>2</td>
<td>23</td>
</tr>
<tr>
<td>Lion</td>
<td>2</td>
<td>0</td>
<td>23</td>
</tr>
<tr>
<td>Bear</td>
<td>23</td>
<td>23</td>
<td>0</td>
</tr>
</tbody>
</table>
<p><img src="svgbob_a4884cea1f03dad70aa6cd67fa7232f3a6772e94.svg" alt="Kroki diagram output"></p>
<p>Note how the <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> has the distance for each pair slotted twice, mirrored across the diagonal of 0s (self distances). For example, the distance between bear and lion is listed twice.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>Just to make it explicit: The ultimate point of this section is to work backwards from a <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> to a <a href="#BM_(phylogenetic%20tree%7Cphylogeny%20tree%7Cphylogenies%7Cevolutionary%20tree)%2Fi">phylogenetic tree</a> (essentially the concept of <a href="#BM_(phylogeny%7Cphylogenetic)%2Fi">phylogeny</a> -- inferring evolutionary history of a set of known / present-day organisms based on how different they are).</p>
</div>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The best way to move forward with this, assuming that you're brand new to it, is to first understand the following four subsections...</p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FTree%20to%20Additive%20Distance%20Matrix)_TOPIC%2F">Algorithms/Phylogeny/Tree to Additive Distance Matrix</a></li>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FTree%20to%20Simple%20Tree)_TOPIC%2F">Algorithms/Phylogeny/Tree to Simple Tree</a></li>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FAdditive%20Distance%20Matrix%20Cardinality)_TOPIC%2F">Algorithms/Phylogeny/Additive Distance Matrix Cardinality</a></li>
</ul>
<p>Then jump to the algorithm you want to learn (subsection) within <a href="#BM_(Algorithms%5C%2FPhylogeny%2FDistance%20Matrix%20to%20Tree)_TOPIC%2F">Algorithms/Phylogeny/Distance Matrix to Tree</a> and work from the prerequisites to the algorithm. Otherwise all the sections in between come off as disjointed because it's building the intermediate knowledge required for the final algorithms.</p>
</div>
<a name="H_Tree%20to%20Additive%20Distance%20Matrix"></a><h3>Tree to Additive Distance Matrix</h3>
<p><a name="BM_(Algorithms%5C%2FPhylogeny%2FTree%20to%20Additive%20Distance%20Matrix)_TOPIC%2F"></a><strong></strong></p>
<p><strong>WHAT</strong>: Given a <a href="#BM_(tree)%2Fi">tree</a>, the <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> generated from that <a href="#BM_(tree)%2Fi">tree</a> is said to be an <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>.</p>
<p><strong>WHY</strong>: The term <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> is derived from the fact that edge weights within the <a href="#BM_(tree)%2Fi">tree</a> are being <em>added</em> together to generate the distances in the <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a>. For example, in the following <a href="#BM_(tree)%2Fi">tree</a> ...</p>
<p><img src="svgbob_82d2f505bd1ac27f99e4162364a2faefde177c9e.svg" alt="Kroki diagram output"></p>
<ul>
<li><code>dist(Cat, Lion)  = dist(Cat, A) + dist(A, Lion)               = 1 + 1     = 2</code></li>
<li><code>dist(Cat, Bear)  = dist(Cat, A) + dist(A, B) + dist(B, Bear)  = 1 + 1 + 2 = 4</code></li>
<li><code>dist(Lion, Bear) = dist(Lion, A) + dist(A, B) + dist(B, Bear) = 1 + 1 + 2 = 4</code></li>
</ul>
<table>
<thead>
<tr>
<th></th>
<th>Cat</th>
<th>Lion</th>
<th>Bear</th>
</tr>
</thead>
<tbody>
<tr>
<td>Cat</td>
<td>0</td>
<td>2</td>
<td>4</td>
</tr>
<tr>
<td>Lion</td>
<td>2</td>
<td>0</td>
<td>4</td>
</tr>
<tr>
<td>Bear</td>
<td>4</td>
<td>4</td>
<td>0</td>
</tr>
</tbody>
</table>
<p>However, <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrices</a> aren't commonly generated from <a href="#BM_(tree)%2Fi">tree</a>s. Rather, they're generated by comparing present-day entities to each other to see how diverged they are (their distance from each other). There's no guarantee that a <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> generated from comparisons will be an <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>. That is, there must exist a <a href="#BM_(tree)%2Fi">tree</a> with edge weights that satisfy that <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> for it to be an <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> (commonly referred to as a <a href="#BM_(tree)%2Fi">tree</a> that fits the <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a>).</p>
<p>In other words, while a...</p>
<ul>
<li>
<p><a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> generated from a <a href="#BM_(tree)%2Fi">tree</a> will always be an <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>, not all <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrices</a> are <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrices</a>. For example, a <a href="#BM_(tree)%2Fi">tree</a> doesn't exist that maps to the following <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> ...</p>
<table>
<thead>
<tr>
<th></th>
<th>Cat</th>
<th>Lion</th>
<th>Bear</th>
<th>Racoon</th>
</tr>
</thead>
<tbody>
<tr>
<td>Cat</td>
<td>0</td>
<td>1</td>
<td>1</td>
<td>1</td>
</tr>
<tr>
<td>Lion</td>
<td>1</td>
<td>0</td>
<td>1</td>
<td>1</td>
</tr>
<tr>
<td>Bear</td>
<td>1</td>
<td>1</td>
<td>0</td>
<td>9</td>
</tr>
<tr>
<td>Racoon</td>
<td>1</td>
<td>1</td>
<td>9</td>
<td>0</td>
</tr>
</tbody>
</table>
</li>
<li>
<p><a href="#BM_(tree)%2Fi">tree</a> maps to exactly one <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>, that <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> maps to many different <a href="#BM_(tree)%2Fi">tree</a>s. For example, the following <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> may map to any of the following <a href="#BM_(tree)%2Fi">tree</a>s ...</p>
<table>
<thead>
<tr>
<th></th>
<th>Cat</th>
<th>Lion</th>
<th>Bear</th>
</tr>
</thead>
<tbody>
<tr>
<td>Cat</td>
<td>0</td>
<td>2</td>
<td>4</td>
</tr>
<tr>
<td>Lion</td>
<td>2</td>
<td>0</td>
<td>4</td>
</tr>
<tr>
<td>Bear</td>
<td>4</td>
<td>4</td>
<td>0</td>
</tr>
</tbody>
</table>
<p><img src="svgbob_213f51f10127b8c8073a3b266428aa0369911bc1.svg" alt="Kroki diagram output"></p>
</li>
</ul>
<p><strong>ALGORITHM</strong>:</p>
<p><a href="ch7_code/src/phylogeny/TreeToAdditiveDistanceMatrix.py">ch7_code/src/phylogeny/TreeToAdditiveDistanceMatrix.py</a> (lines 39 to 69):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">find_path</span>(<span class="hljs-params">g: Graph[N, ND, E, <span class="hljs-built_in">float</span>], n1: N, n2: N</span>) -&gt; <span class="hljs-built_in">list</span>[E]:
    <span class="hljs-keyword">if</span> <span class="hljs-keyword">not</span> g.has_node(n1) <span class="hljs-keyword">or</span> <span class="hljs-keyword">not</span> g.has_node(n2):
        ValueError(<span class="hljs-string">'Node missing'</span>)
    <span class="hljs-keyword">if</span> n1 == n2:
        <span class="hljs-keyword">return</span> []
    queued_edges = <span class="hljs-built_in">list</span>()
    <span class="hljs-keyword">for</span> e <span class="hljs-keyword">in</span> g.get_outputs(n1):
        queued_edges.append((n1, [e]))
    <span class="hljs-keyword">while</span> <span class="hljs-built_in">len</span>(queued_edges) &gt; <span class="hljs-number">0</span>:
        ignore_n, e_list = queued_edges.pop()
        e_last = e_list[-<span class="hljs-number">1</span>]
        active_n = [n <span class="hljs-keyword">for</span> n <span class="hljs-keyword">in</span> g.get_edge_ends(e_last) <span class="hljs-keyword">if</span> n != ignore_n][<span class="hljs-number">0</span>]
        <span class="hljs-keyword">if</span> active_n == n2:
            <span class="hljs-keyword">return</span> e_list
        children = <span class="hljs-built_in">set</span>(g.get_outputs(active_n))
        children.remove(e_last)
        <span class="hljs-keyword">for</span> child_e <span class="hljs-keyword">in</span> children:
            child_ignore_n = active_n
            new_e_list = e_list[:] + [child_e]
            queued_edges.append((child_ignore_n, new_e_list))
    <span class="hljs-keyword">raise</span> ValueError(<span class="hljs-string">f'No path from <span class="hljs-subst">{n1}</span> to <span class="hljs-subst">{n2}</span>'</span>)


<span class="hljs-keyword">def</span> <span class="hljs-title function_">to_additive_distance_matrix</span>(<span class="hljs-params">g: Graph[N, ND, E, <span class="hljs-built_in">float</span>]</span>) -&gt; DistanceMatrix[N]:
    leaves = {n <span class="hljs-keyword">for</span> n <span class="hljs-keyword">in</span> g.get_nodes() <span class="hljs-keyword">if</span> g.get_degree(n) == <span class="hljs-number">1</span>}
    dists = {}
    <span class="hljs-keyword">for</span> l1, l2 <span class="hljs-keyword">in</span> product(leaves, repeat=<span class="hljs-number">2</span>):
        d = <span class="hljs-built_in">sum</span>(g.get_edge_data(e) <span class="hljs-keyword">for</span> e <span class="hljs-keyword">in</span> find_path(g, l1, l2))
        dists[l1, l2] = d
    <span class="hljs-keyword">return</span> DistanceMatrix(dists)
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>The tree...</p>
<p><img src="dot_aa1feda6737daab4bc422403da61d84e.svg" alt="Dot diagram"></p>
<p>... produces the additive distance matrix ...</p>
<table>
<thead><tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr></thead>
<tbody>
<tr>
<td>v0</td>
<td>0.0</td>
<td>13.0</td>
<td>21.0</td>
<td>21.0</td>
<td>22.0</td>
<td>22.0</td>
</tr>
<tr>
<td>v1</td>
<td>13.0</td>
<td>0.0</td>
<td>12.0</td>
<td>12.0</td>
<td>13.0</td>
<td>13.0</td>
</tr>
<tr>
<td>v2</td>
<td>21.0</td>
<td>12.0</td>
<td>0.0</td>
<td>20.0</td>
<td>21.0</td>
<td>21.0</td>
</tr>
<tr>
<td>v3</td>
<td>21.0</td>
<td>12.0</td>
<td>20.0</td>
<td>0.0</td>
<td>7.0</td>
<td>13.0</td>
</tr>
<tr>
<td>v4</td>
<td>22.0</td>
<td>13.0</td>
<td>21.0</td>
<td>7.0</td>
<td>0.0</td>
<td>14.0</td>
</tr>
<tr>
<td>v5</td>
<td>22.0</td>
<td>13.0</td>
<td>21.0</td>
<td>13.0</td>
<td>14.0</td>
<td>0.0</td>
</tr>
</tbody>
</table>
</div>
<p></p>
<a name="H_Tree%20to%20Simple%20Tree"></a><h3>Tree to Simple Tree</h3>
<p><a name="BM_(Algorithms%5C%2FPhylogeny%2FTree%20to%20Simple%20Tree)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FTree%20to%20Additive%20Distance%20Matrix)_TOPIC%2F">Algorithms/Phylogeny/Tree to Additive Distance Matrix</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: Convert a <a href="#BM_(tree)%2Fi">tree</a> into a <a href="#BM_(simple%20tree)%2Fi">simple tree</a>. A <a href="#BM_(simple%20tree)%2Fi">simple tree</a> is an <a href="#BM_(un-rooted%20tree%7Cunrooted%20tree)%2Fi">unrooted tree</a> where ...</p>
<ul>
<li>every internal node has a <a href="#BM_(degree)_GRAPH%2Fi">degree</a> &gt; 2.</li>
<li>every edge has a weight of &gt; 0.</li>
</ul>
<p>The first point just means that the <a href="#BM_(tree)%2Fi">tree</a> can't contain non-splitting internal nodes. By definition a <a href="#BM_(tree)%2Fi">tree</a>'s leaf nodes each have a <a href="#BM_(degree)_GRAPH%2Fi">degree</a> of 1, and this restriction makes it so that each internal node must have a <a href="#BM_(degree)_GRAPH%2Fi">degree</a> &gt; 2 instead of &gt;= 2...</p>
<p><img src="svgbob_8bdf7a40cfcab6612aecbb3bbbf18a9a48455ed4.svg" alt="Kroki diagram output"></p>
<p><img src="svgbob_a5aaa10a8a80878b13db3c7e7a4e5ffee7ffa19c.svg" alt="Kroki diagram output"></p>
<p>In the context of <a href="#BM_(phylogeny%7Cphylogenetic)%2Fi">phylogeny</a>, a <a href="#BM_(simple%20tree)%2Fi">simple tree</a>'s ...</p>
<ul>
<li>leaf nodes represent known entities.</li>
<li>internal nodes represent inferred ancestor entities.</li>
<li>edge weights represent distances between entities.</li>
</ul>
<p><strong>WHY</strong>: <a href="#BM_(simple%20tree)%2Fi">Simple tree</a>s have properties / restrictions that simplify the process of working backwards from a <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> to a <a href="#BM_(tree)%2Fi">tree</a>. In other words, when constructing a <a href="#BM_(tree)%2Fi">tree</a> from a <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a>, the process is simpler if the <a href="#BM_(tree)%2Fi">tree</a> is restricted to being a <a href="#BM_(simple%20tree)%2Fi">simple tree</a>.</p>
<p>The first property is that a unique <a href="#BM_(simple%20tree)%2Fi">simple tree</a> exists for a unique <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> (one-to-one mapping). That is, it isn't possible for...</p>
<ul>
<li>two different <a href="#BM_(simple%20tree)%2Fi">simple tree</a>s to map to the same <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a>.</li>
<li>two different <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrices</a> to map to the same <a href="#BM_(simple%20tree)%2Fi">simple tree</a>.</li>
</ul>
<p>For example, the following <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> will only ever map to the following <a href="#BM_(simple%20tree)%2Fi">simple tree</a> (and vice-versa)...</p>
<table>
<thead>
<tr>
<th></th>
<th>w</th>
<th>u</th>
<th>y</th>
<th>z</th>
</tr>
</thead>
<tbody>
<tr>
<td>w</td>
<td>0</td>
<td>3</td>
<td>8</td>
<td>7</td>
</tr>
<tr>
<td>u</td>
<td>3</td>
<td>0</td>
<td>9</td>
<td>8</td>
</tr>
<tr>
<td>y</td>
<td>8</td>
<td>9</td>
<td>0</td>
<td>5</td>
</tr>
<tr>
<td>z</td>
<td>7</td>
<td>8</td>
<td>5</td>
<td>0</td>
</tr>
</tbody>
</table>
<p><img src="svgbob_a74e35c81b15ac05ccf607f36f8e723935570513.svg" alt="Kroki diagram output"></p>
<p>However, that same <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> can map to an infinite number of non-<a href="#BM_(simple%20tree)%2Fi">simple tree</a>s (and vice-versa)...</p>
<p><img src="svgbob_403c04944e435cd926cbba7d906dc419b29619d7.svg" alt="Kroki diagram output"></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>To clarify: This property / restriction is important because when reconstructing a <a href="#BM_(tree)%2Fi">tree</a> from the <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a>, if you restrict yourself to a <a href="#BM_(simple%20tree)%2Fi">simple tree</a> you'll only ever have 1 <a href="#BM_(tree)%2Fi">tree</a> to reconstruct to. This makes the algorithms simpler. This is discussed further in the cardinality subsection.</p>
</div>
<p>The second property is that the direction of evolution isn't maintained in a <a href="#BM_(simple%20tree)%2Fi">simple tree</a>: It's an <a href="#BM_(un-rooted%20tree%7Cunrooted%20tree)%2Fi">unrooted tree</a> with undirected edges. This is a useful property because, while a <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> may provide enough information to infer common ancestry, it doesn't provide enough information to know the true parent-child relationships between those ancestors. For example, any of the internal nodes in the following <a href="#BM_(simple%20tree)%2Fi">simple tree</a> may be the top-level entity that all other entities are descendants of ...</p>
<p><img src="svgbob_f62876489da68170731df6b6d472bcccbba0af31.svg" alt="Kroki diagram output"></p>
<p>The third property is that weights must be &gt; 0, which is because of the restriction on <a href="#BM_(distance%20metric)%2Fi">distance metric</a>s specified in the parent section: The distance between any two entities must be &gt; 0. That is, it doesn't make sense for the distance between two entities to be ...</p>
<ul>
<li>&lt; 0 because distance represents how diverged the entities are from each other. Having a negative amount of divergence doesn't make sense.</li>
<li>= 0 because then the two nodes between the edge would represent the same entity. Having more than one representation of the same entity in the <a href="#BM_(tree)%2Fi">tree</a> doesn't make sense.</li>
</ul>
<p><img src="svgbob_5a3a67175e47c1737b9aec7ef4704fac8cac6a97.svg" alt="Kroki diagram output"></p>
<p><strong>ALGORITHM</strong>:</p>
<p>The following examples show various real evolutionary paths and their corresponding <a href="#BM_(simple%20tree)%2Fi">simple tree</a>s. Note how the <a href="#BM_(simple%20tree)%2Fi">simple tree</a>s neither fully represent the true lineage nor the direction of evolution (<a href="#BM_(simple%20tree)%2Fi">simple tree</a>s are unrooted and undirected).</p>
<p><img src="svgbob_86f08fc5c51796312a431fa11fa8a143a7525756.svg" alt="Kroki diagram output"></p>
<p>In the first two examples, one present-day entity branched off from another present-day entity. Both entities are still present-day entities (the entity branched off from isn't extinct).</p>
<p>In the fifth example, parent1 split off to the present-day entities entity1 and entity3, then entity2 branched off entity1. All three entities are present-day entities (neither entity1, entity2, nor entity3 is extinct).</p>
<p>In the third and last two examples, the top-level parent doesn't show up because adding it would break the requirement that internal nodes must be splitting (<a href="#BM_(degree)_GRAPH%2Fi">degree</a> &gt; 2). For example, adding parent1 into the <a href="#BM_(simple%20tree)%2Fi">simple tree</a> of the last example above causes parent1 to have a <a href="#BM_(degree)_GRAPH%2Fi">degree</a> = 2...</p>
<p><img src="svgbob_895c08d3866dea4e938d43c7c62da5224abdeb82.svg" alt="Kroki diagram output"></p>
<p>The following algorithm removes nodes of <a href="#BM_(degree)_GRAPH%2Fi">degree</a> = 2, merging its two edges together. This makes it so every internal edge has a <a href="#BM_(degree)_GRAPH%2Fi">degree</a> of &gt; 2...</p>
<p><a href="ch7_code/src/phylogeny/TreeToSimpleTree.py">ch7_code/src/phylogeny/TreeToSimpleTree.py</a> (lines 88 to 105):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">merge_nodes_of_degree2</span>(<span class="hljs-params">g: Graph[N, ND, E, <span class="hljs-built_in">float</span>]</span>) -&gt; <span class="hljs-literal">None</span>:
    <span class="hljs-comment"># Can be made more efficient by not having to re-collect bad nodes each</span>
    <span class="hljs-comment"># iteration. Kept it like this so it's simple to understand what's going on.</span>
    <span class="hljs-keyword">while</span> <span class="hljs-literal">True</span>:
        bad_nodes = {n <span class="hljs-keyword">for</span> n <span class="hljs-keyword">in</span> g.get_nodes() <span class="hljs-keyword">if</span> g.get_degree(n) == <span class="hljs-number">2</span>}
        <span class="hljs-keyword">if</span> <span class="hljs-built_in">len</span>(bad_nodes) == <span class="hljs-number">0</span>:
            <span class="hljs-keyword">return</span>
        bad_n = bad_nodes.pop()
        bad_e1, bad_e2 = <span class="hljs-built_in">tuple</span>(g.get_outputs(bad_n))
        e_id = bad_e1 + bad_e2
        e_n1 = [n <span class="hljs-keyword">for</span> n <span class="hljs-keyword">in</span> g.get_edge_ends(bad_e1) <span class="hljs-keyword">if</span> n != bad_n][<span class="hljs-number">0</span>]
        e_n2 = [n <span class="hljs-keyword">for</span> n <span class="hljs-keyword">in</span> g.get_edge_ends(bad_e2) <span class="hljs-keyword">if</span> n != bad_n][<span class="hljs-number">0</span>]
        e_weight = g.get_edge_data(bad_e1) + g.get_edge_data(bad_e2)
        g.insert_edge(e_id, e_n1, e_n2, e_weight)
        g.delete_edge(bad_e1)
        g.delete_edge(bad_e2)
        g.delete_node(bad_n)
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>The tree...</p>
<p><img src="dot_a6073dfa575690e209a573018def1057.svg" alt="Dot diagram"></p>
<p>... simplifies to ...</p>
<p><img src="dot_811bfab4e79c84353ff668254695e540.svg" alt="Dot diagram"></p>
</div>
<p></p>
<p>The following algorithm tests a <a href="#BM_(tree)%2Fi">tree</a> to see if it meets the requirements of being a <a href="#BM_(simple%20tree)%2Fi">simple tree</a>...</p>
<p><a href="ch7_code/src/phylogeny/TreeToSimpleTree.py">ch7_code/src/phylogeny/TreeToSimpleTree.py</a> (lines 36 to 83):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">is_tree</span>(<span class="hljs-params">g: Graph[N, ND, E, <span class="hljs-built_in">float</span>]</span>) -&gt; <span class="hljs-built_in">bool</span>:
    <span class="hljs-comment"># Check for cycles</span>
    <span class="hljs-keyword">if</span> <span class="hljs-built_in">len</span>(g) == <span class="hljs-number">0</span>:
        <span class="hljs-keyword">return</span> <span class="hljs-literal">False</span>
    walked_edges = <span class="hljs-built_in">set</span>()
    walked_nodes = <span class="hljs-built_in">set</span>()
    queued_edges = <span class="hljs-built_in">set</span>()
    start_n = <span class="hljs-built_in">next</span>(g.get_nodes())
    <span class="hljs-keyword">for</span> e <span class="hljs-keyword">in</span> g.get_outputs(start_n):
        queued_edges.add((start_n, e))
    <span class="hljs-keyword">while</span> <span class="hljs-built_in">len</span>(queued_edges) &gt; <span class="hljs-number">0</span>:
        ignore_n, e = queued_edges.pop()
        active_n = [n <span class="hljs-keyword">for</span> n <span class="hljs-keyword">in</span> g.get_edge_ends(e) <span class="hljs-keyword">if</span> n != ignore_n][<span class="hljs-number">0</span>]
        walked_edges.add(e)
        walked_nodes.update({ignore_n, active_n})
        children = <span class="hljs-built_in">set</span>(g.get_outputs(active_n))
        children.remove(e)
        <span class="hljs-keyword">for</span> child_e <span class="hljs-keyword">in</span> children:
            <span class="hljs-keyword">if</span> child_e <span class="hljs-keyword">in</span> walked_edges:
                <span class="hljs-keyword">return</span> <span class="hljs-literal">False</span>  <span class="hljs-comment"># cyclic -- edge already walked</span>
            child_ignore_n = active_n
            queued_edges.add((child_ignore_n, child_e))
    <span class="hljs-comment"># Check for disconnected graph</span>
    <span class="hljs-keyword">if</span> <span class="hljs-built_in">len</span>(walked_nodes) != <span class="hljs-built_in">len</span>(g):
        <span class="hljs-keyword">return</span> <span class="hljs-literal">False</span>  <span class="hljs-comment"># disconnected -- some nodes not reachable</span>
    <span class="hljs-keyword">return</span> <span class="hljs-literal">True</span>


<span class="hljs-keyword">def</span> <span class="hljs-title function_">is_simple_tree</span>(<span class="hljs-params">g: Graph[N, ND, E, <span class="hljs-built_in">float</span>]</span>) -&gt; <span class="hljs-built_in">bool</span>:
    <span class="hljs-comment"># Check if tree</span>
    <span class="hljs-keyword">if</span> <span class="hljs-keyword">not</span> is_tree(g):
        <span class="hljs-keyword">return</span> <span class="hljs-literal">False</span>
    <span class="hljs-comment"># Test degrees</span>
    <span class="hljs-keyword">for</span> n <span class="hljs-keyword">in</span> g.get_nodes():
        <span class="hljs-comment"># Degree == 0 shouldn't exist if tree</span>
        <span class="hljs-comment"># Degree == 1 is leaf node</span>
        <span class="hljs-comment"># Degree == 2 is a non-splitting internal node (NOT ALLOWED)</span>
        <span class="hljs-comment"># Degree &gt;= 3 is splitting internal node</span>
        degree = g.get_degree(n)
        <span class="hljs-keyword">if</span> degree == <span class="hljs-number">2</span>:
            <span class="hljs-keyword">return</span> <span class="hljs-literal">False</span>
    <span class="hljs-comment"># Test weights</span>
    <span class="hljs-keyword">for</span> e <span class="hljs-keyword">in</span> g.get_edges():
        <span class="hljs-comment"># No non-positive weights</span>
        weight = g.get_edge_data(e)
        <span class="hljs-keyword">if</span> weight &lt;= <span class="hljs-number">0</span>:
            <span class="hljs-keyword">return</span> <span class="hljs-literal">False</span>
    <span class="hljs-keyword">return</span> <span class="hljs-literal">True</span>
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>The tree...</p>
<p><img src="dot_a6073dfa575690e209a573018def1057.svg" alt="Dot diagram"></p>
<p>... is NOT a simple tree</p>
</div>
<p></p>
<a name="H_Additive%20Distance%20Matrix%20Cardinality"></a><h3>Additive Distance Matrix Cardinality</h3>
<p><a name="BM_(Algorithms%5C%2FPhylogeny%2FAdditive%20Distance%20Matrix%20Cardinality)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FTree%20to%20Simple%20Tree)_TOPIC%2F">Algorithms/Phylogeny/Tree to Simple Tree</a></li>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FTree%20to%20Additive%20Distance%20Matrix)_TOPIC%2F">Algorithms/Phylogeny/Tree to Additive Distance Matrix</a></li>
</ul>
</div>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>This was discussed briefly in the <a href="#BM_(simple%20tree)%2Fi">simple tree</a> section, but it's being discussed here in its own section because it's important.</p>
</div>
<p><strong>WHAT</strong>: Determine the cardinality of between an <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> and a type of <a href="#BM_(tree)%2Fi">tree</a>. For, ...</p>
<ul>
<li><a href="#BM_(simple%20tree)%2Fi">simple tree</a>s, there's a 1-to-1 mapping for <a href="#BM_(tree)%2Fi">tree</a> to / from <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>. That is, a unique <a href="#BM_(simple%20tree)%2Fi">simple tree</a> will map to a unique additive distance and vice-versa.</li>
<li>non-<a href="#BM_(simple%20tree)%2Fi">simple tree</a>s, there's a 1-to-many mapping for <a href="#BM_(tree)%2Fi">tree</a> to / from <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>. That is, a non-<a href="#BM_(simple%20tree)%2Fi">simple tree</a> will only ever generate one <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> but that <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> will map to many non-<a href="#BM_(simple%20tree)%2Fi">simple tree</a>s.</li>
</ul>
<p><strong>WHY</strong>: Non-<a href="#BM_(simple%20tree)%2Fi">simple tree</a>s are essentially derived from <a href="#BM_(simple%20tree)%2Fi">simple tree</a>s by splicing nodes in between edges (breaking up an edge into multiple edges). For example, any of the following non-<a href="#BM_(simple%20tree)%2Fi">simple tree</a>s...</p>
<p><img src="svgbob_c19f51e04775f2f929fb819f2fd231788887981a.svg" alt="Kroki diagram output"></p>
<p>... will collapse to the following <a href="#BM_(simple%20tree)%2Fi">simple tree</a> (edges <a href="#BM_(connected)%2Fi">connected</a> by nodes of <a href="#BM_(degree)_GRAPH%2Fi">degree</a> 2 merged by adding weights) ...</p>
<p><img src="svgbob_34ccd200d7f22f62100f4341c9d3d4550b63bda7.svg" alt="Kroki diagram output"></p>
<p>All of the <a href="#BM_(tree)%2Fi">tree</a>s above, both the non-<a href="#BM_(simple%20tree)%2Fi">simple tree</a>s and the <a href="#BM_(simple%20tree)%2Fi">simple tree</a>, will generate the following <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> ...</p>
<table>
<thead>
<tr>
<th></th>
<th>Cat</th>
<th>Lion</th>
<th>Bear</th>
</tr>
</thead>
<tbody>
<tr>
<td>Cat</td>
<td>0</td>
<td>2</td>
<td>4</td>
</tr>
<tr>
<td>Lion</td>
<td>2</td>
<td>0</td>
<td>3</td>
</tr>
<tr>
<td>Bear</td>
<td>4</td>
<td>3</td>
<td>0</td>
</tr>
</tbody>
</table>
<p>Similarly, this <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> will only ever map to the <a href="#BM_(simple%20tree)%2Fi">simple tree</a> shown above or one of its many non-<a href="#BM_(simple%20tree)%2Fi">simple tree</a> derivatives (3 of which are shown above). There is no other <a href="#BM_(simple%20tree)%2Fi">simple tree</a> that this <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> can map to / no other <a href="#BM_(simple%20tree)%2Fi">simple tree</a> that will generate this <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a>. In other words, it isn't possible for...</p>
<ul>
<li>two different <a href="#BM_(simple%20tree)%2Fi">simple tree</a>s to map to the same <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a>.</li>
<li>two different <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrices</a> to map to the same <a href="#BM_(simple%20tree)%2Fi">simple tree</a>.</li>
</ul>
<p>Working backwards from a <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> to a <a href="#BM_(tree)%2Fi">tree</a> is less complex when limiting the <a href="#BM_(tree)%2Fi">tree</a> to a <a href="#BM_(simple%20tree)%2Fi">simple tree</a>, because there's only one <a href="#BM_(simple%20tree)%2Fi">simple tree</a> possible (vs many non-<a href="#BM_(simple%20tree)%2Fi">simple tree</a>s).</p>
<p><strong>ALGORITHM</strong>:</p>
<p>This section is more of a concept than an algorithm. The following just generates an <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> from a <a href="#BM_(tree)%2Fi">tree</a> and says if that <a href="#BM_(tree)%2Fi">tree</a> is unique to that <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> (it should be if it's a <a href="#BM_(simple%20tree)%2Fi">simple tree</a>). There is no code to show for it because it's just calling things from previous sections (generating an <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> and checking if a <a href="#BM_(simple%20tree)%2Fi">simple tree</a>).</p>
<p><a href="ch7_code/src/phylogeny/CardinalityTest.py">ch7_code/src/phylogeny/CardinalityTest.py</a> (lines 15 to 19):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">cardinality_test</span>(<span class="hljs-params">g: Graph[N, ND, E, <span class="hljs-built_in">float</span>]</span>) -&gt; <span class="hljs-built_in">tuple</span>[DistanceMatrix[N], <span class="hljs-built_in">bool</span>]:
    <span class="hljs-keyword">return</span> (
        to_additive_distance_matrix(g),
        is_simple_tree(g)
    )
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>The tree...</p>
<p><img src="dot_aa1feda6737daab4bc422403da61d84e.svg" alt="Dot diagram"></p>
<p>... produces the additive distance matrix ...</p>
<table>
<thead><tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr></thead>
<tbody>
<tr>
<td>v0</td>
<td>0.0</td>
<td>13.0</td>
<td>21.0</td>
<td>21.0</td>
<td>22.0</td>
<td>22.0</td>
</tr>
<tr>
<td>v1</td>
<td>13.0</td>
<td>0.0</td>
<td>12.0</td>
<td>12.0</td>
<td>13.0</td>
<td>13.0</td>
</tr>
<tr>
<td>v2</td>
<td>21.0</td>
<td>12.0</td>
<td>0.0</td>
<td>20.0</td>
<td>21.0</td>
<td>21.0</td>
</tr>
<tr>
<td>v3</td>
<td>21.0</td>
<td>12.0</td>
<td>20.0</td>
<td>0.0</td>
<td>7.0</td>
<td>13.0</td>
</tr>
<tr>
<td>v4</td>
<td>22.0</td>
<td>13.0</td>
<td>21.0</td>
<td>7.0</td>
<td>0.0</td>
<td>14.0</td>
</tr>
<tr>
<td>v5</td>
<td>22.0</td>
<td>13.0</td>
<td>21.0</td>
<td>13.0</td>
<td>14.0</td>
<td>0.0</td>
</tr>
</tbody>
</table>
<p>The tree is simple. This is the ONLY simple tree possible for this additive distance matrix and vice-versa.</p>
</div>
<p></p>
<a name="H_Test%20Additive%20Distance%20Matrix"></a><h3>Test Additive Distance Matrix</h3>
<p><a name="BM_(Algorithms%5C%2FPhylogeny%2FTest%20Additive%20Distance%20Matrix)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FTree%20to%20Additive%20Distance%20Matrix)_TOPIC%2F">Algorithms/Phylogeny/Tree to Additive Distance Matrix</a></li>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FTree%20to%20Simple%20Tree)_TOPIC%2F">Algorithms/Phylogeny/Tree to Simple Tree</a></li>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FAdditive%20Distance%20Matrix%20Cardinality)_TOPIC%2F">Algorithms/Phylogeny/Additive Distance Matrix Cardinality</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: Determine if a <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> is an <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>.</p>
<p><strong>WHY</strong>: Knowing if a <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> is additive helps determine how the <a href="#BM_(tree)%2Fi">tree</a> for that <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> should be constructed. For example, since it's impossible for a non-<a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> to fit a <a href="#BM_(tree)%2Fi">tree</a>, different algorithms are needed to approximate a <a href="#BM_(tree)%2Fi">tree</a> that somewhat fits.</p>
<p><strong>ALGORITHM</strong>:</p>
<p>This algorithm, called the <a href="#BM_(four%20points%3F%20condition%7Cfour%20point%20theorem)%2Fi">four point condition</a> algorithm, tests pairs within each quartet of leaf nodes to ensure that they meet a certain set of conditions. For example, the following <a href="#BM_(tree)%2Fi">tree</a> has the quartet of leaf nodes (v0, v2, v4, v6) ...</p>
<p><img src="dot_d1edbcb3745180a7312db2a6e743e4fe.svg" alt="Dot diagram"></p>
<p>A quartet makes up 3 different pair combinations (pairs of pairs). For example, the example quartet above has the 3 pair combinations ...</p>
<ul>
<li>((v0, v2), (v4, v6))</li>
<li>((v0, v4), (v2, v6))</li>
<li>((v0, v6), (v2, v4))</li>
</ul>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>Order of the pairing doesn't matter at either level. For example, ((v0, v2), (v4, v6)) and ((v6, v4), (v2, v0)) are the same. That's why there are only 3.</p>
</div>
<p>Of these 3 pair combinations, the test checks to see that ...</p>
<ol>
<li>the sum of distances for one is == the sum of distances for another.</li>
<li>the sum of distances for the remaining is &lt;= the sums from the point above.</li>
</ol>
<p>In a <a href="#BM_(tree)%2Fi">tree</a> with edge weights &gt;= 0, every leaf node quartet will pass this test. For example, for leaf node quartet (v0, v2, v4, v6) highlighted in the example <a href="#BM_(tree)%2Fi">tree</a> above ...</p>
<p><img src="dot_d31d36a6fb4e1c4996a83fe60cfef729.svg" alt="Dot diagram"></p>
<p><span style="background-color: violet; color: black; padding: 3px">dist(v0,v2)</span> + <span style="background-color: tan; color: black; padding: 3px">dist(v4,v6)</span> &lt;= <span style="background-color: gold; color: black; padding: 3px">dist(v0,v6)</span> + <span style="background-color: pink; color: black; padding: 3px">dist(v2,v4)</span> == <span style="background-color: turquoise; color: black; padding: 3px">dist(v0,v4)</span> + <span style="background-color: orange; color: black; padding: 3px">dist(v2,v6)</span></p>
<p>Note how the same set of edges are highlighted between the first two diagrams (same distance contributions) while the third diagram has less edges highlighted (missing some distance contributions). This is where the inequality comes from.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>I'm almost certain this inequality should be &lt; instead of &lt;=, because in a <a href="#BM_(phylogenetic%20tree%7Cphylogeny%20tree%7Cphylogenies%7Cevolutionary%20tree)%2Fi">phylogenetic tree</a> you can't have an edge weight of 0, right? An edge weight of 0 would indicate that the nodes at each end of an edge are the same entity.</p>
</div>
<p>All of the information required for the above calculation is available in the <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a>...</p>
<p><a href="ch7_code/src/phylogeny/FourPointCondition.py">ch7_code/src/phylogeny/FourPointCondition.py</a> (lines 21 to 47):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">four_point_test</span>(<span class="hljs-params">dm: DistanceMatrix[N], l0: N, l1: N, l2: N, l3: N</span>) -&gt; <span class="hljs-built_in">bool</span>:
    <span class="hljs-comment"># Pairs of leaf node pairs</span>
    pair_combos = (
        ((l0, l1), (l2, l3)),
        ((l0, l2), (l1, l3)),
        ((l0, l3), (l1, l2))
    )
    <span class="hljs-comment"># Different orders to test pair_combos to see if they match conditions</span>
    test_orders = (
        (<span class="hljs-number">0</span>, <span class="hljs-number">1</span>, <span class="hljs-number">2</span>),
        (<span class="hljs-number">0</span>, <span class="hljs-number">2</span>, <span class="hljs-number">1</span>),
        (<span class="hljs-number">1</span>, <span class="hljs-number">0</span>, <span class="hljs-number">2</span>),
        (<span class="hljs-number">1</span>, <span class="hljs-number">2</span>, <span class="hljs-number">0</span>),
        (<span class="hljs-number">2</span>, <span class="hljs-number">0</span>, <span class="hljs-number">1</span>),
        (<span class="hljs-number">2</span>, <span class="hljs-number">1</span>, <span class="hljs-number">0</span>)
    )
    <span class="hljs-comment"># Find at least one order of pair combos that passes the test</span>
    <span class="hljs-keyword">for</span> p1_idx, p2_idx, p3_idx <span class="hljs-keyword">in</span> test_orders:
        p1_1, p1_2 = pair_combos[p1_idx]
        p2_1, p2_2 = pair_combos[p2_idx]
        p3_1, p3_2 = pair_combos[p3_idx]
        s1 = dm[p1_1] + dm[p1_2]
        s2 = dm[p2_1] + dm[p2_2]
        s3 = dm[p3_1] + dm[p3_2]
        <span class="hljs-keyword">if</span> s1 &lt;= s2 == s3:
            <span class="hljs-keyword">return</span> <span class="hljs-literal">True</span>
    <span class="hljs-keyword">return</span> <span class="hljs-literal">False</span>
</code></pre>
<p>If a <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> was derived from a <a href="#BM_(tree)%2Fi">tree</a> / fits a <a href="#BM_(tree)%2Fi">tree</a>, its leaf node quartets will also pass this test. That is, if all leaf node quartets in a <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> pass the above test, the <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> is an <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> ...</p>
<p><a href="ch7_code/src/phylogeny/FourPointCondition.py">ch7_code/src/phylogeny/FourPointCondition.py</a> (lines 52 to 64):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">is_additive</span>(<span class="hljs-params">dm: DistanceMatrix[N]</span>) -&gt; <span class="hljs-built_in">bool</span>:
    <span class="hljs-comment"># Recall that an additive distance matrix of size &lt;= 3 is guaranteed to be an additive distance</span>
    <span class="hljs-comment"># matrix (try it and see -- any distances you use will always end up fitting a tree). Thats why</span>
    <span class="hljs-comment"># you need at least 4 leaf nodes to test.</span>
    <span class="hljs-keyword">if</span> dm.n &lt; <span class="hljs-number">4</span>:
        <span class="hljs-keyword">return</span> <span class="hljs-literal">True</span>
    leaves = dm.leaf_ids()
    <span class="hljs-keyword">for</span> quartet <span class="hljs-keyword">in</span> combinations(leaves, r=<span class="hljs-number">4</span>):
        passed = four_point_test(dm, *quartet)
        <span class="hljs-keyword">if</span> <span class="hljs-keyword">not</span> passed:
            <span class="hljs-keyword">return</span> <span class="hljs-literal">False</span>
    <span class="hljs-keyword">return</span> <span class="hljs-literal">True</span>
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>The distance matrix...</p>
<table>
<thead><tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
</tr></thead>
<tbody>
<tr>
<td>v0</td>
<td>0.0</td>
<td>3.0</td>
<td>8.0</td>
<td>7.0</td>
</tr>
<tr>
<td>v1</td>
<td>3.0</td>
<td>0.0</td>
<td>9.0</td>
<td>8.0</td>
</tr>
<tr>
<td>v2</td>
<td>8.0</td>
<td>9.0</td>
<td>0.0</td>
<td>5.0</td>
</tr>
<tr>
<td>v3</td>
<td>7.0</td>
<td>8.0</td>
<td>5.0</td>
<td>0.0</td>
</tr>
</tbody>
</table>
<p>... is additive.</p>
</div>
<p></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>Could the differences found by this algorithm help determine how "close" a <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> is to being an <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>?</p>
</div>
<a name="H_Find%20Limb%20Length"></a><h3>Find Limb Length</h3>
<p><a name="BM_(Algorithms%5C%2FPhylogeny%2FFind%20Limb%20Length)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FTree%20to%20Additive%20Distance%20Matrix)_TOPIC%2F">Algorithms/Phylogeny/Tree to Additive Distance Matrix</a></li>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FTree%20to%20Simple%20Tree)_TOPIC%2F">Algorithms/Phylogeny/Tree to Simple Tree</a></li>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FAdditive%20Distance%20Matrix%20Cardinality)_TOPIC%2F">Algorithms/Phylogeny/Additive Distance Matrix Cardinality</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: Given an <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>, there exists a unique <a href="#BM_(simple%20tree)%2Fi">simple tree</a> that fits that matrix. Compute the <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a> of any leaf node in that <a href="#BM_(simple%20tree)%2Fi">simple tree</a> just from the <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>.</p>
<p><strong>WHY</strong>: This is one of the operations required to construct the unique <a href="#BM_(simple%20tree)%2Fi">simple tree</a> for an <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>.</p>
<p><strong>ALGORITHM</strong>:</p>
<p>To conceptualize how this algorithm works, consider the following <a href="#BM_(simple%20tree)%2Fi">simple tree</a> and its corresponding <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>...</p>
<p><img src="dot_2ce9085dbbfd0f88d09667f2cefdce21.svg" alt="Dot diagram"></p>
<table>
<thead>
<tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
<th>v6</th>
</tr>
</thead>
<tbody>
<tr>
<td>v0</td>
<td>0</td>
<td>13</td>
<td>19</td>
<td>20</td>
<td>29</td>
<td>40</td>
<td>36</td>
</tr>
<tr>
<td>v1</td>
<td>13</td>
<td>0</td>
<td>10</td>
<td>11</td>
<td>20</td>
<td>31</td>
<td>27</td>
</tr>
<tr>
<td>v2</td>
<td>19</td>
<td>10</td>
<td>0</td>
<td>11</td>
<td>20</td>
<td>31</td>
<td>27</td>
</tr>
<tr>
<td>v3</td>
<td>20</td>
<td>11</td>
<td>11</td>
<td>0</td>
<td>21</td>
<td>32</td>
<td>28</td>
</tr>
<tr>
<td>v4</td>
<td>29</td>
<td>20</td>
<td>20</td>
<td>21</td>
<td>0</td>
<td>17</td>
<td>13</td>
</tr>
<tr>
<td>v5</td>
<td>40</td>
<td>31</td>
<td>31</td>
<td>32</td>
<td>17</td>
<td>0</td>
<td>6</td>
</tr>
<tr>
<td>v6</td>
<td>36</td>
<td>27</td>
<td>27</td>
<td>28</td>
<td>13</td>
<td>6</td>
<td>0</td>
</tr>
</tbody>
</table>
<p>In this <a href="#BM_(simple%20tree)%2Fi">simple tree</a>, consider a path between leaf nodes that travels over v2's parent (v2 itself excluded). For example, <span style="background-color: pink; color: black; padding: 3px">path(v1,v5)</span> travels over v2's parent...</p>
<p><img src="dot_276f3df9c3fc7d23f14a0984964d0686.svg" alt="Dot diagram"></p>
<p>Now, consider the paths between each of the two nodes in the path above (v1 and v5) and v2: <span style="background-color: orange; color: black; padding: 3px">path(v1,v2)</span> + <span style="background-color: blue; color: white; padding: 3px">path(v2,v5)</span> ...</p>
<p><img src="dot_188fe43ede6da8ad09ce44fa84eabf77.svg" alt="Dot diagram"></p>
<p>Notice how the edges highlighted between <span style="background-color: pink; color: black; padding: 3px">path(v1,v5)</span> and <span style="background-color: orange; color: black; padding: 3px">path(v1,v2)</span> + <span style="background-color: blue; color: white; padding: 3px">path(v2,v5)</span> would be the same had it not been for the two highlights on v2's <a href="#BM_(limb)%2Fi">limb</a>. Adding <span style="background-color: red; color: white; padding: 3px">2 * path(v2,i0)</span> to <span style="background-color: pink; color: black; padding: 3px">path(v1,v5)</span> makes it so that each edge is highlighted an equal number of times ...</p>
<p><img src="dot_be3424051e002c75fa44c9de8ac7357c.svg" alt="Dot diagram"></p>
<p><span style="background-color: orange; color: black; padding: 3px">path(v1,v2)</span> + <span style="background-color: blue; color: white; padding: 3px">path(v2,v5)</span> = <span style="background-color: pink; color: black; padding: 3px">path(v1,v5)</span> + <span style="background-color: red; color: white; padding: 3px">2 * path(v2,i1)</span></p>
<p>Contrast the above to what happens when the pair of leaf nodes selected DOESN'T travel through v2's parent. For example, <span style="background-color: purple; color: white; padding: 3px">path(v4,v5)</span> doesn't travel through v2's parent ...</p>
<p><img src="dot_41b02580dfdd626fd77316ad51612077.svg" alt="Dot diagram"></p>
<p><span style="background-color: brown; color: white; padding: 3px">path(v4,v2)</span> + <span style="background-color: blue; color: white; padding: 3px">path(v2,v5)</span> &gt; <span style="background-color: purple; color: white; padding: 3px">path(v4,v5)</span> + <span style="background-color: red; color: white; padding: 3px">2 * path(v2,i1)</span></p>
<p>Even when <span style="background-color: purple; color: white; padding: 3px">path(v4,v5)</span> includes <span style="background-color: red; color: white; padding: 3px">2 * path(v2,i1)</span>, less edges are highlighted when compared to <span style="background-color: brown; color: white; padding: 3px">path(v4,v2)</span> + <span style="background-color: blue; color: white; padding: 3px">path(v2,v5)</span>. Specifically, edge(i1,i2) is highlighted zero times vs two times.</p>
<p>The above two examples give way to the following two formulas: Given a <a href="#BM_(simple%20tree)%2Fi">simple tree</a> with distinct leaf nodes {L, A, B} and L's parent Lp ...</p>
<ul>
<li>path(L,A) + path(L,B) = path(A,B) + 2 * path(L,Lp) -- if path(A,B) travels through Lp</li>
<li>path(L,A) + path(L,B) &gt; path(A,B) + 2 * path(L,Lp) -- if path(A,B) doesn't travel through Lp</li>
</ul>
<p>These two formulas work just as well with distances instead of paths...</p>
<ul>
<li>dist(L,A) + dist(L,B) = dist(A,B) + 2 * dist(L,Lp) -- if path(A,B) travels through Lp</li>
<li>dist(L,A) + dist(L,B) &gt; dist(A,B) + 2 * dist(L,Lp) -- if path(A,B) doesn't travel through Lp</li>
</ul>
<p>The reason distances work has to do with the fact that <a href="#BM_(simple%20tree)%2Fi">simple tree</a>s require edges weights of &gt; 0, meaning traversing over an edge always increases the overall distance. If ...</p>
<ul>
<li>less edges are highlighted, the distance will be less.</li>
<li>same edges are highlighted, the distance will be equal.</li>
<li>more edges are highlighted, the distance will be more.</li>
</ul>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The Pevzner book has the 2nd formula above as &gt;= instead of &gt;.</p>
<p>I'm assuming they did this because they're letting edge weights be &gt;= 0 instead of &gt; 0, which doesn't make sense because an edge with a weight of 0 means the same entity exists on both ends of the edge. If an edge weight is 0, it'll contribute nothing to the distance, meaning that more edges being highlighted doesn't necessarily mean a larger distance.</p>
</div>
<p>In the above formulas, L's <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a> is represented as dist(L,Lp). Except for dist(L,Lp), all distances in the formulas are between leaf nodes and as such are found in the <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a>. Therefore, the formulas need to be isolated to dist(L,Lp) in order to derive what L's <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a> is ...</p>
<ul>
<li>
<p>dist(L,A) + dist(L,B) = dist(A,B) + 2 * dist(L,Lp) -- if path(A,B) travels through Lp</p>
<pre><code>dist(L,A) + dist(L,B) = dist(A,B) + 2 * dist(L,Lp)
dist(L,A) + dist(L,B) - dist(A,B) = 2 * dist(L,Lp)
(dist(L,A) + dist(L,B) - dist(A,B)) / 2 = dist(L,Lp)
</code></pre>
<p>The following is a conceptualization of the isolation of dist(L,Lp) happening above using the initial equality example from above. Notice how, in the end, v2's <a href="#BM_(limb)%2Fi">limb</a> is highlighted exactly once and nothing else.</p>
<p><img src="dot_1389434a06935635f2c2e770e16c4f9b.svg" alt="Dot diagram"></p>
</li>
<li>
<p>dist(L,A) + dist(L,B) &gt; dist(A,B) + 2 * dist(L,Lp) -- if path(A,B) doesn't travel through Lp</p>
<pre><code>dist(L,A) + dist(L,B) &gt; dist(A,B) + 2 * dist(L,Lp)
dist(L,A) + dist(L,B) - dist(A,B) &gt; 2 * dist(L,Lp)
(dist(L,A) + dist(L,B) - dist(A,B)) / 2 &gt; dist(L,Lp)
</code></pre>
<p>The following is a conceptualization of the isolation of dist(L,Lp) happening above using the initial inequality example from above. Notice how, in the end, v2's <a href="#BM_(limb)%2Fi">limb</a> is highlighted exactly once but other edges are also highlighted. That's why it's &gt; instead of =.</p>
<p><img src="dot_52d2cabe2e82cffa90e6bc6eb96c7766.svg" alt="Dot diagram"></p>
</li>
</ul>
<p>Notice the left-hand side of both solved formulas are the same: (dist(L,A) + dist(L,B) - dist(A,B)) / 2</p>
<p>The algorithm for finding <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a> is essentially an exhaustive test. Of all leaf node pairs (L not included), the one producing the smallest left-hand side result is guaranteed to be L's <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a>. Anything larger will include weights from more edges than just L's <a href="#BM_(limb)%2Fi">limb</a>.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>From the book:</p>
<blockquote>
<p>Exercise Break: The algorithm proposed on the previous step computes <a href="#BM_(limb)%2Fi">Limb</a>Length(j) in O(n2) time (for an n x n <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a>). Design an algorithm that computes <a href="#BM_(limb)%2Fi">Limb</a>Length(j) in O(n) time.</p>
</blockquote>
<p>The answer to this is obvious now that I've gone through and reasoned about things above.</p>
<p>For the <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a> formula to work, you need to find leaf nodes (A, B) whose path travels through leaf node L's parent (Lp). Originally, the book had you try all combinations of leaf nodes (L excluded) and take the minimum. That works, but you don't need to try all possible pairs. Instead, you can just pick any leaf (that isn't L) for A and test against every other node (that isn't L) to find B -- as with the original method, you pick the B that produces the minimum value.</p>
<p>Because a <a href="#BM_(phylogenetic%20tree%7Cphylogeny%20tree%7Cphylogenies%7Cevolutionary%20tree)%2Fi">phylogenetic tree</a> is a <a href="#BM_(connected)%2Fi">connected</a> graph (a path exists between each node and all other nodes), at least 1 path will exist starting from A that travels through Lp.</p>
<pre class="hljs"><code>leaf_nodes.remove(L)  <span class="hljs-comment"># remove L from the set</span>
A = leaf_nodes.pop()  <span class="hljs-comment"># removes and returns an arbitrary leaf node</span>
B = <span class="hljs-built_in">min</span>(leafs, key=<span class="hljs-keyword">lambda</span> x: (dist(L, A) + dist(L, x) - dist(A, x)) / <span class="hljs-number">2</span>)
</code></pre>
<p>For example, imagine that you're trying to find v2's <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a> in the following graph...</p>
<p><img src="dot_2ce9085dbbfd0f88d09667f2cefdce21.svg" alt="Dot diagram"></p>
<p>Pick v4 as your A node, then try the formula with every other leaf node as B (except v2 because that's the node you're trying to get <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a> for + v4 because that's your A node). At least one of path(A, B)'s will cross through v2's parent. Take the minimum, just as you did when you were trying every possible node pair across all leaf nodes in the graph.</p>
</div>
<p><a href="ch7_code/src/phylogeny/FindLimbLength.py">ch7_code/src/phylogeny/FindLimbLength.py</a> (lines 22 to 28):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">find_limb_length</span>(<span class="hljs-params">dm: DistanceMatrix[N], l: N</span>) -&gt; <span class="hljs-built_in">float</span>:
    leaf_nodes = dm.leaf_ids()
    leaf_nodes.remove(l)
    a = leaf_nodes.pop()
    b = <span class="hljs-built_in">min</span>(leaf_nodes, key=<span class="hljs-keyword">lambda</span> x: (dm[l, a] + dm[l, x] - dm[a, x]) / <span class="hljs-number">2</span>)
    <span class="hljs-keyword">return</span> (dm[l, a] + dm[l, b] - dm[a, b]) / <span class="hljs-number">2</span>
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the additive distance matrix...</p>
<table>
<thead><tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
<th>v6</th>
</tr></thead>
<tbody>
<tr>
<td>v0</td>
<td>0.0</td>
<td>13.0</td>
<td>19.0</td>
<td>20.0</td>
<td>29.0</td>
<td>40.0</td>
<td>36.0</td>
</tr>
<tr>
<td>v1</td>
<td>13.0</td>
<td>0.0</td>
<td>10.0</td>
<td>11.0</td>
<td>20.0</td>
<td>31.0</td>
<td>27.0</td>
</tr>
<tr>
<td>v2</td>
<td>19.0</td>
<td>10.0</td>
<td>0.0</td>
<td>11.0</td>
<td>20.0</td>
<td>31.0</td>
<td>27.0</td>
</tr>
<tr>
<td>v3</td>
<td>20.0</td>
<td>11.0</td>
<td>11.0</td>
<td>0.0</td>
<td>21.0</td>
<td>32.0</td>
<td>28.0</td>
</tr>
<tr>
<td>v4</td>
<td>29.0</td>
<td>20.0</td>
<td>20.0</td>
<td>21.0</td>
<td>0.0</td>
<td>17.0</td>
<td>13.0</td>
</tr>
<tr>
<td>v5</td>
<td>40.0</td>
<td>31.0</td>
<td>31.0</td>
<td>32.0</td>
<td>17.0</td>
<td>0.0</td>
<td>6.0</td>
</tr>
<tr>
<td>v6</td>
<td>36.0</td>
<td>27.0</td>
<td>27.0</td>
<td>28.0</td>
<td>13.0</td>
<td>6.0</td>
<td>0.0</td>
</tr>
</tbody>
</table>
<p>The limb for leaf node v2 in its unique simple tree has a weight of 5.0</p>
</div>
<p></p>
<a name="H_Test%20Same%20Subtree"></a><h3>Test Same Subtree</h3>
<p><a name="BM_(Algorithms%5C%2FPhylogeny%2FTest%20Same%20Subtree)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FFind%20Limb%20Length)_TOPIC%2F">Algorithms/Phylogeny/Find Limb Length</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: Splitting a <a href="#BM_(simple%20tree)%2Fi">simple tree</a> on the parent of one of its leaf nodes breaks it up into several <a href="#BM_(subtree%7Csub-tree)%2Fi">subtree</a>s. For example, the following <a href="#BM_(simple%20tree)%2Fi">simple tree</a> has been split on v2's parent, resulting in 4 different <a href="#BM_(subtree%7Csub-tree)%2Fi">subtree</a>s ...</p>
<p><img src="dot_87753d86b95006a2fb6c20e3959bb261.svg" alt="Dot diagram"></p>
<ul>
<li><span style="background-color: green; color: white; padding: 3px">v0, v1, i0</span></li>
<li><span style="background-color: yellow; color: black; padding: 3px">v2</span></li>
<li><span style="background-color: pink; color: black; padding: 3px">v3</span></li>
<li><span style="background-color: cyan; color: black; padding: 3px">v4, i2, i3, v5, v6</span></li>
</ul>
<p>Given just the <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> for a <a href="#BM_(simple%20tree)%2Fi">simple tree</a> (not the <a href="#BM_(simple%20tree)%2Fi">simple tree</a> itself), determine if two <em>leaf nodes</em> belong to the same <a href="#BM_(subtree%7Csub-tree)%2Fi">subtree</a> had that <a href="#BM_(simple%20tree)%2Fi">simple tree</a> been split on some leaf node's parent.</p>
<p><strong>WHY</strong>: This is one of the operations required to construct the unique <a href="#BM_(simple%20tree)%2Fi">simple tree</a> for an <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>.</p>
<p><strong>ALGORITHM</strong>:</p>
<p>The algorithm is essentially the formulas from the <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a> algorithm. Recall that those formulas are ...</p>
<ul>
<li>
<p>dist(L,A) + dist(L,B) = dist(A,B) + 2 * dist(L,Lp) -- if path(A,B) travels through Lp</p>
<pre><code>dist(L,A) + dist(L,B) = dist(A,B) + 2 * dist(L,Lp)
dist(L,A) + dist(L,B) - dist(A,B) = 2 * dist(L,Lp)
(dist(L,A) + dist(L,B) - dist(A,B)) / 2 = dist(L,Lp)
</code></pre>
</li>
<li>
<p>dist(L,A) + dist(L,B) &gt; dist(A,B) + 2 * dist(L,Lp) -- if path(A,B) doesn't travel through Lp</p>
<pre><code>dist(L,A) + dist(L,B) &gt; dist(A,B) + 2 * dist(L,Lp)
dist(L,A) + dist(L,B) - dist(A,B) &gt; 2 * dist(L,Lp)
(dist(L,A) + dist(L,B) - dist(A,B)) / 2 &gt; dist(L,Lp)
</code></pre>
</li>
</ul>
<p>To conceptualize how this algorithm works, consider the following <a href="#BM_(simple%20tree)%2Fi">simple tree</a> and its corresponding <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>...</p>
<p><img src="dot_2ce9085dbbfd0f88d09667f2cefdce21.svg" alt="Dot diagram"></p>
<table>
<thead>
<tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
<th>v6</th>
</tr>
</thead>
<tbody>
<tr>
<td>v0</td>
<td>0</td>
<td>13</td>
<td>19</td>
<td>20</td>
<td>29</td>
<td>40</td>
<td>36</td>
</tr>
<tr>
<td>v1</td>
<td>13</td>
<td>0</td>
<td>10</td>
<td>11</td>
<td>20</td>
<td>31</td>
<td>27</td>
</tr>
<tr>
<td>v2</td>
<td>19</td>
<td>10</td>
<td>0</td>
<td>11</td>
<td>20</td>
<td>31</td>
<td>27</td>
</tr>
<tr>
<td>v3</td>
<td>20</td>
<td>11</td>
<td>11</td>
<td>0</td>
<td>21</td>
<td>32</td>
<td>28</td>
</tr>
<tr>
<td>v4</td>
<td>29</td>
<td>20</td>
<td>20</td>
<td>21</td>
<td>0</td>
<td>17</td>
<td>13</td>
</tr>
<tr>
<td>v5</td>
<td>40</td>
<td>31</td>
<td>31</td>
<td>32</td>
<td>17</td>
<td>0</td>
<td>6</td>
</tr>
<tr>
<td>v6</td>
<td>36</td>
<td>27</td>
<td>27</td>
<td>28</td>
<td>13</td>
<td>6</td>
<td>0</td>
</tr>
</tbody>
</table>
<p>Consider what happens when you break the edges on v2's parent (i1). The <a href="#BM_(tree)%2Fi">tree</a> breaks into 4 distinct <a href="#BM_(subtree%7Csub-tree)%2Fi">subtree</a>s (colored below as green, yellow, pink, and cyan)...</p>
<p><img src="dot_87753d86b95006a2fb6c20e3959bb261.svg" alt="Dot diagram"></p>
<p>If the two leaf nodes chosen are ...</p>
<ul>
<li>
<p>within the same <a href="#BM_(subtree%7Csub-tree)%2Fi">subtree</a>, the path will <em>never</em> travel through v2's parent (i1), meaning that the second formula evaluates to true. For example, since <span style="background-color: cyan; color: black; padding: 3px">v4</span> and <span style="background-color: cyan; color: black; padding: 3px">v5</span> are within the same subset, <span style="background-color: purple; color: white; padding: 3px">path(v4,v5)</span> doesn't travel through v2's parent ...</p>
<p><img src="dot_e6f808aaac07dac8adda771710b9569c.svg" alt="Dot diagram"></p>
<p><span style="background-color: blue; color: white; padding: 3px">dist(v2,v4)</span> + <span style="background-color: orange; color: black; padding: 3px">dist(v2,v5)</span> &gt; <span style="background-color: purple; color: white; padding: 3px">dist(v4,v5)</span> + <span style="background-color: red; color: white; padding: 3px">2 * dist(v2,i1)</span></p>
</li>
<li>
<p>not within the same <a href="#BM_(subtree%7Csub-tree)%2Fi">subtree</a>, the path will <em>always</em> travel through v2's parent (i1), meaning that the first formula evaluates to true. For example, since <span style="background-color: green; color: white; padding: 3px">v1</span> and <span style="background-color: cyan; color: black; padding: 3px">v5</span> are within different subsets, <span style="background-color: pink; color: black; padding: 3px">path(v1,v5)</span> doesn't travel through v2's parent ...</p>
<p><img src="dot_9ccca274456c5e920c6e4db0a94c7266.svg" alt="Dot diagram"></p>
<p><span style="background-color: orange; color: black; padding: 3px">path(v1,v2)</span> + <span style="background-color: blue; color: white; padding: 3px">path(v2,v5)</span> = <span style="background-color: pink; color: black; padding: 3px">path(v1,v5)</span> + <span style="background-color: red; color: white; padding: 3px">2 * path(v2,i1)</span></p>
</li>
</ul>
<p><a href="ch7_code/src/phylogeny/SubtreeDetect.py">ch7_code/src/phylogeny/SubtreeDetect.py</a> (lines 23 to 32):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">is_same_subtree</span>(<span class="hljs-params">dm: DistanceMatrix[N], l: N, a: N, b: N</span>) -&gt; <span class="hljs-built_in">bool</span>:
    l_weight = find_limb_length(dm, l)
    test_res = (dm[l, a] + dm[l, b] - dm[a, b]) / <span class="hljs-number">2</span>
    <span class="hljs-keyword">if</span> test_res == l_weight:
        <span class="hljs-keyword">return</span> <span class="hljs-literal">False</span>
    <span class="hljs-keyword">elif</span> test_res &gt; l_weight:
        <span class="hljs-keyword">return</span> <span class="hljs-literal">True</span>
    <span class="hljs-keyword">else</span>:
        <span class="hljs-keyword">raise</span> ValueError(<span class="hljs-string">'???'</span>)  <span class="hljs-comment"># not additive distance matrix?</span>
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the additive distance matrix...</p>
<table>
<thead><tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
<th>v6</th>
</tr></thead>
<tbody>
<tr>
<td>v0</td>
<td>0.0</td>
<td>13.0</td>
<td>19.0</td>
<td>20.0</td>
<td>29.0</td>
<td>40.0</td>
<td>36.0</td>
</tr>
<tr>
<td>v1</td>
<td>13.0</td>
<td>0.0</td>
<td>10.0</td>
<td>11.0</td>
<td>20.0</td>
<td>31.0</td>
<td>27.0</td>
</tr>
<tr>
<td>v2</td>
<td>19.0</td>
<td>10.0</td>
<td>0.0</td>
<td>11.0</td>
<td>20.0</td>
<td>31.0</td>
<td>27.0</td>
</tr>
<tr>
<td>v3</td>
<td>20.0</td>
<td>11.0</td>
<td>11.0</td>
<td>0.0</td>
<td>21.0</td>
<td>32.0</td>
<td>28.0</td>
</tr>
<tr>
<td>v4</td>
<td>29.0</td>
<td>20.0</td>
<td>20.0</td>
<td>21.0</td>
<td>0.0</td>
<td>17.0</td>
<td>13.0</td>
</tr>
<tr>
<td>v5</td>
<td>40.0</td>
<td>31.0</td>
<td>31.0</td>
<td>32.0</td>
<td>17.0</td>
<td>0.0</td>
<td>6.0</td>
</tr>
<tr>
<td>v6</td>
<td>36.0</td>
<td>27.0</td>
<td>27.0</td>
<td>28.0</td>
<td>13.0</td>
<td>6.0</td>
<td>0.0</td>
</tr>
</tbody>
</table>
<p>Had the tree been split on leaf node v2's parent, leaf nodes v1 and v5 would reside in
different subtrees.</p>
</div>
<p></p>
<a name="H_Trim"></a><h3>Trim</h3>
<p><a name="BM_(Algorithms%5C%2FPhylogeny%2FTrim)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FTree%20to%20Additive%20Distance%20Matrix)_TOPIC%2F">Algorithms/Phylogeny/Tree to Additive Distance Matrix</a></li>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FTree%20to%20Simple%20Tree)_TOPIC%2F">Algorithms/Phylogeny/Tree to Simple Tree</a></li>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FAdditive%20Distance%20Matrix%20Cardinality)_TOPIC%2F">Algorithms/Phylogeny/Additive Distance Matrix Cardinality</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: Remove a <a href="#BM_(limb)%2Fi">limb</a> from an <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>, just as it would get removed from its corresponding unique <a href="#BM_(simple%20tree)%2Fi">simple tree</a>.</p>
<p><strong>WHY</strong>: This is one of the operations required to construct the unique <a href="#BM_(simple%20tree)%2Fi">simple tree</a> for an <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>.</p>
<p><strong>ALGORITHM</strong>:</p>
<p>Recall that for any <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>, there exists a unique <a href="#BM_(simple%20tree)%2Fi">simple tree</a> that fits that matrix. For example, the following <a href="#BM_(simple%20tree)%2Fi">simple tree</a> is unique to the following <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a>...</p>
<p><img src="dot_4e5daffbacf71191e85a5881e8367790.svg" alt="Dot diagram"></p>
<table>
<thead>
<tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
</tr>
</thead>
<tbody>
<tr>
<td>v0</td>
<td>0</td>
<td>13</td>
<td>21</td>
<td>22</td>
</tr>
<tr>
<td>v1</td>
<td>13</td>
<td>0</td>
<td>12</td>
<td>13</td>
</tr>
<tr>
<td>v2</td>
<td>21</td>
<td>12</td>
<td>0</td>
<td>13</td>
</tr>
<tr>
<td>v3</td>
<td>22</td>
<td>13</td>
<td>13</td>
<td>0</td>
</tr>
</tbody>
</table>
<p>Trimming v2 off that <a href="#BM_(simple%20tree)%2Fi">simple tree</a> would result in ...</p>
<p><img src="dot_8970363f8d351a28a22328fbfd029f13.svg" alt="Dot diagram"></p>
<table>
<thead>
<tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v3</th>
</tr>
</thead>
<tbody>
<tr>
<td>v0</td>
<td>0</td>
<td>13</td>
<td>22</td>
</tr>
<tr>
<td>v1</td>
<td>13</td>
<td>0</td>
<td>13</td>
</tr>
<tr>
<td>v3</td>
<td>22</td>
<td>13</td>
<td>0</td>
</tr>
</tbody>
</table>
<p>Notice how when v2 gets trimmed off, the ...</p>
<ul>
<li><a href="#BM_(simple%20tree)%2Fi">simple tree</a> merges path(i0, v2) into a single edge. <a href="#BM_(simple%20tree)%2Fi">Simple tree</a>s can't have nodes with <a href="#BM_(degree)_GRAPH%2Fi">degree</a> 2 (train of non-branching edges not allowed).</li>
<li><a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> row and column for v3 disappear. All other leaf nodes remain with the same distances.</li>
</ul>
<p>As such, removing the row and column for some leaf node in an <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> is equivalent to removing its <a href="#BM_(limb)%2Fi">limb</a> from the corresponding unique <a href="#BM_(simple%20tree)%2Fi">simple tree</a> then merging together any edges <a href="#BM_(connected)%2Fi">connected</a> by nodes of <a href="#BM_(degree)_GRAPH%2Fi">degree</a> 2.</p>
<p><a href="ch7_code/src/phylogeny/Trimmer.py">ch7_code/src/phylogeny/Trimmer.py</a> (lines 26 to 37):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">trim_distance_matrix</span>(<span class="hljs-params">dm: DistanceMatrix[N], leaf: N</span>) -&gt; <span class="hljs-literal">None</span>:
    dm.delete(leaf)  <span class="hljs-comment"># remove row+col for leaf</span>


<span class="hljs-keyword">def</span> <span class="hljs-title function_">trim_tree</span>(<span class="hljs-params">tree: Graph[N, ND, E, <span class="hljs-built_in">float</span>], leaf: N</span>) -&gt; <span class="hljs-literal">None</span>:
    <span class="hljs-keyword">if</span> tree.get_degree(leaf) != <span class="hljs-number">1</span>:
        <span class="hljs-keyword">raise</span> ValueError(<span class="hljs-string">'Not a leaf node'</span>)
    edge = <span class="hljs-built_in">next</span>(tree.get_outputs(leaf))
    tree.delete_edge(edge)
    tree.delete_node(leaf)
    merge_nodes_of_degree2(tree)  <span class="hljs-comment"># make sure its a simple tree</span>
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the additive distance matrix...</p>
<table>
<thead><tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
</tr></thead>
<tbody>
<tr>
<td>v0</td>
<td>0.0</td>
<td>13.0</td>
<td>21.0</td>
<td>22.0</td>
</tr>
<tr>
<td>v1</td>
<td>13.0</td>
<td>0.0</td>
<td>12.0</td>
<td>13.0</td>
</tr>
<tr>
<td>v2</td>
<td>21.0</td>
<td>12.0</td>
<td>0.0</td>
<td>13.0</td>
</tr>
<tr>
<td>v3</td>
<td>22.0</td>
<td>13.0</td>
<td>13.0</td>
<td>0.0</td>
</tr>
</tbody>
</table>
<p>... trimming leaf node v2 results in ...</p>
<table>
<thead><tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v3</th>
</tr></thead>
<tbody>
<tr>
<td>v0</td>
<td>0.0</td>
<td>13.0</td>
<td>22.0</td>
</tr>
<tr>
<td>v1</td>
<td>13.0</td>
<td>0.0</td>
<td>13.0</td>
</tr>
<tr>
<td>v3</td>
<td>22.0</td>
<td>13.0</td>
<td>0.0</td>
</tr>
</tbody>
</table>
</div>
<p></p>
<a name="H_Bald"></a><h3>Bald</h3>
<p><a name="BM_(Algorithms%5C%2FPhylogeny%2FBald)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FFind%20Limb%20Length)_TOPIC%2F">Algorithms/Phylogeny/Find Limb Length</a></li>
</ul>
</div>
<p> <!-- Needs to be disabled because v3's is conflicting with 3' -->
 <!-- Needs to be disabled because v5's is conflicting with 5' --></p>
<p><strong>WHAT</strong>: Set a <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a> to 0 in an <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>, just as it would be set to 0 in its corresponding unique <a href="#BM_(simple%20tree)%2Fi">simple tree</a>. Technically, a <a href="#BM_(simple%20tree)%2Fi">simple tree</a> can't have edge weights that are &lt;= 0. This is a special case, typically used as an intermediate operation of some larger algorithm.</p>
<p><strong>WHY</strong>: This is one of the operations required to construct the unique <a href="#BM_(simple%20tree)%2Fi">simple tree</a> for an <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>.</p>
<p><strong>ALGORITHM</strong>:</p>
<p>Recall that for any <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>, there exists a unique <a href="#BM_(simple%20tree)%2Fi">simple tree</a> that fits that matrix. For example, the following <a href="#BM_(simple%20tree)%2Fi">simple tree</a> is unique to the following <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a>...</p>
<p><img src="dot_e304cd9917c7e5b9d72c4e876e260cba.svg" alt="Dot diagram"></p>
<table>
<thead>
<tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr>
</thead>
<tbody>
<tr>
<td>v0</td>
<td>0</td>
<td>13</td>
<td>21</td>
<td>21</td>
<td>22</td>
<td>22</td>
</tr>
<tr>
<td>v1</td>
<td>13</td>
<td>0</td>
<td>12</td>
<td>12</td>
<td>13</td>
<td>13</td>
</tr>
<tr>
<td>v2</td>
<td>21</td>
<td>12</td>
<td>0</td>
<td>20</td>
<td>21</td>
<td>21</td>
</tr>
<tr>
<td>v3</td>
<td>21</td>
<td>12</td>
<td>20</td>
<td>0</td>
<td>7</td>
<td>13</td>
</tr>
<tr>
<td>v4</td>
<td>22</td>
<td>13</td>
<td>21</td>
<td>7</td>
<td>0</td>
<td>14</td>
</tr>
<tr>
<td>v5</td>
<td>22</td>
<td>13</td>
<td>21</td>
<td>13</td>
<td>14</td>
<td>0</td>
</tr>
</tbody>
</table>
<p>Setting v5's limb length to 0 (balding v5) would result in ...</p>
<p><img src="dot_504f11240750f6a2add064fe593ca205.svg" alt="Dot diagram"></p>
<table>
<thead>
<tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr>
</thead>
<tbody>
<tr>
<td>v0</td>
<td>0</td>
<td>13</td>
<td>21</td>
<td>21</td>
<td>22</td>
<td>15</td>
</tr>
<tr>
<td>v1</td>
<td>13</td>
<td>0</td>
<td>12</td>
<td>12</td>
<td>13</td>
<td>6</td>
</tr>
<tr>
<td>v2</td>
<td>21</td>
<td>12</td>
<td>0</td>
<td>20</td>
<td>21</td>
<td>14</td>
</tr>
<tr>
<td>v3</td>
<td>21</td>
<td>12</td>
<td>20</td>
<td>0</td>
<td>7</td>
<td>6</td>
</tr>
<tr>
<td>v4</td>
<td>22</td>
<td>13</td>
<td>21</td>
<td>7</td>
<td>0</td>
<td>7</td>
</tr>
<tr>
<td>v5</td>
<td>15</td>
<td>6</td>
<td>14</td>
<td>6</td>
<td>7</td>
<td>0</td>
</tr>
</tbody>
</table>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>Can a <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a> be 0 in a <a href="#BM_(simple%20tree)%2Fi">simple tree</a>? I don't think so, but the book seems to imply that it's possible. But, if the distance between the two nodes on an edge is 0, wouldn't that make them the same organism? Maybe this is just a temporary thing for this algorithm.</p>
</div>
<p>Notice how of the two <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrices</a>, all distances are the same except for v5's distances. Each v5 distance in the balded distance matrix is equivalent to the corresponding distance in the original distance matrix subtracted by v5's original limb length...</p>
<table>
<thead>
<tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr>
</thead>
<tbody>
<tr>
<td>v0</td>
<td>0</td>
<td>13</td>
<td>21</td>
<td>21</td>
<td>22</td>
<td>22 - 7 = 15</td>
</tr>
<tr>
<td>v1</td>
<td>13</td>
<td>0</td>
<td>12</td>
<td>12</td>
<td>13</td>
<td>13 - 7 = 6</td>
</tr>
<tr>
<td>v2</td>
<td>21</td>
<td>12</td>
<td>0</td>
<td>20</td>
<td>21</td>
<td>21 - 7 = 14</td>
</tr>
<tr>
<td>v3</td>
<td>21</td>
<td>12</td>
<td>20</td>
<td>0</td>
<td>7</td>
<td>13 - 7 = 6</td>
</tr>
<tr>
<td>v4</td>
<td>22</td>
<td>13</td>
<td>21</td>
<td>7</td>
<td>0</td>
<td>14 - 7 = 7</td>
</tr>
<tr>
<td>v5</td>
<td>22 - 7 = 15</td>
<td>13 - 7 = 6</td>
<td>21 - 7 = 14</td>
<td>13 - 7 = 6</td>
<td>14 - 7 = 7</td>
<td>0</td>
</tr>
</tbody>
</table>
<p>Whereas v5 was originally contributing 7 to distances, after balding it contributes 0.</p>
<p>As such, subtracting some leaf node's <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a> from its distances in an <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> is equivalent to balding that leaf node's <a href="#BM_(limb)%2Fi">limb</a> in its corresponding <a href="#BM_(simple%20tree)%2Fi">simple tree</a>.</p>
<p><a href="ch7_code/src/phylogeny/Balder.py">ch7_code/src/phylogeny/Balder.py</a> (lines 25 to 38):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">bald_distance_matrix</span>(<span class="hljs-params">dm: DistanceMatrix[N], leaf: N</span>) -&gt; <span class="hljs-literal">None</span>:
    limb_len = find_limb_length(dm, leaf)
    <span class="hljs-keyword">for</span> n <span class="hljs-keyword">in</span> dm.leaf_ids_it():
        <span class="hljs-keyword">if</span> n == leaf:
            <span class="hljs-keyword">continue</span>
        dm[leaf, n] -= limb_len


<span class="hljs-keyword">def</span> <span class="hljs-title function_">bald_tree</span>(<span class="hljs-params">tree: Graph[N, ND, E, <span class="hljs-built_in">float</span>], leaf: N</span>) -&gt; <span class="hljs-literal">None</span>:
    <span class="hljs-keyword">if</span> tree.get_degree(leaf) != <span class="hljs-number">1</span>:
        <span class="hljs-keyword">raise</span> ValueError(<span class="hljs-string">'Not a leaf node'</span>)
    limb = <span class="hljs-built_in">next</span>(tree.get_outputs(leaf))
    tree.update_edge_data(limb, <span class="hljs-number">0.0</span>)
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the additive distance matrix...</p>
<table>
<thead><tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr></thead>
<tbody>
<tr>
<td>v0</td>
<td>0.0</td>
<td>13.0</td>
<td>21.0</td>
<td>21.0</td>
<td>22.0</td>
<td>22.0</td>
</tr>
<tr>
<td>v1</td>
<td>13.0</td>
<td>0.0</td>
<td>12.0</td>
<td>12.0</td>
<td>13.0</td>
<td>13.0</td>
</tr>
<tr>
<td>v2</td>
<td>21.0</td>
<td>12.0</td>
<td>0.0</td>
<td>20.0</td>
<td>21.0</td>
<td>21.0</td>
</tr>
<tr>
<td>v3</td>
<td>21.0</td>
<td>12.0</td>
<td>20.0</td>
<td>0.0</td>
<td>7.0</td>
<td>13.0</td>
</tr>
<tr>
<td>v4</td>
<td>22.0</td>
<td>13.0</td>
<td>21.0</td>
<td>7.0</td>
<td>0.0</td>
<td>14.0</td>
</tr>
<tr>
<td>v5</td>
<td>22.0</td>
<td>13.0</td>
<td>21.0</td>
<td>13.0</td>
<td>14.0</td>
<td>0.0</td>
</tr>
</tbody>
</table>
<p>... trimming leaf node v5 results in ...</p>
<table>
<thead><tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr></thead>
<tbody>
<tr>
<td>v0</td>
<td>0.0</td>
<td>13.0</td>
<td>21.0</td>
<td>21.0</td>
<td>22.0</td>
<td>15.0</td>
</tr>
<tr>
<td>v1</td>
<td>13.0</td>
<td>0.0</td>
<td>12.0</td>
<td>12.0</td>
<td>13.0</td>
<td>6.0</td>
</tr>
<tr>
<td>v2</td>
<td>21.0</td>
<td>12.0</td>
<td>0.0</td>
<td>20.0</td>
<td>21.0</td>
<td>14.0</td>
</tr>
<tr>
<td>v3</td>
<td>21.0</td>
<td>12.0</td>
<td>20.0</td>
<td>0.0</td>
<td>7.0</td>
<td>6.0</td>
</tr>
<tr>
<td>v4</td>
<td>22.0</td>
<td>13.0</td>
<td>21.0</td>
<td>7.0</td>
<td>0.0</td>
<td>7.0</td>
</tr>
<tr>
<td>v5</td>
<td>15.0</td>
<td>6.0</td>
<td>14.0</td>
<td>6.0</td>
<td>7.0</td>
<td>0.0</td>
</tr>
</tbody>
</table>
</div>
<p></p>
<p> <!-- Re-enabling. Needed to be disabled because v3's is conflicting with 3' -->
 <!-- Re-enabling. Needed to be disabled because v5's is conflicting with 5' --></p>
<a name="H_Un-trim%20Tree"></a><h3>Un-trim Tree</h3>
<p><a name="BM_(Algorithms%5C%2FPhylogeny%2FUn-trim%20Tree)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FFind%20Limb%20Length)_TOPIC%2F">Algorithms/Phylogeny/Find Limb Length</a></li>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FTest%20Same%20Subtree)_TOPIC%2F">Algorithms/Phylogeny/Test Same Subtree</a></li>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FTrim)_TOPIC%2F">Algorithms/Phylogeny/Trim</a></li>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FBald)_TOPIC%2F">Algorithms/Phylogeny/Bald</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: Given an ...</p>
<ol>
<li><a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> for <a href="#BM_(simple%20tree)%2Fi">simple tree</a> T</li>
<li><a href="#BM_(simple%20tree)%2Fi">simple tree</a> T with <a href="#BM_(limb)%2Fi">limb</a> L <a href="#BM_(trim distance matrix|trimmed distance matrix|trim distance matrices|trimmed distance matrices)/i">trimmed</a> off</li>
</ol>
<p>... this algorithm determines where <a href="#BM_(limb)%2Fi">limb</a> L should be added in the given <a href="#BM_(simple%20tree)%2Fi">simple tree</a> such that it fits the <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>. For example, the following <a href="#BM_(simple%20tree)%2Fi">simple tree</a> would map to the following <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> had v2's <a href="#BM_(limb)%2Fi">limb</a> branched out from some specific location...</p>
<p><img src="dot_d59e5cc5d23e349cc4668d23c9e4f80c.svg" alt="Dot diagram"></p>
<table>
<thead>
<tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
</tr>
</thead>
<tbody>
<tr>
<td>v0</td>
<td>0</td>
<td>13</td>
<td>21</td>
<td>22</td>
</tr>
<tr>
<td>v1</td>
<td>13</td>
<td>0</td>
<td>12</td>
<td>13</td>
</tr>
<tr>
<td>v2</td>
<td>21</td>
<td>12</td>
<td>0</td>
<td>13</td>
</tr>
<tr>
<td>v3</td>
<td>22</td>
<td>13</td>
<td>13</td>
<td>0</td>
</tr>
</tbody>
</table>
<p>That specific location is what this algorithm determines. It could be that v2's <a href="#BM_(limb)%2Fi">limb</a> needs to branch from either ...</p>
<ul>
<li>
<p>an internal node ...</p>
<p><img src="dot_3ec114a5508a03ba155cfc398897b924.svg" alt="Dot diagram"></p>
</li>
<li>
<p>an edge, breaking that edge into two by attaching an internal node in between...</p>
<p><img src="dot_371eab0575eeea1e0be669c219297f02.svg" alt="Dot diagram"></p>
</li>
</ul>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>Attaching a new <a href="#BM_(limb)%2Fi">limb</a> to an existing leaf node is never possible because...</p>
<ol>
<li>it'll turn that existing leaf node to an internal node, which doesn't make sense because in the context of <a href="#BM_(phylogenetic%20tree%7Cphylogeny%20tree%7Cphylogenies%7Cevolutionary%20tree)%2Fi">phylogenetic tree</a>s leaf nodes identify known entities.</li>
<li>it will cease to be a <a href="#BM_(simple%20tree)%2Fi">simple tree</a> -- <a href="#BM_(simple%20tree)%2Fi">simple tree</a>s can't have nodes of <a href="#BM_(degree)_GRAPH%2Fi">degree</a> 2 (train of edges not allowed).</li>
</ol>
</div>
<p><strong>WHY</strong>: This is one of the operations required to construct the unique <a href="#BM_(simple%20tree)%2Fi">simple tree</a> for an <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>.</p>
<p><strong>ALGORITHM</strong>:</p>
<p> <!-- Needs to be disabled because v5's is conflicting with 5' --></p>
<p>The <a href="#BM_(simple%20tree)%2Fi">simple tree</a> below would fit the <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> below had v5's limb been added to it somewhere ...</p>
<p><img src="dot_474a7e4cf5bc49a86d6321e9fb6ca532.svg" alt="Dot diagram"></p>
<table>
<thead>
<tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr>
</thead>
<tbody>
<tr>
<td>v0</td>
<td>0</td>
<td>13</td>
<td>21</td>
<td>21</td>
<td>22</td>
<td>22</td>
</tr>
<tr>
<td>v1</td>
<td>13</td>
<td>0</td>
<td>12</td>
<td>12</td>
<td>13</td>
<td>13</td>
</tr>
<tr>
<td>v2</td>
<td>21</td>
<td>12</td>
<td>0</td>
<td>20</td>
<td>21</td>
<td>21</td>
</tr>
<tr>
<td>v3</td>
<td>21</td>
<td>12</td>
<td>20</td>
<td>0</td>
<td>7</td>
<td>13</td>
</tr>
<tr>
<td>v4</td>
<td>22</td>
<td>13</td>
<td>21</td>
<td>7</td>
<td>0</td>
<td>14</td>
</tr>
<tr>
<td>v5</td>
<td>22</td>
<td>13</td>
<td>21</td>
<td>13</td>
<td>14</td>
<td>0</td>
</tr>
</tbody>
</table>
<p>There's enough information available in this <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> to determine ...</p>
<ul>
<li>v5's limb length: 7 (calculated using the limb length algorithm)</li>
<li>a pair of nodes whose path travels over v5's parent: v0 and v3 (calculated using the same subset algorithm)</li>
</ul>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>Recall that same subset algorithm says that two leaf nodes in DIFFERENT subsets are guaranteed to travel over v5's parent.</p>
</div>
<p>The key to this algorithm is figuring out where along that path (v0 to v3) v5's limb (limb length of 7) should be injected. Imagine that you already had the answer in front of you: v5's limb should be added 4 units from i0 towards i2 ...</p>
<p><img src="dot_7fd5d55328c43bca2193b9548728df02.svg" alt="Dot diagram"></p>
<p>Consider the answer above with v5's limb <a href="#BM_(bald distance matrix|balded distance matrix|bald distance matrices|balded distance matrices)/i">balded</a>...</p>
<p><img src="dot_782638a05800732e25c548f888d031b4.svg" alt="Dot diagram"></p>
<table>
<thead>
<tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr>
</thead>
<tbody>
<tr>
<td>v0</td>
<td>0</td>
<td>13</td>
<td>21</td>
<td>21</td>
<td>22</td>
<td>22 - 7 = 15</td>
</tr>
<tr>
<td>v1</td>
<td>13</td>
<td>0</td>
<td>12</td>
<td>12</td>
<td>13</td>
<td>13 - 7 = 6</td>
</tr>
<tr>
<td>v2</td>
<td>21</td>
<td>12</td>
<td>0</td>
<td>20</td>
<td>21</td>
<td>21 - 7 = 14</td>
</tr>
<tr>
<td>v3</td>
<td>21</td>
<td>12</td>
<td>20</td>
<td>0</td>
<td>7</td>
<td>13 - 7 = 6</td>
</tr>
<tr>
<td>v4</td>
<td>22</td>
<td>13</td>
<td>21</td>
<td>7</td>
<td>0</td>
<td>14 - 7 = 7</td>
</tr>
<tr>
<td>v5</td>
<td>22 - 7 = 15</td>
<td>13 - 7 = 6</td>
<td>21 - 7 = 14</td>
<td>13 - 7 = 6</td>
<td>14 - 7 = 7</td>
<td>0</td>
</tr>
</tbody>
</table>
<p>Since v5's limb length is 0, it doesn't contribute to the distance of any path to / from v5. As such, the distance of any path to / from v5 is actually the distance to / from its parent. For example, ...</p>
<ul>
<li>dist(v0,v5) = dist(v0,i1) + 0 = dist(v0,i1)</li>
<li>dist(v4,v5) = dist(v4,i1) + 0 = dist(v4,i1)</li>
<li>dist(v2,v5) = dist(v2,i1) + 0 = dist(v2,i1)</li>
<li>etc..</li>
</ul>
<p><img src="dot_beaa27212249bfd3c9e90e3ff22371b8.svg" alt="Dot diagram"></p>
<p>Essentially, the <a href="#BM_(bald%20distance%20matrix%7Cbalded%20distance%20matrix%7Cbald%20distance%20matrices%7Cbalded%20distance%20matrices)%2Fi">balded distance matrix</a> is enough to tell you that the path from v0 to v5's parent has a distance of 15. The balded tree itself isn't required.</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">find_pair_traveling_thru_leaf_parent</span>(<span class="hljs-params">dist_mat: DistanceMatrix[N], leaf_node: N</span>) -&gt; <span class="hljs-built_in">tuple</span>[N, N]:
    leaf_set = dist_mat.leaf_ids() - {leaf_node}
    <span class="hljs-keyword">for</span> l1, l2 <span class="hljs-keyword">in</span> product(leaf_set, repeat=<span class="hljs-number">2</span>):
        <span class="hljs-keyword">if</span> <span class="hljs-keyword">not</span> is_same_subtree(dist_mat, leaf_node, l1, l2):
            <span class="hljs-keyword">return</span> l1, l2
    <span class="hljs-keyword">raise</span> ValueError(<span class="hljs-string">'Not found'</span>)


<span class="hljs-keyword">def</span> <span class="hljs-title function_">find_distance_to_leaf_parent</span>(<span class="hljs-params">dist_mat: DistanceMatrix[N], from_leaf_node: N, to_leaf_node: N</span>) -&gt; <span class="hljs-built_in">float</span>:
    balded_dist_mat = dist_mat.copy()
    bald_distance_matrix(balded_dist_mat, to_leaf_node)
    <span class="hljs-keyword">return</span> balded_dist_mat[from_leaf_node, to_leaf_node]
</code></pre>
<p>In the original <a href="#BM_(simple%20tree)%2Fi">simple tree</a>, walking a distance of 15 on the path from v0 to v3 takes you to where v5's parent should be. Since there is no internal node there, one is first added by breaking the edge before attaching v5's limb to it ...</p>
<p><img src="dot_b37fe50f63d1a3ab5722ac6b80a41430.svg" alt="Dot diagram"></p>
<p>Had there been an internal node already there, the <a href="#BM_(limb)%2Fi">limb</a> would get attached to that existing internal node.</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">walk_until_distance</span>(<span class="hljs-params">
        tree: Graph[N, ND, E, <span class="hljs-built_in">float</span>],
        n_start: N,
        n_end: N,
        dist: <span class="hljs-built_in">float</span>
</span>) -&gt; <span class="hljs-type">Union</span>[
    <span class="hljs-built_in">tuple</span>[<span class="hljs-type">Literal</span>[<span class="hljs-string">'NODE'</span>], N],
    <span class="hljs-built_in">tuple</span>[<span class="hljs-type">Literal</span>[<span class="hljs-string">'EDGE'</span>], E, N, N, <span class="hljs-built_in">float</span>, <span class="hljs-built_in">float</span>]
]:
    path = find_path(tree, n_start, n_end)
    last_edge_end = n_start
    dist_walked = <span class="hljs-number">0.0</span>
    <span class="hljs-keyword">for</span> edge <span class="hljs-keyword">in</span> path:
        ends = tree.get_edge_ends(edge)
        n1 = last_edge_end
        n2 = <span class="hljs-built_in">next</span>(n <span class="hljs-keyword">for</span> n <span class="hljs-keyword">in</span> ends <span class="hljs-keyword">if</span> n != last_edge_end)
        weight = tree.get_edge_data(edge)
        dist_walked_with_weight = dist_walked + weight
        <span class="hljs-keyword">if</span> dist_walked_with_weight &gt; dist:
            <span class="hljs-keyword">return</span> <span class="hljs-string">'EDGE'</span>, edge, n1, n2, dist_walked, weight
        <span class="hljs-keyword">elif</span> dist_walked_with_weight == dist:
            <span class="hljs-keyword">return</span> <span class="hljs-string">'NODE'</span>, n2
        dist_walked = dist_walked_with_weight
        last_edge_end = n2
    <span class="hljs-keyword">raise</span> ValueError(<span class="hljs-string">'Bad inputs'</span>)
</code></pre>
<p><a href="ch7_code/src/phylogeny/UntrimTree.py">ch7_code/src/phylogeny/UntrimTree.py</a> (lines 110 to 148):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">untrim_tree</span>(<span class="hljs-params">
        dist_mat: DistanceMatrix[N],
        trimmed_tree: Graph[N, ND, E, <span class="hljs-built_in">float</span>],
        gen_node_id: <span class="hljs-type">Callable</span>[[], N],
        gen_edge_id: <span class="hljs-type">Callable</span>[[], E]
</span>) -&gt; <span class="hljs-literal">None</span>:
    <span class="hljs-comment"># Which node was trimmed?</span>
    n_trimmed = find_trimmed_leaf(dist_mat, trimmed_tree)
    <span class="hljs-comment"># Find a pair whose path that goes through the trimmed node's parent</span>
    n_start, n_end = find_pair_traveling_thru_leaf_parent(dist_mat, n_trimmed)
    <span class="hljs-comment"># What's the distance from n_start to the trimmed node's parent?</span>
    parent_dist = find_distance_to_leaf_parent(dist_mat, n_start, n_trimmed)
    <span class="hljs-comment"># Walk the path from n_start to n_end, stopping once walk dist reaches parent_dist (where trimmed node's parent is)</span>
    res = walk_until_distance(trimmed_tree, n_start, n_end, parent_dist)
    stopped_on = res[<span class="hljs-number">0</span>]
    <span class="hljs-keyword">if</span> stopped_on == <span class="hljs-string">'NODE'</span>:
        <span class="hljs-comment"># It stopped on an existing internal node -- the limb should be added to this node</span>
        parent_n = res[<span class="hljs-number">1</span>]
    <span class="hljs-keyword">elif</span> stopped_on == <span class="hljs-string">'EDGE'</span>:
        <span class="hljs-comment"># It stopped on an edge -- a new internal node should be injected to break the edge, then the limb should extend</span>
        <span class="hljs-comment"># from that node.</span>
        edge, n1, n2, walked_dist, edge_weight = res[<span class="hljs-number">1</span>:]
        parent_n = gen_node_id()
        trimmed_tree.insert_node(parent_n)
        n1_to_parent_id = gen_edge_id()
        n1_to_parent_weight = parent_dist - walked_dist
        trimmed_tree.insert_edge(n1_to_parent_id, n1, parent_n, n1_to_parent_weight)
        parent_to_n2_id = gen_edge_id()
        parent_to_n2_weight = edge_weight - n1_to_parent_weight
        trimmed_tree.insert_edge(parent_to_n2_id, parent_n, n2, parent_to_n2_weight)
        trimmed_tree.delete_edge(edge)
    <span class="hljs-keyword">else</span>:
        <span class="hljs-keyword">raise</span> ValueError(<span class="hljs-string">'???'</span>)
    <span class="hljs-comment"># Add the limb</span>
    limb_e = gen_edge_id()
    limb_len = find_limb_length(dist_mat, n_trimmed)
    trimmed_tree.insert_node(n_trimmed)
    trimmed_tree.insert_edge(limb_e, parent_n, n_trimmed, limb_len)
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the additive distance matrix for simple tree T...</p>
<table>
<thead><tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr></thead>
<tbody>
<tr>
<td>v0</td>
<td>0.0</td>
<td>13.0</td>
<td>21.0</td>
<td>21.0</td>
<td>22.0</td>
<td>22.0</td>
</tr>
<tr>
<td>v1</td>
<td>13.0</td>
<td>0.0</td>
<td>12.0</td>
<td>12.0</td>
<td>13.0</td>
<td>13.0</td>
</tr>
<tr>
<td>v2</td>
<td>21.0</td>
<td>12.0</td>
<td>0.0</td>
<td>20.0</td>
<td>21.0</td>
<td>21.0</td>
</tr>
<tr>
<td>v3</td>
<td>21.0</td>
<td>12.0</td>
<td>20.0</td>
<td>0.0</td>
<td>7.0</td>
<td>13.0</td>
</tr>
<tr>
<td>v4</td>
<td>22.0</td>
<td>13.0</td>
<td>21.0</td>
<td>7.0</td>
<td>0.0</td>
<td>14.0</td>
</tr>
<tr>
<td>v5</td>
<td>22.0</td>
<td>13.0</td>
<td>21.0</td>
<td>13.0</td>
<td>14.0</td>
<td>0.0</td>
</tr>
</tbody>
</table>
<p>... and simple tree trim(T, v5)...</p>
<p><img src="dot_947428a9fad00efc08c05fcfe2286cfb.svg" alt="Dot diagram"></p>
<p>... , v5 is injected at the appropriate location to become simple tree T (un-trimmed) ...</p>
<p><img src="dot_732ee38514b240e67ad8b3b760b395eb.svg" alt="Dot diagram"></p>
</div>
<p></p>
<p> <!-- Re-enable 5' now that section is finished --></p>
<a name="H_Find%20Neighbours"></a><h3>Find Neighbours</h3>
<p><a name="BM_(Algorithms%5C%2FPhylogeny%2FFind%20Neighbours)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FTree%20to%20Additive%20Distance%20Matrix)_TOPIC%2F">Algorithms/Phylogeny/Tree to Additive Distance Matrix</a></li>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FTree%20to%20Simple%20Tree)_TOPIC%2F">Algorithms/Phylogeny/Tree to Simple Tree</a></li>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FAdditive%20Distance%20Matrix%20Cardinality)_TOPIC%2F">Algorithms/Phylogeny/Additive Distance Matrix Cardinality</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: Given a <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a>, if the <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> is ...</p>
<ul>
<li>an <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>, this algorithm finds a pair of leaf nodes guaranteed to be <a href="#BM_(neighbour)%2Fi">neighbour</a>s in its corresponding unique <a href="#BM_(simple%20tree)%2Fi">simple tree</a>.</li>
<li>a non-<a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> (but close to being additive), this algorithm approximates a pair of leaf nodes that are likely to be <a href="#BM_(neighbour)%2Fi">neighbour</a>s.</li>
</ul>
<p><strong>WHY</strong>: This operation is required for <em>approximating</em> a <a href="#BM_(simple%20tree)%2Fi">simple tree</a> for a non-<a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>.</p>
<p><strong>ALGORITHM</strong>:</p>
<p>The algorithm essentially boils down to edge counting. Consider the following example <a href="#BM_(simple%20tree)%2Fi">simple tree</a>...</p>
<p><img src="dot_fa269cbcc76261c503f77dc3a6e5b8b6.svg" alt="Dot diagram"></p>
<p>If you were to choose a leaf node, then gather the paths from that leaf node to all other leaf nodes, the <a href="#BM_(limb)%2Fi">limb</a> for ...</p>
<ul>
<li>the chosen leaf node gets encountered  <code>leaf_count - 1</code> times.</li>
<li>each non-chosen leaf node gets encountered once.</li>
</ul>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">edge_count</span>(<span class="hljs-params">self, l1: N</span>) -&gt; Counter[E]:
    <span class="hljs-comment"># Collect paths from l1 to all other leaf nodes</span>
    path_collection = []
    <span class="hljs-keyword">for</span> l2 <span class="hljs-keyword">in</span> self.leaf_nodes:
        <span class="hljs-keyword">if</span> l1 == l2:
            <span class="hljs-keyword">continue</span>
        path = self.path(l1, l2)
        path_collection.append(path)
    <span class="hljs-comment"># Count edges across all paths</span>
    edge_counts = Counter()
    <span class="hljs-keyword">for</span> path <span class="hljs-keyword">in</span> path_collection:
        edge_counts.update(path)
    <span class="hljs-comment"># Return edge counts</span>
    <span class="hljs-keyword">return</span> edge_counts
</code></pre>
<p>For example, given that the <a href="#BM_(tree)%2Fi">tree</a> has 6 leaf nodes, <code>edge_count(v1)</code> counts v1's <a href="#BM_(limb)%2Fi">limb</a> 5 times while all other <a href="#BM_(limb)%2Fi">limb</a>s are counted once...</p>
<p><img src="dot_be00dd5022c82978ec0451b8a93c238a.svg" alt="Dot diagram"></p>
<table>
<thead>
<tr>
<th></th>
<th>(i0,i1)</th>
<th>(i1,i2)</th>
<th>(v0,i0)</th>
<th>(v1,i0)</th>
<th>(v2,i0)</th>
<th>(v3,i2)</th>
<th>(v4,i2)</th>
<th>(v5,i1)</th>
</tr>
</thead>
<tbody>
<tr>
<td>edge_count(v1)</td>
<td>3</td>
<td>2</td>
<td>1</td>
<td>5</td>
<td>1</td>
<td>1</td>
<td>1</td>
<td>1</td>
</tr>
</tbody>
</table>
<p>If you were to choose a pair of leaf nodes and add their <code>edge_count()</code>s together, the <a href="#BM_(limb)%2Fi">limb</a> for ...</p>
<ul>
<li>each chosen leaf node gets encountered  <code>leaf_count</code> times.</li>
<li>each non-chosen leaf node gets encountered twice.</li>
</ul>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">combine_edge_count</span>(<span class="hljs-params">self, l1: N, l2: N</span>) -&gt; Counter[E]:
    c1 = self.edge_count(l1)
    c2 = self.edge_count(l2)
    <span class="hljs-keyword">return</span> c1 + c2
</code></pre>
<p>For example, <code>combine_edge_count(v1,v2)</code> counts v1's <a href="#BM_(limb)%2Fi">limb</a> 6 times, v2's <a href="#BM_(limb)%2Fi">limb</a> 6 times, and every other <a href="#BM_(limb)%2Fi">limb</a> 2 times ...</p>
<p><img src="dot_980ac0c4600f11bdf470191f270e14f3.svg" alt="Dot diagram"></p>
<table>
<thead>
<tr>
<th></th>
<th>(i0,i1)</th>
<th>(i1,i2)</th>
<th>(v0,i0)</th>
<th>(v1,i0)</th>
<th>(v2,i0)</th>
<th>(v3,i2)</th>
<th>(v4,i2)</th>
<th>(v5,i1)</th>
</tr>
</thead>
<tbody>
<tr>
<td>edge_count(v1)</td>
<td>3</td>
<td>2</td>
<td>1</td>
<td>5</td>
<td>1</td>
<td>1</td>
<td>1</td>
<td>1</td>
</tr>
<tr>
<td>edge_count(v2)</td>
<td>3</td>
<td>2</td>
<td>1</td>
<td>1</td>
<td>5</td>
<td>1</td>
<td>1</td>
<td>1</td>
</tr>
<tr>
<td></td>
<td>-------</td>
<td>-------</td>
<td>-------</td>
<td>-------</td>
<td>-------</td>
<td>-------</td>
<td>-------</td>
<td>-------</td>
</tr>
<tr>
<td></td>
<td>6</td>
<td>4</td>
<td>2</td>
<td>6</td>
<td>6</td>
<td>2</td>
<td>2</td>
<td>2</td>
</tr>
</tbody>
</table>
<p>The key to this algorithm is to normalize <a href="#BM_(limb)%2Fi">limb</a> counts returned by <code>combine_counts()</code> such that each chosen <a href="#BM_(limb)%2Fi">limb</a>'s count equals to each non-chosen <a href="#BM_(limb)%2Fi">limb</a>'s count. That is, each chosen <a href="#BM_(limb)%2Fi">limb</a> count needs to be reduced from <code>leaf_count</code> to 2.</p>
<p>To do this, each edge in the path between the chosen pair must be subtracted <code>leaf_count - 2</code> times from <code>combine_edge_count()</code>'s result.</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">combine_edge_count_and_normalize</span>(<span class="hljs-params">self, l1: N, l2: N</span>) -&gt; Counter[E]:
    edge_counts = self.combine_edge_count(l1, l2)
    path_edges = self.path(l1, l2)
    <span class="hljs-keyword">for</span> e <span class="hljs-keyword">in</span> path_edges:
        edge_counts[e] -= self.leaf_count - <span class="hljs-number">2</span>
    <span class="hljs-keyword">return</span> edge_counts
</code></pre>
<p>Continuing with the example above, the chosen pair (v1 and v2) each have a <a href="#BM_(limb)%2Fi">limb</a> count of 6 while all other <a href="#BM_(limb)%2Fi">limb</a>s have a count of 2. <code>combine_edge_count_and_normalize(v1,v2)</code> subtracts each edge in path(v1,v2) 4 times from the counts...</p>
<p><img src="dot_df662efca0290cfec12e5b3356609f34.svg" alt="Dot diagram"></p>
<table>
<thead>
<tr>
<th></th>
<th>(i0,i1)</th>
<th>(i1,i2)</th>
<th>(v0,i0)</th>
<th>(v1,i0)</th>
<th>(v2,i0)</th>
<th>(v3,i2)</th>
<th>(v4,i2)</th>
<th>(v5,i1)</th>
</tr>
</thead>
<tbody>
<tr>
<td>edge_count(v1)</td>
<td>3</td>
<td>2</td>
<td>1</td>
<td>5</td>
<td>1</td>
<td>1</td>
<td>1</td>
<td>1</td>
</tr>
<tr>
<td>edge_count(v2)</td>
<td>3</td>
<td>2</td>
<td>1</td>
<td>1</td>
<td>5</td>
<td>1</td>
<td>1</td>
<td>1</td>
</tr>
<tr>
<td>-4 * path(v1,v2)</td>
<td></td>
<td></td>
<td></td>
<td>-4</td>
<td>-4</td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td></td>
<td>-------</td>
<td>-------</td>
<td>-------</td>
<td>-------</td>
<td>-------</td>
<td>-------</td>
<td>-------</td>
<td>-------</td>
</tr>
<tr>
<td></td>
<td>6</td>
<td>4</td>
<td>2</td>
<td>2</td>
<td>2</td>
<td>2</td>
<td>2</td>
<td>2</td>
</tr>
</tbody>
</table>
<p>The insight here is that, if the chosen pair ...</p>
<ul>
<li>are <a href="#BM_(neighbour)%2Fi">neighbour</a>s, the path between them will only ever have 2 edges: their <a href="#BM_(limb)%2Fi">limb</a>s.</li>
<li>aren't <a href="#BM_(neighbour)%2Fi">neighbour</a>s, the path between them will have more than 2 edges: their <a href="#BM_(limb)%2Fi">limb</a>s AND internal edges.</li>
</ul>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">neighbour_check</span>(<span class="hljs-params">self, l1: N, l2: N</span>) -&gt; <span class="hljs-built_in">bool</span>:
    path_edges = self.path(l1, l2)
    <span class="hljs-keyword">return</span> <span class="hljs-built_in">len</span>(path_edges) == <span class="hljs-number">2</span>
</code></pre>
<p>For example, ...</p>
<ul>
<li>v1 and v2 are <a href="#BM_(neighbour)%2Fi">neighbour</a>s and as such path(v1,v2) contains only their <a href="#BM_(limb)%2Fi">limb</a>s: [(v1,i0), (v2,i0)].</li>
<li>v1 and v5 aren't <a href="#BM_(neighbour)%2Fi">neighbour</a>s and as such path(v1,v5) has internal edges in addition to their <a href="#BM_(limb)%2Fi">limb</a>s: [(v1,i0), (i0,i1), (v5,i1)].</li>
</ul>
<p><img src="dot_5a3df09ff196ebbf702909f1cb634c93.svg" alt="Dot diagram"></p>
<p>That means if the pair aren't <a href="#BM_(neighbour)%2Fi">neighbour</a>s, <code>combine_edge_count_and_normalize()</code> will normalize <a href="#BM_(limb)%2Fi">limb</a> counts for the pair in addition to reducing internal edge counts. For example, since v1 and v5 aren't <a href="#BM_(neighbour)%2Fi">neighbour</a>s, <code>combine_edge_count_and_normalize(v1,v5)</code> subtracts 4 from the <a href="#BM_(limb)%2Fi">limb</a> counts of v1 and v5 as well as (i0,i1)'s count ...</p>
<p><img src="dot_88d8e6262e9c3de072f7484bf17ab489.svg" alt="Dot diagram"></p>
<table>
<thead>
<tr>
<th></th>
<th>(i0,i1)</th>
<th>(i1,i2)</th>
<th>(v0,i0)</th>
<th>(v1,i0)</th>
<th>(v2,i0)</th>
<th>(v3,i2)</th>
<th>(v4,i2)</th>
<th>(v5,i1)</th>
</tr>
</thead>
<tbody>
<tr>
<td>edge_count(v1)</td>
<td>3</td>
<td>2</td>
<td>1</td>
<td>5</td>
<td>1</td>
<td>1</td>
<td>1</td>
<td>1</td>
</tr>
<tr>
<td>edge_count(v5)</td>
<td>3</td>
<td>2</td>
<td>1</td>
<td>1</td>
<td>1</td>
<td>1</td>
<td>1</td>
<td>5</td>
</tr>
<tr>
<td>-4 * path(v1,v5)</td>
<td>-4</td>
<td></td>
<td></td>
<td>-4</td>
<td></td>
<td></td>
<td></td>
<td>-4</td>
</tr>
<tr>
<td></td>
<td>-------</td>
<td>-------</td>
<td>-------</td>
<td>-------</td>
<td>-------</td>
<td>-------</td>
<td>-------</td>
<td>-------</td>
</tr>
<tr>
<td></td>
<td>2</td>
<td>4</td>
<td>2</td>
<td>2</td>
<td>2</td>
<td>2</td>
<td>2</td>
<td>2</td>
</tr>
</tbody>
</table>
<p>Notice how (i0,i1) was reduced to 2 in the example above. It turns out that any internal edges in the path between the chosen pair get reduced to a count of 2, just like the chosen pair's <a href="#BM_(limb)%2Fi">limb</a> counts.</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">reduced_to_2_check</span>(<span class="hljs-params">self, l1: N, l2: N</span>) -&gt; <span class="hljs-built_in">bool</span>:
    p = self.path(l1, l2)
    c = self.combine_edge_count_and_normalize(l1, l2)
    <span class="hljs-keyword">return</span> <span class="hljs-built_in">all</span>(c[edge] == <span class="hljs-number">2</span> <span class="hljs-keyword">for</span> edge <span class="hljs-keyword">in</span> p)  <span class="hljs-comment"># if counts for all edges in p reduced to 2</span>
</code></pre>
<p>To understand why, consider what's happening in the example. For <code>edge_count(v1)</code>, notice how the count of each internal edge is consistent with the number of leaf nodes it leads to ...</p>
<p><img src="dot_be00dd5022c82978ec0451b8a93c238a.svg" alt="Dot diagram"></p>
<p>That is, <code>edge_count(v1)</code> counts the internal edge ...</p>
<ul>
<li>(i0,i1) 3 times, which is the number of nodes it leads to: [v5, v4, v3].</li>
<li>(i1,i2) 2 times, which is the number of nodes it leads to: [v4, v3].</li>
</ul>
<p>Breaking an internal edge divides a <a href="#BM_(tree)%2Fi">tree</a> into two <a href="#BM_(subtree%7Csub-tree)%2Fi">sub-tree</a>s. In the case of (i1,i2), the <a href="#BM_(tree)%2Fi">tree</a> separates into two <a href="#BM_(subtree%7Csub-tree)%2Fi">sub-tree</a>s where the...</p>
<ul>
<li>i1 side has 4 leaf nodes: [v0, v1, v2, v5].</li>
<li>i2 side has 2 leaf nodes: [v4, v3].</li>
</ul>
<p><img src="dot_2faf30aa6f812fb44557068e9fa76660.svg" alt="Dot diagram"></p>
<p>Running <code>edge_count()</code> for any leaf node on the...</p>
<ul>
<li>i1 side will count (i1,i2) exactly 2 times (the number of leaf nodes on the i2 side).</li>
<li>i2 side will count (i1,i2) exactly 4 times (the number of leaf nodes on the i1 side).</li>
</ul>
<p>For example, since ...</p>
<ul>
<li>v0 is on the i1 side, <code>edge_count(v0)</code> counts (i1,i2) 2 times.</li>
<li>v1 is on the i1 side, <code>edge_count(v1)</code> counts (i1,i2) 2 times.</li>
<li>v2 is on the i1 side, <code>edge_count(v2)</code> counts (i1,i2) 2 times.</li>
<li>v3 is on the i2 side, <code>edge_count(v3)</code> counts (i1,i2) 4 times.</li>
<li>v4 is on the i2 side, <code>edge_count(v4)</code> counts (i1,i2) 4 times.</li>
<li>v5 is on the i1 side, <code>edge_count(v5)</code> counts (i1,i2) 2 times.</li>
</ul>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">segregate_leaves</span>(<span class="hljs-params">self, internal_edge: E</span>) -&gt; <span class="hljs-built_in">dict</span>[N, N]:
    leaf_to_end = {}  <span class="hljs-comment"># leaf -&gt; one of the ends of internal_edge</span>
    e1, e2 = self.tree.get_edge_ends(internal_edge)
    <span class="hljs-keyword">for</span> l1 <span class="hljs-keyword">in</span> self.leaf_nodes:
        <span class="hljs-comment"># If path from l1 to e1 ends with internal_edge, it means that it had to</span>
        <span class="hljs-comment"># walk over the internal edge to get to e1, which ultimately means that l1</span>
        <span class="hljs-comment"># it isn't on the e1 side / it's on the e2 side. Otherwise, it's on the e1</span>
        <span class="hljs-comment"># side.</span>
        p = self.path(l1, e1)
        <span class="hljs-keyword">if</span> p[-<span class="hljs-number">1</span>] != internal_edge:
            leaf_to_end[l1] = e1
        <span class="hljs-keyword">else</span>:
            leaf_to_end[l1] = e2
    <span class="hljs-keyword">return</span> leaf_to_end
</code></pre>
<p>If the chosen pair are on opposite sides, <code>combine_edge_count()</code> will count (i1,i2) 6 times, which is the same number of times that the chosen pair's <a href="#BM_(limb)%2Fi">limb</a>s get counted (the number of leaf nodes in the <a href="#BM_(tree)%2Fi">tree</a>). For example, <code>combine_edge_count(v1,v3)</code> counts (i1,i2) 6 times, because v1 sits on the i1 side (adds 2 to the count) and v3 sits on the i2 side (adds 4 to the count)...</p>
<p><img src="dot_8040d6d051a27cf0ce53ee8c9254d562.svg" alt="Dot diagram"></p>
<table>
<thead>
<tr>
<th></th>
<th>(i0,i1)</th>
<th>(i1,i2)</th>
<th>(v0,i0)</th>
<th>(v1,i0)</th>
<th>(v2,i0)</th>
<th>(v3,i2)</th>
<th>(v4,i2)</th>
<th>(v5,i1)</th>
</tr>
</thead>
<tbody>
<tr>
<td>edge_count(v1)</td>
<td>3</td>
<td>2</td>
<td>1</td>
<td>5</td>
<td>1</td>
<td>1</td>
<td>1</td>
<td>1</td>
</tr>
<tr>
<td>edge_count(v3)</td>
<td>3</td>
<td>4</td>
<td>1</td>
<td>1</td>
<td>1</td>
<td>5</td>
<td>1</td>
<td>1</td>
</tr>
<tr>
<td></td>
<td>-------</td>
<td>-------</td>
<td>-------</td>
<td>-------</td>
<td>-------</td>
<td>-------</td>
<td>-------</td>
<td>-------</td>
</tr>
<tr>
<td></td>
<td>6</td>
<td>6</td>
<td>2</td>
<td>6</td>
<td>2</td>
<td>6</td>
<td>2</td>
<td>2</td>
</tr>
</tbody>
</table>
<p>This will always be the case for any <a href="#BM_(simple%20tree)%2Fi">simple tree</a>: If a chosen pair aren't <a href="#BM_(neighbour)%2Fi">neighbour</a>s, the path between them always travels over at least one internal edge. <code>combine_edge_count()</code> will always count each edge in the path <code>leaf_count</code> times. In the above example, path(v1,v3) travels over internal edges (i0,i1) and (i1,i2) and as such both those edges in addition to the <a href="#BM_(limb)%2Fi">limb</a>s of v1 and v3 have a count of 6.</p>
<p>Just like how <code>combine_edge_count_and_normalize()</code> reduces the counts of the chosen pair's <a href="#BM_(limb)%2Fi">limb</a>s to 2, so will it reduce the count of the internal edges in the path of the chosen pair to 2. That is, all edges in the path between the chosen pair get reduced to a count of 2.</p>
<p>For example, path(v1,v3) has the edges [(v1,i0), (i0,i1), (i1, i2), (v3, i2)]. <code>combine_edge_count_and_normalize(v1,v3)</code> reduces the count of each edge in that path to 2 ...</p>
<p><img src="dot_361948ed39492a22bb69e4393387ba39.svg" alt="Dot diagram"></p>
<table>
<thead>
<tr>
<th></th>
<th>(i0,i1)</th>
<th>(i1,i2)</th>
<th>(v0,i0)</th>
<th>(v1,i0)</th>
<th>(v2,i0)</th>
<th>(v3,i2)</th>
<th>(v4,i2)</th>
<th>(v5,i1)</th>
</tr>
</thead>
<tbody>
<tr>
<td>edge_count(v1)</td>
<td>3</td>
<td>2</td>
<td>1</td>
<td>5</td>
<td>1</td>
<td>1</td>
<td>1</td>
<td>1</td>
</tr>
<tr>
<td>edge_count(v3)</td>
<td>3</td>
<td>4</td>
<td>1</td>
<td>1</td>
<td>1</td>
<td>5</td>
<td>1</td>
<td>1</td>
</tr>
<tr>
<td>-4 * path(v1,v3)</td>
<td>-4</td>
<td>-4</td>
<td></td>
<td>-4</td>
<td></td>
<td>-4</td>
<td></td>
<td></td>
</tr>
<tr>
<td></td>
<td>-------</td>
<td>-------</td>
<td>-------</td>
<td>-------</td>
<td>-------</td>
<td>-------</td>
<td>-------</td>
<td>-------</td>
</tr>
<tr>
<td></td>
<td>2</td>
<td>2</td>
<td>2</td>
<td>2</td>
<td>2</td>
<td>2</td>
<td>2</td>
<td>2</td>
</tr>
</tbody>
</table>
<p>The ultimate idea is that, for any leaf node pair in a <a href="#BM_(simple%20tree)%2Fi">simple tree</a>, <code>combine_edge_count_and_normalize()</code> will have a count of ...</p>
<ul>
<li>2 for <a href="#BM_(limb)%2Fi">limb</a>s.</li>
<li>&gt; 2 for internal edges.</li>
</ul>
<p>In other words, internal edges are the only differentiating factor in <code>combine_edge_count_and_normalize()</code>'s result. Non-<a href="#BM_(neighbour)%2Fi">neighbour</a>ing pairs will have certain internal edge counts reduced to 2 while <a href="#BM_(neighbour)%2Fi">neighbour</a>ing pairs keep internal edge counts &gt; 2. In a ...</p>
<ul>
<li>worst case scenario, all internal edges get reduced to 2.</li>
<li>best case scenario, all internal edges are kept &gt; 2.</li>
</ul>
<p>The pair with the highest total count is guaranteed to be a <a href="#BM_(neighbour)%2Fi">neighbour</a>ing pair because lesser total counts may have had their internal edges reduced.</p>
<p><a href="ch7_code/src/phylogeny/NeighbourJoiningMatrix_EdgeCountExplainer.py">ch7_code/src/phylogeny/NeighbourJoiningMatrix_EdgeCountExplainer.py</a> (lines 126 to 136):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">neighbour_detect</span>(<span class="hljs-params">self</span>) -&gt; <span class="hljs-built_in">tuple</span>[<span class="hljs-built_in">int</span>, <span class="hljs-built_in">tuple</span>[N, N]]:
    found_pair = <span class="hljs-literal">None</span>
    found_total_count = -<span class="hljs-number">1</span>
    <span class="hljs-keyword">for</span> l1, l2 <span class="hljs-keyword">in</span> combinations(self.leaf_nodes, r=<span class="hljs-number">2</span>):
        normalized_counts = self.combine_edge_count_and_normalize(l1, l2)
        total_count = <span class="hljs-built_in">sum</span>(c <span class="hljs-keyword">for</span> c <span class="hljs-keyword">in</span> normalized_counts.values())
        <span class="hljs-keyword">if</span> total_count &gt; found_total_count:
            found_pair = l1, l2
            found_total_count = total_count
    <span class="hljs-keyword">return</span> found_total_count, found_pair
</code></pre>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The graph in the example run below is the same as the graph used above. It may look different because node positions may have shifted around.</p>
</div>
<div style="border:1px solid black;">
<p></p>
<p>Given the tree...</p>
<p><img src="dot_23ac97214b740d5a897adafcbfec4be2.svg" alt="Dot diagram"></p>
<p>neighbour_detect reported that v4 and v3 have the highest total edge count of 26 and as such are guaranteed to be neighbours.</p>
<p>For each leaf pair in the tree, <code>combine_count_and_normalize()</code> totals are ...</p>
<table>
<thead><tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr></thead>
<tbody>
<tr>
<td>v0</td>
<td>0</td>
<td>22</td>
<td>22</td>
<td>16</td>
<td>16</td>
<td>18</td>
</tr>
<tr>
<td>v1</td>
<td>22</td>
<td>0</td>
<td>22</td>
<td>16</td>
<td>16</td>
<td>18</td>
</tr>
<tr>
<td>v2</td>
<td>22</td>
<td>22</td>
<td>0</td>
<td>16</td>
<td>16</td>
<td>18</td>
</tr>
<tr>
<td>v3</td>
<td>16</td>
<td>16</td>
<td>16</td>
<td>0</td>
<td>26</td>
<td>20</td>
</tr>
<tr>
<td>v4</td>
<td>16</td>
<td>16</td>
<td>16</td>
<td>26</td>
<td>0</td>
<td>20</td>
</tr>
<tr>
<td>v5</td>
<td>18</td>
<td>18</td>
<td>18</td>
<td>20</td>
<td>20</td>
<td>0</td>
</tr>
</tbody>
</table>
<p><img src="dot_a8144dba9b116da1f435fa70958c182b.svg" alt="Dot diagram"></p>
</div>
<p></p>
<p>This same reasoning is applied to edge weights. That is, instead of just counting edges, the reasoning works the same if you were to multiply edge weights by those counts.</p>
<p>In the edge count version of this algorithm, <code>edge_count()</code> gets the paths from a leaf node to all other leaf nodes and counts up the number of times each edge is encountered. In the edge weight multiplicity version, instead of counting how many times each edge gets encountered, each time an edge gets encountered it increases the multiplicity of its weight ...</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">edge_multiple</span>(<span class="hljs-params">self, l1: N</span>) -&gt; Counter[E]:
    <span class="hljs-comment"># Collect paths from l1 to all other leaf nodes</span>
    path_collection = []
    <span class="hljs-keyword">for</span> l2 <span class="hljs-keyword">in</span> self.leaf_nodes:
        <span class="hljs-keyword">if</span> l1 == l2:
            <span class="hljs-keyword">continue</span>
        path = self.path(l1, l2)
        path_collection.append(path)
    <span class="hljs-comment"># Sum edge weights across all paths</span>
    edge_weight_sums = Counter()
    <span class="hljs-keyword">for</span> path <span class="hljs-keyword">in</span> path_collection:
        <span class="hljs-keyword">for</span> edge <span class="hljs-keyword">in</span> path:
            edge_weight_sums[edge] += self.tree.get_edge_data(edge)
    <span class="hljs-comment"># Return edge weight sums</span>
    <span class="hljs-keyword">return</span> edge_weight_sums
</code></pre>
<p><img src="dot_f8a0c44f51d391e609208c8102149f44.svg" alt="Dot diagram"></p>
<table>
<thead>
<tr>
<th></th>
<th>(i0,i1)</th>
<th>(i1,i2)</th>
<th>(v0,i0)</th>
<th>(v1,i0)</th>
<th>(v2,i0)</th>
<th>(v3,i2)</th>
<th>(v4,i2)</th>
<th>(v5,i1)</th>
</tr>
</thead>
<tbody>
<tr>
<td>edge_count(v1)</td>
<td>3</td>
<td>2</td>
<td>1</td>
<td>5</td>
<td>1</td>
<td>1</td>
<td>1</td>
<td>1</td>
</tr>
<tr>
<td>edge_multiple(v1)</td>
<td>3*4=12</td>
<td>2*3=6</td>
<td>1*11=11</td>
<td>5*2=10</td>
<td>1*10=10</td>
<td>1*3=3</td>
<td>1*4=4</td>
<td>1*7=7</td>
</tr>
</tbody>
</table>
<p>Similarly, where in the edge count version <code>combine_edge_count()</code> adds together the <code>edge_count()</code>s for two leaf nodes, the edge weight multiplicity version should add together the <code>edge_multiple()</code>s for two leaf nodes instead...</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">combine_edge_multiple</span>(<span class="hljs-params">self, l1: N, l2: N</span>) -&gt; Counter[E]:
    c1 = self.edge_multiple(l1)
    c2 = self.edge_multiple(l2)
    <span class="hljs-keyword">return</span> c1 + c2
</code></pre>
<p><img src="dot_cfaace9a1489262016ebdce7c5a60c33.svg" alt="Dot diagram"></p>
<table>
<thead>
<tr>
<th></th>
<th>(i0,i1)</th>
<th>(i1,i2)</th>
<th>(v0,i0)</th>
<th>(v1,i0)</th>
<th>(v2,i0)</th>
<th>(v3,i2)</th>
<th>(v4,i2)</th>
<th>(v5,i1)</th>
</tr>
</thead>
<tbody>
<tr>
<td>combine_edge_count(v1)</td>
<td>6</td>
<td>4</td>
<td>2</td>
<td>6</td>
<td>6</td>
<td>2</td>
<td>2</td>
<td>2</td>
</tr>
<tr>
<td>combine_edge_multiple(v1)</td>
<td>6*4=24</td>
<td>4*3=12</td>
<td>2*11=22</td>
<td>6*2=20</td>
<td>6*10=60</td>
<td>2*3=6</td>
<td>2*4=8</td>
<td>2*7=14</td>
</tr>
</tbody>
</table>
<p>Similarly, where in the edge count version <code>combine_edge_count_and_normalize()</code> reduces all <a href="#BM_(limb)%2Fi">limb</a>s and possibly some internal edges from <code>combine_edge_count()</code> to a count of 2, the edge multiplicity version reduces weights for those same <a href="#BM_(limb)%2Fi">limb</a>s and edges to a multiple of 2...</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">combine_edge_multiple_and_normalize</span>(<span class="hljs-params">self, l1: N, l2: N</span>) -&gt; Counter[E]:
    edge_multiples = self.combine_edge_multiple(l1, l2)
    path_edges = self.path(l1, l2)
    <span class="hljs-keyword">for</span> e <span class="hljs-keyword">in</span> path_edges:
        edge_multiples[e] -= (self.leaf_count - <span class="hljs-number">2</span>) * self.tree.get_edge_data(e)
    <span class="hljs-keyword">return</span> edge_multiples
</code></pre>
<p><img src="dot_b88b756b70453fbaa4fe6f126b97cd97.svg" alt="Dot diagram"></p>
<table>
<thead>
<tr>
<th></th>
<th>(i0,i1)</th>
<th>(i1,i2)</th>
<th>(v0,i0)</th>
<th>(v1,i0)</th>
<th>(v2,i0)</th>
<th>(v3,i2)</th>
<th>(v4,i2)</th>
<th>(v5,i1)</th>
</tr>
</thead>
<tbody>
<tr>
<td>combine_edge_count_and_normalize(v1,v2)</td>
<td>6</td>
<td>4</td>
<td>2</td>
<td>2</td>
<td>2</td>
<td>2</td>
<td>2</td>
<td>2</td>
</tr>
<tr>
<td>combine_edge_multiple_and_normalize(v1,v2)</td>
<td>6*4=24</td>
<td>4*3=12</td>
<td>2*11=22</td>
<td>2*2=20</td>
<td>2*10=60</td>
<td>2*3=6</td>
<td>2*4=8</td>
<td>2*7=14</td>
</tr>
</tbody>
</table>
<p>Similar to <code>combine_edge_count_and_normalize()</code>, for any leaf node pair in a <a href="#BM_(simple%20tree)%2Fi">simple tree</a> <code>combine_edge_multiple_and_normalize()</code> will have an <em>edge weight multiple</em> of ...</p>
<ul>
<li>2 for <a href="#BM_(limb)%2Fi">limb</a>s.</li>
<li>&gt; 2 for internal edges.</li>
</ul>
<p>In other words, internal edge weight multiples are the only differentiating factor in <code>combine_edge_multiple_and_normalize()</code>'s result. Non-<a href="#BM_(neighbour)%2Fi">neighbour</a>ing pairs will have certain internal edge weight multiples reduced to 2 while <a href="#BM_(neighbour)%2Fi">neighbour</a>ing pairs keep internal edge weight multiples &gt; 2. In a ...</p>
<ul>
<li>worst case scenario, all internal edge weight multiples get reduced to 2.</li>
<li>best case scenario, all internal edge weight multiples are kept &gt; 2.</li>
</ul>
<p>The pair with the highest combined multiple is guaranteed to be a <a href="#BM_(neighbour)%2Fi">neighbour</a>ing pair because lesser combined multiples may have had their internal edge multiples reduced.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>Still confused?</p>
<p>Given a <a href="#BM_(simple%20tree)%2Fi">simple tree</a>, <code>combine_edge_multiple(A, B)</code> will make it so that...</p>
<ul>
<li><a href="#BM_(limb)%2Fi">limb</a> A has a weight multiplicity of <code>leaf_count</code>.</li>
<li><a href="#BM_(limb)%2Fi">limb</a> B has a weight multiplicity of <code>leaf_count</code>.</li>
<li>other <a href="#BM_(limb)%2Fi">limb</a>s each have a weight multiplicity of 2.</li>
<li>internal edges each have a weight multiplicity of &gt; 2.</li>
</ul>
<p>For example, the following diagrams visualize edge weight multiplicities produced by <code>combine_edge_multiple()</code> for various pairs in a 4 leaf <a href="#BM_(simple%20tree)%2Fi">simple tree</a>. Note how the selected pair's <a href="#BM_(limb)%2Fi">limb</a>s have a multiplicity of 4, other <a href="#BM_(limb)%2Fi">limb</a>s have a multiplicity of 2, and internal edges have a multiplicity of 4...</p>
<p><img src="dot_449bba2253f538278bfaad93a18d5cf0.svg" alt="Dot diagram"></p>
<p><code>combine_edge_multiple_and_normalize(A, B)</code> normalizes these multiplicities such that ...</p>
<ul>
<li><a href="#BM_(limb)%2Fi">limb</a> A's weight multiplicity reduces to 2.</li>
<li><a href="#BM_(limb)%2Fi">limb</a> B's weight multiplicity reduces to 2.</li>
<li>other <a href="#BM_(limb)%2Fi">limb</a>s keep their weight multiplicities at 2.</li>
<li>if the pair are <a href="#BM_(neighbour)%2Fi">neighbour</a>s, each internal edge multiplicity remains at &gt; 2.</li>
<li>if the pair aren't <a href="#BM_(neighbour)%2Fi">neighbour</a>s, at least one internal edge multiplicity reduces to 2 while others remain at &gt; 2.</li>
</ul>
<table>
<thead>
<tr>
<th></th>
<th><a href="#BM_(limb)%2Fi">limb</a> multiplicity</th>
<th>internal edge multiplicity</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="#BM_(neighbour)%2Fi">neighbour</a>ing pairs</td>
<td>all = 2</td>
<td>all &gt;= 2</td>
</tr>
<tr>
<td>non-<a href="#BM_(neighbour)%2Fi">neighbour</a>ing pairs</td>
<td>all = 2</td>
<td>at least one = 2, others &gt; 2</td>
</tr>
</tbody>
</table>
<p>Since <a href="#BM_(limb)%2Fi">limb</a>s always contribute the same regardless of whether the pair is <a href="#BM_(neighbour)%2Fi">neighbour</a>ing or not (2*weight), they can be ignored. That leaves internal edge contributions as the only thing differentiating between <a href="#BM_(neighbour)%2Fi">neighbour</a>ing and non-<a href="#BM_(neighbour)%2Fi">neighbour</a>ing pairs.</p>
<p>A <a href="#BM_(simple%20tree)%2Fi">simple tree</a> with 2 or more leaf nodes is guaranteed to have at least 1 <a href="#BM_(neighbour)%2Fi">neighbour</a>ing pair. The pair producing the largest result is the one with maxed out contributions from its multiplied internal edges weights, meaning that none of those contributions were for internal edges reduced to 2*weight. Lesser results MAY be lesser because normalization reduced some of their internal edge weights to 2*weight, but the largest result you know for certain has all of its internal edge weights &gt; 2*weight.</p>
</div>
<p><a href="ch7_code/src/phylogeny/NeighbourJoiningMatrix_EdgeMultiplicityExplainer.py">ch7_code/src/phylogeny/NeighbourJoiningMatrix_EdgeMultiplicityExplainer.py</a> (lines 97 to 107):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">neighbour_detect</span>(<span class="hljs-params">self</span>) -&gt; <span class="hljs-built_in">tuple</span>[<span class="hljs-built_in">int</span>, <span class="hljs-built_in">tuple</span>[N, N]]:
    found_pair = <span class="hljs-literal">None</span>
    found_total_count = -<span class="hljs-number">1</span>
    <span class="hljs-keyword">for</span> l1, l2 <span class="hljs-keyword">in</span> combinations(self.leaf_nodes, r=<span class="hljs-number">2</span>):
        normalized_counts = self.combine_edge_multiple_and_normalize(l1, l2)
        total_count = <span class="hljs-built_in">sum</span>(c <span class="hljs-keyword">for</span> c <span class="hljs-keyword">in</span> normalized_counts.values())
        <span class="hljs-keyword">if</span> total_count &gt; found_total_count:
            found_pair = l1, l2
            found_total_count = total_count
    <span class="hljs-keyword">return</span> found_total_count, found_pair
</code></pre>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The graph in the example run below is the same as the graph used above. It may look different because node positions may have shifted around.</p>
</div>
<div style="border:1px solid black;">
<p></p>
<p>Given the tree...</p>
<p><img src="dot_7ed965464ba60a747b81f636f4046bfa.svg" alt="Dot diagram"></p>
<p>neighbour_detect reported that v3 and v4 have the highest total edge sum of 122 and as such are guaranteed to be neighbours.</p>
<p>For each leaf pair in the tree, <code>combine_count_and_normalize()</code> totals are ...</p>
<table>
<thead><tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr></thead>
<tbody>
<tr>
<td>v0</td>
<td>0</td>
<td>110</td>
<td>110</td>
<td>88</td>
<td>88</td>
<td>94</td>
</tr>
<tr>
<td>v1</td>
<td>110</td>
<td>0</td>
<td>110</td>
<td>88</td>
<td>88</td>
<td>94</td>
</tr>
<tr>
<td>v2</td>
<td>110</td>
<td>110</td>
<td>0</td>
<td>88</td>
<td>88</td>
<td>94</td>
</tr>
<tr>
<td>v3</td>
<td>88</td>
<td>88</td>
<td>88</td>
<td>0</td>
<td>122</td>
<td>104</td>
</tr>
<tr>
<td>v4</td>
<td>88</td>
<td>88</td>
<td>88</td>
<td>122</td>
<td>0</td>
<td>104</td>
</tr>
<tr>
<td>v5</td>
<td>94</td>
<td>94</td>
<td>94</td>
<td>104</td>
<td>104</td>
<td>0</td>
</tr>
</tbody>
</table>
<p><img src="dot_5bd0f8447524227f21b79b84cb7cf16d.svg" alt="Dot diagram"></p>
</div>
<p></p>
<p>The matrix produced in the example above is called a <a href="#BM_(neighbour%20joining%20matrix)%2Fi">neighbour joining matrix</a>. The summation of <code>combine_edge_multiple_and_normalize()</code> performed in each matrix slot is rewritable as a set of addition and subtraction operations between leaf node distances. For example, recall that <code>combine_edge_multiple_and_normalize(v1,v2)</code> in the example graph breaks down to <code>edge_multiple(v1) + edge_multiple(v2) - (leaf_count - 2) * path(v1,v2)</code>. The sum of ...</p>
<ul>
<li>
<p><code>edge_multiple(v1)</code> breaks down to...</p>
<pre><code>dist(v1,v0) + dist(v1,v2) + dist(v1,v3) + dist(v1,v4) + dist(v1,v5)
</code></pre>
</li>
<li>
<p><code>edge_multiple(v2)</code> breaks down to...</p>
<pre><code>dist(v2,v0) + dist(v2,v1) + dist(v2,v3) + dist(v2,v4) + dist(v2,v5)
</code></pre>
</li>
<li>
<p><code>combine_edge_multiple(v2)</code> is simply the sum of the two summations above:</p>
<pre><code>dist(v1,v0) + dist(v1,v2) + dist(v1,v3) + dist(v1,v4) + dist(v1,v5) +
dist(v2,v0) + dist(v2,v1) + dist(v2,v3) + dist(v2,v4) + dist(v2,v5)
</code></pre>
</li>
<li>
<p><code>combine_edge_multiple_and_normalize(v1,v2)</code> is simply the above summation but with <code>dist(v1,v2)</code> removed 4 times:</p>
</li>
</ul>
<pre><code>dist(v1,v0) + dist(v1,v2) + dist(v1,v3) + dist(v1,v4) + dist(v1,v5) +
dist(v2,v0) + dist(v2,v1) + dist(v2,v3) + dist(v2,v4) + dist(v2,v5) -
dist(v1,v2) - dist(v1,v2) - dist(v1,v2) - dist(v1,v2)
</code></pre>
<p>Since only leaf node distances are being used in the summation calculation, a <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> suffices as the input. The actual <a href="#BM_(simple%20tree)%2Fi">simple tree</a> isn't required.</p>
<p><a href="ch7_code/src/phylogeny/NeighbourJoiningMatrix.py">ch7_code/src/phylogeny/NeighbourJoiningMatrix.py</a> (lines 21 to 49):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">total_distance</span>(<span class="hljs-params">dist_mat: DistanceMatrix[N]</span>) -&gt; <span class="hljs-built_in">dict</span>[N, <span class="hljs-built_in">float</span>]:
    ret = {}
    <span class="hljs-keyword">for</span> l1 <span class="hljs-keyword">in</span> dist_mat.leaf_ids():
        ret[l1] = <span class="hljs-built_in">sum</span>(dist_mat[l1, l2] <span class="hljs-keyword">for</span> l2 <span class="hljs-keyword">in</span> dist_mat.leaf_ids())
    <span class="hljs-keyword">return</span> ret


<span class="hljs-keyword">def</span> <span class="hljs-title function_">neighbour_joining_matrix</span>(<span class="hljs-params">dist_mat: DistanceMatrix[N]</span>) -&gt; DistanceMatrix[N]:
    tot_dists = total_distance(dist_mat)
    n = dist_mat.n
    ret = dist_mat.copy()
    <span class="hljs-keyword">for</span> l1, l2 <span class="hljs-keyword">in</span> product(dist_mat.leaf_ids(), repeat=<span class="hljs-number">2</span>):
        <span class="hljs-keyword">if</span> l1 == l2:
            <span class="hljs-keyword">continue</span>
        ret[l1, l2] = tot_dists[l1] + tot_dists[l2] - (n - <span class="hljs-number">2</span>) * dist_mat[l1, l2]
    <span class="hljs-keyword">return</span> ret


<span class="hljs-keyword">def</span> <span class="hljs-title function_">find_neighbours</span>(<span class="hljs-params">dist_mat: DistanceMatrix[N]</span>) -&gt; <span class="hljs-built_in">tuple</span>[N, N]:
    nj_mat = neighbour_joining_matrix(dist_mat)
    found_pair = <span class="hljs-literal">None</span>
    found_nj_val = -<span class="hljs-number">1</span>
    <span class="hljs-keyword">for</span> l1, l2 <span class="hljs-keyword">in</span> product(nj_mat.leaf_ids_it(), repeat=<span class="hljs-number">2</span>):
        <span class="hljs-keyword">if</span> nj_mat[l1, l2] &gt; found_nj_val:
            found_pair = l1, l2
            found_nj_val = nj_mat[l1, l2]
    <span class="hljs-keyword">assert</span> found_pair <span class="hljs-keyword">is</span> <span class="hljs-keyword">not</span> <span class="hljs-literal">None</span>
    <span class="hljs-keyword">return</span> found_pair
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the following distance matrix...</p>
<table>
<thead><tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr></thead>
<tbody>
<tr>
<td>v0</td>
<td>0.0</td>
<td>13.0</td>
<td>21.0</td>
<td>21.0</td>
<td>22.0</td>
<td>22.0</td>
</tr>
<tr>
<td>v1</td>
<td>13.0</td>
<td>0.0</td>
<td>12.0</td>
<td>12.0</td>
<td>13.0</td>
<td>13.0</td>
</tr>
<tr>
<td>v2</td>
<td>21.0</td>
<td>12.0</td>
<td>0.0</td>
<td>20.0</td>
<td>21.0</td>
<td>21.0</td>
</tr>
<tr>
<td>v3</td>
<td>21.0</td>
<td>12.0</td>
<td>20.0</td>
<td>0.0</td>
<td>7.0</td>
<td>13.0</td>
</tr>
<tr>
<td>v4</td>
<td>22.0</td>
<td>13.0</td>
<td>21.0</td>
<td>7.0</td>
<td>0.0</td>
<td>14.0</td>
</tr>
<tr>
<td>v5</td>
<td>22.0</td>
<td>13.0</td>
<td>21.0</td>
<td>13.0</td>
<td>14.0</td>
<td>0.0</td>
</tr>
</tbody>
</table>
<p>... the neighbour joining matrix is ...</p>
<table>
<thead><tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr></thead>
<tbody>
<tr>
<td>v0</td>
<td>0.0</td>
<td>110.0</td>
<td>110.0</td>
<td>88.0</td>
<td>88.0</td>
<td>94.0</td>
</tr>
<tr>
<td>v1</td>
<td>110.0</td>
<td>0.0</td>
<td>110.0</td>
<td>88.0</td>
<td>88.0</td>
<td>94.0</td>
</tr>
<tr>
<td>v2</td>
<td>110.0</td>
<td>110.0</td>
<td>0.0</td>
<td>88.0</td>
<td>88.0</td>
<td>94.0</td>
</tr>
<tr>
<td>v3</td>
<td>88.0</td>
<td>88.0</td>
<td>88.0</td>
<td>0.0</td>
<td>122.0</td>
<td>104.0</td>
</tr>
<tr>
<td>v4</td>
<td>88.0</td>
<td>88.0</td>
<td>88.0</td>
<td>122.0</td>
<td>0.0</td>
<td>104.0</td>
</tr>
<tr>
<td>v5</td>
<td>94.0</td>
<td>94.0</td>
<td>94.0</td>
<td>104.0</td>
<td>104.0</td>
<td>0.0</td>
</tr>
</tbody>
</table>
</div>
<p></p>
<a name="H_Find%20Neighbour%20Limb%20Lengths"></a><h3>Find Neighbour Limb Lengths</h3>
<p><a name="BM_(Algorithms%5C%2FPhylogeny%2FFind%20Neighbour%20Limb%20Lengths)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FFind%20Limb%20Length)_TOPIC%2F">Algorithms/Phylogeny/Find Limb Length</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: Given a <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> and a pair of leaf nodes identified as being <a href="#BM_(neighbour)%2Fi">neighbour</a>s, if the <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> is ...</p>
<ul>
<li>an <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>, this algorithm finds the <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a>s of those <a href="#BM_(neighbour)%2Fi">neighbour</a>s.</li>
<li>a non-<a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> (but close to being additive), this algorithm approximates the <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a>s of those <a href="#BM_(neighbour)%2Fi">neighbour</a>s.</li>
</ul>
<p><strong>WHY</strong>: This operation is required for <em>approximating</em> a <a href="#BM_(simple%20tree)%2Fi">simple tree</a> for a non-<a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>.</p>
<p>Recall that the standard <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a> finding algorithm determines the <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a> of L by testing distances between leaf nodes to deduce a pair whose path crosses over L's parent. That won't work here because non-<a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrices</a> have inconsistent distances -- non-additive means no <a href="#BM_(tree)%2Fi">tree</a> exists that fits its distances.</p>
<a name="H_Average%20Algorithm"></a><h4>Average Algorithm</h4>
<p><a name="BM_(Algorithms%5C%2FPhylogeny%2FFind%20Neighbour%20Limb%20Lengths%2FAverage%20Algorithm)_TOPIC%2F"></a><strong></strong></p>
<p><strong>ALGORITHM</strong>:</p>
<p>The algorithm is an extension of the standard <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a> finding algorithm, essentially running the same computation multiple times and averaging out the results. For example, v1 and v2 are <a href="#BM_(neighbour)%2Fi">neighbour</a>s in the following <a href="#BM_(simple%20tree)%2Fi">simple tree</a>...</p>
<p><img src="dot_fa269cbcc76261c503f77dc3a6e5b8b6.svg" alt="Dot diagram"></p>
<p>Since they're <a href="#BM_(neighbour)%2Fi">neighbour</a>s, they share the same parent node, meaning that the path from...</p>
<ul>
<li>v1 to any other leaf node travels over v2's parent.</li>
<li>v2 to any other leaf node travels over v1's parent.</li>
</ul>
<p><img src="dot_733281bf7d0b32a752b24ae8e837ae49.svg" alt="Dot diagram"></p>
<p>Recall that to find the <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a> for L, the standard <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a> algorithm had to perform a minimum test to find a pair of leaf nodes whose path travelled over the L's parent. Since this algorithm takes in two <em><a href="#BM_(neighbour)%2Fi">neighbour</a>ing</em> leaf nodes, that test isn't required here. The path from L's <a href="#BM_(neighbour)%2Fi">neighbour</a> to every other node always travels over L's parent.</p>
<p>Since the path from L's <a href="#BM_(neighbour)%2Fi">neighbour</a> to every other node always travels over L's parent, the core computation from the standard algorithm is performed multiple times and averaged to produce an approximate <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a>: 0.5 * (dist(L,N) + dist(L,X) - dist(N,X)),  where ...</p>
<ul>
<li>N is L's <a href="#BM_(neighbour)%2Fi">neighbour</a>.</li>
<li>X is a leaf node that isn't L or N.</li>
</ul>
<p>The averaging makes it so that if the input <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> were ...</p>
<ul>
<li>additive, it'd produce the correct <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a>.</li>
<li>non-additive, it'd approximate a <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a> that's probably good enough (assuming the <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> is close to being additive).</li>
</ul>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>Still confused? Think about it like this: When the <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> is non-additive, each X has a different "view" of what the <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a> should be. You're averaging their views to get a single <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a> value.</p>
</div>
<p><a href="ch7_code/src/phylogeny/FindNeighbourLimbLengths.py">ch7_code/src/phylogeny/FindNeighbourLimbLengths.py</a> (lines 21 to 40):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">view_of_limb_length_using_neighbour</span>(<span class="hljs-params">dm: DistanceMatrix[N], l: N, l_neighbour: N, l_from: N</span>) -&gt; <span class="hljs-built_in">float</span>:
    <span class="hljs-keyword">return</span> (dm[l, l_neighbour] + dm[l, l_from] - dm[l_neighbour, l_from]) / <span class="hljs-number">2</span>


<span class="hljs-keyword">def</span> <span class="hljs-title function_">approximate_limb_length_using_neighbour</span>(<span class="hljs-params">dm: DistanceMatrix[N], l: N, l_neighbour: N</span>) -&gt; <span class="hljs-built_in">float</span>:
    leaf_nodes = dm.leaf_ids()
    leaf_nodes.remove(l)
    leaf_nodes.remove(l_neighbour)
    lengths = []
    <span class="hljs-keyword">for</span> l_from <span class="hljs-keyword">in</span> leaf_nodes:
        length = view_of_limb_length_using_neighbour(dm, l, l_neighbour, l_from)
        lengths.append(length)
    <span class="hljs-keyword">return</span> mean(lengths)


<span class="hljs-keyword">def</span> <span class="hljs-title function_">find_neighbouring_limb_lengths</span>(<span class="hljs-params">dm: DistanceMatrix[N], l1: N, l2: N</span>) -&gt; <span class="hljs-built_in">tuple</span>[<span class="hljs-built_in">float</span>, <span class="hljs-built_in">float</span>]:
    l1_len = approximate_limb_length_using_neighbour(dm, l1, l2)
    l2_len = approximate_limb_length_using_neighbour(dm, l2, l1)
    <span class="hljs-keyword">return</span> l1_len, l2_len
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given distance matrix...</p>
<table>
<thead><tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr></thead>
<tbody>
<tr>
<td>v0</td>
<td>0.0</td>
<td>13.0</td>
<td>21.0</td>
<td>21.0</td>
<td>22.0</td>
<td>22.0</td>
</tr>
<tr>
<td>v1</td>
<td>13.0</td>
<td>0.0</td>
<td>12.0</td>
<td>12.0</td>
<td>13.0</td>
<td>13.0</td>
</tr>
<tr>
<td>v2</td>
<td>21.0</td>
<td>12.0</td>
<td>0.0</td>
<td>20.0</td>
<td>21.0</td>
<td>21.0</td>
</tr>
<tr>
<td>v3</td>
<td>21.0</td>
<td>12.0</td>
<td>20.0</td>
<td>0.0</td>
<td>7.0</td>
<td>13.0</td>
</tr>
<tr>
<td>v4</td>
<td>22.0</td>
<td>13.0</td>
<td>21.0</td>
<td>7.0</td>
<td>0.0</td>
<td>14.0</td>
</tr>
<tr>
<td>v5</td>
<td>22.0</td>
<td>13.0</td>
<td>21.0</td>
<td>13.0</td>
<td>14.0</td>
<td>0.0</td>
</tr>
</tbody>
</table>
<p>... and given that v1 and v2 are neighbours, the limb length for leaf node ...</p>
<ul>
<li>v1 is approximated to be 2.0</li>
<li>v2 is approximated to be 10.0</li>
</ul>
</div>
<p></p>
<a name="H_Optimized%20Average%20Algorithm"></a><h4>Optimized Average Algorithm</h4>
<p><a name="BM_(Algorithms%5C%2FPhylogeny%2FFind%20Neighbour%20Limb%20Lengths%2FOptimized%20Average%20Algorithm)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FFind%20Neighbour%20Limb%20Lengths%2FAverage%20Algorithm)_TOPIC%2F">Algorithms/Phylogeny/Find Neighbour Limb Lengths/Average Algorithm</a></li>
</ul>
</div>
<p><strong>ALGORITHM</strong>:</p>
<p>The unoptimized algorithm performs the computation once for each leaf node in the pair. This is inefficient in that it's repeating a lot of the same operations twice. This algorithm removes a lot of that duplicate work.</p>
<p>The unoptimized algorithm maps to the formula ...</p>
<div style="overflow: auto"><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><mfrac><mn>1</mn><mrow><mi>n</mi><mo>−</mo><mn>2</mn></mrow></mfrac><mo>⋅</mo><munder><mo>∑</mo><mrow><mi>k</mi><mo>∈</mo><mi>S</mi><mo>−</mo><mo stretchy="false">{</mo><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn><mo stretchy="false">}</mo></mrow></munder><mfrac><mrow><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn></mrow></msub><mo>+</mo><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub><mo>−</mo><msub><mi>D</mi><mrow><mi>l</mi><mn>2</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub></mrow><mn>2</mn></mfrac></mrow><annotation encoding="application/x-tex">\frac{1}{n-2} \cdot \sum_{k \isin S-\{l1,l2\}}{\frac{D_{l1,l2} + D_{l1,k} - D_{l2,k}}{2}}
</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:2.09077em;vertical-align:-0.7693300000000001em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.32144em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord mathnormal">n</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.7693300000000001em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.876335em;vertical-align:-1.516005em;"></span><span class="mop op-limits"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.050005em;"><span style="top:-1.808995em;margin-left:0em;"><span class="pstrut" style="height:3.05em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03148em;">k</span><span class="mrel mtight">∈</span><span class="mord mathnormal mtight" style="margin-right:0.05764em;">S</span><span class="mbin mtight">−</span><span class="mopen mtight">{</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span><span class="mclose mtight">}</span></span></span></span><span style="top:-3.0500049999999996em;"><span class="pstrut" style="height:3.05em;"></span><span><span class="mop op-symbol large-op">∑</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.516005em;"><span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord"><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.36033em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span></span></span></div><p>... where ...</p>
<ul>
<li>l1 and l2 are the <a href="#BM_(neighbour)%2Fi">neighbour</a>ing leaf nodes,</li>
<li>S is the set of leaf nodes in the <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a>,</li>
<li>n is the size of S.</li>
</ul>
<p>Just like the code, the formula removes l1 and l2 from the set of leaf nodes (S) for the average's summation. The number of leaf nodes (n) is subtracted by 2 for the average's division because l1 and l2 aren't included. To optimize, consider what happens when you re-organize the formula as follows...</p>
<ol>
<li>
<p>Break up the division in the summation...</p>
<div style="overflow: auto"><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><mfrac><mn>1</mn><mrow><mi>n</mi><mo>−</mo><mn>2</mn></mrow></mfrac><mo>⋅</mo><munder><mo>∑</mo><mrow><mi>k</mi><mo>∈</mo><mi>S</mi><mo>−</mo><mo stretchy="false">{</mo><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn><mo stretchy="false">}</mo></mrow></munder><mrow><mo stretchy="false">(</mo><mfrac><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn></mrow></msub><mn>2</mn></mfrac><mo>+</mo><mfrac><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub><mn>2</mn></mfrac><mo>−</mo><mfrac><msub><mi>D</mi><mrow><mi>l</mi><mn>2</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub><mn>2</mn></mfrac><mo stretchy="false">)</mo></mrow></mrow><annotation encoding="application/x-tex">\frac{1}{n-2} \cdot \sum_{k \isin S-\{l1,l2\}}{(\frac{D_{l1,l2}}{2} + \frac{D_{l1,k}}{2} - \frac{D_{l2,k}}{2})}
</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:2.09077em;vertical-align:-0.7693300000000001em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.32144em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord mathnormal">n</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.7693300000000001em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.876335em;vertical-align:-1.516005em;"></span><span class="mop op-limits"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.050005em;"><span style="top:-1.808995em;margin-left:0em;"><span class="pstrut" style="height:3.05em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03148em;">k</span><span class="mrel mtight">∈</span><span class="mord mathnormal mtight" style="margin-right:0.05764em;">S</span><span class="mbin mtight">−</span><span class="mopen mtight">{</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span><span class="mclose mtight">}</span></span></span></span><span style="top:-3.0500049999999996em;"><span class="pstrut" style="height:3.05em;"></span><span><span class="mop op-symbol large-op">∑</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.516005em;"><span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord"><span class="mopen">(</span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.36033em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.36033em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.36033em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mclose">)</span></span></span></span></span></span></div></li>
<li>
<p>Pull out <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mfrac><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn></mrow></msub><mn>2</mn></mfrac></mrow><annotation encoding="application/x-tex">\frac{D_{l1,l2}}{2}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.3364109999999998em;vertical-align:-0.345em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.9914109999999998em;"><span style="top:-2.6550000000000002em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">2</span></span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.51308em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3448em;"><span style="top:-2.3487714285714287em;margin-left:-0.02778em;margin-right:0.07142857142857144em;"><span class="pstrut" style="height:2.5em;"></span><span class="sizing reset-size3 size1 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.29011428571428566em;"><span></span></span></span></span></span></span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.345em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span> as a term of its own...</p>
<div style="overflow: auto"><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><mfrac><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn></mrow></msub><mn>2</mn></mfrac><mo>+</mo><mfrac><mn>1</mn><mrow><mi>n</mi><mo>−</mo><mn>2</mn></mrow></mfrac><mo>⋅</mo><munder><mo>∑</mo><mrow><mi>k</mi><mo>∈</mo><mi>S</mi><mo>−</mo><mo stretchy="false">{</mo><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn><mo stretchy="false">}</mo></mrow></munder><mrow><mo stretchy="false">(</mo><mfrac><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub><mn>2</mn></mfrac><mo>−</mo><mfrac><msub><mi>D</mi><mrow><mi>l</mi><mn>2</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub><mn>2</mn></mfrac><mo stretchy="false">)</mo></mrow></mrow><annotation encoding="application/x-tex">\frac{D_{l1,l2}}{2} + \frac{1}{n-2} \cdot \sum_{k \isin S-\{l1,l2\}}{(\frac{D_{l1,k}}{2} - \frac{D_{l2,k}}{2})}
</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:2.04633em;vertical-align:-0.686em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.36033em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.09077em;vertical-align:-0.7693300000000001em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.32144em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord mathnormal">n</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.7693300000000001em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.876335em;vertical-align:-1.516005em;"></span><span class="mop op-limits"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.050005em;"><span style="top:-1.808995em;margin-left:0em;"><span class="pstrut" style="height:3.05em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03148em;">k</span><span class="mrel mtight">∈</span><span class="mord mathnormal mtight" style="margin-right:0.05764em;">S</span><span class="mbin mtight">−</span><span class="mopen mtight">{</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span><span class="mclose mtight">}</span></span></span></span><span style="top:-3.0500049999999996em;"><span class="pstrut" style="height:3.05em;"></span><span><span class="mop op-symbol large-op">∑</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.516005em;"><span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord"><span class="mopen">(</span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.36033em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.36033em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mclose">)</span></span></span></span></span></span></div><div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>Confused about what's happening above? Think about it like this...</p>
<ul>
<li>mean([0+0.5, 0+1, 0+0.25]) = 0.833 = 0+mean([0.5, 1, 0.25])</li>
<li>mean([1+0.5, 1+1, 1+0.25]) = 1.833 = 1+mean([0.5, 1, 0.25])</li>
<li>mean([2+0.5, 2+1, 2+0.25]) = 2.833 = 2+mean([0.5, 1, 0.25])</li>
<li>mean([3+0.5, 3+1, 3+0.25]) = 3.833 = 3+mean([0.5, 1, 0.25])</li>
<li>...</li>
</ul>
<p>If you're including some constant amount for each element in the averaging, the result of the average will include that constant amount. In the case above, <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mfrac><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn></mrow></msub><mn>2</mn></mfrac></mrow><annotation encoding="application/x-tex">\frac{D_{l1,l2}}{2}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.3364109999999998em;vertical-align:-0.345em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.9914109999999998em;"><span style="top:-2.6550000000000002em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">2</span></span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.51308em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3448em;"><span style="top:-2.3487714285714287em;margin-left:-0.02778em;margin-right:0.07142857142857144em;"><span class="pstrut" style="height:2.5em;"></span><span class="sizing reset-size3 size1 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.29011428571428566em;"><span></span></span></span></span></span></span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.345em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span> is a     constant being added at each element of the average.</p>
</div>
</li>
<li>
<p>Combine the terms in the summation back together ...</p>
<div style="overflow: auto"><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><mfrac><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn></mrow></msub><mn>2</mn></mfrac><mo>+</mo><mfrac><mn>1</mn><mrow><mi>n</mi><mo>−</mo><mn>2</mn></mrow></mfrac><mo>⋅</mo><munder><mo>∑</mo><mrow><mi>k</mi><mo>∈</mo><mi>S</mi><mo>−</mo><mo stretchy="false">{</mo><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn><mo stretchy="false">}</mo></mrow></munder><mfrac><mrow><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub><mo>−</mo><msub><mi>D</mi><mrow><mi>l</mi><mn>2</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub></mrow><mn>2</mn></mfrac></mrow><annotation encoding="application/x-tex">\frac{D_{l1,l2}}{2} + \frac{1}{n-2} \cdot \sum_{k \isin S-\{l1,l2\}}{\frac{D_{l1,k} - D_{l2,k}}{2}}
</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:2.04633em;vertical-align:-0.686em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.36033em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.09077em;vertical-align:-0.7693300000000001em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.32144em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord mathnormal">n</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.7693300000000001em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.876335em;vertical-align:-1.516005em;"></span><span class="mop op-limits"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.050005em;"><span style="top:-1.808995em;margin-left:0em;"><span class="pstrut" style="height:3.05em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03148em;">k</span><span class="mrel mtight">∈</span><span class="mord mathnormal mtight" style="margin-right:0.05764em;">S</span><span class="mbin mtight">−</span><span class="mopen mtight">{</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span><span class="mclose mtight">}</span></span></span></span><span style="top:-3.0500049999999996em;"><span class="pstrut" style="height:3.05em;"></span><span><span class="mop op-symbol large-op">∑</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.516005em;"><span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord"><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.36033em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span></span></span></div></li>
<li>
<p>Factor out <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mfrac><mn>1</mn><mn>2</mn></mfrac></mrow><annotation encoding="application/x-tex">\frac{1}{2}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.190108em;vertical-align:-0.345em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.845108em;"><span style="top:-2.6550000000000002em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">2</span></span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.394em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">1</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.345em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span> from the entire equation...</p>
<div style="overflow: auto"><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><mfrac><mn>1</mn><mn>2</mn></mfrac><mo>⋅</mo><mo stretchy="false">(</mo><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn></mrow></msub><mo>+</mo><mfrac><mn>1</mn><mrow><mi>n</mi><mo>−</mo><mn>2</mn></mrow></mfrac><mo>⋅</mo><munder><mo>∑</mo><mrow><mi>k</mi><mo>∈</mo><mi>S</mi><mo>−</mo><mo stretchy="false">{</mo><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn><mo stretchy="false">}</mo></mrow></munder><mrow><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub><mo>−</mo><msub><mi>D</mi><mrow><mi>l</mi><mn>2</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub></mrow><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">\frac{1}{2} \cdot (D_{l1,l2} + \frac{1}{n-2} \cdot \sum_{k \isin S-\{l1,l2\}}{D_{l1,k} - D_{l2,k}})
</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:2.00744em;vertical-align:-0.686em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.32144em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:1.036108em;vertical-align:-0.286108em;"></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.09077em;vertical-align:-0.7693300000000001em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.32144em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord mathnormal">n</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.7693300000000001em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.5660100000000003em;vertical-align:-1.516005em;"></span><span class="mop op-limits"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.050005em;"><span style="top:-1.808995em;margin-left:0em;"><span class="pstrut" style="height:3.05em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03148em;">k</span><span class="mrel mtight">∈</span><span class="mord mathnormal mtight" style="margin-right:0.05764em;">S</span><span class="mbin mtight">−</span><span class="mopen mtight">{</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span><span class="mclose mtight">}</span></span></span></span><span style="top:-3.0500049999999996em;"><span class="pstrut" style="height:3.05em;"></span><span><span class="mop op-symbol large-op">∑</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.516005em;"><span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord"><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span><span class="mclose">)</span></span></span></span></span></div><div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>Confused about what's happening above? It's just distributing and pulling out. For example, given the formula 5/2 + x*(3/2 + 5/2 + 9/2) ...</p>
<ol>
<li>5/2 + 3x/2 + 5x/2 + 9x/2 -- distribute x</li>
<li>1/2 * (5 + 3x + 5x + 9x) -- all terms are divided by 2 now, so pull out 1/2</li>
<li>1/2 * (5 + x*(3 + 5 + 9)) -- pull x back out</li>
</ol>
</div>
</li>
<li>
<p>Break up the summation into two simpler summations ...</p>
<div style="overflow: auto"><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><mfrac><mn>1</mn><mn>2</mn></mfrac><mo>⋅</mo><mo stretchy="false">(</mo><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn></mrow></msub><mo>+</mo><mfrac><mn>1</mn><mrow><mi>n</mi><mo>−</mo><mn>2</mn></mrow></mfrac><mo>⋅</mo><mo stretchy="false">(</mo><munder><mo>∑</mo><mrow><mi>k</mi><mo>∈</mo><mi>S</mi><mo>−</mo><mo stretchy="false">{</mo><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn><mo stretchy="false">}</mo></mrow></munder><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub><mo>−</mo><munder><mo>∑</mo><mrow><mi>k</mi><mo>∈</mo><mi>S</mi><mo>−</mo><mo stretchy="false">{</mo><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn><mo stretchy="false">}</mo></mrow></munder><msub><mi>D</mi><mrow><mi>l</mi><mn>2</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub><mo stretchy="false">)</mo><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">\frac{1}{2} \cdot (D_{l1,l2} + \frac{1}{n-2} \cdot (\sum_{k \isin S-\{l1,l2\}}{D_{l1,k}} - \sum_{k \isin S-\{l1,l2\}}{D_{l2,k}}))
</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:2.00744em;vertical-align:-0.686em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.32144em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:1.036108em;vertical-align:-0.286108em;"></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.09077em;vertical-align:-0.7693300000000001em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.32144em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord mathnormal">n</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.7693300000000001em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.5660100000000003em;vertical-align:-1.516005em;"></span><span class="mopen">(</span><span class="mop op-limits"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.050005em;"><span style="top:-1.808995em;margin-left:0em;"><span class="pstrut" style="height:3.05em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03148em;">k</span><span class="mrel mtight">∈</span><span class="mord mathnormal mtight" style="margin-right:0.05764em;">S</span><span class="mbin mtight">−</span><span class="mopen mtight">{</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span><span class="mclose mtight">}</span></span></span></span><span style="top:-3.0500049999999996em;"><span class="pstrut" style="height:3.05em;"></span><span><span class="mop op-symbol large-op">∑</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.516005em;"><span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord"><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.5660100000000003em;vertical-align:-1.516005em;"></span><span class="mop op-limits"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.050005em;"><span style="top:-1.808995em;margin-left:0em;"><span class="pstrut" style="height:3.05em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03148em;">k</span><span class="mrel mtight">∈</span><span class="mord mathnormal mtight" style="margin-right:0.05764em;">S</span><span class="mbin mtight">−</span><span class="mopen mtight">{</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span><span class="mclose mtight">}</span></span></span></span><span style="top:-3.0500049999999996em;"><span class="pstrut" style="height:3.05em;"></span><span><span class="mop op-symbol large-op">∑</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.516005em;"><span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord"><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span><span class="mclose">)</span><span class="mclose">)</span></span></span></span></span></div><div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>Confused about what's happening above? Think about it like this...</p>
<p>(9-1)+(8-2)+(7-3) = 9+8+7-1-2-3 = 24+(-6) = 24-6 = sum([9,8,7])-sum([1,2,3])</p>
<p>It's just re-ordering the operations so that it can be represented as two sums. It's perfectly valid.</p>
</div>
</li>
</ol>
<p>The above formula calculates the <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a> for l1. To instead find the formula for l2, just swap l1 and l2 ...</p>
<div style="overflow: auto"><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><mi>l</mi><mi>e</mi><mi>n</mi><mo stretchy="false">(</mo><mi>l</mi><mn>1</mn><mo stretchy="false">)</mo><mo>=</mo><mfrac><mn>1</mn><mn>2</mn></mfrac><mo>⋅</mo><mo stretchy="false">(</mo><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn></mrow></msub><mo>+</mo><mfrac><mn>1</mn><mrow><mi>n</mi><mo>−</mo><mn>2</mn></mrow></mfrac><mo>⋅</mo><mo stretchy="false">(</mo><munder><mo>∑</mo><mrow><mi>k</mi><mo>∈</mo><mi>S</mi><mo>−</mo><mo stretchy="false">{</mo><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn><mo stretchy="false">}</mo></mrow></munder><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub><mo>−</mo><munder><mo>∑</mo><mrow><mi>k</mi><mo>∈</mo><mi>S</mi><mo>−</mo><mo stretchy="false">{</mo><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn><mo stretchy="false">}</mo></mrow></munder><msub><mi>D</mi><mrow><mi>l</mi><mn>2</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub><mo stretchy="false">)</mo><mo stretchy="false">)</mo><mspace linebreak="newline" height="0.5em"></mspace><mi>l</mi><mi>e</mi><mi>n</mi><mo stretchy="false">(</mo><mi>l</mi><mn>2</mn><mo stretchy="false">)</mo><mo>=</mo><mfrac><mn>1</mn><mn>2</mn></mfrac><mo>⋅</mo><mo stretchy="false">(</mo><msub><mi>D</mi><mrow><mi>l</mi><mn>2</mn><mo separator="true">,</mo><mi>l</mi><mn>1</mn></mrow></msub><mo>+</mo><mfrac><mn>1</mn><mrow><mi>n</mi><mo>−</mo><mn>2</mn></mrow></mfrac><mo>⋅</mo><mo stretchy="false">(</mo><munder><mo>∑</mo><mrow><mi>k</mi><mo>∈</mo><mi>S</mi><mo>−</mo><mo stretchy="false">{</mo><mi>l</mi><mn>2</mn><mo separator="true">,</mo><mi>l</mi><mn>1</mn><mo stretchy="false">}</mo></mrow></munder><msub><mi>D</mi><mrow><mi>l</mi><mn>2</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub><mo>−</mo><munder><mo>∑</mo><mrow><mi>k</mi><mo>∈</mo><mi>S</mi><mo>−</mo><mo stretchy="false">{</mo><mi>l</mi><mn>2</mn><mo separator="true">,</mo><mi>l</mi><mn>1</mn><mo stretchy="false">}</mo></mrow></munder><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub><mo stretchy="false">)</mo><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">len(l1) = \frac{1}{2} \cdot (D_{l1,l2} + \frac{1}{n-2} \cdot (\sum_{k \isin S-\{l1,l2\}}{D_{l1,k}} - \sum_{k \isin S-\{l1,l2\}}{D_{l2,k}}))
\\[0.5em]
len(l2) = \frac{1}{2} \cdot (D_{l2,l1} + \frac{1}{n-2} \cdot (\sum_{k \isin S-\{l2,l1\}}{D_{l2,k}} - \sum_{k \isin S-\{l2,l1\}}{D_{l1,k}}))
</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord mathnormal">e</span><span class="mord mathnormal">n</span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord">1</span><span class="mclose">)</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:2.00744em;vertical-align:-0.686em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.32144em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:1.036108em;vertical-align:-0.286108em;"></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.09077em;vertical-align:-0.7693300000000001em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.32144em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord mathnormal">n</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.7693300000000001em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.5660100000000003em;vertical-align:-1.516005em;"></span><span class="mopen">(</span><span class="mop op-limits"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.050005em;"><span style="top:-1.808995em;margin-left:0em;"><span class="pstrut" style="height:3.05em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03148em;">k</span><span class="mrel mtight">∈</span><span class="mord mathnormal mtight" style="margin-right:0.05764em;">S</span><span class="mbin mtight">−</span><span class="mopen mtight">{</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span><span class="mclose mtight">}</span></span></span></span><span style="top:-3.0500049999999996em;"><span class="pstrut" style="height:3.05em;"></span><span><span class="mop op-symbol large-op">∑</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.516005em;"><span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord"><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.5660100000000003em;vertical-align:-1.516005em;"></span><span class="mop op-limits"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.050005em;"><span style="top:-1.808995em;margin-left:0em;"><span class="pstrut" style="height:3.05em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03148em;">k</span><span class="mrel mtight">∈</span><span class="mord mathnormal mtight" style="margin-right:0.05764em;">S</span><span class="mbin mtight">−</span><span class="mopen mtight">{</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span><span class="mclose mtight">}</span></span></span></span><span style="top:-3.0500049999999996em;"><span class="pstrut" style="height:3.05em;"></span><span><span class="mop op-symbol large-op">∑</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.516005em;"><span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord"><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span><span class="mclose">)</span><span class="mclose">)</span></span><span class="mspace newline" style="margin-top:0.5em;"></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord mathnormal">e</span><span class="mord mathnormal">n</span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord">2</span><span class="mclose">)</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:2.00744em;vertical-align:-0.686em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.32144em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:1.036108em;vertical-align:-0.286108em;"></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.09077em;vertical-align:-0.7693300000000001em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.32144em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord mathnormal">n</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.7693300000000001em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.5660100000000003em;vertical-align:-1.516005em;"></span><span class="mopen">(</span><span class="mop op-limits"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.050005em;"><span style="top:-1.808995em;margin-left:0em;"><span class="pstrut" style="height:3.05em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03148em;">k</span><span class="mrel mtight">∈</span><span class="mord mathnormal mtight" style="margin-right:0.05764em;">S</span><span class="mbin mtight">−</span><span class="mopen mtight">{</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mclose mtight">}</span></span></span></span><span style="top:-3.0500049999999996em;"><span class="pstrut" style="height:3.05em;"></span><span><span class="mop op-symbol large-op">∑</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.516005em;"><span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord"><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.5660100000000003em;vertical-align:-1.516005em;"></span><span class="mop op-limits"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.050005em;"><span style="top:-1.808995em;margin-left:0em;"><span class="pstrut" style="height:3.05em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03148em;">k</span><span class="mrel mtight">∈</span><span class="mord mathnormal mtight" style="margin-right:0.05764em;">S</span><span class="mbin mtight">−</span><span class="mopen mtight">{</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mclose mtight">}</span></span></span></span><span style="top:-3.0500049999999996em;"><span class="pstrut" style="height:3.05em;"></span><span><span class="mop op-symbol large-op">∑</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.516005em;"><span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord"><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span><span class="mclose">)</span><span class="mclose">)</span></span></span></span></span></div><p>Note how the two are almost exactly the same. <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn></mrow></msub><mo>=</mo><msub><mi>D</mi><mrow><mi>l</mi><mn>2</mn><mo separator="true">,</mo><mi>l</mi><mn>1</mn></mrow></msub></mrow><annotation encoding="application/x-tex">D_{l1,l2} = D_{l2,l1}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.969438em;vertical-align:-0.286108em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:0.969438em;vertical-align:-0.286108em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span></span></span>, and <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>S</mi><mo>−</mo><mo stretchy="false">{</mo><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn><mo stretchy="false">}</mo><mo>=</mo><mi>S</mi><mo>−</mo><mo stretchy="false">{</mo><mi>l</mi><mn>2</mn><mo separator="true">,</mo><mi>l</mi><mn>1</mn><mo stretchy="false">}</mo></mrow><annotation encoding="application/x-tex">S-{l1,l2} = S-{l2,l1}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.76666em;vertical-align:-0.08333em;"></span><span class="mord mathnormal" style="margin-right:0.05764em;">S</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mopen">{</span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord">1</span><span class="mpunct">,</span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord">2</span><span class="mclose">}</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:0.76666em;vertical-align:-0.08333em;"></span><span class="mord mathnormal" style="margin-right:0.05764em;">S</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mopen">{</span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord">2</span><span class="mpunct">,</span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord">1</span><span class="mclose">}</span></span></span></span>, and both summations are still there. The only exception is the order in which the summations are being subtracted ...</p>
<div style="overflow: auto"><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><mi>l</mi><mi>e</mi><mi>n</mi><mo stretchy="false">(</mo><mi>l</mi><mn>1</mn><mo stretchy="false">)</mo><mo>=</mo><mfrac><mn>1</mn><mn>2</mn></mfrac><mo>⋅</mo><mo stretchy="false">(</mo><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn></mrow></msub><mo>+</mo><mfrac><mn>1</mn><mrow><mi>n</mi><mo>−</mo><mn>2</mn></mrow></mfrac><mo>⋅</mo><mo stretchy="false">(</mo><mstyle mathcolor="#7f7f00"><munder><mo>∑</mo><mrow><mi>k</mi><mo>∈</mo><mi>S</mi><mo>−</mo><mo stretchy="false">{</mo><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn><mo stretchy="false">}</mo></mrow></munder><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub></mstyle><mo>−</mo><mstyle mathcolor="#007f7f"><munder><mo>∑</mo><mrow><mi>k</mi><mo>∈</mo><mi>S</mi><mo>−</mo><mo stretchy="false">{</mo><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn><mo stretchy="false">}</mo></mrow></munder><msub><mi>D</mi><mrow><mi>l</mi><mn>2</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub></mstyle><mo stretchy="false">)</mo><mo stretchy="false">)</mo><mspace linebreak="newline" height="0.5em"></mspace><mi>l</mi><mi>e</mi><mi>n</mi><mo stretchy="false">(</mo><mi>l</mi><mn>2</mn><mo stretchy="false">)</mo><mo>=</mo><mfrac><mn>1</mn><mn>2</mn></mfrac><mo>⋅</mo><mo stretchy="false">(</mo><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn></mrow></msub><mo>+</mo><mfrac><mn>1</mn><mrow><mi>n</mi><mo>−</mo><mn>2</mn></mrow></mfrac><mo>⋅</mo><mo stretchy="false">(</mo><mstyle mathcolor="#007f7f"><munder><mo>∑</mo><mrow><mi>k</mi><mo>∈</mo><mi>S</mi><mo>−</mo><mo stretchy="false">{</mo><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn><mo stretchy="false">}</mo></mrow></munder><msub><mi>D</mi><mrow><mi>l</mi><mn>2</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub></mstyle><mo>−</mo><mstyle mathcolor="#7f7f00"><munder><mo>∑</mo><mrow><mi>k</mi><mo>∈</mo><mi>S</mi><mo>−</mo><mo stretchy="false">{</mo><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn><mo stretchy="false">}</mo></mrow></munder><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub></mstyle><mo stretchy="false">)</mo><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">len(l1) = \frac{1}{2} \cdot (D_{l1,l2} + \frac{1}{n-2} \cdot (\textcolor{#7f7f00}{\sum_{k \isin S-\{l1,l2\}}{D_{l1,k}}} - \textcolor{#007f7f}{\sum_{k \isin S-\{l1,l2\}}{D_{l2,k}}}))
\\[0.5em]
len(l2) = \frac{1}{2} \cdot (D_{l1,l2} + \frac{1}{n-2} \cdot (\textcolor{#007f7f}{\sum_{k \isin S-\{l1,l2\}}{D_{l2,k}}} - \textcolor{#7f7f00}{\sum_{k \isin S-\{l1,l2\}}{D_{l1,k}}}))
</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord mathnormal">e</span><span class="mord mathnormal">n</span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord">1</span><span class="mclose">)</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:2.00744em;vertical-align:-0.686em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.32144em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:1.036108em;vertical-align:-0.286108em;"></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.09077em;vertical-align:-0.7693300000000001em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.32144em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord mathnormal">n</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.7693300000000001em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.5660100000000003em;vertical-align:-1.516005em;"></span><span class="mopen">(</span><span class="mop op-limits" style="color:#7f7f00;"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.050005em;"><span style="top:-1.808995em;margin-left:0em;"><span class="pstrut" style="height:3.05em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#7f7f00;"><span class="mord mtight" style="color:#7f7f00;"><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#7f7f00;">k</span><span class="mrel mtight" style="color:#7f7f00;">∈</span><span class="mord mathnormal mtight" style="margin-right:0.05764em;color:#7f7f00;">S</span><span class="mbin mtight" style="color:#7f7f00;">−</span><span class="mopen mtight" style="color:#7f7f00;">{</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#7f7f00;">l</span><span class="mord mtight" style="color:#7f7f00;">1</span><span class="mpunct mtight" style="color:#7f7f00;">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#7f7f00;">l</span><span class="mord mtight" style="color:#7f7f00;">2</span><span class="mclose mtight" style="color:#7f7f00;">}</span></span></span></span><span style="top:-3.0500049999999996em;"><span class="pstrut" style="height:3.05em;"></span><span><span class="mop op-symbol large-op" style="color:#7f7f00;">∑</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.516005em;"><span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord" style="color:#7f7f00;"><span class="mord" style="color:#7f7f00;"><span class="mord mathnormal" style="margin-right:0.02778em;color:#7f7f00;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#7f7f00;"><span class="mord mtight" style="color:#7f7f00;"><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#7f7f00;">l</span><span class="mord mtight" style="color:#7f7f00;">1</span><span class="mpunct mtight" style="color:#7f7f00;">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#7f7f00;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.5660100000000003em;vertical-align:-1.516005em;"></span><span class="mop op-limits" style="color:#007f7f;"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.050005em;"><span style="top:-1.808995em;margin-left:0em;"><span class="pstrut" style="height:3.05em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#007f7f;"><span class="mord mtight" style="color:#007f7f;"><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#007f7f;">k</span><span class="mrel mtight" style="color:#007f7f;">∈</span><span class="mord mathnormal mtight" style="margin-right:0.05764em;color:#007f7f;">S</span><span class="mbin mtight" style="color:#007f7f;">−</span><span class="mopen mtight" style="color:#007f7f;">{</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#007f7f;">l</span><span class="mord mtight" style="color:#007f7f;">1</span><span class="mpunct mtight" style="color:#007f7f;">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#007f7f;">l</span><span class="mord mtight" style="color:#007f7f;">2</span><span class="mclose mtight" style="color:#007f7f;">}</span></span></span></span><span style="top:-3.0500049999999996em;"><span class="pstrut" style="height:3.05em;"></span><span><span class="mop op-symbol large-op" style="color:#007f7f;">∑</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.516005em;"><span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord" style="color:#007f7f;"><span class="mord" style="color:#007f7f;"><span class="mord mathnormal" style="margin-right:0.02778em;color:#007f7f;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#007f7f;"><span class="mord mtight" style="color:#007f7f;"><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#007f7f;">l</span><span class="mord mtight" style="color:#007f7f;">2</span><span class="mpunct mtight" style="color:#007f7f;">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#007f7f;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span><span class="mclose">)</span><span class="mclose">)</span></span><span class="mspace newline" style="margin-top:0.5em;"></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord mathnormal">e</span><span class="mord mathnormal">n</span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord">2</span><span class="mclose">)</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:2.00744em;vertical-align:-0.686em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.32144em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:1.036108em;vertical-align:-0.286108em;"></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.09077em;vertical-align:-0.7693300000000001em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.32144em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord mathnormal">n</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.7693300000000001em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.5660100000000003em;vertical-align:-1.516005em;"></span><span class="mopen">(</span><span class="mop op-limits" style="color:#007f7f;"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.050005em;"><span style="top:-1.808995em;margin-left:0em;"><span class="pstrut" style="height:3.05em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#007f7f;"><span class="mord mtight" style="color:#007f7f;"><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#007f7f;">k</span><span class="mrel mtight" style="color:#007f7f;">∈</span><span class="mord mathnormal mtight" style="margin-right:0.05764em;color:#007f7f;">S</span><span class="mbin mtight" style="color:#007f7f;">−</span><span class="mopen mtight" style="color:#007f7f;">{</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#007f7f;">l</span><span class="mord mtight" style="color:#007f7f;">1</span><span class="mpunct mtight" style="color:#007f7f;">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#007f7f;">l</span><span class="mord mtight" style="color:#007f7f;">2</span><span class="mclose mtight" style="color:#007f7f;">}</span></span></span></span><span style="top:-3.0500049999999996em;"><span class="pstrut" style="height:3.05em;"></span><span><span class="mop op-symbol large-op" style="color:#007f7f;">∑</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.516005em;"><span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord" style="color:#007f7f;"><span class="mord" style="color:#007f7f;"><span class="mord mathnormal" style="margin-right:0.02778em;color:#007f7f;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#007f7f;"><span class="mord mtight" style="color:#007f7f;"><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#007f7f;">l</span><span class="mord mtight" style="color:#007f7f;">2</span><span class="mpunct mtight" style="color:#007f7f;">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#007f7f;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.5660100000000003em;vertical-align:-1.516005em;"></span><span class="mop op-limits" style="color:#7f7f00;"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.050005em;"><span style="top:-1.808995em;margin-left:0em;"><span class="pstrut" style="height:3.05em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#7f7f00;"><span class="mord mtight" style="color:#7f7f00;"><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#7f7f00;">k</span><span class="mrel mtight" style="color:#7f7f00;">∈</span><span class="mord mathnormal mtight" style="margin-right:0.05764em;color:#7f7f00;">S</span><span class="mbin mtight" style="color:#7f7f00;">−</span><span class="mopen mtight" style="color:#7f7f00;">{</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#7f7f00;">l</span><span class="mord mtight" style="color:#7f7f00;">1</span><span class="mpunct mtight" style="color:#7f7f00;">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#7f7f00;">l</span><span class="mord mtight" style="color:#7f7f00;">2</span><span class="mclose mtight" style="color:#7f7f00;">}</span></span></span></span><span style="top:-3.0500049999999996em;"><span class="pstrut" style="height:3.05em;"></span><span><span class="mop op-symbol large-op" style="color:#7f7f00;">∑</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.516005em;"><span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord" style="color:#7f7f00;"><span class="mord" style="color:#7f7f00;"><span class="mord mathnormal" style="margin-right:0.02778em;color:#7f7f00;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#7f7f00;"><span class="mord mtight" style="color:#7f7f00;"><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#7f7f00;">l</span><span class="mord mtight" style="color:#7f7f00;">1</span><span class="mpunct mtight" style="color:#7f7f00;">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#7f7f00;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span><span class="mclose">)</span><span class="mclose">)</span></span></span></span></span></div><p>Consider what happens when you re-organize the formula for l2 as follows...</p>
<ol>
<li>
<p>Convert the summation subtraction to an addition of a negative...</p>
<div style="overflow: auto"><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><mi>l</mi><mi>e</mi><mi>n</mi><mo stretchy="false">(</mo><mi>l</mi><mn>2</mn><mo stretchy="false">)</mo><mo>=</mo><mfrac><mn>1</mn><mn>2</mn></mfrac><mo>⋅</mo><mo stretchy="false">(</mo><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn></mrow></msub><mo>+</mo><mfrac><mn>1</mn><mrow><mi>n</mi><mo>−</mo><mn>2</mn></mrow></mfrac><mo>⋅</mo><mo stretchy="false">(</mo><mstyle mathcolor="#007f7f"><munder><mo>∑</mo><mrow><mi>k</mi><mo>∈</mo><mi>S</mi><mo>−</mo><mo stretchy="false">{</mo><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn><mo stretchy="false">}</mo></mrow></munder><msub><mi>D</mi><mrow><mi>l</mi><mn>2</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub></mstyle><mo>+</mo><mo stretchy="false">(</mo><mo>−</mo><mstyle mathcolor="#7f7f00"><munder><mo>∑</mo><mrow><mi>k</mi><mo>∈</mo><mi>S</mi><mo>−</mo><mo stretchy="false">{</mo><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn><mo stretchy="false">}</mo></mrow></munder><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub></mstyle><mo stretchy="false">)</mo><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">len(l2) = \frac{1}{2} \cdot (D_{l1,l2} + \frac{1}{n-2} \cdot (\textcolor{#007f7f}{\sum_{k \isin S-\{l1,l2\}}{D_{l2,k}}} + (- \textcolor{#7f7f00}{\sum_{k \isin S-\{l1,l2\}}{D_{l1,k}}}))
</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord mathnormal">e</span><span class="mord mathnormal">n</span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord">2</span><span class="mclose">)</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:2.00744em;vertical-align:-0.686em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.32144em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:1.036108em;vertical-align:-0.286108em;"></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.09077em;vertical-align:-0.7693300000000001em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.32144em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord mathnormal">n</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.7693300000000001em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.5660100000000003em;vertical-align:-1.516005em;"></span><span class="mopen">(</span><span class="mop op-limits" style="color:#007f7f;"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.050005em;"><span style="top:-1.808995em;margin-left:0em;"><span class="pstrut" style="height:3.05em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#007f7f;"><span class="mord mtight" style="color:#007f7f;"><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#007f7f;">k</span><span class="mrel mtight" style="color:#007f7f;">∈</span><span class="mord mathnormal mtight" style="margin-right:0.05764em;color:#007f7f;">S</span><span class="mbin mtight" style="color:#007f7f;">−</span><span class="mopen mtight" style="color:#007f7f;">{</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#007f7f;">l</span><span class="mord mtight" style="color:#007f7f;">1</span><span class="mpunct mtight" style="color:#007f7f;">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#007f7f;">l</span><span class="mord mtight" style="color:#007f7f;">2</span><span class="mclose mtight" style="color:#007f7f;">}</span></span></span></span><span style="top:-3.0500049999999996em;"><span class="pstrut" style="height:3.05em;"></span><span><span class="mop op-symbol large-op" style="color:#007f7f;">∑</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.516005em;"><span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord" style="color:#007f7f;"><span class="mord" style="color:#007f7f;"><span class="mord mathnormal" style="margin-right:0.02778em;color:#007f7f;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#007f7f;"><span class="mord mtight" style="color:#007f7f;"><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#007f7f;">l</span><span class="mord mtight" style="color:#007f7f;">2</span><span class="mpunct mtight" style="color:#007f7f;">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#007f7f;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.5660100000000003em;vertical-align:-1.516005em;"></span><span class="mopen">(</span><span class="mord">−</span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mop op-limits" style="color:#7f7f00;"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.050005em;"><span style="top:-1.808995em;margin-left:0em;"><span class="pstrut" style="height:3.05em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#7f7f00;"><span class="mord mtight" style="color:#7f7f00;"><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#7f7f00;">k</span><span class="mrel mtight" style="color:#7f7f00;">∈</span><span class="mord mathnormal mtight" style="margin-right:0.05764em;color:#7f7f00;">S</span><span class="mbin mtight" style="color:#7f7f00;">−</span><span class="mopen mtight" style="color:#7f7f00;">{</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#7f7f00;">l</span><span class="mord mtight" style="color:#7f7f00;">1</span><span class="mpunct mtight" style="color:#7f7f00;">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#7f7f00;">l</span><span class="mord mtight" style="color:#7f7f00;">2</span><span class="mclose mtight" style="color:#7f7f00;">}</span></span></span></span><span style="top:-3.0500049999999996em;"><span class="pstrut" style="height:3.05em;"></span><span><span class="mop op-symbol large-op" style="color:#7f7f00;">∑</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.516005em;"><span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord" style="color:#7f7f00;"><span class="mord" style="color:#7f7f00;"><span class="mord mathnormal" style="margin-right:0.02778em;color:#7f7f00;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#7f7f00;"><span class="mord mtight" style="color:#7f7f00;"><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#7f7f00;">l</span><span class="mord mtight" style="color:#7f7f00;">1</span><span class="mpunct mtight" style="color:#7f7f00;">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#7f7f00;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span><span class="mclose">)</span><span class="mclose">)</span></span></span></span></span></div></li>
<li>
<p>Swap the order of the summation addition...</p>
<div style="overflow: auto"><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><mi>l</mi><mi>e</mi><mi>n</mi><mo stretchy="false">(</mo><mi>l</mi><mn>2</mn><mo stretchy="false">)</mo><mo>=</mo><mfrac><mn>1</mn><mn>2</mn></mfrac><mo>⋅</mo><mo stretchy="false">(</mo><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn></mrow></msub><mo>+</mo><mfrac><mn>1</mn><mrow><mi>n</mi><mo>−</mo><mn>2</mn></mrow></mfrac><mo>⋅</mo><mo stretchy="false">(</mo><mo>−</mo><mstyle mathcolor="#007f7f"><munder><mo>∑</mo><mrow><mi>k</mi><mo>∈</mo><mi>S</mi><mo>−</mo><mo stretchy="false">{</mo><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn><mo stretchy="false">}</mo></mrow></munder><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub></mstyle><mo>+</mo><mstyle mathcolor="#7f7f00"><munder><mo>∑</mo><mrow><mi>k</mi><mo>∈</mo><mi>S</mi><mo>−</mo><mo stretchy="false">{</mo><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn><mo stretchy="false">}</mo></mrow></munder><msub><mi>D</mi><mrow><mi>l</mi><mn>2</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub></mstyle><mo stretchy="false">)</mo><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">len(l2) = \frac{1}{2} \cdot (D_{l1,l2} + \frac{1}{n-2} \cdot (-\textcolor{#007f7f}{\sum_{k \isin S-\{l1,l2\}}{D_{l1,k}}} + \textcolor{#7f7f00}{\sum_{k \isin S-\{l1,l2\}}{D_{l2,k}}}))
</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord mathnormal">e</span><span class="mord mathnormal">n</span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord">2</span><span class="mclose">)</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:2.00744em;vertical-align:-0.686em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.32144em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:1.036108em;vertical-align:-0.286108em;"></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.09077em;vertical-align:-0.7693300000000001em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.32144em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord mathnormal">n</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.7693300000000001em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.5660100000000003em;vertical-align:-1.516005em;"></span><span class="mopen">(</span><span class="mord">−</span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mop op-limits" style="color:#007f7f;"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.050005em;"><span style="top:-1.808995em;margin-left:0em;"><span class="pstrut" style="height:3.05em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#007f7f;"><span class="mord mtight" style="color:#007f7f;"><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#007f7f;">k</span><span class="mrel mtight" style="color:#007f7f;">∈</span><span class="mord mathnormal mtight" style="margin-right:0.05764em;color:#007f7f;">S</span><span class="mbin mtight" style="color:#007f7f;">−</span><span class="mopen mtight" style="color:#007f7f;">{</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#007f7f;">l</span><span class="mord mtight" style="color:#007f7f;">1</span><span class="mpunct mtight" style="color:#007f7f;">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#007f7f;">l</span><span class="mord mtight" style="color:#007f7f;">2</span><span class="mclose mtight" style="color:#007f7f;">}</span></span></span></span><span style="top:-3.0500049999999996em;"><span class="pstrut" style="height:3.05em;"></span><span><span class="mop op-symbol large-op" style="color:#007f7f;">∑</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.516005em;"><span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord" style="color:#007f7f;"><span class="mord" style="color:#007f7f;"><span class="mord mathnormal" style="margin-right:0.02778em;color:#007f7f;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#007f7f;"><span class="mord mtight" style="color:#007f7f;"><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#007f7f;">l</span><span class="mord mtight" style="color:#007f7f;">1</span><span class="mpunct mtight" style="color:#007f7f;">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#007f7f;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.5660100000000003em;vertical-align:-1.516005em;"></span><span class="mop op-limits" style="color:#7f7f00;"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.050005em;"><span style="top:-1.808995em;margin-left:0em;"><span class="pstrut" style="height:3.05em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#7f7f00;"><span class="mord mtight" style="color:#7f7f00;"><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#7f7f00;">k</span><span class="mrel mtight" style="color:#7f7f00;">∈</span><span class="mord mathnormal mtight" style="margin-right:0.05764em;color:#7f7f00;">S</span><span class="mbin mtight" style="color:#7f7f00;">−</span><span class="mopen mtight" style="color:#7f7f00;">{</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#7f7f00;">l</span><span class="mord mtight" style="color:#7f7f00;">1</span><span class="mpunct mtight" style="color:#7f7f00;">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#7f7f00;">l</span><span class="mord mtight" style="color:#7f7f00;">2</span><span class="mclose mtight" style="color:#7f7f00;">}</span></span></span></span><span style="top:-3.0500049999999996em;"><span class="pstrut" style="height:3.05em;"></span><span><span class="mop op-symbol large-op" style="color:#7f7f00;">∑</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.516005em;"><span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord" style="color:#7f7f00;"><span class="mord" style="color:#7f7f00;"><span class="mord mathnormal" style="margin-right:0.02778em;color:#7f7f00;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#7f7f00;"><span class="mord mtight" style="color:#7f7f00;"><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#7f7f00;">l</span><span class="mord mtight" style="color:#7f7f00;">2</span><span class="mpunct mtight" style="color:#7f7f00;">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#7f7f00;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span><span class="mclose">)</span><span class="mclose">)</span></span></span></span></span></div></li>
<li>
<p>Factor out -1 from summation addition ...</p>
<div style="overflow: auto"><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><mi>l</mi><mi>e</mi><mi>n</mi><mo stretchy="false">(</mo><mi>l</mi><mn>2</mn><mo stretchy="false">)</mo><mo>=</mo><mfrac><mn>1</mn><mn>2</mn></mfrac><mo>⋅</mo><mo stretchy="false">(</mo><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn></mrow></msub><mo>+</mo><mfrac><mn>1</mn><mrow><mi>n</mi><mo>−</mo><mn>2</mn></mrow></mfrac><mo>⋅</mo><mo>−</mo><mn>1</mn><mo>⋅</mo><mo stretchy="false">(</mo><mstyle mathcolor="#7f7f00"><munder><mo>∑</mo><mrow><mi>k</mi><mo>∈</mo><mi>S</mi><mo>−</mo><mo stretchy="false">{</mo><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn><mo stretchy="false">}</mo></mrow></munder><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub></mstyle><mo>−</mo><mstyle mathcolor="#007f7f"><munder><mo>∑</mo><mrow><mi>k</mi><mo>∈</mo><mi>S</mi><mo>−</mo><mo stretchy="false">{</mo><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn><mo stretchy="false">}</mo></mrow></munder><msub><mi>D</mi><mrow><mi>l</mi><mn>2</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub></mstyle><mo stretchy="false">)</mo><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">len(l2) = \frac{1}{2} \cdot (D_{l1,l2} + \frac{1}{n-2} \cdot -1 \cdot (\textcolor{#7f7f00}{\sum_{k \isin S-\{l1,l2\}}{D_{l1,k}}} - \textcolor{#007f7f}{\sum_{k \isin S-\{l1,l2\}}{D_{l2,k}}}))
</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord mathnormal">e</span><span class="mord mathnormal">n</span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord">2</span><span class="mclose">)</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:2.00744em;vertical-align:-0.686em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.32144em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:1.036108em;vertical-align:-0.286108em;"></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.09077em;vertical-align:-0.7693300000000001em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.32144em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord mathnormal">n</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.7693300000000001em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.72777em;vertical-align:-0.08333em;"></span><span class="mord">−</span><span class="mord">1</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.5660100000000003em;vertical-align:-1.516005em;"></span><span class="mopen">(</span><span class="mop op-limits" style="color:#7f7f00;"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.050005em;"><span style="top:-1.808995em;margin-left:0em;"><span class="pstrut" style="height:3.05em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#7f7f00;"><span class="mord mtight" style="color:#7f7f00;"><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#7f7f00;">k</span><span class="mrel mtight" style="color:#7f7f00;">∈</span><span class="mord mathnormal mtight" style="margin-right:0.05764em;color:#7f7f00;">S</span><span class="mbin mtight" style="color:#7f7f00;">−</span><span class="mopen mtight" style="color:#7f7f00;">{</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#7f7f00;">l</span><span class="mord mtight" style="color:#7f7f00;">1</span><span class="mpunct mtight" style="color:#7f7f00;">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#7f7f00;">l</span><span class="mord mtight" style="color:#7f7f00;">2</span><span class="mclose mtight" style="color:#7f7f00;">}</span></span></span></span><span style="top:-3.0500049999999996em;"><span class="pstrut" style="height:3.05em;"></span><span><span class="mop op-symbol large-op" style="color:#7f7f00;">∑</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.516005em;"><span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord" style="color:#7f7f00;"><span class="mord" style="color:#7f7f00;"><span class="mord mathnormal" style="margin-right:0.02778em;color:#7f7f00;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#7f7f00;"><span class="mord mtight" style="color:#7f7f00;"><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#7f7f00;">l</span><span class="mord mtight" style="color:#7f7f00;">1</span><span class="mpunct mtight" style="color:#7f7f00;">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#7f7f00;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.5660100000000003em;vertical-align:-1.516005em;"></span><span class="mop op-limits" style="color:#007f7f;"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.050005em;"><span style="top:-1.808995em;margin-left:0em;"><span class="pstrut" style="height:3.05em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#007f7f;"><span class="mord mtight" style="color:#007f7f;"><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#007f7f;">k</span><span class="mrel mtight" style="color:#007f7f;">∈</span><span class="mord mathnormal mtight" style="margin-right:0.05764em;color:#007f7f;">S</span><span class="mbin mtight" style="color:#007f7f;">−</span><span class="mopen mtight" style="color:#007f7f;">{</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#007f7f;">l</span><span class="mord mtight" style="color:#007f7f;">1</span><span class="mpunct mtight" style="color:#007f7f;">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#007f7f;">l</span><span class="mord mtight" style="color:#007f7f;">2</span><span class="mclose mtight" style="color:#007f7f;">}</span></span></span></span><span style="top:-3.0500049999999996em;"><span class="pstrut" style="height:3.05em;"></span><span><span class="mop op-symbol large-op" style="color:#007f7f;">∑</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.516005em;"><span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord" style="color:#007f7f;"><span class="mord" style="color:#007f7f;"><span class="mord mathnormal" style="margin-right:0.02778em;color:#007f7f;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#007f7f;"><span class="mord mtight" style="color:#007f7f;"><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#007f7f;">l</span><span class="mord mtight" style="color:#007f7f;">2</span><span class="mpunct mtight" style="color:#007f7f;">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#007f7f;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span><span class="mclose">)</span><span class="mclose">)</span></span></span></span></span></div></li>
<li>
<p>Simplify ...</p>
<div style="overflow: auto"><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><mi>l</mi><mi>e</mi><mi>n</mi><mo stretchy="false">(</mo><mi>l</mi><mn>2</mn><mo stretchy="false">)</mo><mo>=</mo><mfrac><mn>1</mn><mn>2</mn></mfrac><mo>⋅</mo><mo stretchy="false">(</mo><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn></mrow></msub><mo>+</mo><mo>−</mo><mfrac><mn>1</mn><mrow><mi>n</mi><mo>−</mo><mn>2</mn></mrow></mfrac><mo>⋅</mo><mo stretchy="false">(</mo><mstyle mathcolor="#7f7f00"><munder><mo>∑</mo><mrow><mi>k</mi><mo>∈</mo><mi>S</mi><mo>−</mo><mo stretchy="false">{</mo><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn><mo stretchy="false">}</mo></mrow></munder><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub></mstyle><mo>−</mo><mstyle mathcolor="#007f7f"><munder><mo>∑</mo><mrow><mi>k</mi><mo>∈</mo><mi>S</mi><mo>−</mo><mo stretchy="false">{</mo><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn><mo stretchy="false">}</mo></mrow></munder><msub><mi>D</mi><mrow><mi>l</mi><mn>2</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub></mstyle><mo stretchy="false">)</mo><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">len(l2) = \frac{1}{2} \cdot (D_{l1,l2} + - \frac{1}{n-2} \cdot (\textcolor{#7f7f00}{\sum_{k \isin S-\{l1,l2\}}{D_{l1,k}}} - \textcolor{#007f7f}{\sum_{k \isin S-\{l1,l2\}}{D_{l2,k}}}))
</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord mathnormal">e</span><span class="mord mathnormal">n</span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord">2</span><span class="mclose">)</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:2.00744em;vertical-align:-0.686em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.32144em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:1.036108em;vertical-align:-0.286108em;"></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.09077em;vertical-align:-0.7693300000000001em;"></span><span class="mord">−</span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.32144em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord mathnormal">n</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.7693300000000001em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.5660100000000003em;vertical-align:-1.516005em;"></span><span class="mopen">(</span><span class="mop op-limits" style="color:#7f7f00;"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.050005em;"><span style="top:-1.808995em;margin-left:0em;"><span class="pstrut" style="height:3.05em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#7f7f00;"><span class="mord mtight" style="color:#7f7f00;"><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#7f7f00;">k</span><span class="mrel mtight" style="color:#7f7f00;">∈</span><span class="mord mathnormal mtight" style="margin-right:0.05764em;color:#7f7f00;">S</span><span class="mbin mtight" style="color:#7f7f00;">−</span><span class="mopen mtight" style="color:#7f7f00;">{</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#7f7f00;">l</span><span class="mord mtight" style="color:#7f7f00;">1</span><span class="mpunct mtight" style="color:#7f7f00;">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#7f7f00;">l</span><span class="mord mtight" style="color:#7f7f00;">2</span><span class="mclose mtight" style="color:#7f7f00;">}</span></span></span></span><span style="top:-3.0500049999999996em;"><span class="pstrut" style="height:3.05em;"></span><span><span class="mop op-symbol large-op" style="color:#7f7f00;">∑</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.516005em;"><span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord" style="color:#7f7f00;"><span class="mord" style="color:#7f7f00;"><span class="mord mathnormal" style="margin-right:0.02778em;color:#7f7f00;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#7f7f00;"><span class="mord mtight" style="color:#7f7f00;"><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#7f7f00;">l</span><span class="mord mtight" style="color:#7f7f00;">1</span><span class="mpunct mtight" style="color:#7f7f00;">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#7f7f00;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.5660100000000003em;vertical-align:-1.516005em;"></span><span class="mop op-limits" style="color:#007f7f;"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.050005em;"><span style="top:-1.808995em;margin-left:0em;"><span class="pstrut" style="height:3.05em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#007f7f;"><span class="mord mtight" style="color:#007f7f;"><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#007f7f;">k</span><span class="mrel mtight" style="color:#007f7f;">∈</span><span class="mord mathnormal mtight" style="margin-right:0.05764em;color:#007f7f;">S</span><span class="mbin mtight" style="color:#007f7f;">−</span><span class="mopen mtight" style="color:#007f7f;">{</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#007f7f;">l</span><span class="mord mtight" style="color:#007f7f;">1</span><span class="mpunct mtight" style="color:#007f7f;">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#007f7f;">l</span><span class="mord mtight" style="color:#007f7f;">2</span><span class="mclose mtight" style="color:#007f7f;">}</span></span></span></span><span style="top:-3.0500049999999996em;"><span class="pstrut" style="height:3.05em;"></span><span><span class="mop op-symbol large-op" style="color:#007f7f;">∑</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.516005em;"><span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord" style="color:#007f7f;"><span class="mord" style="color:#007f7f;"><span class="mord mathnormal" style="margin-right:0.02778em;color:#007f7f;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#007f7f;"><span class="mord mtight" style="color:#007f7f;"><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#007f7f;">l</span><span class="mord mtight" style="color:#007f7f;">2</span><span class="mpunct mtight" style="color:#007f7f;">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#007f7f;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span><span class="mclose">)</span><span class="mclose">)</span></span></span></span></span></div></li>
<li>
<p>Simplify ...</p>
<div style="overflow: auto"><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><mi>l</mi><mi>e</mi><mi>n</mi><mo stretchy="false">(</mo><mi>l</mi><mn>2</mn><mo stretchy="false">)</mo><mo>=</mo><mfrac><mn>1</mn><mn>2</mn></mfrac><mo>⋅</mo><mo stretchy="false">(</mo><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn></mrow></msub><mo>−</mo><mfrac><mn>1</mn><mrow><mi>n</mi><mo>−</mo><mn>2</mn></mrow></mfrac><mo>⋅</mo><mo stretchy="false">(</mo><mstyle mathcolor="#7f7f00"><munder><mo>∑</mo><mrow><mi>k</mi><mo>∈</mo><mi>S</mi><mo>−</mo><mo stretchy="false">{</mo><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn><mo stretchy="false">}</mo></mrow></munder><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub></mstyle><mo>−</mo><mstyle mathcolor="#007f7f"><munder><mo>∑</mo><mrow><mi>k</mi><mo>∈</mo><mi>S</mi><mo>−</mo><mo stretchy="false">{</mo><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn><mo stretchy="false">}</mo></mrow></munder><msub><mi>D</mi><mrow><mi>l</mi><mn>2</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub></mstyle><mo stretchy="false">)</mo><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">len(l2) = \frac{1}{2} \cdot (D_{l1,l2} - \frac{1}{n-2} \cdot (\textcolor{#7f7f00}{\sum_{k \isin S-\{l1,l2\}}{D_{l1,k}}} - \textcolor{#007f7f}{\sum_{k \isin S-\{l1,l2\}}{D_{l2,k}}}))
</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord mathnormal">e</span><span class="mord mathnormal">n</span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord">2</span><span class="mclose">)</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:2.00744em;vertical-align:-0.686em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.32144em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:1.036108em;vertical-align:-0.286108em;"></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.09077em;vertical-align:-0.7693300000000001em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.32144em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord mathnormal">n</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.7693300000000001em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.5660100000000003em;vertical-align:-1.516005em;"></span><span class="mopen">(</span><span class="mop op-limits" style="color:#7f7f00;"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.050005em;"><span style="top:-1.808995em;margin-left:0em;"><span class="pstrut" style="height:3.05em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#7f7f00;"><span class="mord mtight" style="color:#7f7f00;"><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#7f7f00;">k</span><span class="mrel mtight" style="color:#7f7f00;">∈</span><span class="mord mathnormal mtight" style="margin-right:0.05764em;color:#7f7f00;">S</span><span class="mbin mtight" style="color:#7f7f00;">−</span><span class="mopen mtight" style="color:#7f7f00;">{</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#7f7f00;">l</span><span class="mord mtight" style="color:#7f7f00;">1</span><span class="mpunct mtight" style="color:#7f7f00;">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#7f7f00;">l</span><span class="mord mtight" style="color:#7f7f00;">2</span><span class="mclose mtight" style="color:#7f7f00;">}</span></span></span></span><span style="top:-3.0500049999999996em;"><span class="pstrut" style="height:3.05em;"></span><span><span class="mop op-symbol large-op" style="color:#7f7f00;">∑</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.516005em;"><span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord" style="color:#7f7f00;"><span class="mord" style="color:#7f7f00;"><span class="mord mathnormal" style="margin-right:0.02778em;color:#7f7f00;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#7f7f00;"><span class="mord mtight" style="color:#7f7f00;"><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#7f7f00;">l</span><span class="mord mtight" style="color:#7f7f00;">1</span><span class="mpunct mtight" style="color:#7f7f00;">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#7f7f00;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.5660100000000003em;vertical-align:-1.516005em;"></span><span class="mop op-limits" style="color:#007f7f;"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.050005em;"><span style="top:-1.808995em;margin-left:0em;"><span class="pstrut" style="height:3.05em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#007f7f;"><span class="mord mtight" style="color:#007f7f;"><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#007f7f;">k</span><span class="mrel mtight" style="color:#007f7f;">∈</span><span class="mord mathnormal mtight" style="margin-right:0.05764em;color:#007f7f;">S</span><span class="mbin mtight" style="color:#007f7f;">−</span><span class="mopen mtight" style="color:#007f7f;">{</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#007f7f;">l</span><span class="mord mtight" style="color:#007f7f;">1</span><span class="mpunct mtight" style="color:#007f7f;">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#007f7f;">l</span><span class="mord mtight" style="color:#007f7f;">2</span><span class="mclose mtight" style="color:#007f7f;">}</span></span></span></span><span style="top:-3.0500049999999996em;"><span class="pstrut" style="height:3.05em;"></span><span><span class="mop op-symbol large-op" style="color:#007f7f;">∑</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.516005em;"><span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord" style="color:#007f7f;"><span class="mord" style="color:#007f7f;"><span class="mord mathnormal" style="margin-right:0.02778em;color:#007f7f;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#007f7f;"><span class="mord mtight" style="color:#007f7f;"><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#007f7f;">l</span><span class="mord mtight" style="color:#007f7f;">2</span><span class="mpunct mtight" style="color:#007f7f;">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#007f7f;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span><span class="mclose">)</span><span class="mclose">)</span></span></span></span></span></div></li>
</ol>
<p>After this reorganization, the two match up almost exactly. The only difference is that an addition has been swapped to a subtraction...</p>
<div style="overflow: auto"><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><mi>l</mi><mi>e</mi><mi>n</mi><mo stretchy="false">(</mo><mi>l</mi><mn>1</mn><mo stretchy="false">)</mo><mo>=</mo><mfrac><mn>1</mn><mn>2</mn></mfrac><mo>⋅</mo><mo stretchy="false">(</mo><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn></mrow></msub><mstyle mathcolor="#ff0000"><mo>+</mo></mstyle><mfrac><mn>1</mn><mrow><mi>n</mi><mo>−</mo><mn>2</mn></mrow></mfrac><mo>⋅</mo><mo stretchy="false">(</mo><mstyle mathcolor="#7f7f00"><munder><mo>∑</mo><mrow><mi>k</mi><mo>∈</mo><mi>S</mi><mo>−</mo><mo stretchy="false">{</mo><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn><mo stretchy="false">}</mo></mrow></munder><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub></mstyle><mo>−</mo><mstyle mathcolor="#007f7f"><munder><mo>∑</mo><mrow><mi>k</mi><mo>∈</mo><mi>S</mi><mo>−</mo><mo stretchy="false">{</mo><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn><mo stretchy="false">}</mo></mrow></munder><msub><mi>D</mi><mrow><mi>l</mi><mn>2</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub></mstyle><mo stretchy="false">)</mo><mo stretchy="false">)</mo><mspace linebreak="newline" height="0.5em"></mspace><mi>l</mi><mi>e</mi><mi>n</mi><mo stretchy="false">(</mo><mi>l</mi><mn>2</mn><mo stretchy="false">)</mo><mo>=</mo><mfrac><mn>1</mn><mn>2</mn></mfrac><mo>⋅</mo><mo stretchy="false">(</mo><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn></mrow></msub><mstyle mathcolor="#ff0000"><mo>−</mo></mstyle><mfrac><mn>1</mn><mrow><mi>n</mi><mo>−</mo><mn>2</mn></mrow></mfrac><mo>⋅</mo><mo stretchy="false">(</mo><mstyle mathcolor="#7f7f00"><munder><mo>∑</mo><mrow><mi>k</mi><mo>∈</mo><mi>S</mi><mo>−</mo><mo stretchy="false">{</mo><mi>l</mi><mn>2</mn><mo separator="true">,</mo><mi>l</mi><mn>1</mn><mo stretchy="false">}</mo></mrow></munder><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub></mstyle><mo>−</mo><mstyle mathcolor="#007f7f"><munder><mo>∑</mo><mrow><mi>k</mi><mo>∈</mo><mi>S</mi><mo>−</mo><mo stretchy="false">{</mo><mi>l</mi><mn>2</mn><mo separator="true">,</mo><mi>l</mi><mn>1</mn><mo stretchy="false">}</mo></mrow></munder><msub><mi>D</mi><mrow><mi>l</mi><mn>2</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub></mstyle><mo stretchy="false">)</mo><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">len(l1) = \frac{1}{2} \cdot (D_{l1,l2} \textcolor{#ff0000}{+} \frac{1}{n-2} \cdot (\textcolor{#7f7f00}{\sum_{k \isin S-\{l1,l2\}}{D_{l1,k}}} - \textcolor{#007f7f}{\sum_{k \isin S-\{l1,l2\}}{D_{l2,k}}}))
\\[0.5em]
len(l2) = \frac{1}{2} \cdot (D_{l1,l2} \textcolor{#ff0000}{-} \frac{1}{n-2} \cdot (\textcolor{#7f7f00}{\sum_{k \isin S-\{l2,l1\}}{D_{l1,k}}} - \textcolor{#007f7f}{\sum_{k \isin S-\{l2,l1\}}{D_{l2,k}}}))
</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord mathnormal">e</span><span class="mord mathnormal">n</span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord">1</span><span class="mclose">)</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:2.00744em;vertical-align:-0.686em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.32144em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:1.036108em;vertical-align:-0.286108em;"></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin" style="color:#ff0000;">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.09077em;vertical-align:-0.7693300000000001em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.32144em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord mathnormal">n</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.7693300000000001em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.5660100000000003em;vertical-align:-1.516005em;"></span><span class="mopen">(</span><span class="mop op-limits" style="color:#7f7f00;"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.050005em;"><span style="top:-1.808995em;margin-left:0em;"><span class="pstrut" style="height:3.05em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#7f7f00;"><span class="mord mtight" style="color:#7f7f00;"><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#7f7f00;">k</span><span class="mrel mtight" style="color:#7f7f00;">∈</span><span class="mord mathnormal mtight" style="margin-right:0.05764em;color:#7f7f00;">S</span><span class="mbin mtight" style="color:#7f7f00;">−</span><span class="mopen mtight" style="color:#7f7f00;">{</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#7f7f00;">l</span><span class="mord mtight" style="color:#7f7f00;">1</span><span class="mpunct mtight" style="color:#7f7f00;">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#7f7f00;">l</span><span class="mord mtight" style="color:#7f7f00;">2</span><span class="mclose mtight" style="color:#7f7f00;">}</span></span></span></span><span style="top:-3.0500049999999996em;"><span class="pstrut" style="height:3.05em;"></span><span><span class="mop op-symbol large-op" style="color:#7f7f00;">∑</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.516005em;"><span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord" style="color:#7f7f00;"><span class="mord" style="color:#7f7f00;"><span class="mord mathnormal" style="margin-right:0.02778em;color:#7f7f00;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#7f7f00;"><span class="mord mtight" style="color:#7f7f00;"><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#7f7f00;">l</span><span class="mord mtight" style="color:#7f7f00;">1</span><span class="mpunct mtight" style="color:#7f7f00;">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#7f7f00;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.5660100000000003em;vertical-align:-1.516005em;"></span><span class="mop op-limits" style="color:#007f7f;"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.050005em;"><span style="top:-1.808995em;margin-left:0em;"><span class="pstrut" style="height:3.05em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#007f7f;"><span class="mord mtight" style="color:#007f7f;"><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#007f7f;">k</span><span class="mrel mtight" style="color:#007f7f;">∈</span><span class="mord mathnormal mtight" style="margin-right:0.05764em;color:#007f7f;">S</span><span class="mbin mtight" style="color:#007f7f;">−</span><span class="mopen mtight" style="color:#007f7f;">{</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#007f7f;">l</span><span class="mord mtight" style="color:#007f7f;">1</span><span class="mpunct mtight" style="color:#007f7f;">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#007f7f;">l</span><span class="mord mtight" style="color:#007f7f;">2</span><span class="mclose mtight" style="color:#007f7f;">}</span></span></span></span><span style="top:-3.0500049999999996em;"><span class="pstrut" style="height:3.05em;"></span><span><span class="mop op-symbol large-op" style="color:#007f7f;">∑</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.516005em;"><span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord" style="color:#007f7f;"><span class="mord" style="color:#007f7f;"><span class="mord mathnormal" style="margin-right:0.02778em;color:#007f7f;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#007f7f;"><span class="mord mtight" style="color:#007f7f;"><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#007f7f;">l</span><span class="mord mtight" style="color:#007f7f;">2</span><span class="mpunct mtight" style="color:#007f7f;">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#007f7f;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span><span class="mclose">)</span><span class="mclose">)</span></span><span class="mspace newline" style="margin-top:0.5em;"></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord mathnormal">e</span><span class="mord mathnormal">n</span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord">2</span><span class="mclose">)</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:2.00744em;vertical-align:-0.686em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.32144em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:1.036108em;vertical-align:-0.286108em;"></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin" style="color:#ff0000;">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.09077em;vertical-align:-0.7693300000000001em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.32144em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord mathnormal">n</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.7693300000000001em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.5660100000000003em;vertical-align:-1.516005em;"></span><span class="mopen">(</span><span class="mop op-limits" style="color:#7f7f00;"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.050005em;"><span style="top:-1.808995em;margin-left:0em;"><span class="pstrut" style="height:3.05em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#7f7f00;"><span class="mord mtight" style="color:#7f7f00;"><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#7f7f00;">k</span><span class="mrel mtight" style="color:#7f7f00;">∈</span><span class="mord mathnormal mtight" style="margin-right:0.05764em;color:#7f7f00;">S</span><span class="mbin mtight" style="color:#7f7f00;">−</span><span class="mopen mtight" style="color:#7f7f00;">{</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#7f7f00;">l</span><span class="mord mtight" style="color:#7f7f00;">2</span><span class="mpunct mtight" style="color:#7f7f00;">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#7f7f00;">l</span><span class="mord mtight" style="color:#7f7f00;">1</span><span class="mclose mtight" style="color:#7f7f00;">}</span></span></span></span><span style="top:-3.0500049999999996em;"><span class="pstrut" style="height:3.05em;"></span><span><span class="mop op-symbol large-op" style="color:#7f7f00;">∑</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.516005em;"><span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord" style="color:#7f7f00;"><span class="mord" style="color:#7f7f00;"><span class="mord mathnormal" style="margin-right:0.02778em;color:#7f7f00;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#7f7f00;"><span class="mord mtight" style="color:#7f7f00;"><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#7f7f00;">l</span><span class="mord mtight" style="color:#7f7f00;">1</span><span class="mpunct mtight" style="color:#7f7f00;">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#7f7f00;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.5660100000000003em;vertical-align:-1.516005em;"></span><span class="mop op-limits" style="color:#007f7f;"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.050005em;"><span style="top:-1.808995em;margin-left:0em;"><span class="pstrut" style="height:3.05em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#007f7f;"><span class="mord mtight" style="color:#007f7f;"><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#007f7f;">k</span><span class="mrel mtight" style="color:#007f7f;">∈</span><span class="mord mathnormal mtight" style="margin-right:0.05764em;color:#007f7f;">S</span><span class="mbin mtight" style="color:#007f7f;">−</span><span class="mopen mtight" style="color:#007f7f;">{</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#007f7f;">l</span><span class="mord mtight" style="color:#007f7f;">2</span><span class="mpunct mtight" style="color:#007f7f;">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#007f7f;">l</span><span class="mord mtight" style="color:#007f7f;">1</span><span class="mclose mtight" style="color:#007f7f;">}</span></span></span></span><span style="top:-3.0500049999999996em;"><span class="pstrut" style="height:3.05em;"></span><span><span class="mop op-symbol large-op" style="color:#007f7f;">∑</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.516005em;"><span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord" style="color:#007f7f;"><span class="mord" style="color:#007f7f;"><span class="mord mathnormal" style="margin-right:0.02778em;color:#007f7f;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#007f7f;"><span class="mord mtight" style="color:#007f7f;"><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#007f7f;">l</span><span class="mord mtight" style="color:#007f7f;">2</span><span class="mpunct mtight" style="color:#007f7f;">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#007f7f;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span><span class="mclose">)</span><span class="mclose">)</span></span></span></span></span></div><p>The point of this optimization is that the summation calculation only needs to be performed once. The result can be used to calculate the <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a> for both of the <a href="#BM_(neighbour)%2Fi">neighbour</a>ing leaf nodes...</p>
<div style="overflow: auto"><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><mi>r</mi><mi>e</mi><mi>s</mi><mo>=</mo><mfrac><mn>1</mn><mrow><mi>n</mi><mo>−</mo><mn>2</mn></mrow></mfrac><mo>⋅</mo><mo stretchy="false">(</mo><mstyle mathcolor="#7f7f00"><munder><mo>∑</mo><mrow><mi>k</mi><mo>∈</mo><mi>S</mi><mo>−</mo><mo stretchy="false">{</mo><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn><mo stretchy="false">}</mo></mrow></munder><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub></mstyle><mo>−</mo><mstyle mathcolor="#007f7f"><munder><mo>∑</mo><mrow><mi>k</mi><mo>∈</mo><mi>S</mi><mo>−</mo><mo stretchy="false">{</mo><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn><mo stretchy="false">}</mo></mrow></munder><msub><mi>D</mi><mrow><mi>l</mi><mn>2</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub></mstyle><mo stretchy="false">)</mo><mspace linebreak="newline" height="0.5em"></mspace><mi>l</mi><mi>e</mi><mi>n</mi><mo stretchy="false">(</mo><mi>l</mi><mn>1</mn><mo stretchy="false">)</mo><mo>=</mo><mfrac><mn>1</mn><mn>2</mn></mfrac><mo>⋅</mo><mo stretchy="false">(</mo><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn></mrow></msub><mstyle mathcolor="#ff0000"><mo>+</mo></mstyle><mi>r</mi><mi>e</mi><mi>s</mi><mo stretchy="false">)</mo><mspace linebreak="newline" height="0.5em"></mspace><mi>l</mi><mi>e</mi><mi>n</mi><mo stretchy="false">(</mo><mi>l</mi><mn>2</mn><mo stretchy="false">)</mo><mo>=</mo><mfrac><mn>1</mn><mn>2</mn></mfrac><mo>⋅</mo><mo stretchy="false">(</mo><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn></mrow></msub><mstyle mathcolor="#ff0000"><mo>−</mo></mstyle><mi>r</mi><mi>e</mi><mi>s</mi><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">res = \frac{1}{n-2} \cdot (\textcolor{#7f7f00}{\sum_{k \isin S-\{l1,l2\}}{D_{l1,k}}} - \textcolor{#007f7f}{\sum_{k \isin S-\{l1,l2\}}{D_{l2,k}}})
\\[0.5em]
len(l1) = \frac{1}{2} \cdot (D_{l1,l2} \textcolor{#ff0000}{+} res)
\\[0.5em]
len(l2) = \frac{1}{2} \cdot (D_{l1,l2} \textcolor{#ff0000}{-} res)
</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.43056em;vertical-align:0em;"></span><span class="mord mathnormal" style="margin-right:0.02778em;">r</span><span class="mord mathnormal">e</span><span class="mord mathnormal">s</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:2.09077em;vertical-align:-0.7693300000000001em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.32144em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord mathnormal">n</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.7693300000000001em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.5660100000000003em;vertical-align:-1.516005em;"></span><span class="mopen">(</span><span class="mop op-limits" style="color:#7f7f00;"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.050005em;"><span style="top:-1.808995em;margin-left:0em;"><span class="pstrut" style="height:3.05em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#7f7f00;"><span class="mord mtight" style="color:#7f7f00;"><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#7f7f00;">k</span><span class="mrel mtight" style="color:#7f7f00;">∈</span><span class="mord mathnormal mtight" style="margin-right:0.05764em;color:#7f7f00;">S</span><span class="mbin mtight" style="color:#7f7f00;">−</span><span class="mopen mtight" style="color:#7f7f00;">{</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#7f7f00;">l</span><span class="mord mtight" style="color:#7f7f00;">1</span><span class="mpunct mtight" style="color:#7f7f00;">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#7f7f00;">l</span><span class="mord mtight" style="color:#7f7f00;">2</span><span class="mclose mtight" style="color:#7f7f00;">}</span></span></span></span><span style="top:-3.0500049999999996em;"><span class="pstrut" style="height:3.05em;"></span><span><span class="mop op-symbol large-op" style="color:#7f7f00;">∑</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.516005em;"><span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord" style="color:#7f7f00;"><span class="mord" style="color:#7f7f00;"><span class="mord mathnormal" style="margin-right:0.02778em;color:#7f7f00;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#7f7f00;"><span class="mord mtight" style="color:#7f7f00;"><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#7f7f00;">l</span><span class="mord mtight" style="color:#7f7f00;">1</span><span class="mpunct mtight" style="color:#7f7f00;">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#7f7f00;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.5660100000000003em;vertical-align:-1.516005em;"></span><span class="mop op-limits" style="color:#007f7f;"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.050005em;"><span style="top:-1.808995em;margin-left:0em;"><span class="pstrut" style="height:3.05em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#007f7f;"><span class="mord mtight" style="color:#007f7f;"><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#007f7f;">k</span><span class="mrel mtight" style="color:#007f7f;">∈</span><span class="mord mathnormal mtight" style="margin-right:0.05764em;color:#007f7f;">S</span><span class="mbin mtight" style="color:#007f7f;">−</span><span class="mopen mtight" style="color:#007f7f;">{</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#007f7f;">l</span><span class="mord mtight" style="color:#007f7f;">1</span><span class="mpunct mtight" style="color:#007f7f;">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#007f7f;">l</span><span class="mord mtight" style="color:#007f7f;">2</span><span class="mclose mtight" style="color:#007f7f;">}</span></span></span></span><span style="top:-3.0500049999999996em;"><span class="pstrut" style="height:3.05em;"></span><span><span class="mop op-symbol large-op" style="color:#007f7f;">∑</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.516005em;"><span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord" style="color:#007f7f;"><span class="mord" style="color:#007f7f;"><span class="mord mathnormal" style="margin-right:0.02778em;color:#007f7f;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#007f7f;"><span class="mord mtight" style="color:#007f7f;"><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#007f7f;">l</span><span class="mord mtight" style="color:#007f7f;">2</span><span class="mpunct mtight" style="color:#007f7f;">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#007f7f;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span><span class="mclose">)</span></span><span class="mspace newline" style="margin-top:0.5em;"></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord mathnormal">e</span><span class="mord mathnormal">n</span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord">1</span><span class="mclose">)</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:2.00744em;vertical-align:-0.686em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.32144em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:1.036108em;vertical-align:-0.286108em;"></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin" style="color:#ff0000;">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal" style="margin-right:0.02778em;">r</span><span class="mord mathnormal">e</span><span class="mord mathnormal">s</span><span class="mclose">)</span></span><span class="mspace newline" style="margin-top:0.5em;"></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord mathnormal">e</span><span class="mord mathnormal">n</span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord">2</span><span class="mclose">)</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:2.00744em;vertical-align:-0.686em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.32144em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:1.036108em;vertical-align:-0.286108em;"></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin" style="color:#ff0000;">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal" style="margin-right:0.02778em;">r</span><span class="mord mathnormal">e</span><span class="mord mathnormal">s</span><span class="mclose">)</span></span></span></span></span></div><p>Depending on your architecture, this optimized form can be tweaked even further for better performance. Recall that the distance of anything to itself is always zero, meaning that...</p>
<ul>
<li><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>1</mn></mrow></msub></mrow><annotation encoding="application/x-tex">D_{l1,l1}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.969438em;vertical-align:-0.286108em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span></span></span> doesn't contribute anything to the first summation.</li>
<li><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>D</mi><mrow><mi>l</mi><mn>2</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn></mrow></msub></mrow><annotation encoding="application/x-tex">D_{l2,l2}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.969438em;vertical-align:-0.286108em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span></span></span> doesn't contribute anything to the second summation.</li>
</ul>
<p>If the cost of removing those terms from their respective summations is higher than the cost of keeping them in (adding that extra 0), you might as well not remove them...</p>
<div style="overflow: auto"><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><mi>r</mi><mi>e</mi><mi>s</mi><mo>=</mo><mfrac><mn>1</mn><mrow><mi>n</mi><mo>−</mo><mn>2</mn></mrow></mfrac><mo>⋅</mo><mo stretchy="false">(</mo><mstyle mathcolor="#7f7f00"><munder><mo>∑</mo><mrow><mi>k</mi><mo>∈</mo><mi>S</mi><mo>−</mo><mo stretchy="false">{</mo><mi>l</mi><mn>2</mn><mo stretchy="false">}</mo></mrow></munder><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub></mstyle><mo>−</mo><mstyle mathcolor="#007f7f"><munder><mo>∑</mo><mrow><mi>k</mi><mo>∈</mo><mi>S</mi><mo>−</mo><mo stretchy="false">{</mo><mi>l</mi><mn>1</mn><mo stretchy="false">}</mo></mrow></munder><msub><mi>D</mi><mrow><mi>l</mi><mn>2</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub></mstyle><mo stretchy="false">)</mo><mspace linebreak="newline" height="0.5em"></mspace><mi>l</mi><mi>e</mi><mi>n</mi><mo stretchy="false">(</mo><mi>l</mi><mn>1</mn><mo stretchy="false">)</mo><mo>=</mo><mfrac><mn>1</mn><mn>2</mn></mfrac><mo>⋅</mo><mo stretchy="false">(</mo><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn></mrow></msub><mstyle mathcolor="#ff0000"><mo>+</mo></mstyle><mi>r</mi><mi>e</mi><mi>s</mi><mo stretchy="false">)</mo><mspace linebreak="newline" height="0.5em"></mspace><mi>l</mi><mi>e</mi><mi>n</mi><mo stretchy="false">(</mo><mi>l</mi><mn>2</mn><mo stretchy="false">)</mo><mo>=</mo><mfrac><mn>1</mn><mn>2</mn></mfrac><mo>⋅</mo><mo stretchy="false">(</mo><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn></mrow></msub><mstyle mathcolor="#ff0000"><mo>−</mo></mstyle><mi>r</mi><mi>e</mi><mi>s</mi><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">res = \frac{1}{n-2} \cdot (\textcolor{#7f7f00}{\sum_{k \isin S-\{l2\}}{D_{l1,k}}} - \textcolor{#007f7f}{\sum_{k \isin S-\{l1\}}{D_{l2,k}}})
\\[0.5em]
len(l1) = \frac{1}{2} \cdot (D_{l1,l2} \textcolor{#ff0000}{+} res)
\\[0.5em]
len(l2) = \frac{1}{2} \cdot (D_{l1,l2} \textcolor{#ff0000}{-} res)
</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.43056em;vertical-align:0em;"></span><span class="mord mathnormal" style="margin-right:0.02778em;">r</span><span class="mord mathnormal">e</span><span class="mord mathnormal">s</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:2.09077em;vertical-align:-0.7693300000000001em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.32144em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord mathnormal">n</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.7693300000000001em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.5660100000000003em;vertical-align:-1.516005em;"></span><span class="mopen">(</span><span class="mop op-limits" style="color:#7f7f00;"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.050005em;"><span style="top:-1.808995em;margin-left:0em;"><span class="pstrut" style="height:3.05em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#7f7f00;"><span class="mord mtight" style="color:#7f7f00;"><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#7f7f00;">k</span><span class="mrel mtight" style="color:#7f7f00;">∈</span><span class="mord mathnormal mtight" style="margin-right:0.05764em;color:#7f7f00;">S</span><span class="mbin mtight" style="color:#7f7f00;">−</span><span class="mopen mtight" style="color:#7f7f00;">{</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#7f7f00;">l</span><span class="mord mtight" style="color:#7f7f00;">2</span><span class="mclose mtight" style="color:#7f7f00;">}</span></span></span></span><span style="top:-3.0500049999999996em;"><span class="pstrut" style="height:3.05em;"></span><span><span class="mop op-symbol large-op" style="color:#7f7f00;">∑</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.516005em;"><span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord" style="color:#7f7f00;"><span class="mord" style="color:#7f7f00;"><span class="mord mathnormal" style="margin-right:0.02778em;color:#7f7f00;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#7f7f00;"><span class="mord mtight" style="color:#7f7f00;"><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#7f7f00;">l</span><span class="mord mtight" style="color:#7f7f00;">1</span><span class="mpunct mtight" style="color:#7f7f00;">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#7f7f00;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.5660100000000003em;vertical-align:-1.516005em;"></span><span class="mop op-limits" style="color:#007f7f;"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.050005em;"><span style="top:-1.808995em;margin-left:0em;"><span class="pstrut" style="height:3.05em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#007f7f;"><span class="mord mtight" style="color:#007f7f;"><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#007f7f;">k</span><span class="mrel mtight" style="color:#007f7f;">∈</span><span class="mord mathnormal mtight" style="margin-right:0.05764em;color:#007f7f;">S</span><span class="mbin mtight" style="color:#007f7f;">−</span><span class="mopen mtight" style="color:#007f7f;">{</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#007f7f;">l</span><span class="mord mtight" style="color:#007f7f;">1</span><span class="mclose mtight" style="color:#007f7f;">}</span></span></span></span><span style="top:-3.0500049999999996em;"><span class="pstrut" style="height:3.05em;"></span><span><span class="mop op-symbol large-op" style="color:#007f7f;">∑</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.516005em;"><span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord" style="color:#007f7f;"><span class="mord" style="color:#007f7f;"><span class="mord mathnormal" style="margin-right:0.02778em;color:#007f7f;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#007f7f;"><span class="mord mtight" style="color:#007f7f;"><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#007f7f;">l</span><span class="mord mtight" style="color:#007f7f;">2</span><span class="mpunct mtight" style="color:#007f7f;">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#007f7f;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span><span class="mclose">)</span></span><span class="mspace newline" style="margin-top:0.5em;"></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord mathnormal">e</span><span class="mord mathnormal">n</span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord">1</span><span class="mclose">)</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:2.00744em;vertical-align:-0.686em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.32144em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:1.036108em;vertical-align:-0.286108em;"></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin" style="color:#ff0000;">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal" style="margin-right:0.02778em;">r</span><span class="mord mathnormal">e</span><span class="mord mathnormal">s</span><span class="mclose">)</span></span><span class="mspace newline" style="margin-top:0.5em;"></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord mathnormal">e</span><span class="mord mathnormal">n</span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord">2</span><span class="mclose">)</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:2.00744em;vertical-align:-0.686em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.32144em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:1.036108em;vertical-align:-0.286108em;"></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin" style="color:#ff0000;">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal" style="margin-right:0.02778em;">r</span><span class="mord mathnormal">e</span><span class="mord mathnormal">s</span><span class="mclose">)</span></span></span></span></span></div><p>Similarly, removing both l2 from the first summation and l1 from the second summation doesn't actually change the result. The first summation will add <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn></mrow></msub></mrow><annotation encoding="application/x-tex">D_{l1,l2}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.969438em;vertical-align:-0.286108em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span></span></span> but the second summation will remove <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn></mrow></msub></mrow><annotation encoding="application/x-tex">D_{l1,l2}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.969438em;vertical-align:-0.286108em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span></span></span>, resulting in an overall contribution of 0. If the cost of removing those terms from their respective summations is higher than the cost of keeping them in, you might as well not remove them...</p>
<div style="overflow: auto"><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><mi>r</mi><mi>e</mi><mi>s</mi><mo>=</mo><mfrac><mn>1</mn><mrow><mi>n</mi><mo>−</mo><mn>2</mn></mrow></mfrac><mo>⋅</mo><mo stretchy="false">(</mo><mstyle mathcolor="#7f7f00"><munder><mo>∑</mo><mrow><mi>k</mi><mo>∈</mo><mi>S</mi></mrow></munder><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub></mstyle><mo>−</mo><mstyle mathcolor="#007f7f"><munder><mo>∑</mo><mrow><mi>k</mi><mo>∈</mo><mi>S</mi></mrow></munder><msub><mi>D</mi><mrow><mi>l</mi><mn>2</mn><mo separator="true">,</mo><mi>k</mi></mrow></msub></mstyle><mo stretchy="false">)</mo><mspace linebreak="newline" height="0.5em"></mspace><mi>l</mi><mi>e</mi><mi>n</mi><mo stretchy="false">(</mo><mi>l</mi><mn>1</mn><mo stretchy="false">)</mo><mo>=</mo><mfrac><mn>1</mn><mn>2</mn></mfrac><mo>⋅</mo><mo stretchy="false">(</mo><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn></mrow></msub><mstyle mathcolor="#ff0000"><mo>+</mo></mstyle><mi>r</mi><mi>e</mi><mi>s</mi><mo stretchy="false">)</mo><mspace linebreak="newline" height="0.5em"></mspace><mi>l</mi><mi>e</mi><mi>n</mi><mo stretchy="false">(</mo><mi>l</mi><mn>2</mn><mo stretchy="false">)</mo><mo>=</mo><mfrac><mn>1</mn><mn>2</mn></mfrac><mo>⋅</mo><mo stretchy="false">(</mo><msub><mi>D</mi><mrow><mi>l</mi><mn>1</mn><mo separator="true">,</mo><mi>l</mi><mn>2</mn></mrow></msub><mstyle mathcolor="#ff0000"><mo>−</mo></mstyle><mi>r</mi><mi>e</mi><mi>s</mi><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">res = \frac{1}{n-2} \cdot (\textcolor{#7f7f00}{\sum_{k \isin S}{D_{l1,k}}} - \textcolor{#007f7f}{\sum_{k \isin S}{D_{l2,k}}})
\\[0.5em]
len(l1) = \frac{1}{2} \cdot (D_{l1,l2} \textcolor{#ff0000}{+} res)
\\[0.5em]
len(l2) = \frac{1}{2} \cdot (D_{l1,l2} \textcolor{#ff0000}{-} res)
</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.43056em;vertical-align:0em;"></span><span class="mord mathnormal" style="margin-right:0.02778em;">r</span><span class="mord mathnormal">e</span><span class="mord mathnormal">s</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:2.09077em;vertical-align:-0.7693300000000001em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.32144em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord mathnormal">n</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.7693300000000001em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.3794880000000003em;vertical-align:-1.329483em;"></span><span class="mopen">(</span><span class="mop op-limits" style="color:#7f7f00;"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.050005em;"><span style="top:-1.8478869999999998em;margin-left:0em;"><span class="pstrut" style="height:3.05em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#7f7f00;"><span class="mord mtight" style="color:#7f7f00;"><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#7f7f00;">k</span><span class="mrel mtight" style="color:#7f7f00;">∈</span><span class="mord mathnormal mtight" style="margin-right:0.05764em;color:#7f7f00;">S</span></span></span></span><span style="top:-3.0500049999999996em;"><span class="pstrut" style="height:3.05em;"></span><span><span class="mop op-symbol large-op" style="color:#7f7f00;">∑</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.329483em;"><span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord" style="color:#7f7f00;"><span class="mord" style="color:#7f7f00;"><span class="mord mathnormal" style="margin-right:0.02778em;color:#7f7f00;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#7f7f00;"><span class="mord mtight" style="color:#7f7f00;"><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#7f7f00;">l</span><span class="mord mtight" style="color:#7f7f00;">1</span><span class="mpunct mtight" style="color:#7f7f00;">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#7f7f00;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:2.3794880000000003em;vertical-align:-1.329483em;"></span><span class="mop op-limits" style="color:#007f7f;"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.050005em;"><span style="top:-1.8478869999999998em;margin-left:0em;"><span class="pstrut" style="height:3.05em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#007f7f;"><span class="mord mtight" style="color:#007f7f;"><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#007f7f;">k</span><span class="mrel mtight" style="color:#007f7f;">∈</span><span class="mord mathnormal mtight" style="margin-right:0.05764em;color:#007f7f;">S</span></span></span></span><span style="top:-3.0500049999999996em;"><span class="pstrut" style="height:3.05em;"></span><span><span class="mop op-symbol large-op" style="color:#007f7f;">∑</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.329483em;"><span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord" style="color:#007f7f;"><span class="mord" style="color:#007f7f;"><span class="mord mathnormal" style="margin-right:0.02778em;color:#007f7f;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight" style="color:#007f7f;"><span class="mord mtight" style="color:#007f7f;"><span class="mord mathnormal mtight" style="margin-right:0.01968em;color:#007f7f;">l</span><span class="mord mtight" style="color:#007f7f;">2</span><span class="mpunct mtight" style="color:#007f7f;">,</span><span class="mord mathnormal mtight" style="margin-right:0.03148em;color:#007f7f;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span><span class="mclose">)</span></span><span class="mspace newline" style="margin-top:0.5em;"></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord mathnormal">e</span><span class="mord mathnormal">n</span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord">1</span><span class="mclose">)</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:2.00744em;vertical-align:-0.686em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.32144em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:1.036108em;vertical-align:-0.286108em;"></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin" style="color:#ff0000;">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal" style="margin-right:0.02778em;">r</span><span class="mord mathnormal">e</span><span class="mord mathnormal">s</span><span class="mclose">)</span></span><span class="mspace newline" style="margin-top:0.5em;"></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord mathnormal">e</span><span class="mord mathnormal">n</span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord">2</span><span class="mclose">)</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:2.00744em;vertical-align:-0.686em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.32144em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:1.036108em;vertical-align:-0.286108em;"></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361079999999999em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span><span class="mord mtight">2</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin" style="color:#ff0000;">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal" style="margin-right:0.02778em;">r</span><span class="mord mathnormal">e</span><span class="mord mathnormal">s</span><span class="mclose">)</span></span></span></span></span></div><p><a href="ch7_code/src/phylogeny/FindNeighbourLimbLengths_Optimized.py">ch7_code/src/phylogeny/FindNeighbourLimbLengths_Optimized.py</a> (lines 21 to 28):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">find_neighbouring_limb_lengths</span>(<span class="hljs-params">dm: DistanceMatrix[N], l1: N, l2: N</span>) -&gt; <span class="hljs-built_in">tuple</span>[<span class="hljs-built_in">float</span>, <span class="hljs-built_in">float</span>]:
    l1_dist_sum = <span class="hljs-built_in">sum</span>(dm[l1, k] <span class="hljs-keyword">for</span> k <span class="hljs-keyword">in</span> dm.leaf_ids())
    l2_dist_sum = <span class="hljs-built_in">sum</span>(dm[l2, k] <span class="hljs-keyword">for</span> k <span class="hljs-keyword">in</span> dm.leaf_ids())
    res = (l1_dist_sum - l2_dist_sum) / (dm.n - <span class="hljs-number">2</span>)
    l1_len = (dm[l1, l2] + res) / <span class="hljs-number">2</span>
    l2_len = (dm[l1, l2] - res) / <span class="hljs-number">2</span>
    <span class="hljs-keyword">return</span> l1_len, l2_len
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given distance matrix...</p>
<table>
<thead><tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr></thead>
<tbody>
<tr>
<td>v0</td>
<td>0.0</td>
<td>13.0</td>
<td>21.0</td>
<td>21.0</td>
<td>22.0</td>
<td>22.0</td>
</tr>
<tr>
<td>v1</td>
<td>13.0</td>
<td>0.0</td>
<td>12.0</td>
<td>12.0</td>
<td>13.0</td>
<td>13.0</td>
</tr>
<tr>
<td>v2</td>
<td>21.0</td>
<td>12.0</td>
<td>0.0</td>
<td>20.0</td>
<td>21.0</td>
<td>21.0</td>
</tr>
<tr>
<td>v3</td>
<td>21.0</td>
<td>12.0</td>
<td>20.0</td>
<td>0.0</td>
<td>7.0</td>
<td>13.0</td>
</tr>
<tr>
<td>v4</td>
<td>22.0</td>
<td>13.0</td>
<td>21.0</td>
<td>7.0</td>
<td>0.0</td>
<td>14.0</td>
</tr>
<tr>
<td>v5</td>
<td>22.0</td>
<td>13.0</td>
<td>21.0</td>
<td>13.0</td>
<td>14.0</td>
<td>0.0</td>
</tr>
</tbody>
</table>
<p>... and given that v1 and v2 are neighbours, the limb length for leaf node ...</p>
<ul>
<li>v1 is approximated to be 2.0</li>
<li>v2 is approximated to be 10.0</li>
</ul>
</div>
<p></p>
<a name="H_Expose%20Neighbour%20Parent"></a><h3>Expose Neighbour Parent</h3>
<p><a name="BM_(Algorithms%5C%2FPhylogeny%2FExpose%20Neighbour%20Parent)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FBald)_TOPIC%2F">Algorithms/Phylogeny/Bald</a></li>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FFind%20Neighbour%20Limb%20Lengths%2FAverage%20Algorithm)_TOPIC%2F">Algorithms/Phylogeny/Find Neighbour Limb Lengths/Average Algorithm</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: Given a <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> and a pair of leaf nodes identified as being <a href="#BM_(neighbour)%2Fi">neighbour</a>s, this algorithm removes those <a href="#BM_(neighbour)%2Fi">neighbour</a>s from the <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> and brings their parent to the forefront (as a leaf node in the <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a>). If the <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> is a non-<a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> (but close to being additive), this algorithm approximates the shared parent.</p>
<p><strong>WHY</strong>: This operation is required for <em>approximating</em> a <a href="#BM_(simple%20tree)%2Fi">simple tree</a> for a non-<a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>.</p>
<a name="H1_Average%20Algorithm"></a><h4>Average Algorithm</h4>
<p><a name="BM_(Algorithms%5C%2FPhylogeny%2FMerge%20Neighbours%2FAverage%20Algorithm)_TOPIC%2F"></a><strong></strong></p>
<p><strong>ALGORITHM</strong>:</p>
<p>At a high-level, this algorithm essentially boils down to balding each of the <a href="#BM_(neighbour)%2Fi">neighbour</a>s and combining them together. For example, v0 and v1 are <a href="#BM_(neighbour)%2Fi">neighbour</a>s in the following <a href="#BM_(simple%20tree)%2Fi">simple tree</a>...</p>
<p><img src="dot_fa269cbcc76261c503f77dc3a6e5b8b6.svg" alt="Dot diagram"></p>
<table>
<thead>
<tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr>
</thead>
<tbody>
<tr>
<td>v0</td>
<td>0</td>
<td>13</td>
<td>21</td>
<td>21</td>
<td>22</td>
<td>22</td>
</tr>
<tr>
<td>v1</td>
<td>13</td>
<td>0</td>
<td>12</td>
<td>12</td>
<td>13</td>
<td>13</td>
</tr>
<tr>
<td>v2</td>
<td>21</td>
<td>12</td>
<td>0</td>
<td>20</td>
<td>21</td>
<td>21</td>
</tr>
<tr>
<td>v3</td>
<td>21</td>
<td>12</td>
<td>20</td>
<td>0</td>
<td>7</td>
<td>13</td>
</tr>
<tr>
<td>v4</td>
<td>22</td>
<td>13</td>
<td>21</td>
<td>7</td>
<td>0</td>
<td>14</td>
</tr>
<tr>
<td>v5</td>
<td>22</td>
<td>13</td>
<td>21</td>
<td>13</td>
<td>14</td>
<td>0</td>
</tr>
</tbody>
</table>
<p>Balding both v0 and v1 results in ...</p>
<p><img src="dot_d70127552cc7a7b3da6a33d760e69ea9.svg" alt="Dot diagram"></p>
<table>
<thead>
<tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr>
</thead>
<tbody>
<tr>
<td>v0</td>
<td><span style="background-color: red; color: white; padding: 3px">0</span></td>
<td><span style="background-color: red; color: white; padding: 3px">0</span></td>
<td><span style="background-color: red; color: white; padding: 3px">10</span></td>
<td><span style="background-color: red; color: white; padding: 3px">10</span></td>
<td><span style="background-color: red; color: white; padding: 3px">11</span></td>
<td><span style="background-color: red; color: white; padding: 3px">11</span></td>
</tr>
<tr>
<td>v1</td>
<td><span style="background-color: red; color: white; padding: 3px">0</span></td>
<td><span style="background-color: red; color: white; padding: 3px">0</span></td>
<td><span style="background-color: red; color: white; padding: 3px">10</span></td>
<td><span style="background-color: red; color: white; padding: 3px">10</span></td>
<td><span style="background-color: red; color: white; padding: 3px">11</span></td>
<td><span style="background-color: red; color: white; padding: 3px">11</span></td>
</tr>
<tr>
<td>v2</td>
<td><span style="background-color: red; color: white; padding: 3px">10</span></td>
<td><span style="background-color: red; color: white; padding: 3px">10</span></td>
<td>0</td>
<td>20</td>
<td>21</td>
<td>21</td>
</tr>
<tr>
<td>v3</td>
<td><span style="background-color: red; color: white; padding: 3px">10</span></td>
<td><span style="background-color: red; color: white; padding: 3px">10</span></td>
<td>20</td>
<td>0</td>
<td>7</td>
<td>13</td>
</tr>
<tr>
<td>v4</td>
<td><span style="background-color: red; color: white; padding: 3px">11</span></td>
<td><span style="background-color: red; color: white; padding: 3px">11</span></td>
<td>21</td>
<td>7</td>
<td>0</td>
<td>14</td>
</tr>
<tr>
<td>v5</td>
<td><span style="background-color: red; color: white; padding: 3px">11</span></td>
<td><span style="background-color: red; color: white; padding: 3px">11</span></td>
<td>21</td>
<td>13</td>
<td>14</td>
<td>0</td>
</tr>
</tbody>
</table>
<p>Merging together balded v0 and balded v1 is done by iterating over the other leaf nodes and averaging their balded distances (e.g. the merged distance to v2 is calculated as dist(v0,v2) + dist(v1,v2) / 2)...</p>
<p><img src="dot_3d01f4a0fab8e57bf0bfa96673b676e3.svg" alt="Dot diagram"></p>
<table>
<thead>
<tr>
<th></th>
<th>M</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr>
</thead>
<tbody>
<tr>
<td>M</td>
<td><span style="background-color: red; color: white; padding: 3px">0</span></td>
<td><span style="background-color: red; color: white; padding: 3px">(10+10)/2=10</span></td>
<td><span style="background-color: red; color: white; padding: 3px">(10+10)/2=10</span></td>
<td><span style="background-color: red; color: white; padding: 3px">(11+11)/2=11</span></td>
<td><span style="background-color: red; color: white; padding: 3px">(11+11)/2=11</span></td>
</tr>
<tr>
<td>v2</td>
<td><span style="background-color: red; color: white; padding: 3px">(10+10)/2=10</span></td>
<td>0</td>
<td>20</td>
<td>21</td>
<td>21</td>
</tr>
<tr>
<td>v3</td>
<td><span style="background-color: red; color: white; padding: 3px">(10+10)/2=10</span></td>
<td>20</td>
<td>0</td>
<td>7</td>
<td>13</td>
</tr>
<tr>
<td>v4</td>
<td><span style="background-color: red; color: white; padding: 3px">(11+11)/2=11</span></td>
<td>21</td>
<td>7</td>
<td>0</td>
<td>14</td>
</tr>
<tr>
<td>v5</td>
<td><span style="background-color: red; color: white; padding: 3px">(11+11)/2=11</span></td>
<td>21</td>
<td>13</td>
<td>14</td>
<td>0</td>
</tr>
</tbody>
</table>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>Notice how when both v0 and v1 are balded, their distances to other leaf nodes are exactly the same. So, why average it instead of just taking the distinct value? Because averaging helps with understanding the revised form of the algorithm explained in another section.</p>
</div>
<p>This algorithm is essentially removing two <a href="#BM_(neighbour)%2Fi">neighbour</a>ing leaf nodes and bringing their shared parent to the forefront (into the <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> as a leaf node). In the example above, the new leaf node M represents internal node i0 because the distance between M and i0 is 0.</p>
<p><a href="ch7_code/src/phylogeny/ExposeNeighbourParent_AdditiveExplainer.py">ch7_code/src/phylogeny/ExposeNeighbourParent_AdditiveExplainer.py</a> (lines 22 to 37):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">expose_neighbour_parent</span>(<span class="hljs-params">
        dm: DistanceMatrix[N],
        l1: N,
        l2: N,
        gen_node_id: <span class="hljs-type">Callable</span>[[], N]
</span>) -&gt; N:
    bald_distance_matrix(dm, l1)
    bald_distance_matrix(dm, l2)
    m_id = gen_node_id()
    m_dists = {x: (dm[l1, x] + dm[l2, x]) / <span class="hljs-number">2</span> <span class="hljs-keyword">for</span> x <span class="hljs-keyword">in</span> dm.leaf_ids_it()}
    m_dists[m_id] = <span class="hljs-number">0</span>
    dm.insert(m_id, m_dists)
    dm.delete(l1)
    dm.delete(l2)
    <span class="hljs-keyword">return</span> m_id
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given additive distance matrix...</p>
<table>
<thead><tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr></thead>
<tbody>
<tr>
<td>v0</td>
<td>0.0</td>
<td>13.0</td>
<td>21.0</td>
<td>21.0</td>
<td>22.0</td>
<td>22.0</td>
</tr>
<tr>
<td>v1</td>
<td>13.0</td>
<td>0.0</td>
<td>12.0</td>
<td>12.0</td>
<td>13.0</td>
<td>13.0</td>
</tr>
<tr>
<td>v2</td>
<td>21.0</td>
<td>12.0</td>
<td>0.0</td>
<td>20.0</td>
<td>21.0</td>
<td>21.0</td>
</tr>
<tr>
<td>v3</td>
<td>21.0</td>
<td>12.0</td>
<td>20.0</td>
<td>0.0</td>
<td>7.0</td>
<td>13.0</td>
</tr>
<tr>
<td>v4</td>
<td>22.0</td>
<td>13.0</td>
<td>21.0</td>
<td>7.0</td>
<td>0.0</td>
<td>14.0</td>
</tr>
<tr>
<td>v5</td>
<td>22.0</td>
<td>13.0</td>
<td>21.0</td>
<td>13.0</td>
<td>14.0</td>
<td>0.0</td>
</tr>
</tbody>
</table>
<p>... and given that v0 and v1 are neighbours, balding and merging v0 and v1 results in ...</p>
<table>
<thead><tr>
<th></th>
<th>N1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr></thead>
<tbody>
<tr>
<td>N1</td>
<td>0.0</td>
<td>10.0</td>
<td>10.0</td>
<td>11.0</td>
<td>11.0</td>
</tr>
<tr>
<td>v2</td>
<td>10.0</td>
<td>0.0</td>
<td>20.0</td>
<td>21.0</td>
<td>21.0</td>
</tr>
<tr>
<td>v3</td>
<td>10.0</td>
<td>20.0</td>
<td>0.0</td>
<td>7.0</td>
<td>13.0</td>
</tr>
<tr>
<td>v4</td>
<td>11.0</td>
<td>21.0</td>
<td>7.0</td>
<td>0.0</td>
<td>14.0</td>
</tr>
<tr>
<td>v5</td>
<td>11.0</td>
<td>21.0</td>
<td>13.0</td>
<td>14.0</td>
<td>0.0</td>
</tr>
</tbody>
</table>
</div>
<p></p>
<p>The problem with the above algorithm is that balding a <a href="#BM_(limb)%2Fi">limb</a> can't be done on a non-<a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>. That is, since a <a href="#BM_(tree)%2Fi">tree</a> doesn't exist for a non-<a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>, it's impossible to get a definitive <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a> to use for balding. In such cases, a <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a> for each path being balded can be approximated. For example, the following non-<a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> is a slightly tweaked version of the <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> in the initial example where v0 and v1 are <a href="#BM_(neighbour)%2Fi">neighbour</a>s...</p>
<table>
<thead>
<tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr>
</thead>
<tbody>
<tr>
<td>v0</td>
<td>0</td>
<td>14</td>
<td>22</td>
<td>20</td>
<td>23</td>
<td>22</td>
</tr>
<tr>
<td>v1</td>
<td>14</td>
<td>0</td>
<td>12</td>
<td>10</td>
<td>12</td>
<td>14</td>
</tr>
<tr>
<td>v2</td>
<td>22</td>
<td>12</td>
<td>0</td>
<td>20</td>
<td>22</td>
<td>20</td>
</tr>
<tr>
<td>v3</td>
<td>20</td>
<td>10</td>
<td>20</td>
<td>0</td>
<td>8</td>
<td>12</td>
</tr>
<tr>
<td>v4</td>
<td>23</td>
<td>12</td>
<td>22</td>
<td>8</td>
<td>0</td>
<td>15</td>
</tr>
<tr>
<td>v5</td>
<td>22</td>
<td>14</td>
<td>20</td>
<td>12</td>
<td>15</td>
<td>0</td>
</tr>
</tbody>
</table>
<p>
</p>
<p>Assuming v0 and v1 are still <a href="#BM_(neighbour)%2Fi">neighbour</a>s, the <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a> for v0 based on ...</p>
<ul>
<li>v2's view: 0.5 * (dist(v0,v1) + dist(v0,v2) - dist(v1,v2)) = 0.5 * (14 + 22 - 12) = 12</li>
<li>v3's view: 0.5 * (dist(v0,v1) + dist(v0,v3) - dist(v1,v3)) = 0.5 * (14 + 20 - 10) = 12</li>
<li>v4's view: 0.5 * (dist(v0,v1) + dist(v0,v4) - dist(v1,v4)) = 0.5 * (14 + 23 - 12) = 12.5</li>
<li>v5's view: 0.5 * (dist(v0,v1) + dist(v0,v5) - dist(v1,v5)) = 0.5 * (14 + 22 - 14) = 11</li>
</ul>
<p>Similarly, assuming v0 and v1 are still <a href="#BM_(neighbour)%2Fi">neighbour</a>s, the <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a> for v1 based on ...</p>
<ul>
<li>v2's view: 0.5 * (dist(v0,v1) + dist(v1,v2) - dist(v0,v2)) = 0.5 * (14 + 12 - 22) = 2</li>
<li>v3's view: 0.5 * (dist(v0,v1) + dist(v1,v3) - dist(v0,v3)) = 0.5 * (14 + 10 - 20) = 2</li>
<li>v4's view: 0.5 * (dist(v0,v1) + dist(v1,v4) - dist(v0,v4)) = 0.5 * (14 + 12 - 23) = 1.5</li>
<li>v5's view: 0.5 * (dist(v0,v1) + dist(v1,v5) - dist(v0,v5)) = 0.5 * (14 + 14 - 22) = 3</li>
</ul>
<p>
</p>
<p>Note how the <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a>s above are very close to the corresponding <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a>s in the original un-tweaked <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>: 12 for v0, 2 for v1.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>Confused where the above computations are coming from? See "view" of a <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a> is described in <a href="#BM_(Algorithms%5C%2FPhylogeny%2FFind%20Neighbour%20Limb%20Lengths%2FAverage%20Algorithm)_TOPIC%2F">Algorithms/Phylogeny/Find Neighbour Limb Lengths/Average Algorithm</a>.</p>
</div>
<p>To bald a <a href="#BM_(limb)%2Fi">limb</a> in the <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a>, each leaf node needs its view of the <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a> subtracted from its distance. Balding v0 and v1 results in ...</p>
<table>
<thead>
<tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr>
</thead>
<tbody>
<tr>
<td>v0</td>
<td><span style="background-color: red; color: white; padding: 3px">0</span></td>
<td><span style="background-color: red; color: white; padding: 3px">?????</span></td>
<td><span style="background-color: red; color: white; padding: 3px">22-12=10</span></td>
<td><span style="background-color: red; color: white; padding: 3px">20-12=8</span></td>
<td><span style="background-color: red; color: white; padding: 3px">23-12.5=10.5</span></td>
<td><span style="background-color: red; color: white; padding: 3px">22-11=11</span></td>
</tr>
<tr>
<td>v1</td>
<td><span style="background-color: red; color: white; padding: 3px">?????</span></td>
<td><span style="background-color: red; color: white; padding: 3px">0</span></td>
<td><span style="background-color: red; color: white; padding: 3px">12-2=10</span></td>
<td><span style="background-color: red; color: white; padding: 3px">10-2=8</span></td>
<td><span style="background-color: red; color: white; padding: 3px">12-1.5=10.5</span></td>
<td><span style="background-color: red; color: white; padding: 3px">14-3=11</span></td>
</tr>
<tr>
<td>v2</td>
<td><span style="background-color: red; color: white; padding: 3px">22-12=10</span></td>
<td><span style="background-color: red; color: white; padding: 3px">12-2=10</span></td>
<td>0</td>
<td>20</td>
<td>22</td>
<td>20</td>
</tr>
<tr>
<td>v3</td>
<td><span style="background-color: red; color: white; padding: 3px">20-12=8</span></td>
<td><span style="background-color: red; color: white; padding: 3px">10-2=8</span></td>
<td>20</td>
<td>0</td>
<td>8</td>
<td>12</td>
</tr>
<tr>
<td>v4</td>
<td><span style="background-color: red; color: white; padding: 3px">23-12.5=10.5</span></td>
<td><span style="background-color: red; color: white; padding: 3px">12-1.5=10.5</span></td>
<td>22</td>
<td>8</td>
<td>0</td>
<td>15</td>
</tr>
<tr>
<td>v5</td>
<td><span style="background-color: red; color: white; padding: 3px">22-11=11</span></td>
<td><span style="background-color: red; color: white; padding: 3px">14-3=11</span></td>
<td>20</td>
<td>12</td>
<td>15</td>
<td>0</td>
</tr>
</tbody>
</table>
<p>Merging together v0 and v1 happens just as it did before, by averaging together the balded distances for each leaf node...</p>
<table>
<thead>
<tr>
<th></th>
<th>M</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr>
</thead>
<tbody>
<tr>
<td>M</td>
<td><span style="background-color: red; color: white; padding: 3px">0</span></td>
<td><span style="background-color: red; color: white; padding: 3px">22-12=10</span></td>
<td><span style="background-color: red; color: white; padding: 3px">20-12=8</span></td>
<td><span style="background-color: red; color: white; padding: 3px">23-12.5=10.5</span></td>
<td><span style="background-color: red; color: white; padding: 3px">22-11=11</span></td>
</tr>
<tr>
<td>v2</td>
<td><span style="background-color: red; color: white; padding: 3px">(10+10)/2=10</span></td>
<td>0</td>
<td>20</td>
<td>22</td>
<td>20</td>
</tr>
<tr>
<td>v3</td>
<td><span style="background-color: red; color: white; padding: 3px">(8+8)/2=8</span></td>
<td>20</td>
<td>0</td>
<td>8</td>
<td>12</td>
</tr>
<tr>
<td>v4</td>
<td><span style="background-color: red; color: white; padding: 3px">(10.5+10.5)/2=10.5</span></td>
<td>22</td>
<td>8</td>
<td>0</td>
<td>15</td>
</tr>
<tr>
<td>v5</td>
<td><span style="background-color: red; color: white; padding: 3px">(11+11)/2=11</span></td>
<td>20</td>
<td>12</td>
<td>15</td>
<td>0</td>
</tr>
</tbody>
</table>
<p>Note that dist(v0,v1) is unknown in the balded matrix (denoted by a bunch of question marks). That doesn't matter because dist(v0,v1) merges into dist(M,M), which must always be 0 (the distance from anything to itself is always 0).</p>
<p><a href="ch7_code/src/phylogeny/ExposeNeighbourParent.py">ch7_code/src/phylogeny/ExposeNeighbourParent.py</a> (lines 23 to 50):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">expose_neighbour_parent</span>(<span class="hljs-params">
        dm: DistanceMatrix[N],
        l1: N,
        l2: N,
        gen_node_id: <span class="hljs-type">Callable</span>[[], N]
</span>) -&gt; N:
    <span class="hljs-comment"># bald</span>
    l1_len_views = {}
    l2_len_views = {}
    <span class="hljs-keyword">for</span> x <span class="hljs-keyword">in</span> dm.leaf_ids_it():
        <span class="hljs-keyword">if</span> x == l1 <span class="hljs-keyword">or</span> x == l2:
            <span class="hljs-keyword">continue</span>
        l1_len_views[x] = view_of_limb_length_using_neighbour(dm, l1, l2, x)
        l2_len_views[x] = view_of_limb_length_using_neighbour(dm, l2, l1, x)
    <span class="hljs-keyword">for</span> x <span class="hljs-keyword">in</span> dm.leaf_ids_it():
        <span class="hljs-keyword">if</span> x == l1 <span class="hljs-keyword">or</span> x == l2:
            <span class="hljs-keyword">continue</span>
        dm[l1, x] = dm[l1, x] - l1_len_views[x]
        dm[l2, x] = dm[l2, x] - l2_len_views[x]
    <span class="hljs-comment"># merge</span>
    m_id = gen_node_id()
    m_dists = {x: (dm[l1, x] + dm[l2, x]) / <span class="hljs-number">2</span> <span class="hljs-keyword">for</span> x <span class="hljs-keyword">in</span> dm.leaf_ids_it()}
    m_dists[m_id] = <span class="hljs-number">0</span>
    dm.insert(m_id, m_dists)
    dm.delete(l1)
    dm.delete(l2)
    <span class="hljs-keyword">return</span> m_id
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given NON- additive distance matrix...</p>
<table>
<thead><tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr></thead>
<tbody>
<tr>
<td>v0</td>
<td>0.0</td>
<td>14.0</td>
<td>22.0</td>
<td>20.0</td>
<td>23.0</td>
<td>22.0</td>
</tr>
<tr>
<td>v1</td>
<td>14.0</td>
<td>0.0</td>
<td>12.0</td>
<td>10.0</td>
<td>12.0</td>
<td>14.0</td>
</tr>
<tr>
<td>v2</td>
<td>22.0</td>
<td>12.0</td>
<td>0.0</td>
<td>20.0</td>
<td>22.0</td>
<td>20.0</td>
</tr>
<tr>
<td>v3</td>
<td>20.0</td>
<td>10.0</td>
<td>20.0</td>
<td>0.0</td>
<td>8.0</td>
<td>12.0</td>
</tr>
<tr>
<td>v4</td>
<td>23.0</td>
<td>12.0</td>
<td>22.0</td>
<td>8.0</td>
<td>0.0</td>
<td>15.0</td>
</tr>
<tr>
<td>v5</td>
<td>22.0</td>
<td>14.0</td>
<td>20.0</td>
<td>12.0</td>
<td>15.0</td>
<td>0.0</td>
</tr>
</tbody>
</table>
<p>... and given that v0 and v1 are neighbours, balding and merging v0 and v1 results in ...</p>
<table>
<thead><tr>
<th></th>
<th>N1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr></thead>
<tbody>
<tr>
<td>N1</td>
<td>0.0</td>
<td>10.0</td>
<td>8.0</td>
<td>10.5</td>
<td>11.0</td>
</tr>
<tr>
<td>v2</td>
<td>10.0</td>
<td>0.0</td>
<td>20.0</td>
<td>22.0</td>
<td>20.0</td>
</tr>
<tr>
<td>v3</td>
<td>8.0</td>
<td>20.0</td>
<td>0.0</td>
<td>8.0</td>
<td>12.0</td>
</tr>
<tr>
<td>v4</td>
<td>10.5</td>
<td>22.0</td>
<td>8.0</td>
<td>0.0</td>
<td>15.0</td>
</tr>
<tr>
<td>v5</td>
<td>11.0</td>
<td>20.0</td>
<td>12.0</td>
<td>15.0</td>
<td>0.0</td>
</tr>
</tbody>
</table>
</div>
<p></p>
<a name="H_Inverse%20Algorithm"></a><h4>Inverse Algorithm</h4>
<p><a name="BM_(Algorithms%5C%2FPhylogeny%2FMerge%20Neighbours%2FInverse%20Algorithm)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FMerge%20Neighbours%2FAverage%20Algorithm)_TOPIC%2F">Algorithms/Phylogeny/Merge Neighbours/Average Algorithm</a></li>
</ul>
</div>
<p><strong>ALGORITHM</strong>:</p>
<p>This algorithm flips around the idea of finding a <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a> to perform the same thing as the averaging algorithm. Instead of finding a <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a>, it finds everything in the path EXCEPT for the <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a>.</p>
<p>For example, consider the following <a href="#BM_(simple%20tree)%2Fi">simple tree</a> and corresponding <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> ...</p>
<p><img src="dot_fa269cbcc76261c503f77dc3a6e5b8b6.svg" alt="Dot diagram"></p>
<table>
<thead>
<tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr>
</thead>
<tbody>
<tr>
<td>v0</td>
<td>0</td>
<td>13</td>
<td>21</td>
<td>21</td>
<td>22</td>
<td>22</td>
</tr>
<tr>
<td>v1</td>
<td>13</td>
<td>0</td>
<td>12</td>
<td>12</td>
<td>13</td>
<td>13</td>
</tr>
<tr>
<td>v2</td>
<td>21</td>
<td>12</td>
<td>0</td>
<td>20</td>
<td>21</td>
<td>21</td>
</tr>
<tr>
<td>v3</td>
<td>21</td>
<td>12</td>
<td>20</td>
<td>0</td>
<td>7</td>
<td>13</td>
</tr>
<tr>
<td>v4</td>
<td>22</td>
<td>13</td>
<td>21</td>
<td>7</td>
<td>0</td>
<td>14</td>
</tr>
<tr>
<td>v5</td>
<td>22</td>
<td>13</td>
<td>21</td>
<td>13</td>
<td>14</td>
<td>0</td>
</tr>
</tbody>
</table>
<p>Assume that you hadn't already seen the <a href="#BM_(tree)%2Fi">tree</a> but somehow already knew that v0 and v1 are <a href="#BM_(neighbour)%2Fi">neighbour</a>s. Consider what happens when you use the standard <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a> algorithm to find v0's <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a> from v3 ...</p>
<ul>
<li>len(v0) = (dist(v0,v1) + dist(v0,v3) - dist(v1,v3)) / 2</li>
<li>len(v0) = (13 + 21 - 12) / 2</li>
<li>len(v0) = 11</li>
</ul>
<p><img src="dot_5d4a13520a122532adfe3c8c79f81d70.svg" alt="Dot diagram"></p>
<p>By slightly tweaking the terms in the expression above, it's possible to instead find the distance between the <a href="#BM_(neighbour)%2Fi">neighbour</a>ing pair's parent (i0) and v3 ...</p>
<ul>
<li>inverse_len(v3) = (dist(v0,v3) + dist(v1,v3) - dist(v0,v1)) / 2</li>
<li>inverse_len(v3) = (21 + 12 - 13) / 2</li>
<li>inverse_len(v3) = 10 = dist(v0,v3) - len(v0) = dist(v1,v3) - len(v1)</li>
</ul>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>All the same distances are being used in this new computation, they're just being added / subtracted in a different order.</p>
</div>
<p><img src="dot_5fbdc713d9750b2a93916259f64882a6.svg" alt="Dot diagram"></p>
<p>The inverse_len function above in abstracted form is 0.5 * (dist(L,X) + dist(N,X) - dist(L,N)), where ...</p>
<ul>
<li>L and N are <a href="#BM_(neighbour)%2Fi">neighbour</a>s.</li>
<li>X is a leaf node that isn't L or N.</li>
</ul>
<p>Note that the distance calculated by the inverse_len example above is exactly the same distance you'd get for v3 when balding and merging v0 and v1 using the averaging algorithm. That is, instead of using the averaging algorithm to bald and merge the <a href="#BM_(neighbour)%2Fi">neighbour</a>ing pair, you can just inject inverse_len's result for each leaf node into the <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> and remove the <a href="#BM_(neighbour)%2Fi">neighbour</a>ing pair.</p>
<p>The inverse_len for leaf node ...</p>
<ul>
<li>v2: 0.5 * (dist(v0,v2) + dist(v1,v2) - dist(v0,v1)) = 0.5 * (21 + 12 - 13) = 10</li>
<li>v3: 0.5 * (dist(v0,v3) + dist(v1,v3) - dist(v0,v1)) = 0.5 * (21 + 12 - 13) = 10</li>
<li>v4: 0.5 * (dist(v0,v4) + dist(v1,v4) - dist(v0,v1)) = 0.5 * (22 + 13 - 13) = 11</li>
<li>v5: 0.5 * (dist(v0,v5) + dist(v1,v5) - dist(v0,v1)) = 0.5 * (22 + 13 - 13) = 11</li>
</ul>
<table>
<thead>
<tr>
<th></th>
<th>M</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr>
</thead>
<tbody>
<tr>
<td>M</td>
<td><span style="background-color: red; color: white; padding: 3px">0</span></td>
<td><span style="background-color: red; color: white; padding: 3px">(21+12-13)/2=10</span></td>
<td><span style="background-color: red; color: white; padding: 3px">(21+12-13)/2=10</span></td>
<td><span style="background-color: red; color: white; padding: 3px">(21+12-13)/2=10</span></td>
<td><span style="background-color: red; color: white; padding: 3px">(21+12-13)/2=10</span></td>
</tr>
<tr>
<td>v2</td>
<td><span style="background-color: red; color: white; padding: 3px">(21+12-13)/2=10</span></td>
<td>0</td>
<td>20</td>
<td>21</td>
<td>21</td>
</tr>
<tr>
<td>v3</td>
<td><span style="background-color: red; color: white; padding: 3px">(21+12-13)/2=10</span></td>
<td>20</td>
<td>0</td>
<td>7</td>
<td>13</td>
</tr>
<tr>
<td>v4</td>
<td><span style="background-color: red; color: white; padding: 3px">(21+12-13)/2=10</span></td>
<td>21</td>
<td>7</td>
<td>0</td>
<td>14</td>
</tr>
<tr>
<td>v5</td>
<td><span style="background-color: red; color: white; padding: 3px">(22+13-13)/2=11</span></td>
<td>21</td>
<td>13</td>
<td>14</td>
<td>0</td>
</tr>
</tbody>
</table>
<p><img src="dot_3d01f4a0fab8e57bf0bfa96673b676e3.svg" alt="Dot diagram"></p>
<p>In fact, inverse_len is just the simplified expression form of the averaging algorithm. Consider the steps you have to go through for each leaf node to bald and merge the <a href="#BM_(neighbour)%2Fi">neighbour</a>ing pair v0 and v1 using the averaging algorithm. For example, to figure out the balded distance between v3 and the merged node, the steps are ...</p>
<p></p>
<ol>
<li>
<p>Get v3's view of v0's limb length:</p>
<p>len(v0) = 0.5 * (dist(v0,v1) + dist(v0,v3) - dist(v1,v3))</p>
</li>
<li>
<p>Get v3's view of v1's limb length:</p>
<p>len(v1) = 0.5 * (dist(v1,v0) + dist(v1,v3) - dist(v0,v3))</p>
</li>
<li>
<p>Bald v0 for v3 using step 1's result:</p>
<p>bald_dist(v0,v5) = dist(v0,v3) - len(v0)</p>
</li>
<li>
<p>Bald v1 for v3 using step 2's result:</p>
<p>bald_dist(v1,v5) = dist(v1,v3) - len(v1)</p>
</li>
<li>
<p>Average results from step 3 and 4 to produce the merged node's distance for v3:</p>
<p>merge(v0,v1) = (bald_dist(v0,v5) + bald_dist(v1,v5)) / 2</p>
</li>
</ol>
<p></p>
<p>Consider what happens when you combine all of the above steps together as a single expression ...</p>
<div style="overflow: auto"><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><mfrac><mrow><msub><mi>D</mi><mrow><mi>v</mi><mn>0</mn><mo separator="true">,</mo><mi>v</mi><mn>3</mn></mrow></msub><mo>−</mo><mo stretchy="false">(</mo><mn>0.5</mn><mo>⋅</mo><mo stretchy="false">(</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>0</mn><mo separator="true">,</mo><mi>v</mi><mn>1</mn></mrow></msub><mo>+</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>0</mn><mo separator="true">,</mo><mi>v</mi><mn>3</mn></mrow></msub><mo>−</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>1</mn><mo separator="true">,</mo><mi>v</mi><mn>3</mn></mrow></msub><mo stretchy="false">)</mo><mo stretchy="false">)</mo><mo>+</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>1</mn><mo separator="true">,</mo><mi>v</mi><mn>3</mn></mrow></msub><mo>−</mo><mo stretchy="false">(</mo><mn>0.5</mn><mo>⋅</mo><mo stretchy="false">(</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>1</mn><mo separator="true">,</mo><mi>v</mi><mn>0</mn></mrow></msub><mo>+</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>1</mn><mo separator="true">,</mo><mi>v</mi><mn>3</mn></mrow></msub><mo>−</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>0</mn><mo separator="true">,</mo><mi>v</mi><mn>3</mn></mrow></msub><mo stretchy="false">)</mo><mo stretchy="false">)</mo></mrow><mn>2</mn></mfrac></mrow><annotation encoding="application/x-tex">\frac{D_{v0,v3} - (0.5 \cdot (D_{v0,v1} + D_{v0,v3} - D_{v1,v3})) + D_{v1,v3} - (0.5 \cdot (D_{v1,v0} + D_{v1,v3} - D_{v0,v3}))}{2}
</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:2.113em;vertical-align:-0.686em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.427em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">0</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">3</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mopen">(</span><span class="mord">0</span><span class="mord">.</span><span class="mord">5</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">0</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">1</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">0</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">3</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">3</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mclose">)</span><span class="mclose">)</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">3</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mopen">(</span><span class="mord">0</span><span class="mord">.</span><span class="mord">5</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">0</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">3</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">0</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">3</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mclose">)</span><span class="mclose">)</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span></span></div><p>Simplifying that expression results in ...</p>
<div style="overflow: auto"><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><mfrac><mrow><msub><mi>D</mi><mrow><mi>v</mi><mn>0</mn><mo separator="true">,</mo><mi>v</mi><mn>3</mn></mrow></msub><mo>−</mo><mo stretchy="false">(</mo><mn>0.5</mn><mo>⋅</mo><mo stretchy="false">(</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>0</mn><mo separator="true">,</mo><mi>v</mi><mn>1</mn></mrow></msub><mo>+</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>0</mn><mo separator="true">,</mo><mi>v</mi><mn>3</mn></mrow></msub><mo>−</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>1</mn><mo separator="true">,</mo><mi>v</mi><mn>3</mn></mrow></msub><mo stretchy="false">)</mo><mo stretchy="false">)</mo><mo>+</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>1</mn><mo separator="true">,</mo><mi>v</mi><mn>3</mn></mrow></msub><mo>−</mo><mo stretchy="false">(</mo><mn>0.5</mn><mo>⋅</mo><mo stretchy="false">(</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>1</mn><mo separator="true">,</mo><mi>v</mi><mn>0</mn></mrow></msub><mo>+</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>1</mn><mo separator="true">,</mo><mi>v</mi><mn>3</mn></mrow></msub><mo>−</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>0</mn><mo separator="true">,</mo><mi>v</mi><mn>3</mn></mrow></msub><mo stretchy="false">)</mo><mo stretchy="false">)</mo></mrow><mn>2</mn></mfrac><mspace linebreak="newline" height="0.5em"></mspace><mfrac><mrow><msub><mi>D</mi><mrow><mi>v</mi><mn>0</mn><mo separator="true">,</mo><mi>v</mi><mn>3</mn></mrow></msub><mo>−</mo><mo stretchy="false">(</mo><mn>0.5</mn><mo>⋅</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>0</mn><mo separator="true">,</mo><mi>v</mi><mn>1</mn></mrow></msub><mo>+</mo><mn>0.5</mn><mo>⋅</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>0</mn><mo separator="true">,</mo><mi>v</mi><mn>3</mn></mrow></msub><mo>−</mo><mn>0.5</mn><mo>⋅</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>1</mn><mo separator="true">,</mo><mi>v</mi><mn>3</mn></mrow></msub><mo stretchy="false">)</mo><mo stretchy="false">)</mo><mo>+</mo><mo stretchy="false">(</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>1</mn><mo separator="true">,</mo><mi>v</mi><mn>3</mn></mrow></msub><mo>−</mo><mo stretchy="false">(</mo><mn>0.5</mn><mo>⋅</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>1</mn><mo separator="true">,</mo><mi>v</mi><mn>0</mn></mrow></msub><mo>+</mo><mn>0.5</mn><mo>⋅</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>1</mn><mo separator="true">,</mo><mi>v</mi><mn>3</mn></mrow></msub><mo>−</mo><mn>0.5</mn><mo>⋅</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>0</mn><mo separator="true">,</mo><mi>v</mi><mn>3</mn></mrow></msub><mo stretchy="false">)</mo></mrow><mn>2</mn></mfrac><mspace linebreak="newline" height="0.5em"></mspace><mfrac><mrow><msub><mi>D</mi><mrow><mi>v</mi><mn>0</mn><mo separator="true">,</mo><mi>v</mi><mn>3</mn></mrow></msub><mo>−</mo><mn>0.5</mn><mo>⋅</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>0</mn><mo separator="true">,</mo><mi>v</mi><mn>1</mn></mrow></msub><mo>−</mo><mn>0.5</mn><mo>⋅</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>0</mn><mo separator="true">,</mo><mi>v</mi><mn>3</mn></mrow></msub><mo>+</mo><mn>0.5</mn><mo>⋅</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>1</mn><mo separator="true">,</mo><mi>v</mi><mn>3</mn></mrow></msub><mo>+</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>1</mn><mo separator="true">,</mo><mi>v</mi><mn>3</mn></mrow></msub><mo>−</mo><mn>0.5</mn><mo>⋅</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>1</mn><mo separator="true">,</mo><mi>v</mi><mn>0</mn></mrow></msub><mo>−</mo><mn>0.5</mn><mo>⋅</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>1</mn><mo separator="true">,</mo><mi>v</mi><mn>3</mn></mrow></msub><mo>+</mo><mn>0.5</mn><mo>⋅</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>0</mn><mo separator="true">,</mo><mi>v</mi><mn>3</mn></mrow></msub></mrow><mn>2</mn></mfrac><mspace linebreak="newline" height="0.5em"></mspace><mfrac><mrow><msub><mi>D</mi><mrow><mi>v</mi><mn>0</mn><mo separator="true">,</mo><mi>v</mi><mn>3</mn></mrow></msub><mo>−</mo><mn>0.5</mn><mo>⋅</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>0</mn><mo separator="true">,</mo><mi>v</mi><mn>1</mn></mrow></msub><mo>+</mo><mn>0.5</mn><mo>⋅</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>1</mn><mo separator="true">,</mo><mi>v</mi><mn>3</mn></mrow></msub><mo>+</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>1</mn><mo separator="true">,</mo><mi>v</mi><mn>3</mn></mrow></msub><mo>−</mo><mn>0.5</mn><mo>⋅</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>1</mn><mo separator="true">,</mo><mi>v</mi><mn>0</mn></mrow></msub><mo>−</mo><mn>0.5</mn><mo>⋅</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>1</mn><mo separator="true">,</mo><mi>v</mi><mn>3</mn></mrow></msub></mrow><mn>2</mn></mfrac><mspace linebreak="newline" height="0.5em"></mspace><mfrac><mrow><msub><mi>D</mi><mrow><mi>v</mi><mn>0</mn><mo separator="true">,</mo><mi>v</mi><mn>3</mn></mrow></msub><mo>−</mo><mn>0.5</mn><mo>⋅</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>0</mn><mo separator="true">,</mo><mi>v</mi><mn>1</mn></mrow></msub><mo>+</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>1</mn><mo separator="true">,</mo><mi>v</mi><mn>3</mn></mrow></msub><mo>−</mo><mn>0.5</mn><mo>⋅</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>1</mn><mo separator="true">,</mo><mi>v</mi><mn>0</mn></mrow></msub></mrow><mn>2</mn></mfrac><mspace linebreak="newline" height="0.5em"></mspace><mfrac><mrow><msub><mi>D</mi><mrow><mi>v</mi><mn>0</mn><mo separator="true">,</mo><mi>v</mi><mn>3</mn></mrow></msub><mo>+</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>1</mn><mo separator="true">,</mo><mi>v</mi><mn>3</mn></mrow></msub><mo>−</mo><mn>1</mn><mo>⋅</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>1</mn><mo separator="true">,</mo><mi>v</mi><mn>0</mn></mrow></msub></mrow><mn>2</mn></mfrac><mspace linebreak="newline" height="0.5em"></mspace><mfrac><mrow><msub><mi>D</mi><mrow><mi>v</mi><mn>0</mn><mo separator="true">,</mo><mi>v</mi><mn>3</mn></mrow></msub><mo>+</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>1</mn><mo separator="true">,</mo><mi>v</mi><mn>3</mn></mrow></msub><mo>−</mo><msub><mi>D</mi><mrow><mi>v</mi><mn>1</mn><mo separator="true">,</mo><mi>v</mi><mn>0</mn></mrow></msub></mrow><mn>2</mn></mfrac></mrow><annotation encoding="application/x-tex">\frac{D_{v0,v3} - (0.5 \cdot (D_{v0,v1} + D_{v0,v3} - D_{v1,v3})) + D_{v1,v3} - (0.5 \cdot (D_{v1,v0} + D_{v1,v3} - D_{v0,v3}))}{2}
\\[0.5em]
\frac{D_{v0,v3} - (0.5 \cdot D_{v0,v1} + 0.5 \cdot D_{v0,v3} - 0.5 \cdot D_{v1,v3})) + (D_{v1,v3} - (0.5 \cdot D_{v1,v0} + 0.5 \cdot D_{v1,v3} - 0.5 \cdot D_{v0,v3})}{2}
\\[0.5em]
\frac{D_{v0,v3} - 0.5 \cdot D_{v0,v1} - 0.5 \cdot D_{v0,v3} + 0.5 \cdot D_{v1,v3} + D_{v1,v3} - 0.5 \cdot D_{v1,v0} - 0.5 \cdot D_{v1,v3} + 0.5 \cdot D_{v0,v3}}{2}
\\[0.5em]
\frac{D_{v0,v3} - 0.5 \cdot D_{v0,v1} + 0.5 \cdot D_{v1,v3} + D_{v1,v3} - 0.5 \cdot D_{v1,v0} - 0.5 \cdot D_{v1,v3}}{2}
\\[0.5em]
\frac{D_{v0,v3} - 0.5 \cdot D_{v0,v1} + D_{v1,v3} - 0.5 \cdot D_{v1,v0}}{2}
\\[0.5em]
\frac{D_{v0,v3} + D_{v1,v3} - 1 \cdot D_{v1,v0}}{2}
\\[0.5em]
\frac{D_{v0,v3} + D_{v1,v3} - D_{v1,v0}}{2}
</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:2.113em;vertical-align:-0.686em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.427em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">0</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">3</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mopen">(</span><span class="mord">0</span><span class="mord">.</span><span class="mord">5</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">0</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">1</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">0</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">3</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">3</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mclose">)</span><span class="mclose">)</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">3</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mopen">(</span><span class="mord">0</span><span class="mord">.</span><span class="mord">5</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">0</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">3</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">0</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">3</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mclose">)</span><span class="mclose">)</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span><span class="mspace newline" style="margin-top:0.5em;"></span><span class="base"><span class="strut" style="height:2.113em;vertical-align:-0.686em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.427em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">0</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">3</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mopen">(</span><span class="mord">0</span><span class="mord">.</span><span class="mord">5</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">0</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">1</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">0</span><span class="mord">.</span><span class="mord">5</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">0</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">3</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">0</span><span class="mord">.</span><span class="mord">5</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">3</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mclose">)</span><span class="mclose">)</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">3</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mopen">(</span><span class="mord">0</span><span class="mord">.</span><span class="mord">5</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">0</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">0</span><span class="mord">.</span><span class="mord">5</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">3</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">0</span><span class="mord">.</span><span class="mord">5</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">0</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">3</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mclose">)</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span><span class="mspace newline" style="margin-top:0.5em;"></span><span class="base"><span class="strut" style="height:2.04633em;vertical-align:-0.686em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.36033em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">0</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">3</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">0</span><span class="mord">.</span><span class="mord">5</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">0</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">1</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">0</span><span class="mord">.</span><span class="mord">5</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">0</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">3</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">0</span><span class="mord">.</span><span class="mord">5</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">3</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">3</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">0</span><span class="mord">.</span><span class="mord">5</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">0</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">0</span><span class="mord">.</span><span class="mord">5</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">3</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">0</span><span class="mord">.</span><span class="mord">5</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">0</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">3</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span><span class="mspace newline" style="margin-top:0.5em;"></span><span class="base"><span class="strut" style="height:2.04633em;vertical-align:-0.686em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.36033em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">0</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">3</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">0</span><span class="mord">.</span><span class="mord">5</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">0</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">1</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">0</span><span class="mord">.</span><span class="mord">5</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">3</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">3</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">0</span><span class="mord">.</span><span class="mord">5</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">0</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">0</span><span class="mord">.</span><span class="mord">5</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">3</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span><span class="mspace newline" style="margin-top:0.5em;"></span><span class="base"><span class="strut" style="height:2.04633em;vertical-align:-0.686em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.36033em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">0</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">3</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">0</span><span class="mord">.</span><span class="mord">5</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">0</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">1</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">3</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">0</span><span class="mord">.</span><span class="mord">5</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">0</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span><span class="mspace newline" style="margin-top:0.5em;"></span><span class="base"><span class="strut" style="height:2.04633em;vertical-align:-0.686em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.36033em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">0</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">3</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">3</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">1</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">0</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span><span class="mspace newline" style="margin-top:0.5em;"></span><span class="base"><span class="strut" style="height:2.04633em;vertical-align:-0.686em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.36033em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">2</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">0</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">3</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">3</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.301108em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">1</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span><span class="mord mtight">0</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span></span></div><p>The simplified form of the expression is exactly the computation that the inverse_len example ran for v3 ...</p>
<ul>
<li>inverse_len(v3) = (dist(v0,v3) + dist(v1,v3) - dist(v0,v1)) / 2</li>
<li>inverse_len(v3) = (21 + 12 - 13) / 2</li>
<li>inverse_len(v3) = 10 = dist(v0,v3) - len(v0) = dist(v1,v3) - len(v1)</li>
</ul>
<p>Since this algorithm is doing the same thing as the averaging algorithm, it'll work on non-<a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrices</a> in the exact same way as the averaging algorithm. It's just the averaging algorithm in simplified / optimized form. For example, the following non-<a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> is a slightly tweaked version of the <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> in the initial example where v0 and v1 are <a href="#BM_(neighbour)%2Fi">neighbour</a>s...</p>
<table>
<thead>
<tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr>
</thead>
<tbody>
<tr>
<td>v0</td>
<td>0</td>
<td>14</td>
<td>22</td>
<td>20</td>
<td>23</td>
<td>22</td>
</tr>
<tr>
<td>v1</td>
<td>14</td>
<td>0</td>
<td>12</td>
<td>10</td>
<td>12</td>
<td>14</td>
</tr>
<tr>
<td>v2</td>
<td>22</td>
<td>12</td>
<td>0</td>
<td>20</td>
<td>22</td>
<td>20</td>
</tr>
<tr>
<td>v3</td>
<td>20</td>
<td>10</td>
<td>20</td>
<td>0</td>
<td>8</td>
<td>12</td>
</tr>
<tr>
<td>v4</td>
<td>23</td>
<td>12</td>
<td>22</td>
<td>8</td>
<td>0</td>
<td>15</td>
</tr>
<tr>
<td>v5</td>
<td>22</td>
<td>14</td>
<td>20</td>
<td>12</td>
<td>15</td>
<td>0</td>
</tr>
</tbody>
</table>
<p>Assuming v0 and v1 are still <a href="#BM_(neighbour)%2Fi">neighbour</a>s, the merged distance for ...</p>
<ul>
<li>v2: 0.5 * (dist(v0,v2) + dist(v1,v2) - dist(v0,v1)) = 0.5 * (22 + 12 - 14) = 10</li>
<li>v3: 0.5 * (dist(v0,v3) + dist(v1,v3) - dist(v0,v1)) = 0.5 * (20 + 10 - 14) = 8</li>
<li>v4: 0.5 * (dist(v0,v4) + dist(v1,v4) - dist(v0,v1)) = 0.5 * (23 + 12 - 14) = 10.5</li>
<li>v5: 0.5 * (dist(v0,v5) + dist(v1,v5) - dist(v0,v1)) = 0.5 * (22 + 14 - 14) = 11</li>
</ul>
<table>
<thead>
<tr>
<th></th>
<th>M</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr>
</thead>
<tbody>
<tr>
<td>M</td>
<td><span style="background-color: red; color: white; padding: 3px">0</span></td>
<td><span style="background-color: red; color: white; padding: 3px">(22+12-14)/2=10</span></td>
<td><span style="background-color: red; color: white; padding: 3px">(20+10-14)/2=8</span></td>
<td><span style="background-color: red; color: white; padding: 3px">(23+12-14)/2=10.5</span></td>
<td><span style="background-color: red; color: white; padding: 3px">(22+14-14)/2=11</span></td>
</tr>
<tr>
<td>v2</td>
<td><span style="background-color: red; color: white; padding: 3px">(22+12-14)/2=10</span></td>
<td>0</td>
<td>20</td>
<td>22</td>
<td>20</td>
</tr>
<tr>
<td>v3</td>
<td><span style="background-color: red; color: white; padding: 3px">(20+10-14)/2=8</span></td>
<td>20</td>
<td>0</td>
<td>8</td>
<td>12</td>
</tr>
<tr>
<td>v4</td>
<td><span style="background-color: red; color: white; padding: 3px">(23+12-14)/2=10.5</span></td>
<td>22</td>
<td>8</td>
<td>0</td>
<td>15</td>
</tr>
<tr>
<td>v5</td>
<td><span style="background-color: red; color: white; padding: 3px">(22+14-14)/2=11</span></td>
<td>20</td>
<td>12</td>
<td>15</td>
<td>0</td>
</tr>
</tbody>
</table>
<p><a href="ch7_code/src/phylogeny/ExposeNeighbourParent_Optimized.py">ch7_code/src/phylogeny/ExposeNeighbourParent_Optimized.py</a> (lines 22 to 35):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">expose_neighbour_parent</span>(<span class="hljs-params">
        dm: DistanceMatrix[N],
        l1: N,
        l2: N,
        gen_node_id: <span class="hljs-type">Callable</span>[[], N]
</span>) -&gt; N:
    m_id = gen_node_id()
    m_dists = {x: (dm[l1, x] + dm[l2, x] - dm[l1, l2]) / <span class="hljs-number">2</span> <span class="hljs-keyword">for</span> x <span class="hljs-keyword">in</span> dm.leaf_ids_it()}
    m_dists[m_id] = <span class="hljs-number">0</span>
    dm.insert(m_id, m_dists)
    dm.delete(l1)
    dm.delete(l2)
    <span class="hljs-keyword">return</span> m_id
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given NON- additive distance matrix...</p>
<table>
<thead><tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr></thead>
<tbody>
<tr>
<td>v0</td>
<td>0.0</td>
<td>14.0</td>
<td>22.0</td>
<td>20.0</td>
<td>23.0</td>
<td>22.0</td>
</tr>
<tr>
<td>v1</td>
<td>14.0</td>
<td>0.0</td>
<td>12.0</td>
<td>10.0</td>
<td>12.0</td>
<td>14.0</td>
</tr>
<tr>
<td>v2</td>
<td>22.0</td>
<td>12.0</td>
<td>0.0</td>
<td>20.0</td>
<td>22.0</td>
<td>20.0</td>
</tr>
<tr>
<td>v3</td>
<td>20.0</td>
<td>10.0</td>
<td>20.0</td>
<td>0.0</td>
<td>8.0</td>
<td>12.0</td>
</tr>
<tr>
<td>v4</td>
<td>23.0</td>
<td>12.0</td>
<td>22.0</td>
<td>8.0</td>
<td>0.0</td>
<td>15.0</td>
</tr>
<tr>
<td>v5</td>
<td>22.0</td>
<td>14.0</td>
<td>20.0</td>
<td>12.0</td>
<td>15.0</td>
<td>0.0</td>
</tr>
</tbody>
</table>
<p>... and given that v0 and v1 are neighbours, balding and merging v0 and v1 results in ...</p>
<table>
<thead><tr>
<th></th>
<th>N1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr></thead>
<tbody>
<tr>
<td>N1</td>
<td>0.0</td>
<td>10.0</td>
<td>8.0</td>
<td>10.5</td>
<td>11.0</td>
</tr>
<tr>
<td>v2</td>
<td>10.0</td>
<td>0.0</td>
<td>20.0</td>
<td>22.0</td>
<td>20.0</td>
</tr>
<tr>
<td>v3</td>
<td>8.0</td>
<td>20.0</td>
<td>0.0</td>
<td>8.0</td>
<td>12.0</td>
</tr>
<tr>
<td>v4</td>
<td>10.5</td>
<td>22.0</td>
<td>8.0</td>
<td>0.0</td>
<td>15.0</td>
</tr>
<tr>
<td>v5</td>
<td>11.0</td>
<td>20.0</td>
<td>12.0</td>
<td>15.0</td>
<td>0.0</td>
</tr>
</tbody>
</table>
</div>
<p></p>
<a name="H_Distance%20Matrix%20to%20Tree"></a><h3>Distance Matrix to Tree</h3>
<p><a name="BM_(Algorithms%5C%2FPhylogeny%2FDistance%20Matrix%20to%20Tree)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FTree%20to%20Additive%20Distance%20Matrix)_TOPIC%2F">Algorithms/Phylogeny/Tree to Additive Distance Matrix</a></li>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FTree%20to%20Simple%20Tree)_TOPIC%2F">Algorithms/Phylogeny/Tree to Simple Tree</a></li>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FAdditive%20Distance%20Matrix%20Cardinality)_TOPIC%2F">Algorithms/Phylogeny/Additive Distance Matrix Cardinality</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: Given a <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a>, convert that <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> into an <a href="#BM_(phylogenetic%20tree%7Cphylogeny%20tree%7Cphylogenies%7Cevolutionary%20tree)%2Fi">evolutionary tree</a>. Different algorithms are presented that either ...</p>
<ul>
<li>find the unique <a href="#BM_(simple%20tree)%2Fi">simple tree</a> for an <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>,</li>
<li>approximate a <a href="#BM_(simple%20tree)%2Fi">simple tree</a> for a non-<a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>,</li>
<li>approximate a <a href="#BM_(tree)%2Fi">tree</a> for a <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> regardless of if it's additive or not.</li>
</ul>
<p><strong>WHY</strong>: Recall that converting a <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> to a <a href="#BM_(tree)%2Fi">tree</a> is the end goal of <a href="#BM_(phylogeny%7Cphylogenetic)%2Fi">phylogeny</a>. Given the distances between a set of known / present-day entities, these algorithms will infer their evolutionary relationships.</p>
<a name="H_UPGMA%20Algorithm"></a><h4>UPGMA Algorithm</h4>
<p><a name="BM_(Algorithms%5C%2FPhylogeny%2FDistance%20Matrix%20to%20Tree%2FUPGMA%20Algorithm)_TOPIC%2F"></a><strong></strong></p>
<p><strong>ALGORITHM</strong>:</p>
<p><a href="#BM_(unweighted%20pair%20group%20method%20with%20arithmetic%20mean%7CUPGMA)%2Fi">Unweighted pair group method with arithmetic mean</a> (<a href="#BM_(unweighted%20pair%20group%20method%20with%20arithmetic%20mean%7CUPGMA)%2Fi">UPGMA</a>) is a heuristic algorithm used to estimate a binary <a href="#BM_(ultrametric%20tree%7Cultrametric)%2Fi">ultrametric tree</a> for some <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a>.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>A binary <a href="#BM_(ultrametric%20tree%7Cultrametric)%2Fi">ultrametric tree</a> is an <a href="#BM_(ultrametric%20tree%7Cultrametric)%2Fi">ultrametric tree</a> where each internal node only branches to two children. In other words, a binary <a href="#BM_(ultrametric%20tree%7Cultrametric)%2Fi">ultrametric tree</a> is a <a href="#BM_(rooted%20binary%20tree)%2Fi">rooted binary tree</a> where all leaf nodes are equidistant from the root.</p>
</div>
<p>The algorithm assumes that the rate of mutation is consistent (<a href="#BM_(molecular%20clock)%2Fi">molecular clock</a>). For example, ...</p>
<ul>
<li>every minute, around n of m nucleotides mutate.</li>
<li>every hour, around n <a href="#BM_(genome%20rearrangement%7Cchromosomal%20rearrangement%7Cchromosome%20rearrangement)%2Fi">genome rearrangement</a> reversals occur per <a href="#BM_(genome)%2Fi">genome</a> segment of size m.</li>
<li>etc..</li>
</ul>
<p>This assumption is what makes the <a href="#BM_(tree)%2Fi">tree</a> <a href="#BM_(ultrametric%20tree%7Cultrametric)%2Fi">ultrametric</a>. A set of present day species (leaf nodes) are assumed to all have the same amount of mutation (distance) from their shared ancestor (shared internal node).</p>
<p><img src="svgbob_9c45e3a8fedf42d570c570b7c417173e7ed57970.svg" alt="Kroki diagram output"></p>
<p>For example, assume the present year is 2000. Four present day species share a common ancestor from the year 1800. The age difference between each of these four species and their shared ancestor is the same: 2000 - 1800 = 200 years.</p>
<p>Since the rate of mutation is assumed to be consistent, all four present day species should have roughly the same amount of mutation when compared against their shared ancestor: 200 years worth of mutation. Assume the number of <a href="#BM_(genome%20rearrangement%7Cchromosomal%20rearrangement%7Cchromosome%20rearrangement)%2Fi">genome rearrangement</a> reversals is being used as the measure of mutation. If the rate of reversals expected per 100 years is 2, the distance between each of the four present day species and their shared ancestor would be 4: 2 reversals per century * 2 centuries = 4 reversals.</p>
<p><img src="svgbob_55b2b7d647c259a686085e583dce32da8851a4b6.svg" alt="Kroki diagram output"></p>
<p>In the example above, ...</p>
<ul>
<li>present day species a and b (leaf nodes) have the same amount of mutation (distance) from shared ancestor f (shared internal node): 1.5 reversals</li>
<li>present day species c and d (leaf nodes) have the same amount of mutation (distance) from shared ancestor e (shared internal node): 1 reversal</li>
<li>present day species a and c (leaf nodes) have the same amount of mutation (distance) from shared ancestor g (shared internal node): 4 reversals</li>
<li>present day species a and d (leaf nodes) have the same amount of mutation (distance) from shared ancestor g (shared internal node): 4 reversals</li>
<li>etc..</li>
</ul>
<p>Given a <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a>, <a href="#BM_(unweighted%20pair%20group%20method%20with%20arithmetic%20mean%7CUPGMA)%2Fi">UPGMA</a> estimates an <a href="#BM_(ultrametric%20tree%7Cultrametric)%2Fi">ultrametric tree</a> for that matrix by iteratively picking two available nodes and connecting them with a new internal node, where available node is defined as a node without a parent. The process stops once a single available node remains (that node being the root node).</p>
<p><img src="svgbob_3d816b9d61032209afd8294044f1caf010c46928.svg" alt="Kroki diagram output"></p>
<p>Which two nodes are selected per iteration is based on <a href="#BM_(cluster)%2Fi">cluster</a>ing. In the beginning, each leaf node in the <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> is its own <a href="#BM_(cluster)%2Fi">cluster</a>: Ca={a}, Cb={b}, Cc={c}, and Cd={d}.</p>
<table>
<thead>
<tr>
<th></th>
<th>Ca={a}</th>
<th>Cb={b}</th>
<th>Cc={c}</th>
<th>Cd={d}</th>
</tr>
</thead>
<tbody>
<tr>
<td>Ca={a}</td>
<td>0</td>
<td>3</td>
<td>4</td>
<td>3</td>
</tr>
<tr>
<td>Cb={b}</td>
<td>3</td>
<td>0</td>
<td>4</td>
<td>5</td>
</tr>
<tr>
<td>Cc={c}</td>
<td>4</td>
<td>4</td>
<td>0</td>
<td>2</td>
</tr>
<tr>
<td>Cd={d}</td>
<td>3</td>
<td>5</td>
<td>2</td>
<td>0</td>
</tr>
</tbody>
</table>
<p>The two <a href="#BM_(cluster)%2Fi">cluster</a>s with the minimum distance are chosen to connect in the <a href="#BM_(tree)%2Fi">tree</a>. In the example <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> above, the minimum distance is between Cc and Cd (distance of 2), meaning that Cc and Cd should be <a href="#BM_(connected)%2Fi">connected</a> together with a new internal node.</p>
<p><img src="svgbob_33e1b85dbaff91ad5a6f7d584fbcd812e2c84362.svg" alt="Kroki diagram output"></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>Note what's happening here. The assumption being made that the leaf nodes for the minimum <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> value are always <a href="#BM_(neighbour)%2Fi">neighbour</a>s. Not always true, but probably good enough as a starting point. For example, the following <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> and <a href="#BM_(tree)%2Fi">tree</a> would identify v0 and v2 as <a href="#BM_(neighbour)%2Fi">neighbour</a>s when in fact they aren't ...</p>
<p><img src="svgbob_d8273b1705b944b7eedb9a09ae7d33d6c994f20d.svg" alt="Kroki diagram output"></p>
<table>
<thead>
<tr>
<th></th>
<th>a</th>
<th>b</th>
<th>c</th>
<th>d</th>
</tr>
</thead>
<tbody>
<tr>
<td>a</td>
<td>0</td>
<td>91</td>
<td>3</td>
<td>92</td>
</tr>
<tr>
<td>b</td>
<td>91</td>
<td>0</td>
<td>92</td>
<td>181</td>
</tr>
<tr>
<td>c</td>
<td>3</td>
<td>92</td>
<td>0</td>
<td>91</td>
</tr>
<tr>
<td>d</td>
<td>92</td>
<td>181</td>
<td>91</td>
<td>0</td>
</tr>
</tbody>
</table>
<p>It may be a good idea to use <a href="#BM_(Algorithms%5C%2FPhylogeny%2FFind%20Neighbours)_TOPIC%2F">Algorithms/Phylogeny/Find Neighbours</a> to short circuit this restriction, possibly producing a better heuristic. But, the original algorithm doesn't call for it.</p>
</div>
<p>This new internal node represents a shared ancestor. The distance of 2 represents the total amount of mutation that any species in Cc must undergo to become a species in Cd (and vice-versa). Since the assumption is that the rate of mutation is steady, it's assumed that the species in Cc and species in Cd all have an equal amount of mutation from their shared ancestor:</p>
<ul>
<li>mut(Ce) = dist(Cc, Cd) / 2 = 2 / 2 = 1</li>
<li>dist(Ce, Cc) = mut(Ce) - mut(Cc) = 1 - 0 = 1</li>
<li>dist(Ce, Cd) = mut(Ce) - mut(Cd) = 1 - 0 = 1</li>
</ul>
<p><img src="svgbob_43abd429f61152576f3c8e328872adaa265aa9d1.svg" alt="Kroki diagram output"></p>
<p>The <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> then gets modified by merging together the recently <a href="#BM_(connected)%2Fi">connected</a> <a href="#BM_(cluster)%2Fi">cluster</a>s. The new <a href="#BM_(cluster)%2Fi">cluster</a> combines the leaf nodes from both <a href="#BM_(cluster)%2Fi">cluster</a>s: Ce={c,d}, where new <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> distances for that <a href="#BM_(cluster)%2Fi">cluster</a> are computed using the formula...</p>
<div style="overflow: auto"><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><msub><mi>D</mi><mrow><msub><mi>C</mi><mn>1</mn></msub><mo separator="true">,</mo><msub><mi>C</mi><mn>2</mn></msub></mrow></msub><mo>=</mo><mfrac><mrow><munder><mo>∑</mo><mrow><mi>i</mi><mo>∈</mo><msub><mi>C</mi><mn>1</mn></msub></mrow></munder><munder><mo>∑</mo><mrow><mi>j</mi><mo>∈</mo><msub><mi>C</mi><mn>2</mn></msub></mrow></munder><msub><mi>D</mi><mrow><mi>i</mi><mo separator="true">,</mo><mi>j</mi></mrow></msub></mrow><mrow><mi mathvariant="normal">∣</mi><msub><mi>C</mi><mn>1</mn></msub><mi mathvariant="normal">∣</mi><mo>⋅</mo><mi mathvariant="normal">∣</mi><msub><mi>C</mi><mn>2</mn></msub><mi mathvariant="normal">∣</mi></mrow></mfrac></mrow><annotation encoding="application/x-tex">D_{C_1,C_2} = \frac{
  \sum_{i \in C_1} \sum_{j \in C_2} D_{i,j}
  }{
    |C_1| \cdot |C_2|
  }
</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.969438em;vertical-align:-0.286108em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.328331em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.07153em;">C</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31731428571428577em;"><span style="top:-2.357em;margin-left:-0.07153em;margin-right:0.07142857142857144em;"><span class="pstrut" style="height:2.5em;"></span><span class="sizing reset-size3 size1 mtight"><span class="mord mtight">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.143em;"><span></span></span></span></span></span></span><span class="mpunct mtight">,</span><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.07153em;">C</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31731428571428577em;"><span style="top:-2.357em;margin-left:-0.07153em;margin-right:0.07142857142857144em;"><span class="pstrut" style="height:2.5em;"></span><span class="sizing reset-size3 size1 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.143em;"><span></span></span></span></span></span></span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:2.511818em;vertical-align:-0.936em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.575818em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">∣</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.07153em;">C</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.07153em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mord">∣</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">∣</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.07153em;">C</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.07153em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mord">∣</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.825818em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mop"><span class="mop op-symbol small-op" style="position:relative;top:-0.0000050000000000050004em;">∑</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.17862099999999992em;"><span style="top:-2.40029em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">i</span><span class="mrel mtight">∈</span><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.07153em;">C</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31731428571428577em;"><span style="top:-2.357em;margin-left:-0.07153em;margin-right:0.07142857142857144em;"><span class="pstrut" style="height:2.5em;"></span><span class="sizing reset-size3 size1 mtight"><span class="mord mtight">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.143em;"><span></span></span></span></span></span></span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.39981em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mop"><span class="mop op-symbol small-op" style="position:relative;top:-0.0000050000000000050004em;">∑</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.17862099999999992em;"><span style="top:-2.40029em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.05724em;">j</span><span class="mrel mtight">∈</span><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.07153em;">C</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31731428571428577em;"><span style="top:-2.357em;margin-left:-0.07153em;margin-right:0.07142857142857144em;"><span class="pstrut" style="height:2.5em;"></span><span class="sizing reset-size3 size1 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.143em;"><span></span></span></span></span></span></span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.43581800000000004em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">D</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.311664em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">i</span><span class="mpunct mtight">,</span><span class="mord mathnormal mtight" style="margin-right:0.05724em;">j</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.286108em;"><span></span></span></span></span></span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.936em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span></span></div><p><a href="ch7_code/src/phylogeny/UPGMA.py">ch7_code/src/phylogeny/UPGMA.py</a> (lines 64 to 70):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">cluster_dist</span>(<span class="hljs-params">dm_orig: DistanceMatrix[N], c_set: ClusterSet, c1: <span class="hljs-built_in">str</span>, c2: <span class="hljs-built_in">str</span></span>) -&gt; <span class="hljs-built_in">float</span>:
    c1_set = c_set[c1]  <span class="hljs-comment"># this should be a set of leaf nodes from the ORIGINAL unmodified distance matrix</span>
    c2_set = c_set[c2]  <span class="hljs-comment"># this should be a set of leaf nodes from the ORIGINAL unmodified distance matrix</span>
    numerator = <span class="hljs-built_in">sum</span>(dm_orig[i, j] <span class="hljs-keyword">for</span> i, j <span class="hljs-keyword">in</span> product(c1_set, c2_set))  <span class="hljs-comment"># sum it all up</span>
    denominator = <span class="hljs-built_in">len</span>(c1_set) * <span class="hljs-built_in">len</span>(c2_set)  <span class="hljs-comment"># number of additions that occurred</span>
    <span class="hljs-keyword">return</span> numerator / denominator
</code></pre>
<table>
<thead>
<tr>
<th></th>
<th>Ca={a}</th>
<th>Cb={b}</th>
<th>Ce={c,d}</th>
</tr>
</thead>
<tbody>
<tr>
<td>Ca={a}</td>
<td>0</td>
<td>3</td>
<td>3.5</td>
</tr>
<tr>
<td>Cb={b}</td>
<td>3</td>
<td>0</td>
<td>7.5</td>
</tr>
<tr>
<td>Ce={c,d}</td>
<td>3.5</td>
<td>7.5</td>
<td>0</td>
</tr>
</tbody>
</table>
<p>This process repeats at each iteration until a single <a href="#BM_(cluster)%2Fi">cluster</a> remains. At the next iteration, Ca and Cb have the minimum distance in the previous <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> (distance of 3), meaning that Ca and Cb should be <a href="#BM_(connected)%2Fi">connected</a> with a new internal internal node:</p>
<ul>
<li>mut(Cf) = dist(Ca, Cb) / 2 = 3 / 2 = 1.5</li>
<li>dist(Cf, Ca) = mut(Cf) - mut(Ca) = 1.5 - 0 = 1.5</li>
<li>dist(Cf, Cb) = mut(Cf) - mut(Cb) = 1.5 - 0 = 1.5</li>
</ul>
<p><img src="svgbob_962e2ba14c7994a6bd258ad28c99dfca555738f5.svg" alt="Kroki diagram output"></p>
<table>
<thead>
<tr>
<th></th>
<th>Cf={a,b}</th>
<th>Ce={c,d}</th>
</tr>
</thead>
<tbody>
<tr>
<td>Cf={a,b}</td>
<td>0</td>
<td>4</td>
</tr>
<tr>
<td>Ce={c,d}</td>
<td>4</td>
<td>0</td>
</tr>
</tbody>
</table>
<p>At the next iteration, Ce and Cf have the minimum distance in the previous <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> (distance of 4), meaning that Ce and Cf should be <a href="#BM_(connected)%2Fi">connected</a> together with a new internal node:</p>
<ul>
<li>mut(Cg) = dist(Cf, Ce) / 2 = 4 / 2 = 2</li>
<li>dist(Cg, Cf) = mut(Cg) - mut(Cf) = 2 - 1.5 = 0.5</li>
<li>dist(Cg, Ce) = mut(Cg) - mut(Ce) = 2 - 1 = 1</li>
</ul>
<p><img src="svgbob_98ab2c91ad7bd2b3913d79cfa991112d4dcecaa7.svg" alt="Kroki diagram output"></p>
<table>
<thead>
<tr>
<th></th>
<th>Cg={a,b,c,d}</th>
</tr>
</thead>
<tbody>
<tr>
<td>Cg={a,b,c,d}</td>
<td>0</td>
</tr>
</tbody>
</table>
<p>The process is complete. Only a single <a href="#BM_(cluster)%2Fi">cluster</a> remains (representing the root) / the <a href="#BM_(ultrametric%20tree%7Cultrametric)%2Fi">ultrametric tree</a> is fully generated.</p>
<p><img src="svgbob_d60f7033a6d15c3337efd4d2055adb9ef9dfb1f8.svg" alt="Kroki diagram output"></p>
<p>Note that the generated <a href="#BM_(ultrametric%20tree%7Cultrametric)%2Fi">ultrametric tree</a> above is an estimation. The <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> for the example above isn't an <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>, meaning a unique <a href="#BM_(simple%20tree)%2Fi">simple tree</a> doesn't exist for it. Even if it were an <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>, an <a href="#BM_(ultrametric%20tree%7Cultrametric)%2Fi">ultrametric tree</a> is a <a href="#BM_(rooted%20tree)%2Fi">rooted tree</a>, meaning it'll never qualify as the <a href="#BM_(simple%20tree)%2Fi">simple tree</a> unique to that <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> (root node has <a href="#BM_(degree)_GRAPH%2Fi">degree</a> of 2 which isn't allowed in a <a href="#BM_(simple%20tree)%2Fi">simple tree</a>).</p>
<p>In addition, some distances in the generated <a href="#BM_(ultrametric%20tree%7Cultrametric)%2Fi">ultrametric tree</a> are wildly off from the original <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> distances. For example, ...</p>
<ul>
<li>dist(a,d)=8 in the generated <a href="#BM_(ultrametric%20tree%7Cultrametric)%2Fi">ultrametric tree</a>.</li>
<li>dist(a,d)=5 in the original <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a>.</li>
</ul>
<p>Part of this may have to do with the assumption that the closest two nodes in the <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> are neighbors in the <a href="#BM_(ultrametric%20tree%7Cultrametric)%2Fi">ultrametric tree</a>.</p>
<p><a href="ch7_code/src/phylogeny/UPGMA.py">ch7_code/src/phylogeny/UPGMA.py</a> (lines 74 to 143):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">find_clusters_with_min_dist</span>(<span class="hljs-params">dm: DistanceMatrix[N], c_set: ClusterSet</span>) -&gt; <span class="hljs-built_in">tuple</span>[N, N, <span class="hljs-built_in">float</span>]:
    <span class="hljs-keyword">assert</span> c_set.active_count() &gt; <span class="hljs-number">1</span>
    min_n1_id = <span class="hljs-literal">None</span>
    min_n2_id = <span class="hljs-literal">None</span>
    min_dist = <span class="hljs-literal">None</span>
    <span class="hljs-keyword">for</span> n1, n2 <span class="hljs-keyword">in</span> product(c_set.active(), repeat=<span class="hljs-number">2</span>):
        <span class="hljs-keyword">if</span> n1 == n2:
            <span class="hljs-keyword">continue</span>
        d = dm[n1, n2]
        <span class="hljs-keyword">if</span> min_dist <span class="hljs-keyword">is</span> <span class="hljs-literal">None</span> <span class="hljs-keyword">or</span> d &lt; min_dist:
            min_n1_id = n1
            min_n2_id = n2
            min_dist = d
    <span class="hljs-keyword">assert</span> min_n1_id <span class="hljs-keyword">is</span> <span class="hljs-keyword">not</span> <span class="hljs-literal">None</span> <span class="hljs-keyword">and</span> min_n2_id <span class="hljs-keyword">is</span> <span class="hljs-keyword">not</span> <span class="hljs-literal">None</span> <span class="hljs-keyword">and</span> min_dist <span class="hljs-keyword">is</span> <span class="hljs-keyword">not</span> <span class="hljs-literal">None</span>
    <span class="hljs-keyword">return</span> min_n1_id, min_n2_id, min_dist


<span class="hljs-keyword">def</span> <span class="hljs-title function_">cluster_merge</span>(<span class="hljs-params">
        dm: DistanceMatrix[N],
        dm_orig: DistanceMatrix[N],
        c_set: ClusterSet,
        old_id1: N,
        old_id2: N,
        new_id: N
</span>) -&gt; <span class="hljs-literal">None</span>:
    c_set.merge(new_id, old_id1, old_id2)  <span class="hljs-comment"># create new cluster w/ elements from old - old_ids deactived+new_id actived</span>
    new_dists = {}
    <span class="hljs-keyword">for</span> existing_id <span class="hljs-keyword">in</span> dm.leaf_ids():
        <span class="hljs-keyword">if</span> existing_id == old_id1 <span class="hljs-keyword">or</span> existing_id == old_id2:
            <span class="hljs-keyword">continue</span>
        new_dist = cluster_dist(dm_orig, c_set, new_id, existing_id)
        new_dists[existing_id] = new_dist
    dm.merge(new_id, old_id1, old_id2, new_dists)  <span class="hljs-comment"># remove old ids and replace with new_id that has new distances</span>


<span class="hljs-keyword">def</span> <span class="hljs-title function_">upgma</span>(<span class="hljs-params">dm: DistanceMatrix[N]</span>) -&gt; <span class="hljs-built_in">tuple</span>[Graph, N]:
    g = Graph()
    c_set = ClusterSet(dm)  <span class="hljs-comment"># primed with leaf nodes (all active)</span>
    <span class="hljs-keyword">for</span> node <span class="hljs-keyword">in</span> dm.leaf_ids_it():
        g.insert_node(node, <span class="hljs-number">0</span>)  <span class="hljs-comment"># initial node weights (each leaf node has an age of 0)</span>
    dm_orig = dm.copy()
    <span class="hljs-comment"># set node ages</span>
    next_node_id = <span class="hljs-number">0</span>
    next_edge_id = <span class="hljs-number">0</span>
    <span class="hljs-keyword">while</span> c_set.active_count() &gt; <span class="hljs-number">1</span>:
        min_n1_id, min_n2_id, min_dist = find_clusters_with_min_dist(dm, c_set)
        new_node_id = next_node_id
        new_node_age = min_dist / <span class="hljs-number">2</span>
        g.insert_node(<span class="hljs-string">f'C<span class="hljs-subst">{new_node_id}</span>'</span>, new_node_age)
        next_node_id += <span class="hljs-number">1</span>
        g.insert_edge(<span class="hljs-string">f'E<span class="hljs-subst">{next_edge_id}</span>'</span>, min_n1_id, <span class="hljs-string">f'C<span class="hljs-subst">{new_node_id}</span>'</span>)
        next_edge_id += <span class="hljs-number">1</span>
        g.insert_edge(<span class="hljs-string">f'E<span class="hljs-subst">{next_edge_id}</span>'</span>, min_n2_id, <span class="hljs-string">f'C<span class="hljs-subst">{new_node_id}</span>'</span>)
        next_edge_id += <span class="hljs-number">1</span>
        cluster_merge(dm, dm_orig, c_set, min_n1_id, min_n2_id, <span class="hljs-string">f'C<span class="hljs-subst">{new_node_id}</span>'</span>)
    <span class="hljs-comment"># set amount of age added by each edge</span>
    nodes_by_age = <span class="hljs-built_in">sorted</span>([(n, g.get_node_data(n)) <span class="hljs-keyword">for</span> n <span class="hljs-keyword">in</span> g.get_nodes()], key=<span class="hljs-keyword">lambda</span> x: x[<span class="hljs-number">1</span>])
    set_edges = <span class="hljs-built_in">set</span>()  <span class="hljs-comment"># edges that have already had their weights set</span>
    <span class="hljs-keyword">for</span> child_n, child_age <span class="hljs-keyword">in</span> nodes_by_age:
        <span class="hljs-keyword">for</span> e <span class="hljs-keyword">in</span> g.get_outputs(child_n):
            <span class="hljs-keyword">if</span> e <span class="hljs-keyword">in</span> set_edges:
                <span class="hljs-keyword">continue</span>
            parent_n = [n <span class="hljs-keyword">for</span> n <span class="hljs-keyword">in</span> g.get_edge_ends(e) <span class="hljs-keyword">if</span> n != child_n].pop()
            parent_age = g.get_node_data(parent_n)
            weight = parent_age - child_age
            g.update_edge_data(e, weight)
            set_edges.add(e)
    root_id = c_set.active().pop()
    <span class="hljs-keyword">return</span> g, root_id
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the distance matrix ...</p>
<table>
<thead><tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr></thead>
<tbody>
<tr>
<td>v0</td>
<td>0.0</td>
<td>13.0</td>
<td>21.0</td>
<td>21.0</td>
<td>22.0</td>
<td>22.0</td>
</tr>
<tr>
<td>v1</td>
<td>13.0</td>
<td>0.0</td>
<td>12.0</td>
<td>12.0</td>
<td>13.0</td>
<td>13.0</td>
</tr>
<tr>
<td>v2</td>
<td>21.0</td>
<td>12.0</td>
<td>0.0</td>
<td>20.0</td>
<td>21.0</td>
<td>21.0</td>
</tr>
<tr>
<td>v3</td>
<td>21.0</td>
<td>12.0</td>
<td>20.0</td>
<td>0.0</td>
<td>7.0</td>
<td>13.0</td>
</tr>
<tr>
<td>v4</td>
<td>22.0</td>
<td>13.0</td>
<td>21.0</td>
<td>7.0</td>
<td>0.0</td>
<td>14.0</td>
</tr>
<tr>
<td>v5</td>
<td>22.0</td>
<td>13.0</td>
<td>21.0</td>
<td>13.0</td>
<td>14.0</td>
<td>0.0</td>
</tr>
</tbody>
</table>
<p>... the UPGMA generated tree is ...</p>
<p><img src="dot_246c4b1c76ce4a0e7566cdecd1db58ad.svg" alt="Dot diagram"></p>
</div>
<p></p>
<a name="H_Additive%20Phylogeny%20Algorithm"></a><h4>Additive Phylogeny Algorithm</h4>
<p><a name="BM_(Algorithms%5C%2FPhylogeny%2FDistance%20Matrix%20to%20Tree%2FAdditive%20Phylogeny%20Algorithm)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FTrim)_TOPIC%2F">Algorithms/Phylogeny/Trim</a></li>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FUn-trim%20Tree)_TOPIC%2F">Algorithms/Phylogeny/Un-trim Tree</a></li>
</ul>
</div>
<p><strong>ALGORITHM</strong>:</p>
<p><a href="#BM_(additive%20phylogeny)%2Fi">Additive phylogeny</a> is a recursive algorithm that finds the unique <a href="#BM_(simple%20tree)%2Fi">simple tree</a> for some <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>. At each recursive step, the algorithm trims off a single leaf node from the <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a>, stopping once the <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> consists of only two leaf nodes. The <a href="#BM_(simple%20tree)%2Fi">simple tree</a> for any 2x2 <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> is obvious as ...</p>
<ul>
<li>it only consists of 2 nodes</li>
<li>it only consists of a single edge (nodes of <a href="#BM_(degree)_GRAPH%2Fi">degree</a> 2 not allowed in <a href="#BM_(simple%20tree)%2Fi">simple tree</a>s, meaning train of non-branching edges not allowed)</li>
</ul>
<p>For example, the following 2x2 <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> has the following <a href="#BM_(simple%20tree)%2Fi">simple tree</a>...</p>
<table>
<thead>
<tr>
<th></th>
<th>v0</th>
<th>v1</th>
</tr>
</thead>
<tbody>
<tr>
<td>v0</td>
<td>0</td>
<td>14</td>
</tr>
<tr>
<td>v1</td>
<td>14</td>
<td>0</td>
</tr>
</tbody>
</table>
<p><img src="svgbob_64427b0b4402aafe7a3102b5777d5a031cf020b7.svg" alt="Kroki diagram output"></p>
<p><a href="ch7_code/src/phylogeny/AdditivePhylogeny.py">ch7_code/src/phylogeny/AdditivePhylogeny.py</a> (lines 34 to 49):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">to_obvious_graph</span>(<span class="hljs-params">
        dm: DistanceMatrix[N],
        gen_edge_id: <span class="hljs-type">Callable</span>[[], E]
</span>) -&gt; Graph[N, ND, E, <span class="hljs-built_in">float</span>]:
    <span class="hljs-keyword">if</span> dm.n != <span class="hljs-number">2</span>:
        <span class="hljs-keyword">raise</span> ValueError(<span class="hljs-string">'Distance matrix must only contain 2 leaf nodes'</span>)
    l1, l2 = dm.leaf_ids()
    g = Graph()
    g.insert_node(l1)
    g.insert_node(l2)
    g.insert_edge(
        gen_edge_id(),
        l1,
        l2,
        dm[l1, l2]
    )
    <span class="hljs-keyword">return</span> g
</code></pre>
<p>As the algorithm returns from each recursive step, it has 2 pieces of information:</p>
<ul>
<li>an <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> containing trimmed leaf node L.</li>
<li>a <a href="#BM_(simple%20tree)%2Fi">simple tree</a> not containing trimmed leaf node L.</li>
</ul>
<p>That's enough information to know where on the returned <a href="#BM_(tree)%2Fi">tree</a> L's <a href="#BM_(limb)%2Fi">limb</a> should be added and what L's <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a> should be (un-trimming the <a href="#BM_(tree)%2Fi">tree</a>). At the end, the algorithm will have constructed the entire <a href="#BM_(simple%20tree)%2Fi">simple tree</a> for the <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>.</p>
<p><a href="ch7_code/src/phylogeny/AdditivePhylogeny.py">ch7_code/src/phylogeny/AdditivePhylogeny.py</a> (lines 55 to 68):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">additive_phylogeny</span>(<span class="hljs-params">
        dm: DistanceMatrix[N],
        gen_node_id: <span class="hljs-type">Callable</span>[[], N],
        gen_edge_id: <span class="hljs-type">Callable</span>[[], E]
</span>) -&gt; Graph:
    <span class="hljs-keyword">if</span> dm.n == <span class="hljs-number">2</span>:
        <span class="hljs-keyword">return</span> to_obvious_graph(dm, gen_edge_id)
    n = <span class="hljs-built_in">next</span>(dm.leaf_ids_it())
    dm_untrimmed = dm.copy()
    trim_distance_matrix(dm, n)
    g = additive_phylogeny(dm, gen_node_id, gen_edge_id)
    untrim_tree(dm_untrimmed, g, gen_node_id, gen_edge_id)
    <span class="hljs-keyword">return</span> g
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the distance matrix ...</p>
<table><thead><tr><th></th><th>v0</th><th>v1</th><th>v2</th><th>v3</th><th>v4</th><th>v5</th></tr></thead><tbody><tr><td>v0</td><td>0.0</td><td>13.0</td><td>21.0</td><td>21.0</td><td>22.0</td><td>22.0</td></tr><tr><td>v1</td><td>13.0</td><td>0.0</td><td>12.0</td><td>12.0</td><td>13.0</td><td>13.0</td></tr><tr><td>v2</td><td>21.0</td><td>12.0</td><td>0.0</td><td>20.0</td><td>21.0</td><td>21.0</td></tr><tr><td>v3</td><td>21.0</td><td>12.0</td><td>20.0</td><td>0.0</td><td>7.0</td><td>13.0</td></tr><tr><td>v4</td><td>22.0</td><td>13.0</td><td>21.0</td><td>7.0</td><td>0.0</td><td>14.0</td></tr><tr><td>v5</td><td>22.0</td><td>13.0</td><td>21.0</td><td>13.0</td><td>14.0</td><td>0.0</td></tr></tbody></table>
<p>Trimmed v0 to produce distance matrix ...</p>
<table><thead><tr><th></th><th>v1</th><th>v2</th><th>v3</th><th>v4</th><th>v5</th></tr></thead><tbody><tr><td>v1</td><td>0.0</td><td>12.0</td><td>12.0</td><td>13.0</td><td>13.0</td></tr><tr><td>v2</td><td>12.0</td><td>0.0</td><td>20.0</td><td>21.0</td><td>21.0</td></tr><tr><td>v3</td><td>12.0</td><td>20.0</td><td>0.0</td><td>7.0</td><td>13.0</td></tr><tr><td>v4</td><td>13.0</td><td>21.0</td><td>7.0</td><td>0.0</td><td>14.0</td></tr><tr><td>v5</td><td>13.0</td><td>21.0</td><td>13.0</td><td>14.0</td><td>0.0</td></tr></tbody></table>
<p>Trimmed v1 to produce distance matrix ...</p>
<table><thead><tr><th></th><th>v2</th><th>v3</th><th>v4</th><th>v5</th></tr></thead><tbody><tr><td>v2</td><td>0.0</td><td>20.0</td><td>21.0</td><td>21.0</td></tr><tr><td>v3</td><td>20.0</td><td>0.0</td><td>7.0</td><td>13.0</td></tr><tr><td>v4</td><td>21.0</td><td>7.0</td><td>0.0</td><td>14.0</td></tr><tr><td>v5</td><td>21.0</td><td>13.0</td><td>14.0</td><td>0.0</td></tr></tbody></table>
<p>Trimmed v3 to produce distance matrix ...</p>
<table><thead><tr><th></th><th>v2</th><th>v4</th><th>v5</th></tr></thead><tbody><tr><td>v2</td><td>0.0</td><td>21.0</td><td>21.0</td></tr><tr><td>v4</td><td>21.0</td><td>0.0</td><td>14.0</td></tr><tr><td>v5</td><td>21.0</td><td>14.0</td><td>0.0</td></tr></tbody></table>
<p>Trimmed v2 to produce distance matrix ...</p>
<table><thead><tr><th></th><th>v4</th><th>v5</th></tr></thead><tbody><tr><td>v4</td><td>0.0</td><td>14.0</td></tr><tr><td>v5</td><td>14.0</td><td>0.0</td></tr></tbody></table>
<p>Obvious simple tree...</p>
<p><img src="dot_a1744096a16de60264a14880f84d70be.svg" alt="Dot diagram"></p>
<p>Attached v2 to produce tree...</p>
<p><img src="dot_c055a665c75a3139921a8a461681bb6c.svg" alt="Dot diagram"></p>
<p>Attached v3 to produce tree...</p>
<p><img src="dot_d245ba2437e23017778f31eb1645d0b3.svg" alt="Dot diagram"></p>
<p>Attached v1 to produce tree...</p>
<p><img src="dot_fc28272d8de4992c3f02e4e24c2d38ac.svg" alt="Dot diagram"></p>
<p>Attached v0 to produce tree...</p>
<p><img src="dot_640af463f84c93a65fd967cd10901ebf.svg" alt="Dot diagram"></p>
</div>
<p></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The book is inconsistent about whether <a href="#BM_(simple%20tree)%2Fi">simple tree</a>s can have internal edges of weight 0. Early in the book it says that it can and later on it says that it goes back on that and says internal edges of weight 0 aren't actually allowed. I'd already implied as much given that they'd be the same organism at both ends, and this algorithm explicitly won't allow it in that if it walks up to a node, it'll branch off that node (an additional edge weight of 0 won't extend past that node).</p>
</div>
<a name="H_Neighbour%20Joining%20Phylogeny%20Algorithm"></a><h4>Neighbour Joining Phylogeny Algorithm</h4>
<p><a name="BM_(Algorithms%5C%2FPhylogeny%2FDistance%20Matrix%20to%20Tree%2FNeighbour%20Joining%20Phylogeny%20Algorithm)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FFind%20Neighbours)_TOPIC%2F">Algorithms/Phylogeny/Find Neighbours</a></li>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FFind%20Neighbour%20Limb%20Lengths)_TOPIC%2F">Algorithms/Phylogeny/Find Neighbour Limb Lengths</a></li>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FExpose%20Neighbour%20Parent)_TOPIC%2F">Algorithms/Phylogeny/Expose Neighbour Parent</a></li>
</ul>
</div>
<p><strong>ALGORITHM</strong>:</p>
<p><a href="#BM_(neighbour%20joining%20phylogeny)%2Fi">Neighbour joining phylogeny</a> is a recursive algorithm that either...</p>
<ul>
<li>finds the unique <a href="#BM_(simple%20tree)%2Fi">simple tree</a> for some <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>.</li>
<li>approximates a <a href="#BM_(simple%20tree)%2Fi">simple tree</a> for some non-<a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> (close to being additive).</li>
</ul>
<p>At each recursive step, the algorithm finds a pair of <a href="#BM_(neighbour)%2Fi">neighbour</a>ing leaf nodes in the <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> and exposes their shared parent (<a href="#BM_(neighbour)%2Fi">neighbour</a>s replaced with parent in the <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a>), stopping once the <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> consists of only two leaf nodes. The <a href="#BM_(simple%20tree)%2Fi">simple tree</a> for any 2x2 <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> is obvious as ...</p>
<ul>
<li>it only consists of 2 nodes</li>
<li>it only consists of a single edge (nodes of <a href="#BM_(degree)_GRAPH%2Fi">degree</a> 2 not allowed in <a href="#BM_(simple%20tree)%2Fi">simple tree</a>s, meaning train of non-branching edges not allowed)</li>
</ul>
<p>For example, the following 2x2 <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> has the following <a href="#BM_(simple%20tree)%2Fi">simple tree</a>...</p>
<table>
<thead>
<tr>
<th></th>
<th>v0</th>
<th>v1</th>
</tr>
</thead>
<tbody>
<tr>
<td>v0</td>
<td>0</td>
<td>14</td>
</tr>
<tr>
<td>v1</td>
<td>14</td>
<td>0</td>
</tr>
</tbody>
</table>
<p><img src="svgbob_64427b0b4402aafe7a3102b5777d5a031cf020b7.svg" alt="Kroki diagram output"></p>
<p><a href="ch7_code/src/phylogeny/NeighbourJoiningPhylogeny.py">ch7_code/src/phylogeny/NeighbourJoiningPhylogeny.py</a> (lines 48 to 63):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">to_obvious_graph</span>(<span class="hljs-params">
        dm: DistanceMatrix[N],
        gen_edge_id: <span class="hljs-type">Callable</span>[[], E]
</span>) -&gt; Graph:
    <span class="hljs-keyword">if</span> dm.n != <span class="hljs-number">2</span>:
        <span class="hljs-keyword">raise</span> ValueError(<span class="hljs-string">'Distance matrix must only contain 2 leaf nodes'</span>)
    l1, l2 = dm.leaf_ids()
    g = Graph()
    g.insert_node(l1)
    g.insert_node(l2)
    g.insert_edge(
        gen_edge_id(),
        l1,
        l2,
        dm[l1, l2]
    )
    <span class="hljs-keyword">return</span> g
</code></pre>
<p>As the algorithm returns from each recursive step, it has 3 pieces of information:</p>
<ul>
<li>there exists two <a href="#BM_(neighbour)%2Fi">neighbour</a>ing leaf nodes L and N.</li>
<li>a <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> containing L and N.</li>
<li>a <a href="#BM_(simple%20tree)%2Fi">simple tree</a> missing L and N, but containing their shared parent P.</li>
</ul>
<p>That's enough information to know where L and N should be added on to the <a href="#BM_(tree)%2Fi">tree</a> (node P) and what their <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a>s are. At the end, the algorithm will have constructed the entire <a href="#BM_(simple%20tree)%2Fi">simple tree</a> for the <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>.</p>
<p><a href="ch7_code/src/phylogeny/NeighbourJoiningPhylogeny.py">ch7_code/src/phylogeny/NeighbourJoiningPhylogeny.py</a> (lines 69 to 86):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">neighbour_joining_phylogeny</span>(<span class="hljs-params">
        dm: DistanceMatrix,
        gen_node_id: <span class="hljs-type">Callable</span>[[], N],
        gen_edge_id: <span class="hljs-type">Callable</span>[[], E]
</span>) -&gt; Graph:
    <span class="hljs-keyword">if</span> dm.n == <span class="hljs-number">2</span>:
        <span class="hljs-keyword">return</span> to_obvious_graph(dm, gen_edge_id)
    l1, l2 = find_neighbours(dm)
    l1_len, l2_len = find_neighbouring_limb_lengths(dm, l1, l2)
    dm_trimmed = dm.copy()
    p = expose_neighbour_parent(dm_trimmed, l1, l2, gen_node_id)  <span class="hljs-comment"># p added to dm_trimmed while l1, l2 removed</span>
    g = neighbour_joining_phylogeny(dm_trimmed, gen_node_id, gen_edge_id)
    g.insert_node(l1)
    g.insert_node(l2)
    g.insert_edge(gen_edge_id(), p, l1, l1_len)
    g.insert_edge(gen_edge_id(), p, l2, l2_len)
    <span class="hljs-keyword">return</span> g
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given NON- additive distance matrix...</p>
<table>
<thead><tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr></thead>
<tbody>
<tr>
<td>v0</td>
<td>0.0</td>
<td>14.0</td>
<td>22.0</td>
<td>20.0</td>
<td>23.0</td>
<td>22.0</td>
</tr>
<tr>
<td>v1</td>
<td>14.0</td>
<td>0.0</td>
<td>12.0</td>
<td>10.0</td>
<td>12.0</td>
<td>14.0</td>
</tr>
<tr>
<td>v2</td>
<td>22.0</td>
<td>12.0</td>
<td>0.0</td>
<td>20.0</td>
<td>22.0</td>
<td>20.0</td>
</tr>
<tr>
<td>v3</td>
<td>20.0</td>
<td>10.0</td>
<td>20.0</td>
<td>0.0</td>
<td>8.0</td>
<td>12.0</td>
</tr>
<tr>
<td>v4</td>
<td>23.0</td>
<td>12.0</td>
<td>22.0</td>
<td>8.0</td>
<td>0.0</td>
<td>15.0</td>
</tr>
<tr>
<td>v5</td>
<td>22.0</td>
<td>14.0</td>
<td>20.0</td>
<td>12.0</td>
<td>15.0</td>
<td>0.0</td>
</tr>
</tbody>
</table>
<p>Replaced neighbours ('v3', 'v4') with their parent N1 to produce distance matrix ...</p>
<table><thead><tr><th></th><th>N1</th><th>v0</th><th>v1</th><th>v2</th><th>v5</th></tr></thead><tbody><tr><td>N1</td><td>0.0</td><td>17.5</td><td>7.0</td><td>17.0</td><td>9.5</td></tr><tr><td>v0</td><td>17.5</td><td>0.0</td><td>14.0</td><td>22.0</td><td>22.0</td></tr><tr><td>v1</td><td>7.0</td><td>14.0</td><td>0.0</td><td>12.0</td><td>14.0</td></tr><tr><td>v2</td><td>17.0</td><td>22.0</td><td>12.0</td><td>0.0</td><td>20.0</td></tr><tr><td>v5</td><td>9.5</td><td>22.0</td><td>14.0</td><td>20.0</td><td>0.0</td></tr></tbody></table>
<p>Replaced neighbours ('N1', 'v5') with their parent N2 to produce distance matrix ...</p>
<table><thead><tr><th></th><th>N2</th><th>v0</th><th>v1</th><th>v2</th></tr></thead><tbody><tr><td>N2</td><td>0.0</td><td>15.0</td><td>5.75</td><td>13.75</td></tr><tr><td>v0</td><td>15.0</td><td>0.0</td><td>14.0</td><td>22.0</td></tr><tr><td>v1</td><td>5.75</td><td>14.0</td><td>0.0</td><td>12.0</td></tr><tr><td>v2</td><td>13.75</td><td>22.0</td><td>12.0</td><td>0.0</td></tr></tbody></table>
<p>Replaced neighbours ('v1', 'v2') with their parent N3 to produce distance matrix ...</p>
<table><thead><tr><th></th><th>N2</th><th>N3</th><th>v0</th></tr></thead><tbody><tr><td>N2</td><td>0.0</td><td>3.75</td><td>15.0</td></tr><tr><td>N3</td><td>3.75</td><td>0.0</td><td>12.0</td></tr><tr><td>v0</td><td>15.0</td><td>12.0</td><td>0.0</td></tr></tbody></table>
<p>Replaced neighbours ('v0', 'N2') with their parent N4 to produce distance matrix ...</p>
<table><thead><tr><th></th><th>N3</th><th>N4</th></tr></thead><tbody><tr><td>N3</td><td>0.0</td><td>0.375</td></tr><tr><td>N4</td><td>0.375</td><td>0.0</td></tr></tbody></table>
<p>Obvious simple tree...</p>
<p><img src="dot_be14fd16e403f27a505730a47fb24818.svg" alt="Dot diagram"></p>
<p>Attached ('v0', 'N2') to N4 to produce tree...</p>
<p><img src="dot_cd1212ec72d7435abefd6c91f344c9dd.svg" alt="Dot diagram"></p>
<p>Attached ('v1', 'v2') to N3 to produce tree...</p>
<p><img src="dot_b3390ad411b2114cc14ad8ab972c789e.svg" alt="Dot diagram"></p>
<p>Attached ('N1', 'v5') to N2 to produce tree...</p>
<p><img src="dot_ccb89c35883c4cecfdbd8e7aeceeeb80.svg" alt="Dot diagram"></p>
<p>Attached ('v3', 'v4') to N1 to produce tree...</p>
<p><img src="dot_d82377b2f7aa851fa37e18b80217a112.svg" alt="Dot diagram"></p>
</div>
<p></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The book is inconsistent about whether <a href="#BM_(simple%20tree)%2Fi">simple tree</a>s can have internal edges of weight 0. Early in the book it says that it can and later on it says that it goes back on that and says internal edges of weight 0 aren't actually allowed. I'd already implied as much given that they'd be the same organism at both ends, but I'm unsure if this algorithm will allow it if fed in a non-<a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>. It should never happend with an <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>.</p>
</div>
<a name="H_Evolutionary%20Algorithm"></a><h4>Evolutionary Algorithm</h4>
<p><a name="BM_(Algorithms%5C%2FPhylogeny%2FDistance%20Matrix%20to%20Tree%2FEvolutionary%20Algorithm)_TOPIC%2F"></a><strong></strong></p>
<p><strong>ALGORITHM</strong>:</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>This is essentially a hammer, ignoring much of the logic and techniques derived in prior sections. There is no code for this section because writing it involves doing things like writing a generic linear systems solver, evolutionary algorithms framework, etc... There are Python packages you can use if you really want to do this, but this section is more describing the overarching idea.</p>
<p>The logic and techniques in prior sections typically work much better and much faster than doing something like this, but this doesn't require as much reasoning / thinking. This idea was first hinted at in the Pevzner book when first describing how to assign weights for non-<a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrices</a>.</p>
</div>
<p>Given an <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>, if you already know the structure of the <a href="#BM_(tree)%2Fi">tree</a>, with edge weights that satisfy that <a href="#BM_(tree)%2Fi">tree</a> are derivable from that <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a>. For example, given the following <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> and <a href="#BM_(tree)%2Fi">tree</a> structure...</p>
<table>
<thead>
<tr>
<th></th>
<th>Cat</th>
<th>Lion</th>
<th>Bear</th>
</tr>
</thead>
<tbody>
<tr>
<td>Cat</td>
<td>0</td>
<td>2</td>
<td>4</td>
</tr>
<tr>
<td>Lion</td>
<td>2</td>
<td>0</td>
<td>3</td>
</tr>
<tr>
<td>Bear</td>
<td>4</td>
<td>3</td>
<td>0</td>
</tr>
</tbody>
</table>
<p><img src="svgbob_9efe0131b95aacb3aa51fbb4cd8f18843707a13e.svg" alt="Kroki diagram output"></p>
<p>... the distances between species must have been calculated as follows:</p>
<ul>
<li>dist(Cat, Lion) = dist(Cat, A1) + dist(A1, Lion)`</li>
<li>dist(Cat, Bear) = dist(Cat, A1) + dist(A1, A2) + dist(A2, Bear)`</li>
<li>dist(Lion, Bear) = dist(Lion, A1) + dist(A1, A2) + dist(A2, Bear)`</li>
</ul>
<p>This is a system of linear equations that may be solved using standard algebra. For example, each <code>dist()</code> above is representable as either a variable or a constant...</p>
<ul>
<li>2 = dist(Cat, Lion)  = dist(Lion, Cat)</li>
<li>4 = dist(Cat, Bear)  = dist(Bear, Cat)</li>
<li>3 = dist(Lion, Bear) = dist(Bear, Lion)</li>
<li>w = dist(Cat, A1)    = dist(A1, Cat)</li>
<li>x = dist(Lion, A1)   = dist(A1, Lion)</li>
<li>y = dist(A1, A2)     = dist(A2, A1)</li>
<li>z = dist(A2, Bear)   = dist(Bear, A2)</li>
</ul>
<p>... , which converts each calculation above to the following equations ...</p>
<ul>
<li>2 = w + x</li>
<li>4 = w + y + z</li>
<li>3 = x + y + z</li>
</ul>
<p><img src="svgbob_2f13ccbb0caa3ef9d6c3abe089b7f6bd3f123ce5.svg" alt="Kroki diagram output"></p>
<p>Solving this system of linear equations results in. ..</p>
<ul>
<li>x = 0.5</li>
<li>w = 1.5</li>
<li>z = 2.5 - y`</li>
</ul>
<p>As such, the example <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> is an <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive matrix</a> because there exists a <a href="#BM_(tree)%2Fi">tree</a> that satisfies it. Any of the following edge weights will work with this <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a>...</p>
<ul>
<li>x = 0.5, w = 1.5, y = 0.5, z = 2.0</li>
<li>x = 0.5, w = 1.5, y = 1.0, z = 1.5</li>
<li>x = 0.5, w = 1.5, y = 1.5, z = 1.0</li>
<li>...</li>
</ul>
<p>The example above tests against a <a href="#BM_(tree)%2Fi">tree</a> that's a non-<a href="#BM_(simple%20tree)%2Fi">simple tree</a> (A2 is an internal node with <a href="#BM_(degree)_GRAPH%2Fi">degree</a> of 2). If you limit your search to <a href="#BM_(simple%20tree)%2Fi">simple tree</a>s and find nothing, there won't be any non-<a href="#BM_(simple%20tree)%2Fi">simple tree</a>s either: Non-<a href="#BM_(simple%20tree)%2Fi">simple tree</a>s are essentially <a href="#BM_(simple%20tree)%2Fi">simple tree</a>s that have had edges broken up by splicing nodes in between (<a href="#BM_(degree)_GRAPH%2Fi">degree</a> 2 nodes).</p>
<p>The non-<a href="#BM_(simple%20tree)%2Fi">simple tree</a> example above collapsed into a <a href="#BM_(simple%20tree)%2Fi">simple tree</a>:</p>
<p><img src="svgbob_da0e7825576bb7895069ddd56443653ece61dac9.svg" alt="Kroki diagram output"></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The path A1-A2-Bear has been collapsed into A1-Bear, where the weight of the newly collapsed edge is represented by a (formerly y+z). Using the same <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>, the <a href="#BM_(simple%20tree)%2Fi">simple tree</a> above gets solved to <code>w = 2, x = 1, a = 2</code>.</p>
</div>
<p>If the <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> isn't additive, something like <a href="#BM_(sum%20of%20squared%20errors%7Csum%20of%20square%20error%7Csum%20of%20errors%20squared)%2Fi">sum of errors squared</a> may be used to converge on an approximate set of weights that work. Similarly, evolutionary algorithms may be used in addition to approximating weights to find a <a href="#BM_(simple%20tree)%2Fi">simple tree</a> that's close enough to the</p>
<a name="H_Sequence%20Inference"></a><h3>Sequence Inference</h3>
<p><a name="BM_(Algorithms%5C%2FPhylogeny%2FSequence%20Inference)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FDistance%20Matrix%20to%20Tree)_TOPIC%2F">Algorithms/Phylogeny/Distance Matrix to Tree</a></li>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment)_TOPIC%2F">Algorithms/Sequence Alignment</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: It's possible to infer the <a href="#BM_(sequence)%2Fi">sequence</a>s for shared ancestors in a <a href="#BM_(phylogenetic%20tree%7Cphylogeny%20tree%7Cphylogenies%7Cevolutionary%20tree)%2Fi">phylogenetic tree</a>. Specifically, given a <a href="#BM_(phylogenetic%20tree%7Cphylogeny%20tree%7Cphylogenies%7Cevolutionary%20tree)%2Fi">phylogenetic tree</a>, each node in may have a <a href="#BM_(sequence)%2Fi">sequence</a> assigned to it, where a ...</p>
<ul>
<li>leaf node is assigned the <a href="#BM_(sequence)%2Fi">sequence</a> of the known entity it's for.</li>
<li>internal node is assigned an inferred <a href="#BM_(sequence)%2Fi">sequence</a> based on <a href="#BM_(neighbour)%2Fi">neighbour</a>ing nodes.</li>
</ul>
<p><img src="svgbob_fab0e3e719365ddd7a071bc539f7b3d7aa117f9d.svg" alt="Kroki diagram output"></p>
<p><strong>WHY</strong>: Inferring ancestral <a href="#BM_(sequence)%2Fi">sequence</a>s may help provide additional insight / clues as to the evolution that took place.</p>
<a name="H_Small%20Parsimony%20Algorithm"></a><h4>Small Parsimony Algorithm</h4>
<p><a name="BM_(Algorithms%5C%2FPhylogeny%2FSequence%20Inference%2FSmall%20Parsimony%20Algorithm)_TOPIC%2F"></a><strong></strong></p>
<p><strong>ALGORITHM</strong>:</p>
<p>Given a <a href="#BM_(phylogenetic%20tree%7Cphylogeny%20tree%7Cphylogenies%7Cevolutionary%20tree)%2Fi">phylogenetic tree</a> and the <a href="#BM_(sequence)%2Fi">sequence</a>s for its leaf nodes (known entities), this algorithm infers <a href="#BM_(sequence)%2Fi">sequence</a>s for its internal nodes (ancestor entities) based on how likely it is for <a href="#BM_(sequence)%2Fi">sequence</a> elements to change from one type to another. The <a href="#BM_(sequence)%2Fi">sequence</a> / <a href="#BM_(sequence)%2Fi">sequence</a> element most likely to be there is said to be the most <a href="#BM_(parsimony%7Cparsimonious)%2Fi">parsimonious</a>.</p>
<p>The algorithm only works on <a href="#BM_(sequence)%2Fi">sequence</a>s of matching length.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>If you're interested to see why it's called <a href="#BM_(small%20parsimony%7Clarge%20parsimony)%2Fi">small parsimony</a>, see the next section which describes <a href="#BM_(small%20parsimony%7Clarge%20parsimony)%2Fi">small parsimony</a> vs <a href="#BM_(small%20parsimony%7Clarge%20parsimony)%2Fi">large parsimony</a>.</p>
</div>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The Pevzner book says that if the <a href="#BM_(sequence)%2Fi">sequence</a>s for known entities aren't the same length, common practice is to align them (e.g. multiple alignment) and remove any <a href="#BM_(indel)%2Fi">indel</a>s before continuing. Once <a href="#BM_(indel)%2Fi">indel</a>s are removed, the <a href="#BM_(sequence)%2Fi">sequence</a>s will all become the same length.</p>
<p><img src="svgbob_b265b2e8fe1ea65e1b7d208b4895c873721367c3.svg" alt="Kroki diagram output"></p>
<p>I'm not sure why <a href="#BM_(indel)%2Fi">indel</a>s can't just be included as an option (e.g. A, C, T, G, and -)? There's probably a reason. Maybe because <a href="#BM_(indel)%2Fi">indel</a>s that happen in bursts are likely from <a href="#BM_(genome%20rearrangement%7Cchromosomal%20rearrangement%7Cchromosome%20rearrangement)%2Fi">genome rearrangement</a> mutations instead of <a href="#BM_(point%20mutation)%2Fi">point mutation</a>s and including them muddies the waters? I don't know.</p>
</div>
<p>The algorithm works by building a distance map for each index of each node's <a href="#BM_(sequence)%2Fi">sequence</a>. Each map defines the distance if that specific index were to contain that specific element. The shorter the distance, the more likely it is for that index to contain that specific element. For example, ...</p>
<table>
<thead>
<tr>
<th></th>
<th>A</th>
<th>C</th>
<th>T</th>
<th>G</th>
</tr>
</thead>
<tbody>
<tr>
<td>0</td>
<td>1.0</td>
<td><strong>0.0</strong></td>
<td>4.0</td>
<td>3.0</td>
</tr>
<tr>
<td>1</td>
<td>2.0</td>
<td>2.0</td>
<td><strong>1.0</strong></td>
<td>3.0</td>
</tr>
<tr>
<td>2</td>
<td>1.0</td>
<td>1.0</td>
<td><strong>0.0</strong></td>
<td>1.0</td>
</tr>
<tr>
<td>3</td>
<td>2.0</td>
<td>3.0</td>
<td>1.0</td>
<td><strong>0.0</strong></td>
</tr>
<tr>
<td>4</td>
<td>1.0</td>
<td>1.0</td>
<td><strong>0.0</strong></td>
<td>1.0</td>
</tr>
<tr>
<td>5</td>
<td>1.0</td>
<td><strong>0.0</strong></td>
<td>1.0</td>
<td>2.0</td>
</tr>
</tbody>
</table>
<p>These maps are built from the ground up, starting at leaf nodes and working their way "upward" through the internal nodes of the <a href="#BM_(tree)%2Fi">tree</a>. Since the <a href="#BM_(sequence)%2Fi">sequence</a>s at leaf nodes are known (leaf nodes represent known entities), building their maps is fairly straightforward: 0.0 distance for the element at that index and ∞ distance for all other elements. For example, the <a href="#BM_(sequence)%2Fi">sequence</a> ACTGCT would generate the following mappings at each index ...</p>
<table>
<thead>
<tr>
<th></th>
<th>A</th>
<th>C</th>
<th>T</th>
<th>G</th>
</tr>
</thead>
<tbody>
<tr>
<td>0</td>
<td>0.0</td>
<td>∞</td>
<td>∞</td>
<td>∞</td>
</tr>
<tr>
<td>1</td>
<td>∞</td>
<td>0.0</td>
<td>∞</td>
<td>∞</td>
</tr>
<tr>
<td>2</td>
<td>∞</td>
<td>∞</td>
<td>0.0</td>
<td>∞</td>
</tr>
<tr>
<td>3</td>
<td>∞</td>
<td>∞</td>
<td>∞</td>
<td>0.0</td>
</tr>
<tr>
<td>4</td>
<td>∞</td>
<td>0.0</td>
<td>∞</td>
<td>∞</td>
</tr>
<tr>
<td>5</td>
<td>∞</td>
<td>∞</td>
<td>0.0</td>
<td>∞</td>
</tr>
</tbody>
</table>
<p><a href="ch7_code/src/sequence_phylogeny/SmallParsimony.py">ch7_code/src/sequence_phylogeny/SmallParsimony.py</a> (lines 188 to 198):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">distance_for_leaf_element_types</span>(<span class="hljs-params">
        elem_type_dst: <span class="hljs-built_in">str</span>,
        elem_types: <span class="hljs-built_in">str</span> = <span class="hljs-string">'ACTG'</span>
</span>) -&gt; <span class="hljs-built_in">dict</span>[<span class="hljs-built_in">str</span>, <span class="hljs-built_in">float</span>]:
    dist_set = {}
    <span class="hljs-keyword">for</span> e <span class="hljs-keyword">in</span> elem_types:
        <span class="hljs-keyword">if</span> e == elem_type_dst:
            dist_set[e] = <span class="hljs-number">0.0</span>
        <span class="hljs-keyword">else</span>:
            dist_set[e] = math.inf
    <span class="hljs-keyword">return</span> dist_set
</code></pre>
<p>Once all the downstream <a href="#BM_(neighbour)%2Fi">neighbour</a>s of an internal node have mappings, its mappings can be built by determining the minimized distance to reach each element. For example, imagine an internal node with 3 downstream <a href="#BM_(neighbour)%2Fi">neighbour</a>s...</p>
<p><img src="svgbob_12f104bbc0334c9194e786ece20223413b90d0cf.svg" alt="Kroki diagram output"></p>
<p>To determine A's value for the mapping at index 3, pull in index 3 from all downstream nodes...</p>
<p><img src="svgbob_2238c6763afef29be8f39f22341b36a0c6bea1fb.svg" alt="Kroki diagram output"></p>
<p>For each downstream index 3 mapping, walk over each element and add in the distance from A to that element, then select the minimum value ...</p>
<pre class="hljs"><code>n2_val = <span class="hljs-built_in">min</span>(
    N2[<span class="hljs-number">3</span>][<span class="hljs-string">'A'</span>] + dist_metric(<span class="hljs-string">'A'</span>, <span class="hljs-string">'A'</span>),  <span class="hljs-comment"># N2[3]['A']=2</span>
    N2[<span class="hljs-number">3</span>][<span class="hljs-string">'C'</span>] + dist_metric(<span class="hljs-string">'A'</span>, <span class="hljs-string">'C'</span>),  <span class="hljs-comment"># N2[3]['C']=4</span>
    N2[<span class="hljs-number">3</span>][<span class="hljs-string">'T'</span>] + dist_metric(<span class="hljs-string">'A'</span>, <span class="hljs-string">'T'</span>),  <span class="hljs-comment"># N2[3]['T']=1</span>
    N2[<span class="hljs-number">3</span>][<span class="hljs-string">'G'</span>] + dist_metric(<span class="hljs-string">'A'</span>, <span class="hljs-string">'G'</span>)   <span class="hljs-comment"># N2[3]['G']=4</span>
)
n3_val = <span class="hljs-built_in">min</span>(
    N3[<span class="hljs-number">3</span>][<span class="hljs-string">'A'</span>] + dist_metric(<span class="hljs-string">'A'</span>, <span class="hljs-string">'A'</span>),  <span class="hljs-comment"># N3[3]['A']=2</span>
    N3[<span class="hljs-number">3</span>][<span class="hljs-string">'C'</span>] + dist_metric(<span class="hljs-string">'A'</span>, <span class="hljs-string">'C'</span>),  <span class="hljs-comment"># N3[3]['C']=3</span>
    N3[<span class="hljs-number">3</span>][<span class="hljs-string">'T'</span>] + dist_metric(<span class="hljs-string">'A'</span>, <span class="hljs-string">'T'</span>),  <span class="hljs-comment"># N3[3]['T']=1</span>
    N3[<span class="hljs-number">3</span>][<span class="hljs-string">'G'</span>] + dist_metric(<span class="hljs-string">'A'</span>, <span class="hljs-string">'G'</span>)   <span class="hljs-comment"># N3[3]['G']=0</span>
)
n4_val = <span class="hljs-built_in">min</span>(
    N4[<span class="hljs-number">3</span>][<span class="hljs-string">'A'</span>] + dist_metric(<span class="hljs-string">'A'</span>, <span class="hljs-string">'A'</span>),  <span class="hljs-comment"># N4[3]['A']=1</span>
    N4[<span class="hljs-number">3</span>][<span class="hljs-string">'C'</span>] + dist_metric(<span class="hljs-string">'A'</span>, <span class="hljs-string">'C'</span>),  <span class="hljs-comment"># N4[3]['C']=3</span>
    N4[<span class="hljs-number">3</span>][<span class="hljs-string">'T'</span>] + dist_metric(<span class="hljs-string">'A'</span>, <span class="hljs-string">'T'</span>),  <span class="hljs-comment"># N4[3]['T']=1</span>
    N4[<span class="hljs-number">3</span>][<span class="hljs-string">'G'</span>] + dist_metric(<span class="hljs-string">'A'</span>, <span class="hljs-string">'G'</span>)   <span class="hljs-comment"># N4[3]['G']=0</span>
)
</code></pre>
<p>The sum of all values generated above produces the distance for A in the mapping. You can think of this distance as the minimum cost of transitioning to / from A ...</p>
<pre class="hljs"><code>N1[<span class="hljs-number">3</span>][<span class="hljs-string">'A'</span>] = n2_val + n3_val + n4_val
</code></pre>
<p>This same process is repeated for the remaining elements in the mapping (C, T, and G) to generate the full mapping for index 3.</p>
<p><a href="ch7_code/src/sequence_phylogeny/SmallParsimony.py">ch7_code/src/sequence_phylogeny/SmallParsimony.py</a> (lines 150 to 183):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">distance_for_internal_element_types</span>(<span class="hljs-params">
        downstream_dist_sets: Iterable[<span class="hljs-built_in">dict</span>[<span class="hljs-built_in">str</span>, <span class="hljs-built_in">float</span>]],
        dist_metric: <span class="hljs-type">Callable</span>[[<span class="hljs-built_in">str</span>, <span class="hljs-built_in">str</span>], <span class="hljs-built_in">float</span>],
        elem_types: <span class="hljs-built_in">str</span> = <span class="hljs-string">'ACTG'</span>
</span>) -&gt; <span class="hljs-built_in">dict</span>[<span class="hljs-built_in">str</span>, <span class="hljs-built_in">float</span>]:
    dist_set = {}
    <span class="hljs-keyword">for</span> elem_type <span class="hljs-keyword">in</span> elem_types:
        dist = distance_for_internal_element_type(
            downstream_dist_sets,
            dist_metric,
            elem_type,
            elem_types
        )
        dist_set[elem_type] = dist
    <span class="hljs-keyword">return</span> dist_set


<span class="hljs-keyword">def</span> <span class="hljs-title function_">distance_for_internal_element_type</span>(<span class="hljs-params">
        downstream_dist_sets: Iterable[<span class="hljs-built_in">dict</span>[<span class="hljs-built_in">str</span>, <span class="hljs-built_in">float</span>]],
        dist_metric: <span class="hljs-type">Callable</span>[[<span class="hljs-built_in">str</span>, <span class="hljs-built_in">str</span>], <span class="hljs-built_in">float</span>],
        elem_type_dst: <span class="hljs-built_in">str</span>,
        elem_types: <span class="hljs-built_in">str</span> = <span class="hljs-string">'ACTG'</span>
</span>) -&gt; <span class="hljs-built_in">float</span>:
    min_dists = []
    <span class="hljs-keyword">for</span> downstream_dist_set <span class="hljs-keyword">in</span> downstream_dist_sets:
        possible_dists = []
        <span class="hljs-keyword">for</span> elem_type_src <span class="hljs-keyword">in</span> elem_types:
            downstream_dist = downstream_dist_set[elem_type_src]
            transition_cost = dist_metric(elem_type_src, elem_type_dst)
            dist = downstream_dist + transition_cost
            possible_dists.append(dist)
        min_dist = <span class="hljs-built_in">min</span>(possible_dists)
        min_dists.append(min_dist)
    <span class="hljs-keyword">return</span> <span class="hljs-built_in">sum</span>(min_dists)
</code></pre>
<p>The algorithm builds these maps from the ground up, starting at leaf nodes and working their way "upward" through the internal nodes of the <a href="#BM_(tree)%2Fi">tree</a>. Since <a href="#BM_(phylogenetic%20tree%7Cphylogeny%20tree%7Cphylogenies%7Cevolutionary%20tree)%2Fi">phylogenetic tree</a>s are typically <a href="#BM_(un-rooted%20tree%7Cunrooted%20tree)%2Fi">unrooted tree</a>s, a node needs to be selected as the root such that the algorithm can work upward to that root. The inferred <a href="#BM_(sequence)%2Fi">sequence</a>s for internal nodes will very likely be different depending on which node is selected as root.</p>
<p><img src="svgbob_69d53a1f9394aef6a3e6deac1675d302e17f4aad.svg" alt="Kroki diagram output"></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The Pevzner book claims this is <a href="#BM_(dynamic%20programming%20algorithm%7Cdynamic%20programming)%2Fi">dynamic programming</a>. This is somewhat similar to how the backtracking <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">sequence alignment path</a> finding algorithm works (they're both graphs).</p>
</div>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>If the <a href="#BM_(tree)%2Fi">tree</a> is unrooted, the Pevzner book says to pick an edge and inject a fake root into it, then remove it once the <a href="#BM_(sequence)%2Fi">sequence</a>s have been inferred. It says that if the <a href="#BM_(tree)%2Fi">tree</a> is a binary <a href="#BM_(tree)%2Fi">tree</a> and <a href="#BM_(hamming%20distance)%2Fi">hamming distance</a> is used as the metric, the same element type will win at every index of every node (lowest distance) regardless of which edge the fake root was injected into. At least I think that's what it says -- maybe it means the <a href="#BM_(parsimony%20score)%2Fi">parsimony score</a> will be the same (<a href="#BM_(parsimony%20score)%2Fi">parsimony score</a> discussed in next section).</p>
<p>If the <a href="#BM_(tree)%2Fi">tree</a> isn't binary and/or something other than <a href="#BM_(hamming%20distance)%2Fi">hamming distance</a> is chosen as the metric, will this still be the case? If it isn't, I can't see how doing that is any better than just picking some internal node to be the root.</p>
<p>So which node should be selected as root? The <a href="#BM_(tree)%2Fi">tree</a> structure being used for this algorithm very likely came from a <a href="#BM_(phylogenetic%20tree%7Cphylogeny%20tree%7Cphylogenies%7Cevolutionary%20tree)%2Fi">phylogenetic tree</a> built using distances (e.g. <a href="#BM_(additive%20phylogeny)%2Fi">additive phylogeny</a>, <a href="#BM_(neighbour%20joining%20phylogeny)%2Fi">neighbour joining phylogeny</a>, <a href="#BM_(unweighted%20pair%20group%20method%20with%20arithmetic%20mean%7CUPGMA)%2Fi">UPGMA</a>, etc..). Here are a couple of ideas I just thought up:</p>
<ul>
<li>For each leaf node, count the number of nodes in the path to reach that internal node. Sum up the counts and pick the internal node with the largest sum as the root.</li>
<li>For each leaf node, calculate the distance to reach that internal node. Sum up the distances and pick the internal node with the largest sum as the root.</li>
</ul>
<p>I think the second one might not work because all sums will be the same? Maybe instead average the distances to leaf nodes and pick the one with the largest average?</p>
</div>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The algorithm doesn't factor in distances (edge weights). For example, if an internal node has 3 children, and one has a much shorter distance than the others, shouldn't the shorter one's <a href="#BM_(sequence)%2Fi">sequence</a> elements be given more of a preference over the highers (e.g. higher probability of showing up)?</p>
</div>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>In addition to <a href="#BM_(small%20parsimony%7Clarge%20parsimony)%2Fi">small parsimony</a>, there's <a href="#BM_(small%20parsimony%7Clarge%20parsimony)%2Fi">large parsimony</a>.</p>
<p><a href="#BM_(small%20parsimony%7Clarge%20parsimony)%2Fi">Small parsimony</a>: When a <a href="#BM_(tree)%2Fi">tree</a> structure and its leaf node <a href="#BM_(sequence)%2Fi">sequence</a>s are given, derive the internal node <a href="#BM_(sequence)%2Fi">sequence</a>s with the lowest possible distance (most <a href="#BM_(parsimony%7Cparsimonious)%2Fi">parsimonious</a>).</p>
<p><a href="#BM_(small%20parsimony%7Clarge%20parsimony)%2Fi">Large parsimony</a>: When only the leaf node <a href="#BM_(sequence)%2Fi">sequence</a>s are given, derive the combination of <a href="#BM_(tree)%2Fi">tree</a> structure and internal node <a href="#BM_(sequence)%2Fi">sequence</a>s with the lowest possible distance (most <a href="#BM_(parsimony%7Cparsimonious)%2Fi">parsimonious</a>).</p>
<p>Trying to do <a href="#BM_(small%20parsimony%7Clarge%20parsimony)%2Fi">large parsimony</a> explodes the search space (e.g. NP-complete), meaning it isn't realistic to attempt.</p>
</div>
<p><a href="ch7_code/src/sequence_phylogeny/SmallParsimony.py">ch7_code/src/sequence_phylogeny/SmallParsimony.py</a> (lines 54 to 146):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">populate_distance_sets</span>(<span class="hljs-params">
        tree: Graph[N, ND, E, ED],
        root: N,
        seq_length: <span class="hljs-built_in">int</span>,
        get_sequence: <span class="hljs-type">Callable</span>[[N], <span class="hljs-built_in">str</span>],
        set_sequence: <span class="hljs-type">Callable</span>[[N, <span class="hljs-built_in">str</span>], <span class="hljs-literal">None</span>],
        get_dist_set: <span class="hljs-type">Callable</span>[
            [
                N,   <span class="hljs-comment"># node</span>
                <span class="hljs-built_in">int</span>  <span class="hljs-comment"># index within N's sequence</span>
            ],
            <span class="hljs-built_in">dict</span>[<span class="hljs-built_in">str</span>, <span class="hljs-built_in">float</span>]
        ],
        set_dist_set: <span class="hljs-type">Callable</span>[
            [
                N,    <span class="hljs-comment"># node</span>
                <span class="hljs-built_in">int</span>,  <span class="hljs-comment"># index within N's sequence</span>
                <span class="hljs-built_in">dict</span>[<span class="hljs-built_in">str</span>, <span class="hljs-built_in">float</span>]
            ],
            <span class="hljs-literal">None</span>
        ],
        dist_metric: <span class="hljs-type">Callable</span>[[<span class="hljs-built_in">str</span>, <span class="hljs-built_in">str</span>], <span class="hljs-built_in">float</span>],
        elem_types: <span class="hljs-built_in">str</span> = <span class="hljs-string">'ACTG'</span>
</span>) -&gt; <span class="hljs-literal">None</span>:
    neighbours_unprocessed = Counter()
    <span class="hljs-keyword">for</span> n <span class="hljs-keyword">in</span> tree.get_nodes():
        neighbours_unprocessed[n] = tree.get_degree(n)
    leaf_nodes = {n <span class="hljs-keyword">for</span> n, c <span class="hljs-keyword">in</span> neighbours_unprocessed.items() <span class="hljs-keyword">if</span> c == <span class="hljs-number">1</span>}
    internal_nodes = {n <span class="hljs-keyword">for</span> n, c <span class="hljs-keyword">in</span> neighbours_unprocessed.items() <span class="hljs-keyword">if</span> c &gt; <span class="hljs-number">1</span>}

    <span class="hljs-comment"># Add +1 to the unprocessed count of the node deemed to be root. This</span>
    <span class="hljs-comment"># will make it so that it gets processed last.</span>
    <span class="hljs-keyword">assert</span> root <span class="hljs-keyword">in</span> neighbours_unprocessed
    neighbours_unprocessed[root] += <span class="hljs-number">1</span>

    <span class="hljs-comment"># Build dist sets for leaf nodes</span>
    <span class="hljs-keyword">for</span> n <span class="hljs-keyword">in</span> leaf_nodes:
        <span class="hljs-comment"># Build and set dist set for each element</span>
        seq = get_sequence(n)
        <span class="hljs-keyword">for</span> idx, elem <span class="hljs-keyword">in</span> <span class="hljs-built_in">enumerate</span>(seq):
            dist_set = distance_for_leaf_element_types(elem, elem_types)
            set_dist_set(n, idx, dist_set)
        <span class="hljs-comment"># Decrement waiting count for upstream neighbour</span>
        <span class="hljs-keyword">for</span> edge <span class="hljs-keyword">in</span> tree.get_outputs(n):
            n_upstream = tree.get_edge_end(edge, n)
            neighbours_unprocessed[n_upstream] -= <span class="hljs-number">1</span>
        <span class="hljs-comment"># Remove from pending nodes</span>
        neighbours_unprocessed.pop(n)

    <span class="hljs-comment"># Build dist sets for internal nodes (walking up from leaf nodes)</span>
    <span class="hljs-keyword">while</span> <span class="hljs-literal">True</span>:
        <span class="hljs-comment"># Get next node ready to be processed</span>
        ready = {n <span class="hljs-keyword">for</span> n, c <span class="hljs-keyword">in</span> neighbours_unprocessed.items() <span class="hljs-keyword">if</span> c == <span class="hljs-number">1</span>}
        <span class="hljs-keyword">if</span> <span class="hljs-keyword">not</span> ready:
            <span class="hljs-keyword">break</span>
        n = ready.pop()
        <span class="hljs-comment"># For each index, pull distance sets for outputs of n (that have them) and</span>
        <span class="hljs-comment"># use them to build a distance set for n.</span>
        <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(seq_length):
            downstream_dist_sets = []
            <span class="hljs-keyword">for</span> edge <span class="hljs-keyword">in</span> tree.get_outputs(n):
                n_downstream = tree.get_edge_end(edge, n)
                <span class="hljs-comment"># If it's root, treat all edges as pointing to downstream nodes</span>
                <span class="hljs-comment"># If it's not root, only nodes already processed are downstream nodes</span>
                <span class="hljs-keyword">if</span> n != root <span class="hljs-keyword">and</span> n_downstream <span class="hljs-keyword">in</span> neighbours_unprocessed:
                    <span class="hljs-keyword">continue</span>  <span class="hljs-comment"># Skip -- not root + not processed = actually upstream node</span>
                dist_set = get_dist_set(n_downstream, i)
                downstream_dist_sets.append(dist_set)
            dist_set = distance_for_internal_element_types(
                downstream_dist_sets,
                dist_metric,
                elem_types
            )
            set_dist_set(n, i, dist_set)
        <span class="hljs-comment"># Mark neighbours as processed</span>
        <span class="hljs-keyword">for</span> edge <span class="hljs-keyword">in</span> tree.get_outputs(n):
            n_upstream = tree.get_edge_end(edge, n)
            <span class="hljs-keyword">if</span> n_upstream <span class="hljs-keyword">in</span> neighbours_unprocessed:
                neighbours_unprocessed[n_upstream] -= <span class="hljs-number">1</span>
        <span class="hljs-comment"># Remove from pending nodes</span>
        neighbours_unprocessed.pop(n)

    <span class="hljs-comment"># Set sequences for internal nodes based on dist sets</span>
    <span class="hljs-keyword">for</span> n <span class="hljs-keyword">in</span> internal_nodes:
        seq = <span class="hljs-string">''</span>
        <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(seq_length):
            elem, _ = <span class="hljs-built_in">min</span>(
                ((elem, dist) <span class="hljs-keyword">for</span> elem, dist <span class="hljs-keyword">in</span> get_dist_set(n, i).items()),
                key=<span class="hljs-keyword">lambda</span> x: x[<span class="hljs-number">1</span>]  <span class="hljs-comment"># sort on dist</span>
            )
            seq += elem
        set_sequence(n, seq)
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>The tree...</p>
<p><img src="dot_a68c13de3d0e4bcf34af631d2256b12f.svg" alt="Dot diagram"></p>
<p>... with i0 set as its root and the distances ...</p>
<table>
<thead><tr>
<th></th>
<th>A</th>
<th>C</th>
<th>T</th>
<th>G</th>
</tr></thead>
<tbody>
<tr>
<td>A</td>
<td>0.0</td>
<td>1.0</td>
<td>1.0</td>
<td>1.0</td>
</tr>
<tr>
<td>C</td>
<td>1.0</td>
<td>0.0</td>
<td>1.0</td>
<td>1.0</td>
</tr>
<tr>
<td>T</td>
<td>1.0</td>
<td>1.0</td>
<td>0.0</td>
<td>1.0</td>
</tr>
<tr>
<td>G</td>
<td>1.0</td>
<td>1.0</td>
<td>1.0</td>
<td>0.0</td>
</tr>
</tbody>
</table>
<p>... has the following inferred ancestor sequences ...</p>
<p><img src="dot_3b2597d68f417ffbcd12f9ea9eb45c69.svg" alt="Dot diagram"></p>
</div>
<p></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The <a href="#BM_(distance%20metric)%2Fi">distance metric</a> used in the example execution above is <a href="#BM_(hamming%20distance)%2Fi">hamming distance</a>. If you're working with proteins, a more appropriate matrix might be a <a href="#BM_%5Cb(BLOSUM)%5Cd*%5Cb%2F">BLOSUM</a> matrix (e.g. <a href="#BM_%5Cb(BLOSUM)%5Cd*%5Cb%2F">BLOSUM</a>62). Whatever you use, just make sure to negate the values if appropriate -- it should be such that the lower the distance the stronger the affinity.</p>
</div>
<a name="H_Nearest%20Neighbour%20Interchange%20Algorithm"></a><h4>Nearest Neighbour Interchange Algorithm</h4>
<p><a name="BM_(Algorithms%5C%2FPhylogeny%2FSequence%20Inference%2FNearest%20Neighbour%20Algorithm)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FSequence%20Inference%2FSmall%20Parsimony%20Algorithm)_TOPIC%2F">Algorithms/Phylogeny/Sequence Inference/Small Parsimony Algorithm</a></li>
</ul>
</div>
<p><strong>ALGORITHM</strong>:</p>
<p>The problem with <a href="#BM_(small%20parsimony%7Clarge%20parsimony)%2Fi">small parsimony</a> is that inferred <a href="#BM_(sequence)%2Fi">sequence</a>s vary greatly based on both the given <a href="#BM_(tree)%2Fi">tree</a> structure and the element <a href="#BM_(distance%20metric)%2Fi">distance metric</a> used. Specifically, there are many ways in which...</p>
<ol>
<li>a <a href="#BM_(phylogenetic%20tree%7Cphylogeny%20tree%7Cphylogenies%7Cevolutionary%20tree)%2Fi">phylogenetic tree</a> structure can be generated (e.g. <a href="#BM_(unweighted%20pair%20group%20method%20with%20arithmetic%20mean%7CUPGMA)%2Fi">UPGMA</a>, <a href="#BM_(neighbour%20joining%20phylogeny)%2Fi">neighbour joining phylogeny</a>, etc..).</li>
<li>a distance can be generated between two elements of a <a href="#BM_(sequence)%2Fi">sequence</a> (e.g. <a href="#BM_%5Cb(PAM)%5Cd*%5Cb%2F">PAM</a>250, <a href="#BM_%5Cb(BLOSUM)%5Cd*%5Cb%2F">BLOSUM</a>62, <a href="#BM_(hamming%20distance)%2Fi">hamming distance</a>, etc...).</li>
</ol>
<p>Oftentimes the combination of <a href="#BM_(tree)%2Fi">tree</a> structure and internal node <a href="#BM_(sequence)%2Fi">sequence</a>s may be reasonable, but they likely aren't optimal (see <a href="#BM_(small%20parsimony%7Clarge%20parsimony)%2Fi">large parsimony</a>).</p>
<p>Given a <a href="#BM_(phylogenetic%20tree%7Cphylogeny%20tree%7Cphylogenies%7Cevolutionary%20tree)%2Fi">phylogenetic tree</a> where <a href="#BM_(small%20parsimony%7Clarge%20parsimony)%2Fi">small parsimony</a> has been applied, it's possible to derive a <a href="#BM_(parsimony%20score)%2Fi">parsimony score</a>: a measure of how likely the scenario is based on <a href="#BM_(parsimony%7Cparsimonious)%2Fi">parsimony</a>. For each edge, compute a weight by taking the two <a href="#BM_(sequence)%2Fi">sequence</a>s at its ends and summing the distances between the element pairs at each index. For example, ...</p>
<p><img src="svgbob_16d74c50ad8227599d75c5e09d62af6f5a25930f.svg" alt="Kroki diagram output"></p>
<p>The sum of edge weights is the <a href="#BM_(parsimony%20score)%2Fi">parsimony score</a> of the <a href="#BM_(tree)%2Fi">tree</a> (lower sum is better). For example, the following <a href="#BM_(tree)%2Fi">tree</a> has a <a href="#BM_(parsimony%20score)%2Fi">parsimony score</a> of 4...</p>
<p><img src="svgbob_f901ad93961054da4bd1b48a748231e62eaecec6.svg" alt="Kroki diagram output"></p>
<p><a href="ch7_code/src/sequence_phylogeny/NearestNeighbourInterchange.py">ch7_code/src/sequence_phylogeny/NearestNeighbourInterchange.py</a> (lines 114 to 141):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">parsimony_score</span>(<span class="hljs-params">
        tree: Graph[N, ND, E, ED],
        seq_length: <span class="hljs-built_in">int</span>,
        get_dist_set: <span class="hljs-type">Callable</span>[
            [
                N,  <span class="hljs-comment"># node</span>
                <span class="hljs-built_in">int</span>  <span class="hljs-comment"># index within N's sequence</span>
            ],
            <span class="hljs-built_in">dict</span>[<span class="hljs-built_in">str</span>, <span class="hljs-built_in">float</span>]
        ],
        set_edge_score: <span class="hljs-type">Callable</span>[[E, <span class="hljs-built_in">float</span>], <span class="hljs-literal">None</span>],
        dist_metric: <span class="hljs-type">Callable</span>[[<span class="hljs-built_in">str</span>, <span class="hljs-built_in">str</span>], <span class="hljs-built_in">float</span>]
</span>) -&gt; <span class="hljs-built_in">float</span>:
    total_score = <span class="hljs-number">0.0</span>
    edges = <span class="hljs-built_in">set</span>(tree.get_edges())  <span class="hljs-comment"># iterator to set -- avoids concurrent modification bug</span>
    <span class="hljs-keyword">for</span> e <span class="hljs-keyword">in</span> edges:
        n1, n2 = tree.get_edge_ends(e)
        e_score = <span class="hljs-number">0.0</span>
        <span class="hljs-keyword">for</span> idx <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(seq_length):
            n1_ds = get_dist_set(n1, idx)
            n2_ds = get_dist_set(n2, idx)
            n1_elem = <span class="hljs-built_in">min</span>(n1_ds, key=<span class="hljs-keyword">lambda</span> k: n1_ds[k])
            n2_elem = <span class="hljs-built_in">min</span>(n2_ds, key=<span class="hljs-keyword">lambda</span> k: n2_ds[k])
            e_score += dist_metric(n1_elem, n2_elem)
        set_edge_score(e, e_score)
        total_score += e_score
    <span class="hljs-keyword">return</span> total_score
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>The tree...</p>
<p><img src="dot_a68c13de3d0e4bcf34af631d2256b12f.svg" alt="Dot diagram"></p>
<p>... has a parsimony score of 4.0...</p>
<p><img src="dot_9c5b11c92142545ea08ba670225b5958.svg" alt="Dot diagram"></p>
</div>
<p></p>
<p>The nearest <a href="#BM_(neighbour)%2Fi">neighbour</a> interchange algorithm is a greedy heuristic which attempts to perturb the <a href="#BM_(tree)%2Fi">tree</a> to produce a lower <a href="#BM_(parsimony%20score)%2Fi">parsimony score</a>. The core operation of this algorithm is to pick an internal edge within the <a href="#BM_(tree)%2Fi">tree</a> and swap <a href="#BM_(neighbour)%2Fi">neighbour</a>s between the nodes at each end ...</p>
<p><img src="svgbob_a209a19385ac0e0a95667e64b6fa185c896caa9f.svg" alt="Kroki diagram output"></p>
<p>These swaps aren't just the nodes themselves, but the entire <a href="#BM_(subtree%7Csub-tree)%2Fi">sub-tree</a>s under those nodes. For example, ...</p>
<p><img src="svgbob_d02488346a63826fb1e96ed8e19ecc587aee966c.svg" alt="Kroki diagram output"></p>
<p><a href="ch7_code/src/sequence_phylogeny/NearestNeighbourInterchange.py">ch7_code/src/sequence_phylogeny/NearestNeighbourInterchange.py</a> (lines 49 to 110):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">list_nn_swap_options</span>(<span class="hljs-params">
        tree: Graph[N, ND, E, ED],
        edge: E
</span>) -&gt; <span class="hljs-built_in">set</span>[
    <span class="hljs-built_in">tuple</span>[
        <span class="hljs-built_in">frozenset</span>[E],  <span class="hljs-comment"># side1 edges</span>
        <span class="hljs-built_in">frozenset</span>[E]   <span class="hljs-comment"># side2 edges</span>
    ]
]:
    n1, n2 = tree.get_edge_ends(edge)
    n1_edges = <span class="hljs-built_in">set</span>(tree.get_outputs(n1))
    n2_edges = <span class="hljs-built_in">set</span>(tree.get_outputs(n2))
    n1_edges.remove(edge)
    n2_edges.remove(edge)
    n1_edges = <span class="hljs-built_in">frozenset</span>(n1_edges)
    n2_edges = <span class="hljs-built_in">frozenset</span>(n2_edges)
    n1_edge_cnt = <span class="hljs-built_in">len</span>(n1_edges)
    n2_edge_cnt = <span class="hljs-built_in">len</span>(n2_edges)
    both_edges = n1_edges | n2_edges
    ret = <span class="hljs-built_in">set</span>()
    <span class="hljs-keyword">for</span> n1_edges_perturbed <span class="hljs-keyword">in</span> combinations(both_edges, n1_edge_cnt):
        n1_edges_perturbed = <span class="hljs-built_in">frozenset</span>(n1_edges_perturbed)
        n2_edges_perturbed = <span class="hljs-built_in">frozenset</span>(both_edges.difference(n1_edges_perturbed))
        <span class="hljs-keyword">if</span> (n1_edges_perturbed, n2_edges_perturbed) <span class="hljs-keyword">in</span> ret:
            <span class="hljs-keyword">continue</span>
        <span class="hljs-keyword">if</span> (n2_edges_perturbed, n1_edges_perturbed) <span class="hljs-keyword">in</span> ret:
            <span class="hljs-keyword">continue</span>
        <span class="hljs-keyword">if</span> {n1_edges_perturbed, n2_edges_perturbed} == {n1_edges, n2_edges}:
            <span class="hljs-keyword">continue</span>
        ret.add((n1_edges_perturbed, n2_edges_perturbed))
    <span class="hljs-keyword">return</span> ret


<span class="hljs-keyword">def</span> <span class="hljs-title function_">nn_swap</span>(<span class="hljs-params">
    tree: Graph[N, ND, E, ED],
    edge: E,
    side1: <span class="hljs-built_in">frozenset</span>[E],
    side2: <span class="hljs-built_in">frozenset</span>[E]
</span>) -&gt; <span class="hljs-built_in">tuple</span>[
    <span class="hljs-built_in">frozenset</span>[E],  <span class="hljs-comment"># orig edges for side A</span>
    <span class="hljs-built_in">frozenset</span>[E]   <span class="hljs-comment"># orig edges for side B</span>
]:
    n1, n2 = tree.get_edge_ends(edge)
    n1_edges = <span class="hljs-built_in">set</span>(tree.get_outputs(n1))
    n2_edges = <span class="hljs-built_in">set</span>(tree.get_outputs(n2))
    n1_edges.remove(edge)
    n2_edges.remove(edge)
    <span class="hljs-keyword">assert</span> n1_edges | n2_edges == side1 | side2
    edge_details = {}
    <span class="hljs-keyword">for</span> e <span class="hljs-keyword">in</span> side1 | side2:
        end1, end2, data = tree.get_edge(e)
        end = {end1, end2}.difference({n1, n2}).pop()
        edge_details[e] = (end, data)
        tree.delete_edge(e)
    <span class="hljs-keyword">for</span> e <span class="hljs-keyword">in</span> side1:
        end, data = edge_details[e]
        tree.insert_edge(e, n1, end, data)
    <span class="hljs-keyword">for</span> e <span class="hljs-keyword">in</span> side2:
        end, data = edge_details[e]
        tree.insert_edge(e, n2, end, data)
    <span class="hljs-keyword">return</span> <span class="hljs-built_in">frozenset</span>(n1_edges), <span class="hljs-built_in">frozenset</span>(n2_edges)  <span class="hljs-comment"># return original edges</span>
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>The tree...</p>
<p><img src="dot_9b81e948066b89df103053f4b105da53.svg" alt="Dot diagram"></p>
<p>... can have any of the following nearest neighbour swaps on edge i0-i1...</p>
<p><img src="dot_91f37c7fea12ac74f5870ade915e7a0f.svg" alt="Dot diagram"></p>
<p><img src="dot_9e3e0c70bd5bc42ee008c4c120a23165.svg" alt="Dot diagram"></p>
<p><img src="dot_618f5a59247efef5120ce07be88aa291.svg" alt="Dot diagram"></p>
</div>
<p></p>
<p>Given a <a href="#BM_(tree)%2Fi">tree</a>, this algorithm goes over each internal edge and tries all possible <a href="#BM_(neighbour)%2Fi">neighbour</a> swaps on that edge in the hopes of driving down the <a href="#BM_(parsimony%20score)%2Fi">parsimony score</a>. After all possible swaps are performed on every internal edge, the swap that produced the lowest <a href="#BM_(parsimony%20score)%2Fi">parsimony score</a> is chosen. If that <a href="#BM_(parsimony%20score)%2Fi">parsimony score</a> is lower than the <a href="#BM_(parsimony%20score)%2Fi">parsimony score</a> for the original <a href="#BM_(tree)%2Fi">tree</a>, the swap is applied to the original and the process repeats.</p>
<p><a href="ch7_code/src/sequence_phylogeny/NearestNeighbourInterchange.py">ch7_code/src/sequence_phylogeny/NearestNeighbourInterchange.py</a> (lines 145 to 250):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">nn_interchange</span>(<span class="hljs-params">
        tree: Graph[N, ND, E, ED],
        root: N,
        seq_length: <span class="hljs-built_in">int</span>,
        get_sequence: <span class="hljs-type">Callable</span>[[N], <span class="hljs-built_in">str</span>],
        set_sequence: <span class="hljs-type">Callable</span>[[N, <span class="hljs-built_in">str</span>], <span class="hljs-literal">None</span>],
        get_dist_set: <span class="hljs-type">Callable</span>[
            [
                N,  <span class="hljs-comment"># node</span>
                <span class="hljs-built_in">int</span>  <span class="hljs-comment"># index within N's sequence</span>
            ],
            <span class="hljs-built_in">dict</span>[<span class="hljs-built_in">str</span>, <span class="hljs-built_in">float</span>]
        ],
        set_dist_set: <span class="hljs-type">Callable</span>[
            [
                N,  <span class="hljs-comment"># node</span>
                <span class="hljs-built_in">int</span>,  <span class="hljs-comment"># index within N's sequence</span>
                <span class="hljs-built_in">dict</span>[<span class="hljs-built_in">str</span>, <span class="hljs-built_in">float</span>]
            ],
            <span class="hljs-literal">None</span>
        ],
        dist_metric: <span class="hljs-type">Callable</span>[[<span class="hljs-built_in">str</span>, <span class="hljs-built_in">str</span>], <span class="hljs-built_in">float</span>],
        set_edge_score: <span class="hljs-type">Callable</span>[[E, <span class="hljs-built_in">float</span>], <span class="hljs-literal">None</span>],
        elem_types: <span class="hljs-built_in">str</span> = <span class="hljs-string">'ACTG'</span>,
        update_callback: <span class="hljs-type">Optional</span>[<span class="hljs-type">Callable</span>[[Graph, <span class="hljs-built_in">float</span>], <span class="hljs-literal">None</span>]] = <span class="hljs-literal">None</span>
</span>) -&gt; <span class="hljs-built_in">tuple</span>[<span class="hljs-built_in">float</span>, <span class="hljs-built_in">float</span>]:
    input_score = <span class="hljs-literal">None</span>
    output_score = <span class="hljs-literal">None</span>
    <span class="hljs-keyword">while</span> <span class="hljs-literal">True</span>:
        populate_distance_sets(
            tree,
            root,
            seq_length,
            get_sequence,
            set_sequence,
            get_dist_set,
            set_dist_set,
            dist_metric,
            elem_types
        )
        orig_score = parsimony_score(
            tree,
            seq_length,
            get_dist_set,
            set_edge_score,
            dist_metric
        )
        <span class="hljs-keyword">if</span> input_score <span class="hljs-keyword">is</span> <span class="hljs-literal">None</span>:
            input_score = orig_score
        output_score = orig_score
        <span class="hljs-keyword">if</span> update_callback <span class="hljs-keyword">is</span> <span class="hljs-keyword">not</span> <span class="hljs-literal">None</span>:
            update_callback(tree, output_score)  <span class="hljs-comment"># notify caller that the graph updated</span>
        swap_scores = []
        edges = <span class="hljs-built_in">set</span>(tree.get_edges())  <span class="hljs-comment"># bug -- avoids concurrent modification problems</span>
        <span class="hljs-keyword">for</span> edge <span class="hljs-keyword">in</span> edges:
            <span class="hljs-comment"># is it a limb? if so, skip it -- we want internal edges only</span>
            n1, n2 = tree.get_edge_ends(edge)
            <span class="hljs-keyword">if</span> tree.get_degree(n1) == <span class="hljs-number">1</span> <span class="hljs-keyword">or</span> tree.get_degree(n2) == <span class="hljs-number">1</span>:
                <span class="hljs-keyword">continue</span>
            <span class="hljs-comment"># get all possible nn swaps for this internal edge</span>
            options = list_nn_swap_options(tree, edge)
            <span class="hljs-comment"># for each possible swap...</span>
            <span class="hljs-keyword">for</span> swapped_side1, swapped_side2 <span class="hljs-keyword">in</span> options:
                <span class="hljs-comment"># swap</span>
                orig_side1, orig_side2 = nn_swap(
                    tree,
                    edge,
                    swapped_side1,
                    swapped_side2
                )
                <span class="hljs-comment"># small parsimony</span>
                populate_distance_sets(
                    tree,
                    root,
                    seq_length,
                    get_sequence,
                    set_sequence,
                    get_dist_set,
                    set_dist_set,
                    dist_metric,
                    elem_types
                )
                <span class="hljs-comment"># score and store</span>
                score = parsimony_score(
                    tree,
                    seq_length,
                    get_dist_set,
                    set_edge_score,
                    dist_metric
                )
                swap_scores.append((score, edge, swapped_side1, swapped_side2))
                <span class="hljs-comment"># unswap (back to original tree)</span>
                nn_swap(
                    tree,
                    edge,
                    orig_side1,
                    orig_side2
                )
        <span class="hljs-comment"># if swap producing the lowest parsimony score is lower than original, apply that</span>
        <span class="hljs-comment"># swap and try again, otherwise we're finished</span>
        score, edge, side1, side2 = <span class="hljs-built_in">min</span>(swap_scores, key=<span class="hljs-keyword">lambda</span> x: x[<span class="hljs-number">0</span>])
        <span class="hljs-keyword">if</span> score &gt;= orig_score:
            <span class="hljs-keyword">return</span> input_score, output_score
        <span class="hljs-keyword">else</span>:
            nn_swap(tree, edge, side1, side2)
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>The tree...</p>
<p><img src="dot_9b81e948066b89df103053f4b105da53.svg" alt="Dot diagram"></p>
<p>... with i0 set as its root and the distances ...</p>
<table>
<thead><tr>
<th></th>
<th>A</th>
<th>C</th>
<th>T</th>
<th>G</th>
</tr></thead>
<tbody>
<tr>
<td>A</td>
<td>0.0</td>
<td>1.0</td>
<td>1.0</td>
<td>1.0</td>
</tr>
<tr>
<td>C</td>
<td>1.0</td>
<td>0.0</td>
<td>1.0</td>
<td>1.0</td>
</tr>
<tr>
<td>T</td>
<td>1.0</td>
<td>1.0</td>
<td>0.0</td>
<td>1.0</td>
</tr>
<tr>
<td>G</td>
<td>1.0</td>
<td>1.0</td>
<td>1.0</td>
<td>0.0</td>
</tr>
</tbody>
</table>
<p>... has the following inferred ancestor sequences after using nearest neighbour interchange ...</p>
<p>graph score: 9.0</p>
<p><img src="dot_5c63144e975c84bc65b983dbf3f880c1.svg" alt="Dot diagram"></p>
<p>graph score: 6.0</p>
<p><img src="dot_dd926b5ceec12ba73951730b6d4c315f.svg" alt="Dot diagram"></p>
<p>After applying the nearest neighbour interchange heuristic, the tree updated to have a parismony score of 6.0 vs the original score of 9.0</p>
</div>
<p></p>
<a name="H_Gene%20Expression"></a><h2>Gene Expression</h2>
<p><a name="BM_(Algorithms%5C%2FGene%20Expression)_TOPIC%2F"></a><strong></strong></p>
<p><a href="#BM_(gene%20expression)%2Fi">Gene expression</a> is the biological process by which a <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a> (segment of DNA) is synthesized into a <a href="#BM_(gene%20product)%2Fi">gene product</a> (e.g. protein).</p>
<p><img src="svgbob_4c3340936a17e2840e08c6fbcd2b26b03b1c812b.svg" alt="Kroki diagram output"></p>
<p>When a <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a> <a href="#BM_(encode%7Cencoding%7Cdecode%7Cdecoding)%2Fi">encode</a>s for ...</p>
<ul>
<li>a protein, the <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a> is <a href="#BM_(transcription%7Ctranscribed%7Ctranscribe)%2Fi">transcribed</a> to mRNA which in turn is <a href="#BM_(translation%7Ctranslated%7Ctranslate)%2Fi">translated</a> to that protein.</li>
<li>functional RNA, the <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a> is <a href="#BM_(transcription%7Ctranscribed%7Ctranscribe)%2Fi">transcribed</a> to a type of RNA that isn't mRNA (only mRNA is <a href="#BM_(translation%7Ctranslated%7Ctranslate)%2Fi">translated</a> to a protein).</li>
</ul>
<p>A snapshot of all <a href="#BM_(RNA%20transcript%7Ctranscript)%2Fi">RNA transcript</a>s within a cell at a given point in time, called a <a href="#BM_(transcriptome)%2Fi">transcriptome</a>, can be captured using <a href="#BM_(RNA%20sequencing)%2Fi">RNA sequencing</a> technology. Both the <a href="#BM_(RNA%20transcript%7Ctranscript)%2Fi">RNA transcript</a> <a href="#BM_(sequence)%2Fi">sequence</a>s and the counts of those <a href="#BM_(sequence)%2Fi">sequence</a>s (number of instances) are captured. Given that an <a href="#BM_(RNA%20transcript%7Ctranscript)%2Fi">RNA transcript</a> is simply a <a href="#BM_(transcription%7Ctranscribed%7Ctranscribe)%2Fi">transcribed</a> "copy" of the DNA it came from (it identifies the <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a>), a snapshot indirectly shows the amount of <a href="#BM_(gene%20expression)%2Fi">gene expression</a> taking place for each <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a> at the time that snapshot was taken.</p>
<table>
<thead>
<tr>
<th></th>
<th>Count</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> / RNA A</td>
<td>100</td>
</tr>
<tr>
<td><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> / RNA B</td>
<td>70</td>
</tr>
<tr>
<td><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> / RNA C</td>
<td>110</td>
</tr>
<tr>
<td>...</td>
<td>...</td>
</tr>
</tbody>
</table>
<p>Biologists typically capture multiple snapshots to help identify which <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">genes</a> are influenced by or responsible for some change. The counts from each snapshot are placed together to form a matrix called a <a href="#BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi">gene expression matrix</a>, where each row in the matrix is called a <a href="#BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi">gene expression vector</a>. For example, the following <a href="#BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi">gene expression matrix</a> captures snapshots at regular intervals to help identify which <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">genes</a> are effected by a drug. Notice that the <a href="#BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi">gene expression vector</a> for B lowers after the drug is administered while C's elevates ...</p>
<table>
<thead>
<tr>
<th></th>
<th>1hr before drug given</th>
<th>0hr before/after drug given</th>
<th>1hr after drug given</th>
<th>...</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> A</td>
<td>100</td>
<td>100</td>
<td>100</td>
<td>...</td>
</tr>
<tr>
<td><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> B</td>
<td>100</td>
<td>70</td>
<td>50</td>
<td>...</td>
</tr>
<tr>
<td><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> C</td>
<td>100</td>
<td>110</td>
<td>140</td>
<td>...</td>
</tr>
<tr>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
</tbody>
</table>
<p>If a <a href="#BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi">gene expression vector</a> elevates/lowers across the set of snapshots, it may indicate that the <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a> is either responsible for or influenced by what happened.</p>
<p><img src="svgbob_244463cf241f3165a435a9043e21e7e80fc58e80.svg" alt="Kroki diagram output"></p>
<p>Similarly, if two or more <a href="#BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi">gene expression vector</a>s elevate/lower in a similar pattern, it could mean that the <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">genes</a> they represent either perform similar functions or are <a href="#BM_(co-regulate)%2Fi">co-regulate</a>d (e.g. each <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a> is influenced by the same <a href="#BM_(transcription%20factor)%2Fi">transcription factor</a>).</p>
<p><img src="svgbob_7fe990231108b57361d81844080fe92f2afe6105.svg" alt="Kroki diagram output"></p>
<p>In addition to the interval-based <a href="#BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi">gene expression matrices</a> described above, it's also common for <a href="#BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi">gene expression matrices</a> to state-based. State-based <a href="#BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi">gene expression matrices</a> combine snapshots from multiple organisms of the same species in different states. For example, there exists some set of <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">genes</a> that are <a href="#BM_(transcription%7Ctranscribed%7Ctranscribe)%2Fi">transcribed</a> more / less when comparing a ...</p>
<ul>
<li>cancerous blood cell vs non-cancerous blood cell.</li>
<li>rose cell that's bloomed vs rose cell that's un-bloomed.</li>
<li>bacteria cell when flagella is moving vs bacteria cell when flagella is still.</li>
<li>mouse cell that's inflamed vs mouse cell that's un-inflamed.</li>
</ul>
<table>
<thead>
<tr>
<th></th>
<th>blood cancer patient1</th>
<th>blood cancer patient2</th>
<th>cancer-free patient3</th>
<th>...</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> A</td>
<td>100</td>
<td>100</td>
<td>100</td>
<td>...</td>
</tr>
<tr>
<td><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> B</td>
<td>100</td>
<td>70</td>
<td>50</td>
<td>...</td>
</tr>
<tr>
<td><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> C</td>
<td>100</td>
<td>110</td>
<td>140</td>
<td>...</td>
</tr>
<tr>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
</tbody>
</table>
<p>Biologists determine which <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">genes</a> <em>may</em> be related to a change by grouping together similar <a href="#BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi">gene expression vector</a>s. The goal is for the <a href="#BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi">gene expression vector</a>s in each group to be more similar to each other than to those in other groups. The process of grouping items together in this way is called <a href="#BM_(cluster)%2Fi">cluster</a>ing, and each group formed by the process is called a <a href="#BM_(cluster)%2Fi">cluster</a>. For example, the <a href="#BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi">gene expression matrix</a> below clearly forms two <a href="#BM_(cluster)%2Fi">cluster</a>s if the <a href="#BM_(similarity%20metric)%2Fi">similarity metric</a> is the <a href="#BM_(euclidean%20distance)%2Fi">euclidean distance</a> between points...</p>
<table>
<thead>
<tr>
<th></th>
<th>1hr before</th>
<th>1hr after</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> A</td>
<td>5</td>
<td>1</td>
</tr>
<tr>
<td><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> B</td>
<td>20</td>
<td>1</td>
</tr>
<tr>
<td><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> C</td>
<td>24</td>
<td>4</td>
</tr>
<tr>
<td><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> D</td>
<td>1</td>
<td>2</td>
</tr>
<tr>
<td><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> E</td>
<td>3</td>
<td>4</td>
</tr>
<tr>
<td><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> F</td>
<td>22</td>
<td>4</td>
</tr>
</tbody>
</table>
<p><img src="svgbob_87187f74bb020ad404fc5577bd9e7827afd790af.svg" alt="Kroki diagram output"></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The goal described above is referred to as the <a href="#BM_(good%20clustering%20principle)%2Fi">good clustering principle</a>.</p>
</div>
<p>In the above example, <a href="#BM_(cluster)%2Fi">cluster</a> 1 reveals <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">genes</a> that weren't impacted while <a href="#BM_(cluster)%2Fi">cluster</a> 2 reveals <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">genes</a> that had their expression drastically lowered. Real-world <a href="#BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi">gene expression matrices</a> are often much more complex. Specifically, ...</p>
<ol>
<li>there are often more than two columns to a <a href="#BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi">gene expression matrix</a> (more than two dimensions), meaning that the <a href="#BM_(cluster)%2Fi">cluster</a>ing becomes non-trivial.</li>
<li><a href="#BM_(RNA%20sequencing)%2Fi">RNA sequencing</a> is an inherently a biased / noisy process, meaning that certain <a href="#BM_(RNA%20transcript%7Ctranscript)%2Fi">RNA transcript</a> counts elevating or lowering could be bad data.</li>
<li><a href="#BM_(RNA%20transcript%7Ctranscript)%2Fi">RNA transcript</a> counts can fluctuate due to normal cell operations (e.g. <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">genes</a> regulated by <a href="#BM_(circadian%20clock)%2Fi">circadian clock</a>), meaning that certain <a href="#BM_(RNA%20transcript%7Ctranscript)%2Fi">RNA transcript</a> counts elevating or lowering doesn't necessarily mean that they're relevant. This especially becomes a problem in state-based <a href="#BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi">gene expression matrices</a> where variables can't be as tightly controlled (e.g. in the blood cancer example above,the samples include people at different stages of cancer, could have been taken at different times of day, etc..).</li>
</ol>
<p>Prior to <a href="#BM_(cluster)%2Fi">cluster</a>ing, <a href="#BM_(RNA%20sequencing)%2Fi">RNA sequencing</a> outputs typically have to go through several rounds of processing (cleanup / normalization) to limit the impact of the last two points above. For example, to limit the impact of small fluctuations caused by either noisy data or normal cell operations, biologists often take the logarithm of the data. Doing so removes normal sized fluctuations but keeps drastic ones, which is often the ones biologists are interested in.</p>
<p><img src="svgbob_b482629285163bf79f8d3b8d936f4efc0cc87fa5.svg" alt="Kroki diagram output"></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>This section doesn't cover de-noising or de-biasing. It only covers <a href="#BM_(cluster)%2Fi">cluster</a>ing and common similarity / <a href="#BM_(distance%20metric)%2Fi">distance metric</a>s for real-valued vectors (which are what <a href="#BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi">gene expression vector</a>s are). Note that <a href="#BM_(cluster)%2Fi">cluster</a>ing can be used with data types other than vectors. For example, you can <a href="#BM_(cluster)%2Fi">cluster</a> protein <a href="#BM_(sequence)%2Fi">sequence</a>s where the <a href="#BM_(similarity%20metric)%2Fi">similarity metric</a> is the <a href="#BM_%5Cb(BLOSUM)%5Cd*%5Cb%2F">BLOSUM</a>62 score.</p>
</div>
<a name="H_Euclidean%20Distance%20Metric"></a><h3>Euclidean Distance Metric</h3>
<p><a name="BM_(Algorithms%5C%2FGene%20Expression%2FEuclidean%20Distance%20Metric)_TOPIC%2F"></a><strong></strong></p>
<p><strong>WHAT</strong>: Given two n-dimensional vectors, compute the distance between those vectors if traveling directly from one to the other in a straight line, referred to as the <a href="#BM_(euclidean%20distance)%2Fi">euclidean distance</a>.</p>
<p><img src="svgbob_812b661882246e5c5673044d47b2a743122f1f75.svg" alt="Kroki diagram output"></p>
<p><strong>WHY</strong>: This is one of many common metrics used for <a href="#BM_(cluster)%2Fi">cluster</a>ing <a href="#BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi">gene expression vector</a>s. One way to think about it is that it checks to see how closely component plots of the vectors match up. For example, ...</p>
<table>
<thead>
<tr>
<th></th>
<th>hour0</th>
<th>hour1</th>
<th>hour2</th>
<th>hour3</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> A</td>
<td>2</td>
<td>10</td>
<td>2</td>
<td>10</td>
</tr>
<tr>
<td><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> B</td>
<td>2</td>
<td>8</td>
<td>2</td>
<td>8</td>
</tr>
<tr>
<td><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> C</td>
<td>2</td>
<td>2</td>
<td>2</td>
<td>10</td>
</tr>
</tbody>
</table>
<p><img src="svgbob_9e826f4a4ade56f78f3c26b7b256969a30621136.svg" alt="Kroki diagram output"></p>
<ul>
<li><code>dist((2,10,2,10), (2,8,2,8)) = 2.82</code></li>
<li><code>dist((2,10,2,10), (2,2,2,10)) = 8</code></li>
<li><code>dist((2,8,2,8), (2,2,2,10)) = 6.325</code></li>
</ul>
<p><strong>ALGORITHM</strong>:</p>
<p>The algorithm extends the basic 2D distance formula from highschool math to multiple dimensions. Recall that two compute the distance between two...</p>
<ul>
<li>2D points is <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msqrt><mrow><mo stretchy="false">(</mo><msub><mi>x</mi><mn>2</mn></msub><mo>−</mo><msub><mi>x</mi><mn>1</mn></msub><msup><mo stretchy="false">)</mo><mn>2</mn></msup><mo>+</mo><mo stretchy="false">(</mo><msub><mi>y</mi><mn>2</mn></msub><mo>−</mo><msub><mi>y</mi><mn>1</mn></msub><msup><mo stretchy="false">)</mo><mn>2</mn></msup></mrow></msqrt></mrow><annotation encoding="application/x-tex">\sqrt{(x_2 - x_1)^2 + (y_2 - y_1)^2}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.24em;vertical-align:-0.30499999999999994em;"></span><span class="mord sqrt"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.935em;"><span class="svg-align" style="top:-3.2em;"><span class="pstrut" style="height:3.2em;"></span><span class="mord" style="padding-left:1em;"><span class="mopen">(</span><span class="mord"><span class="mord mathnormal">x</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal">x</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mclose"><span class="mclose">)</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.740108em;"><span style="top:-2.9890000000000003em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.03588em;">y</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.03588em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.03588em;">y</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.03588em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mclose"><span class="mclose">)</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.740108em;"><span style="top:-2.9890000000000003em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span></span></span><span style="top:-2.8950000000000005em;"><span class="pstrut" style="height:3.2em;"></span><span class="hide-tail" style="min-width:1.02em;height:1.28em;"><svg width="400em" height="1.28em" viewBox="0 0 400000 1296" preserveAspectRatio="xMinYMin slice"><path d="M263,681c0.7,0,18,39.7,52,119
c34,79.3,68.167,158.7,102.5,238c34.3,79.3,51.8,119.3,52.5,120
c340,-704.7,510.7,-1060.3,512,-1067
l0 -0
c4.7,-7.3,11,-11,19,-11
H40000v40H1012.3
s-271.3,567,-271.3,567c-38.7,80.7,-84,175,-136,283c-52,108,-89.167,185.3,-111.5,232
c-22.3,46.7,-33.8,70.3,-34.5,71c-4.7,4.7,-12.3,7,-23,7s-12,-1,-12,-1
s-109,-253,-109,-253c-72.7,-168,-109.3,-252,-110,-252c-10.7,8,-22,16.7,-34,26
c-22,17.3,-33.3,26,-34,26s-26,-26,-26,-26s76,-59,76,-59s76,-60,76,-60z
M1001 80h400000v40h-400000z"></path></svg></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.30499999999999994em;"><span></span></span></span></span></span></span></span></span>.</li>
<li>3D points is <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msqrt><mrow><mo stretchy="false">(</mo><msub><mi>x</mi><mn>2</mn></msub><mo>−</mo><msub><mi>x</mi><mn>1</mn></msub><msup><mo stretchy="false">)</mo><mn>2</mn></msup><mo>+</mo><mo stretchy="false">(</mo><msub><mi>y</mi><mn>2</mn></msub><mo>−</mo><msub><mi>y</mi><mn>1</mn></msub><msup><mo stretchy="false">)</mo><mn>2</mn></msup><mo>+</mo><mo stretchy="false">(</mo><msub><mi>z</mi><mn>2</mn></msub><mo>−</mo><msub><mi>z</mi><mn>1</mn></msub><msup><mo stretchy="false">)</mo><mn>2</mn></msup></mrow></msqrt></mrow><annotation encoding="application/x-tex">\sqrt{(x_2 - x_1)^2 + (y_2 - y_1)^2 + (z_2 - z_1)^2}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.24em;vertical-align:-0.30499999999999994em;"></span><span class="mord sqrt"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.935em;"><span class="svg-align" style="top:-3.2em;"><span class="pstrut" style="height:3.2em;"></span><span class="mord" style="padding-left:1em;"><span class="mopen">(</span><span class="mord"><span class="mord mathnormal">x</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal">x</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mclose"><span class="mclose">)</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.740108em;"><span style="top:-2.9890000000000003em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.03588em;">y</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.03588em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.03588em;">y</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.03588em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mclose"><span class="mclose">)</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.740108em;"><span style="top:-2.9890000000000003em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.04398em;">z</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.04398em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.04398em;">z</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.04398em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mclose"><span class="mclose">)</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.740108em;"><span style="top:-2.9890000000000003em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span></span></span><span style="top:-2.8950000000000005em;"><span class="pstrut" style="height:3.2em;"></span><span class="hide-tail" style="min-width:1.02em;height:1.28em;"><svg width="400em" height="1.28em" viewBox="0 0 400000 1296" preserveAspectRatio="xMinYMin slice"><path d="M263,681c0.7,0,18,39.7,52,119
c34,79.3,68.167,158.7,102.5,238c34.3,79.3,51.8,119.3,52.5,120
c340,-704.7,510.7,-1060.3,512,-1067
l0 -0
c4.7,-7.3,11,-11,19,-11
H40000v40H1012.3
s-271.3,567,-271.3,567c-38.7,80.7,-84,175,-136,283c-52,108,-89.167,185.3,-111.5,232
c-22.3,46.7,-33.8,70.3,-34.5,71c-4.7,4.7,-12.3,7,-23,7s-12,-1,-12,-1
s-109,-253,-109,-253c-72.7,-168,-109.3,-252,-110,-252c-10.7,8,-22,16.7,-34,26
c-22,17.3,-33.3,26,-34,26s-26,-26,-26,-26s76,-59,76,-59s76,-60,76,-60z
M1001 80h400000v40h-400000z"></path></svg></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.30499999999999994em;"><span></span></span></span></span></span></span></span></span>.</li>
</ul>
<p>In n dimensional space, this is calculated as <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msqrt><mrow><msubsup><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></msubsup><mrow><mo stretchy="false">(</mo><msub><mi>v</mi><mi>i</mi></msub><mo>−</mo><msub><mi>w</mi><mi>i</mi></msub><msup><mo stretchy="false">)</mo><mn>2</mn></msup></mrow></mrow></msqrt></mrow><annotation encoding="application/x-tex">\sqrt{\sum_{i=1}^n{(v_i - w_i)^2}}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.24em;vertical-align:-0.302709em;"></span><span class="mord sqrt"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.937291em;"><span class="svg-align" style="top:-3.2em;"><span class="pstrut" style="height:3.2em;"></span><span class="mord" style="padding-left:1em;"><span class="mop"><span class="mop op-symbol small-op" style="position:relative;top:-0.0000050000000000050004em;">∑</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.804292em;"><span style="top:-2.40029em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">i</span><span class="mrel mtight">=</span><span class="mord mtight">1</span></span></span></span><span style="top:-3.2029em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">n</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.29971000000000003em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord"><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.03588em;">v</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:-0.03588em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02691em;">w</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:-0.02691em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mclose"><span class="mclose">)</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.740108em;"><span style="top:-2.9890000000000003em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span></span></span></span><span style="top:-2.897291em;"><span class="pstrut" style="height:3.2em;"></span><span class="hide-tail" style="min-width:1.02em;height:1.28em;"><svg width="400em" height="1.28em" viewBox="0 0 400000 1296" preserveAspectRatio="xMinYMin slice"><path d="M263,681c0.7,0,18,39.7,52,119
c34,79.3,68.167,158.7,102.5,238c34.3,79.3,51.8,119.3,52.5,120
c340,-704.7,510.7,-1060.3,512,-1067
l0 -0
c4.7,-7.3,11,-11,19,-11
H40000v40H1012.3
s-271.3,567,-271.3,567c-38.7,80.7,-84,175,-136,283c-52,108,-89.167,185.3,-111.5,232
c-22.3,46.7,-33.8,70.3,-34.5,71c-4.7,4.7,-12.3,7,-23,7s-12,-1,-12,-1
s-109,-253,-109,-253c-72.7,-168,-109.3,-252,-110,-252c-10.7,8,-22,16.7,-34,26
c-22,17.3,-33.3,26,-34,26s-26,-26,-26,-26s76,-59,76,-59s76,-60,76,-60z
M1001 80h400000v40h-400000z"></path></svg></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.302709em;"><span></span></span></span></span></span></span></span></span>, where v and w are two n dimensional points.</p>
<p><a href="ch8_code/src/metrics/EuclideanDistance.py">ch8_code/src/metrics/EuclideanDistance.py</a> (lines 9 to 14):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">euclidean_distance</span>(<span class="hljs-params">v: <span class="hljs-type">Sequence</span>[<span class="hljs-built_in">float</span>], w: <span class="hljs-type">Sequence</span>[<span class="hljs-built_in">float</span>], dims: <span class="hljs-built_in">int</span></span>):
    x = <span class="hljs-number">0.0</span>
    <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(dims):
        x += (w[i] - v[i]) ** <span class="hljs-number">2</span>
    <span class="hljs-keyword">return</span> sqrt(x)
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the vectors ...</p>
<ul>
<li>[0.0, 1.0, 2.0]</li>
<li>[2.0, 1.0, 0.0]</li>
</ul>
<p>Their euclidean distance is 2.8284271247461903</p>
</div>
<p></p>
<a name="H_Manhattan%20Distance%20Metric"></a><h3>Manhattan Distance Metric</h3>
<p><a name="BM_(Algorithms%5C%2FGene%20Expression%2FManhattan%20Distance%20Metric)_TOPIC%2F"></a><strong></strong></p>
<p><strong>WHAT</strong>: Given two n-dimensional vectors, compute the distance between those vectors if traveling only via the axis of the coordinate system, referred to as the <a href="#BM_(manhattan%20distance%7Ctaxi%5B%5Cs%5C-%5D%3Fcab%20distance%7Ccity%5B%5Cs%5C-%5Dblock%20distance)%2Fi">manhattan distance</a>.</p>
<p><img src="svgbob_19c8269cf6dc61112d64c53c1045bfe1e1f518e3.svg" alt="Kroki diagram output"></p>
<p><strong>WHY</strong>: This is one of many common metrics used for <a href="#BM_(cluster)%2Fi">cluster</a>ing <a href="#BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi">gene expression vector</a>s.  One way to think about it is that it checks to see how closely component plots of the vectors match up. For example, ...</p>
<table>
<thead>
<tr>
<th></th>
<th>hour0</th>
<th>hour1</th>
<th>hour2</th>
<th>hour3</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> A</td>
<td>2</td>
<td>10</td>
<td>2</td>
<td>10</td>
</tr>
<tr>
<td><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> B</td>
<td>2</td>
<td>8</td>
<td>2</td>
<td>8</td>
</tr>
<tr>
<td><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> C</td>
<td>2</td>
<td>2</td>
<td>2</td>
<td>10</td>
</tr>
</tbody>
</table>
<p><img src="svgbob_9e826f4a4ade56f78f3c26b7b256969a30621136.svg" alt="Kroki diagram output"></p>
<ul>
<li><code>dist((2,10,2,10), (2,8,2,8)) = 4</code></li>
<li><code>dist((2,10,2,10), (2,2,2,10)) = 8</code></li>
<li><code>dist((2,8,2,8), (2,2,2,10)) = 8</code></li>
</ul>
<p><strong>ALGORITHM</strong>:</p>
<p>The algorithm sums the absolute differences between the elements at each index: <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msubsup><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></msubsup><mrow><mi mathvariant="normal">∣</mi><msub><mi>v</mi><mi>i</mi></msub><mo>−</mo><msub><mi>w</mi><mi>i</mi></msub><mi mathvariant="normal">∣</mi></mrow></mrow><annotation encoding="application/x-tex">\sum_{i=1}^n{|v_i - w_i|}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.104002em;vertical-align:-0.29971000000000003em;"></span><span class="mop"><span class="mop op-symbol small-op" style="position:relative;top:-0.0000050000000000050004em;">∑</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.804292em;"><span style="top:-2.40029em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">i</span><span class="mrel mtight">=</span><span class="mord mtight">1</span></span></span></span><span style="top:-3.2029em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">n</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.29971000000000003em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord"><span class="mord">∣</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.03588em;">v</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:-0.03588em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02691em;">w</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:-0.02691em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mord">∣</span></span></span></span></span>, where v and w are two n dimensional points. The absolute differences are used because a distance can't be negative.</p>
<p><a href="ch8_code/src/metrics/ManhattanDistance.py">ch8_code/src/metrics/ManhattanDistance.py</a> (lines 9 to 14):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">manhattan_distance</span>(<span class="hljs-params">v: <span class="hljs-type">Sequence</span>[<span class="hljs-built_in">float</span>], w: <span class="hljs-type">Sequence</span>[<span class="hljs-built_in">float</span>], dims: <span class="hljs-built_in">int</span></span>):
    x = <span class="hljs-number">0.0</span>
    <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(dims):
        x += <span class="hljs-built_in">abs</span>(w[i] - v[i])
    <span class="hljs-keyword">return</span> x
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the vectors ...</p>
<ul>
<li>[0.0, 1.0, 2.0]</li>
<li>[2.0, 1.0, 0.0]</li>
</ul>
<p>Their manhattan distance is 4.0</p>
</div>
<p></p>
<a name="H_Cosine%20Similarity%20Metric"></a><h3>Cosine Similarity Metric</h3>
<p><a name="BM_(Algorithms%5C%2FGene%20Expression%2FCosine%20Similarity%20Metric)_TOPIC%2F"></a><strong></strong></p>
<p><strong>WHAT</strong>: Given two n-dimensional vectors, compute the cosine of the single between them, referred to as the <a href="#BM_(cosine%20similarity%7Ccosine%20distance)%2Fi">cosine similarity</a>.</p>
<p><img src="svgbob_504c5b0c7c25a852e40496991ce6164fb25bfd05.svg" alt="Kroki diagram output"></p>
<p>This metric only factors in the angles between vectors, not their magnitudes (lengths). For example, imagine the following 2-dimensional <a href="#BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi">gene expression vector</a>s ...</p>
<table>
<thead>
<tr>
<th></th>
<th>before</th>
<th>after</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> U</td>
<td>9</td>
<td>9</td>
</tr>
<tr>
<td><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> T</td>
<td>15</td>
<td>32</td>
</tr>
<tr>
<td><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> C</td>
<td>3</td>
<td>0</td>
</tr>
<tr>
<td><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> J</td>
<td>21</td>
<td>21</td>
</tr>
</tbody>
</table>
<ul>
<li>
<p><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> U's count remained unchanged while <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a> C's count lowered to zero. The angle between those vectors is 45deg.</p>
<p><img src="svgbob_a53e618e5b8ad9142adca2dcbcd1acbcaa59ed0d.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> U's count remained unchanged while <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a> T's count approximately doubled. The angle between those vectors is roughly 20deg.</p>
<p><img src="svgbob_6ffa484453c5b156d0cd353c713d36d401fc604c.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> C's count lowered to zero but <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a> T's count approximately doubled. The angle between those vectors is roughly 65deg.</p>
<p><img src="svgbob_4fada2461e1ece3179922b290d9b31f2c453acfb.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> U and <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a> J's counts are different but both remained unchanged. The angle between those vectors is 0deg.</p>
<p><img src="svgbob_04b95b704a55c5b0910b9ccee54209850edd8882.svg" alt="Kroki diagram output"></p>
</li>
</ul>
<p>What's being compared is the trajectory at which the counts changed (angle between vectors), not the counts themselves (vector magnitudes). Given two <a href="#BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi">gene expression vector</a>s, if they grew/shrunk at ...</p>
<ul>
<li>exactly the same trajectory (angle of 0deg), they'll have maximum similarity: 0deg.</li>
<li>completely opposite trajectories, they'll have maximum dissimilarity: 180deg.</li>
</ul>
<p>Since the algorithm is calculating the cosine of the angle, the metric returns a result from from 1 to -1 instead of 0deg to 180deg, where ...</p>
<ul>
<li>maximum similarity is cos(0deg) = 1.</li>
<li>minimum similarity is cos(180deg) = -1.</li>
</ul>
<p><strong>WHY</strong>: Imagine the following two 4-dimensional <a href="#BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi">gene expression vector</a>s...</p>
<table>
<thead>
<tr>
<th></th>
<th>hour0</th>
<th>hour1</th>
<th>hour2</th>
<th>hour3</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> A</td>
<td>2</td>
<td>10</td>
<td>2</td>
<td>10</td>
</tr>
<tr>
<td><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> B</td>
<td>1</td>
<td>5</td>
<td>1</td>
<td>5</td>
</tr>
</tbody>
</table>
<p>Plotting out each component of the <a href="#BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi">gene expression vector</a>s above reveals that <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a> B's expression is a scaled down version of <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a> A's expression ...</p>
<p><img src="svgbob_3825eb8406ae4f47e2f738a066b7e507a34bffb4.svg" alt="Kroki diagram output"></p>
<p>The cosine of the angle between <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a> A's expression and <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a> B's expression is 1.0 (maximum similarity). This will always be the case as long as one <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a>'s expression is a linearly scaled version of the other <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a>'s expression. For example, the <a href="#BM_(cosine%20similarity%7Ccosine%20distance)%2Fi">cosine similarity</a> of ...</p>
<ul>
<li>(1, 5, 1, 5) and (1, 5, 1, 5) is 1.0 -- scaled 1x from first to second</li>
<li>(1, 5, 1, 5) and (2, 10, 2, 10) is 1.0 -- scaled 2x from first to second</li>
<li>(1, 5, 1, 5) and (3, 15, 3, 15) is 1.0 -- scaled 3x from first to second</li>
<li>(1, 5, 1, 5) and (1.5, 7.5, 1.5, 7.5) is 1.0 -- scaled 1.5x from first to second</li>
<li>(1, 5, 1, 5) and (0.5, 2.5, 0.5, 2.5) is 1.0 -- scaled 0.5x from first to second</li>
</ul>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>Still confused? Scaling makes sense if you think of it in terms of angles. The vectors (5,5) vs (10,10) have the same angle. Any vector with the same angle is just a scaled version of the other -- each of it's components are scaled by the same constant...</p>
<ul>
<li>(10,10) is (5,5) with each component scaled by 2.0x</li>
<li>(5,5) is (10,10) with each component scaled by 0.5x</li>
</ul>
<p><img src="svgbob_719a091a73ddb79c940edce9e59ee352db129d25.svg" alt="Kroki diagram output"></p>
</div>
<p>While <a href="#BM_(cosine%20similarity%7Ccosine%20distance)%2Fi">cosine similarity</a> does take into account scaling of components, <em>it doesn't support shifting of components</em>. Imagine the following 4-dimensional <a href="#BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi">gene expression vector</a>s...</p>
<table>
<thead>
<tr>
<th></th>
<th>hour0</th>
<th>hour1</th>
<th>hour2</th>
<th>hour3</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> A</td>
<td>2</td>
<td>10</td>
<td>2</td>
<td>10</td>
</tr>
<tr>
<td><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> B</td>
<td>1</td>
<td>5</td>
<td>1</td>
<td>5</td>
</tr>
<tr>
<td><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> C</td>
<td>5</td>
<td>9</td>
<td>5</td>
<td>9</td>
</tr>
</tbody>
</table>
<p>Plotting out each component of the <a href="#BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi">gene expression vector</a>s above reveals that...</p>
<ul>
<li>B is a scaled down version of A (by 0.5x).</li>
<li>C is a shifted version of B (by +4).</li>
<li>C is a scaled down and shifted version of A (by 0.5x and +4).</li>
</ul>
<p><img src="svgbob_843812b9fc06fffd9e97dae300501423d03b1e2e.svg" alt="Kroki diagram output"></p>
<p>All <a href="#BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi">gene expression vector</a>s follow the same pattern, notice that ...</p>
<ul>
<li>B is A where each component has been scaled by 0.5x, <a href="#BM_(cosine%20similarity%7Ccosine%20distance)%2Fi">cosine similarity</a> is 1.0 (maximum similarity).</li>
<li>C is A where each component has been scaled by 0.5x then shifted up by 4, <a href="#BM_(cosine%20similarity%7Ccosine%20distance)%2Fi">cosine similarity</a> is 0.992.</li>
<li>C is B where each component has been shifted up by 4, <a href="#BM_(cosine%20similarity%7Ccosine%20distance)%2Fi">cosine similarity</a> is 0.992.</li>
</ul>
<p>Even though the patterns are the same across all three <a href="#BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi">gene expression vector</a>s, <a href="#BM_(cosine%20similarity%7Ccosine%20distance)%2Fi">cosine similarity</a> gets thrown off in the presence of shifting.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>If you're trying to determine if the the components of the <a href="#BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi">gene expression vector</a>s follow the same pattern regardless of scale, the lack of shift support seems to make this unusable. The <a href="#BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi">gene expression vector</a>s may follow a similarly scaled patterns but it seems likely that each pattern is at an arbitrary offset (shift). So then what's the point of this? Why did the book mention if for <a href="#BM_(gene%20expression)%2Fi">gene expression</a> analysis?</p>
<p><a href="#BM_(pearson%20correlation%20coefficient%7Cpearson%20correlation%7Cpearson%20distance%7Cpearson%20similarity)%2Fi">Pearson similarity</a> seems to factor in both scaling and shifting. Use that instead.</p>
</div>
<p><strong>ALGORITHM</strong>:</p>
<p>Given the vectors A and B, the formula for the algorithm is as follows ...</p>
<div style="overflow: auto"><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><mi>c</mi><mi>o</mi><mi>s</mi><mo stretchy="false">(</mo><mi>θ</mi><mo stretchy="false">)</mo><mo>=</mo><mfrac><mrow><mi>A</mi><mo>⋅</mo><mi>B</mi></mrow><mrow><mi mathvariant="normal">∣</mi><mi mathvariant="normal">∣</mi><mi>A</mi><mi mathvariant="normal">∣</mi><mi mathvariant="normal">∣</mi><mtext> </mtext><mi mathvariant="normal">∣</mi><mi mathvariant="normal">∣</mi><mi>B</mi><mi mathvariant="normal">∣</mi><mi mathvariant="normal">∣</mi></mrow></mfrac></mrow><annotation encoding="application/x-tex">cos(\theta) = \frac{A \cdot B}{||A|| \: ||B||}
</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal">c</span><span class="mord mathnormal">o</span><span class="mord mathnormal">s</span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.02778em;">θ</span><span class="mclose">)</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:2.29633em;vertical-align:-0.936em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.36033em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">∣</span><span class="mord">∣</span><span class="mord mathnormal">A</span><span class="mord">∣</span><span class="mord">∣</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">∣</span><span class="mord">∣</span><span class="mord mathnormal" style="margin-right:0.05017em;">B</span><span class="mord">∣</span><span class="mord">∣</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord mathnormal">A</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord mathnormal" style="margin-right:0.05017em;">B</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.936em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span></span></div><p>The formula is confusing in that the ...</p>
<ul>
<li>central dot is the dot-product between vectors: <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>A</mi><mo>⋅</mo><mi>B</mi><mo>=</mo><msubsup><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></msubsup><mrow><msub><mi>A</mi><mi>i</mi></msub><mo>⋅</mo><msub><mi>B</mi><mi>i</mi></msub></mrow></mrow><annotation encoding="application/x-tex">A \cdot B = \sum_{i=1}^n {A_i \cdot B_i}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.68333em;vertical-align:0em;"></span><span class="mord mathnormal">A</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.68333em;vertical-align:0em;"></span><span class="mord mathnormal" style="margin-right:0.05017em;">B</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:1.104002em;vertical-align:-0.29971000000000003em;"></span><span class="mop"><span class="mop op-symbol small-op" style="position:relative;top:-0.0000050000000000050004em;">∑</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.804292em;"><span style="top:-2.40029em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">i</span><span class="mrel mtight">=</span><span class="mord mtight">1</span></span></span></span><span style="top:-3.2029em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">n</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.29971000000000003em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord"><span class="mord"><span class="mord mathnormal">A</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.05017em;">B</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:-0.05017em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span></span></span></span></span>.</li>
<li>double pipe encapsulation represents the magnitude of a vector: <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi mathvariant="normal">∣</mi><mi mathvariant="normal">∣</mi><mi>A</mi><mi mathvariant="normal">∣</mi><mi mathvariant="normal">∣</mi><mo>=</mo><msqrt><mrow><msubsup><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></msubsup><msubsup><mi>A</mi><mi>i</mi><mn>2</mn></msubsup></mrow></msqrt></mrow><annotation encoding="application/x-tex">||A|| = \sqrt{\sum_{i=1}^{n} {A_i^2}}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord">∣</span><span class="mord">∣</span><span class="mord mathnormal">A</span><span class="mord">∣</span><span class="mord">∣</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:1.24em;vertical-align:-0.302709em;"></span><span class="mord sqrt"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.937291em;"><span class="svg-align" style="top:-3.2em;"><span class="pstrut" style="height:3.2em;"></span><span class="mord" style="padding-left:1em;"><span class="mop"><span class="mop op-symbol small-op" style="position:relative;top:-0.0000050000000000050004em;">∑</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.804292em;"><span style="top:-2.40029em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">i</span><span class="mrel mtight">=</span><span class="mord mtight">1</span></span></span></span><span style="top:-3.2029em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">n</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.29971000000000003em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord"><span class="mord"><span class="mord mathnormal">A</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.7959080000000001em;"><span style="top:-2.4231360000000004em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span><span style="top:-3.0448000000000004em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.27686399999999994em;"><span></span></span></span></span></span></span></span></span></span><span style="top:-2.897291em;"><span class="pstrut" style="height:3.2em;"></span><span class="hide-tail" style="min-width:1.02em;height:1.28em;"><svg width="400em" height="1.28em" viewBox="0 0 400000 1296" preserveAspectRatio="xMinYMin slice"><path d="M263,681c0.7,0,18,39.7,52,119
c34,79.3,68.167,158.7,102.5,238c34.3,79.3,51.8,119.3,52.5,120
c340,-704.7,510.7,-1060.3,512,-1067
l0 -0
c4.7,-7.3,11,-11,19,-11
H40000v40H1012.3
s-271.3,567,-271.3,567c-38.7,80.7,-84,175,-136,283c-52,108,-89.167,185.3,-111.5,232
c-22.3,46.7,-33.8,70.3,-34.5,71c-4.7,4.7,-12.3,7,-23,7s-12,-1,-12,-1
s-109,-253,-109,-253c-72.7,-168,-109.3,-252,-110,-252c-10.7,8,-22,16.7,-34,26
c-22,17.3,-33.3,26,-34,26s-26,-26,-26,-26s76,-59,76,-59s76,-60,76,-60z
M1001 80h400000v40h-400000z"></path></svg></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.302709em;"><span></span></span></span></span></span></span></span></span>.</li>
<li>two double pipe encapsulations is a multiplication of vector magnitudes: <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi mathvariant="normal">∣</mi><mi mathvariant="normal">∣</mi><mi>A</mi><mi mathvariant="normal">∣</mi><mi mathvariant="normal">∣</mi><mtext> </mtext><mi mathvariant="normal">∣</mi><mi mathvariant="normal">∣</mi><mi>B</mi><mi mathvariant="normal">∣</mi><mi mathvariant="normal">∣</mi><mo>=</mo><mi mathvariant="normal">∣</mi><mi mathvariant="normal">∣</mi><mi>A</mi><mi mathvariant="normal">∣</mi><mi mathvariant="normal">∣</mi><mo>⋅</mo><mi mathvariant="normal">∣</mi><mi mathvariant="normal">∣</mi><mi>B</mi><mi mathvariant="normal">∣</mi><mi mathvariant="normal">∣</mi></mrow><annotation encoding="application/x-tex">||A|| : ||B|| = ||A|| \cdot ||B||</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord">∣</span><span class="mord">∣</span><span class="mord mathnormal">A</span><span class="mord">∣</span><span class="mord">∣</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord">∣</span><span class="mord">∣</span><span class="mord mathnormal" style="margin-right:0.05017em;">B</span><span class="mord">∣</span><span class="mord">∣</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord">∣</span><span class="mord">∣</span><span class="mord mathnormal">A</span><span class="mord">∣</span><span class="mord">∣</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord">∣</span><span class="mord">∣</span><span class="mord mathnormal" style="margin-right:0.05017em;">B</span><span class="mord">∣</span><span class="mord">∣</span></span></span></span></li>
</ul>
<div style="overflow: auto"><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><mi>c</mi><mi>o</mi><mi>s</mi><mo stretchy="false">(</mo><mi>θ</mi><mo stretchy="false">)</mo><mo>=</mo><mfrac><mrow><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><mrow><msub><mi>A</mi><mi>i</mi></msub><mo>⋅</mo><msub><mi>B</mi><mi>i</mi></msub></mrow></mrow><mrow><msqrt><mrow><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><msubsup><mi>A</mi><mi>i</mi><mn>2</mn></msubsup></mrow></msqrt><mo>⋅</mo><msqrt><mrow><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><msubsup><mi>B</mi><mi>i</mi><mn>2</mn></msubsup></mrow></msqrt></mrow></mfrac></mrow><annotation encoding="application/x-tex">cos(\theta) = \frac{\sum_{i=1}^n {A_i \cdot B_i}}{\sqrt{\sum_{i=1}^n {A_i^2}} \cdot \sqrt{\sum_{i=1}^n {B_i^2}}}
</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal">c</span><span class="mord mathnormal">o</span><span class="mord mathnormal">s</span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.02778em;">θ</span><span class="mclose">)</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:2.624002em;vertical-align:-1.13em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.494002em;"><span style="top:-2.1727090000000002em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord sqrt"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.937291em;"><span class="svg-align" style="top:-3.2em;"><span class="pstrut" style="height:3.2em;"></span><span class="mord" style="padding-left:1em;"><span class="mop"><span class="mop op-symbol small-op" style="position:relative;top:-0.0000050000000000050004em;">∑</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.804292em;"><span style="top:-2.40029em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">i</span><span class="mrel mtight">=</span><span class="mord mtight">1</span></span></span></span><span style="top:-3.2029em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">n</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.29971000000000003em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord"><span class="mord"><span class="mord mathnormal">A</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.7959080000000001em;"><span style="top:-2.4231360000000004em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span><span style="top:-3.0448000000000004em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.27686399999999994em;"><span></span></span></span></span></span></span></span></span></span><span style="top:-2.897291em;"><span class="pstrut" style="height:3.2em;"></span><span class="hide-tail" style="min-width:1.02em;height:1.28em;"><svg width="400em" height="1.28em" viewBox="0 0 400000 1296" preserveAspectRatio="xMinYMin slice"><path d="M263,681c0.7,0,18,39.7,52,119
c34,79.3,68.167,158.7,102.5,238c34.3,79.3,51.8,119.3,52.5,120
c340,-704.7,510.7,-1060.3,512,-1067
l0 -0
c4.7,-7.3,11,-11,19,-11
H40000v40H1012.3
s-271.3,567,-271.3,567c-38.7,80.7,-84,175,-136,283c-52,108,-89.167,185.3,-111.5,232
c-22.3,46.7,-33.8,70.3,-34.5,71c-4.7,4.7,-12.3,7,-23,7s-12,-1,-12,-1
s-109,-253,-109,-253c-72.7,-168,-109.3,-252,-110,-252c-10.7,8,-22,16.7,-34,26
c-22,17.3,-33.3,26,-34,26s-26,-26,-26,-26s76,-59,76,-59s76,-60,76,-60z
M1001 80h400000v40h-400000z"></path></svg></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.302709em;"><span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord sqrt"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.937291em;"><span class="svg-align" style="top:-3.2em;"><span class="pstrut" style="height:3.2em;"></span><span class="mord" style="padding-left:1em;"><span class="mop"><span class="mop op-symbol small-op" style="position:relative;top:-0.0000050000000000050004em;">∑</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.804292em;"><span style="top:-2.40029em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">i</span><span class="mrel mtight">=</span><span class="mord mtight">1</span></span></span></span><span style="top:-3.2029em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">n</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.29971000000000003em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord"><span class="mord"><span class="mord mathnormal" style="margin-right:0.05017em;">B</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.7959080000000001em;"><span style="top:-2.4231360000000004em;margin-left:-0.05017em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span><span style="top:-3.0448000000000004em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.27686399999999994em;"><span></span></span></span></span></span></span></span></span></span><span style="top:-2.897291em;"><span class="pstrut" style="height:3.2em;"></span><span class="hide-tail" style="min-width:1.02em;height:1.28em;"><svg width="400em" height="1.28em" viewBox="0 0 400000 1296" preserveAspectRatio="xMinYMin slice"><path d="M263,681c0.7,0,18,39.7,52,119
c34,79.3,68.167,158.7,102.5,238c34.3,79.3,51.8,119.3,52.5,120
c340,-704.7,510.7,-1060.3,512,-1067
l0 -0
c4.7,-7.3,11,-11,19,-11
H40000v40H1012.3
s-271.3,567,-271.3,567c-38.7,80.7,-84,175,-136,283c-52,108,-89.167,185.3,-111.5,232
c-22.3,46.7,-33.8,70.3,-34.5,71c-4.7,4.7,-12.3,7,-23,7s-12,-1,-12,-1
s-109,-253,-109,-253c-72.7,-168,-109.3,-252,-110,-252c-10.7,8,-22,16.7,-34,26
c-22,17.3,-33.3,26,-34,26s-26,-26,-26,-26s76,-59,76,-59s76,-60,76,-60z
M1001 80h400000v40h-400000z"></path></svg></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.302709em;"><span></span></span></span></span></span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.6897100000000003em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mop"><span class="mop op-symbol small-op" style="position:relative;top:-0.0000050000000000050004em;">∑</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.804292em;"><span style="top:-2.40029em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">i</span><span class="mrel mtight">=</span><span class="mord mtight">1</span></span></span></span><span style="top:-3.2029em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">n</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.29971000000000003em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord"><span class="mord"><span class="mord mathnormal">A</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.05017em;">B</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:-0.05017em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.13em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span></span></div><div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>What is the formula actually calculating / what's the reasoning behind the formula? The only part I understand is the magnitude calculation, which is just the <a href="#BM_(euclidean%20distance)%2Fi">euclidean distance</a> between the origin and the coordinates of a vector. For example, the magnitude between (0,0) and (5,7) is calculated as sqrt((5-0)^2 + (7-0)^2). Since the components of the origin are all always going to be 0, it can be shortened to sqrt(5^2 + 7^2).</p>
<p>The rest of it I don't understand. What is the <a href="#BM_(dot%20product)%2Fi">dot product</a> actually calculating? And why multiply the magnitudes and divide? How does that result in the cosine of the angle?</p>
</div>
<p><a href="ch8_code/src/metrics/CosineSimilarity.py">ch8_code/src/metrics/CosineSimilarity.py</a> (lines 9 to 25):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">cosine_similarity</span>(<span class="hljs-params">v: <span class="hljs-type">Sequence</span>[<span class="hljs-built_in">float</span>], w: <span class="hljs-type">Sequence</span>[<span class="hljs-built_in">float</span>], dims: <span class="hljs-built_in">int</span></span>):
    vec_dp = <span class="hljs-built_in">sum</span>(v[i] * w[i] <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(dims))
    v_mag = sqrt(<span class="hljs-built_in">sum</span>(v[i] ** <span class="hljs-number">2</span> <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(dims)))
    w_mag = sqrt(<span class="hljs-built_in">sum</span>(w[i] ** <span class="hljs-number">2</span> <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(dims)))
    <span class="hljs-keyword">return</span> vec_dp / (v_mag * w_mag)


<span class="hljs-keyword">def</span> <span class="hljs-title function_">cosine_distance</span>(<span class="hljs-params">v: <span class="hljs-type">Sequence</span>[<span class="hljs-built_in">float</span>], w: <span class="hljs-type">Sequence</span>[<span class="hljs-built_in">float</span>], dims: <span class="hljs-built_in">int</span></span>):
    <span class="hljs-comment"># To turn cosine similarity into a distance metric, subtract 1.0 from it. By</span>
    <span class="hljs-comment"># subtracting 1.0, you're changing the bounds from [1.0, -1.0] to [0.0, 2.0].</span>
    <span class="hljs-comment">#</span>
    <span class="hljs-comment"># Recall that any distance metric must return 0 when the items being compared</span>
    <span class="hljs-comment"># are the same and increases the more different they get. By subtracting 1.0,</span>
    <span class="hljs-comment"># you're matching that distance metric requirement: 0.0 when totally similar</span>
    <span class="hljs-comment"># and 2.0 for totally dissimilar.</span>
    <span class="hljs-keyword">return</span> <span class="hljs-number">1.0</span> - cosine_similarity(v, w, dims)
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the vectors ...</p>
<ul>
<li>[9.0, 9.0, 9.0]</li>
<li>[90.0, 90.0, 95.0]</li>
</ul>
<p>Their cosine similarity is 0.9996695853205689</p>
</div>
<p></p>
<a name="H_Pearson%20Similarity%20Metric"></a><h3>Pearson Similarity Metric</h3>
<p><a name="BM_(Algorithms%5C%2FGene%20Expression%2FPearson%20Similarity%20Metric)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FGene%20Expression%2FCosine%20Similarity%20Metric)_TOPIC%2F">Algorithms/Gene Expression/Cosine Similarity Metric</a></li>
</ul>
</div>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>A lot of what's below is my understanding of what's going on, which I'm almost certain is flawed. I've put up a <a href="https://stats.stackexchange.com/q/558913">question asking for help</a>.</p>
</div>
<p><strong>WHAT</strong>: Given two n-dimensional vectors, ...</p>
<ol>
<li>
<p>pair together each index to produce a set of 2D points.</p>
<p><img src="svgbob_8a51306e40078da605b817e49a9cfc8efd9ef259.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>fit a straight line to those points.</p>
<p><img src="svgbob_34537ac0704f3bc4cc192e942b87c516be82d143.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>quantify the proximity of those points to the fitted line, where the proximity of larger points contribute more to a strong similarity than smaller points. The quantity ranges from 0.0 (loose proximity) and 1.0 (tight proximity), negating the if the slope of the fitted line is negative.</p>
<p><img src="svgbob_56335bffa65afe36045fbbc42b3c280b1f643c9a.svg" alt="Kroki diagram output"></p>
</li>
</ol>
<p><strong>WHY</strong>: Imagine the following 4-dimensional <a href="#BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi">gene expression vector</a>s...</p>
<table>
<thead>
<tr>
<th></th>
<th>hour0</th>
<th>hour1</th>
<th>hour2</th>
<th>hour3</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> A</td>
<td>2</td>
<td>10</td>
<td>2</td>
<td>10</td>
</tr>
<tr>
<td><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> B</td>
<td>1</td>
<td>5</td>
<td>1</td>
<td>5</td>
</tr>
<tr>
<td><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> C</td>
<td>5</td>
<td>9</td>
<td>5</td>
<td>9</td>
</tr>
</tbody>
</table>
<p>Plotting out each component of the <a href="#BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi">gene expression vector</a>s above reveals that ...</p>
<ul>
<li>B is a scaled down version of A (by 0.5x).</li>
<li>C is a shifted version of B (by +4).</li>
<li>C is a scaled down and shifted version of A (by 0.5x and +4).</li>
</ul>
<p><img src="svgbob_bf01aa75a9035951e9d9462b6a70567479a339e8.svg" alt="Kroki diagram output"></p>
<p><a href="#BM_(pearson%20correlation%20coefficient%7Cpearson%20correlation%7Cpearson%20distance%7Cpearson%20similarity)%2Fi">Pearson similarity</a> returns 1.0 (maximum similarity) for all possible comparison of the three <a href="#BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi">gene expression vector</a>s above. Note that this isn't the case with <a href="#BM_(cosine%20similarity%7Ccosine%20distance)%2Fi">cosine similarity</a>. <a href="#BM_(cosine%20similarity%7Ccosine%20distance)%2Fi">Cosine similarity</a> gets thrown off in the presence of shifting while <a href="#BM_(pearson%20correlation%20coefficient%7Cpearson%20correlation%7Cpearson%20distance%7Cpearson%20similarity)%2Fi">pearson similarity</a> does not.</p>
<table>
<thead>
<tr>
<th></th>
<th><a href="#BM_(cosine%20similarity%7Ccosine%20distance)%2Fi">Cosine similarity</a></th>
<th><a href="#BM_(pearson%20correlation%20coefficient%7Cpearson%20correlation%7Cpearson%20distance%7Cpearson%20similarity)%2Fi">Pearson similarity</a></th>
</tr>
</thead>
<tbody>
<tr>
<td>B vs A</td>
<td>1.0</td>
<td>1.0</td>
</tr>
<tr>
<td>C vs A</td>
<td>0.992</td>
<td>1.0</td>
</tr>
<tr>
<td>C vs B</td>
<td>0.992</td>
<td>1.0</td>
</tr>
</tbody>
</table>
<p>Similarly, comparing a <a href="#BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi">gene expression vector</a> with a mirror (across the X-axis) that has been scaled and / or shifted will result in a <a href="#BM_(pearson%20correlation%20coefficient%7Cpearson%20correlation%7Cpearson%20distance%7Cpearson%20similarity)%2Fi">pearson similarity</a> of -1.0.</p>
<p><img src="svgbob_be0bbfb08a95364703d513e64acc89a395b495e4.svg" alt="Kroki diagram output"></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>If you're trying to determine if the the components of the <a href="#BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi">gene expression vector</a>s follow the same pattern regardless of scale OR offset, this is the similarity to use. They may have similar patterns even though they're scaled differently or offset differently. For example, both <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">genes</a> below may be influenced by the same <a href="#BM_(transcription%20factor)%2Fi">transcription factor</a>, but their base expression rates are different so the <a href="#BM_(transcription%20factor)%2Fi">transcription factor</a> influences their <a href="#BM_(gene%20expression)%2Fi">gene expression</a> proportionally.</p>
<p><img src="svgbob_357557bde44ac14c02730aba47f77f1f29bde3a2.svg" alt="Kroki diagram output"></p>
</div>
<p><strong>ALGORITHM</strong>:</p>
<p>Given the vectors A and B, the formula for the algorithm is as follows ...</p>
<div style="overflow: auto"><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><msub><mi>r</mi><mrow><mi>A</mi><mi>B</mi></mrow></msub><mo>=</mo><mfrac><mrow><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><mrow><mo stretchy="false">(</mo><msub><mi>A</mi><mi>i</mi></msub><mo>−</mo><mi>a</mi><mi>v</mi><mi>g</mi><mo stretchy="false">(</mo><mi>A</mi><mo stretchy="false">)</mo><mo stretchy="false">)</mo><mo stretchy="false">(</mo><msub><mi>B</mi><mi>i</mi></msub><mo>−</mo><mi>a</mi><mi>v</mi><mi>g</mi><mo stretchy="false">(</mo><mi>B</mi><mo stretchy="false">)</mo><mo stretchy="false">)</mo></mrow></mrow><mrow><msqrt><mrow><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><mrow><mo stretchy="false">(</mo><msub><mi>A</mi><mi>i</mi></msub><mo>−</mo><mi>a</mi><mi>v</mi><mi>g</mi><mo stretchy="false">(</mo><mi>A</mi><mo stretchy="false">)</mo><msup><mo stretchy="false">)</mo><mn>2</mn></msup></mrow></mrow></msqrt><mo>⋅</mo><msqrt><mrow><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><mrow><mo stretchy="false">(</mo><msub><mi>B</mi><mi>i</mi></msub><mo>−</mo><mi>a</mi><mi>v</mi><mi>g</mi><mo stretchy="false">(</mo><mi>B</mi><mo stretchy="false">)</mo><msup><mo stretchy="false">)</mo><mn>2</mn></msup></mrow></mrow></msqrt></mrow></mfrac></mrow><annotation encoding="application/x-tex">r_{AB} = \frac{\sum_{i=1}^n {(A_i - avg(A))(B_i - avg(B))}}{\sqrt{\sum_{i=1}^n {(A_i - avg(A))^2}} \cdot \sqrt{\sum_{i=1}^n {(B_i - avg(B))^2}}}
</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.58056em;vertical-align:-0.15em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">r</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.32833099999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">A</span><span class="mord mathnormal mtight" style="margin-right:0.05017em;">B</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:2.624002em;vertical-align:-1.13em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.494002em;"><span style="top:-2.1727090000000002em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord sqrt"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.937291em;"><span class="svg-align" style="top:-3.2em;"><span class="pstrut" style="height:3.2em;"></span><span class="mord" style="padding-left:1em;"><span class="mop"><span class="mop op-symbol small-op" style="position:relative;top:-0.0000050000000000050004em;">∑</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.804292em;"><span style="top:-2.40029em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">i</span><span class="mrel mtight">=</span><span class="mord mtight">1</span></span></span></span><span style="top:-3.2029em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">n</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.29971000000000003em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord"><span class="mopen">(</span><span class="mord"><span class="mord mathnormal">A</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord mathnormal">a</span><span class="mord mathnormal" style="margin-right:0.03588em;">v</span><span class="mord mathnormal" style="margin-right:0.03588em;">g</span><span class="mopen">(</span><span class="mord mathnormal">A</span><span class="mclose">)</span><span class="mclose"><span class="mclose">)</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.740108em;"><span style="top:-2.9890000000000003em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span></span></span></span><span style="top:-2.897291em;"><span class="pstrut" style="height:3.2em;"></span><span class="hide-tail" style="min-width:1.02em;height:1.28em;"><svg width="400em" height="1.28em" viewBox="0 0 400000 1296" preserveAspectRatio="xMinYMin slice"><path d="M263,681c0.7,0,18,39.7,52,119
c34,79.3,68.167,158.7,102.5,238c34.3,79.3,51.8,119.3,52.5,120
c340,-704.7,510.7,-1060.3,512,-1067
l0 -0
c4.7,-7.3,11,-11,19,-11
H40000v40H1012.3
s-271.3,567,-271.3,567c-38.7,80.7,-84,175,-136,283c-52,108,-89.167,185.3,-111.5,232
c-22.3,46.7,-33.8,70.3,-34.5,71c-4.7,4.7,-12.3,7,-23,7s-12,-1,-12,-1
s-109,-253,-109,-253c-72.7,-168,-109.3,-252,-110,-252c-10.7,8,-22,16.7,-34,26
c-22,17.3,-33.3,26,-34,26s-26,-26,-26,-26s76,-59,76,-59s76,-60,76,-60z
M1001 80h400000v40h-400000z"></path></svg></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.302709em;"><span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord sqrt"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.937291em;"><span class="svg-align" style="top:-3.2em;"><span class="pstrut" style="height:3.2em;"></span><span class="mord" style="padding-left:1em;"><span class="mop"><span class="mop op-symbol small-op" style="position:relative;top:-0.0000050000000000050004em;">∑</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.804292em;"><span style="top:-2.40029em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">i</span><span class="mrel mtight">=</span><span class="mord mtight">1</span></span></span></span><span style="top:-3.2029em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">n</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.29971000000000003em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord"><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.05017em;">B</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:-0.05017em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord mathnormal">a</span><span class="mord mathnormal" style="margin-right:0.03588em;">v</span><span class="mord mathnormal" style="margin-right:0.03588em;">g</span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.05017em;">B</span><span class="mclose">)</span><span class="mclose"><span class="mclose">)</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.740108em;"><span style="top:-2.9890000000000003em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span></span></span></span><span style="top:-2.897291em;"><span class="pstrut" style="height:3.2em;"></span><span class="hide-tail" style="min-width:1.02em;height:1.28em;"><svg width="400em" height="1.28em" viewBox="0 0 400000 1296" preserveAspectRatio="xMinYMin slice"><path d="M263,681c0.7,0,18,39.7,52,119
c34,79.3,68.167,158.7,102.5,238c34.3,79.3,51.8,119.3,52.5,120
c340,-704.7,510.7,-1060.3,512,-1067
l0 -0
c4.7,-7.3,11,-11,19,-11
H40000v40H1012.3
s-271.3,567,-271.3,567c-38.7,80.7,-84,175,-136,283c-52,108,-89.167,185.3,-111.5,232
c-22.3,46.7,-33.8,70.3,-34.5,71c-4.7,4.7,-12.3,7,-23,7s-12,-1,-12,-1
s-109,-253,-109,-253c-72.7,-168,-109.3,-252,-110,-252c-10.7,8,-22,16.7,-34,26
c-22,17.3,-33.3,26,-34,26s-26,-26,-26,-26s76,-59,76,-59s76,-60,76,-60z
M1001 80h400000v40h-400000z"></path></svg></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.302709em;"><span></span></span></span></span></span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.6897100000000003em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mop"><span class="mop op-symbol small-op" style="position:relative;top:-0.0000050000000000050004em;">∑</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.804292em;"><span style="top:-2.40029em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">i</span><span class="mrel mtight">=</span><span class="mord mtight">1</span></span></span></span><span style="top:-3.2029em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">n</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.29971000000000003em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord"><span class="mopen">(</span><span class="mord"><span class="mord mathnormal">A</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord mathnormal">a</span><span class="mord mathnormal" style="margin-right:0.03588em;">v</span><span class="mord mathnormal" style="margin-right:0.03588em;">g</span><span class="mopen">(</span><span class="mord mathnormal">A</span><span class="mclose">)</span><span class="mclose">)</span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.05017em;">B</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:-0.05017em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord mathnormal">a</span><span class="mord mathnormal" style="margin-right:0.03588em;">v</span><span class="mord mathnormal" style="margin-right:0.03588em;">g</span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.05017em;">B</span><span class="mclose">)</span><span class="mclose">)</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.13em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span></span></div><div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>Much like <a href="#BM_(cosine%20similarity%7Ccosine%20distance)%2Fi">cosine similarity</a>, I can't pinpoint what exactly it is that the formula is calculating / the reasoning behind the calculations. The only part I somewhat understand where it's getting the <a href="#BM_(euclidean%20distance)%2Fi">euclidean distance</a> to the average of each vector.</p>
<p>The rest of it I don't understand.</p>
</div>
<p><a href="ch8_code/src/metrics/PearsonSimilarity.py">ch8_code/src/metrics/PearsonSimilarity.py</a> (lines 10 to 28):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">pearson_similarity</span>(<span class="hljs-params">v: <span class="hljs-type">Sequence</span>[<span class="hljs-built_in">float</span>], w: <span class="hljs-type">Sequence</span>[<span class="hljs-built_in">float</span>], dims: <span class="hljs-built_in">int</span></span>):
    v_avg = mean(v)
    w_avg = mean(w)
    vec_avg_diffs_dp = <span class="hljs-built_in">sum</span>((v[i] - v_avg) * (w[i] - w_avg) <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(dims))
    dist_to_v_avg = sqrt(<span class="hljs-built_in">sum</span>((v[i] - v_avg) ** <span class="hljs-number">2</span> <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(dims)))
    dist_to_w_avg = sqrt(<span class="hljs-built_in">sum</span>((w[i] - w_avg) ** <span class="hljs-number">2</span> <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(dims)))
    <span class="hljs-keyword">return</span> vec_avg_diffs_dp / (dist_to_v_avg * dist_to_w_avg)


<span class="hljs-keyword">def</span> <span class="hljs-title function_">pearson_distance</span>(<span class="hljs-params">v: <span class="hljs-type">Sequence</span>[<span class="hljs-built_in">float</span>], w: <span class="hljs-type">Sequence</span>[<span class="hljs-built_in">float</span>], dims: <span class="hljs-built_in">int</span></span>):
    <span class="hljs-comment"># To turn pearson similarity into a distance metric, subtract 1.0 from it. By</span>
    <span class="hljs-comment"># subtracting 1.0, you're changing the bounds from [1.0, -1.0] to [0.0, 2.0].</span>
    <span class="hljs-comment">#</span>
    <span class="hljs-comment"># Recall that any distance metric must return 0 when the items being compared</span>
    <span class="hljs-comment"># are the same and increases the more different they get. By subtracting 1.0,</span>
    <span class="hljs-comment"># you're matching that distance metric requirement: 0.0 when totally similar</span>
    <span class="hljs-comment"># and 2.0 for totally dissimilar.</span>
    <span class="hljs-keyword">return</span> <span class="hljs-number">1.0</span> - pearson_similarity(v, w, dims)
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given the vectors ...</p>
<ul>
<li>[2, 10, 4, 12, 6, 14]</li>
<li>[5, 9, 6, 10, 7, 11]</li>
</ul>
<p>Their pearson similarity is 0.9999999999999999</p>
</div>
<p></p>
<a name="H_K-Centers%20Clustering"></a><h3>K-Centers Clustering</h3>
<p><a name="BM_(Algorithms%5C%2FGene%20Expression%2FK-Centers%20Clustering)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FGene%20Expression%2FEuclidean%20Distance%20Metric)_TOPIC%2F">Algorithms/Gene Expression/Euclidean Distance Metric</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: Given a list of n-dimensional points (vectors), choose a predefined number of points (k), called centers. Each center identifies a <a href="#BM_(cluster)%2Fi">cluster</a>, where the points closest to that center (<a href="#BM_(euclidean%20distance)%2Fi">euclidean distance</a>) are that <a href="#BM_(cluster)%2Fi">cluster</a>'s members. The goal is to choose centers such that, out of all possible <a href="#BM_(cluster)%2Fi">cluster</a> center to member distances, the farthest distance is the minimum it could possibly be out of all possible choices for centers.</p>
<p>In terms of a scoring function, the score being minimized is ...</p>
<div style="overflow: auto"><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><mi>s</mi><mi>c</mi><mi>o</mi><mi>r</mi><mi>e</mi><mo>=</mo><mi>m</mi><mi>a</mi><mi>x</mi><mo stretchy="false">(</mo><mi>d</mi><mo stretchy="false">(</mo><msub><mi>P</mi><mn>1</mn></msub><mo separator="true">,</mo><mi>C</mi><mo stretchy="false">)</mo><mo separator="true">,</mo><mi>d</mi><mo stretchy="false">(</mo><msub><mi>P</mi><mn>2</mn></msub><mo separator="true">,</mo><mi>C</mi><mo stretchy="false">)</mo><mo separator="true">,</mo><mi mathvariant="normal">.</mi><mi mathvariant="normal">.</mi><mi mathvariant="normal">.</mi><mo separator="true">,</mo><mi>d</mi><mo stretchy="false">(</mo><msub><mi>P</mi><mi>n</mi></msub><mo separator="true">,</mo><mi>C</mi><mo stretchy="false">)</mo><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">score = max(d(P_1, C), d(P_2, C), ..., d(P_n, C))
</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.43056em;vertical-align:0em;"></span><span class="mord mathnormal">s</span><span class="mord mathnormal">c</span><span class="mord mathnormal">o</span><span class="mord mathnormal" style="margin-right:0.02778em;">r</span><span class="mord mathnormal">e</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal">m</span><span class="mord mathnormal">a</span><span class="mord mathnormal">x</span><span class="mopen">(</span><span class="mord mathnormal">d</span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.13889em;">P</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.13889em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mpunct">,</span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord mathnormal" style="margin-right:0.07153em;">C</span><span class="mclose">)</span><span class="mpunct">,</span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord mathnormal">d</span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.13889em;">P</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.13889em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mpunct">,</span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord mathnormal" style="margin-right:0.07153em;">C</span><span class="mclose">)</span><span class="mpunct">,</span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord">.</span><span class="mord">.</span><span class="mord">.</span><span class="mpunct">,</span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord mathnormal">d</span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.13889em;">P</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.151392em;"><span style="top:-2.5500000000000003em;margin-left:-0.13889em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">n</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mpunct">,</span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord mathnormal" style="margin-right:0.07153em;">C</span><span class="mclose">)</span><span class="mclose">)</span></span></span></span></span></div><ul>
<li>P is the set of points.</li>
<li>C is the set of centers.</li>
<li>n is the number of points in P</li>
<li>d returns the <a href="#BM_(euclidean%20distance)%2Fi">euclidean distance</a> between a point and its closest center.</li>
<li>max is the maximum function.</li>
</ul>
<pre class="hljs"><code><span class="hljs-comment"># d() function from the formula</span>
<span class="hljs-keyword">def</span> <span class="hljs-title function_">dist_to_closest_center</span>(<span class="hljs-params">data_pt, center_pts</span>):
  center_pt = <span class="hljs-built_in">min</span>(
    center_pts,
    key=<span class="hljs-keyword">lambda</span> cp: dist(data_pt, cp)
  )
  <span class="hljs-keyword">return</span> dist(center_pt, data_pt)

<span class="hljs-comment"># scoring function (what's trying to be minimized)</span>
<span class="hljs-keyword">def</span> <span class="hljs-title function_">k_centers_score</span>(<span class="hljs-params">data_pts, center_pts</span>):
  <span class="hljs-keyword">return</span> <span class="hljs-built_in">max</span>(dist_to_closest_center(p, center_pts) <span class="hljs-keyword">for</span> p <span class="hljs-keyword">in</span> data_pts)
</code></pre>
<p><img src="svgbob_049e214f3ef4fbb258190aeac55bfbf167a4b32b.svg" alt="Kroki diagram output"></p>
<p><strong>WHY</strong>: This is one of the methods used for <a href="#BM_(cluster)%2Fi">cluster</a>ing <a href="#BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi">gene expression vector</a>s. Because it's limited to use <a href="#BM_(euclidean%20distance)%2Fi">euclidean distance</a> as the metric, it's essentially <a href="#BM_(cluster)%2Fi">cluster</a>ing by how close the component plots match up. For example, ...</p>
<table>
<thead>
<tr>
<th></th>
<th>hour0</th>
<th>hour1</th>
<th>hour2</th>
<th>hour3</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> A</td>
<td>2</td>
<td>10</td>
<td>2</td>
<td>10</td>
</tr>
<tr>
<td><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> B</td>
<td>2</td>
<td>8</td>
<td>2</td>
<td>8</td>
</tr>
<tr>
<td><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> C</td>
<td>2</td>
<td>2</td>
<td>2</td>
<td>10</td>
</tr>
</tbody>
</table>
<p><img src="svgbob_9e826f4a4ade56f78f3c26b7b256969a30621136.svg" alt="Kroki diagram output"></p>
<ul>
<li><code>dist((2,10,2,10), (2,8,2,8)) = 2.82</code></li>
<li><code>dist((2,10,2,10), (2,2,2,10)) = 8</code></li>
<li><code>dist((2,8,2,8), (2,2,2,10)) = 6.325</code></li>
</ul>
<p>In addition to only being able to use <a href="#BM_(euclidean%20distance)%2Fi">euclidean distance</a>, another limitation is that it requires knowing the number of <a href="#BM_(cluster)%2Fi">cluster</a>s (k) beforehand. Other <a href="#BM_(cluster)%2Fi">cluster</a>ing algorithms exist that don't have either restriction.</p>
<p><strong>ALGORITHM</strong>:</p>
<p>Solving <a href="#BM_(k-centers%3F%20clustering%7C%5Cd%2B-centers%3F%20clustering%7Ck-centers%3F%7C%5Cd%2B-centers%3F)%2Fi">k-centers</a> for any non-trivial input isn't possible because the search space is too huge. Because of this, heuristics are used. A common <a href="#BM_(k-centers%3F%20clustering%7C%5Cd%2B-centers%3F%20clustering%7Ck-centers%3F%7C%5Cd%2B-centers%3F)%2Fi">k-centers</a> heuristic is the <a href="#BM_(farthest%20first%20traversal)%2Fi">farthest first traversal</a> algorithm. The algorithm iteratively builds out more centers by inspecting the <a href="#BM_(euclidean%20distance)%2Fi">euclidean distance</a>s from points to existing centers. At each step, the algorithm ...</p>
<ol>
<li>gets the closest center for each point,</li>
<li>picks the point with the farthest <a href="#BM_(euclidean%20distance)%2Fi">euclidean distance</a> and sets it as the new center.</li>
</ol>
<p>The algorithm initially primes the list of centers with a randomly chosen point and stops executing once it has k points.</p>
<p><a href="ch8_code/src/clustering/KCenters_FarthestFirstTraversal.py">ch8_code/src/clustering/KCenters_FarthestFirstTraversal.py</a> (lines 119 to 169):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">find_closest_center</span>(<span class="hljs-params">data_pt, center_pts</span>):
    center_pt = <span class="hljs-built_in">min</span>(
        center_pts,
        key=<span class="hljs-keyword">lambda</span> cp: dist(data_pt, cp)
    )
    <span class="hljs-keyword">return</span> center_pt, dist(center_pt, data_pt)


<span class="hljs-keyword">def</span> <span class="hljs-title function_">centers_to_clusters</span>(<span class="hljs-params">
        centers: <span class="hljs-built_in">list</span>[<span class="hljs-type">Sequence</span>[<span class="hljs-built_in">float</span>]],
        vectors: <span class="hljs-built_in">list</span>[<span class="hljs-type">Sequence</span>[<span class="hljs-built_in">float</span>]]
</span>) -&gt; <span class="hljs-built_in">dict</span>[<span class="hljs-built_in">tuple</span>[<span class="hljs-built_in">float</span>], <span class="hljs-built_in">list</span>[<span class="hljs-type">Sequence</span>[<span class="hljs-built_in">float</span>]]]:
    mapping = {<span class="hljs-built_in">tuple</span>(ct_pt): [] <span class="hljs-keyword">for</span> ct_pt <span class="hljs-keyword">in</span> centers}
    <span class="hljs-keyword">for</span> pt <span class="hljs-keyword">in</span> vectors:
        ct_pt, _ = find_closest_center(pt, centers)
        ct_pt = <span class="hljs-built_in">tuple</span>(ct_pt)
        mapping[ct_pt].append(pt)
    <span class="hljs-keyword">return</span> mapping


<span class="hljs-keyword">def</span> <span class="hljs-title function_">k_centers_farthest_first_traversal</span>(<span class="hljs-params">
        k: <span class="hljs-built_in">int</span>,
        vectors: <span class="hljs-built_in">list</span>[<span class="hljs-type">Sequence</span>[<span class="hljs-built_in">float</span>]],
        dims: <span class="hljs-built_in">int</span>,
        iteration_callback: <span class="hljs-type">Callable</span>[  <span class="hljs-comment"># callback func to invoke on each iteration</span>
            [
                <span class="hljs-built_in">dict</span>[<span class="hljs-built_in">tuple</span>[<span class="hljs-built_in">float</span>], <span class="hljs-built_in">list</span>[<span class="hljs-type">Sequence</span>[<span class="hljs-built_in">float</span>]]]
            ],
            <span class="hljs-literal">None</span>
        ] | <span class="hljs-literal">None</span> = <span class="hljs-literal">None</span>
</span>) -&gt; <span class="hljs-built_in">dict</span>[<span class="hljs-built_in">tuple</span>[<span class="hljs-built_in">float</span>], <span class="hljs-built_in">list</span>[<span class="hljs-type">Sequence</span>[<span class="hljs-built_in">float</span>]]]:
    <span class="hljs-comment"># choose an initial center</span>
    centers = [random.choice(vectors)]
    <span class="hljs-comment"># notify of cluster for first iteration</span>
    mapping = centers_to_clusters(centers, vectors)
    iteration_callback(mapping)
    <span class="hljs-comment"># iterate</span>
    <span class="hljs-keyword">while</span> <span class="hljs-built_in">len</span>(centers) &lt; k:
        <span class="hljs-comment"># get next center</span>
        dists = {}
        <span class="hljs-keyword">for</span> pt <span class="hljs-keyword">in</span> vectors:
            _, d = find_closest_center(pt, centers)
            pt = <span class="hljs-built_in">tuple</span>(pt)
            dists[pt] = d
        farthest_closest_center_pt = <span class="hljs-built_in">max</span>(dists, key=<span class="hljs-keyword">lambda</span> x: dists[x])
        centers.append(farthest_closest_center_pt)
        <span class="hljs-comment"># notify of the current iteration's cluster</span>
        mapping = centers_to_clusters(centers, vectors)
        iteration_callback(mapping)
    <span class="hljs-keyword">return</span> mapping
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given k=3 and vectors=[[2, 2], [2, 4], [2.5, 6], [3.5, 2], [4, 3], [4, 5], [4.5, 4], [7, 2], [7.5, 3], [8, 1], [9, 2], [8, 7], [8.5, 8], [9, 6], [10, 7]]...</p>
<p>The farthest first travel heuristic produced the clusters at each iteration ...</p>
<ul>
<li>
<p>Iteration 0</p>
<ul>
<li>cluster center (8.5, 8)=[[2, 2], [2, 4], [2.5, 6], [3.5, 2], [4, 3], [4, 5], [4.5, 4], [7, 2], [7.5, 3], [8, 1], [9, 2], [8, 7], [8.5, 8], [9, 6], [10, 7]]</li>
</ul>
<p><img src="ch8_007dfb27f4db5c8496443fd990231612_plot0.svg" alt="k-centers 2D plot"></p>
</li>
<li>
<p>Iteration 1</p>
<ul>
<li>cluster center (8.5, 8)=[[7.5, 3], [9, 2], [8, 7], [8.5, 8], [9, 6], [10, 7]]</li>
<li>cluster center (2, 2)=[[2, 2], [2, 4], [2.5, 6], [3.5, 2], [4, 3], [4, 5], [4.5, 4], [7, 2], [8, 1]]</li>
</ul>
<p><img src="ch8_007dfb27f4db5c8496443fd990231612_plot1.svg" alt="k-centers 2D plot"></p>
</li>
<li>
<p>Iteration 2</p>
<ul>
<li>cluster center (8.5, 8)=[[8, 7], [8.5, 8], [9, 6], [10, 7]]</li>
<li>cluster center (2, 2)=[[2, 2], [2, 4], [2.5, 6], [3.5, 2], [4, 3], [4, 5], [4.5, 4]]</li>
<li>cluster center (8, 1)=[[7, 2], [7.5, 3], [8, 1], [9, 2]]</li>
</ul>
<p><img src="ch8_007dfb27f4db5c8496443fd990231612_plot2.svg" alt="k-centers 2D plot"></p>
</li>
</ul>
</div>
<p></p>
<p>One problem that should be noted with this heuristic is that, when outliers are present, it'll likely place those outliers into their own <a href="#BM_(cluster)%2Fi">cluster</a>.</p>
<p><img src="svgbob_339254b3bba7afdccd5bbaa3b7bf6a4db8926346.svg" alt="Kroki diagram output"></p>
<a name="H_K-Means%20Clustering"></a><h3>K-Means Clustering</h3>
<p><a name="BM_(Algorithms%5C%2FGene%20Expression%2FK-Means%20Clustering)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FGene%20Expression%2FEuclidean%20Distance%20Metric)_TOPIC%2F">Algorithms/Gene Expression/Euclidean Distance Metric</a></li>
<li><a href="#BM_(Algorithms%5C%2FGene%20Expression%2FK-Centers%20Clustering)_TOPIC%2F">Algorithms/Gene Expression/K-Centers Clustering</a></li>
</ul>
</div>
<p><strong>WHAT</strong>: <a href="#BM_(k-means%3F%7C%5Cd%2B-means%3F%7Csquared%20error%20distortion)%2Fi">K-means</a> is <a href="#BM_(k-centers%3F%20clustering%7C%5Cd%2B-centers%3F%20clustering%7Ck-centers%3F%7C%5Cd%2B-centers%3F)%2Fi">k-centers</a> except the scoring function is different. Recall that the scoring function for <a href="#BM_(k-centers%3F%20clustering%7C%5Cd%2B-centers%3F%20clustering%7Ck-centers%3F%7C%5Cd%2B-centers%3F)%2Fi">k-centers</a> is ...</p>
<div style="overflow: auto"><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><mi>s</mi><mi>c</mi><mi>o</mi><mi>r</mi><mi>e</mi><mo>=</mo><mi>m</mi><mi>a</mi><mi>x</mi><mo stretchy="false">(</mo><mi>d</mi><mo stretchy="false">(</mo><msub><mi>P</mi><mn>1</mn></msub><mo separator="true">,</mo><mi>C</mi><mo stretchy="false">)</mo><mo separator="true">,</mo><mi>d</mi><mo stretchy="false">(</mo><msub><mi>P</mi><mn>2</mn></msub><mo separator="true">,</mo><mi>C</mi><mo stretchy="false">)</mo><mo separator="true">,</mo><mi mathvariant="normal">.</mi><mi mathvariant="normal">.</mi><mi mathvariant="normal">.</mi><mo separator="true">,</mo><mi>d</mi><mo stretchy="false">(</mo><msub><mi>P</mi><mi>n</mi></msub><mo separator="true">,</mo><mi>C</mi><mo stretchy="false">)</mo><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">score = max(d(P_1, C), d(P_2, C), ..., d(P_n, C))
</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.43056em;vertical-align:0em;"></span><span class="mord mathnormal">s</span><span class="mord mathnormal">c</span><span class="mord mathnormal">o</span><span class="mord mathnormal" style="margin-right:0.02778em;">r</span><span class="mord mathnormal">e</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal">m</span><span class="mord mathnormal">a</span><span class="mord mathnormal">x</span><span class="mopen">(</span><span class="mord mathnormal">d</span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.13889em;">P</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.13889em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mpunct">,</span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord mathnormal" style="margin-right:0.07153em;">C</span><span class="mclose">)</span><span class="mpunct">,</span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord mathnormal">d</span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.13889em;">P</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.13889em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mpunct">,</span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord mathnormal" style="margin-right:0.07153em;">C</span><span class="mclose">)</span><span class="mpunct">,</span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord">.</span><span class="mord">.</span><span class="mord">.</span><span class="mpunct">,</span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord mathnormal">d</span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.13889em;">P</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.151392em;"><span style="top:-2.5500000000000003em;margin-left:-0.13889em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">n</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mpunct">,</span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord mathnormal" style="margin-right:0.07153em;">C</span><span class="mclose">)</span><span class="mclose">)</span></span></span></span></span></div><p>... where ...</p>
<ul>
<li>P is the set of points.</li>
<li>C is the set of centers.</li>
<li>n is the number of points in P</li>
<li>d returns the <a href="#BM_(euclidean%20distance)%2Fi">euclidean distance</a> between a point and its closest center.</li>
<li>max is the maximum function.</li>
</ul>
<p>The scoring function for <a href="#BM_(k-means%3F%7C%5Cd%2B-means%3F%7Csquared%20error%20distortion)%2Fi">k-means</a>, called <a href="#BM_(k-means%3F%7C%5Cd%2B-means%3F%7Csquared%20error%20distortion)%2Fi">squared error distortion</a>, is as follows ...</p>
<div style="overflow: auto"><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><mi>s</mi><mi>c</mi><mi>o</mi><mi>r</mi><mi>e</mi><mo>=</mo><mfrac><mrow><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><mrow><mi>d</mi><mo stretchy="false">(</mo><msub><mi>P</mi><mi>i</mi></msub><mo separator="true">,</mo><mi>C</mi><msup><mo stretchy="false">)</mo><mn>2</mn></msup></mrow></mrow><mi>n</mi></mfrac></mrow><annotation encoding="application/x-tex">score = \frac{\sum_{i=1}^{n} {d(P_i, C)^2}}{n}
</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.43056em;vertical-align:0em;"></span><span class="mord mathnormal">s</span><span class="mord mathnormal">c</span><span class="mord mathnormal">o</span><span class="mord mathnormal" style="margin-right:0.02778em;">r</span><span class="mord mathnormal">e</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:2.189818em;vertical-align:-0.686em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.5038179999999999em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord mathnormal">n</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.6897100000000003em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mop"><span class="mop op-symbol small-op" style="position:relative;top:-0.0000050000000000050004em;">∑</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.804292em;"><span style="top:-2.40029em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">i</span><span class="mrel mtight">=</span><span class="mord mtight">1</span></span></span></span><span style="top:-3.2029em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">n</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.29971000000000003em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord"><span class="mord mathnormal">d</span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.13889em;">P</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:-0.13889em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mpunct">,</span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord mathnormal" style="margin-right:0.07153em;">C</span><span class="mclose"><span class="mclose">)</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8141079999999999em;"><span style="top:-3.063em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span></span></div><div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The formula is taking the squares of d() and averaging them.</p>
</div>
<pre class="hljs"><code><span class="hljs-comment"># d() function from the formula</span>
<span class="hljs-keyword">def</span> <span class="hljs-title function_">dist_to_closest_center</span>(<span class="hljs-params">data_pt, center_pts</span>):
  center_pt = <span class="hljs-built_in">min</span>(
    center_pts,
    key=<span class="hljs-keyword">lambda</span> cp: dist(data_pt, cp)
  )
  <span class="hljs-keyword">return</span> dist(center_pt, data_pt)

<span class="hljs-comment"># scoring function (what's trying to be minimized)</span>
<span class="hljs-keyword">def</span> <span class="hljs-title function_">k_means_score</span>(<span class="hljs-params">data_pts, center_pts</span>):
  res = []
  <span class="hljs-keyword">for</span> data_pt <span class="hljs-keyword">in</span> data_pts:
    dist_to = dist_to_closest_center(data_pt, center_pts)
    res.append(dist_to ** <span class="hljs-number">2</span>)
  <span class="hljs-keyword">return</span> <span class="hljs-built_in">sum</span>(res) / <span class="hljs-built_in">len</span>(res)
</code></pre>
<p>Compared to <a href="#BM_(k-centers%3F%20clustering%7C%5Cd%2B-centers%3F%20clustering%7Ck-centers%3F%7C%5Cd%2B-centers%3F)%2Fi">k-centers</a>, <a href="#BM_(cluster)%2Fi">cluster</a> membership is still decided by the distance to its closest <a href="#BM_(cluster)%2Fi">cluster</a> (d in the formula above). It's the placement of centers that's different.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>There's a version of <a href="#BM_(k-centers%3F%20clustering%7C%5Cd%2B-centers%3F%20clustering%7Ck-centers%3F%7C%5Cd%2B-centers%3F)%2Fi">k-centers</a> / <a href="#BM_(k-means%3F%7C%5Cd%2B-means%3F%7Csquared%20error%20distortion)%2Fi">k-means</a> for <a href="#BM_(similarity%20metric)%2Fi">similarity metric</a>s / <a href="#BM_(distance%20metric)%2Fi">distance metric</a>s other than <a href="#BM_(euclidean%20distance)%2Fi">euclidean distance</a>. It's called k-medods but I haven't have a chance to look at it yet and it wasn't covered by the book.</p>
<ul>
<li><a href="https://stats.stackexchange.com/a/81496">link 1</a></li>
<li><a href="https://stats.stackexchange.com/a/32942">link 2</a></li>
</ul>
</div>
<p><strong>WHY</strong>: <a href="#BM_(k-means%3F%7C%5Cd%2B-means%3F%7Csquared%20error%20distortion)%2Fi">K-means</a> is more resilient to outliers than <a href="#BM_(k-centers%3F%20clustering%7C%5Cd%2B-centers%3F%20clustering%7Ck-centers%3F%7C%5Cd%2B-centers%3F)%2Fi">k-centers</a>. For example, consider finding a single center (k=1) for the following 1-D points: [0, 0.5, 1, 1.5, 10]. The last point (10) is an outlier. Without that outlier, <a href="#BM_(k-centers%3F%20clustering%7C%5Cd%2B-centers%3F%20clustering%7Ck-centers%3F%7C%5Cd%2B-centers%3F)%2Fi">k-centers</a> has a center of 0.75 ...</p>
<p><img src="svgbob_5d89462992602e4a4d116f317249afe9b4f5f378.svg" alt="Kroki diagram output"></p>
<p>With that outlier, the <a href="#BM_(k-centers%3F%20clustering%7C%5Cd%2B-centers%3F%20clustering%7Ck-centers%3F%7C%5Cd%2B-centers%3F)%2Fi">k-centers</a> has a center of 5, which is a drastic shift from the original 0.75 shown above ...</p>
<p><img src="svgbob_bcb7698e47f537fb89d02916b019dc04cad01ece.svg" alt="Kroki diagram output"></p>
<p><a href="#BM_(k-means%3F%7C%5Cd%2B-means%3F%7Csquared%20error%20distortion)%2Fi">K-means</a> combats this shift by applying weighting: The idea is that the 4 real points should have a stronger influence on the center than the one outlier point, essentially dragging it back to towards them. Using <a href="#BM_(k-means%3F%7C%5Cd%2B-means%3F%7Csquared%20error%20distortion)%2Fi">k-means</a>, the center is 2.6 ...</p>
<p><img src="svgbob_09d7cd1a38aa2c3314c87c3b98adf995282ac311.svg" alt="Kroki diagram output"></p>
<p>Note that the scoring functions for <a href="#BM_(k-means%3F%7C%5Cd%2B-means%3F%7Csquared%20error%20distortion)%2Fi">k-means</a> and <a href="#BM_(k-centers%3F%20clustering%7C%5Cd%2B-centers%3F%20clustering%7Ck-centers%3F%7C%5Cd%2B-centers%3F)%2Fi">k-centers</a> produce vastly different scores, but the scores themselves don't matter. What matters is the minimization of the score. The diagram below shows the scores for both <a href="#BM_(k-means%3F%7C%5Cd%2B-means%3F%7Csquared%20error%20distortion)%2Fi">k-means</a> and <a href="#BM_(k-centers%3F%20clustering%7C%5Cd%2B-centers%3F%20clustering%7Ck-centers%3F%7C%5Cd%2B-centers%3F)%2Fi">k-centers</a> as the center shifts from 10 to 0 ...</p>
<p><img src="svgbob_26fcdb256e94f273fe5fceadb585adb25da2cb6c.svg" alt="Kroki diagram output"></p>
<p><strong>ALGORITHM</strong>:</p>
<p>Similar to <a href="#BM_(k-centers%3F%20clustering%7C%5Cd%2B-centers%3F%20clustering%7Ck-centers%3F%7C%5Cd%2B-centers%3F)%2Fi">k-centers</a>, solving <a href="#BM_(k-means%3F%7C%5Cd%2B-means%3F%7Csquared%20error%20distortion)%2Fi">k-means</a> for any non-trivial input isn't possible because the search space is too huge. Because of this, heuristics are used. A common <a href="#BM_(k-means%3F%7C%5Cd%2B-means%3F%7Csquared%20error%20distortion)%2Fi">k-means</a> heuristic is <a href="#BM_(lloyd%20algorithm%7Clloyd's%20algorithm%7Ck-means%3F%5C%2B%5C%2B%20initializer%7Ck-means%3F%5C%2B%5C%2B)%2Fi">Lloyd's algorithm</a>. The algorithm randomly picks k points to set as the centers and iteratively refines those centers. At each step, the algorithm ...</p>
<ol>
<li>
<p>converts centers to <a href="#BM_(cluster)%2Fi">cluster</a>s,</p>
<p>The point closest to a center becomes a member of that <a href="#BM_(cluster)%2Fi">cluster</a>. Ties are broken arbitrarily.</p>
</li>
<li>
<p>converts <a href="#BM_(cluster)%2Fi">cluster</a>s to centers.</p>
<p>The <a href="#BM_(cluster)%2Fi">cluster</a>s from step 1 are turned into new centers. Each component of a center becomes the average of that dimension across <a href="#BM_(cluster)%2Fi">cluster</a> members, referred to as the <a href="#BM_(center%20of%20gravity)%2Fi">center of gravity</a>.</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">center_of_gravity</span>(<span class="hljs-params">data_pts, dim</span>):
    center = []
    <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(dim):
        val = mean(pt[i] <span class="hljs-keyword">for</span> pt <span class="hljs-keyword">in</span> data_pts)
        center.append(val)
    <span class="hljs-keyword">return</span> center
</code></pre>
</li>
</ol>
<p>The algorithm will converge to stable centers, at which point it stops iterating.</p>
<p><a href="ch8_code/src/clustering/KMeans_Lloyds.py">ch8_code/src/clustering/KMeans_Lloyds.py</a> (lines 138 to 169):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">k_means_lloyds</span>(<span class="hljs-params">
        k: <span class="hljs-built_in">int</span>,
        vectors: <span class="hljs-built_in">list</span>[<span class="hljs-type">Sequence</span>[<span class="hljs-built_in">float</span>]],
        centers_init: <span class="hljs-built_in">list</span>[<span class="hljs-type">Sequence</span>[<span class="hljs-built_in">float</span>]],
        dims: <span class="hljs-built_in">int</span>,
        iteration_callback: <span class="hljs-type">Callable</span>[  <span class="hljs-comment"># callback func to invoke on each iteration</span>
            [
                <span class="hljs-built_in">dict</span>[<span class="hljs-built_in">tuple</span>[<span class="hljs-built_in">float</span>], <span class="hljs-built_in">list</span>[<span class="hljs-type">Sequence</span>[<span class="hljs-built_in">float</span>]]]
            ],
            <span class="hljs-literal">None</span>
        ] | <span class="hljs-literal">None</span> = <span class="hljs-literal">None</span>
</span>) -&gt; <span class="hljs-built_in">dict</span>[<span class="hljs-built_in">tuple</span>[<span class="hljs-built_in">float</span>], <span class="hljs-built_in">list</span>[<span class="hljs-type">Sequence</span>[<span class="hljs-built_in">float</span>]]]:
    old_centers = []
    centers = centers_init[:]
    <span class="hljs-keyword">while</span> centers != old_centers:
        mapping = {<span class="hljs-built_in">tuple</span>(ct_pt): [] <span class="hljs-keyword">for</span> ct_pt <span class="hljs-keyword">in</span> centers}
        <span class="hljs-comment"># centers to clusters</span>
        <span class="hljs-keyword">for</span> pt <span class="hljs-keyword">in</span> vectors:
            ct_pt, _ = find_closest_center(pt, centers)
            ct_pt = <span class="hljs-built_in">tuple</span>(ct_pt)
            mapping[ct_pt].append(pt)
        <span class="hljs-comment"># clusters to centers</span>
        old_centers = centers
        centers = []
        <span class="hljs-keyword">for</span> pts <span class="hljs-keyword">in</span> mapping.values():
            new_ct_pt = center_of_gravity(pts, dims)
            new_ct_pt = <span class="hljs-built_in">tuple</span>(new_ct_pt)
            centers.append(new_ct_pt)
        <span class="hljs-comment"># notify of current iteration's cluster</span>
        iteration_callback(mapping)
    <span class="hljs-keyword">return</span> mapping
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given k=3 and vectors=[[2, 2], [2, 4], [2.5, 6], [3.5, 2], [4, 3], [4, 5], [4.5, 4], [17, 12], [17.5, 13], [18, 11], [19, 12], [18, -7], [18.5, -8], [19, -6], [20, -7]]...</p>
<p>The llyod's algorithm heuristic produced the clusters at each iteration ...</p>
<ul>
<li>
<p>Iteration 0</p>
<ul>
<li>cluster center (2, 2)=[[2, 2], [3.5, 2], [4, 3], [18, -7], [18.5, -8], [19, -6], [20, -7]]</li>
<li>cluster center (2, 4)=[[2, 4], [4.5, 4]]</li>
<li>cluster center (2.5, 6)=[[2.5, 6], [4, 5], [17, 12], [17.5, 13], [18, 11], [19, 12]]</li>
</ul>
<p><img src="ch8_722c596c7a5f2cad02f1f095ebc56d59_plot0.svg" alt="k-centers 2D plot"></p>
</li>
<li>
<p>Iteration 1</p>
<ul>
<li>cluster center (12.142857142857142, -3)=[[18, -7], [18.5, -8], [19, -6], [20, -7]]</li>
<li>cluster center (3.25, 4)=[[2, 2], [2, 4], [2.5, 6], [3.5, 2], [4, 3], [4, 5], [4.5, 4]]</li>
<li>cluster center (13.0, 9.833333333333334)=[[17, 12], [17.5, 13], [18, 11], [19, 12]]</li>
</ul>
<p><img src="ch8_722c596c7a5f2cad02f1f095ebc56d59_plot1.svg" alt="k-centers 2D plot"></p>
</li>
<li>
<p>Iteration 2</p>
<ul>
<li>cluster center (18.875, -7)=[[18, -7], [18.5, -8], [19, -6], [20, -7]]</li>
<li>cluster center (3.2142857142857144, 3.7142857142857144)=[[2, 2], [2, 4], [2.5, 6], [3.5, 2], [4, 3], [4, 5], [4.5, 4]]</li>
<li>cluster center (17.875, 12)=[[17, 12], [17.5, 13], [18, 11], [19, 12]]</li>
</ul>
<p><img src="ch8_722c596c7a5f2cad02f1f095ebc56d59_plot2.svg" alt="k-centers 2D plot"></p>
</li>
</ul>
</div>
<p></p>
<p>At each iteration, the <a href="#BM_(cluster)%2Fi">cluster</a> members captured (step 1) will drag the new center towards them (step 2). After so many iterations, each center will be at a point where further iterations won't capture a different set of members, meaning that the centers will stay where they're at (converged).</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>I said "ties are broken arbitrarily" (step 1) because that's what the Pevzner book says. This isn't entirely true? I think it's possible to get into a situation where a tied point ping-pongs back and forth between <a href="#BM_(cluster)%2Fi">cluster</a>s. So, maybe what actually needs to happen is you need to break ties consistently -- it doesn't matter how, just that its consistent (e.g. the center closest to origin + smallest angle from origin always wins the tied member).</p>
<p>Also, if the centers haven't converged, the dragged center is guaranteed to decrease the <a href="#BM_(k-means%3F%7C%5Cd%2B-means%3F%7Csquared%20error%20distortion)%2Fi">squared error distortion</a> when compared to the previous center. But, does that mean that a set of converged centers are optimal in terms of <a href="#BM_(k-means%3F%7C%5Cd%2B-means%3F%7Csquared%20error%20distortion)%2Fi">squared error distortion</a>? I don't think so. Even if a <a href="#BM_(cluster)%2Fi">cluster</a> converged to all the correct members, could it be that the center can be slightly tweaked to get the <a href="#BM_(k-means%3F%7C%5Cd%2B-means%3F%7Csquared%20error%20distortion)%2Fi">squared error distortion</a> down even more? Probably.</p>
</div>
<p>The hope with the heuristic is that, at each iteration, enough true <a href="#BM_(cluster)%2Fi">cluster</a> members are captured (step 1) to drag the new center (step 2) closer to the true center. One way to increase the odds that this heuristic converges on a good solution is the initial center selection: You can increase the chance of converging to a good solution by <em>probabilistically</em> selecting initial centers that are far from each other, referred to as <a href="#BM_(lloyd%20algorithm%7Clloyd's%20algorithm%7Ck-means%3F%5C%2B%5C%2B%20initializer%7Ck-means%3F%5C%2B%5C%2B)%2Fi">k-means++ initializer</a>.</p>
<ol>
<li>The 1st center is chosen from the list of points at random.</li>
<li>The 2nd center is chosen by selecting a point that's likely to be much farther away from center 1 than most other points.</li>
<li>The 3rd center is chosen by selecting a point that's likely to be much farther away from center 1 and 2 than most other points.</li>
<li>...</li>
</ol>
<p>The probability of selecting a point as the next center is proportional to its squared distance to the existing centers.</p>
<p><a href="ch8_code/src/clustering/KMeans_Lloyds.py">ch8_code/src/clustering/KMeans_Lloyds.py</a> (lines 246 to 264):</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">k_means_PP_initializer</span>(<span class="hljs-params">
        k: <span class="hljs-built_in">int</span>,
        vectors: <span class="hljs-built_in">list</span>[<span class="hljs-type">Sequence</span>[<span class="hljs-built_in">float</span>]],
</span>):
    centers = [random.choice(vectors)]
    <span class="hljs-keyword">while</span> <span class="hljs-built_in">len</span>(centers) &lt; k:
        choice_points = []
        choice_weights = []
        <span class="hljs-keyword">for</span> v <span class="hljs-keyword">in</span> vectors:
            <span class="hljs-keyword">if</span> v <span class="hljs-keyword">in</span> centers:
                <span class="hljs-keyword">continue</span>
            _, d = find_closest_center(v, centers)
            choice_weights.append(d)
            choice_points.append(v)
        total = <span class="hljs-built_in">sum</span>(choice_weights)
        choice_weights = [w / total <span class="hljs-keyword">for</span> w <span class="hljs-keyword">in</span> choice_weights]
        c_pt = random.choices(choice_points, weights=choice_weights, k=<span class="hljs-number">1</span>).pop(<span class="hljs-number">0</span>)
        centers.append(c_pt)
    <span class="hljs-keyword">return</span> centers
</code></pre>
<div style="border:1px solid black;">
<p></p>
<p>Given k=3 and vectors=[[2, 2], [2, 4], [2.5, 6], [3.5, 2], [4, 3], [4, 5], [4.5, 4], [17, 12], [17.5, 13], [18, 11], [19, 12], [18, -7], [18.5, -8], [19, -6], [20, -7]]...</p>
<p>The llyod's algorithm heuristic produced the clusters at each iteration ...</p>
<ul>
<li>
<p>Iteration 0</p>
<ul>
<li>cluster center (4.5, 4)=[[2.5, 6], [4, 3], [4, 5], [4.5, 4], [18, -7], [18.5, -8], [19, -6], [20, -7]]</li>
<li>cluster center (19, 12)=[[17, 12], [17.5, 13], [18, 11], [19, 12]]</li>
<li>cluster center (2, 2)=[[2, 2], [2, 4], [3.5, 2]]</li>
</ul>
<p><img src="ch8_388fae171efe60a95e37b946e7ae343d_plot0.svg" alt="k-centers 2D plot"></p>
</li>
<li>
<p>Iteration 1</p>
<ul>
<li>cluster center (11.3125, -1.25)=[[18, -7], [18.5, -8], [19, -6], [20, -7]]</li>
<li>cluster center (17.875, 12)=[[17, 12], [17.5, 13], [18, 11], [19, 12]]</li>
<li>cluster center (2.5, 2.6666666666666665)=[[2, 2], [2, 4], [2.5, 6], [3.5, 2], [4, 3], [4, 5], [4.5, 4]]</li>
</ul>
<p><img src="ch8_388fae171efe60a95e37b946e7ae343d_plot1.svg" alt="k-centers 2D plot"></p>
</li>
<li>
<p>Iteration 2</p>
<ul>
<li>cluster center (18.875, -7)=[[18, -7], [18.5, -8], [19, -6], [20, -7]]</li>
<li>cluster center (17.875, 12)=[[17, 12], [17.5, 13], [18, 11], [19, 12]]</li>
<li>cluster center (3.2142857142857144, 3.7142857142857144)=[[2, 2], [2, 4], [2.5, 6], [3.5, 2], [4, 3], [4, 5], [4.5, 4]]</li>
</ul>
<p><img src="ch8_388fae171efe60a95e37b946e7ae343d_plot2.svg" alt="k-centers 2D plot"></p>
</li>
</ul>
</div>
<p></p>
<p>Even with <a href="#BM_(lloyd%20algorithm%7Clloyd's%20algorithm%7Ck-means%3F%5C%2B%5C%2B%20initializer%7Ck-means%3F%5C%2B%5C%2B)%2Fi">k-means++ initializer</a>, <a href="#BM_(lloyd%20algorithm%7Clloyd's%20algorithm%7Ck-means%3F%5C%2B%5C%2B%20initializer%7Ck-means%3F%5C%2B%5C%2B)%2Fi">Lloyd's algorithm</a> isn't guaranteed to always converge to a good solution. The typically workflow is to run it multiple times, where the run producing centers with the lowest <a href="#BM_(k-means%3F%7C%5Cd%2B-means%3F%7Csquared%20error%20distortion)%2Fi">squared error distortion</a> is the one accepted.</p>
<p>Futhermore, <a href="#BM_(lloyd%20algorithm%7Clloyd's%20algorithm%7Ck-means%3F%5C%2B%5C%2B%20initializer%7Ck-means%3F%5C%2B%5C%2B)%2Fi">Lloyd's algorithm</a> may fail to converge to a good solution when the <a href="#BM_(cluster)%2Fi">cluster</a>s aren't globular and / or aren't of similar densities. Below are example <a href="#BM_(cluster)%2Fi">cluster</a>s that are obvious to a human but problematic for the algorithm.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The Pevzner book explicitly calls out <a href="#BM_(lloyd%20algorithm%7Clloyd's%20algorithm%7Ck-means%3F%5C%2B%5C%2B%20initializer%7Ck-means%3F%5C%2B%5C%2B)%2Fi">Lloyd's algorithm</a> for this, but I'm thinking this is more to do with the scoring function for <a href="#BM_(k-means%3F%7C%5Cd%2B-means%3F%7Csquared%20error%20distortion)%2Fi">k-means</a> (what's trying to be minimized)? I think the same problem applies to the scoring function for <a href="#BM_(k-centers%3F%20clustering%7C%5Cd%2B-centers%3F%20clustering%7Ck-centers%3F%7C%5Cd%2B-centers%3F)%2Fi">k-centers</a> and the furthest first traveled heuristic?</p>
<p>The examples below are taken directly from the Pevzner book.</p>
</div>
<p><img src="svgbob_5a95bee7c208bcd365a2c0338c8cf2d7900cdce7.svg" alt="Kroki diagram output"></p>
<a name="H_Soft%20K-Means%20Clustering"></a><h3>Soft K-Means Clustering</h3>
<p><a name="BM_(Algorithms%5C%2FGene%20Expression%2FSoft%20K-Means%20Clustering)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FGene%20Expression%2FK-Means%20Clustering)_TOPIC%2F">Algorithms/Gene Expression/K-Means Clustering</a></li>
</ul>
</div>
<p><strong>WHAT</strong>:</p>
<p><strong>WHY</strong>:</p>
<p><strong>HOW</strong>:</p>
<a name="H_Hierarchial%20Clustering"></a><h3>Hierarchial Clustering</h3>
<a name="H_Soft%20Hierarchial%20Clustering"></a><h3>Soft Hierarchial Clustering</h3>
<p>THIS IS NOT FROM THE BOOK, BUT I REASONED ABOUT THIS MYSELF. IMPLEMENT IT AND WRITE ABOUT IT HERE.)</p>
<p>Soft <a href="#BM_(hierarchial%20cluster)%2Fi">hierarchial cluster</a>ing - Build out a <a href="#BM_(neighbour%20joining%20phylogeny)%2Fi">neighbour joining phylogeny</a> <a href="#BM_(tree)%2Fi">tree</a>. Each internal node is a <a href="#BM_(cluster)%2Fi">cluster</a>. The distance between that internal node to all leaf nodes can be used to define the probability that the leaf node belongs to that <a href="#BM_(cluster)%2Fi">cluster</a>? This makes sense because <a href="#BM_(neighbour%20joining%20phylogeny)%2Fi">neighbour joining phylogeny</a> produces <a href="#BM_(un-rooted%20tree%7Cunrooted%20tree)%2Fi">unrooted tree</a>s (<a href="#BM_(simple%20tree)%2Fi">simple tree</a>s). If it were a <a href="#BM_(rooted%20tree)%2Fi">rooted tree</a>, you could say that internal node X leaf nodes A, B, and C -- meaning that A, B and C are members of <a href="#BM_(cluster)%2Fi">cluster</a> X. But, because it's unrooted, technically any leaf node in the graph could be a member of <a href="#BM_(cluster)%2Fi">cluster</a> X.</p>
<p>This seems like it'd be very useful. It's easy to understand (if you understand <a href="#BM_(neighbour%20joining%20phylogeny)%2Fi">neighbour joining phylogeny</a> / <a href="#BM_(additive%20phylogeny)%2Fi">additive phylogeny</a>)</p>
<a name="H_CAST%20Clustering"></a><h3>CAST Clustering</h3>
<a name="H_Stories"></a><h1>Stories</h1>
<a name="H_Bacterial%20Genome%20Replication"></a><h2>Bacterial Genome Replication</h2>
<p><a name="BM_(Stories%5C%2FBacterial%20Genome%20Replication)_TOPIC%2F"></a><strong></strong></p>
<p>Bacteria are known to have a single chromosome of circular / looping DNA. On that DNA, the <a href="#BM_(replication%20origin)%2Fi">replication origin</a> (<a href="#BM_%5Cb(ori)%5Cb%2Fi">ori</a>) is the region in which DNA replication starts, while the <a href="#BM_(replication%20terminus)%2Fi">replication terminus</a> (<a href="#BM_%5Cb(ter)%5Cb%2Fi">ter</a>) is where it ends. The <a href="#BM_%5Cb(ori)%5Cb%2Fi">ori</a> and <a href="#BM_%5Cb(ter)%5Cb%2Fi">ter</a> and usually placed on opposite ends of each other.</p>
<p><img src="svgbob_308a4af4e5d6864abeb06f6652581aec0ff50fb1.svg" alt="Kroki diagram output"></p>
<p>The replication process begins by a <a href="#BM_(replication%20fork)%2Fi">replication fork</a> opening at the <a href="#BM_%5Cb(ori)%5Cb%2Fi">ori</a>. As replication happens, that fork widens until the point it reaches <a href="#BM_%5Cb(ter)%5Cb%2Fi">ter</a>...</p>
<p><img src="svgbob_fc7cf2c734fa0cab15b5746e25ad07e25f85ec64.svg" alt="Kroki diagram output"></p>
<p>For each forked <a href="#BM_(single%20stranded%20DNA%7Csingle-stranded%20DNA)%2Fi">single-stranded DNA</a>, <a href="#BM_(DNA%20polymerase)%2Fi">DNA polymerase</a>s attach on and synthesize a new <a href="#BM_(reverse%20complement)%2Fi">reverse complement</a> strand so that it turns back into <a href="#BM_(double%20stranded%20DNA%7Cdouble-stranded%20DNA)%2Fi">double-stranded DNA</a>....</p>
<p><img src="svgbob_971abfeff166ada9787cd484db9bfbe43120973a.svg" alt="Kroki diagram output"></p>
<p>The process of synthesizing a <a href="#BM_(reverse%20complement)%2Fi">reverse complement</a> strand is different based on the section of DNA that <a href="#BM_(DNA%20polymerase)%2Fi">DNA polymerase</a> is operating on. For each <a href="#BM_(single%20stranded%20DNA%7Csingle-stranded%20DNA)%2Fi">single-stranded DNA</a>, if the direction of that DNA strand is traveling from ...</p>
<ul>
<li><a href="#BM_%5Cb(ori)%5Cb%2Fi">ori</a> to <a href="#BM_%5Cb(ter)%5Cb%2Fi">ter</a>, it's called a <a href="#BM_(forward%20half-strand)%2Fi">forward half-strand</a>.</li>
<li><a href="#BM_%5Cb(ter)%5Cb%2Fi">ter</a> to <a href="#BM_%5Cb(ori)%5Cb%2Fi">ori</a>, it's called a <a href="#BM_(reverse%20half-strand%7Cbackward%20half-strand%7Cbackwards%20half-strand)%2Fi">reverse half-strand</a>.</li>
</ul>
<p><img src="svgbob_ad11e0114c63955ab620cbe9a8dee61a85caa1f9.svg" alt="Kroki diagram output"></p>
<p>Since <a href="#BM_(DNA%20polymerase)%2Fi">DNA polymerase</a> can only walk over DNA in the reverse direction (<a href="#BM_(3')%2Fi">3'</a> to <a href="#BM_(5')%2Fi">5'</a>), the 2 <a href="#BM_(reverse%20half-strand%7Cbackward%20half-strand%7Cbackwards%20half-strand)%2Fi">reverse half-strand</a>s will quickly get walked over in one shot. A <a href="#BM_(primer)%2Fi">primer</a> gets attached to the <a href="#BM_%5Cb(ori)%5Cb%2Fi">ori</a>, then a <a href="#BM_(DNA%20polymerase)%2Fi">DNA polymerase</a> attaches to that <a href="#BM_(primer)%2Fi">primer</a> to begin synthesis of a new strand. Synthesis continues until the <a href="#BM_%5Cb(ter)%5Cb%2Fi">ter</a> is reached...</p>
<p><img src="svgbob_83b0a64a4614fa906388ef166341e3fd293c6dc0.svg" alt="Kroki diagram output"></p>
<p>For the <a href="#BM_(forward%20half-strand)%2Fi">forward half-strand</a>s, the process is much slower. Since <a href="#BM_(DNA%20polymerase)%2Fi">DNA polymerase</a> can only walk DNA in the reverse direction, the <a href="#BM_(forward%20half-strand)%2Fi">forward half-strand</a>s get replicated in small segments. That is, as the <a href="#BM_(replication%20fork)%2Fi">replication fork</a> continues to grow, every ~2000 nucleotides a new <a href="#BM_(primer)%2Fi">primer</a> attaches to the end of the fork on the forward strands. A new <a href="#BM_(DNA%20polymerase)%2Fi">DNA polymerase</a> attaches to each <a href="#BM_(primer)%2Fi">primer</a> and walks in the reverse direction (towards the <a href="#BM_%5Cb(ori)%5Cb%2Fi">ori</a>) to synthesize a small segment of DNA. That small segment of DNA is called an <a href="#BM_(Okazaki%20fragment)%2Fi">Okazaki fragment</a>...</p>
<p><img src="svgbob_4c651d9aa100c5b586103bb0d42cc91ff27da474.svg" alt="Kroki diagram output"></p>
<p>The <a href="#BM_(replication%20fork)%2Fi">replication fork</a> will keep widening until the original 2 strands split off. <a href="#BM_(DNA%20polymerase)%2Fi">DNA polymerase</a> will have made sure that for each separated strand, a newly synthesized <a href="#BM_(reverse%20complement)%2Fi">reverse complement</a> is paired to it. The end result is 2 daughter chromosome where each chromosome has gaps...</p>
<p><img src="svgbob_eac405c5709e835b492078d1ea557a7199d8f722.svg" alt="Kroki diagram output"></p>
<p>The <a href="#BM_(Okazaki%20fragment)%2Fi">Okazaki fragment</a>s synthesized on the forward strands end up getting sewn together by <a href="#BM_(DNA%20ligase)%2Fi">DNA ligase</a>...</p>
<p><img src="svgbob_5c3304d08cc6bbc226df966d6e1c119636331723.svg" alt="Kroki diagram output"></p>
<p>There are now two complete copies of the DNA.</p>
<a name="H_Find%20Ori%20and%20Ter"></a><h3>Find Ori and Ter</h3>
<p><a name="BM_(Stories%5C%2FBacterial%20Genome%20Replication%2FFind%20Ori%20and%20Ter)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FGC%20Skew)_TOPIC%2F">Algorithms/GC Skew</a></li>
</ul>
</div>
<p>Since the <a href="#BM_(forward%20half-strand)%2Fi">forward half-strand</a> gets its <a href="#BM_(reverse%20complement)%2Fi">reverse complement</a> synthesized at a much slower rate than the <a href="#BM_(reverse%20half-strand%7Cbackward%20half-strand%7Cbackwards%20half-strand)%2Fi">reverse half-strand</a>, it stays single stranded for a much longer time. <a href="#BM_(single%20stranded%20DNA%7Csingle-stranded%20DNA)%2Fi">Single-stranded DNA</a> is 100 times more susceptible to mutations than <a href="#BM_(double%20stranded%20DNA%7Cdouble-stranded%20DNA)%2Fi">double-stranded DNA</a>. Specifically, in <a href="#BM_(single%20stranded%20DNA%7Csingle-stranded%20DNA)%2Fi">single-stranded DNA</a>, C has a greater tendency to mutate to T. This process of mutation is referred to as deanimation.</p>
<p><img src="svgbob_f52124bb3b29a8388ed6a1fc5f597b26a2514f42.svg" alt="Kroki diagram output"></p>
<p>The <a href="#BM_(reverse%20half-strand%7Cbackward%20half-strand%7Cbackwards%20half-strand)%2Fi">reverse half-strand</a> spends much less time as a <a href="#BM_(single%20stranded%20DNA%7Csingle-stranded%20DNA)%2Fi">single-stranded DNA</a>. As such, it experiences much less C to T mutations.</p>
<p><img src="svgbob_7d778227822729bde64ce4a36b5cc9da206869c1.svg" alt="Kroki diagram output"></p>
<p>Ultimately, that means that a single strand will have a different nucleotide distribution between its <a href="#BM_(forward%20half-strand)%2Fi">forward half-strand</a> vs its <a href="#BM_(reverse%20half-strand%7Cbackward%20half-strand%7Cbackwards%20half-strand)%2Fi">backward half-strand</a>. If the half-strand being targeted for replication is the ...</p>
<ul>
<li><a href="#BM_(forward%20half-strand)%2Fi">forward half-strand</a>, some Cs get replaced with Ts. As such, its synthesized <a href="#BM_(reverse%20half-strand%7Cbackward%20half-strand%7Cbackwards%20half-strand)%2Fi">reverse half-strand</a> will have less Gs.</li>
<li><a href="#BM_(reverse%20half-strand%7Cbackward%20half-strand%7Cbackwards%20half-strand)%2Fi">reverse half-strand</a>, most Cs are kept. As such, its synthesized <a href="#BM_(forward%20half-strand)%2Fi">forward half-strand</a> will keep its Gs.</li>
</ul>
<p>To simplify, the ...</p>
<ul>
<li><a href="#BM_(forward%20half-strand)%2Fi">forward half-strand</a>: loses Cs, keeps Gs.</li>
<li><a href="#BM_(reverse%20half-strand%7Cbackward%20half-strand%7Cbackwards%20half-strand)%2Fi">reverse half-strand</a>: keeps Cs, loses Gs.</li>
</ul>
<p>You can use a GC skew diagram to help pinpoint where the <a href="#BM_%5Cb(ori)%5Cb%2Fi">ori</a> and <a href="#BM_%5Cb(ter)%5Cb%2Fi">ter</a> might be. The plot will typically form a peak where the <a href="#BM_%5Cb(ter)%5Cb%2Fi">ter</a> is (more G vs C) and form a valley where the <a href="#BM_%5Cb(ori)%5Cb%2Fi">ori</a> is (less G vs C). For example, the GC skew diagram for E. coli bacteria shows a distinct peak and distinct valley.</p>
<div style="border:1px solid black;">
<p></p>
<p>Calculating skew for: ...</p>
<p>Result: [0, 0, 1, 0,...</p>
<p><img src="skew_3b3ef8e0e7ef6297fdc10c525d5b0fbc.png" alt="GC Skew Plot"></p>
<p>Min position (ori): 4719166</p>
<p>Max position (ter): 2073768</p>
</div>
<p></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The material talks about how not all bacteria have a single peak and single valley. Some may have multiple. The reasoning for this still hasn't been discovered. It was speculated at one point that some bacteria may have multiple <a href="#BM_%5Cb(ori)%5Cb%2Fi">ori</a> / <a href="#BM_%5Cb(ter)%5Cb%2Fi">ter</a> regions.</p>
</div>
<a name="H_Find%20the%20DnaA%20Box"></a><h3>Find the DnaA Box</h3>
<p><a name="BM_(Stories%5C%2FBacterial%20Genome%20Replication%2FFind%20the%20DnaA%20Box)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Stories%5C%2FBacterial%20Genome%20Replication%2FFind%20Ori%20and%20Ter)_TOPIC%2F">Stories/Bacterial Genome Replication/Find Ori and Ter</a></li>
<li><a href="#BM_(Algorithms%5C%2FK-mer%2FFind%20Repeating%20in%20Window)_TOPIC%2F">Algorithms/K-mer/Find Repeating in Window</a></li>
<li><a href="#BM_(Algorithms%5C%2FGC%20Skew)_TOPIC%2F">Algorithms/GC Skew</a></li>
</ul>
</div>
<p>Within the <a href="#BM_%5Cb(ori)%5Cb%2Fi">ori</a> region, there exists several copies of some <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> pattern. These copies are referred to as <a href="#BM_(DnaA%20box)%2Fi">DnaA box</a>es.</p>
<p><img src="svgbob_d3e7c213995dec8a967fa89f40ebae73ed6e53ac.svg" alt="Kroki diagram output"></p>
<p>The DnaA protein binds to a <a href="#BM_(DnaA%20box)%2Fi">DnaA box</a> to activate the process of DNA replication. Through experiments, biologists have determined that <a href="#BM_(DnaA%20box)%2Fi">DnaA box</a>es are typical <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">9-mer</a>s. The <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">9-mer</a>s may not match exactly -- the DnaA protein may bind to ...</p>
<ul>
<li>the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">9-mer</a> itself.</li>
<li>slight variations of the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">9-mer</a>.</li>
<li>the <a href="#BM_(reverse%20complement)%2Fi">reverse complement</a> of the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">9-mer</a>.</li>
<li>slight variations of the <a href="#BM_(reverse%20complement)%2Fi">reverse complement</a> of the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">9-mer</a>.</li>
</ul>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The reason why multiple copies of the <a href="#BM_(DnaA%20box)%2Fi">DnaA box</a> exist probably has to do with DNA mutation. If one of the copies mutates to a point where the DnaA protein no longer binds to it, it can still bind to the other copies.</p>
</div>
<p>In the example below, the general vicinity of E. coli's <a href="#BM_%5Cb(ori)%5Cb%2Fi">ori</a> is found using GC skew, then that general vicinity is searched for repeating <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">9-mer</a>s. These repeating <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">9-mer</a>s are potential <a href="#BM_(DnaA%20box)%2Fi">DnaA box</a> candidates.</p>
<div style="border:1px solid black;">
<p></p>
<p>Calculating skew for: ...</p>
<p>Result: [0, 0, 1, 0,...</p>
<p><img src="skew_3b3ef8e0e7ef6297fdc10c525d5b0fbc.png" alt="GC Skew Plot"></p>
<p>Ori vicinity (min pos): 4719166</p>
<p>In the ori vicinity, found clusters of k=9 (at least 3 occurrences in window of 500) in ... at...</p>
<ul>
<li>KmerCluster(kmer='CGGTGCTGG', start_index=9, occurrence_count=3)</li>
<li>KmerCluster(kmer='CCGCGCTGG', start_index=9, occurrence_count=3)</li>
<li>KmerCluster(kmer='TCGGCGGTA', start_index=52, occurrence_count=3)</li>
<li>KmerCluster(kmer='CTCGGCGGT', start_index=53, occurrence_count=3)</li>
<li>KmerCluster(kmer='CTGAAGATC', start_index=98, occurrence_count=3)</li>
<li>KmerCluster(kmer='AGGCGGTTC', start_index=160, occurrence_count=3)</li>
<li>KmerCluster(kmer='CAGGCGGAT', start_index=161, occurrence_count=3)</li>
<li>KmerCluster(kmer='CAGGCGGCT', start_index=161, occurrence_count=3)</li>
<li>KmerCluster(kmer='CAGGCGGGT', start_index=161, occurrence_count=3)</li>
<li>KmerCluster(kmer='CAGGCGGTT', start_index=161, occurrence_count=4)</li>
<li>KmerCluster(kmer='TCAGGCGGT', start_index=162, occurrence_count=4)</li>
<li>KmerCluster(kmer='ATCAGGCGG', start_index=163, occurrence_count=4)</li>
<li>KmerCluster(kmer='CATCAGGCG', start_index=164, occurrence_count=3)</li>
<li>KmerCluster(kmer='CCATCAGGC', start_index=165, occurrence_count=3)</li>
<li>KmerCluster(kmer='CGGCGATGG', start_index=199, occurrence_count=3)</li>
<li>KmerCluster(kmer='GGCGGTATG', start_index=210, occurrence_count=3)</li>
<li>KmerCluster(kmer='GGCGGTTCG', start_index=210, occurrence_count=3)</li>
<li>KmerCluster(kmer='GGCAGTACG', start_index=210, occurrence_count=3)</li>
<li>KmerCluster(kmer='CAGGCGGTT', start_index=212, occurrence_count=3)</li>
<li>KmerCluster(kmer='TCAGGCGGC', start_index=213, occurrence_count=3)</li>
<li>KmerCluster(kmer='TCAGGCGGA', start_index=213, occurrence_count=3)</li>
<li>KmerCluster(kmer='TCAGGCGGT', start_index=213, occurrence_count=3)</li>
<li>KmerCluster(kmer='TCAGGCGGG', start_index=213, occurrence_count=3)</li>
<li>KmerCluster(kmer='CTCAGGCGG', start_index=214, occurrence_count=3)</li>
<li>KmerCluster(kmer='ATCAGGCGG', start_index=214, occurrence_count=3)</li>
<li>KmerCluster(kmer='TTCAGGCGG', start_index=214, occurrence_count=3)</li>
<li>KmerCluster(kmer='GTCAGGCGG', start_index=214, occurrence_count=3)</li>
<li>KmerCluster(kmer='GGTCAGGCG', start_index=215, occurrence_count=3)</li>
<li>KmerCluster(kmer='CGGATCCTG', start_index=232, occurrence_count=3)</li>
<li>KmerCluster(kmer='CGGATCGTT', start_index=232, occurrence_count=3)</li>
<li>KmerCluster(kmer='GCGGATCCT', start_index=233, occurrence_count=3)</li>
<li>KmerCluster(kmer='CCGGATCGT', start_index=233, occurrence_count=3)</li>
<li>KmerCluster(kmer='AGCCGGATC', start_index=235, occurrence_count=3)</li>
<li>KmerCluster(kmer='GATCCTAAA', start_index=282, occurrence_count=3)</li>
<li>KmerCluster(kmer='TTTTGATAC', start_index=427, occurrence_count=3)</li>
<li>KmerCluster(kmer='TTCTTTTGA', start_index=430, occurrence_count=3)</li>
<li>KmerCluster(kmer='TGGCTGGGG', start_index=460, occurrence_count=3)</li>
<li>KmerCluster(kmer='ATCACCATT', start_index=500, occurrence_count=3)</li>
<li>KmerCluster(kmer='TCCTTTTTA', start_index=513, occurrence_count=3)</li>
<li>KmerCluster(kmer='ATCCTTTTT', start_index=514, occurrence_count=3)</li>
<li>KmerCluster(kmer='GATCCTTAT', start_index=515, occurrence_count=3)</li>
<li>KmerCluster(kmer='GATCCTTTT', start_index=515, occurrence_count=3)</li>
<li>KmerCluster(kmer='GGATCCTTT', start_index=516, occurrence_count=3)</li>
<li>KmerCluster(kmer='CTGATCCTT', start_index=517, occurrence_count=3)</li>
<li>KmerCluster(kmer='CGGATCATT', start_index=517, occurrence_count=3)</li>
<li>KmerCluster(kmer='CCGATCCTT', start_index=517, occurrence_count=3)</li>
<li>KmerCluster(kmer='AGGATCCTT', start_index=517, occurrence_count=3)</li>
<li>KmerCluster(kmer='CCGGATCCC', start_index=518, occurrence_count=3)</li>
<li>KmerCluster(kmer='CAGGATCCT', start_index=518, occurrence_count=3)</li>
<li>KmerCluster(kmer='CCCGGATCC', start_index=519, occurrence_count=3)</li>
<li>KmerCluster(kmer='TTATCCAGA', start_index=522, occurrence_count=3)</li>
<li>KmerCluster(kmer='CCAGGTTTT', start_index=529, occurrence_count=3)</li>
<li>KmerCluster(kmer='TCATTCTCA', start_index=611, occurrence_count=3)</li>
<li>KmerCluster(kmer='TCCCAGGTT', start_index=618, occurrence_count=3)</li>
<li>KmerCluster(kmer='ACAGATCTT', start_index=644, occurrence_count=3)</li>
<li>KmerCluster(kmer='AAACAGATC', start_index=646, occurrence_count=3)</li>
<li>KmerCluster(kmer='TCCAAATAA', start_index=653, occurrence_count=3)</li>
<li>KmerCluster(kmer='TTATTGATC', start_index=730, occurrence_count=3)</li>
<li>KmerCluster(kmer='GTTGTTGAG', start_index=731, occurrence_count=3)</li>
<li>KmerCluster(kmer='AGGATCAAC', start_index=769, occurrence_count=3)</li>
<li>KmerCluster(kmer='AGATCAACC', start_index=836, occurrence_count=3)</li>
</ul>
</div>
<p></p>
<a name="H_Transcription%20Factors"></a><h2>Transcription Factors</h2>
<p><a name="BM_(Stories%5C%2FTranscription%20Factors)_TOPIC%2F"></a><strong></strong></p>
<p>A <a href="#BM_(transcription%20factor)%2Fi">transcription factor</a> / <a href="#BM_(regulatory%20protein)%2Fi">regulatory protein</a> is an enzyme that influences the rate of <a href="#BM_(gene%20expression)%2Fi">gene expression</a> for some set of <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">genes</a>. As the saturation of a <a href="#BM_(transcription%20factor)%2Fi">transcription factor</a> changes, so does the rate of <a href="#BM_(gene%20expression)%2Fi">gene expression</a> for the set of <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">genes</a> that it influences.</p>
<p><a href="#BM_(transcription%20factor)%2Fi">Transcription factor</a>s bind to DNA near the <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">genes</a> they influence: a <a href="#BM_(transcription%20factor%20binding%20site)%2Fi">transcription factor binding site</a> is located in a <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a>'s <a href="#BM_(upstream%20region)%2Fi">upstream region</a> and the <a href="#BM_(sequence)%2Fi">sequence</a> at that location is a fuzzy nucleotide <a href="#BM_(sequence)%2Fi">sequence</a> of length 8 to 12 called a <a href="#BM_(regulatory%20motif)%2Fi">regulatory motif</a>. The simplest way to think of a <a href="#BM_(regulatory%20motif)%2Fi">regulatory motif</a> is a regex pattern without quantifiers. For example, the regex <code>[AT]TT[GC]CCCTA</code> may match to ATTGCCCTA, ATTCCCCTA, TTTGCCCTA, and TTTCCCCTA. The regex itself is the <a href="#BM_(motif)%2Fi">motif</a>, while the <a href="#BM_(sequence)%2Fi">sequence</a>s being matched are <a href="#BM_(motif%20member)%2Fi">motif member</a>s.</p>
<p><img src="svgbob_c77523791e9f610ea50acd1a596be075095bcceb.svg" alt="Kroki diagram output"></p>
<p>The production of <a href="#BM_(transcription%20factor)%2Fi">transcription factor</a>s may be tied to certain internal or external conditions. For example, imagine a flower where the petals...</p>
<ul>
<li>bunch together at night time when sunlight is hidden and temperature is lower.</li>
<li>spread out at day time when sunlight is available and temperature is higher.</li>
</ul>
<p>The external conditions of sunlight and temperature causes the saturation of some <a href="#BM_(transcription%20factor)%2Fi">transcription factor</a>s to change. Those <a href="#BM_(transcription%20factor)%2Fi">transcription factor</a>s influence the rate of <a href="#BM_(gene%20expression)%2Fi">gene expression</a> for the <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">genes</a> that control the bunching and spreading of the petals.</p>
<p><img src="svgbob_02d5e3f967efc5f1f94cccd5f97a6610035cb8fd.svg" alt="Kroki diagram output"></p>
<a name="H_Find%20Regulatory%20Motif"></a><h3>Find Regulatory Motif</h3>
<p><a name="BM_(Stories%5C%2FTranscription%20Factors%2FFind%20Regulatory%20Motif)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FMotif%2FFind%20Motif%20Matrix)_TOPIC%2F">Algorithms/Motif/Find Motif Matrix</a></li>
</ul>
</div>
<p>Given an organism, it's suspected that some physical change in that organism is linked to a <a href="#BM_(transcription%20factor)%2Fi">transcription factor</a>. However, it isn't known ...</p>
<ul>
<li>which <a href="#BM_(transcription%20factor)%2Fi">transcription factor</a> (if any).</li>
<li>what the <a href="#BM_(regulatory%20motif)%2Fi">regulatory motif</a> for that <a href="#BM_(transcription%20factor)%2Fi">transcription factor</a> is.</li>
</ul>
<p>A special device is used to take snapshots of the organism's mRNA at different points in time: <a href="#BM_(DNA%20microarray)%2Fi">DNA microarray</a> / RNA <a href="#BM_(sequencer)%2Fi">sequencer</a>. Specifically, two snapshots are taken:</p>
<ol>
<li>When the physical change is expressed.</li>
<li>When the physical change isn't expressed.</li>
</ol>
<p>Comparing these snapshots identifies which <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">genes</a> have noticeably differing rates of <a href="#BM_(gene%20expression)%2Fi">gene expression</a>. If these <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">genes</a> (or a subset of these <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">genes</a>) were influenced by the same <a href="#BM_(transcription%20factor)%2Fi">transcription factor</a>, their <a href="#BM_(upstream%20region)%2Fi">upstream region</a>s would contain <a href="#BM_%5Cb(membership%7Cmember)_MOTIF%2Fi">member</a>s of that <a href="#BM_(transcription%20factor)%2Fi">transcription factor</a>'s <a href="#BM_(regulatory%20motif)%2Fi">regulatory motif</a>.</p>
<p>Since neither the <a href="#BM_(transcription%20factor)%2Fi">transcription factor</a> nor its <a href="#BM_(regulatory%20motif)%2Fi">regulatory motif</a> are known, there is no specific <a href="#BM_(motif)%2Fi">motif</a> to search for in the <a href="#BM_(upstream%20region)%2Fi">upstream region</a>s. But, because <a href="#BM_(motif%20member)%2Fi">motif member</a>s are typically similar to each other, <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> finding algorithms can be used on these <a href="#BM_(upstream%20region)%2Fi">upstream region</a>s to find sets of similar <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s. These similar <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s may all be <a href="#BM_%5Cb(membership%7Cmember)_MOTIF%2Fi">member</a>s of the same <a href="#BM_(transcription%20factor)%2Fi">transcription factor</a>'s <a href="#BM_(regulatory%20motif)%2Fi">regulatory motif</a>.</p>
<p><img src="svgbob_bc9f1ba4918b4a4995ec308b823f93eb3da45567.svg" alt="Kroki diagram output"></p>
<p>In the example below, a set of <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">genes</a> in baker's yeast (<a href="#BM_(Saccharomyces)%2Fi">Saccharomyces</a> cerevisiae) are suspected of being influenced by the same <a href="#BM_(transcription%20factor)%2Fi">transcription factor</a>. These <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">genes</a> are searched for a common <a href="#BM_(motif)%2Fi">motif</a>. Assuming one is found, it could be the <a href="#BM_(motif)%2Fi">motif</a> of the suspected <a href="#BM_(transcription%20factor)%2Fi">transcription factor</a>.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The example below hard codes k to 18, but you typically don't know what k should be set to beforehand. The Pevzner book doesn't discuss how to work around this problem. A strategy for finding k may be to run the <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a> finding algorithm multiple times, but with a different k each time. For each <a href="#BM_%5Cb(membership%7Cmember)_MOTIF%2Fi">member</a>, if the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s selected across the runs came from the same general vicinity of the <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a>'s <a href="#BM_(upstream%20region)%2Fi">upstream region</a>, those <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s may either be picking ...</p>
<ul>
<li>the actual <a href="#BM_%5Cb(membership%7Cmember)_MOTIF%2Fi">member</a>.</li>
<li>a part of the actual <a href="#BM_%5Cb(membership%7Cmember)_MOTIF%2Fi">member</a>.</li>
<li>a part of the actual <a href="#BM_%5Cb(membership%7Cmember)_MOTIF%2Fi">member</a> with some junk prepended/appended to it.</li>
</ul>
</div>
<div style="border:1px solid black;">
<p></p>
<p>Organism is baker's yeast. Suspected genes influenced by transcription factor: THI12, YHL017W, SYN8, YCG1, UBX5, and KEI1.</p>
<p>Searching for 18-mer across a set of 6 gene upstream regions...</p>
<p>GAAAAGAAAGAAAAAGGA<br>GAAAAGAAAAAGAAAAAA<br>GAAAGAAAAAGAAAAAAA<br>AAAAGGAAAAAAAGAAGA<br>GAAATGAAAAGGAACAGT<br>AAAATCAAAAAAATAAAT</p>
<p>Score is: 22</p>
</div>
<p></p>
<a name="H_Non-ribosomal%20Peptides"></a><h2>Non-ribosomal Peptides</h2>
<p><a name="BM_(Stories%5C%2FNon-ribosomal%20Peptides)_TOPIC%2F"></a><strong></strong></p>
<p>A <a href="#BM_(peptide)%2Fi">peptide</a> is a miniature protein consisting of a chain of <a href="#BM_(amino%20acid)%2Fi">amino acid</a>s anywhere between 2 to 100 <a href="#BM_(amino%20acid)%2Fi">amino acid</a>s in length.</p>
<p><img src="svgbob_65d4227aac16dd6290813a8c18042c837e078f9d.svg" alt="Kroki diagram output"></p>
<p>Most <a href="#BM_(peptide)%2Fi">peptide</a>s are synthesized through the <a href="#BM_(central%20dogma%20of%20molecular%20biology)%2Fi">central dogma of molecular biology</a>: a segment of the DNA that <a href="#BM_(encode%7Cencoding%7Cdecode%7Cdecoding)%2Fi">encode</a>s the <a href="#BM_(peptide)%2Fi">peptide</a> is <a href="#BM_(transcription%7Ctranscribed%7Ctranscribe)%2Fi">transcribed</a> to mRNA, which in turn is <a href="#BM_(translation%7Ctranslated%7Ctranslate)%2Fi">translated</a> to a <a href="#BM_(peptide)%2Fi">peptide</a> by a ribosome.</p>
<p><img src="svgbob_3b69a590a67c8d8db6ca97d50bdeec9cb4c1ffee.svg" alt="Kroki diagram output"></p>
<p><a href="#BM_(non-ribosomal%20peptide)%2Fi">Non-ribosomal peptide</a>s (<a href="#BM_%5Cb(NRP)s%3F%5Cb%2F">NRP</a>) however, aren't synthesized via the <a href="#BM_(central%20dogma%20of%20molecular%20biology)%2Fi">central dogma of molecular biology</a>. Instead, giant proteins typically found in bacteria and fungi called <a href="#BM_(non-ribosomal%20peptide%20synthetase%7CNRP%20synthetase)%2Fi">NRP synthetase</a> build out these <a href="#BM_(peptide)%2Fi">peptide</a>s by growing them one <a href="#BM_(amino%20acid)%2Fi">amino acid</a> at a time.</p>
<p><img src="svgbob_beb44b4a49d8f6713eefec469487c6b717b811f4.svg" alt="Kroki diagram output"></p>
<p>Each segment of an <a href="#BM_(non-ribosomal%20peptide%20synthetase%7CNRP%20synthetase)%2Fi">NRP synthetase</a> protein responsible for the outputting a single <a href="#BM_(amino%20acid)%2Fi">amino acid</a> is called an <a href="#BM_(adenylation%20domain%7CA%20domain%7CA-domain)%2Fi">adenylation domain</a>. The example above has 5 <a href="#BM_(adenylation%20domain%7CA%20domain%7CA-domain)%2Fi">adenylation domain</a>s, each of which is responsible for outputting a single <a href="#BM_(amino%20acid)%2Fi">amino acid</a> of the <a href="#BM_(peptide)%2Fi">peptide</a> it produces.</p>
<p><a href="#BM_%5Cb(NRP)s%3F%5Cb%2F">NRP</a>s may be <a href="#BM_(cyclopeptide|cyclic peptide)/i">cyclic</a>. Common use-cases for <a href="#BM_%5Cb(NRP)s%3F%5Cb%2F">NRP</a>s:</p>
<ul>
<li><a href="#BM_(antibiotic)%2Fi">antibiotic</a>s</li>
<li>anti-tumor agents</li>
<li>immunosuppressors</li>
<li>communication between bacteria (quorum sensing)</li>
</ul>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>According to the Wikipedia article on <a href="#BM_%5Cb(NRP)s%3F%5Cb%2F">NRP</a>s, there exist a wide range of <a href="#BM_(peptide)%2Fi">peptide</a>s that are not synthesized by ribosomes but the term <a href="#BM_(non-ribosomal%20peptide)%2Fi">non-ribosomal peptide</a> typically refers to the ones synthesized by <a href="#BM_(non-ribosomal%20peptide%20synthetase%7CNRP%20synthetase)%2Fi">NRP synthetase</a>s.</p>
</div>
<a name="H_Find%20Sequence"></a><h3>Find Sequence</h3>
<p><a name="BM_(Stories%5C%2FNon-ribosomal%20Peptides%2FFind%20Sequence)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FPeptide%20Sequence%2FSpectrum%20Sequence)_TOPIC%2F">Algorithms/Peptide Sequence/Spectrum Sequence</a></li>
</ul>
</div>
<p>Unlike ribosomal <a href="#BM_(peptide)%2Fi">peptide</a>s, <a href="#BM_%5Cb(NRP)s%3F%5Cb%2F">NRP</a>s aren't <a href="#BM_(encode%7Cencoding%7Cdecode%7Cdecoding)%2Fi">encode</a>d in the organism's DNA. As such, their <a href="#BM_(sequence)%2Fi">sequence</a> can't be inferred by directly by looking through the organism's DNA <a href="#BM_(sequence)%2Fi">sequence</a>.</p>
<p>Instead, a sample of the <a href="#BM_%5Cb(NRP)s%3F%5Cb%2F">NRP</a> needs to be isolated and passed through a <a href="#BM_(mass%20spectrometer%7Cmass%20spectrometry)%2Fi">mass spectrometer</a>. A <a href="#BM_(mass%20spectrometer%7Cmass%20spectrometry)%2Fi">mass spectrometer</a> is a device that shatters and bins molecules by their mass-to-charge ratio: Given a sample of molecules, the device randomly shatters each molecule in the sample (forming ions), then bins each ion by its mass-to-charge ratio (<span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mfrac><mi>m</mi><mi>z</mi></mfrac></mrow><annotation encoding="application/x-tex">\frac{m}{z}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.040392em;vertical-align:-0.345em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.695392em;"><span style="top:-2.6550000000000002em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.04398em;">z</span></span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.394em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">m</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.345em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span>).</p>
<p>The output of a <a href="#BM_(mass%20spectrometer%7Cmass%20spectrometry)%2Fi">mass spectrometer</a> is a plot called a <a href="#BM_(spectrum)_MS%2Fi">spectrum</a>. The plot's ...</p>
<ul>
<li>x-axis is the mass-to-charge ratio.</li>
<li>y-axis is the intensity of that mass-to-charge ratio (how much more / less did that mass-to-charge appear compared to the others).</li>
</ul>
<p><img src="svgbob_2b5ef43ae7572c2c6bf63687c53860a71830d160.svg" alt="Kroki diagram output"></p>
<p>For example, given a sample containing multiple instances of the <a href="#BM_(linear%20peptide)%2Fi">linear peptide</a> NQY, the <a href="#BM_(mass%20spectrometer%7Cmass%20spectrometry)%2Fi">mass spectrometer</a> will take each instance of NQY and randomly break the bonds between its <a href="#BM_(amino%20acid)%2Fi">amino acid</a>s:</p>
<p><img src="svgbob_33c109cffe376e25dfa6ba85b256a1bb307dcd76.svg" alt="Kroki diagram output"></p>
<p>Each <a href="#BM_(subpeptide)%2Fi">subpeptide</a> then will have its mass-to-charge ratio measured, which in turn gets converted to a set of potential masses by performing basic math. With these potential masses, it's possible to infer which <a href="#BM_(amino%20acid)%2Fi">amino acid</a>s make up the <a href="#BM_(peptide)%2Fi">peptide</a> as well as the <a href="#BM_(peptide)%2Fi">peptide</a> <a href="#BM_(sequence)%2Fi">sequence</a>.</p>
<p>In the example below, <a href="#BM_(peptide)%2Fi">peptide</a> <a href="#BM_(sequence)%2Fi">sequence</a>s are inferred from a noisy <a href="#BM_(spectrum)_MS%2Fi">spectrum</a> for the <a href="#BM_(cyclopeptide%7Ccyclic%20peptide)%2Fi">cyclopeptide</a> Viomycin. The elements of each inferred <a href="#BM_(peptide)%2Fi">peptide</a> <a href="#BM_(sequence)%2Fi">sequence</a> are <a href="#BM_(amino%20acid)%2Fi">amino acid</a> masses rather than the <a href="#BM_(amino%20acid)%2Fi">amino acid</a>s themselves (e.g. instead of S being output at a position, the mass of S is output -- 87). Since the <a href="#BM_(spectrum)_MS%2Fi">spectrum</a> is noisy, the inferred <a href="#BM_(peptide)%2Fi">peptide</a> <a href="#BM_(sequence)%2Fi">sequence</a>s are also noisy (e.g. instead of an <a href="#BM_(amino%20acid)%2Fi">amino acid</a> mass 87 showing up as exactly 87 in the <a href="#BM_(peptide)%2Fi">peptide</a> <a href="#BM_(sequence)%2Fi">sequence</a>, it may show up as 87.2, 86.9, etc...).</p>
<p>Note that the correct <a href="#BM_(peptide)%2Fi">peptide</a> <a href="#BM_(sequence)%2Fi">sequence</a> isn't guaranteed to be inferred. Also, since Viomycin is a <a href="#BM_(cyclopeptide%7Ccyclic%20peptide)%2Fi">cyclopeptide</a>, the correct <a href="#BM_(peptide)%2Fi">peptide</a> may be inferred in a wrapped form (e.g. the <a href="#BM_(cyclopeptide%7Ccyclic%20peptide)%2Fi">cyclopeptide</a> 128-113-57 may show up as 128-113-57, 113-57-128, or 57-128-113).</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>I artificially generated a <a href="#BM_(spectrum)_MS%2Fi">spectrum</a> for Viomycin from the <a href="#BM_(sequence)%2Fi">sequence</a> listed on <a href="https://www.genome.jp/dbget-bin/www_bget?cpd:C01540">KEGG</a>.</p>
<blockquote>
<p><a href="#BM_(sequence)%2Fi">Sequence</a> 	 0 beta-Lys  1 Dpr  2 Ser  3 Ser  4 Ala  5 Cpd  (Cyclization: 1-5)</p>
<p><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> 	0 vioO [UP:Q6WZ98] vioM [UP:Q6WZA0]; 1 vioF [UP:Q6WZA7]; 2-3 vioA
[UP:Q6WZB2]; 4 vioI [UP:Q6WZA4]; 5 vioG [UP:Q84CG4]</p>
<p>Organism 	 Streptomyces vinaceus</p>
<p>Type 	<a href="#BM_%5Cb(NRP)s%3F%5Cb%2F">NRP</a></p>
</blockquote>
<p>The problem is that I have no idea what the 5th <a href="#BM_(amino%20acid)%2Fi">amino acid</a> is: Cpd (I arbitrarily put it's mass as 200) and I'm unsure of the mapping I found for Dpr (2,3-diaminopropionic acid has mass of 104). The <a href="#BM_(peptide)%2Fi">peptide</a> <a href="#BM_(sequence)%2Fi">sequence</a> being searched for in the example below is 128-104-87-87-71-200.</p>
</div>
<div style="border:1px solid black;">
<p></p>
<p>Given the ...</p>
<ul>
<li>experimental spectrum: [0.0, 71.2, 87.1, 104.0, 127.9, 158.0, 174.0, 191.0, 200.0, 232.0, 245.0, 271.0, 319.0, 328.0, 358.0, 399.0, 406.0, 432.0, 445.0, 477.1, 490.0, 492.0, 503.1, 530.5, 548.7, 572.9, 590.1, 606.1, 677.0]</li>
<li>experimental spectrum mass noise: ±0.3</li>
<li>assumed peptide type: cyclic</li>
<li>assumed peptide length: 6</li>
<li>assumed peptide mass: any of the last 1 experimental spectrum masses</li>
<li>score backlog: 0</li>
<li>leaderboard size: 30</li>
</ul>
<p>Top 24 captured mino acid masses (rounded to 1): [86.8, 86.9, 87.0, 87.1, 71.1, 71.2, 70.9, 71.0, 128.3, 199.8, 199.9, 200.0, 200.1, 103.7, 103.9, 104.0, 104.1, 127.9, 128.0]</p>
<p>For peptides between 673.1 and 680.9...</p>
<ul>
<li>Score 26: 104.0-128.0-200.0-71.0-87.0-86.8</li>
<li>Score 26: 104.0-128.0-200.0-71.0-87.0-86.9</li>
<li>Score 26: 104.0-128.0-200.0-71.0-87.0-87.0</li>
<li>Score 26: 104.0-128.0-200.0-71.0-87.0-87.1</li>
<li>Score 26: 104.0-128.0-200.0-71.1-87.0-86.8</li>
<li>Score 26: 104.0-128.0-200.0-71.1-87.0-86.9</li>
<li>Score 26: 104.0-128.0-200.0-71.1-87.0-87.0</li>
<li>Score 26: 104.0-128.0-200.0-71.1-87.0-87.1</li>
<li>Score 26: 104.1-127.9-200.0-71.1-86.9-86.8</li>
<li>Score 26: 104.1-127.9-200.0-71.1-86.9-86.9</li>
<li>Score 26: 104.1-127.9-200.0-71.1-86.9-87.0</li>
<li>Score 26: 104.1-127.9-200.0-71.1-86.9-87.1</li>
<li>Score 26: 104.0-128.0-200.0-71.0-87.1-86.8</li>
<li>Score 26: 104.0-128.0-200.0-71.0-87.1-86.9</li>
<li>Score 26: 104.0-128.0-200.0-71.0-87.1-87.0</li>
<li>Score 26: 104.0-128.0-200.0-71.0-87.1-87.1</li>
<li>Score 26: 104.0-127.9-200.1-71.0-87.0-86.8</li>
<li>Score 26: 104.0-127.9-200.1-71.0-87.0-86.9</li>
<li>Score 26: 104.0-127.9-200.1-71.0-87.0-87.0</li>
<li>Score 26: 104.0-127.9-200.1-71.0-87.0-87.1</li>
<li>Score 26: 104.1-127.9-200.0-71.0-87.0-86.8</li>
<li>Score 26: 104.1-127.9-200.0-71.0-87.0-86.9</li>
<li>Score 26: 104.1-127.9-200.0-71.0-87.0-87.0</li>
<li>Score 26: 104.1-127.9-200.0-71.0-87.0-87.1</li>
<li>Score 26: 104.0-127.9-200.0-71.0-87.0-86.9</li>
<li>Score 26: 104.0-127.9-200.0-71.0-87.0-87.0</li>
<li>Score 26: 104.0-127.9-200.0-71.0-87.0-87.1</li>
<li>Score 26: 104.1-127.9-200.0-71.1-87.0-86.8</li>
<li>Score 26: 104.1-127.9-200.0-71.1-87.0-86.9</li>
<li>Score 26: 104.1-127.9-200.0-71.1-87.0-87.0</li>
<li>Score 26: 104.1-127.9-200.0-71.1-87.0-87.1</li>
<li>Score 26: 104.0-127.9-200.0-71.1-87.0-86.8</li>
<li>Score 26: 104.0-127.9-200.0-71.1-87.0-86.9</li>
<li>Score 26: 104.0-127.9-200.0-71.1-87.0-87.0</li>
<li>Score 26: 104.0-127.9-200.0-71.1-87.0-87.1</li>
<li>Score 26: 104.0-128.0-200.0-71.1-86.9-86.8</li>
<li>Score 26: 104.0-128.0-200.0-71.1-86.9-86.9</li>
<li>Score 26: 104.0-128.0-200.0-71.1-86.9-87.0</li>
<li>Score 26: 104.0-128.0-200.0-71.1-86.9-87.1</li>
<li>Score 26: 104.0-127.9-200.1-71.0-87.1-86.8</li>
<li>Score 26: 104.0-127.9-200.1-71.0-87.1-86.9</li>
<li>Score 26: 104.0-127.9-200.1-71.0-87.1-87.0</li>
<li>Score 26: 104.0-127.9-200.1-71.0-87.1-87.1</li>
<li>Score 26: 104.0-127.9-200.0-71.0-87.1-86.8</li>
<li>Score 26: 104.0-127.9-200.0-71.0-87.1-86.9</li>
<li>Score 26: 104.0-127.9-200.0-71.0-87.1-87.0</li>
<li>Score 26: 104.0-127.9-200.0-71.0-87.1-87.1</li>
<li>Score 26: 104.0-127.9-200.0-71.1-87.1-86.8</li>
<li>Score 26: 104.0-127.9-200.0-71.1-87.1-86.9</li>
<li>Score 26: 104.0-127.9-200.0-71.1-87.1-87.0</li>
<li>Score 26: 104.0-127.9-200.0-71.1-87.1-87.1</li>
<li>Score 26: 104.0-127.9-200.1-71.1-87.0-86.8</li>
<li>Score 26: 104.0-127.9-200.1-71.1-87.0-86.9</li>
<li>Score 26: 104.0-127.9-200.1-71.1-87.0-87.0</li>
<li>Score 26: 104.0-127.9-200.1-71.1-87.0-87.1</li>
<li>Score 26: 104.0-127.9-200.0-71.1-86.9-86.9</li>
<li>Score 26: 104.0-127.9-200.0-71.1-86.9-87.0</li>
<li>Score 26: 104.0-127.9-200.0-71.1-86.9-87.1</li>
<li>Score 26: 71.2-200.0-127.9-104.0-87.0-86.8</li>
<li>Score 26: 71.2-200.0-127.9-104.0-87.0-86.9</li>
<li>Score 26: 71.2-200.0-127.9-104.0-87.0-87.0</li>
<li>Score 26: 71.2-200.0-127.9-104.0-87.0-87.1</li>
<li>Score 26: 104.0-127.9-200.0-71.2-87.0-86.8</li>
<li>Score 26: 104.0-127.9-200.0-71.2-87.0-86.9</li>
<li>Score 26: 104.0-127.9-200.0-71.2-87.0-87.0</li>
<li>Score 26: 104.0-127.9-200.0-71.2-87.0-87.1</li>
<li>Score 26: 104.0-128.0-200.0-71.1-87.1-86.8</li>
<li>Score 26: 104.0-128.0-200.0-71.1-87.1-86.9</li>
<li>Score 26: 104.0-128.0-200.0-71.1-87.1-87.0</li>
<li>Score 26: 104.1-127.9-200.0-71.2-86.9-86.8</li>
<li>Score 26: 104.1-127.9-200.0-71.2-86.9-86.9</li>
<li>Score 26: 104.1-127.9-200.0-71.2-86.9-87.0</li>
<li>Score 26: 104.1-127.9-200.0-71.2-86.9-87.1</li>
<li>Score 26: 71.2-200.0-127.9-104.1-86.9-86.8</li>
<li>Score 26: 71.2-200.0-127.9-104.1-86.9-86.9</li>
<li>Score 26: 71.2-200.0-127.9-104.1-86.9-87.0</li>
<li>Score 26: 71.2-200.0-127.9-104.1-86.9-87.1</li>
<li>Score 26: 104.1-127.9-200.0-71.0-87.1-86.8</li>
<li>Score 26: 104.1-127.9-200.0-71.0-87.1-86.9</li>
<li>Score 26: 104.1-127.9-200.0-71.0-87.1-87.0</li>
<li>Score 26: 104.1-127.9-200.0-71.0-87.1-87.1</li>
<li>Score 26: 104.0-128.0-200.0-71.2-87.0-86.8</li>
<li>Score 26: 104.0-128.0-200.0-71.2-87.0-86.9</li>
<li>Score 26: 104.0-128.0-200.0-71.2-87.0-87.0</li>
<li>Score 26: 104.0-127.9-200.1-71.1-86.9-86.8</li>
<li>Score 26: 104.0-127.9-200.1-71.1-86.9-86.9</li>
<li>Score 26: 104.0-127.9-200.1-71.1-86.9-87.0</li>
<li>Score 26: 104.0-127.9-200.1-71.1-86.9-87.1</li>
<li>Score 26: 104.1-127.9-200.0-71.1-87.1-86.8</li>
<li>Score 26: 104.1-127.9-200.0-71.1-87.1-86.9</li>
<li>Score 26: 104.1-127.9-200.0-71.1-87.1-87.0</li>
<li>Score 26: 104.0-127.9-200.1-71.1-87.1-86.8</li>
<li>Score 26: 104.0-127.9-200.1-71.1-87.1-86.9</li>
<li>Score 26: 104.0-127.9-200.1-71.1-87.1-87.0</li>
<li>Score 26: 104.0-128.0-200.0-71.2-86.9-86.8</li>
<li>Score 26: 104.0-128.0-200.0-71.2-86.9-86.9</li>
<li>Score 26: 104.0-128.0-200.0-71.2-86.9-87.0</li>
<li>Score 26: 104.0-128.0-200.0-71.2-86.9-87.1</li>
<li>Score 26: 104.0-127.9-200.0-71.2-87.1-86.8</li>
<li>Score 26: 104.0-127.9-200.0-71.2-87.1-86.9</li>
<li>Score 26: 104.0-127.9-200.0-71.2-87.1-87.0</li>
<li>Score 26: 104.0-127.9-200.0-71.2-87.1-87.1</li>
<li>Score 26: 104.0-127.9-200.0-71.2-86.9-86.8</li>
<li>Score 26: 104.0-127.9-200.0-71.2-86.9-86.9</li>
<li>Score 26: 104.0-127.9-200.0-71.2-86.9-87.0</li>
<li>Score 26: 104.0-127.9-200.0-71.2-86.9-87.1</li>
<li>Score 26: 71.2-200.0-127.9-104.0-86.9-86.8</li>
<li>Score 26: 71.2-200.0-127.9-104.0-86.9-86.9</li>
<li>Score 26: 71.2-200.0-127.9-104.0-86.9-87.0</li>
<li>Score 26: 71.2-200.0-127.9-104.0-86.9-87.1</li>
<li>Score 26: 71.2-200.0-127.9-104.0-87.1-86.8</li>
<li>Score 26: 71.2-200.0-127.9-104.0-87.1-86.9</li>
<li>Score 26: 71.2-200.0-127.9-104.0-87.1-87.0</li>
</ul>
</div>
<p></p>
<a name="H_Genome%20Rearrangement"></a><h2>Genome Rearrangement</h2>
<p><a name="BM_(Stories%5C%2FGenome%20Rearrangement)_TOPIC%2F"></a><strong></strong></p>
<p><a href="#BM_(genome%20rearrangement%7Cchromosomal%20rearrangement%7Cchromosome%20rearrangement)%2Fi">Genome rearrangement</a> is form of mutation where chromosomes go through structural changes. These structural changes include chromosome segments getting ...</p>
<ul>
<li>
<p>shuffled into a different order (<a href="#BM_(translocation)%2Fi">translocation</a>, <a href="#BM_(fission)%2Fi">fission</a>, <a href="#BM_(%5Cbfusion%7C%5Cbfuse)%2Fi">fusion</a>) or direction (reversal).</p>
<p>For example, a segment of chromosome breaks off and rejoins, but each end of that segment joins back up at a different point.</p>
<p><img src="svgbob_dfd16cae74c24c2767d35903c11b7f5050f886c4.svg" alt="Kroki diagram output"></p>
<p><img src="svgbob_03cbcaf05f055c39a43c9a1cc326b685879d4beb.svg" alt="Kroki diagram output"></p>
<p><img src="svgbob_a3e31548f4a9e9497f1251dc3f6017c22b2d5858.svg" alt="Kroki diagram output"></p>
<p><img src="svgbob_d1769d28e9ed5c384f15ea5f0bde362ee6eab157.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>deleted.</p>
<p>For example, a segment of a chromosome breaks off and DNA repair mechanisms close the gap.</p>
<p><img src="svgbob_a9ce396e59a5a339165477856ba34d65ad885310.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>duplicated.</p>
<p>For example, a segment of a chromosome breaks off and DNA repair mechanisms close the gap, but that broken off segment may still re-attach at a different location.</p>
<p><img src="svgbob_302aff8adacb0e3118dfe66ec50ed3015862d7ee.svg" alt="Kroki diagram output"></p>
</li>
</ul>
<p>There are fragile regions of chromosomes where breakages are more likely to occur. For example, the ABL-BCR <a href="#BM_(%5Cbfusion%7C%5Cbfuse)%2Fi">fusion</a> protein, a protein implicated in the development of a cancer known as chronic myeloid leukemia, is the result of human chromosomes 9 and 22 breaking and fusing back together in a different order: Chromosome 9 contains the <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a> for ABL while chromosome 22 contains the <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a> for BCR and both <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">genes</a> are in fragile regions of their respective chromosome. If those fragile chromosome regions both break but <a href="#BM_(%5Cbfusion%7C%5Cbfuse)%2Fi">fuse</a> back together in the wrong order, the ABL-BCR <a href="#BM_(chimeric%20gene)%2Fi">chimeric gene</a> is formed.</p>
<p>As shown with the ABL-BCR <a href="#BM_(%5Cbfusion%7C%5Cbfuse)%2Fi">fusion</a> protein example above, <a href="#BM_(genome%20rearrangement%7Cchromosomal%20rearrangement%7Cchromosome%20rearrangement)%2Fi">genome rearrangement</a>s often result in the sterility or death of the organism. However, when a species branches off from an existing one, <a href="#BM_(genome%20rearrangement%7Cchromosomal%20rearrangement%7Cchromosome%20rearrangement)%2Fi">genome rearrangement</a>s are likely responsible for at least some of the divergence. That is, the two related <a href="#BM_(genome)%2Fi">genome</a>s will share long stretches of similar <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">genes</a>, but these long stretches will appear as if they had been randomly cut-and-paste and / or randomly reversed when compared to the other. For example, humans and mice have a shared ancestry and as such share a vast number of long stretches (around 280).</p>
<p>These long stretches of similar <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">genes</a> are called <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s. For example, the following <a href="#BM_(genome%20rearrangement%7Cchromosomal%20rearrangement%7Cchromosome%20rearrangement)%2Fi">genome rearrangement</a> mutations result in 4 <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s shared between the two <a href="#BM_(genome)%2Fi">genome</a>s ...</p>
<p><img src="svgbob_d5f827e20974efd41c9ac222d75bd5443f3d190d.svg" alt="Kroki diagram output"></p>
<ul>
<li><a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">Synteny block</a> 1: <code>[G1, G2]</code></li>
<li><a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">Synteny block</a> 2: <code>[G3]</code></li>
<li><a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">Synteny block</a> 3: <code>[G4, G5, G6]</code> (although they're reversed)</li>
<li><a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">Synteny block</a> 4: <code>[G7, G8, G9]</code></li>
</ul>
<a name="H_Find%20Synteny%20Blocks"></a><h3>Find Synteny Blocks</h3>
<p><a name="BM_(Stories%5C%2FGenome%20Rearrangement%2FFind%20Synteny%20Blocks)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FSynteny%2FGenomic%20Dot%20Plot)_TOPIC%2F">Algorithms/Synteny/Genomic Dot Plot</a></li>
<li><a href="#BM_(Algorithms%5C%2FSynteny%2FSynteny%20Graph)_TOPIC%2F">Algorithms/Synteny/Synteny Graph</a></li>
</ul>
</div>
<p><a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">Synteny block</a>s are identified by first finding matching <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s and <a href="#BM_(reverse%20complement)%2Fi">reverse complement</a> matching <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s, then combining matches that are close together (<a href="#BM_(cluster)%2Fi">cluster</a>ing) to reveal the long stretches of matches that make up <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s.</p>
<p>The visual manifestation of this concept is the <a href="#BM_(genomic%20dot-plot%7Cgenomic%20dot%20plot)%2Fi">genomic dot-plot</a> and <a href="#BM_(synteny%20graph)%2Fi">synteny graph</a>. A <a href="#BM_(genomic%20dot-plot%7Cgenomic%20dot%20plot)%2Fi">genomic dot-plot</a> is a 2D plot where each axis is assigned to one of the <a href="#BM_(genome)%2Fi">genome</a>s and a dot is placed at each coordinate containing a match, while a <a href="#BM_(synteny%20graph)%2Fi">synteny graph</a> is the <a href="#BM_(cluster)%2Fi">cluster</a>ed form of a <a href="#BM_(genomic%20dot-plot%7Cgenomic%20dot%20plot)%2Fi">genomic dot-plot</a> that reveals <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s.</p>
<p><img src="svgbob_eb50ab9882ae13623c74006244614b059e22fa47.svg" alt="Kroki diagram output"></p>
<p>The <a href="#BM_(synteny%20graph)%2Fi">synteny graph</a> above reveals that 4 <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s are shared between the <a href="#BM_(genome)%2Fi">genome</a>s. One of the <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s is a normal match (C) while three are matching against their <a href="#BM_(reverse%20complement)%2Fi">reverse complement</a>s (A, B, and D).</p>
<p><img src="svgbob_725cfe74948aa476c335f99e888801f679a0059a.svg" alt="Kroki diagram output"></p>
<p>In the example below, two species of the Mycoplasma bacteria are analyzed to find the <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s between them. The output reveals that pretty much the entirety of both <a href="#BM_(genome)%2Fi">genome</a>s are shared, just in a different order.</p>
<div style="border:1px solid black;">
<p></p>
<p>Finding synteny blocks for...</p>
<ul>
<li>k=20</li>
<li>cyclic=True</li>
<li>show_graph=True</li>
<li>genome1 reference: [[G1C1, /input/ch6_code/src/Mycoplasma bovis - GCA_000696015.1_ASM69601v1_genomic.fna.xz]]</li>
<li>genome2 reference: [[G2C1, /input/ch6_code/src/Mycoplasma agalactiae 14628 - GCA_000266865.1_ASM26686v1_genomic.fna.xz]]</li>
</ul>
<p>NOTE: Nucleotide codes that aren't ACGT get filtered out of the genomes.</p>
<p>Generating genomic dotplot...</p>
<p><img src="dotplot_aa18ecc0bccea2e4d491f745259e60c6.png" alt="Genomic Dot Plot"></p>
<p>Clustering genomic dotplot to snyteny graph...</p>
<ul>
<li>Merging radius=25 angle_half_maw=90...</li>
<li>Merging radius=50 angle_half_maw=90...</li>
<li>Merging radius=100 angle_half_maw=90...</li>
<li>Merging radius=200 angle_half_maw=90...</li>
<li>Merging radius=400 angle_half_maw=90...</li>
<li>Merging radius=800 angle_half_maw=90...</li>
<li>Merging radius=1600 angle_half_maw=90...</li>
<li>Merging radius=3200 angle_half_maw=90...</li>
<li>Filtering max_filter_length=3200.0 max_merge_distance=20000.0...</li>
<li>Merging radius=6400 angle_half_maw=90...</li>
<li>Merging radius=12800 angle_half_maw=90...</li>
<li>Culling below length=10000.0...</li>
</ul>
<p>Generating synteny graph...</p>
<p><img src="syntenygraph_aa18ecc0bccea2e4d491f745259e60c6.png" alt="Synteny Graph"></p>
<p>Mapping synteny graph matches to IDs using x-axis genome...</p>
<ul>
<li>G2C1_B0 = {'y': ('G1C1', 456122, 490876), 'x': ('G2C1', 26146, 55215), 'type': 'REVERSE_COMPLEMENT'}</li>
<li>G2C1_B1 = {'y': ('G1C1', 403833, 428055), 'x': ('G2C1', 55602, 78341), 'type': 'REVERSE_COMPLEMENT'}</li>
<li>G2C1_B2 = {'y': ('G1C1', 356150, 384121), 'x': ('G2C1', 88486, 113645), 'type': 'REVERSE_COMPLEMENT'}</li>
<li>G2C1_B3 = {'y': ('G1C1', 596687, 770257), 'x': ('G2C1', 119776, 270535), 'type': 'NORMAL'}</li>
<li>G2C1_B4 = {'y': ('G1C1', 772154, 835264), 'x': ('G2C1', 287182, 351655), 'type': 'NORMAL'}</li>
<li>G2C1_B5 = {'y': ('G1C1', 835398, 948473), 'x': ('G2C1', 385360, 498503), 'type': 'NORMAL'}</li>
<li>G2C1_B6 = {'y': ('G1C1', 90, 262351), 'x': ('G2C1', 498629, 739753), 'type': 'NORMAL'}</li>
<li>G2C1_B7 = {'y': ('G1C1', 276781, 355692), 'x': ('G2C1', 740078, 815603), 'type': 'NORMAL'}</li>
<li>G2C1_B8 = {'y': ('G1C1', 529629, 591336), 'x': ('G2C1', 816444, 878201), 'type': 'REVERSE_COMPLEMENT'}</li>
<li>G2C1_B9 = {'y': ('G1C1', 491198, 509162), 'x': ('G2C1', 880483, 897381), 'type': 'REVERSE_COMPLEMENT'}</li>
</ul>
</div>
<p></p>
<a name="H_Find%20Reversal%20Path"></a><h3>Find Reversal Path</h3>
<p><a name="BM_(Stories%5C%2FGenome%20Rearrangement%2FFind%20Reversal%20Path)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Stories%5C%2FGenome%20Rearrangement%2FFind%20Synteny%20Blocks)_TOPIC%2F">Stories/Genome Rearrangement/Find Synteny Blocks</a></li>
<li><a href="#BM_(Algorithms%5C%2FSynteny%2FReversal%20Path%2FBreakpoint%20Graph%20Algorithm)_TOPIC%2F">Algorithms/Synteny/Reversal Path/Breakpoint Graph Algorithm</a></li>
</ul>
</div>
<p>A reversal is the most common type of <a href="#BM_(genome%20rearrangement%7Cchromosomal%20rearrangement%7Cchromosome%20rearrangement)%2Fi">genome rearrangement</a> mutation: A segment of chromosome breaks off and ends up re-attaching, but with the ends swapped.</p>
<p><img src="svgbob_49b402ac745f2c2e4400c3b2de9a7ad449cb1c2b.svg" alt="Kroki diagram output"></p>
<p>The theory is that <a href="#BM_(genome%20rearrangement%7Cchromosomal%20rearrangement%7Cchromosome%20rearrangement)%2Fi">genome rearrangement</a>s between two species take the <a href="#BM_(parsimony%7Cparsimonious)%2Fi">parsimonious</a> path (or close to it). Since reversals are the most common form of <a href="#BM_(genome%20rearrangement%7Cchromosomal%20rearrangement%7Cchromosome%20rearrangement)%2Fi">genome rearrangement</a> mutation, by calculating a <a href="#BM_(parsimony%7Cparsimonious)%2Fi">parsimonious</a> reversal path (smallest set of reversals) it's possible to get an idea of how the two species branched off.</p>
<p>Note that there may be many <a href="#BM_(parsimony%7Cparsimonious)%2Fi">parsimonious</a> reversal paths between two <a href="#BM_(genome)%2Fi">genome</a>s with shared <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s.</p>
<p><img src="svgbob_035933ac760c4093244bf1755e35100274fda88b.svg" alt="Kroki diagram output"></p>
<p>Given a <a href="#BM_(parsimony%7Cparsimonious)%2Fi">parsimonious</a> reversal path, it may be that one of the <a href="#BM_(genome)%2Fi">genome</a>s in the reversal path is the parent species (or close to it).</p>
<p><img src="svgbob_e20200ac846a9d4bd80722ede4c92f15023e6797.svg" alt="Kroki diagram output"></p>
<p>In the example below, two species of the Mycoplasma bacteria are analyzed to find a <a href="#BM_(parsimony%7Cparsimonious)%2Fi">parsimonious</a> reversal path using the <a href="#BM_(breakpoint%20graph)_GR%2Fi">breakpoint graph</a> algorithm. The output reveals that only 1 reversal is responsible for the change in species. As such, it's very likely that one species broke off from the other rather than there being a shared parent species.</p>
<div style="border:1px solid black;">
<p></p>
<p>Solving a parsimonious reversal path for...</p>
<ul>
<li>k=20</li>
<li>cyclic=True</li>
<li>show_graph=True</li>
<li>genome1 reference: [[G1C1, /input/ch6_code/src/Mycoplasma bovis - GCA_000696015.1_ASM69601v1_genomic.fna.xz]]</li>
<li>genome2 reference: [[G2C1, /input/ch6_code/src/Mycoplasma agalactiae 14628 - GCA_000266865.1_ASM26686v1_genomic.fna.xz]]</li>
</ul>
<p>NOTE: Nucleotide codes that aren't ACGT get filtered out of the genomes.</p>
<p>Generating genomic dotplot...</p>
<p><img src="dotplot_aa18ecc0bccea2e4d491f745259e60c6.png" alt="Genomic Dot Plot"></p>
<p>Clustering genomic dotplot to snyteny graph...</p>
<ul>
<li>Merging radius=25 angle_half_maw=90...</li>
<li>Merging radius=50 angle_half_maw=90...</li>
<li>Merging radius=100 angle_half_maw=90...</li>
<li>Merging radius=200 angle_half_maw=90...</li>
<li>Merging radius=400 angle_half_maw=90...</li>
<li>Merging radius=800 angle_half_maw=90...</li>
<li>Merging radius=1600 angle_half_maw=90...</li>
<li>Merging radius=3200 angle_half_maw=90...</li>
<li>Filtering max_filter_length=3200.0 max_merge_distance=20000.0...</li>
<li>Merging radius=6400 angle_half_maw=90...</li>
<li>Merging radius=12800 angle_half_maw=90...</li>
<li>Culling below length=10000.0...</li>
</ul>
<p>Generating synteny graph...</p>
<p><img src="syntenygraph_aa18ecc0bccea2e4d491f745259e60c6.png" alt="Synteny Graph"></p>
<p>Mapping synteny graph matches to IDs using x-axis genome...</p>
<ul>
<li>G2C1_B0 = {'y': ('G1C1', 456122, 490875), 'x': ('G2C1', 26147, 55215), 'type': 'REVERSE_COMPLEMENT'}</li>
<li>G2C1_B1 = {'y': ('G1C1', 403833, 428051), 'x': ('G2C1', 55606, 78341), 'type': 'REVERSE_COMPLEMENT'}</li>
<li>G2C1_B2 = {'y': ('G1C1', 356150, 384124), 'x': ('G2C1', 88483, 113645), 'type': 'REVERSE_COMPLEMENT'}</li>
<li>G2C1_B3 = {'y': ('G1C1', 596687, 770257), 'x': ('G2C1', 119776, 270535), 'type': 'NORMAL'}</li>
<li>G2C1_B4 = {'y': ('G1C1', 772155, 835261), 'x': ('G2C1', 287183, 351652), 'type': 'NORMAL'}</li>
<li>G2C1_B5 = {'y': ('G1C1', 835398, 948473), 'x': ('G2C1', 385360, 498503), 'type': 'NORMAL'}</li>
<li>G2C1_B6 = {'y': ('G1C1', 90, 262351), 'x': ('G2C1', 498629, 739753), 'type': 'NORMAL'}</li>
<li>G2C1_B7 = {'y': ('G1C1', 276780, 355692), 'x': ('G2C1', 740077, 815603), 'type': 'NORMAL'}</li>
<li>G2C1_B8 = {'y': ('G1C1', 529629, 591336), 'x': ('G2C1', 816444, 878201), 'type': 'REVERSE_COMPLEMENT'}</li>
<li>G2C1_B9 = {'y': ('G1C1', 491197, 509162), 'x': ('G2C1', 880483, 897382), 'type': 'REVERSE_COMPLEMENT'}</li>
</ul>
<p>Generating permutations for genomes...</p>
<ul>
<li>genome1_permutations=[['+G2C1_B6', '+G2C1_B7', '-G2C1_B2', '-G2C1_B1', '-G2C1_B0', '-G2C1_B9', '-G2C1_B8', '+G2C1_B3', '+G2C1_B4', '+G2C1_B5']]</li>
<li>genome2_permutations=[['+G2C1_B0', '+G2C1_B1', '+G2C1_B2', '+G2C1_B3', '+G2C1_B4', '+G2C1_B5', '+G2C1_B6', '+G2C1_B7', '+G2C1_B8', '+G2C1_B9']]</li>
</ul>
<p>Generating reversal path on genomes that are cyclic=True...</p>
<ul>
<li>
<p>INITIAL red_p_list=[['+G2C1_B0', '+G2C1_B1', '+G2C1_B2', '-G2C1_B7', '-G2C1_B6', '-G2C1_B5', '-G2C1_B4', '-G2C1_B3', '+G2C1_B8', '+G2C1_B9']]</p>
<p><img src="dot_f8ab9badd4c1e04246b0acafcf6925a8.svg" alt="Dot diagram"></p>
</li>
<li>
<p>red_p_list=[['+G2C1_B0', '+G2C1_B1', '+G2C1_B2', '+G2C1_B3', '+G2C1_B4', '+G2C1_B5', '+G2C1_B6', '+G2C1_B7', '+G2C1_B8', '+G2C1_B9']]</p>
<p><img src="dot_da97992266966ec432f5101440600199.svg" alt="Dot diagram"></p>
</li>
</ul>
</div>
<p></p>
<a name="H_Evolutionary%20History"></a><h2>Evolutionary History</h2>
<p><a name="BM_(Stories%5C%2FEvolutionary%20History)_TOPIC%2F"></a><strong></strong></p>
<p>When scientists work with biological entities, those entities are either present day entities or relics of extinct entities (<a href="#BM_(paleontology%7Cpalaeontology%7Cpal%C3%A6ontology)%2Fi">paleontology</a>). In certain cases, it's reasonable to assume the shared ancestry of a set of present day entities by comparing their features to those of extinct relics. For example, ...</p>
<ul>
<li>modern day big cats (e.g. leopards and tigers) probably derived from saber-toothed tigers.</li>
<li>modern day elephants (e.g. african elephants and indian elephants) probably derived from mammoths.</li>
</ul>
<p><img src="svgbob_939dacd2d3d9f2ce35d2b21b8f8c6557c90034f2.svg" alt="Kroki diagram output"></p>
<p>In most cases however, there are no relics. For example, extinct viruses or bacteria typically don't leave much evidence around in the same way that ...</p>
<ul>
<li>extinct dinosaur species have their bones found underground.</li>
<li>extinct insect species are found encased in amber.</li>
<li>extinct arctic species are found encased in ice.</li>
</ul>
<p>In such cases, it's still possible to infer the evolutionary history of a set of present day species by comparing their features to see how diverged they are. Those features could be phenotypic features (e.g. behavioural or physical features) or molecular features (e.g. DNA <a href="#BM_(sequence)%2Fi">sequence</a>s, protein <a href="#BM_(sequence)%2Fi">sequence</a>s, organelles and other cell features).</p>
<p>The process of inferring evolutionary history by comparing features for divergence is called <a href="#BM_(phylogeny%7Cphylogenetic)%2Fi">phylogeny</a>. <a href="#BM_(phylogeny%7Cphylogenetic)%2Fi">Phylogeny</a> algorithms provide insight into ...</p>
<ul>
<li>the shared ancestry of a set of present day entities.</li>
<li>how similar each present day entity is to its ancestor.</li>
<li>what the features of the ancestor likely were.</li>
</ul>
<p><img src="svgbob_d1db0bb621930ac2127d7c814dbb9c537324c8f1.svg" alt="Kroki diagram output"></p>
<p>Oftentimes, <a href="#BM_(phylogeny%7Cphylogenetic)%2Fi">phylogeny</a> produces much more accurate results than simply eye-balling it (as was done in the initial example), but ultimately the quality of the result is dependent on what features are being measured and the metric used for measurement. Prior to <a href="#BM_(sequencing%7Csequenced)%2Fi">sequencing</a> technology, most <a href="#BM_(phylogeny%7Cphylogenetic)%2Fi">phylogeny</a> was done by comparing phenotypic features (e.g. <a href="#BM_(character%20table%7Ccharacter%20vector)%2Fi">character table</a>s). Common practice now is to use molecular features (e.g. DNA <a href="#BM_(sequencing%7Csequenced)%2Fi">sequencing</a>) since those have more information that's definitive and less biased (e.g. phenotypic features are subject to human interpretation).</p>
<a name="H_Find%20Evolutionary%20Tree"></a><h3>Find Evolutionary Tree</h3>
<p><a name="BM_(Stories%5C%2FEvolutionary%20History%2FFind%20Evolutionary%20Tree)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FDistance%20Matrix%20to%20Tree)_TOPIC%2F">Algorithms/Phylogeny/Distance Matrix to Tree</a></li>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FGlobal%20Alignment)_TOPIC%2F">Algorithms/Sequence Alignment/Global Alignment</a></li>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FProtein%20Scoring%2FBLOSUM%20Scoring%20Matrix)_TOPIC%2F">Algorithms/Sequence Alignment/Protein Scoring/BLOSUM Scoring Matrix</a></li>
</ul>
</div>
<p>Evolutionary history is often displayed as a <a href="#BM_(tree)%2Fi">tree</a> called a <a href="#BM_(phylogenetic%20tree%7Cphylogeny%20tree%7Cphylogenies%7Cevolutionary%20tree)%2Fi">phylogenetic tree</a>, where leaf nodes represent known entities and internal nodes represent inferred ancestor entities. Depending on the <a href="#BM_(phylogeny%7Cphylogenetic)%2Fi">phylogeny</a> algorithm used, the <a href="#BM_(tree)%2Fi">tree</a> may be either a <a href="#BM_(rooted%20tree)%2Fi">rooted tree</a> or an <a href="#BM_(un-rooted%20tree%7Cunrooted%20tree)%2Fi">unrooted tree</a>. The difference is that a <a href="#BM_(rooted%20tree)%2Fi">rooted tree</a> infers parent-child relationships of ancestors while an <a href="#BM_(un-rooted%20tree%7Cunrooted%20tree)%2Fi">unrooted tree</a> does not.</p>
<p><img src="svgbob_afb68d67568f9c3c2ef2342f2ef9f6b05b17fc74.svg" alt="Kroki diagram output"></p>
<p>In the example above, the <a href="#BM_(rooted%20tree)%2Fi">rooted tree</a> (left diagram) shows ancestors B and C as branching off (evolving) from their common ancestor A. The <a href="#BM_(un-rooted%20tree%7Cunrooted%20tree)%2Fi">unrooted tree</a> (right diagram) shows ancestors B and C but doesn't provide infer which branched off the other. It could be that ancestor B ultimately descended from C or vice versa.</p>
<p><a href="#BM_(SARS)%2F">SARS</a>-<a href="#BM_(COVID%7CCoV)%2F">CoV</a>-2 is the virus that causes <a href="#BM_(COVID%7CCoV)%2F">COVID</a>-19. The example below measures <a href="#BM_(SARS)%2F">SARS</a>-<a href="#BM_(COVID%7CCoV)%2F">CoV</a>-2 spike protein <a href="#BM_(sequence)%2Fi">sequence</a>s collected from different patients to produce its evolutionary history. The metric used to measure how diverged two <a href="#BM_(sequence)%2Fi">sequence</a>s are from each other is <a href="#BM_(global%20alignment%7Cglobal%20sequence%20alignment)%2Fi">global alignment</a> using a <a href="#BM_%5Cb(BLOSUM)%5Cd*%5Cb%2F">BLOSUM</a>80 scoring matrix. Once divergences (distances) are calculated, the <a href="#BM_(neighbour%20joining%20phylogeny)%2Fi">neighbour joining phylogeny</a> algorithm is used to generate a <a href="#BM_(phylogenetic%20tree%7Cphylogeny%20tree%7Cphylogenies%7Cevolutionary%20tree)%2Fi">phylogenetic tree</a>.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p><a href="#BM_%5Cb(BLOSUM)%5Cd*%5Cb%2F">BLOSUM</a>80 was chosen because <a href="#BM_(SARS)%2F">SARS</a>-<a href="#BM_(COVID%7CCoV)%2F">CoV</a>-2 is a relatively new virus (~2 years). I don't know if it was a good choice because I've been told viruses mutate more rapidly, so maybe <a href="#BM_%5Cb(BLOSUM)%5Cd*%5Cb%2F">BLOSUM</a>62 would have been a better choice.</p>
<p>The original NCBI dataset had 30k to 40k unique spike <a href="#BM_(sequence)%2Fi">sequence</a>s. I couldn't justify sticking all of that into the git repo (too large) so I whittled it down to a random sample of 1000.</p>
<p>From that 1000, only a small sample are selected to run the code. The problem is that <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a>s are computationally expensive and Python is slow. Doing a <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a> between two spike protein <a href="#BM_(sequence)%2Fi">sequence</a>s on my VM takes a long time (~4 seconds per alignment), so for the full 1000 <a href="#BM_(sequence)%2Fi">sequence</a>s the total running time would end up being ~4 years (if I calculated it correctly - single core).</p>
</div>
<div style="border:1px solid black;">
<p></p>
<p>Given a random sample of 6 sequences from 1000_unique_sarscov2_spike_seqs.json.xz and the following alignment weights...</p>
<pre><code>   A  R  N  D  C  Q  E  G  H  I  L  K  M  F  P  S  T  W  Y  V  B  J  Z  X  *
A  5 -2 -2 -2 -1 -1 -1  0 -2 -2 -2 -1 -1 -3 -1  1  0 -3 -2  0 -2 -2 -1 -1 -6
R -2  6 -1 -2 -4  1 -1 -3  0 -3 -3  2 -2 -4 -2 -1 -1 -4 -3 -3 -1 -3  0 -1 -6
N -2 -1  6  1 -3  0 -1 -1  0 -4 -4  0 -3 -4 -3  0  0 -4 -3 -4  5 -4  0 -1 -6
D -2 -2  1  6 -4 -1  1 -2 -2 -4 -5 -1 -4 -4 -2 -1 -1 -6 -4 -4  5 -5  1 -1 -6
C -1 -4 -3 -4  9 -4 -5 -4 -4 -2 -2 -4 -2 -3 -4 -2 -1 -3 -3 -1 -4 -2 -4 -1 -6
Q -1  1  0 -1 -4  6  2 -2  1 -3 -3  1  0 -4 -2  0 -1 -3 -2 -3  0 -3  4 -1 -6
E -1 -1 -1  1 -5  2  6 -3  0 -4 -4  1 -2 -4 -2  0 -1 -4 -3 -3  1 -4  5 -1 -6
G  0 -3 -1 -2 -4 -2 -3  6 -3 -5 -4 -2 -4 -4 -3 -1 -2 -4 -4 -4 -1 -5 -3 -1 -6
H -2  0  0 -2 -4  1  0 -3  8 -4 -3 -1 -2 -2 -3 -1 -2 -3  2 -4 -1 -4  0 -1 -6
I -2 -3 -4 -4 -2 -3 -4 -5 -4  5  1 -3  1 -1 -4 -3 -1 -3 -2  3 -4  3 -4 -1 -6
L -2 -3 -4 -5 -2 -3 -4 -4 -3  1  4 -3  2  0 -3 -3 -2 -2 -2  1 -4  3 -3 -1 -6
K -1  2  0 -1 -4  1  1 -2 -1 -3 -3  5 -2 -4 -1 -1 -1 -4 -3 -3 -1 -3  1 -1 -6
M -1 -2 -3 -4 -2  0 -2 -4 -2  1  2 -2  6  0 -3 -2 -1 -2 -2  1 -3  2 -1 -1 -6
F -3 -4 -4 -4 -3 -4 -4 -4 -2 -1  0 -4  0  6 -4 -3 -2  0  3 -1 -4  0 -4 -1 -6
P -1 -2 -3 -2 -4 -2 -2 -3 -3 -4 -3 -1 -3 -4  8 -1 -2 -5 -4 -3 -2 -4 -2 -1 -6
S  1 -1  0 -1 -2  0  0 -1 -1 -3 -3 -1 -2 -3 -1  5  1 -4 -2 -2  0 -3  0 -1 -6
T  0 -1  0 -1 -1 -1 -1 -2 -2 -1 -2 -1 -1 -2 -2  1  5 -4 -2  0 -1 -1 -1 -1 -6
W -3 -4 -4 -6 -3 -3 -4 -4 -3 -3 -2 -4 -2  0 -5 -4 -4 11  2 -3 -5 -3 -3 -1 -6
Y -2 -3 -3 -4 -3 -2 -3 -4  2 -2 -2 -3 -2  3 -4 -2 -2  2  7 -2 -3 -2 -3 -1 -6
V  0 -3 -4 -4 -1 -3 -3 -4 -4  3  1 -3  1 -1 -3 -2  0 -3 -2  4 -4  2 -3 -1 -6
B -2 -1  5  5 -4  0  1 -1 -1 -4 -4 -1 -3 -4 -2  0 -1 -5 -3 -4  5 -4  0 -1 -6
J -2 -3 -4 -5 -2 -3 -4 -5 -4  3  3 -3  2  0 -4 -3 -1 -3 -2  2 -4  3 -3 -1 -6
Z -1  0  0  1 -4  4  5 -3  0 -4 -3  1 -1 -4 -2  0 -1 -3 -3 -3  0 -3  5 -1 -6
X -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -6
* -6 -6 -6 -6 -6 -6 -6 -6 -6 -6 -6 -6 -6 -6 -6 -6 -6 -6 -6 -6 -6 -6 -6 -6  1
</code></pre>
<p>The tree generated by neighbour joining phylogeny is (distances measured using global alignment, edge lengths scaled to 0.1) ...</p>
<p><img src="dot_0232d175f5f7d43bab86f93f5c44847a.svg" alt="Dot diagram"></p>
</div>
<p></p>
<a name="H_Find%20Ancestral%20Features"></a><h3>Find Ancestral Features</h3>
<p><a name="BM_(Stories%5C%2FEvolutionary%20History%2FFind%20Ancestral%20Features)_TOPIC%2F"></a><strong></strong></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BM_(Stories%5C%2FEvolutionary%20History%2FFind%20Evolutionary%20Tree)_TOPIC%2F">Stories/Evolutionary History/Find Evolutionary Tree</a></li>
<li><a href="#BM_(Algorithms%5C%2FSequence%20Alignment%2FMultiple%20Alignment)_TOPIC%2F">Algorithms/Sequence Alignment/Multiple Alignment</a></li>
<li><a href="#BM_(Algorithms%5C%2FPhylogeny%2FSequence%20Inference)_TOPIC%2F">Algorithms/Phylogeny/Sequence Inference</a></li>
</ul>
</div>
<p>An unknown ancestor's features are probabilistically inferrable via the features of entities that descend from it.</p>
<p><img src="svgbob_301dc5669f5493f72306ed61f5f34720826895f1.svg" alt="Kroki diagram output"></p>
<p>The example above infers phenotypic features for the common ancestor of leopard and tiger. If a feature is present and the same in both, it's safe to assume that it was present in their common ancestor as well (e.g. 4 legs). Otherwise, there's still some smaller chance that the feature was present, possibly with some variability in how it manifested (e.g. type of coat pattern).</p>
<p>With the advent of <a href="#BM_(sequencing%7Csequenced)%2Fi">sequencing</a> technology, the practice of using phenotypic features for <a href="#BM_(phylogeny%7Cphylogenetic)%2Fi">phylogeny</a> was superseded by <a href="#BM_(sequencing%7Csequenced)%2Fi">sequencing</a> data. When <a href="#BM_(sequence)%2Fi">sequence</a>s are used, the features are the <a href="#BM_(sequence)%2Fi">sequence</a>s themselves, meaning that the <a href="#BM_(sequence)%2Fi">sequence</a> of the common ancestor is what gets inferred.</p>
<p><img src="svgbob_250f42ea364b8fe4b11cb8bd97fb264f2049daf9.svg" alt="Kroki diagram output"></p>
<p>The example below infers the spike protein <a href="#BM_(sequence)%2Fi">sequence</a>s for the ancestors of <a href="#BM_(SARS)%2F">SARS</a>-<a href="#BM_(COVID%7CCoV)%2F">CoV</a>-2 variants. First, a <a href="#BM_(phylogenetic%20tree%7Cphylogeny%20tree%7Cphylogenies%7Cevolutionary%20tree)%2Fi">phylogenetic tree</a> is generated using <a href="#BM_%5Cb(BLOSUM)%5Cd*%5Cb%2F">BLOSUM</a>80 as the <a href="#BM_(distance%20metric)%2Fi">distance metric</a>. Then, the <a href="#BM_(sequence)%2Fi">sequence</a>s are all aligned together using <a href="#BM_%5Cb(BLOSUM)%5Cd*%5Cb%2F">BLOSUM</a>80 (multiple alignment, not pairwise alignment as was used for the <a href="#BM_(distance%20metric)%2Fi">distance metric</a>). The <a href="#BM_(sequence)%2Fi">sequence</a>s of ancestors are inferred using those aligned <a href="#BM_(sequence)%2Fi">sequence</a>s.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>This is badly cobbled together code. It's taking the code from the previous section's example and embedding/duct-taping even more pieces from the <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a> module just to get a running example. In a perfect world I would just import the <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a> module, but that module lives in a separate container. This is the best I can do.</p>
<p>Running this is even slower than the previous section's example, so the sample size has been reduced even further.</p>
</div>
<div style="border:1px solid black;">
<p></p>
<p>Given a random sample of 3 sequences from 1000_unique_sarscov2_spike_seqs.json.xz and the following alignment weights...</p>
<pre><code>   A  R  N  D  C  Q  E  G  H  I  L  K  M  F  P  S  T  W  Y  V  B  J  Z  X
A  5 -2 -2 -2 -1 -1 -1  0 -2 -2 -2 -1 -1 -3 -1  1  0 -3 -2  0 -2 -2 -1 -1
R -2  6 -1 -2 -4  1 -1 -3  0 -3 -3  2 -2 -4 -2 -1 -1 -4 -3 -3 -1 -3  0 -1
N -2 -1  6  1 -3  0 -1 -1  0 -4 -4  0 -3 -4 -3  0  0 -4 -3 -4  5 -4  0 -1
D -2 -2  1  6 -4 -1  1 -2 -2 -4 -5 -1 -4 -4 -2 -1 -1 -6 -4 -4  5 -5  1 -1
C -1 -4 -3 -4  9 -4 -5 -4 -4 -2 -2 -4 -2 -3 -4 -2 -1 -3 -3 -1 -4 -2 -4 -1
Q -1  1  0 -1 -4  6  2 -2  1 -3 -3  1  0 -4 -2  0 -1 -3 -2 -3  0 -3  4 -1
E -1 -1 -1  1 -5  2  6 -3  0 -4 -4  1 -2 -4 -2  0 -1 -4 -3 -3  1 -4  5 -1
G  0 -3 -1 -2 -4 -2 -3  6 -3 -5 -4 -2 -4 -4 -3 -1 -2 -4 -4 -4 -1 -5 -3 -1
H -2  0  0 -2 -4  1  0 -3  8 -4 -3 -1 -2 -2 -3 -1 -2 -3  2 -4 -1 -4  0 -1
I -2 -3 -4 -4 -2 -3 -4 -5 -4  5  1 -3  1 -1 -4 -3 -1 -3 -2  3 -4  3 -4 -1
L -2 -3 -4 -5 -2 -3 -4 -4 -3  1  4 -3  2  0 -3 -3 -2 -2 -2  1 -4  3 -3 -1
K -1  2  0 -1 -4  1  1 -2 -1 -3 -3  5 -2 -4 -1 -1 -1 -4 -3 -3 -1 -3  1 -1
M -1 -2 -3 -4 -2  0 -2 -4 -2  1  2 -2  6  0 -3 -2 -1 -2 -2  1 -3  2 -1 -1
F -3 -4 -4 -4 -3 -4 -4 -4 -2 -1  0 -4  0  6 -4 -3 -2  0  3 -1 -4  0 -4 -1
P -1 -2 -3 -2 -4 -2 -2 -3 -3 -4 -3 -1 -3 -4  8 -1 -2 -5 -4 -3 -2 -4 -2 -1
S  1 -1  0 -1 -2  0  0 -1 -1 -3 -3 -1 -2 -3 -1  5  1 -4 -2 -2  0 -3  0 -1
T  0 -1  0 -1 -1 -1 -1 -2 -2 -1 -2 -1 -1 -2 -2  1  5 -4 -2  0 -1 -1 -1 -1
W -3 -4 -4 -6 -3 -3 -4 -4 -3 -3 -2 -4 -2  0 -5 -4 -4 11  2 -3 -5 -3 -3 -1
Y -2 -3 -3 -4 -3 -2 -3 -4  2 -2 -2 -3 -2  3 -4 -2 -2  2  7 -2 -3 -2 -3 -1
V  0 -3 -4 -4 -1 -3 -3 -4 -4  3  1 -3  1 -1 -3 -2  0 -3 -2  4 -4  2 -3 -1
B -2 -1  5  5 -4  0  1 -1 -1 -4 -4 -1 -3 -4 -2  0 -1 -5 -3 -4  5 -4  0 -1
J -2 -3 -4 -5 -2 -3 -4 -5 -4  3  3 -3  2  0 -4 -3 -1 -3 -2  2 -4  3 -3 -1
Z -1  0  0  1 -4  4  5 -3  0 -4 -3  1 -1 -4 -2  0 -1 -3 -3 -3  0 -3  5 -1
X -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1
</code></pre>
<p>INDEL=-6.0</p>
<p>The tree generated by neighbour joining phylogeny ALONG WITH INFERRED ANCESTRAL SEQUENCES is (distances measured using global alignment, edge lengths scaled to 0.1) ...</p>
<p><img src="dot_efed26f7f13639f895adeed724a5c3a9.svg" alt="Dot diagram"></p>
</div>
<p></p>
<a name="H_Ideas"></a><h1>Ideas</h1>
<ul>
<li>
<p>CPU optimized C++ <a href="#BM_(global%20alignment%7Cglobal%20sequence%20alignment)%2Fi">global alignment</a> - Simple <a href="#BM_(global%20alignment%7Cglobal%20sequence%20alignment)%2Fi">global alignment</a> is C++ with all optimizations turned on AND multi-threading or fibers that optimize work size to fit in cache lines.</p>
</li>
<li>
<p>GPU optimized C++ <a href="#BM_(global%20alignment%7Cglobal%20sequence%20alignment)%2Fi">global alignment</a> - Simple <a href="#BM_(global%20alignment%7Cglobal%20sequence%20alignment)%2Fi">global alignment</a> in Nvidia's HPC SDK C++ where GPU "thread" is optimized to fit in caches. Maybe do the divide-and-conquer variant as well. (divide and conquer might be a good idea because it'll work on super fat <a href="#BM_(sequence)%2Fi">sequence</a>s)</p>
</li>
<li>
<p>GPU optimized C++ probabilistic multiple alignment - Probabilistic multiple alignment in Nvidia's HPC SDK C++ where GPU "thread" is optimized to fit in caches.</p>
</li>
<li>
<p>Deep-learning <a href="#BM_(regulatory%20motif)%2Fi">Regulatory Motif</a> Detection - Try training a deep learning model to "find" <a href="#BM_(regulatory%20motif)%2Fi">regulatory motif</a>s for new <a href="#BM_(transcription%20factor)%2Fi">transcription factor</a>s based on past training data.</p>
</li>
<li>
<p><a href="#BM_(global%20alignment%7Cglobal%20sequence%20alignment)%2Fi">Global alignment</a> that takes <a href="#BM_(genome%20rearrangement%7Cchromosomal%20rearrangement%7Cchromosome%20rearrangement)%2Fi">genome rearrangement</a>s into account - multiple chromosomes, chromosomes becoming circular or linear, reversals, <a href="#BM_(fission)%2Fi">fission</a>s, <a href="#BM_(%5Cbfusion%7C%5Cbfuse)%2Fi">fusion</a>s, copies, etc..</p>
</li>
<li>
<p>Organism lookup by <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> - Two-tiered database containing <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s. The first tier is an "inverse index" of <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s that rarely appear across all organisms (unique or almost unique to the <a href="#BM_(genome)%2Fi">genome</a>) exposed as either a trie / hashtable (for exact lookups) or possibly as a list where highly optimized miniature alignments get performed (for fuzzy lookups -- SIMD + things fit nicely into cache lines). It widdles down the list of organism for the second tier, which is a full on database search for each matches across all <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s.</p>
</li>
<li>
<p><a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">K-mer</a> <a href="#BM_(hierarchial%20cluster)%2Fi">hierarchial cluster</a>ing - Hierarchical <a href="#BM_(cluster)%2Fi">cluster</a> together similar <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s using either <a href="#BM_(pearson%20correlation%20coefficient%7Cpearson%20correlation%7Cpearson%20distance%7Cpearson%20similarity)%2Fi">pearson similarity</a>/<a href="#BM_(pearson%20correlation%20coefficient%7Cpearson%20correlation%7Cpearson%20distance%7Cpearson%20similarity)%2Fi">pearson distance</a> [between one/zero vector of sub-<a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s] or <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a> distance to form its <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> / <a href="#BM_(similarity%20matrix%7Csimilarity%20matrices)%2Fi">similarity matrix</a>. This is useful for when you're trying to identify which organism a <a href="#BM_(sequence)%2Fi">sequence</a> belongs to by searching for its <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s in a database. The <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s that make up the database would be <a href="#BM_(cluster)%2Fi">cluster</a>ed, and <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s that closely <a href="#BM_(cluster)%2Fi">cluster</a> together under a branch of the <a href="#BM_(hierarchial%20cluster)%2Fi">hierarchial cluster</a> <a href="#BM_(tree)%2Fi">tree</a> are those you'd be more cautious with -- the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> may have matched but it could have actually been a corrupted form of one of the other <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s in the <a href="#BM_(cluster)%2Fi">cluster</a> (<a href="#BM_(sequencing%20error)%2Fi">sequencing error</a>).</p>
<p>This logic also applies to spell checking. Words that <a href="#BM_(cluster)%2Fi">cluster</a> together closely are more likely to be mis-identified by a standard spellchecker, meaning individual <a href="#BM_(cluster)%2Fi">cluster</a>s should have their own spell checking strategies? If you're going to do this with words, use a factor in QWERTY keyboard key distances into the similarity / <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a>.</p>
</li>
<li>
<p><a href="#BM_(hierarchial%20cluster)%2Fi">Hierarchial cluster</a>ing explorer - Generate a <a href="#BM_(neighbour%20joining%20phylogeny)%2Fi">neighbour joining phylogeny</a> <a href="#BM_(tree)%2Fi">tree</a> based on <a href="#BM_(pearson%20correlation%20coefficient%7Cpearson%20correlation%7Cpearson%20distance%7Cpearson%20similarity)%2Fi">pearson distance</a> of <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a> distance, then visualize the <a href="#BM_(tree)%2Fi">tree</a> and provide the user with "interesting" internal nodes (<a href="#BM_(cluster)%2Fi">cluster</a>s). In this case, "interesting" would be any internal node where the distance to most leaf nodes is within some threshold / average / variance / etc... Also, maybe provide an "idealized" view of the <a href="#BM_(cluster)%2Fi">cluster</a>ed data for each internal node (e.g. average the vectors for the leaves to produce the vector for the internal node).</p>
<p>Another idea is to take the generated <a href="#BM_(tree)%2Fi">tree</a> and convert it back into <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a>. If the data isn't junk, the <a href="#BM_(distance%20metric)%2Fi">distance metric</a> isn't junk, and the data is <a href="#BM_(cluster)%2Fi">cluster</a>able on that <a href="#BM_(distance%20metric)%2Fi">distance metric</a>, the generated <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> should match closely to the input <a href="#BM_(distance%20metric)%2Fi">distance metric</a>. The tool can warn the user if it doesn't.</p>
</li>
<li>
<p>Soft <a href="#BM_(hierarchial%20cluster)%2Fi">hierarchial cluster</a>ing - Build out a <a href="#BM_(neighbour%20joining%20phylogeny)%2Fi">neighbour joining phylogeny</a> <a href="#BM_(tree)%2Fi">tree</a>. Each internal node is a <a href="#BM_(cluster)%2Fi">cluster</a>. The distance between that internal node to all leaf nodes can be used to define the probability that the leaf node belongs to that <a href="#BM_(cluster)%2Fi">cluster</a>? This makes sense because <a href="#BM_(neighbour%20joining%20phylogeny)%2Fi">neighbour joining phylogeny</a> produces <a href="#BM_(un-rooted%20tree%7Cunrooted%20tree)%2Fi">unrooted tree</a>s (<a href="#BM_(simple%20tree)%2Fi">simple tree</a>s). If it were a <a href="#BM_(rooted%20tree)%2Fi">rooted tree</a>, you could say that internal node X leaf nodes A, B, and C -- meaning that A, B and C are members of <a href="#BM_(cluster)%2Fi">cluster</a> X. But, because it's unrooted, technically any leaf node in the graph could be a member of <a href="#BM_(cluster)%2Fi">cluster</a> X.</p>
<p>This relates to the idea above (<a href="#BM_(hierarchial%20cluster)%2Fi">hierarchial cluster</a>ing explorer) -- You can identify "interesting <a href="#BM_(cluster)%2Fi">cluster</a>s" using this (e.g. a small group tightly <a href="#BM_(cluster)%2Fi">cluster</a>ed together) and return it to the user for inspection.</p>
</li>
<li>
<p><a href="#BM_(hierarchial%20cluster)%2Fi">Hierarchial cluster</a>ing as a means of detecting outliers - <a href="#BM_(cluster)%2Fi">Cluster</a> data using <a href="#BM_(neighbour)%2Fi">neighbour</a>ing join <a href="#BM_(phylogeny%7Cphylogenetic)%2Fi">phylogeny</a>. How far is each leaf node to its parent internal node? Find any that are grossly over the average / <a href="#BM_(k-means%3F%7C%5Cd%2B-means%3F%7Csquared%20error%20distortion)%2Fi">squared error distortion</a> / some other metric? Report it. Try other ways as well (e.g. pick a root and see how far it is from the root -- root picked using some metric like avg distance between leaf nodes / 2 or <a href="#BM_(k-means%3F%7C%5Cd%2B-means%3F%7Csquared%20error%20distortion)%2Fi">squared error distortion</a>).</p>
<p>This relates to the idea above (soft <a href="#BM_(hierarchial%20cluster)%2Fi">hierarchial cluster</a>ing) -- You may be able to identify outliers using soft <a href="#BM_(hierarchial%20cluster)%2Fi">hierarchial cluster</a>ing using this (e.g. the probability of being a part of some internal node is way farther than any of the other leaf nodes).</p>
</li>
</ul>
<a name="H_Terminology"></a><h1>Terminology</h1>
<ul>
<li>
<p><a name="BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi"></a><strong>k-mer</strong> - A <a href="#BM_(substring)%2Fi">substring</a> of length k within some larger biological <a href="#BM_(sequence)%2Fi">sequence</a> (e.g. DNA or <a href="#BM_(amino%20acid)%2Fi">amino acid</a> chain). For example, in the DNA <a href="#BM_(sequence)%2Fi">sequence</a> GAAATC, the following <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>'s exist:</p>
<table>
<thead>
<tr>
<th>k</th>
<th><a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s</th>
</tr>
</thead>
<tbody>
<tr>
<td>1</td>
<td>G A A A T C</td>
</tr>
<tr>
<td>2</td>
<td>GA AA AA AT TC</td>
</tr>
<tr>
<td>3</td>
<td>GAA AAA AAT ATC</td>
</tr>
<tr>
<td>4</td>
<td>GAAA AAAT AATC</td>
</tr>
<tr>
<td>5</td>
<td>GAAAT AAATC</td>
</tr>
<tr>
<td>6</td>
<td>GAAATC</td>
</tr>
</tbody>
</table>
</li>
<li>
<p><a name="BM_(%5C(%5Cd%2B%2C%5Cs*%5Cd%2B%5C)-mer%7Ckd-mer%7Ckdmer%7C%5C(k%2C%5Cs*d%5C)-mer)%2Fi"></a><strong>kd-mer</strong> - A <a href="#BM_(substring)%2Fi">substring</a> of length 2k + d within some larger biological <a href="#BM_(sequence)%2Fi">sequence</a> (e.g. DNA or <a href="#BM_(amino%20acid)%2Fi">amino acid</a> chain) where the first k elements and the last k elements are known but the d elements in between isn't known.</p>
<p>When identifying a <a href="#BM_(%5C(%5Cd%2B%2C%5Cs*%5Cd%2B%5C)-mer%7Ckd-mer%7Ckdmer%7C%5C(k%2C%5Cs*d%5C)-mer)%2Fi">kd-mer</a> with a specific k and d, the proper syntax is <a href="#BM_(%5C(%5Cd%2B%2C%5Cs*%5Cd%2B%5C)-mer%7Ckd-mer%7Ckdmer%7C%5C(k%2C%5Cs*d%5C)-mer)%2Fi">(k, d)-mer</a>. For example, <a href="#BM_(%5C(%5Cd%2B%2C%5Cs*%5Cd%2B%5C)-mer%7Ckd-mer%7Ckdmer%7C%5C(k%2C%5Cs*d%5C)-mer)%2Fi">(1, 2)-mer</a> represents a <a href="#BM_(%5C(%5Cd%2B%2C%5Cs*%5Cd%2B%5C)-mer%7Ckd-mer%7Ckdmer%7C%5C(k%2C%5Cs*d%5C)-mer)%2Fi">kd-mer</a> with k=1 and d=2. In the DNA <a href="#BM_(sequence)%2Fi">sequence</a> GAAATC, the following <a href="#BM_(%5C(%5Cd%2B%2C%5Cs*%5Cd%2B%5C)-mer%7Ckd-mer%7Ckdmer%7C%5C(k%2C%5Cs*d%5C)-mer)%2Fi">(1, 2)-mer</a>'s exist: <code>G--A</code>, <code>A--T</code>, <code>A--C</code>.</p>
<p>See <a href="#BM_(read-pair%7Cread%20pair)%2Fi">read-pair</a>.</p>
</li>
<li>
<p><a name="BM_(5')%2Fi"></a><strong>5'</strong> (<a name="BM_(5%20prime)%2Fi"></a><strong>5 prime</strong>) / <a name="BM_(3')%2Fi"></a><strong>3'</strong> (<a name="BM_(3%20prime)%2Fi"></a><strong>3 prime</strong>) - <a href="#BM_(5')%2Fi">5'</a> (<a href="#BM_(5%20prime)%2Fi">5 prime</a>) and <a href="#BM_(3')%2Fi">3'</a> (<a href="#BM_(3%20prime)%2Fi">3 prime</a>) describe the opposite ends of DNA. The chemical structure at each end is what defines if it's <a href="#BM_(5')%2Fi">5'</a> or <a href="#BM_(3')%2Fi">3'</a> -- each end is guaranteed to be different from the other. The forward direction on DNA is defined as <a href="#BM_(5')%2Fi">5'</a> to <a href="#BM_(3')%2Fi">3'</a>, while the backwards direction is <a href="#BM_(3')%2Fi">3'</a> to <a href="#BM_(5')%2Fi">5'</a>.</p>
<p>Two complementing DNA strands will always be attached in opposite directions.</p>
<p><img src="svgbob_a2c87a8f72a4ea21f474350dc6a429e386ae8222.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a name="BM_(DNA%20polymerase)%2Fi"></a><strong>DNA polymerase</strong> - An enzyme that replicates a strand of DNA. That is, <a href="#BM_(DNA%20polymerase)%2Fi">DNA polymerase</a> walks over a single strand of DNA bases (not the strand of base pairs) and  generates a strand of complements. Before <a href="#BM_(DNA%20polymerase)%2Fi">DNA polymerase</a> can attach itself and start replicating DNA, it requires a <a href="#BM_(primer)%2Fi">primer</a>.</p>
<p><img src="svgbob_ac1f3f5c9441e395290e542abbc9256bed49302f.svg" alt="Kroki diagram output"></p>
<p><a href="#BM_(DNA%20polymerase)%2Fi">DNA polymerase</a> is unidirectional, meaning that it can only walk a DNA strand in one direction: reverse (<a href="#BM_(3')%2Fi">3'</a> to <a href="#BM_(5')%2Fi">5'</a>)</p>
</li>
<li>
<p><a name="BM_(primer)%2Fi"></a><strong>primer</strong> - A <a href="#BM_(primer)%2Fi">primer</a> is a short strand of RNA that binds to some larger strand of DNA (single bases, not a strand of base pairs) and allows DNA synthesis to  happen. That is, the <a href="#BM_(primer)%2Fi">primer</a> acts as the entry point for special enzymes <a href="#BM_(DNA%20polymerase)%2Fi">DNA polymerase</a>s. <a href="#BM_(DNA%20polymerase)%2Fi">DNA polymerase</a>s bind to the <a href="#BM_(primer)%2Fi">primer</a> to get access to the strand.</p>
</li>
<li>
<p><a name="BM_(replication%20fork)%2Fi"></a><strong>replication fork</strong> - The process of DNA replication requires that DNA's 2 complementing strands be unwound and split open. The area where the DNA starts to  split is called the <a href="#BM_(replication%20fork)%2Fi">replication fork</a>. In bacteria, the <a href="#BM_(replication%20fork)%2Fi">replication fork</a> starts at the <a href="#BM_(replication%20origin)%2Fi">replication origin</a> and keeps expanding until it reaches the <a href="#BM_(replication%20terminus)%2Fi">replication terminus</a>.  Special enzymes called <a href="#BM_(DNA%20polymerase)%2Fi">DNA polymerase</a>s walk over each unwound strand and create complementing strands.</p>
<p><img src="svgbob_fc7cf2c734fa0cab15b5746e25ad07e25f85ec64.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a name="BM_(replication%20origin)%2Fi"></a><strong>replication origin</strong> (<a name="BM_%5Cb(ori)%5Cb%2Fi"></a><strong>ori</strong>) - The point in DNA at which replication starts.</p>
</li>
<li>
<p><a name="BM_(replication%20terminus)%2Fi"></a><strong>replication terminus</strong> (<a name="BM_%5Cb(ter)%5Cb%2Fi"></a><strong>ter</strong>) - The point in DNA at which replication ends.</p>
</li>
<li>
<p><a name="BM_(forward%20half-strand)%2Fi"></a><strong>forward half-strand</strong> / <a name="BM_(reverse%20half-strand%7Cbackward%20half-strand%7Cbackwards%20half-strand)%2Fi"></a><strong>reverse half-strand</strong> - Bacteria are known to have a single chromosome of circular / looping DNA. In this DNA, the <a href="#BM_(replication%20origin)%2Fi">replication origin</a> (<a href="#BM_%5Cb(ori)%5Cb%2Fi">ori</a>) is the region of DNA where replication starts, while the <a href="#BM_(replication%20terminus)%2Fi">replication terminus</a> (<a href="#BM_%5Cb(ter)%5Cb%2Fi">ter</a>) is where replication ends.</p>
<p><img src="svgbob_fb9438945a109c884df7ccc79d1d8051fd358a87.svg" alt="Kroki diagram output"></p>
<p>If you split up the DNA based on <a href="#BM_%5Cb(ori)%5Cb%2Fi">ori</a> and <a href="#BM_%5Cb(ter)%5Cb%2Fi">ter</a> being cutting points, you end up with 4 distinct strands. Given that the direction of a strand is <a href="#BM_(5')%2Fi">5'</a> to <a href="#BM_(3')%2Fi">3'</a>, if the direction of the strand starts at...</p>
<ul>
<li>
<p><a href="#BM_%5Cb(ori)%5Cb%2Fi">ori</a> and ends at <a href="#BM_%5Cb(ter)%5Cb%2Fi">ter</a>, it's called the <a href="#BM_(forward%20half-strand)%2Fi">forward half-strand</a>.</p>
<p><img src="svgbob_d97cf1f12de4bd77dbaf9eca708b0e9120e50b33.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a href="#BM_%5Cb(ter)%5Cb%2Fi">ter</a> and ends at <a href="#BM_%5Cb(ori)%5Cb%2Fi">ori</a>, it's called the <a href="#BM_(reverse%20half-strand%7Cbackward%20half-strand%7Cbackwards%20half-strand)%2Fi">reverse half-strand</a>.</p>
<p><img src="svgbob_0016146d0b3f1f6edf2724e9978d52de7057f5f4.svg" alt="Kroki diagram output"></p>
</li>
</ul>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<ul>
<li><a href="#BM_(forward%20half-strand)%2Fi">Forward half-strand</a> is the same as <a href="#BM_(lagging%20half-strand)%2Fi">lagging half-strand</a>.</li>
<li><a href="#BM_(reverse%20half-strand%7Cbackward%20half-strand%7Cbackwards%20half-strand)%2Fi">Reverse half-strand</a> is the same as <a href="#BM_(leading%20half-strand)%2Fi">leading half-strand</a>.</li>
</ul>
</div>
</li>
<li>
<p><a name="BM_(leading%20half-strand)%2Fi"></a><strong>leading half-strand</strong> / <a name="BM_(lagging%20half-strand)%2Fi"></a><strong>lagging half-strand</strong> - Given the 2 strands that make up a DNA molecule, the strand that goes in the...</p>
<ul>
<li>reverse direction (<a href="#BM_(3')%2Fi">3'</a> to <a href="#BM_(5')%2Fi">5'</a>) is called the <a href="#BM_(leading%20half-strand)%2Fi">leading half-strand</a>.</li>
<li>forward direction (<a href="#BM_(5')%2Fi">5'</a> to <a href="#BM_(3')%2Fi">3'</a>) is called the <a href="#BM_(lagging%20half-strand)%2Fi">lagging half-strand</a>.</li>
</ul>
<p>This nomenclature has to do with <a href="#BM_(DNA%20polymerase)%2Fi">DNA polymerase</a>. Since <a href="#BM_(DNA%20polymerase)%2Fi">DNA polymerase</a> can only walk in the reverse direction (<a href="#BM_(3')%2Fi">3'</a> to <a href="#BM_(5')%2Fi">5'</a>), it synthesizes the <a href="#BM_(leading%20half-strand)%2Fi">leading half-strand</a> in one shot. For the <a href="#BM_(lagging%20half-strand)%2Fi">lagging half-strand</a> (<a href="#BM_(5')%2Fi">5'</a> to <a href="#BM_(3')%2Fi">3'</a>), multiple <a href="#BM_(DNA%20polymerase)%2Fi">DNA polymerase</a>s have to used to synthesize DNA, each binding to the lagging strand and walking backwards a small amount to generate a small fragment of DNA (<a href="#BM_(Okazaki%20fragment)%2Fi">Okazaki fragment</a>). the process is much slower for the <a href="#BM_(lagging%20half-strand)%2Fi">lagging half-strand</a>, that's why it's called lagging.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<ul>
<li><a href="#BM_(leading%20half-strand)%2Fi">Leading half-strand</a> is the same as <a href="#BM_(reverse%20half-strand%7Cbackward%20half-strand%7Cbackwards%20half-strand)%2Fi">reverse half-strand</a>.</li>
<li><a href="#BM_(lagging%20half-strand)%2Fi">Lagging half-strand</a> is the same as <a href="#BM_(forward%20half-strand)%2Fi">forward half-strand</a>.</li>
</ul>
</div>
</li>
<li>
<p><a name="BM_(Okazaki%20fragment)%2Fi"></a><strong>Okazaki fragment</strong> - A small fragment of DNA generated by <a href="#BM_(DNA%20polymerase)%2Fi">DNA polymerase</a> for <a href="#BM_(forward%20half-strand)%2Fi">forward half-strand</a>s. DNA synthesis for the <a href="#BM_(forward%20half-strand)%2Fi">forward half-strand</a>s can only happen in small pieces. As the fork open ups every ~2000 nucleotides, <a href="#BM_(DNA%20polymerase)%2Fi">DNA polymerase</a> attaches to the end of the fork on the <a href="#BM_(forward%20half-strand)%2Fi">forward half-strand</a> and walks in reverse to generate that small segment (<a href="#BM_(DNA%20polymerase)%2Fi">DNA polymerase</a> can only walk in the reverse direction).</p>
</li>
<li>
<p><a name="BM_(DNA%20ligase)%2Fi"></a><strong>DNA ligase</strong> - An enzyme that sews together short segments of DNA called <a href="#BM_(Okazaki%20fragment)%2Fi">Okazaki fragment</a>s by binding the phosphate group on the end of one strand with the deoxyribose group on the other strand.</p>
</li>
<li>
<p><a name="BM_(DnaA%20box)%2Fi"></a><strong>DnaA box</strong> - A <a href="#BM_(sequence)%2Fi">sequence</a> in the <a href="#BM_%5Cb(ori)%5Cb%2Fi">ori</a> that the DnaA protein (responsible for DNA replication) binds to.</p>
</li>
<li>
<p><a name="BM_(single%20stranded%20DNA%7Csingle-stranded%20DNA)%2Fi"></a><strong>single stranded DNA</strong> - A single strand of DNA, not bound to a strand of its <a href="#BM_(reverse%20complement)%2Fi">reverse complement</a>s.</p>
<p><img src="svgbob_03d83c0a0dd86c50ead48a0a8bc87eebf4353039.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a name="BM_(double%20stranded%20DNA%7Cdouble-stranded%20DNA)%2Fi"></a><strong>double stranded DNA</strong> - Two strands of DNA bound together, where each strand is the <a href="#BM_(reverse%20complement)%2Fi">reverse complement</a> of the other.</p>
<p><img src="svgbob_b05df825cb791aaf67d96b4e629e42c02b75f509.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a name="BM_(reverse%20complement)%2Fi"></a><strong>reverse complement</strong> - Given <a href="#BM_(double%20stranded%20DNA%7Cdouble-stranded%20DNA)%2Fi">double-stranded DNA</a>, each ...</p>
<ul>
<li>strand's direction is opposite of the other,</li>
<li>strand position has a nucleotide that complements the nucleotide at that same position on the other stand (A ⟷ T and C ⟷ G)</li>
</ul>
<p><img src="svgbob_b05df825cb791aaf67d96b4e629e42c02b75f509.svg" alt="Kroki diagram output"></p>
<p>The <a href="#BM_(reverse%20complement)%2Fi">reverse complement</a> means that a stretch of <a href="#BM_(single%20stranded%20DNA%7Csingle-stranded%20DNA)%2Fi">single-stranded DNA</a> has its direction reversed (<a href="#BM_(5')%2Fi">5'</a> and <a href="#BM_(3')%2Fi">3'</a> switched) and nucleotides complemented.</p>
</li>
<li>
<p><a name="BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi"></a><strong>gene</strong> - A segment of DNA that contains the instructions for either a protein or functional RNA.</p>
</li>
<li>
<p><a name="BM_(gene%20product)%2Fi"></a><strong>gene product</strong> - The final synthesized material resulting from the instructions that make up a <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a>. That synthesized material either being a protein or functional RNA.</p>
</li>
<li>
<p><a name="BM_(transcription%7Ctranscribed%7Ctranscribe)%2Fi"></a><strong>transcription</strong> - The process of transcribing a <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a> to RNA. Specifically, the enzyme RNA polymerase copies the segment of DNA that makes up that <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a> to a strand of RNA.</p>
<p><img src="svgbob_0fed5f4215fb0a5da3d93da589650b84df3f9c09.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a name="BM_(translation%7Ctranslated%7Ctranslate)%2Fi"></a><strong>translation</strong> - The process of translating mRNA to protein. Specifically, a ribosome takes in the mRNA generated by <a href="#BM_(transcription%7Ctranscribed%7Ctranscribe)%2Fi">transcription</a> and outputs the protein that it codes for.</p>
<p><img src="svgbob_aeaffd032a869fcce0f71f969f22f2191880a88c.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a name="BM_(gene%20expression)%2Fi"></a><strong>gene expression</strong> - The process by which a <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a> is synthesized into a <a href="#BM_(gene%20product)%2Fi">gene product</a>. When the <a href="#BM_(gene%20product)%2Fi">gene product</a> is...</p>
<ul>
<li>a protein, the <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a> is <a href="#BM_(transcription%7Ctranscribed%7Ctranscribe)%2Fi">transcribed</a> to mRNA which in turn is <a href="#BM_(translation%7Ctranslated%7Ctranslate)%2Fi">translated</a> to a protein.</li>
<li>functional RNA, the <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a> is <a href="#BM_(transcription%7Ctranscribed%7Ctranscribe)%2Fi">transcribed</a> to a type of RNA that isn't mRNA (only mRNA is <a href="#BM_(translation%7Ctranslated%7Ctranslate)%2Fi">translated</a> to a protein).</li>
</ul>
<p><img src="svgbob_4c3340936a17e2840e08c6fbcd2b26b03b1c812b.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a name="BM_(regulatory%20gene)%2Fi"></a><strong>regulatory gene</strong> / <a name="BM_(regulatory%20protein)%2Fi"></a><strong>regulatory protein</strong> - The proteins <a href="#BM_(encode%7Cencoding%7Cdecode%7Cdecoding)%2Fi">encode</a>d by these <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">genes</a> affect <a href="#BM_(gene%20expression)%2Fi">gene expression</a> for certain other <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">genes</a>. That is, a <a href="#BM_(regulatory%20protein)%2Fi">regulatory protein</a> can cause certain other <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">genes</a> to be expressed more (promote <a href="#BM_(gene%20expression)%2Fi">gene expression</a>) or less (repress <a href="#BM_(gene%20expression)%2Fi">gene expression</a>).</p>
<p><a href="#BM_(regulatory%20gene)%2Fi">Regulatory gene</a>s are often controlled by external factors (e.g. sunlight, nutrients, temperature, etc..)</p>
</li>
<li>
<p><a name="BM_(feedback%20loop)%2Fi"></a><strong>feedback loop</strong> / <a name="BM_(negative%20feedback%20loop)%2Fi"></a><strong>negative feedback loop</strong> / <a name="BM_(positive%20feedback%20loop)%2Fi"></a><strong>positive feedback loop</strong> - A <a href="#BM_(feedback%20loop)%2Fi">feedback loop</a> is a system where the output (or some part of the output) is fed back into the system to either promote or repress further outputs.</p>
<p><img src="svgbob_3ba8fa4d05a4b9ef40e13dd6d9c7d60ecb9705ad.svg" alt="Kroki diagram output"></p>
<p>A <a href="#BM_(positive%20feedback%20loop)%2Fi">positive feedback loop</a> amplifies the output while a <a href="#BM_(negative%20feedback%20loop)%2Fi">negative feedback loop</a> regulates the output. <a href="#BM_(negative%20feedback%20loop)%2Fi">Negative feedback loop</a>s in particular are important in biology because they allow organisms to maintain homeostasis / equilibrium (keep a consistent internal state). For example, the system that regulates core temperatures in a human is a <a href="#BM_(negative%20feedback%20loop)%2Fi">negative feedback loop</a>. If a human's core temperature gets too...</p>
<ul>
<li>low, they shiver to drive the temperature up.</li>
<li>high, they sweat to drive the temperature down.</li>
</ul>
<p>In the example above, the output is the core temperature. The body monitors its core temperature and employs mechanisms to bring it back to normal if it goes out of range (e.g. sweat, shiver). The outside temperature is influencing the body's core temperature as well as the internal shivering / sweating mechanisms the body employs.</p>
<p><img src="svgbob_894d108e9621a523aaed2c3a3d5d13d20871f1f3.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a name="BM_(circadian%20clock)%2Fi"></a><strong>circadian clock</strong> / <a name="BM_(circadian%20oscillator)%2Fi"></a><strong>circadian oscillator</strong> - A biological clock that synchronizes roughly around the earth's day-night cycle. This internal clock helps many species regulate their physical and behavioural attributes. For example, hunt during the night vs sleep during the day (e.g. nocturnal owls).</p>
</li>
<li>
<p><a name="BM_(upstream%20region)%2Fi"></a><strong>upstream region</strong> - The area just before some interval of DNA. Since the direction of DNA is <a href="#BM_(5')%2Fi">5'</a> to <a href="#BM_(3')%2Fi">3'</a>, this area is towards the <a href="#BM_(5')%2Fi">5'</a> end (upper end).</p>
</li>
<li>
<p><a name="BM_(downstream%20region)%2Fi"></a><strong>downstream region</strong> - The area just after some interval of DNA. Since the direction of DNA is <a href="#BM_(5')%2Fi">5'</a> to <a href="#BM_(3')%2Fi">3'</a>, this area is towards the <a href="#BM_(3')%2Fi">3'</a> end (lower end).</p>
</li>
<li>
<p><a name="BM_(transcription%20factor)%2Fi"></a><strong>transcription factor</strong> - A <a href="#BM_(regulatory%20protein)%2Fi">regulatory protein</a> that controls the rate of <a href="#BM_(transcription%7Ctranscribed%7Ctranscribe)%2Fi">transcription</a> for some <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a> that it has influence over (the copying of DNA to mRNA). The protein binds to a specific <a href="#BM_(sequence)%2Fi">sequence</a> in the <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a>'s <a href="#BM_(upstream%20region)%2Fi">upstream region</a>.</p>
</li>
<li>
<p><a name="BM_(motif)%2Fi"></a><strong>motif</strong> - A pattern that matches against many different <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s, where those matched <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s have some shared biological significance. The pattern matches a fixed k where each position may have alternate forms. The simplest way to think of a <a href="#BM_(motif)%2Fi">motif</a> is a regex pattern without quantifiers. For example, the regex <code>[AT]TT[GC]C</code> may match to ATTGC, ATTCC, TTTGC, and TTTCC.</p>
</li>
<li>
<p><a name="BM_(motif%20member)%2Fi"></a><strong>motif member</strong> <a name="BM_%5Cb(membership%7Cmember)_MOTIF%2Fi"></a><strong></strong> - A specific nucleotide <a href="#BM_(sequence)%2Fi">sequence</a> that matches a <a href="#BM_(motif)%2Fi">motif</a>. For example, given a <a href="#BM_(motif)%2Fi">motif</a> represented by the regex <code>[AT]TT[GC]C</code>, the <a href="#BM_(sequence)%2Fi">sequence</a>s ATTGC, ATTCC, TTTGC, and TTTCC would be its <a href="#BM_%5Cb(membership%7Cmember)_MOTIF%2Fi">member</a>s.</p>
</li>
<li>
<p><a name="BM_(motif%20matrix%7Cmotif%20matrices)%2Fi"></a><strong>motif matrix</strong> - A set of <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s stacked on top of each other in a matrix, where the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s are either...</p>
<ul>
<li><a href="#BM_%5Cb(membership%7Cmember)_MOTIF%2Fi">member</a>s of the same <a href="#BM_(motif)%2Fi">motif</a>,</li>
<li>or suspected <a href="#BM_%5Cb(membership%7Cmember)_MOTIF%2Fi">member</a>s of the same <a href="#BM_(motif)%2Fi">motif</a>.</li>
</ul>
<p>For example, the <a href="#BM_(motif)%2Fi">motif</a> <code>[AT]TT[GC]C</code> has the following matrix:</p>
<table>
<thead>
<tr>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
</tr>
</thead>
<tbody>
<tr>
<td>A</td>
<td>T</td>
<td>T</td>
<td>G</td>
<td>C</td>
</tr>
<tr>
<td>A</td>
<td>T</td>
<td>T</td>
<td>C</td>
<td>C</td>
</tr>
<tr>
<td>T</td>
<td>T</td>
<td>T</td>
<td>G</td>
<td>C</td>
</tr>
<tr>
<td>T</td>
<td>T</td>
<td>T</td>
<td>C</td>
<td>C</td>
</tr>
</tbody>
</table>
</li>
<li>
<p><a name="BM_(regulatory%20motif)%2Fi"></a><strong>regulatory motif</strong> - The <a href="#BM_(motif)%2Fi">motif</a> of a <a href="#BM_(transcription%20factor)%2Fi">transcription factor</a>, typically 8 to 12 nucleotides in length.</p>
</li>
<li>
<p><a name="BM_(transcription%20factor%20binding%20site)%2Fi"></a><strong>transcription factor binding site</strong> - The physical binding site for a <a href="#BM_(transcription%20factor)%2Fi">transcription factor</a>. A <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a> that's regulated by a <a href="#BM_(transcription%20factor)%2Fi">transcription factor</a> needs a <a href="#BM_(sequence)%2Fi">sequence</a> located in its <a href="#BM_(upstream%20region)%2Fi">upstream region</a> that the <a href="#BM_(transcription%20factor)%2Fi">transcription factor</a> can bind to: a <a href="#BM_(motif%20member)%2Fi">motif member</a> of that <a href="#BM_(transcription%20factor)%2Fi">transcription factor</a>'s <a href="#BM_(regulatory%20motif)%2Fi">regulatory motif</a>.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>A <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a>'s <a href="#BM_(upstream%20region)%2Fi">upstream region</a> is the 600 to 1000 nucleotides preceding the start of the <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a>.</p>
</div>
</li>
<li>
<p><a name="BM_(complementary%20DNA)%2Fi"></a><strong>complementary DNA (cDNA)</strong> <a name="BM_(cDNA)%2F"></a><strong></strong> - A single strand of DNA generated from mRNA. The enzyme reverse <a href="#BM_(RNA%20transcript%7Ctranscript)%2Fi">transcript</a>ase scans over the mRNA and creates the complementing single DNA strand.</p>
<p><img src="svgbob_97c5d076bf0a1cd98b75bbb88b72fc29c3b1b93d.svg" alt="Kroki diagram output"></p>
<p>The mRNA portion breaks off, leaving the <a href="#BM_(single%20stranded%20DNA%7Csingle-stranded%20DNA)%2Fi">single-stranded DNA</a>.</p>
<p><img src="svgbob_97e71090d31bdaa40278b0e705aef4e3555be8e3.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a name="BM_(DNA%20microarray)%2Fi"></a><strong>DNA microarray</strong> - A device used to compare <a href="#BM_(gene%20expression)%2Fi">gene expression</a>. This works by measuring 2 mRNA samples against each other: a control sample and an experimental sample. The samples could be from...</p>
<ul>
<li>the same organism but at different times.</li>
<li>diseased and healthy versions of the same organism.</li>
<li>etc..</li>
</ul>
<p>Both mRNA samples are converted to <a href="#BM_(cDNA)%2F">cDNA</a> and are given fluorescent dyes. The control sample gets dyed green while the experimental sample gets dyed red.</p>
<p><img src="svgbob_6ce607591d81d0fd6c7507408e97fb8146e83baf.svg" alt="Kroki diagram output"></p>
<p>A sheet is broken up into multiple regions, where each region has the <a href="#BM_(cDNA)%2F">cDNA</a> for one specific <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a> from the control sample printed.</p>
<p><img src="svgbob_421210658cf3b4d0040b4815e879948ebfd7ccc5.svg" alt="Kroki diagram output"></p>
<p>The idea is that once the experimental <a href="#BM_(cDNA)%2F">cDNA</a> is introduced to that region, it should bind to the control <a href="#BM_(cDNA)%2F">cDNA</a> that's been printed to form <a href="#BM_(double%20stranded%20DNA%7Cdouble-stranded%20DNA)%2Fi">double-stranded DNA</a>. The color emitted in a region should correspond to the amount of <a href="#BM_(gene%20expression)%2Fi">gene expression</a> for the <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a> that region represents. For example, if a region on the sheet is fully yellow, it means that the <a href="#BM_(gene%20expression)%2Fi">gene expression</a> for that <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a> is roughly equal (red mixed with green is yellow).</p>
</li>
<li>
<p><a name="BM_(greedy%20algorithm)%2Fi"></a><strong>greedy algorithm</strong> - An algorithm that tries to speed things up by taking the locally optimal choice at each step. That is, the algorithm doesn't look more than 1 step ahead.</p>
<p>For example, imagine a chess playing AI that had a strategy of trying to eliminate the other player's most valuable piece at each turn. It would be considered greedy because it only looks 1 move ahead before taking action. Normal chess AIs / players look many moves ahead before taking action. As such, the greedy AI may be fast but it would very likely lose most matches.</p>
</li>
<li>
<p><a name="BM_(Cromwell's%20rule)%2Fi"></a><strong>Cromwell's rule</strong> - When a probability is based on past events, 0.0 and 1.0 shouldn't be used. That is, if you've...</p>
<ul>
<li>never seen an even occur in the past, it doesn't mean that there's a 0.0 probability of it occurring next.</li>
<li>always seen an event occur in the past, it doesn't mean that there's a 1.0 probability of it occurring next.</li>
</ul>
<p>Unless you're dealing with hard logical statements where prior occurrences don't come in to play (e.g. 1+1=2), you should include a small chance that some extremely unlikely event may happen. The example tossed around is "the probability that the sun will not rise tomorrow." Prior recorded observations show that the sun has always risen, but that doesn't mean that there's a 1.0 probability of the sun rising tomorrow (e.g. some extremely unlikely cataclysmic event may prevent the sun from rising).</p>
</li>
<li>
<p><a name="BM_(Laplace's%20rule%20of%20succession%7CLaplace's%20rule)%2Fi"></a><strong>Laplace's rule of succession</strong> - If some independent true/false event occurs n times, and s of those n times were successes, it's natural for people to assume the probability of success is <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mfrac><mi>s</mi><mi>n</mi></mfrac></mrow><annotation encoding="application/x-tex">\frac{s}{n}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.040392em;vertical-align:-0.345em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.695392em;"><span style="top:-2.6550000000000002em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">n</span></span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.394em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">s</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.345em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span>. However, if the number of successes is 0, the probability would be 0.0. <a href="#BM_(Cromwell's%20rule)%2Fi">Cromwell's rule</a> states that when a probability is based off past events, 0.0 and 1.0 shouldn't be used. As such, a more appropriate / meaningful measure of probability is <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mfrac><mrow><mi>s</mi><mo>+</mo><mn>1</mn></mrow><mrow><mi>n</mi><mo>+</mo><mn>2</mn></mrow></mfrac></mrow><annotation encoding="application/x-tex">\frac{s+1}{n+2}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.2484389999999999em;vertical-align:-0.403331em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.845108em;"><span style="top:-2.655em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">n</span><span class="mbin mtight">+</span><span class="mord mtight">2</span></span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.394em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">s</span><span class="mbin mtight">+</span><span class="mord mtight">1</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.403331em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span>.</p>
<p>For example, imagine you're sitting on a park bench having lunch. Of the 8 birds you've seen since starting your lunch, all have been pigeons. If you were to calculate the probability that the next bird you'll see a crow, <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mfrac><mn>0</mn><mn>8</mn></mfrac></mrow><annotation encoding="application/x-tex">\frac{0}{8}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.190108em;vertical-align:-0.345em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.845108em;"><span style="top:-2.6550000000000002em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">8</span></span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.394em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">0</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.345em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span> would be flawed because it states that there's no chance that the next bird will be a crow (there obviously is a chance, but it may be a small chance). Instead, applying <a href="#BM_(Laplace's%20rule%20of%20succession%7CLaplace's%20rule)%2Fi">Laplace's rule</a> allows for the small probability that a crow may be seen next: <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mfrac><mrow><mn>0</mn><mo>+</mo><mn>1</mn></mrow><mrow><mn>8</mn><mo>+</mo><mn>2</mn></mrow></mfrac></mrow><annotation encoding="application/x-tex">\frac{0+1}{8+2}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.2484389999999999em;vertical-align:-0.403331em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.845108em;"><span style="top:-2.655em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">8</span><span class="mbin mtight">+</span><span class="mord mtight">2</span></span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.394em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">0</span><span class="mbin mtight">+</span><span class="mord mtight">1</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.403331em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span>.</p>
<p><a href="#BM_(Laplace's%20rule%20of%20succession%7CLaplace's%20rule)%2Fi">Laplace's rule of succession</a> is more meaningful when the number of trials (n) is small.</p>
</li>
<li>
<p><a name="BM_(pseudocount)%2Fi"></a><strong>pseudocount</strong> - When a zero is replaced with a small number to prevent unfair scoring. See <a href="#BM_(Laplace's%20rule%20of%20succession%7CLaplace's%20rule)%2Fi">Laplace's rule of succession</a>.</p>
</li>
<li>
<p><a name="BM_(randomized%20algorithm)%2Fi"></a><strong>randomized algorithm</strong> - An algorithm that uses a source of randomness as part of its logic. <a href="#BM_(randomized%20algorithm)%2Fi">Randomized algorithm</a>s come in two forms: <a href="#BM_(Las%20Vegas%20algorithm)%2Fi">Las Vegas algorithm</a>s and <a href="#BM_(Monte%20Carlo%20algorithm)%2Fi">Monte Carlo algorithm</a>s</p>
</li>
<li>
<p><a name="BM_(Las%20Vegas%20algorithm)%2Fi"></a><strong>Las Vegas algorithm</strong> - A <a href="#BM_(randomized%20algorithm)%2Fi">randomized algorithm</a> that delivers a guaranteed exact solution. That is, even though the algorithm makes random decisions it is guaranteed to converge on the exact solution to the problem its trying to solve (not an approximate solution).</p>
<p>An example of a <a href="#BM_(Las%20Vegas%20algorithm)%2Fi">Las Vegas algorithm</a> is randomized quicksort (randomness is applied when choosing the pivot).</p>
</li>
<li>
<p><a name="BM_(Monte%20Carlo%20algorithm)%2Fi"></a><strong>Monte Carlo algorithm</strong> - A <a href="#BM_(randomized%20algorithm)%2Fi">randomized algorithm</a> that delivers an approximate solution. Because these algorithms are quick, they're typically run many times. The approximation considered the best out of all runs is the one that gets chosen as the solution.</p>
<p>An example of a <a href="#BM_(Monte%20Carlo%20algorithm)%2Fi">Monte Carlo algorithm</a> is a genetic algorithm to optimize the weights of a deep neural network. That is, a step of the optimization requires running n different neural networks to see which gives the best result, then replacing those n networks with n copies of the best performing network where each copy has randomly tweaked weights. At some point the algorithm will stop producing incrementally better results.</p>
<p>Perform the optimization (the entire thing, not just a single step) thousands of times and pick the best network.</p>
</li>
<li>
<p><a name="BM_(consensus%20string%7Cconsensus%20sequence)%2Fi"></a><strong>consensus string</strong> - The <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> generated by selecting the most abundant column at each index of a <a href="#BM_(motif%20matrix%7Cmotif%20matrices)%2Fi">motif matrix</a>.</p>
<table>
<thead>
<tr>
<th></th>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> 1</td>
<td>A</td>
<td>T</td>
<td>T</td>
<td>G</td>
<td>C</td>
</tr>
<tr>
<td><a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> 2</td>
<td>A</td>
<td>T</td>
<td>T</td>
<td>C</td>
<td>C</td>
</tr>
<tr>
<td><a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> 3</td>
<td>T</td>
<td>T</td>
<td>T</td>
<td>G</td>
<td>C</td>
</tr>
<tr>
<td><a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> 4</td>
<td>T</td>
<td>T</td>
<td>T</td>
<td>C</td>
<td>C</td>
</tr>
<tr>
<td><a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> 5</td>
<td>A</td>
<td>T</td>
<td>T</td>
<td>C</td>
<td>G</td>
</tr>
<tr>
<td>consensus</td>
<td>A</td>
<td>T</td>
<td>T</td>
<td>C</td>
<td>C</td>
</tr>
</tbody>
</table>
<p>The generated <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> may also use a <a href="#BM_(hybrid%20alphabet%7Calternate%20alphabet%7Calternative%20alphabet)%2Fi">hybrid alphabet</a>. The <a href="#BM_(consensus%20string%7Cconsensus%20sequence)%2Fi">consensus string</a> for the same matrix above using <a href="#BM_(IUPAC%20nucleotide%20code)%2Fi">IUPAC nucleotide code</a>s: <code>WTTSS</code>.</p>
</li>
<li>
<p><a name="BM_(entropy)%2Fi"></a><strong>entropy</strong> - The uncertainty associated with a random variable. Given some set of outcomes for a variable, it's calculated as <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo>−</mo><msubsup><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></msubsup><mi>P</mi><mo stretchy="false">(</mo><msub><mi>x</mi><mi>i</mi></msub><mo stretchy="false">)</mo><mi>l</mi><mi>o</mi><mi>g</mi><mi>P</mi><mo stretchy="false">(</mo><msub><mi>x</mi><mi>i</mi></msub><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">-\sum_{i=1}^{n} P(x_i) log P(x_i)</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.104002em;vertical-align:-0.29971000000000003em;"></span><span class="mord">−</span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mop"><span class="mop op-symbol small-op" style="position:relative;top:-0.0000050000000000050004em;">∑</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.804292em;"><span style="top:-2.40029em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">i</span><span class="mrel mtight">=</span><span class="mord mtight">1</span></span></span></span><span style="top:-3.2029em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">n</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.29971000000000003em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord mathnormal" style="margin-right:0.13889em;">P</span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal">x</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mclose">)</span><span class="mord mathnormal" style="margin-right:0.01968em;">l</span><span class="mord mathnormal">o</span><span class="mord mathnormal" style="margin-right:0.03588em;">g</span><span class="mord mathnormal" style="margin-right:0.13889em;">P</span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal">x</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mclose">)</span></span></span></span>.</p>
<p>This definition is for information theory. In other contexts (e.g. physics, economics), this term has a different meaning.</p>
</li>
<li>
<p><a name="BM_(genome)%2Fi"></a><strong>genome</strong> - In the context of a ...</p>
<ul>
<li>specific organism, <a href="#BM_(genome)%2Fi">genome</a> refers to all of the DNA for that organism (e.g. a specific E. coli cell).</li>
<li>specific species, <a href="#BM_(genome)%2Fi">genome</a> refers to the idealized DNA for that species (e.g. all E. coli).</li>
</ul>
<p>DNA of individual cells mutate all the time. For example, even in multi-cell organism, two cells from the same mouse may not have the exact same DNA.</p>
</li>
<li>
<p><a name="BM_(sequence)%2Fi"></a><strong>sequence</strong> - The ordered elements that make up some biological entity. For example, a ...</p>
<ul>
<li>DNA <a href="#BM_(sequence)%2Fi">sequence</a> contains the set of nucleotides and their positions for that DNA strand.</li>
<li><a href="#BM_(peptide)%2Fi">peptide</a> <a href="#BM_(sequence)%2Fi">sequence</a> contains the set of <a href="#BM_(amino%20acid)%2Fi">amino acid</a>s and their positions for that <a href="#BM_(peptide)%2Fi">peptide</a>.</li>
</ul>
</li>
<li>
<p><a name="BM_(sequencing%7Csequenced)%2Fi"></a><strong>sequencing</strong> - The process of determining which nucleotides are assigned to which positions in a strand of DNA or RNA.</p>
<p>The machinery used for DNA <a href="#BM_(sequencing%7Csequenced)%2Fi">sequencing</a> is called a <a href="#BM_(sequencer)%2Fi">sequencer</a>. A <a href="#BM_(sequencer)%2Fi">sequencer</a> takes multiple copies of the same DNA, breaks that DNA up into smaller fragments, and scans in those <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s. Each <a href="#BM_(fragment)_SEQ%2Fi">fragment</a> is typically the same length but has a unique starting offset. Because the starting offsets are all different, the original larger DNA <a href="#BM_(sequence)%2Fi">sequence</a> can be guessed at by finding <a href="#BM_(fragment)_SEQ%2Fi">fragment</a> with overlapping regions and stitching them together.</p>
<table>
<thead>
<tr>
<th></th>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
<th>6</th>
<th>7</th>
<th>8</th>
<th>9</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="#BM_%5Cb(read)_SEQ%2Fi">read</a> 1</td>
<td></td>
<td></td>
<td></td>
<td></td>
<td>C</td>
<td>T</td>
<td>T</td>
<td>C</td>
<td>T</td>
<td>T</td>
</tr>
<tr>
<td><a href="#BM_%5Cb(read)_SEQ%2Fi">read</a> 2</td>
<td></td>
<td></td>
<td></td>
<td>G</td>
<td>C</td>
<td>T</td>
<td>T</td>
<td>C</td>
<td>T</td>
<td></td>
</tr>
<tr>
<td><a href="#BM_%5Cb(read)_SEQ%2Fi">read</a> 3</td>
<td></td>
<td></td>
<td>T</td>
<td>G</td>
<td>C</td>
<td>T</td>
<td>T</td>
<td>C</td>
<td></td>
<td></td>
</tr>
<tr>
<td><a href="#BM_%5Cb(read)_SEQ%2Fi">read</a> 4</td>
<td></td>
<td>T</td>
<td>T</td>
<td>G</td>
<td>C</td>
<td>T</td>
<td>T</td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><a href="#BM_%5Cb(read)_SEQ%2Fi">read</a> 5</td>
<td>A</td>
<td>T</td>
<td>T</td>
<td>G</td>
<td>C</td>
<td>T</td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td>reconstructed</td>
<td>A</td>
<td>T</td>
<td>T</td>
<td>G</td>
<td>C</td>
<td>T</td>
<td>T</td>
<td>C</td>
<td>T</td>
<td>T</td>
</tr>
</tbody>
</table>
</li>
<li>
<p><a name="BM_(sequencer)%2Fi"></a><strong>sequencer</strong> - A machine that performs DNA or <a href="#BM_(RNA%20sequencing)%2Fi">RNA sequencing</a>.</p>
</li>
<li>
<p><a name="BM_(sequencing%20error)%2Fi"></a><strong>sequencing error</strong> - An error caused by a <a href="#BM_(sequencer)%2Fi">sequencer</a> returning a <a href="#BM_(fragment)_SEQ%2Fi">fragment</a> where a nucleotide was misinterpreted at one or more positions (e.g. offset 3 was actually a C but it got scanned in as a G).</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>This term may also be used in reference to homopolymer errors, known to happen with nanopore technology. From <a href="https://news.ycombinator.com/item?id=25459670">here</a>...</p>
<blockquote>
<p>A homopolymer is when you have stretches of the same nucleotide, and the error is miscounting the number of them. e.g: GAAAC could be called as "GAAC" or "GAAAAC" or even "GAAAAAAAC".</p>
</blockquote>
</div>
</li>
<li>
<p><a name="BM_%5Cb(read)_SEQ%2Fi"></a><strong>read</strong> - A segment of <a href="#BM_(genome)%2Fi">genome</a> scanned in during the process of <a href="#BM_(sequencing%7Csequenced)%2Fi">sequencing</a>.</p>
</li>
<li>
<p><a name="BM_(read-pair%7Cread%20pair)%2Fi"></a><strong>read-pair</strong> - A segment of <a href="#BM_(genome)%2Fi">genome</a> scanning in during the process of <a href="#BM_(sequencing%7Csequenced)%2Fi">sequencing</a>, where the middle of the segment is unknown. That is, the first k elements and the last k elements are known, but the d elements in between aren't known. The total size of the segment is 2k + d.</p>
<p><a href="#BM_(sequencer)%2Fi">Sequencer</a>s provide <a href="#BM_(read-pair%7Cread%20pair)%2Fi">read-pair</a>s as an alternative to longer <a href="#BM_%5Cb(read)_SEQ%2Fi">read</a>s because the longer a <a href="#BM_%5Cb(read)_SEQ%2Fi">read</a> is the more errors it contains.</p>
<p>See <a href="#BM_(%5C(%5Cd%2B%2C%5Cs*%5Cd%2B%5C)-mer%7Ckd-mer%7Ckdmer%7C%5C(k%2C%5Cs*d%5C)-mer)%2Fi">kd-mer</a>.</p>
</li>
<li>
<p><a name="BM_(fragment)_SEQ%2Fi"></a><strong>fragment</strong> - A scanned <a href="#BM_(sequence)%2Fi">sequence</a> returned by a <a href="#BM_(sequencer)%2Fi">sequencer</a>. Represented as either a <a href="#BM_%5Cb(read)_SEQ%2Fi">read</a> or a <a href="#BM_(read-pair%7Cread%20pair)%2Fi">read-pair</a>.</p>
</li>
<li>
<p><a name="BM_(assembly%7Cassemble)%2Fi"></a><strong>assembly</strong> - The process of stitching together overlapping <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s to guess the <a href="#BM_(sequence)%2Fi">sequence</a> of the original larger DNA <a href="#BM_(sequence)%2Fi">sequence</a> that those <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s came from.</p>
</li>
<li>
<p><a name="BM_(hybrid%20alphabet%7Calternate%20alphabet%7Calternative%20alphabet)%2Fi"></a><strong>hybrid alphabet</strong> - When representing a <a href="#BM_(sequence)%2Fi">sequence</a> that isn't fully conserved, it may be more appropriate to use an alphabet where each letter can represent more than 1 nucleotide. For example, the <a href="#BM_(IUPAC%20nucleotide%20code)%2Fi">IUPAC nucleotide code</a>s provides the following alphabet:</p>
<ul>
<li>A = A</li>
<li>C = C</li>
<li>T = T</li>
<li>G = G</li>
<li>W = A or T</li>
<li>S = G or C</li>
<li>K = G or T</li>
<li>Y = C or T</li>
<li>...</li>
</ul>
<p>If the <a href="#BM_(sequence)%2Fi">sequence</a> being represented can be either AAAC or AATT, it may be easier to represent a single string of AAWY.</p>
</li>
<li>
<p><a name="BM_(IUPAC%20nucleotide%20code)%2Fi"></a><strong>IUPAC nucleotide code</strong> - A <a href="#BM_(hybrid%20alphabet%7Calternate%20alphabet%7Calternative%20alphabet)%2Fi">hybrid alphabet</a> with the following mapping:</p>
<table>
<thead>
<tr>
<th>Letter</th>
<th>Base</th>
</tr>
</thead>
<tbody>
<tr>
<td>A</td>
<td>Adenine</td>
</tr>
<tr>
<td>C</td>
<td>Cytosine</td>
</tr>
<tr>
<td>G</td>
<td>Guanine</td>
</tr>
<tr>
<td>T (or U)</td>
<td>Thymine (or Uracil)</td>
</tr>
<tr>
<td>R</td>
<td>A or G</td>
</tr>
<tr>
<td>Y</td>
<td>C or T</td>
</tr>
<tr>
<td>S</td>
<td>G or C</td>
</tr>
<tr>
<td>W</td>
<td>A or T</td>
</tr>
<tr>
<td>K</td>
<td>G or T</td>
</tr>
<tr>
<td>M</td>
<td>A or C</td>
</tr>
<tr>
<td>B</td>
<td>C or G or T</td>
</tr>
<tr>
<td>D</td>
<td>A or G or T</td>
</tr>
<tr>
<td>H</td>
<td>A or C or T</td>
</tr>
<tr>
<td>V</td>
<td>A or C or G</td>
</tr>
<tr>
<td>N</td>
<td>any base</td>
</tr>
<tr>
<td>. or -</td>
<td>gap</td>
</tr>
</tbody>
</table>
<p><a href="https://www.bioinformatics.org/sms/iupac.html">Source</a>.</p>
</li>
<li>
<p><a name="BM_(%5Cblogo%7Csequence%20logo)%2Fi"></a><strong>sequence logo</strong> - A graphical representation of how conserved a <a href="#BM_(sequence)%2Fi">sequence</a>'s positions are. Each position has its possible nucleotides stacked on top of each other, where the height of each nucleotide is based on how conserved it is. The more conserved a position is, the taller that column will be.</p>
<p>Typically applied to DNA or RNA, and May also be applied to other biological <a href="#BM_(sequence)%2Fi">sequence</a> types (e.g. <a href="#BM_(amino%20acid)%2Fi">amino acid</a>s).</p>
<p>The following is an example of a <a href="#BM_(%5Cblogo%7Csequence%20logo)%2Fi">logo</a> generated from a <a href="#BM_(motif)%2Fi">motif</a> <a href="#BM_(sequence)%2Fi">sequence</a>:</p>
<div style="border:1px solid black;">
<p></p>
<p>Generating logo for the following motif matrix...</p>
<p>TCGGGGGTTTTT<br>CCGGTGACTTAC<br>ACGGGGATTTTC<br>TTGGGGACTTTT<br>AAGGGGACTTCC<br>TTGGGGACTTCC<br>TCGGGGATTCAT<br>TCGGGGATTCCT<br>TAGGGGAACTAC<br>TCGGGTATAACC</p>
<p>Result...</p>
<p><img src="motif_logo_01b7e8763a84973751af52163d52f5bd.svg" alt="Motif Logo"></p>
</div>
<p></p>
</li>
<li>
<p><a name="BM_(transposon%7Ctransposable%20element%7Cjumping%20gene)%2Fi"></a><strong>transposon</strong> - A DNA <a href="#BM_(sequence)%2Fi">sequence</a> that can change its position within a <a href="#BM_(genome)%2Fi">genome</a>, altering the <a href="#BM_(genome)%2Fi">genome</a> size. They come in two flavours:</p>
<ul>
<li>Class I (retro<a href="#BM_(transposon%7Ctransposable%20element%7Cjumping%20gene)%2Fi">transposon</a>) - Behaves similarly to copy-and-paste where the <a href="#BM_(sequence)%2Fi">sequence</a> is duplicated. DNA is <a href="#BM_(transcription%7Ctranscribed%7Ctranscribe)%2Fi">transcribed</a> to RNA, followed by that RNA being reverse <a href="#BM_(transcription%7Ctranscribed%7Ctranscribe)%2Fi">transcribed</a> back to DNA by an enzyme called reverse <a href="#BM_(RNA%20transcript%7Ctranscript)%2Fi">transcript</a>ase.</li>
<li>Class II (DNA <a href="#BM_(transposon%7Ctransposable%20element%7Cjumping%20gene)%2Fi">transposon</a>) - Behaves similarly to cut-and-paste where the <a href="#BM_(sequence)%2Fi">sequence</a> is moved. DNA is physically cut out by an enzyme called transposases and placed back in at some other location.</li>
</ul>
<p>Often times, <a href="#BM_(transposon%7Ctransposable%20element%7Cjumping%20gene)%2Fi">transposon</a>s cause disease. For example, ...</p>
<ul>
<li>insertion of a <a href="#BM_(transposon%7Ctransposable%20element%7Cjumping%20gene)%2Fi">transposon</a> into a <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a> will likely disable that <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a>.</li>
<li>after a <a href="#BM_(transposon%7Ctransposable%20element%7Cjumping%20gene)%2Fi">transposon</a> leaves a <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a>, the gap likely won't be repaired correctly.</li>
</ul>
</li>
<li>
<p><a name="BM_(adjacency%20list)%2Fi"></a><strong>adjacency list</strong> - An internal representation of a graph where each node has a list of pointers to other nodes that it can forward to.</p>
<p><img src="svgbob_5b40f65c78645193c91742f02934854ae8846ede.svg" alt="Kroki diagram output"></p>
<p>The graph above represented as an <a href="#BM_(adjacency%20list)%2Fi">adjacency list</a> would be...</p>
<table>
<thead>
<tr>
<th>From</th>
<th>To</th>
</tr>
</thead>
<tbody>
<tr>
<td>A</td>
<td>B</td>
</tr>
<tr>
<td>B</td>
<td>C</td>
</tr>
<tr>
<td>C</td>
<td>D,E</td>
</tr>
<tr>
<td>D</td>
<td>F</td>
</tr>
<tr>
<td>E</td>
<td>D,F</td>
</tr>
<tr>
<td>F</td>
<td></td>
</tr>
</tbody>
</table>
</li>
<li>
<p><a name="BM_(adjacency%20matrix)%2Fi"></a><strong>adjacency matrix</strong> - An internal representation of a graph where a matrix defines the number of times that each node forwards to every other node.</p>
<p><img src="svgbob_5b40f65c78645193c91742f02934854ae8846ede.svg" alt="Kroki diagram output"></p>
<p>The graph above represented as an <a href="#BM_(adjacency%20matrix)%2Fi">adjacency matrix</a> would be...</p>
<table>
<thead>
<tr>
<th></th>
<th>A</th>
<th>B</th>
<th>C</th>
<th>D</th>
<th>E</th>
<th>F</th>
</tr>
</thead>
<tbody>
<tr>
<td>A</td>
<td>0</td>
<td>1</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
<tr>
<td>B</td>
<td>0</td>
<td>0</td>
<td>1</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
<tr>
<td>C</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>1</td>
<td>1</td>
<td>0</td>
</tr>
<tr>
<td>D</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>1</td>
</tr>
<tr>
<td>E</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>1</td>
<td>0</td>
<td>1</td>
</tr>
<tr>
<td>F</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
</tbody>
</table>
</li>
<li>
<p><a name="BM_(Hamiltonian%20path%7CHamilton%20path)%2Fi"></a><strong>Hamiltonian path</strong> - A path in a graph that visits every node exactly once.</p>
<p>The graph below has the <a href="#BM_(Hamiltonian%20path%7CHamilton%20path)%2Fi">Hamiltonian path</a> ABCEDF.</p>
<p><img src="svgbob_5b40f65c78645193c91742f02934854ae8846ede.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a name="BM_(Eulerian%20path)%2Fi"></a><strong>Eulerian path</strong> <a name="BM_(Eulerian)_PATH%2Fi"></a><strong></strong> - A path in a graph that visits every edge exactly once.</p>
<p>In the graph below, the <a href="#BM_(Eulerian%20path)%2Fi">Eulerian path</a> is (A,B), (B,C), (C,D), (D,E), (E,C), (C,D), (D,F).</p>
<p><img src="svgbob_87660b191d813d39c396985391b70507c8c1236b.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a name="BM_(Eulerian%20cycle)%2Fi"></a><strong>Eulerian cycle</strong> <a name="BM_(Eulerian)_CYCLE%2Fi"></a><strong></strong> - An <a href="#BM_(Eulerian%20path)%2Fi">Eulerian path</a> that forms a cycle. That is, a path in a graph that is a cycle and visits every edge exactly once.</p>
<p>The graph below has an <a href="#BM_(Eulerian%20cycle)%2Fi">Eulerian cycle</a> of (A,B), (B,C) (C,D), (D,F), (F,C), (C,A).</p>
<p><img src="svgbob_244dd1930eb16ef58be2261f1ffb9277ad2d793a.svg" alt="Kroki diagram output"></p>
<p>If a graph contains an <a href="#BM_(Eulerian%20cycle)%2Fi">Eulerian cycle</a>, it's said to be an <a href="#BM_(Eulerian%20graph)%2Fi">Eulerian graph</a>.</p>
</li>
<li>
<p><a name="BM_(Eulerian%20graph)%2Fi"></a><strong>Eulerian graph</strong> <a name="BM_(Eulerian)_GRAPH%2Fi"></a><strong></strong> - For a graph to be <a href="#BM_(Eulerian)_GRAPH%2Fi">Eulerian</a>, it must have an <a href="#BM_(Eulerian%20cycle)%2Fi">Eulerian cycle</a>: a path in a graph that is a cycle and visits every edge exactly once. For a graph to have an <a href="#BM_(Eulerian%20cycle)%2Fi">Eulerian cycle</a>, it must be both <a href="#BM_(balanced)_GRAPH%2Fi">balanced</a> and <a href="#BM_(strongly%20connected)%2Fi">strongly connected</a>.</p>
<p><img src="svgbob_d0af75d07c2969afc5f4088d18d3b7a327f26ef2.svg" alt="Kroki diagram output"></p>
<p>Note how in the graph above, ...</p>
<ul>
<li>
<p>every node is reachable from every other node (<a href="#BM_(strongly%20connected)%2Fi">strongly connected</a>),</p>
</li>
<li>
<p>every node has an <a href="#BM_(outdegree)%2Fi">outdegree</a> equal to its <a href="#BM_(indegree)%2Fi">indegree</a> (<a href="#BM_(balanced)_GRAPH%2Fi">balanced</a>).</p>
<table>
<thead>
<tr>
<th>Node</th>
<th><a href="#BM_(indegree)%2Fi">Indegree</a></th>
<th><a href="#BM_(outdegree)%2Fi">Outdegree</a></th>
</tr>
</thead>
<tbody>
<tr>
<td>A</td>
<td>1</td>
<td>1</td>
</tr>
<tr>
<td>B</td>
<td>1</td>
<td>1</td>
</tr>
<tr>
<td>C</td>
<td>2</td>
<td>2</td>
</tr>
<tr>
<td>D</td>
<td>1</td>
<td>1</td>
</tr>
<tr>
<td>F</td>
<td>1</td>
<td>1</td>
</tr>
</tbody>
</table>
</li>
</ul>
<p>In contrast, the following graphs are not <a href="#BM_(Eulerian%20graph)%2Fi">Eulerian graph</a>s (no <a href="#BM_(Eulerian%20cycle)%2Fi">Eulerian cycle</a>s exist):</p>
<ul>
<li>
<p><a href="#BM_(strongly%20connected)%2Fi">Strongly connected</a> but not <a href="#BM_(balanced)_GRAPH%2Fi">balanced</a>.</p>
<p><img src="svgbob_efdff1b3ec66000b9acf2ddc84c2d3684d318683.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a href="#BM_(balanced)_GRAPH%2Fi">Balanced</a> but not <a href="#BM_(strongly%20connected)%2Fi">strongly connected</a>.</p>
<p><img src="svgbob_48abea7e0377d91bfd845a85b2ee9615e9852bfd.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a href="#BM_(balanced)_GRAPH%2Fi">Balanced</a> but <a href="#BM_(disconnected)%2Fi">disconnected</a> (not <a href="#BM_(strongly%20connected)%2Fi">strongly connected</a>).</p>
<p><img src="svgbob_1f6d7503fdb3f3654b3c64f041254352c8d9a43e.svg" alt="Kroki diagram output"></p>
</li>
</ul>
</li>
<li>
<p><a name="BM_(disconnected)%2Fi"></a><strong>disconnected</strong> / <a name="BM_(connected)%2Fi"></a><strong>connected</strong> - A graph is <a href="#BM_(disconnected)%2Fi">disconnected</a> if you can break it out into 2 or more distinct sub-graphs without breaking any paths. In other words, the graph contains at least two nodes which aren't contained in any path.</p>
<p>The graph below is <a href="#BM_(disconnected)%2Fi">disconnected</a> because there is no path that contains E, F, G, or H and A, B, C, or D.</p>
<p><img src="svgbob_754a446a49b8209971f339d76f4fcc492acbb615.svg" alt="Kroki diagram output"></p>
<p>The graph below is <a href="#BM_(connected)%2Fi">connected</a>.</p>
<p><img src="svgbob_028c9291a4412afebe0b605b464987e2c9bf75ef.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a name="BM_(strongly%20connected)%2Fi"></a><strong>strongly connected</strong> - A graph is <a href="#BM_(strongly%20connected)%2Fi">strongly connected</a> if every node is reachable from every other node.</p>
<p>The graph below is <strong>not</strong> <a href="#BM_(strongly%20connected)%2Fi">strongly connected</a> because neither A nor B is reachable by C, D, E, or F.</p>
<p><img src="svgbob_5b40f65c78645193c91742f02934854ae8846ede.svg" alt="Kroki diagram output"></p>
<p>The graph below is <a href="#BM_(strongly%20connected)%2Fi">strongly connected</a> because all nodes are reachable from all nodes.</p>
<p><img src="svgbob_244dd1930eb16ef58be2261f1ffb9277ad2d793a.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a name="BM_(indegree)%2Fi"></a><strong>indegree</strong> / <a name="BM_(outdegree)%2Fi"></a><strong>outdegree</strong> - The number of edges leading into / out of a node of a directed graph.</p>
<p>The node below has an <a href="#BM_(indegree)%2Fi">indegree</a> of 3 and an <a href="#BM_(outdegree)%2Fi">outdegree</a> of 1.</p>
<p><img src="svgbob_7159ebaf654fe57b46d2f14a29a73d50e280afe6.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a name="BM_(balanced%20node)%2Fi"></a><strong>balanced node</strong> <a name="BM_(balanced)_NODE%2Fi"></a><strong></strong> - A node of a directed graph that has an equal <a href="#BM_(indegree)%2Fi">indegree</a> and <a href="#BM_(outdegree)%2Fi">outdegree</a>. That is, the number of edges coming in is equal to the number of edges going out.</p>
<p>The node below has an <a href="#BM_(indegree)%2Fi">indegree</a> and <a href="#BM_(outdegree)%2Fi">outdegree</a> of 1. It is <a href="#BM_(balanced)_NODE%2Fi">balanced</a>.</p>
<p><img src="svgbob_7c49a4fbe8fc85173147545e7a459bca61c84f26.svg" alt="Kroki diagram output"></p>
<p></p>
</li>
<li>
<p><a name="BM_(balanced%20graph)%2Fi"></a><strong>balanced graph</strong> <a name="BM_(balanced)_GRAPH%2Fi"></a><strong></strong> - A directed graph where ever node is <a href="#BM_(balanced)_NODE%2Fi">balanced</a>.</p>
<p>The graph below is <a href="#BM_(balanced)_GRAPH%2Fi">balanced</a> because all nodes are <a href="#BM_(balanced)_NODE%2Fi">balanced</a>.</p>
<p><img src="svgbob_244dd1930eb16ef58be2261f1ffb9277ad2d793a.svg" alt="Kroki diagram output"></p>
<table>
<thead>
<tr>
<th>Node</th>
<th><a href="#BM_(indegree)%2Fi">Indegree</a></th>
<th><a href="#BM_(outdegree)%2Fi">Outdegree</a></th>
</tr>
</thead>
<tbody>
<tr>
<td>A</td>
<td>1</td>
<td>1</td>
</tr>
<tr>
<td>B</td>
<td>1</td>
<td>1</td>
</tr>
<tr>
<td>C</td>
<td>2</td>
<td>2</td>
</tr>
<tr>
<td>D</td>
<td>1</td>
<td>1</td>
</tr>
<tr>
<td>F</td>
<td>1</td>
<td>1</td>
</tr>
</tbody>
</table>
<p></p>
</li>
<li>
<p><a name="BM_(overlap%20graph)%2Fi"></a><strong>overlap graph</strong> - A graph representing the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s making up a string. Specifically, the graph is built in 2 steps:</p>
<ol>
<li>
<p>Each node is a <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>.</p>
<p><img src="svgbob_ba61508d2d4f3dfef58ce5783f8e22f49ddec891.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>Each edge is between overlapping <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s (nodes), where the ...</p>
<ul>
<li>source node has the overlap in its suffix .</li>
<li>destination node has the overlap in its prefix.</li>
</ul>
<p><img src="svgbob_679d24959920acb04c059d6e53dee545fff77300.svg" alt="Kroki diagram output"></p>
</li>
</ol>
<p><a href="#BM_(overlap%20graph)%2Fi">Overlap graph</a>s used for <a href="#BM_(genome)%2Fi">genome</a> <a href="#BM_(assembly%7Cassemble)%2Fi">assembly</a>.</p>
</li>
<li>
<p><a name="BM_(de%20Bruijn%20graph)%2Fi"></a><strong>de Bruijn graph</strong> - A special graph representing the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s making up a string. Specifically, the graph is built in 2 steps:</p>
<ol>
<li>
<p>Each <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> is represented as an edge connecting 2 nodes. The ...</p>
<ul>
<li>source node represents the first 0 to n-1 elements of the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>,</li>
<li>destination node represents last 1 to n elements of the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>,</li>
<li>and edge represents the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>.</li>
</ul>
<p>For example, ...</p>
<p><img src="svgbob_8e6112b97be535a52f2be925bb5b56b546699aaa.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>Each node representing the same value is merged together to form the graph.</p>
<p>For example, ...</p>
<p><img src="svgbob_cfa519e3b0b53ea602d7a16720f54ea9c85f3f08.svg" alt="Kroki diagram output"></p>
</li>
</ol>
<p><a href="#BM_(de%20Bruijn%20graph)%2Fi">De Bruijn graph</a>s are used for <a href="#BM_(genome)%2Fi">genome</a> <a href="#BM_(assembly%7Cassemble)%2Fi">assembly</a>. It's much faster to <a href="#BM_(assembly%7Cassemble)%2Fi">assemble</a> a <a href="#BM_(genome)%2Fi">genome</a> from a <a href="#BM_(de%20Bruijn%20graph)%2Fi">de Bruijn graph</a> than it is to from an <a href="#BM_(overlap%20graph)%2Fi">overlap graph</a>s.</p>
<p><a href="#BM_(de%20Bruijn%20graph)%2Fi">De Bruijn graph</a>s were originally invented to solve the <a href="#BM_(k-universal%7C%5Cd%2B-universal)%2Fi">k-universal</a> string problem.</p>
</li>
<li>
<p><a name="BM_(k-universal%7C%5Cd%2B-universal)%2Fi"></a><strong>k-universal</strong> - For some alphabet and k, a string is considered <a href="#BM_(k-universal%7C%5Cd%2B-universal)%2Fi">k-universal</a> if it contains every <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> for that alphabet exactly once. For example, for an alphabet containing only 0 and 1 (binary) and k=3, a <a href="#BM_(k-universal%7C%5Cd%2B-universal)%2Fi">3-universal</a> string would be 0001110100 because it contains every <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">3-mer</a> exactly once:</p>
<ul>
<li>000: <strong>000</strong>1110100</li>
<li>001: 0<strong>001</strong>110100</li>
<li>010: 000111<strong>010</strong>0</li>
<li>011: 00<strong>011</strong>10100</li>
<li>100: 0001110<strong>100</strong></li>
<li>101: 00011<strong>101</strong>00</li>
<li>110: 0001<strong>110</strong>100</li>
<li>111: 000<strong>111</strong>0100</li>
</ul>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>This is effectively <a href="#BM_(assembly%7Cassemble)%2Fi">assembly</a>. There are a set of <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s and they're being stitched together to form a larger string. The only difference is that the elements aren't nucleotides.</p>
</div>
<p><a href="#BM_(de%20Bruijn%20graph)%2Fi">De Bruijn graph</a>s were invented in an effort to construct <a href="#BM_(k-universal%7C%5Cd%2B-universal)%2Fi">k-universal</a> strings for arbitrary values of k. For example, given the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s in the example above (000, 001, ...), a <a href="#BM_(k-universal%7C%5Cd%2B-universal)%2Fi">k-universal</a> string can be found by constructing a <a href="#BM_(de%20Bruijn%20graph)%2Fi">de Bruijn graph</a> from the <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s and finding a <a href="#BM_(Eulerian%20cycle)%2Fi">Eulerian cycle</a> in that graph.</p>
<p><img src="svgbob_f05ffef51444183bdc007d88010e832876608ed7.svg" alt="Kroki diagram output"></p>
<p>There are multiple <a href="#BM_(Eulerian%20cycle)%2Fi">Eulerian cycle</a>s in the graph, meaning that there are multiple <a href="#BM_(k-universal%7C%5Cd%2B-universal)%2Fi">3-universal</a> strings:</p>
<ul>
<li>0001110100</li>
<li>0011101000</li>
<li>1110001011</li>
<li>1100010111</li>
<li>...</li>
</ul>
<p>For larger values of k (e.g. 20), finding <a href="#BM_(k-universal%7C%5Cd%2B-universal)%2Fi">k-universal</a> strings would be too computationally intensive without <a href="#BM_(de%20Bruijn%20graph)%2Fi">De Bruijn graph</a>s and <a href="#BM_(Eulerian%20cycle)%2Fi">Eulerian cycle</a>s.</p>
</li>
<li>
<p><a name="BM_(coverage)_SEQ%2Fi"></a><strong>coverage</strong> - Given a <a href="#BM_(substring)%2Fi">substring</a> from some larger <a href="#BM_(sequence)%2Fi">sequence</a> that was reconstructed from a set of <a href="#BM_(fragment)_SEQ%2Fi">fragment</a>s, the <a href="#BM_(coverage)_SEQ%2Fi">coverage</a> of that <a href="#BM_(substring)%2Fi">substring</a> is the number of <a href="#BM_%5Cb(read)_SEQ%2Fi">read</a>s used to construct it. The <a href="#BM_(substring)%2Fi">substring</a> length is typically 1: the <a href="#BM_(coverage)_SEQ%2Fi">coverage</a> for each position of the <a href="#BM_(sequence)%2Fi">sequence</a>.</p>
<p><img src="svgbob_8dd35ba9a6498739b4dd8f3759fd49dec3d1e9c8.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a name="BM_(read%20breaking%7Cread-breaking%7Cbreaking%20reads)%2Fi"></a><strong>read breaking</strong> - The concept of taking multiple <a href="#BM_%5Cb(read)_SEQ%2Fi">read</a>s and breaking them up into smaller <a href="#BM_%5Cb(read)_SEQ%2Fi">read</a>s.</p>
<p><img src="svgbob_d9e103b16312cfe660d19df9305a9361ac1241ac.svg" alt="Kroki diagram output"></p>
<p>When <a href="#BM_(read%20breaking%7Cread-breaking%7Cbreaking%20reads)%2Fi">read breaking</a>, smaller <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a>s result in better <a href="#BM_(coverage)_SEQ%2Fi">coverage</a> but also make the <a href="#BM_(de%20Bruijn%20graph)%2Fi">de Bruijn graph</a> more tangled. The more tangled the <a href="#BM_(de%20Bruijn%20graph)%2Fi">de Bruijn graph</a> is, the harder it is to infer the full <a href="#BM_(sequence)%2Fi">sequence</a>.</p>
<p>In the example above, the average <a href="#BM_(coverage)_SEQ%2Fi">coverage</a>...</p>
<ul>
<li>for the left-hand side (original) is 2.1.</li>
<li>for the right-hand side (broken) is 4.</li>
</ul>
<p>See also: <a href="#BM_(read-pair%20breaking%7Cread%20pair%20breaking%7Cbreaking%20read-pairs%7Cbreaking%20read%20pairs)%2Fi">read-pair breaking</a>.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>What purpose does this actually serve? Mimicking 1 long <a href="#BM_%5Cb(read)_SEQ%2Fi">read</a> as n shorter <a href="#BM_%5Cb(read)_SEQ%2Fi">read</a>s isn't equivalent to actually having <a href="#BM_(sequencing%7Csequenced)%2Fi">sequenced</a> those n shorter <a href="#BM_%5Cb(read)_SEQ%2Fi">read</a>s. For example, what if the longer <a href="#BM_%5Cb(read)_SEQ%2Fi">read</a> being broken up has an error? That error replicates when breaking into n shorter <a href="#BM_%5Cb(read)_SEQ%2Fi">read</a>s, which gives a false sense of having good <a href="#BM_(coverage)_SEQ%2Fi">coverage</a> and makes it seem as if it wasn't an error.</p>
</div>
</li>
<li>
<p><a name="BM_(read-pair%20breaking%7Cread%20pair%20breaking%7Cbreaking%20read-pairs%7Cbreaking%20read%20pairs)%2Fi"></a><strong>read-pair breaking</strong> - The concept of taking multiple <a href="#BM_(read-pair%7Cread%20pair)%2Fi">read-pair</a>s and breaking them up into <a href="#BM_(read-pair%7Cread%20pair)%2Fi">read-pair</a>s with a smaller k.</p>
<p><img src="svgbob_eb8f1251b4e293a9662ade2c0a4452a7c7884f2f.svg" alt="Kroki diagram output"></p>
<p>When <a href="#BM_(read-pair%20breaking%7Cread%20pair%20breaking%7Cbreaking%20read-pairs%7Cbreaking%20read%20pairs)%2Fi">read-pair breaking</a>, a smaller k results in better <a href="#BM_(coverage)_SEQ%2Fi">coverage</a> but also make the <a href="#BM_(de%20Bruijn%20graph)%2Fi">de Bruijn graph</a> more tangled. The more tangled the <a href="#BM_(de%20Bruijn%20graph)%2Fi">de Bruijn graph</a> is, the harder it is to infer the full <a href="#BM_(sequence)%2Fi">sequence</a>.</p>
<p>In the example above, the average <a href="#BM_(coverage)_SEQ%2Fi">coverage</a>...</p>
<ul>
<li>for the left-hand side (original) is 1.6.</li>
<li>for the right-hand side (broken) is 2.5.</li>
</ul>
<p>See also: <a href="#BM_(read%20breaking%7Cread-breaking%7Cbreaking%20reads)%2Fi">read breaking</a>.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>What purpose does this actually serve? Mimicking 1 long <a href="#BM_(read-pair%7Cread%20pair)%2Fi">read-pair</a> as n shorter <a href="#BM_(read-pair%7Cread%20pair)%2Fi">read-pair</a>s isn't equivalent to actually having <a href="#BM_(sequencing%7Csequenced)%2Fi">sequenced</a> those n shorter <a href="#BM_(read-pair%7Cread%20pair)%2Fi">read-pair</a>s. For example, what if the longer <a href="#BM_(read-pair%7Cread%20pair)%2Fi">read-pair</a> being broken up has an error? That error replicates when breaking into n shorter <a href="#BM_(read-pair%7Cread%20pair)%2Fi">read-pair</a>s, which gives a false sense of having good <a href="#BM_(coverage)_SEQ%2Fi">coverage</a> and makes it seem as if it wasn't an error.</p>
</div>
</li>
<li>
<p><a name="BM_(contig)s%3F%5Cb%2Fi"></a><strong>contig</strong> - An unambiguous stretch of DNA derived by searching an <a href="#BM_(overlap%20graph)%2Fi">overlap graph</a> / <a href="#BM_(de%20Bruijn%20graph)%2Fi">de Bruijn graph</a> for paths that are the longest possible stretches of non-branching nodes (<a href="#BM_(indegree)%2Fi">indegree</a> and <a href="#BM_(outdegree)%2Fi">outdegree</a> of 1). Each stretch will be a path that's either  ...</p>
<ul>
<li>
<p>a line: each node has an <a href="#BM_(indegree)%2Fi">indegree</a> and <a href="#BM_(outdegree)%2Fi">outdegree</a> of 1.</p>
<p><img src="svgbob_8f50e668ad56e2eeff3c39e6aa89792f6f9444f9.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>a cycle: each node has an <a href="#BM_(indegree)%2Fi">indegree</a> and <a href="#BM_(outdegree)%2Fi">outdegree</a> of 1 and it loops.</p>
<p><img src="svgbob_9a5c6dc10a8a016d6f43d4c9bdd6243d0ee5b332.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>a line sandwiched between branching nodes: nodes in between have an <a href="#BM_(indegree)%2Fi">indegree</a> and <a href="#BM_(outdegree)%2Fi">outdegree</a> of 1 but either...</p>
<ul>
<li>starts at a node where <a href="#BM_(indegree)%2Fi">indegree</a> != 1 but <a href="#BM_(outdegree)%2Fi">outdegree</a> == 1 (incoming branch),</li>
<li>or ends at a node where <a href="#BM_(indegree)%2Fi">indegree</a> == 1 but <a href="#BM_(outdegree)%2Fi">outdegree</a> != 1 (outgoing branch),</li>
<li>or both.</li>
</ul>
<p><img src="svgbob_057bcd0e831c50ca2384b8da9ea9cd1f46186b4a.svg" alt="Kroki diagram output"></p>
</li>
</ul>
<p>Real-world complications with DNA <a href="#BM_(sequencing%7Csequenced)%2Fi">sequencing</a> make de Bruijn / <a href="#BM_(overlap%20graph)%2Fi">overlap graph</a>s too tangled to guess a full <a href="#BM_(genome)%2Fi">genome</a>: both strands of <a href="#BM_(double%20stranded%20DNA%7Cdouble-stranded%20DNA)%2Fi">double-stranded DNA</a> are <a href="#BM_(sequencing%7Csequenced)%2Fi">sequenced</a> and mixed into the graph, <a href="#BM_(sequencing%20error)%2Fi">sequencing error</a>s make into the graph, repeats regions of the <a href="#BM_(genome)%2Fi">genome</a> can't be reliably handled by the graph, poor <a href="#BM_(coverage)_SEQ%2Fi">coverage</a>, etc.. As such, biologists / bioinformaticians have no choice but to settle on <a href="#BM_(contig)s%3F%5Cb%2Fi">contig</a>s.</p>
<p><img src="svgbob_db2c0ed38b3144b9db68bc5e39f4b97ce471f510.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a name="BM_(ribonucleotide)%2Fi"></a><strong>ribonucleotide</strong> - Elements that make up RNA, similar to how nucleotides are the elements that make up DNA.</p>
<ul>
<li>A = Adenine (same as nucleotide)</li>
<li>C = Cytosine (same as nucleotide)</li>
<li>G = Guanine (same as nucleotide)</li>
<li>U = Uracil (replace nucleotide Thymine)</li>
</ul>
</li>
<li>
<p><a name="BM_(antibiotic)%2Fi"></a><strong>antibiotic</strong> - A substance (typically an enzyme) for killing, preventing, or inhibiting the grow of bacterial infections.</p>
</li>
<li>
<p><a name="BM_(amino%20acid)%2Fi"></a><strong>amino acid</strong> - The building blocks of <a href="#BM_(peptide)%2Fi">peptide</a>s / proteins, similar to how nucleotides are the building blocks of DNA.</p>
<p>See <a href="#BM_(proteinogenic%20amino%20acid)%2Fi">proteinogenic amino acid</a> for the list of 20 <a href="#BM_(amino%20acid)%2Fi">amino acid</a>s used during the <a href="#BM_(translation%7Ctranslated%7Ctranslate)%2Fi">translation</a>.</p>
</li>
<li>
<p><a name="BM_(proteinogenic%20amino%20acid)%2Fi"></a><strong>proteinogenic amino acid</strong> - <a href="#BM_(amino%20acid)%2Fi">Amino acid</a>s that are used during <a href="#BM_(translation%7Ctranslated%7Ctranslate)%2Fi">translation</a>. These are the 20 <a href="#BM_(amino%20acid)%2Fi">amino acid</a>s that the ribosome <a href="#BM_(translation%7Ctranslated%7Ctranslate)%2Fi">translate</a>s from <a href="#BM_(codon%7Cgenetic%20code)%2Fi">codon</a>s. In contrast, there are many other non-<a href="#BM_(proteinogenic%20amino%20acid)%2Fi">proteinogenic amino acid</a>s that are used for <a href="#BM_(non-ribosomal%20peptide)%2Fi">non-ribosomal peptide</a>s.</p>
<p>The term "proteinogenic" means "protein creating".</p>
<table>
<thead>
<tr>
<th>1 Letter Code</th>
<th>3 Letter Code</th>
<th><a href="#BM_(amino%20acid)%2Fi">Amino Acid</a></th>
<th>Mass (<a href="#BM_(dalton)%2Fi">dalton</a>s)</th>
</tr>
</thead>
<tbody>
<tr>
<td>A</td>
<td>Ala</td>
<td>Alanine</td>
<td>71.04</td>
</tr>
<tr>
<td>C</td>
<td>Cys</td>
<td>Cysteine</td>
<td>103.01</td>
</tr>
<tr>
<td>D</td>
<td>Asp</td>
<td>Aspartic acid</td>
<td>115.03</td>
</tr>
<tr>
<td>E</td>
<td>Glu</td>
<td>Glutamic acid</td>
<td>129.04</td>
</tr>
<tr>
<td>F</td>
<td>Phe</td>
<td>Phenylalanine</td>
<td>147.07</td>
</tr>
<tr>
<td>G</td>
<td>Gly</td>
<td>Glycine</td>
<td>57.02</td>
</tr>
<tr>
<td>H</td>
<td>His</td>
<td>Histidine</td>
<td>137.06</td>
</tr>
<tr>
<td>I</td>
<td>Ile</td>
<td>Isoleucine</td>
<td>113.08</td>
</tr>
<tr>
<td>K</td>
<td>Lys</td>
<td>Lysine</td>
<td>128.09</td>
</tr>
<tr>
<td>L</td>
<td>Leu</td>
<td>Leucine</td>
<td>113.08</td>
</tr>
<tr>
<td>M</td>
<td>Met</td>
<td>Methionine</td>
<td>131.04</td>
</tr>
<tr>
<td>N</td>
<td>Asn</td>
<td>Asparagine</td>
<td>114.04</td>
</tr>
<tr>
<td>P</td>
<td>Pro</td>
<td>Proline</td>
<td>97.05</td>
</tr>
<tr>
<td>Q</td>
<td>Gln</td>
<td>Glutamine</td>
<td>128.06</td>
</tr>
<tr>
<td>R</td>
<td>Arg</td>
<td>Arginine</td>
<td>156.1</td>
</tr>
<tr>
<td>S</td>
<td>Ser</td>
<td>Serine</td>
<td>87.03</td>
</tr>
<tr>
<td>T</td>
<td>Thr</td>
<td>Threonine</td>
<td>101.05</td>
</tr>
<tr>
<td>V</td>
<td>Val</td>
<td>Valine</td>
<td>99.07</td>
</tr>
<tr>
<td>W</td>
<td>Trp</td>
<td>Tryptophan</td>
<td>186.08</td>
</tr>
<tr>
<td>Y</td>
<td>Tyr</td>
<td>Tyrosine</td>
<td>163.06</td>
</tr>
</tbody>
</table>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The masses are monoisotopic masses.</p>
</div>
</li>
<li>
<p><a name="BM_(peptide)%2Fi"></a><strong>peptide</strong> - A short <a href="#BM_(amino%20acid)%2Fi">amino acid</a> chain of at least size two. <a href="#BM_(peptide)%2Fi">Peptide</a>s are considered miniature proteins, but when something should be called a <a href="#BM_(peptide)%2Fi">peptide</a> vs a protein is loosely defined: the cut-off is anywhere between 50 to 100 <a href="#BM_(amino%20acid)%2Fi">amino acid</a>s.</p>
</li>
<li>
<p><a name="BM_(polypeptide)%2Fi"></a><strong>polypeptide</strong> - A <a href="#BM_(peptide)%2Fi">peptide</a> of at least size 10.</p>
</li>
<li>
<p><a name="BM_(amino%20acid%20residue)%2Fi"></a><strong>amino acid residue</strong> - The part of an <a href="#BM_(amino%20acid)%2Fi">amino acid</a> that makes it unique from all others.</p>
<p>When two or more <a href="#BM_(amino%20acid)%2Fi">amino acid</a>s combine to make a <a href="#BM_(peptide)%2Fi">peptide</a>/protein, specific elements are removed from each <a href="#BM_(amino%20acid)%2Fi">amino acid</a>. What remains of each <a href="#BM_(amino%20acid)%2Fi">amino acid</a> is the <a href="#BM_(amino%20acid%20residue)%2Fi">amino acid residue</a>.</p>
</li>
<li>
<p><a name="BM_(cyclopeptide%7Ccyclic%20peptide)%2Fi"></a><strong>cyclopeptide</strong> - A <a href="#BM_(peptide)%2Fi">peptide</a> that doesn't have a start / end. It loops.</p>
<p><img src="svgbob_0de8076d92d299650e3473fc869328891e1ef452.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a name="BM_(linear%20peptide)%2Fi"></a><strong>linear peptide</strong> - A <a href="#BM_(peptide)%2Fi">peptide</a> that has a start and an end. It doesn't loop.</p>
<p><img src="svgbob_1a5d3942fce4284a9ad77c8e4624b7b21573543d.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a name="BM_(subpeptide)%2Fi"></a><strong>subpeptide</strong> - A <a href="#BM_(peptide)%2Fi">peptide</a> derived taking some contiguous piece of a larger <a href="#BM_(peptide)%2Fi">peptide</a>.</p>
<p>A <a href="#BM_(subpeptide)%2Fi">subpeptide</a> can have a length == 1 where a <a href="#BM_(peptide)%2Fi">peptide</a> must have a length &gt; 1. As such, in the case where the <a href="#BM_(subpeptide)%2Fi">subpeptide</a> has a length ...</p>
<ul>
<li>== 1, it isn't considered a <a href="#BM_(peptide)%2Fi">peptide</a>.</li>
<li>&gt; 1, it is considered a <a href="#BM_(peptide)%2Fi">peptide</a>.</li>
</ul>
</li>
<li>
<p><a name="BM_(central%20dogma%20of%20molecular%20biology)%2Fi"></a><strong>central dogma of molecular biology</strong> - The overall concept of <a href="#BM_(transcription%7Ctranscribed%7Ctranscribe)%2Fi">transcription</a> and <a href="#BM_(translation%7Ctranslated%7Ctranslate)%2Fi">translation</a>: Instructions for making a protein are copied from DNA to RNA, then RNA feeds into the ribosome to make that protein (DNA → RNA → Protein).</p>
<p>Most, not all, <a href="#BM_(peptide)%2Fi">peptide</a>s are synthesized as described above. <a href="#BM_(non-ribosomal%20peptide)%2Fi">Non-ribosomal peptide</a>s are synthesized outside of the <a href="#BM_(transcription%7Ctranscribed%7Ctranscribe)%2Fi">transcription</a> and <a href="#BM_(translation%7Ctranslated%7Ctranslate)%2Fi">translation</a>.</p>
</li>
<li>
<p><a name="BM_(non-ribosomal%20peptide)%2Fi"></a><strong>non-ribosomal peptide</strong> <a name="BM_%5Cb(NRP)s%3F%5Cb%2F"></a><strong></strong> - A <a href="#BM_(peptide)%2Fi">peptide</a> that was synthesized by a protein called <a href="#BM_(non-ribosomal%20peptide%20synthetase%7CNRP%20synthetase)%2Fi">NRP synthetase</a> rather than synthesized by a ribosome. <a href="#BM_(non-ribosomal%20peptide%20synthetase%7CNRP%20synthetase)%2Fi">NRP synthetase</a> builds <a href="#BM_(peptide)%2Fi">peptide</a>s one <a href="#BM_(amino%20acid)%2Fi">amino acid</a> at a time without relying on <a href="#BM_(transcription%7Ctranscribed%7Ctranscribe)%2Fi">transcription</a> or <a href="#BM_(translation%7Ctranslated%7Ctranslate)%2Fi">translation</a>.</p>
<p><a href="#BM_(non-ribosomal%20peptide)%2Fi">Non-ribosomal peptide</a>s may be cyclic. Common use-cases for <a href="#BM_(non-ribosomal%20peptide)%2Fi">non-ribosomal peptide</a>s:</p>
<ul>
<li><a href="#BM_(antibiotic)%2Fi">antibiotic</a>s</li>
<li>anti-tumor agents</li>
<li>immunosuppressors</li>
<li>communication between bacteria (quorum sensing)</li>
</ul>
</li>
<li>
<p><a name="BM_(non-ribosomal%20peptide%20synthetase%7CNRP%20synthetase)%2Fi"></a><strong>non-ribosomal peptide synthetase</strong> - A protein responsible for the production of a <a href="#BM_(non-ribosomal%20peptide)%2Fi">non-ribosomal peptide</a>.</p>
</li>
<li>
<p><a name="BM_(adenylation%20domain%7CA%20domain%7CA-domain)%2Fi"></a><strong>adenylation domain</strong> - A segment of an <a href="#BM_(non-ribosomal%20peptide%20synthetase%7CNRP%20synthetase)%2Fi">NRP synthetase</a> protein responsible for the outputting a single <a href="#BM_(amino%20acid)%2Fi">amino acid</a>. For example, the <a href="#BM_(non-ribosomal%20peptide%20synthetase%7CNRP%20synthetase)%2Fi">NRP synthetase</a> responsible for producing Tyrocidine has 10 <a href="#BM_(adenylation%20domain%7CA%20domain%7CA-domain)%2Fi">adenylation domain</a>s, each of which is responsible for outputting a single <a href="#BM_(amino%20acid)%2Fi">amino acid</a> of Tyrocidine.</p>
</li>
<li>
<p><a name="BM_(mass%20spectrometer%7Cmass%20spectrometry)%2Fi"></a><strong>mass spectrometer</strong> - A device that randomly shatters molecules into pieces and measures the mass-to-charge of those pieces. The output of the device is a plot called a <a href="#BM_(spectrum)_MS%2Fi">spectrum</a>.</p>
<p>Note that <a href="#BM_(mass%20spectrometer%7Cmass%20spectrometry)%2Fi">mass spectrometer</a>s have various real-world practical problems. Specifically, they ...</p>
<ul>
<li>may not capture all possible pieces from the intended molecules (missing mass-to-charge ratios).</li>
<li>may capture pieces from unintended molecules (faulty mass-to-charge ratios).</li>
<li>will likely introduce noise into the pieces they capture.</li>
</ul>
</li>
<li>
<p><a name="BM_(spectrum)_MS%2Fi"></a><strong>spectrum</strong> - The output of a <a href="#BM_(mass%20spectrometer%7Cmass%20spectrometry)%2Fi">mass spectrometer</a>. The...</p>
<ul>
<li>x-axis is the mass-to-charge ratio.</li>
<li>y-axis is the intensity of that mass-to-charge ratio (how much more / less did that mass-to-charge appear compared to the others).</li>
</ul>
<p><img src="svgbob_2b5ef43ae7572c2c6bf63687c53860a71830d160.svg" alt="Kroki diagram output"></p>
<p>Note that <a href="#BM_(mass%20spectrometer%7Cmass%20spectrometry)%2Fi">mass spectrometer</a>s have various real-world practical problems. Specifically, they ...</p>
<ul>
<li>may not capture all possible pieces from the intended molecules (missing mass-to-charge ratios).</li>
<li>may capture pieces from unintended molecules (faulty mass-to-charge ratios).</li>
<li>will likely introduce noise into the pieces they capture.</li>
</ul>
<p>As such, these plots aren't exact.</p>
</li>
<li>
<p><a name="BM_(experimental%20spectrum)%2Fi"></a><strong>experimental spectrum</strong> - List of potential fragment masses derived from a <a href="#BM_(spectrum)_MS%2Fi">spectrum</a>. That is, the molecules fed into the <a href="#BM_(mass%20spectrometer%7Cmass%20spectrometry)%2Fi">mass spectrometer</a> were randomly fragmented and each fragment had its mass-to-charge ratio measured. From there, each mass-to-charge ratio was converted to a set of potential masses.</p>
<p>The masses in an <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> ...</p>
<ul>
<li>may not capture all possible fragments for the intended molecule (missing masses).</li>
<li>may capture fragments from unintended molecules (faulty masses).</li>
<li>will likely contain noise.</li>
</ul>
<p>In the context of <a href="#BM_(peptide)%2Fi">peptide</a>s, the <a href="#BM_(mass%20spectrometer%7Cmass%20spectrometry)%2Fi">mass spectrometer</a> is expected to fragment based on the bonds holding the individual <a href="#BM_(amino%20acid)%2Fi">amino acid</a>s together. For example, given the <a href="#BM_(linear%20peptide)%2Fi">linear peptide</a> NQY, the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> may include the masses for [N, Q, ?, ?, QY, ?, NQY] (? indicate faulty masses, Y and NQ missing).</p>
</li>
<li>
<p><a name="BM_(theoretical%20spectrum)%2Fi"></a><strong>theoretical spectrum</strong> - List of all of possible fragment masses for a molecule in addition to 0 and the mass of the entire molecule. This is what the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> would be in a perfect world: no missing masses, no faulty masses, no noise, only a single possible mass for each mass-to-charge ratio.</p>
<p>In the context of <a href="#BM_(peptide)%2Fi">peptide</a>s, the <a href="#BM_(mass%20spectrometer%7Cmass%20spectrometry)%2Fi">mass spectrometer</a> is expected to fragment based on the bonds holding the individual <a href="#BM_(amino%20acid)%2Fi">amino acid</a>s together. For example, given the <a href="#BM_(linear%20peptide)%2Fi">linear peptide</a> NQY, the <a href="#BM_(theoretical%20spectrum)%2Fi">theoretical spectrum</a> will include the masses for [0, N, Q, Y, NQ, QY, NQY]. It shouldn't include masses for partial <a href="#BM_(amino%20acid)%2Fi">amino acid</a>s. For example, it shouldn't include NQY breaking into 2 pieces by splitting Q, such that one half has N and part of Q, and the other has the remaining part of Q with Y.</p>
</li>
<li>
<p><a name="BM_(spectrum%20convolution)%2Fi"></a><strong>spectrum convolution</strong> - An operation used to derive <a href="#BM_(amino%20acid)%2Fi">amino acid</a> masses that probably come from the <a href="#BM_(peptide)%2Fi">peptide</a> used to generate that <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>. That is, it generates a list of <a href="#BM_(amino%20acid)%2Fi">amino acid</a> masses that could have been for the <a href="#BM_(peptide)%2Fi">peptide</a> that generated the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>.</p>
<p>The operation derives <a href="#BM_(amino%20acid)%2Fi">amino acid</a> masses by subtracting <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> masses from each other. For example, the following <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> is for the <a href="#BM_(linear%20peptide)%2Fi">linear peptide</a> NQY: [113.9, 115.1, 136.2, 162.9, 242.0, 311.1, 346.0, 405.2]. Performing 242.0 - 113.9 results in 128.1, which is very close to the mass for <a href="#BM_(amino%20acid)%2Fi">amino acid</a> Y.</p>
<p>Note how the mass for Y was derived from the masses in <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> even though it's missing from the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a> itself:</p>
<ul>
<li>Mass of N is 114. 2 masses are close to 114 in the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>: [113.9, 115.1].</li>
<li>Mass of Q is 163. 1 mass is close to 163 in the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>: [162.9].</li>
<li>Mass of Y is 128. 0 masses are close to 128 in the <a href="#BM_(experimental%20spectrum)%2Fi">experimental spectrum</a>: [].</li>
</ul>
</li>
<li>
<p><a name="BM_(dalton)%2Fi"></a><strong>dalton</strong> <a name="BM_%5Cb%5Cd%2B(%3F%3A%5C.%5Cd%2B)%3F(Da)%5Cb%2Fi"></a><strong></strong> - A unit of measurement used in physics and chemistry. 1 <a href="#BM_(dalton)%2Fi">Dalton</a> is approximately the mass of a single proton / neutron, derived by taking the mass of a carbon-12 atom and dividing it by 12.</p>
</li>
<li>
<p><a name="BM_(codon%7Cgenetic%20code)%2Fi"></a><strong>codon</strong> - A <a href="#BM_(sequence)%2Fi">sequence</a> of 3 <a href="#BM_(ribonucleotide)%2Fi">ribonucleotide</a>s that maps to an <a href="#BM_(amino%20acid)%2Fi">amino acid</a> or a stop marker. During <a href="#BM_(translation%7Ctranslated%7Ctranslate)%2Fi">translation</a>, the ribosome <a href="#BM_(translation%7Ctranslated%7Ctranslate)%2Fi">translate</a>s the RNA to a protein 3 <a href="#BM_(ribonucleotide)%2Fi">ribonucleotide</a>s at a time:</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The stop marker tells the ribosome to stop translating / the protein is complete.</p>
</div>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The <a href="#BM_(codon%7Cgenetic%20code)%2Fi">codon</a>s are listed as <a href="#BM_(ribonucleotide)%2Fi">ribonucleotide</a>s (RNA). For nucleotides (DNA), swap U with T.</p>
</div>
<table>
<thead>
<tr>
<th>1 Letter Code</th>
<th>3 Letter Code</th>
<th><a href="#BM_(amino%20acid)%2Fi">Amino Acid</a></th>
<th><a href="#BM_(codon%7Cgenetic%20code)%2Fi">Codon</a>s</th>
</tr>
</thead>
<tbody>
<tr>
<td>A</td>
<td>Ala</td>
<td>Alanine</td>
<td>GCA, GCC, GCG, GCU</td>
</tr>
<tr>
<td>C</td>
<td>Cys</td>
<td>Cysteine</td>
<td>UGC, UGU</td>
</tr>
<tr>
<td>D</td>
<td>Asp</td>
<td>Aspartic acid</td>
<td>GAC, GAU</td>
</tr>
<tr>
<td>E</td>
<td>Glu</td>
<td>Glutamic acid</td>
<td>GAA, GAG</td>
</tr>
<tr>
<td>F</td>
<td>Phe</td>
<td>Phenylalanine</td>
<td>UUC, UUU</td>
</tr>
<tr>
<td>G</td>
<td>Gly</td>
<td>Glycine</td>
<td>GGA, GGC, GGG, GGU</td>
</tr>
<tr>
<td>H</td>
<td>His</td>
<td>Histidine</td>
<td>CAC, CAU</td>
</tr>
<tr>
<td>I</td>
<td>Ile</td>
<td>Isoleucine</td>
<td>AUA, AUC, AUU</td>
</tr>
<tr>
<td>K</td>
<td>Lys</td>
<td>Lysine</td>
<td>AAA, AAG</td>
</tr>
<tr>
<td>L</td>
<td>Leu</td>
<td>Leucine</td>
<td>CUA, CUC, CUG, CUU, UUA, UUG</td>
</tr>
<tr>
<td>M</td>
<td>Met</td>
<td>Methionine</td>
<td>AUG</td>
</tr>
<tr>
<td>N</td>
<td>Asn</td>
<td>Asparagine</td>
<td>AAC, AAU</td>
</tr>
<tr>
<td>P</td>
<td>Pro</td>
<td>Proline</td>
<td>CCA, CCC, CCG, CCU</td>
</tr>
<tr>
<td>Q</td>
<td>Gln</td>
<td>Glutamine</td>
<td>CAA, CAG</td>
</tr>
<tr>
<td>R</td>
<td>Arg</td>
<td>Arginine</td>
<td>AGA, AGG, CGA, CGC, CGG, CGU</td>
</tr>
<tr>
<td>S</td>
<td>Ser</td>
<td>Serine</td>
<td>AGC, AGU, UCA, UCC, UCG, UCU</td>
</tr>
<tr>
<td>T</td>
<td>Thr</td>
<td>Threonine</td>
<td>ACA, ACC, ACG, ACU</td>
</tr>
<tr>
<td>V</td>
<td>Val</td>
<td>Valine</td>
<td>GUA, GUC, GUG, GUU</td>
</tr>
<tr>
<td>W</td>
<td>Trp</td>
<td>Tryptophan</td>
<td>UGG</td>
</tr>
<tr>
<td>Y</td>
<td>Tyr</td>
<td>Tyrosine</td>
<td>UAC, UAU</td>
</tr>
<tr>
<td>*</td>
<td>*</td>
<td><strong>STOP</strong></td>
<td>UAA, UAG, UGA</td>
</tr>
</tbody>
</table>
</li>
<li>
<p><a name="BM_(reading%20frame)%2Fi"></a><strong>reading frame</strong> - The different ways of dividing a DNA string into <a href="#BM_(codon%7Cgenetic%20code)%2Fi">codon</a>s. Specifically, there are 6 different ways that a DNA string can be divided into <a href="#BM_(codon%7Cgenetic%20code)%2Fi">codon</a>s:</p>
<ul>
<li>You can start dividing at index 0, 1, or 2.</li>
<li>You can divide either the DNA string itself or the <a href="#BM_(reverse%20complement)%2Fi">reverse complement</a>ing DNA string.</li>
</ul>
<p>For example, given the string ATGTTCCATTAA, the following <a href="#BM_(codon%7Cgenetic%20code)%2Fi">codon</a> division are possible:</p>
<table>
<thead>
<tr>
<th>DNA</th>
<th>Start Index</th>
<th>Discard Prefix</th>
<th><a href="#BM_(codon%7Cgenetic%20code)%2Fi">Codon</a>s</th>
<th>Discard Suffix</th>
</tr>
</thead>
<tbody>
<tr>
<td>ATGTTCCATTAA</td>
<td>0</td>
<td></td>
<td>ATG, TTC, CAT, TAA</td>
<td></td>
</tr>
<tr>
<td>ATGTTCCATTAA</td>
<td>1</td>
<td>A</td>
<td>TGT, TCC, ATT</td>
<td>AA</td>
</tr>
<tr>
<td>ATGTTCCATTAA</td>
<td>2</td>
<td>AT</td>
<td>GTT, CCA, TTA</td>
<td>A</td>
</tr>
<tr>
<td>TTAATGGAACAT</td>
<td>0</td>
<td></td>
<td>TTA, ATG, GAA, CAT</td>
<td></td>
</tr>
<tr>
<td>TTAATGGAACAT</td>
<td>1</td>
<td>T</td>
<td>TAA, TGG, AAC</td>
<td>AT</td>
</tr>
<tr>
<td>TTAATGGAACAT</td>
<td>2</td>
<td>TT</td>
<td>AAT, GGA, ACA</td>
<td>T</td>
</tr>
</tbody>
</table>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>TTAATGGAACAT is the <a href="#BM_(reverse%20complement)%2Fi">reverse complement</a> of ATGTTCCATTAA.</p>
</div>
</li>
<li>
<p><a name="BM_(encode%7Cencoding%7Cdecode%7Cdecoding)%2Fi"></a><strong>encode</strong> - When a DNA string or its <a href="#BM_(reverse%20complement)%2Fi">reverse complement</a> is made up of the <a href="#BM_(codon%7Cgenetic%20code)%2Fi">codon</a>s required for an <a href="#BM_(amino%20acid)%2Fi">amino acid</a> <a href="#BM_(sequence)%2Fi">sequence</a>. For example, ACAGTA <a href="#BM_(encode%7Cencoding%7Cdecode%7Cdecoding)%2Fi">encode</a>s for the <a href="#BM_(amino%20acid)%2Fi">amino acid</a> <a href="#BM_(sequence)%2Fi">sequence</a>...</p>
<ul>
<li>Threonine-Valine</li>
<li>Tyrosine-Cysteine (derived from <a href="#BM_(reverse%20complement)%2Fi">reverse complement</a>)</li>
</ul>
</li>
<li>
<p><a name="BM_(branch-and-bound%20algorithm%7Cbranch%20and%20bound%20algorithm)%2Fi"></a><strong>branch-and-bound algorithm</strong> - A bruteforce algorithm that enumerates candidates to explore at each step but also discards untenable candidates using various checks. The enumeration of candidates is the branching step, while the culling of untenable candidates is the bounding step.</p>
</li>
<li>
<p><a name="BM_(subsequence)%2Fi"></a><strong>subsequence</strong> - A <a href="#BM_(sequence)%2Fi">sequence</a> derived by traversing some other <a href="#BM_(sequence)%2Fi">sequence</a> in order and choosing which elements to keep vs delete. For example, can is a <a href="#BM_(subsequence)%2Fi">subsequence</a> of cation.</p>
<p><img src="svgbob_c25309d3982681fa348879d4c3fe065899f82016.svg" alt="Kroki diagram output"></p>
<p>Not to be confused with <a href="#BM_(substring)%2Fi">substring</a>. A <a href="#BM_(substring)%2Fi">substring</a> may also be a <a href="#BM_(subsequence)%2Fi">subsequence</a>, but a <a href="#BM_(subsequence)%2Fi">subsequence</a> won't necessarily be a <a href="#BM_(substring)%2Fi">substring</a>.</p>
</li>
<li>
<p><a name="BM_(substring)%2Fi"></a><strong>substring</strong> - A <a href="#BM_(sequence)%2Fi">sequence</a> derived by taking a contiguous part of some other <a href="#BM_(sequence)%2Fi">sequence</a> (order of elements maintained). For example, cat is a <a href="#BM_(substring)%2Fi">substring</a> of cation.</p>
<p><img src="svgbob_1c969351b840ffdbd3172ad117673e5506efbd17.svg" alt="Kroki diagram output"></p>
<p>Not to be confused with <a href="#BM_(subsequence)%2Fi">subsequence</a>. A <a href="#BM_(substring)%2Fi">substring</a> may also be a <a href="#BM_(subsequence)%2Fi">subsequence</a>, but a <a href="#BM_(subsequence)%2Fi">subsequence</a> won't necessarily be a <a href="#BM_(substring)%2Fi">substring</a>.</p>
</li>
<li>
<p><a name="BM_(topological%20order)%2Fi"></a><strong>topological order</strong> - A 1-dimensional ordering of nodes in a <a href="#BM_(directed%20acyclic%20graph)%2Fi">directed acyclic graph</a> in which each node is ahead of all of its predecessors / parents. In other words, the node is ahead of all other nodes that connect to it.</p>
<p>For example, the graph ...</p>
<p><img src="svgbob_6a2f3c0925fa0e1b69f5e2c59e731dc23f8c54e5.svg" alt="Kroki diagram output"></p>
<p>... the <a href="#BM_(topological%20order)%2Fi">topological order</a> is either [A, B, C, D, E] or [A, B, C, E, D]. Both are correct.</p>
</li>
<li>
<p><a name="BM_(longest%20common%20subsequence)%2Fi"></a><strong>longest common subsequence</strong>  <a name="BM_%5Cb(LCS)es%3F%5Cb%2F"></a><strong></strong> - A common <a href="#BM_(subsequence)%2Fi">subsequence</a> between a set of strings of which is the longest out of all possible common <a href="#BM_(subsequence)%2Fi">subsequence</a>s. There may be more than one per set.</p>
<p>For example, AACCTTGG and ACACTGTGA share a <a href="#BM_(longest%20common%20subsequence)%2Fi">longest common subsequence</a> of...</p>
<ul>
<li>
<p>ACCTGG...</p>
<p><img src="svgbob_f1eadf7cf7bbf51eeb3471a6a5610a06f4dc2d1b.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>AACTGG...</p>
<p><img src="svgbob_a9b0360f8eeb866179f567437d7cfc36c167251a.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>etc..</p>
</li>
</ul>
</li>
<li>
<p><a name="BM_(sequence%20alignment)%2Fi"></a><strong>sequence alignment</strong> - Given a set of <a href="#BM_(sequence)%2Fi">sequence</a>s, a <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a> is a set of operations applied to each position in an effort to line up the <a href="#BM_(sequence)%2Fi">sequence</a>s. These operations include:</p>
<ul>
<li>insert/delete (<a href="#BM_(indel)%2Fi">indel</a> for short).</li>
<li>replace (also referred to as mismatch).</li>
<li>keep matching (also referred to as match).</li>
</ul>
<p>For example, the <a href="#BM_(sequence)%2Fi">sequence</a>s MAPLE and TABLE may be aligned by performing...</p>
<table>
<thead>
<tr>
<th>String 1</th>
<th>String 2</th>
<th>Operation</th>
</tr>
</thead>
<tbody>
<tr>
<td>M</td>
<td></td>
<td>Insert/delete</td>
</tr>
<tr>
<td></td>
<td>T</td>
<td>Insert/delete</td>
</tr>
<tr>
<td>A</td>
<td>A</td>
<td>Keep matching</td>
</tr>
<tr>
<td>P</td>
<td>B</td>
<td>Replace</td>
</tr>
<tr>
<td>L</td>
<td>L</td>
<td>Keep matching</td>
</tr>
<tr>
<td>E</td>
<td>E</td>
<td>Keep matching</td>
</tr>
</tbody>
</table>
<p>Or, MAPLE and TABLE may be aligned by performing...</p>
<table>
<thead>
<tr>
<th>String 1</th>
<th>String 2</th>
<th>Operation</th>
</tr>
</thead>
<tbody>
<tr>
<td>M</td>
<td>T</td>
<td>Replace</td>
</tr>
<tr>
<td>A</td>
<td>A</td>
<td>Keep matching</td>
</tr>
<tr>
<td>P</td>
<td>B</td>
<td>Replace</td>
</tr>
<tr>
<td>L</td>
<td>L</td>
<td>Keep matching</td>
</tr>
<tr>
<td>E</td>
<td>E</td>
<td>Keep matching</td>
</tr>
</tbody>
</table>
<p>The names of these operations make more sense if you were to think of alignment instead as <strong>transformation</strong>. The first example above in the context of <strong>transforming</strong> MAPLE to TABLE may be thought of as:</p>
<table>
<thead>
<tr>
<th>From</th>
<th>To</th>
<th>Operation</th>
<th>Result</th>
</tr>
</thead>
<tbody>
<tr>
<td>M</td>
<td></td>
<td>Delete M</td>
<td></td>
</tr>
<tr>
<td></td>
<td>T</td>
<td>Insert T</td>
<td>T</td>
</tr>
<tr>
<td>A</td>
<td>A</td>
<td>Keep matching A</td>
<td>TA</td>
</tr>
<tr>
<td>P</td>
<td>B</td>
<td>Replace P to B</td>
<td>TAB</td>
</tr>
<tr>
<td>L</td>
<td>L</td>
<td>Keep matching L</td>
<td>TABL</td>
</tr>
<tr>
<td>E</td>
<td>E</td>
<td>Keep matching E</td>
<td>TABLE</td>
</tr>
</tbody>
</table>
<p>The shorthand form of representing <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a>s is to stack each <a href="#BM_(sequence)%2Fi">sequence</a>. The example above may be written as...</p>
<table>
<thead>
<tr>
<th></th>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
</tr>
</thead>
<tbody>
<tr>
<td>String 1</td>
<td>M</td>
<td></td>
<td>A</td>
<td>P</td>
<td>L</td>
<td>E</td>
</tr>
<tr>
<td>String 2</td>
<td></td>
<td>T</td>
<td>A</td>
<td>B</td>
<td>L</td>
<td>E</td>
</tr>
</tbody>
</table>
<p>All possible <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a>s are represented using an <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a>. A path through the <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a> (called <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a>) represents one possible way to align the set of <a href="#BM_(sequence)%2Fi">sequence</a>s.</p>
</li>
<li>
<p><a name="BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi"></a><strong>alignment graph</strong> - A directed graph representing all possible <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a>s for some set of <a href="#BM_(sequence)%2Fi">sequence</a>s. For example, the graph showing all the different ways that MAPLE and TABLE may be aligned ...</p>
<p><img src="svgbob_8fd0bc1c44c43582501d24079827c8db51860ff1.svg" alt="Kroki diagram output"></p>
<p>A path in this graph from source (top-left) to sink (bottom-right) represents an alignment.</p>
</li>
<li>
<p><a name="BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi"></a><strong>alignment path</strong> - A path in an <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">alignment graph</a> that represents one possible <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a>. For example, the following <a href="#BM_(sequence%20alignment%20graph%20path%7Csequence%20alignment%20path%7Calignment%20graph%20path%7Calignment%20path)%2Fi">alignment path</a> ...</p>
<p><img src="svgbob_dab8cfd9932f94ef62dada363fa927912ddc6a7d.svg" alt="Kroki diagram output"></p>
<p>is for the <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a>...</p>
<table>
<thead>
<tr>
<th></th>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
<th>6</th>
<th>7</th>
</tr>
</thead>
<tbody>
<tr>
<td>String 1</td>
<td>-</td>
<td>-</td>
<td>M</td>
<td>A</td>
<td>P</td>
<td>-</td>
<td>L</td>
<td>E</td>
</tr>
<tr>
<td>String 2</td>
<td>T</td>
<td>A</td>
<td>-</td>
<td>B</td>
<td>L</td>
<td>E</td>
<td>-</td>
<td>-</td>
</tr>
</tbody>
</table>
</li>
<li>
<p><a name="BM_(indel)%2Fi"></a><strong>indel</strong> - In the context of <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a>, <a href="#BM_(indel)%2Fi">indel</a> is short-hand for insert/delete. For example, the following <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a> has 2 <a href="#BM_(indel)%2Fi">indel</a>s in the very beginning...</p>
<table>
<thead>
<tr>
<th>String 1</th>
<th>String 2</th>
<th>Operation</th>
</tr>
</thead>
<tbody>
<tr>
<td>M</td>
<td></td>
<td><a href="#BM_(indel)%2Fi">Indel</a></td>
</tr>
<tr>
<td></td>
<td>T</td>
<td><a href="#BM_(indel)%2Fi">Indel</a></td>
</tr>
<tr>
<td>A</td>
<td>A</td>
<td>Keep matching</td>
</tr>
<tr>
<td>P</td>
<td>B</td>
<td>Replace</td>
</tr>
<tr>
<td>L</td>
<td>L</td>
<td>Keep matching</td>
</tr>
<tr>
<td>E</td>
<td>E</td>
<td>Keep matching</td>
</tr>
</tbody>
</table>
<p>The term insert/delete makes sense if you were to think of the set of operations as a <strong>transformation</strong> rather than an alignment. For example, the example above in the context of <strong>transforming</strong> MAPLE to TABLE:</p>
<table>
<thead>
<tr>
<th>From</th>
<th>To</th>
<th>Operation</th>
<th>Result</th>
</tr>
</thead>
<tbody>
<tr>
<td>M</td>
<td></td>
<td>Delete M</td>
<td></td>
</tr>
<tr>
<td></td>
<td>T</td>
<td>Insert T</td>
<td>T</td>
</tr>
<tr>
<td>A</td>
<td>A</td>
<td>Keep matching A</td>
<td>TA</td>
</tr>
<tr>
<td>P</td>
<td>B</td>
<td>Replace P to B</td>
<td>TAB</td>
</tr>
<tr>
<td>L</td>
<td>L</td>
<td>Keep matching L</td>
<td>TABL</td>
</tr>
<tr>
<td>E</td>
<td>E</td>
<td>Keep matching E</td>
<td>TABLE</td>
</tr>
</tbody>
</table>
</li>
<li>
<p><a name="BM_(oncogene)%2Fi"></a><strong>oncogene</strong> - A <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a> that has the potential to cause cancer. In tumor cells, these <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">genes</a> are often mutated or expressed at higher levels.</p>
<p>Most normal cells will undergo apoptosis when critical functions are altered and malfunctioning. Activated <a href="#BM_(oncogene)%2Fi">oncogene</a>s may cause those cells to survive and proliferate instead.</p>
</li>
<li>
<p><a name="BM_(hamming%20distance)%2Fi"></a><strong>hamming distance</strong> - Given two strings, the <a href="#BM_(hamming%20distance)%2Fi">hamming distance</a> is the number of positional mismatches between them. For example, the <a href="#BM_(hamming%20distance)%2Fi">hamming distance</a> between ACTTTGTT and AGTTTCTT is 2.</p>
<table>
<thead>
<tr>
<th></th>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
<th>6</th>
<th>7</th>
</tr>
</thead>
<tbody>
<tr>
<td>String 1</td>
<td>A</td>
<td>C</td>
<td>T</td>
<td>T</td>
<td>T</td>
<td>G</td>
<td>T</td>
<td>T</td>
</tr>
<tr>
<td>String 2</td>
<td>A</td>
<td>G</td>
<td>T</td>
<td>T</td>
<td>T</td>
<td>C</td>
<td>T</td>
<td>T</td>
</tr>
<tr>
<td>Results</td>
<td>✓</td>
<td>✗</td>
<td>✓</td>
<td>✓</td>
<td>✓</td>
<td>✗</td>
<td>✓</td>
<td>✓</td>
</tr>
</tbody>
</table>
</li>
<li>
<p><a name="BM_(dynamic%20programming%20algorithm%7Cdynamic%20programming)%2Fi"></a><strong>dynamic programming</strong> - An algorithm that solves a problem by recursively breaking it down into smaller sub-problems, where the result of each recurrence computation is stored in some lookup table such that it can be re-used if it were ever encountered again (essentially trading space for speed). The lookup table may be created beforehand or as a cache that gets filled as the algorithm runs.</p>
<p>For example, imagine a money system where coins are represented in 1, 12, and 13 cent denominations. You can use recursion to find the minimum number of coins to represent some monetary value such as $0.17:</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">min_coins</span>(<span class="hljs-params">value</span>):
  <span class="hljs-keyword">if</span> value == <span class="hljs-number">0.01</span> <span class="hljs-keyword">or</span> value == <span class="hljs-number">0.12</span> <span class="hljs-keyword">or</span> value == <span class="hljs-number">0.13</span>:
    <span class="hljs-keyword">return</span> <span class="hljs-number">1</span>
  <span class="hljs-keyword">else</span>:
    <span class="hljs-keyword">return</span> <span class="hljs-built_in">min</span>([
      min_coins(value - <span class="hljs-number">0.01</span>) + <span class="hljs-number">1</span>,
      min_coins(value - <span class="hljs-number">0.12</span>) + <span class="hljs-number">1</span>,
      min_coins(value - <span class="hljs-number">0.13</span>) + <span class="hljs-number">1</span>
    ])
</code></pre>
<p><img src="svgbob_011d5f745d519fd5c1a5354faa9961f538f726fc.svg" alt="Kroki diagram output"></p>
<p>The recursive graph above shows how $0.17 can be produced from a minimum of 5 coins: 1 x 13 cent denomination and 4 x 1 cent denomination. However, it recomputes identical parts of the graph multiple times. For example, <code>min_coins(3)</code> is independently computed 5 times. With <a href="#BM_(dynamic%20programming%20algorithm%7Cdynamic%20programming)%2Fi">dynamic programming</a>, it would only be computed once and the result would be re-used each subsequent time <code>min_coins(3)</code> gets encountered.</p>
</li>
<li>
<p><a name="BM_(manhattan%20tourist%20problem)%2Fi"></a><strong>manhattan tourist problem</strong> - The <a href="#BM_(manhattan%20tourist%20problem)%2Fi">Manhattan tourist problem</a> is an allegory to help explain <a href="#BM_(alignment%20graph%7Csequence%20alignment%20graph)%2Fi">sequence alignment graph</a>s. Whereas in <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a>s you're finding a path through the graph from source to sink that has the maximum weight, in the <a href="#BM_(manhattan%20tourist%20problem)%2Fi">Manhattan tourist problem</a> you're finding a path from 59th St and 8th Ave to 42nd St and 3rd Ave with the most tourist sights to see. It's essentially the same problem as <a href="#BM_(global%20alignment%7Cglobal%20sequence%20alignment)%2Fi">global alignment</a>:</p>
<ul>
<li>The graph is the s<a href="#BM_(tree)%2Fi">tree</a>t layout of Manhattan.</li>
<li>The only options at each intersection are to move right or down.</li>
<li>The source node is the intersection of 59th St and 8th Ave.</li>
<li>The sink node is the intersection of 42nd St and 3rd Ave.</li>
<li>The number of tourist sights from one intersection to the next is the weight of an edge.</li>
</ul>
<p><img src="svgbob_903ed0d46089376b0a56d7285560ca67078091e7.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a name="BM_(point%20accepted%20mutation%7Cpercent%20accepted%20mutation)%2Fi"></a><strong>point accepted mutation</strong> <a name="BM_%5Cb(PAM)%5Cd*%5Cb%2F"></a><strong></strong> - A scoring matrix used for <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a>s of proteins. The scoring matrix is calculated by inspecting / extrapolating mutations as homologous proteins evolve. Specifically, mutations in the DNA <a href="#BM_(sequence)%2Fi">sequence</a> that <a href="#BM_(encode%7Cencoding%7Cdecode%7Cdecoding)%2Fi">encode</a> some protein may change the resulting <a href="#BM_(amino%20acid)%2Fi">amino acid</a> <a href="#BM_(sequence)%2Fi">sequence</a> for that protein. Those mutations that...</p>
<ul>
<li>impair the ability of the protein to function aren't likely to survive, and as such are given a low score.</li>
<li>keep the protein functional are likely to survive, and as such are given a normal or high score.</li>
</ul>
</li>
<li>
<p><a name="BM_(blocks%20substitution%20matrix%7Cblocks%20substitution%20matrices%7Cblocks%20amino%20acid%20substitution%20matrices%7Cblocks%20amino%20acid%20substitution%20matrix)%2Fi"></a><strong>blocks amino acid substitution matrix</strong> <a name="BM_%5Cb(BLOSUM)%5Cd*%5Cb%2F"></a><strong></strong> - A scoring matrix used for <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a>s of proteins. The scoring matrix is calculated by scanning a protein database for highly conserved regions between similar proteins, where the mutations between those highly conserved regions define the scores. Specifically, those highly conserved regions are identified based on <a href="#BM_(local%20alignment%7Clocal%20sequence%20alignment)%2Fi">local alignment</a>s without support for <a href="#BM_(indel)%2Fi">indel</a>s (gaps not allowed). Non-matching positions in that alignment define potentially acceptable mutations.</p>
</li>
<li>
<p><a name="BM_(point%20mutation)%2Fi"></a><strong>point mutation</strong> - A mutation in DNA (or RNA) where a single nucleotide base is either changed, inserted, or deleted.</p>
</li>
<li>
<p><a name="BM_(directed%20acyclic%20graph)%2Fi"></a><strong>directed acyclic graph</strong> <a name="BM_%5Cb(DAG)%5Cb%2F"></a><strong></strong> - A graph where the edges are directed (have a direction) and no cycles exist in the graph.</p>
<p>For example, the following is a <a href="#BM_(directed%20acyclic%20graph)%2Fi">directed acyclic graph</a>...</p>
<p><img src="svgbob_d6ba29b704329aea40b096d1beeeec7da7d062f5.svg" alt="Kroki diagram output"></p>
<p>The following graph isn't a <a href="#BM_(directed%20acyclic%20graph)%2Fi">directed acyclic graph</a> because the edges don't have a direction (no arrowhead means you can travel in either direction)...</p>
<p><img src="svgbob_64d614efe5a1433237c817bfba588e85082acf69.svg" alt="Kroki diagram output"></p>
<p>The following graph isn't a <a href="#BM_(directed%20acyclic%20graph)%2Fi">directed acyclic graph</a> because it contains a cycle between D and B...</p>
<p><img src="svgbob_36208e0955503cf8020cbe27be7ca286ebcf55e2.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a name="BM_(divide-and-conquer%20algorithm%7Cdivide%20and%20conquer%20algorithm)%2Fi"></a><strong>divide-and-conquer algorithm</strong> - An algorithm that solves a problem by recursively breaking it down into two or more smaller sub-problems, up until the point where each sub-problem is small enough / simple enough to solve. Examples include quicksort and merge sort.</p>
<p>See <a href="#BM_(dynamic%20programming%20algorithm%7Cdynamic%20programming)%2Fi">dynamic programming</a>.</p>
</li>
<li>
<p><a name="BM_(global%20alignment%7Cglobal%20sequence%20alignment)%2Fi"></a><strong>global alignment</strong> - A form of <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a> that finds the highest scoring alignment between a set of <a href="#BM_(sequence)%2Fi">sequence</a>s. The <a href="#BM_(sequence)%2Fi">sequence</a>s are aligned in their entirety. For example, the <a href="#BM_(sequence)%2Fi">sequence</a>s TRELLO and MELLOW have the following <a href="#BM_(global%20alignment%7Cglobal%20sequence%20alignment)%2Fi">global alignment</a>...</p>
<table>
<thead>
<tr>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
<th>6</th>
</tr>
</thead>
<tbody>
<tr>
<td>T</td>
<td>R</td>
<td>E</td>
<td>L</td>
<td>L</td>
<td>O</td>
<td>-</td>
</tr>
<tr>
<td>-</td>
<td>M</td>
<td>E</td>
<td>L</td>
<td>L</td>
<td>O</td>
<td>W</td>
</tr>
</tbody>
</table>
<p>This is the form of <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a> that most people think about when they hear "<a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a>."</p>
</li>
<li>
<p><a name="BM_(local%20alignment%7Clocal%20sequence%20alignment)%2Fi"></a><strong>local alignment</strong> - A form of <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a> that isolates the alignment to a <a href="#BM_(substring)%2Fi">substring</a> of each <a href="#BM_(sequence)%2Fi">sequence</a>. The <a href="#BM_(substring)%2Fi">substring</a>s that score the highest are the ones selected. For example, the <a href="#BM_(sequence)%2Fi">sequence</a>s TRELLO and MELLOW have the following <a href="#BM_(local%20alignment%7Clocal%20sequence%20alignment)%2Fi">local alignment</a>...</p>
<table>
<thead>
<tr>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
</tr>
</thead>
<tbody>
<tr>
<td>E</td>
<td>L</td>
<td>L</td>
<td>O</td>
</tr>
<tr>
<td>E</td>
<td>L</td>
<td>L</td>
<td>O</td>
</tr>
</tbody>
</table>
<p>... because out of all <a href="#BM_(substring)%2Fi">substring</a>s in TRELLO and all <a href="#BM_(substring)%2Fi">substring</a>s in MELLOW, ELLO (from TRELLO) scores the highest against ELLO (from MELLOW).</p>
</li>
<li>
<p><a name="BM_(fitting%20alignment%7Cfitting%20sequence%20alignment)%2Fi"></a><strong>fitting alignment</strong> - A form of 2-way <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a> that isolates the alignment such that the entirety of one <a href="#BM_(sequence)%2Fi">sequence</a> is aligned against a <a href="#BM_(substring)%2Fi">substring</a> of the other <a href="#BM_(sequence)%2Fi">sequence</a>. The <a href="#BM_(substring)%2Fi">substring</a> producing the highest score is the one that's selected. For example, the <a href="#BM_(sequence)%2Fi">sequence</a>s ELO and MELLOW have the following <a href="#BM_(fitting%20alignment%7Cfitting%20sequence%20alignment)%2Fi">fitting alignment</a>...</p>
<table>
<thead>
<tr>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
</tr>
</thead>
<tbody>
<tr>
<td>E</td>
<td>L</td>
<td>-</td>
<td>O</td>
</tr>
<tr>
<td>E</td>
<td>L</td>
<td>L</td>
<td>O</td>
</tr>
</tbody>
</table>
<p>... because out of all the <a href="#BM_(substring)%2Fi">substring</a>s in MELLOW, the <a href="#BM_(substring)%2Fi">substring</a> ELLO scores the highest against ELO.</p>
</li>
<li>
<p><a name="BM_(overlap%20alignment%7Coverlap%20sequence%20alignment)%2Fi"></a><strong>overlap alignment</strong> - A form of 2-way <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a> that isolates the alignment to a suffix of the first <a href="#BM_(sequence)%2Fi">sequence</a>s and a prefix of the second <a href="#BM_(sequence)%2Fi">sequence</a>. The prefix and suffix producing the highest score are the ones selected . For example, the <a href="#BM_(sequence)%2Fi">sequence</a>s BURRITO and RICOTTA have the following <a href="#BM_(overlap%20alignment%7Coverlap%20sequence%20alignment)%2Fi">overlap alignment</a>...</p>
<table>
<thead>
<tr>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
</tr>
</thead>
<tbody>
<tr>
<td>R</td>
<td>I</td>
<td>T</td>
<td>-</td>
<td>O</td>
</tr>
<tr>
<td>R</td>
<td>I</td>
<td>-</td>
<td>C</td>
<td>O</td>
</tr>
</tbody>
</table>
<p>... because out of all the suffixes in BURRITO and the prefixes in RICOTTA, RITO and RICO score the highest.</p>
</li>
<li>
<p><a name="BM_(levenshtein%20distance%7Cstring%20distance)%2Fi"></a><strong>Levenshtein distance</strong> - An application of <a href="#BM_(global%20alignment%7Cglobal%20sequence%20alignment)%2Fi">global alignment</a> where the final weight represents the minimum number of operations required to transform one <a href="#BM_(sequence)%2Fi">sequence</a> to another (via swaps, insertions, and deletions). Matches are scored 0, while mismatches and <a href="#BM_(indel)%2Fi">indel</a>s are scored -1. For example, TRELLO and MELLOW have the <a href="#BM_(levenshtein%20distance%7Cstring%20distance)%2Fi">Levenshtein distance</a> of 3...</p>
<table>
<thead>
<tr>
<th></th>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
<th>6</th>
<th></th>
</tr>
</thead>
<tbody>
<tr>
<td></td>
<td>T</td>
<td>R</td>
<td>E</td>
<td>L</td>
<td>L</td>
<td>O</td>
<td>-</td>
<td></td>
</tr>
<tr>
<td></td>
<td>-</td>
<td>M</td>
<td>E</td>
<td>L</td>
<td>L</td>
<td>O</td>
<td>W</td>
<td></td>
</tr>
<tr>
<td>Score</td>
<td>-1</td>
<td>-1</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>-1</td>
<td>Total: -3</td>
</tr>
</tbody>
</table>
<p>Negate the total score to get the minimum number of operations. In the example above, the final score of -3 maps to a minimum of 3 operations.</p>
</li>
<li>
<p><a name="BM_(genome%20rearrangement%7Cchromosomal%20rearrangement%7Cchromosome%20rearrangement)%2Fi"></a><strong>genome rearrangement</strong> - A type of mutation where chromosomes go through structural changes, typically caused by either</p>
<ul>
<li>breakages in the chromosome, where the broken ends possibly rejoin but in a different order / direction.</li>
<li>DNA replication error.</li>
<li>DNA repair error.</li>
</ul>
<p>The different classes of rearrangement include...</p>
<ul>
<li>
<p>reversal / inversion: A break at two different locations followed by rejoining of the broken ends in different order. and rejoin in different order</p>
<p><img src="svgbob_c74a74fa3c923d49b533b6a31889fba8c857805a.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a href="#BM_(translocation)%2Fi">translocation</a>:</p>
<p><img src="svgbob_1a60b8a84c9cf4c8e262eb77c4db7fd6cdcb613c.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>deletion:</p>
<p><img src="svgbob_99d9b70d106d2088d8c2b416ae27d9528b9bd80f.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>duplication:</p>
<p><img src="svgbob_89a0a11786d7852db04220f4567d8243d5c22975.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>chromosome <a href="#BM_(%5Cbfusion%7C%5Cbfuse)%2Fi">fusion</a>:</p>
<p><img src="svgbob_5dda88b1068989e48adff91a638bd6e8122a5269.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>chromosome <a href="#BM_(fission)%2Fi">fission</a>:</p>
<p><img src="svgbob_2bf9b82f60b2bbaf95ba7e19855dd43c0b57008f.svg" alt="Kroki diagram output"></p>
</li>
</ul>
<p>The segments of the <a href="#BM_(genome)%2Fi">genome</a> that were moved around are referred to as <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s.</p>
</li>
<li>
<p><a name="BM_(chimeric%20gene)%2Fi"></a><strong>chimeric gene</strong> - A <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a> born from two separate <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">genes</a> getting <a href="#BM_(%5Cbfusion%7C%5Cbfuse)%2Fi">fuse</a>d together. A <a href="#BM_(chimeric%20gene)%2Fi">chimeric gene</a> may have been created via <a href="#BM_(genome%20rearrangement%7Cchromosomal%20rearrangement%7Cchromosome%20rearrangement)%2Fi">genome rearrangement</a> <a href="#BM_(translocation)%2Fi">translocation</a>s.</p>
<p>An example of a <a href="#BM_(chimeric%20gene)%2Fi">chimeric gene</a> is the <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a> coding for ABL-BCR <a href="#BM_(%5Cbfusion%7C%5Cbfuse)%2Fi">fusion</a> protein: A <a href="#BM_(%5Cbfusion%7C%5Cbfuse)%2Fi">fusion</a> of two smaller <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">genes</a> (coding for ABL and BCR individually) caused by chromosomes 9 and 22 breaking and re-attaching in a different order. The ABL-BCR <a href="#BM_(%5Cbfusion%7C%5Cbfuse)%2Fi">fusion</a> protein has been implicated in the development of a cancer known as chronic myeloid leukemia.</p>
</li>
<li>
<p><a name="BM_(reversal%20distance)%2Fi"></a><strong>reversal distance</strong> - The <em>minimum</em> number of <a href="#BM_(genome%20rearrangement%7Cchromosomal%20rearrangement%7Cchromosome%20rearrangement)%2Fi">genome rearrangement</a> reversals required to transform <a href="#BM_(genome)%2Fi">genome</a> P to <a href="#BM_(genome)%2Fi">genome</a> Q. The minimum is chosen because of <a href="#BM_(parsimony%7Cparsimonious)%2Fi">parsimony</a>.</p>
<p>The short-hand for this is <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>d</mi><mrow><mi>r</mi><mi>e</mi><mi>v</mi></mrow></msub><mo stretchy="false">(</mo><mi>P</mi><mo separator="true">,</mo><mi>Q</mi><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">d_{rev}(P, Q)</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord"><span class="mord mathnormal">d</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.151392em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight" style="margin-right:0.02778em;">r</span><span class="mord mathnormal mtight">e</span><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.13889em;">P</span><span class="mpunct">,</span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord mathnormal">Q</span><span class="mclose">)</span></span></span></span>.</p>
</li>
<li>
<p><a name="BM_(sex-chromosome%20dosage%20compensation%7Csex%20chromosome%20dosage%20compensation%7Cdosage%20compensation)%2Fi"></a><strong>dosage compensation</strong> - The mechanism by which sex chromosome <a href="#BM_(gene%20expression)%2Fi">gene expression</a> is equalized between different sexes of the same species.</p>
<p>For example, mammals have two sexes...</p>
<ul>
<li>males, identified by one X chromosome and one Y chromosome.</li>
<li>females, identified by two X chromosomes.</li>
</ul>
<p>Since females have two X chromosomes, it would make sense for females to have double the <a href="#BM_(gene%20expression)%2Fi">gene expression</a> for X chromosome <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">genes</a>. However, many X chromosome <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">genes</a> have nothing to do with sex and if their expression were doubled it would lead to disease. As such, female mammals randomly shut down one of two X chromosomes so as to keep X chromosome <a href="#BM_(gene%20expression)%2Fi">gene expression</a> levels roughly equivalent to that of males.</p>
<p>For mammals, this mechanism means that X chromosomes are mostly conserved because an X chromosome that's gone through <a href="#BM_(genome%20rearrangement%7Cchromosomal%20rearrangement%7Cchromosome%20rearrangement)%2Fi">genome rearrangement</a> likely won't survive: If a <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a> jumps off an X chromosome its <a href="#BM_(gene%20expression)%2Fi">gene expression</a> may double, leading to problems.</p>
<p>Different species have different mechanisms for equalization. For example, some species will double the <a href="#BM_(gene%20expression)%2Fi">gene expression</a> on the male's single X chromosome rather than deactivating one of the female's two X chromosomes. Other hermaphroditic species may scale down X chromosome <a href="#BM_(gene%20expression)%2Fi">gene expression</a> when multiple X chromosomes are present.</p>
</li>
<li>
<p><a name="BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi"></a><strong>synteny</strong> - Intervals within two sets of chromosomes that have similar <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">genes</a> which are either in ...</p>
<ul>
<li>
<p>the same order.</p>
<p><img src="svgbob_d3e4eb729d41d644a981794c1b1d200e2ac32fac.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p>reverse order, where each <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a>'s <a href="#BM_(sequence)%2Fi">sequence</a> is also reversed.</p>
<p><img src="svgbob_feb089f8283ebeb0a53ce28c3d58c61f4291646f.svg" alt="Kroki diagram output"></p>
</li>
</ul>
<p>The idea is that as evolution branches out a single ancestor species to different sub-species, <a href="#BM_(genome%20rearrangement%7Cchromosomal%20rearrangement%7Cchromosome%20rearrangement)%2Fi">genome rearrangement</a>s (reversals, <a href="#BM_(translocation)%2Fi">translocation</a>s, etc..) are responsible for some of those mutations. As chromosomes break and rejoin back together in different order, the stretches between breakage points remain largely the same. For example, it's assumed that mice and humans have the same ancestor species because of the high number of <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s between their <a href="#BM_(genome)%2Fi">genome</a>s (most human <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">genes</a> have a mouse counterpart).</p>
<p><img src="svgbob_f4076143294d6480fb4b883c6cf2d02cdbf9117b.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a name="BM_(parsimony%7Cparsimonious)%2Fi"></a><strong>parsimony</strong> - The scientific concept of choosing the fewest number of steps / shortest path / simplest scenario / simplest explanation that fits the evidence available.</p>
</li>
<li>
<p><a name="BM_(genomic%20dot-plot%7Cgenomic%20dot%20plot)%2Fi"></a><strong>genomic dot-plot</strong> - Given two <a href="#BM_(genome)%2Fi">genome</a>s, create a 2D plot where each axis is assigned to one of the <a href="#BM_(genome)%2Fi">genome</a>s and a dot is placed at each coordinate containing a match, where a match is either a shared <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> or a <a href="#BM_(%5Cd%2B-mer%7Ck-mer%7Ckmer)%2Fi">k-mer</a> and its <a href="#BM_(reverse%20complement)%2Fi">reverse complement</a>. Matches may also be fuzzily found (e.g. within some <a href="#BM_(hamming%20distance)%2Fi">hamming distance</a> rather).</p>
<p>For example, ...</p>
<p><img src="svgbob_3e7a2fb0e16cdd00ff9eaefd665998612d1db15c.svg" alt="Kroki diagram output"></p>
<p><a href="#BM_(genomic%20dot-plot%7Cgenomic%20dot%20plot)%2Fi">Genomic dot-plot</a>s are typically used in building <a href="#BM_(synteny%20graph)%2Fi">synteny graph</a>s: Graphs that reveal shared <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s (shared stretches of DNA). <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">Synteny block</a>s exist because <a href="#BM_(genome%20rearrangement%7Cchromosomal%20rearrangement%7Cchromosome%20rearrangement)%2Fi">genome rearrangement</a>s account for a large percentage of mutations between two species that branched off from the same parent (given that they aren't too far removed -- e.g. mouse vs human).</p>
</li>
<li>
<p><a name="BM_(synteny%20graph)%2Fi"></a><strong>synteny graph</strong> - Given the <a href="#BM_(genomic%20dot-plot%7Cgenomic%20dot%20plot)%2Fi">genomic dot-plot</a> for two <a href="#BM_(genome)%2Fi">genome</a>s, <a href="#BM_(cluster)%2Fi">cluster</a> together points so as to reveal <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s. For example, ...</p>
<p><img src="svgbob_eb50ab9882ae13623c74006244614b059e22fa47.svg" alt="Kroki diagram output"></p>
<p>... reveals that 4 <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s are shared between the <a href="#BM_(genome)%2Fi">genome</a>s. One of the <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s is a normal match (C) while three are matching against their <a href="#BM_(reverse%20complement)%2Fi">reverse complement</a>s (A, B, and D)...</p>
<p><img src="svgbob_725cfe74948aa476c335f99e888801f679a0059a.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a name="BM_(adjacent%7Cadjacency%7Cadjacencies%7Cbreakpoint)_GR%2Fi"></a><strong>breakpoint</strong> - Given two <a href="#BM_(genome)%2Fi">genome</a>s that share <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s, where one <a href="#BM_(genome)%2Fi">genome</a> has the <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s in desired order and direction while the other does not, an ...</p>
<ul>
<li>
<p><a href="#BM_(adjacent%7Cadjacency%7Cadjacencies%7Cbreakpoint)_GR%2Fi">adjacency</a> is when two <a href="#BM_(neighbour)%2Fi">neighbour</a>ing <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s in the undesired <a href="#BM_(genome)%2Fi">genome</a> are following each other just as they do in the desired <a href="#BM_(genome)%2Fi">genome</a>.</p>
<p><img src="svgbob_0eeb1c6bdbb8c831beacb96ed553a24786b51106.svg" alt="Kroki diagram output"></p>
<p><img src="svgbob_38f85e31e85b1a0e4922855af392c0423a9cda9a.svg" alt="Kroki diagram output"></p>
<p><img src="svgbob_0688b38f16f6f4a626db94a63389ec54f36fac30.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a href="#BM_(adjacent%7Cadjacency%7Cadjacencies%7Cbreakpoint)_GR%2Fi">breakpoint</a> is when two <a href="#BM_(neighbour)%2Fi">neighbour</a>ing <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s in the undesired <a href="#BM_(genome)%2Fi">genome</a> don't fit the definition of an <a href="#BM_(adjacent%7Cadjacency%7Cadjacencies%7Cbreakpoint)_GR%2Fi">adjacency</a>: They aren't following each other just as they do in the desired <a href="#BM_(genome)%2Fi">genome</a>.</p>
<p><img src="svgbob_499a43209f898efb42fff1bb55780d8c5001df23.svg" alt="Kroki diagram output"></p>
<p><img src="svgbob_05d2980e44e429d3d51e6349e29ff7734c7097ff.svg" alt="Kroki diagram output"></p>
</li>
</ul>
<p><a href="#BM_(adjacent%7Cadjacency%7Cadjacencies%7Cbreakpoint)_GR%2Fi">Breakpoint</a>s and <a href="#BM_(adjacent%7Cadjacency%7Cadjacencies%7Cbreakpoint)_GR%2Fi">adjacencies</a> are useful because they identify desirable points for reversals (<a href="#BM_(genome%20rearrangement%7Cchromosomal%20rearrangement%7Cchromosome%20rearrangement)%2Fi">genome rearrangement</a>), giving way to algorithms that find / estimate the <a href="#BM_(reversal%20distance)%2Fi">reversal distance</a>. For example, a contiguous train of <a href="#BM_(adjacent%7Cadjacency%7Cadjacencies%7Cbreakpoint)_GR%2Fi">adjacencies</a> in an undesired <a href="#BM_(genome)%2Fi">genome</a> may identify the boundaries for a single reversal that gets the undesired <a href="#BM_(genome)%2Fi">genome</a> closer to the desired <a href="#BM_(genome)%2Fi">genome</a>.</p>
<p><img src="svgbob_185405cd7f39c08a1a02020363ae24096ff5b234.svg" alt="Kroki diagram output"></p>
<p>The number of <a href="#BM_(adjacent%7Cadjacency%7Cadjacencies%7Cbreakpoint)_GR%2Fi">breakpoint</a>s and <a href="#BM_(adjacent%7Cadjacency%7Cadjacencies%7Cbreakpoint)_GR%2Fi">adjacencies</a> always equals one less than the number of <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s.</p>
<p><img src="svgbob_04ffee42b4f3665b1043ccb7aa7273eff21cb2fb.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a name="BM_(breakpoint%20graph)_GR%2Fi"></a><strong>breakpoint graph</strong> - An undirected graph representing the order and orientation of <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s shared between two <a href="#BM_(genome)%2Fi">genome</a>s.</p>
<p>For example, the following two <a href="#BM_(genome)%2Fi">genome</a>s share the <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s A, B, C, and D...</p>
<p><img src="svgbob_7e9c33210faa97dd82efb19ecbdd7b9300236fba.svg" alt="Kroki diagram output"></p>
<p>The <a href="#BM_(breakpoint%20graph)_GR%2Fi">breakpoint graph</a> for the above two <a href="#BM_(genome)%2Fi">genome</a>s is basically just a merge of the above diagrams. The set of <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s shared between both <a href="#BM_(genome)%2Fi">genome</a>s (A, B, C, and D) become dashed edges where each edge's...</p>
<ul>
<li>arrow end is a tail node (t suffix).</li>
<li>non-arrow end is a head node (h suffix).</li>
</ul>
<p>Gap regions between <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s are represented by solid colored edges, either red or blue depending on which <a href="#BM_(genome)%2Fi">genome</a> it is.</p>
<p>If the <a href="#BM_(genome)%2Fi">genome</a>s are linear, gap region edges are created between the nodes and the edges and a special termination node.</p>
<p><img src="dot_68956ed2f489ac4cb837efb6bd1d2a65.svg" alt="Dot diagram"></p>
<p>In the above <a href="#BM_(breakpoint%20graph)_GR%2Fi">breakpoint graph</a>, the blue edges represent <a href="#BM_(genome)%2Fi">genome</a> 2's gap regions while the red edges represent <a href="#BM_(genome)%2Fi">genome</a> 1's gap regions. The set of edges representing <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s is shared between them.</p>
<p><a href="#BM_(breakpoint%20graph)_GR%2Fi">Breakpoint graph</a>s build on the concept of <a href="#BM_(adjacent%7Cadjacency%7Cadjacencies%7Cbreakpoint)_GR%2Fi">breakpoint</a>s to compute a <a href="#BM_(parsimony%7Cparsimonious)%2Fi">parsimonious</a> path of <a href="#BM_(%5Cbfusion%7C%5Cbfuse)%2Fi">fusion</a>, <a href="#BM_(fission)%2Fi">fission</a>, and reversal mutations (<a href="#BM_(genome%20rearrangement%7Cchromosomal%20rearrangement%7Cchromosome%20rearrangement)%2Fi">genome rearrangement</a>s) that transforms one <a href="#BM_(genome)%2Fi">genome</a> into the other (see <a href="#BM_%5Cb(2-break%7C2%20break%7Ctwo%20break%7Ctwo-break)s%3F%5Cb%2Fi">2-break</a>). Conventionally, blue edges represent the final desired path while red edges represent the path being transformed. As such, <a href="#BM_(breakpoint%20graph)_GR%2Fi">breakpoint graph</a>s typically order <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s so that blue edges are uniformly sandwiched between <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a>s / red edges get chaotically scattered around.</p>
</li>
<li>
<p><a name="BM_%5Cb(2-break%7C2%20break%7Ctwo%20break%7Ctwo-break)s%3F%5Cb%2Fi"></a><strong>2-break</strong> - Given a <a href="#BM_(breakpoint%20graph)_GR%2Fi">breakpoint graph</a>, a <a href="#BM_%5Cb(2-break%7C2%20break%7Ctwo%20break%7Ctwo-break)s%3F%5Cb%2Fi">2-break</a> operation breaks the two red edges at a <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a> boundary and re-wires them such that one of the red edges matches the blue edge at that boundary.</p>
<p>For example, the two red edges highlighted below share the same <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny block</a> boundary and can be re-wired such that one of the edges matches the blue edge at that <a href="#BM_(synteny%20block%7Csynteny%20region%7Csyntenic%20region%7Csynteny%7Csyntenic%7Csyntenies)%2Fi">synteny</a> boundary ...</p>
<p><img src="dot_2ca5243fc24500a47a2859ee21a4a61a.svg" alt="Dot diagram"></p>
<p>Each <a href="#BM_%5Cb(2-break%7C2%20break%7Ctwo%20break%7Ctwo-break)s%3F%5Cb%2Fi">2-break</a> operation on a <a href="#BM_(breakpoint%20graph)_GR%2Fi">breakpoint graph</a> represents a <a href="#BM_(%5Cbfusion%7C%5Cbfuse)%2Fi">fusion</a>, <a href="#BM_(fission)%2Fi">fission</a>, or reversal mutation (<a href="#BM_(genome%20rearrangement%7Cchromosomal%20rearrangement%7Cchromosome%20rearrangement)%2Fi">genome rearrangement</a>). Continually applying <a href="#BM_%5Cb(2-break%7C2%20break%7Ctwo%20break%7Ctwo-break)s%3F%5Cb%2Fi">2-break</a>s until all red edges match blue edges reveals a <a href="#BM_(parsimony%7Cparsimonious)%2Fi">parsimonious</a> path of such mutations that transforms the red <a href="#BM_(genome)%2Fi">genome</a> to the blue <a href="#BM_(genome)%2Fi">genome</a>.</p>
</li>
<li>
<p><a name="BM_(permutation)_GR%2Fi"></a><strong>permutation</strong> - A list representing a single chromosome in one of the two <a href="#BM_(genome)%2Fi">genome</a>s that make up a <a href="#BM_(breakpoint%20graph)_GR%2Fi">breakpoint graph</a>. The entire <a href="#BM_(breakpoint%20graph)_GR%2Fi">breakpoint graph</a> is representable as two sets of <a href="#BM_(permutation)_GR%2Fi">permutation</a>s, where each <a href="#BM_(genome)%2Fi">genome</a> in the <a href="#BM_(breakpoint%20graph)_GR%2Fi">breakpoint graph</a> is a set.</p>
<p><a href="#BM_(permutation)_GR%2Fi">Permutation</a> sets are commonly used for tersely representing <a href="#BM_(breakpoint%20graph)_GR%2Fi">breakpoint graph</a>s as text. For example, given the following <a href="#BM_(breakpoint%20graph)_GR%2Fi">breakpoint graph</a> ...</p>
<p><img src="dot_bed1d61715ed515600f51c8a4b395968.svg" alt="Dot diagram"></p>
<p>... , the <a href="#BM_(permutation)_GR%2Fi">permutation</a> set representing the red <a href="#BM_(genome)%2Fi">genome</a> may be any of the following ...</p>
<ul>
<li><code>{[-D, -B, +C, -A]}</code></li>
<li><code>{[+A, -C, +B, +D]}</code></li>
<li><code>{[-B, +C, -A, -D]}</code></li>
<li><code>{[-C, +B, +D, +A]}</code></li>
<li><code>{[+C, -A, -D, -B]}</code></li>
<li>...</li>
</ul>
<p>All representations above are equivalent.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>See <a href="#BM_(Algorithms%5C%2FSynteny%2FReversal%20Path%2FBreakpoint%20List%20Algorithm)_TOPIC%2F">Algorithms/Synteny/Reversal Path/Breakpoint List Algorithm</a> for a full explanation of how to read <a href="#BM_(permutation)_GR%2Fi">permutation</a>s / how to convert from and to <a href="#BM_(breakpoint%20graph)_GR%2Fi">breakpoint graph</a>s.</p>
</div>
</li>
<li>
<p><a name="BM_(%5Cbfusion%7C%5Cbfuse)%2Fi"></a><strong>fusion</strong> - Joining two or more things together to form a single entity. For example, two chromosomes may join together to form a single chromosome (<a href="#BM_(genome%20rearrangement%7Cchromosomal%20rearrangement%7Cchromosome%20rearrangement)%2Fi">genome rearrangement</a>).</p>
</li>
<li>
<p><a name="BM_(fission)%2Fi"></a><strong>fission</strong> - Splitting a single entity into two or more parts. For example, a single chromosome may break into multiple pieces where each piece becomes its own chromosome (<a href="#BM_(genome%20rearrangement%7Cchromosomal%20rearrangement%7Cchromosome%20rearrangement)%2Fi">genome rearrangement</a>).</p>
</li>
<li>
<p><a name="BM_(translocation)%2Fi"></a><strong>translocation</strong> - Changing location. For example, part of a chromosome may transfer to another chromosome (<a href="#BM_(genome%20rearrangement%7Cchromosomal%20rearrangement%7Cchromosome%20rearrangement)%2Fi">genome rearrangement</a>).</p>
</li>
<li>
<p><a name="BM_(severe%20acute%20respiratory%20syndrome)%2Fi"></a><strong>severe acute respiratory syndrome</strong> <a name="BM_(SARS)%2F"></a><strong></strong> - A deadly <a href="#BM_(coronavirus)%2Fi">coronavirus</a> that emerged from China around early 2003.</p>
</li>
<li>
<p><a name="BM_(coronavirus)%2Fi"></a><strong>coronavirus</strong> <a name="BM_(COVID%7CCoV)%2F"></a><strong></strong> - A family of viruses that attack the respiratory tracts of mammals and birds. The name comes from the fact that the outer spikes of the virus resemble the corona of the sun (crown of the sun / outermost part of the sun's atmosphere protruding out).</p>
<p>The common cold, <a href="#BM_(SARS)%2F">SARS</a>, and <a href="#BM_(COVID%7CCoV)%2F">COVID</a>-19 are examples of <a href="#BM_(coronavirus)%2Fi">coronavirus</a>es.</p>
</li>
<li>
<p><a name="BM_(human%20immunodeficiency%20virus)%2Fi"></a><strong>human immunodeficiency virus</strong> <a name="BM_(HIV%7CAIDS)%2F"></a><strong></strong> - A virus that over time causes acquired <a href="#BM_(immunodeficiency%7Cimmunocompromisation%7Cimmunocompromised)%2Fi">immunodeficiency</a> syndrome (<a href="#BM_(HIV%7CAIDS)%2F">AIDS</a>).</p>
</li>
<li>
<p><a name="BM_(immunodeficiency%7Cimmunocompromisation%7Cimmunocompromised)%2Fi"></a><strong>immunodeficiency</strong> - A state in which the immune system's ability to fight infectious disease and cancer is compromised or entirely absent.</p>
</li>
<li>
<p><a name="BM_(DNA%20virus)%2Fi"></a><strong>DNA virus</strong> - A virus with a DNA <a href="#BM_(genome)%2Fi">genome</a>. Depending on the type of virus, the <a href="#BM_(genome)%2Fi">genome</a> may be <a href="#BM_(single%20stranded%20DNA%7Csingle-stranded%20DNA)%2Fi">single-stranded DNA</a> or <a href="#BM_(double%20stranded%20DNA%7Cdouble-stranded%20DNA)%2Fi">double-stranded DNA</a>.</p>
<p>Herpes, chickenpox, and smallpox are examples of <a href="#BM_(DNA%20virus)%2Fi">DNA virus</a>es.</p>
</li>
<li>
<p><a name="BM_(RNA%20virus)%2Fi"></a><strong>RNA virus</strong> - A virus with a RNA <a href="#BM_(genome)%2Fi">genome</a>. RNA replication has a higher rate than DNA replication, meaning that <a href="#BM_(RNA%20virus)%2Fi">RNA virus</a>es mutate faster than <a href="#BM_(DNA%20virus)%2Fi">DNA virus</a>es.</p>
<p><a href="#BM_(coronavirus)%2Fi">Coronavirus</a>es, <a href="#BM_(HIV%7CAIDS)%2F">HIV</a>, and influenza are examples of <a href="#BM_(RNA%20virus)%2Fi">RNA virus</a>es.</p>
</li>
<li>
<p><a name="BM_(phylogeny%7Cphylogenetic)%2Fi"></a><strong>phylogeny</strong> - The concept of inferring the evolutionary history among some set of species (shared ancestry) by inspecting properties of those species (e.g. relatedness of phenotypic or genotypic characteristics).</p>
<p><img src="svgbob_cabf6297c9c4eca18762400d14f3c05340a62b31.svg" alt="Kroki diagram output"></p>
<p>In the example above, cat and lion are descendants of some shared ancestor species. Likewise, that ancestor and bears are likely descendants from some other higher up species.</p>
</li>
<li>
<p><a name="BM_(phylogenetic%20tree%7Cphylogeny%20tree%7Cphylogenies%7Cevolutionary%20tree)%2Fi"></a><strong>phylogenetic tree</strong> - A <a href="#BM_(tree)%2Fi">tree</a> showing the degree in which biological species or entities (e.g. viruses) are related. Such <a href="#BM_(tree)%2Fi">tree</a>s help infer relationships such as common ancestry or which animal a virus jumped to humans from (e.g. virus A and B are related but A is only present in bats while B just showed up in humans).</p>
<p><img src="svgbob_f7b1705852841b00a5ecb758d6e39af634d6898e.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a name="BM_(distance%20metric)%2Fi"></a><strong>distance metric</strong> - A metric used to measure how different a pair of entities are to each other. Examples include...</p>
<ul>
<li><a href="#BM_(hamming%20distance)%2Fi">hamming distance</a> between DNA / protein <a href="#BM_(sequence)%2Fi">sequence</a>s.</li>
<li><a href="#BM_(global%20alignment%7Cglobal%20sequence%20alignment)%2Fi">global alignment</a> score between DNA / protein <a href="#BM_(sequence)%2Fi">sequence</a>s.</li>
<li><a href="#BM_%5Cb(2-break%7C2%20break%7Ctwo%20break%7Ctwo-break)s%3F%5Cb%2Fi">two-break</a> count (<a href="#BM_(reversal%20distance)%2Fi">reversal distance</a>).</li>
<li>number of similar physical or behavioural attributes.</li>
<li><a href="#BM_(euclidean%20distance)%2Fi">euclidean distance</a> between two vectors.</li>
<li><a href="#BM_(pearson%20correlation%20coefficient%7Cpearson%20correlation%7Cpearson%20distance%7Cpearson%20similarity)%2Fi">pearson distance</a> between two vectors.</li>
<li>etc..</li>
</ul>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>See also: <a href="#BM_(similarity%20metric)%2Fi">similarity metric</a>.</p>
</div>
</li>
<li>
<p><a name="BM_(distance%20matrix%7Cdistance%20matrices)%2Fi"></a><strong>distance matrix</strong> - Given a set of n different entities, a <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> is an n-by-n matrix where each element contains the distance between the entities for that cell. For example, for the species snake, lizard, bird, and crocodile ...</p>
<table>
<thead>
<tr>
<th></th>
<th>Snake</th>
<th>Lizard</th>
<th>Bird</th>
<th>Crocodile</th>
</tr>
</thead>
<tbody>
<tr>
<td>Snake</td>
<td>0</td>
<td>2</td>
<td>6</td>
<td>4</td>
</tr>
<tr>
<td>Lizard</td>
<td>2</td>
<td>0</td>
<td>6</td>
<td>4</td>
</tr>
<tr>
<td>Bird</td>
<td>6</td>
<td>6</td>
<td>0</td>
<td>5</td>
</tr>
<tr>
<td>Crocodile</td>
<td>4</td>
<td>4</td>
<td>5</td>
<td>0</td>
</tr>
</tbody>
</table>
<p>The <a href="#BM_(distance%20metric)%2Fi">distance metric</a> can be anything so long as it meets the following properties:</p>
<ul>
<li>Must produce a non-negative distance -- dist(A,B) &gt;= 0</li>
<li>Must produce the same distance regardless of species order -- dist(A,B) == dist(B,A)</li>
<li>Must produce a distance that satisfies the triangle inequality property -- if dist(B,C) = x, then dist(A,B) + dist(A,C) &gt;= x</li>
</ul>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>I think what the last bullet point means is that the distance will be &gt;= if you travel to it indirectly (hop over to it instead of taking a straight path). For example, if dist(B,C) = 5, then dist(A,B) + dist(A,C) must be &gt;= 5.</p>
<p><img src="svgbob_3f0c5795b674de5612a12db473d621ed2ed53e92.svg" alt="Kroki diagram output"></p>
<p>A, B, and C are species.</p>
</div>
<p>Common <a href="#BM_(distance%20metric)%2Fi">distance metric</a>s include...</p>
<ul>
<li><a href="#BM_(hamming%20distance)%2Fi">hamming distance</a> between the DNA <a href="#BM_(sequence)%2Fi">sequence</a>s.</li>
<li><a href="#BM_(levenshtein%20distance%7Cstring%20distance)%2Fi">levenshtein distance</a> between the DNA <a href="#BM_(sequence)%2Fi">sequence</a>s.</li>
<li><a href="#BM_%5Cb(2-break%7C2%20break%7Ctwo%20break%7Ctwo-break)s%3F%5Cb%2Fi">two-break</a> count (<a href="#BM_(reversal%20distance)%2Fi">reversal distance</a>).</li>
</ul>
<p><a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">Distance matrices</a> are used to generate <a href="#BM_(phylogenetic%20tree%7Cphylogeny%20tree%7Cphylogenies%7Cevolutionary%20tree)%2Fi">phylogenetic tree</a>s. A single <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> may fit many different <a href="#BM_(tree)%2Fi">tree</a>s or it's possible that it fits no <a href="#BM_(tree)%2Fi">tree</a> at all. For example, the <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> above fits the <a href="#BM_(tree)%2Fi">tree</a>...</p>
<p><img src="svgbob_5a2ddd5247c86bd4e9d05f814cb689158dca59b0.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a name="BM_(tree)%2Fi"></a><strong>tree</strong> - In graph theory, a <a href="#BM_(tree)%2Fi">tree</a> is an acyclic undirected graph in which any two nodes are <a href="#BM_(connected)%2Fi">connected</a> by exactly one path (nodes branch outward / never converge).</p>
<p><a href="#BM_(tree)%2Fi">Tree</a>s come in two forms: <a href="#BM_(rooted%20tree)%2Fi">rooted tree</a>s and <a href="#BM_(un-rooted%20tree%7Cunrooted%20tree)%2Fi">unrooted tree</a>s. In graph theory, a <a href="#BM_(tree)%2Fi">tree</a> typically refers to an <a href="#BM_(un-rooted%20tree%7Cunrooted%20tree)%2Fi">unrooted tree</a>.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>This is different from the computer science definition of <a href="#BM_(tree)%2Fi">tree</a>, which is an abstract data type representing a hierarchy (always a single root that flows downwards), typically generalized as a <em>directed</em> acyclic graph as opposed to an un<a href="#BM_(directed%20acyclic%20graph)%2Fi">directed acyclic graph</a>.</p>
</div>
</li>
<li>
<p><a name="BM_(un-rooted%20tree%7Cunrooted%20tree)%2Fi"></a><strong>unrooted tree</strong> - A <a href="#BM_(tree)%2Fi">tree</a> without a root node...</p>
<p><img src="svgbob_3ae717329571359a328c5697332e94ce675003fb.svg" alt="Kroki diagram output"></p>
<p>An <a href="#BM_(un-rooted%20tree%7Cunrooted%20tree)%2Fi">unrooted tree</a> may be turned into a <a href="#BM_(rooted%20tree)%2Fi">rooted tree</a> by choosing any non-leaf node (internal node) to be the root node.</p>
</li>
<li>
<p><a name="BM_(rooted%20tree)%2Fi"></a><strong>rooted tree</strong> - A <a href="#BM_(tree)%2Fi">tree</a> with a root node...</p>
<p><img src="svgbob_4a12a5e83db150bd3a7d62bc553c9ece9b5d54b2.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a name="BM_(subtree%7Csub-tree)%2Fi"></a><strong>subtree</strong> - Given a node in a <a href="#BM_(tree)%2Fi">tree</a>, that node and all of its descendants comprise a <a href="#BM_(subtree%7Csub-tree)%2Fi">subtree</a>. For example, the following <a href="#BM_(tree)%2Fi">tree</a> has the <a href="#BM_(subtree%7Csub-tree)%2Fi">subtree</a> ...</p>
<p><img src="svgbob_045b94a42c0e447555bf437dbf0da870b4ab806c.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a name="BM_(degree)_GRAPH%2Fi"></a><strong>degree</strong> - The number of edges leading into / out of a node of an undirected graph.</p>
<p>The node below has a <a href="#BM_(degree)_GRAPH%2Fi">degree</a> of 3.</p>
<p><img src="svgbob_60f0bcaccb25a424cce90b06db0fd197ae921dda.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a name="BM_(simple%20tree)%2Fi"></a><strong>simple tree</strong> - An <a href="#BM_(un-rooted%20tree%7Cunrooted%20tree)%2Fi">unrooted tree</a> where ...</p>
<ul>
<li>every internal node has a <a href="#BM_(degree)_GRAPH%2Fi">degree</a> &gt; 2.</li>
<li>every edge has a weight of &gt; 0.</li>
</ul>
<p><img src="svgbob_a5aaa10a8a80878b13db3c7e7a4e5ffee7ffa19c.svg" alt="Kroki diagram output"></p>
<p>In the context of <a href="#BM_(phylogeny%7Cphylogenetic)%2Fi">phylogeny</a>, a <a href="#BM_(simple%20tree)%2Fi">simple tree</a>'s ...</p>
<ul>
<li>leaf nodes represent known entities.</li>
<li>internal nodes represent inferred ancestor entities.</li>
<li>edge weights represent distances between entities.</li>
</ul>
<p>The restrictions placed on <a href="#BM_(simple%20tree)%2Fi">simple tree</a>s simplify the process of working backwards from a <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> to a <a href="#BM_(phylogenetic%20tree%7Cphylogeny%20tree%7Cphylogenies%7Cevolutionary%20tree)%2Fi">phylogenetic tree</a>.</p>
</li>
<li>
<p><a name="BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi"></a><strong>additive distance matrix</strong> - Given a <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a>, if there exists a <a href="#BM_(tree)%2Fi">tree</a> with edge weights that satisfy that <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> (referred to as fit), that <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> is said to be an <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>.</p>
<p>For example, the following <a href="#BM_(tree)%2Fi">tree</a> fits the following <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> ...</p>
<table>
<thead>
<tr>
<th></th>
<th>Cat</th>
<th>Lion</th>
<th>Bear</th>
</tr>
</thead>
<tbody>
<tr>
<td>Cat</td>
<td>0</td>
<td>2</td>
<td>3</td>
</tr>
<tr>
<td>Lion</td>
<td>2</td>
<td>0</td>
<td>3</td>
</tr>
<tr>
<td>Bear</td>
<td>3</td>
<td>3</td>
<td>0</td>
</tr>
</tbody>
</table>
<p><img src="svgbob_e5ec5970ccdbca871edbc896779d5d1131ed9055.svg" alt="Kroki diagram output"></p>
<p>The term additive is used because the weights of all edges along the path between leaves (i, j) add to <code>dist(i, j)</code> in the <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a>. Not all <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrices</a> are additive. For example, no <a href="#BM_(simple%20tree)%2Fi">simple tree</a> exists that satisfies the following <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a>...</p>
<table>
<thead>
<tr>
<th></th>
<th>S1</th>
<th>S2</th>
<th>S3</th>
<th>S4</th>
</tr>
</thead>
<tbody>
<tr>
<td>S1</td>
<td>0</td>
<td>3</td>
<td>4</td>
<td>3</td>
</tr>
<tr>
<td>S2</td>
<td>3</td>
<td>0</td>
<td>4</td>
<td>5</td>
</tr>
<tr>
<td>S3</td>
<td>4</td>
<td>4</td>
<td>0</td>
<td>2</td>
</tr>
<tr>
<td>S4</td>
<td>3</td>
<td>5</td>
<td>2</td>
<td>0</td>
</tr>
</tbody>
</table>
<ul>
<li>
<p>Test <a href="#BM_(simple%20tree)%2Fi">simple tree</a> 1:</p>
<p><img src="svgbob_74cdc79c5f9b82937ba473b3e3e632aab5cb6b4e.svg" alt="Kroki diagram output"></p>
<pre><code>dist(S1, S2) is 3 = w + x
dist(S1, S3) is 4 = w + y
dist(S1, S4) is 3 = w + z
dist(S2, S3) is 4 = x + y
dist(S2, S4) is 5 = x + z
dist(S3, S4) is 2 = y + z
</code></pre>
<p>Attempting to solve this produces inconsistent results. Solved values for each variable don't work across all equations present.</p>
</li>
<li>
<p>Test <a href="#BM_(simple%20tree)%2Fi">simple tree</a> 2:</p>
<p><img src="svgbob_310a023ae046f0c94fc708941d7e494797cf787b.svg" alt="Kroki diagram output"></p>
<pre><code>dist(S1, S2) is 3 = w + x
dist(S1, S3) is 4 = w + u + y
dist(S1, S4) is 3 = w + u + z
dist(S2, S3) is 4 = x + u + y
dist(S2, S4) is 5 = x + u + z
dist(S3, S4) is 2 = y + z
</code></pre>
<p>Attempting to solve this produces inconsistent results. Solved values for each variable don't work across all equations present.</p>
</li>
<li>
<p>Test <a href="#BM_(simple%20tree)%2Fi">simple tree</a> 3:</p>
<p><img src="svgbob_27b6be0c6cb6193b60f5164ee5e83ce0f0389d38.svg" alt="Kroki diagram output"></p>
<pre><code>dist(S1, S2) is 4 = w + u + y
dist(S1, S3) is 3 = w + x
dist(S1, S4) is 3 = w + u + z
dist(S2, S3) is 4 = x + u + y
dist(S2, S4) is 2 = y + z
dist(S3, S4) is 5 = x + u + z
</code></pre>
<p>Attempting to solve this produces inconsistent results. Solved values for each variable don't work across all equations present.</p>
</li>
<li>
<p>etc..</p>
</li>
</ul>
</li>
<li>
<p><a name="BM_(neighbour)%2Fi"></a><strong>neighbour</strong> - Given two leaf nodes in a <a href="#BM_(tree)%2Fi">tree</a>, those leaf nodes are said to be <a href="#BM_(neighbour)%2Fi">neighbour</a>s if they share they connect to the same internal node. For example, leaf nodes A and B are <a href="#BM_(neighbour)%2Fi">neighbour</a>s in the following <a href="#BM_(tree)%2Fi">tree</a> because they both connect to internal node D ...</p>
<p><img src="svgbob_80372a14f2a248e2ffdfdbfef223b5b15c39f7b1.svg" alt="Kroki diagram output"></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>A leaf node will only ever have 1 parent, by definition of a <a href="#BM_(tree)%2Fi">tree</a>.</p>
</div>
</li>
<li>
<p><a name="BM_(limb)%2Fi"></a><strong>limb</strong> - Given a leaf node in a <a href="#BM_(tree)%2Fi">tree</a>, that leaf node's <a href="#BM_(limb)%2Fi">limb</a> is the edge between it and its parent (node it's <a href="#BM_(connected)%2Fi">connected</a> to). For example, the following <a href="#BM_(tree)%2Fi">tree</a> has the following <a href="#BM_(limb)%2Fi">limb</a>s ...</p>
<p><img src="svgbob_80372a14f2a248e2ffdfdbfef223b5b15c39f7b1.svg" alt="Kroki diagram output"></p>
<ul>
<li>(A, D)</li>
<li>(B, D)</li>
<li>(G, I)</li>
</ul>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>A leaf node will only ever have 1 parent, by definition of a <a href="#BM_(tree)%2Fi">tree</a>.</p>
</div>
</li>
<li>
<p><a name="BM_(limb%20length%7Climb%20weight)%2Fi"></a><strong>limb length</strong> - Given a leaf node in a <a href="#BM_(tree)%2Fi">tree</a>, the leaf node's <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a> is the weight assigned to its <a href="#BM_(limb)%2Fi">limb</a>. For example, node A has a <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a> of 2 in the following <a href="#BM_(tree)%2Fi">tree</a>...</p>
<p><img src="svgbob_54ab99476bd3bb4646b1899b68122441df1dc495.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a name="BM_(four%20points%3F%20condition%7Cfour%20point%20theorem)%2Fi"></a><strong>four point condition</strong> - An algorithm for determining if a <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> is an <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>. Given four leaf nodes, the algorithm checks different permutations of those leaf nodes to see if any pass a test, where that test builds node pairings from the quartet and checks their distances to see if they meet a specific condition...</p>
<pre class="hljs"><code><span class="hljs-keyword">for</span> a, b, c, d <span class="hljs-keyword">in</span> permutations(quartet, r=<span class="hljs-number">4</span>):  <span class="hljs-comment"># find one perm that passes the following test</span>
    s1 = dist_mat[a][b] + dist_mat[c][d]  <span class="hljs-comment"># sum of dists for (a,b) and (c,d)</span>
    s2 = dist_mat[a][c] + dist_mat[b][d]  <span class="hljs-comment"># sum of dists for (a,c) and (b,d)</span>
    s3 = dist_mat[a][d] + dist_mat[b][c]  <span class="hljs-comment"># sum of dists for (a,d) and (b,c)</span>
    <span class="hljs-keyword">if</span> s1 &lt;= s2 == s3:
        <span class="hljs-keyword">return</span> <span class="hljs-literal">True</span>
<span class="hljs-keyword">return</span> <span class="hljs-literal">False</span>
</code></pre>
<p>If all possible leaf node quartets pass the above test, the <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> is an <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> (was derived from a <a href="#BM_(tree)%2Fi">tree</a> / fits a <a href="#BM_(tree)%2Fi">tree</a>).</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>See <a href="#BM_(Algorithms%5C%2FPhylogeny%2FTest%20Additive%20Distance%20Matrix)_TOPIC%2F">Algorithms/Phylogeny/Test Additive Distance Matrix</a> for a full explanation of how this algorithm works.</p>
</div>
</li>
<li>
<p><a name="BM_(trim%20distance%20matrix%7Ctrimmed%20distance%20matrix%7Ctrim%20distance%20matrices%7Ctrimmed%20distance%20matrices)%2Fi"></a><strong>trimmed distance matrix</strong> - A <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> where a leaf node's row and column have been removed. This is equivalent to removing the leaf node's <a href="#BM_(limb)%2Fi">limb</a> in the corresponding <a href="#BM_(simple%20tree)%2Fi">simple tree</a> and merging together any edges <a href="#BM_(connected)%2Fi">connected</a> by nodes of <a href="#BM_(degree)_GRAPH%2Fi">degree</a> 2.</p>
<p>For example, removing v2 from the following <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a>...</p>
<table>
<thead>
<tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th><span style="background-color: red; color: white; padding: 3px">v2</span></th>
<th>v3</th>
</tr>
</thead>
<tbody>
<tr>
<td>v0</td>
<td>0</td>
<td>13</td>
<td>21</td>
<td>22</td>
</tr>
<tr>
<td>v1</td>
<td>13</td>
<td>0</td>
<td>12</td>
<td>13</td>
</tr>
<tr>
<td><span style="background-color: red; color: white; padding: 3px">v2</span></td>
<td>21</td>
<td>12</td>
<td>0</td>
<td>13</td>
</tr>
<tr>
<td>v3</td>
<td>22</td>
<td>13</td>
<td>13</td>
<td>0</td>
</tr>
</tbody>
</table>
<p><img src="dot_4e5daffbacf71191e85a5881e8367790.svg" alt="Dot diagram"></p>
<p>... results in v2's row and column being removed ...</p>
<table>
<thead>
<tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v3</th>
</tr>
</thead>
<tbody>
<tr>
<td>v0</td>
<td>0</td>
<td>13</td>
<td>22</td>
</tr>
<tr>
<td>v1</td>
<td>13</td>
<td>0</td>
<td>13</td>
</tr>
<tr>
<td>v3</td>
<td>22</td>
<td>13</td>
<td>0</td>
</tr>
</tbody>
</table>
<p><img src="dot_8970363f8d351a28a22328fbfd029f13.svg" alt="Dot diagram"></p>
</li>
<li>
<p><a name="BM_(bald%20distance%20matrix%7Cbalded%20distance%20matrix%7Cbald%20distance%20matrices%7Cbalded%20distance%20matrices)%2Fi"></a><strong>balded distance matrix</strong> - An <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> where the distances in a leaf node's row and column have been subtracted by that leaf node's <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a>. This is equivalent to setting the leaf node's <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a> to 0 in the corresponding <a href="#BM_(simple%20tree)%2Fi">simple tree</a>.</p>
<p> <!-- Needs to be disabled because v3's is conflicting with 3' --></p>
<p>For example, balding v5's limb length in the following distance matrix ...</p>
<table>
<thead>
<tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr>
</thead>
<tbody>
<tr>
<td>v0</td>
<td>0</td>
<td>13</td>
<td>21</td>
<td>21</td>
<td>22</td>
<td>22</td>
</tr>
<tr>
<td>v1</td>
<td>13</td>
<td>0</td>
<td>12</td>
<td>12</td>
<td>13</td>
<td>13</td>
</tr>
<tr>
<td>v2</td>
<td>21</td>
<td>12</td>
<td>0</td>
<td>20</td>
<td>21</td>
<td>21</td>
</tr>
<tr>
<td>v3</td>
<td>21</td>
<td>12</td>
<td>20</td>
<td>0</td>
<td>7</td>
<td>13</td>
</tr>
<tr>
<td>v4</td>
<td>22</td>
<td>13</td>
<td>21</td>
<td>7</td>
<td>0</td>
<td>14</td>
</tr>
<tr>
<td>v5</td>
<td>22</td>
<td>13</td>
<td>21</td>
<td>13</td>
<td>14</td>
<td>0</td>
</tr>
</tbody>
</table>
<p><img src="dot_e304cd9917c7e5b9d72c4e876e260cba.svg" alt="Dot diagram"></p>
<p>... results in ...</p>
<table>
<thead>
<tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr>
</thead>
<tbody>
<tr>
<td>v0</td>
<td>0</td>
<td>13</td>
<td>21</td>
<td>21</td>
<td>22</td>
<td>15</td>
</tr>
<tr>
<td>v1</td>
<td>13</td>
<td>0</td>
<td>12</td>
<td>12</td>
<td>13</td>
<td>6</td>
</tr>
<tr>
<td>v2</td>
<td>21</td>
<td>12</td>
<td>0</td>
<td>20</td>
<td>21</td>
<td>14</td>
</tr>
<tr>
<td>v3</td>
<td>21</td>
<td>12</td>
<td>20</td>
<td>0</td>
<td>7</td>
<td>6</td>
</tr>
<tr>
<td>v4</td>
<td>22</td>
<td>13</td>
<td>21</td>
<td>7</td>
<td>0</td>
<td>7</td>
</tr>
<tr>
<td>v5</td>
<td>15</td>
<td>6</td>
<td>14</td>
<td>6</td>
<td>7</td>
<td>0</td>
</tr>
</tbody>
</table>
<p><img src="dot_504f11240750f6a2add064fe593ca205.svg" alt="Dot diagram"></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>Technically, an edge weight of 0 is a violation of the <a href="#BM_(simple%20tree)%2Fi">simple tree</a> requirement of having edge weights &gt; 0. This is a special case.</p>
</div>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>How do you know the <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a> from just the <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a>? See the algorithm to determine <a href="#BM_(limb%20length%7Climb%20weight)%2Fi">limb length</a> for any leaf from just the <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a>.</p>
</div>
<p> <!-- Re-enabling. Needed to be disabled because v3's is conflicting with 3' --></p>
</li>
<li>
<p><a name="BM_(additive%20phylogeny)%2Fi"></a><strong>additive phylogeny</strong> - A recursive algorithm that finds the unique <a href="#BM_(simple%20tree)%2Fi">simple tree</a> for some <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>. The algorithm <a href="#BM_(trim distance matrix|trimmed distance matrix|trim distance matrices|trimmed distance matrices)/i">trim</a>s a single leaf node at each recursive step until the <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> has a size of two. The <a href="#BM_(simple%20tree)%2Fi">simple tree</a> for any two leaf nodes is those two nodes <a href="#BM_(connected)%2Fi">connected</a> by a single edge. Using that <a href="#BM_(tree)%2Fi">tree</a> as its base, the algorithm recurses out of each step by finding where that step's trimmed node exists on the <a href="#BM_(tree)%2Fi">tree</a> and attaching it on.</p>
<p>At the end, the algorithm will have constructed the entire <a href="#BM_(simple%20tree)%2Fi">simple tree</a> for the <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>. For example, ...</p>
<ul>
<li>
<p>Initial <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> ...</p>
<table>
<thead>
<tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
</tr>
</thead>
<tbody>
<tr>
<td>v0</td>
<td>0</td>
<td>13</td>
<td>21</td>
<td>22</td>
</tr>
<tr>
<td>v1</td>
<td>13</td>
<td>0</td>
<td>12</td>
<td>13</td>
</tr>
<tr>
<td>v2</td>
<td>21</td>
<td>12</td>
<td>0</td>
<td>13</td>
</tr>
<tr>
<td>v3</td>
<td>22</td>
<td>13</td>
<td>13</td>
<td>0</td>
</tr>
</tbody>
</table>
</li>
<li>
<p>Trim v1 to produce <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> ...</p>
<table>
<thead>
<tr>
<th></th>
<th>v0</th>
<th>v2</th>
<th>v3</th>
</tr>
</thead>
<tbody>
<tr>
<td>v0</td>
<td>0</td>
<td>21</td>
<td>22</td>
</tr>
<tr>
<td>v2</td>
<td>21</td>
<td>0</td>
<td>13</td>
</tr>
<tr>
<td>v3</td>
<td>22</td>
<td>13</td>
<td>0</td>
</tr>
</tbody>
</table>
</li>
<li>
<p>Trim v0 to produce <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> ...</p>
<table>
<thead>
<tr>
<th></th>
<th>v2</th>
<th>v3</th>
</tr>
</thead>
<tbody>
<tr>
<td>v2</td>
<td>0</td>
<td>13</td>
</tr>
<tr>
<td>v3</td>
<td>13</td>
<td>0</td>
</tr>
</tbody>
</table>
</li>
<li>
<p><a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">Distance matrix</a> maps to the obvious <a href="#BM_(simple%20tree)%2Fi">simple tree</a>...</p>
<p><img src="dot_17295d277521c7d7445601414800311c.svg" alt="Dot diagram"></p>
</li>
<li>
<p>Attach v0 to produce <a href="#BM_(tree)%2Fi">tree</a>...</p>
<p><img src="dot_ee7b26791f6d3ee197982f570c83e294.svg" alt="Dot diagram"></p>
</li>
<li>
<p>Attach v1 to produce <a href="#BM_(tree)%2Fi">tree</a>...</p>
<p><img src="dot_a9a8b9059c8bf67a1d36b651e907a38a.svg" alt="Dot diagram"></p>
</li>
</ul>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>See <a href="#BM_(Algorithms%5C%2FPhylogeny%2FDistance%20Matrix%20to%20Tree%2FAdditive%20Phylogeny%20Algorithm)_TOPIC%2F">Algorithms/Phylogeny/Distance Matrix to Tree/Additive Phylogeny Algorithm</a> for a full explanation of how this algorithm works.</p>
</div>
</li>
<li>
<p><a name="BM_(sum%20of%20squared%20errors%7Csum%20of%20square%20error%7Csum%20of%20errors%20squared)%2Fi"></a><strong>sum of squared errors</strong> - <a href="#BM_(sum%20of%20squared%20errors%7Csum%20of%20square%20error%7Csum%20of%20errors%20squared)%2Fi">Sum of errors squared</a> is an algorithm used to quantify how far off some estimation / prediction is.</p>
<p>Given a set of real values and a set of predicted values, the error is the difference between the real and predicted values at each data point. For example...</p>
<table>
<thead>
<tr>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
</tr>
</thead>
<tbody>
<tr>
<td>Real</td>
<td>5</td>
<td>4</td>
<td>7</td>
<td>8</td>
<td>5</td>
<td>4</td>
</tr>
<tr>
<td>Predicted</td>
<td>4</td>
<td>5</td>
<td>7</td>
<td>6</td>
<td>4</td>
<td>4</td>
</tr>
<tr>
<td>Error</td>
<td>1</td>
<td>-1</td>
<td>0</td>
<td>2</td>
<td>1</td>
<td>0</td>
</tr>
</tbody>
</table>
<p>The algorithm squares each error and sums them together:</p>
<pre class="hljs"><code>res = <span class="hljs-number">0</span>
<span class="hljs-keyword">for</span> r_val, p_val <span class="hljs-keyword">in</span> <span class="hljs-built_in">zip</span>(real, predicted):
  err = r_val - p_val
  res += err ** <span class="hljs-number">2</span>
</code></pre>
<p>The algorithm as a formula: <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msubsup><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>0</mn></mrow><mi>n</mi></msubsup><mrow><mo stretchy="false">(</mo><msub><mi>r</mi><mi>i</mi></msub><mo>−</mo><msub><mi>p</mi><mi>i</mi></msub><msup><mo stretchy="false">)</mo><mn>2</mn></msup></mrow></mrow><annotation encoding="application/x-tex">\sum_{i=0}^{n}{(r_i - p_i)^2}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.113818em;vertical-align:-0.29971000000000003em;"></span><span class="mop"><span class="mop op-symbol small-op" style="position:relative;top:-0.0000050000000000050004em;">∑</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.804292em;"><span style="top:-2.40029em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">i</span><span class="mrel mtight">=</span><span class="mord mtight">0</span></span></span></span><span style="top:-3.2029em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">n</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.29971000000000003em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord"><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02778em;">r</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal">p</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mclose"><span class="mclose">)</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8141079999999999em;"><span style="top:-3.063em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span></span></span></span></span></p>
</li>
<li>
<p><a name="BM_(speciation)%2Fi"></a><strong>speciation</strong> - The evolutionary process by which a species splits into distinct child species.</p>
<p>In <a href="#BM_(phylogenetic%20tree%7Cphylogeny%20tree%7Cphylogenies%7Cevolutionary%20tree)%2Fi">phylogenetic tree</a>s, internal nodes branching are assumed to be <a href="#BM_(speciation)%2Fi">speciation</a> events. That is, an event where the ancestral species represented by that node splits into distinct child species.</p>
</li>
<li>
<p><a name="BM_(unrooted%20binary%20tree)%2Fi"></a><strong>unrooted binary tree</strong> - In the context of <a href="#BM_(phylogeny%7Cphylogenetic)%2Fi">phylogeny</a>, an <a href="#BM_(unrooted%20binary%20tree)%2Fi">unrooted binary tree</a> is a <a href="#BM_(simple%20tree)%2Fi">simple tree</a> where internal nodes must have a <a href="#BM_(degree)_GRAPH%2Fi">degree</a> of 3...</p>
<p><img src="svgbob_b29e6fe881c0b485a2b8b1eac26f15987cb45a9a.svg" alt="Kroki diagram output"></p>
<p>In other words, an edge leading to an internal node is guaranteed to branch <em>exactly</em> twice.</p>
<p><img src="svgbob_43a3ec6298299c551b148002f27fa4910a77a171.svg" alt="Kroki diagram output"></p>
<p>Contrast that to normal <a href="#BM_(simple%20tree)%2Fi">simple tree</a>s where internal nodes can have any <a href="#BM_(degree)_GRAPH%2Fi">degree</a> greater than 2...</p>
<p><img src="svgbob_17ac9bddd64ed68c7e7e56c709f3a2967622a61a.svg" alt="Kroki diagram output"></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>Recall that <a href="#BM_(simple%20tree)%2Fi">simple tree</a>s are unrooted to begin with and can't have nodes with <a href="#BM_(degree)_GRAPH%2Fi">degree</a> 2 (train of non-branching edges not allowed).</p>
</div>
</li>
<li>
<p><a name="BM_(rooted%20binary%20tree)%2Fi"></a><strong>rooted binary tree</strong> - In the context of <a href="#BM_(phylogeny%7Cphylogenetic)%2Fi">phylogeny</a>, a <a href="#BM_(rooted%20binary%20tree)%2Fi">rooted binary tree</a> is an <a href="#BM_(unrooted%20binary%20tree)%2Fi">unrooted binary tree</a> with a root node injected in between one of its edges. The injected root node will always end up as an internal node of <a href="#BM_(degree)_GRAPH%2Fi">degree</a> of 2, breaking the constraint of ...</p>
<ul>
<li><a href="#BM_(unrooted%20binary%20tree)%2Fi">unrooted binary tree</a>s that states internal nodes must have <a href="#BM_(degree)_GRAPH%2Fi">degree</a> of exactly 3.</li>
<li><a href="#BM_(simple%20tree)%2Fi">simple tree</a>s that states internal nodes must have <a href="#BM_(degree)_GRAPH%2Fi">degree</a> of greater than 2.</li>
</ul>
<p><img src="svgbob_9d8f5b3c109dc96946bf1d24e21ac3304a29070c.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a name="BM_(ultrametric%20tree%7Cultrametric)%2Fi"></a><strong>ultrametric tree</strong> - A <a href="#BM_(rooted%20tree)%2Fi">rooted tree</a> where all leaf nodes are equidistant from the root.</p>
<p><img src="svgbob_10fa2bf98dc55b5346b2ebb7f45c2e56a81730fe.svg" alt="Kroki diagram output"></p>
<p>In the example above, all leaf nodes are a distance of 4 from the root.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>Does an <a href="#BM_(ultrametric%20tree%7Cultrametric)%2Fi">ultrametric tree</a> have to be a <a href="#BM_(rooted%20binary%20tree)%2Fi">rooted binary tree</a>? I think the answer is no: <a href="#BM_(unweighted%20pair%20group%20method%20with%20arithmetic%20mean%7CUPGMA)%2Fi">UPGMA</a> generated <a href="#BM_(rooted%20binary%20tree)%2Fi">rooted binary tree</a>s, but <a href="#BM_(ultrametric%20tree%7Cultrametric)%2Fi">ultrametric tree</a>s in general just have to be <a href="#BM_(rooted%20tree)%2Fi">rooted tree</a>s / they don't have to be binary.</p>
</div>
</li>
<li>
<p><a name="BM_(molecular%20clock)%2Fi"></a><strong>molecular clock</strong> - The assumption that the rate of mutation is more-or-less consistent. For example, ...</p>
<ul>
<li>every minute, around n of m nucleotides get mutated.</li>
<li>every hour, around n <a href="#BM_(genome%20rearrangement%7Cchromosomal%20rearrangement%7Cchromosome%20rearrangement)%2Fi">genome rearrangement</a> reversals occur per chromosome segment of size m.</li>
<li>etc..</li>
</ul>
<p>This assumption is used for some <a href="#BM_(phylogeny%7Cphylogenetic)%2Fi">phylogeny</a> algorithms (e.g. <a href="#BM_(unweighted%20pair%20group%20method%20with%20arithmetic%20mean%7CUPGMA)%2Fi">UPGMA</a>).</p>
</li>
<li>
<p><a name="BM_(unweighted%20pair%20group%20method%20with%20arithmetic%20mean%7CUPGMA)%2Fi"></a><strong>unweighted pair group method with arithmetic mean (UPGMA)</strong> - A heuristic algorithm used to estimate a binary <a href="#BM_(ultrametric%20tree%7Cultrametric)%2Fi">ultrametric tree</a> for some <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a>.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>A binary <a href="#BM_(ultrametric%20tree%7Cultrametric)%2Fi">ultrametric tree</a> is an <a href="#BM_(ultrametric%20tree%7Cultrametric)%2Fi">ultrametric tree</a> where each internal node only branches to two children. In other words, a binary <a href="#BM_(ultrametric%20tree%7Cultrametric)%2Fi">ultrametric tree</a> is a <a href="#BM_(rooted%20binary%20tree)%2Fi">rooted binary tree</a> where all leaf nodes are equidistant from the root.</p>
</div>
<p>The algorithm assumes that the rate of mutation is consistent (<a href="#BM_(molecular%20clock)%2Fi">molecular clock</a>). This assumption is what makes the <a href="#BM_(tree)%2Fi">tree</a> <a href="#BM_(ultrametric%20tree%7Cultrametric)%2Fi">ultrametric</a>. A set of present day species (leaf nodes) are assumed to all have the same amount of mutation (distance) from their shared ancestor (shared internal node).</p>
<p><img src="svgbob_9c45e3a8fedf42d570c570b7c417173e7ed57970.svg" alt="Kroki diagram output"></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>See <a href="#BM_(Algorithms%5C%2FPhylogeny%2FDistance%20Matrix%20to%20Tree%2FUPGMA%20Algorithm)_TOPIC%2F">Algorithms/Phylogeny/Distance Matrix to Tree/UPGMA Algorithm</a> for a full explanation of how this algorithm works.</p>
</div>
</li>
<li>
<p><a name="BM_(neighbour%20joining%20matrix)%2Fi"></a><strong>neighbour joining matrix</strong> - A matrix produced by transforming a <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> such that each element is calculated as total_dist(a) + total_dist(b) - (n - 2) * dist(a, b), where...</p>
<ul>
<li>n is the number of leaf nodes in the <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a>.</li>
<li>a and b are the leaf nodes being referenced in the <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a>.</li>
<li>dist(a, b) returns the distance between leaf nodes a and b in the <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a>.</li>
<li>total dist(a) returns the sum of all distances to / from leaf node a.</li>
<li>total dist(b) returns the sum of all distances to / from leaf node b.</li>
</ul>
<p>The maximum element in the <a href="#BM_(neighbour%20joining%20matrix)%2Fi">neighbour joining matrix</a> is guaranteed to be for two <a href="#BM_(neighbour)%2Fi">neighbour</a>ing leaf nodes. For example, the following <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> produces the following <a href="#BM_(neighbour%20joining%20matrix)%2Fi">neighbour joining matrix</a>...</p>
<table>
<thead>
<tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr>
</thead>
<tbody>
<tr>
<td>v0</td>
<td>0</td>
<td>13</td>
<td>21</td>
<td>21</td>
<td>22</td>
<td>22</td>
</tr>
<tr>
<td>v1</td>
<td>13</td>
<td>0</td>
<td>12</td>
<td>12</td>
<td>13</td>
<td>13</td>
</tr>
<tr>
<td>v2</td>
<td>21</td>
<td>12</td>
<td>0</td>
<td>20</td>
<td>21</td>
<td>21</td>
</tr>
<tr>
<td>v3</td>
<td>21</td>
<td>12</td>
<td>20</td>
<td>0</td>
<td>7</td>
<td>13</td>
</tr>
<tr>
<td>v4</td>
<td>22</td>
<td>13</td>
<td>21</td>
<td>7</td>
<td>0</td>
<td>14</td>
</tr>
<tr>
<td>v5</td>
<td>22</td>
<td>13</td>
<td>21</td>
<td>13</td>
<td>14</td>
<td>0</td>
</tr>
</tbody>
</table>
<table>
<thead>
<tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
<th>v4</th>
<th>v5</th>
</tr>
</thead>
<tbody>
<tr>
<td>v0</td>
<td>0</td>
<td>110</td>
<td>110</td>
<td>88</td>
<td>88</td>
<td>94</td>
</tr>
<tr>
<td>v1</td>
<td>110</td>
<td>0</td>
<td>110</td>
<td>88</td>
<td>88</td>
<td>94</td>
</tr>
<tr>
<td>v2</td>
<td>110</td>
<td>110</td>
<td>0</td>
<td>88</td>
<td>88</td>
<td>94</td>
</tr>
<tr>
<td>v3</td>
<td>88</td>
<td>88</td>
<td>88</td>
<td>0</td>
<td>122</td>
<td>104</td>
</tr>
<tr>
<td>v4</td>
<td>88</td>
<td>88</td>
<td>88</td>
<td>122</td>
<td>0</td>
<td>104</td>
</tr>
<tr>
<td>v5</td>
<td>94</td>
<td>94</td>
<td>94</td>
<td>104</td>
<td>104</td>
<td>0</td>
</tr>
</tbody>
</table>
<p>The maximum element is for (v3, v4), meaning that v3 and v4 are <a href="#BM_(neighbour)%2Fi">neighbour</a>ing leaf nodes.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>See <a href="#BM_(Algorithms%5C%2FPhylogeny%2FFind%20Neighbours)_TOPIC%2F">Algorithms/Phylogeny/Find Neighbours</a> for a full explanation of how this algorithm works.</p>
</div>
</li>
<li>
<p><a name="BM_(neighbour%20joining%20phylogeny)%2Fi"></a><strong>neighbour joining phylogeny</strong> - A recursive algorithm that can either...</p>
<ul>
<li>find the unique <a href="#BM_(simple%20tree)%2Fi">simple tree</a> for an <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>.</li>
<li>approximate a <a href="#BM_(simple%20tree)%2Fi">simple tree</a> for a non-<a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>.</li>
</ul>
<p>The algorithm finds and <a href="#BM_(trim distance matrix|trimmed distance matrix|trim distance matrices|trimmed distance matrices)/i">replace</a>s a pair of <a href="#BM_(neighbour)%2Fi">neighbour</a>ing leaf nodes in the <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> with their shared parent at each recursive step (parent is promoted to a leaf node) until the <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> has a size of two. The <a href="#BM_(simple%20tree)%2Fi">simple tree</a> for any two leaf nodes is those two nodes <a href="#BM_(connected)%2Fi">connected</a> by a single edge. Using that <a href="#BM_(tree)%2Fi">tree</a> as its base, the algorithm recurses out of each step by attaching the <a href="#BM_(neighbour)%2Fi">neighbour</a>s removed from the distance at that step to their parent in the <a href="#BM_(tree)%2Fi">tree</a>.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The term <a href="#BM_(neighbour)%2Fi">neighbour</a>ing means having a shared parent in the <a href="#BM_(tree)%2Fi">tree</a>, not next to each other in the <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a>.</p>
</div>
<p>At the end, the algorithm will have constructed the entire <a href="#BM_(simple%20tree)%2Fi">simple tree</a> for the <a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a>. For example, ...</p>
<ul>
<li>
<p>Initial non-<a href="#BM_(additive%20matrix%7Cadditive%20matrices%7Cadditive%20distance%20matrix%7Cadditive%20distance%20matrices)%2Fi">additive distance matrix</a> ...</p>
<table>
<thead>
<tr>
<th></th>
<th>v0</th>
<th>v1</th>
<th>v2</th>
<th>v3</th>
</tr>
</thead>
<tbody>
<tr>
<td>v0</td>
<td>0</td>
<td>16</td>
<td>22</td>
<td>22</td>
</tr>
<tr>
<td>v1</td>
<td>16</td>
<td>0</td>
<td>13</td>
<td>12</td>
</tr>
<tr>
<td>v2</td>
<td>22</td>
<td>13</td>
<td>0</td>
<td>11</td>
</tr>
<tr>
<td>v3</td>
<td>22</td>
<td>12</td>
<td>11</td>
<td>0</td>
</tr>
</tbody>
</table>
</li>
<li>
<p>Replace <a href="#BM_(neighbour)%2Fi">neighbour</a>s (v1, v0) with their parent N1 to produce <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> ...</p>
<table>
<thead>
<tr>
<th></th>
<th>N1</th>
<th>v2</th>
<th>v3</th>
</tr>
</thead>
<tbody>
<tr>
<td>N1</td>
<td>0</td>
<td>22</td>
<td>22</td>
</tr>
<tr>
<td>v2</td>
<td>9.5</td>
<td>0</td>
<td>11</td>
</tr>
<tr>
<td>v3</td>
<td>9</td>
<td>11</td>
<td>0</td>
</tr>
</tbody>
</table>
</li>
<li>
<p>Replace <a href="#BM_(neighbour)%2Fi">neighbour</a>s (v2, v3) with their parent N2 to produce <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrix</a> ...</p>
<table>
<thead>
<tr>
<th></th>
<th>N1</th>
<th>N2</th>
</tr>
</thead>
<tbody>
<tr>
<td>N1</td>
<td>0</td>
<td>3.75</td>
</tr>
<tr>
<td>N2</td>
<td>3.75</td>
<td>0</td>
</tr>
</tbody>
</table>
</li>
<li>
<p><a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">Distance matrix</a> maps to the obvious <a href="#BM_(simple%20tree)%2Fi">simple tree</a>...</p>
<p><img src="dot_9f4177ab91402e113243c9985f953563.svg" alt="Dot diagram"></p>
</li>
<li>
<p>Attach (v2, v3) to N2 to produce <a href="#BM_(tree)%2Fi">tree</a>...</p>
<p><img src="dot_3bcebffa84a180db2bd3ef759fd52701.svg" alt="Dot diagram"></p>
</li>
<li>
<p>Attach (v1, v0) to N1 to produce <a href="#BM_(tree)%2Fi">tree</a>...</p>
<p><img src="dot_9a3e57b6b197a70fe4edf94c50972151.svg" alt="Dot diagram"></p>
</li>
</ul>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>See <a href="#BM_(Algorithms%5C%2FPhylogeny%2FDistance%20Matrix%20to%20Tree%2FNeighbour%20Joining%20Phylogeny%20Algorithm)_TOPIC%2F">Algorithms/Phylogeny/Distance Matrix to Tree/Neighbour Joining Phylogeny Algorithm</a> for a full explanation of how this algorithm works.</p>
</div>
</li>
<li>
<p><a name="BM_(paleontology%7Cpalaeontology%7Cpal%C3%A6ontology)%2Fi"></a><strong>paleontology</strong> - The scientific study of ancient organisms: dinosaurs, prehistoric plants, prehistoric insects, prehistoric fungi, etc...</p>
</li>
<li>
<p><a name="BM_(anatomy%7Canatomical)%2Fi"></a><strong>anatomy</strong> - The study of the identification and description of structures in organisms.</p>
</li>
<li>
<p><a name="BM_(physiology%7Cphysiological)%2Fi"></a><strong>physiology</strong> - The study of organism function.</p>
</li>
<li>
<p><a name="BM_(character%20table%7Ccharacter%20vector)%2Fi"></a><strong>character table</strong> - A matrix where the columns represent biological entities and the rows represent characteristics of those entities, where those characteristics are typically <a href="#BM_(anatomy%7Canatomical)%2Fi">anatomical</a>ly or <a href="#BM_(physiology%7Cphysiological)%2Fi">physiological</a>ly.</p>
<table>
<thead>
<tr>
<th></th>
<th>wings</th>
<th>sucks blood</th>
<th>number of legs</th>
</tr>
</thead>
<tbody>
<tr>
<td>house fly</td>
<td>2</td>
<td>no</td>
<td>6</td>
</tr>
<tr>
<td>mosquito</td>
<td>2</td>
<td>yes</td>
<td>6</td>
</tr>
<tr>
<td>snail</td>
<td>0</td>
<td>no</td>
<td>0</td>
</tr>
</tbody>
</table>
<p><a href="#BM_(character%20table%7Ccharacter%20vector)%2Fi">Character table</a>s were commonly used for <a href="#BM_(phylogeny%7Cphylogenetic)%2Fi">phylogeny</a> before discovering that DNA can be used to compare the relatedness of organisms.</p>
<p>A row in a <a href="#BM_(character%20table%7Ccharacter%20vector)%2Fi">character table</a> is referred to as a <a href="#BM_(character%20table%7Ccharacter%20vector)%2Fi">character vector</a>. Prior to the advent of <a href="#BM_(sequencer)%2Fi">sequencer</a>s, scientists would treat <a href="#BM_(character%20table%7Ccharacter%20vector)%2Fi">character vector</a>s as <a href="#BM_(sequence)%2Fi">sequence</a>s for generating <a href="#BM_(phylogenetic%20tree%7Cphylogeny%20tree%7Cphylogenies%7Cevolutionary%20tree)%2Fi">phylogenetic tree</a>s or doing comparisons between organisms.</p>
</li>
<li>
<p><a name="BM_(mitochondrial%20DNA%7CmtDNA)%2Fi"></a><strong>mitochondrial DNA</strong> - DNA unique to the mitochondria. This DNA is unique to the mitochondria, different from the DNA of the cell that the mitochondria lives in. The mitochondria is suspected of being bacteria that made it into the cell and survived, forming a symbiotic relationship.</p>
<p><a href="#BM_(mitochondrial%20DNA%7CmtDNA)%2Fi">Mitochondrial DNA</a> is inherited fully from the mother. It isn't a mix of parental DNA as the cell DNA is.</p>
</li>
<li>
<p><a name="BM_(small%20parsimony%7Clarge%20parsimony)%2Fi"></a><strong>small parsimony</strong> - In the context of <a href="#BM_(phylogenetic%20tree%7Cphylogeny%20tree%7Cphylogenies%7Cevolutionary%20tree)%2Fi">phylogenetic tree</a>s, ...</p>
<ul>
<li><a href="#BM_(small%20parsimony%7Clarge%20parsimony)%2Fi">small parsimony</a>: When a <a href="#BM_(tree)%2Fi">tree</a> structure and its leaf node <a href="#BM_(sequence)%2Fi">sequence</a>s are given, derive the internal node <a href="#BM_(sequence)%2Fi">sequence</a>s with the lowest possible distance (most <a href="#BM_(parsimony%7Cparsimonious)%2Fi">parsimonious</a>).</li>
<li><a href="#BM_(small%20parsimony%7Clarge%20parsimony)%2Fi">large parsimony</a>: When only the leaf node <a href="#BM_(sequence)%2Fi">sequence</a>s are given, derive the combination of <a href="#BM_(tree)%2Fi">tree</a> structure and internal node <a href="#BM_(sequence)%2Fi">sequence</a>s with the lowest possible distance (most <a href="#BM_(parsimony%7Cparsimonious)%2Fi">parsimonious</a>).</li>
</ul>
<p><img src="svgbob_4869b4c66c4290aef524561ee17190d9cff0640c.svg" alt="Kroki diagram output"></p>
<p><a href="#BM_(small%20parsimony%7Clarge%20parsimony)%2Fi">Large parsimony</a> isn't a process that's normally done because the search space explodes in size (e.g. NP-complete). Instead, <a href="#BM_(small%20parsimony%7Clarge%20parsimony)%2Fi">small parsimony</a> is used on a <a href="#BM_(tree)%2Fi">tree</a> generated using an algorithm like <a href="#BM_(unweighted%20pair%20group%20method%20with%20arithmetic%20mean%7CUPGMA)%2Fi">UPGMA</a> or <a href="#BM_(neighbour%20joining%20phylogeny)%2Fi">neighbour joining phylogeny</a>.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The <a href="#BM_(parsimony%20score)%2Fi">parsimony score</a> algorithm is what's typically used to evaluate how well a combination of <a href="#BM_(tree)%2Fi">tree</a> structure + ancestral <a href="#BM_(sequence)%2Fi">sequence</a>s do.</p>
</div>
</li>
<li>
<p><a name="BM_(parsimony%20score)%2Fi"></a><strong>parsimony score</strong> - Given a <a href="#BM_(phylogenetic%20tree%7Cphylogeny%20tree%7Cphylogenies%7Cevolutionary%20tree)%2Fi">phylogenetic tree</a> with <a href="#BM_(sequence)%2Fi">sequence</a>s for both leaf nodes (known entities) and internal nodes (inferred ancestral entities), the <a href="#BM_(parsimony%20score)%2Fi">parsimony score</a> is a measure of how far off a parent's <a href="#BM_(sequence)%2Fi">sequence</a> is from its children (and vice versa). The idea is that the most <a href="#BM_(parsimony%7Cparsimonious)%2Fi">parsimonious</a> evolutionary path is the one that's most likely to have occurred. As such, the less far off <a href="#BM_(sequence)%2Fi">sequence</a>s are, the more likely it is that the actual ancestral lineage and ancestral <a href="#BM_(sequence)%2Fi">sequence</a>s match what's depicted in the <a href="#BM_(tree)%2Fi">tree</a>.</p>
<p><img src="svgbob_68590e7d4bd005a4f80364ba3967a7124b6b6527.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a name="BM_(Saccharomyces)%2Fi"></a><strong>Saccharomyces</strong> - A genus of yeast used for producing alcohol and bread.</p>
</li>
<li>
<p><a name="BM_(diauxic%20shift)%2Fi"></a><strong>diauxic shift</strong> - A change in the metabolism of <a href="#BM_(Saccharomyces)%2Fi">Saccharomyces</a> cerevisiae. When glucose is present, <a href="#BM_(Saccharomyces)%2Fi">Saccharomyces</a> cerevisiae consumes that glucose for energy and produces ethanol as a byproduct. Once all available glucose in the environment has been depleted, it inverts its metabolism to instead consume the ethanol byproduct it produced earlier.</p>
<p>The consumption of ethanol only happens in the presence of oxygen. Without oxygen, <a href="#BM_(Saccharomyces)%2Fi">Saccharomyces</a> cerevisiae enters hibernation until either glucose or oxygen become available. This property is what allows for the making of wine: Wine production typically involves sealed barrels that prevent oxygen from entering.</p>
</li>
<li>
<p><a name="BM_(whole%20genome%20duplication)%2Fi"></a><strong>whole genome duplication</strong> <a name="BM_%5Cb(WGD)s%3F%5Cb%2F"></a><strong></strong> - A rare evolutionary event in which the entire <a href="#BM_(genome)%2Fi">genome</a> of an organism is duplicated.</p>
<p>After duplication, much of the functionality provided by the <a href="#BM_(genome)%2Fi">genome</a> becomes redundant. The organism evolves much more rapidly because a mutation in one copy of a <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a> won't necessarily make the organism less fit (duplicate copy still exists). It's typical for a <a href="#BM_(whole%20genome%20duplication)%2Fi">whole genome duplication</a> to be quickly followed up by massive amounts of <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a> mutations, <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a> loss, and <a href="#BM_(genome%20rearrangement%7Cchromosomal%20rearrangement%7Cchromosome%20rearrangement)%2Fi">genome rearrangement</a>s.</p>
</li>
<li>
<p><a name="BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi"></a><strong>gene expression matrix</strong> - A matrix where each column represents a point in time, each row represents a <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a>, and each cell is a number representing the amount of <a href="#BM_(gene%20expression)%2Fi">gene expression</a> taking place for that <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">gene</a> (row) at that time (column).</p>
<table>
<thead>
<tr>
<th></th>
<th>5 AM</th>
<th>6 AM</th>
<th>7 AM</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> 1</td>
<td>1.0</td>
<td>1.0</td>
<td>1.0</td>
</tr>
<tr>
<td><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> 2</td>
<td>1.0</td>
<td>0.7</td>
<td>0.5</td>
</tr>
<tr>
<td><a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">Gene</a> 3</td>
<td>1.0</td>
<td>1.1</td>
<td>1.4</td>
</tr>
</tbody>
</table>
<p>Each row in a <a href="#BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi">gene expression matrix</a> is called a <a href="#BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi">gene expression vector</a>.</p>
</li>
<li>
<p><a name="BM_(co-regulate)%2Fi"></a><strong>co-regulate</strong> - A set of <a href="#BM_(%5Cbgenes%5Cb%7C%5Cbgene%5Cb)%2Fi">genes</a> whose <a href="#BM_(gene%20expression)%2Fi">gene expression</a> is regulated by the same <a href="#BM_(transcription%20factor)%2Fi">transcription factor</a>.</p>
</li>
<li>
<p><a name="BM_(RNA%20transcript%7Ctranscript)%2Fi"></a><strong>RNA transcript</strong> - RNA output of <a href="#BM_(transcription%7Ctranscribed%7Ctranscribe)%2Fi">transcription</a>.</p>
</li>
<li>
<p><a name="BM_(transcriptome)%2Fi"></a><strong>transcriptome</strong> - All <a href="#BM_(RNA%20transcript%7Ctranscript)%2Fi">RNA transcript</a>s within a cell at a specific point in time.</p>
</li>
<li>
<p><a name="BM_(good%20clustering%20principle)%2Fi"></a><strong>good clustering principle</strong> - The idea that items within the same <a href="#BM_(cluster)%2Fi">cluster</a> should be more similar to each other than items in other <a href="#BM_(cluster)%2Fi">cluster</a>s.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>This was originally introduced as "every pair of points within the same <a href="#BM_(cluster)%2Fi">cluster</a> should be closer to each other than any points each from different <a href="#BM_(cluster)%2Fi">cluster</a>s", where "closer" was implying <a href="#BM_(euclidean%20distance)%2Fi">euclidean distance</a>. I think the idea was intended to be abstracted out to the definition above since the items may not be thought of as "points" but as vectors or <a href="#BM_(sequence)%2Fi">sequence</a>s + you can choose a <a href="#BM_(similarity%20metric)%2Fi">similarity metric</a> other than <a href="#BM_(euclidean%20distance)%2Fi">euclidean distance</a>.</p>
</div>
</li>
<li>
<p><a name="BM_(euclidean%20distance)%2Fi"></a><strong>euclidean distance</strong> - The distance between two points if traveling directly from one to the other in a straight line.</p>
<p><img src="svgbob_812b661882246e5c5673044d47b2a743122f1f75.svg" alt="Kroki diagram output"></p>
<p>In 2 dimensional space, this is calculated as <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msqrt><mrow><mo stretchy="false">(</mo><msub><mi>x</mi><mn>2</mn></msub><mo>−</mo><msub><mi>x</mi><mn>1</mn></msub><msup><mo stretchy="false">)</mo><mn>2</mn></msup><mo>+</mo><mo stretchy="false">(</mo><msub><mi>y</mi><mn>2</mn></msub><mo>−</mo><msub><mi>y</mi><mn>1</mn></msub><msup><mo stretchy="false">)</mo><mn>2</mn></msup></mrow></msqrt></mrow><annotation encoding="application/x-tex">\sqrt{(x_2 - x_1)^2 + (y_2 - y_1)^2}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.24em;vertical-align:-0.30499999999999994em;"></span><span class="mord sqrt"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.935em;"><span class="svg-align" style="top:-3.2em;"><span class="pstrut" style="height:3.2em;"></span><span class="mord" style="padding-left:1em;"><span class="mopen">(</span><span class="mord"><span class="mord mathnormal">x</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal">x</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mclose"><span class="mclose">)</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.740108em;"><span style="top:-2.9890000000000003em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.03588em;">y</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.03588em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.03588em;">y</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.03588em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mclose"><span class="mclose">)</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.740108em;"><span style="top:-2.9890000000000003em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span></span></span><span style="top:-2.8950000000000005em;"><span class="pstrut" style="height:3.2em;"></span><span class="hide-tail" style="min-width:1.02em;height:1.28em;"><svg width="400em" height="1.28em" viewBox="0 0 400000 1296" preserveAspectRatio="xMinYMin slice"><path d="M263,681c0.7,0,18,39.7,52,119
c34,79.3,68.167,158.7,102.5,238c34.3,79.3,51.8,119.3,52.5,120
c340,-704.7,510.7,-1060.3,512,-1067
l0 -0
c4.7,-7.3,11,-11,19,-11
H40000v40H1012.3
s-271.3,567,-271.3,567c-38.7,80.7,-84,175,-136,283c-52,108,-89.167,185.3,-111.5,232
c-22.3,46.7,-33.8,70.3,-34.5,71c-4.7,4.7,-12.3,7,-23,7s-12,-1,-12,-1
s-109,-253,-109,-253c-72.7,-168,-109.3,-252,-110,-252c-10.7,8,-22,16.7,-34,26
c-22,17.3,-33.3,26,-34,26s-26,-26,-26,-26s76,-59,76,-59s76,-60,76,-60z
M1001 80h400000v40h-400000z"></path></svg></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.30499999999999994em;"><span></span></span></span></span></span></span></span></span>.</p>
<p>In 3 dimensional space, this is calculated as <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msqrt><mrow><mo stretchy="false">(</mo><msub><mi>x</mi><mn>2</mn></msub><mo>−</mo><msub><mi>x</mi><mn>1</mn></msub><msup><mo stretchy="false">)</mo><mn>2</mn></msup><mo>+</mo><mo stretchy="false">(</mo><msub><mi>y</mi><mn>2</mn></msub><mo>−</mo><msub><mi>y</mi><mn>1</mn></msub><msup><mo stretchy="false">)</mo><mn>2</mn></msup><mo>+</mo><mo stretchy="false">(</mo><msub><mi>z</mi><mn>2</mn></msub><mo>−</mo><msub><mi>z</mi><mn>1</mn></msub><msup><mo stretchy="false">)</mo><mn>2</mn></msup></mrow></msqrt></mrow><annotation encoding="application/x-tex">\sqrt{(x_2 - x_1)^2 + (y_2 - y_1)^2 + (z_2 - z_1)^2}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.24em;vertical-align:-0.30499999999999994em;"></span><span class="mord sqrt"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.935em;"><span class="svg-align" style="top:-3.2em;"><span class="pstrut" style="height:3.2em;"></span><span class="mord" style="padding-left:1em;"><span class="mopen">(</span><span class="mord"><span class="mord mathnormal">x</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal">x</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mclose"><span class="mclose">)</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.740108em;"><span style="top:-2.9890000000000003em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.03588em;">y</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.03588em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.03588em;">y</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.03588em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mclose"><span class="mclose">)</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.740108em;"><span style="top:-2.9890000000000003em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.04398em;">z</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.04398em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.04398em;">z</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.04398em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mclose"><span class="mclose">)</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.740108em;"><span style="top:-2.9890000000000003em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span></span></span><span style="top:-2.8950000000000005em;"><span class="pstrut" style="height:3.2em;"></span><span class="hide-tail" style="min-width:1.02em;height:1.28em;"><svg width="400em" height="1.28em" viewBox="0 0 400000 1296" preserveAspectRatio="xMinYMin slice"><path d="M263,681c0.7,0,18,39.7,52,119
c34,79.3,68.167,158.7,102.5,238c34.3,79.3,51.8,119.3,52.5,120
c340,-704.7,510.7,-1060.3,512,-1067
l0 -0
c4.7,-7.3,11,-11,19,-11
H40000v40H1012.3
s-271.3,567,-271.3,567c-38.7,80.7,-84,175,-136,283c-52,108,-89.167,185.3,-111.5,232
c-22.3,46.7,-33.8,70.3,-34.5,71c-4.7,4.7,-12.3,7,-23,7s-12,-1,-12,-1
s-109,-253,-109,-253c-72.7,-168,-109.3,-252,-110,-252c-10.7,8,-22,16.7,-34,26
c-22,17.3,-33.3,26,-34,26s-26,-26,-26,-26s76,-59,76,-59s76,-60,76,-60z
M1001 80h400000v40h-400000z"></path></svg></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.30499999999999994em;"><span></span></span></span></span></span></span></span></span>.</p>
<p>In n dimensional space, this is calculated as <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msqrt><mrow><msubsup><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></msubsup><mrow><mo stretchy="false">(</mo><msub><mi>v</mi><mi>i</mi></msub><mo>−</mo><msub><mi>w</mi><mi>i</mi></msub><msup><mo stretchy="false">)</mo><mn>2</mn></msup></mrow></mrow></msqrt></mrow><annotation encoding="application/x-tex">\sqrt{\sum_{i=1}^n{(v_i - w_i)^2}}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.24em;vertical-align:-0.302709em;"></span><span class="mord sqrt"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.937291em;"><span class="svg-align" style="top:-3.2em;"><span class="pstrut" style="height:3.2em;"></span><span class="mord" style="padding-left:1em;"><span class="mop"><span class="mop op-symbol small-op" style="position:relative;top:-0.0000050000000000050004em;">∑</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.804292em;"><span style="top:-2.40029em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">i</span><span class="mrel mtight">=</span><span class="mord mtight">1</span></span></span></span><span style="top:-3.2029em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">n</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.29971000000000003em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord"><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.03588em;">v</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:-0.03588em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02691em;">w</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:-0.02691em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mclose"><span class="mclose">)</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.740108em;"><span style="top:-2.9890000000000003em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span></span></span></span><span style="top:-2.897291em;"><span class="pstrut" style="height:3.2em;"></span><span class="hide-tail" style="min-width:1.02em;height:1.28em;"><svg width="400em" height="1.28em" viewBox="0 0 400000 1296" preserveAspectRatio="xMinYMin slice"><path d="M263,681c0.7,0,18,39.7,52,119
c34,79.3,68.167,158.7,102.5,238c34.3,79.3,51.8,119.3,52.5,120
c340,-704.7,510.7,-1060.3,512,-1067
l0 -0
c4.7,-7.3,11,-11,19,-11
H40000v40H1012.3
s-271.3,567,-271.3,567c-38.7,80.7,-84,175,-136,283c-52,108,-89.167,185.3,-111.5,232
c-22.3,46.7,-33.8,70.3,-34.5,71c-4.7,4.7,-12.3,7,-23,7s-12,-1,-12,-1
s-109,-253,-109,-253c-72.7,-168,-109.3,-252,-110,-252c-10.7,8,-22,16.7,-34,26
c-22,17.3,-33.3,26,-34,26s-26,-26,-26,-26s76,-59,76,-59s76,-60,76,-60z
M1001 80h400000v40h-400000z"></path></svg></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.302709em;"><span></span></span></span></span></span></span></span></span>, where v and w are two n dimensional points.</p>
</li>
<li>
<p><a name="BM_(manhattan%20distance%7Ctaxi%5B%5Cs%5C-%5D%3Fcab%20distance%7Ccity%5B%5Cs%5C-%5Dblock%20distance)%2Fi"></a><strong>manhattan distance</strong> - The distance between two points if traveling only via the axis of the coordinate system.</p>
<p><img src="svgbob_19c8269cf6dc61112d64c53c1045bfe1e1f518e3.svg" alt="Kroki diagram output"></p>
<p>In n dimensional space, this is calculated as <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msubsup><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></msubsup><mrow><mi mathvariant="normal">∣</mi><msub><mi>v</mi><mi>i</mi></msub><mo>−</mo><msub><mi>w</mi><mi>i</mi></msub><mi mathvariant="normal">∣</mi></mrow></mrow><annotation encoding="application/x-tex">\sum_{i=1}^n{|v_i - w_i|}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.104002em;vertical-align:-0.29971000000000003em;"></span><span class="mop"><span class="mop op-symbol small-op" style="position:relative;top:-0.0000050000000000050004em;">∑</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.804292em;"><span style="top:-2.40029em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">i</span><span class="mrel mtight">=</span><span class="mord mtight">1</span></span></span></span><span style="top:-3.2029em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">n</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.29971000000000003em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord"><span class="mord">∣</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.03588em;">v</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:-0.03588em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.02691em;">w</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:-0.02691em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mord">∣</span></span></span></span></span>, where v and w are two n dimensional points.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The name <a href="#BM_(manhattan%20distance%7Ctaxi%5B%5Cs%5C-%5D%3Fcab%20distance%7Ccity%5B%5Cs%5C-%5Dblock%20distance)%2Fi">manhattan distance</a> is an allegory to the city blocks of manhattan, where your options (most of the time) are to move either left/right or up/down. Other names for this same metric are <a href="#BM_(manhattan%20distance%7Ctaxi%5B%5Cs%5C-%5D%3Fcab%20distance%7Ccity%5B%5Cs%5C-%5Dblock%20distance)%2Fi">taxi-cab distance</a> and <a href="#BM_(manhattan%20distance%7Ctaxi%5B%5Cs%5C-%5D%3Fcab%20distance%7Ccity%5B%5Cs%5C-%5Dblock%20distance)%2Fi">city block distance</a>.</p>
</div>
</li>
<li>
<p><a name="BM_(cluster)%2Fi"></a><strong>clustering</strong> - Grouping together a set of objects such that objects within the same group are more similar to each other than objects in other groups. Each group is referred to as a <a href="#BM_(cluster)%2Fi">cluster</a>.</p>
</li>
<li>
<p><a name="BM_(k-centers%3F%20clustering%7C%5Cd%2B-centers%3F%20clustering%7Ck-centers%3F%7C%5Cd%2B-centers%3F)%2Fi"></a><strong>k-centers clustering</strong> - A form of <a href="#BM_(cluster)%2Fi">cluster</a>ing where a point, called a center, defines <a href="#BM_(cluster)%2Fi">cluster</a> membership. k different centers are chosen (one for each <a href="#BM_(cluster)%2Fi">cluster</a>), and the points closest to each center (<a href="#BM_(euclidean%20distance)%2Fi">euclidean distance</a>) make up members of that <a href="#BM_(cluster)%2Fi">cluster</a>. The goal is to choose centers such that, out of all possible <a href="#BM_(cluster)%2Fi">cluster</a> center to member distances, the farthest distance is the minimum it could possibly be out of all possible choices for centers.</p>
<p>In terms of a scoring function, the score being minimized is ...</p>
<div style="overflow: auto"><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><mi>s</mi><mi>c</mi><mi>o</mi><mi>r</mi><mi>e</mi><mo>=</mo><mi>m</mi><mi>a</mi><mi>x</mi><mo stretchy="false">(</mo><mi>d</mi><mo stretchy="false">(</mo><msub><mi>P</mi><mn>1</mn></msub><mo separator="true">,</mo><mi>C</mi><mo stretchy="false">)</mo><mo separator="true">,</mo><mi>d</mi><mo stretchy="false">(</mo><msub><mi>P</mi><mn>2</mn></msub><mo separator="true">,</mo><mi>C</mi><mo stretchy="false">)</mo><mo separator="true">,</mo><mi mathvariant="normal">.</mi><mi mathvariant="normal">.</mi><mi mathvariant="normal">.</mi><mo separator="true">,</mo><mi>d</mi><mo stretchy="false">(</mo><msub><mi>P</mi><mi>n</mi></msub><mo separator="true">,</mo><mi>C</mi><mo stretchy="false">)</mo><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">score = max(d(P_1, C), d(P_2, C), ..., d(P_n, C))
</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.43056em;vertical-align:0em;"></span><span class="mord mathnormal">s</span><span class="mord mathnormal">c</span><span class="mord mathnormal">o</span><span class="mord mathnormal" style="margin-right:0.02778em;">r</span><span class="mord mathnormal">e</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal">m</span><span class="mord mathnormal">a</span><span class="mord mathnormal">x</span><span class="mopen">(</span><span class="mord mathnormal">d</span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.13889em;">P</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.13889em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mpunct">,</span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord mathnormal" style="margin-right:0.07153em;">C</span><span class="mclose">)</span><span class="mpunct">,</span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord mathnormal">d</span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.13889em;">P</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.13889em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mpunct">,</span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord mathnormal" style="margin-right:0.07153em;">C</span><span class="mclose">)</span><span class="mpunct">,</span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord">.</span><span class="mord">.</span><span class="mord">.</span><span class="mpunct">,</span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord mathnormal">d</span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.13889em;">P</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.151392em;"><span style="top:-2.5500000000000003em;margin-left:-0.13889em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">n</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mpunct">,</span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord mathnormal" style="margin-right:0.07153em;">C</span><span class="mclose">)</span><span class="mclose">)</span></span></span></span></span></div><ul>
<li>P is the set of points.</li>
<li>C is the set of centers.</li>
<li>n is the number of points in P</li>
<li>d returns the <a href="#BM_(euclidean%20distance)%2Fi">euclidean distance</a> between a point and its closest center.</li>
<li>max is the maximum function.</li>
</ul>
<pre class="hljs"><code><span class="hljs-comment"># d() function from the formula</span>
<span class="hljs-keyword">def</span> <span class="hljs-title function_">dist_to_closest_center</span>(<span class="hljs-params">data_pt, center_pts</span>):
  center_pt = <span class="hljs-built_in">min</span>(
    center_pts,
    key=<span class="hljs-keyword">lambda</span> cp: dist(data_pt, cp)
  )
  <span class="hljs-keyword">return</span> dist(center_pt, data_pt)

<span class="hljs-comment"># scoring function (what's trying to be minimized)</span>
<span class="hljs-keyword">def</span> <span class="hljs-title function_">k_centers_score</span>(<span class="hljs-params">data_pts, center_pts</span>):
  <span class="hljs-keyword">return</span> <span class="hljs-built_in">max</span>(dist_to_closest_center(p, center_pts) <span class="hljs-keyword">for</span> p <span class="hljs-keyword">in</span> data_pts)
</code></pre>
<p><img src="svgbob_5fa140a687d17ef8490f92bae2a74d597d2d56f9.svg" alt="Kroki diagram output"></p>
<p>For a non-trivial input, the search space is too massive for a straight-forward algorithm to work. As such, heuristics are commonly used instead.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>See <a href="#BM_(farthest%20first%20traversal)%2Fi">farthest first traversal</a> heuristic.</p>
</div>
</li>
<li>
<p><a name="BM_(farthest%20first%20traversal)%2Fi"></a><strong>farthest first traversal</strong> - A heuristic commonly used for <a href="#BM_(k-centers%3F%20clustering%7C%5Cd%2B-centers%3F%20clustering%7Ck-centers%3F%7C%5Cd%2B-centers%3F)%2Fi">k-centers clustering</a>. The algorithm iteratively builds out more centers by inspecting the <a href="#BM_(euclidean%20distance)%2Fi">euclidean distance</a>s from points to existing centers. At each step, the algorithm ...</p>
<ol>
<li>gets the closest center for each point,</li>
<li>picks the point with the farthest <a href="#BM_(euclidean%20distance)%2Fi">euclidean distance</a> and sets it as the new center.</li>
</ol>
<p>The algorithm initially primes the list of centers with a randomly chosen point and stops executing once it has k points.</p>
<p><img src="svgbob_52c67bd6a373b1ba175fc4205574ee9cfc99b6fb.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a name="BM_(k-means%3F%7C%5Cd%2B-means%3F%7Csquared%20error%20distortion)%2Fi"></a><strong>k-means clustering</strong> - A form of <a href="#BM_(cluster)%2Fi">cluster</a>ing where a point, called a center, defines <a href="#BM_(cluster)%2Fi">cluster</a> membership. k different centers are chosen (one for each <a href="#BM_(cluster)%2Fi">cluster</a>), and the points closest to each center (<a href="#BM_(euclidean%20distance)%2Fi">euclidean distance</a>) make up members of that <a href="#BM_(cluster)%2Fi">cluster</a>. The goal is to choose centers such that, out of all possible <a href="#BM_(cluster)%2Fi">cluster</a> center to member distances, the formula below is the minimum it could possibly be out of all possible choices for centers.</p>
<p>The formula below is referred to as <a href="#BM_(k-means%3F%7C%5Cd%2B-means%3F%7Csquared%20error%20distortion)%2Fi">squared error distortion</a>.</p>
<div style="overflow: auto"><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><mi>s</mi><mi>c</mi><mi>o</mi><mi>r</mi><mi>e</mi><mo>=</mo><mfrac><mrow><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><mrow><mi>d</mi><mo stretchy="false">(</mo><msub><mi>P</mi><mi>i</mi></msub><mo separator="true">,</mo><mi>C</mi><msup><mo stretchy="false">)</mo><mn>2</mn></msup></mrow></mrow><mi>n</mi></mfrac></mrow><annotation encoding="application/x-tex">score = \frac{\sum_{i=1}^{n} {d(P_i, C)^2}}{n}
</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.43056em;vertical-align:0em;"></span><span class="mord mathnormal">s</span><span class="mord mathnormal">c</span><span class="mord mathnormal">o</span><span class="mord mathnormal" style="margin-right:0.02778em;">r</span><span class="mord mathnormal">e</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:2.189818em;vertical-align:-0.686em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.5038179999999999em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord mathnormal">n</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.6897100000000003em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mop"><span class="mop op-symbol small-op" style="position:relative;top:-0.0000050000000000050004em;">∑</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.804292em;"><span style="top:-2.40029em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">i</span><span class="mrel mtight">=</span><span class="mord mtight">1</span></span></span></span><span style="top:-3.2029em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">n</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.29971000000000003em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord"><span class="mord mathnormal">d</span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.13889em;">P</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:-0.13889em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mpunct">,</span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord mathnormal" style="margin-right:0.07153em;">C</span><span class="mclose"><span class="mclose">)</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8141079999999999em;"><span style="top:-3.063em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span></span></div><ul>
<li>P is the set of points.</li>
<li>C is the set of centers.</li>
<li>n is the number of points in P</li>
<li>d returns the <a href="#BM_(euclidean%20distance)%2Fi">euclidean distance</a> between a point and its closest center.</li>
<li>max is the maximum function.</li>
</ul>
<pre class="hljs"><code><span class="hljs-comment"># d() function from the formula</span>
<span class="hljs-keyword">def</span> <span class="hljs-title function_">find_closest_center</span>(<span class="hljs-params">data_pt, center_pts</span>):
  center_pt = <span class="hljs-built_in">min</span>(
    center_pts,
    key=<span class="hljs-keyword">lambda</span> cp: dist(data_pt, cp)
  )
  <span class="hljs-keyword">return</span> center_pt, dist(center_pt, data_pt)

<span class="hljs-comment"># scoring function (what's trying to be minimized) -- taking squares of d() and averaging</span>
<span class="hljs-keyword">def</span> <span class="hljs-title function_">squared_error_distortion</span>(<span class="hljs-params">data_pts, center_pts</span>):
  res = []
  <span class="hljs-keyword">for</span> data_pt <span class="hljs-keyword">in</span> data_pts:
    closest_center_pt, dist_to = find_closest_center(data_pt, center_pts)
    res.append(dist_to ** <span class="hljs-number">2</span>)
  <span class="hljs-keyword">return</span> <span class="hljs-built_in">sum</span>(res) / <span class="hljs-built_in">len</span>(res)
</code></pre>
<p>Compared to <a href="#BM_(k-centers%3F%20clustering%7C%5Cd%2B-centers%3F%20clustering%7Ck-centers%3F%7C%5Cd%2B-centers%3F)%2Fi">k-centers</a>, <a href="#BM_(k-means%3F%7C%5Cd%2B-means%3F%7Csquared%20error%20distortion)%2Fi">k-means</a> more appropriately handles outliers. If outliers are present, <a href="#BM_(k-centers%3F%20clustering%7C%5Cd%2B-centers%3F%20clustering%7Ck-centers%3F%7C%5Cd%2B-centers%3F)%2Fi">k-centers</a>'s metric causes the <a href="#BM_(cluster)%2Fi">cluster</a> center to be wildly offset while <a href="#BM_(k-means%3F%7C%5Cd%2B-means%3F%7Csquared%20error%20distortion)%2Fi">k-mean</a>'s metric will only be mildly offset. In the example below, the best scoring <a href="#BM_(cluster)%2Fi">cluster</a> center for <a href="#BM_(k-centers%3F%20clustering%7C%5Cd%2B-centers%3F%20clustering%7Ck-centers%3F%7C%5Cd%2B-centers%3F)%2Fi">k-centers</a> is wildly offset by outlier Z while <a href="#BM_(k-means%3F%7C%5Cd%2B-means%3F%7Csquared%20error%20distortion)%2Fi">k-means</a> isn't offset as much.</p>
<p><img src="svgbob_50f9ef0bee677ed328acdaec5260e8530cf4b1be.svg" alt="Kroki diagram output"></p>
<p>For a non-trivial input, the search space is too massive for a straight-forward algorithm to work. As such, heuristics are commonly used instead.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>See <a href="#BM_(lloyd%20algorithm%7Clloyd's%20algorithm%7Ck-means%3F%5C%2B%5C%2B%20initializer%7Ck-means%3F%5C%2B%5C%2B)%2Fi">Lloyd's algorithm</a>.</p>
</div>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>The Pevzner book states that <a href="#BM_(k-means%3F%7C%5Cd%2B-means%3F%7Csquared%20error%20distortion)%2Fi">k-means</a> <a href="#BM_(cluster)%2Fi">cluster</a>ing often doesn't produce good results if the clumps have a non-globular shape (e.g. elongated) or have widely different densities. This probably applies to <a href="#BM_(k-centers%3F%20clustering%7C%5Cd%2B-centers%3F%20clustering%7Ck-centers%3F%7C%5Cd%2B-centers%3F)%2Fi">k-centers</a> as well</p>
</div>
</li>
<li>
<p><a name="BM_(center%20of%20gravity)%2Fi"></a><strong>center of gravity</strong> - In terms of <a href="#BM_(cluster)%2Fi">cluster</a>ing using <a href="#BM_(euclidean%20distance)%2Fi">euclidean distance</a>, the <a href="#BM_(center%20of%20gravity)%2Fi">center of gravity</a> for a set of points is the average of each individual dimension / coordinates.</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">center_of_gravity</span>(<span class="hljs-params">data_pts, dim</span>):
   center = []
   <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> <span class="hljs-built_in">range</span>(dim):
     val = mean(pt[i] <span class="hljs-keyword">for</span> pt <span class="hljs-keyword">in</span> data_pts)
     center.append(val)
   <span class="hljs-keyword">return</span> center
</code></pre>
<p>For example, for the 2D points (0,5), (1,3), and (2,2), the <a href="#BM_(center%20of%20gravity)%2Fi">center of gravity</a> is (1, 3.3333). The coordinate / dimension for ...</p>
<ul>
<li>x is calculated as <code>0+1+2/3 = 1</code>.</li>
<li>y is calculated as <code>5+3+2/3 = 3.3333</code>.</li>
</ul>
</li>
<li>
<p><a name="BM_(lloyd%20algorithm%7Clloyd's%20algorithm%7Ck-means%3F%5C%2B%5C%2B%20initializer%7Ck-means%3F%5C%2B%5C%2B)%2Fi"></a><strong>Lloyd's algorithm</strong> - A heuristic for determining centers in <a href="#BM_(k-means%3F%7C%5Cd%2B-means%3F%7Csquared%20error%20distortion)%2Fi">k-means</a> <a href="#BM_(cluster)%2Fi">cluster</a>ing. The algorithm begins by choosing centers by arbitrarily selecting k points from the points being <a href="#BM_(cluster)%2Fi">cluster</a>ed, then iteratively...</p>
<ol>
<li>assigns each point to its nearest center (ties broken arbitrarily),</li>
<li>calculates a new set of centers by calculating the <a href="#BM_(center%20of%20gravity)%2Fi">center of gravity</a> from the points in each.</li>
</ol>
<p>The algorithm stops once the new centers are the same as the centers from the previous iteration (stops converging).</p>
<p>Since this algorithm is a heuristic, it doesn't always converge to a good solution. The algorithm typically runs multiple times, where the run producing centers with the lowest <a href="#BM_(k-means%3F%7C%5Cd%2B-means%3F%7Csquared%20error%20distortion)%2Fi">squared error distortion</a> is accepted. An enhancement to the algorithm, called <a href="#BM_(lloyd%20algorithm%7Clloyd's%20algorithm%7Ck-means%3F%5C%2B%5C%2B%20initializer%7Ck-means%3F%5C%2B%5C%2B)%2Fi">k-means++ initializer</a>, increases the chances of converging to a good solution by <em>probabilistically</em> selecting initial centers that are far from each other:</p>
<ol>
<li>The 1st center is chosen from the list of points at random.</li>
<li>The 2nd center is chosen by selecting a point that's likely to be much farther away from center 1 than most other points.</li>
<li>The 3rd center is chosen by selecting a point that's likely to be much farther away from center 1 and 2 than most other points.</li>
<li>...</li>
</ol>
<p>The probability of selecting a point as the next center is proportional to its squared distance to the existing centers.</p>
<pre class="hljs"><code><span class="hljs-keyword">def</span> <span class="hljs-title function_">k_means_PP_initializer</span>(<span class="hljs-params">data_pts, k</span>):
  centers = [random.choice(data_pts)]
  <span class="hljs-keyword">while</span> <span class="hljs-built_in">len</span>(centers) &lt; k:
    choice_points = []
    choice_weights = []
    <span class="hljs-keyword">for</span> pt <span class="hljs-keyword">in</span> data_pts:
      <span class="hljs-keyword">if</span> pt <span class="hljs-keyword">in</span> centers:
        <span class="hljs-keyword">continue</span>
      _, d = find_closest_center(pt, centers)
      choice_weights.append(d)
      choice_points.append(pt)
    total = <span class="hljs-built_in">sum</span>(choice_weights)
    choice_weights = [w / total <span class="hljs-keyword">for</span> w <span class="hljs-keyword">in</span> choice_weights]
    c_pt = random.choices(choice_points, weights=choice_weights, k=<span class="hljs-number">1</span>).pop(<span class="hljs-number">0</span>)
    centers.append(c_pt)
  <span class="hljs-keyword">return</span> centers
</code></pre>
</li>
<li>
<p><a name="BM_(hard%20clustering%7Csoft%20clustering)%2Fi"></a><strong>hard clustering / soft clustering</strong> - In the context of <a href="#BM_(cluster)%2Fi">cluster</a>ing, ...</p>
<ul>
<li>
<p><a href="#BM_(hard%20clustering%7Csoft%20clustering)%2Fi">soft clustering</a> algorithms assign each data point to a set of probabilities, where each probability is how likely it is for that data point to be assigned to a <a href="#BM_(cluster)%2Fi">cluster</a>.</p>
<p><img src="svgbob_a9ecf872b60c1d4716510c88c16191c37ea93595.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a href="#BM_(hard%20clustering%7Csoft%20clustering)%2Fi">hard clustering</a> algorithms assign each data point to exactly one <a href="#BM_(cluster)%2Fi">cluster</a>.</p>
<p><img src="svgbob_59de88c1a6509780a865b5dee966022210b55bd7.svg" alt="Kroki diagram output"></p>
</li>
</ul>
</li>
<li>
<p><a name="BM_(dot%20product)%2Fi"></a><strong>dot product</strong> - Given two equal sized vectors, the <a href="#BM_(dot%20product)%2Fi">dot product</a> of those vectors is calculated by first multiplying the pair at each index, then summing the result of those multiplications together. For example, the <a href="#BM_(dot%20product)%2Fi">dot product</a> (1, 2, 3) and (4, 5, 6) is 1*4+2*5+3*6.</p>
<p>The notation for <a href="#BM_(dot%20product)%2Fi">dot product</a> is a central dot in between the two vector: <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>a</mi><mo>⋅</mo><mi>b</mi></mrow><annotation encoding="application/x-tex">a \cdot b</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.44445em;vertical-align:0em;"></span><span class="mord mathnormal">a</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.69444em;vertical-align:0em;"></span><span class="mord mathnormal">b</span></span></span></span>.</p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>Central dot is also common place for standard multiplication.</p>
<p>In geometry, the <a href="#BM_(dot%20product)%2Fi">dot product</a> of two vectors is used to get the angle between those vector.</p>
</div>
</li>
<li>
<p><a name="BM_(conditional%20probability)%2Fi"></a><strong>conditional probability</strong> - The probability of an event occurring given that another event has already occurred.</p>
<p>The notation for <a href="#BM_(conditional%20probability)%2Fi">conditional probability</a> is Pr(B|A), where B is the event occurring and A is the event that's already occurred.</p>
<p>If B and A are...</p>
<ul>
<li>independent events, Pr(B|A) is simply Pr(B).</li>
<li>dependent events, Pr(B|A) is calculated as the probability that both B and A happen divided by the probability that just A happens: Pr(A∩B) / Pr(A).</li>
</ul>
</li>
<li>
<p><a name="BM_(similarity%20metric)%2Fi"></a><strong>similarity metric</strong> - A metric used to measure how similar a pair of entities are to each other. Where as a <a href="#BM_(distance%20metric)%2Fi">distance metric</a> must start at 0 for total similarity and grows based on how different the entities are, a <a href="#BM_(similarity%20metric)%2Fi">similarity metric</a> has no requirements for bounds on similarity or dissimilarity. Examples of <a href="#BM_(similarity%20metric)%2Fi">similarity metric</a>s include ...</p>
<ul>
<li><a href="#BM_(pearson%20correlation%20coefficient%7Cpearson%20correlation%7Cpearson%20distance%7Cpearson%20similarity)%2Fi">pearson similarity</a> for <a href="#BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi">gene expression vector</a>s.</li>
<li><a href="#BM_(cosine%20similarity%7Ccosine%20distance)%2Fi">cosine similarity</a> for <a href="#BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi">gene expression vector</a>s.</li>
<li><a href="#BM_%5Cb(BLOSUM)%5Cd*%5Cb%2F">BLOSUM</a> / <a href="#BM_%5Cb(PAM)%5Cd*%5Cb%2F">PAM</a> matrices for protein <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a>s.</li>
<li>etc..</li>
</ul>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>This topic was only briefly discussed, so I don't know for sure what the properties/requirements are for a <a href="#BM_(similarity%20metric)%2Fi">similarity metric</a> other than higher = more similar. Contrast this to <a href="#BM_(distance%20metric)%2Fi">distance metric</a>s, where it explicitly mentions the requirements that need to be followed (e.g. triangle inequality property). For <a href="#BM_(similarity%20metric)%2Fi">similarity metric</a>s, it didn't say if there's some upper-bound to similarity or if totally similar entities have to score the same. For example, does <code>similarity(snake,snake) == similarity(bird,bird)</code> have to be true or can it be that <code>similarity(snake,snake) &gt; similarity(bird,bird)</code>?</p>
<p>I saw on Wikipedia that <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a> scoring matrices like <a href="#BM_%5Cb(PAM)%5Cd*%5Cb%2F">PAM</a> and <a href="#BM_%5Cb(BLOSUM)%5Cd*%5Cb%2F">BLOSUM</a> are <a href="#BM_(similarity%20matrix%7Csimilarity%20matrices)%2Fi">similarity matrices</a>, so that implies that totally similar entities don't have to be the same score. For example, in <a href="#BM_%5Cb(BLOSUM)%5Cd*%5Cb%2F">BLOSUM</a>62 <code>similarity(A,A) = 4</code> but <code>similarity(R,R) = 5</code>.</p>
<p>Also, does a <a href="#BM_(similarity%20metric)%2Fi">similarity metric</a> have to be symmetric? For example, <code>similarity(snake,bird) == similarity(bird,snake)</code>. I think it does have to be symmetric.</p>
</div>
</li>
<li>
<p><a name="BM_(similarity%20matrix%7Csimilarity%20matrices)%2Fi"></a><strong>similarity matrix</strong> - Given a set of n different entities, a <a href="#BM_(similarity%20matrix%7Csimilarity%20matrices)%2Fi">similarity matrix</a> is an n-by-n matrix where each element contains the similarity measure between the entities for that cell. For example, for the species snake, lizard, bird, and crocodile ...</p>
<table>
<thead>
<tr>
<th></th>
<th>Snake</th>
<th>Lizard</th>
<th>Bird</th>
<th>Crocodile</th>
</tr>
</thead>
<tbody>
<tr>
<td>Snake</td>
<td>1.0</td>
<td>0.8</td>
<td>0.4</td>
<td>0.6</td>
</tr>
<tr>
<td>Lizard</td>
<td>0.8</td>
<td>1.0</td>
<td>0.4</td>
<td>0.6</td>
</tr>
<tr>
<td>Bird</td>
<td>0.4</td>
<td>0.4</td>
<td>1.0</td>
<td>0.5</td>
</tr>
<tr>
<td>Crocodile</td>
<td>0.6</td>
<td>0.6</td>
<td>0.5</td>
<td>1.0</td>
</tr>
</tbody>
</table>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p>This topic was only briefly discussed, so I have no idea what properties are required other than: 0 = completely dissimilar / orthogonal and anything higher than that is more similar. It didn't say if there's some upper-bound to similarity or if totally similar entities have to score the same. For example, does <code>similarity(snake,snake) == similarity(bird,bird)</code> have to be true or can it be that <code>similarity(snake,snake) &gt; similarity(bird,bird)</code>? I saw on Wikipedia that <a href="#BM_(sequence%20alignment)%2Fi">sequence alignment</a> scoring matrices like <a href="#BM_%5Cb(PAM)%5Cd*%5Cb%2F">PAM</a> and <a href="#BM_%5Cb(BLOSUM)%5Cd*%5Cb%2F">BLOSUM</a> are <a href="#BM_(similarity%20matrix%7Csimilarity%20matrices)%2Fi">similarity matrices</a>, so that implies that totally similar entities don't have to be the same score. For example, in <a href="#BM_%5Cb(BLOSUM)%5Cd*%5Cb%2F">BLOSUM</a>62 <code>similarity(A,A) = 4</code> but <code>similarity(R,R) =5</code>.</p>
<p>There may be other properties involved, such as how the triangle inequality property is a thing for <a href="#BM_(distance%20matrix%7Cdistance%20matrices)%2Fi">distance matrices</a> / <a href="#BM_(distance%20metric)%2Fi">distance metric</a>s.</p>
</div>
</li>
<li>
<p><a name="BM_(pearson%20correlation%20coefficient%7Cpearson%20correlation%7Cpearson%20distance%7Cpearson%20similarity)%2Fi"></a><strong>pearson correlation coefficient</strong> - A metric used to quantify how correlated two vectors are.</p>
<div style="overflow: auto"><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><mfrac><mrow><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>m</mi></munderover><mo stretchy="false">(</mo><msub><mi>x</mi><mi>i</mi></msub><mo>−</mo><mi>a</mi><mi>v</mi><mi>g</mi><mo stretchy="false">(</mo><mi>x</mi><mo stretchy="false">)</mo><mo stretchy="false">)</mo><mo>⋅</mo><mo stretchy="false">(</mo><msub><mi>y</mi><mi>i</mi></msub><mo>−</mo><mi>a</mi><mi>v</mi><mi>g</mi><mo stretchy="false">(</mo><mi>y</mi><mo stretchy="false">)</mo><mo stretchy="false">)</mo></mrow><msqrt><mrow><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>m</mi></munderover><mo stretchy="false">(</mo><msub><mi>x</mi><mi>i</mi></msub><mo>−</mo><mi>a</mi><mi>v</mi><mi>g</mi><mo stretchy="false">(</mo><mi>x</mi><mo stretchy="false">)</mo><msup><mo stretchy="false">)</mo><mn>2</mn></msup><mo>⋅</mo><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>m</mi></munderover><mo stretchy="false">(</mo><msub><mi>y</mi><mi>i</mi></msub><mo>−</mo><mi>a</mi><mi>v</mi><mi>g</mi><mo stretchy="false">(</mo><mi>y</mi><mo stretchy="false">)</mo><msup><mo stretchy="false">)</mo><mn>2</mn></msup></mrow></msqrt></mfrac></mrow><annotation encoding="application/x-tex">\frac{
  \sum_{i=1}^{m}(x_i-avg(x))\cdot(y_i-avg(y))
}{
  \sqrt{
    \sum_{i=1}^{m}(x_i-avg(x))^2\cdot\sum_{i=1}^{m}(y_i-avg(y))^2
  }
}
</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:2.624002em;vertical-align:-1.13em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.494002em;"><span style="top:-2.1727090000000002em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord sqrt"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.937291em;"><span class="svg-align" style="top:-3.2em;"><span class="pstrut" style="height:3.2em;"></span><span class="mord" style="padding-left:1em;"><span class="mop"><span class="mop op-symbol small-op" style="position:relative;top:-0.0000050000000000050004em;">∑</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.804292em;"><span style="top:-2.40029em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">i</span><span class="mrel mtight">=</span><span class="mord mtight">1</span></span></span></span><span style="top:-3.2029em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">m</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.29971000000000003em;"><span></span></span></span></span></span></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal">x</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord mathnormal">a</span><span class="mord mathnormal" style="margin-right:0.03588em;">v</span><span class="mord mathnormal" style="margin-right:0.03588em;">g</span><span class="mopen">(</span><span class="mord mathnormal">x</span><span class="mclose">)</span><span class="mclose"><span class="mclose">)</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.740108em;"><span style="top:-2.9890000000000003em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mop"><span class="mop op-symbol small-op" style="position:relative;top:-0.0000050000000000050004em;">∑</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.804292em;"><span style="top:-2.40029em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">i</span><span class="mrel mtight">=</span><span class="mord mtight">1</span></span></span></span><span style="top:-3.2029em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">m</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.29971000000000003em;"><span></span></span></span></span></span></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.03588em;">y</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:-0.03588em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord mathnormal">a</span><span class="mord mathnormal" style="margin-right:0.03588em;">v</span><span class="mord mathnormal" style="margin-right:0.03588em;">g</span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.03588em;">y</span><span class="mclose">)</span><span class="mclose"><span class="mclose">)</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.740108em;"><span style="top:-2.9890000000000003em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span></span></span><span style="top:-2.897291em;"><span class="pstrut" style="height:3.2em;"></span><span class="hide-tail" style="min-width:1.02em;height:1.28em;"><svg width="400em" height="1.28em" viewBox="0 0 400000 1296" preserveAspectRatio="xMinYMin slice"><path d="M263,681c0.7,0,18,39.7,52,119
c34,79.3,68.167,158.7,102.5,238c34.3,79.3,51.8,119.3,52.5,120
c340,-704.7,510.7,-1060.3,512,-1067
l0 -0
c4.7,-7.3,11,-11,19,-11
H40000v40H1012.3
s-271.3,567,-271.3,567c-38.7,80.7,-84,175,-136,283c-52,108,-89.167,185.3,-111.5,232
c-22.3,46.7,-33.8,70.3,-34.5,71c-4.7,4.7,-12.3,7,-23,7s-12,-1,-12,-1
s-109,-253,-109,-253c-72.7,-168,-109.3,-252,-110,-252c-10.7,8,-22,16.7,-34,26
c-22,17.3,-33.3,26,-34,26s-26,-26,-26,-26s76,-59,76,-59s76,-60,76,-60z
M1001 80h400000v40h-400000z"></path></svg></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.302709em;"><span></span></span></span></span></span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.6897100000000003em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mop"><span class="mop op-symbol small-op" style="position:relative;top:-0.0000050000000000050004em;">∑</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.804292em;"><span style="top:-2.40029em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">i</span><span class="mrel mtight">=</span><span class="mord mtight">1</span></span></span></span><span style="top:-3.2029em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">m</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.29971000000000003em;"><span></span></span></span></span></span></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal">x</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord mathnormal">a</span><span class="mord mathnormal" style="margin-right:0.03588em;">v</span><span class="mord mathnormal" style="margin-right:0.03588em;">g</span><span class="mopen">(</span><span class="mord mathnormal">x</span><span class="mclose">)</span><span class="mclose">)</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.03588em;">y</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:-0.03588em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord mathnormal">a</span><span class="mord mathnormal" style="margin-right:0.03588em;">v</span><span class="mord mathnormal" style="margin-right:0.03588em;">g</span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.03588em;">y</span><span class="mclose">)</span><span class="mclose">)</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.13em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span></span></div><p>In the above formula, x and y are the two input vectors and avg is the average function. The result of the formula is a number between -1 and 1, where ...</p>
<ul>
<li>-1 represents a total negative correlation.</li>
<li>0 represents no correlation.</li>
<li>1 represents a total positive correlation.</li>
</ul>
<p>The formula may be modified to become a <a href="#BM_(distance%20metric)%2Fi">distance metric</a> as follows: <code>1 - pearson_correlation(x, y)</code>. Where as the <a href="#BM_(pearson%20correlation%20coefficient%7Cpearson%20correlation%7Cpearson%20distance%7Cpearson%20similarity)%2Fi">pearson correlation coefficient</a> varies between -1 and 1, the <a href="#BM_(pearson%20correlation%20coefficient%7Cpearson%20correlation%7Cpearson%20distance%7Cpearson%20similarity)%2Fi">pearson distance</a> varies between 0 (totally similar) and 2 (totally dissimilar).</p>
</li>
<li>
<p><a name="BM_(similarity%20graph)%2Fi"></a><strong>similarity graph</strong> - A transformation of a <a href="#BM_(similarity%20matrix%7Csimilarity%20matrices)%2Fi">similarity matrix</a> into a graph, where the entities that make up the <a href="#BM_(similarity%20matrix%7Csimilarity%20matrices)%2Fi">similarity matrix</a> are represented as nodes and edges between nodes are only made if the similarity exceeds a certain threshold.</p>
<p>The <a href="#BM_(similarity%20graph)%2Fi">similarity graph</a> below was generated using the accompanying <a href="#BM_(similarity%20matrix%7Csimilarity%20matrices)%2Fi">similarity matrix</a> and threshold of 0.45.</p>
<table>
<thead>
<tr>
<th></th>
<th>Snake</th>
<th>Lizard</th>
<th>Bird</th>
<th>Crocodile</th>
</tr>
</thead>
<tbody>
<tr>
<td>Snake</td>
<td>1.0</td>
<td>0.8</td>
<td>0.4</td>
<td>0.6</td>
</tr>
<tr>
<td>Lizard</td>
<td>0.8</td>
<td>1.0</td>
<td>0.4</td>
<td>0.6</td>
</tr>
<tr>
<td>Bird</td>
<td>0.4</td>
<td>0.4</td>
<td>1.0</td>
<td>0.5</td>
</tr>
<tr>
<td>Crocodile</td>
<td>0.6</td>
<td>0.6</td>
<td>0.5</td>
<td>1.0</td>
</tr>
</tbody>
</table>
<p><img src="svgbob_9d808e8cab6e4c6130daf068aabd299ef244ef53.svg" alt="Kroki diagram output"></p>
<p><a href="#BM_(similarity%20graph)%2Fi">Similarity graph</a>s are used for <a href="#BM_(cluster)%2Fi">cluster</a>ing (e.g. <a href="#BM_(gene%20expression%20matrix%7Cexpression%20matrix%7Cgene%20expression%20matrices%7Cexpression%20matrices%7Cgene%20expression%20data%7Cexpression%20data%7Cgene%20expression%20vector%7Cexpression%20vector)%2Fi">gene expression vector</a>s). Assuming <a href="#BM_(cluster)%2Fi">cluster</a>s exist and the <a href="#BM_(similarity%20metric)%2Fi">similarity metric</a> captures them, there should be some threshold where the edges produced in the <a href="#BM_(similarity%20graph)%2Fi">similarity graph</a> form <a href="#BM_(clique)%2Fi">clique</a>s or corrupted <a href="#BM_(clique)%2Fi">clique</a>s (a set of nodes that would form a <a href="#BM_(clique)%2Fi">clique</a> if it weren't for a few extra edges and/or missing edges).</p>
</li>
<li>
<p><a name="BM_(clique)%2Fi"></a><strong>clique</strong> - A set of nodes in a graph where every possible node pairing has an edge.</p>
<p><img src="svgbob_331f1a046e4a700b7f330a8c4875cfa3415603f4.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a name="BM_(clique%20graph)%2Fi"></a><strong>clique graph</strong> - A graph consisting only of <a href="#BM_(clique)%2Fi">clique</a>s.</p>
<p><img src="svgbob_146037c85efe776db75ed0bb94cf76542654c3df.svg" alt="Kroki diagram output"></p>
</li>
<li>
<p><a name="BM_(RNA%20sequencing)%2Fi"></a><strong>RNA sequencing</strong> <a name="BM_(RNA-Seq)%2F"></a><strong></strong> - A technique which uses next-generation <a href="#BM_(sequencing%7Csequenced)%2Fi">sequencing</a> to reveal the presence and quantity of RNA in a biological sample at some given moment.</p>
</li>
<li>
<p><a name="BM_(hierarchial%20cluster)%2Fi"></a><strong>hierarchial cluster</strong> - A form of tiered <a href="#BM_(cluster)%2Fi">cluster</a>ing where <a href="#BM_(cluster)%2Fi">cluster</a>s are represented as a <a href="#BM_(tree)%2Fi">tree</a>. Each node represents a <a href="#BM_(cluster)%2Fi">cluster</a> (leaf nodes being a <a href="#BM_(cluster)%2Fi">cluster</a> of size 1), where the <a href="#BM_(cluster)%2Fi">cluster</a>s represented by a parent node is the combination of the <a href="#BM_(cluster)%2Fi">cluster</a>s represented by its children.</p>
<p><img src="svgbob_f007a8eb3073e01e8c6a0486dbddd83959b0a388.svg" alt="Kroki diagram output"></p>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>⚠️NOTE️️️⚠️</strong></p>
<p><a href="#BM_(hierarchial%20cluster)%2Fi">Hierarchial cluster</a>ing has its roots in <a href="#BM_(phylogeny%7Cphylogenetic)%2Fi">phylogeny</a>. The <a href="#BM_(similarity%20metric)%2Fi">similarity metric</a> to build <a href="#BM_(cluster)%2Fi">cluster</a>s is massages into a <a href="#BM_(distance%20metric)%2Fi">distance metric</a>, which is then used to form a <a href="#BM_(tree)%2Fi">tree</a> that represents the <a href="#BM_(cluster)%2Fi">cluster</a>s.</p>
</div>
</li>
<li>
<p><a name="BM_(cosine%20similarity%7Ccosine%20distance)%2Fi"></a><strong>cosine similarity</strong> - A <a href="#BM_(similarity%20metric)%2Fi">similarity metric</a> that measures if two vectors grew/shrunk in a similar trajectories (similar angles).</p>
<p><img src="svgbob_b7fe91c303f6d60163c6d99da6f196ea8e1f0d4d.svg" alt="Kroki diagram output"></p>
<p>The metric computes the trajectory as the cosine of an angle between vectors. In the example above, T and U have different magnitudes than A and B, but the angle between T and U is exactly the same as the angle between A and B: 20deg. The <a href="#BM_(cosine%20similarity%7Ccosine%20distance)%2Fi">cosine similarity</a> of both pairs is cos(20deg) = 0.939. Had the angle been ...</p>
<ul>
<li>smaller (more similar trajectory), the cosine would get closer to 1.</li>
<li>larger (less similar trajectory), the cosine would get closer to -1.</li>
</ul>
<p>The formula may be modified to become a <a href="#BM_(distance%20metric)%2Fi">distance metric</a> as follows: <code>1 - cosine_similarity(x, y)</code>. Where as the <a href="#BM_(cosine%20similarity%7Ccosine%20distance)%2Fi">cosine similarity</a> varies between -1 and 1, the <a href="#BM_(cosine%20similarity%7Ccosine%20distance)%2Fi">cosine distance</a> varies between 0 (totally similar) and 2 (totally dissimilar).</p>
</li>
</ul>
<p>
</p>
<p>
</p>
<p>
</p>
<p>
</p>
<p>
</p>
<p>

</p>
<p>
</p>
<p>

</p>
<p>
</p>
<p>
</p>
<p></p>
<p></p>
<p></p>
<p>
</p>

        </body></html>